{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8e576922",
   "metadata": {},
   "outputs": [],
   "source": [
    "embed_api = \"sk-6f27c88843124ca5a0bfc4d6b500b011\"\n",
    "embed_url = \"https://dashscope.aliyuncs.com/compatible-mode/v1\"\n",
    "embed_model = \"text-embedding-v4\"\n",
    "\n",
    "openai_key = 'sk-NYPE5WNGTK781Nox801a5934A9F84a03BfC525359eBaE7B6'\n",
    "openai_base = 'http://10.224.28.80:3000/v1'\n",
    "openai_chat_model = 'deepseek-r1-hs'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8e35ccf",
   "metadata": {},
   "source": [
    "# Set ups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a9ae6eb",
   "metadata": {},
   "source": [
    "## Set up embedding models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "92ae6d64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1024\n"
     ]
    }
   ],
   "source": [
    "from langchain_openai import OpenAIEmbeddings \n",
    "from langchain_community.embeddings import DashScopeEmbeddings\n",
    "\n",
    "# Set up embedding model\n",
    "\n",
    "# embed_api = \"\"\n",
    "# embed_url = \"\"\n",
    "# embed_model = \"text-embedding-v4\" # qwen text-embedding-v4 is recommended\n",
    "\n",
    "#-------------------\n",
    "# Note:\n",
    "# The docker version of RefCodeDB is built with Qwen3-Embedding by default\n",
    "#                                       see https://qwenlm.github.io/zh/blog/qwen3-embedding/\n",
    "# If you need to use another embedded model, please rebuild RefCodeDB with your model.\n",
    "#-------------------\n",
    "\n",
    "use_modelscope = True\n",
    "\n",
    "if use_modelscope:\n",
    "    embedding = DashScopeEmbeddings(\n",
    "        model = embed_model,\n",
    "        dashscope_api_key=embed_api)\n",
    "\n",
    "else:\n",
    "    embedding = OpenAIEmbeddings(\n",
    "        api_key=embed_api,\n",
    "        base_url=embed_url,\n",
    "        model = embed_model\n",
    "        )\n",
    "\n",
    "# Test embedding \n",
    "vector = embedding.embed_query(\"test\")\n",
    "print(len(vector))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3ab6d6c",
   "metadata": {},
   "source": [
    "## Pull or Create docker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12ead0ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pull RefCodeDB from docker hub\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cc688f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only if you did not pull RefcodeDB docker, and wanna build it from scratch, run from here, otherwise, skip this cell\n",
    "# Run docker container\n",
    "\n",
    "!docker run --name BIA-RefCodeDB -e POSTGRES_USER=Refcodedb -e POSTGRES_PASSWORD=Refcodedb -e POSTGRES_DB=RefCodeDB -p 6688:5432 -d pgvector/pgvector:pg17"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92224b92",
   "metadata": {},
   "source": [
    "## Connect to vector database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b69723e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from langchain_postgres.vectorstores import PGVector\n",
    "# Connect to Docker Postgres\n",
    "connection = \"postgresql+psycopg://Refcodedb:Refcodedb@localhost:6688/RefCodeDB\"  \n",
    "collection_name = \"RefCodeDB\"\n",
    "\n",
    "vector_store = PGVector(\n",
    "    embeddings=embedding,\n",
    "    collection_name=collection_name,\n",
    "    connection=connection,\n",
    "    use_jsonb=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49c9c8a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "ab44fc50",
   "metadata": {},
   "source": [
    "# Build RefcodeDB from raw\n",
    "\n",
    "!!! if you don't wanna to build it yourself, skip this part."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1b5a27f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from langchain_core.documents import Document\n",
    "# for vector plotting\n",
    "from sklearn.preprocessing import StandardScaler, LabelEncoder\n",
    "import umap.umap_ as umap\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.colors import ListedColormap\n",
    "from langchain_core.messages import HumanMessage, SystemMessage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5327c54e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup LLM\n",
    "# Use Openai API for example, you can use other LLM provider as well\n",
    "from langchain_openai import ChatOpenAI\n",
    "\n",
    "# I will recommend use Openai API, I haven't tested any APIs from other suppliers yet\n",
    "def call_chatllm_openai(key, api_base, model_name):\n",
    "    llm = ChatOpenAI(\n",
    "        api_key = key,\n",
    "        base_url= api_base,\n",
    "        model = model_name,\n",
    "        temperature= 0,\n",
    "        max_retries = 3,\n",
    "        )\n",
    "    return llm\n",
    "\n",
    "# # Setup up api keys \n",
    "# openai_key = \"\"\n",
    "# openai_base = \"\"\n",
    "# openai_chat_model = \"\" \n",
    "# openai_code_model = \"\"\n",
    "\n",
    "# Here, I recommend using an LLM with stronger coding capabilities as the code model, which is set to perform code-related work in all Ghostcoder graphs. Meanwhile, it is better to use LLM with stronger reasoning ability as the chat model.\n",
    "chat_model = call_chatllm_openai(openai_key, openai_base, openai_chat_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e9f08225",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For data processing\n",
    "def translate(code_block,max_retry=3, prompt_dir='prompts/code2instru.md'):\n",
    "    with open(prompt_dir,'r') as f:\n",
    "        prompt = f.read()\n",
    "\n",
    "    message = [\n",
    "        SystemMessage(content=prompt),\n",
    "        HumanMessage(content=code_block)\n",
    "    ]\n",
    "\n",
    "    max_retry =3\n",
    "\n",
    "    i = 0\n",
    "    while i < max_retry:\n",
    "        try:\n",
    "            response = chat_model.invoke(message)\n",
    "            break\n",
    "        except Exception as e:\n",
    "            i+=1\n",
    "            if i == max_retry:\n",
    "                print(f\"Error generating instruction: {e}\")\n",
    "                raise  \n",
    "    return response.content\n",
    "\n",
    "def summarize(instruction, task, max_retry=3, prompt_dir = 'prompts/instru2descrip.md'):\n",
    "    with open(prompt_dir,'r') as f:\n",
    "        prompt = f.read()\n",
    "\n",
    "    human_message = \"Following instruction is for a \"+task+\" task:\\n\" + instruction\n",
    "    \n",
    "    message = [\n",
    "        SystemMessage(content=prompt),\n",
    "        HumanMessage(content=human_message)\n",
    "    ]\n",
    "\n",
    "    max_retry =3\n",
    "\n",
    "    i = 0\n",
    "    while i < max_retry:\n",
    "        try:\n",
    "            response = chat_model.invoke(message)\n",
    "            break\n",
    "        except Exception as e:\n",
    "            i+=1\n",
    "            if i == max_retry:\n",
    "                print(f\"Error generating description: {e}\")\n",
    "                raise  \n",
    "    return  response.content\n",
    "\n",
    "\n",
    "# For sematic embedding\n",
    "async def acal_EmbVec(RefcodeDB, cate):\n",
    "    try:\n",
    "        vecs = await embedding.aembed_documents(RefcodeDB[cate].tolist())\n",
    "    except:\n",
    "        vecs = []\n",
    "        for _, row in RefcodeDB.iterrows():\n",
    "            vec = embedding.embed_query(row[cate])\n",
    "            vecs.append(vec)\n",
    "    return vecs\n",
    "\n",
    "def cal_EmbVec(RefcodeDB, cate):\n",
    "    vecs = []\n",
    "    for text in RefcodeDB[cate].tolist():\n",
    "        vec = embedding.embed_query(text)\n",
    "        vecs.append(vec)\n",
    "    return vecs\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5050a4da",
   "metadata": {},
   "source": [
    "## Data processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d517dd59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RefcodeDB has collected 54 reference code blocks, for 12 tasks\n"
     ]
    }
   ],
   "source": [
    "# You can add or remove code blocks from RefcodeDB by you own. \n",
    "# Here we show the process of generating task instructions and descriptions from code blocks.\n",
    "\n",
    "# Load RefcodeDB tables\n",
    "table_dir = \"rawtable/RefCodeDB.csv\"\n",
    "RefcodeDB = pd.read_csv(table_dir)\n",
    "\n",
    "print(\"RefcodeDB has collected {} reference code blocks, for {} tasks.\".format(RefcodeDB.shape[0],len(RefcodeDB['Task'].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0af98f9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Translate code block to task instruction, then summarize to description\n",
    "for idx, row in RefcodeDB.iterrows():\n",
    "    code = row['CodeBlock']\n",
    "    task = row['Task']\n",
    "    inst = translate(code)\n",
    "    dscp = summarize(inst, task)\n",
    "    RefcodeDB.at[idx, 'Instruction'] = inst\n",
    "    RefcodeDB.at[idx, 'Description'] = dscp\n",
    "    \n",
    "RefcodeDB.to_csv(\"RefCodeDB.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "83524493",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Translate code block to task instruction, then summarize to description\n",
    "for idx, row in RefcodeDB.iterrows():\n",
    "    #code = row['CodeBlock']\n",
    "    task = row['Task']\n",
    "    #inst = translate(code)\n",
    "    inst = row['Instruction']\n",
    "    dscp = summarize(inst, task)\n",
    "    #RefcodeDB.at[idx, 'Instruction'] = inst\n",
    "    RefcodeDB.at[idx, 'Description'] = dscp\n",
    "    \n",
    "RefcodeDB.to_csv(\"RefCodeDB.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8625de06",
   "metadata": {},
   "source": [
    "## Evaluate your embedding model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03f3858c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_cluster_umap(vectors, label, Title, scaled = True):\n",
    "    # Calculate UMAP embedding\n",
    "    X = np.array(vectors)\n",
    "    X_scaled = StandardScaler().fit_transform(X)\n",
    "    reducer = umap.UMAP(\n",
    "        n_neighbors=15,       \n",
    "        min_dist=0.1,         \n",
    "        n_components=2,       \n",
    "        random_state=42      \n",
    "    )\n",
    "    if scaled:\n",
    "        embedding = reducer.fit_transform(X_scaled)\n",
    "    else:\n",
    "        embedding = reducer.fit_transform(X)\n",
    "\n",
    "    # Map labels color\n",
    "    labels = RefcodeDB[label].tolist()\n",
    "    le = LabelEncoder()\n",
    "    color_labels = le.fit_transform(labels)\n",
    "    unique_labels = le.classes_\n",
    "    cmap = plt.get_cmap('Spectral', len(unique_labels))\n",
    "\n",
    "    # Plot umap scatter\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    scatter = plt.scatter(\n",
    "        embedding[:, 0], \n",
    "        embedding[:, 1], \n",
    "        s=50, \n",
    "        alpha=0.7,\n",
    "        c = color_labels,\n",
    "        cmap=cmap\n",
    "        )\n",
    "    \n",
    "    ax = plt.gca()\n",
    "    ax.set_aspect('equal', 'box') \n",
    "    ax.set_box_aspect(1) \n",
    "\n",
    "    # Legend\n",
    "    handles, _ = scatter.legend_elements(prop=\"colors\", alpha=0.6)\n",
    "    plt.legend(\n",
    "        handles, \n",
    "        unique_labels, \n",
    "        title=\"Task Categories\",\n",
    "        loc=\"center left\",  \n",
    "        bbox_to_anchor=(1, 0.5),\n",
    "        frameon=False,\n",
    "        framealpha=0.8\n",
    "    )\n",
    "\n",
    "    # Add Title\n",
    "    plt.title(Title, fontsize=14)\n",
    "    plt.xlabel('UMAP 1', fontsize=12)\n",
    "    plt.ylabel('UMAP 2', fontsize=12)\n",
    "    #plt.grid(alpha=0.2)\n",
    "\n",
    "    # Adjust layout\n",
    "    plt.tight_layout()\n",
    "    plt.subplots_adjust(right=0.8)\n",
    "    plt.show()\n",
    "\n",
    "def embedding_heatmap(vectors, RefcodeDB, Title, scaled = True):\n",
    "    X = np.array(vectors)\n",
    "    X_scaled = StandardScaler().fit_transform(X)\n",
    "    # Group label\n",
    "    languages = RefcodeDB['Language'].tolist()\n",
    "    tasks = RefcodeDB['Task'].tolist()\n",
    "\n",
    "    # Calculate similarity score\n",
    "    similarity_matrix = np.zeros((48, 48))\n",
    "    for i in range(48):\n",
    "        for j in range(48):\n",
    "            if scaled:\n",
    "                similarity_matrix[i, j] = np.dot(X_scaled[i], X_scaled[j])\n",
    "            else:\n",
    "                similarity_matrix[i, j] = np.dot(X[i], X[j])\n",
    "            \n",
    "\n",
    "    plt.figure(figsize=(15, 14))\n",
    "    ax = plt.gca()\n",
    "\n",
    "    # Plot heatmap\n",
    "    im = ax.imshow(similarity_matrix, cmap='viridis', interpolation='nearest')\n",
    "\n",
    "    # Add group label\n",
    "    ax_lang_top = ax.inset_axes([0, 1.05, 1, 0.05], transform=ax.transAxes)\n",
    "    unique_langs = sorted(set(languages))\n",
    "    lang_colors = sns.color_palette(\"Set3\", len(unique_langs))\n",
    "    lang_indices = [unique_langs.index(lang) for lang in languages]\n",
    "    ax_lang_top.imshow([lang_indices], cmap=ListedColormap(lang_colors), aspect='auto')\n",
    "    ax_lang_top.set_xticks([])\n",
    "    ax_lang_top.set_yticks([])\n",
    "    ax_lang_top.set_title('Language Groups', fontsize=12, pad=5)\n",
    "\n",
    "    ax_task_below = ax.inset_axes([0, 1.00, 1, 0.05], transform=ax.transAxes)  \n",
    "    unique_tasks = sorted(set(tasks))\n",
    "    task_colors = sns.color_palette(\"Set2\", len(unique_tasks))\n",
    "    task_indices = [unique_tasks.index(task) for task in tasks]\n",
    "    ax_task_below.imshow([task_indices], cmap=ListedColormap(task_colors), aspect='auto')\n",
    "    ax_task_below.set_xticks([])\n",
    "    ax_task_below.set_yticks([])\n",
    "    ax_task_below.set_title('Task Groups', fontsize=12, pad=5)\n",
    "\n",
    "\n",
    "    # Legend for groups\n",
    "    lang_patches = [Patch(color=lang_colors[i], label=lang) for i, lang in enumerate(unique_langs)]\n",
    "    task_patches = [Patch(color=task_colors[i], label=task) for i, task in enumerate(unique_tasks)]\n",
    "    plt.colorbar(im, label='Dot Product Similarity')\n",
    "    plt.legend(\n",
    "        handles=lang_patches + task_patches, \n",
    "        title='Group Legend',\n",
    "        loc='upper center',  \n",
    "        bbox_to_anchor=(0.5, -0.10),  \n",
    "        ncol=4,  \n",
    "        frameon=False,\n",
    "        framealpha=0.8,\n",
    "        fontsize=12)\n",
    "\n",
    "\n",
    "    # Add title\n",
    "    plt.title(Title, fontsize=18, y=1.15)\n",
    "    plt.xlabel('Vector Index', fontsize=12)\n",
    "    plt.ylabel('Vector Index', fontsize=12)\n",
    "    plt.grid(False)\n",
    "\n",
    "    # Adjust fig\n",
    "    plt.tight_layout()\n",
    "    plt.subplots_adjust(bottom=0.25)  \n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71d9eaa5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qixin/miniconda3/envs/BIA/lib/python3.11/site-packages/umap/umap_.py:1945: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3AAAAJOCAYAAAD27eW+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAADp2UlEQVR4nOzdd1xTV/8H8M9NIGGGJUtEgSIKCuKsoxWtA2fVah114WwdtXXU1taB9tdabd1trVUBa4daV617Yq17FxUXgijiAgUZEkjO7w9KHmMCgoAI/byfVx6bc88953tvAsmXe+45khBCgIiIiIiIiF56srIOgIiIiIiIiAqHCRwREREREVE5wQSOiIiIiIionGACR0REREREVE4wgSMiIiIiIionmMARERERERGVE0zgiIiIiIiIygkmcEREREREROUEEzgiIiIiIqJyggkclbiff/4ZgYGBsLKygiRJCA0NLeuQ8uXh4QEPD48y6TsiIgKSJCEiIqLQ+0iShBYtWuiVhYSEQJIkxMXFlWh8L9KtW7fQr18/uLm5QSaTQZKksg6pTIWGhkKSJERGRpZqP8beT/RyelHviZdJixYt/vO/C4iIjGECV0yRkZGQJAkhISFFrpP34SRJErZv357v/vXq1dPVO3LkSL71pk6dCkmSoFQqkZSUlG+9J/uVJAkymQx2dnZo3rw5IiIiIITId99nOXToEPr374+MjAyMGjUK06ZNe+YXRA8PD714jD0ePnz43DHRyy8kJAS//vorWrZsiSlTpmDatGkF1s9Lfp98D6tUKnh6eqJLly5YtGgRkpOTX1D0L6+K9gU4JiYG48aNQ2BgIGxtbWFqagpHR0cEBQXhiy++wM2bN8s6xAoj73OrsA/+IYCI6MUxKesACDAxMUFYWBjatWtnsO3s2bM4ffo0TExMkJOTk28bWq0WK1asgCRJUKvV+Pnnn/HBBx8U2O/48eNhZWUFjUaDa9euYf369Thw4ABOnjyJRYsWPdexbN26FQDw008/oXHjxoXeTy6XY/LkyfluNzMze654/gtmzpyJTz75BG5ubmUdynNRq9XYs2cP2rZti59//rlI+7Zq1QqvvfYaACAtLQ03b97EgQMHsGnTJkybNg1LlizB22+/XRphl6rRo0ejd+/eqFq1aqn2Ex0dDQsLi1Lto6TMnTsXH3/8MTQaDRo3boz+/ftDpVIhOTkZx44dw5QpUzBjxgwcP34cAQEBZR1uuefh4WHwh5S4uDisWLECderUQdeuXQ3qExHRi8EE7iXQvn17/PHHH0hKSoKDg4PetuXLl8PExARt2rTBtm3b8m1j165diI+Px4gRI/DTTz9h+fLlz0zgJkyYABcXF93zqKgovPrqq/juu+8wbtw4eHp6FvlYbt26BQB67RaGiYnJSz3U8mXm6uoKV1fXsg7jud2+fRtarbbI7xkAaN26NT755BO9Mo1Gg4iICLz//vvo06cPbGxs0LZt25IK94WoVKkSKlWqVOr91KxZs9T7KAlLlizB+PHj8corr2DNmjWoV6+eQZ3o6GhMmTIFqampZRBhxePh4WHwOzkyMhIrVqxAYGAgf18TEZUhDqF8CQwaNAhqtRq//PKLXrlarcavv/6Kjh07wsnJqcA2li9fDgAYOXIkunXrhqioKBw/frxIcfj7+yMoKAhCCJw8eVJXfujQIXTs2BH29vYwMzNDzZo1ERoaioyMDF2dvOE24eHhAABPT0/d0JqSljcsLCsrC59++imqVq0Kc3Nz1K9fH7t37wYAPHr0CGPGjIGbmxvMzMzQpEkTnDhxIt82Hzx4gGHDhsHZ2Rnm5uZo1KgRNm3aZLSuEAJhYWFo1qwZVCoVLCws0KBBA4SFhRmtn5ycjPfeew/Ozs6wsLBAw4YNsWHDhgKPcdmyZahduzbMzMzg7u6OiRMn4vHjx0brGrsHLu/1CA0NxalTpxAcHAxra2vY2NigW7du+d4vt379ejRo0ADm5uZwdnbGsGHD8ODBgyLfK5iRkYHQ0FDUrFkTZmZmsLe3R8eOHXHo0CG9ei1atEC1atUAQHcFubj3TcrlcgwZMgQ//PADNBoNxo0bZzAsWK1WY+7cuahXrx4sLS1hbW2N119/3ehrnpKSgqlTp8LPzw9WVlawsbFBzZo1MWjQINy4cUOvrhACK1asQPPmzWFrawsLCwtUr14d7733HuLj4/WOO+89PHXqVHh7e8PU1FR33Mbud4qLi9MNxT537hzat28PGxsbqFQqdO7cGRcuXNCLRZIk7N+/X/ffeY8nh3LnN/QtKSkJY8eOhaenJ5RKJZycnNCrVy+DPgD999/3338PX19fmJmZoVq1apg+fTq0Wq3R16mwHjx4gIkTJ8LMzAzbtm0zmrwBgK+vL9auXWtw5T/vGBMSEhASEgIXFxfIZDK9c7tixQo0btwYVlZWsLKyQuPGjbFixQqDPgq6b/XJnzlj/d+4cQO9evWCg4MDLC0t0aJFC4Ofh8JaunQpatWqBTMzM1StWhWTJk3S+/1w7do1yGQydOzY0ej+Dx48gJmZGerUqfNc/T/t5MmTGD16NGrXrg0bGxuYm5vD398fX331FbKzsw3qX7lyBYMGDYKnpyfMzMxQqVIl1KtXD+PHjy9Uf7/++isUCgXq16+Pu3fvlsgxEBGVJ7wC9xJo0qQJfH19ER4ejjFjxujKN27ciKSkJAwePBjr16/Pd/+kpCT88ccfqFu3LmrXro0BAwbg559/xvLly9GwYcNixbZu3Tr07t0bCoUCvXr1gpOTE3bv3o3p06dj586d2LdvH5RKpW64zcaNG3H27Fl88MEHsLW1LVbfz9KrVy9ERUXhzTffRGZmJn755Rd06tQJhw4dwrvvvovHjx+jR48euHfvHlavXo3g4GDExsZCpVLptaNWq9G6dWtkZmZi4MCBePjwIVatWoWuXbti5cqV6Nu3r66uEAL9+vXDr7/+Ch8fH7zzzjtQKBTYtWsXhgwZggsXLuCbb77R1c/IyECLFi0QFRWFJk2aICgoSPdFLr+rQp9//jmmTp2qS6BMTU2xevVqREdHF/kcnThxAl9//TVatGiBd999F6dPn8bGjRsRFRWFc+fO6Q1NDQsLw5AhQ2Bra4sBAwbAxsYGW7duRZs2bZCdnQ1TU9NC9ZmVlYVWrVrhyJEjqFevHj788EPcvXsXq1evxs6dO7F69Wq89dZbAHK//AcGBmLBggV6w7JK4n6afv36Ydq0aTh//jzOnTsHf39/XXzt2rVDZGQk6tatiyFDhiA7OxtbtmzR3T83evRoALmvd3BwMI4ePYpmzZqhXbt2kMlkiIuLw4YNGzBw4EC4u7vr6vbp0werV6+Gm5sb+vTpA5VKhbi4OKxevRrt2rUzGBL51ltv4ezZswgODoa9vT28vLyeeVzXrl1Ds2bN0KhRI4wcORJXrlzBhg0b8Pfff+PQoUPw9fUFAEybNg0RERG4fv263lC4wMDAAttPSkpC48aNcfXqVbRo0QK9e/dGXFwc1q5diy1btmDXrl1o0qSJwX4fffQRIiMj0alTJ7Rt2xYbN25EaGgo1Go1vvjii2ceV37Wrl2L1NRU9O/fH9WrV39mfRMTw4+1pKQkNGnSBPb29ujVqxfUarXu98DYsWMxf/58uLm5YciQIZAkCevWrUNISAjOnj2LuXPnPnfseR48eIBmzZrB1dUVw4cPR0JCAlavXo2WLVtix44dRXq/z5kzB5GRkejVqxc6deqErVu34quvvsLp06exbds2SJIELy8vtG7dGtu3b8fNmzdRpUoVvTZWrlyJrKwsDBs2rNjHBuQmlH/++SeaN2+ODh06ICMjA5GRkZg0aRKOHz+OdevW6ereunULjRo1Qnp6Ojp27IhevXohLS0NV65cwaJFizBnzpwC+1qwYAHGjh2Lli1bYuPGjbC2ti6RYyAiKlcEFcu+ffsEADFw4MAi1wkKChIARGJiopg9e7YAIE6dOqXb3rZtW+Hs7Cyys7PFwIEDBQBx+PBhg/bnzZsnAIi5c+cKIYTQaDSiSpUqQqVSifT0dIP6T/b7pH/++UeYm5sLSZJEbGysSE1NFba2tkKpVIqzZ8/q6mm1WvHOO+8IAOLzzz/XayMvztjY2HzPx9OqVasm5HK5mDZtmtHH4sWLjcbfrFkzkZaWpitftWqVACBsbW3F22+/LbKzs3XbZs2apXeOnuwbgHjjjTeEWq3WlUdHRwtzc3Nha2srUlNTdeU//vijACCGDBmi135WVpbo3LmzACBOnDihK582bZoAIIYNG6bX744dOwQAAUCEh4fryq9cuSJMTEyEm5ubuHPnjq48JSVF1KhRQwAQQUFBem0ZO+d57zkAYtWqVXr1+/fvLwCI3377TVf24MEDYWVlJaytrUVMTIyuPDs7W7Ru3VoAENWqVROFMWPGDAFA9O3bV2i1Wl352bNnhVKpFHZ2dnrnNDY29pk/Q08LDw8XAMTMmTMLrJd3rMuXL9eVffrppwKACA0N1YsvNTVVNGjQQCgUCpGQkCCEyP2ZACC6detm0Pbjx4/Fo0ePdM+/++47AUC0atVKZGRk6NXNyMgQSUlJuud57+HAwEC98jx575t9+/bpyvLOEwAxefJkvforVqzQvY+flNdPfoy9nwYPHiwAiEmTJumVb9++XQAQ1atXFxqNRlee9/7z9PQUt27d0pXfu3dP2NraCmtra5GVlZVvDM8yaNAgAUCEhYU91/5552zQoEEiJydHb9tff/0lAAhfX1/x8OFDXfnDhw9FzZo1BQBx4MABXXne++7Jn9k8eT9z06ZNM9p///799d5vkZGRQpIk4e3trXc+85P3njAzMxPnzp3TlWdnZ4s2bdoIAOKnn37Slf/+++8CgJg+fbpBWwEBAcLMzEw8ePDgmf0aO8anf1bj4uIMzq1Wq9W9l/7++29d+cKFCwUAsWDBAoP27927p/f86ffvpEmTBADx9ttvF+s9RURU3jGBK6aSSuBu374tTExMxOjRo4UQQsTHxwuZTCY++ugjIYQoMIHz9/cXcrlcLyH7+OOPBQCxYsUKg/p5/Y4fP15MmzZNTJ48WbzzzjvCzMxMABBjxowRQgjx008/CQBixIgRBm3Ex8cLExMT8corr+iVP28Cl/clx9ijTp06RuOPjIzUK8/JyRGmpqYCgLh+/bpBvMZeg7y+Dx48aBDXqFGjBACxcuVKXVlAQICwtLQUmZmZBvXzvuyPHz9eV+bp6SkUCoVBsiyEEK1atTL4Mjh9+nQBQMyZM8eg/sqVK4ucwDVv3tygnbxt48aN05VFREQIAGLs2LEG9Q8fPlykBM7Ly0uYmpqKGzduGGx79913Dc5paSZweT8Hs2bNEkLk/nHDzs5OeHt7632ZzrNp0yYBQCxatEgI8b/X9J133nlmTH5+fkIul4vLly8/s27ee/iPP/4wur2gBM7Ozk7vDxdC5H5Zrl27tgAg4uPjDfrJz9Pvp6ysLGFubi4cHByM/vEnODjYIKnJe/8ZS7Dytv3zzz/5xvAs7du3FwDE9u3bDbadPHnS4A8+f/75p8ExKhQKg+RAiP8lq6tXrzbY9ttvv+n+WJPneRM4uVyu97rk6dixo8H5zE9+fwwSQojjx4/r/niQR61WC2dnZ+Hh4aH3Xj927JgAIPr16/fMPp9WmM+7J508eVL3x5I8eQncjz/++Mz9896/OTk5YsiQIbrPo8IkvEREFRmHUL4knJ2d0aFDB/z666/45ptvEBERAa1Wi0GDBhW437FjxxAVFYV27drpTQIxcOBAzJo1C2FhYRgwYIDRffOGqkiSBJVKhYYNG2LIkCG6+qdPnwZgfDibu7s7XnnlFVy6dAmPHj0q9jAWpVKZ7z1e+albt67ec7lcDicnJ6SnpxsMVcub5CMhIcGgHVNTU6MzZr7++uv47rvvcObMGfTr1w8ZGRmIiopC5cqV8dVXXxnUz7vX4+LFiwBy78OLjY2Fn5+f0Qk6Xn/9dezZs0ev7OzZs7ptxuoXlbH7hfKGUz25NENev02bNjWo36hRI6PD0oxJTU3FtWvX4OvrazBsC8h9Ly1ZskR3TkubeOret0uXLuHBgweoXLkypk+fblD/3r17AP73Gvr6+sLf3x+//vorbty4ga5du+L1119HvXr1IJfLdfulp6fjwoUL8Pb2LtQwvzyNGjUq8jHVrVsXlpaWemWSJOG1117DuXPncPbsWd2wzqK6ePEiMjMz0aJFC6OzU7Zo0QI7duzAmTNndLN/5inse62onn4Nn3Tq1CmD1/Hdd99Fp06d9Mo8PT2NTgpT0O+4vLIzZ84ULWAjqlWrZvQ1ef3117Flyxaj5zM/xn4P5N23+mSspqamGDx4MGbOnIldu3bphmzn3S89dOjQ5zgS49RqNb799lusWrUKFy9eRFpamt7rlje5FQB06tQJn3zyCUaNGoVdu3ahXbt2eO211+Dj45Nv+2+99ZZuZllOnkJExHvgik0my50HpqAb9fO25dXNz6BBg7Bp0yZs2LABERERaNy4se5+lvzkTZzRv39/vXJfX180aNAA+/fvx9WrV+Ht7W2wb2JiYoEz/+XN5ubs7Gx0u4uLCy5duoTU1NQyuQ/h6XvZgNz7X2xsbIyWAzB6Q72Dg4PR1ybvuFNSUgDk3scihEBCQoLRL/950tPT9fbLbwIaY+e1oH3yex0KUtC50Gg0urK819rR0dGgvkwmK/SMiIV5zwD/O87SlpiYCOB/x5W3Ntz58+dx/vz5fPfLew1NTEywd+9ehIaGYv369bpJFipVqoT3338fn332GeRyuS5BKepSDs/zmj7r/VScc1uc16+w77WiyovF2B9fhg4dqktEIiMj0bJlywLbeFpqaipkMpnR972zszNkMlmJvFdL8jXLry0nJyeDczRs2DB89dVXWLZsGdq2bYuMjAz89ttv8PHxQVBQUKH7fJYePXrgzz//hI+Pj+5eaVNTUzx8+BALFixAVlaWrq6npycOHz6M6dOnY9u2bfj9998BADVq1MDnn39udNmPAwcOwNzcHO3bty+xmImIyjPOQllMeV9aClo4+/79+3p189OpUyc4OztjwoQJuHbtGgYPHlxg/bwPYwDo27evwcKqebMu5jc74rPkJUh37twxuj2v3FgiVZ4kJSUZTcDzji/vdcs7zvr160PkDj82+ti3b59e/fxmSTN2XvP6MrZPfq9DSciLNe8K1JO0Wq3uPVzYdl6G94xWq8Vff/0FALrJfPL67d69e4GvYd5sqkBusvbtt98iISEBFy5cwLfffgsHBwdMmzYNs2fPBvC/181YklGQ55ml9Vnvp2f9ninIy/T65cm7Kpz3c/U88jvPKpUKWq3W6Pv+7t270Gq1esea94ceY2tyFpSEleRrll9bd+/eNWjH09MTbdq0wR9//IH79+9jzZo1SE1NLdGrb8ePH8eff/6J4OBgXLhwAUuXLsUXX3yB0NBQ9O7d2+g+AQEBWLduHZKTk3H48GFMnToVd+7cQa9evXDw4EGD+nv27IG5uTnatm2LI0eOlFjsRETlFRO4YqpRowYUCgWOHz+e70Lbhw8fBoBnLi5rYmKCfv36ISEhARYWFujVq1eB9fNmZwsMDMSQIUOMPkxNTbFixYrn+gt43hDFJ6fbzpOQkICYmBh4eXmV+1nAsrOzjX4pOHDgAID/zdpnbW0NX19fREdHF2pImEqlgqenJ65evYrbt2/n2/6T8qb1NrbNWFlJyevX2LTmx44dK3AR+SepVCp4eXnh6tWrRpOZvGntnzUTYklYuXIlrl+/Dn9/f9SqVQtA7pVplUqFEydOGL0aWxBJkuDr66sb+gVAt+yAlZUV/Pz8EBsbiytXrpTsgTzl9OnTuiuET8r74vvk1PB5wzwL+/Oft+zD8ePH9ZYJyfMiX788PXr0gLW1NX7//fcSP7cF/Y4zdqx2dnYAjCfqecMxjbl+/brBkhOA4e+YwjD2e+DEiRPIzMw02s7w4cOhVqt164Oamppi4MCBhe7vWWJiYgAAHTt21BtWnF+sT8obvj59+nQsXLgQQghs3rzZoF7dunWxZ88emJiYIDg4mEkcEf3nMYErJjMzM/Ts2RP37t3D//3f/xlsj4qKwrJly2BtbY1u3bo9s72PPvoIGzZswI4dO575V+68exnmzZuHZcuWGX106tQJt27dKnAR8Px06dIFNjY2CA8P1xtuJoTApEmTkJ2drbemVHk2ZcoUvS/0Fy9eRFhYGGxsbNClSxdd+ZgxY5CRkYFhw4YZ/RIdGxurt8Za//79oVarMXXqVL16O3fuNLj/DQDeeecdyOVyzJ07V+8v7ampqUbfXyWlS5cusLKywrJlyxAbG6srz8nJwZQpU4rU1sCBA5GdnY1Jkybp3Qdz7tw5hIeHw8bGRrdcQGnQaDQICwvDiBEjdOcy7wqMiYkJRowYgevXr2PChAlGk7hz587pzn1sbKzRtc/yrpyYm5vrykaNGgWNRoORI0ciMzNTr/7jx491wzeL68GDBwb3YP7000+IiorCG2+8oXevlb29PQDg5s2bhWpboVCgT58+uH//PmbOnKm3bffu3di2bRu8vb3RrFmzYh2Dh4eHwdqF+bG3t8fMmTORlZWF9u3b55soPc99dnmJzPTp0/UWAE9NTdUNk34y2alXrx4kScKqVav07tm9cuUKFixYkG8/Go0Gn332md7Pw/79+7F161Z4e3sbvfc0PytXrtT7fZyTk4NPP/3UINY8Xbp0gYuLC+bMmYO///4bb7755jPXFS2KvHUc//77b73y8+fPG7yHgNwrdgWNMHjyZ+pJgYGB2Lt3L0xNTdG2bdvnXkOPiKgi4D1wJWDOnDk4evQopk+fjs2bNyMoKAhmZma4fPkyNm3aBCEEfvnll0Kti+bs7FyoL7dXr17FX3/9BS8vrwLvZRg0aBA2bNiA5cuXG9zY/ywqlQpLly5Fnz598Oqrr6JXr15wdHTEnj17cOLECTRq1AgfffRRkdrMT05OToE3p4eEhBRpIemicHV1xcOHDxEYGIiOHTsiJSUFv/32Gx4/foylS5fqXWF89913ceTIEaxYsQIHDx5E69atUblyZdy5cwcXL17E0aNH8euvv+pinThxItavX4+lS5fi/PnzaN68OW7cuIE1a9agY8eO2LJli14s3t7emDp1KqZNm4aAgAD07NkTJiYmWLduHfz9/XHp0qVSOQe2traYO3cuhg8fjnr16qFXr166deCUSiUqV678zHs480ycOBFbtmzBypUrER0djVatWunW4svOzsZPP/1UYldtd+/erfsinZGRgZs3b+Kvv/5CQkIC7O3tsXLlSrRu3Vpvn+nTp+PUqVNYuHAhtmzZgqCgIDg6OiIhIQFRUVE4e/YsDh8+DCcnJ5w9exbdunVDw4YNUbt2bbi4uCAhIQEbN26EXC7XW3h4xIgR2L9/P9asWYPq1avjzTffhEqlQnx8PHbs2IHly5eXSOL6+uuvY+HChThy5AgaNmyIy5cvY8OGDbCxscG3336rV/eNN97A2rVr8fbbb6NDhw4wMzODv79/vgs8A8CsWbOwf/9+/N///R8OHTqEV199VbcOnIWFBcLDwwv9XshP3pDlwk6OM2rUKKSnp+PTTz9F/fr10bhxYzRo0ADW1tZISkpCdHQ0Dhw4AKVSWaS1L5s3b473338fixYtQu3atXVDa9evX48bN25gzJgxaN68ua6+m5sbevXqhVWrVqF+/fpo164d7t69iw0bNqBdu3Z66509KSAgAJGRkWjcuDHeeOMN3Lp1C6tWrYKpqSmWLl1apPPZunVrNG7cGL1794a9vT22bt2Kc+fOITg42OjEQCYmJhg8eDC+/PJLACU7eQmQOxFPo0aNsGbNGiQmJqJx48aIj4/Hpk2b0LFjR6xdu1av/i+//ILvv/8eLVq0gLe3N1QqFS5cuICtW7eiUqVKBd46UKdOHezduxetWrVCu3btsH379iIlv0REFcaLm/CyYnv48KGYNm2aqFOnjrC0tBSmpqbC3d1dvPPOO3pruz0pv/XYjHl6GYFPPvlEwMg6bE/Lzs4Wzs7OwsTERNy+fbvI/QqRu1ZS+/btha2trVAoFMLHx0dMmTLFYCrzJ+MsyWUE8NR06gVNjV6tWrV8p7uHkSn48+onJSWJoUOHCicnJ6FUKkWDBg3yneJdCCFWr14tWrduLezs7ISpqalwc3MTLVq0EHPmzDGYrjwpKUkMHz5cODo6CjMzM1G/fn2xfv36AqckX7p0qfDz8xMKhUJUqVJFTJgwQWRkZBR5GYGnpzQXouBp+3///XdRt25doVQqhZOTkxg6dKhISkoSVlZWBss5FCQtLU1MmTJF+Pj4CIVCIWxtbUX79u2NTpdenGUE8h6SJAkrKyvh4eEhOnfuLBYtWiSSk5Pz3T8nJ0csWbJENGvWTKhUKqFUKkXVqlVFu3btxOLFi3Xv7Rs3bohPPvlENG7cWDg5OQmFQiGqVq0qevToIY4ePWrQrlarFcuWLRONGzcWlpaWwsLCQlSvXl289957RZrev6BlBAYOHCj++ecf0a5dO2FtbS2srKxEx44d9dYGy5OdnS0mTpwoqlatKkxMTAzOs7H3kxC563GNGTNGVKtWTZiamopKlSqJHj16iKioKIO6Bf3MGzuOBw8eCJlMJpo1a5bv8efn0qVLYsyYMcLf31+oVCphYmIiKlWqJF5//XUxY8YMo1P153eMTwoLCxMNGzYUFhYWwsLCQjRs2DDfdefS09PF+++/L5ydnYVSqRQBAQHil19+KXAZgaCgIHH9+nXx9ttvCzs7O2Fubi6aN2+utz7aszx5LpcsWSL8/PyEUqkUVapUEZ988onB2oNPunTpkgAgqlatWqwp+PNbRuDu3bti8ODBonLlysLMzEz4+/uL7777Tly7ds2g/pEjR8S7774rateuLWxtbYW5ubmoXr26GDNmjMHrl9/PydmzZ0WlSpWElZVVoZZgICKqaCQhCpijmYj+865evYrq1aujZ8+eWL16dVmH858VFxcHT09PDBw4EBEREWUdznPbvHkzOnfujC1btqBDhw5lHU6pkyQJQUFBRu+ze1HWrFmDXr16Yfr06QbDuYmIqPzhPXBEBCD33qonp/sGgMzMTIwdOxYASvW+NfrvOHDgAOrUqfOfSN5eBkIIzJ07FyYmJhgyZEhZh0NERCWA98AREYDcSRWGDBmCtm3bomrVqrh//z727t2LuLg4vPHGG8+cFZWoMGbNmoVZs2aVdRgVXlRUFDZv3oxDhw7h6NGjeO+994q8TiEREb2cmMAREQCgVq1aaNOmDQ4ePIiNGzcCyJ1U5fPPP8eECROKPXEFEb04J0+exKeffgpbW1sMGDAAX3/9dVmHREREJYT3wBEREREREZUT/JM6ERERERFROcEEjoiIiIiIqJzgPXBGaLVa3Lp1C9bW1pAkqazDISIiomIQQuDRo0eoXLky7+clonKPCZwRt27dgru7e1mHQURERCXoxo0bqFKlSlmHQURULEzgjLC2tgaQ+4tepVKVcTRERERUHKmpqXB3d9d9vhMRlWdM4IzIGzapUqmYwBEREVUQvC2CiCqCCjcQPDQ0FJIk6T1cXFzKOiwiIiIiIqJiq5BX4GrVqoXdu3frnsvl8jKMhoiIiIiIqGRUyATOxMSEV92IiIiIiKjCqXBDKAHgypUrqFy5Mjw9PdG7d29cu3atwPpZWVlITU3VexAREREREb1sKlwC9+qrr+Knn37Cjh07sHTpUty+fRtNmzZFUlJSvvvMnDkTNjY2ugeXECAiIiIiopeRJIQQZR1EaUpPT8crr7yCiRMnYty4cUbrZGVlISsrS/c8b7rhlJQUzkJJRERUzqWmpsLGxoaf60RUIVTIe+CeZGlpCX9/f1y5ciXfOkqlEkql8gVGRUREREREVHQVbgjl07KyshAdHQ1XV9eyDoWIiIiIiKhYKlwCN2HCBOzfvx+xsbE4evQoevTogdTUVAwcOLCsQyMiIiIiIiqWCjeE8ubNm+jTpw/u378PR0dHNG7cGEeOHEG1atXKOjQiIiIiIqJiqXAJ3KpVq8o6BCIiIiIiolJR4YZQEhERERERVVRM4IiIiIiIiMoJJnBERERERETlRIW7B46IqCLLzMlGbEoysjQaWJoq4GVjBxOZvKzDIiIioheECRwRUTmg1miw70YMTt5NQGrWY2gBmEgSHMwt0axyNbzq4g5Jkso6TCIiIiplTOCIiF5y2VoN1lz6B/8k3YaZXA4HMwvIZTJkazRIzszAHzEX8PBxJoI9fJjEERERVXBM4IiIXnIn7yTgXNId2CnMoDTJ/bWdoxFIfihBrVZCLXKwN/s6fOwd4WVjX8bREhERUWliAkdE9BLTCi2O374JSQKUJiYQQuD6LS3ibmmQ+VgAQoIAABnw7cNLmNG+IazM+KudiIioouIslERELyl1jhZ7Lyfi/J0k3E/JxtU7aTh9SY0LMTnIygLMlRKsLCRYmUuQQcKZ2HTM23UZ6Vk5ZR06ERERlRL+mZaI6CWQmvUYZ+4l4sy9BJjLM1FNlQMpW4uHGYC3bTYu3pNwNxW4kyiDTALMzeSABGRpNNAIASEXkMsknLh+HzvPW6NbPfeyPiQiIiIqBUzgiIjK2PXUB1h96R+kqtPRyj0HVVVaSEKLHAF4VQLqAXhQTWDBbhNAK0EyFUjPzoFcJiAASAC0QkAmSXiU/Ri/nIhBUM1KsLcwL+MjIyIiopLGIZRERGXo4eNMrLr0D5Iz09HRE/Cx0yIrR8LtNCA5U8K9TCDpsYCNGaDUSlCYApIkAGih0QpIACRJgiRJUMpNYKaQITldjRX//AOt0Jb14REREVEJYwJHRFSGTt+7haTMdNSupISbVTbSc+RIUwtIef8TErRCQlIGIEHA0lQLSFoIAAKARiuQo9VCJslgIpNBLkmQSzJcT3mIaykPyvrwiIiIqIQxgSMiKiNCCJy6ewumcjk8bNSQSUCOVoJWK3LHRQIAJEAAGi3gYK2FBAkKk9zNuipCglYDCAFk5wBKUwkyUy2i7t8uk+MiIiKi0sN74IiIyohGaJGRrYZCJoe1Ihsa8e8GCbmX1/6lFQAg4OOkxeVEwFIGWJpLkMsArVZCmhpIzRJQ5whk5wDurnKYKSQ8eJz54g+KiIiIShWvwBERlRG5JIOJTA6N0EKrlXRX1ExlMggBCAhotFoIkZvN1XTNQWDVHGg1ckArwVQCLEwBZysBV0vgcZYWVhYSqrnKoRECSrm87A6OiIiISgWvwBERlRFJkuBfyRl/3YzF7QwTOFlkAxAwNZFBrdFCo/l3lkkZIGkleDsJVLHLQkYOcCXRBA8y8pK+3GGVPs4aeHqaQK4QePQYqG5XqUyPj4iIiEoeEzgiojJU18kNJ+8kIOp+NnzsZLAw0SItO+8K3L8zTgrATgkoTIF7jyS081ejUXU1Lt8yQUaWHAoTAQ/HHDTy0ODaIxkib2bB3swctR2cy/rwiIiIqIQxgSMiKkNuVip09KqJP2Oi8XdCDl5z08JMlo1MuYBamzvO3VwmwdEKyFQDdx/JAEhwstbA2U8NmSRBBgkarYCQgMpWGVApVHjLuzYsTBVlfXhERERUwpjAERGVsQbOVWBvZoGjifH488oNBDpp4WwJmEoSNBoZUjOA1EyBpDQJGq0ESQJyskwgM9FAMtVCKxOQZIBGSLA3U2CgXx24WXP4JBERUUXEBI6I6CXgZWMPlcwSE9dl4tCZLNgos+BgKSFHI8ONJBnGt3/07wLeucsFCCGQo5ZB5Mhhago4WCpQWWUCE0kOMxPbsj0YIiIiKjVM4IiIXhLRialIyxSwMrPA1dsC1+UyyP6dK/jybVM08FAjNSP33jitNrdcIZfB3dYCDtZKCJEGmeQISeKvdiIiooqKywgQEb0k1BotJAC25qYwM5UhW6PVbTseq0C2VoK9lYBMAuQyCWYKOWpWVqGStRIQWQDkMJFVLrP4iYiIqPQxgSMieknYWSggSRKyNQIuNuaQAKiztYAALt02we7zSijkAi62WpgrBFxUpjAz0UIr0iGggYmsGmSSQ1kfBhEREZUijrMhInpJ1HazgYuNGe49yoKzygwae4FbDzPxOFsDQMKm0wrcSAJa1MxGgLscjioJQA5kki1MZFUgl1wgSdKzuiEiIqJyjAkcEdFLQmEiQ6c6lRF24BoeZKhRyUoBG3NTJKerkZ6VA3WOFudvyVHJuio6+FeBJGkgSSaQYMXEjYiI6D+CCRwR0Uukla8TUjPV2HTmFm4kZ0JpKoNMAhQALHK0qF1ZhSGNqsFEblnWoRIREVEZYAJHRPQSEdCgbYAc/tVscCEhDScuaZF8MRXS3XQotQJpN1Kx5Owd+AW44LU3XoFLZVVZh0xEREQvEBM4IqKXgBBaPMiKwcOsq8gWGYBcoKqdwKUYDTISJViaW8LKSglJAjIzs3Hs4HXEXLqPd4Y0gLuHXVmHT0RERC8IZ6EkIipjQgjcy4zCvcf/IEc8hqlkAYVchRM7JSTG5cDaPhtmNlkwUchgYiqHtcoMzq7WeJCcgfW/nUVOtqasD4GIiIheECZwRERlLCPnHh6qr0EGBUxllpAkOVKTNYi9kA0LKzlMTE2g0T5GjjZTt49MJsHOwQJ3bqXiysV7ZRg9ERERvUhM4IiIyliq+jq0IgcmMqWu7FZsNrIyBcytJEiQAEjI1mZAPLGfQiGHRqPFtStJLzxmIiIiKhtM4IiIylhGzj3IJFO9spyc3H/zlgeQJBmEyIEQTw+XlJCtznkBURIREdHLgAkcEVEZEkIAEP9eZfsfS5UMkgzIyf7fNTfxxP8/ua/K1vxFhEpEREQvASZwRERlSJIkKOQ20IrsJ0oFqnhpYGMPpKfkABAQQgtJkkGS5Lpa6WlqmFsoUKuO6wuPm4iIiMoGEzgiojJmo6gKSBK0IgcQmYAmCaYmjxD4mgZCCKQ9zIYQOTCRFLordRkZaqSnqRFQrzKcXa3L+AiIiIjoReE6cERE+ch5kILH128CGg1MXZwgmcjx6PAppJ+9AG12NhSuTlA1aQDLgJqQTJ7/16mVaWVYmDghXX0TJsiGDLlX2mo1BDLTNTj1lxYP70gwVz5GuiwV6mwJpqZy1G3kjk7da5fgERMREdHLjgkcEdFTsu8nI2njDjw6dgaaR+mAEBA5OdBkPoakMIWJhQUgl+Hx1Tg8OnQSlnX84PJef5iorJ6rP5lkAlezANzJuYV0LZADLSC0AIA6LSS84qdA4j9K3IhVQysEXNyrIbChOzy9HSCTSc9onYiIiCoSJnBERE/IvpeEm98swePYGzCxtoLCxRE5qY+QGXMdyMmBZGICmUwGucoapq7OEOpspJ34B7eXrITbuOGQ5PJnd2KEifYBKsst8VhuhzSRCY1WA5kkg6XMHBbuZpCqSoDIBkQOYO4JyG1L9sCJiIioXGACR0T0hLu//oGMizGQm5tBk5EBbXY21LfvAdm5U/ULtRrq2/cgPUiBLNEUClcnyB3skP7PRWREX4Vl7RrP17H2ESRJgrnMDOYwy6eSCSAeA9o0JnBERET/UUzgiIj+9ej4GTzYsgfa7GxoU9NyJ+zX5ED8m7xBLgNkckArAJkMIjsHWTcSoaziCqHOxqMjp54/gSMiIiIqBM5CSUQEICvhNm4tioAmMxOSQgGZuRIycyWEVvu/SuJ/a7BJWi1kSgUgBNS37wIyGdR37j9/AHLr3CXenujDUA4gmQAyzjpJRET0X8UEjogIQPLm3chJfph7j5s891ejyMnRT6i0Ty+iDUgKUwh1NjTpGZCZK58/ALkrIFPmLiNgjBCA9jEgtwNkqufvh4iIiMo1JnBE9J+XnfwQacfPQq6ygiSTIPISNa0WgAQ8OdGjJveKnCTL/fUpSRKEEBA52bD0r/n8QciUgKL6v/2m62ahBJA7cYk2DZCZAQofQOLMk0RERP9VvAeOiP7z1Il3oMl8DFNHe2gePoI26zEkpRK6zE2S/S+hErn3v0mmub8+BQCh0UBuaQHrRnWLF4hJVQAyQH0VEBm5fYl/+5fbAko/Tl5CRET0H8cEjoj+8yRJ9m+qJoOpSyVkxSdAq84G8tZYk6TcR96wSVMTQJKgzcmBVp0NSSaDfafWMLEp5r1pkgSYugMmrkDO3dwkDhIgswHk9rmJHBEREf2nMYEjov88hZsL5FaW0DxKg4m9LYRGC3XindzZJyX8O5Ty32GVcjkkmQxCrYaQJMjNlFBWcUWlnp1LLiDJBDCtXHLtERERUYXBBI6I/vNMbKxh3aQ+krfuhYnKGgpHB5jYWCPnQQpyHqRAk/oIQquFZGoKixqvABKgVauheZQBuZUFnIf0gdw8v7XbiIiIiEoOEzgiIgD2nVsjI/oKsmJvwMTOBjIrSyicHWHqYAd14h1kJz2EzNwMOQ9Tc6/KSRLMqrqhUs9OsG4UWNbhExER0X8EEzgiIgCmDnaoMv5d3PlpLTLOX0JOQqputkdTp0pwGtADiqpuUMffAoQWClcnWAT4QmZqWsaRExER0X+JJESBq8b+J6WmpsLGxgYpKSlQqbjeEtF/TdaNW8iIvgqRnQMTOxUsA2tBbmFe1mER0XPi5zoRVSS8AkdE9BSle2Uo3TmJCBEREb18OCc1ERERERFROcEEjoiIiIiIqJxgAkdERERERFROMIEjIiIiIiIqJ5jAERERERERlRNM4IiIiIiIiMoJJnBERERERETlBBM4IiIiIiKicoIJHBERERERUTnBBI6IiIiIiKicYAJHRERERERUTjCBIyIiIiIiKieYwBEREREREZUTTOCIiIiIiIjKiQqbwH3//ffw9PSEmZkZ6tevjwMHDpR1SERERERERMVSIRO41atX48MPP8Rnn32G06dP4/XXX0f79u0RHx9f1qERERERERE9N0kIIco6iJL26quvol69eli8eLGuzNfXF127dsXMmTOfuX9qaipsbGyQkpIClUpVmqESERFRKePnOhFVJBXuCpxarcbJkyfRtm1bvfK2bdvi0KFDZRQVERERERFR8ZmUdQAl7f79+9BoNHB2dtYrd3Z2xu3bt43uk5WVhaysLN3z1NTUUo2RiIiIiIjoeVS4K3B5JEnSey6EMCjLM3PmTNjY2Oge7u7uLyJEIiIiIiKiIqlwCVylSpUgl8sNrrbdvXvX4KpcnkmTJiElJUX3uHHjxosIlYiIiIiIqEgqXAKnUChQv3597Nq1S698165daNq0qdF9lEolVCqV3oOIiIiIiOhlU+HugQOAcePGoX///mjQoAGaNGmCH3/8EfHx8XjvvffKOjQiIiIiIqLnViETuF69eiEpKQkzZsxAYmIiateuja1bt6JatWplHRoREREREdFzq5DrwBUX14shIiKqOPi5TkQVSYW7B46IiIiIiKiiYgJHRERERERUTjCBIyIiIiIiKieYwBEREREREZUTTOCIiIiIiIjKCSZwRERERERE5QQTOCIiIiIionKCCRwREREREVE5wQSOiIiIiIionGACR0REREREVE4wgSMiIiIiIionmMARERERERGVE0zgiIiIiIiIygkmcEREREREROUEEzgiIiIiIqJyggkcERERERFROcEEjoiIiIiIqJxgAkdERERERFROMIEjIiIiIiIqJ5jAERERERERlRNM4IiIiIiIiMoJJnBERERERETlBBM4IiIiIiKicoIJHBERERERUTnBBI6IiIiIiKicYAJHRERERERUTjCBIyIiIiIiKieYwBEREREREZUTTOCIiIiIqNwLCQlB165dyzqMlwbPR8XFBI6IiIiIXghJkgp8hISEvPCY9u3bhw4dOsDBwQEWFhbw8/PD+PHjkZCQUOg2WrRogQ8//LD0gnwOCxYsQERERFmHQaWACRwRERERvRCJiYm6x/z586FSqfTKFixY8ELjWbJkCVq3bg0XFxesW7cOFy5cwA8//ICUlBTMmTPnhcZSUjQaDbRaLWxsbGBra1vW4VApYAJHRERERC+Ei4uL7mFjYwNJknTPTU1N8d5776FKlSqwsLCAv78/fvvtN739165dC39/f5ibm8PBwQGtW7dGenq60b5OnjwJJycnfPHFF0a337x5E2PGjMGYMWMQFhaGFi1awMPDA82bN8eyZcswdepUAEBSUhL69OmTb1whISHYv38/FixYoLuSGBcXBwC4cOECOnToACsrKzg7O6N///64f/++bt9Hjx6hb9++sLS0hKurK+bNm2dwNe/BgwcYMGAA7OzsYGFhgfbt2+PKlSu67REREbC1tcXmzZvh5+cHpVKJ69evGwyhFEJg9uzZ8PLygrm5OerUqYO1a9fq9dO3b184OjrC3Nwc1atXR3h4eMEvKJUJJnBEREREVOYeP36M+vXrY/PmzTh37hyGDx+O/v374+jRowByr9716dMHgwcPRnR0NCIjI/HWW29BCGHQVmRkJFq1aoXp06fjs88+M9rf77//DrVajYkTJxrdnnf16llxLViwAE2aNMGwYcN0VxLd3d2RmJiIoKAgBAYG4sSJE9i+fTvu3LmDnj176voYN24cDh48iE2bNmHXrl04cOAATp06pRdHSEgITpw4gU2bNuHw4cMQQqBDhw7Izs7W1cnIyMDMmTOxbNkynD9/Hk5OTgbHM3nyZISHh2Px4sU4f/48xo4di379+mH//v0AgClTpuDChQvYtm0boqOjsXjxYlSqVCm/l4vKkElZB0BERERE5ObmhgkTJuiev//++9i+fTt+//13vPrqq0hMTEROTg7eeustVKtWDQDg7+9v0M4ff/yB/v37Y8mSJejTp0++/V25cgUqlQqurq7FisvGxgYKhQIWFhZwcXHR1Vu8eDHq1auHL7/8UlcWFhYGd3d3XL58Ga6urlixYgV+/fVXtGrVCgAQHh6OypUr68W4adMmHDx4EE2bNgUA/PLLL3B3d8fGjRvx9ttvAwCys7Px/fffo06dOkaPIT09HXPnzsXevXvRpEkTAICXlxf+/vtvLFmyBEFBQYiPj0fdunXRoEEDAICHh0eB54XKDhM4IiIiIipzGo0GX331FVavXo2EhARkZWUhKysLlpaWAIA6deqgVatW8Pf3R3BwMNq2bYsePXrAzs5O18bRo0exefNm/P777+jWrVuB/QkhIElSsePKz8mTJ7Fv3z5YWVkZbIuJiUFmZiays7PRqFEjXbmNjQ1q1Kihex4dHQ0TExO8+uqrujIHBwfUqFED0dHRujKFQoGAgIB8Y7lw4QIeP36MNm3a6JWr1WrUrVsXADBixAh0794dp06dQtu2bdG1a1dd0kgvFyZwRERERFTm5syZg3nz5mH+/Pnw9/eHpaUlPvzwQ6jVagCAXC7Hrl27cOjQIezcuROLFi3CZ599hqNHj8LT0xMA8Morr8DBwQFhYWHo2LEjFApFvv35+PggJSUFiYmJBV6Fe1Zc+dFqtejcuTNmzZplsM3V1VV3H9vTSeSTQ0KNDQ/NK39yP3Nz8wKTUa1WCwDYsmUL3Nzc9LYplUoAQPv27XH9+nVs2bIFu3fvRqtWrTBq1Ch88803BR0mlQHeA0dEREREZe7AgQPo0qUL+vXrhzp16sDLy0tvsg4gN9lp1qwZpk+fjtOnT0OhUGDDhg267ZUqVcLevXsRExODXr166d0n9rQePXpAoVBg9uzZRrc/fPiw0HEpFApoNBq9snr16uH8+fPw8PCAt7e33sPS0hKvvPIKTE1NcezYMd0+qampem37+fkhJydHd78dkDupyuXLl+Hr65vvsT0tb3KT+Ph4g1jc3d119RwdHRESEoKff/4Z8+fPx48//ljoPujFYQJHRERERGXO29tbd4UtOjoa7777Lm7fvq3bfvToUXz55Zc4ceIE4uPjsX79ety7d88gkXFycsLevXtx8eJF9OnTBzk5OUb7c3d3x7x587BgwQIMGTIE+/fvx/Xr13Hw4EG8++67+PzzzwsVF5B7v9jRo0cRFxeH+/fvQ6vVYtSoUUhOTkafPn1w7NgxXLt2DTt37sTgwYOh0WhgbW2NgQMH4qOPPsK+fftw/vx5DB48GDKZTHc1rXr16ujSpQuGDRuGv//+G2fPnkW/fv3g5uaGLl26FPrcWltbY8KECRg7dixWrFiBmJgYnD59Gt999x1WrFgBAJg6dSr++OMPXL16FefPn8fmzZuLlCTSi8MEjoiIiIjK3JQpU1CvXj0EBwejRYsWcHFx0ZsGX6VS4a+//kKHDh3g4+ODyZMnY86cOWjfvr1BWy4uLti7dy+ioqLQt29fg6tjeUaOHImdO3ciISEB3bp1Q82aNTF06FCoVCrdxCXPigsAJkyYALlcDj8/Pzg6OiI+Ph6VK1fGwYMHodFoEBwcjNq1a+ODDz6AjY0NZLLcr+Bz585FkyZN0KlTJ7Ru3RrNmjWDr68vzMzMdG2Hh4ejfv366NSpE5o0aQIhBLZu3QpTU9Mind/PP/8cU6dOxcyZM+Hr64vg4GD8+eefuuGnCoUCkyZNQkBAAJo3bw65XI5Vq1YVqQ96MSSR3+Da/7DU1FTY2NggJSUFKpWqrMMhIiKiYuDnOpUX6enpcHNzw5w5czBkyJCyDodeUpzEhIiIiIioDJw+fRoXL15Eo0aNkJKSghkzZgBAkYZH0n8PEzgiIiIiojLyzTff4NKlS1AoFKhfvz4OHDjABbSpQEzgiIiIiIjKQN26dXHy5MmyDoPKGU5iQkREREREVE4wgSMiIiIiIionmMARERERERGVE0zgiIiIiIiIygkmcEREREREROUEEzgiIiIiIqJyggkcEREREb00QkJC0LVr17IOg+ilxQSOiIiIqJiEVouUyzdw7/hFpFy+AaHVvrC+Dx06BLlcjnbt2r2wPp8UGRkJSZLw8OHDIu0XFxcHSZJw5swZvfIFCxYgIiKixOIjqmi4kDcRERFRMSSdvoKrP+3EwwvXoXmshtxMAVu/avAe0BYOdauXev9hYWF4//33sWzZMsTHx6Nq1aql3mdpsrGxKesQiF5qvAJHRERE9JySTl/BmRkrcf/EZSjsVbCu7gaFvQr3T1zGmRkrkXT6Sqn2n56ejjVr1mDEiBHo1KmT3pWrvCtje/bsQYMGDWBhYYGmTZvi0qVLujqhoaEIDAzEypUr4eHhARsbG/Tu3RuPHj3S1cnKysKYMWPg5OQEMzMzvPbaazh+/DiA3KtoLVu2BADY2dlBkiSEhIQAALZv347XXnsNtra2cHBwQKdOnRATE6Nr19PTEwBQt25dSJKEFi1aADAcQllQ/4U9TqKKhAkcERER0XMQWi2u/rQTj++nwMa3KhQqC8jkcihUFrDxrYrH91NwdeWuUh1OuXr1atSoUQM1atRAv379EB4eDiGEXp3PPvsMc+bMwYkTJ2BiYoLBgwfrbY+JicHGjRuxefNmbN68Gfv378dXX32l2z5x4kSsW7cOK1aswKlTp+Dt7Y3g4GAkJyfD3d0d69atAwBcunQJiYmJWLBgAYDc5HLcuHE4fvw49uzZA5lMhm7dukH77/k4duwYAGD37t1ITEzE+vXrjR5jQf0X5TiJKgomcERERETPIfVqAh5euA6LKo6QJElvmyRJsKjiiIfn45B6NaHUYli+fDn69esHAGjXrh3S0tKwZ88evTpffPEFgoKC4Ofnh08++QSHDh3C48ePddu1Wi0iIiJQu3ZtvP766+jfv7+ujfT0dCxevBhff/012rdvDz8/PyxduhTm5uZYvnw55HI57O3tAQBOTk5wcXHRDYHs3r073nrrLVSvXh2BgYFYvnw5oqKicOHCBQCAo6MjAMDBwQEuLi66dp70rP6LcpxEFQUTOCIiIqLnoE5Jh+axGiaWSqPbTSyU0DxWQ52SXir9X7p0CceOHUPv3r1z+zMxQa9evRAWFqZXLyAgQPffrq6uAIC7d+/qyjw8PGBtba1XJ297TEwMsrOz0axZM912U1NTNGrUCNHR0QXGFxMTg3feeQdeXl5QqVS6IZPx8fGFPsai9P+s4ySqKDiJCREREdFzUNhYQm6mQE56FhQqC4PtORlZkJspoLCxLJX+ly9fjpycHLi5uenKhBAwNTXFgwcPdGWmpqa6/867Uqh9Yljnk9vz6uRtzxuO+fQVRiGEQdnTOnfuDHd3dyxduhSVK1eGVqtF7dq1oVarC32MRen/WcdJVFHwChwRERHRc1B5u8HWrxoybt4zuO9MCIGMm/dgW8sDKm+3fFp4fjk5Ofjpp58wZ84cnDlzRvc4e/YsqlWrhl9++aVE+vH29oZCocDff/+tK8vOzsaJEyfg6+sLAFAoFAAAjUajq5OUlITo6GhMnjwZrVq1gq+vr15Smd9+z9M/0X8Nr8ARERERPQdJJoP3gLZIi7uDlOh4WFRxhImFEjkZWci4eQ9mjjbw7t8Gkqzk/16+efNmPHjwAEOGDDGYdr9Hjx5Yvnw55s2bV+x+LC0tMWLECHz00Uewt7dH1apVMXv2bGRkZGDIkCEAgGrVqkGSJGzevBkdOnSAubk57Ozs4ODggB9//BGurq6Ij4/HJ598ote2k5MTzM3NsX37dlSpUgVmZmYGx1KY/on+a3gFjoiIiOg5OdStjsCp/VGpgQ/Uyal4dDUB6uRUVGpYA4FT+pfaOnDLly9H69atja6Z1r17d5w5cwanTp0qkb6++uordO/eHf3790e9evVw9epV7NixA3Z2dgAANzc3TJ8+HZ988gmcnZ0xevRoyGQyrFq1CidPnkTt2rUxduxYfP3113rtmpiYYOHChViyZAkqV66MLl26PFf/RP81knj6mj8hNTUVNjY2SElJgUqlKutwiIiIqBhexOe60GqRejUB6pR0KGwsofJ2K5Urb0REHEJJREREVEySTAYbH/eyDoOI/gP4pyEiIiIiIqJyggkcERERERFROcEEjoiIiIiIqJxgAkdERERERFROVLgEzsPDA5Ik6T2eXneEiIiIiIioPKqQs1DOmDEDw4YN0z23srIqw2iIiIiIiIhKRoVM4KytreHi4lLWYRAREREREZWoCjeEEgBmzZoFBwcHBAYG4osvvoBarS7rkIiIiIiIiIqtwiVwH3zwAVatWoV9+/Zh9OjRmD9/PkaOHFngPllZWUhNTdV7EBEREVV0oaGhCAwM1D0PCQlB165dS7XPFi1a4MMPP9Q99/DwwPz580u1z/JGkiRs3LixVPuIiIiAra1tqfZBpaNcJHChoaEGE5M8/Thx4gQAYOzYsQgKCkJAQACGDh2KH374AcuXL0dSUlK+7c+cORM2Nja6h7u7+4s6NCIiIqoAhFaLrPgEZFy4gqz4BAitttT7vH37Nt5//314eXlBqVTC3d0dnTt3xp49e0q9bypdiYmJaN++fYm1ZyxJ7tWrFy5fvlxifdCLUy7ugRs9ejR69+5dYB0PDw+j5Y0bNwYAXL16FQ4ODkbrTJo0CePGjdM9T01NZRJHREREhZJx6RoebNuLx7E3INRqSAoFzDzdYdf+DVjU8CqVPuPi4tCsWTPY2tpi9uzZCAgIQHZ2Nnbs2IFRo0bh4sWLpdIvvRgvYi4Hc3NzmJubl3o/VPLKxRW4SpUqoWbNmgU+zMzMjO57+vRpAICrq2u+7SuVSqhUKr0HERER0bNkXLqGO2GrkHnxKuQqayiquEGuskbmxau4E7YKGZeulUq/I0eOhCRJOHbsGHr06AEfHx/UqlUL48aNw5EjR3T1UlJSMHz4cDg5OUGlUuGNN97A2bNni9V3VlYWJk6cCHd3dyiVSlSvXh3Lly/Xbb9w4QI6dOgAKysrODs7o3///rh//36x+gwLC0OtWrWgVCrh6uqK0aNH67bFx8ejS5cusLKygkqlQs+ePXHnzh3d9rxhomFhYahatSqsrKwwYsQIaDQazJ49Gy4uLnBycsIXX3yh16ckSViyZAk6deoECwsL+Pr64vDhw7h69SpatGgBS0tLNGnSBDExMbp9jA1B/fDDD9GiRQvd8xYtWmDMmDGYOHEi7O3t4eLigtDQUIO+nxxCefPmTfTu3Rv29vawtLREgwYNcPToUQBATEwMunTpAmdnZ1hZWaFhw4bYvXu3Xn/Xr1/H2LFjdSPXAONDKBcvXoxXXnkFCoUCNWrUwMqVKw3iWrZsGbp16wYLCwtUr14dmzZtMv6iUakpFwlcYR0+fBjz5s3DmTNnEBsbizVr1uDdd9/Fm2++iapVq5Z1eERERFSBCK0WD7bthSYlFYpq7pBbWkCSyyC3tICimjs0Kal4sG1fiQ+nTE5Oxvbt2zFq1ChYWloabM/7Ui6EQMeOHXH79m1s3boVJ0+eRL169dCqVSskJyc/d/8DBgzAqlWrsHDhQkRHR+OHH37QLdmUmJiIoKAgBAYG4sSJE9i+fTvu3LmDnj17Pnd/ixcvxqhRozB8+HBERUVh06ZN8Pb21h1j165dkZycjP3792PXrl2IiYlBr1699NqIiYnBtm3bsH37dvz2228ICwtDx44dcfPmTezfvx+zZs3C5MmT9ZJfAPj8888xYMAAnDlzBjVr1sQ777yDd999F5MmTdLdvvNkMllYK1asgKWlJY4ePYrZs2djxowZ2LVrl9G6aWlpCAoKwq1bt7Bp0yacPXsWEydOhPbf91VaWho6dOiA3bt34/Tp0wgODkbnzp0RHx8PAFi/fj2qVKmCGTNmIDExEYmJiUb72bBhAz744AOMHz8e586dw7vvvotBgwZh3759evWmT5+Onj174p9//kGHDh3Qt2/fYr2fqOjKxRDKwlIqlVi9ejWmT5+OrKwsVKtWDcOGDcPEiRPLOjQiIiKqYNQ3E/E49gZMHCvprmrkkSQJJo6V8Dg2HuqbiVBWdSuxfq9evQohBGrWrFlgvX379iEqKgp3796FUqkEAHzzzTfYuHEj1q5di+HDhxe578uXL2PNmjXYtWsXWrduDQDw8vrfMNHFixejXr16+PLLL3VlYWFhcHd3x+XLl+Hj41PkPv/v//4P48ePxwcffKAra9iwIQBg9+7d+OeffxAbG6u7/WXlypWoVasWjh8/rqun1WoRFhYGa2tr+Pn5oWXLlrh06RK2bt0KmUyGGjVqYNasWYiMjNTdfgMAgwYN0iWfH3/8MZo0aYIpU6YgODgYQO7keYMGDSryMQUEBGDatGkAgOrVq+Pbb7/Fnj170KZNG4O6v/76K+7du4fjx4/D3t4eAHQJLADUqVMHderU0TtfGzZswKZNmzB69GjY29tDLpc/c5mtb775BiEhIbrJ//Ku5n7zzTdo2bKlrl5ISAj69OkDAPjyyy+xaNEiHDt2DO3atSvyeaDnU6ESuHr16hn85YSIiIioNGjSMiDUasjyuY1DZqZEzv0kaNIySrRfIQQAGCSNTzt58iTS0tIM5gDIzMzUG/aXn19++QXvvvuu7vm2bduQmJgIuVyOoKCgfPvct2+f7orck2JiYoqcwN29exe3bt1Cq1atjG6Pjo6Gu7u73twFfn5+sLW1RXR0tC6B8/DwgLW1ta6Os7Mz5HI5ZDKZXtndu3f12g8ICNDbDgD+/v56ZY8fP0ZqamqRbsF5sl0g91afp/vOc+bMGdStW1eXvD0tPT0d06dPx+bNm3Hr1i3k5OQgMzNTdwWusKKjow2S+mbNmmHBggX5xm5paQlra+t8Y6fSUaESOCIiIqIXRW5lAUmhgPbxY8gtLQy2ax9nQVIoILcy3FYc1atXhyRJiI6OLnDKf61WC1dXV0RGRhpsK8z08W+++SZeffVV3XM3Nze9e6vy67Nz586YNWuWwbaC5iPIz7Mm2RBCGE1kny43NTXV2y5JktEy7VPDXZ+sk9eesbK8/WQymS7BzpOdnW0QX2H6zvOsc/DRRx9hx44d+Oabb+Dt7Q1zc3P06NHjudZBfvpcGju/RYmdSkeFugeOiIiI6EVRVHGFmac7cu7dN/jSLoRAzr37MPOsCkWVoicuBbG3t0dwcDC+++47pKenG2x/+PAhgNyRSbdv34aJiQm8vb31HpUqVXpmP9bW1nr7mJubw9/fH1qtFvv37ze6T7169XD+/Hl4eHgY9Gnsfr3CxODh4ZHv0gh+fn6Ij4/HjRs3dGUXLlxASkoKfH19i9xfcTk6OhrcY3bmzJlitRkQEIAzZ87ke5/ZgQMHEBISgm7dusHf3x8uLi6Ii4vTq6NQKKDRaArsx9fXF3///bde2aFDh8rkPFLBmMARERERPQdJJoNd+zcgt1FBff0GNOkZEBoNNOkZUF+/AbmNCnbtW0KSlfzXre+//x4ajQaNGjXCunXrcOXKFURHR2PhwoVo0qQJAKB169Zo0qQJunbtih07diAuLg6HDh3C5MmTdRNwFJWHhwcGDhyIwYMHY+PGjYiNjUVkZCTWrFkDABg1ahSSk5PRp08fHDt2DNeuXcPOnTsxePDgZyYQ+QkNDcWcOXOwcOFCXLlyBadOncKiRYt0xxgQEIC+ffvi1KlTOHbsGAYMGICgoCA0aNDguforjjfeeAMnTpzATz/9hCtXrmDatGk4d+5csdrs06cPXFxc0LVrVxw8eBDXrl3DunXrcPjwYQC598OtX78eZ86cwdmzZ/HOO+8YXBHz8PDAX3/9hYSEhHxnBP3oo48QERGBH374AVeuXMHcuXOxfv16TJgwoVjxU8ljAkdERET0nCxqeMF5cG+Y1/SGJvUR1DdvQZP6COY1q8N5cO9SWwfO09MTp06dQsuWLTF+/HjUrl0bbdq0wZ49e7B48WIAuUPbtm7diubNm2Pw4MHw8fFB7969ERcXp7uf63ksXrwYPXr0wMiRI1GzZk0MGzZMdyWwcuXKOHjwIDQaDYKDg1G7dm188MEHsLGx0bvfrCgGDhyI+fPn4/vvv0etWrXQqVMnXLlyRXeMGzduhJ2dHZo3b47WrVvDy8sLq1evfu7jK47g4GBMmTIFEydORMOGDfHo0SMMGDCgWG0qFArs3LkTTk5O6NChA/z9/fHVV19BLpcDAObNmwc7Ozs0bdoUnTt3RnBwMOrVq6fXxowZMxAXF4dXXnkFjo6ORvvp2rUrFixYgK+//hq1atXCkiVLEB4errcEAr0cJPH0NX9CamoqbGxskJKSwjXhiIiIyrkX8bkutFqobyZCk5YBuZUFFFVcS+XKGxERJzEhIiIiKiZJJivRpQKIiPLDPw0RERERERGVE0zgiIiIiIiIygkmcEREREREROUEEzgiIiIiIqJyggkcERERERFROcEEjoiIiIiIqJxgAkdERERERFROMIEjIiIiIiIqJ5jAEREREVVQkiRh48aNZR2GUaGhoQgMDCzrMIjKHSZwRERERMUktFpobt9ETuwlaG7fhNBqX0i/t2/fxvvvvw8vLy8olUq4u7ujc+fO2LNnT4n3FRkZCUmS8PDhwxJpb8KECaUSJ1FFZ1LWARARERGVZznxV6E+tAuaxOtAdjZgagq5azUomraBSVXvUus3Li4OzZo1g62tLWbPno2AgABkZ2djx44dGDVqFC5evFhqfReHEAIajQZWVlawsrIq63CIyh1egSMiIiJ6TjnxV/H4z5+Rc/0KJEsVJOfKkCxVyLl+Jbc8/mqp9T1y5EhIkoRjx46hR48e8PHxQa1atTBu3DgcOXLEoL6xK2hnzpyBJEmIi4sDAFy/fh2dO3eGnZ0dLC0tUatWLWzduhVxcXFo2bIlAMDOzg6SJCEkJARAbkI2e/ZseHl5wdzcHHXq1MHatWsN+t2xYwcaNGgApVKJAwcOGAyhDAkJQdeuXfHNN9/A1dUVDg4OGDVqFLKzs3V1EhMT0bFjR5ibm8PT0xO//vorPDw8MH/+/BI7r0QvO16BIyIiInoOQquF+tAuaNNSIXN1hyRJuRvMLSAzc4c28QbUh3dBXsULkqxk/2aenJyM7du344svvoClpaXBdltb2+dqd9SoUVCr1fjrr79gaWmJCxcuwMrKCu7u7li3bh26d++OS5cuQaVSwdzcHAAwefJkrF+/HosXL0b16tXx119/oV+/fnB0dERQUJCu7YkTJ+Kbb76Bl5cXbG1tsX//foP+9+3bB1dXV+zbtw9Xr15Fr169EBgYiGHDhgEABgwYgPv37yMyMhKmpqYYN24c7t69+1zHSlReMYEjIiIieg7au7egSbwOmV2l/yVv/5IkCTK7StDcug7t3VuQu1Qp0b6vXr0KIQRq1qxZou3Gx8eje/fu8Pf3BwB4eXnpttnb2wMAnJycdAlieno65s6di71796JJkya6ff7++28sWbJEL4GbMWMG2rRpU2D/dnZ2+PbbbyGXy1GzZk107NgRe/bswbBhw3Dx4kXs3r0bx48fR4MGDQAAy5YtQ/Xq1Uvs+InKAyZwRERERM9BZKbn3vOmVBqvoFQCD7Nz65V030IAgEHiWFxjxozBiBEjsHPnTrRu3Rrdu3dHQEBAvvUvXLiAx48fGyRmarUadevW1SvLS7oKUqtWLcjlct1zV1dXREVFAQAuXboEExMT1KtXT7fd29sbdnZ2hTo2ooqC98ARERERPQfJ3BIwNQWysoxXyMoCTE1z65Ww6tWrQ5IkREdHF3of2b/DOPOSPwB695cBwNChQ3Ht2jX0798fUVFRaNCgARYtWpRvm9p/Z9vcsmULzpw5o3tcuHBB7z44AEaHej7N1NRU77kkSbo+noz7SfmVE1VUTOCIiIiInoPMqTLkrtWgfXDfIIkQQkD74D7klatB5lS5xPu2t7dHcHAwvvvuO6SnG17hMzbVv6OjI4DciUDynDlzxqCeu7s73nvvPaxfvx7jx4/H0qVLAQAKhQIAoNFodHX9/PygVCoRHx8Pb29vvYe7u3txDtFAzZo1kZOTg9OnT+vKrl69WmLLGhCVF0zgiIiIiJ6DJJNB0bQNZFYqaBNvQGRmQGg1EJkZ0CbegMxKBUWTNiU+gUme77//HhqNBo0aNcK6detw5coVREdHY+HChbr70Z6Ul1SFhobi8uXL2LJlC+bMmaNX58MPP8SOHTsQGxuLU6dOYe/evfD19QUAVKtWDZIkYfPmzbh37x7S0tJgbW2NCRMmYOzYsVixYgViYmJw+vRpfPfdd1ixYkWJHm/NmjXRunVrDB8+HMeOHcPp06cxfPhwmJubl/hQUqKXGRM4IiIioudkUtUbZp37waRadYj0VIg7tyDSU2HiUT23vBTXgfP09MSpU6fQsmVLjB8/HrVr10abNm2wZ88eLF682KC+qakpfvvtN1y8eBF16tTBrFmz8H//9396dTQaDUaNGgVfX1+0a9cONWrUwPfffw8AcHNzw/Tp0/HJJ5/A2dkZo0ePBgB8/vnnmDp1KmbOnAlfX18EBwfjzz//hKenZ4kf808//QRnZ2c0b94c3bp1w7Bhw2BtbQ0zM7MS74voZSUJDhw2kJqaChsbG6SkpEClUpV1OERERFQML+JzXWi10N69BZGZDsncEjKnyqV25Y3+5+bNm3B3d8fu3bvRqlWrsg6H6IXgLJRERERExSTJZCW+VAAZ2rt3L9LS0uDv74/ExERMnDgRHh4eaN68eVmHRvTCMIEjIiIionIhOzsbn376Ka5duwZra2s0bdoUv/zyi8HslUQVGYdQGsEhlERERBUHP9eJqCLh4GwiIiIiIqJyggkcERERERFROcEEjoiIiIiIqJxgAkdERERERFROMIEjIiIiIiIqJ5jAERERERERlRNM4IiIiIjopRIaGorAwMAC68TFxUGSJJw5c+aFxJQnIiICtra2L7RPY0r6+D08PDB//vwSaYtKFxM4IiIiomISQguRkghxLyb3X6Et9T5v376N999/H15eXlAqlXB3d0fnzp2xZ8+eUu+7tE2YMEHvOEJCQtC1a1e9Ou7u7khMTETt2rVfcHQV0/HjxzF8+PCyDoMKwaSsAyAiIiIqz8T9OGiv/gXxIAHQZgMyU0h2bpB5N4dUyaNU+oyLi0OzZs1ga2uL2bNnIyAgANnZ2dixYwdGjRqFixcvlkq/L4qVlRWsrKwKrCOXy+Hi4vKCIqr4HB0dyzoEKiRegSMiIiJ6TuJ+HLSn10HcvwYorQBrF0BpBXH/2r/lcaXS78iRIyFJEo4dO4YePXrAx8cHtWrVwrhx43DkyBFdvfj4eHTp0gVWVlZQqVTo2bMn7ty5o9ueN1QxLCwMVatWhZWVFUaMGAGNRoPZs2fDxcUFTk5O+OKLL/T6lyQJixcvRvv27WFubg5PT0/8/vvvenWioqLwxhtvwNzcHA4ODhg+fDjS0tJ02yMjI9GoUSNYWlrC1tYWzZo1w/Xr1/XiyvvvFStW4I8//oAkSZAkCZGRkUaHEO7fvx+NGjWCUqmEq6srPvnkE+Tk5Oi2t2jRAmPGjMHEiRNhb28PFxcXhIaG6sU9d+5c+Pv7w9LSEu7u7hg5cqRe3IXx8ccfw8fHBxYWFvDy8sKUKVOQnZ1tcN5XrlwJDw8P2NjYoHfv3nj06JGuzvbt2/Haa6/B1tYWDg4O6NSpE2JiYoz2J4SAt7c3vvnmG73yc+fOQSaT6fYLDQ1F1apVoVQqUblyZYwZM0ZX9+khlAXVpbLFBI6IiIjoOQihzb3ylvUIsHGDpDCHJJNBUpgDNm4QWY9yt5fwcMrk5GRs374do0aNgqWlpcH2vPuzhBDo2rUrkpOTsX//fuzatQsxMTHo1auXXv2YmBhs27YN27dvx2+//YawsDB07NgRN2/exP79+zFr1ixMnjxZLzEEgClTpqB79+44e/Ys+vXrhz59+iA6OhoAkJGRgXbt2sHOzg7Hjx/H77//jt27d2P06NEAgJycHHTt2hVBQUH4559/cPjwYQwfPhySJBkcz4QJE9CzZ0+0a9cOiYmJSExMRNOmTQ3qJSQkoEOHDmjYsCHOnj2LxYsXY/ny5fi///s/vXorVqyApaUljh49itmzZ2PGjBnYtWuXbrtMJsPChQtx7tw5rFixAnv37sXEiRML8cr8j7W1NSIiInDhwgUsWLAAS5cuxbx58wzO+8aNG7F582Zs3rwZ+/fvx1dffaXbnp6ejnHjxuH48ePYs2cPZDIZunXrBq3W8P0kSRIGDx6M8PBwvfKwsDC8/vrreOWVV7B27VrMmzcPS5YswZUrV7Bx40b4+/sbjb8odakMCDKQkpIiAIiUlJSyDoWIiIiKqbQ+17UPb4nsrV+K7D0LRc6BpQaP7D0LRfbWL4X24a0S7ffo0aMCgFi/fn2B9Xbu3CnkcrmIj4/XlZ0/f14AEMeOHRNCCDFt2jRhYWEhUlNTdXWCg4OFh4eH0Gg0urIaNWqImTNn6p4DEO+9955ef6+++qoYMWKEEEKIH3/8UdjZ2Ym0tDTd9i1btgiZTCZu374tkpKSBAARGRlpNPZp06aJOnXq6J4PHDhQdOnSRa9ObGysACBOnz4thBDi008/FTVq1BBarVZX57vvvhNWVla6YwkKChKvvfaaXjsNGzYUH3/8sdE4hBBizZo1wsHBQfc8PDxc2NjY5FvfmNmzZ4v69evrHd/T5/2jjz4Sr776ar5t3L17VwAQUVFRQgjD479165aQy+Xi6NGjQggh1Gq1cHR0FBEREUIIIebMmSN8fHyEWq022n61atXEvHnzClWXyhavwBERERE9D3VG7j1vJkrj200UudvVGSXarRACAIxerXpSdHQ03N3d4e7urivz8/ODra2t7koZkDt0ztraWvfc2dkZfn5+kMlkemV3797Va79JkyYGz/PajY6ORp06dfSuEDZr1gxarRaXLl2Cvb09QkJCEBwcjM6dO2PBggVITEws7CnI93ibNGmid16aNWuGtLQ03Lx5U1cWEBCgt5+rq6vese3btw9t2rSBm5sbrK2tMWDAACQlJSE9Pb3QsaxduxavvfYaXFxcYGVlhSlTpiA+Pl6vztPn/ek4YmJi8M4778DLywsqlQqenp4AYNDOk/t37NgRYWFhAIDNmzfj8ePHePvttwEAb7/9NjIzM+Hl5YVhw4Zhw4YNesNLn1SUuvTiMYEjIiIieh4KC0BmCuRkGd+eo87drrAo0W6rV68OSZL0kjBjhBBGk7yny01NTfW2S5JktMzY0L2n5bWbX99P1gkPD8fhw4fRtGlTrF69Gj4+PgbDNIvCWJ/Gkt2Cju369evo0KEDateujXXr1uHkyZP47rvvAEDvHraCHDlyBL1790b79u2xefNmnD59Gp999hnUarVevWed486dOyMpKQlLly7F0aNHcfToUQAwaOdJQ4cOxapVq5CZmYnw8HD06tULFha57z93d3dcunQJ3333HczNzTFy5Eg0b97c6HEVpS69eEzgiIiIiJ6HyhmSnRuQkaRLFPIIIYCMpNztKucS7dbe3h7BwcH47rvvjF4VevjwIYDcq23x8fG4ceOGbtuFCxeQkpICX1/fYsfxdLJ15MgR1KxZU9f3mTNn9OI7ePAgZDIZfHx8dGV169bFpEmTcOjQIdSuXRu//vqr0b4UCgU0Gk2B8fj5+eHQoUN6r8WhQ4dgbW0NNze3Qh3TiRMnkJOTgzlz5qBx48bw8fHBrVu3CrVvnoMHD6JatWr47LPP0KBBA1SvXl03OUthJSUlITo6GpMnT0arVq3g6+uLBw8ePHO/Dh06wNLSEosXL8a2bdswePBgve3m5uZ48803sXDhQkRGRuLw4cOIiooy2lZR6tKLxQSOiIiI6DlIkix3qQClNZCSAKHOhNBqINSZQEoCJDNV7nap5L9uff/999BoNGjUqBHWrVuHK1euIDo6GgsXLtQNbWzdujUCAgLQt29fnDp1CseOHcOAAQMQFBSEBg0aFDuG33//HWFhYbh8+TKmTZuGY8eO6SYp6du3L8zMzDBw4ECcO3cO+/btw/vvv4/+/fvD2dkZsbGxmDRpEg4fPozr169j586duHz5cr6JpYeHB/755x9cunQJ9+/fN3olaOTIkbhx4wbef/99XLx4EX/88QemTZuGcePG6Q0HLcgrr7yCnJwcLFq0CNeuXcPKlSvxww8/FOm8eHt7Iz4+HqtWrUJMTAwWLlyIDRs2FKkNOzs7ODg44Mcff8TVq1exd+9ejBs37pn7yeVyhISEYNKkSfD29tYb5hoREYHly5fj3LlzumMzNzdHtWrVDNopSl168ZjAERERET0nqZIHZHW7Q6rkBWSlAY/uAFlpkCp5QRb4VqmtA+fp6YlTp06hZcuWGD9+PGrXro02bdpgz549WLx4cW5skoSNGzfCzs4OzZs3R+vWreHl5YXVq1eXSAzTp0/HqlWrEBAQgBUrVuCXX36Bn58fAMDCwgI7duxAcnIyGjZsiB49eqBVq1b49ttvddsvXryI7t27w8fHB8OHD8fo0aPx7rvvGu1r2LBhqFGjBho0aABHR0ccPHjQoI6bmxu2bt2KY8eOoU6dOnjvvfcwZMgQTJ48udDHFBgYiLlz52LWrFmoXbs2fvnlF8ycObNI56VLly4YO3YsRo8ejcDAQBw6dAhTpkwpUhsymQyrVq3CyZMnUbt2bYwdOxZff/11ofYdMmQI1Gq1wdU3W1tbLF26FM2aNUNAQAD27NmDP//8Ew4ODgZtFKUuvXiSePqaPyE1NRU2NjZISUmBSqUq63CIiIioGF7E57oQWiD1Tu6EJQqL3OGVpXDl7WUhSRI2bNiArl27lnUo9JSDBw+iRYsWuHnzJpydS3b4Lr0cTMo6ACIiIqLyTpJkgI1rWYdB/2FZWVm4ceMGpkyZgp49ezJ5q8Aq7p+GiIiIiIj+I3777TfUqFEDKSkpmD17dlmHQ6WIQyiN4BBKIiKiioOf60RUkfAKHBERERERUTnBBI6IiIiIiKicYAJHRERERERUTjCBIyIiIiIiKieKlMAlJCTg6NGjePDggdHt9+/fx08//VQigREREREREZG+QiVwWq0WQ4cORdWqVdG0aVO4uLjgo48+QnZ2tl69mJgYDBo0qFQCJSIiIiIi+q8rVAK3fPlyREREYOTIkYiIiECfPn0wb948tGrVCo8ePSrtGImIiIjoOUmShI0bN5Z1GAWKjIyEJEl4+PDhC+03NDQUgYGBuuchISHo2rXrC42hqF5UjB4eHpg/f36p90NFV6gEbsmSJfjwww+xaNEi9O/fHxEREdi6dSuioqLQpk0bpKSklHacRERERC8tIbQQj+9DpN/M/VdoS7W/kJAQSJIESZJgamoKZ2dntGnTBmFhYdBq9ftOTExE+/btSzWe4mratCkSExNhY2NTpnEsWLAAERERuuctWrTAhx9+WGbxvAgRERGwtbU1KD9+/DiGDx/+4gOiZypUAnflyhWDH/y2bdtiz549uHr1Klq3bp3vfXFEREREFZnIuAUkbAdubtb7V2TcKtV+27Vrh8TERMTFxWHbtm1o2bIlPvjgA3Tq1Ak5OTm6ei4uLlAqlaUaS3EpFAq4uLhAkqQyjcPGxsZoMlPS1Gp1qfdRXI6OjrCwsCjrMMiIQiVwCoXC4H43AKhXrx527dqF2NhYtGrVCklJSSUeIBEREdHLSmTcAm7vAzISALkFoHTI/TcjAbi9r1STOKVSCRcXF7i5uaFevXr49NNP8ccff2Dbtm16V5GeHEIZFxcHSZKwZs0avP766zA3N0fDhg1x+fJlHD9+HA0aNICVlRXatWuHe/fu6fUXHh4OX19fmJmZoWbNmvj+++912/LaXb9+PVq2bAkLCwvUqVMHhw8f1tW5fv06OnfuDDs7O1haWqJWrVrYunUrAONDKNetW4datWpBqVTCw8MDc+bM0YvHw8MDX375JQYPHgxra2tUrVoVP/74o16djz/+GD4+PrCwsICXlxemTJli9DttnieHJ4aEhGD//v1YsGCB7mpnbGwsvL298c033+jtd+7cOchkMsTExBTY7syZM1G5cmX4+PgAyJ0gsFevXrCzs4ODgwO6dOmCuLg43X4ajQbjxo2Dra0tHBwcMHHiRAghDM7D00MdAwMDERoaqnv+8OFDDB8+HM7OzjAzM0Pt2rWxefNmREZGYtCgQUhJSdEdY95+T7cbHx+PLl26wMrKCiqVCj179sSdO3d02/OGo65cuRIeHh6wsbFB7969ebtVKShUAle9enW9H8An1a1bF7t27UJ8fDwGDhxYosERERERvayE0ALJZ4CcDEDpCMiVgCTL/VfpmFuefKbUh1M+6Y033kCdOnWwfv36AutNmzYNkydPxqlTp2BiYoI+ffpg4sSJWLBgAQ4cOICYmBhMnTpVV3/p0qX47LPP8MUXXyA6OhpffvklpkyZghUrVui1+9lnn2HChAk4c+YMfHx80KdPH93VwFGjRiErKwt//fUXoqKiMGvWLFhZWRmN7+TJk+jZsyd69+6NqKgohIaGYsqUKXqJKQDMmTMHDRo0wOnTpzFy5EiMGDECFy9e1G23trZGREQELly4gAULFmDp0qWYN29eoc7lggUL0KRJEwwbNgyJiYlITExE1apVMXjwYISHh+vVDQsLw+uvv45XXnkl3/b27NmD6Oho7Nq1C5s3b0ZGRgZatmwJKysr/PXXX/j77791yXPeFbo5c+YgLCwMy5cvx99//43k5GRs2LChUPHn0Wq1aN++PQ4dOoSff/4ZFy5cwFdffQW5XI6mTZti/vz5UKlUumOcMGGCQRtCCHTt2hXJycnYv38/du3ahZiYGPTq1UuvXkxMDDZu3IjNmzdj8+bN2L9/P7766qsixUuFIArhs88+E9WqVRM5OTn51jl9+rRwcHAQMpmsME2+1FJSUgQAkZKSUtahEBERUTGV1ue6NvOe0F4JF9prq4X2+kbDx7XVudsz75Vov0IIMXDgQNGlSxej23r16iV8fX11zwGIDRs2CCGEiI2NFQDEsmXLdNt/++03AUDs2bNHVzZz5kxRo0YN3XN3d3fx66+/6vXz+eefiyZNmuTb7vnz5wUAER0dLYQQwt/fX4SGhhqNed++fQKAePDggRBCiHfeeUe0adNGr85HH30k/Pz8dM+rVasm+vXrp3uu1WqFk5OTWLx4sdE+hBBi9uzZon79+rrn06ZNE3Xq1NE9f/q8BgUFiQ8++ECvjVu3bgm5XC6OHj0qhBBCrVYLR0dHERERkW+/AwcOFM7OziIrK0tXtnz5clGjRg2h1Wp1ZVlZWcLc3Fzs2LFDCCGEq6ur+Oqrr3Tbs7OzRZUqVfRirFatmpg3b55ef3Xq1BHTpk0TQgixY8cOIZPJxKVLl4zGFh4eLmxsbAzKn2x3586dQi6Xi/j4eN32vNf32LFjQojcc2lhYSFSU1N1dT766CPx6quv5nte6PmYFCbJGz16NJo0aYK0tLR8by4NDAzEoUOH8r1SR0RERFShaB4D2hzA1NT4dpkpkJ2TW+8FEkI8816ygIAA3X87OzsDAPz9/fXK7t69CwC4d+8ebty4gSFDhmDYsGG6Ojk5OQbfC59s19XVFQBw9+5d1KxZE2PGjMGIESOwc+dOtG7dGt27d9er/6To6Gh06dJFr6xZs2aYP38+NBoN5HK5QX+SJMHFxUUXNwCsXbsW8+fPx9WrV5GWloacnByoVKoCz82zuLq6omPHjggLC0OjRo2wefNmPH78GG+//XaB+/n7+0OhUOienzx5ElevXoW1tbVevcePHyMmJgYpKSlITExEkyZNdNtMTEzQoEEDg2GUBTlz5gyqVKmiG7b5PKKjo+Hu7g53d3ddmZ+fH2xtbREdHY2GDRsCyB12+eTxuLq66r0eVDIKlcC5uLigY8eOz6zn4+NTrDcHERERUbkhNwNkJoA2O3fY5NO02bnb5WYvNKzo6Gh4enoWWMf0iaQzL9l7uixvNsu8f5cuXYpXX31Vr528RKqgdvP2Hzp0KIKDg7Flyxbs3LkTM2fOxJw5c/D+++8bxGcsCTWWtJg+lTw/GfeRI0fQu3dvTJ8+HcHBwbCxscGqVasM7qV7HkOHDkX//v0xb948hIeHo1evXs+c8MPS0lLvuVarRf369fHLL78Y1HV0dCx0LDKZzODcPHmfn7m5eaHbyk9+fxR4uryg14NKTqHugSMiIiKipyjtATMnIDsFeDq5ECK33Mwpt94LsnfvXkRFRaF79+4l1qazszPc3Nxw7do1eHt76z2elSg+zd3dHe+99x7Wr1+P8ePHY+nSpUbr+fn54e+//9YrO3ToEHx8fAySxvwcPHgQ1apVw2effYYGDRqgevXquH79epHiVSgU0Gg0BuUdOnSApaUlFi9ejG3btmHw4MFFahfInQzwypUrcHJyMjivNjY2sLGxgaurK44cOaLbJycnBydPntRrx9HREYmJibrnqampiI2N1T0PCAjAzZs3cfny5SId45P8/PwQHx+PGzdu6MouXLiAlJQU+Pr6Fum4qfiYwBERERE9B0mSAfaBgIkFkHUP0GQBQpv7b9Y9wMQSsA/MrVcKsrKycPv2bSQkJODUqVP48ssv0aVLF3Tq1AkDBgwo0b5CQ0Mxc+ZMLFiwAJcvX0ZUVBTCw8Mxd+7cQrfx4YcfYseOHYiNjcWpU6ewd+/efL/8jx8/Hnv27MHnn3+Oy5cvY8WKFfj222+NTrCRH29vb8THx2PVqlWIiYnBwoULizwBiIeHB44ePYq4uDjcv39fdzVJLpcjJCQEkyZNgre3t94wx8Lq27cvKlWqhC5duuDAgQOIjY3F/v378cEHH+DmzZsAgA8++ABfffUVNmzYgIsXL2LkyJEGi52/8cYbWLlyJQ4cOIBz585h4MCBekluUFAQmjdvju7du+tmj9+2bRu2b9+uO8a0tDTs2bMH9+/fR0ZGhkGsrVu3RkBAAPr27YtTp07h2LFjGDBgAIKCgtCgQYMiHzsVDxM4IiIiouckWVQGXFoCFm6AJgPISsr918INcGmRu72UbN++Ha6urvDw8EC7du2wb98+LFy4EH/88Uehr1IV1tChQ7Fs2TJERETA398fQUFBiIiIKNIVOI1Gg1GjRsHX1xft2rVDjRo19JYieFK9evWwZs0arFq1CrVr18bUqVMxY8YMhISEFLq/Ll26YOzYsRg9erRuroYpU6YUen8AmDBhAuRyOfz8/ODo6Ij4+HjdtiFDhkCtVj/X1TcAsLCwwF9//YWqVavirbfegq+vLwYPHozMzEzdfXrjx4/HgAEDEBISgiZNmsDa2hrdunXTa2fSpElo3rw5OnXqhA4dOqBr164Gs2GuW7cODRs2RJ8+feDn54eJEyfqrro1bdoU7733Hnr16gVHR0fMnj3bINa8pSjs7OzQvHlztG7dGl5eXli9evVzHTsVjySKchdkGfviiy+wZcsWnDlzBgqFwuAvEEDuGhWjRo3C3r17YW5ujnfeeQfffPON3k2jz5KamgobGxukpKQU+0ZXIiIiKlsv4nNdCC2QlZw7YYncDFDal9qVN3o5HDx4EC1atMDNmzd1E8EQvQiFmsTkZaFWq/H222+jSZMmWL58ucF2jUaDjh07wtHREX///TeSkpIwcOBACCGwaNGiMoiYiIiI/gskSQaYVSrrMOgFyMrKwo0bNzBlyhT07NmTyRu9cIVO4DIzM7Fx40Zcv34djo6OePPNN4s0Q05JmD59OgAYLOKYZ+fOnbhw4QJu3LiBypVzhyzMmTMHISEh+OKLL3g1jYiIiIiK5bfffsOQIUMQGBiIlStXlnU49B9UqATu1q1baN68OWJjY3XTlE6YMAHbtm1D48aNSzXAojh8+DBq166tS94AIDg4GFlZWTh58iRatmxpdL+srCxkZWXpnqemppZ6rERERERU/oSEhBTpXjyiklaowdmTJ09GQkICJk+ejC1btmD+/PlQKBQYMWJEacdXJLdv3za4jG1nZweFQoHbt2/nu9/MmTN107Xa2NjoLVJIRERERET0sihUArdr1y58+umnmD59Otq3b48xY8YgLCwM//zzD+7cuVOsAEJDQyFJUoGPEydOFLq9wiwy+LRJkyYhJSVF93hyjQsiIiIiIqKXRaGGUN6+fRvNmzfXK2vRogWEELhz506xbt4cPXo0evfuXWAdDw+PQrXl4uKCo0eP6pU9ePAA2dnZBcaoVCqhVCoL1QcREREREVFZKVQCp9FoYG5urldmZmYGIHdF+OKoVKkSKlUqmVmbmjRpgi+++AKJiYlwdXUFkDuxiVKpRP369UukDyIiIiIiorJS6FkoL126BBOT/1XPW/zv4sWLBnXr1atXAqEZio+PR3JyMuLj46HRaHDmzBkAgLe3N6ysrNC2bVv4+fmhf//++Prrr5GcnIwJEyZg2LBhnIGSiIiIiIjKvUIt5C2TyQp1b1ne87zkrqSFhIRgxYoVBuX79u1DixYtAOQmeSNHjjRYyLsoQyS5kDcREVHFwc91IqpICpXAGUuaCjJw4MDnDuhlwF/0REREFQc/14smNDQUGzdu1I10ehZJkrBhwwZ07dq1VOOqCEJCQvDw4UNs3Lix2G1FRETgww8/xMOHD4vdFpUvhRpCWd4TMiIiIqJSJQSgfQQgG4ApILMGCpgBu7jyG5UUHByM7du3F6vtCRMm4P333y9WGy+Loiaj5UmvXr3QoUOHsg6DykCh74F7Fq1Wiy1btiAsLAwbNmwoqWaJiIiIXm6aB4A6FtCmAtAAkAMyFaDwBOR2pdZtu3btEB4erldWnFm1hRDQaDSwsrKClZVVccOjUmZubm4wySD9NxRqHbiCXL58GZ988gmqVKmCLl26YNeuXSURFxEREdHLT/MAyDoHaJIBSQFI1rn/apL/LX9Qal0rlUq4uLjoPezs/pcwSpKEZcuWoVu3brCwsED16tWxadMm3fbIyEhIkoQdO3agQYMGUCqVOHDgAEJDQxEYGKjXV1hYGGrVqgWlUglXV1eMHj1ab/v9+/cL1U/dunVhbm6ON954A3fv3sW2bdvg6+sLlUqFPn36ICMjQ7efEAKzZ8+Gl5cXzM3NUadOHaxdu9ag3T179qBBgwawsLBA06ZNcenSJQC5QwynT5+Os2fP6tYWjoiIMHoujx8/jjZt2qBSpUqwsbFBUFAQTp06pVfnWedTo9FgyJAh8PT0hLm5OWrUqIEFCxbk+/r99NNPcHBwQFZWll559+7dMWDAAADA2bNn0bJlS1hbW0OlUqF+/fq69ZEjIiJga2ur26+gulSxPFcCl5GRgYiICDRv3hy+vr74+uuv4eDggIULFyIhIaGkYyQiIiJ6+Qjx75W3rNwrbpJp7rBJyTT3uTYrd/uzpxsoNdOnT0fPnj3xzz//oEOHDujbty+Sk5P16kycOBEzZ85EdHQ0AgICDNpYvHgxRo0aheHDhyMqKgqbNm2Ct7d3kfsJDQ3Ft99+i0OHDuHGjRvo2bMn5s+fj19//RVbtmzBrl27sGjRIl39yZMnIzw8HIsXL8b58+cxduxY9OvXD/v379dr97PPPsOcOXNw4sQJmJiYYPDgwQByhxiOHz8etWrVQmJiIhITE9GrVy+j5+nRo0cYOHAgDhw4gCNHjqB69ero0KEDHj16VOjj1Gq1qFKlCtasWYMLFy5g6tSp+PTTT7FmzRqjfb799tvQaDR6SeD9+/exefNmDBo0CADQt29fVKlSBcePH8fJkyfxySefwNTU1Gh7RalL5ZwogsOHD4thw4YJlUolJEkSKpVK9OvXT8hkMrF///6iNPVSS0lJEQBESkpKWYdCRERExVRqn+s5KUKk7RMi/aAQGccMH+kHc7fnlPz3iYEDBwq5XC4sLS31HjNmzNDVASAmT56se56WliYkSRLbtm0TQgixb98+AUBs3LhRr+1p06aJOnXq6J5XrlxZfPbZZ/nGUth+du/eraszc+ZMAUDExMToyt59910RHBysa8PMzEwcOnRIr68hQ4aIPn365Nvuli1bBACRmZlp9FgKKycnR1hbW4s///yz0MdpzMiRI0X37t11zwcOHCi6dOmiez5ixAjRvn173fP58+cLLy8vodVqhRBCWFtbi4iICKNth4eHCxsbG93zgupSxVKoe+Dmzp2L5cuX4+LFixBC4LXXXsOQIUPw9ttvQ61W45dffimd7JKIiIjopZWN3Hve8vs6Jf93e3ap9N6yZUssXrxYr8ze3l7v+ZNX1CwtLWFtbY27d+/q1WnQoEG+fdy9exe3bt1Cq1atCoylMP08WcfZ2RkWFhbw8vLSKzt27BgA4MKFC3j8+DHatGmj14ZarUbdunXzbdfV1VUXd9WqVQuM+Ul3797F1KlTsXfvXty5cwcajQYZGRmIj48v0nH+8MMPWLZsGa5fv47MzEyo1WqD4ahPGjZsGBo2bIiEhAS4ubkhPDwcISEhumW6xo0bh6FDh2LlypVo3bo13n77bbzyyitG2ypKXSrfCpXATZgwAZIkoWPHjpg3b57emyE7u3R+KRERERG93EyRm6Tl/PvfT/t3QhOj24rP0tLSYCjj054eQidJErRarUE7+SnsJBmF6efJOpIkFbhP3r9btmyBm5ubXr2nJ2p5ut0n9y+skJAQ3Lt3D/Pnz0e1atWgVCrRpEkTqNXqfPt6OuY1a9Zg7NixmDNnDpo0aQJra2t8/fXXOHr0aL791q1bF3Xq1MFPP/2E4OBgREVF4c8//9RtDw0NxTvvvIMtW7Zg27ZtmDZtGlatWoVu3boZtFWUulS+FSqBCwwMxJkzZ7BlyxbcuXMHgwcPRp8+fbiWChEREf13yaxz73XTJP97D9wTywYIAWgzALl9br1yytraGh4eHtizZw9atmz5wvr18/ODUqlEfHw8goKCnrsdhUIBjUbzzHoHDhzA999/r5uW/8aNG7h//36R+jpw4ACaNm2KkSNH6spiYmKeud/QoUMxb948JCQkoHXr1nB3d9fb7uPjAx8fH4wdOxZ9+vRBeHh4vklZUepS+VWoSUxOnTqF06dPY8SIEbh69SpGjBgBV1dXDBgwAJGRkaUcIhEREdFLSJJylwqQKXOXEBDZgNDm/qtNzS1XeJbaenBZWVm4ffu23qOoSUdhhIaGYs6cOVi4cCGuXLmCU6dO6U02Uhqsra0xYcIEjB07FitWrEBMTAxOnz6N7777zuj6d/nx8PBAbGwszpw5g/v37xvM+JjH29sbK1euRHR0NI4ePYq+ffsWeYp+b29vnDhxAjt27MDly5cxZcoUHD9+/Jn79e3bFwkJCVi6dKluAhYAyMzMxOjRoxEZGYnr16/j4MGDOH78OHx9fQ3aKEpdKv8KPQtlnTp18O233+LWrVv4+eef0bhxY/zyyy946623IEkS1q1bh9u3b5dmrEREREQvF7kdoKyde6VNqAGRlvuv3P7f8tJbB2779u1wdXXVe7z22msl3s/AgQMxf/58fP/996hVqxY6deqEK1eulHg/T/v8888xdepUzJw5E76+vggODsaff/4JT0/PQrfRvXt3tGvXDi1btoSjoyN+++03o/XCwsLw4MED1K1bF/3798eYMWPg5ORUpHjfe+89vPXWW+jVqxdeffVVJCUl6V2Ny49KpUL37t1hZWWFrl276srlcjmSkpIwYMAA+Pj4oGfPnmjfvj2mT59u0EZR6lL5Jwnx/HPbxsXFYfny5VixYgVu3rwJU1NTdO7cWW+NjvIoNTUVNjY2SElJ4TBRIiKicu6FfK4LAWgfIXfCEtPcYZOldOWNKp42bdrA19cXCxcuLOtQqBwoVgKXRwiB7du3Y/ny5diyZQsyMzNLIrYywwSOiIio4uDnOr2skpOTsXPnTvTt2xcXLlxAjRo1yjokKgcKNYnJs0iShPbt26N9+/ZISkoqiSaJiIiIiCq0evXq4cGDB5g1axaTNyq0EkngnuTg4FDSTRIRERERVThxcXFlHQKVQ4VK4ORyeaEblCQJOTk5zx0QERERERERGVeoBE4IARsbG7z55pscO05ERERERFRGCpXAhYSEYO3atVi3bh3eeustDBkypFiLKhIREREREVHRFWoduLCwMCQmJmLBggW4du0aWrZsCW9vb3z55Ze4detWacdIREREREREKMJC3paWlhgyZAj+/vtvREdH46233sKiRYtQrVo1dOzYEXv27CnNOImIiIiIiP7zCp3APalGjRqYPXs2bt68iSlTpmDHjh347rvvSjo2IiIiIqIii4uLgyRJOHPmTKn2ExISgq5du5ZqH0RPe64ELiUlBYsXL0bjxo0RGhqK6tWro0ePHiUdGxEREREZERISAkmSDB5Xr14tkfYjIiJga2tbIm1VBPklhAsWLEBERESZxET/XUVaB27Pnj0ICwvDxo0bIZfL8fbbb2PBggVo2rRpacVHRERE9NITQkAgDUJkQ5JMIcEKkiSVap/t2rVDeHi4Xpmjo2Op9vk8srOzYWpqWtZhlAobG5uyDoH+gwp1BW7GjBnw8vJC27ZtER8fj2+//RaJiYlYvnw5kzciIiL6T9OIh1Br/kGW5izU2qjcfzX/QCMelmq/SqUSLi4ueo+8tXv//PNP1K9fH2ZmZvDy8sL06dP11umdO3cu/P39YWlpCXd3d4wcORJpaWkAgMjISAwaNAgpKSm6K3uhoaEActf73bhxo14ctra2uqtQeVeq1qxZgxYtWsDMzAw///wzACA8PBy+vr4wMzNDzZo18f333xd4fGvXroW/vz/Mzc3h4OCA1q1bIz09Xbe9qO1duHABHTp0gJWVFZydndG/f3/cv39ft12r1WLWrFnw9vaGUqlE1apV8cUXXwAAPD09AQB169aFJElo0aIFAMMhlFlZWRgzZgycnJxgZmaG1157DcePH9dtj4yMhCRJ2LNnDxo0aAALCws0bdoUly5dKjB2oicV6gpcaGgoVCoVhg4diho1auDBgwdYsmSJ0bqSJGHs2LElGiQRERHRy0gjHiJbEw0BNSSYATAHkAMtHiJbkwHIfSGXbF9oTDt27EC/fv2wcOFCvP7664iJicHw4cMBANOmTQMAyGQyLFy4EB4eHoiNjcXIkSMxceJEfP/992jatCnmz5+PqVOn6hILKyurIsXw8ccfY86cOQgPD4dSqcTSpUsxbdo0fPvtt6hbty5Onz6NYcOGwdLSEgMHDjTYPzExEX369MHs2bPRrVs3PHr0CAcOHIAQAgCeq72goCAMGzYMc+fORWZmJj7++GP07NkTe/fuBQBMmjQJS/+/vfsOr6LK/zj+nttvekIgCRIIXTpSVAQpNkSXlcXeFhRRV7GssvjbVRBFxYZ97QqIDSu7lhURARUsgLCiFOkgJNT0ctuc3x+Ru4YECCWE4Of1PPcx98yZM985d+Tmm3PmzAsv8Oijj9KrVy+ys7NZvnw5AN999x3HH388n332Ge3atcPj8VR53qNGjeLdd99l8uTJNGnShAcffJD+/fuzatUqUlJSovVuv/12JkyYQP369bn22mu58sormTt37n71sfyOmWqwLKvaL4fDUZ0mj2j5+fkGMPn5+bUdioiIiBykmvpet23blIUWm5LQbFMaWmDKwgujr9LQAlMSmm3KQouNbduH9LjGGDNkyBDjdDpNbGxs9HXeeecZY4w5+eSTzX333Veh/pQpU0xGRsYe23vrrbdMvXr1ou8nTpxoEhMTK9UDzPvvv1+hLDEx0UycONEYY8zatWsNYB577LEKdTIzM83rr79eoWzcuHGmR48eVcazcOFCA5h169ZVuX1f7e2KY9GiRcYYY0aPHm3OOOOMCvU3btxoALNixQpTUFBgvF6veeGFF6o83u7t7TJkyBBzzjnnGGOMKSoqMm6327z22mvR7cFg0DRs2NA8+OCDxhhjZs2aZQDz2WefRet89NFHBjClpaVVHltkd9UagVu7dm3NZI8iIiIidZShCJtCLHyV7nezLAuMD5tCDEVYxB/y4/fr149nnnkm+j42NhaAhQsXMn/+/Oj0P4BIJEJZWRklJSXExMQwa9Ys7rvvPpYuXUpBQQHhcJiysjKKi4uj7RyMbt26RX/etm0bGzduZNiwYQwfPjxaHg6H93gPWadOnTj11FPp0KED/fv354wzzuC8884jOTn5gNpbuHAhs2bNqnIkcfXq1eTl5REIBDj11FMP9JRZvXo1oVCInj17RsvcbjfHH388y5Ytq1C3Y8eO0Z8zMjIA2Lp1K40bNz7g48vvR7USuCZNmtR0HCIiIiJ1ijEhIEL5tMmqOIGy8no1sJ5JbGwsLVq0qFRu2zZ33XUXgwcPrrTN5/Oxfv16zjrrLK699lrGjRtHSkoKX331FcOGDSMUCu31mJZlRacx7lLVPr9NAm3bBsqnPZ5wwgkV6u26Z293TqeTGTNmMG/ePD799FOefPJJbr/9dr799ltiYmL2uz3bthk4cCAPPPBApW0ZGRmsWbOmyv32x65+2T2ZN8ZUKvvtoi67tu3qJ5F92a9VKEVERESknGW5KU/SwkBVqyxGAOev9Q6fLl26sGLFiiqTO4AFCxYQDoeZMGECDkf5enZvvfVWhToej4dIJFJp3/r165OdnR19v3LlSkpKSvYaT1paGscccwxr1qzh0ksvrfZ5WJZFz5496dmzJ2PGjKFJkya8//773HLLLfvdXpcuXXj33XfJysrC5ar862/Lli3x+/3MnDmTq666qtL2Xfe8VdUnu7Ro0QKPx8NXX33FJZdcApQntwsWLODmm2+uVpwi1VGtBO6UU07Z4za3201aWhqnn346F198cZX/U4iIiIgcbSzicBCPTR4YV4VRFmMMhjIcJGGxfwuAHKwxY8bwhz/8gczMTM4//3wcDgc//PADS5Ys4Z577qF58+aEw2GefPJJBg4cyNy5c3n22WcrtJGVlUVRUREzZ86kU6dOxMTEEBMTwymnnMJTTz3FiSeeiG3b3HbbbdV6RMDYsWO58cYbSUhIYMCAAQQCARYsWEBubi633HJLpfrffvstM2fO5IwzzqBBgwZ8++23bNu2jTZt2hxQe9dffz0vvPACF198MX/7299ITU1l1apVvPnmm7zwwgv4fD5uu+02Ro0ahcfjoWfPnmzbto2ffvqJYcOG0aBBA/x+P5988gmNGjXC5/NVmq4ZGxvLX/7yF/72t7+RkpJC48aNefDBBykpKWHYsGH78xGK7FW1HiOwdetWtm3bVuVrw4YN/Pvf/2bIkCH06tVrn3+FERERETkaWJaFy9kEC0/0GXDG2OX/pQgLT/n2Gn4e3O769+/Phx9+yIwZM+jevTsnnngijzzySPSWmM6dO/PII4/wwAMP0L59e1577TXGjx9foY2TTjqJa6+9lgsvvJD69evz4IMPAjBhwgQyMzPp3bs3l1xyCSNHjoxOadybq666ihdffJFJkybRoUMH+vTpw6RJk6LL8+8uISGBL774grPOOotWrVpxxx13MGHCBAYMGHBA7TVs2JC5c+cSiUTo378/7du356abbiIxMTE6Cjl69GhuvfVWxowZQ5s2bbjwwgvZunUrAC6XiyeeeILnnnuOhg0bcs4551R5nPvvv59zzz2Xyy+/nC5durBq1SqmT59OcnLyPvtIpLoss/tE5gP0n//8hwsvvJAbbrihwk2zdVFBQQGJiYnk5+eTkJBQ2+GIiIjIQajp7/WIySMcWY9NIbumTTqIx+VsctgfISAiR79qjcBVx4ABA7jtttt49913D1WTIiIiIkc8p5WEx9kRr7MTHkeH8v86Oyp5E5EaccgSOIDjjz+e9evXH8omRURERI54lmXhsOJxOlJwWPGHfdqkiPx+HNIErrS0VIuYiIiIiIiI1JBDmsB9/PHHtG3b9lA2KSIiIiIiIr+q1nDZzp0797gtEomwadMmpk6dyosvvsjTTz99yIITERERERGR/6lWApeamlqtudwjRozg6quvPuigREREREREpLJqJXBjxozZYwLncrlIS0vjlFNOoVmzZoc0OBEREREREfmfQ/YcuKOJngMnIiJy9ND3uogcTQ7pIiYiIiIiIiJSc5TAiYiIiMhRZfbs2ViWRV5eXm2H8rtiWRbTpk2r7TD2y6RJk0hKSjokba1btw7Lsli8ePEhaW9PlMCJiIiI1EE5OTncdNNNtGjRAp/PR1paGr169eLZZ5+lpKSktsOrc/r27cvNN99c22EclMOVQEjVMjMzyc7Opn379jV6HD11W0REROQgGWMI2gVETBCn5cHjSKjWCt4Has2aNfTs2ZOkpCTuu+8+OnToQDgc5ueff+bll1+mYcOG/PGPf6yx4x+IYDCIx+Op7TBq3O/lPH/r93jOVXE6naSnp9f4cTQCJyIiInIQSsPb2VzyLZuK57G5+Lvy/5Z8S2l4e40d87rrrsPlcrFgwQIuuOAC2rRpQ4cOHTj33HP56KOPGDhwYLRufn4+V199NQ0aNCAhIYFTTjmF//73v9HtY8eOpXPnzkyZMoWsrCwSExO56KKLKCwsjNYxxvDggw/SrFkz/H4/nTp14p133tlrjFlZWdxzzz0MHTqUxMREhg8fDsC8efPo3bs3fr+fzMxMbrzxRoqLi6P7ZWdnc/bZZ+P3+2natCmvv/46WVlZPPbYY0DVo0x5eXlYlsXs2bOrjGXHjh1cfPHFNGrUiJiYGDp06MAbb7wR3T506FDmzJnD448/jmVZWJbFunXrAJgzZw7HH388Xq+XjIwM/u///o9wOBzdt2/fvowYMYJbbrmF1NRUTj/9dK688kr+8Ic/VIghHA6Tnp7Oyy+/XGWMu6byTZ8+nTZt2hAXF8eZZ55JdnZ2hXoTJ06kTZs2+Hw+jj322ArPYG7atCkAxx13HJZl0bdvX5YsWYLD4WD79vLrMTc3F4fDwfnnnx/db/z48fTo0SP6/kDOuSp33303aWlp0c/q6aefpmXLltER4/POO6/K/WDfn9muOG688UZGjRpFSkoK6enpjB07tkKdRx55hA4dOhAbG0tmZibXXXcdRUVFVR5z3bp1OBwOFixYUKH8ySefpEmTJhhjyM3N5dJLL6V+/fr4/X5atmzJxIkTo/v/9trcW92DoQRORERE5ACVhrezpWQRpeHtOC0vXkc8TstbofxQ27FjB59++inXX389sbGxVdbZNfpnjOHss88mJyeHjz/+mIULF9KlSxdOPfVUdu7cGa2/evVqpk2bxocffsiHH37InDlzuP/++6Pb77jjDiZOnMgzzzzDTz/9xF//+lcuu+wy5syZs9dYH3roIdq3b8/ChQsZPXo0S5YsoX///gwePJgffviBqVOn8tVXXzFixIjoPn/+85/ZvHkzs2fP5t133+X5559n69atB9NllJWV0bVrVz788EN+/PFHrr76ai6//HK+/fZbAB5//HF69OjB8OHDyc7OJjs7m8zMTDZt2sRZZ51F9+7d+e9//8szzzzDSy+9xD333FOh/cmTJ+NyuZg7dy7PPfccV111FZ988kmF5Ovjjz+mqKiICy64YI9xlpSU8PDDDzNlyhS++OILNmzYwMiRI6PbX3jhBW6//Xbuvfdeli1bxn333cfo0aOZPHkyAN999x0An332GdnZ2bz33nu0b9+eevXqRT+rL774gnr16vHFF19E2509ezZ9+vQBOOBz/i1jDDfddBMvvfQSX331FZ07d2bBggXceOON3H333axYsYJPPvmE3r17H/Bn9ts4YmNj+fbbb3nwwQe5++67mTFjRnS7w+HgiSee4Mcff2Ty5Ml8/vnnjBo1qspjZmVlcdppp1VKsiZOnMjQoUOxLIvRo0ezdOlS/vOf/7Bs2TKeeeYZUlNTq2xvf+ruFyOV5OfnG8Dk5+fXdigiIiJykGrqe922bfNL0ddmZd4HZkPBF2Zj4ZfR14aCL8zKvA/MpqKvjW3bh/S433zzjQHMe++9V6G8Xr16JjY21sTGxppRo0YZY4yZOXOmSUhIMGVlZRXqNm/e3Dz33HPGGGPuvPNOExMTYwoKCqLb//a3v5kTTjjBGGNMUVGR8fl8Zt68eRXaGDZsmLn44ov3GGeTJk3MoEGDKpRdfvnl5uqrr65Q9uWXXxqHw2FKS0vNsmXLDGDmz58f3b5y5UoDmEcffdQYY8zatWsNYBYtWhStk5ubawAza9YsY4wxs2bNMoDJzc3dY3xnnXWWufXWW6Pv+/TpY2666aYKdf7xj3+Y1q1bV/gM//nPf5q4uDgTiUSi+3Xu3LlS+23btjUPPPBA9P2gQYPM0KFD9xjPxIkTDWBWrVpV4VhpaWnR95mZmeb111+vsN+4ceNMjx49jDFV940xxgwePNiMGDHCGGPMzTffbG699VaTmppqfvrpJxMKhUxcXJz5z3/+c9DnDJi3337bXHbZZebYY481GzdujG579913TUJCQoXrbH9V9Zn16tWrQp3u3bub2267bY9tvPXWW6ZevXrR9xMnTjSJiYnR91OnTjXJycnR/2cWL15sLMsya9euNcYYM3DgQHPFFVdU2fbu/b+3ugdD98CJiIiIHICgXUAgkovbEVPpfjfLsnA7YiiL5BK0C/A6Ew/58Xc/5nfffYdt21x66aUEAgEAFi5cSFFREfXq1atQt7S0lNWrV0ffZ2VlER8fH32fkZERHfVaunQpZWVllabJBYNBjjvuuL3G2K1btwrvFy5cyKpVq3jttdeiZcYYbNtm7dq1/Pzzz7hcLrp06RLd3qJFC5KTk/d6nH2JRCLcf//9TJ06lU2bNhEIBAgEAnscwdxl2bJl9OjRo0Jf9+zZk6KiIn755RcaN25c5XkCXHXVVTz//POMGjWKrVu38tFHHzFz5sy9Hi8mJobmzZtH3//2c9i2bRsbN25k2LBh0emoUD41MzFx79dX3759ef7554Hy6ZHjxo1j7dq1zJkzh/z8fEpLS+nZs+dBnzPAX//6V7xeL998802F0abTTz+dJk2a0KxZM84880zOPPNM/vSnPxETE1NlO9X9zDp27Fjh/W/7DGDWrFncd999LF26lIKCAsLhMGVlZRQXF1f5+Q8aNIgRI0bw/vvvc9FFF/Hyyy/Tr18/srKyAPjLX/7Cueeey/fff88ZZ5zBoEGDOOmkk6o8h/2puz+UwImIiIgcgIgJYhsbt1X1r1MOXIRMKRETPKTHbdGiBZZlsXz58grlzZo1A8Dv90fLbNsmIyOjynvDfrt0utvtrrDNsixs2462AfDRRx9xzDHHVKjn9Xr3GuvuvyDbts0111zDjTfeWKlu48aNWbFiRZXtGGOiPzscjkploVBor3FMmDCBRx99lMceeyx6P9TNN99MMLj3z8YYUylR3nXc35ZXlQj8+c9/5v/+7//4+uuv+frrr8nKyuLkk0/e6/Gq+hx2HW/X5/DCCy9wwgknVKjndDr32m7fvn256aabWLVqFT/++CMnn3wyq1evZs6cOeTl5dG1a9doAn8w5wzlidobb7zB9OnTufTSS6Pl8fHxfP/998yePZtPP/2UMWPGMHbsWObPn1/lMv7V/cz2du2uX7+es846i2uvvZZx48aRkpLCV199xbBhw/Z4zXg8Hi6//HImTpzI4MGDef3116P3XwIMGDCA9evX89FHH/HZZ59x6qmncv311/Pwww9Xamt/6u4PJXAiIiIiB8BpeXBYDmzCOKm8Ap9NGIflwGkd2tX56tWrx+mnn85TTz3FDTfcsNdRpC5dupCTk4PL5YqOIOyvtm3b4vV62bBhQ/Q+qQPVpUsXfvrpJ1q0aFHl9mOPPZZwOMyiRYvo2rUrAKtWrarwPLf69esD5Yud7BoB3Ney+V9++SXnnHMOl112GVCeDK1cuZI2bdpE63g8HiKRSIX92rZty7vvvlshqZk3bx7x8fGVktnd1atXj0GDBjFx4kS+/vprrrjiir3W35e0tDSOOeYY1qxZUyEx+q1dK0Hufh677oO755576NSpEwkJCfTp04fx48eTm5tb4XM9mHMG+OMf/8jAgQO55JJLcDqdXHTRRdFtLpeL0047jdNOO40777yTpKQkPv/8cwYPHlypnep8ZvuyYMECwuEwEyZMiCb+b7311j73u+qqq2jfvj1PP/00oVCoUnz169dn6NChDB06lJNPPpm//e1ve0zK9qdudWkRExEREZED4HEk4HUmE7JLKowGQfmIRcguwedMxuNIOOTHfvrppwmHw3Tr1o2pU6eybNkyVqxYwauvvsry5cujIzKnnXYaPXr0YNCgQUyfPp1169Yxb9487rjjjkor7e1JfHw8I0eO5K9//SuTJ09m9erVLFq0iH/+85/RxTOq67bbbuPrr7/m+uuvZ/HixaxcuZJ///vf3HDDDUB5Anfaaadx9dVX891337Fo0SKuvvpq/H5/NJnw+/2ceOKJ3H///SxdupQvvviCO+64Y6/HbdGiBTNmzGDevHksW7aMa665hpycnAp1srKy+Pbbb1m3bh3bt2/Htm2uu+46Nm7cyA033MDy5cv517/+xZ133sktt9wSTQj25qqrrmLy5MksW7aMIUOG7FdfVWXs2LGMHz+exx9/nJ9//pklS5YwceJEHnnkEQAaNGiA3+/nk08+YcuWLeTn5wPlo1K9e/fm1VdfpW/fvkD51MNgMMjMmTOjZcBBnzPAn/70J6ZMmcIVV1wRXa30ww8/5IknnmDx4sWsX7+eV155Bdu2ad26dZVtVOcz25fmzZsTDod58sknWbNmDVOmTOHZZ5/d535t2rThxBNP5LbbbuPiiy+uMKo9ZswY/vWvf7Fq1Sp++uknPvzwwz0mlftTd38ogRMRERE5AJZlkeJtgcvyErDziZggxthETJCAnY/L8pHsbVEjz4Nr3rw5ixYt4rTTTuPvf/87nTp1olu3bjz55JOMHDmScePGRWP8+OOP6d27N1deeSWtWrXioosuYt26daSlpVX7eOPGjWPMmDGMHz+eNm3a0L9/fz744IPosvXV1bFjR+bMmcPKlSs5+eSTOe644xg9ejQZGRnROq+88gppaWn07t2bP/3pTwwfPpz4+Hh8Pl+0zssvv0woFKJbt27cdNNNlVZI3N3o0aPp0qUL/fv3p2/fvqSnpzNo0KAKdUaOHInT6aRt27bUr1+fDRs2cMwxx/Dxxx/z3Xff0alTJ6699lqGDRu2z4Rxl9NOO42MjAz69+9Pw4YNq99Re3DVVVfx4osvMmnSJDp06ECfPn2YNGlS9HNwuVw88cQTPPfcczRs2JBzzjknum+/fv2IRCLRZM2yrOiUzl69ekXrHew573LeeecxefJkLr/8ct577z2SkpJ47733OOWUU2jTpg3PPvssb7zxBu3ataty/+p8ZvvSuXNnHnnkER544AHat2/Pa6+9xvjx46u177BhwwgGg1x55ZUVyj0eD3//+9/p2LEjvXv3xul08uabb1bZxv7U3R+W2f1PRkJBQQGJiYnk5+eTkHDo/2omIiIih09Nf6+XhrezM7CKQCQX29g4LAc+ZzLJ3hb4XYdgyfDfuV9++YXMzMzoPUR1SUlJCQ0bNuTll1+ucpqgHLnuvfde3nzzTZYsWVLboVSie+BEREREDoLflUpDZz2CdgERE8RpefA4Empk5O334PPPP6eoqIgOHTqQnZ3NqFGjyMrK2uszw440tm2Tk5PDhAkTSExM5I9//GNthyTVVFRUxLJly3jyySejI9lHGiVwIiIiIgfJsqwaeVTA71EoFOIf//gHa9asIT4+npNOOonXXnut0mqDR7INGzbQtGlTGjVqxKRJk3C59Ct3XTFixAjeeOMNBg0aVGn65JFCUyiroCmUIiIiRw99r4vI0USLmIiIiIiIiNQRSuBERERERETqCCVwIiIiIiIidYQSOBERERERkTpCCZyIiIiIiEgdoQRORERERESkjlACJyIiIiJHjNmzZ2NZFnl5ebUdisgRSQmciIiISB00dOhQBg0aVO36lmUxbdq0GovnQPTt25ebb765QtlJJ51EdnY2iYl6MLpIVfRYeBEREZGDZIwhP1hI0A7hcbhJ9MRjWVZth1VrQqEQbrf7gPb1eDykp6cf4ohEjh4agRMRERE5CNvLdjJvy0LmbpnP11u+Z+6W+czbspDtZTsPWwx9+/blxhtvZNSoUaSkpJCens7YsWOj27OysgD405/+hGVZ0fcAH3zwAV27dsXn89GsWTPuuusuwuFwdPvy5cvp1asXPp+Ptm3b8tlnn1UYzVu3bh2WZfHWW2/Rt29ffD4fr776Kjt27ODiiy+mUaNGxMTE0KFDB954441ou0OHDmXOnDk8/vjjWJaFZVmsW7euyimU7777Lu3atcPr9ZKVlcWECRMqnH9WVhb33XcfV155JfHx8TRu3Jjnn3/+kPWvyJFECZyIiIjIAdpetpOF25awvWwnXoeXRE88Xoe3QvnhMnnyZGJjY/n222958MEHufvuu5kxYwYA8+fPB2DixIlkZ2dH30+fPp3LLruMG2+8kaVLl/Lcc88xadIk7r33XgBs22bQoEHExMTw7bff8vzzz3P77bdXefzbbruNG2+8kWXLltG/f3/Kysro2rUrH374IT/++CNXX301l19+Od9++y0Ajz/+OD169GD48OFkZ2eTnZ1NZmZmpXYXLlzIBRdcwEUXXcSSJUsYO3Yso0ePZtKkSRXqTZgwgW7durFo0SKuu+46/vKXv7B8+fJD0rciR5I6lcDde++9nHTSScTExJCUlFRlnV1/wfnt69lnnz28gYqIiMhRzxjDirw1lEUCJHkS8DjdOCwLj9NNkieBskiAn/PWYIw5LPF07NiRO++8k5YtW/LnP/+Zbt26MXPmTADq168PQFJSEunp6dH39957L//3f//HkCFDaNasGaeffjrjxo3jueeeA+DTTz9l9erVvPLKK3Tq1IlevXpFk7vd3XzzzQwePJimTZvSsGFDjjnmGEaOHEnnzp1p1qwZN9xwA/379+ftt98GIDExEY/HQ0xMDOnp6aSnp+N0Oiu1+8gjj3DqqacyevRoWrVqxdChQxkxYgQPPfRQhXpnnXUW1113HS1atOC2224jNTWV2bNnH5K+FTmS1Kl74ILBIOeffz49evTgpZde2mO9iRMncuaZZ0bf6yZYEREROdTyg4XkBfOJdcVUut/NsixiXTHkBvPJDxaS5E2o8Xg6duxY4X1GRgZbt27d6z4LFy5k/vz5FZKySCRCWVkZJSUlrFixgszMzAr3pB1//PFVttWtW7cK7yORCPfffz9Tp05l06ZNBAIBAoEAsbGx+3Vey5Yt45xzzqlQ1rNnTx577DEikUg06fvt+VuWRXp6+j7PX6QuqlMJ3F133QVQach8d7v+uiQiIiJSU4J2iLBtE+uq+tcpl8NJcdgmaIcOSzy7LxpiWRa2be91H9u2ueuuuxg8eHClbT6fD2NMtRdj2T0xmzBhAo8++iiPPfYYHTp0IDY2lptvvplgMFit9napKoaqRjUP5PxF6qI6NYWyukaMGEFqairdu3fn2Wef1f+8IiIicsh5HG5cDgdhO1zl9rAdweVw4HEc2GqMh5rb7SYSiVQo69KlCytWrKBFixaVXg6Hg2OPPZYNGzawZcuW6D677p/bly+//JJzzjmHyy67jE6dOtGsWTNWrlxZoY7H46kU0+7atm3LV199VaFs3rx5tGrVqsoplyJHuzo1Alcd48aN49RTT8Xv9zNz5kxuvfVWtm/fzh133LHHfXYN6e9SUFBwOEIVERGROizRE0+SJ5HtZTtxOxIqjBIZYygOl1Dfl0KiJ74Wo/yfrKwsZs6cSc+ePfF6vSQnJzNmzBj+8Ic/kJmZyfnnn4/D4eCHH35gyZIl3HPPPZx++uk0b96cIUOG8OCDD1JYWBhdxGRfI3MtWrTg3XffZd68eSQnJ/PII4+Qk5NDmzZtKsT07bffsm7dOuLi4khJSanUzq233kr37t0ZN24cF154IV9//TVPPfUUTz/99KHtIJE6otZH4MaOHVvlwiO/fS1YsKDa7d1xxx306NGDzp07c+utt3L33XdXusl1d+PHjycxMTH6qmoFJBEREZHfsiyL1knN8Dm95AULCEZC2MYmGAmRFyzA5/TSKqnZEfM8uAkTJjBjxgwyMzM57rjjAOjfvz8ffvghM2bMoHv37px44ok88sgjNGnSBACn08m0adMoKiqie/fuXHXVVdE/ivt8vr0eb/To0XTp0oX+/fvTt29f0tPTKz14fOTIkTidTtq2bUv9+vXZsGFDpXa6dOnCW2+9xZtvvkn79u0ZM2YMd999N0OHDj34ThGpgyxzuJZG2oPt27ezffv2vdbJysqq8I/EpEmTuPnmmys8H2RP5s6dS69evcjJySEtLa3KOlWNwGVmZpKfn09CQs3fdCwiIiI1p6CggMTExBr7Xt9etpMVeWvIC+YTtm1cDgfJnkRaJTUj1Vd5RKmu2/W71apVq2jevHlthyPyu1PrUyhTU1NJTU2tsfYXLVqEz+fb42MHALxeL16vt8ZiEBERkaNXqi+FemnJ5AcLCdohPA43iZ74I2bk7WC9//77xMXF0bJlS1atWsVNN91Ez549lbyJ1JJaT+D2x4YNG9i5cycbNmwgEomwePFioHyOdVxcHB988AE5OTn06NEDv9/PrFmzuP3227n66quVoImIiEiNsSzrsDwqoDYUFhYyatQoNm7cSGpqKqeddhoTJkyo7bBEfrdqfQrl/hg6dCiTJ0+uVD5r1iz69u3LJ598wt///ndWrVqFbds0a9aMq666iuuvvx7XHpb4rUpNT7UQERGRw0ff6yJyNKlTCdzhon/oRUREjh76XheRo0mtr0IpIiIiIiIi1aMETkREREREpI5QAiciIiIiIlJHKIETERERERGpI5TAiYiIiIiI1BFK4ERERETksJg9ezaWZZGXlwfApEmTSEpKqvHjWpbFtGnTavw4IoeDEjgRERGRg2QbQ3ZxIWvyd5JdXIh9GJ7SNHToUCzL4v77769QPm3aNCzLqvHjHwoXXnghP//88yFrb+zYsXTu3LlSeXZ2NgMGDDhkxxGpTdV/urWIiIiIVLKuIJevNq1jc1EBITuC2+GkYVwCvY7JIishuUaP7fP5eOCBB7jmmmtITj40xwoGg3g8nkPS1r74/X78fn+NHyc9Pb3GjyFyuGgETkREROQArSvIZdqqn1iXn0uc20N6TDxxbg/r8n8tL8it0eOfdtpppKenM378+D3Weffdd2nXrh1er5esrCwmTJhQYXtWVhb33HMPQ4cOJTExkeHDh0enNn744Ye0bt2amJgYzjvvPIqLi5k8eTJZWVkkJydzww03EIlEom29+uqrdOvWjfj4eNLT07nkkkvYunXrHmPbfQplVlYWlmVVeu1y22230apVK2JiYmjWrBmjR48mFApF27rrrrv473//G91v0qRJQOUplEuWLOGUU07B7/dTr149rr76aoqKiqLbhw4dyqBBg3j44YfJyMigXr16XH/99dFjidQmJXAiIiIiB8A2hq82raMoGKRhbDx+lxuHZeF3uWkYG09RMMjcTetqdDql0+nkvvvu48knn+SXX36ptH3hwoVccMEFXHTRRSxZsoSxY8cyevToaGKzy0MPPUT79u1ZuHAho0ePBqCkpIQnnniCN998k08++YTZs2czePBgPv74Yz7++GOmTJnC888/zzvvvBNtJxgMMm7cOP773/8ybdo01q5dy9ChQ6t9PvPnzyc7O5vs7Gx++eUXTjzxRE4++eTo9vj4eCZNmsTSpUt5/PHHeeGFF3j00UeB8umYt956K+3atYu2ceGFF1Y6RklJCWeeeSbJycnMnz+ft99+m88++4wRI0ZUqDdr1ixWr17NrFmzmDx5MpMmTarUbyK1QVMoRURERA7AlpIiNhcVkOLzV7rnzLIsUnx+NhUVsKWkiIzY+BqL409/+hOdO3fmzjvv5KWXXqqw7ZFHHuHUU0+NJmWtWrVi6dKlPPTQQxUSq1NOOYWRI0dG33/11VeEQiGeeeYZmjdvDsB5553HlClT2LJlC3FxcbRt25Z+/foxa9asaKJ05ZVXRtto1qwZTzzxBMcffzxFRUXExcXt81zq168f/fmmm24iOzub+fPnR8vuuOOO6M9ZWVnceuutTJ06lVGjRuH3+4mLi8Plcu11yuRrr71GaWkpr7zyCrGxsQA89dRTDBw4kAceeIC0tDQAkpOTeeqpp3A6nRx77LGcffbZzJw5k+HDh+/zPERqkkbgRERERA5AaThEyI7gdVb993CP00nIjlAarvlpdw888ACTJ09m6dKlFcqXLVtGz549K5T17NmTlStXVpj62K1bt0ptxsTERJM3gLS0NLKysiokYmlpaRWmSC5atIhzzjmHJk2aEB8fT9++fQHYsGHDfp3P888/z0svvcS//vWvCkndO++8Q69evUhPTycuLo7Ro0fvd9vLli2jU6dO0eQNyvvEtm1WrFgRLWvXrh1OpzP6PiMjY6/TQUUOFyVwIiIiIgfA73LjdjgJRMJVbg9Gyhc08bvcNR5L79696d+/P//4xz8qlBtjKo0OmiqmdP42mdnF7a4Yt2VZVZbZtg1AcXExZ5xxBnFxcbz66qvMnz+f999/HyifWllds2fP5oYbbuCVV16hU6dO0fJvvvmGiy66iAEDBvDhhx+yaNEibr/99v1qG6ruk9+ezy57O1eR2qQplCIiIiIHIC0mjoZxCazLz6VhrKvCL//GGHaWldI0MZm0mH1PHTwU7r//fjp37kyrVq2iZW3btuWrr76qUG/evHm0atWqwujSobB8+XK2b9/O/fffT2ZmJgALFizYrzZWrVrFueeeyz/+8Q8GDx5cYdvcuXNp0qQJt99+e7Rs/fr1Fep4PJ4KI4tVadu2LZMnT6a4uDiauM6dOxeHw1Gh70SOVBqBExERETkADsui1zFZxHk8bC4upDQcImJsSsMhNhcXEu/x0vOYLByH6ZlsHTp04NJLL+XJJ5+Mlt16663MnDmTcePG8fPPPzN58mSeeuqpCve7HSqNGzfG4/Hw5JNPsmbNGv79738zbty4au9fWlrKwIED6dy5M1dffTU5OTnRF0CLFi3YsGEDb775JqtXr+aJJ56IjvDtkpWVxdq1a1m8eDHbt28nEAhUOs6ll16Kz+djyJAh/Pjjj8yaNYsbbriByy+/PHr/m8iRTAmciIiIyAHKSkhmUIt2ZCUmUxQKsqWkiKJQkKaJyZzTom2NPwdud+PGjaswRbJLly689dZbvPnmm7Rv354xY8Zw991379fKkNVVv359Jk2axNtvv03btm25//77efjhh6u9/5YtW1i+fDmff/45DRs2JCMjI/oCOOecc/jrX//KiBEj6Ny5M/PmzYsuzrLLueeey5lnnkm/fv2oX78+b7zxRqXjxMTEMH36dHbu3En37t0577zzOPXUU3nqqacOrgNEDhPLVDUR+neuoKCAxMRE8vPzSUhIqO1wRERE5CAcju912xi2lBRRGg7hd7lJi4k7bCNvIvL7onvgRERERA6Sw7Jq9FEBIiK7aAqliIiIiIhIHaEETkREREREpI5QAiciIiIiIlJHKIETERERERGpI5TAiYiIiIiI1BFK4EREREREROoIJXAiIiIiIiJ1hBI4ERERERGROkIJnIiIiIjs1dixY+ncuXP0/dChQxk0aFCtxXOkmD17NpZlkZeXV9uhyO+IEjgRERGRg2Qbw4YdxSzdnM+GHcXYxtT4MTdu3MiwYcNo2LAhHo+HJk2acNNNN7Fjx44aP/bjjz/OpEmTou/79u3LzTffXOPHPRTqUqwiVXHVdgAiIiIiddmKnAI++iGbdduLCYRtvC4HWamxnN0xg9bpCTVyzDVr1tCjRw9atWrFG2+8QdOmTfnpp5/429/+xn/+8x+++eYbUlJSauTYAImJiTXW9pHAGEMkEsHl0q/KcuTRCJyIiIjIAVqRU8BLX6xleXYBCT43mckxJPjcLM8uL1+RU1Ajx73++uvxeDx8+umn9OnTh8aNGzNgwAA+++wzNm3axO233x6ta1kW06ZNq7B/UlJShRG02267jVatWhETE0OzZs0YPXo0oVBoj8f/7RTKoUOHMmfOHB5//HEsy8KyLNauXUuLFi14+OGHK+z3448/4nA4WL169R7bfvnll2nXrh1er5eMjAxGjBgR3bZhwwbOOecc4uLiSEhI4IILLmDLli3R7bumek6ZMoWsrCwSExO56KKLKCws3GOs69ati06FnD59Ot26dcPr9fLll18SCAS48cYbadCgAT6fj169ejF//vw9xi5yOCiBExERETkAtjF89EM2eaVBsurFEut14XRYxHpdZNWLJa80yMc/ZB/y6ZQ7d+5k+vTpXHfddfj9/grb0tPTufTSS5k6dSpmP44bHx/PpEmTWLp0KY8//jgvvPACjz76aLX2ffzxx+nRowfDhw8nOzub7OxsGjduzJVXXsnEiRMr1H355Zc5+eSTad68eZVtPfPMM1x//fVcffXVLFmyhH//+9+0aNECKB8VGzRoEDt37mTOnDnMmDGD1atXc+GFF1ZoY/Xq1UybNo0PP/yQDz/8kDlz5nD//ffvMdbMzMzovqNGjWL8+PEsW7aMjh07MmrUKN59910mT57M999/T4sWLejfvz87d+6sdt+KHGoaFxYRERE5AL/sLGHd9mIaxPuwLKvCNsuyaBDvY+32Yn7ZWULjerGH7LgrV67EGEObNm2q3N6mTRtyc3PZtm0bDRo0qFabd9xxR/TnrKwsbr31VqZOncqoUaP2uW9iYiIej4eYmBjS09Oj5VdccQVjxozhu+++4/jjjycUCvHqq6/y0EMP7bGte+65h1tvvZWbbropWta9e3cAPvvsM3744QfWrl0bTbqmTJlCu3btmD9/frSebdtMmjSJ+Ph4AC6//HJmzpzJvffeu8dYd7n77rs5/fTTASguLuaZZ55h0qRJDBgwAIAXXniBGTNm8NJLL/G3v/1tn30jUhM0AiciIiJyAIoCYQJhG5/bWeV2n9tJIGxTFAgf1rh2jbx5PJ5q7/POO+/Qq1cv0tPTiYuLY/To0WzYsOGg4sjIyODss8/m5ZdfBuDDDz+krKyM888/v8r6W7duZfPmzZx66qlVbl+2bBmZmZkVRszatm1LUlISy5Yti5ZlZWVFk7ddcWzdurVaMXfr1i368+rVqwmFQvTs2TNa5na7Of744yscT+RwUwInIiIicgDivC68LgdloUiV28tCEbwuB3HeQzvhqUWLFliWxdKlS6vcvnz5curXr09SUhJQPhq4+3TK397f9s0333DRRRcxYMAAPvzwQxYtWsTtt99OMBg86Fivuuoq3nzzTUpLS5k4cSIXXnghMTExVdbdfTro7owxlUY6qyp3u90VtluWhW3b1Yo3NvZ/I6W7+mz3Y+4pDpHDRQmciIiIyAFolBJDVmosWwvLKiVIxhi2FpbRNDWWRilVJywHql69epx++uk8/fTTlJaWVtiWk5PDa6+9xtChQ6Nl9evXJzs7O/p+5cqVlJSURN/PnTuXJk2acPvtt9OtWzdatmzJ+vXr9ysmj8dDJFI5kT3rrLOIjY3lmWee4T//+Q9XXnnlHtuIj48nKyuLmTNnVrm9bdu2bNiwgY0bN0bLli5dSn5+/h6nk+5PrLtr0aIFHo+Hr776KloWCoVYsGDBfh1P5FBTAiciIiJyAByWxdkdM0jye1i3o5jiQJiIbSgOhFm3o5gkv4ezOmbgqIHRmqeeeopAIED//v354osv2LhxI5988gmnn346rVq1YsyYMdG6p5xyCk899RTff/89CxYs4Nprr60wStWiRQs2bNjAm2++yerVq3niiSd4//339yuerKwsvv32W9atW8f27dujI15Op5OhQ4fy97//nRYtWtCjR4+9tjN27FgmTJjAE088wcqVK/n+++958sknATjttNPo2LEjl156Kd9//z3fffcdf/7zn+nTp0+FqY8HGuvuYmNj+ctf/sLf/vY3PvnkE5YuXcrw4cMpKSlh2LBh1T6eyKGmBE5ERETkALVOT2BY76Ycm5FAQVmIjbklFJSFaJNRXl5Tz4Fr2bIl8+fPp1mzZlxwwQU0adKEAQMG0KpVK+bOnUtcXFy07oQJE8jMzKR3795ccskljBw5ssI0xnPOOYe//vWvjBgxgs6dOzNv3jxGjx69X/GMHDkSp9NJ27ZtqV+/foX754YNG0YwGNzr6NsuQ4YM4bHHHuPpp5+mXbt2/OEPf2DlypXA/x6HkJycTO/evTnttNNo1qwZU6dOPWSx7u7+++/n3HPP5fLLL6dLly6sWrWK6dOnk5ycvF/HFDmULLM/a8z+ThQUFJCYmEh+fj4JCTXzD6+IiIgcHofje902hl92llAUCBPnddEoJaZGRt725s477+SRRx7h008/3edI1+E0d+5c+vbtyy+//EJaWlpthyNS5+kxAiIiIiIHyWFZh/RRAQfirrvuik4PPOGEE3A4aneiVSAQYOPGjYwePZoLLrhAyZvIIaIETkREROQoccUVV9R2CFFvvPEGw4YNo3PnzkyZMqW2wxE5amgKZRU0hVJEROTooe91ETmaaBETERERERGROkIJnIiIiIiISB2hBE5ERERERKSOUAInIiIiIiJSRyiBExERERERqSOUwImIiIiIiNQRSuBEREREhKFDhzJo0KDaDuOQmT17NpZlkZeXt1/7jR07lrS0NCzLYtq0aTUSm8jB0HPgqqDnxYiIiBw9Dsf3um0btmYXUlISJCbGQ4OMeBwOq0aOZVl7b3fIkCFMmjRpv9vNz8/HGENSUtKBBbaboUOHkpeXV2tJUDAYZOfOndFkrDqWLVtG27Ztef/99znxxBNJTk7G6/XWcKQi+8dV2wGIiIiI1GXr1+zkq1mr2bwxn2AwgsfjpGFmIr36NadJs5RDfrzs7Ozoz1OnTmXMmDGsWLEiWub3+yvUD4VCuN3ufbabmJh46II8hILBIB6PZ7/383g8pKen79c+q1evBuCcc86pdtJXler2uciB0BRKERERkQO0fs1Opk39gbWrdhIb5yW9YTyxcV7WriovX79m5yE/Znp6evSVmJiIZVnR92VlZSQlJfHWW2/Rt29ffD4fr776Kjt27ODiiy+mUaNGxMTE0KFDB954440K7e4+hdIYw4MPPkizZs3w+/106tSJd955p8I+P/30E2effTYJCQnEx8dz8skns3r1asaOHcvkyZP517/+hWVZWJbF7NmzAViyZAmnnHIKfr+fevXqcfXVV1NUVFQpjvHjx9OwYUNatWrF3XffTYcOHSr1RdeuXRkzZkyV/bT7FMpJkyaRlJTE9OnTadOmDXFxcZx55pnRhHjs2LEMHDgQAIfDUSGBmzhxIm3atMHn83Hsscfy9NNPR7etW7cOy7Iq9Xl193vvvffo168fMTExdOrUia+//rrCecydO5c+ffoQExNDcnIy/fv3Jzc3t9qfkRx9NAInIiIicgBs2/DVrNUUFgRo2Cgh+gu/P8ZBQ38Cm38p4KtZq8nMSq6x6ZR7cttttzFhwgQmTpyI1+ulrKyMrl27ctttt5GQkMBHH33E5ZdfTrNmzTjhhBOqbOOOO+7gvffe45lnnqFly5Z88cUXXHbZZdSvX58+ffqwadMmevfuTd++ffn8889JSEhg7ty5hMNhRo4cybJlyygoKGDixIkApKSkUFJSwplnnsmJJ57I/Pnz2bp1K1dddRUjRoyoMO1z5syZJCQkMGPGjOi0zrvuuov58+fTvXt3AH744QcWLVrE22+/Xe1+KSkp4eGHH2bKlCk4HA4uu+wyRo4cyWuvvcbIkSPJysriiiuuqDDK+cILL3DnnXfy1FNPcdxxx7Fo0SKGDx9ObGwsQ4YM2WOfV3e/22+/nYcffpiWLVty++23c/HFF7Nq1SpcLheLFy/m1FNP5corr+SJJ57A5XIxa9YsIpFItT4jOTopgRMRERE5AFuzC9m8MZ+UejGVpttZlkVKvRg2b8xna3Yh6ccc3nvqb775ZgYPHlyhbOTIkdGfb7jhBj755BPefvvtKhO44uJiHnnkET7//HN69OgBQLNmzfjqq6947rnn6NOnD//85z9JTEzkzTffjE4XbNWqVbQNv99PIBCoMI1x8uTJlJaW8sorrxAbGwvAU089xcCBA3nggQdIS0sDIDY2lhdffLHC1Mn+/fszceLEaAI3ceJE+vTpQ7NmzardL6FQiGeffZbmzZsDMGLECO6++24A4uLiovf//TbmcePGMWHChGh/Nm3alKVLl/Lcc89VSMR27/Pq7jdy5EjOPvtsAO666y7atWvHqlWrOPbYY3nwwQfp1q1bhZG7du3aAdX7jOTopARORERE5ACUlAQJBiOk+JxVbvd4neTujFBSEjzMkUG3bt0qvI9EItx///1MnTqVTZs2EQgECAQC0SRqd0uXLqWsrIzTTz+9QnkwGOS4444DYPHixZx88sn7da/XsmXL6NSpU4Xj9uzZE9u2WbFiRTSB69ChQ6X73oYPH86VV17JI488gtPp5LXXXmPChAnVPjZATExMNHkDyMjIYOvWrXusv23bNjZu3MiwYcMYPnx4tDwcDle6Z/C3fb4/+3Xs2LFCPABbt27l2GOPZfHixZx//vlVxladz0iOTkrgRERERA5ATIwHj8dJoCyCP6bysgLBQPmCJjEx+78Ax8HaPTGbMGECjz76KI899hgdOnQgNjaWm2++mWCw6uTStm0APvroI4455pgK23atyrj7YinVYYzZ4+Igvy2vKrEcOHAgXq+X999/H6/XSyAQ4Nxzz92v4++ebFqWxd4WZN/VDy+88EKlkUqns2Li/tuY92e/38a0qw927b+3Pq7OZyRHJyVwIiIiIgegQUY8DTMTWbtqJw39CRUSEGMMO3eU0LRFCg0y4msxynJffvkl55xzDpdddhlQ/sv/ypUradOmTZX127Zti9frZcOGDXucitexY0cmT568xxUXPR5P9F6t37Y7efJkiouLownP3LlzcTgcFaZfVsXlcjFkyJDoPWYXXXQRMTEx+zz3g5GWlsYxxxzDmjVruPTSS2t8v9117NiRmTNnctddd1XaVp3PSI5OSuBEREREDoDDYdGrX3O2by1m8y8FpNSLweN1EgxE2LmjhPgEL736NT/sC5hUpUWLFrz77rvMmzeP5ORkHnnkEXJycvaYwMXHxzNy5Ej++te/Yts2vXr1oqCggHnz5hEXF8eQIUMYMWIETz75JBdddBF///vfSUxM5JtvvuH444+ndevWZGVlMX36dFasWEG9evVITEzk0ksv5c4772TIkCGMHTuWbdu2ccMNN3D55ZdHp0/uzVVXXRWNee7cuYe0j/Zk7Nix3HjjjSQkJDBgwAACgQALFiwgNzeXW2655ZDv91t///vf6dChA9dddx3XXnstHo+HWbNmcf7555OamrrPz0iOTnqMgIiIiMgBatIshUEXdqRpixSKiwJsyS6kuChA0xbl5TXxHLgDMXr0aLp06UL//v3p27cv6enpFR4ZUJVx48YxZswYxo8fT5s2bejfvz8ffPABTZs2BaBevXp8/vnnFBUV0adPH7p27coLL7wQHY0bPnw4rVu3plu3btSvX5+5c+cSExPD9OnT2blzJ927d+e8887j1FNP5amnnqrWebRs2ZKTTjqJ1q1b73H1zEPtqquu4sUXX2TSpEl06NCBPn36MGnSpGg/HOr9fqtVq1Z8+umn/Pe//+X444+nR48e/Otf/8LlKh+D2ddnJEcny+xt4u/vVEFBAYmJieTn55OQcHhXjRIREZFD63B8r9u2YWt2ISUlQWJiPDTIiD8iRt72x8UXX4zT6Yw+w+xIZIzh2GOP5Zprrqn2KJbI0UZTKEVEREQOksNhHfZHBRwq4XCYn3/+ma+//pprrrmmtsPZo61btzJlyhQ2bdrEFVdcUdvhiNQaJXAiIiIiv2M//vgjJ510Ev369ePaa6+t7XD2KC0tjdTUVJ5//nmSk5NrOxyRWqMETkREROR3rHPnzpSUlNR2GPuku35EymkRExERERERkTpCCZyIiIiIiEgdoQRORERERESkjlACJyIiIiIiUkcogRMREREREakj6kwCt27dOoYNG0bTpk3x+/00b96cO++8k2AwWKHehg0bGDhwILGxsaSmpnLjjTdWqiMiIiIiIlIX1ZnHCCxfvhzbtnnuuedo0aIFP/74I8OHD6e4uJiHH34YgEgkwtlnn039+vX56quv2LFjB0OGDMEYw5NPPlnLZyAiIiIiInJwLFOHH6rx0EMP8cwzz7BmzRoA/vOf//CHP/yBjRs30rBhQwDefPNNhg4dytatW0lISKhWuwUFBSQmJpKfn1/tfUREROTIpO91ETma1JkplFXJz88nJSUl+v7rr7+mffv20eQNoH///gQCARYuXLjHdgKBAAUFBRVeIiIiIiIiR5o6m8CtXr2aJ598kmuvvTZalpOTQ1paWoV6ycnJeDwecnJy9tjW+PHjSUxMjL4yMzNrLG4REREREZEDVesJ3NixY7Esa6+vBQsWVNhn8+bNnHnmmZx//vlcddVVFbZZllXpGMaYKst3+fvf/05+fn70tXHjxkNzciIiIiIiIodQrS9iMmLECC666KK91snKyor+vHnzZvr160ePHj14/vnnK9RLT0/n22+/rVCWm5tLKBSqNDL3W16vF6/Xu//Bi4iIiIiIHEa1nsClpqaSmpparbqbNm2iX79+dO3alYkTJ+JwVBxA7NGjB/feey/Z2dlkZGQA8Omnn+L1eunateshj11ERERERORwqjOrUG7evJk+ffrQuHFjXnnlFZxOZ3Rbeno6UP4Ygc6dO5OWlsZDDz3Ezp07GTp0KIMGDdqvxwhotSoREZGjh77XReRoUusjcNX16aefsmrVKlatWkWjRo0qbNuVgzqdTj766COuu+46evbsid/v55JLLok+J05ERERERKQuqzMjcIeT/lInIiJy9ND3uogcTWp9FUoRERERERGpHiVwIiIiIiIidYQSOBERERERkTpCCZyIiIiIiEgdUWdWoRQROaRMCOwiwIAVCw5vbUckIiIisk9K4ETk98Uug9BaCG8GEywvs9zgTAdPU3DE1m58IiIiInuhBE5Efj/sUihbCJF8sFxg+crLTQjC6yCyHXxdwKllxkVEROTIpHvgROT3I7AME8nDWD6M5QXLWf5y+MCKB7sYAj+CsWs7UhEREZEqaQRORI56xhjsyEas8AYMYSAIxgLjwWH5sCw3WBY4/GDnQyQXXPVqO2wRERGRSjQCJyJHNWMMYbOeSGgJEKT8nz0LMEAZtsnHNoHyypYLsMHeUVvhioiIiOyVEjgROarZZgfhyOpf3znA2pXAOQAnYDCmCGPC5VUMYCK1EaqIiIjIPimBE5GjWtjehCFSfp8bgDG71XACNoay8m0W/1vcREREROQIo3vgROSoZUwQ2+zEwoPtcGEsJ5axMTh3q2lhTBCDCws3uNJqJV4RERGRfdEInIgctcoXLDHsmjppO2PL35tIxZE4s6ssBO5G4IippYhFRERE9k4jcCJy1LJwU/53qgjgIuKKBWyc4eLyMlNeC2zAgeXOBE/rWotXREREZF+UwInIUcuy3Dis+kTMJjAeLMsi4orHdvpwhEuxTAiMwXYYHO7mOFztyx8nICIiInKEUgInIkc1t7MRdngrhlIwfizLwlgeIh4PxhgMxViWH5ezhZI3EREROeLpHjgROao5rETczjZYuDAUY5tSjAlgm5Jo8uZxtMNh+Ws7VBEREZF90giciBz1XI50HFYcYTubiNkCJozDisHpyMBppSt5ExERkTpDCZyI/C44rDg8zpYY0wIwWJYmIIiIiEjdowRORH5XLMuifOVJERERkbpHf4IWERERERGpI5TAiYiIiIiI1BFK4EREREREROoIJXAiIiIiIiJ1hBI4ERERERGROkKrUIrIkcGEILwFTCngAEcSOFPA0oqRIiIiIrsogROR2mUMhNZDaM2vyZtVXmY5wJEI3rbgTKrtKEVERESOCJpCKSK1K7QGgsvABMGKBUccOOPB8kIkF8q+h0hBbUcpIiIickRQAicitccugeAayqdMxpSPuu1iucqTObsEgiv32owxhpKcnRSuzSaYX1SzMYuIiIjUIk2hFJHaE84GgmDFVb3dssDhg8h2sAvBEV9hszGGTdPns+6dL8hdsgYTsXH6PDQ8rQtNL+xHUpsmNX8OIiIiIoeREjgRqT12AWDtY6ESNxAAu6hCAmeM4ccJb7F6ymfYkQje5HgcbifhkgBr3pzF5pmL6HrvMNJ7d6zpsxARERE5bDSFUkTqpHXvfcnKiZ/givWR0CwDb3Ic7jg//gZJJLQ8hmBuId/fOYmS7B21HaqIiIjIIaMETkRqjyMBMOWrTu5RCPj1fjhg+4IVLPjHi3z9l8coWptDwapfKFyTTbgkEN3DsizistIozd7JLx9/W6OnICIiInI4aQqliNQ4YwybNuazZXMBBmiQHk9mkyQsV0b5IiamDCx/VTuCXQauDHDEs2rKDJY+8R6BnYWES8qw3E7sYISi9Vso3ZJL4rGN8SaXJ3qWw4HT62LzZwtpNeysw3vCIiIiIjVECZyI1Khf1ucy/YNlrF+zk2AgggE8HieZWcmcfvaxNG3cHIIrylebtHz/W4nShMEuLV+d0tOSnC9+4KfH3sFyOIjNrE9gez4OjwvL6QDchEvKyF++gXrHtcDp8wDg8LgI5hfX2rmLiIiIHGpK4ESkxmxYl8trL84nb2cpiUk+kpL9hMM2+fllLPshh03r87j86u60aN7m15G4EjD8+iBvJziTwduWgnVFLLz9RQpWbsLp85SPvEUiWBHr1wQOXH4f4ZIySrfmEde4AQCRshD+tORa7AERERGRQ0sJnIjUCNs2fPjuj+TnlpKWEUckbNiSXUhBfoBIxMYYQ0FBGU9P+Iq/3NqLlq1PhvDW8iQOBziTwJHCuve+5Ifxr5O7dB2Ww4EdCmHKgtjBMJFACE+SA4fTWb6YpcNB2a8JnB0KYyIRGp11Ym13hYiIiMgho0VMRKRGrF21g80b834ddTNs3JDLzh0lgMHjceLzuXC7neTllvLyP79m+dKd4D4GPC3B0xyc9cj5agk/3P8G4ZIyHG4XLr8Xp9eDK8aLK9YLxhDKL8HYvy6C4rCwQ2HscITCtTkktGzEMWd0rdV+EBERETmUlMCJSI3YtDGPcMjG63OxfWsRpSUhvF4XLpcTy7IAC7fbictlUVwY5N9Tf6C0NBTd3xjD6lc/I1xcSswxqVgOB8a2o9vLEzkfxrYJFZYQ+XVUzg6FKVybQ2KrRnR/8Bo8iXt4SLiIiIhIHaQplCJSI2zbgAWhUITCgjJcLsevidvuLGJiPezcWcJP/82m24mNASham82O71fiq5+E0+fFkxRXvnCJ+3//bLlivJhIBKfXg+V0YAfDJLRsRNsbB9PorOPx1Us8TGcrIiIicngogRORGlEvNRYLi6LCAJFI+bTJ3Zlfn//m9bkoKw2zfs3OaAJXtqOASCCIt14ClgUxDesRzC0kEgjh9LqjbThcLpw+N57kePzpKfR59R/EZNQ7PCcpIiIicphpCqWI1IjWbRuQUj+GosIgQJWjb6GQjdvtID7Bi8MB4dD/pki6/F4spxMTCgPgq5dAfLMMMBAuLiMSDGGHIkRCYUIlAfxpyXS77yolbyIiInJUUwInIjXC43Vx6pmtcbud2BFDJBKJbjPGEApGAEO91FgcDgvbNqSkxkTrJLbOJD4rjbLt+dGy2Eb1Se7YFH96CpZlYdsRLIdF5tkn0uvlUdQ/oc3hPEURERGRw04JnIjUmOOOb8Tgizvi87soK4sQKAsRKAsTDJYnXg3S40lK8VNcFMTnd9PhuIbRfR1uF1nn9cEORwgW/O9h3N6kOJLaNCb1hDbEZtQjtXtrTnj0OuKz0mvjFEVEREQOK90DJyI1xrIsTji5KU6XkzcmLiAYCOPzu/H53SQkeHG6HJQUBykuCnLCyVmkN0yosH/TC/uRt3QdG/41j8COQnypCVguJ+HiMgI7C/HVT6LLXVfgb6CHdYuIiMjvgxI4ETlgkYjN6hXb2bg+FztiSE6NoW2HdGJiPRXqdevRGIfD4uNpP1FUECAcipCbW0okbPD6XHQ9qRE9z84kP1hAnDsWp1W+4InD5eS4u64gpVML1r37BYWrN2NHbFx+D00Gn0zzS04luX3T2jh1ERERkVphmV3LwElUQUEBiYmJ5Ofnk5CQsO8dRH6H1q/Zyb/fXkLOpgIikf8tPpKY5Kf3aS04qW/TSguXFOSVsmRRNutW7yActkmo7ya2eZiSxFzCpvweuVh3DE3jM2me0AS3439/Y7IjEYrXbyESCOFLTcRXP+mwnKeI1H36XheRo4kSuCroH3qRvdu4LpcpL8ynIK+U5BQ/Hm95ohWJ2OTnlWHbhv4D29D7tBZ7bGNHWS5fb/meonAxHocHt8OFAQKRALaxyYhpQI+0rhWSOBGRA6HvdRE5mug3IxHZL8YYPvt4Ofm5paRlxFUYZXM6HaTUiyEvt5Q5M1bSsUtDklLKV5YsLQ2xctlWiouCuDwOfolfRdBXQoI7vkIbboeLsrIQPyz6hc0mRIuUxjRtUY969WMP+7mKiIiIHGmUwInIfsn+pYC1q3aSkOSt8tluBvDHu9iaU8incxfTvXcmq78uYsHcjeTnlgEQjIQIewI0bOunTT+D21/ejjGG9QtKWPtdMcW5YSw2sMi1nZgYD8e2T2fAOW1ISPIfztMVEREROaIogROR/bJ1SyGBQJikZF+lbWEToSBYSCASpMyOsHTNRn7e9As5iyLEeL3UT03A5XKQW5ZPXr5h48IyyvIMXc5NwumxWPlFEavmFuFwWMQmuTDOCAkeD5FSi/nfrCc7J5fLrulOapKmQImIiMjvk54DJyL7xaLyqBtAxNjsDORRGglgWRYOHNhFLrYuieDyG0xckBJTggGMBd54BzHJTratCbBxcSkFOWHWfluMy2MRk+zE6bYwBoJ2kIC7BOLL+HlVDhPf/4zvti5mZ1neYT1vERERkSOBEjgR2S9pGfH4fC5KS0IVyotCxYQiYVyWC8uUJ3nBEhs7BP44NxYWxaESwnYYp+XAGHC6LBxO2Li4hA3/LSJUZuOJLf9nyTYGG5viUClBO4TT7cDtdZDzU5Cft63ni5xv2VScc9jPX0RERKQ2aQqliOyXtIbxNG1ZjyWLNxF0lWEAy4LScBkOyyr/Od/GG+sgXGbj8pUnc07LQcgOUxopw+fyURIuxcbG5YPC3ABl4QDGZQgZgyPiwGBjjMHhcOB0lCd13liLskIbq8BH0Bdi4bYlJHkSiHXH1GKPiIiIiBw+GoETkf1SFC4mvnspJi7ItpwiCktKKAwWE7TDhMIRineGsW1Dox4uwkSI2BFCkTAhO0zk13vkioLFOC0nQTtMmDDGGLDBsiyMMYRNmIixsSwrmrwBlM/eNFjhAHHYlAQL2FCwrpZ6QkREROTwUwInItVWGi5jXs5CQkmFdDk3idQmfuwyB4GdFmU7DKV5NlZshGP6QXyHIL5Um1DAJmzChE0Ygyl/1psdIGAHMcYmXAYuv0VMBkRC5Y+ljN5nZ359/fpzqCSEyxUmxr8DK7ADZ7iIX7bOx2z5EhMJ1EaXiIiIiBxWmkIpItW2rvAXcgP5xHvicDRycOLlPnJ/CbHzlwAFgSJcCTbxTQ0er4Xb4SK1nUXe2hCRIDg95W1YgMtyEbGDmAjYQYsG3dzEN7bI/TmICTjw+MunWxpswsEynJEIJmITLHDQpJONN9YNloXDQIgI7FgIgZ2YRgOwHJ5a7SMRERGRmqQETkSqxTY2aws34nQ4cVjlg/eWwyKlsYeUxh5yAzaFoSKgfBqkMYb4pjYJTSF/NbhjwekrX4EyYmzCQZtgAfjrQ/Nu8SQk+ilul8+mH0qxDBiPwbIMtrEhGKGkyEFMXJjGjQuJ5HpxJicQDNs4NxYx68n/Ei7+lpgm35A5+GwantYFd7zuixMREZGjjxI4EamWQCRIWSSAx+GucrvLcpb/YMDGxo5EsFyQdaaDjTNt8tdAuAQwNg4rDE5IyHTS5HQHJiaM5bBoPyABp9ti838LCBTY2A4HDuPABSSmhmh3QgmxCTamqISiYpsSy4V5cxmsL8RyGIpyVrB1YTarX29K94euIT4r/bD1j4iIiMjhoARORKrFssrvTDPRm9L+J1xSRun2nRADOAGHRSgSAZcDp9ci6ywHwe1OCtZAoNjG53MR0zhCwjEObMsQiAQoCZficrpof7qfxmm/sHmThzzjw+2wOSYmn5T6ARwWmBAEwpDvdeDcXEDKthKcGXFgIoCN7Uohd8ka5o96jpNfHoU7zn+Ye0pERESk5iiBE5Fq8To8JHniKQzuINbhJGwsAsZJuKSM3B/WEIoFyx+LCYOjNIgVsokk+yBscPu8+NLdJKQbwnaEBE8sJaESAnYIg8G2bfICBViWhTscISbZ0LS+TZEvhKssgG25CYQsLGOIuFyUlYFjdR4pry/GaXnB6Sx/loEBh9tJXFYaeT+tY/NnC2kyqFdtd52IiIjIIaMETkT2zRisyBa6xgcJh8pwWgEMFnlhNyvX7iBcUITT4SZixYAB59YiHEUhbJ8L43IQKS7DmejCYLAsCwcWEWNHR/OclhOn5cRgCGATSorDHTGkr8+h5YIl5DRtyNZGDQi7XMTsLIJ31+P9biOeSJiQ30/E48dEIjjdTtx+J05PeUL3y8ffKoETERGRo4oSOBHZO2MguBpCq4h3RCh0eCgJB3E6oJ6zjLgmLjyRWFaXGiwbCNtYxeHy0bDcUuy0WOyIwQ6FsF0WLstFcbi0PJnb9bgADDZ29JARhwNXOEy7bxbjLw2Q9ONKjv1hGVgW+dthwadluDyGooCDQE4Jxior39HhxLVpJbGN6uP0eSjZtP2wd5eIiIhITdJz4ERk7yI7ILQacGA54on3JBPjjiNiHOSXRTAui7atE0k1NvH//hnn2jzCabGEYz2Y7cWQXwYeB0EiRIxNyA5RGikjbCI4LAdJnnji3LG4LCdOy4Hf6SGhoAR3MEhBYgwQAjtSHosxOJ3lPxflQ2mhjTEGpxtcMW6cPi+R0gD5K3+hNGcnzlhfrXWbiIiISE1QAiciexfeSPmNbeXJkGVZJHjiqO+rR8zWIoKr8vAVBmm9ajuu15fi++dC3POzwWVBWjwEbQiEsY3BmP8tgrJr9K0kXIbX6aW+vx4N/KmkxKQSiwNjDPkpiViWEywH/LrKZUw8OByGYBk43RZOp1W+3eHBcjpx+r3gsAjsKCCxdWbt9JmIiIhIDakzCdy6desYNmwYTZs2xe/307x5c+68806CwWCFepZlVXo9++yztRS1SB1nQhDeDlblh2M7AG8ggCO3jPDOUup1ro9tgSeniNiJPxD3wDfEPvs9vi/WY+UHCOeEycv34nXE4LKceJ1e3A4XEWOTH8jHNpH/tZ3oBQcYwITC0efKmQiY8K+TLy0wNoAFDk95EgcYY8Auv9fODoUPSzeJiIiIHC515h645cuXY9s2zz33HC1atODHH39k+PDhFBcX8/DDD1eoO3HiRM4888zo+8TExMMdrsjRwUQAE02OKrAsPH4XTo+DSFkEXBYujwM7UH4vm3tbEc7cQnJPb4IjYpOeHKLQKiEcjAF3+cO+LcvC5XBSkh9i9dICHEE3LleQeg1tHPExxIR/Teoiv94f57AoNTHYFOHy2oQDhohlcLoNjoiNidjYoTCW24mvQRL5yzYcnn4SEREROUzqTAJ35plnVkjKmjVrxooVK3jmmWcqJXBJSUmkp+sBviIHzXKD5SofibN222RZOOISiEkIEHI5Kd1eRqQsjGU5cFo2HitEaVIckYYJ+OwQTq+fZAIYE6KQGMImghV2svHLENuWRoiUFuN0OMGO4PQ6SWth0Sc5EUecHzxuLI+bcMRi9af5FOdFsH59HJ2xLUJFZTicDhx+D74GScQek0q4NKgROBERETnq1JkplFXJz88nJSWlUvmIESNITU2le/fuPPvss9i2XcXe/xMIBCgoKKjwEhHK7ztzZZTfA2cqPsDbGIMVn4g33o0/xcum2b8QygsRyg0QziujZGcE2+vCHevAUz+GME4CthMLQ5wJYCI2qz8JkL0wgjHgT4H4VBf+eoADtv0IH21shd2iKdg24bwSfvhgB1t+LsGifFDQHevGk5KAO8aH5XTgSYglqU1jPElxhEvKSGhxTK10m4iIiEhNqTMjcLtbvXo1Tz75JBMmTKhQPm7cOE499VT8fj8zZ87k1ltvZfv27dxxxx17bGv8+PHcddddNR2ySN3kzoTwZjDFQCzG2FCaD6V5EA7hTPXjKCwmLTGfgh7J2DtL8RIiJg5iOzlZHO8m7PNG/7Exlgu/sdm62rDzZ4MnzuDwWmBZhEwYp8NJvZgwXo/Fqg0OlrbsyPH9G7L6jSVs31RETLIbh8dQkh/G8pQvWOL0e7DCToK5RZRty8cd78dyOMgc2KMWO05ERETk0LOM2e3P6ofZ2LFj95k8zZ8/n27dukXfb968mT59+tCnTx9efPHFve47YcIE7r77bvLz8/dYJxAIEAgEou8LCgrIzMwkPz+fhISEap6JyFEsvB0CP2AixVCSjwmVgcOJ5XRgykJEflxJ4JccivNDWNvLcFgQjIunJOsYlrRvQaHbhTdSPtIW63Xhsmzmv2+xdb0Lfz0HERPB5/Tid/rwubw4S3MIB4PsyPeQ0QBGDIEvb51D3s+5xDUqH13L/bmQSNDC6fdiOcrnd4ZLynDF+nAnxNKw33Gc8MQInB53LXeeiNS2goICEhMT9b0uIkeFWk/gtm/fzvbte3/YblZWFj5f+RLmmzdvpl+/fpxwwglMmjQJh2Pvs0Dnzp1Lr169yMnJIS0trVox6R96kcpMpATz87vgLACvHyJAXikmrwzCNoHSEgq3ZePwOilObEBxYgo4HOS6XayN8WFsg6cwiKMkTCRs+PaDGHB7cMU5cDtdpPpSsLAozC8jb0chpSUB7IgBy+LU4yPYU78g3pTijXeA5SAUiid/1TbCxWXl0zstsMMRLIeDFn8+g673XYU3Ka62u01EjgD6XheRo0mtT6FMTU0lNTW1WnU3bdpEv3796Nq1KxMnTtxn8gawaNEifD4fSUlJBxmpyO9cYS72imXg8mJ5Yipt9vpjILkepnQL2Pk4txdhW04sXzyFW2LI9ngJuBxYZRFM2BA2QDCEq9RJUlIylrHYklNI3o4SDOB0urEIY0cM3y0xWCkd6JS/jHRHCLwpeOL81OuSRGBnIcHcQuxwBBOO4EmKo/vD1+KO0UO8RURE5OhT6wlcdW3evJm+ffvSuHFjHn74YbZt2xbdtmvFyQ8++ICcnBx69OiB3+9n1qxZ3H777Vx99dV4vd7aCl3kqGBylkM4AP6kqrfbETyhQgiW4SkIEnGUT120C7YRv6KUyFI3RSdkEWhbD+OxiPOHKdzkwCrIoyQ1gklLJW9HCU6nA6er/I8zQRt8fouMzFjW54X5r92OZN92fL8uQelwOvDXT8Rfv/xRIQVrsknu2AyXX/+/i4iIyNGpziRwn376KatWrWLVqlU0atSowrZds0DdbjdPP/00t9xyC7Zt06xZM+6++26uv/762ghZ5OgSKgPKHx+wO2NsyNsEgaLye+PcHlxeN1huCtblEZdk0/W4MD/PXUtgxipwWBh/SxY5jwGnm9JteZRQPmK2K3mzbYOxISklFqcvlvrpYbZssNkc9tHMXVophnBpAIwh608nVxmjiIiIyNGg1u+BOxJprrxIZfbymdg/fgzxaZUSJFNaUJ7AOZxghyGuHjgC2IEQ23/Mw7IgNtnBLytg00oHeFOw3QkssBqQTQxWIECJPx5fnBfLsrBtQzgUwR/rIbNxEk6XAzscYcMPG0nYsYUeJgd3YiyWVf5A8GBe+eqT6b07csLjIzQCJyIV6HtdRI4mdfo5cCJy+FjpbcDlg2BJ5Y2leYApX0zE4YSYFPBnEI74yxcicTgIhyG1sRNHbBqWOx4nhi5mG40pxDic2MYQDEYIBiLYtiEuwccxmYnRETmHy0ncMfVwHdOAcEmAgpWbKVi1iYKVm4gEQjT+40l0f+gaJW8iIiJyVKszUyhFpJYlZmClt8Zs/C/G6cZyeQAw8Ov0SgtMBPwpWM7y+98sTyKWcxs43URscPsM3hgnZb/OgHRj08VsJ3FHkP82bEtMkh+fz0VsnAevz1VppM820LRPe/rceQZbvvyBYEEJnsRY0vt00kO7RURE5HdBCZyIVItlWTiOG4wdLMVsXYmxHODxl2dwkRBgwJ8I8Q2i+7hifTjcbuxQBLxOjAFjdpt+aQz18rfSpPNx5Lkd1KsfW+Xxw6EIAB2Oa0hi60wSW2fW1KmKiIiIHLE0hVJEqs3yxuHoMQRH1/OwkhuVT5nElCdt3jhIOgbrN4/3cLic+NOTMeEILleEQKmLYMBZoc2yrXl4E2PpM7AdLreT/NxSdr81NxKx2b6tmIxGibTtmH44TlVERETkiKQROBHZL5bbh9X0REzWCeWPFQDMjvXYX0+EYCl4K46gxTaqTzi/ABMuYfPPXuyIwXJYRAIhSrfkgm1oM2IQrc7uSMgfx2cfL2dLdhE+vwunwyIQCBOOGBo2SuLCIV3x+ty1cdoiIiIiRwQlcCJyQCzLAvevD8tOa4XV9ATM6nmYcAD8CVgOF8a2scLFJDaOY8fmJHLWlhLYkQ2WheV0ENu4AS2H9Cfr/D5YlsXJpzanSbNkvv/uF5YvySEcsclITaTLiZl07HIMcfFaoERERER+3/QYgSpouWGR/WfsCPbKLzCr55WvSrnrnxZvHFbmcTja9adkayHbF/yMHQjhq59I/R7tcPk8VbdnDMY2OJya6S0iB0ff6yJyNNEInIgcEpbDibN1P0yzHpitKyFQDC4vVoPmWL7yX5hij/ERe0z96rVnWVhOPZBbRERE5LeUwInIIWW5fVjHdKjtMERERESOSpqbJCIiIiIiUkcogRMREREREakjlMCJiIiIiIjUEUrgRERERERE6gglcCIiIiIiInWEEjgREREREZE6QgmciIiIiIhIHaEETkREREREpI5QAiciIiIiIlJHKIETERERERGpI5TAiYiIiIiI1BFK4EREREREROoIJXAiIiIiIiJ1hBI4ERERERGROkIJnIiIiIiISB3hqu0AjkTGGAAKCgpqORIRERE5WLu+z3d9v4uI1GVK4KpQWFgIQGZmZi1HIiIiIodKYWEhiYmJtR2GiMhBsYz+HFWJbdts3ryZ+Ph4LMuq7XBqXEFBAZmZmWzcuJGEhITaDqfOUL/tP/XZ/lOfHRj12/47mvvMGENhYSENGzbE4dDdIyJSt2kErgoOh4NGjRrVdhiHXUJCwlH3pX04qN/2n/ps/6nPDoz6bf8drX2mkTcROVroz1AiIiIiIiJ1hBI4ERERERGROkIJnOD1ernzzjvxer21HUqdon7bf+qz/ac+OzDqt/2nPhMRqRu0iImIiIiIiEgdoRE4ERERERGROkIJnIiIiIiISB2hBE5ERERERKSOUAInPP300zRt2hSfz0fXrl358ssvazukI9bYsWOxLKvCKz09vbbDOuJ88cUXDBw4kIYNG2JZFtOmTauw3RjD2LFjadiwIX6/n759+/LTTz/VTrBHiH312dChQytdeyeeeGLtBHuEGD9+PN27dyc+Pp4GDRowaNAgVqxYUaGOrrWKqtNnutZERI5sSuB+56ZOncrNN9/M7bffzqJFizj55JMZMGAAGzZsqO3Qjljt2rUjOzs7+lqyZElth3TEKS4uplOnTjz11FNVbn/wwQd55JFHeOqpp5g/fz7p6emcfvrpFBYWHuZIjxz76jOAM888s8K19/HHHx/GCI88c+bM4frrr+ebb75hxowZhMNhzjjjDIqLi6N1dK1VVJ0+A11rIiJHMq1C+Tt3wgkn0KVLF5555ploWZs2bRg0aBDjx4+vxciOTGPHjmXatGksXry4tkOpMyzL4v3332fQoEFA+YhIw4YNufnmm7ntttsACAQCpKWl8cADD3DNNdfUYrRHht37DMpHRfLy8iqNzMn/bNu2jQYNGjBnzhx69+6ta60adu8z0LUmInKk0wjc71gwGGThwoWcccYZFcrPOOMM5s2bV0tRHflWrlxJw4YNadq0KRdddBFr1qyp7ZDqlLVr15KTk1PhuvN6vfTp00fX3T7Mnj2bBg0a0KpVK4YPH87WrVtrO6QjSn5+PgApKSmArrXq2L3PdtG1JiJy5FIC9zu2fft2IpEIaWlpFcrT0tLIycmppaiObCeccAKvvPIK06dP54UXXiAnJ4eTTjqJHTt21HZodcaua0vX3f4ZMGAAr732Gp9//jkTJkxg/vz5nHLKKQQCgdoO7YhgjOGWW26hV69etG/fHtC1ti9V9RnoWhMROdK5ajsAqX2WZVV4b4ypVCblBgwYEP25Q4cO9OjRg+bNmzN58mRuueWWWoys7tF1t38uvPDC6M/t27enW7duNGnShI8++ojBgwfXYmRHhhEjRvDDDz/w1VdfVdqma61qe+ozXWsiIkc2jcD9jqWmpuJ0Oiv9JXrr1q2V/mItVYuNjaVDhw6sXLmytkOpM3at2qnr7uBkZGTQpEkTXXvADTfcwL///W9mzZpFo0aNouW61vZsT31WFV1rIiJHFiVwv2Mej4euXbsyY8aMCuUzZszgpJNOqqWo6pZAIMCyZcvIyMio7VDqjKZNm5Kenl7hugsGg8yZM0fX3X7YsWMHGzdu/F1fe8YYRowYwXvvvcfnn39O06ZNK2zXtVbZvvqsKrrWRESOLJpC+Tt3yy23cPnll9OtWzd69OjB888/z4YNG7j22mtrO7Qj0siRIxk4cCCNGzdm69at3HPPPRQUFDBkyJDaDu2IUlRUxKpVq6Lv165dy+LFi0lJSaFx48bcfPPN3HfffbRs2ZKWLVty3333ERMTwyWXXFKLUdeuvfVZSkoKY8eO5dxzzyUjI4N169bxj3/8g9TUVP70pz/VYtS16/rrr+f111/nX//6F/Hx8dGRtsTERPx+P5Zl6Vrbzb76rKioSNeaiMiRzsjv3j//+U/TpEkT4/F4TJcuXcycOXNqO6Qj1oUXXmgyMjKM2+02DRs2NIMHDzY//fRTbYd1xJk1a5YBKr2GDBlijDHGtm1z5513mvT0dOP1ek3v3r3NkiVLajfoWra3PispKTFnnHGGqV+/vnG73aZx48ZmyJAhZsOGDbUddq2qqr8AM3HixGgdXWsV7avPdK2JiBz59Bw4ERERERGROkL3wImIiIiIiNQRSuBERERERETqCCVwIiIiIiIidYQSOBERERERkTpCCZyIiIiIiEgdoQRORERERESkjlACJyIiIiIiUkcogRMREREREakjlMCJSK0bO3YslmWxffv2Kre3b9+evn37Rt+vW7cOy7KwLIuxY8dWuc+VV14ZrbMnXbp0wbIsHn744Sq3T5o0KdqGZVm4XC4aNWrEFVdcwaZNm/Z5Xq+88goXXXQRrVu3xuFwkJWVtc99RERERPZGCZyI1Fnx8fFMmjQJ27YrlBcVFfH222+TkJCwx30XL17MokWLAHjppZf2epyJEyfy9ddfM2PGDIYPH84bb7zBySefTHFx8V73mzJlCj/99BPHH388zZs3r+ZZiYiIiOyZEjgRqbMuvPBC1q9fz8yZMyuUT506lUgkwh//+Mc97vviiy8CcPbZZ7N8+XLmzZu3x7rt27fnxBNPpF+/ftx5552MGjWKtWvXMm3atL3GN336dJYsWcKUKVNo3bp19U9MREREZA+UwIlIndW6dWtOOukkXn755QrlL7/8MoMHDyYxMbHK/crKynj99dfp2rUrjz76aHSf6jrxxBMBWL9+/V7rORz6J1ZEREQOLf12ISJ12rBhw5g2bRq5ubkArFixgnnz5jFs2LA97vPee++Rm5vLlVdeScuWLenVqxdTp06lqKioWsdctWoVAPXr1z/4ExARERHZD0rgRKROu+CCC3C5XLz++utA+f1sTZs2rbDoye5eeuklfD4fl1xyCVCeBBYVFfHWW29VWT8SiRAOhykqKuKjjz7innvuIT4+fq9TNEVERERqghI4EanT4uLiOP/883n55ZcJh8O88sorXHHFFXtcfXLt2rXMmjWLwYMHk5SUBMD5559PfHz8HqdRnnjiibjdbuLj4/nDH/5Aeno6//nPf0hLS6up0xIRERGpkqu2AxARcbnK/ymKRCJVbg+Hw7jd7j3uP2zYMHr16sW9997Ltm3bGDp06B7rvvzyyxhjOO+888jLy4uW//GPf+S1115j+fLlHHvssRX2eeWVV2jTpg0ul4u0tDQyMjKqf3IiIiIih5ASOBGpdbtGsjZt2lRpVMsYQ3Z2Nt26ddvj/j179qR169bcfffdnH766WRmZlZZz7ZtJk2aBMDgwYOrrPPyyy/z4IMPVihr06bNXo8vIiIicrgogRORWnfKKadgWRZTp06lS5cuFbZ98sknFBQUcNppp+21jTvuuIN33nmH66+/fo91pk+fzi+//ML111/PeeedV2n7iBEjeOWVV7jvvvuio4IiIiIiRxL9hiIita558+aMGDGChx56iLy8PM466yz8fj/z58/n/vvvp1u3btEFR/bksssu47LLLttrnZdeegmXy8U//vEPGjZsWGn7Nddcw4033shHH33EOeecc1DnBLB06VKWLl0KQE5ODiUlJbzzzjsAtG3blrZt2x70MUREROT3RYuYiMgR4fHHH+fpp5/m+++/55JLLmHgwIFMnjyZ66+/nlmzZuHxeA6q/e3bt/PBBx/whz/8ocrkDeDyyy/H7/fz0ksvHdSxdnnrrbc4//zzOf/881m4cCHbtm2Lvt/TipciIiIie2MZY0xtByEiIiIiIiL7phE4ERERERGROkIJnIiIiIiISB2hBE5ERERERKSOUAInIiIiIiJSRyiBExERERERqSOUwImIiIiIiNQRSuBERERERETqCCVwIiIiIiIidYQSOBERERERkTpCCZyIiIiIiEgdoQRORERERESkjlACJyIiIiIiUkf8P8NVtQJ18pM+AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qixin/miniconda3/envs/BIA/lib/python3.11/site-packages/umap/umap_.py:1945: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtkAAAJOCAYAAABiLhtlAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAACCOElEQVR4nOzdd3xTVePH8c9N0qa7BQote28QBESGMhwMcT8oiCJLVBR9FNGfCxk+bnHh40Tm48CJAxcquFAEBEWWbLBQNm3pbnJ+f6QJhKalQEqgfN+vV15t7jz35jb55vTccyxjjEFERERERILGFuoCiIiIiIiUNwrZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtmnkf/973+0bt2amJgYLMti3LhxoS5SserUqUOdOnVCsu9p06ZhWRbTpk0r9TqWZdGtWze/aYMHD8ayLDZt2hTU8p1I27Zt47rrrqN69erYbDYsywp1kUJq3LhxWJbF/Pnzy3Q/ga4nOTmdqGtCRE495TZkz58/H8uyGDx48FEv061bNyzLwrIsvvzyy2LXb9OmjW+5X3/9tdjlHnroISzLwul0smfPnmKXO3S/lmVhs9moUKECXbp0Ydq0aRhjil33SBYsWMDAgQPJysri1ltvZezYsUf8EK9Tp45feQI99u/ff8xlkpPf4MGDeeutt+jevTtjxoxh7NixJS7v/YJy6DUcFxdH3bp1ueyyy5g0aRJ79+49QaU/eXn/1suL9evXM2rUKFq3bk1CQgJhYWFUrlyZrl278sgjj/DPP/+Euojljvdv7fHHHw91UUSkGI5QF+Bk5nA4mDJlCr169Soy748//mDp0qU4HA4KCgqK3Ybb7Wb69OlYlkVeXh7/+9//+Pe//13ifu+66y5iYmJwuVxs2LCBDz/8kB9//JElS5YwadKkYzqWzz//HIAZM2bQoUOHUq9nt9t58MEHi50fERFxTOU5HTz22GPce++9VK9ePdRFOSZ5eXl8++239OjRg//9739Hte7555/POeecA8CBAwf4559/+PHHH/nkk08YO3Ysr776KldddVVZFLtMjRw5kv79+1OrVq0y3c+qVauIiooq030EyzPPPMP//d//4XK56NChAwMHDiQuLo69e/fy22+/MWbMGCZMmMCiRYs444wzQl1cEZETRiG7BL179+bjjz9mz549VKpUyW/eG2+8gcPh4MILL+SLL74odhtz585ly5YtjBgxghkzZvDGG28cMWSPHj2a5ORk3/Ply5dz9tln89///pdRo0ZRt27doz6Wbdu2AfhttzQcDsdJ3azkZFa1alWqVq0a6mIcs9TUVNxu91FfMwAXXHAB9957r980l8vFtGnTuO2227jmmmuIj4+nR48ewSruCZGYmEhiYmKZ76dJkyZlvo9gePXVV7nrrruoX78+7777Lm3atCmyzKpVqxgzZgzp6ekhKKGISOiU2+YiwTBkyBDy8vJ48803/abn5eXx1ltv0adPH6pUqVLiNt544w0AbrnlFq644gqWL1/OokWLjqocLVu2pGvXrhhjWLJkiW/6ggUL6NOnDxUrViQiIoImTZowbtw4srKyfMt4m8RMnToVgLp16/r+lR9s3n+B5+bmcv/991OrVi0iIyNp27Yt33zzDQAZGRncfvvtVK9enYiICDp27MjixYuL3ea+ffsYPnw4SUlJREZG0r59ez755JOAyxpjmDJlCp07dyYuLo6oqCjatWvHlClTAi6/d+9ebr75ZpKSkoiKiuKss87io48+KvEYJ0+eTIsWLYiIiKBmzZrcc8895OTkBFw2UJts7+sxbtw4fv/9d3r27ElsbCzx8fFcccUVxbbf/vDDD2nXrh2RkZEkJSUxfPhw9u3bd9Rt17Oyshg3bhxNmjQhIiKCihUr0qdPHxYsWOC3XLdu3ahduzaA7z8xx9uO3263M2zYMF555RVcLhejRo0q0gQqLy+PZ555hjZt2hAdHU1sbCznnntuwNc8LS2Nhx56iGbNmhETE0N8fDxNmjRhyJAhbN261W9ZYwzTp0+nS5cuJCQkEBUVRcOGDbn55pvZsmWL33F7r+GHHnqIBg0aEBYW5jvuQO1vN23a5Gt29tdff9G7d2/i4+OJi4vjkksuYeXKlX5lsSyL77//3ve793Fos7Xi2mTv2bOHO++8k7p16+J0OqlSpQr9+vUrsg/wv/5eeuklmjZtSkREBLVr12b8+PG43e6Ar1Np7du3j3vuuYeIiAi++OKLgAEboGnTprz//vtF/oPmPcaUlBQGDx5McnIyNpvN79xOnz6dDh06EBMTQ0xMDB06dGD69OlF9lHSfRSH/s0F2v/WrVvp168flSpVIjo6mm7duhX5eyit119/nebNmxMREUGtWrW47777/N4fNmzYgM1mo0+fPgHX37dvHxEREbRq1eqY9l+cJUuWMHLkSFq0aEF8fDyRkZG0bNmSxx9/nPz8/CLLe99XMjMzGTVqFNWrV8fpdHLGGWfw/vvvB9zHpk2b6NevHxUrViQmJoauXbvyww8/BPybOZbX66OPPuKaa66hQYMGREVFER8fz7nnnssHH3xQ7HG/+uqrvtfj0Pfr4v6+MjIyGDt2LM2bNycyMpKEhAR69erFTz/9VOw+REqimuwSdOzYkaZNmzJ16lRuv/123/TZs2ezZ88ehg4dyocffljs+nv27OHjjz/mzDPPpEWLFlx//fX873//44033uCss846rrJ98MEH9O/fn/DwcPr160eVKlX45ptvGD9+PF9//TXz5s3D6XRSp04dxo4dy+zZs/njjz/497//TUJCwnHt+0j69evH8uXLufTSS8nOzubNN9/k4osvZsGCBdx0003k5OTQt29fdu3axaxZs+jZsycbN24kLi7Obzt5eXlccMEFZGdnM2jQIPbv388777zD5ZdfzsyZM7n22mt9yxpjuO6663jrrbdo1KgRAwYMIDw8nLlz5zJs2DBWrlzJ008/7Vs+KyuLbt26sXz5cjp27EjXrl19H7bF1a4+/PDDPPTQQ76QGxYWxqxZs1i1atVRn6PFixfz1FNP0a1bN2666SaWLl3K7NmzWb58OX/99ZdfM5wpU6YwbNgwEhISuP7664mPj+fzzz/nwgsvJD8/n7CwsFLtMzc3l/PPP59ff/2VNm3acMcdd7Bz505mzZrF119/zaxZs7jyyisBT0Br3bo1zz//PK1ateLyyy8HCMrNeNdddx1jx45lxYoV/PXXX7Rs2dJXvl69ejF//nzOPPNMhg0bRn5+PnPmzPG15x45ciTgeb179uzJwoUL6dy5M7169cJms7Fp0yY++ugjBg0aRM2aNX3LXnPNNcyaNYvq1atzzTXXEBcXx6ZNm5g1axa9evUq0vzjyiuv5I8//qBnz55UrFiRevXqHfG4NmzYQOfOnWnfvj233HILa9eu5aOPPuKnn35iwYIFNG3aFICxY8cybdo0Nm/e7NfGvXXr1iVuf8+ePXTo0IF169bRrVs3+vfvz6ZNm3j//feZM2cOc+fOpWPHjkXWu/vuu5k/fz4XX3wxPXr0YPbs2YwbN468vDweeeSRIx5Xcd5//33S09MZOHAgDRs2POLyDkfRj5s9e/bQsWNHKlasSL9+/cjLy/O9D9x5550899xzVK9enWHDhmFZFh988AGDBw/mjz/+4Jlnnjnmsnvt27ePzp07U7VqVW688UZSUlKYNWsW3bt356uvvjqq633ixInMnz+ffv36cfHFF/P555/z+OOPs3TpUr744gssy6JevXpccMEFfPnll/zzzz/UqFHDbxszZ84kNzeX4cOHH/exHer111/n008/pUuXLlx00UVkZWUxf/587rvvPhYtWhQwqObn59OjRw/27t3LlVdeSVZWFu+88w5XX301X375pd/7ZEpKCp06dWL79u1cdNFFtGrVijVr1tCjRw+6d+8elGO47777CA8P55xzzqFq1ars2rWLTz75hL59+/LCCy9w2223+S3/0EMP8fDDD/teW4fDwXvvvcfq1asDbn/v3r106dKFFStWcO6559KzZ0/S0tL4+OOP6d69O++9957vfVCk1Ew5NW/ePAOYQYMGHfUyXbt2NYDZvn27efLJJw1gfv/9d9/8Hj16mKSkJJOfn28GDRpkAPPLL78U2f6zzz5rAPPMM88YY4xxuVymRo0aJi4uzmRmZhZZ/tD9HurPP/80kZGRxrIss3HjRpOenm4SEhKM0+k0f/zxh285t9ttBgwYYADz8MMP+23DW86NGzcWez4OV7t2bWO3283YsWMDPl5++eWA5e/cubM5cOCAb/o777xjAJOQkGCuuuoqk5+f75v3xBNP+J2jQ/cNmPPOO8/k5eX5pq9atcpERkaahIQEk56e7pv+2muvGcAMGzbMb/u5ubnmkksuMYBZvHixb/rYsWMNYIYPH+6336+++soABjBTp071TV+7dq1xOBymevXqZseOHb7paWlppnHjxgYwXbt29dtWoHPuveYA88477/gtP3DgQAOYt99+2zdt3759JiYmxsTGxpr169f7pufn55sLLrjAAKZ27dqmNCZMmGAAc+211xq32+2b/scffxin02kqVKjgd043btx4xL+hw02dOtUA5rHHHitxOe+xvvHGG75p999/vwHMuHHj/MqXnp5u2rVrZ8LDw01KSooxxvM3AZgrrriiyLZzcnJMRkaG7/l///tfA5jzzz/fZGVl+S2blZVl9uzZ43vuvYZbt27tN93Le93MmzfPN817ngDz4IMP+i0/ffp033V8KO9+ihPoeho6dKgBzH333ec3/csvvzSAadiwoXG5XL7p3uuvbt26Ztu2bb7pu3btMgkJCSY2Ntbk5uYWW4YjGTJkiAHMlClTjml97zkbMmSIKSgo8Jv3ww8/GMA0bdrU7N+/3zd9//79pkmTJgYwP/74o2+697o79G/Wy/s3N3bs2ID7HzhwoN/1Nn/+fGNZlmnQoIHf+SyO95qIiIgwf/31l296fn6+ufDCCw1gZsyY4Zv+3nvvGcCMHz++yLbOOOMMExERYfbt23fE/Zb2b80YYzZt2lTkHLvdbt819dNPP/nN877/XnbZZX7XyDfffGMA07NnT7/lr7vuOgOYp556KmAZD/+bOZbX69D3P6+MjAzTsmVLEx8f7/eZumbNGmO3202tWrXM7t27/ZZv3rx5wL8v72fn4ddzamqqqVmzpqlcubLJzs4uUgaRkihkHyFkp6amGofDYUaOHGmMMWbLli3GZrOZu+++2xhjSgzZLVu2NHa73S80/9///Z8BzPTp04ss793vXXfdZcaOHWsefPBBM2DAABMREWEAc/vttxtjjJkxY4YBzIgRI4psY8uWLcbhcJj69ev7TT/WkO19gwz0aNWqVcDyz58/3296QUGBCQsLM4DZvHlzkfIGeg28+/7555+LlOvWW281gJk5c6Zv2hlnnGGio6MDvgl6A9ldd93lm1a3bl0THh5e5AuNMcacf/75RT4Axo8fbwAzceLEIsvPnDnzqEN2ly5dimzHO2/UqFG+adOmTTOAufPOO4ss/8svvxxVyK5Xr54JCwszW7duLTLvpptuKnJOyzJke/8OnnjiCWOM5wtohQoVTIMGDfwCj9cnn3xiADNp0iRjzMHXdMCAAUcsU7NmzYzdbjd///33EZf1XsMff/xxwPklhewKFSr4fbk0xhNkWrRoYQCzZcuWIvspzuHXU25uromMjDSVKlUK+AW9Z8+eRYKn9/oLFIK98/78889iy3AkvXv3NoD58ssvi8xbsmRJkS/ln376aZFjDA8PN7t27Sqyvjf8zZo1q8i8t99+2/eF2utYQ7bdbvd7Xbz69OlT5HwWp7gv7MYYs2jRIt8XPK+8vDyTlJRk6tSp43et//bbbwYw11133RH3aczRheziLFmyxPfF9lDe998NGzYUWad27dqmYsWKvuc5OTnG6XSapKSkIl/a3G6370vR8Ybs4kycOLHI5864ceMMYJ5//vkiy3uvn0P/vnbt2mXsdrvf63SoF154wQBFrmGRI1FzkSNISkrioosu4q233uLpp59m2rRpuN1uhgwZUuJ6v/32G8uXL6dXr15+N44NGjSIJ554gilTpnD99dcHXHfixImAp81gXFwcZ511FsOGDfMtv3TpUiDwv+5r1qxJ/fr1WbNmDRkZGcTGxh7LYfs4nc5i2xwX58wzz/R7brfbqVKlCpmZmUX+Le+9MTAlJaXIdsLCwgL2hHLuuefy3//+l2XLlnHdddeRlZXF8uXLqVatWsDurLxtDr3/JszIyGDjxo00a9Ys4E195557Lt9++63ftD/++MM3L9DyRytQ+1Xvv44P7RbRu99OnToVWb59+/YB/wUfSHp6Ohs2bKBp06ZF/kUNnmvp1Vdf9Z3TsmYOa4u9Zs0a9u3bR7Vq1Rg/fnyR5Xft2gUcfA2bNm1Ky5Yteeutt9i6dSuXX3455557Lm3atMFut/vWy8zMZOXKlTRo0KBUTRq82rdvf9THdOaZZxIdHe03zbIszjnnHP766y/++OMPXxOWo7V69Wqys7Pp1q1bwF5HunXrxldffcWyZct8vbp4lfZaO1qHv4aH+v3334u8jjfddBMXX3yx37S6desGvJG0pPc477Rly5YdXYEDqF27dsDX5Nxzz2XOnDkBz2dxAr0PeO+jOLSsYWFhDB06lMcee4y5c+f6ml1479+54YYbjuFISpaXl8eLL77IO++8w+rVqzlw4IDf6+e9Mf5QCQkJAW+yr1GjBr/88ovv+Zo1a8jNzaVdu3aEh4f7LWtZFh07diy2icbR2LlzJ48//jhffPEFmzdvJjs722/+ocdQ0vtmoGmLFi3C5XKRk5MT8L6TtWvXAp6/w8OvYZGSlNuQbbN57uks6eYe7zzvssUZMmQIn3zyCR999BHTpk2jQ4cOvvaVxfHebDdw4EC/6U2bNqVdu3Z8//33rFu3jgYNGhRZd/v27SX26OC9Sz8pKSng/OTkZNasWUN6evpxh+xjcXjbavC0x4yPjw84HQh4802lSpUCvjbe405LSwM87SqNMaSkpAQMaF6ZmZl+6xV302qg81rSOsW9DiUp6Vy4XC7fNO9rXbly5SLL22y2Uvd0UZprBg4eZ1nbvn07cPC4vH1nr1ixghUrVhS7nvc1dDgcfPfdd4wbN44PP/yQu+66C/D0/nHbbbfxwAMPYLfbfSHyaLtRPJbX9EjX0/Gc2+N5/Up7rR0tb1kCfUG+4YYbfGFx/vz5xbbLLe540tPTsdlsAa/7pKQkbDZbUK7VYL5mxW2rSpUqRc7R8OHDefzxx5k8eTI9evQgKyuLt99+m0aNGtG1a9dS77O0+vbty6effkqjRo189/CEhYWxf/9+nn/+eXJzc4usE+i6Ac+1c+jnaknvUXBsf0uH27t3L2eddRZbtmyhc+fOXHDBBSQkJGC321m2bBkff/yx3zGUVKZA5fG+//z888/8/PPPxZbD+/4jUlrltncR7xtESYO/7N6922/Z4lx88cUkJSUxevRoNmzYwNChQ0tc3vuGCXDttdcWGcDF25tGcb1eHIk3xO7YsSPgfO/0QGH3VLJnz56AX5K8x+d93bzH2bZtW4ynCVTAx7x58/yW37lzZ8D9Bjqv3n0FWqe41yEYvGX11uQeyu12+67h0m7nZLhm3G43P/zwA4DvBmDvfv/1r3+V+Bp6e8kBT6B+8cUXSUlJYeXKlbz44otUqlSJsWPH8uSTTwIHX7dAQbAkx9L7zpGupyO9z5TkZHr9vLw1gt6/q2NR3HmOi4vD7XYHvO537tyJ2+32O1bvl/FAYxaUFJSD+ZoVt62dO3cW2U7dunW58MIL+fjjj9m9ezfvvvsu6enpZVKLvWjRIj799FN69uzJypUref3113nkkUcYN24c/fv3P+7tl/QeBYGv2aN9vd544w22bNnCf/7zH3766ScmTZrEww8/zLhx4wL+t7OkMgUqj3f5u+66q8T3nyMNxiVyuHIbshs3bkx4eDiLFi0qdrAY77+8jjRAgsPh4LrrriMlJYWoqCj69etX4vLeu+5bt27NsGHDAj7CwsKYPn36MdUkeZtjBBrGNyUlhfXr11OvXr2Q1GIHU35+fsCRNH/88UfgYG8MsbGxNG3alFWrVpXq39/eEQjXrVtHampqsds/lLdLrUDzAk0LFu9+A3Up9ttvv5U4ENKh4uLiqFevHuvWrQsYOL1dyh2ph4tgmDlzJps3b6Zly5Y0b94c8PyHJy4ujsWLFwf8r0ZJLMuiadOm3HrrrcydOxfA1+VfTEwMzZo1Y+PGjb5/+ZaVpUuXBqzp8taMHdotm7dJS2n//r1dLi5atMivi06vE/n6efXt25fY2Fjee++9oJ/bkt7jAh1rhQoVgMBfprxNTwLZvHlzke4eoeh7TGkEeh9YvHgx2dnZAbdz4403kpeX5xs/ISwsjEGDBpV6f6W1fv16APr06ePXlKq4Mh+txo0b43Q6WbJkCXl5eX7zjDEB38OP9vXyHsOll15aZF5J79eB3jcDTTvrrLOwLMuvGYxIMJTbkB0REcHVV1/Nrl27+M9//lNk/vLly5k8eTKxsbFcccUVR9ze3XffzUcffcRXX311xNoib9u6Z599lsmTJwd8XHzxxWzbtq3EgWyKc9lllxEfH8/UqVP9/rVujOG+++4jPz+/xOHkTyVjxozxC12rV69mypQpxMfHc9lll/mm33777WRlZTF8+PCAQWfjxo1+fVAPHDiQvLw8HnroIb/lvv766yLtsQEGDBiA3W7nmWee8auxSk9PD3h9Bctll11GTEwMkydPZuPGjb7pBQUFjBkz5qi2NWjQIPLz87nvvvv82mP+9ddfTJ06lfj4+DLtosrlcjFlyhRGjBjhO5femkyHw8GIESPYvHkzo0ePDhi0//rrL9+537hxY8C+ob21VJGRkb5pt956Ky6Xi1tuuaVIO86cnJygDfO+b9++IvcEzJgxg+XLl3Peeef5tf2tWLEiQKmHGw8PD+eaa65h9+7dPPbYY37zvvnmG7744gsaNGhA586dj+sY6tSpU6Rv9+JUrFiRxx57jNzcXHr37l1smD2Wdt/esDl+/Hi/QWzS09N9TcIODaRt2rTBsizeeecdv3tI1q5dy/PPP1/sflwuFw888IDf38P333/P559/ToMGDQK23y3OzJkz/d6PCwoKuP/++4uU1euyyy4jOTmZiRMn8tNPP3HppZcecdyFY+Ht7/7wvp5XrFhR5Fo6Fk6nk759+5KamsoLL7zgN2/GjBkBuzg92teruGN46623fKMZH6p///7YbDaeeeYZv/9mZ2ZmBuy2Mjk5mauvvpoFCxbw1FNPBbzfYOHChQG/4IqUpNy2yQbPDYQLFy5k/PjxfPbZZ3Tt2pWIiAj+/vtvPvnkE4wxvPnmm6XqNzopKalUAWTdunX88MMP1KtXr8S2dUOGDOGjjz7ijTfeOOobKeLi4nj99de55pprOPvss+nXrx+VK1fm22+/ZfHixbRv35677777qLZZnIKCghIHIBk8ePBRDYZyNKpWrcr+/ftp3bo1ffr0IS0tjbfffpucnBxef/11v5r6m266iV9//ZXp06fz888/c8EFF1CtWjV27NjB6tWrWbhwIW+99ZavrPfccw8ffvghr7/+OitWrKBLly5s3bqVd999lz59+jBnzhy/sjRo0ICHHnqIsWPHcsYZZ3D11VfjcDj44IMPaNmyJWvWrCmTc5CQkMAzzzzDjTfeSJs2bejXr5+vn2yn00m1atWOeE+B1z333MOcOXOYOXMmq1at4vzzz/f1VZ6fn8+MGTOC9t+Pb775xvfhmZWVxT///MMPP/xASkoKFStWZObMmVxwwQV+64wfP57ff/+dF154gTlz5tC1a1cqV65MSkoKy5cv548//uCXX36hSpUq/PHHH1xxxRWcddZZtGjRguTkZFJSUpg9ezZ2u93XRhtgxIgRfP/997z77rs0bNiQSy+9lLi4OLZs2cJXX33FG2+8EZQvF+eeey4vvPACv/76K2eddRZ///03H330EfHx8bz44ot+y5533nm8//77XHXVVVx00UVERETQsmXLYgcpAXjiiSf4/vvv+c9//sOCBQs4++yzff1kR0VFMXXq1FJfC8XxNs8q7Q21t956K5mZmdx///20bduWDh060K5dO2JjY9mzZw+rVq3ixx9/xOl0HtXYAF26dOG2225j0qRJtGjRwteM6MMPP2Tr1q3cfvvtdOnSxbd89erV6devH++88w5t27alV69e7Ny5k48++ohevXoVO2DJGWecwfz58+nQoQPnnXce27Zt45133iEsLIzXX3/9qM7nBRdcQIcOHejfvz8VK1bk888/56+//qJnz54BbyZ2OBwMHTqURx99FDj2Gx5L6vt5wIABnH/++bRv3553332X7du306FDB7Zs2cInn3xCnz59ih1c5mg89thjfPPNN9x9993MmzeP1q1bs2bNGj777DN69erFl19+6Xcuj/b1GjhwIE888QS33XYb8+bNo3bt2vz555988803XHnllUXGq2jcuDH33nsvjz76KC1btuSqq67C4XDw4Ycf0rJlS/76668ir+1LL73EmjVruOeee5g5cyYdO3YkPj6erVu3smTJEtauXcv27dsD3ngsUqyy78AktPbv32/Gjh1rWrVqZaKjo01YWJipWbOmGTBggF/f14cqrr/qQA7vwu/ee+81ULSf6sPl5+ebpKQk43A4TGpq6lHv1xhPX7K9e/c2CQkJJjw83DRq1MiMGTOmSDdih5YzmF34cVi3TCV1S1a7du1iu5ojQPd33uX37NljbrjhBlOlShXjdDpNu3btiu1ezRhjZs2aZS644AJToUIFExYWZqpXr266detmJk6cWKSrsD179pgbb7zRVK5c2URERJi2bduaDz/8sMTupV5//XXTrFkzEx4ebmrUqGFGjx5tsrKyjroLv0DdU5XUZd57771nzjzzTON0Ok2VKlXMDTfcYPbs2WNiYmKKdKVYkgMHDpgxY8aYRo0amfDwcJOQkGB69+4dsKuy4+nCz/uwLMvExMSYOnXqmEsuucRMmjTJ7N27t9j1CwoKzKuvvmo6d+5s4uLijNPpNLVq1TK9evUyL7/8su/a3rp1q7n33ntNhw4dTJUqVUx4eLipVauW6du3r1m4cGGR7brdbjN58mTToUMHEx0dbaKiokzDhg3NzTfffFRd65XUhd+gQYPMn3/+aXr16mViY2NNTEyM6dOnj1/fyV75+fnmnnvuMbVq1TIOh6PIeQ50PRnj6Wrs9ttvN7Vr1zZhYWEmMTHR9O3b1yxfvrzIsiX9zQc6jn379hmbzWY6d+5c7PEXZ82aNeb22283LVu2NHFxccbhcJjExERz7rnnmgkTJgTsJq+4YzzUlClTzFlnnWWioqJMVFSUOeuss4rtlzszM9PcdtttJikpyTidTnPGGWeYN998s8Qu/Lp27Wo2b95srrrqKlOhQgUTGRlpunTpUqTf6JIcei5fffVV06xZM+N0Ok2NGjXMvffeW6Rv9kOtWbPGAKZWrVql6pP7UIf/rQV6PPvss8YYY3bu3GmGDh1qqlWrZiIiIkzLli3Nf//7X7Nhw4aAf+MlvV8X9zeyYcMGc9VVV5n4+HgTFRVlzj33XPP999+bkSNHGsAsXbrUb/mjfb2WLVtmevToYSpUqGBiY2NN165dzTfffFPi+/VLL71kmjZt6vd+vXXrVgOePsAPl5WVZZ588knTtm1bEx0dbSIjI03dunXN5ZdfbmbMmOE3BoNIaVjGlNAPk4ictNatW0fDhg25+uqrmTVrVqiLc9ratGkTdevWZdCgQQGHiT5VfPbZZ1xyySXMmTOHiy66KNTFKXOWZdG1a9eA7b5PlHfffZd+/foxfvz4Ik3XyotzzjmHX375hbS0NGJiYkJdHL755hsuvPBC7rnnHp544olQF0fKuXLbJlukvNi3b1+RLrays7O58847ATTUrwTFjz/+SKtWrU6LgH0yMMbwzDPP4HA4GDZsWKiLc9y83XIe6s033/Q13zvRAXvXrl1Fbizev38/9913H6D3TTkxynWbbJHy4Pvvv2fYsGH06NGDWrVqsXv3br777js2bdrEeeedd8TebkRK44knnlDN3gmwfPlyPvvsMxYsWMDChQu5+eabj7of95NRixYtOPPMM2nWrJmv/+r58+cTGxvL008/fcLL8+abb/L0009z3nnnUa1aNbZv386XX37Jzp07GTx4MB07djzhZZLTj0K2yEmuefPmXHjhhfz888/Mnj0b8NyI+fDDDzN69OjjvtlNRE6cJUuWcP/995OQkMD111/PU089FeoiBcXNN9/Mp59+yuLFi8nMzKRy5coMGDCAMWPG0KRJkxNenk6dOtG2bVu++eYb9u7di91up2nTpowZM4ZbbrnlhJdHTk9qky0iIiIiEmSqAhMRERERCTKFbBERERGRIFOb7ADcbjfbtm0jNjbWNyKdiIiInJqMMWRkZBzVAF4ix0shO4Bt27b5DYEsIiIip76tW7dSo0aNUBdDThMK2QF4h5beunUrcXFxIS6NiIiIHI/09HRq1qzp+3wXOREUsgPwNhGJi4tTyBYRESkn1ARUTqRy1zBp3LhxWJbl90hOTg51sURERETkNFIua7KbN2/ON99843tut9tDWBoREREROd2Uy5DtcDhUey0iIiIiIVPumosArF27lmrVqlG3bl369+/Phg0bSlw+NzeX9PR0v4eIiIiIyLEqdyH77LPPZsaMGXz11Ve8/vrrpKam0qlTJ/bs2VPsOo899hjx8fG+h7rvExEREZHjYRljTKgLUZYyMzOpX78+99xzD6NGjQq4TG5uLrm5ub7n3q5+0tLS1LuIiIjIKS49PZ34+Hh9rssJVS7bZB8qOjqali1bsnbt2mKXcTqdOJ3OE1gqERERESnPyl1zkcPl5uayatUqqlatGuqiiIiIiMhpotyF7NGjR/P999+zceNGFi5cSN++fUlPT2fQoEGhLpqIiIiInCbKXXORf/75h2uuuYbdu3dTuXJlOnTowK+//krt2rVDXTQREREROU2Uu5D9zjvvhLoIIiIiInKaK3fNRUREREREQk0hW0REREQkyBSyRURERESCrNy1yRYRKd8KgAzABYQBsai+RETk5KOQLSJySnAB24DdQD5g8IRrJ5AEVAGskJVORET8KWSLiJz03MAGYC9gxxOsbYXTc4HNQB5QAwVtEZGTg/7HKCJy0tsF7MMTrr0B2+Cp3bYKf9+OpxmJiIicDBSyRUROagZPyLbw1GIbPLXXGUAmkFX4PBtYj6fNtoiIhJpCtojIScsN7ADSOBisD+AJ1m4OBm9b4e8ZwDoUtEVEQk9tskVETgp5wB48Nzbm4gnKBXjCtDdQu/DUZAM4KMjMJ3tnBrl7szBuQ1ick+gaBrszBpujRgiOQUREvBSyRURCLgPPjY3Z+AfrQxkOvakxM2UfGRv3Y/JdWDYbWFBwIA8MuPP3EV09BmfFhBNUfhEROZyai4iIhFQuBwO2wT9cH95TiKcWO3tXFhnr94LbjSM6HHtkGPYIB47oCFzZLiy7ixXPvYnb5TohRyAiIkUpZIuIhNQeIAfPwDIFHOw5xMs/aBtjyNyahnEb7BEODgZzy7OuZcMW5mDvHxvYvWjNCTkCEREpSiFbRCRkDJ422DYODjBzhH6uDbjzXNid9oCzHVEOXLkuMrelkfL14uAWV0RESk1tskVEQsbgqb224wnZRx5Ixu1yY3PYyNufgzEGy7JhC7djdwI2N45oBzsX7MCda8hK2V22xRcRkWIpZIuIhExhEw9fl3vmkOnmsGmwb8VO0tfvI6FpIgW5LgoycjHGhTvfhTEOYuvGkJ2aTeoPqZgCN44o5wk7EhER8aeQLSISMhZQAUjF83ZcgCdUe4ZMN243xuUGm0XW9gwWjZ5L9s5MGg49k6TONYlKigYDxgJ3rou9y1LZ/FEK2TvyAEjq3CJUByYictpTyBYRCalEPO2yC/DWYBu3hSsnH1dOAZbDhmWz+Pv1paSv20dM7XjWvLKEtVOXUf3CejgrReHKKWD34u2kfr+J6BrJYAxRNRKpdmHb0B6aiMhpTCFbRCSkooFawBagAOMuwJWXj7vAjd1pxwB5+3NI/X4zxuVm31+7iGtYkcyt6ax6cRGW3dObCJbBXWDI2JhKpTMbcua4wYTHx4T20ERETmPqXUREJOQqAw2BJHL25OLKdWEPt2MMFGTmk7k1g6x/0slLz6MgM4/0dXuJb1qZ2PoVsTsduPMLcOe6wLIRHh9Nu8dvIKlT81AflIjIaU012SIiJ4U4srbn8+OQLwmPt4hvHI9ls+POd5G+di+5+3I8fWNHhlGQmU9+Wg4xteKIrhFHfmYBuMMoyHRjc4ZTsXWDUB+MiMhpTyFbROQksXvxGjI378VWP5G9y3bg+Wejp1s/Z8VoslMzsJwWGEPe/hwiEiOxbHbCY2OBcHJ2pVCjayvCoiNDeRgiIoKai4iInDTcuflgWVg2J56+sw8OsR6ZFIdlt+HKdQEWxmUKlykM2LvTsDvDqXX5OSEpu4iI+FPIFhE5SURUTsCyWbhy8wBvH9duwBBeIZLoWgngNrjzXBgDrjw7+Rm5ZGzYTkFmDg2H9SbpnJahOwAREfFRcxERkZNE5Q5Nia6dRFbKbmJqVSmcmg24sSyIqR2PcbnJ3pGBLdxJ1rZMbGEOKp7ZgHr9ulOjTwcs68ijRoqISNlTyBYROUnYneE0HNSTZf+ZSfaOfURUScCyHHiGXC8gPyMb47LRaPhFNBpyMfnp2YTFRhHXqIbCtYjISUYhW0TkJFLnqq7k7s3g7ymfk742BUdkOFg2sFzENahA7cs70mj4lYRFx4W6qCIiUgKFbBGRk0hengt3p3bEVapG2rK/id6+hVrnVCS5a3Uik2NxRDmxrHVAApAMRIW2wCIiEpBCtojIScDtcrPg+40s+H4j+/dm4XYbomKiufLfbalWL5zIqEjs9vDCpV3ALiAdaABoZEcRkZONQraISIgZY/hi9kp+mr8Bh8NGhUqROBx2OnaF5KqQsjWXqCioXjMcm92Gp2MoB56bIjcBzVBnUSIiJxe9K4uIhNj6v3fz60+biIoKo2KlKBwOO9GxULueRV6uhd1m50BGLvv3ZR+yloWnm78sIC00BRcRkWIpZIuIhNjS37aSl+ciJtbpm5ZUFcKdkJMNNpuFZVmk7cvBGHPImnbAABknusgiInIECtkiIiG2Ye0eIiL8W+85vPm5kN1hkZdXQH6eK8AWAk0TEZFQUsgWEQkhYwwulxvbYf1cZ2WBMWCzH7Z8wGfhiIjIyUUhW0QkhCzLIrlaHDk5+YdMNWz/J4+MNBfR0QYwuF0Gu92Gw3Ho23YBnhsgK57QMouIyJEpZIuIhFibs2sCnuYgkAdk4HJlsfLPHLDcREa5cLldxCWEY7d737YL8IwEWRGIDFHJRUSkOOrCT0SkWHl4eu8wQASeHj324unNw104rRIQx/HUWTQ7oyoNmlTm75XbiE8wRERaWJadv1e5sDtyad4qjORqduLiXYXloXB/iUCtY96viIiUHYVsEZEicoFtwD48NcbgCdUuDvZRbQGZeEJ3HFAPCDumvYWH2+k3qAUfvb2Hv1dlkbbfYFku3AZ278xj965ILruqAg6Hu7AcSXgCdmxhOURE5GSjkC0i4icXWIsnQDvw1Fbn4xn4xXCwxjq8cJ4b2A9sBBpyrKE3JjaT64YnsnWzxao/D3Agw0VEpI0GjaNp0CQau93CE/hdQBU0yqOIyMlNIVtExM8WPMOV2/EEWjeQw8GePNx4gng+nsAdgSdwpxWuF3+M+83Gsixq1YmiVp2oYpax42nCkoNCtojIyU0hW0TEZy+wA0+Q9jYTMfh3nGcd8tzgqeGOLFxnL8ceskVEpDxR7yIiIoAnLK/HE65teGqNDxsRBg577n0LzcETvnOPY//e2uvD93coV2GZ1JuIiMjJTiFbRASA7XiaYlgcbFddUuD1sjhY820/wrIlqYCn2UleMfNN4bwYDgZyERE5WSlki4iQh+fmxTA8b4uHNgcpjnXYTzeeXkaOVThQrXCfOYXb83LhqWkPB2qgHkVERE5+apMtIkIOnprocDw3NLo5cpA9dL4hOCMvVsET8rdxsOmJt0eTGDx9YuuGRxGRU4FCtogIcLCZSASeAV8ODdqH3uwIntDrneYq/L0qx9pPtn8ZKuMJ6/s5GLSj8dSSqwZbRORUoZAtIkIknrfDAjxBORL/JhuBmo24C6d7b0SsEcTy2PGMJCkiIqcqhWwREcLw1B6nFv7uLPyZV/gowL/ZhoWnBtuF5220Dsd306OIiJQ3CtkiIgAk4xlMJgtP22w7Bweayebg4DMFh6wTBVTn+Ntii4hIeaOQLSICeGqvGwGb8YRtb3d+4AnbdQp/ekd/jMAz8Iw6aRIRkaIUskVEfLxBOwvIwBOmw/CEab1diohI6elTQ0SkiCg04IuIiBwP/Z9TRERERCTIFLJFRERERIJMIVtEREREJMgUskVEREREgkwhW0REREQkyBSyRURERESCTCFbRERERCTIFLJFRERERIJMIVtEREREJMgUskVEREREgkwhW0REREQkyBSyRURERESCTCFbRERERCTIFLJFRERERIKs3Ibsl156ibp16xIREUHbtm358ccfQ10kERERETlNlMuQPWvWLO644w4eeOABli5dyrnnnkvv3r3ZsmVLqIsmIiIiIqcByxhjQl2IYDv77LNp06YNL7/8sm9a06ZNufzyy3nssceOuH56ejrx8fGkpaURFxdXlkUVERGRMqbPdQmFcleTnZeXx5IlS+jRo4ff9B49erBgwYIQlUpERERETieOUBcg2Hbv3o3L5SIpKclvelJSEqmpqQHXyc3NJTc31/c8PT29TMsoIiIiIuVbuavJ9rIsy++5MabINK/HHnuM+Ph436NmzZonoogiIiIiUk6Vu5CdmJiI3W4vUmu9c+fOIrXbXvfddx9paWm+x9atW09EUUVERESknCp3ITs8PJy2bdsyd+5cv+lz586lU6dOAddxOp3ExcX5PUREREREjlW5a5MNMGrUKAYOHEi7du3o2LEjr732Glu2bOHmm28OddFERERE5DRQLkN2v3792LNnDxMmTGD79u20aNGCzz//nNq1a4e6aCIiIiJyGiiX/WQfL/WnKSIiUn7oc11Cody1yRYRERERCTWFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSArdyG7Tp06WJbl97j33ntDXSwREREROY04Ql2AsjBhwgSGDx/uex4TExPC0oiIiIjI6aZchuzY2FiSk5NDXQwREREROU2Vu+YiAE888QSVKlWidevWPPLII+Tl5YW6SCIiIiJyGil3Ndn//ve/adOmDRUqVOC3337jvvvuY+PGjUyePLnYdXJzc8nNzfU9T09PPxFFFREREZFyyjLGmFAX4kjGjRvH+PHjS1xm0aJFtGvXrsj0Dz74gL59+7J7924qVap0VNtPS0sjLi7u2AotIiIiJ4X09HTi4+P1uS4n1CkRsnfv3s3u3btLXKZOnTpEREQUmZ6SkkKNGjX49ddfOfvsswOuG6gmu2bNmvpjFBERKQcUsiUUTonmIomJiSQmJh7TukuXLgWgatWqxS7jdDpxOp3HtH0RERERkcOdEiG7tH755Rd+/fVXunfvTnx8PIsWLeLOO+/k0ksvpVatWqEunoiIiIicJspVyHY6ncyaNYvx48eTm5tL7dq1GT58OPfcc0+oiyYiIiIip5FyFbLbtGnDr7/+GupiiIiIiMhprlz2ky0iIiIiEkoK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiepwYMHc/nll4e6GCeNU+l8KGSLiIiIHAfLskp8DB48+ISXad68eVx00UVUqlSJqKgomjVrxl133UVKSkqpt9GtWzfuuOOOsivkMXj++eeZNm1aqItRKgrZIiIiIsdh+/btvsdzzz1HXFyc37Tnn3/+hJbn1Vdf5YILLiA5OZkPPviAlStX8sorr5CWlsbEiRNPaFmCxeVy4Xa7iY+PJyEhIdTFKRWFbBEREZHjkJyc7HvEx8djWZbveVhYGDfffDM1atQgKiqKli1b8vbbb/ut//7779OyZUsiIyOpVKkSF1xwAZmZmQH3tWTJEqpUqcIjjzwScP4///zD7bffzu23386UKVPo1q0bderUoUuXLkyePJmHHnoIgD179nDNNdcUW67Bgwfz/fff8/zzz/tq5Ddt2gTAypUrueiii4iJiSEpKYmBAweye/du37oZGRlce+21REdHU7VqVZ599tkiteL79u3j+uuvp0KFCkRFRdG7d2/Wrl3rmz9t2jQSEhL47LPPaNasGU6nk82bNxdpLmKM4cknn6RevXpERkbSqlUr3n//fb/9XHvttVSuXJnIyEgaNmzI1KlTS35Bg0QhW0RERKSM5OTk0LZtWz777DP++usvbrzxRgYOHMjChQsBTy34Nddcw9ChQ1m1ahXz58/nyiuvxBhTZFvz58/n/PPPZ/z48TzwwAMB9/fee++Rl5fHPffcE3C+txb4SOV6/vnn6dixI8OHD/fVyNesWZPt27fTtWtXWrduzeLFi/nyyy/ZsWMHV199tW8fo0aN4ueff+aTTz5h7ty5/Pjjj/z+++9+5Rg8eDCLFy/mk08+4ZdffsEYw0UXXUR+fr5vmaysLB577DEmT57MihUrqFKlSpHjefDBB5k6dSovv/wyK1as4M477+S6667j+++/B2DMmDGsXLmSL774glWrVvHyyy+TmJhY3MsVXEaKSEtLM4BJS0sLdVFERETkOJ3Iz/WpU6ea+Pj4Epe56KKLzF133WWMMWbJkiUGMJs2bQq47KBBg8xll11mZs+ebWJjY81bb71V4rZHjBhh4uLijqnsh5bLGGO6du1q/v3vf/stM2bMGNOjRw+/aVu3bjWAWbNmjUlPTzdhYWHmvffe883fv3+/iYqK8m3r77//NoD5+eeffcvs3r3bREZGmnfffdcY4zmPgFm2bJnfvrznwxhjDhw4YCIiIsyCBQv8lhk2bJi55pprjDHGXHLJJWbIkCFHfzKCwHFioryIiIjI6cflcvH4448za9YsUlJSyM3NJTc3l+joaABatWrF+eefT8uWLenZsyc9evSgb9++VKhQwbeNhQsX8tlnn/Hee+9xxRVXlLg/YwyWZR13uYqzZMkS5s2bR0xMTJF569evJzs7m/z8fNq3b++bHh8fT+PGjX3PV61ahcPh4Oyzz/ZNq1SpEo0bN2bVqlW+aeHh4ZxxxhnFlmXlypXk5ORw4YUX+k3Py8vjzDPPBGDEiBH861//4vfff6dHjx5cfvnldOrUqcRjDBaFbBEREZEyMnHiRJ599lmee+45WrZsSXR0NHfccQd5eXkA2O125s6dy4IFC/j666+ZNGkSDzzwAAsXLqRu3boA1K9fn0qVKjFlyhT69OlDeHh4sftr1KgRaWlpbN++napVqx5zuYrjdru55JJLeOKJJ4rMq1q1qq9d9eFB3xzS/MUEaArjnX7oepGRkSV+YXC73QDMmTOH6tWr+81zOp0A9O7dm82bNzNnzhy++eYbzj//fG699Vaefvrpkg4zKNQmW0RERKSM/Pjjj1x22WVcd911tGrVinr16vnd4AeeQNq5c2fGjx/P0qVLCQ8P56OPPvLNT0xM5LvvvmP9+vX069fPr93y4fr27Ut4eDhPPvlkwPn79+8vdbnCw8NxuVx+09q0acOKFSuoU6cODRo08HtER0dTv359wsLC+O2333zrpKen+227WbNmFBQU+Np/g+dGzL///pumTZsWe2yH894QuWXLliJlqVmzpm+5ypUrM3jwYP73v//x3HPP8dprr5V6H8fjqEJ2SkoKCxcuZN++fQHn7969mxkzZgSlYCIiIiKnugYNGvhqqletWsVNN91Eamqqb/7ChQt59NFHWbx4MVu2bOHDDz9k165dRcJmlSpV+O6771i9ejXXXHMNBQUFAfdXs2ZNnn32WZ5//nmGDRvG999/z+bNm/n555+56aabePjhh0tVLoA6deqwcOFCNm3axO7du3G73dx6663s3buXa665ht9++40NGzbw9ddfM3ToUFwuF7GxsQwaNIi7776befPmsWLFCoYOHYrNZvPVSjds2JDLLruM4cOH89NPP/HHH39w3XXXUb16dS677LJSn9vY2FhGjx7NnXfeyfTp01m/fj1Lly7lv//9L9OnTwfgoYce4uOPP2bdunWsWLGCzz777KiC/PEoVch2u93ccMMN1KpVi06dOpGcnMzdd99d5JvU+vXrGTJkSJkUVERERORUM2bMGNq0aUPPnj3p1q0bycnJfl3QxcXF8cMPP3DRRRfRqFEjHnzwQSZOnEjv3r2LbCs5OZnvvvuO5cuXc+211xapZfa65ZZb+Prrr0lJSeGKK66gSZMm3HDDDcTFxTF69OhSlQtg9OjR2O12mjVrRuXKldmyZQvVqlXj559/xuVy0bNnT1q0aMG///1v4uPjsdk8sfKZZ56hY8eOXHzxxVxwwQV07tyZpk2bEhER4dv21KlTadu2LRdffDEdO3bEGMPnn39OWFjYUZ3fhx9+mIceeojHHnuMpk2b0rNnTz799FNfU5vw8HDuu+8+zjjjDLp06YLdbuedd945qn0cK8sU1zDmEK+//jojRoxgxIgRtG/fnm+//Zb//e9/dOrUiTlz5hAbGwt4vo116tSp2Bf9VJGenk58fDxpaWnExcWFujgiIiJyHPS5HlqZmZlUr16diRMnMmzYsFAX54QpVU32q6++yh133MGkSZMYOHAg06ZN4/PPP2f58uVceOGFpKWllXU5RUREROQUsHTpUt5++23Wr1/P77//zrXXXgtwVE1ByoNShey1a9cW+bdFjx49+Pbbb1m3bh0XXHBBse20RUREROT08vTTT9OqVSvf6JU//vjjiRsE5iRRqi78wsPDA97J2qZNG+bOncuFF17I+eefz3/+85+gF1BERERETh1nnnkmS5YsCXUxQq5UNdkNGzbkl19+CTjvzDPPZO7cuWzZsoVBgwYFtXAiIiIiIqeiUoXs8847j+nTpxd7Q+OZZ57JN998U2zn4iIiIiIip5NSNRcZOXIkHTt25MCBA8THxwdcpnXr1ixYsKDYGm8RERERkdNFqbrwO92oqx8REZHyQ5/rEgoaVl1EREREJMgUskVEREREgkwhW0REREQkyE6pkP3II4/QqVMnoqKiSEhICLjMli1buOSSS4iOjiYxMZHbb7+dvLy8E1tQEREROa0Yt5u0v7eya9Fq0v7einG7y3yfgwcPxrIsLMvC4XBQq1YtRowYoQECTxKl6l3kZJGXl8dVV11Fx44deeONN4rMd7lc9OnTh8qVK/PTTz+xZ88eBg0ahDGGSZMmhaDEIiIiUt7tWbqWdTO+Zv/Kzbhy8rBHhJPQrDYNru9BpTMblum+e/XqxdSpUykoKGDlypUMHTqU/fv38/bbb5fpfuXISh2ys7OzmT17Nps3b6Zy5cpceumlVK5cuSzLVsT48eMBmDZtWsD5X3/9NStXrmTr1q1Uq1YNgIkTJzJ48GAeeeQR3VEsIiIiQbVn6VqWTZhJzu40ompUxhHtpCAzl92L/+bAph20fmhgmQZtp9NJcnIyADVq1KBfv37F5iQ5sUoVsrdt20aXLl3YuHGjb8CZ0aNH88UXX9ChQ4cyLeDR+OWXX2jRooUvYAP07NmT3NxclixZQvfu3QOul5ubS25uru95enp6mZdVRERETm3G7WbdjK/J2Z1GfNNaWJYFQHhcFGFNa5G2agvrZs6lYqv6WLayb6G7YcMGvvzyS8LCwsp8X3JkpXrFH3zwQVJSUnjwwQeZM2cOzz33HOHh4YwYMaKsy3dUUlNTSUpK8ptWoUIFwsPDSU1NLXa9xx57jPj4eN+jZs2aZV1UEREROcWlr0th/8rNRNWo7AvYXpZlEVWjMvtXbCJ9XUqZleGzzz4jJiaGyMhI6tevz8qVK/m///u/MtuflF6pQvbcuXO5//77GT9+PL179+b2229nypQp/Pnnn+zYseO4CjBu3Dhfo/3iHosXLy719g6/yAGMMQGne913332kpaX5Hlu3bj2mYxEREZHTR15aJq6cPBzRzoDzHVFOXDl55KVlllkZunfvzrJly1i4cCG33XYbPXv25Lbbbiuz/Unplaq5SGpqKl26dPGb1q1bN4wx7Nixo0jt8dEYOXIk/fv3L3GZOnXqlGpbycnJLFy40G/avn37yM/PL7GMTqcTpzPwH4iIiIhIIOHx0dgjwinIzCU8LqrI/IKsXOwR4YTHR5dZGaKjo2nQoAEAL7zwAt27d2f8+PE8/PDDZbZPKZ1ShWyXy0VkZKTftIiICAAKCgqOqwCJiYkkJiYe1za8OnbsyCOPPML27dupWrUq4LkZ0ul00rZt26DsQ0RERAQgrkF1EprVZvfivwk7pE02eP6LnvXPLhLPakxcg+onrExjx46ld+/ejBgxwu8eNTnxSt27yJo1a3A4Di7ucrkAWL16dZFl27RpE4SiFbVlyxb27t3Lli1bcLlcLFu2DIAGDRoQExNDjx49aNasGQMHDuSpp55i7969jB49muHDh6tnEREREQkqy2ajwfU9OLBpB2mrtnh6F4lyUpCVS9Y/u4ioHE+DgReekJsevbp160bz5s159NFHefHFF0/YfqUoy3i7CymBzWYrVVtn73NvAA+2wYMHM3369CLT582bR7du3QBPEL/lllv47rvviIyMZMCAATz99NNH1RwkPT2d+Ph40tLSFM5FREROcWX9uR6wn+zmdWgw8MIy7b5v8ODB7N+/n9mzZ/tNf+uttxgyZAjr1q1TZw4hVKqQHSjYlmTQoEHHXKCTgUK2iIhI+XEiPteN2036uhTy0jIJj48mrkH1E1qDLSefUoXs041CtoiISPmhz3UJhaB9xXK73Xz66adcccUVwdqkiIiIiMgpqdQ3Phbn77//ZsqUKcyYMYPU1FSioop2YSMiIiIicjo5pprsrKwspk2bRpcuXWjatClPPfUUlSpV4oUXXiAlpexGNRIRERERORUcVU32r7/+ypQpU5g1axYZGRnExsYyYMAA3nrrLf773/8WGbBGREREROR0VKqa7GeeeYbmzZvTuXNnJk+eTKtWrZg6dSrbt29n0qRJ6N5JEREREZGDSlWTPXr0aCzLok+fPjz77LPUr1/fNy8/P7/MCiciIiIicioqVU1269atMcYwZ84cBgwYwKuvvkp6enpZl01ERERE5JRUqpD9+++/s3TpUkaMGMG6desYMWIEVatW5frrr2f+/PllXEQRERERkVNLqXsXadWqFS+++CLbtm3jf//7Hx06dODNN9/kyiuvxLIsPvjgA1JTU8uyrCIiIiJShjZt2oRlWSxbtizURTnlHXUXfk6nkwEDBvDtt9+yfv167r//fqpXr86kSZOoXbs2ffv2LYtyioiIiJzEDJAFpBf+PPk7hRg8eDCXX355qItRbh3XiI916tTh4YcfZvPmzcyZM4dLLrmEOXPmBKtsIiIiIqeADGAdsBZYX/hzXeF0OV0FZVh1y7Lo3bs377//Pv/8808wNikiIiJyCsgANgEH8HTaFln480Dh9LIL2t26dWPkyJGMHDmShIQEKlWqxIMPPogxhgkTJtCyZcsi67Rt25aHHnqIcePGMX36dD7++GMsy8KyLL/77DZs2ED37t2JioqiVatW/PLLL37b+eCDD2jevDlOp5M6deowceJEv/l16tTh0UcfZejQocTGxlKrVi1ee+21MjkPJ6ughOxDVapUKdibFBERETkJGSAVKOBguLY4GLYLCueXXdOR6dOn43A4WLhwIS+88ALPPvsskydPZujQoaxcuZJFixb5lv3zzz9ZunQpgwcPZvTo0Vx99dX06tWL7du3s337djp16uRb9oEHHmD06NEsW7aMRo0acc0111BQUADAkiVLuPrqq+nfvz/Lly9n3LhxjBkzhmnTpvmVbeLEibRr146lS5dyyy23MGLECFavXl1m5+JkY5lSjCRjt9tLv0HL8r0Ip6r09HTi4+NJS0sjLi4u1MURERGR41B2n+tZeJqGOAg89EhB4aMhEBXE/Xp069aNnTt3smLFCizLAuDee+/lk08+YeXKlVx00UXUqVOHl156CYA777yTZcuWMW/ePMDTJnv//v3Mnj3bt81NmzZRt25dJk+ezLBhwwBYuXIlzZs3Z9WqVTRp0oRrr72WXbt28fXXX/vWu+eee5gzZw4rVqwAPDXZ5557LjNnzgTAGENycjLjx4/n5ptvDvq5OBmVajAaYwzx8fFceumlCp0iIiIigCdAu4HiKiPtQF7hcmWjQ4cOvoAN0LFjRyZOnIjL5WL48OEMHTqUZ555BrvdzptvvlmkWUdxzjjjDN/vVatWBWDnzp00adKEVatWcdlll/kt37lzZ5577jlcLpevcvbQbViWRXJyMjt37jzmYz3VlCpkDx48mPfff58PPviAK6+8kmHDhtG1a9eyLpuIiIjIScyBp+Wti8CRylU4v1RxK+guueQSnE4nH330EU6nk9zcXP71r3+Vat2wsDDf794Q73a7AU/l66HB3jutpG14t+PdxumgVG2yp0yZwvbt23n++ed9DeEbNGjAo48+yrZt28q6jCIiIiInoUg8zUDyKNru2hROjypcrmz8+uuvRZ43bNgQu92Ow+Fg0KBBTJ06lalTp9K/f3+iog42WwkPD8flch31Pps1a8ZPP/3kN23BggU0atToqJoYl3elvvExOjqaYcOG8dNPP7Fq1SquvPJKX9/Yffr04dtvvy3LcoqIiIicZCwgGU9NdTaeZiGm8Gd24fTkwuXKxtatWxk1ahRr1qzh7bffZtKkSfz73//2zb/hhhv47rvv+OKLLxg6dKjfunXq1OHPP/9kzZo17N69m/z8/FLt86677uLbb7/l4Ycf5u+//2b69Om8+OKLjB49OqjHdqo7pt5FGjduzJNPPsk///zDmDFj+Oqrr/jvf/8b7LKJiIiInORigTpADAfDdUHh8zqF88vO9ddfT3Z2Nu3bt+fWW2/ltttu48Ybb/TNb9iwIZ06daJx48acffbZfusOHz6cxo0b065dOypXrszPP/9cqn22adOGd999l3feeYcWLVrw0EMPMWHCBAYPHhzMQzvlHVMjobS0NN566y2mTJnCkiVLaNy4sUZ6FBERkdNULJ5Q7Q3Y3i78yq4G2yssLIznnnuOl19+OeB8Yww7duzgpptuKjKvcuXKfj2EHLrOoRISEopM+9e//lVi++5NmzYVmXa6DdV+VCH722+/ZcqUKcyePRu73c5VV13F888/79evooiIiMjpx6Isuuk7Hjt37mTmzJmkpKQwZMiQUBfntFOqkD1hwgSmTZvG5s2b6dSpEy+++CJXX3010dHRZV0+ERERETkGSUlJJCYm8tprr1GhQoVQF+e0U6rBaGw2G3FxcfTr14/GjRuXvEHL4s477wxaAUNBg9GIiIiUH/pcl1Aodcgu9QYt65i6gzmZ6I9RRESk/NDnuoRCqZqLbNy4sazLISIiIiJSbpQqZNeuXbusyyEiIiIiUm4cUz/ZIiIiIiJSvFLVZJ933nnFzgsLCyMpKYkLL7yQa665BofjmLreFhEREREpN0p142OLFi2wrMAdqhcUFLB9+3bS09Np37493333HVFRJ1c/kUdLN0iIiIiUH/pcl1AoVXORv/76i+XLlwd8rFq1iv379zNnzhxWrlzJI488UtZlFhEREZEysGnTJizLOilGZ5w2bRoJCQlluo9u3bpxxx13lMm2g9Ymu3fv3vzf//0fH3zwQbA2KSIiInJKcLsNqSnpbFi7m9SUdNzuIzYUCLnBgwdz+eWXh7oY5VZQG1C3b9+e//znP8HcpIiIiMhJbfOGvfw0bz3btqaRl+ciPNxOtZrxnNO9PrXrVQx18SREgtq7SHZ2tm58FBERkdPG5g17mT3rTzau20t0jJPkarFExzjZuM4zffOGvWWy327dujFy5EhGjhxJQkIClSpV4sEHH8R7q92ECRNo2bJlkfXatm3LQw89xLhx45g+fToff/wxlmVhWRbz58/3Lbdhwwa6d+9OVFQUrVq14pdffvHbzgcffEDz5s1xOp3UqVOHiRMn+s2vU6cOjz76KEOHDiU2NpZatWrx2muvHdOxzp49m0aNGhEREcGFF17I1q1bffPWr1/PZZddRlJSEjExMZx11ll88803fuu/9NJLNGzYkIiICJKSkujbt6/ffLfbzT333EPFihVJTk5m3Lhxx1TOwwU1ZH/++ec0a9YsmJsUEREROSm53Yaf5q0nIz2XajXiiIwKw2azERkVRrUacWSk5/LTvPVl1nRk+vTpOBwOFi5cyAsvvMCzzz7L5MmTARg6dCgrV65k0aJFvuX//PNPli5dyuDBgxk9ejRXX301vXr1Yvv27Wzfvp1OnTr5ln3ggQcYPXo0y5Yto1GjRlxzzTUUFBQAsGTJEq6++mr69+/P8uXLGTduHGPGjGHatGl+5Zs4cSLt2rVj6dKl3HLLLYwYMYLVq1cf1TFmZWXxyCOPMH36dH7++WfS09Pp37+/b/6BAwe46KKL+Oabb1i6dCk9e/bkkksuYcuWLQAsXryY22+/nQkTJrBmzRq+/PJLunTpUuQ8RkdHs3DhQp588kkmTJjA3Llzj6qcgZSqd5G9e4v/FuZyuUhJSWHWrFk89dRTvPTSS9x4443HXbBQ0l3IIiIi5UdZfa6npqQz8/XfiI5xEhkVVmR+dlY+mQdyGTi8PcnVg5snunXrxs6dO1mxYoWvB7h7772XTz75hJUrVwJw0UUXUadOHV566SUA7rzzTpYtW8a8efMAT5vs/fv3M3v2bN92N23aRN26dZk8eTLDhg0DYOXKlTRv3pxVq1bRpEkTrr32Wnbt2sXXX3/tW++ee+5hzpw5rFixAvDUZJ977rnMnDkTAGMMycnJjB8/nptvvrlUxzht2jSGDBnCr7/+ytlnnw3A6tWradq0KQsXLqR9+/YB12vevDkjRoxg5MiRfPjhhwwZMoR//vmH2NjYgOfR5XLx448/+qa1b9+e8847j8cff7xU5SxOqWqyExMTqVy5csBHcnIybdu25cknn+TWW2895QO2iIiISGlkZeWRl+fCGWEPOD/caScvz0VWVl6Z7L9Dhw5+XSx37NiRtWvX4nK5ABg+fDhvv/02OTk55Ofn8+abbzJ06NBSbfuMM87w/V61alUAdu7cCcCqVavo3Lmz3/KdO3f22/fh27Asi+TkZN82SsvhcNCuXTvf8yZNmpCQkMCqVasAyMzM5J577qFZs2YkJCQQExPD6tWrfTXZF154IbVr16ZevXoMHDiQN998k6ysrGKP1Xu8R1vOgGUvzUIPPfRQsf1kOxwOkpKSOO+886hXr95xF0hERETkVBAVFU54uJ3cHBeRUUXrLfNyPTdBRkWFh6B0cMkll+B0Ovnoo49wOp3k5ubyr3/9q1TrhoUdrJn3ZkC32w14aqUPz4WBGkYcug3vdrzbOBqBMqh32t13381XX33F008/TYMGDYiMjKRv377k5Xm+2MTGxvL7778zf/58vv76a1979EWLFvm6BwxWOQ9XqpAdrAbgIiIiIuVFlaqxVKsZz8Z1e6kWGecXBo0x7N2TRd0GFalStWgzhWD49ddfizxv2LAhdrunZt3hcDBo0CCmTp2K0+mkf//+fgMGhoeH+9U8l1azZs346aef/KYtWLCARo0a+fYdLAUFBSxevNjXNGTNmjXs37+fJk2aAPDjjz8yePBgrrjiCsDTRnvTpk1+23A4HFxwwQVccMEFjB07loSEBL777juuvPLKoJb1cOoKREREROQY2GwW53Svz+6dmWz7J52KlaI8TURyXezdk0VsnJNzutfHZgvcGuB4bd26lVGjRnHTTTfx+++/M2nSpCK9fNxwww00bdoUgJ9//tlvXp06dfjqq69Ys2YNlSpVIj4+vlT7veuuuzjrrLN4+OGH6devH7/88gsvvviir+13MIWFhXHbbbfxwgsvEBYWxsiRI+nQoYMvdDdo0IAPP/yQSy65BMuyGDNmjF8t9GeffcaGDRvo0qULFSpU4PPPP8ftdtO4ceOgl/VwQe1dREREROR0UrteRS7vdwZ1G1Qk80AuO7ZnkHkgl7oNPNPLsp/s66+/nuzsbNq3b8+tt97KbbfdVuTeuIYNG9KpUycaN27su3nQa/jw4TRu3Jh27dpRuXLlIiG8OG3atOHdd9/lnXfeoUWLFjz00ENMmDCBwYMHH1X5u3XrdsR1oqKi+L//+z8GDBhAx44diYyM5J133vHNf/bZZ6lQoQKdOnXikksuoWfPnrRp08Y3PyEhgQ8//JDzzjuPpk2b8sorr/D222/TvHnzoyrrsShV7yKnG/UuIiIiUn6ciM91t9uwc3sGWVl5REWFU6VqbJnVYIMnoLZu3ZrnnnuuxOWMMTRp0oSbbrqJUaNGlVl5jkWdOnUYN27cUYfzU4Wai4iIiIgcJ5vNCno3fcdr586dzJw5k5SUFIYMGRLq4vhZvXo1sbGxXH/99aEuSplRyBYREREph5KSkkhMTOS1116jQoUKoS6OnyZNmrB8+fJQF6NMKWSLiIiInGIOHQK9OGoRHFq68VFEREREJMgUskVEREREgkwhW0REREQkyBSyRURERESCTCFbRERERCTIFLJFRERERIJMIVtEREREJMgUskVEREREgkwhW0REREQkyBSyRURERESCTCFbRERERCTIFLJFRERERIJMIVtEREREJMgUskVEREREgkwhW0REREQkyBSyRURERESCTCFbRERERCTIFLJFRERERIJMIVtEREREJMgUskVEREREgkwhW0REREQkyE6pkP3II4/QqVMnoqKiSEhICLiMZVlFHq+88sqJLaiIiIiInNYcoS7A0cjLy+Oqq66iY8eOvPHGG8UuN3XqVHr16uV7Hh8ffyKKJyIiIiICnGIhe/z48QBMmzatxOUSEhJITk4+ASUSERERESnqlGouUlojR44kMTGRs846i1deeQW32x3qIomIiIjIaeSUqskujYcffpjzzz+fyMhIvv32W+666y52797Ngw8+WOw6ubm55Obm+p6np6efiKKKiIiISDkV8prscePGBbxZ8dDH4sWLS729Bx98kI4dO9K6dWvuuusuJkyYwFNPPVXiOo899hjx8fG+R82aNY/3sERERETkNGYZY0woC7B79252795d4jJ16tQhIiLC93zatGnccccd7N+//4jb//nnnznnnHNITU0lKSkp4DKBarJr1qxJWloacXFxpTsQEREROSmlp6cTHx+vz3U5oULeXCQxMZHExMQy2/7SpUuJiIgotss/AKfTidPpLLMyiIiIiMjpJeQh+2hs2bKFvXv3smXLFlwuF8uWLQOgQYMGxMTE8Omnn5KamkrHjh2JjIxk3rx5PPDAA9x4440K0SIiIiJywpxSIfuhhx5i+vTpvudnnnkmAPPmzaNbt26EhYXx0ksvMWrUKNxuN/Xq1WPChAnceuutoSqyiIiIiJyGQt4m+2SktlsiIiLlhz7XJRRC3ruIiIiIiEh5o5AtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBppAtIiIiIhJkCtkiIiIiIkGmkC0iIiIiEmQK2SIiIiIiQaaQLSIiIiISZArZIiIiIiJBdsqE7E2bNjFs2DDq1q1LZGQk9evXZ+zYseTl5fktt2XLFi655BKio6NJTEzk9ttvL7KMiIiIiEhZcoS6AKW1evVq3G43r776Kg0aNOCvv/5i+PDhZGZm8vTTTwPgcrno06cPlStX5qeffmLPnj0MGjQIYwyTJk0K8RGIiIiIyOnCMsaYUBfiWD311FO8/PLLbNiwAYAvvviCiy++mK1bt1KtWjUA3nnnHQYPHszOnTuJi4sr1XbT09OJj48nLS2t1OuIiIjIyUmf6xIKp0xzkUDS0tKoWLGi7/kvv/xCixYtfAEboGfPnuTm5rJkyZJit5Obm0t6errfQ0RERETkWJ2yIXv9+vVMmjSJm2++2TctNTWVpKQkv+UqVKhAeHg4qampxW7rscceIz4+3veoWbNmmZVbRERERMq/kIfscePGYVlWiY/Fixf7rbNt2zZ69erFVVddxQ033OA3z7KsIvswxgSc7nXfffeRlpbme2zdujU4ByciIiIip6WQ3/g4cuRI+vfvX+IyderU8f2+bds2unfvTseOHXnttdf8lktOTmbhwoV+0/bt20d+fn6RGu5DOZ1OnE7n0RdeRERERCSAkIfsxMREEhMTS7VsSkoK3bt3p23btkydOhWbzb8ivmPHjjzyyCNs376dqlWrAvD111/jdDpp27Zt0MsuIiIiIhLIKdO7yLZt2+jatSu1atVixowZ2O1237zk5GTA04Vf69atSUpK4qmnnmLv3r0MHjyYyy+//Ki68NNdyCIiIuWHPtclFEJek11aX3/9NevWrWPdunXUqFHDb573e4LdbmfOnDnccsstdO7cmcjISAYMGODrR1tERERE5EQ4ZWqyTyR94xURESk/9LkuoRDy3kVERERERMobhWwRERERkSBTyBYRERERCTKFbBERERGRIDtlehcREQmuAiC78HcnEB7CsoiISHmjkC0ip5k8IBXYgydoA9iBCkBVICJE5RIRkfJEIVtETiO5wFogE0+w9tZeFwA7gXSgIRAVktKJiEj5oTbZInIa2QocAMIKH7bCRzgQCeQAmwANHyAiIsdHNdkichoweGqqdwAuIB+w8LwFOgt/WoW/ZwIZgAasEBGRY6eabBEp5wyeNtgb8DQLsQofBk/77AOFP8HThMSNp9mIiIjIsVPIFpFyLg1I4WC4th3y0164TDaeGm4v94ksoIiIlEMK2SJSzu3GE6DDOFiDfSgbnlCdd8g8decnIiLHR22yRaQcy8dTkx3GwXbXpvDnoazCZe2Fy1U4gWUUEZHySDXZIlKOuTgYqr03NoKn5vrwGm03njbbiYcsJyIicmwUskWkHHNwsDkIeMLzoUH70AdAZaDmiSygiIiUUwrZIlKOOYAEPDXU3hrtCCAGT9i2c/AGyOpAXfS2KCIiwaA22SJSzlUB9uEZ7dHJwf6xHXiCdw6eGx2rU7SttoiIyLFRlY2IlHMxQB08tdXZeMJ2Hp5wnY0nYNdD7bBFRCSYVJMtIqeBSniGTd8D7MVzQ6QTz02OlVDAFhGRYFPIFpHTRFThowaeZiL6R56IiJQdhWwROc14u/MTEREpO6rKEREREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTL1LiIiJ4kCYD+ewWIsPIPIxKKeQERE5FSkkC0iIWaAHUAqnpEYvdNsePq1ro0ncIuIiJw61FxEREIsFdiKpybbiWdkxiggDDgArAMyQ1Y6ERGRY6GQLSIhlAtsx9MkxIn/W5IdT+DOBVKOsB1TuFw2nrAuIiISWmouIiIhtAfIxxOmA7Hw1Gin4wnQhy9ngL3ALjy13d5mJhWAKkB08IssIiJSCqrJFpEQyuLIw5w7ABeekH0og6eZyQY8IdwqXNYAO4E1eG6kFBEROfEUskXklOTK24G7YDvGbcNTwx2Gp4lJeOHzAmAjnmYkIiIiJ5aai4hICEXhaTJiKL42uwDPW5WnqcjuxWvY/NFPJHUJI65hPDk7cohMqkBkcgUcURGF61hABJ7a771A1bI8CBERkSJUky0iJ4DB02Z6V+HjQOG0SnhqoPNKWC8fiAMiWTdzLgtGPMeu35YSUyuG/LQ83PkFHNicyt4/1pO778Ah61p43uL2ltExiYiIFE812SJSxg4A/xT+dBVOs+Hp+7oGnlrmf/A06wjj4Hd/V+E0J1Cd1B/+ZMVz72PZbMQ3qY49IgxXjsEeEQ4YCrJySVu9hUpnNiicBp6g7d2niIjIiaOQLSJl6ACwFk9NdXjhw42ndnofntrtBkAtPF35edtPe3sJiQVqkb5+P0semEz62hRPgLblUZDVCFuYhTsfwMIR6aQgK5fsnfuJqVWlcDtuPMFdRETkxFJzEREpIwbYjCdgR+J5u8nGE7yz8QTtTGAVnvbTLYB6QHWgJtAYaMqmD35nfr8J7F7yN8btxp2fT9qqHez4aTM2JxiX27M7y8KyWeTs3F+4/8LpVCr7QxURETmMQraIlJEMPF30OfEE7gN4aqoNnl5AHHjegvLwdLd3AEjEE7KrAXGk/ricPx9/m4KsHGxhDhyRTuzOcBxRTrZ8so78jDwcMTYwxrNLmw13fn7hPnLwhPuKJ+6QRURECilki0gZ8Q4OY8dTc+0q/P3Qtx0bnnbT+cAmDh2t0RjD+v99Q0FmNlHVE7FsNozb7Zu/Z8kOVr24BFdOPpHJETgrOgmPDyOqWgwHB66pj1rFiYhIKOjTR0TKSGHtsq8Ndknf6R14arT3AZUBOLBxO3t+X0tE5QTsEU7CE2LI3Z2GLezg29Y/n29g7x+p1OnbnOo9GoBxYdyRQG0O9lwiIiJy4qkmW0TKiLfP6jyK7wfbG8S9IzVm+Obk7EnHlZuHPdKJZUFUtUpYdhuu3Hy/LWSs28+qSYv54fo5LL53MfaI1kAyCtgiIhJKCtkiUkbi8bTHLihhGW8vIg48Idz45jginVh2Oybfs35EpThi61UFAwWZObjy8nHnu3DlF5CflUtkUgXaPjKMqKq60VFEREJPIVtEyogdz02MNjzh2X3IvEOfR3AwYDt9S8Q3rklsnSRydqf5pkXXqEyFM+oSmVwRy7Jwu11YNouafTpwzpR7qHx207I9JBERkVJSyBaRMlQJqIsncLvx3PzoKvzdwnNzYjie2m47h/YEYgtzUKdvV9wFLvLSM33TnQkxJDStReLZTYmuWonEsxpz9rO3EFsn+UQdlIiIyBHpxkcRKUMWUKXw5zo84drbfV9Y4fQCPDdGVgGi/Nau2687+1duYsvHC8jdk0FEYhyWw05BZg65ezOIqJxAm/FDiKxS4QQek4iIyJEpZIvIMXMXuNj160r2Ld+Iu8BFdI3KVD2vNeHxMYctWRlPoN6CJ1C78e8zuxKePrKz8DQf8fyTzeawc+b4IVRs1YBNH/xAxvptuF1uHJHh1L7yXOoPOJ8KLeqemIMVERE5CpYxxhx5sdNLeno68fHxpKWlERcXF+riiJyU9ixdyx+PvEn63//gLijA23tIRFIFGg3tTf3rLsCyDu9RJA/Yi6cXEW/AdhU+9zYhceIJ3FUK53u4XS4yN+/AlZtPRGI8EZUTyvT4RKT80Oe6hIJqskXkqO1dvoGFo14me8deoqsl4ojy3LDoLnCRnbqPv56ehTu/gEZDex+2Zjie7vWS8YzwuB7PyIze5iMUPt8CpAMN8AZtm91ObL1qZXxkIiIiwaEbH0XkqBhjWPXibLK37yGuXjVfwAZP847oGonYwh38/cbnZG3fc8iaBcAeYAewk4MB23vzo73wEVH4fA+e4dZ3FS4nIiJy6lDIFpGjkrZ6C3uW/E1kUgUsW6ABZiAyKY68vftI+XIusBvYCvyFJ1hvLvyZhqfJyKEt1gyettqZeEL5bmADsKJwnbwyOSYREZFgU3MRETkqGRu2U5CZQ2TVigHmuoFsLFsBWAWk/f03UKNwehieWms7nqYihoOjQUYXrp+DJ2SDpw7ADTgwBkzBDgpy9mEK6uGsEGjfIiIiJw+FbBE5OpZF4Lul3XhqoF14bmC0iK5dgYODzuRzsDmIVfiw4amxzsXzdnRowAZjIGfnfrK2p1OQnUNU1Wi2zP6JnN1x1OvfXT2LiIjISUvNRUTkqMQ1qEZYTCT5aZmHzcnFE7BtGBdgIKlz9cJ5djyh+uAyHt7mJnl4arGNb5oxhvR1e0hbvY28/QfADe5cF0nnVOOfLxbw803PsH3e0jI7ThERkeOhkC0iRyWuYQ0qt29M9o49GHcWnr6ts/AEaE8NdXZqBs5K0cQ3rsTBIO0dXj2fgwPRmEOm5XLoqJDZO9LJSknHcjhwREVgC3NQkOXGWSmSKp3qkZ+exdLxM8jatvtEHr6IiEipKGSLyFGxrFyajjyDqBpRZKzfRv6BDIzJAVy48vI5sGUvbrebJiPaYw+HosOpZ+OptfY2FXEftgeDMW6ytqUBFrawg63avN362+xuYmonkJ26i38+X1DGRywiInL0FLJF5CjkAetIaBpNh+evoFLb2uTuyyVj/T7S1+0jc+t+IpMiaf1AZ+oNaMzBmmo3/mE6H0/ADqwgK5+CA/nYwv3fohzRdlw5ueTs2oNly8HuNKR8Mx9Pv9rFb09ERORE042PInIUduO5uTGCCi1q0GXGtez5/R/2Ld+GuyCT6BoxJHetXdh3tnc0x8Nrqimcl3/Yc/DeNGlcNjAUdhFYGJ4tQ1iMk9TvU8jPKADsniYkGTnAdjw15A04dJRIERGRUFHIFpFScuMJ2Ta8/wSzbDYS29UisV0tPOHbO2jMoTXYhws03TsgTRaQh81pBxsYlwvLBrZwOxGJUeTsyGT7vE2F69tw5eQTkRSDJ4jvKFy/BlABvb2JiEgo6VNIREqpAE/tc3FvG4c27Th8kJnDeUO296ZIb7d/UYCFw5lDfKOK5GfmYwu3Q4HhwKY01r/1Fzk7MgGDO78A43ZR86J6h2wvA8/gNdFAfTzhXURE5MRTyBaRUvIG4kDh2UXR0RhdAdb39ihi52Aw9vYukocnqDuBPPIPFJDy1Xpc2QVk78wkfc1ujNuzb3e+i4yN+4lvUplqPRpwsOcSg6fnkkw8YbsxepsTEZFQ0KePiJSSA0/NcDoHB5KxAS6M+wCYArAMGHC73Fg2O5YdLAs8odpb0+3CG6QPtssuwBOMvUHcTVz9RPZV28nyJ34id182jqgwbGF2CrLzMQVu4pskctZT5xMed3httQ3PgDeZwD6gchmdDxERkeIpZItIKRhgPwf7s87DE4YduAoKoCAPV24BjugwAFzZBbjz8wiPd4LDhmV52lAfHGzGE6QPOtjO2xPCDeCi9uVNSGieyNY5a9k+dwMF2fnE1q9ArUsaUb1XfZwVIjnY1OTQbXu3tReFbBERCQXLeDueFZ/09HTi4+NJS0sjLi4u1MURCTGDp/eOFA72eZ0PWBhjMG4XrtwCMBZ2px1jDAUHCgCDLcyGPdKBZbNhWXYONhUB/55HbPi3z+aQ596fgW6iPJwn+HtqysFzM+UZR3vAIlLO6HNdQkH9ZIvIEaQD2/C8XUQBMXiajdgwLjeWZeFwhoEbslKzcGUVYI+wY9kt3Pku3LkHa6Y9Qdl7A6X7kG1GcLA2Oxz/0SAP72O7OBYH23pnc7CNt4iIyImnTyAROYJdeAJyeOFzC4jEmFhydmbjznPhLnCTl55L9rYMDmzeT97+HCybhT3CAcZg3Ibi/2eWh6f2Oa7wEYN/SzbrkEdJ7BxsKuK9mTLq6A9XREQkCE6ZkL1p0yaGDRtG3bp1iYyMpH79+owdO5a8PP8eDSzLKvJ45ZVXQlRqkVNdAZ6a7LCiswzk7M7C7TK4C9yExYYBBleOm8ytGaSv28+BTelk78nGuAzufMPBmmpb4Ta9Nc9Z+NdWm2J+L443YB/LuiIiIsF3ytz4uHr1atxuN6+++ioNGjTgr7/+Yvjw4WRmZvL000/7LTt16lR69erlex4fH3+iiytSTrjxBNUA38cti4LMAlzZBThiwsBtgc3yZWVXdj55+7Kp2KoypsDFvuWphCdEElW9ImEx3vbZniYe2TvS2PHj3+TtL8AR46BK50RiasZzMCR7f3qCdGZKGjt/3kp+Rh5hcU6Szq1LVHIsB5uYWBwc3EZEROTEO2VCdq9evfyCc7169VizZg0vv/xykZCdkJBAcnLyiS6iSDnkwFNLXMDhbxeWZRFeIZaslAPENa6AZT/YD7Y7z0V+Zi7R1WOxR4aRlZKBK89N1rZ0sndlktiuJo4IGwXZblY+/wtbPl1N3r4cz76Mm7D4MKqdX5+W93QjPN57U6RFfkYey5/+iZSv15K/P9fbPyDhCQupeXEjmt/RAUdUNJ4bH71twUVERE68U6a5SCBpaWlUrFixyPSRI0eSmJjIWWedxSuvvILbXfJNU7m5uaSnp/s9RAQ8bxEVCRxYDVFVK5CfVUB+eh65+3OISooiMjmKiCqRxNZNAJtF3v4csndnY3c6cESHg9uQsX4PBdkF/D7mG9ZOX4pxGWLrJhDXoBKx9Stg2W1sen8Vv436ivxMz94KsnNZdPeXbJz1F5ZlEVu/AnENKhJbNxGMjXUzl7P4vh9x5TnxfCFwoxEfRUQkVE6ZmuzDrV+/nkmTJjFx4kS/6Q8//DDnn38+kZGRfPvtt9x1113s3r2bBx98sNhtPfbYY4wfP76siyxyiqoM7AFy8PQC4r2pMJewOBeV2yWTtz+b9TP+xBhwVowk/0AumVvS2b9qN2c/04OwmHBcWQUA2CPCydmZxd5lO0j5ah1RyTGExXp7FHFj2WxEVIrGERXBzl82s2X2eupf24qtn64m9cctRNeIwxEVhrc7QMtuI6JyLPbIcLZ/u5ZtX6+h5sVNCste6USfLBEREeAk6Cd73LhxRwy4ixYtol27dr7n27Zto2vXrnTt2pXJkyeXuO7EiROZMGECaWlpxS6Tm5tLbm6u73l6ejo1a9ZUf5oiPmnARjxB29v9Hnj6yrYoyHSRt/8AO37ewsJ/f0l+ei7h8ZFEVY+n7WPdiG9ciezUA4XrOHDnFbD86Z9J+3sv8Q0qFm4vDE87ageettQFHNiURlzjKnR9exA/Dvwf+/7aTmzdQ2vWvd3+eZqNpK/fTZUOteg8+XIsqwLQkFP8H3YiEgTqJ1tCIeQ12SNHjqR///4lLlOnTh3f79u2baN79+507NiR11577Yjb79ChA+np6ezYsYOkpKSAyzidTpxOZ8B5IgIQDzQBVuPpcs87THoYlhVOWIwNR3QstS6LImdXJn+//gfRNStg2SxSv99KbL0EwitGYArAER2OO99F5j9phMd6/+7C8HTdZxVu33MDY3iFMDK37Cbrn+24cnMJT4jgYP/aEYXLunylDI8PZ//qHbhyY3FE1EMBW0REQiXkITsxMZHExMRSLZuSkkL37t1p27YtU6dOxWY78gfo0qVLiYiIICEh4ThLKnK6c+G5ATKWQG8dlhWO3RlDjd4NCm+KjMSdW8D+VXvY+UsKtS9vhC3cDgbc+S7P6JC++yWi8ATsbDzDtsOhNdS2sAJa3d+J1a/+Ts6OXDwB21vz7R3cxoBVWBNu6hOw20EREZETJOQhu7S2bdtGt27dqFWrFk8//TS7du3yzfP2JPLpp5+SmppKx44diYyMZN68eTzwwAPceOONqqkWOW5p+A9Kczg3luUipnYCdfo2xZ3vCdDVLqyPIzqMgmwX7vR8wqLDwBhi61Vg79JUqGzwBGt74U/vgDKQl5ZPTO0KRFRJBLdFoyGtWTnpbwoyCwr3aeEJ02GFy6dTqU1d7BH6excRkdA6ZUL2119/zbp161i3bh01atTwm+dtVh4WFsZLL73EqFGjcLvd1KtXjwkTJnDrrbeGosgi5Yy3WUagkRcNnnbU+Vg2G3kZBWT9k4Ejykl0rWjCYsLIP5BP5j+ZZGTsBwtq9mnFnqU7KcgqwBF1MFh7f7pyC3Dnuqh9ZWvsYTFgyyYyuYAKLeLZtXBPkRIUZOeCMdS54lws60ijQ4qIiJStkN/4eDLSDRIigWwDtuLpFu/wEJsHZOLtJzsvw2LfnylgXFQ4ozIANodF5j8ZZO/IIa5+NcIrxvHbqI/Y/u1aIipHEF4hAssKwxjIT88hOzWDSu1q0umlqwiPj8Rd4CI7dRs7Fmxl9YsrCYuPxrIsjDHk7T9Azq40krucwdnPj8QRqZpsETlIn+sSCqdMTbaIhFoCnqDtouhbR94hv9sIj42lYusI8vbuw2a3cOW5MDYbUdXjiUyqSnh8DABnPXEpfzw6l+3frSZj3V6w7Bi3ISzGSbUejWk9pifh8Z6+rm0OO5HJFYlvmE9B1lKyd6Zh2fAsHxdFrUs70eqBaxWwRUTkpKCa7AD0jVckEAOsx9NnthNPG2rv9HQOduvnxHMjI0ABxmSAscCysCyD58ZJO4fK2PAPqd+vI2+/G0dMJEnn1CW+SVKAZh/ZGJNA+t+R7PjxT/LSswiPjya5ayviGlQP/iGLSLmgz3UJBdVki0gpWUBtPL15eEdFdeAJ2e7Cn078R1m0YVk2sEr6Lu+5CTK2XkdKHqXRE+ItqyLxjROJb1zzmI9ERESkrKkTWRE5CmF4BnipA0QXTrM42KWetys+Lxue3ki8Qdw7eMyh8gvXrY5/P9mHcuMZCCcKqBCUIxERESlLqskWkaNkB6rgGW7dhScYZwBr8dRyH94/tRNPkC7Av5mIm4NtuasDyYW/p+DpL9uOd6h1zyMKqM/hTU1ERERORgrZInKMLA6+hcTjCd07OThEug1PjbQ3XHtvSMwpnO6tAU8uXNcCquJps70b2M/B5iOVgYpogBkRETlVKGSLSBBYQC08TUN24hlUxhukHXjCc3U8Nd8ZHAzicRStmY4pfHibjKjPaxEROfUoZItIkNiAakASntEhvTXYsRwcJdLBwRrtI1G4FhGRU5dCtogEmR1P0w4REZHTl3oXEREREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyR6gLcDIyxgCQnp4e4pKIiIjI8fJ+nns/30VOBIXsADIyMgCoWbNmiEsiIiIiwZKRkUF8fHyoiyGnCcvoa10Rbrebbdu2ERsbi2VZoS5OmUtPT6dmzZps3bqVuLi4UBfnlKHzdvR0zo6eztmx0Xk7euX5nBljyMjIoFq1athsaikrJ4ZqsgOw2WzUqFEj1MU44eLi4srdG+uJoPN29HTOjp7O2bHReTt65fWcqQZbTjR9nRMRERERCTKFbBERERGRIFPIFpxOJ2PHjsXpdIa6KKcUnbejp3N29HTOjo3O29HTORMJLt34KCIiIiISZKrJFhEREREJMoVsEREREZEgU8gWEREREQkyhWzhpZdeom7dukRERNC2bVt+/PHHUBfppDVu3Dgsy/J7JCcnh7pYJ50ffviBSy65hGrVqmFZFrNnz/abb4xh3LhxVKtWjcjISLp168aKFStCU9iTxJHO2eDBg4tcex06dAhNYU8Sjz32GGeddRaxsbFUqVKFyy+/nDVr1vgto2vNX2nOma41keBQyD7NzZo1izvuuIMHHniApUuXcu6559K7d2+2bNkS6qKdtJo3b8727dt9j+XLl4e6SCedzMxMWrVqxYsvvhhw/pNPPskzzzzDiy++yKJFi0hOTubCCy8kIyPjBJf05HGkcwbQq1cvv2vv888/P4ElPPl8//333Hrrrfz666/MnTuXgoICevToQWZmpm8ZXWv+SnPOQNeaSFAYOa21b9/e3HzzzX7TmjRpYu69994QlejkNnbsWNOqVatQF+OUApiPPvrI99ztdpvk5GTz+OOP+6bl5OSY+Ph488orr4SghCefw8+ZMcYMGjTIXHbZZSEpz6li586dBjDff/+9MUbXWmkcfs6M0bUmEiyqyT6N5eXlsWTJEnr06OE3vUePHixYsCBEpTr5rV27lmrVqlG3bl369+/Phg0bQl2kU8rGjRtJTU31u+6cTiddu3bVdXcE8+fPp0qVKjRq1Ijhw4ezc+fOUBfppJKWlgZAxYoVAV1rpXH4OfPStSZy/BSyT2O7d+/G5XKRlJTkNz0pKYnU1NQQlerkdvbZZzNjxgy++uorXn/9dVJTU+nUqRN79uwJddFOGd5rS9fd0enduzdvvvkm3333HRMnTmTRokWcd9555ObmhrpoJwVjDKNGjeKcc86hRYsWgK61Iwl0zkDXmkiwOEJdAAk9y7L8nhtjikwTj969e/t+b9myJR07dqR+/fpMnz6dUaNGhbBkpx5dd0enX79+vt9btGhBu3btqF27NnPmzOHKK68MYclODiNHjuTPP//kp59+KjJP11pgxZ0zXWsiwaGa7NNYYmIidru9SI3Ozp07i9T8SGDR0dG0bNmStWvXhroopwxvbyy67o5P1apVqV27tq494LbbbuOTTz5h3rx51KhRwzdd11rxijtngehaEzk2CtmnsfDwcNq2bcvcuXP9ps+dO5dOnTqFqFSnltzcXFatWkXVqlVDXZRTRt26dUlOTva77vLy8vj+++913R2FPXv2sHXr1tP62jPGMHLkSD788EO+++476tat6zdf11pRRzpngehaEzk2ai5ymhs1ahQDBw6kXbt2dOzYkddee40tW7Zw8803h7poJ6XRo0dzySWXUKtWLXbu3Ml//vMf0tPTGTRoUKiLdlI5cOAA69at8z3fuHEjy5Yto2LFitSqVYs77riDRx99lIYNG9KwYUMeffRRoqKiGDBgQAhLHVolnbOKFSsybtw4/vWvf1G1alU2bdrE/fffT2JiIldccUUISx1at956K2+99RYff/wxsbGxvhrr+Ph4IiMjsSxL19phjnTODhw4oGtNJFhC2LOJnCT++9//mtq1a5vw8HDTpk0bv66cxF+/fv1M1apVTVhYmKlWrZq58sorzYoVK0JdrJPOvHnzDFDkMWjQIGOMp2u1sWPHmuTkZON0Ok2XLl3M8uXLQ1voECvpnGVlZZkePXqYypUrm7CwMFOrVi0zaNAgs2XLllAXO6QCnS/ATJ061beMrjV/RzpnutZEgscyxpgTGepFRERERMo7tckWEREREQkyhWwRERERkSBTyBYRERERCTKFbBERERGRIFPIFhEREREJMoVsEREREZEgU8gWEREREQkyhWwRERERkSBTyBaRkBs3bhyWZbF79+6A81u0aEG3bt18zzdt2oRlWViWxbhx4wKuM3ToUN8yxWnTpg2WZfH0008HnD9t2jTfNizLwuFwUKNGDYYMGUJKSsoRj2vGjBn079+fxo0bY7PZqFOnzhHXERGR8kEhW0ROWbGxsUybNg232+03/cCBA7z33nvExcUVu+6yZctYunQpAG+88UaJ+5k6dSq//PILc+fOZfjw4bz99tuce+65ZGZmlrjezJkzWbFiBe3bt6d+/fqlPCoRESkPFLJF5JTVr18/Nm/ezLfffus3fdasWbhcLi699NJi1508eTIAffr0YfXq1SxYsKDYZVu0aEGHDh3o3r07Y8eO5Z577mHjxo3Mnj27xPJ99dVXLF++nJkzZ9K4cePSH5iIiJzyFLJF5JTVuHFjOnXqxJQpU/ymT5kyhSuvvJL4+PiA6+Xk5PDWW2/Rtm1bnn32Wd86pdWhQwcANm/eXOJyNpveYkVETlf6BBCRU9qwYcOYPXs2+/btA2DNmjUsWLCAYcOGFbvOhx9+yL59+xg6dCgNGzbknHPOYdasWRw4cKBU+1y3bh0AlStXPv4DEBGRckkhW0ROaVdffTUOh4O33noL8LSvrlu3rt+Nkod74403iIiIYMCAAYAnqB84cIB333034PIul4uCggIOHDjAnDlz+M9//kNsbGyJzVFEROT0ppAtIqe0mJgYrrrqKqZMmUJBQQEzZsxgyJAhxfYqsnHjRubNm8eVV15JQkICAFdddRWxsbHFNhnp0KEDYWFhxMbGcvHFF5OcnMwXX3xBUlJSWR2WiIic4hyhLoCIiMPheStyuVwB5xcUFBAWFlbs+sOGDeOcc87hkUceYdeuXQwePLjYZadMmYIxhr59+7J//37f9EsvvZQ333yT1atX06RJE791ZsyYQdOmTXE4HCQlJVG1atXSH5yIiJyWFLJFJOS8NcIpKSlFaoeNMWzfvp127doVu37nzp1p3LgxEyZM4MILL6RmzZoBl3O73UybNg2AK6+8MuAyU6ZM4cknn/Sb1rRp0xL3LyIicjiFbBEJufPOOw/Lspg1axZt2rTxm/fll1+Snp7OBRdcUOI2HnzwQd5//31uvfXWYpf56quv+Oeff7j11lvp27dvkfkj/7+9O0ZNJArgOPxfCMJcIW26qT2FJFhNE8baQvEAHkACqdLYWSRlsBNrDyBazhkCSZHK1u1sdrO7rAMJ4fu6YR5vZqr58eDNjMd5enrKbDY7ra4DwP/wFgE+3dXVVcbjce7v7/P+/p5er5eiKLLdbnN3d5dut3vapPiRuq5T1/UfxywWi1xcXGQ6neby8vKX88PhMJPJJOv1Ov1+/6xnSpKmadI0TZLk5eUlh8Mhy+UySVKWZcqyPPsaAHxNNj4CX8LDw0Pm83n2+31ub29zc3OTx8fHjEajbDabdDqds+Z/e3vLarXK9fX1bwM7SQaDQYqi+OsfIP/V8/NzqqpKVVXZ7XZ5fX09HX/0JRMAvocfx+Px+Nk3AQAA34mVbAAAaJnIBgCAlolsAABomcgGAICWiWwAAGiZyAYAgJaJbAAAaJnIBgCAlolsAABomcgGAICWiWwAAGiZyAYAgJb9BI0LdP/AoTpOAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABKwAAAVKCAYAAADEzeNiAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdeZyNdf/H8feZfcastrGNQUmkECkkVJREcldK1tAiJSp3ShktFDWUohXRvt53ixbZWvCzlqQ9DJqxM9YZc87394f7XM0xi5lxca4z5/V8PK4H5zrX9bk+13Wuc13nfOb7/R6XMcYIAAAAAAAAcIgQfycAAAAAAAAA5EfBCgAAAAAAAI5CwQoAAAAAAACOQsEKAAAAAAAAjkLBCgAAAAAAAI5CwQoAAAAAAACOQsEKAAAAAAAAjkLBCgAAAAAAAI5CwQoAAAAAAACOQsEKCBAzZ86Uy+VSnTp1/J2Kj7S0NLlcLrVv397fqZQL7du3l8vlUlpamr9TKdTChQvlcrnkcrnKtH5x53Ewn0t16tSRy+XSzJkzT/m2izvuTnhN/HlsADt988036tKli6pUqaLQ0FC5XC51797d32mVCyd6b7JD//795XK51L9/f7/lAADlDQUroAy8X+KOnaKiolSrVi1169ZN77zzjowx/k4Vx/AWhEoyBWPhBGV35MgRTZ8+XVdccYVq1qypyMhIJSQk6IwzzlCHDh00atQoffbZZzp8+LC/Uy030tLSlJaWpg0bNvhl+/m/JC9cuPC4y3uLb077Qjt58mSlpaXp+++/93cq5dbSpUt18cUXa86cOdq5c6cqVqyo5ORkJSUllWj9/Oda/ikmJkbVqlVTkyZN1LdvX02bNk27du06yXsTXBYuXKi0tDSK5gDgB2H+TgAIdMnJydb/9+7dqy1btmjLli36+OOPNXPmTH344YeKjIw84e0kJCSoQYMGqlmz5gnHslPlypXVoEED1a5d29+plEp4eLgqVqxY7DLHex72CtRzSZI2bdqkK664QmvXrrXmRUREKDQ0VH/++ad+//13LVy4UI8//rgWLFhQoBh62mmnKSoqSgkJCac4c+cf9+KOzdixYyUdLUQ7rfVpIJk8ebI2btyoOnXqqGnTpv5Op1yaPHmy8vLy1KZNG3300UcndH9JSkpSRESEJCkvL087d+7U1q1btWbNGs2ePVvDhw/XzTffrPHjx6tChQp27YKjxcTEqEGDBicl9sKFCzV27Fi1a9eu2GJz9erV1aBBA1WvXv2k5AEAwYiCFXCCsrKyrP97PB79/PPPGj58uObOnavPPvtMo0eP1sSJE094O1dffbWuvvrqE45jt6FDh2ro0KH+TqPUWrduXaIWETh1AvVccrvduuqqq7R27VrFxMRo1KhR6tevn2rVqiWXy6WcnBytWbNGc+bM0ezZswuNMW/evFOc9T+cftz9eWwAu/z444+SpOuvv/6E/xjywQcfFCh6b9iwQd9++62mTZumxYsXa8qUKZo/f76++eabErfiCmQtW7bUL7/84tccxo8fr/Hjx/s1BwAob+gSCNgoJCREZ511lj766COdfvrpkqQXXnhBeXl5fs4MwMkyf/58rV69WpL0yiuvaPTo0UpJSbHGUomMjNR5552nMWPG6Pfff1erVq38mS4APzh48KAkKTY29qTEr1Onjnr37q3vvvtO6enpkqSffvpJPXv2PCnbAwDgVKBgBZwEUVFRuvbaayVJ+/bts/7qt3fvXr311lu68cYbdfbZZ6tixYqKiopSamqqevXqpaVLlxYZszSDVb///vvq1KmTqlatqpCQEKWlpWnFihVyuVwKCwvT3r17C8S45ZZbrDExPvnkkwLPv/nmm3K5XAW6DR1vUOYvvvhCPXr0UK1atRQREaH4+HjVq1dPnTp10pNPPlnkWBuHDx/WM888o3bt2qly5cqKiIhQtWrV1L17d33++edFHqeTacOGDdYx2rBhgzZu3KjBgwerdu3aioqK0mmnnabRo0frwIED1jpr165V7969lZKSoqioKNWvX1+PPvqojhw5ctzt5ebm6vHHH9c555yjChUqKCkpSR07dtRnn3123HX//PNP3XHHHWrYsKFiY2MVExOjhg0b6q677lJGRkax6/7yyy+68cYbVa1aNUVFRalevXq64447tHXr1uMfJB0dq6V79+6qXLmyoqOj1aBBAz3wwAPav39/sesVdy4dO5jte++9p/bt26tixYqKiYlR06ZN9fTTT8vj8RQZ3xijGTNmqFWrVoqLi1NCQoLOP/98vfjiizLGlHnA3Pzj/lx11VXFLutyuQrtIlzcwOL5x0jauXOnRowYodNOO03R0dFKTU3V0KFDtX37dmv5jRs36rbbblPdunUVFRWl2rVr6+6779a+ffsKzamsA6sfOnRIH330kQYPHqymTZuqSpUqioyMVI0aNdS9e/diz9Njr2cLFixQ9+7dVb16dYWGhvq8BoUdG+9r5dWhQwefcX28ce+77z65XC6dddZZxe5Ldna2YmNjHTG4+969e/XYY4/p/PPPV1JSkiIjI5WSkqIbbrih2HvEr7/+qokTJ+rSSy+1zo/4+Hg1a9ZMo0eP1o4dOwqs433tN27cKEkaMGBAgTGSvE7F9e9E7pHHnsfvvPOO2rVrp4oVK6pChQpq3ry5nn32Wbnd7mKP//GsXr1affv2VWpqqqKiopSUlKTWrVtr8uTJysnJKbB8/mMmFTzGJ2P8teHDh+v222+XJM2dO7fYVooLFy7UDTfcYL2OCQkJatmypSZMmODzWh6rrPf2AwcOKD093bq3R0ZGqlatWmrXrp2eeuqpAveZ/D9EcuTIET311FNq0aKFEhMTfcaOK27Q9WOvN3PnzlXnzp1VpUoVRUdH66yzztKjjz5aYHxB7znv7Xq8aNGiAu+Pwq5Lxd1DFi5cqGuvvdYa57By5cq65JJLNGPGjCLPzWPP7Xnz5lmD90dFRalhw4YaO3Ys4yMCKJ8MgFIbM2aMkWSKews999xz1jLfffddgfUkmdjYWBMZGWk9drlc5umnny403owZM4wkk5qaWmQ+7dq1MyNGjLBiJSUlmdDQUDNmzBjjdrtNYmKikWT++9//Fohx+umnW3kMHz68wPODBg0ykkzfvn2L3Paxxo4d67O/MTExJjY21mfeggULCqz322+/mfr16/scl4SEBJ/1brvttkKP0/G0a9euyHyPZ/369db233//fet4xsfHm9DQUOu5tm3bmtzcXPPJJ5+YmJgYI8kkJCQYl8tlLdOzZ89i8xs1apRp27atkWTCwsKsbXmnMWPGFJnniy++aMLDw61lIyMjTXR0tPU4Pj7efPnll4Wu+9lnn/mck7GxsSYqKspIMtWrVzfTp08v9tx/5ZVXTEhIiLVMQkKCiYiIMJLMmWeeadLT00t0Hh+rX79+RpLp16+fuf32240kExISUuC4HHt+euXl5ZmePXv6nFNJSUlWrjfccIPPNkpjwoQJVtzffvutVOt6paamGklmxowZBZ7zxn711VdNrVq1jCRToUIF67hKMg0bNjS7d+82y5YtM5UrV7Ze57CwMGuZNm3amLy8vALxizvuxT3nvSZ5p+joaOt890533313ofub/3r29NNPW++NhIQEEx4e7vMaFHZs7rzzTpOcnGxtJykpySQnJ1tTixYtjDHG/PXXX1bsb775psjjP23aNGv7Bw4cKHK5Yy1YsKDYa9mxvPtS1Dm2dOlSn/0KDQ01cXFxPuftuHHjio3tXS4xMdHnmlOzZk3zyy+/+KwzceJEk5ycbL0P4uPjfY5jcnKyteypuP6dyD0y/7k6cuTIQt/nksxll11mDh8+fNzXqjCTJk3y2Q/v+ep9fM4555i///7bZx3vcSzqGGdkZJRo26U91/7++28rtz59+hR4/siRI9Z9Pf/xzv9aNmjQwGzYsKHAumW9t69cudKkpKRYy4SEhJikpCSfYzpp0iSfdbz3xH//+9+mdevW1j0xKSnJZzv5j8+x8l9vnn32WWt7iYmJPtfIZs2amV27dlnrZWRkmOTkZFOhQgUjyYSHhxd4f7z11lvW8se7hwwfPrzAezT/8b744otNdnZ2gfXyn9sTJkwwLper0Pd4hw4dCr3GA0Ago2AFlEFJClb33nuvtczPP/9sjDn6pWj48OFm6dKlZvfu3cYYYzwej/nrr7/MsGHDjMvlMqGhoWbVqlUF4pWkYOX9wDhy5Eizbds2Y4wxhw8ftj5wduvWzUgyw4YN81l/06ZN1gdpSaZJkyYFtnHaaacV+oW6qC+0GzZssD6gjxgxwmzZssV6bs+ePeabb74xQ4YMMStWrPBZb/fu3aZOnTrWh7evv/7a+nKxZ88ek56ebu3n5MmTC+R5PHYVrBITE80ll1xifvrpJ2OMMQcPHjTPPPOM9eFz9OjRJiEhwfTs2dM6/vv27TMPPPCAFWPu3LlF5peQkGAiIyPN888/bw4dOmSMOfrh+ZprrrHWL6zw+OGHH1ofrO+77z6zYcMG4/F4jMfjMb/88ou59tprrdd648aNPutu2rTJOgfOOecc83//93/GGGPcbrf57LPPTK1atXwKRMdauXKl9eG/ffv21nmfm5tr3nzzTZOYmGitX9aCVVJSkomIiDDp6elm7969xhhjduzY4fPFa968eQXWHz9+vPX8iBEjzI4dO4wxxuzdu9eMGzfO+mJb3JeNoixcuNDnC8fmzZtLtb4xJStYJSYmmqZNm5qlS5caY/45rt6iwNChQ01qaqq5+OKLzdq1a40xxhw6dMhMmTLFOi9feumlAvHLWrD68MMPzc0332wWLFhgHU9jjn5RHjt2rPVlubDz1Hs9i4qKMqGhoaZ///7WF/e8vDzzxx9/lOrYFPcF/vLLLzdS0cVMY4w599xzrWNYGnYWrNavX2+9P6655hqzcuVKc+TIEWOMMVu3bjUPPvig9f768MMPC6zfs2dPM2XKFPPHH3+YnJwcY4wxOTk55quvvjItW7Y0ksy5555bbF6FHeP8+Z3s69+J3CO956r3jxtDhw617oN79+41jzzyiPXlvrA/yhzPxx9/bOV+1VVXmb/++ssYc/QYz5o1yyostm7dutCiQUmOcXFKe64ZY6zXvXbt2gWeGzZsmJFkkpOTzdSpU83OnTuNMUevKwsWLDDNmjWzzhm3222tV9Z7e0ZGhlVMT0lJMW+99ZZVHD58+LD58ccfTVpamnnttdd81vPeE2NjY01sbKyZMWOGOXjwoDHm6LXfm3dJClYxMTEmPDzcXHvttdb15tChQ+b555+3CqNXX311gfWLuw7mV1zBasqUKVZ+N998s8nMzDTGGLN//34zadIk671dWDHXu/3ExEQTEhJiRo0aZbZv326MOXpuP/TQQ1bsV155pdgcASDQULACyuB4Bau9e/eaGjVqGEmmYsWKPh/2iuNtOTJw4MACz5WkYOX9AFmUyZMnG0nm7LPP9pn/6quvWtutWbOmcblcPl9AMzIyrPjH/rW1qA9yb7/9tpFkzjjjjBLs+T/uuece64u/98vasT744AMjyVSuXLnIZYri/fBb2F9Ki/vLqTG+X9jOOuusQv9K36dPH2uZjh07Go/HU2AZb8upwl5nb35FffB0u93moosuMpJMo0aNfJ7LyckxNWvWPO6H1qIKl7fddpuRZCpVqmS2bt1aYL0ff/zRpzXBsTp37my95t4vFPl9/vnn1rplLVgV94WvefPmRpIZNGiQz/wDBw5YhbjCjnn+bZelYGWMMR07drTWDw0NNa1atTJ33XWXmT17dolaXZWkKJOcnOzzvvR68MEHS3xeXnLJJQWeK2vB6ngmTpxY5Dbzt87q0aNHsXFOtGD1n//8x0hHW4B5iyD5rVy50oqzZs2a4+2Wj/xfko9t5VXY5P2iX9g55i1GF9YaxsvbQrGwPyoUZ9++fVbLrcJampW2YHWyrn/HU9w9Mv97uKhjOHr0aCMdbaGTv9BSEo0aNTKSzIUXXlhoQeqjjz6ytv/uu+8WeN4fBavBgwdb6+S/V/7444/G5XKZmJiYIs/57Oxsq0Vn/gJpWe/tvXv3tu4vJW1VZozvPfGjjz4qcrmSFKy817LCPpO9/PLL1jLLli3zee5EC1YHDx40FStWNNLR1ryFeeaZZ6ztL1++vNDtS0W3ru7Ro4eRZC699NJicwSAQMMYVoCN9uzZo3nz5uniiy/W33//LUkaNmyYQkJK9lbr0qWLJOnbb78t0/ZDQkL073//u8jnO3ToIOnouCL5x7xZsGCBJOniiy9Whw4dZIyx5klHB5WWpLp16yo1NbVEuSQmJko6OoZXceNg5GeM0fTp0yVJd999t8LCCv8h0+7duys+Pl47duzQypUrSxT7WEeOHNHWrVuLnQ4dOlTk+sOHDy90LKLLLrvM+r93/JyillmzZk2R8VNSUjRgwIAC80NCQjR69GhJ0rp166xfnpKkzz77TFu2bFFycnKh63r17dtX0tExSLyMMXr77bclSbfeequqVq1aYL3GjRvrmmuuKTTmnj17rHj33nuvoqOjCyxz2WWXnfCA4ykpKVb+x+rWrZukgsf1iy++UHZ2tiTpgQceKHTdu+++WzExMWXO68MPP9SQIUMUHh4ut9utJUuWaPLkyerTp4/OOOMM1alTR2PHjrXyKIvBgwerUqVKBebnP+dGjBhR7HlZ3DlnN+/1bMmSJcWOGzRq1KiTmseVV16plJQUHTp0qNBfaXzxxRclSa1atdLZZ59d5u3s3r37uNeUosZY27Vrlz744ANJR68bRfGe+z/88EOJx5STjg703a5dO0llv7/kd7Kvf0Up6T3yoYceKnS+99qUl5en999/v8TbXbNmjdatWydJevDBBxUaGlpgma5du6ply5aSjo756AT5f40w/5hSr7zyiowx6tKlS5HnfFxcnLp37y7J915Rlnv7gQMHrPvLfffdp5SUlNLshiTprLPOUteuXUu93rFGjx5d6GeyAQMGqFatWpKkt95664S3k9/cuXOt45+WllboMkOGDFH16tUlFX3+REZG6p577in0Oe/4iafyGg8ApwIFK+AE5R98MykpSZdeeqlVROndu3eBL8h//fWX7rnnHjVv3lyJiYkKDQ211r/iiiskSZs3by5TLqeffnqhhQavs88+W5UrVy5QkDq2YCX9U6TK/3/vcyXRsmVLVa5cWZmZmTr//PP17LPP6pdffpExpsh11q1bZ32o69+/v6pVq1boVL16dWsAb+9gwaXVrl07maOtTIucihs41fvF5FjJycnW/88777xil9m9e3eR8b0DzRbmoosusop5K1assOZ7v8Tt3r1b1atXL/L4DR48WJLvsVu/fr117C+++OIi8yrquVWrVllfxsuyfkmdd955RRaAa9SoIUkFBvtdtWqVJKl27dqqW7duoevGxcWpefPmZc6rQoUKeu6557R582a9+OKL6tOnjxo2bGh9sd24caPS0tLUtGlT/fnnn2Xaxsk+58pi69atGjNmjFq1aqVKlSopLCzMup41atRI0tFfRytqu9HR0Tr33HNtzelYoaGhGjRokCTppZde8nnuwIED1pfDm2+++YS2s2DBguNeU4oq+C9ZssTn/VPUezf/4PGFXfs++eQT9ezZU/Xq1VOFChV87k/vvPOOpLLfX/I7mefiid4jU1JSrF/pPVZ8fLz1Ps9/7Twe77JhYWFW4a8wHTt2LHXsk6mo+633XvHZZ58Vea5Vq1ZNM2bMkOR7rpXl3r5ixQproP2yFp3atGlTpvXyCwsLU9u2bQt9LiQkxBrU3O7XzxsvJSVFZ5xxRqHLhIaGWvfHorZ/1llnFfkrk0Xd/wAg0BXefAFAieX/gO79xZdmzZrpxhtvLFDg+fDDD3XDDTf4/JJQfHy8oqKi5HK5lJubq927d5f4r5bHKq5YJcn6lZn33ntP8+fP13XXXae//vpLGzduVMOGDVWtWjXrA1P+gpW3oFWaglViYqLefPNN9erVSz/99JPuuOMOSVJCQoIuuugiXXfdderZs6fCw8Otdbyt0iT5tAArjvenwk+1uLi4QufnbxV2vGWK+6XAmjVrFvlcZGSkKlWqpK1bt2rbtm3WfO/xy83NLVHri/wtyPLHKW7b3r9AH+tE1y+poo6pVPRx9Z5L3g/0RSku75KqWrWqBg8ebBUF9+/fr/nz52vixIn69ttvtX79el1//fVavnx5qWPbcc7l5eWVertFWbJkia644grt2bPHmuf9RUqXyyW32239Mt2BAwdUuXLlAjEqVapU4haoJ2LQoEF65JFH9OOPP2rp0qW64IILJB1tSZGdna3ExERdd911Jz2PouS/9pW05VT+a5/H41Hv3r19WmaEhYUpKSlJERERko7+At/hw4fLfH/J72Rd/+y4Rx7vfex9Pv8163i8y3p/1a4o3utbaWKfTPmLgvlbW3nPt/379x/311sl33OtLPf2rKws6/8lbaV9rON9vimJ471+ZTk3SsIb73jn5vHOn5Lc/+y8xgOAE9DCCjhBWVlZ1rRx40atXLlSL7/8coHizs6dO9W/f3/l5OTo4osv1sKFC3Xw4EHt3btXW7duVVZWlt59990TyqWwbgrHOrYFlfdfb6GqTp06qlu3rn799Vf9/fff+uOPP7Rp0yafdUvq0ksv1fr16zVr1iz169dP9evX1969e/Xxxx+rT58+atasmbZs2WItn7/bUFZW1nFbKxyvFVQgK6p1VXG8x+/yyy8v0bEr6i/iZdm2k3n383j7VVwLgbKKjY1Vt27dtGjRIuv9s2LFCn3//fe2b+tUysvL0w033KA9e/aoadOmmjNnjrKzs7Vv3z7rerZ06VJr+aKObUmuWXaoUaOG1WXU2wVQ+qfFVe/evU+oS+iJ8r53o6OjS/ze9bYGkY528XrzzTcVGhqqhx56SL///rtycnK0a9cu6/7k7c57Ms5zO9h1jzyZ16+SxnbKNfSHH36QdLRIlL+Y6D3fHn/88RKdawsXLvSJW9p7ux3suFb4+3UJtPMHAJyAghVwini/0CUlJenjjz9Wu3btCozzk/+vkCeLtzD1+++/a/PmzQUKVpJvUcv7/BlnnFGmFigVKlRQnz59NHPmTP3222/avHmznnjiCUVFRfn8dVaSqlWrZv0//9hMwai4Li85OTnauXOnJN+/OnuPX1mOXf44xW27qC8h+dcv7ouK3V9iSsKbW/5WLIU53vMnIiQkxOqWJkm//vrrSdvWqbBkyRJt3LhRoaGh+uSTT9S5c+cCf/0/Fdez0rj11lslSW+//bays7P1448/6v/+7/8knXh3wBPlfe8eOnRIf/zxR6nX9465M2jQII0dO1ann356gZZrTns9jmXXPfJ4XR6916DStNjxLrt9+3af1l9FbbtKlSoljn2yZGZmavXq1ZLkU9yUTuxe4VWae7t3bCap7N347XC8168s50ZJeON5//hXFCedPwDgFBSsgFPE+0GlQYMGRf4l/6uvvjrpeZx55pnWh8f58+dr4cKFPmM3SPLpFliW7oDFqVmzpkaOHKm7775b0tHBSL0aN26s+Ph4SfYPehpoFi1aVGRLiG+++cZq9t+iRQtrvneMjy1btpR6YOW6detaXUbyj292rPxdRfM799xzrS/IZVn/ZPKOkbRx40Zt2LCh0GX2799f5gH8Syr/2CPFdUsJBN7rWZUqVYosZJ+K65m3JUJJWg1deumlOv3003Xw4EG9/vrrVuuqEx1s3Q6tW7e29qUs1z7v69GsWbNCn9+/f79VnCuM973rz9ZXdt0jN23aVOQ4cfv27bPe5/mvncfjXTYvL0+LFi0qcjlvfkWN33UqPfbYY1a3y2NbInvvFZ9++mmJugSWRHH39hYtWlhdUz/++GNbtlcWeXl5Rd4bjTH6+uuvJRU8N070/eGNt3nzZv3222+FLuN2u617pxPOHwBwCgpWwCmSkJAgSfrtt990+PDhAs9///33euONN05JLt7i09SpU5WZmakmTZr4jG+Rv4WVtytAaQtWxf0VU5L1l/P8zfzDwsJ00003SZJeffXV4xZdyvPgohkZGXr11VcLzPd4PBo3bpwkqWHDhj5ftLt27WoVI4cNG3bc8b3yHz+Xy2WN4fP8889bYw/lt27dOr333nuFxkpMTFSnTp0kSU8++WSh5/hXX32lxYsXF5vTydCpUyerEOo9dseaNGlSmcdDW7t2bYlajs2aNcv6f1GFhUDhvZ55f/3uWJs3b9Yzzzxz0vPwvq75x9Eqisvl0i233CLp6LXvtddek+T/1lXS0RYY3l/5mjhxYpFfar2OvfZ5Xw9vF7BjPfLII9q3b1+R8UpzHE8WO++RjzzySKHzn3rqKR06dEhhYWHq0aNHiXM755xzrB8RePTRRwv91cs5c+ZYRcEbbrihxLFPhsmTJ+u5556TdLSL+LEtrAYPHiyXy6U9e/bo3nvvLTbWkSNHfIpaZbm3x8TE6Prrr5d0tBvi8VoanUyPPfZYob/W+eqrryojI0OS1LNnT5/nTvT90bFjR+sXXov6lcAXXnjBauXr7/MHAJyEghVwinTq1EkhISHatWuXbrzxRusLbm5urt555x116tSp2AE17eQtPnk/XB/7y201atRQgwYNtHHjRqsLxrEfeI/niSeeUOfOnTV79myfLho5OTl65513NHHiREmyfvXJ68EHH9Rpp52mvLw8XX755UpPT/cZgH3v3r36/PPP1a9fvyJ/7ac8SEhI0G233aaXXnrJ+vK2adMm3XDDDdZfYR977DGfdaKiojR16lS5XC6tWrVKbdq00RdffKHc3FxrmfXr1+uFF15Qy5YtNXXqVJ/1R40apbi4OO3YsUMdO3a0fqnIGKMvv/xSnTt3Lnacn0ceeUShoaH65Zdf1KVLF6vbW15ent555x1dd9111k+in0oVKlTQv//9b0lHxywaOXKk9YV/3759euKJJ5SWlqakpKQyxV+4cKHq1aunnj176t1331VmZqb13OHDh/Xtt9+qW7duev/99yVJ11xzTZkHHnaKCy+8UBUqVJAxRtddd51VYHG73friiy+K/ZVLOzVu3FiS9Prrr5eo4DhgwABFRkZq7dq12r17txITEwt8OfWXp556SpUqVVJ2drYuvPBCTZ8+XXv37rWe37Fjhz744AP16NGjwBfayy+/XNLR8/vFF1+03vNZWVkaPny4JkyYYH1hLoz3OL733nu2/5JkSdl1j0xISNCrr76qYcOGWYX3ffv2ady4cVYh6/bbby91F/cnnnhC0tEWrtdcc43Wr18v6WhB5/XXX7dek9atW6t79+6lim2HjIwMvf7667rwwgs1fPhwSUd/GTj/QPxeTZs21V133SXp6B8orr32Wn3//fdWCyK3260ffvhBjzzyiE477TSfMffKem9/7LHHVLlyZe3cuVNt2rTRO++8Y/3wR05OjtasWaN7771Xs2fPtu2YHCsmJkbffvutevXqZeV++PBhvfTSS7rtttskSVdddVWBX8H0vj9++umnMv3RJTo62ipUvfnmm7r11lutQv/Bgwc1ZcoU6/Xo2bPnCf1iLQCUOwZAqY0ZM8ZIMqV9C/373/+21pNkEhISTHh4uJFk6tata15//fUi486YMcNIMqmpqUXm065duxLl8eeff/rk8emnnxZY5tZbb7Web9SoUZGxitp2/mMkyURHR5uKFSsal8tlzWvYsKHJzMwsEPOvv/4yTZo08Vk/MTHRxMfH+8w7/fTTS7S/+bVr185IMuHh4SY5Ofm4U37r16+3tr1+/fpC4y9YsOC450Zxr6U3v1GjRpkLL7zQyjUpKcln30ePHl1k/Ndee83ExMRYy4aFhZlKlSqZyMhInxiPPvpogXU/+eQTn+Xi4uJMdHS0kWSqV69upk+fXuz+vfDCCz6vcUJCghXvzDPPNOnp6WU6j/v162ckmX79+hW538Ud1yNHjphrrrnGyiskJMQkJSWZ0NBQI8n06dPH9O3b10gyt9xyS5HbKMzzzz/vc1wlmaioqAKvmSTTqVMns3fv3gIxUlNTjSQzY8aMAs95112wYEGh2z/R87K4417cc9OmTfPZt9jYWBMVFWUkmcqVK5uPPvqoyLyKe62OVdyxmT17trWN8PBwU7NmTZOammratGlTZLzevXtb6wwdOvS42y9O/uNa1OuTn3dfijqPV61aZerUqWPFdLlcJikpycTGxvoc60svvdRnvd27d5szzzzT5/xOTEy03ou33HJLse+hRYsWWcuGhoaa6tWrm9TUVJ/X51Rc/07kHpn/XB05cqR1HCpWrGi9z73H7tChQ0XmV5z09HSf61tiYqKJiIiwHp999tlmy5Ytha5b3HlcEvmPbVJSknWPqlSpknWMvFNkZKQZNmyYOXDgQJHx8vLyzF133VXgulWpUiUTFhbmM//bb7+11juRe/vKlStNzZo1reVCQ0NNUlKSz7qTJk3yWcd7TxwzZkyJj8+x8p9zzz77rLW9pKQkn2PXpEkTs2PHjgLrHzlyxDRo0MDn+HvfH++++6613PHuU8OHDy/w3s5/rDt06GCys7MLrFeSz3glee8BQCCihRVwCj3++OOaNWuWWrZsqejoaB05ckSnn3667r//fq1evVo1atQ4JXnUq1fPauERFhZWaEulwgZhL42bb75ZL774om644QY1btxYMTEx1oC6bdu21eTJk7Vq1Sqfgda96tatqxUrVmjWrFm68sorVb16dR04cEC5ubmqW7eurr76ak2fPl1LliwpdV5eR44csbozFTf5S0REhObNm6dx48apQYMGysnJUUJCgi655BJ9+umnRXZ5kaQbb7xRf/zxh0aPHq0WLVooNjZWe/bsUVRUlJo2baqhQ4fqq6++slod5delSxetWrVK119/vapWrarc3FwlJydr6NChWr16terWrVts3jfffLO+++47de3aVRUrVlROTo5SU1M1atQoLVu2rMytmE5UWFiY3nnnHb388svW+y8vL08tWrTQyy+/rFmzZlndPUrbCuyWW27RDz/8oCeeeEJXXXWVTj/9dIWGhmrv3r2Ki4tTo0aN1LdvX82ZM0dffPGF1b0k0N1666369NNP1b59e8XGxiovL081a9bUHXfcoR9++OGUjAvVu3dvzZ49WxdeeKFiYmKUmZmpjRs3Fjvw9rXXXmv93wndAfNr1qyZ1q1bp2effVaXXnqpKleurH379snj8ah+/frq1auX3nrrLX3wwQc+6yUmJmrx4sW66667VKdOHYWGhiosLEzt27fXm2++qeeff77Y7V500UX69NNPdemllyohIUFbt27Vxo0bT/kA2XbdI5944gm99dZbatOmjTwejyIiItS0aVM9/fTT+vzzzxUVFVWm/IYPH64VK1aod+/eSklJ0cGDBxUdHa0LLrhA6enpWrZs2Sm5j+/evdu6Rx04cEBJSUk655xz1KdPH02bNk2ZmZmaPHlysS1iQ0NDNWnSJK1atUo333yzGjRoYF23kpKS1KZNG6Wlpen777+3xrySTuzefu655+rnn3/W448/rgsuuEBxcXE6cOCAatWqpfbt2ys9PV29evU6KcfM6/bbb9cXX3yhyy+/XCEhIQoJCdGZZ56phx9+WEuWLCm0JWJYWJjmzZunQYMGqU6dOjpw4ID1/ijNGGDp6emaP3++/vWvfyk5OVn79+9XXFycOnTooOnTp2vu3LmnrKU9AAQKlzEO/X1jAABOEWOMateurc2bN2vWrFnq06ePv1PCSXLHHXfo2WefVatWrfwyphpOjrS0NI0dO1bt2rWzxl4EJGnmzJkaMGCAUlNTi/zhDQCAM9HCCgAQ9LzjsYSFhemSSy7xdzo4SbKzs63B771j1gAAAMCZKFgBAILCDTfcoPfee8/nFxC3bt2qxx9/XIMHD5Yk9e3b95R1zcWplZOTo2HDhik7O1spKSmOGWwdAAAAhQvzdwIAAJwKn332md566y1JR38tKjw83OdX2Nq2batJkyb5Kz2cJJMnT9bkyZO1bds261fJ0tPTFRER4efMAAAAUBxaWAEAgsIzzzyj66+/Xg0aNFBkZKQOHjyoKlWqqGPHjnrllVc0b968cjMgOv6xZ88ebdy4UcYYNW3aVG+//bauueYaf6cFAACA42DQdQAAAAAAADgKLawAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAADjJZs6cKZfLpRUrVvg7lXJr69atuv/++9W0aVPFx8crIiJCtWrVUo8ePfTRRx/J7Xb7O0UAAACUQpi/EwAAADgRS5cuVbdu3WSM0W233aYLLrhAsbGxysjI0Mcff6wePXrohRde0MCBA/2dKgAAAEqIghUAAAhYe/bsUffu3RUbG6vvvvtO1atX93m+d+/eWrNmjXbu3FlsnEOHDikqKkoul+tkpgsAAIASoksgAAAOcPjwYd19991q2rSpEhISVLFiRbVq1Ur//e9/Cyzrcrk0dOhQzZ49Ww0bNlRMTIyaNGmiTz75pMCy//3vf3XOOecoMjJS9erV09NPP620tDSfwsyGDRvkcrk0c+bMQreVlpZmPf7jjz80YMAA1a9fXzExMapZs6a6du2qH3/8scC6P/30kzp16qSYmBhVqVJFt99+uz799FO5XC4tXLjQZ9mvvvpKl1xyieLj4xUTE6M2bdpo3rx5xz1uL730krZu3aoJEyYUKFZ5nXPOOerQoYP12NtF88svv9RNN92kKlWqKCYmRjk5OfJ4PJowYYLOPPNMRUZGqmrVqurbt682b97sE7NOnTrq379/gW21b99e7du3tx4vXLhQLpdLr732mkaMGKFq1aopOjpa7dq10+rVq33W/euvv3T99derRo0aioyMVHJysi655BJ9//33xz0OAAAA5Q0trAAAcICcnBzt2rVL99xzj2rWrKnc3Fx99dVX6tGjh2bMmKG+ffv6LP/pp59q+fLlevjhhxUbG6sJEybo6quv1q+//qp69epJkj7//HP16NFDF110kd5++23l5eXpySef1NatW8uc599//61KlSrp8ccfV5UqVbRr1y69+uqrOv/887V69Wo1aNBAkpSZmal27dqpQoUKmjZtmqpWrao333xTQ4cOLRDztddeU9++fXXVVVfp1VdfVXh4uF544QVddtll+uKLL3TJJZcUmc/cuXMVGhqqK664otT7ctNNN6lLly6aPXu2Dhw4oPDwcN1222168cUXNXToUF155ZXasGGDHnzwQS1cuFCrVq1S5cqVS70dSbr//vt17rnn6uWXX9bevXuVlpam9u3ba/Xq1dbrdcUVV8jtdmvChAmqXbu2duzYocWLF2vPnj1l2iYAAEAgo2AFAIADJCQkaMaMGdZjt9utSy65RLt379bkyZMLFKwOHTqkr776SnFxcZKkc889VzVq1NA777yj++67T5L00EMPqWbNmvriiy8UEREhSbr88stVp06dMud50UUX6aKLLvLJs0uXLjrrrLP0wgsvKD09XZI0adIk7dq1S19//bUaNWokSercubMuv/xybdiwwVr/4MGDGjZsmK688kp9+OGH1vwrrrhC5557ru6//3793//9X5H5bNq0yWohlZ/H45HH47Eeh4SEKCTEt2H5JZdcohdeeMF6/Msvv+jFF1/UkCFDNGXKFGt+s2bNdP7552vSpEl67LHHSnKYCqhSpYo+/PBDq2XbhRdeqPr162v8+PF66aWXtHPnTv3666+aPHmyevfuba3Xo0ePMm0PAAAg0NElEAAAh3j33XfVpk0bxcbGKiwsTOHh4XrllVf0888/F1i2Q4cOVrFKkpKTk1W1alVt3LhRknTgwAGtWLFC3bt3t4pVkhQbG6uuXbuWOce8vDyNGzdOjRo1UkREhMLCwhQREaHff//dJ89FixapcePGVrHK64YbbvB5vHjxYu3atUv9+vVTXl6eNXk8Hl1++eVavny5Dhw4UOo8R4wYofDwcGvq1q1bgWX+9a9/+TxesGCBJBXo6teyZUs1bNiwRF0Ui9KrVy+fbpipqalq3bq1tc2KFSvqtNNO08SJE5Wenq7Vq1f7FNwAAACCDQUrAAAc4IMPPtB1112nmjVr6rXXXtOSJUu0fPly3XTTTTp8+HCB5StVqlRgXmRkpA4dOiRJ2r17t4wxSk5OLrBcYfNKasSIEXrwwQfVvXt3ffzxx/q///s/LV++XE2aNLG2LUk7d+4s0ba93ROvueYanwJTeHi4nnjiCRljtGvXriLzqV27trZv366DBw/6zL/77ru1fPlyLV++vMixrY6d7x2YvbDla9SocdyB24tTrVq1Qud5Y7pcLs2bN0+XXXaZJkyYoHPPPVdVqlTRnXfeqX379pV5uwAAAIGKLoEAADjAa6+9prp16+rtt9/2aYmTk5NTpnhJSUlyuVyFjleVlZXl8zgqKqrQbRVWoPGONzVu3Dif+Tt27FBiYqL1uFKlSiXatndMqClTpuiCCy4odF+KK7B17NhRX375pebMmaNrrrnGmp+SkqKUlBRJ8mlhlt+xvwjoLQJmZmaqVq1aPs/9/fffPuNXRUVFFfra7Nixo9Bxro7db++8/IXH1NRUvfLKK5Kk3377Te+8847S0tKUm5ur559/vtB9AAAAKK9oYQUAgAO4XC5FRET4FFGysrIK/ZXAkqhQoYJatGih//znP8rNzbXm79+/v8CvCSYnJysqKkpr1qzxmV/ULxRGRkb6zPv000+1ZcsWn3nt2rXT2rVrtW7dOp/5b731ls/jNm3aKDExUevWrVOLFi0KnYoqOEnSoEGDlJycrJEjRyozM7PI5Uri4osvlnS0KJff8uXL9fPPP/sM/l6nTp0Cx+u3337Tr7/+WmjsN998U8YY6/HGjRu1ePFin18UzO+MM87Q6NGjdfbZZ2vVqlVl2R0AAICARgsrAABOkfnz5/sMOO51xRVX6Morr9QHH3ygIUOG6JprrtGmTZv0yCOPqHr16vr999/LtL2HH35YXbp00WWXXaZhw4bJ7XZr4sSJio2N9elm53K51Lt3b02fPl2nnXaamjRpomXLlumNN94oEPPKK6/UzJkzdeaZZ+qcc87RypUrNXHixAItku666y5Nnz5dnTt31sMPP6zk5GS98cYb+uWXXyTJGgA9NjZWU6ZMUb9+/bRr1y5dc801qlq1qrZv364ffvhB27dv17Rp04rcx8TERP3nP/9R165d1aRJE91222264IILFBsbq507d+rrr79WVlaWWrdufdzj1aBBA918882aMmWKQkJC1LlzZ+tXAlNSUjR8+HBr2T59+qh3794aMmSI/vWvf2njxo2aMGGCqlSpUmjsbdu26eqrr9bgwYO1d+9ejRkzRlFRURo1apQkac2aNRo6dKiuvfZa1a9fXxEREZo/f77WrFljDaIPAAAQVAwAADipZsyYYSQVOa1fv94YY8zjjz9u6tSpYyIjI03Dhg3NSy+9ZMaMGWOOvV1LMrfffnuB7aSmppp+/fr5zPvwww/N2WefbSIiIkzt2rXN448/bu68806TlJTks9zevXvNoEGDTHJysqlQoYLp2rWr2bBhg5FkxowZYy23e/duM3DgQFO1alUTExNjLrzwQvPNN9+Ydu3amXbt2vnEXLt2rbn00ktNVFSUqVixohk4cKB59dVXjSTzww8/+Cy7aNEi06VLF1OxYkUTHh5uatasabp06WLefffdEh3jrKwsM2rUKHPOOeeYChUqmPDwcFOjRg3TtWtXM2vWLHPkyJECr8fy5csLxHG73eaJJ54wZ5xxhgkPDzeVK1c2vXv3Nps2bfJZzuPxmAkTJph69eqZqKgo06JFCzN//vwCx2HBggVGkpk9e7a58847TZUqVUxkZKRp27atWbFihbXc1q1bTf/+/c2ZZ55pKlSoYGJjY80555xjJk2aZPLy8kp0DAAAAMoTlzH52qcDAIBy7ciRI2ratKlq1qypL7/88pRv/+abb9abb76pnTt3FtvVr7xYuHChOnTooHfffddnjC0AAAAUjy6BAACUYwMHDlTHjh1VvXp1ZWVl6fnnn9fPP/+sp59++qRv++GHH1aNGjVUr149a+ysl19+WaNHjw6KYhUAAADKjoIVAADl2L59+3TPPfdo+/btCg8P17nnnqs5c+bo0ksvPenbDg8P18SJE7V582bl5eWpfv36Sk9P17Bhw076tgEAABDY6BIIAAAAAAAARwnxdwIAAAAAAABAfhSsAAAAAAAA4CiMYVUKHo9Hf//9t+Li4uRyufydDgAAAAAAPowx2rdvn2rUqKGQENqoIHBRsCqFv//+WykpKf5OAwAAAACAYm3atEm1atXydxpAmVGwKoW4uDhJ0qZNMxQfH+PnbJzl2XWVTjjG0EY7bcjEHnbsj2TPPjkpF7twvhRU3vZHctY+ASXF+X9y2XV87cBrhED09fz6tsRZU22XLXHswHvRftnZB5WSMsD6/goEKgpWpeDtBhgfH0PB6hhRsRVOOEZ8/CEbMrGHHfsj2bNPTsrFLpwvBZW3/ZGctU9ASXH+n1x2HV878BohEFWoEGtLnKjYHFvi2IH34snDMDYIdHRoBQAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjULACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjhPk7AQAAUDYuV9cSLbdgwTi1b3+2Ldu7/fYuevbZW8u0fk7OEb388pd6551vtHZthrKzDyo+PkZnn52qG264SL16tVNcXMwJ5wkAAIDAR8EKAIAAtWTJRJ/HjzzythYsWKP58x/zmd+oUe1TmVahtm/fq8svH6O1azeqX79LdOedXVW1aqJ27szW/PlrNHLkTH377TrNnn23v1MFAACAA1CwAgAgQF1wwZk+j6tUiVdISEiB+U7Qu/dT+vHHjfrqq0d00UWNfZ7r3r2Vxoy5QZ99trLYGG63W3l5HkVGhp/MVAEAAOAAjGEFAEA59txzn+qii+5T1aq9VaHCNTr77KGaMOF9HTmS57Pc6tV/6sorx6pq1d6KjLxaNWr0U5cuY7V5844iYxtjdP/9sxQe3l0vvfRFkcstX/6bvvxytW6++bICxSqvSpXi1bt3B+vxhg1b5XJ11YQJ7+vRR99W3boDFRnZQwsWrJEkffTR/6lVq3sUE/MvxcVdp44dH9SSJb/4xOzff5Lq1BlYYFtpaW8U6E7pcnXV0KHPa8k7H2r8lddp5LkXacJVN2j1Z3N9lss9dFgfPfmMHru8h/7dvJ1Gt+mkST0HaNWcL4vcfwAAAJQeLawAACjH/vwzU716tVPdusmKiAjTDz+s12OPvaNfftms6dOHSZIOHDisjh0fVN26yXruuVuVnJyorKzdWrDgR+3bd6jQuDk5R9S//2R9+ulyffzxg7r88uZF5jB37veSpG7dWpY6/2ee+VhnnFFTTz55k+LjY1S/fg298cZC3XjjU+rUqZnefPNe5eQc0YQJH6h9+1GaN+9RXXjhWaXejnS0CJYbsU6X3z5YEdFRWvzWB3pt5EMKCQ1Vk04XH11m4tNa+cnnunzozarZsIFyDx1S1u9/6eCevWXaJgAAAApHwQoAgHIsPX2Q9X+Px6O2bc9SpUpxGjDgaT311EAlJcXql182a+fOfXrllTt11VUXWMtfd13bQmPu2rVPV131qNav36pvvnlCTZrULTaHTZuOttJKTa3qM98YI7fbYz12uaTQ0FCfZaKiIvTFF2MVHh5m7UObNiN19tl19NlnaQoJOdpY/IorWui0027Wv//9qr77bsLxDkuhduzI1r/nvKq4yhUlSQ3bttbEq2/UnKenWQWr9avX6IxWLdWu7w3Weo0ualOm7QEAAKBodAkEAKAcW736T3Xr9ogqVeql0NCrFB7eXX37TpLb7dFvv22RJJ1+enUlJcXq3/+eqeef/0zr1mUUGW/9+q1q1epeZWcf1NKlTx63WFWc//53qcLDu1tTQsL1BZbp1q2lVaySpF9/3aK//96lPn06WMUqSYqNjda//tVKS5f+qoMHD5cpn0suaWIVqyQpJDRUTS+/VDsyNmtP1jZJUu2zG+mXb5fqk0lT9cfyVTpyuGzbAgAAQPEoWAEAUE5lZGxT27b3acuWnXr66cH65pvHtXx5up577lZJ0qFDuZKkhIQKWrRovJo2raf775+ls866XTVq9NOYMa8XGOtq2bLf9NtvW9SzZ1vVqlW5RHnUrl1FkrRx4zaf+e3bn63ly9O1fHm6rrzyvELXrV69os/jnTuz/zc/qcCyNWpUksfj0e7dB0qU17GqVSsY01vAOrj3aJe/7veNUIebemvt/EWadtPtGn3hZZp+57+1feOmMm0TAAAAhaNLIAAA5dR//rNUBw4c1gcf3O/THe/77/8qsOzZZ9fRW2+NlDFGa9Zs0MyZX+nhh99SdHSE7rvvWmu5nj3bqlq1JD3wwGx5PEajR/c8bh4dOzbV/ffP0kcfLVOnTuda8xMTY9WiRX1JUqVKcYWu63L5Pq5UKV6SlJm5u8Cyf/+9UyEhIUpKqiDpaHfCnJwjBZbbsSO70G1lZe1Ww2Pm7duxS5IUk5AgSYqMidbltw/W5bcP1r4du/Tzt0s0Z/JUvTL0Ht338duFxgUAAEDp0cIKAIByyvW/ak9kZLg1zxijl14q+hftXC6XmjSpq0mTBisxsYJWrfqzwDKjR/fU5MmD9dBDr2vUqFePm0eLFvXVqVMzvfTSF/rmm5/KsCf/aNCgpmrWrKQ33lgkY4w1/8CBw3r//SVq1aqBYmKiJEl16lTVtm17tXXrP8Wt3Nwj+uKLVYXGnjfvB6tAJUket1vff/6VKqXUVGK1qgWWj6tcUS27d1Gzzh21fUOGcg/RPRAAAMAutLACAKCc6tixqSIiwnTDDRM1cmQPHT58RNOmzdHu3ft9lvvkk2WaOnWOune/QPXqVZMxRh98sER79hxQx47NCo09bFg3xcZG6eabn9P+/Yf1zDM3WwWywrz22t267LKHdOmlo9W//yW67LJzVbVqgrKzD2nNmvX66qsfFB8fc9x9CgkJ0YQJ/XXjjU/pyisf1i23XK6cnCOaOPED7dmzX48/3s9atmfPtnrooTd0/fUTde+9PXT4cK6eeeZjn4He86tcOV7TBg1Vx1sGWL8SuG39RvWe+Ii1zNO9BqrRRW1U/YzTFR0fp21/bdCKTz5XapPGioiOOm7+AAAAKBkKVgAAlFNnnpmi998fpdGjX1OPHuNVqVKcevVqpxEjuqtz5zRrufr1aygxsYImTPhAf/+9UxER4WrQoKZmzrxL/fpdUmT8gQM7qUKFKPXpk64DBw7r5Zfv8BkIPb8qVRK0ZMmTeumlL/T229/onXe+1f79hxUfH6OzzqqtO+/sqkGDOpVov3r1aq8KFaI0fvx76tlzgkJDQ3TBBQ20YME4tW79T6e+unWr6b//fUD33z9b11wzXtWrV9SIEd21fftejR37ZoG43bqdr6ykRvpsygvak7lVlVJq6sbH09Ts8kutZU5v2Vw/LfxWi2a/pSOHc5RQtYpadO2sS2/uX6LcAQAAUDIuk789PYqVnZ2thIQE7d37don+ChxM0teWbODd4oxovMOGTOxhx/5I9uyTk3KxC+dLQeVtfyRn7RNwPC5XV91+exfVu3W0LfE4/wtn1/XFDrxGCEQL5jawJc7q6jttiWMH3ov2y84+qISEntq7d6/i4+P9nQ5QZoxhBQAAAAAAAEehYAUAAAAAAABHYQwrAAAQ9Iz5WJKUvtbPiQAAAEASLawAAAAAAADgMBSsAAAAAAAA4Ch0CSwFt9stSdq8eQe/EniMPVs9Jxxjc+IuGzKxhx37I9mzT07KxS6cLwWVt/2RnLVPQElx/p9cdh1fO/AaIRBt355gS5w9IbttiWMH3ov2y84+KOmf769AoHIZY4y/kwgUy5cvV8uWLf2dBgAAAAAAxVq2bJnOO+88f6cBlBkFq1LYvXu3KlasqAfTP1FUdAV/p1Pu3LbrFX+nYLttLZr4OwXLt5W+9XcKttq6/C5/p2Cr3+v84e8UANjo/uUpJxzj3SqRNmQi/aveihOOEfJ/Z9uQCYqS3OWeE47h/rzpiSdSDoVe/r2/U7DV5388aEucz+KzbYnjFHZcc8uT/YcP6Nxx/9KuXbuUlJTk73SAMqNLYCmEhoZKkqKiKygqOtbP2ZQ/8VER/k7BdocqRPs7BUtMbLi/U7BVeXsPRlSgmzFQnsRFnfgftqKio2zIRIqz4V4UYsP+oGjxsaEnHMNdDj9H2SHUhmPrJDEx9nxeiKiQZ0scp7Djmlseeb+/AoGKQdcBAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADKia+//lpdu3ZVjRo15HK59J///Oe46yxatEjNmzdXVFSU6tWrp+eff/7kJ3ocFKwAAAAAAADKiQMHDqhJkyZ69tlnS7T8+vXrdcUVV6ht27ZavXq17r//ft155516//33T3KmxQvz69YBAAAAAABgm86dO6tz584lXv75559X7dq1NXnyZElSw4YNtWLFCj355JP617/+dZKyPD5aWAEAAAAAAASpJUuWqFOnTj7zLrvsMq1YsUJHjhzxU1a0sCoVt9stSTp86ICfMymfsg/n+jsF2+07cMjfKVgORvnvQnMyHD60398p2Cr3wEF/pwDARvsOn/hnhcOH8mzIxJ57UYgN+4Oixex3n3AMdzn8HGWHUBuOrZMcPGjP54XcsPL1ucOOa255sv9/x8P7/RX/OHz4sHJzA/N6aYyRy+XymRcZGanIyMgTjp2VlaXk5GSfecnJycrLy9OOHTtUvXr1E95GWVCwKoU//vhDkvTIiCv9nEn59IC/EzgpXvN3AuXYf/2dAAAUaaa/E8infN5fy5mH7Aiy1o4g5U+5ewP09HcCjjTT3wk41B9//KHzzjvP32k4xuHDh1U3NVZZ2wKzkBcbG6v9+33/aD9mzBilpaXZEv/YYpgxptD5pxIFq1I4/fTTJUk/LkpRfFyoLTFbf23fTSf1LWNbLEnaeL3/TkwEhtOf3uvvFIp1JDnetljhW7Nti3UyBNO+2snlsfe6mVs9wdZ4wSJ8m70tJo9UjbU1Hpwhc4A9Lc68aj9p31/Yj1SKsS0WnONg1Qh/p3DKRO629/0VdtC+ePtrR9kWS5ISf9hla7z9DZLsC+a253NJ3pHD+mFBuvX9FUfl5uYqa5tbG1fWUXxcYI2OlL3Po9TmG7Rp0ybFx//zud+O1lWSVK1aNWVlZfnM27Ztm8LCwlSpUiVbtlEWFKxKITT0aJGqVvVw207wsIr2fbmJirL3i1dYRQpWKF5UuL3nnN1Co+x7f4WH2xbqpAimfbWTy+2xNV6Ija9DMAmPsPdDY2hUnK3x4Azhlez9Qh0VkWNbrNCoCrbFgnO4Y+z5IhgIog7ZO3REmNu+92tuhWjbYklSVIS915IjMYm2xXLZWLCS/vn+Cl/xcSG2NUA51eLj430KVnZp1aqVPv74Y595X375pVq0aKFwP345oGAFAAAAAACCgkdGHtn7R8uTzaPSFTP3799vDWkkSevXr9f333+vihUrqnbt2ho1apS2bNmiWbNmSZJuvfVWPfvssxoxYoQGDx6sJUuW6JVXXtGbb75p636UFgUrAAAAAACAcmLFihXq0KGD9XjEiBGSpH79+mnmzJnKzMxURkaG9XzdunU1Z84cDR8+XM8995xq1KihZ555Rv/6179Oee75BV3BaurUqZo4caIyMzN11llnafLkyWrbtq2/0wIAAAAAADhh7du3twZNL8zMmTMLzGvXrp1WrVp1ErMqvcAaaewEvf3227rrrrv0wAMPaPXq1Wrbtq06d+7sU1kEAAAAAACAfwVVwSo9PV0DBw7UoEGD1LBhQ02ePFkpKSmaNm2av1MDAAAAAAAnmdt4AnIKRkFTsMrNzdXKlSvVqVMnn/mdOnXS4sWLC10nJydH2dnZPhMAAAAAAABOrqApWO3YsUNut1vJyck+85OTk5WVlVXoOuPHj1dCQoI1paSknIpUAQAAAAAAglrQFKy8XC6Xz2NjTIF5XqNGjdLevXutadOmTaciRQAAAAAAgKAWNL8SWLlyZYWGhhZoTbVt27YCra68IiMjFRkZeSrSAwAAAAAAJ5lHRh4V/Qt6ThRo+dolaFpYRUREqHnz5po7d67P/Llz56p169Z+ygoAAAAAAADHCpoWVpI0YsQI9enTRy1atFCrVq304osvKiMjQ7feequ/UwMAAAAAAMD/BFXBqmfPntq5c6cefvhhZWZmqnHjxpozZ45SU1P9nRoAAAAAADjJPPLI4+8kSinwMrZHUBWsJGnIkCEaMmSIv9MAAAAAAABAEYJmDCsAAAAAAAAEBgpWAAAAAAAAcJSg6xIIAAAAAACCk9sYuY3xdxqlEmj52oWCVRmcs6C3QqKjbIn1V8fptsSRpEa/2zs2118dp9oaD+VPk1XOHg/uSJx9scL3RdsX7CQIpn21U8Q+e2/+B2q4bI0XLKqusvfjyLZzw22NB2eokbTZ3oCZ+2wLtaVHZdtiwTkq/uL2dwqnTHh2rq3xQg4fsS1WzFabv7J67B28OuyAfedJSJ49n0tcecFz7qJ8o0sgAAAAAAAAHIWCFQAAAAAAAByFLoEAAAAAACAoeGTkUWCNCRVo+dqFFlYAAAAAAABwFApWAAAAAAAAcBS6BAIAAAAAgKDgkZE7wLrY0SUQAAAAAAAAcAAKVgAAAAAAAHAUClYAAAAAAABwFMawAgAAAAAAQcEjE3BjQgVavnahhRUAAAAAAAAchYIVAAAAAAAAHIWCFQAAAAAAAByFMawAAAAAAEBQcBsjtwmsMaECLV+70MIKAAAAAAAAjkLBCgAAAAAAAI5Cl0AAAAAAABAUPP+bAkmg5WsXWlgBAAAAAADAUWhhVQZ1ZxuFhdkz6Fmj34fYEkeS1g2ZalssSWo01b7cUD61vXG1v1Mo1gu1ltgW65bNrWyLdTIE077aad6iprbG++NGe6/DwaLZo/beb+y+H8IZWo+41dZ4ke51tsXinCufurTu5u8UTpmcOpVtjRe6Psu2WJEh9raxyK2ZaGu8mN+22xbLXSnOljgud44tcQB/o4UVAAAAAAAAHIUWVgAAAAAAICi4ZeSWPT2mTpVAy9cutLACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjMIYVAAAAAAAICm5zdAokgZavXWhhBQAAAAAAAEehYAUAAAAAAABHoUsgAAAAAAAICp7/TYEk0PK1Cy2sAAAAAAAA4CgUrAAAAAAAAOAoFKwAAAAAAADgKIxhBQAAAAAAgoJHLrnl8ncapeIJsHztQgsrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CmNYAQAAAACAoOAxR6dAEmj52oUWVgAAAAAAAHAUClYAAAAAAABwFLoElsH6Pi6FRNvzs5J/dZxqSxxJajR1iG2xJGndEPtyQ/nU5Al7zzm7NYprZlus8H22hTopgmlf7ZS0z9721XZfh4NF1Q1HbI3H61A+VR+02d6AX4TaFopzrnyq2MLt7xROmQoZB22NZ6pXsi1WTuVo22JJUtTG3bbGO3hGFdtiheTZ87kkz6Y4gL9RsAIAAAAAAEHBLZfcsqcByqkSaPnahS6BAAAAAAAAcBQKVgAAAAAAAHAUugQCAAAAAICgQJfAwEELKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOApjWAEAAAAAgKDgMS55TGCNCRVo+dqFFlYAAAAAAABwFApWAAAAAAAAcBQKVgAAAAAAAHAUxrACAAAAAABBwS2X3AqsMaECLV+70MIKAAAAAAAAjkLBCgAAAAAAAI5Cl0AAAAAAABAU3AqRO8Da7rj9nYCfBNarBAAAAAAAgHKPghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFAwxiWPcfk7jVIxAZavXWhhBQAAAAAAAEehYAUAAAAAAABHoWAFAAAAAAAAR2EMKwAAAAAAEBTccsmtwBoTKtDytQstrAAAAAAAAOAoFKwAAAAAAADgKHQJBAAAAAAAQcFtQuQ2gdV2x238nYF/BNarBAAAAAAAgHKPghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFDwyCVPgLXd8Sg4B7EKrFcJAAAAAAAA5R4FKwAAAAAAADgKBSsAAAAAAAA4CmNYAQAAAACAoOCWS265/J1GqQRavnahhRUAAAAAAAAchYIVAAAAAAAAHIUugQAAAAAAICi4TYjcJrDa7riN8XcKfhFYrxIAAAAAAADKPQpWAAAAAAAAcBQKVgAAAAAAAHAUxrACAAAAAABBwSOXPHL5O41SCbR87UILKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOApjWAEAAAAAgKDgUYjcAdZ2xyPj7xT8IrBeJQAAAAAAAJR7FKwAAAAAAADKmalTp6pu3bqKiopS8+bN9c033xS7/HPPPaeGDRsqOjpaDRo00KxZs05RpoWjSyAAAAAAAAgKbhMitwmstjtuU/ougW+//bbuuusuTZ06VW3atNELL7ygzp07a926dapdu3aB5adNm6ZRo0bppZde0nnnnadly5Zp8ODBSkpKUteuXe3YjVILrFcJAAAAAAAAxUpPT9fAgQM1aNAgNWzYUJMnT1ZKSoqmTZtW6PKzZ8/WLbfcop49e6pevXq6/vrrNXDgQD3xxBOnOPN/ULACAAAAAABwuOzsbJ8pJyen0OVyc3O1cuVKderUyWd+p06dtHjx4kLXycnJUVRUlM+86OhoLVu2TEeOHLFnB0qJghUAAAAAAIDDpaSkKCEhwZrGjx9f6HI7duyQ2+1WcnKyz/zk5GRlZWUVus5ll12ml19+WStXrpQxRitWrND06dN15MgR7dixw/Z9KQnGsAIAAAAAAEHBoxB5AqztjkdHx7DatGmT4uPjrfmRkZHFrudyuXweG2MKzPN68MEHlZWVpQsuuEDGGCUnJ6t///6aMGGCQkNDT3APyiawXiUAAAAAAIAgFB8f7zMVVbCqXLmyQkNDC7Sm2rZtW4FWV17R0dGaPn26Dh48qA0bNigjI0N16tRRXFycKleubPu+lAQFKwAAAAAAgHIiIiJCzZs319y5c33mz507V61bty523fDwcNWqVUuhoaF66623dOWVVyokxD+lI7oEAgAAAAAAlCMjRoxQnz591KJFC7Vq1UovvviiMjIydOutt0qSRo0apS1btmjWrFmSpN9++03Lli3T+eefr927dys9PV1r167Vq6++6rd9oGAFAAAAAACCgtu45DaFj+PkVGXJt2fPntq5c6cefvhhZWZmqnHjxpozZ45SU1MlSZmZmcrIyPhnG263nnrqKf36668KDw9Xhw4dtHjxYtWpU8eu3Sg1ClYAAAAAAADlzJAhQzRkyJBCn5s5c6bP44YNG2r16tWnIKuSYwwrAAAAAAAAOAotrAAAAAAAQFBwK0TuAGu745bxdwp+EVivEgAAAAAAAMo9ClYAAAAAAABwFApWAAAAAAAAcBTGsAIAAAAAAEHBY0LkMYHVdsdjGMMKAAAAAAAA8DsKVgAAAAAAAHAUClYAAAAAAABwFMawAgAAAAAAQcGtELkDrO2OW4xhBQAAAAAAAPgdBSsAAAAAAAA4Cl0CAQAAAABAUPBIchuXv9MoFY+/E/ATWlgBAAAAAADAUShYAQAAAAAAwFEoWAEAAAAAAMBRGMMKAAAAAAAEBY9C5AmwtjuBlq9dgnOvAQAAAAAA4FgUrAAAAAAAAOAoFKwAAAAAAADgKIxhBQAAAAAAgoLbhMhtAqvtTqDla5fg3GsAAAAAAAA4FgUrAAAAAAAAOApdAgEAAAAAQFDwyCWPXP5Oo1QCLV+70MIKAAAAAAAAjkLBCgAAAAAAAI5CwQoAAAAAAACOwhhWAAAAAAAgKLhNiNwmsNruBFq+dgnOvQYAAAAAAIBjUbACAAAAAACAo1CwAgAAAAAAgKMwhhUAAAAAAAgKboXIHWBtdwItX7sE514DAAAAAADAsShYAQAAAAAAwFEoWAEAAAAAAMBRGMMKAAAAAAAEBY9xyWNc/k6jVAItX7vQwgoAAAAAAACOQsEKAAAAAAAAjkKXQAAAAAAAEBQ8CpE7wNrueAIsX7sE514DAAAAAADAsShYAQAAAAAAwFEoWAEAAAAAAMBRGMMKAAAAAAAEBY8JkccEVtudQMvXLsG51wAAAAAAAHAsClYAAAAAAABwFApWAAAAAAAAcBTGsAIAAAAAAEHBLZfccvk7jVIJtHztQgsrAAAAAAAAOAoFKwAAAAAAADgKXQIBAAAAAEBQ8JgQeUxgtd0JtHztEpx7DQAAAAAAAMeiYAUAAAAAAABHoWAFAAAAAAAAR2EMKwAAAAAAEBTcktxy+TuNUnH7OwE/oYUVAAAAAAAAHIWCFQAAAAAAAByFghUAAAAAAAAchTGsAAAAAABAUPCYEHlMYLXdCbR87RKcew0AAAAAAADHomAFAAAAAAAAR6FLIAAAAAAACApuEyJ3gHWxC7R87RKcew0AAAAAAADHomAFAAAAAAAAR6FgBQAAAAAAAEdhDCsAAAAAABAUjFzyyOXvNErFBFi+dqGFFQAAAAAAAByFghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFBwmxC5TWC13Qm0fO0SnHsNAAAAAAAAx6JgBQAAAAAAAEehSyAAAAAAAAgKHuOSx7j8nUapBFq+dqGFFQAAAAAAAByFghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFBwK0TuAGu7E2j52iU49xoAAAAAAACORcEKAAAAAAAAjkLBCgAAAAAAAI7CGFZlcMaEvQoLPWxLrCarhtgSR5La3rjatliS1OQJ+3JD+fTDv6f6O4Vi3bK5lW2xXqi1xLZYJ0Mw7audGr5g73Xuoi72XoeDxZq/mtgar203Xofy6K97z7Q1njm0zbZYnHPl0+ZXK/k7hVMmt14VewMu+9G2UNH16tgWS5LcFWNtjRe97E/7glW353UIdefYEqe88hiXPMbl7zRKJdDytQstrAAAAAAAAOAoFKwAAAAAAADgKHQJBAAAAAAAQcGjEHkCrO1OoOVrl+DcawAAAAAAADgWBSsAAAAAAAA4SrkpWH399dfq2rWratSoIZfLpf/85z8+zxtjlJaWpho1aig6Olrt27fXTz/95J9kAQAAAAAAUKRyU7A6cOCAmjRpomeffbbQ5ydMmKD09HQ9++yzWr58uapVq6aOHTtq3759pzhTAAAAAADgD27jCsgpGJWbQdc7d+6szp07F/qcMUaTJ0/WAw88oB49ekiSXn31VSUnJ+uNN97QLbfccipTBQAAAAAAQDHKTQur4qxfv15ZWVnq1KmTNS8yMlLt2rXT4sWL/ZgZAAAAAAAAjlVuWlgVJysrS5KUnJzsMz85OVkbN24scr2cnBzl5ORYj7Ozs09OggAAAAAAALAERcHKy+Xy7fdpjCkwL7/x48dr7NixJzstAAAAAABwCniMS54AGxMq0PK1S1B0CaxWrZqkf1paeW3btq1Aq6v8Ro0apb1791rTpk2bTmqeAAAAAAAACJKCVd26dVWtWjXNnTvXmpebm6tFixapdevWRa4XGRmp+Ph4nwkAAAAAAAAnV7npErh//3798ccf1uP169fr+++/V8WKFVW7dm3dddddGjdunOrXr6/69etr3LhxiomJUa9evfyYNQAAAAAAOFWMCZHHBFbbHRNg+dql3BSsVqxYoQ4dOliPR4wYIUnq16+fZs6cqZEjR+rQoUMaMmSIdu/erfPPP19ffvml4uLi/JUyAAAAAAAAClFuClbt27eXMabI510ul9LS0pSWlnbqkgIAAAAAAECpBWe7MgAAAAAAADhWuWlhBQAAAAAAUBy3XHLL5e80SiXQ8rULLawAAAAAAADKmalTp6pu3bqKiopS8+bN9c033xS7/Ouvv64mTZooJiZG1atX14ABA7Rz585TlG1BFKwAAAAAAADKkbffflt33XWXHnjgAa1evVpt27ZV586dlZGRUejy3377rfr27auBAwfqp59+0rvvvqvly5dr0KBBpzjzf1CwAgAAAAAAKEfS09M1cOBADRo0SA0bNtTkyZOVkpKiadOmFbr80qVLVadOHd15552qW7euLrzwQt1yyy1asWLFKc78HxSsAAAAAABAUPAYyWNcATaVbh9zc3O1cuVKderUyWd+p06dtHjx4kLXad26tTZv3qw5c+bIGKOtW7fqvffeU5cuXcp6qE8YBSsAAAAAAACHy87O9plycnIKXW7Hjh1yu91KTk72mZ+cnKysrKxC12ndurVef/119ezZUxEREapWrZoSExM1ZcoU2/ejpChYAQAAAAAAOFxKSooSEhKsafz48cUu73L5/rqgMabAPK9169bpzjvv1EMPPaSVK1fq888/1/r163Xrrbfaln9phfltywAAAAAAAKeQx4TIYwKr7Y43302bNik+Pt6aHxkZWejylStXVmhoaIHWVNu2bSvQ6spr/PjxatOmje69915J0jnnnKMKFSqobdu2evTRR1W9enU7dqVUAutVAgAAAAAACELx8fE+U1EFq4iICDVv3lxz5871mT937ly1bt260HUOHjyokBDfElFoaKikoy2z/IGCFQAAAAAAQDkyYsQIvfzyy5o+fbp+/vlnDR8+XBkZGVYXv1GjRqlv377W8l27dtUHH3ygadOm6a+//tJ3332nO++8Uy1btlSNGjX8sg90CQQAAAAAAChHevbsqZ07d+rhhx9WZmamGjdurDlz5ig1NVWSlJmZqYyMDGv5/v37a9++fXr22Wd19913KzExURdffLGeeOIJf+0CBSsAAAAAABAcPHLJo8IHHneqsuY7ZMgQDRkypNDnZs6cWWDeHXfcoTvuuKNM2zoZ6BIIAAAAAAAAR6FgBQAAAAAAAEehYAUAAAAAAABHYQwrAAAAAAAQFNzGJbcJrDGsAi1fu9DCCgAAAAAAAI5CwQoAAAAAAACOQpdAAAAAAAAQFDwmRB4TWG13Ai1fuwTnXgMAAAAAAMCxKFgBAAAAAADAUShYAQAAAAAAwFEYw6oMjlSPlwmLsidWnC1hJEkv1FpiXzBJjeKa2RoP5c8tm1v5O4Vi2fmeYF/LpyOxxtZ4dl+Hg0WzRHvvN7wO5VPr6k1sjZcYEWFbrBdqfW1bLDhHl9Bu/k7hlDEul63xQqtUsS2WJz7GtliSlBdr33tfkiIT422L5Y6xJzeP22NLnPLKI5c8xt5z/mTzKLDytQstrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CiMYQUAAAAAAIKCkSvgxoQyAZavXWhhBQAAAAAAAEehYAUAAAAAAABHoWAFAAAAAAAAR2EMKwAAAAAAEBQ8xiWPCawxoQItX7vQwgoAAAAAAACOQsEKAAAAAAAAjkKXQAAAAAAAEBQ8JkQeE1htdwItX7sE514DAAAAAADAsShYAQAAAAAAwFEoWAEAAAAAAMBRGMMKAAAAAAAEBY9xyWNc/k6jVAItX7vQwgoAAAAAAACOQsEKAAAAAAAAjkLBCgAAAAAAAI7CGFZlEJ6ZrbDQHHti7Yu2JY4k3bK5lW2xJCl8n63hUA69UGuJv1Molp3vCfa1fGq4v5mt8ey+DgeLqD0eW+PxOpRPFTLt+ezlZXJzbYvFOVdOue29NjmZyxhb47m3b7ctVlhcBdtiSVJYmL1tNsyebNtihURH2hPHfcSWOOWVRy55FFhjQgVavnahhRUAAAAAAAAchYIVAAAAAAAAHIUugQAAAAAAICh4jEseE1hd7AItX7vQwgoAAAAAAACOQsEKAAAAAAAAjkLBCgAAAAAAAI7CGFYAAAAAACAoMIZV4KCFFQAAAAAAAByFghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFBgDKvAQQsrAAAAAAAAOAoFKwAAAAAAADgKXQIBAAAAAEBQoEtg4KCFFQAAAAAAAByFghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFAwkjwKrDGhjL8T8BNaWAEAAAAAAMBRKFgBAAAAAADAUShYAQAAAAAAwFEYwwoAAAAAAAQFj3HJYwJrDKtAy9cutLACAAAAAACAo1CwAgAAAAAAgKPQJRAAAAAAAAQFugQGDlpYAQAAAAAAwFEoWAEAAAAAAMBRKFgBAAAAAADAURjDCgAAAAAABAXGsAoctLACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjMIYVAAAAAAAICoxhFThoYQUAAAAAAABHoWAFAAAAAAAAR6FLIAAAAAAACArGuGQCrItdoOVrF1pYAQAAAAAAwFEoWAEAAAAAAMBRKFgBAAAAAADAURjDqgyOVI+XCYuyJ1acLWEkSS/UWmJfMEmN4prZGg/lzy2bW/k7hWLZ+Z5gX8unI7HG1nh2X4eDRbNEe+83vA7lU+vqTWyNlxgRYVusF2p9bVssOEeX0G7+TuGUMS57x8cJrVLFtlie+BjbYklSXqx9731JikyMty2WO8ae3Dxujy1xyiuPXPIosMaECrR87UILKwAAAAAAADgKBSsAAAAAAAA4CgUrAAAAAAAAOApjWAEAAAAAgKDgMS55TGCNCRVo+dqFFlYAAAAAAABwFApWAAAAAAAAcBS6BAIAAAAAgKBgjEsmwLrYBVq+dqGFFQAAAAAAAByFghUAAAAAAAAchYIVAAAAAAAAHIUxrAAAAAAAQFDwGJc8ATYmVKDlaxdaWAEAAAAAAMBRKFgBAAAAAADAUShYAQAAAAAAwFEYwwoAAAAAAAQFY1wyATYmVKDlaxdaWAEAAAAAAMBRKFgBAAAAAADAUegSCAAAAAAAgoIxLnkCrItdsHYJpGBVBuGZ2QoLzbEn1r5oW+JI0i2bW9kWS5LC99kaDuXQC7WW+DuFYtn5nmBfy6eG+5vZGs/u63CwiNrjsTUer0P5VCHTns9eXiY317ZYnHPllNvea5OTuYyxNZ57+3bbYoXFVbAtliSFhdnbycjsybYtVkh0pD1x3EdsiQP4G10CAQAAAAAA4CgUrAAAAAAAAOAodAkEAAAAAABBwUiyuRfsSRdg6dqGFlYAAAAAAABwFApWAAAAAAAAcBQKVgAAAAAAAHAUxrACAAAAAABBwSOXXHL5O41S8QRYvnahhRUAAAAAAAAchYIVAAAAAAAAHIUugQAAAAAAICgY45IxgdXFLtDytQstrAAAAAAAAMqZqVOnqm7duoqKilLz5s31zTffFLls//795XK5CkxnnXXWKczYFwUrAAAAAACAcuTtt9/WXXfdpQceeECrV69W27Zt1blzZ2VkZBS6/NNPP63MzExr2rRpkypWrKhrr732FGf+DwpWAAAAAAAA5Uh6eroGDhyoQYMGqWHDhpo8ebJSUlI0bdq0QpdPSEhQtWrVrGnFihXavXu3BgwYcIoz/wdjWAEAAAAAgKDgMS65AmxMKM//8s3OzvaZHxkZqcjIyALL5+bmauXKlbrvvvt85nfq1EmLFy8u0TZfeeUVXXrppUpNTS1j1ieOFlYAAAAAAAAOl5KSooSEBGsaP358ocvt2LFDbrdbycnJPvOTk5OVlZV13O1kZmbqs88+06BBg2zJu6xoYQUAAAAAAOBwmzZtUnx8vPW4sNZV+blcvi3JjDEF5hVm5syZSkxMVPfu3cuUp10oWAEAAAAAADhcfHy8T8GqKJUrV1ZoaGiB1lTbtm0r0OrqWMYYTZ8+XX369FFERMQJ5Xui6BIIAAAAAACCgjGBOZVGRESEmjdvrrlz5/rMnzt3rlq3bl3suosWLdIff/yhgQMHlvbQ2o4WVgAAAAAAAOXIiBEj1KdPH7Vo0UKtWrXSiy++qIyMDN16662SpFGjRmnLli2aNWuWz3qvvPKKzj//fDVu3NgfafugYAUAAAAAAFCO9OzZUzt37tTDDz+szMxMNW7cWHPmzLF+9S8zM1MZGRk+6+zdu1fvv/++nn76aX+kXAAFKwAAAAAAgHJmyJAhGjJkSKHPzZw5s8C8hIQEHTx48CRnVXIUrAAAAAAAQFAwxiVjjv9LeU4SaPnahUHXAQAAAAAA4CgUrAAAAAAAAOAodAkEAAAAAABBgS6BgYMWVgAAAAAAAHAUClYAAAAAAABwFApWAAAAAAAAcBTGsAIAAAAAAEHBY1xyBdiYUJ4Ay9cutLACAAAAAACAo1CwAgAAAAAAgKNQsAIAAAAAAICjMIYVAAAAAAAICsYcnQJJoOVrF1pYAQAAAAAAwFEoWAEAAAAAAMBR6BIIAAAAAACCwtEugS5/p1EqdAkEAAAAAAAAHICCFQAAAAAAAByFghUAAAAAAAAchTGsAAAAAABAUDDGFYBjWAVWvnahhRUAAAAAAAAchYIVAAAAAAAAHIWCFQAAAAAAAByFMawAAAAAAEBQMP+bAkmg5WsXWlgBAAAAAADAUShYAQAAAAAAwFHoEggAAAAAAIKCMS4Z4/J3GqUSaPnahRZWAAAAAAAAcBQKVgAAAAAAAHAUClYAAAAAAABwFMawAgAAAAAAwcH8bwokgZavTWhhBQAAAAAAAEehYAUAAAAAAABHoWAFAAAAAAAAR2EMKwAAAAAAEByMS8a4/J1F6QRavjahhRUAAAAAAAAchYIVAAAAAAAAHIUugQAAAAAAICgYc3QKJIGWr10oWJWBy+ORSx5bYkXss+/Mm7eoqW2xJCnJxtxQPjV8YYi/UyjWkVj7zuGG+5vZFutkCKZ9tdPPt0y1Nd7pr99ma7xgEVHX3nEZ7L4fwhnO2LbD1ngetz2f5SRp3qLmtsWCc5wRt9PfKZwyYXsO2xrPVbmSbbHMvgO2xZIkJVSwNZwrMd62WJ4wezpAGRcdqVA+cCYDAAAAAADAUShYAQAAAAAAwFHoEggAAAAAAIKCMS4ZY+9wBCdboOVrF1pYAQAAAAAAwFEoWAEAAAAAAMBRKFgBAAAAAADAURjDCgAAAAAABAfjOjoFkkDL1ya0sAIAAAAAAICjULACAAAAAACAo9AlEAAAAAAABAVjjk6BJNDytQstrAAAAAAAAOAoFKwAAAAAAADgKBSsAAAAAAAA4CiMYQUAAAAAAIKD+d8USAItX5vQwgoAAAAAAACOQsEKAAAAAAAAjkLBCgAAAAAAAI7CGFYAAAAAACAoGOOSMS5/p1EqgZavXWhhBQAAAAAAAEehhVUZ5NZMlCcsypZYB2rYVyn948aptsWSpEZTh9gaD+XPRV1W+zuFYr1Qa4ltsW7Z3Mq2WCdDMO2rnU5//TZb4/1x4zRb4wWLZo/ae79ZdzuvQ3nUeuWttsZL3LLVtli898unLs9183cKp0xuncq2xov42752EaZmFdtiSVJeQqSt8UL37Lctlomw5+u5cefZEgfwNwpWAAAAAAAgeBh/J4CSoEsgAAAAAAAAHIWCFQAAAAAAAByFghUAAAAAAADKrH379po1a5YOHTpkW0wKVgAAAAAAICgY4wrIyemaN2+ukSNHqlq1aho8eLCWLl16wjHLTcFq/PjxOu+88xQXF6eqVauqe/fu+vXXX32WMcYoLS1NNWrUUHR0tNq3b6+ffvrJTxkDAAAAAAAEvqeeekpbtmzRrFmztH37dl100UVq1KiRnnzySW3dWrZf5i03BatFixbp9ttv19KlSzV37lzl5eWpU6dOOnDggLXMhAkTlJ6ermeffVbLly9XtWrV1LFjR+3bt8+PmQMAAAAAAAS20NBQXXXVVfrPf/6jLVu2qFevXnrwwQeVkpKi7t27a/78+aWKF3aS8jzlPv/8c5/HM2bMUNWqVbVy5UpddNFFMsZo8uTJeuCBB9SjRw9J0quvvqrk5GS98cYbuuWWW/yRNgAAAAAAQLmxbNkyzZgxQ2+++aaqVq2q/v37KzMzU127dtVtt92mJ598skRxyk0Lq2Pt3btXklSxYkVJ0vr165WVlaVOnTpZy0RGRqpdu3ZavHhxoTFycnKUnZ3tMwEAAAAAgABlAnRyuG3btumpp55S48aN1bZtW23fvl1vvfWWNmzYoLFjx+rFF1/Uf//7Xz3//PMljlluWljlZ4zRiBEjdOGFF6px48aSpKysLElScnKyz7LJycnauHFjoXHGjx+vsWPHntxkAQAAAAAAAlitWrV02mmn6aabblL//v1VpUqVAsu0bNlS5513XoljlsuC1dChQ7VmzRp9++23BZ5zuXxH1zfGFJjnNWrUKI0YMcJ6nJ2drZSUFHuTBQAAAAAACGDz5s1T27Zti10mPj5eCxYsKHHMctcl8I477tBHH32kBQsWqFatWtb8atWqSfqnpZXXtm3bCrS68oqMjFR8fLzPBAAAAAAAApUrQCdnGzNmjPbs2VNgfnZ2ti6++OIyxSw3BStjjIYOHaoPPvhA8+fPV926dX2er1u3rqpVq6a5c+da83Jzc7Vo0SK1bt36VKcLAAAAAABQLixatEi5ubkF5h8+fFjffPNNmWKWmy6Bt99+u9544w3997//VVxcnNWSKiEhQdHR0XK5XLrrrrs0btw41a9fX/Xr19e4ceMUExOjXr16+Tl7AAAAAACAwLJmzRpJRxsRrVu3zqdXm9vt1ueff66aNWuWKXa5KVhNmzZNktS+fXuf+TNmzFD//v0lSSNHjtShQ4c0ZMgQ7d69W+eff76+/PJLxcXFneJsAQAAAAAAAlvTpk3lcrnkcrkK7foXHR2tKVOmlCl2uSlYGXP833l0uVxKS0tTWlrayU8IAAAAAAA4i/nfFEgcnO/69etljFG9evW0bNkyn18HjIiIUNWqVRUaGlqm2OWmYAUAAAAAAIBTJzU1VZLk8Xhsj03BCgAAAAAAAKXy0UcfqXPnzgoPD9dHH31U7LLdunUrdXwKVgAAAAAAACiV7t27KysrS1WrVlX37t2LXM7lcsntdpc6PgUrAAAAAAAQHBjDyjb5uwGejC6BIbZHBAAAAAAAQFA4cuSIOnTooN9++83WuBSsAAAAAAAAUCbh4eFau3atXC6XrXEpWAEAAAAAgOBgXIE5OVzfvn31yiuv2BqTMawAAAAAAABQZrm5uXr55Zc1d+5ctWjRQhUqVPB5Pj09vdQxKVgBAAAAAACgzNauXatzzz1XkgqMZVXWroIUrAAAAAAAAFBmCxYssD0mBSsAAAAAABAUjDk6BZJAy9cufitYbd26VcnJyUU+v2LFCrVo0eIUZgQAAAAAAICyWL58ud59911lZGQoNzfX57kPPvig1PH89iuBTZs21fz58wt97umnn9aFF154ijMCAAAAAABAab311ltq06aN1q1bpw8//FBHjhzRunXrNH/+fCUkJJQppt8KVo0aNdJll12mtLQ0mf+1b9u7d6969Oih4cOHa/Dgwf5KDQAAAAAAACU0btw4TZo0SZ988okiIiL09NNP6+eff9Z1112n2rVrlymm3wpWX331le6//349+uijuvTSS/XJJ5+oadOmWrBggd577z1NmTLFX6kBAAAAAIDyyATo5HB//vmnunTpIkmKjIzUgQMH5HK5NHz4cL344otlium3gpXL5dLYsWP15ZdfasmSJbrqqqsUFxen1atXq0ePHv5KCwAAAAAAAKVQsWJF7du3T5JUs2ZNrV27VpK0Z88eHTx4sEwx/VawkqTs7Gw999xzOnz4sKpWrar169fr66+/9mdKAAAAAAAAKIW2bdtq7ty5kqTrrrtOw4YN0+DBg3XDDTfokksuKVNMv/1K4OrVq3Xddddp+/bteuedd3TFFVfolltu0YABA7Ro0SI999xzioqK8ld6AAAAAAAAKIFnn31Whw8fliSNGjVK4eHh+vbbb9WjRw89+OCDZYrpt4JV69at1bBhQ33++ec67bTTJEmzZ89Wu3btNGzYMC1fvlxr1qzxV3oAAAAAAKC8Ma6jUyAJgHwrVqxo/T8kJEQjR47UyJEjTyim3wpW/fv31+TJkxUZGekzf9CgQTr//PPVs2dPP2UGAAAAAACA4mRnZ5d42fj4+FLH91vBatq0aUU+d/bZZ2vFihWnMBsAAAAAAACUVGJiolyu4lt/GWPkcrnkdrtLHd9vBSuvL774QgsXLtSOHTv04IMPqnbt2lq+fLnq1KmjmJgYf6cHAAAAAADKCZc5OgUSp+a7YMGCkxrfbwWrgwcP6qqrrtK8efOsitxtt92m2rVr68knn1RKSoqefPJJf6UHAAAAAACAIrRr1+6kxvdbweqBBx7QihUr9P7776tjx44+/Rk7deqkKVOm+Cs1AAAAAAAAFGPNmjVq3LixQkJCjvujeeecc06p44eUNbET9e677+qRRx7R1VdfrejoaJ/nateurYyMDD9lBgAAAAAAENimTp2qunXrKioqSs2bN9c333xT7PI5OTl64IEHlJqaqsjISJ122mmaPn16kcs3bdpUO3bssP7frFkzNW3atMDUrFmzMuXvtxZW27dv11lnnVXocyEhITp06NApzggAAAAAAJRr5n9TIClDvm+//bbuuusuTZ06VW3atNELL7ygzp07a926dapdu3ah61x33XXaunWrXnnlFZ1++unatm2b8vLyitzG+vXrVaVKFev/dvNbwapmzZr68ccf1aFDhwLPrVmzRnXr1vVDVgAAAAAAAIEtPT1dAwcO1KBBgyRJkydP1hdffKFp06Zp/PjxBZb//PPPtWjRIv3111+qWLGiJKlOnTrFbiM1NbXQ/9vFbwWrHj166LHHHlPbtm2tvowul0sbN27UpEmTNGDAAH+lBgAAAAAAEJByc3O1cuVK3XfffT7zO3XqpMWLFxe6zkcffaQWLVpowoQJmj17tipUqKBu3brpkUceKTCMU1G2bNmi7777Ttu2bZPH4/F57s477yz1fvitYDVmzBjNmzdPLVu2VOPGjeVyuTRgwAD9+eefatCgQYEDCwAAAAAAEKyys7N9HkdGRioyMrLAcjt27JDb7VZycrLP/OTkZGVlZRUa+6+//tK3336rqKgoffjhh9qxY4eGDBmiXbt2FTuOldeMGTN06623KiIiQpUqVZLL5bKec7lcZSpY+W3Q9bi4OC1evFiPPPKIYmNjddpppykmJkajRo3S119/XeIKHgAAAAAAQIkYV2BOklJSUpSQkGBNhXXtyy9/0UiSjDEF5nl5PB65XC69/vrratmypa644gqlp6dr5syZJRpj/KGHHtJDDz2kvXv3asOGDVq/fr01/fXXXyV8cXz5rYWVJEVHR+u+++6jNRUAAAAAAEAxNm3apPj4eOtxYa2rJKly5coKDQ0t0Jpq27ZtBVpdeVWvXl01a9ZUQkKCNa9hw4Yyxmjz5s2qX79+sbkdPHhQ119/vUJC7GsX5bcWVgAAAAAAACiZ+Ph4n6moglVERISaN2+uuXPn+syfO3euWrduXeg6bdq00d9//639+/db83777TeFhISoVq1ax81t4MCBevfdd0uxN8d3SltYXXzxxSVe1uVyad68eScxGwAAAAAAEFTM/6ZAUoZ8R4wYoT59+qhFixZq1aqVXnzxRWVkZOjWW2+VJI0aNUpbtmzRrFmzJEm9evXSI488ogEDBmjs2LHasWOH7r33Xt10000lGrJp/PjxuvLKK/X555/r7LPPVnh4uM/z6enppd6HU1qw8vaJ9Pr111+VlZWl1NRUVatWTVlZWdq4caOqV6+uBg0anMrUAAAAAAAAyoWePXtq586devjhh5WZmanGjRtrzpw5Sk1NlSRlZmYqIyPDWj42NlZz587VHXfcoRYtWqhSpUq67rrr9Oijj5Zoe+PGjdMXX3xh1XKOHXS9LE5pwWrhwoXW/z///HMNHjxY3333nVq1amXNX7x4sXr27Kl77rnnVKYGAAAAAABQbgwZMkRDhgwp9LmZM2cWmHfmmWcW6EZYUunp6Zo+fbr69+9fpvUL47cxrEaPHq20tDSfYpUktW7dWmPGjNEDDzzgp8wAAAAAAABQUpGRkWrTpo2tMf1WsPrpp5+UkpJS6HO1a9fWL7/8coozAgAAAAAA5ZoJ0Mnhhg0bpilTptga85R2CcwvOTlZ77//vjp16lTguXfffbfIn1oEAAAAAACAcyxbtkzz58/XJ598orPOOqvAoOsffPBBqWP6rWA1ZMgQ3Xfffdq1a5d69eplDbr++uuv68MPP9T48eP9lRoAAAAAAABKKDExUT169LA1pt8KViNHjtTBgwc1YcIEq9JmjFFUVJQeeOABjRw50l+pAQAAAAAAoIRmzJhhe0y/FawkKS0tTcOHD9eSJUu0c+dOVapUSRdccIESExP9mRYAAAAAACiPAmRMKB+Blq9N/FqwkqSEhARdfvnl/k4DAAAAAAAAJXTuuedq3rx5SkpKUrNmzeRyuYpcdtWqVaWO79eClTFGy5cv18aNG3Xo0KECz/ft29cPWQEAAAAAAKA4V111lSIjIyVJ3bt3tz2+3wpWv/32m7p166bff/9dxhRs3+ZyuShYAQAAAAAA+xjX0SmQODTfMWPGFPp/u/itYHX77bfr8OHDevvtt3XOOedYVTkAAAAAAAAEJm+t58CBA+rYsaPq169fpjh+K1gtW7ZML730kq655hp/pQAAAAAAAIAyuvfee5Wbm6unn35akpSbm6sLLrhA69atU0xMjEaOHKkvv/xSrVu3LnXsELuTLanY2FjFx8f7a/MAAAAAAAA4AZ999pkuueQS6/Hrr7+ujIwM/f7779q9e7euvfZaPfbYY2WK7beC1YABA/TGG2/4a/MAAAAAACDIuExgTk6VkZGhRo0aWY+//PJLXXPNNUpNTZXL5dKwYcO0evXqMsX2W5fAxo0b680331S3bt3UtWtXVapUqcAyPXr08ENmAAAAAAAAOJ6QkBCfH9JbunSpHnzwQetxYmKidu/eXabYfitY9erVS5K0fv16ffLJJ//P3p3H2Vj+fxx/32bMbiyDYTKZse9LKEN2Q5aQVJaERhGFL0W0GBKSRMlWMZZI30KLLcX4KqOQforSRsQgSibLmOX+/THNyWmGZsZtzn3mvJ6Px/V4ONd9n8/53Gfc97nnM9d1nSzbDcNQWlpafqcFAAAAAACAHKhWrZref/99jRw5Uvv27dPhw4fVqlUrx/aff/5ZoaGheYrtsoLVli1bXPXSAAAAAAAAuEaPPfaYevXqpbVr12rfvn3q2LGjIiMjHdvXrVunm2++OU+xXVawatGihateGgAAAAAAeCLzr+ZObJzvnXfeqXXr1mnt2rVq166dHnnkEaftAQEBGjJkSJ5iu6xgBQAAAAAAAPfWtm1btW3bNttt48ePz3PcfC1YDRs2LMf7GoahWbNmXcdsAAAAAAAAYEf5WrCaPXt2jvelYAUAAAAAAOCZ8rVglZ6enp8vBwAAAAAAADdUyNUJAAAAAAAAAJejYAUAAAAAAIA8a926tc6cOZOl/+zZs2rdunWeYvItgQAAAAAAwCMYkgzT1VnkjuHqBHIgPj5ely5dytJ/8eJFbdu2LU8xKVgBAAAAAAAg1/bu3ev49/79+3X8+HHH47S0NG3YsEE33HBDnmJTsAIAAAAAAECu1atXT4ZhyDCMbKf++fv76+WXX85TbJcUrNLS0vTjjz+qdOnSKlasmCtSuCaFT/wpb68US2KV/sK6H0H9SUMsiyVJpQ9Zc4wouPb+VNfVKVxV/WL1LYvld8be33LqScdqJZ9IawdYW30d9hR7npxjaTx+DgVT2V2J1gYsbN97MNhDmH5xdQr5xki19rPfCPCzLliytb+T+P7wh6Xx0v84a1msQn4+1sRJyzotC7heDh48KNM0VaFCBX3++ecqVaqUY5uPj49Kly4tLy+vPMV2ScHKNE3VqFFD77//vjp06OCKFAAAAAAAgKcxjYzmTmycb/ny5SVJ6enW/9HbJd8S6O3trTJlylyXAwIAAAAAAED+mTJlihYuXJilf+HChXruuefyFNMlBStJ6tmzp5YsWeKqlwcAAAAAAIAF5s+fr2rVqmXpr1mzpubNm5enmC5bdL1evXpauXKlWrdure7du6ts2bIyDOdhbt27d3dRdgAAAAAAoMAx/2ruxA3yPX78uMqWLZulv1SpUkpMzNs6lC4rWN13332SpKNHjyo+Pj7LdsMwlJaWls9ZAQAAAAAAIDfCw8P16aefKjIy0qn/008/VVhYWJ5iuqxgtWXLFle9NAAAAAAAACwycOBAjRgxQikpKWrdurUk6eOPP9bo0aM1atSoPMV0WcGqRYsWrnppAAAAAAAAWGT06NH67bffNGTIEF26dEmS5OfnpzFjxmjs2LF5iumyglWmpKQkJSQk6PTp0ypZsqQaN26sIkWKuDotAAAAAABQ0LCG1XVhGIaee+45PfXUU/rmm2/k7++vypUry9fXN88xXVqwmj59uiZMmKDz58/LNDN+AoGBgZowYYJGjhzpytQAAAAAAACQC0FBQWrUqJElsVxWsFqyZIlGjx6tDh06qH///goLC9OxY8e0ePFiPfbYYypVqpT69u3rqvQAAAAAAACQA61atZJhGFfcvnnz5lzHdFnB6sUXX1Tv3r21bNkyp/677rpL9957r1588UUKVgAAAAAAADZXr149p8cpKSn68ssv9fXXX6tfv355iumygtW3336rKVOmZLvt3nvv1R133JHPGQEAAAAAgILMMDOaO3GHfF988cVs+2NjY/Xnn3/mKWaha0noWvj7++u3337Ldttvv/0mf3//fM4IAAAAAAAAVrn33nu1cOHCPD3XZQWrZs2aKTY2VseOHXPqP378uCZOnKjmzZu7KDMAAAAAAABcq4SEBPn5+eXpuS6bEjh58mQ1adJElSpVUps2bVS2bFklJiZq8+bNKly4sFatWuWq1AAAAAAAQEFk/tXciRvk2717d6fHpmkqMTFRu3bt0lNPPZWnmC4rWNWsWVM7d+7U+PHjtWXLFp0+fVohISHq1q2bxo8frypVqrgqNQAAAAAAAORQ0aJFnR4XKlRIVatW1cSJE9WuXbs8xXRZwUqSqlSpohUrVrgyBQAAAAAAAFyDRYsWWR7TZWtYtW7dWt9++22227777ju1bt06nzMCAAAAAACAHbhshFV8fLzOnj2b7bakpCRt3bo1nzMCAAAAAAAFGmtYWaZ48eIyDCNH+/7222+5ju/SKYFXkpiYqICAAFencUUpoUEyvfO2yv0/nbypsCVxJGn/kDmWxZKkGnOGWBoPBU+zLntcncJVzS+XYFmsQb9EWRbrevCkY7XSx1vrWRpv/9C5lsbzFPUnWft5s+dJaz8PYQ9Nfhtsabxi605bFov/cwVTp3VdXJ1CvjG9rZ14Y56/aF2s4kUsiyVJl0JDLY3n+7N1v1KnBfpaEic9zabVDRQ4M2fOdPz79OnTmjRpktq3b6+oqIzfJxISErRx40b3WHT93Xff1bvvvut4/Mwzz6hUqVJO+1y4cEHx8fGqX79+fqYGAAAAAACAHOrXr5/j33feeacmTpyohx9+2NE3bNgwzZ49Wx999JH+85//5Dp+vhas9u/fr//+97+SJMMwtHnzZhUq5FzN9/X1Ve3atTVr1qz8TA0AAAAAAAB5sHHjRj333HNZ+tu3b6/HH388TzHzddH1sWPHKikpSUlJSTJNU1u2bHE8zmynTp3Sli1bVKdOnfxMDQAAAAAAFHCG6Z7N7kJCQrR69eos/WvWrFFISEieYrpsDav09HRXvTQAAAAAAAAsMmHCBMXExCg+Pt6xhtWOHTu0YcMGvfbaa3mKma8jrC63Y8cOvfXWW9lue+utt/TZZ5/lc0YAAAAAAADIrf79+2v79u0qVqyYVq1apXfeeUdFixbVp59+qv79++cppstGWI0bN05NmzbV3XffnWXb/v379eqrr2rTpk0uyAwAAAAAABRIppHR3Imb5HvLLbfojTfesCyeywpWe/fu1ahRo7Lddsstt2jOHL4eGAAAAAAAwB2kpaVpzZo1+uabb2QYhmrUqKEuXbrIy8srT/FcVrA6d+6cvL2zf/lChQopKSkpnzMCAAAAAABAbv3www/q2LGjjh49qqpVq8o0TX333XcKDw/X2rVrVbFixVzHdNkaVpGRkdqyZUu227Zs2aLy5cvnc0YAAAAAAADIrWHDhqlixYo6cuSIvvjiC+3Zs0eHDx9WZGSkhg0blqeYLitY9ezZUy+++KIWLVrk1B8XF6eZM2eqV69eLsoMAAAAAAAUSKabNpvbunWrpk2bphIlSjj6QkJCNHXqVG3dujVPMV1WsHr88cfVtGlTxcTEKDAwUJUrV1ZgYKBiYmLUtGlTjR071lWpAQAAAAAAIId8fX2zXdrpzz//lI+PT55iuqxg5ePjo02bNmnx4sXq2rWrIiMj1a1bNy1evFgffvhhng8IAAAAAAAA+adz58568MEH9dlnn8k0TZmmqR07dmjw4MHq0qVLnmK6bNF1SfLy8lLfvn3Vt29fV6YBAAAAAACAPHrppZfUr18/RUVFqXDhwpKk1NRUdenSRbNmzcpTTJcWrCTp22+/1datW3Xq1CnFxMSoTJkyOnbsmIoXLy5/f39XpwcAAAAAAAoIw8xo7sTu+ZqmqT/++EMrVqzQsWPH9M0338g0TdWoUUOVKlXKc1yXFazS0tL04IMPKi4uTqZpyjAMdejQQWXKlNGgQYNUv359TZw40VXpAQAAAAAA4F+YpqnKlStr3759qly58jUVqS7nsjWsnn32WS1fvlzPP/+8vv76a5nm3yXDDh06aMOGDa5KDQAAAAAAADlQqFAhVa5cWadPn7Y2rqXRciEuLk5PPfWURo4cqapVqzpti4yM1MGDB12UGQAAAAAAAHJq2rRpeuyxx/T1119bFtNlUwKPHj2qqKiobLf5+fll+3WIAAAAAAAAeWb+1dyJG+R777336vz586pbt658fHyyrEn+22+/5TqmywpWpUuX1k8//aRWrVpl2XbgwAGVK1fOBVkBAAAAAAAgN1588UUZhmFpTJcVrDp27Khnn31Wt912m8qUKSNJMgxDf/zxh1566SXdfvvtrkoNAAAAAAAAOdSrVy+lpqYqMDDQspguW8Nq4sSJSk1NVY0aNXTnnXfKMAyNGzdOtWrV0sWLF/XUU0+5KjUAAAAAAFAQmZLhZs3OUwJPnTqlTp06KSgoSMHBwWrSpIl++uknS2Lna8Hq119/dfw7NDRUO3fuVK9evbR79255eXnp//7v/9ShQwdt375dJUqUyM/UAAAAAAAAkAtjx47V7t27NWHCBD3//PM6deqUBg0aZEnsfJ0SeMMNN6hLly6KiYnRbbfdptDQUM2bNy8/UwAAAAAAAIAFNm7cqIULF6pjx46SMpZ/qlWrllJSUlS4cOFrip2vI6zuuusurV+/Xp07d1Z4eLiefPJJy4aKAQAAAAAAIP8cO3ZM9evXdzyuVq2afHx8dOzYsWuOna8FqzfeeEOJiYl65ZVXdMMNN2jy5MmqXLmyWrZsqWXLlunixYv5mQ4AAAAAAPAkpps2mzJNU97ezpP3vL29lZ6efs2x833R9eDgYA0ePFifffaZ9u3bp//85z86cOCA7rvvPpUpU0YPPfSQdu7cmd9pAQAAAAAAIBdM01SbNm100003Odr58+d1++23O/XlRb6uYfVP1atX1/Tp0/Xcc89p7dq1WrhwoRYuXKgFCxaoZs2a2rt3ryvTAwAAAAAAwBWMHz8+S1/Xrl0tie3SglUmLy8vdenSRbfccoumTZummTNnat++fa5OCwAAAAAAAFeQXcHKKi4vWKWlpem9997TokWLtGHDBqWmpqpOnTqKiYlxdWoAAAAAAKAgsfmaUNlyt3wt4rKC1b59+7Rw4UItW7ZMp06dUnBwsAYOHKiYmBg1aNDAVWkBAAAAAADAxfK1YHX27FktX75cCxcu1O7duyVJzZs3V0xMjHr06CE/P7/8TAcAAAAAAAA2lK8FqzJlyig5OVlly5bV448/rvvvv18VK1bMzxQAAAAAAICHMsyM5k7cLV+rFMrPF7vtttv03nvv6fDhw3r22WcpVgEAAAAAALi5JUuWKDk5OUv/pUuXtGTJkjzFzNeC1apVq9SpUycVKpSvLwsAAAAAAIDrZMCAAfrjjz+y9CclJWnAgAF5iknlCAAAAAAAAHlmmqYMw8jS/8svv6ho0aJ5iumybwkEAAAAAACA+6pfv74Mw5BhGGrTpo28vf8uM6WlpengwYO67bbb8hSbghUAAAAAAAByrVu3bpKkL7/8Uu3bt1dQUJBjm4+PjyIiInTnnXfmKTYFKwAAAAAAAOTa+PHjJUkRERG655575OfnZ1lsClYAAAAAAADIs379+kmSdu/erW+++UaGYahGjRqqX79+nmNSsAIAAAAAAJ7B/Ku5EzfI9+TJk+rZs6fi4+NVrFgxmaapP/74Q61atdKbb76pUqVK5Tom3xIIAAAAAACAPHvkkUd09uxZ7du3T7/99pt+//13ff311zp79qyGDRuWp5iMsAIAAAAAAECebdiwQR999JGqV6/u6KtRo4ZeeeUVtWvXLk8xKVgBAAAAAACPYJgZzZ24Q77p6ekqXLhwlv7ChQsrPT09TzGZEggAAAAAAIA8a926tYYPH65jx445+o4ePar//Oc/atOmTZ5iUrACAAAAAABAns2ePVtJSUmKiIhQxYoVValSJUVGRiopKUkvv/xynmIyJRAAAAAAAAB5Fh4eri+++EKbNm3St99+K9M0VaNGDbVt2zbPMSlYAQAAAAAAz+EGa0K5q+joaEVHR1sSi4IVAAAAAAAA8iQ9PV1xcXFatWqVDh06JMMwFBkZqR49eqhv374yDCNPcVnDCgAAAAAAALlmmqa6dOmigQMH6ujRo6pdu7Zq1qypn3/+Wf3799cdd9yR59iMsAIAAAAAAECuxcXF6X//+58+/vhjtWrVymnb5s2b1a1bNy1ZskT33XdfrmMzwgoAAAAAAHgG002bTa1YsULjxo3LUqySpNatW+vxxx/XG2+8kafYFKwAAAAAAACQa3v37tVtt912xe0dOnTQ//3f/+UpNgUrAAAAAACAAmbOnDmKjIyUn5+fGjRooG3btl1x3/j4eBmGkaV9++23V32N3377TaGhoVfcHhoaqt9//z1P+bOGFQAAAAAA8AiGmdHcSV7yXblypUaMGKE5c+aoadOmmj9/vjp06KD9+/frxhtvvOLzDhw4oODgYMfjUqVKXfV10tLS5O195dKSl5eXUlNTc38AomAFAAAAAABQoMyYMUMxMTEaOHCgJGnmzJnauHGj5s6dqylTplzxeaVLl1axYsVy/Dqmaap///7y9fXNdntycnKu8r4cBSsAAAAAAACbO3v2rNNjX1/fbAtFly5d0u7du/X444879bdr107bt2+/6mvUr19fFy9eVI0aNfTkk09mu5j65fr16/eveeflGwIlClYAAAAAAAC2Fx4e7vR4/Pjxio2NzbLfqVOnlJaWlmVtqdDQUB0/fjzb2GXLltWCBQvUoEEDJScna+nSpWrTpo3i4+PVvHnzK+a0aNGi3B9IDlGwAgAAAAAAnsH8q7mTv/I9cuSI0/pSV5qGl8kwDOcwppmlL1PVqlVVtWpVx+OoqCgdOXJE06dPv2rB6nriWwIBAAAAAABsLjg42KldqWBVsmRJeXl5ZRlNdfLkyat+o98/NW7cWN9///015XwtKFgBAAAAAAAUED4+PmrQoIE2bdrk1L9p0yY1adIkx3H27NmjsmXLWp1ejjElEAAAAAAAoAAZOXKk+vbtq4YNGyoqKkoLFizQ4cOHNXjwYEnS2LFjdfToUS1ZskRSxrcIRkREqGbNmrp06ZKWLVumd955R++8847LjoGCFQAAAAAA8AiGmdHcSV7yveeee3T69GlNnDhRiYmJqlWrltatW6fy5ctLkhITE3X48GHH/pcuXdKjjz6qo0ePyt/fXzVr1tTatWvVsWNHqw4j1yhYAQAAAAAAFDBDhgzRkCFDst0WFxfn9Hj06NEaPXp0PmSVc6xhBQAAAAAAAFthhBUAAAAAAPAM5l/NnbhbvhZhhBUAAAAAAABshYIVAAAAAAAAbIWCFQAAAAAAAGyFNawAAAAAAIBnYA0rt8EIKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AprWAEAAAAAAI9gmBnNnbhbvlZhhBUAAAAAAABshYIVAAAAAAAAbIUpgQAAAAAAwDOYfzV34m75WoQRVgAAAAAAALAVClYAAAAAAACwFQpWAAAAAAAAsBXWsAIAAAAAAJ6BNazcBiOsAAAAAAAAYCsUrAAAAAAAAGArFKwAAAAAAABgK6xhBQAAAAAAPIJhZjR34m75WoURVgAAAAAAALAVClYAAAAAAACwFaYEAgAAAAAAz2D+1dyJu+VrEUZYAQAAAAAAwFYoWAEAAAAAAMBWKFgBAAAAAADAVljDCgAAAAAAeATDzGjuxN3ytQojrAAAAAAAAGArFKwAAAAAAABgKxSsAAAAAAAAYCusYQUAAAAAADyD+VdzJ+6Wr0UYYQUAAAAAAABboWAFAAAAAAAAW2FKIAAAAAAA8AxMCXQbjLACAAAAAACArVCwAgAAAAAAgK1QsAIAAAAAAICtsIYVAAAAAADwCMZfzZ24W75WYYQVAAAAAAAAbKXAFKzmzp2rOnXqKDg4WMHBwYqKitL69esd203TVGxsrMLCwuTv76+WLVtq3759LswYAAAAAAAA2SkwBaty5cpp6tSp2rVrl3bt2qXWrVura9eujqLUtGnTNGPGDM2ePVs7d+5UmTJlFB0draSkJBdnDgAAAAAAgMsVmILV7bffro4dO6pKlSqqUqWKnn32WQUFBWnHjh0yTVMzZ87UE088oe7du6tWrVpavHixzp8/r+XLl7s6dQAAAAAAkB9MN20eqMAUrC6XlpamN998U+fOnVNUVJQOHjyo48ePq127do59fH191aJFC23fvv2KcZKTk3X27FmnBgAAAAAAgOurQBWsvvrqKwUFBcnX11eDBw/W6tWrVaNGDR0/flySFBoa6rR/aGioY1t2pkyZoqJFizpaeHj4dc0fAAAAAAAABaxgVbVqVX355ZfasWOHHnroIfXr10/79+93bDcM5y+DNE0zS9/lxo4dqz/++MPRjhw5ct1yBwAAAAAAQAZvVydgJR8fH1WqVEmS1LBhQ+3cuVOzZs3SmDFjJEnHjx9X2bJlHfufPHkyy6iry/n6+srX1/f6Jg0AAAAAAPKFYWY0d+Ju+VqlQI2w+ifTNJWcnKzIyEiVKVNGmzZtcmy7dOmStm7dqiZNmrgwQwAAAAAAAPxTgRlhNW7cOHXo0EHh4eFKSkrSm2++qfj4eG3YsEGGYWjEiBGaPHmyKleurMqVK2vy5MkKCAhQ7969XZ06AAAAAAAALlNgClYnTpxQ3759lZiYqKJFi6pOnTrasGGDoqOjJUmjR4/WhQsXNGTIEP3++++65ZZb9OGHH6pIkSIuzhwAAAAAAOQL86/mTtwtX4sUmILV66+/ftXthmEoNjZWsbGx+ZMQAAAAAAAA8qRAr2EFAAAAAAAA90PBCgAAAAAAALZSYKYEAgAAAAAA/CsPXRPK3VCwyoOjD6bKKyDVklhhxX+xJI4kNRk52LJYklR2oHW5oWD66bFqrk7hqpqUrWtZrMDEZMtiXQ+edKxWqnLylKXxmuy29jrsKcruSrQ0XpPf+DkURNtnzLM0Xoe1zSyLZfU9GGziFlcnkH8CTlyyNJ5ZNsSyWOfLBVkWS5KCdv1sabyUyDKWxUoNtObX89RUqjEoGJgSCAAAAAAAAFuhYAUAAAAAAABbYUogAAAAAADwCIaZ0dyJu+VrFUZYAQAAAAAAwFYoWAEAAAAAAMBWmBIIAAAAAAA8g/lXcyfulq9FGGEFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABbYQ0rAAAAAADgEQwzo7kTd8vXKoywAgAAAAAAgK1QsAIAAAAAAICtULACAAAAAACArbCGFQAAAAAA8AzmX82duFu+FmGEFQAAAAAAAGyFghUAAAAAAABshSmBAAAAAADAIxhmRnMn7pavVRhhBQAAAAAAAFuhYAUAAAAAAABbYUpgHkRMviRvL8OaYIlJ1sSR5Ju237JYkqSNXtbGQ4FjXjjp6hSuqpiPj2WxzEuXLIt1PXjSsVopPS3d0njFjp6wNJ7HKGzt7UixdactjQd76LC2maXx1h/YZlmsjtUsui+ErZg3hrk6hXxjJFv72X+0cxnLYpVbc9SyWJJknr9gabxCKWmWxfI/YM3nV2p6siVxAFejYAUAAAAAADyD+VdzJ+6Wr0WYEggAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFthDSsAAAAAAOAZWMPKbTDCCgAAAAAAALZCwQoAAAAAAAC2wpRAAAAAAADgEQwzo7kTd8vXKoywAgAAAAAAgK1QsAIAAAAAAICtULACAAAAAACArbCGFQAAAAAA8AzmX82duFu+FmGEFQAAAAAAAGyFghUAAAAAAABshYIVAAAAAAAAbIU1rAAAAAAAgEcwTFOG6V6LQrlbvlZhhBUAAAAAAABshYIVAAAAAAAAbIUpgXmQUjJAprefJbGOdi9pSRxJ2j9kjmWxJKnGnCGWxkPB06zLHlencFXzy/3PsliDfomyLNb14EnHaqWPtzawNN4PfeZaGs9T1J9k7efNniet/TyEPTQZOdjSeB2rGZbFWvetdddg2EenJl1cnUK+SY6w7ncSSbph2feWxUq/oZRlsSQppXyIpfF8fz5tWay00GLWxEm7KP1iSaiCyfyruRN3y9cijLACAAAAAACArVCwAgAAAAAAgK1QsAIAAAAAAICtsIYVAAAAAADwCIaZ0dyJu+VrFUZYAQAAAAAAFDBz5sxRZGSk/Pz81KBBA23bti1Hz/v000/l7e2tevXqXd8E/wUFKwAAAAAAgAJk5cqVGjFihJ544gnt2bNHzZo1U4cOHXT48OGrPu+PP/7QfffdpzZt2uRTpldGwQoAAAAAAKAAmTFjhmJiYjRw4EBVr15dM2fOVHh4uObOnXvV5w0aNEi9e/dWVFRUPmV6ZRSsAAAAAACAZzDdtOXCpUuXtHv3brVr186pv127dtq+ffsVn7do0SL9+OOPGj9+fO5e8Dph0XUAAAAAAACbO3v2rNNjX19f+fr6Ztnv1KlTSktLU2hoqFN/aGiojh8/nm3s77//Xo8//ri2bdsmb297lIoYYQUAAAAAAGBz4eHhKlq0qKNNmTLlqvsbhuH02DTNLH2SlJaWpt69e2vChAmqUqWKpTlfC3uUzQAAAAAAAK4zw8xo7iQz3yNHjig4ONjRn93oKkkqWbKkvLy8soymOnnyZJZRV5KUlJSkXbt2ac+ePXr44YclSenp6TJNU97e3vrwww/VunVri44m5yhYAQAAAAAA2FxwcLBTwepKfHx81KBBA23atEl33HGHo3/Tpk3q2rVrtnG/+uorp745c+Zo8+bNevvttxUZGXntyecBBSsAAAAAAIACZOTIkerbt68aNmyoqKgoLViwQIcPH9bgwYMlSWPHjtXRo0e1ZMkSFSpUSLVq1XJ6funSpeXn55elPz9RsAIAAAAAAChA7rnnHp0+fVoTJ05UYmKiatWqpXXr1ql8+fKSpMTERB0+fNjFWV4dBSsAAAAAAOAZzL+aO8ljvkOGDNGQIUOy3RYXF3fV58bGxio2NjZvL2wRviUQAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whpWAAAAAADAIxhmRnMn7pavVRhhBQAAAAAAAFuhYAUAAAAAAABbYUogAAAAAADwDOZfzZ24W74WYYQVAAAAAAAAbIWCFQAAAAAAAGyFghUAAAAAAABshTWsAAAAAACAxzA8dE0od8MIKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AprWAEAAAAAAM9gmhnNnbhbvhZhhBUAAAAAAABshYIVAAAAAAAAbIUpgQAAAAAAwCMYZkZzJ+6Wr1UYYQUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFthDSsAAAAAAOAZzL+aO3G3fC3CCCsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKa1gBAAAAAACPYKRnNHfibvlahRFWAAAAAAAAsBUKVgAAAAAAALAVClYAAAAAAACwFdawAgAAAAAAnsH8q7kTd8vXIoywAgAAAAAAgK1QsAIAAAAAAICtMCUQAAAAAAB4BMPMaO7E3fK1CiOsAAAAAAAAYCsUrAAAAAAAAGArFKwAAAAAAABgK6xhBQAAAAAAPINpZjR34m75WoQRVgAAAAAAALAVClYAAAAAAACwFQpWAAAAAAAAsBXWsAIAAAAAAB7BMDOaO3G3fK3CCCsAAAAAAADYCgUrAAAAAAAA2ApTAgEAAAAAgGcw/2ruxN3ytQgjrAAAAAAAAGArFKwAAAAAAABgKxSsAAAAAAAAYCusYQUAAAAAADyCYWY0d+Ju+VqFEVYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAV1rACAAAAAACewTQzmjtxt3wtwggrAAAAAAAA2AoFKwAAAAAAANgKUwIBAAAAAIBHMMyM5k7cLV+rMMIKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whpWAAAAAADAM5h/NXfibvlahBFWAAAAAAAAsBUKVgAAAAAAALAVClYAAAAAAACwFdawAgAAAAAAHsEwM5o7cbd8rcIIKwAAAAAAANgKBSsAAAAAAADYClMCAQAAAACAZ0g3M5o7cbd8LcIIKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AprWAEAAAAAAM9g/tXcibvlaxFGWAEAAAAAAMBWKFgBAAAAAADAVihYAQAAAAAAwFZYwwoAAAAAAHgEQ5LhZmtCGa5OwEUYYQUAAAAAAABboWAFAAAAAAAAW2FKIAAAAAAA8AymmdHcibvlaxFGWAEAAAAAAMBWKFgBAAAAAADAVihYAQAAAAAAwFZYwwoAAAAAAHgEw8xo7sTd8rUKI6wAAAAAAABgKxSsAAAAAAAAYCsUrAAAAAAAAGArrGEFAAAAAAA8g/lXcyfulq9FGGEFAAAAAAAAW6FgBQAAAAAAAFthSiAAAAAAAPAIhmnKMN1rjp275WsVRlgBAAAAAADAVihYAQAAAAAAwFYoWAEAAAAAAMBWWMMKAAAAAAB4hvS/mjtxt3wtwggrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCmtYAQAAAAAAj2CYpgzTdHUaueJu+VqFEVYAAAAAAACwFQpWAAAAAAAAsBWmBAIAAAAAAM9g/tXcibvlaxFGWAEAAAAAABQwc+bMUWRkpPz8/NSgQQNt27btivt+8sknatq0qUJCQuTv769q1arpxRdfzMdss2KEFQAAAAAAQAGycuVKjRgxQnPmzFHTpk01f/58dejQQfv379eNN96YZf/AwEA9/PDDqlOnjgIDA/XJJ59o0KBBCgwM1IMPPuiCI2CEFQAAAAAAQIEyY8YMxcTEaODAgapevbpmzpyp8PBwzZ07N9v969evr169eqlmzZqKiIjQvffeq/bt2191VNb1xgirPDgX6ivvwr6WxCrxbZolcSSpU5MulsWSpBINrcsNBdMvi0NcncJVdfKy8JxIS7cu1nXgScdqpSpFTlsar9Mr1l6HPUWYfrE0Xqd1/BwKpFusDWfeGGZZLKvvwWAPa7e/5+oU8k2NuUMsjRcUUcmyWAEnUi2LJUk+Z1MsjXemUVnLYvmdsia31FQPXfAop0wzo7mTv/I9e/asU7evr698fbPWJi5duqTdu3fr8ccfd+pv166dtm/fnqOX3LNnj7Zv365JkyblMelrxwgrAAAAAAAAmwsPD1fRokUdbcqUKdnud+rUKaWlpSk0NNSpPzQ0VMePH7/qa5QrV06+vr5q2LChhg4dqoEDB1qWf24xwgoAAAAAAMDmjhw5ouDgYMfj7EZXXc4wDKfHpmlm6funbdu26c8//9SOHTv0+OOPq1KlSurVq1fek74GFKwAAAAAAABsLjg42KlgdSUlS5aUl5dXltFUJ0+ezDLq6p8iIyMlSbVr19aJEycUGxvrsoIVUwIBAAAAAIBHMEz3bLnh4+OjBg0aaNOmTU79mzZtUpMmTXIcxzRNJScn5+7FLcQIKwAAAAAAgAJk5MiR6tu3rxo2bKioqCgtWLBAhw8f1uDBgyVJY8eO1dGjR7VkyRJJ0iuvvKIbb7xR1apVkyR98sknmj59uh555BGXHQMFKwAAAAAAgALknnvu0enTpzVx4kQlJiaqVq1aWrduncqXLy9JSkxM1OHDhx37p6ena+zYsTp48KC8vb1VsWJFTZ06VYMGDXLVIVCwAgAAAAAAHsI0M5o7yWO+Q4YM0ZAhQ7LdFhcX5/T4kUceceloquywhhUAAAAAAABshYIVAAAAAAAAbIWCFQAAAAAAAGyFNawAAAAAAIBHMNIzmjtxt3ytwggrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCmtYAQAAAAAAz2CaGc2duFu+FmGEFQAAAAAAAGyFghUAAAAAAABshYIVAAAAAAAAbIU1rAAAAAAAgGcw/2ruxN3ytQgjrAAAAAAAAGArFKwAAAAAAABgK0wJBAAAAAAAHsEwTRmme82xc7d8rcIIKwAAAAAAANgKBSsAAAAAAADYSoEsWE2ZMkWGYWjEiBGOPtM0FRsbq7CwMPn7+6tly5bat2+f65IEAAAAAABAtgpcwWrnzp1asGCB6tSp49Q/bdo0zZgxQ7Nnz9bOnTtVpkwZRUdHKykpyUWZAgAAAACAfGWa7tk8UIEqWP3555/q06ePXn31VRUvXtzRb5qmZs6cqSeeeELdu3dXrVq1tHjxYp0/f17Lly93YcYAAAAAAAD4pwJVsBo6dKg6deqktm3bOvUfPHhQx48fV7t27Rx9vr6+atGihbZv357faQIAAAAAAOAqvF2dgFXefPNNffHFF9q5c2eWbcePH5ckhYaGOvWHhobq559/vmLM5ORkJScnOx6fPXvWomwBAAAAAABwJQVihNWRI0c0fPhwLVu2TH5+flfczzAMp8emaWbpu9yUKVNUtGhRRwsPD7csZwAAAAAAkM9MSelu1jxzCauCUbDavXu3Tp48qQYNGsjb21ve3t7aunWrXnrpJXl7eztGVmWOtMp08uTJLKOuLjd27Fj98ccfjnbkyJHrehwAAAAAAAAoIFMC27Rpo6+++sqpb8CAAapWrZrGjBmjChUqqEyZMtq0aZPq168vSbp06ZK2bt2q55577opxfX195evre11zBwAAAAAAgLMCUbAqUqSIatWq5dQXGBiokJAQR/+IESM0efJkVa5cWZUrV9bkyZMVEBCg3r17uyJlAAAAAACQzwzTlGG61xw7d8vXKgWiYJUTo0eP1oULFzRkyBD9/vvvuuWWW/Thhx+qSJEirk4NAAAAAAAAlymwBav4+Hinx4ZhKDY2VrGxsS7JBwAAAAAAADlTIBZdBwAAAAAAQMFRYEdYAQAAAAAAODEluduaUG6WrlUYYQUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFthDSsAAAAAAOAZTNMN17Bys3wtwggrAAAAAAAA2AoFKwAAAAAAANgKUwIBAAAAAIBnSJdkuDqJXEp3dQKuwQgrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCmtYAQAAAAAAj2CYpgzTdHUaueJu+VqFEVYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAV1rACAAAAAACewTQzmjtxt3wtwggrAAAAAAAA2AoFKwAAAAAAANgKUwIBAAAAAIBnYEqg22CEFQAAAAAAAGyFghUAAAAAAABshYIVAAAAAAAAbIU1rPLA77cUeXt7WRKr8NlLlsSRpOSIkpbFkqTAw+ctjYeC51KFUq5O4apMw7AslmHzeeOedKxW8j5z0dJ4lyy+DnsKIzXd0nimN3+PK4gCTlh3zyRJRrJ978FgDzXmDnF1Cvlm/0NzLI3X6v4HLIuV7m3dPY4kJRf3sTSez9k0y2Kl+Vnz+ZWWwufgVbGGldvgfzIAAAAAAABshYIVAAAAAAAAbIWCFQAAAAAAAGyFNawAAAAAAIBnSJdk7dJo15+1y326DUZYAQAAAAAAwFYoWAEAAAAAAMBWmBIIAAAAAAA8gmGaMkzT1WnkirvlaxVGWAEAAAAAAMBWKFgBAAAAAADAVihYAQAAAAAAwFZYwwoAAAAAAHgG08xo7sTd8rUII6wAAAAAAABgKxSsAAAAAAAAYCsUrAAAAAAAAGArrGEFAAAAAAA8Q7opGW62JlS6m+VrEUZYAQAAAAAAwFYoWAEAAAAAAMBWmBIIAAAAAAA8g2lmNHfibvlahBFWAAAAAAAAsBVGWOWB97lUeXunWhKr0MUUS+JIktfB45bFkiSzbIil8VAAff6VqzO4Kq9SpSyLlfbrr5bFuh486VitZJS09jrnc4y/A+WFEeBnaTzz/EVL48EerL4vOdq5jGWxblj2vWWxYB9BEZVcnUK+aXX/A5bG27LwVctidejQy7JYkmRcsub3uExn6lh3bSr2xUlL4qSmJVsSB3A17qwBAAAAAABgK4ywAgAAAAAAHsIN17CSu+VrDUZYAQAAAAAAwFYoWAEAAAAAAMBWKFgBAAAAAADAVljDCgAAAAAAeAbTDdewcrd8LcIIKwAAAAAAANgKBSsAAAAAAADYClMCAQAAAACAZ0g3JbnZFLt0N8vXIoywAgAAAAAAgK1QsAIAAAAAAICtULACAAAAAACArbCGFQAAAAAA8AxmekZzJ+6Wr0UYYQUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFthDSsAAAAAAOAZTDOjuRN3y9cijLACAAAAAACArTDCKg+SIvzl5eNnSayAE9b9CHwLWVt/TC7pb2k8FDz+FSJcncJVpQcHWBbLu0igZbGuB086ViuZSeesjXdDKUvjeYzkFEvDmcWLWBoP9nC+XJCl8cqtOWpZrHTO/QIp4ESqq1PIN+nehqXxOnToZVms9etXWBZLkqLv7m9pvFR/6967i5ElLImTmnpR+tGSUIBLUbACAAAAAACeId2U5GZT7NLdLF+LMCUQAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whpWAAAAAADAM5hmRnMn7pavRRhhBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAPIOpv9excpuWt0OdM2eOIiMj5efnpwYNGmjbtm1X3HfVqlWKjo5WqVKlFBwcrKioKG3cuDFvL2wRClYAAAAAAAAFyMqVKzVixAg98cQT2rNnj5o1a6YOHTro8OHD2e7/v//9T9HR0Vq3bp12796tVq1a6fbbb9eePXvyOfO/UbACAAAAAAAoQGbMmKGYmBgNHDhQ1atX18yZMxUeHq65c+dmu//MmTM1evRoNWrUSJUrV9bkyZNVuXJlvf/++/mc+d8oWAEAAAAAANjc2bNnnVpycnK2+126dEm7d+9Wu3btnPrbtWun7du35+i10tPTlZSUpBIlSlxz3nlFwQoAAAAAAHgGl69HlccmKTw8XEWLFnW0KVOmZHuIp06dUlpamkJDQ536Q0NDdfz48Ry9TS+88ILOnTunu++++9re72vg7bJXBgAAAAAAQI4cOXJEwcHBjse+vr5X3d8wDKfHpmlm6cvOihUrFBsbq3fffVelS5fOW7IWoGAFAAAAAABgc8HBwU4FqyspWbKkvLy8soymOnnyZJZRV/+0cuVKxcTE6L///a/atm17TfleK6YEAgAAAAAAz5Ce7p4tF3x8fNSgQQNt2rTJqX/Tpk1q0qTJFZ+3YsUK9e/fX8uXL1enTp3y9PZaiRFWAAAAAAAABcjIkSPVt29fNWzYUFFRUVqwYIEOHz6swYMHS5LGjh2ro0ePasmSJZIyilX33XefZs2apcaNGztGZ/n7+6to0aIuOQYKVgAAAAAAAAXIPffco9OnT2vixIlKTExUrVq1tG7dOpUvX16SlJiYqMOHDzv2nz9/vlJTUzV06FANHTrU0d+vXz/FxcXld/qSKFgBAAAAAAAUOEOGDNGQIUOy3fbPIlR8fPz1TyiXKFjlQfE9p+XtdfXV+HMsl3NRr+bSDcUsiyVJfj//bmk8FDxpJYJcncJVpQb5WBbL29veS/550rFaqmigpeFSi1r02eBhfH/4w9J4l/5lMVG4p6BdP1sazzx/wbJYKeVDLIsF+/A5m+LqFPJNcnHr7iMkybiUalms6Lv7WxZLkja9FWdpvHpTsy8G5IVfwhlL4qSmJVsSp8AyzYzmTtwtX4t40G8lAAAAAAAAcAcUrAAAAAAAAGArFKwAAAAAAABgK6xhBQAAAAAAPANrWLkNRlgBAAAAAADAVihYAQAAAAAAwFaYEggAAAAAADxDuinJzabYpbtZvhZhhBUAAAAAAABshYIVAAAAAAAAbIWCFQAAAAAAAGyFNawAAAAAAIBHMM10mWa6q9PIFXfL1yqMsAIAAAAAAICtULACAAAAAACArVCwAgAAAAAAgK2whhUAAAAAAPAMpimlm67OIndMN8vXIoywAgAAAAAAgK1QsAIAAAAAAICtMCUQAAAAAAB4BtOU5GZT7JgSCAAAAAAAALgeI6zyIKl6CXkX9rMklve5NEviSFLAd79aFkuSzlcpZWk8FDz+n//o6hSuyrdYsGWxzDNnLYt1PXjSsVrJsPB9kySvM39aGs9TpP9h7f8535+5vSmIUiLLWBqvUIp192C+P5+2LBbs40yjsq5OId/4nLXufJCkM3VCLIuV6m9YFkuS6k0dYmm8Lx+fY1msJicGWxInNeWi9L0loQCXYoQVAAAAAAAAbIU/QQIAAAAAAM+Qni4Z6a7OIndMN8vXIoywAgAAAAAAgK1QsAIAAAAAAICtULACAAAAAACArbCGFQAAAAAA8AymKcl0dRa5Y7pZvhZhhBUAAAAAAABshYIVAAAAAAAAbIUpgQAAAAAAwCOY6ekyjXRXp5Erpule+VqFEVYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAV1rACAAAAAACewTQlma7OIndMN8vXIoywAgAAAAAAgK1QsAIAAAAAAICtULACAAAAAACArbCGFQAAAAAA8AzppmS42ZpQrGEFAAAAAAAAuB4jrPLASDNlFLKmwlko1bpKaVpIEctiSdbmhgKqbClXZ3BVaQE+lsUq5O9rWazrwZOO1Urp3tb+3cb04WM1Lwr5Wff/V5LSAj3n/7AnSQ209vzyP3DaslhpocUsiwX78DuV4uoU8k2an7Wfh8W+OGlZrIuRJSyLJUl+CWcsjdfkxGDLYm1/cZ4lcc4mpav4O5aEAlyKO2sAAAAAAOAZTFNSuquzyB2mBAIAAAAAAACuR8EKAAAAAAAAtkLBCgAAAAAAALbCGlYAAAAAAMAjmOmmTMO91oQyWcMKAAAAAAAAcD0KVgAAAAAAALAVClYAAAAAAACwFdawAgAAAAAAnsFMl5Tu6ixyx3SzfC3CCCsAAAAAAADYCgUrAAAAAAAA2ApTAgEAAAAAgEcw002ZhunqNHLFNN0rX6swwgoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALbCGlYAAAAAAMAzmOmS0l2dRe6YbpavRRhhBQAAAAAAAFuhYAUAAAAAAABbYUpgLmR+lWRqykXLYhqpadbFSku2LJYkpaZ65ldnIue8LP4/Z7X0NOuGzhZKS7Es1vXgScdqJdOw9u82ZlqqpfE8RaG0S5bGS0/j86sgsvq+JDXdus+wtDTr7g1hH550L5yWYu3nYaqF94ipqdaeX1bmJln7u+HZJGvu587+mREn8/dXOEtViuRmb02qPOf+/HKGyf/iHPvll18UHh7u6jQAAAAAALiqI0eOqFy5cq5OwzYuXryoyMhIHT9+3NWp5EmZMmV08OBB+fn5uTqVfEPBKhfS09N17NgxFSlSRIZhXHG/s2fPKjw8XEeOHFFwcHA+ZgjYD+cD8DfOB8AZ5wTwN84HWMU0TSUlJSksLEyFCrEK0OUuXryoS5esHdmdX3x8fDyqWCUxJTBXChUqlKsKdXBwMB82wF84H4C/cT4AzjgngL9xPsAKRYsWdXUKtuTn5+dxRR93RrkVAAAAAAAAtkLBCgAAAAAAALZCweo68PX11fjx4+Xr6+vqVACX43wA/sb5ADjjnAD+xvkAAM5YdB0AAAAAAAC2wggrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFK4vNmTNHkZGR8vPzU4MGDbRt2zZXpwTki//973+6/fbbFRYWJsMwtGbNGqftpmkqNjZWYWFh8vf3V8uWLbVv3z7XJAtcR1OmTFGjRo1UpEgRlS5dWt26ddOBAwec9uF8gCeZO3eu6tSpo+DgYAUHBysqKkrr1693bOd8gCebMmWKDMPQiBEjHH2cEwCQgYKVhVauXKkRI0boiSee0J49e9SsWTN16NBBhw8fdnVqwHV37tw51a1bV7Nnz852+7Rp0zRjxgzNnj1bO3fuVJkyZRQdHa2kpKR8zhS4vrZu3aqhQ4dqx44d2rRpk1JTU9WuXTudO3fOsQ/nAzxJuXLlNHXqVO3atUu7du1S69at1bVrV8cv4JwP8FQ7d+7UggULVKdOHad+zgkAyGCYpmm6OomC4pZbbtFNN92kuXPnOvqqV6+ubt26acqUKS7MDMhfhmFo9erV6tatm6SMvxSGhYVpxIgRGjNmjCQpOTlZoaGheu655zRo0CAXZgtcX7/++qtKly6trVu3qnnz5pwPgKQSJUro+eef1/3338/5AI/0559/6qabbtKcOXM0adIk1atXTzNnzuQzAgAuwwgri1y6dEm7d+9Wu3btnPrbtWun7du3uygrwB4OHjyo48ePO50fvr6+atGiBecHCrw//vhDUsYv6BLnAzxbWlqa3nzzTZ07d05RUVGcD/BYQ4cOVadOndS2bVunfs4JAPibt6sTKChOnTqltLQ0hYaGOvWHhobq+PHjLsoKsIfMcyC78+Pnn392RUpAvjBNUyNHjtStt96qWrVqSeJ8gGf66quvFBUVpYsXLyooKEirV69WjRo1HL+Acz7Ak7z55pv64osvtHPnzizb+IwAgL9RsLKYYRhOj03TzNIHeCrOD3iahx9+WHv37tUnn3ySZRvnAzxJ1apV9eWXX+rMmTN655131K9fP23dutWxnfMBnuLIkSMaPny4PvzwQ/n5+V1xP84JAGBKoGVKliwpLy+vLKOpTp48meUvJICnKVOmjCRxfsCjPPLII3rvvfe0ZcsWlStXztHP+QBP5OPjo0qVKqlhw4aaMmWK6tatq1mzZnE+wOPs3r1bJ0+eVIMGDeTt7S1vb29t3bpVL730kry9vR3/7zknAICClWV8fHzUoEEDbdq0yal/06ZNatKkiYuyAuwhMjJSZcqUcTo/Ll26pK1bt3J+oMAxTVMPP/ywVq1apc2bNysyMtJpO+cDkHGeJCcncz7A47Rp00ZfffWVvvzyS0dr2LCh+vTpoy+//FIVKlTgnACAvzAl0EIjR45U37591bBhQ0VFRWnBggU6fPiwBg8e7OrUgOvuzz//1A8//OB4fPDgQX355ZcqUaKEbrzxRo0YMUKTJ09W5cqVVblyZU2ePFkBAQHq3bu3C7MGrDd06FAtX75c7777rooUKeL4K3nRokXl7+8vwzA4H+BRxo0bpw4dOig8PFxJSUl68803FR8frw0bNnA+wOMUKVLEsaZhpsDAQIWEhDj6OScAIAMFKwvdc889On36tCZOnKjExETVqlVL69atU/ny5V2dGnDd7dq1S61atXI8HjlypCSpX79+iouL0+jRo3XhwgUNGTJEv//+u2655RZ9+OGHKlKkiKtSBq6LuXPnSpJatmzp1L9o0SL1799fkjgf4FFOnDihvn37KjExUUWLFlWdOnW0YcMGRUdHS+J8AP6JcwIAMhimaZquTgIAAAAAAADIxBpWAAAAAAAAsBUKVgAAAAAAALAVClYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAVClYAAAAAAACwFQpWAAAAAAAAsBUKVgAAuJk77rhD/v7+OnPmzBX36dOnjwoXLqwTJ05Y+tqTJ0/WmjVrLI35bwzD0MMPP2xpzJYtW6ply5aWxgQAAIB1KFgBAOBmYmJidPHiRS1fvjzb7X/88YdWr16tzp07KzQ01NLXdkXBCgAAAJ6HghUAAG6mQ4cOCgsL08KFC7PdvmLFCl24cEExMTH5nFnepKSkKDU11dVpAAAAwEYoWAEA4Ga8vLzUr18/7d69W1999VWW7YsWLVLZsmXVoUMHSdLx48c1aNAglStXTj4+PoqMjNSECROyFImSk5M1ceJEVa9eXX5+fgoJCVGrVq20fft2SRlT886dO6fFixfLMAwZhuE0re7rr79W165dVbx4cfn5+alevXpavHix02vEx8fLMAwtXbpUo0aN0g033CBfX1/98MMPOT7+zBgrVqzQE088obCwMAUHB6tt27Y6cOCA076maWratGkqX768/Pz8dNNNN2n9+vXZxj179qweffRRRUZGysfHRzfccINGjBihc+fOOfYZPHiw/Pz8tHv3bkdfenq62rRpo9DQUCUmJub4OAAAAHBl3q5OAAAA5N7999+vqVOnauHChXrxxRcd/fv379fnn3+uxx9/XF5eXjp+/LhuvvlmFSpUSE8//bQqVqyohIQETZo0SYcOHdKiRYskSampqerQoYO2bdumESNGqHXr1kpNTdWOHTt0+PBhNWnSRAkJCWrdurVatWqlp556SpIUHBwsSTpw4ICaNGmi0qVL66WXXlJISIiWLVum/v3768SJExo9erRT/mPHjlVUVJTmzZunQoUKqXTp0rl+D8aNG6emTZvqtdde09mzZzVmzBjdfvvt+uabb+Tl5SVJmjBhgiZMmKCYmBj16NFDR44c0QMPPKC0tDRVrVrVEev8+fNq0aKFfvnlF40bN0516tTRvn379PTTT+urr77SRx99JMMwNHPmTH322We6++67tXv3bhUrVkwTJkxQfHy8NmzYoLJly+b6OAAAAJANEwAAuKUWLVqYJUuWNC9duuToGzVqlCnJ/O6770zTNM1BgwaZQUFB5s8//+z03OnTp5uSzH379pmmaZpLliwxJZmvvvrqVV8zMDDQ7NevX5b+nj17mr6+vubhw4ed+jt06GAGBASYZ86cMU3TNLds2WJKMps3b57j45RkDh061PE4M0bHjh2d9nvrrbdMSWZCQoJpmqb5+++/m35+fuYdd9zhtN+nn35qSjJbtGjh6JsyZYpZqFAhc+fOnU77vv3226Ykc926dY6+77//3gwODja7detmfvTRR2ahQoXMJ598MsfHAwAAgH/HlEAAANxUTEyMTp06pffee09SxiipZcuWqVmzZqpcubIk6YMPPlCrVq0UFham1NRUR8ucLrh161ZJ0vr16+Xn56f7778/T7ls3rxZbdq0UXh4uFN///79df78eSUkJDj133nnnXl6nct16dLF6XGdOnUkST///LMkKSEhQRcvXlSfPn2c9mvSpInKly/v1PfBBx+oVq1aqlevntP71L59exmGofj4eMe+lSpV0quvvqo1a9aoc+fOatasmWJjY6/5eAAAAPA3ClYAALipHj16qGjRoo5pfevWrdOJEyecFls/ceKE3n//fRUuXNip1axZU5J06tQpSdKvv/6qsLAwFSqUt1uD06dPZzsdLiwszLH9clZMnQsJCXF67OvrK0m6cOGC02uWKVMmy3P/2XfixAnt3bs3y/tUpEgRmabpeJ8yderUSaGhobp48aJGjhzpmIIIAAAAa7CGFQAAbsrf31+9evXSq6++qsTERC1cuFBFihTRXXfd5dinZMmSqlOnjp599tlsY2QWlEqVKqVPPvlE6enpeSpahYSEZLvg+LFjxxx5XM4wjFy/Rl5ykjIWnf+n48ePKyIiwvG4ZMmS8vf3v+I3L/4z/8GDByspKUk1a9bUsGHD1KxZMxUvXty65AEAADwcI6wAAHBjMTExSktL0/PPP69169apZ8+eCggIcGzv3Lmzvv76a1WsWFENGzbM0jILVh06dNDFixcVFxd31dfz9fV1jGC6XJs2bbR582ZHgSrTkiVLFBAQoMaNG1/7weZS48aN5efnpzfeeMOpf/v27Y5pg5k6d+6sH3/8USEhIdm+T5cXt1577TUtW7ZMs2fP1nvvvaczZ85owIAB+XFIAAAAHoMRVgAAuLGGDRuqTp06mjlzpkzTdJoOKEkTJ07Upk2b1KRJEw0bNkxVq1bVxYsXdejQIa1bt07z5s1TuXLl1KtXLy1atEiDBw/WgQMH1KpVK6Wnp+uzzz5T9erV1bNnT0lS7dq1FR8fr/fff19ly5ZVkSJFVLVqVY0fP96xXtbTTz+tEiVK6I033tDatWs1bdo0FS1aNN/fm+LFi+vRRx/VpEmTNHDgQN111106cuSIYmNjs0wJHDFihN555x01b95c//nPf1SnTh2lp6fr8OHD+vDDDzVq1Cjdcsst+uqrrzRs2DD169fPUaR6/fXX1aNHD82cOVMjRozI9+MEAAAoiChYAQDg5mJiYjR8+HDVqFFDt9xyi9O2smXLateuXXrmmWf0/PPP65dfflGRIkUUGRmp2267zTGNzdvbW+vWrdOUKVO0YsUKzZw5U0WKFFHdunV12223OeLNmjVLQ4cOVc+ePXX+/Hm1aNFC8fHxqlq1qrZv365x48Zp6NChunDhgqpXr65Fixapf//++fl2OJk4caICAwM1Z84cLV26VNWqVdO8efM0ffp0p/0CAwO1bds2TZ06VQsWLNDBgwfl7++vG2+8UW3btlVERITOnTunu+++W5GRkZozZ47juXfeeaeGDh2q0aNHq0mTJrr55pvz+zABAAAKHMM0TdPVSQAAAAAAAACZWMMKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAIA8iIiI0c+ZMV6cBAABQIFGwAgCgADh+/LiGDx+uSpUqyc/PT6Ghobr11ls1b948nT9/3tXpXRFFHwAAAGTH29UJAACAa/PTTz+padOmKlasmCZPnqzatWsrNTVV3333nRYuXKiwsDB16dIl2+empKSocOHC+ZwxAAAAcHWMsAIAwM0NGTJE3t7e2rVrl+6++25Vr15dtWvX1p133qm1a9fq9ttvd+xrGIbmzZunrl27KjAwUJMmTZIkzZ07VxUrVpSPj4+qVq2qpUuXOp5z6NAhGYahL7/80tF35swZGYah+Ph4SVJ8fLwMw9DatWtVt25d+fn56ZZbbtFXX311Tcf2/vvvq0GDBvLz81OFChU0YcIEpaamOrZ/++23uvXWW+Xn56caNWroo48+kmEYWrNmjWOfo0eP6p577lHx4sUVEhKirl276tChQ47t/fv3V7du3TR9+nSVLVtWISEhGjp0qFJSUhz7nDx5Urfffrv8/f0VGRmpN95445qOCwAAAFdHwQoAADd2+vRpffjhhxo6dKgCAwOz3ccwDKfH48ePV9euXfXVV1/p/vvv1+rVqzV8+HCNGjVKX3/9tQYNGqQBAwZoy5Ytuc7nscce0/Tp07Vz506VLl1aXbp0cSr85MbGjRt17733atiwYdq/f7/mz5+vuLg4Pfvss5Kk9PR0devWTQEBAfrss8+0YMECPfHEE04xzp8/r1atWikoKEj/+9//9MknnygoKEi33XabLl265Nhvy5Yt+vHHH7VlyxYtXrxYcXFxiouLc2zv37+/Dh06pM2bN+vtt9/WnDlzdPLkyTwdFwAAAP4dBSsAANzYDz/8INM0VbVqVaf+kiVLKigoSEFBQRozZozTtt69e+v+++9XhQoVVL58eU2fPl39+/fXkCFDVKVKFY0cOVLdu3fX9OnTc53P+PHjFR0drdq1a2vx4sU6ceKEVq9enadje/bZZ/X444+rX79+qlChgqKjo/XMM89o/vz5kqQPP/xQP/74o5YsWaK6devq1ltvdRSzMr355psqVKiQXnvtNdWuXVvVq1fXokWLdPjwYcfoMEkqXry4Zs+erWrVqqlz587q1KmTPv74Y0nSd999p/Xr1+u1115TVFSUGjRooNdff10XLlzI03EBAADg37GGFQAABcA/R1F9/vnnSk9PV58+fZScnOy0rWHDhk6Pv/nmGz344INOfU2bNtWsWbNynUdUVJTj3yVKlFDVqlX1zTff5DqOJO3evVs7d+50KkKlpaXp4sWLOn/+vA4cOKDw8HCVKVPGsf3mm2/OEuOHH35QkSJFnPovXryoH3/80fG4Zs2a8vLycjwuW7asYzrjN998I29vb6f3rVq1aipWrFiejgsAAAD/joIVAABurFKlSjIMQ99++61Tf4UKFSRJ/v7+WZ6T3dTBfxa8TNN09BUqVMjRlyk30/z+GTun0tPTNWHCBHXv3j3LNj8/P6ccrxajQYMG2a45VapUKce//7nwvGEYSk9Pl/T3cef1OAAAAJB7TAkEAMCNhYSEKDo6WrNnz9a5c+fyFKN69er65JNPnPq2b9+u6tWrS/q7sJOYmOjYfvkC7JfbsWOH49+///67vvvuO1WrVi1Ped100006cOCAKlWqlKUVKlRI1apV0+HDh3XixAnHc3bu3Jklxvfff6/SpUtniVG0aNEc5VG9enWlpqZq165djr4DBw7ozJkzeTouAAAA/DtGWAEA4ObmzJmjpk2bqmHDhoqNjVWdOnVUqFAh7dy5U99++60aNGhw1ec/9thjuvvuu3XTTTepTZs2ev/997Vq1Sp99NFHkjJGaTVu3FhTp05VRESETp06pSeffDLbWBMnTlRISIhCQ0P1xBNPqGTJkurWrdtVX//o0aNZCmA33nijnn76aXXu3Fnh4eG66667VKhQIe3du1dfffWVJk2apOjoaFWsWFH9+vXTtGnTlJSU5Fh0PXM0VJ8+ffT888+ra9eumjhxosqVK6fDhw9r1apVeuyxx1SuXLl/fX+rVq2q2267TQ888IAWLFggb29vjRgxItvRawAAALAGI6wAAHBzFStW1J49e9S2bVuNHTtWdevWVcOGDfXyyy/r0Ucf1TPPPHPV53fr1k2zZs3S888/r5o1a2r+/PlatGiRWrZs6dhn4cKFSklJUcOGDTV8+HBNmjQp21hTp07V8OHD1aBBAyUmJuq9996Tj4/PVV9/+vTpql+/vlN777331L59e33wwQfatGmTGjVqpMaNG2vGjBkqX768JMnLy0tr1qzRn3/+qUaNGmngwIGOQpqfn58kKSAgQP/73/904403qnv37qpevbruv/9+XbhwQcHBwTl9i7Vo0SKFh4erRYsW6t69ux588EGVLl06x88HAABA7hjm5QtSAAAA5EF8fLxatWql33//3aWLkX/66ae69dZb9cMPP6hixYouywMAAADXhimBAADAba1evVpBQUGqXLmyfvjhBw0fPlxNmzalWAUAAODmKFgBAAC3lZSUpNGjR+vIkSMqWbKk2rZtqxdeeMHVaQEAAOAaMSUQAAAAAAAAtsKi6wAAAAAAALAVClYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAVClYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAVbyuCnL2UrAtpqVaEyjN/L28F+/jm6blxcXEaMGCA47GXl5dKlSqlFi1a6JlnnlHlypWtShMFhHn2tHThT9cm4R8kIzgkz0/fu3evXnzxRcXHxysxMVHe3t6qUqWKevbsqYEDB6pEiRK5ihcbG6sJEybINE1HX8uWLSVJ8fHxec4zL+yUS0FkGIbGjx+v2NjY6xJ/8uTJqlGjhrp16+bUHx8fr1atWmnLli2On6fHSkmU0s64NgevYlLhsrl+2j8/c//J6p9vdteDnGjZsqVOnTqlr7/+2rJc8suxY8e0YMECdevWTfXq1XN1Old1vc/riIgItWzZUnFxcZbHBq6Fu99LZl7LfX19deDAAZUvX95puztfQ/v376/4+HgdOnTI0Xc9ryXnz5/XtGnT1LJlyyzXwcz3+eDBg4qIiLD8tQG7u+aC1dlLyVr0/ZdKy+WNoNW8DEMDKtfLc9FKkhYtWqRq1arp4sWL+vTTT/Xss89qy5Yt+vbbb1W8eHELs4U7M8+eVvqiJ6S0FNcm4lVYhQY8m6cbjVdffVVDhgxR1apV9dhjj6lGjRpKSUnRrl27NG/ePCUkJGj16tXXIWkUBAkJCSpXrtx1iz958mT16NEjS8HqpptuUkJCgmrUqHHdXtstpCRKh7pK5iXX5mH4SBHv5qloJf39mftPVv98Bw4cqNtuu83SmHZ37NgxTZgwQREREbYvWF1vq1evVnBwsKvTAJwUhHvJTMnJyXryySe1dOlSCxOzn+t5LTl//rwmTJggSVkKVp06dVJCQoLKls3bZy3g7q65YHUhLdXlxSpJSjNNXUhLVbDyXrCqVauWGjZsKCnjYpGWlqbx48drzZo1V/1rMDzMhT9df4MhZeRw4U8plzcZCQkJeuihhxQdHa01a9bI1/fvcyY6OlqjRo3Shg0brM4WBUjjxo1d8rrBwcEue21bSTvj+mKVlJFD2pk8F6wu/8y9Hs6fP6+AgACVK1fuuhZYYW/169d3dQpAVm5+L3m52267TcuXL9ejjz6qunXrWpjc3y5cuCB/f//rEjunXHUtKVWqlEqVKuWS1wbsgDWsriLzRvrEiRMuzgSwzuTJk2UYhhYsWOBUrMrk4+OjLl26OPWtXLlSUVFRCgwMVFBQkNq3b689e/ZYmld6erpefvll1atXT/7+/ipWrJgaN26s9957L99zWb58uaKiohQUFKSgoCDVq1dPr7/+utM+CxcuVN26deXn56cSJUrojjvu0DfffOO0T//+/RUUFKRvv/1W7du3V2BgoMqWLaupU6dKknbs2KFbb71VgYGBqlKlihYvXuz0/Li4OBmGoc2bN+uBBx5QSEiIgoODdd999+ncuXM6fvy47r77bhUrVkxly5bVo48+qpSUv2+A4+PjZRhGlmmQhw4dkmEYTsPaM3P94Ycf1LFjRwUFBSk8PFyjRo1ScnKy0/MNw8gyHfDo0aN68MEHFR4eLh8fH4WFhalHjx6O6+fFixc1atQo1atXT0WLFlWJEiUUFRWld999N0vsc+fOafHixTIMQ4ZhOE3nzO543nvvPUVFRSkgIEBFihRRdHS0EhISnPaJjY2VYRjat2+fevXqpaJFiyo0NFT333+//vjjD8GeDMPQww8/rKVLl6p69eoKCAhQ3bp19cEHHzjtl/nz/eKLL9SjRw8VL15cFStWdNr2Tzk5zyVp586datasmQICAlShQgVNnTpV6enpju2Z/y+XL1+uMWPGqGzZsgoKCtLtt9+uEydOKCkpSQ8++KBKliypkiVLasCAAfrzT+dpQKZpas6cOY7rX/HixdWjRw/99NNPTvu1bNlStWrVumpO8fHxatSokSRpwIABjvPoalN4f/31Vw0ZMkQ1atRQUFCQSpcurdatW2vbtm1O+2VeO6ZPn64ZM2YoMjJSQUFBioqK0o4dO5z23bVrl3r27KmIiAj5+/srIiJCvXr10s8//3zFPCRp6dKlMgwjyzksSRMnTlThwoV17NgxSdKePXvUuXNnlS5dWr6+vgoLC1OnTp30yy+/OJ4TERGh/v37Ox6np6dr0qRJqlq1quOzpk6dOpo1a9ZV8wKQvdGjRyskJERjxoy56n4XL17U2LFjFRkZKR8fH91www0aOnSozpw547RfRESEOnfurFWrVql+/fry8/PThAkTLLnWvvLKK2revLlKly6twMBA1a5dW9OmTXO6d7qSf15LWrZs6bi+/rNl3l/l5Np66NAhR0FqwoQJjhiZr5V5L3j59EQpd/ehObm3A+yKgtVVHDx4UJJUpUoVF2cCWCMtLU2bN29WgwYNFB4enqPnTJ48Wb169VKNGjX01ltvaenSpUpKSlKzZs20f/9+y3Lr37+/hg8frkaNGmnlypV688031aVLF6cP6PzI5emnn1afPn0UFhamuLg4rV69Wv369XP6JWvKlCmKiYlRzZo1tWrVKs2aNUt79+5VVFSUvv/+e6d4KSkp6t69uzp16qR3331XHTp00NixYzVu3Dj169dP999/v1avXq2qVauqf//+2r17d5acBg4cqKJFi+rNN9/Uk08+qeXLl+uBBx5Qp06dVLduXb399tvq16+fXnjhBb388st5PvaUlBR16dJFbdq00bvvvqv7779fL774op577rmrPu/o0aNq1KiRVq9erZEjR2r9+vWaOXOmihYtqt9//11SxpSB3377TY8++qjWrFmjFStW6NZbb1X37t21ZMkSR6yEhAT5+/urY8eOSkhIUEJCgubMmXPF116+fLm6du2q4OBgrVixQq+//rp+//13tWzZUp988kmW/e+8805VqVJF77zzjh5//HEtX75c//nPf/L4juFapKWlKTU11amlpaVl2W/t2rWaPXu2Jk6cqHfeecdxY/7PYo4kde/eXZUqVdJ///tfzZs374qvnZPzXJKOHz+uPn366N5779V7773nOH+XLVuWJea4ceN08uRJxcXF6YUXXlB8fLx69eqlO++8U0WLFtWKFSs0evRoLV26VOPGjXN67qBBgzRixAi1bdtWa9as0Zw5c7Rv3z41adIkyx/N/i2nm266SYsWLZIkPfnkk47zaODAgVd8P3777TdJ0vjx47V27VotWrRIFSpUUMuWLbNd9++VV17Rpk2bNHPmTL3xxhs6d+6cOnbs6FT8PXTokKpWraqZM2dq48aNeu6555SYmKhGjRrp1KlTV8zlnnvuUZkyZfTKK6849aempmr+/Pm64447FBYWpnPnzik6OlonTpxwyufGG29UUlLSFeNPmzZNsbGx6tWrl9auXauVK1cqJiYmyy/NAHKmSJEievLJJ7Vx40Zt3rw5231M01S3bt00ffp09e3bV2vXrtXIkSO1ePFitW7dOkvx5IsvvtBjjz2mYcOGacOGDbrzzjsd267lWvvjjz+qd+/eWrp0qT744APFxMTo+eef16BBg3J93HPmzHFcXzNb27Zt5eXlpapVq0rK2bW1bNmyjpkNMTExjlhPPfXUFV87t/ehebm3A+zCkkXXC4rMm+fMNawmTZqk5s2bZxltArirU6dO6fz584qMjMzR/keOHNH48eP18MMP66WXXnL0R0dHq3LlypowYYJWrlx5zXlt27ZNS5cu1RNPPKFJkyY5+i9fdyY/cjl48KAmT56sPn36OP1CGh0d7fj3mTNn9Mwzz6hjx45avny5o79ly5aqXLmyYmNj9cYbbzj6L126pEmTJql79+6O/T744ANNmTJFX3zxhWOIecOGDVW6dGktX75cDRo0cMqrc+fOmj59uiOXhIQErVixQjNmzHAUW9q2bauNGzfqjTfe0MiRI/N0/JcuXdKECRN01113SZLatGmjXbt2afny5Xr66aev+Lynn35ap06d0v/93/+pevXqjv67777b8e+iRYs6fomWMq63bdq00e+//66ZM2fqvvvuk5Qx3bBQoUIqVarUv07/S09P12OPPabatWtr/fr1KlQo428wHTt2VMWKFTVmzBh9+umnTs+JiYnRY489JinjPfvhhx+0cOFCvf7669mOwsH1k93P18vLS6mpzl/icuHCBX300UcqUqSIpIyCTFhYmN566y09/vjjTvv269fPsQ7IleTkPM90+vRprVu3TjfffLOkjP8z8fHxWr58ueP/bKY6deo4/R//9ttvNXPmTA0bNkzPP/+84zUSEhL0xhtvOK5jO3bs0KuvvqoXXnjB6dxt1qyZqlSpohkzZjj9YvFvOQUHB6tWrVqSpIoVK+ZoGm3VqlWdCsNpaWlq3769Dh06pJdeeinLmipFihTRBx98IC8vL0lSWFiYbr75Zq1fv149e/aUJPXo0UM9evRwitm5c2eFhoZq+fLlGjZsWLa5+Pj4aNCgQZoyZYpmzJih0qVLS5JWrVqlY8eO6eGHH3a8v6dPn9brr7+url27Op5/+XUnO59++qlq167tNOKsffv2//IOAbiawYMHa9asWRozZow+//zzLJ+nH374oTZu3Khp06Y5PoOjo6MVHh6ue+65R0uWLNEDDzzg2P/kyZPav3+/06CBzAJPXq+1kjRjxgzHv9PT09WsWTOFhIRowIABeuGFF3K1ZvE/11ucPn26PvroIy1YsEBRUVGScnZt9fX1ddz3lStX7l+v2Xm5D83LvR1gF4ywukzjxo1VuHBhFSlSRLfddpuKFy+ud999V97e1PXgmTZu3KjU1FTdd999TqMg/Pz81KJFi1x/494/R1RkTmFZv369JGno0KH5lkt2Nm3apLS0tKvmkZCQoAsXLjgNC5ek8PBwtW7dWh9//LFTv2EY6tixo+Oxt7e3KlWqpLJlyzqth1CiRAmVLl062+kynTt3dnqcWRTq1KlTlv5/m25zNYZh6Pbbb3fqq1Onzr/GXL9+vVq1auVUrMrOf//7XzVt2lRBQUHy9vZW4cKF9frrr2cZwp5TBw4c0LFjx9S3b19HsUqSgoKCdOedd2rHjh06f/6803P++QeIOnXq6OLFizp58mSeckDeLVmyRDt37nRqn332WZb9WrVq5ShWSVJoaOgVz5XL/wp/JTk5zzOVKVPGURjKdKVzIjfn6W+//eaYqvLBBx/IMAzde++9Tte2MmXKqG7dulmubbnJKTfmzZunm266SX5+fo7z8+OPP872/OzUqZOjWJX5+pKccvjzzz81ZswYVapUSd7e3vL29lZQUJDOnTv3r+f8Qw89JCnjC0IyzZ49W7Vr11bz5s0lSZUqVVLx4sU1ZswYzZs3L8ejbG+++Wb93//9n4YMGaKNGzfq7NmzOXoegCvz8fHRpEmTtGvXLr311ltZtmeOvPrnvdNdd92lwMDALPdOderUueIMl7xea6WMacRdunRRSEiIvLy8VLhwYd13331KS0vTd999l7ODzUbmqK4nn3zSqfAm5e7amhN5uQ/Ny70dYBcUrC6TefO8efNmDRo0SN9884169erl6rQAy5QsWVIBAQGO6a7/JnMqSqNGjVS4cGGntnLlyqtO68hOmzZtnGLcf//9kjLm+Ht5ealMmTL5lkt2fv31V0m66gLNp0+flqRsv60lLCzMsT1TQECA/Pz8nPp8fHxUokSJLM/38fHRxYsXs/T/c18fH58r9mf3/JzKLldfX99/jfnrr7/+66LWq1at0t13360bbrhBy5YtU0JCgnbu3Kn7778/zzn/288iPT3dMSUxU0iI88Kymeu4XbhwIU85IO+qV6+uhg0bOrV/ji6Usv7MpIyfW3Y/s5x8i1JOzvO8vHZuzlNJjv/3J06ckGmaCg0NzXJt27FjR5ZrW25yyqkZM2booYce0i233KJ33nlHO3bs0M6dO3XbbbdlGzcn51Hv3r01e/ZsDRw4UBs3btTnn3+unTt3qlSpUv+aa2hoqO655x7Nnz9faWlp2rt3r7Zt2+YYXSVljNrcunWr6tWrp3HjxqlmzZoKCwvT+PHjr7oezdixYzV9+nTt2LFDHTp0UEhIiGPEAYC869mzp2666SY98cQTWc7B06dPy9vbO8vi4YZhqEyZMlnuna52Lc/rtfbw4cNq1qyZjh49qlmzZmnbtm3auXOnY/pxXq+hW7ZsUf/+/XXffffpmWeecdqW22trTlhxH5qTezvALhg6dJnMm2cp4y+6aWlpeu211/T22287DWsH3JWXl5fatGmj9evX65dffvnXX9hKliwpSXr77bdVvnz5a379+fPnO60tkhm/VKlSSktL0/Hjx694k2J1LtnJvJH65ZdfrrjGV+YvaomJiVm2HTt2zJGnHWTeoPxzbQgrinuXK1WqlNMix9lZtmyZIiMjtXLlSqepAtey6Oe//SwKFSqUq+H9cH85mdaZk/M8P5UsWVKGYWjbtm3ZfhFGdn1WW7ZsmVq2bKm5c+c69V9tLair+eOPP/TBBx9o/PjxTtM2M9eyy4nhw4dr6dKlevfdd7VhwwYVK1ZMffr0cdqndu3aevPNN2Wapvbu3au4uDhNnDhR/v7+WaaLZvL29tbIkSM1cuRInTlzRh999JHGjRun9u3b68iRIwoICMjTMQOezjAMPffcc4qOjtaCBQuctoWEhCg1NVW//vqrU9HKNE0dP37c8UURl8ey2po1a3Tu3DmtWrXK6T7yyy+/zHPMvXv3qlu3bmrRooXTiNBMVl9bJfe6DwWswAirq5g2bZqKFy+up59+2ukbgQB3NnbsWJmmqQceeECXLl3Ksj0lJUXvv/++pIx1Pby9vfXjjz9mGQmR2XKjatWqTs+NiIiQJHXo0EGSsnygX87qXLLTrl07eXl5XTWPqKgo+fv7Z1l0+ZdfftHmzZvVpk2ba87DKpnv7969e536//nNi9eqQ4cO2rJliw4cOHDFfQzDkI+Pj9NN6PHjx7N8S6CU89EiVatW1Q033KDly5fLNE1H/7lz5/TOO+84vjkQuFxOzvP81LlzZ5mmqaNHj2Z7Xatdu3auY+Z25KBhGFkKY3v37s32m/pyGs80zSwxX3vttWwX1s9OgwYN1KRJEz333HN644031L9/fwUGBl7x9erWrasXX3xRxYoV0xdffJGj1yhWrJh69OihoUOH6rfffsvyLVwAcqdt27aKjo7WxIkTnabiZd4b/fPe6Z133tG5c+fy5d4p8/7j8uuSaZrZFppy4vDhw+rQoYMqVKigd955R4ULF872NXNybc3NNdud7kMBKzDC6iqKFy+usWPHavTo0Vq+fLnuvfdeV6cEXLOoqCjNnTtXQ4YMUYMGDfTQQw+pZs2aSklJ0Z49e7RgwQLVqlVLt99+uyIiIjRx4kQ98cQT+umnnxxru504cUKff/65AgMD/3WB45xo1qyZ+vbtq0mTJunEiRPq3LmzfH19tWfPHgUEBOiRRx7Jl1wiIiI0btw4PfPMM7pw4YJ69eqlokWLav/+/Tp16pQmTJigYsWK6amnntK4ceN03333qVevXjp9+rQmTJggPz8/jR8//prfD6uUKVNGbdu21ZQpU1S8eHGVL19eH3/8sVatWmXp60ycOFHr169X8+bNNW7cONWuXVtnzpzRhg0bNHLkSFWrVs3xFdVDhgxRjx49dOTIET3zzDMqW7Zslm+0qV27tuLj4/X++++rbNmyKlKkiOMbdy5XqFAhTZs2TX369FHnzp01aNAgJScn6/nnn9eZM2c0depUS48T1vr666+zLLAuZSwU/s9pI1bKyXmen5o2baoHH3xQAwYM0K5du9S8eXMFBgYqMTFRn3zyiWrXru1Y0ymnKlasKH9/f73xxhuqXr26goKCFBYWprCwsGz379y5s5555hmNHz9eLVq00IEDBzRx4kRFRkZm+zP6N8HBwWrevLmef/55lSxZUhEREdq6datef/11FStWLMdxhg8frnvuuUeGYWjIkCFO2z744APNmTNH3bp1U4UKFWSaplatWqUzZ85ku4B+pttvv121atVSw4YNVapUKf3888+aOXOmypcvr8qVK+f6WAE4e+6559SgQQOdPHlSNWvWlJSxCHr79u01ZswYnT17Vk2bNtXevXs1fvx41a9fX3379r3ueUVHR8vHx0e9evXS6NGjdfHiRc2dOzfL0gE51aFDB505c0azZ8/Wvn37nLZlfo7l9NpapEgRlS9fXu+++67atGmjEiVKOK6d/+RO96GAFShY/YtHHnnE8XXavXr1clpkFHBXDzzwgG6++WbH19oeP35chQsXVpUqVdS7d2+ndULGjh2rGjVqaNasWVqxYoWSk5NVpkwZNWrUSIMHD7Ysp7i4ON100016/fXXFRcXJ39/f9WoUcPpK4nzI5eJEyeqcuXKevnll9WnTx95e3urcuXKTt9oNXbsWJUuXVovvfSSVq5cKX9/f7Vs2VKTJ0+23S88S5cu1SOPPKIxY8YoLS1Nt99+u1asWGHJiLRMN9xwgz7//HONHz9eU6dO1enTp1WqVCndeuutjjUlBgwYoJMnT2revHlauHChKlSooMcff1y//PJLlgLBrFmzNHToUPXs2VPnz5+/6qL6vXv3VmBgoKZMmaJ77rlHXl5eaty4sbZs2aImTZpYdoyw3oABA7Ltf/XVVzVw4MDr+to5Oc/z0/z589W4cWPNnz9fc+bMUXp6usLCwtS0adMsC6znREBAgBYuXKgJEyaoXbt2SklJ0fjx452+Ge9yTzzxhM6fP6/XX39d06ZNU40aNTRv3jytXr06z19osXz5cg0fPlyjR49WamqqmjZtqk2bNmVZGPlqunXrJl9fX7Vq1SrLtbVy5coqVqyYpk2bpmPHjsnHx0dVq1ZVXFyc+vXrd8WYrVq10jvvvKPXXntNZ8+eVZkyZRQdHa2nnnoq2xESAHKnfv366tWrl9M32BmGoTVr1ig2NlaLFi3Ss88+q5IlS6pv376aPHlyvkx9rlatmt555x09+eST6t69u0JCQtS7d2+NHDnSMdI/NzK/6CHzW6Avt2jRIvXv3z9X19bXX39djz32mLp06aLk5GT169dPcXFx2b62O92HAtfKMC+fR5EHZy8la9H3Xyrt2sJcMy/D0IDK9RTsc/0vePBs5tnTSl/0hJR25UVd84VXYRUa8KyM4KwL8AIowFISpUNdJTPrlN58ZfhIEe9Khf99oXMgL95//3116dJFa9eudfq2VcDdcS8JADlzzQUrKaNodSEt90PGreTv5U2xCvnGPHtauvDnv+94PfkHcYMBeKqURCntjGtz8CpGsQrXxf79+/Xzzz9r+PDhCgwM1BdffHFdFmEGXIl7SQD4d5YUrAAAAAArtGzZUp9++qluuukmLV68WNWqVXN1SgAAwAUoWAEAAAAAAMBWCrk6AQAAAAAAAOByFKwAAAAAAABgKxSsAAAAAAAAYCsUrAAAAAAAAGArFKwAAAAAAABgKxSsAAAAAAAAYCsUrAAAAAAAAGAr3taEuSDpkjWh8sxHkr+Lc4CnOPtnsi4kp7o0B39fbwUH+bo0BwCukZqcpPSUiy7NoVBhP3n7FnFpDgDgrriXBIB/Z0HB6oKkLZLSrz3UNSkkqZXsULSaPHmyatSooW7dujn1x8XFacCAAdq5c6caNmzomuRwzc7+maxFa75WWprp0jy8vAwN6Fbrmm409u7dqxdffFHx8fFKTEyUt7e3qlSpop49e2rgwIEqUaKEWrZsKUmKj4+3JvF/mDNnjgICAtS/f//rEt8wDI0fP16xsbHXJT6Q31KTk3Ry9xuSmebaRAwvlW7QJ9dFq8zPwuyMGjVK06dPtyI7J/v379dbb72l/v37KyIiwvL4dnHo0CFFRkZq0aJF1+Waev78eU2bNk0tW7Z0fDZkyvy5Hjx4sEC/x4AVCtK95I4dO/TCCy/ok08+0enTpxUSEqJmzZrpscceU6NGjSzMNnv9+/dXfHy8Dh065OiLiIhQy5YtFRcXJ0k6duyYFixYoG7duqlevXrXPafrbfny5Tp58qRGjBhheWyu5bAbCwpWl+T6YpWUkcMl2aVg1aNHjywFKxQMF5JTXX6DIUlpaaYuJKfm+Sbj1Vdf1ZAhQ1S1alU99thjqlGjhlJSUrRr1y7NmzdPCQkJWr16tcVZZzVnzhyVLFnyuhWsEhISVK5cuesSG3CF9JSLri9WSZKZlpFLHkdZLVq0SNWqVXPqCwsLsyKzLPbv368JEyaoZcuW3IBfg/Pnz2vChAmSlKVg1alTJyUkJKhs2bIuyAxwLwXlXvLll1/WiBEjdPPNN2vatGkqX768Dh8+rFdeeUWNGzfW3Llz9eCDD1qc9b9bvXq1goODHY+PHTumCRMmKCIiosAUrL7++uvrUrAC7MaiKYEA3ElCQoIeeughRUdHa82aNfL1/ftGJTo6WqNGjdKGDRtcmOG1MU1TFy9elL+/vxo3buzqdABko1atWm4/2jglJUWGYcjbm9upUqVKqVSpUq5OA0A++fTTTzVixAh17NhRq1evdroO9uzZU3fccYeGDBmi+vXr58tIq8vVr18/X1/PztLS0pSamup0rw+4ExZdlxQbGyvDMLRnzx51795dwcHBKlq0qO699179+uuvkqSYmBiVKFFC58+fz/L81q1bq2bNmpIyph+dO3dOixcvlmEYMgwjy18hk5KS9NBDD6lkyZIKCQlR9+7ddezYMad90tPTNW3aNFWrVk2+vr4qXbq07rvvPv3yyy9O+7Vs2VK1atXSzp071axZMwUEBKhChQqaOnWq0tPtMPINdjR58mQZhqEFCxZk+wHm4+OjLl26ZPvc+Ph4GYaRZYrgoUOHZBiGY/i1JP3000/q2bOnwsLC5Ovrq9DQULVp00ZffvmlpIwh2/v27dPWrVsd58vlox/Onj2rRx99VJGRkfLx8dENN9ygESNG6Ny5c06vbRiGHn74Yc2bN0/Vq1eXr6+vFi9e7Nh2+XTAuLg4GYahLVu2/Ot5mJycrFGjRqlMmTIKCAhQ8+bNtXv3bkVERFy3EWEAMqxcuVJRUVEKDAxUUFCQ2rdvrz179jjts2vXLvXs2VMRERHy9/dXRESEevXqpZ9//tmxT1xcnO666y5JUqtWrRzXmsxr1ZXO539Oe8u89i1dulSjRo3SDTfcIF9fX/3www+SpI8++kht2rRRcHCwAgIC1LRpU3388cf/epzp6emaNGmSqlatKn9/fxUrVkx16tTRrFmznPb7/vvv1bt3b5UuXVq+vr6qXr26XnnllZy8lTl+7pkzZzRq1ChVqFDBce/RsWNHffvttzp06JCjIDVhwgTH+5j53mVeWy+fliNJCxcuVN26deXn56cSJUrojjvu0DfffOO0T//+/RUUFKQffvhBHTt2VFBQkMLDwzVq1CglJyfn6BgB5K8pU6bIMAzNnTs3S9He29tbc+bMceyX6UrTsjN/F7vcK6+8oubNm6t06dIKDAxU7dq1NW3aNKWkpPxrbpdf1+Pj4x0FswEDBjiuXbGxsVq6dKkMw1BCQkKWGBMnTlThwoWz3Bv+07fffqtevXopNDRUvr6+uvHGG3Xfffc5Xbu+/vprde3aVcWLF5efn5/q1avnuE/NlPkZs2LFCj3xxBMKCwtTcHCw2rZtqwMHDjj2a9mypdauXauff/7ZcSyZ713mvfi0adM0adIkRUZGytfXV1u2bJEkvffee4qKilJAQICKFCmi6OjobI8dsBMKVpe54447VKlSJb399tuKjY3VmjVr1L59e6WkpGj48OH6/ffftXz5cqfn7N+/X1u2bNHQoUMlZYxc8ff3V8eOHZWQkKCEhATHBTvTwIEDVbhwYS1fvlzTpk1TfHy87r33Xqd9HnroIY0ZM0bR0dF677339Mwzz2jDhg1q0qSJTp065bTv8ePH1adPH917771677331KFDB40dO1bLli27Du8S3F1aWpo2b96sBg0aKDw8/Lq+VseOHbV7925NmzZNmzZt0ty5c1W/fn2dOXNGUsaQ7QoVKqh+/fqO8yVzGuL58+fVokULLV68WMOGDdP69es1ZswYxcXFqUuXLjJN56H0a9as0dy5c/X0009r48aNatas2VVzy8l5OGDAAM2cOVMDBgzQu+++qzvvvFN33HGHI38AeZf5V9/LW6bJkyerV69eqlGjht566y0tXbpUSUlJatasmfbv3+/Y79ChQ6patapmzpypjRs36rnnnlNiYqIaNWrk+Kzs1KmTJk+eLCnjF6DMa02nTp3ylPfYsWN1+PBhzZs3T++//75Kly6tZcuWqV27dgoODtbixYv11ltvqUSJEmrfvv2/Fq2mTZum2NhY9erVS2vXrtXKlSsVExPjdJ3Zv3+/GjVqpK+//lovvPCCPvjgA3Xq1EnDhg1zTNG7kpw+NykpSbfeeqvmz5+vAQMG6P3339e8efNUpUoVJSYmqmzZso6RtzExMY738amnnrria0+ZMkUxMTGqWbOmVq1apVmzZmnv3r2KiorS999/77RvSkqKunTpojZt2ujdd9/V/fffrxdffFHPPffcVY8PQP5LS0vTli1b1LBhwysuuxAeHq4GDRroo48+ytMf0X/88Uf17t1bS5cu1QcffKCYmBg9//zzGjRoUK7i3HTTTVq0aJEk6cknn3RcuwYOHKh77rlH/9/enYfXcD2OH3/f7DcLkUTEmpDaIoh9CZVFkMUeVGyxEzstoohES8VOSShiKWqLNXYRqrYo5VdaWhqqhDSqjUgiy/z+8LnzdXOz3Cwk2vN6njxP7twzZ87MnXPOzJlzztjY2Gg04GdkZLBmzRq6d++e51D169ev06xZMy5evEhISAhHjhxh/vz5pKWl8erV6xeS3b59m9atW3Pz5k1WrFhBZGQkDg4O+Pv7ExoaqhHnjBkzuH//PuvWrWPt2rX88ssvdO7cmczM11MBrF69GmdnZ2xsbOR9yd7otGLFCqKjo1m0aBFHjhyhTp06bNu2ja5du1KmTBm2b9/O+vXr+euvv3BxceHcuXMFOqaC8C6JPuxv6NGjh1xwdOjQgQoVKtCvXz927txJv379aNeuHatWrWLYsGHyOl9++SVlypRh4MCBALRs2RIdHR3Kly+f61CkTp06sWLFCvnzs2fPmDp1KvHx8djY2PDzzz+zdu1aAgICWLlypRyuUaNGtGjRgqVLl/L555/LyxMTEzl8+DDNmzcHoH379sTExLBt2zY5XYKg8ueff/Ly5UuqV6/+VreTmJjI7du3WbZsmVpDUI8ePeT/GzVqhFKppEyZMhr5ZcWKFdy4cYNLly7Jw4bc3d2pXLkyvr6+HD16FE9PTzn8ixcv+H//7/9Rrlw5rdKXXz68desW27dvZ9q0afLTQQ8PDypUqEDfvn0LfkAEQVCTUx2Znp7O48ePCQoKYuzYsWp51MPDg5o1axIcHMyOHTsA8PX1xdfXVw6TmZmJj48PFSpUYNu2bYwfP57y5ctTs2ZNABwcHIo8TNje3p5du3bJn1++fMmECRPw8fFRm/fPy8uLxo0bM2PGDC5dupRrfN999x3169dX6wnasWNHtTCTJ0/GzMyMc+fOyfOyeHh4kJaWxhdffMH48eNzLfu0XXfZsmXcvHmTEydO0L59e3n9N8vsJk2aAFClSpV8j+Pz58+ZO3cuXl5eag/7XFxcqFmzJnPmzGHr1q3y8levXhEcHCz3hnN3d+fKlSts27aN2bNn57ktQRDeLW2vJatXr87ly5d59uwZVlZWBdrGkiVL5P+zsrJo27YtlpaWDB48mMWLF2t9vVemTBkcHR2B1+V39rJr5MiRzJ8/nyVLlmBtbQ1AZGQkjx49YuzYsXnGPXnyZPT09Lh8+bLakOh+/frJ/8+ZM4dXr15x+vRp+UGxl5cXz58/Jzg4mJEjR1K2bFk5vIODg1qnA11dXXr37k1sbCwtW7bEwcEBc3NzDA0Ncy2HjYyMOHbsGPr6+sDr4+fs7Ez9+vU5cuQIOjo6cjrs7e2ZNm0a3333Xd4HUhBKiOhh9YY3CxeA3r17o6enJ3ejnDBhAj/88IOcof/55x+2bNnCoEGDMDU11Xo72YdaNWjQAEAewqDaXvYhCs2bN6du3boaT2ttbGzkxqo343xzSIQgvGsWFhbY29uzcOFClixZwrVr1wr0hO3QoUM4Ojri5OSk1gOjY8eOOQ5JdHNz0/riBfLPh2fOnAFelwNv8vX1FfPVCEIx2Lx5M7GxsWp/enp6HDt2jIyMDAYOHKiW942MjGjXrp1a3n/x4gXTpk3jgw8+QE9PDz09PUxNTUlOTtYYdlZcevbsqfb5/PnzPHv2jEGDBqmlNysri06dOhEbG6sxjPlNzZs35/r16wQEBHDs2DH++ecfte9TU1M5deoU3bt3x9jYWG0bXl5epKamcvHixRzjLsi6R44coVatWmqNVUVx4cIFUlJSNK5lqlatipubm8a1jEKhoHPnzmrLxLWMILzfVL3hsw/308a1a9fo0qULlpaW6Orqoq+vz8CBA8nMzOTOnTvFlsbRo0cDr19GpPLll19Sv359Pvzww1zXe/nyJWfOnKF37955zt8XHR2Nu7u7xqgGf39/Xr58qdE7Kr/rU2106dJFbqyC1728Hj16xIABA+TGKgBTU1N69uzJxYsXc5z2RhBKA9Fg9QYbGxu1z3p6elhaWpKYmAhA165dsbOzk7uNbty4keTkZHk4oLYsLS3VPqvmEEpJSQGQt5fTm3YqVaokf59bfKo4VfEJwpusrKwwNjbmt99+e6vbUSgUnDp1io4dOxIaGkrjxo0pX74848ePJykpKd/1nzx5wo0bN9DX11f7MzMzQ5IkjaGxBX0zlbb5sEKFCmrhVOWCIAhFU7duXZo2bar2B6/zPkCzZs008v+OHTvU8r6fnx9ffvklw4YN49ixY1y+fJnY2FjKly//1urA7GWNKr2+vr4a6V2wYAGSJPHs2bNc4wsMDGTRokVcvHgRT09PLC0t5d5F8LosysjIYOXKlRrxe3l5AWiUhyoFWTchIaFY36ha0GsZY2NjjIyM1JYZGhqSmppabGkSBKF4aHstGRcXh1KpLPB104MHD2jbti1//PEHy5cv59tvvyU2Nla+ByvO8r1ChQr06dOHNWvWkJmZyY0bN/j222/z7V31119/kZmZmW+5mZiYmGs5qPr+Tfldn2oj+/byK4+zsrL466+/tI5fEN4l0U3gDfHx8VSuXFn+nJGRQWJiolxw6OjoMGbMGGbMmMHixYtZvXo17u7u1K5du1jTodre48ePNQrBR48eFbhLrSC8SVdXF3d3d44cOcLDhw8LfIOiuqHIPhFuTjdMtra2rF+/HoA7d+6wc+dOuWt0eHh4ntuxsrJCqVSyYcOGXL9/U2Ge3uVFlQ+fPHmSY7kgCMLbocrbu3fvxtbWNtdwf//9N4cOHSIoKIjp06fLy9PS0vJsIMrOyMgox4m9//zzzxzr2+xljSrMypUrcx2ekb3h+016enpMnjyZyZMn8/z5c06ePMmMGTPo2LEjv//+O+XKlUNXV5cBAwbk+oAst2E5BVm3fPnyGi92KYo3r2WyE9cygvB+09XVxc3NLc9ryYcPH/L999/TqVMneVle5e2b9u3bR3JyMpGRkWr1gOqlPcVtwoQJbNmyhf3793P06FHMzc01Rt5kZ2Fhga6ubr7lpqWlZa7lIGhezxaH7PVUfuWxjo5OgUYpCMK7JHpYveHNuRQAdu7cSUZGhtpbgoYNG4aBgQH9+vXj9u3bOba+F7V3k5ubG4DGpOmxsbH89NNPuLu7FzpuQYDXT/QlSWL48OHypJBvSk9P5+DBgzmuq3q7y40bN9SWHzhwIM9t1qpVi5kzZ1K/fn2uXr0qL88tv/j4+HD37l0sLS01emE0bdo0x7fMFCdVN3DVXDkqu3fvVpscWhCE4tWxY0f09PS4e/dujnlf1RNLoVAgSZLGm07XrVsnT06rktcTajs7O43y7M6dO2pvZcqLs7Mz5ubm3Lp1K9f0GhgYaBWXubk5vr6+jBkzhmfPnhEXF4exsTGurq5cu3aNBg0a5Bh/br0XCrKup6cnd+7cITo6Otf0FeRJf6tWrVAqlRrXMg8fPpSHyAiC8P6aPn06kiQREBCgUeZmZmYyevRoMjMzmTBhgrzczs6Op0+fyj1T4fX8dceOHVNbX9Xg8mb5LkmS2rC9gsiv7GrSpAmtW7dmwYIFbN26FX9/f0xMTPKMU6lU0q5dO3bt2pVrL1d4PR9fdHS0xtsGN2/ejLGxcaHmVSzovWbt2rWpXLky27ZtU3tpUXJyMnv27JHfHCgIpZHoYfWGyMhI9PT08PDw4ObNm8yaNYuGDRuqzWFjbm7OwIEDCQsLw9bWVmO+BYD69esTExPDwYMHqVixImZmZgXqhVW7dm1GjBjBypUr0dHRwdPTk7i4OGbNmkXVqlWZNGlSseyv8N/VqlUrwsLCCAgIoEmTJowePZp69eqRnp7OtWvXWLt2LY6Ojjme3zY2NrRv35758+dTrlw5bG1tOXXqFJGRkWrhbty4wdixY+nVqxc1a9bEwMCA6Ohobty4odYbon79+nzzzTfs2LGDGjVqYGRkRP369Zk4cSJ79uzhww8/ZNKkSTRo0ICsrCwePHjA8ePHmTJlCi1atHhrx6hevXr07duXxYsXy08Sb968yeLFiylbtqzaHACCIBQfOzs7QkJC+PTTT7l37x6dOnWiXLlyPHnyhMuXL2NiYkJwcDBlypThww8/ZOHChVhZWWFnZ8eZM2dYv3495ubmanGqJtxdu3YtZmZmGBkZUb16dSwtLRkwYAD9+/cnICCAnj17cv/+fUJDQ/Ock+RNpqamrFy5kkGDBvHs2TN8fX2xtrYmISGB69evk5CQQFhYWK7rd+7cGUdHR5o2bUr58uW5f/8+y5Ytw9bWVp4sfvny5bRp04a2bdsyevRo7OzsSEpK4tdff+XgwYN5NjJpu+7EiRPZsWMHXbt2Zfr06TRv3pyUlBTOnDmDj48Prq6umJmZYWtry/79+3F3d8fCwkI+9tmZm5sza9YsZsyYwcCBA+nbty+JiYkEBwdjZGREUFCQVsdXEITSydnZmWXLljFhwgTatGnD2LFjqVatGg8ePJDfyDpnzhw8PDzkdfr06cPs2bP56KOP+OSTT0hNTWXFihUaDV4eHh4YGBjQt29fpk6dSmpqKmFhYYUetmZvb49SqWTr1q3UrVsXU1NTKlWqpPYGwAkTJtCnTx8UCgUBAQFaxbtkyRLatGlDixYtmD59Oh988AFPnjzhwIEDrFmzBjMzM4KCgjh06BCurq7Mnj0bCwsLtm7dSlRUFKGhoWoTrmurfv36REZGEhYWRpMmTdDR0ZEf5uRER0eH0NBQ+vXrh4+PDyNHjiQtLY2FCxfy/PlzvvjiiwKnQRDeFdFg9YbIyEjmzJlDWFiYPPnnsmXLNJ6M9unTh7CwMEaPHp3jTevy5csZM2YMH330ES9fvtSYJFYbYWFh2Nvbs379elatWkXZsmXp1KkT8+fPF/PnCMVi+PDhNG/eXH5teHx8PPr6+tSqVQs/P788x+5v2bKFcePGMW3aNDIzM+ncuTPbt29XqyxtbGywt7dn9erV/P777ygUCmrUqMHixYsZN26cHC44OJjHjx8zfPhwkpKSsLW1JS4uDhMTE7799lu++OIL1q5dy2+//YZSqaRatWq0b9/+rfewAoiIiKBixYqsX7+epUuX4uTkxM6dO+nUqZPGDbEgCMUnMDAQBwcHli9fzvbt20lLS8PGxoZmzZoxatQoOdy2bduYMGECU6dOJSMjA2dnZ06cOIG3t7dafNWrV2fZsmUsX74cFxcXMjMziYiIwN/fQitJrQAAQwBJREFUHz8/Px49ekR4eDgRERE4OjoSFhZGcHCw1unt378/1apVIzQ0lJEjR5KUlIS1tTVOTk4ak45n5+rqyp49e1i3bh3//PMPNjY2eHh4MGvWLHnSXAcHB65evcrcuXOZOXMmT58+xdzcnJo1a8pzUeVG23VVbxKcM2cOa9euJTg4mHLlytGsWTNGjBghh1u/fj2ffPIJXbp0IS0tjUGDBrFx48Yctx0YGIi1tTUrVqxgx44dKJVKXFxcmDdvntwYJwjC+2vcuHE0bdqUxYsXM2XKFBISEsjKysLIyIioqCiN8ql69ers37+fGTNm4OvrS8WKFZk8eTIJCQlqZW6dOnXYs2cPM2fOpEePHlhaWuLn58fkyZPV3hCtLWNjYzZs2EBwcDAdOnQgPT2doKAgtbezduvWDUNDQ1xdXbUunxo2bMjly5cJCgoiMDCQpKQkbGxscHNzk+8fa9euzfnz55kxYwZjxowhJSWFunXrynVQYUyYMIGbN28yY8YM/v77byRJUus5lRM/Pz9MTEyYP38+ffr0QVdXl5YtW3L69Glat25dqHQIwrugkPI7u/OVApwGtH/719uhA7gCygKvOWfOHIKDg0lISNBqHPGUKVMICwvj999/F41H/0H/vEgjYt+PZGYWMesUka6ugsHdHCljaph/YKHYnD9/HmdnZ7Zu3Yqfn19JJ0f4D8pIS+Lp91tBysw/8Nuk0MW6ST/0DM1KNh2CIAjvmX/zteTmzZsZNGgQU6dOZcGCBcUW79t28OBBunTpkmNDmyAIJacYelgped1QpDkPzrtlQGEaqwri4sWL3Llzh9WrVzNy5EjRWPUfVcbUkMHdHElJK9l5jJSGeqKx6i07ceIEFy5coEmTJiiVSq5fv84XX3xBzZo16dGjR0knT/iP0jM0w7pJP7LSS/btaTr6RqKxShAEoRD+zdeSAwcO5PHjx0yfPh0TExNmz55drPEXt1u3bnH//n2mTJmCk5NToXpwCYLw9hRDD6v3n7Y9rBQKBcbGxnh5eREREYGpqek7TKUgCO/apUuXmDJlCrdu3SIpKQkrKys6duzI/Pnzc3w1sCAIgiAIgvD+cHFx4bvvvqNx48Zs2rSJOnXqlHSSBEF4g2iwEgRBEARBEARBEARBEEoV8ZorQRAEQRAEQRAEQRAEoVQRDVaCIAiCIAiCIAiCIAhCqSIarARBEARBEARBEARBEIRSRTRYCYIgCIIgCIIgCIIgCKWKaLASBEEQBEEQBEEQBEEQShXRYCUIgiAIgiAIgiAIgiCUKnrFEUlqSjrp6ZnFEVWh6evrYqTUL9E0CP8dmf+kkpWSUaJp0FHqoVvGqETTIAhCyUhJSeHVq1clmgYDAwOUSmWJpkEQBOF9Ja4lBUEQ8lfkBqvUlHQunb9PVpZUHOkpNB0dBS1a25aKRqt58+bh4OBAt27d1JZv3LiRwYMHExsbS9OmTUsmcVrw9/dn9+7dvHjx4p1uV6FQMGbMGL788st3ut2CyvwnlYR130NmyZ7z6CooP6xJoS80bty4wdKlS4mJieHx48fo6elRq1YtPvroI4YNG4aFhUUxJ7h0srOzw8XFhY0bNwLw6NEj1q5dS7du3XByclILO2fOHIKDg5GkEv7t/yd72kujt5nG0vZ7vCspKSmcPn2arKysEk2Hjo4Orq6uhW60unHjBsuXLycmJoZHjx4BUKVKFdzc3Bg+fHipridLq6LkicOHD3P58mXmzJlT/AkrZV6+fEloaCguLi64uLiUdHJkdnZ2ODo6cujQoZJOSrFTKBQEBQW9lfPL39+fmJgY4uLiij3ut+V9v5ZUKBRahTt9+nSR8tjbrufPnz/P8ePHmThxIubm5m9lG4VV1PN6x44dhISEcO/ePVJTU7l27ZrGda0gvA+KPCQwPT2zxBurALKypBLv5aUyb9489u3bV9LJEN6SrJSMkr/AAMiUCv1k7quvvqJJkybExsbyySefcPToUfbu3UuvXr0IDw9n6NChxZzY0mvv3r3MmjVL/vzo0SOCg4P54YcfNMIOGzaMCxcuvMPUCXn5r/4er169KvHGKoCsrKxC9/Jas2YNTZo04dKlS0yYMIFDhw4RFRXFxIkTuXnzJs2aNePu3bvFnGIhL4cPHyY4OLikk/FOvHz5kuDgYGJiYko6KUIxmDVrFnv37i3pZBTI+34teeHCBbU/Ly8vlEqlxvLGjRsXKXlvu54/f/48wcHBPH/+/K1to7CKcl4nJCQwYMAA7O3tOXr0KBcuXKBWrVrFnEJBeDeKZUigIAjvjwsXLjB69Gg8PDzYt28fhoaG8nceHh5MmTKFo0ePlmAK361GjRppHbZKlSpUqVLlLaZGKAjxe7yfvvvuOwICAvD29mb37t0YGBjI37m5uTFmzBh27dr13g03TElJwcjISOueB/8VKSkp791v+a69fPkSY2Pjkk7Ge8ve3r6kk/Cf07JlS7XP5cuXR0dHR2N5dgU919/Xer448nRRzus7d+6Qnp5O//79adeuXZHSoSLqOKGk/OcnXZ8zZw4KhYJr167Ro0cPypQpQ9myZenfvz8JCQlyuKFDh2JhYcHLly814nBzc6NevXrA6y6yycnJbNq0CYVCgUKh0OgKm5SUxOjRo7GyssLS0pIePXrIwyFUsrKyCA0NpU6dOhgaGmJtbc3AgQN5+PChWjgXFxccHR2JjY2lbdu2GBsbU6NGDb744osiP4G/efMm7u7umJiYUL58ecaOHaux/6tWreLDDz/E2toaExMT6tevT2hoKOnp6Wrhrl27ho+PD9bW1hgaGlKpUiW8vb019gdgy5Yt1K1bF2NjYxo2bPiv7BpfkubNm4dCoWDt2rVqjVUqBgYGdOnSRf5c0HPxwoULtG7dGqVSiZ2dHREREQBERUXRuHFjjI2NqV+/vkajmLZ5sSBp0ua8s7Ozw9/fH4CYmBiaNWsGwODBg+U8rBrCoEpjYdKibV5NTU1lypQpODk5UbZsWSwsLGjVqhX79+/X+K20pW0+fdtpfPHiBebm5owcOVLju7i4OHR1dVm4cCHw+mLv448/pnr16hgZGWFhYUHTpk3Zvn27vE5Ov0d0dDQuLi5YWlqiVCqpVq0aPXv2zLHsFkrGvHnz0NXVZc2aNWqNVW/q1asXlSpVUlt25coVunTpgoWFBUZGRjRq1IidO3eqhdm4cSMKhYLTp0/nW8/C6yETrVq1wsTEBFNTUzp27Mi1a9fy3QfVdo4fP86QIUMoX748xsbGpKWlFSjer776ilq1amFoaIiDgwPbtm3D398fOzs7OUxMTAwKhUKjN1BcXBwKhSLf4bY7duygQ4cOVKxYEaVSSd26dZk+fTrJyclyGH9/f1atWgUgl3sKhUIeipKamkpgYCDVq1fHwMCAypUrM2bMGI1eCXZ2dvj4+BAZGUmjRo0wMjIiODgYd3d36tSpozGsR5IkPvjgA7y9vfPcB1W8R48epXHjxiiVSurUqcOGDRs0wsbHxzNy5EiqVKmCgYEB1atXJzg4mIyMDPm4lS9fHoDg4GB5X/39/bl58yYKhYJdu3bJ8X3//fcoFAr5Ok+lS5cuNGnSRP5c0Lrg7NmztG7dGmNjY4YMGZLrvq9evRo9PT2CgoLkZWFhYTRs2BBTU1PMzMyoU6cOM2bMyPMYqva3RYsWWFhYUKZMGRo3bsz69es1fhdtj3dCQgIBAQE4ODhgamqKtbU1bm5ufPvtt3mmIy4uDj09PebPn6/x3dmzZ9V+g4SEBEaMGEHVqlUxNDSkfPnyODs7c/LkSXmd7HkGYNeuXbRo0YKyZcvKdVlex1kofnmd69qUS5BzPa9aX5sy9tKlS3Tu3BlLS0uMjIywt7dn4sSJctyffPIJANWrV5fLAlVZW9Q8re29Y25yOq8VCgVjx47N817J39+fNm3aANCnTx+Ne9GC1KVFqeP8/f0xNTXl119/xcvLC1NTU6pWrcqUKVPkeFTS0tIICQmhbt26GBkZYWlpiaurK+fPn5fDSJLE6tWrcXJyQqlUUq5cOXx9fbl3716ex1F4//3nG6xUunfvzgcffMDu3buZM2cO+/bto2PHjvIN3YQJE/jrr7/Ytm2b2nq3bt3i9OnTjBkzBnjde0WpVOLl5SV3h129erXaOsOGDUNfX59t27YRGhpKTEwM/fv3VwszevRopk2bhoeHBwcOHGDu3LkcPXqU1q1b8+eff6qFjY+Pp1+/fvTv358DBw7g6elJYGAgX3/9daGPR3p6Ol5eXri7u7Nv3z7Gjh3LmjVr6NOnj1q4u3fv4ufnx5YtWzh06BBDhw5l4cKFajekycnJeHh48OTJE1atWsWJEydYtmwZ1apVIykpSS2+qKgovvzyS0JCQtizZw8WFhZ0795dFEbFJDMzk+joaJo0aULVqlW1Wqeg5+LgwYMZNmwY+/fvp379+gwZMoSQkBACAwOZOnUqe/bswdTUlG7duuV4A5lfXtQ2TQU571QaN24sN7DNnDlTzsPDhg0rtuOTX15NS0vj2bNnfPzxx+zbt4/t27fTpk0bevTowebNm/P5tXKmTT59F2k0NTVlyJAhbN26lb///lvtu9WrV2NgYCBfzE6ePJmwsDDGjx/P0aNH2bJlC7169SIxMTHX+OPi4vD29sbAwIANGzZw9OhRvvjiC0xMTEp8gnLhtczMTE6fPk3Tpk2pWLGi1uudPn0aZ2dnnj9/Tnh4OPv378fJyYk+ffrk2GCjTT07b948+vbti4ODAzt37mTLli0kJSXRtm1bbt26pVW6hgwZgr6+Plu2bGH37t3o6+trHe/atWsZMWIEDRo0IDIykpkzZ76VYWq//PILXl5erF+/nqNHjzJx4kR27txJ586d5TCzZs3C19cXUB/qU7FiRSRJolu3bixatIgBAwYQFRXF5MmT2bRpE25ubho3HlevXuWTTz6R827Pnj2ZMGECt2/f5tSpU2phjxw5wt27d+VrqLxcv36dKVOmMGnSJPbv30+DBg0YOnQoZ8+elcPEx8fTvHlzjh07xuzZszly5AhDhw5l/vz5DB8+HICKFSvKD0yGDh0q7+usWbOoV68eFStWVGsMOXnyJEqlklu3bsl1VkZGBmfOnKF9+/ZyuILUBY8fP6Z///74+flx+PBhAgICNPZXkiQ+/vhjJk6cyLp16+Thmt988w0BAQG0a9eOvXv3sm/fPiZNmqRxo5+TuLg4Ro4cyc6dO4mMjKRHjx6MGzeOuXPnFup4P3v2DICgoCCioqKIiIigRo0auLi45Hke29nZ0aVLF8LDw8nMVJ/O48svv6RSpUp0794dgAEDBrBv3z5mz57N8ePHWbduHe3bt8+zLrhw4QJ9+vShRo0afPPNN0RFRTF79my50VJ4d3I717Upl3KjbRl77Ngx2rZty4MHD1iyZAlHjhxh5syZPHnyBHhdT4wbNw6AyMhIjWGMRc3T2t47FlR+90qzZs2SH0DMmzdP7V60oHVpUeo4eH0/2aVLF9zd3dm/fz9Dhgxh6dKlLFiwQA6TkZGBp6cnc+fOxcfHh71797Jx40Zat27NgwcP5HAjR45k4sSJtG/fnn379rF69Wpu3rxJ69at5d9U+JeSiuifv1Ok6ON3SsXfP3+nFDj9QUFBEiBNmjRJbfnWrVslQPr666/lZe3atZOcnJzUwo0ePVoqU6aMlJSUJC8zMTGRBg0apLGtiIgICZACAgLUloeGhkqA9PjxY0mSJOmnn37KMdylS5ckQJoxY4ZamgDp0qVLamEdHBykjh07anEENA0aNEgCpOXLl6st//zzzyVAOnfuXI7rZWZmSunp6dLmzZslXV1d6dmzZ5IkSdKVK1ckQNq3b1+e2wWkChUqSP/884+8LD4+XtLR0ZHmz59fqH15G17FJ0mPQ78tFX+v4pPyT/Ab4uPjJUD66KOPtApfmHPxypUr8rLExERJV1dXUiqV0h9//CEv/+GHHyRAWrFihbxM27yobZq0Pe9sbW3V8mtsbKwESBERERphVWksyvEpaF7NyMiQ0tPTpaFDh0qNGjXKM+3ayC2fvqs03r17V9LR0ZGWLl0qL0tJSZEsLS2lwYMHy8scHR2lbt265bkv2X+P3bt3S4D0ww8/5Lne++758+fSwYMHS8Xf8+fPC5T2vMog1Xmk+svKypK/q1OnjtSoUSMpPT1dbR0fHx+pYsWKUmZmpiRJ2tezDx48kPT09KRx48aphUtKSpJsbGyk3r1757kfqu0MHDhQbbm28WZmZko2NjZSixYt1MLdv39f0tfXl2xtbeVlp0+flgDp9OnTamF/++03jbIqe57ILisrS0pPT5fOnDkjAdL169fl78aMGZPjukePHpUAKTQ0VG35jh07JEBau3atvMzW1lbS1dWVbt++rRY2MzNTqlGjhtS1a1e15Z6enpK9vb3ab50TW1tbycjISLp//768LCUlRbKwsJBGjhwpLxs5cqRkamqqFk6SJGnRokUSIN28eVOSJElKSEiQACkoKEhjW/3795dq1Kghf27fvr00fPhwqVy5ctKmTZskSZKk7777TgKk48ePS5JUuLrg1KlTOe6nt7e39PLlS6lnz55S2bJlpZMnT6qFGTt2rGRubp7n8dKGqi4ICQmRLC0t1X4DbY93dqo87O7uLnXv3l3tu+zHW3Ve7927V172xx9/SHp6elJwcLC8zNTUVJo4cWKe+zJo0CC1PKP6vQtaPr1L7/O1ZE4GDRokmZiYqC3L61x/U17lUvYyrSBlt729vWRvby+lpOR+f7hw4UIJkH777Te15cWVp7W9d8xJ9vNakrS/V1Llr127dqmtX9C6tLB1nCr9gLRz5061sF5eXlLt2rXlz5s3b5YA6auvvsr1WFy4cEECpMWLF6st//333yWlUilNnTo113WF95/oYfU//fr1U/vcu3dv9PT0OH36tLxswoQJ/PDDD3z33XcA/PPPP2zZsoVBgwZhamqq9bbeHG4F0KBBAwDu378PIG9TNUxJpXnz5tStW1fjCaWNjQ3NmzfXiFMVX2FlPyZ+fn5q6YPXQ666dOmCpaUlurq66OvrM3DgQDIzM7lz5w4AH3zwAeXKlWPatGmEh4fn+eTa1dUVMzMz+XOFChWwtrYu8r4IhVPQc7FixYpqQyQsLCywtrbGyclJbXhP3bp1AXL8XfPLi9qmqSDnXWG9rby6a9cunJ2dMTU1RU9PD319fdavX89PP/1UqHRqk0/fVRpr1KiBj48Pq1evloehbNu2jcTERMaOHSuHa968OUeOHGH69OnExMSQkpKS7346OTlhYGDAiBEj2LRpk+iZ+Z5p0qQJ+vr68t/ixYsB+PXXX/n555/lsiEjI0P+8/Ly4vHjx9y+fVstrvzq2WPHjpGRkcHAgQPV4jMyMqJdu3Za93Lq2bOn2mdt4719+zbx8fH07t1bbf1q1arh7Oys1ba1de/ePfz8/LCxsZHzv2pOE23KlOjoaECznOvVqxcmJiYa5VyDBg00JvfV0dFh7NixHDp0SH5ifvfuXY4ePUpAQIBWc6I4OTlRrVo1+bORkRG1atVSK5sOHTqEq6srlSpVUjv+np6eAJw5cybf7bi7u3Pv3j1+++03UlNTOXfuHJ06dcLV1ZUTJ04Ar3tdGRoaysNuCloXlCtXDjc3txy3n5iYiJubG5cvX+bcuXO4u7trxPn8+XP69u3L/v37NXp65CU6Opr27dtTtmxZ+VyYPXs2iYmJPH36VC2sNscbIDw8nMaNG2NkZCTXBadOncr33HJxcaFhw4ZyTxBVXAqFghEjRqjt78aNG/nss8+4ePGixlD2nKiG9vfu3ZudO3fyxx9/5LuO8Hbkdq4XtlzStoy9c+cOd+/eZejQoRgZFfxt2sWVp4vr3vFNhb1XKkxdWtg6TkWhUGj0mst+TXnkyBGMjIzyHLJ76NAhFAoF/fv3V9uujY0NDRs2FC/Q+JcTDVb/Y2Njo/ZZT08PS0tLtS7HXbt2xc7OTq5cN27cSHJycoG7dFpaWqp9Vs0jpLohU20zp+ESlSpV0ugGnT0+VZza3ODlRrX/b1IdI9X2Hzx4QNu2bfnjjz9Yvnw53377LbGxsfLxUW2/bNmynDlzBicnJ2bMmEG9evWoVKkSQUFBGhceb2NfhP9jZWWFsbExv/32m1bhC3ouWlhYaIQzMDDQWK6atyY1NVUjfH55Uds0FeS8K6y3kVcjIyPp3bs3lStX5uuvv+bChQvExsYyZMiQHI9XfrTNp+8yjRMmTOCXX36Rb/5WrVpFq1at1N4mtGLFCqZNm8a+fftwdXXFwsKCbt268csvv+Qar729PSdPnsTa2poxY8Zgb2+Pvb09y5cvzzdNwrthZWWFUqnM8cJ627ZtxMbGcuDAAbXlqq7+H3/8sVqDlr6+vjy8JPtNe371rCrOZs2aacS5Y8cOrRsBsud9beNVlQ0VKlTQiDOnZYX14sUL2rZty6VLl/jss8+IiYkhNjaWyMhIQDP/5yQxMRE9PT153icVhUKBjY2NRjmX21DPIUOGoFQqCQ8PB17ne6VSqfW8QtqUTU+ePOHgwYMax141V4w2v6tqmN/Jkyc5d+4c6enpuLm50b59e/km9eTJkzg7O8uTyRe0LshrOOydO3e4dOkSnp6eODo6anw/YMAANmzYwP379+nZsyfW1ta0aNFCLk9zc/nyZTp06AC8njvtu+++IzY2lk8//RQoXF2wZMkSRo8eTYsWLdizZw8XL14kNjaWTp06aXVujR8/nlOnTnH79m3S09P56quv8PX1VbsO2LFjB4MGDWLdunW0atUKCwsLBg4cSHx8fK7xfvjhh+zbt0++sa5SpQqOjo5qcyAK70ZO53pRyiVty1jV3KeFnbC9uPJ0cd07vqmw90qFqUsLW8epGBsbazQYGhoaql0rJiQkUKlSJXR0cm+WePLkCZIkUaFCBY3tXrx4sUAN98L7R7wl8H/i4+OpXLmy/DkjI4PExES1QkFHR4cxY8YwY8YMFi9ezOrVq3F3d6d27drFmhbVNh8/fqxR0D569AgrK6ti3V5Octp/1cWBatm+fftITk4mMjISW1tbOdwPP/ygEV/9+vX55ptvkCSJGzdusHHjRkJCQlAqlUyfPv3t7owg09XVxd3dnSNHjvDw4cN8K/KSOBfzy4sFSdPbPu/exvH5+uuvqV69Ojt27FDrdZB9nhhtFSSfvqs0urm54ejoyJdffompqSlXr17VmHPPxMSE4OBggoODefLkidzbqnPnzvz888+5xt22bVvatm1LZmYmV65cYeXKlUycOJEKFSrw0UcfFW6HhWKjq6uLm5sbx48f5/Hjx2oXww4ODgDyRN8qqnwUGBhIjx49coy3oPWwKs7du3er5YuCyt4zSNt4VWVHTvNuZL8RV13sZ89f2lygR0dH8+jRI2JiYtTeFFWQV7hbWlqSkZFBQkKCWqOVJEnEx8fLvVlUcustVbZsWbnh4eOPPyYiIgI/Pz/Mzc21Tkt+rKysaNCgAZ9//nmO32efyD8nVapUoVatWpw8eRI7OzuaNm2Kubk57u7uBAQEcOnSJS5evCjPKQUFrwvy6lHWqlUrevXqxdChQ4HXE6xnv5EbPHgwgwcPJjk5mbNnzxIUFISPjw937tzJ9bz75ptv0NfX59ChQ2o3kPv27cv3mOTm66+/xsXFhbCwMLXluc0TmZ2fnx/Tpk1j1apVtGzZkvj4eI0beSsrK5YtW8ayZct48OABBw4cYPr06Tx9+jTPNxp37dqVrl27kpaWxsWLF5k/fz5+fn7Y2dnRqlWrgu+sUCg5netFKZe0LWNVZVVOL3fSRnHl6Xd176iNwtSlha3jCqJ8+fKcO3eOrKysXButrKysUCgUfPvttzm+MCqnZcK/h+hh9T9bt25V+7xz504yMjI03vA3bNgwDAwM6NevH7dv31YbwqJS1B5Bqi6l2W/gYmNj+emnnzS6h78t2Y+JatJA1TFRFWJvFhKSJPHVV1/lGqdCoaBhw4YsXboUc3Nzrl69WsypFvITGBiIJEkMHz48x4mo09PTOXjwIFAy52J+ebEwaSrIeZe9J0Ze3sbxUSgUGBgYqF0kxMfHF/otgYXJp+8ijePHjycqKorAwEAqVKhAr169cg1boUIF/P396du3L7dv39bqjX+6urq0aNFCfqopyprSIzAwkMzMTEaNGqVVb8fatWtTs2ZNrl+/TtOmTXP8e3N4hDY6duyInp4ed+/ezTXOwtA23tq1a2NjY6PxZqYHDx6ovRUJkN8SdePGDbXl2Xui5SSn/A+wZs0ajbC5lX2qcix7Obdnzx6Sk5MLVM6NHz+eP//8E19fX54/f57jNVRR+Pj48OOPP2Jvb5/jsVc1WOVXzrdv357o6GhOnDiBh4cHALVq1aJatWrMnj2b9PR0tQnXi7suGDRoEN988w0RERHy8O2cmJiY4OnpyaeffsqrV6+4efNmrnEqFAr09PTQ1dWVl6WkpLBly5YCpS17nNnPrRs3bnDhwgWt1jcyMpKHcC9ZsgQnJ6c8h8RWq1aNsWPH4uHhoXWZbmhoSLt27eRJnrV5C6jwdhWkXMpO2zK2Vq1a2Nvbs2HDhjwfpuVWFhRnntbm3vFdKI669G3UnZ6enqSmpub5xlsfHx8kSeKPP/7IcZv169cv8HaF94foYfU/kZGR6Onp4eHhwc2bN5k1axYNGzbUmF/C3NycgQMHEhYWhq2tbY5vs6hfvz4xMTEcPHiQihUrYmZmVqCW9Nq1azNixAhWrlyJjo4Onp6exMXFMWvWLKpWrcqkSZMKtY8uLi6cOXNG4/XFOTEwMGDx4sW8ePGCZs2acf78eT777DM8PT3lORs8PDwwMDCgb9++TJ06ldTUVMLCwvjrr7/U4jp06BCrV6+mW7du1KhRA0mSiIyM5Pnz5/KFoPDutGrVirCwMAICAmjSpAmjR4+mXr16pKenc+3aNdauXYujoyOdO3d+a+diXvLLi9qmqbDnnb29PUqlkq1bt1K3bl1MTU2pVKlSjk/m38bxUb0SPiAgAF9fX37//Xfmzp1LxYoV8xwOlxtt8+m7TmP//v0JDAzk7NmzzJw5Ux4mqtKiRQt8fHxo0KAB5cqV46effmLLli20atUKY2PjHOMMDw8nOjoab29vqlWrRmpqqvwa9jdvLoWS5ezszKpVqxg3bhyNGzdmxIgR1KtXDx0dHR4/fsyePXsAKFOmjLzOmjVr8PT0pGPHjvj7+1O5cmWePXvGTz/9xNWrV9m1a1eB0mBnZ0dISAiffvop9+7do1OnTpQrV44nT55w+fJluYdfQWkbr46ODsHBwYwcORJfX1+GDBnC8+fPCQ4OpmLFimpPmW1sbGjfvj3z58+nXLly2NracurUKXn4TF5at25NuXLlGDVqFEFBQejr67N161auX7+uEVZ1wb9gwQI8PT3R1dWlQYMGeHh40LFjR6ZNm8Y///yDs7MzN27cICgoiEaNGjFgwACtj0+tWrXo1KkTR44coU2bNjRs2FDrdbUREhLCiRMnaN26NePHj6d27dqkpqYSFxfH4cOHCQ8Pp0qVKpiZmWFra8v+/ftxd3fHwsICKysruXHQ3d2d1atX8+eff7Js2TI5fnd3dyIiIihXrpzafI1voy7w9fXF2NgYX19fUlJS2L59OwYGBgwfPhylUomzszMVK1YkPj6e+fPnU7ZsWY3ebm/y9vZmyZIl+Pn5MWLECBITE1m0aFGReib4+Pgwd+5cgoKCaNeuHbdv3yYkJITq1atr/Ua+gIAAQkND+f7771m3bp3ad3///Teurq74+flRp04dzMzMiI2N5ejRo7n2EAGYPXs2Dx8+xN3dnSpVqvD8+XOWL1+uNk+SUHIKUi5lV5Cye9WqVXTu3JmWLVsyadIkqlWrxoMHDzh27Jj8cFRV7i1fvpxBgwahr69P7dq1izVPa3Pv+K4UtS59G3Vn3759iYiIYNSoUdy+fRtXV1eysrK4dOkSdevW5aOPPsLZ2ZkRI0YwePBgrly5wocffoiJiQmPHz/m3Llz1K9fn9GjRxfl0AilmGiw+p/IyEjmzJlDWFiYPEHcsmXLNG6iAPr06UNYWBijR4/Osevi8uXLGTNmDB999BEvX74s0ASuKmFhYdjb27N+/XpWrVpF2bJl6dSpE/Pnz89x7LI2Xrx4oTE/UG5U3cbHjx/PZ599hlKpZPjw4SxcuFAOU6dOHfbs2cPMmTPp0aMHlpaW+Pn5MXnyZHmCU4CaNWtibm5OaGgojx49wsDAgNq1a7Nx40YGDRpUqH0Rimb48OE0b95cfrVsfHw8+vr61KpVCz8/P7WnP2/jXMyLNnlRmzQV9rwzNjZmw4YNBAcH06FDB9LT0wkKCmLOnDk5hi/u4zN48GCePn1KeHg4GzZsoEaNGkyfPp2HDx8W6gZa23z6rtOoVCrp3LkzX3/9NaNGjdL43s3NjQMHDrB06VJevnxJ5cqVGThwoDzfSk6cnJw4fvw4QUFBxMfHY2pqiqOjIwcOHJDnbhFKh1GjRtGqVSuWL1/O0qVLefToEQqFgipVqtC6dWtOnTqlNoGtq6srly9f5vPPP2fixIn89ddfWFpa4uDgoPFgSVuBgYE4ODiwfPlytm/fTlpaGjY2NjRr1izHc7K44x0xYgQKhYLQ0FC6d++OnZ0d06dPZ//+/Wqv8gbYsmUL48aNY9q0aWRmZtK5c2e2b9+e79NsS0tLoqKimDJlCv3798fExISuXbuyY8cOtTnj4PXwrO+++47Vq1cTEhKCJEn89ttv2NnZsW/fPubMmUNERASff/45VlZWDBgwgHnz5hW4waNPnz4cOXLkrfQyqFixIleuXGHu3LksXLiQhw8fYmZmRvXq1eUbK5X169fzySef0KVLF9LS0hg0aJD8hN/NzQ0dHR2USqXa8LH27dsTERGBq6urxvXf26grvby8OHz4MJ07d6Zr165ERkbStm1bNm7cyM6dO/nrr7+wsrKiTZs2bN68WWOesTe5ubmxYcMGFixYQOfOnalcuTLDhw/H2tpaHn5YUJ9++ikvX75k/fr1hIaG4uDgQHh4OHv37tX6urdy5cq0adOGGzduyC/3UTEyMqJFixZs2bKFuLg40tPTqVatGtOmTWPq1Km5xtmiRQuuXLnCtGnTSEhIwNzcnKZNmxIdHS3PZyaUnIKUSznRtozt2LEjZ8+eJSQkhPHjx5OamkqVKlXUXsrh4uJCYGAgmzZt4quvviIrK4vTp0/LQ12LK0/nd+/4rhRHXVrcdaeenh6HDx9m/vz5bN++nWXLlmFmZkbDhg3p1KmTHG7NmjW0bNmSNWvWsHr1arKysqhUqRLOzs4aLwsS/l0UkjbdbfKQmpLOpfP3ycoqUjRFpqOjoEVrW4yU+gVab86cOQQHB5OQkKD1fDNTpkwhLCyM33///a3csL8NSUlJWFhYsGzZsiJN9CdA5j+pJKz7HjJL9pxHV0H5YU3QLVPwt5+URoXJi8L769WrV9jZ2dGmTRuNYVFC3lJSUjh9+jRZWVklmg4dHR1cXV3liaeFonv+/Dm1atWiW7durF27tqST81b07NmTixcvEhcXh75+wa7ZhH+fp0+fYmtry7hx4wgNDS3p5Lwz4lpSO5MmTWLLli3v/aTa7+O9oyCUFkXuYWWk1KdFa1vS03MeX/+u6OvrFrixqqAuXrzInTt3WL16NSNHjnyvCpyzZ8/KT9OEotEtY0T5YU3IStGuu/vboqPUK7UXGIKQm4SEBG7fvk1ERARPnjwRL10oBKVSiaura45z0L1LBgYGorGqCOLj4/n8889xdXXF0tKS+/fvs3TpUpKSkpgwYUJJJ69YpaWlcfXqVS5fvszevXtZsmSJaKz6j3v48CH37t1j4cKF6Ojo/OvO+fyIa8m8PX36lAsXLhAZGfleT5L/Pt87CkJpUSxDAo2U+m+9sag0UM2d4uPjw2effVbSySkQb29vvL29SzoZ/xq6ZYzQLZN/OEEQ1EVFRTF48GAqVqzI6tWrter+L2hSKpWiseg9Z2hoSFxcHAEBATx79gxjY2NatmxJeHj4v27Y0uPHj2ndujVlypRh5MiRjBs3rqSTJJSwdevWERISgp2dHVu3blV7O/B/hbiWzN3hw4cZO3YsLVu2ZPny5SWdnEJ7n+8dBaG0KPKQQEEQBEEQBEEQBEEQBEEoTiU365sgCIIgCIIgCIIgCIIg5EA0WAmCIAiCIAiCIAiCIAilimiwEgRBEARBEARBEARBEEoV0WAlCIIgCIIgCIIgCIIglCqiwUoQBEEQBEEQBEEQBEEoVUSDlSAIgiAIgiAIgiAIglCqiAYrQRAEQRAEQRAEQRAEoVTRK45InqUm8yIjrTiiKjRTPUMsjEyKFMeKFSuYMGEC9erV48cffyymlBWPefPm4eDgQLdu3Qq1/suXLwkNDcXFxQUXFxe17zZu3MjgwYP57bffsLOzK3Ja/wuS0xNIy0wq0TQY6pphol++wOupfm85HkNDzM3NqVu3Lh06dGDo0KFYW1urrTNnzhyCg4ORJKnI6S5pCoWCoKAg5syZU9JJyTHv+fv7ExMTQ1xcnByuqPn/XXvb58utW7fYuXMn/v7+GmVWTsfv3+h9r3dV535sbCxNmzbVer286rLS6vDhw1y+fDnHMsfOzg4XFxc2btz4ztMlCIIgCIJQ2hW5wepZajKzrhwkQ8oqjvQUmp5Ch7lNOxep0WrDhg0A3Lx5k0uXLtGiRYviSl6RzZs3D19f3yI1WAUHBwNoXOR7e3tz4cIFKlasWMRU/jckpydw6LcJZEnpJZoOHYU+PtWXF6rRCiAiIoI6deqQnp7O06dPOXfuHAsWLGDRokXs2LGD9u3by2GHDRtGp06diivpJerChQtUqVKlpJORq1mzZjFhwgS1ZUXN//82t27dIjg4GBcXF40Gq5yO37/Nv6neLai86rLS6vDhw6xatSrHBqu9e/dSpkyZd58oQRAEQRCE90CRG6xeZKSV+EUzQIaUxYuMNCwo3IXzlStXuH79Ot7e3kRFRbF+/fpS1WD1NpUvX57y5QvX6PFflJaZVOKNVQBZUjppmUmFbrBydHRU69nQs2dPJk2aRJs2bejRowe//PILFSpUAKBKlSqlupGnIFq2bFnSSciTvb19iW07PT0dhUKBnl6xdL4tESV5/N6Vf0u9+756+fIlxsbGxRJXo0aNiiUeQRAEQRCEfyMxh9X/rF+/HoAvvviC1q1b88033/Dy5Uv5+7i4OBQKBYsWLWLJkiVUr14dU1NTWrVqxcWLF9Xi8vf3x9TUlF9//RUvLy9MTU2pWrUqU6ZMIS1NfQjHs2fPCAgIoHLlyhgYGFCjRg0+/fRTtXAKhYLk5GQ2bdqEQqFAoVDIT5YTEhIICAjAwcEBU1NTrK2tcXNz49tvv1VLu6pBKjg4WI7D398feD00Q6FQaAyh2bBhAw0bNsTIyAgLCwu6d+/OTz/9VOh9FUq/atWqsXjxYpKSklizZo28fM6cOSgUCrWwdnZ2+Pj4cOjQIRo1aoRSqaRu3bocOnQIeH1e1a1bFxMTE5o3b86VK1c0tnflyhW6dOmChYUFRkZGNGrUiJ07d6qFUZ2fp0+fZvTo0VhZWWFpaUmPHj149OiRWtjo6GhcXFywtLREqVRSrVo1evbsqZaXFQqFRk+HH3/8ka5du1KuXDmMjIxwcnJi06ZNamFiYmJQKBRs376dTz/9lEqVKlGmTBnat2/P7du31cKeOHGCrl27UqVKFYyMjPjggw8YOXIkf/75Zz6/ABrD3HLL/3Fxcejp6TF//nyNOM6ePYtCoWDXrl25bke1P1u2bGHKlClUrlwZQ0NDfv31VwBOnjyJu7s7ZcqUwdjYGGdnZ06dOqURT1RUFE5OThgaGlK9enUWLVqkEUZVfuY07Cmn3+Pnn3+mb9++VKhQAUNDQ6pVq8bAgQNJS0tj48aN9OrVCwBXV1f5mKjizmmYYGpqKoGBgVSvXh0DAwMqV67MmDFjeP78uVo41Tl99OhRGjdujFKppE6dOnLvW+Ht0aYuya8uA/jll1/w8/PD2toaQ0ND6taty6pVqzS2d/PmTTp06ICxsTHly5dnzJgxREVFoVAoiImJkcO5uLjg6OjI2bNnad26NcbGxgwZMgSAHTt20KFDBypWrCiXf9OnTyc5OVltv1TbV6X3zfrWzs5OLf0ADx48oH///mr7sHjxYrKy/q+RsiDXJIIgCIIgCO8r0WAFpKSksH37dpo1a4ajoyNDhgwhKSkpx5u9VatWceLECZYtW8bWrVtJTk7Gy8uLv//+Wy1ceno6Xbp0wd3dnf379zNkyBCWLl3KggUL5DCpqam4urqyefNmJk+eTFRUFP379yc0NJQePXrI4S5cuIBSqcTLy4sLFy5w4cIFVq9eDbxu8AIICgoiKiqKiIgIatSogYuLi3zRXbFiRY4ePQrA0KFD5ThmzZqV6zGZP38+Q4cOpV69ekRGRrJ8+XJu3LhBq1at+OWXXwq8r8L7w8vLC11dXc6ePZtv2OvXrxMYGMi0adOIjIykbNmy9OjRg6CgINatW8e8efPYunUrf//9Nz4+PqSkpMjrnj59GmdnZ54/f054eDj79+/HycmJPn365NiwMWzYMPT19dm2bRuhoaHExMTQv39/+fu4uDi8vb0xMDBgw4YNHD16lC+++AITExNevXqV6z7cvn2b1q1bc/PmTVasWEFkZCQODg74+/sTGhqqEX7GjBncv3+fdevWsXbtWn755Rc6d+5MZmamHObu3bu0atWKsLAwjh8/zuzZs7l06RJt2rQhPb1gvfNyy/92dnZ06dKF8PBwtW0DfPnll1SqVInu3bvnG39gYCAPHjwgPDycgwcPYm1tzddff02HDh0oU6YMmzZtYufOnVhYWNCxY0e1RqtTp07RtWtXzMzM+Oabb1i4cCE7d+4kIiKiQPv4puvXr9OsWTMuXrxISEgIR44cYf78+aSlpfHq1Su8vb2ZN28e8Lo8Vh0Tb2/vHOOTJIlu3bqxaNEiBgwYQFRUFJMnT2bTpk24ublpNKxfv36dKVOmMGnSJPbv30+DBg0YOnSoVvlBKJr86pL86rJbt27RrFkzfvzxRxYvXsyhQ4fw9vZm/Pjx8jBCgMePH9OuXTtu375NWFgYmzdvJikpibFjx+aYrsePH9O/f3/8/Pw4fPgwAQEBwOvGMS8vL9avX8/Ro0eZOHEiO3fupHPnzvK6s2bNwtfXF0BOb15D8BMSEmjdujXHjx9n7ty5HDhwgPbt2/Pxxx/nmD5tr0kEQRAEQRDeR+/vuI9itHv3bv7++2+GDh0KQJ8+fZg4cSLr169n0KBBamHNzMw4dOgQurq6AFSqVInmzZtz5MgRPvroIzncq1evCA4OlnsCuLu7c+XKFbZt28bs2bMB2LRpEzdu3GDnzp1yOA8PD0xNTZk2bRonTpzAw8ODli1boqOjQ/ny5TWGM9WuXVtuvALIzMykY8eOxMXFsWLFClxcXDA0NKRJkybA66Fd+Q2Jev78OXPnzsXLy4tt27bJy11cXKhZsyZz5sxh69atBdpX4f1hYmKClZWVRu+lnCQmJnLx4kUqV64MvM4PTk5OfPXVV/z666/ysBmFQkG3bt04efKkfDMXEBBAvXr1iI6OloegdezYkT///JMZM2YwcOBAdHT+r029U6dOrFixQv787Nkzpk6dSnx8PDY2Nnz//fekpqaycOFCGjZsKIfz8/PLcx/mzJnDq1evOH36NFWrVgVeN9o9f/6c4OBgRo4cSdmyZeXwDg4OfP311/JnXV1devfuTWxsrJy3Ro0aJX8vSRKtW7fGxcUFW1tbjhw5QpcuXfI9tip55f/x48fj6urKwYMH5fmtHj16xN69e5k1a5ZWQ/vs7e3VGudfvnzJhAkT8PHxYe/evfJyLy8vGjduzIwZM7h06RIAn376KRUqVODEiRMYGRkBr3/Dory8YfLkyejp6XH58mW1ocr9+vUDXpfBNWvWBF7/FvmVZ8ePH+fYsWOEhobyySefAK/L2apVq9KnTx82b97M8OHD5fB//vkn3333HdWqVQPgww8/5NSpU2zbto0PP/yw0Psl5C+/uiS/umzy5MmYmZlx7tw5eV4oDw8P0tLS+OKLLxg/fjzlypVj6dKlPHv2jLNnz+Lg4ACAp6cnnTp1ynGy/mfPnrFr1y7c3NzUls+cOVP+X5IknJ2dqVu3Lu3atePGjRs0aNAAe3t7eWi1NsORlyxZwh9//MGlS5do3rw58DpPZWZmEh4ezsSJE6lVq5YcXttrEkEQBEEQhPeR6GHF6+GASqVSvrgzNTWlV69efPvttxq9iby9veULQ4AGDRoAcP/+fbVwCoVC7SmrKuyb4aKjozExMZGfvqqohgfkNPwmJ+Hh4TRu3BgjIyP09PTQ19fn1KlTGsP3tHXhwgVSUlI0hilUrVoVNzc3jXRps6/C+0Xbt7s5OTnJjVUAdevWBV43br45x4tqueqc+PXXX/n555/lRoiMjAz5z8vLi8ePH2sMs8veyJM97zk5OWFgYMCIESPYtGkT9+7d02ofoqOjcXd3lxurVPz9/Xn58iUXLlwoUDoAnj59yqhRo6hataqcJ21tbQEKnS9z4uLiQsOGDdWGPIWHh6NQKBgxYoRWcfTs2VPt8/nz53n27BmDBg1S+12ysrLo1KkTsbGxJCcnk5ycTGxsLD169JAbq+D1DXT28kBbL1++5MyZM/Tu3bvY5tWLjo4G0CjPevXqhYmJiUZ55uTkJDdWARgZGVGrVi1Rnr0DRalLUlNTOXXqFN27d8fY2FijTElNTZWHyp05cwZHR0e5sUqlb9++OcZdrlw5jcYqgHv37uHn54eNjQ26urro6+vTrl07oPD5PDo6GgcHB7mxSsXf3x9JkuTzWUXbaxJBEARBEIT30X++werXX3/l7NmzeHt7I0kSz58/5/nz53IjUva5SywtLdU+GxoaAqgNdQIwNjZWu4lThU1NTZU/JyYmYmNjozE3kLW1NXp6eiQmJuab/iVLljB69GhatGjBnj17uHjxIrGxsXTq1EkjTdpSbTenIQuVKlXSSJc2+yq8P5KTk0lMTKRSpUr5hrWwsFD7bGBgkOdy1Tnx5MkTAD7++GP09fXV/lTDbbLP95Rf3rO3t+fkyZNYW1szZswY7O3tsbe3Z/ny5XnuQ2JiYq7nuur7gqQjKyuLDh06EBkZydSpUzl16hSXL1+Wb5YLmy9zM378eE6dOsXt27dJT0/nq6++wtfXFxsbG63Wz77vqt/G19dX47dZsGABkiTx7Nkz/vrrL7KysnLcjrbbzu6vv/4iMzOzWCf4T0xMRE9PT6MBTKFQYGNjk+/vC69/4+L+3QRNRalLEhMTycjIYOXKlRrnrZeXF/B/ZUpiYqLc6+lNOS2DnOvCFy9e0LZtWy5dusRnn31GTEwMsbGxREZGAoXP58VdHgmCIAiCILzP/vNDAjds2IAkSezevZvdu3drfL9p0yY+++yzt7JtS0tLLl26hCRJao1WT58+JSMjAysrq3zj+Prrr3FxcSEsLExteVJSUpHSBa/n7cju0aNHWqVLeH9FRUWRmZn5Vl8ZrzqHAgMD1eZre1Pt2rULHG/btm1p27YtmZmZXLlyhZUrVzJx4kQqVKiQ6/AYS0vLXM/1N9OqrR9//JHr16+zceNGtSHFqsnMi5ufnx/Tpk1j1apVtGzZkvj4eMaMGaP1+tkbzFX7u3LlylyHMFWoUEF+o2B8fLzG99mXqRohss8Xlf3m28LCAl1dXR4+fKh1+vNjaWlJRkYGCQkJao1WkiQRHx9Ps2bNim1bQskpV64curq6DBgwINfzv3r16sDrc0LVMPumnM5l0Mwj8Lon1KNHj4iJiZF7VQEaE/kXVHGXR4IgCIIgCO+z/3QPq8zMTDZt2oS9vT2nT5/W+JsyZQqPHz/myJEjb2X77u7uvHjxgn379qkt37x5s/y9Sm5P+BUKhfxEVeXGjRsaw5gK8tS1VatWKJVKtXl6AB4+fCgPnxL+nR48eMDHH39M2bJlGTly5FvbTu3atalZsybXr1+nadOmOf6ZmZkVOn5dXV1atGghD5W7evVqrmHd3d3lm883bd68GWNjY63mnXmT6uY2e758862LBZVXDx8jIyN5GOSSJUtwcnLC2dm50NtydnbG3NycW7du5frbGBgYyG9/jIyMVOsBk5SUxMGDB9XirFChAkZGRty4cUNt+f79+9U+K5VK2rVrx65du/J8o2JByjNVeZW9PNuzZw/JycmiPHvP5NWr2dXVlWvXrtGgQYMcz1vVw5h27drx448/cuvWLbU4vvnmG63TUZB8XtDz9datWxpl1ubNm1EoFLi6umqdRkEQBEEQhPfdf7qH1ZEjR3j06BELFizIsTeJo6MjX375JevXr2fp0qXFvv2BAweyatUqBg0aRFxcHPXr1+fcuXPMmzcPLy8v2rdvL4etX78+MTExHDx4kIoVK2JmZkbt2rXx8fFh7ty5BAUFyW89CgkJoXr16mRkZMjrm5mZYWtry/79+3F3d8fCwgIrK6scJ0c2Nzdn1qxZ8sTXffv2JTExkeDgYIyMjAgKCir2YyG8ez/++KM8x8vTp0/59ttviYiIQFdXl7179xbbHEK5WbNmDZ6ennTs2BF/f38qV67Ms2fP+Omnn7h69WqOb+nMS3h4ONHR0Xh7e1OtWjVSU1PlIb1v5qXsgoKCOHToEK6ursyePRsLCwu2bt1KVFQUoaGhahOua6NOnTrY29szffp0JEnCwsKCgwcPcuLEiQLF86bc8r9KQEAAoaGhfP/996xbt67Q24HXc/itXLmSQYMG8ezZM3x9fbG2tiYhIYHr16+TkJAg9+icO3cunTp1wsPDgylTppCZmcmCBQswMTGR32AKr2/u+/fvz4YNG7C3t6dhw4ZcvnxZ7aUOKkuWLKFNmza0aNGC6dOn88EHH/DkyRMOHDjAmjVrMDMzw9HREYC1a9diZmaGkZER1atXz3E4n4eHBx07dmTatGn8888/ODs7c+PGDYKCgmjUqBEDBgwo0vES3q286rLly5fTpk0b2rZty+jRo7GzsyMpKYlff/2VgwcPyvM/TZw4kQ0bNuDp6UlISAgVKlRg27Zt/PzzzwBqL3vITevWrSlXrhyjRo0iKCgIfX19tm7dyvXr1zXC1q9fH4AFCxbg6emJrq4uDRo0kIdKv2nSpEls3rwZb29vQkJCsLW1JSoqitWrVzN69Gi1CdcFQRAEQRD+7f7TPazWr1+PgYEBgwcPzvF7KysrunfvzqFDh3IcPlBURkZGnD59mn79+rFw4UI8PT3ZuHEjH3/8sTwPhsry5cupWbMmH330Ec2aNZN7v3z66adMmTKF9evX4+3tzbp16wgPD6dNmzY57q+xsTFdunShWbNmzJkzJ9e0BQYGsm7dOq5fv063bt0YO3Ys9erV4/z58/IbuoT32+DBg2nVqhXu7u6MHj2aa9euMW3aNH7++ed38hTf1dWVy5cvY25uzsSJE2nfvj2jR4/m5MmTeTYw5cbJyYmMjAyCgoLw9PRkwIABJCQkcODAATp06JDrerVr1+b8+fPUrl2bMWPG0K1bN3788UciIiLkt8oVhL6+PgcPHqRWrVqMHDmSvn378vTpU06ePFnguFRyy/8qlStXpk2bNlhYWOT7VkRt9O/fn9OnT/PixQtGjhxJ+/btmTBhAlevXlXrkeTh4cG+ffv4559/6NOnD5MnT6Znz54MGTJEI87FixfTv39/QkND6dq1KxcuXODQoUMa4VSNWU2aNCEwMJBOnToxbdo0DA0N5Rv86tWrs2zZMq5fv46LiwvNmjXT6NWlolAo2LdvH5MnTyYiIgIvLy8WLVrEgAEDiI6O1ughI5R+udVlDg4OXL16FUdHR2bOnEmHDh0YOnQou3fvVjtvK1WqxJkzZ6hVqxajRo2iX79+GBgYEBISArx+aJMfS0tLoqKiMDY2pn///gwZMgRTU1N27NihEdbPz49hw4axevVqWrVqRbNmzXJ9C2v58uU5f/48bm5uBAYG4uPjI7/lcuXKlQU/WIIgCIIgCO8xhaTt68By8Sw1mVlXDpIhZRVXmgpFT6HD3KadsTAyKdF0CP9+yekJHPptAllSeommQ0ehj0/15Zjov92eUELp9/TpU2xtbRk3bhyhoaElnRzhLRP17tsxYsQItm/fTmJiYo69nwRBEARBEIR3q8gNVvD64vlFRlr+Ad8iUz3Df81Fs1D6JacnkJZZ+Inti4OhrplorPqPe/jwIffu3WPhwoVER0dz584dKleuXNLJEt4BUe8WTUhICJUqVaJGjRq8ePGCQ4cOsW7dOmbOnCn3tBIEQRAEQRBKVrHMYWVhZIIF7+dFqyAUhol+edFYJJS4devWERISgp2dHVu3bhWNVf8hot4tGn19fRYuXMjDhw/JyMigZs2aLFmyhAkTJpR00gRBEARBEIT/KZYeVoIgCIIgCIIgCIIgCIJQXP7Tk64LgiAIgiAIgiAIgiAIpY9osBIEQRAEQRAEQRAEQRBKFdFgJQiCIAiCIAiCIAiCIJQqosFKEARBEARBEARBEARBKFVEg5UgCIIgCIIgCIIgCIJQqogGK0EQBEEQBEEQBEEQBKFUEQ1WgiAIgiAIgiAIgiAIQqkiGqwEQRAEQRAEQRAEQRCEUkU0WAmCIAiCIAiCIAiCIAilimiwEgRBEARBEARBEARBEEoV0WAlCIIgCIIgCIIgCIIglCqiwUoQBEEQBEEQBEEQBEEoVUSDlSAIgiAIgiAIgiAIglCqiAYrQRAEQRAEQRAEQRAEoVQRDVaCIAiCIAiCIAiCIAhCqSIarARBEARBEARBEARBEIRS5f8DJy+rjjY51P0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1500x1400 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#vecs_Desc = cal_EmbVec(RefcodeDB, \"Description\")\n",
    "plot_cluster_umap(vecs_Desc, \"Task\", \"UMAPof Embedding of Description, Group by Task\")\n",
    "plot_cluster_umap(vecs_Desc, \"Language\", \"UMAPof Embedding of Description, Group by Language\")\n",
    "embedding_heatmap(vecs_Desc, RefcodeDB, \"Pairwise Embedding Similarity Heatmap of Description\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3c5b7f23",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qixin/miniconda3/envs/BIA/lib/python3.11/site-packages/umap/umap_.py:1945: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3IAAAJOCAYAAADyGDWDAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdeVxUVf/A8c+dgRmGfVEWEUUCF1RcUnMpl9y3tDTNSiW3UtPKzLJyq6csS1NbfPqZiFmpletj7mvmvqeJG6K44IqArMPM3N8fPEyODAgJIj7f9+s1L51zzz3n3MvAzHfOpqiqqiKEEEIIIYQQoszQlHYDhBBCCCGEEEIUjQRyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHLivvnhhx+oW7curq6uKIrCxIkTS7tJ+QoODiY4OLhU6o6OjkZRFKKjowt9jqIotGzZ0iYtMjISRVE4e/Zssbbvfrp06RIvvvgigYGBaDQaFEUp7SY9NOy9ZsSDaeLEiSiKwpYtW0q7KfdNy5Yt5fddCCHuQgK5ErJlyxYURSEyMrLIeXLfwBRFYc2aNfmeX79+fWu+Xbt25Ztv/PjxKIqCXq/nxo0b+ea7vV5FUdBoNHh5edG8eXOio6NRVTXfc+9mx44d9O3bl/T0dIYPH86ECRPu+iEyODjYpj32HklJSf+4TeLBFxkZyU8//USrVq0YN24cEyZMKDB/bhD8ySef3Jf2PcgfsB+2D8KxsbGMGjWKunXr4unpiaOjI+XLl6dFixZ89NFHXLhwobSb+NDIfW8q7EO+EBBCiNLhUNoNEPlzcHAgKiqKDh065Dl2+PBhDh48iIODAyaTKd8yLBYL8+bNQ1EUjEYjP/zwA6+99lqB9b755pu4urpiNps5c+YMS5YsYdu2bezfv58vv/zyH13LqlWrAPj+++9p3Lhxoc/TarW8//77+R53cnL6R+35XzB58mTeeecdAgMDS7sp/4jRaGTjxo20a9eOH374obSb89CJiYnB2dm5tJtRKNOmTePtt9/GbDbTuHFj+vbti7u7O4mJiezZs4dx48bxwQcfsHfvXiIiIkq7uWVecHBwni9Nzp49y7x586hTpw7du3fPk18IIcT9J4HcA6xjx44sX76cGzdu4OPjY3Nszpw5ODg40LZtW1avXp1vGevXryc+Pp6hQ4fy/fffM2fOnLsGcqNHj8bf39/6/MiRIzz22GN8/fXXjBo1iipVqhT5Wi5dugRgU25hODg4PNBDMB9kAQEBBAQElHYz/rHLly9jsViK/JoRhVO9evXSbkKhfPvtt7z55ps88sgj/Pzzz9SvXz9PnpiYGMaNG0dKSkoptPDhExwcnOfv7pYtW5g3bx5169aVv8lCCPGAkKGVD7CXXnoJo9HIjz/+aJNuNBr56aef6Ny5M76+vgWWMWfOHACGDRvG008/zZEjR9i7d2+R2lG7dm1atGiBqqrs37/fmr5jxw46d+6Mt7c3Tk5OVK9enYkTJ5Kenm7NkztEZ+7cuQBUqVLFOhynuOUOJcvKyuLdd9+lUqVKGAwGHn30UTZs2ADArVu3GDlyJIGBgTg5OdGkSRP27duXb5k3b95k8ODB+Pn5YTAYaNSoEStWrLCbV1VVoqKiaNasGe7u7jg7O9OgQQOioqLs5k9MTOSVV17Bz88PZ2dnGjZsyNKlSwu8xu+++45atWrh5OREUFAQY8aMITMz025ee3Pkcn8eEydO5MCBA7Rv3x43Nzc8PDx4+umn851Pt2TJEho0aIDBYMDPz4/Bgwdz8+bNIs8lTE9PZ+LEiVSvXh0nJye8vb3p3LkzO3bssMnXsmVLKleuDGDtUb6XeZW57UxLS2PUqFEEBgai1+uJiIjg119/zZM/OTmZ8ePHEx4ejqurKx4eHlSvXp2XXnqJ8+fPW9s4adIkAFq1amVt4+33I7fepKQkRo4cSVBQEA4ODtb5jwXdv/yGRqqqyrx582jevDmenp44OzsTFhbGK6+8Qnx8PJAz/23r1q3W/+c+bh/Gnd+QuBs3bvDGG29QpUoV9Ho9vr6+9O7dm2PHjuXJe/tr7JtvvqFGjRo4OTlRuXJlJk2ahMVisXtthXXz5k3GjBmDk5MTq1evthvEAdSoUYNff/01T29/7jVevHiRyMhI/P390Wg0NkNh582bR+PGjXF1dcXV1ZXGjRszb968PHUUNHf19t8re/WfP3+e3r174+Pjg4uLCy1btszzmi+s2bNnU7NmTZycnKhUqRJjx461+Rtw5swZNBoNnTt3tnv+zZs3cXJyok6dOv+o/jvt37+fV199lVq1auHh4YHBYKB27dp88sknZGdn58l/6tQpXnrpJapUqYKTkxPlypWjfv36vPnmm4Wq76effkKn0/Hoo49y9erVYrkGIYQoq6RH7gHWpEkTatSowdy5cxk5cqQ1fdmyZdy4cYMBAwawZMmSfM+/ceMGy5cvp169etSqVYt+/frxww8/MGfOHBo2bHhPbVu8eDHPPfccOp2O3r174+vry4YNG5g0aRLr1q1j8+bN6PV66xCdZcuWcfjwYV577TU8PT3vqe676d27N0eOHOGpp54iIyODH3/8kS5durBjxw5efvllMjMz6dmzJ9euXWPRokW0b9+euLg43N3dbcoxGo20adOGjIwM+vfvT1JSEgsXLqR79+7Mnz+fF154wZpXVVVefPFFfvrpJ6pWrcrzzz+PTqdj/fr1DBw4kGPHjvH5559b86enp9OyZUuOHDlCkyZNaNGihfXDXrt27exe14cffsj48eOtgZSjoyOLFi0iJiamyPdo3759fPbZZ7Rs2ZKXX36ZgwcPsmzZMo4cOcLRo0dthqxGRUUxcOBAPD096devHx4eHqxatYq2bduSnZ2No6NjoerMysqidevW7Nq1i/r16/P6669z9epVFi1axLp161i0aBHPPPMMkBMg1K1blxkzZtgM5bqXuTjZ2dm0a9eOxMREnnnmGdLT01m4cCG9evVizZo11vuuqirt27dn9+7dNGvWjA4dOqDRaDh79ixLly6lf//+BAUFWYOirVu30r9/f2tAdufrOysriyeffJJbt27RtWtXdDodfn5+/+gaVFWlT58+LFq0iMDAQPr06YO7uztnz55l0aJFdOjQgUqVKjFhwgSio6M5d+6czRC5unXrFlj+jRs3aNy4MadPn6Zly5Y899xznD17ll9//ZXffvuN9evX06RJkzznvfXWW2zZsoUuXbrQrl07li1bxsSJEzEajXz00Uf/6FoBfv31V1JSUujbty9hYWF3ze/gkPct7caNGzRp0gRvb2969+6N0Wi0/q6/8cYbTJ8+ncDAQAYOHIiiKCxevJjIyEgOHz7MtGnT/nHbc928eZNmzZoREBDAkCFDuHjxIosWLaJVq1asXbu2SK/pqVOnsmXLFnr37k2XLl1YtWoVn3zyCQcPHmT16tUoikJISAht2rRhzZo1XLhwgYoVK9qUMX/+fLKyshg8ePA9XxvkBJb/+c9/aN68OZ06dSI9PZ0tW7YwduxY9u7dy+LFi615L126RKNGjUhLS6Nz58707t2b1NRUTp06xZdffsnUqVMLrGvGjBm88cYbtGrVimXLluHm5lYs1yCEEGWWKkrE5s2bVUDt379/kfO0aNFCBdSEhAR1ypQpKqAeOHDAerxdu3aqn5+fmp2drfbv318F1J07d+Yp/4svvlABddq0aaqqqqrZbFYrVqyouru7q2lpaXny317v7f7880/VYDCoiqKocXFxakpKiurp6anq9Xr18OHD1nwWi0V9/vnnVUD98MMPbcrIbWdcXFy+9+NOlStXVrVarTphwgS7j1mzZtltf7NmzdTU1FRr+sKFC1VA9fT0VJ999lk1OzvbeuzTTz+1uUe31w2oTz75pGo0Gq3pMTExqsFgUD09PdWUlBRr+v/93/+pgDpw4ECb8rOystSuXbuqgLpv3z5r+oQJE1RAHTx4sE29a9euVQEVUOfOnWtNP3XqlOrg4KAGBgaqV65csaYnJyer1apVUwG1RYsWNmXZu+e5rzlAXbhwoU3+vn37qoC6YMECa9rNmzdVV1dX1c3NTY2NjbWmZ2dnq23atFEBtXLlymphfPDBByqgvvDCC6rFYrGmHz58WNXr9aqXl5fNPY2Li7vr79Cd5s6dqwLq5MmTbdJzf57dunVTs7KyrOkbNmxQAbV9+/bWtD///FMF1KeffjpP+ZmZmeqtW7esz3N/jps3b7bbntx627Vrp6anp9s9nt/9y3093+7rr79WAbV169Z5yktPT1dv3LhR4Pm3s/eaGTBggAqoY8eOtUlfs2aNCqhhYWGq2Wy2pue+xqpUqaJeunTJmn7t2jXV09NTdXNzs7nfRfXSSy+pgBoVFfWPzs99rb/00kuqyWSyOfb777+rgFqjRg01KSnJmp6UlKRWr15dBdRt27ZZ03NfW7f/XubK/b2aMGGC3fr79u1r85rfsmWLqiiKGhoaanM/85P7OnNyclKPHj1qTc/Ozlbbtm2rAur3339vTf/ll19UQJ00aVKesiIiIlQnJyf15s2bd63X3jXe+ft49uzZPPfWYrFYX0t//PGHNX3mzJkqoM6YMSNP+deuXbN5fufrd+zYsSqgPvvss/f0mhJCiIeJDK18wPXr18+66AnA+fPn2bBhgzW9IFFRUWi1Wvr06QOARqPhhRdeICUlxe5wslyff/45EydOZNy4cbzwwgs0atSIjIwMRowYQXBwMMuWLSMpKYkBAwbYLCyQu1rg7UPH7pXZbGbSpEl2H//+97/tnvPRRx/h4uJifd6zZ08cHR1JSkri888/t7lvuffm8OHDdsv68MMPbXqcqlevzoABA0hKSmL58uXW9K+++goXFxe++uorm/J1Op21R2LBggXW9O+//x6dTscHH3xgU1+7du1o3bp1nnb89NNPmEwmRo0aZTOc1t3dvcDFYPLTvHlzevfubZM2YMAAAJuht8uXLyc1NZVBgwYREhJiTXdwcODDDz8sUp3R0dE4OjryySef2AwZjIiIIDIykps3b9rc05LwxRdfoNPprM9bt25N5cqV7Q43NhgMedL0ej2urq5Frvezzz6zW15Rff3112i1WmbNmpWnPIPBgLe39z8u22g0smDBAnx8fPK8ptq3b0/79u05deqU3SGB48aNs5mPWa5cObp168atW7c4ceLEP27T5cuXAahQoUKeYwcOHGDixIk2j5UrV+bJp9PpmDJlClqt1iY992/UxIkT8fDwsKZ7eHhYezGL4++YVqvlo48+snnNt2jRgk6dOnH69OkiDbHs27cvNWvWtD53cHDg448/BrAZDtqtWzf8/PyYO3euzWrDe/fu5c8//6Rnz57FNjKicuXKee6toigMHz4cwDqs/Xb2fhfKlStnt3yz2cygQYOYPHkyQ4cOZeHChTa/w0II8b9MArkHnJ+fH506deKnn34iKyuL6OhoLBYLL730UoHn7dmzhyNHjtC2bVubxSL69+8PkO+8LcgZvjNp0iQ++ugjfvvtNxo2bEh0dDTTp08H4ODBg4D9YW5BQUE88sgjxMbGcuvWrSJebV56vR5VVe0+Dh06ZPecevXq2TzXarX4+vri6elJpUqVbI7lfvi8ePFinnIcHR3trrD5xBNPAFjrT09P58iRI3h6evLJJ5/k+XC5cOFCAI4fPw7kzNOLi4sjNDTU7kIeueXfLjfQtHfMXtrd2JtrlDsE6/YtHXLrbdq0aZ78jRo1uuuXCblSUlI4c+YMoaGheYZ6wd+vpfx+psXB09PT7kI9FStWtLnmGjVqULt2bX766SeaN2/OtGnT2Lt3L2az+R/V6+TkRO3atf9ps63S0tI4duwYVapUKdQww6I6fvw4GRkZNGrUyO5qlgX9jAr7eiqq24OQOx04cCDPlzv2ArkqVarYDRIK+jtWnK/HypUrExQUlCf9zr8jhWHvdz137urt5Tg6OjJgwADOnj3L+vXrrem5c6YHDRpU6Drvxmg0Mm3aNBo1aoS7u7t1v8dHH30U+HuhK4AuXbrg7OzM8OHD6dWrF1FRUZw8ebLA8p955hnmzJnDhAkT+Oabb9Bo5GOLEELkkjlyJST3zaagyf65x+72xvTSSy+xYsUKli5dSnR0NI0bN6ZGjRoFnpMbqPXt29cmvUaNGjRo0ICtW7dy+vRpQkND85ybkJBQ4EqBuSvD5TfPx9/fnxMnTpCSklIqcxjunOsGOd9c3/6t++3pgN1J+T4+PnZ/NrnXnZycDOTMgVFVlYsXL1oXv7AnLS3N5rz8Fqqxd18LOuefzLcq6F7cHrDk/qzLly+fJ79Go8n3W/Q7FeY1A39fZ0mwd82Qc923/546ODiwadMmJk6cyJIlS6yLMJQrV44RI0bw3nvv5emBKIivr2+xLO6TGxCV1HYS9/IzKuzrqahy22Lvi5ZBgwZZA5ItW7bQqlWrAsu4U0pKChqNxu5r28/PD41GUyyvx7v9nheljvzK8vX1zXOPBg8ezCeffMJ3331Hu3btSE9PZ8GCBVStWpUWLVoUus676dmzJ//5z3+oWrWqdb507giIGTNmkJWVZc1bpUoVdu7cyaRJk1i9ejW//PILANWqVePDDz/k2WefzVP+tm3bMBgMdOzYsdjaLIQQDwv5aquE5H6wKWgD7uvXr9vkzU+XLl3w8/Nj9OjRnDlzxjoELj+5b9gAL7zwQp7NW3NXaSyoV64guYHSlStX7B7PTbcXUJUlN27csBuI515f7s8t9zofffTRfHsPVVVl8+bNNvnzW3HN3n3NrcveOfn9HIpDbluvXbuW55jFYrG+hgtbTll5zZQrV46vvvqKixcvcuzYMb766it8fHyYMGECU6ZMKVJZBQVxGo0m330g7/yAn/sasBfUFIcH8WeU2xOc+7vzT+R3/93d3bFYLHZf21evXsVisdhca+6XOvZ+XgUFY3f7Pb/b3//ClHX16tU85VSpUoW2bduyfPlyrl+/zs8//0xKSkqx9sbt3buX//znP7Rv355jx44xe/ZsPvroIyZOnMhzzz1n95yIiAgWL15MYmIiO3fuZPz48Vy5coXevXuzffv2PPk3btyIwWCgXbt27Nq1q9jaLoQQDwMJ5EpItWrV0Ol07N27N98Pajt37gS46wa2Dg4OvPjii1y8eBFnZ+c8c5vulLvSW926dRk4cKDdh6OjI/PmzftH35bnDl28fQnvXBcvXiQ2NpaQkJAyv6JYdna23Q8O27ZtA/5eAdDNzY0aNWoQExNTqGFk7u7uVKlShdOnT1vnANkr/3a5S4XbO2Yvrbjk1mtvHs+ePXsK3Iz+du7u7oSEhHD69Gm7gUjuUvl3W1XxflMUhRo1ajB8+HDrELXbt5/I7Zn7p71OXl5eXL16Nc99TEtL49SpUzZprq6uhIeHExcXl+eYPUVtW+6WEHv37rXZQiRXafyMevbsiZubG7/88kuhrrkoCvo7Zu9avby8APuBdO4wTXvOnTtn3bLidnf+HSkMe7/r+/btIyMjw245Q4YMwWg0WvcQdXR0tA6vLw6xsbEAdO7cOU8v9d3+LuUOXZ80aRIzZ85EVVW7Q2Pr1avHxo0bcXBwoH379hLMCSHEbSSQKyFOTk706tWLa9eu8a9//SvP8SNHjvDdd9/h5ubG008/fdfy3nrrLZYuXcratWvv+o147jyIL774gu+++87uo0uXLly6dKnAzcTz061bNzw8PJg7dy5//fWXNV1VVcaOHUt2drbNflVl2bhx42yGXR4/fpyoqCg8PDzo1q2bNX3kyJGkp6czePBg6xDK28XFxdns0da3b1+MRiPjx4+3ybdu3To2btyY5/znn38erVbLtGnTbL6VT0lJsfv6Ki7dunXD1dWV7777jri4OGu6yWRi3LhxRSqrf//+ZGdnM3bsWJu5T0ePHmXu3Ll4eHhYtxkoTXFxcXb3TMvtQbl9oYbcxUUuXLjwj+pq0KAB2dnZNntF5v4e2XsdDR8+HLPZzLBhw8jIyLA5lpmZSWJi4j9um06no0+fPly/fp3JkyfbHNuwYQOrV68mNDSUZs2aFfr67AkODs6zv2F+vL29mTx5MllZWXTs2DHfgOmfzMPLDWgmTZpks5F4SkqKdYj07UFP/fr1URSFhQsX2uzbdurUKWbMmJFvPWazmffee8/mNb9161ZWrVpFaGio3fmn+Zk/f77N31yTycS7776bp625unXrhr+/P1OnTuWPP/7gqaeeuuveo0WRu9fjH3/8YZP+119/5XkNQU4PXkGjCvJbEKhu3bps2rQJR0dH2rVr94/34BNCiIeNzJErQVOnTmX37t3WSfgtWrTAycmJkydPsmLFClRV5ccffyzU6mF+fn6F+pB7+vRpfv/9d0JCQgqcB/HSSy+xdOlS5syZQ5cuXYpwVTm9K7Nnz6ZPnz489thj9O7dm/Lly7Nx40b27dtHo0aNeOutt4pUZn5MJlOBm0BHRkYWaUPqoggICCApKYm6devSuXNnkpOTWbBgAZmZmcyePdumx/Hll19m165dzJs3j+3bt9OmTRsqVKjAlStXOH78OLt37+ann36ytnXMmDEsWbKE2bNn89dff9G8eXPOnz/Pzz//TOfOnfntt99s2hIaGsr48eOZMGECERER9OrVCwcHBxYvXkzt2rXvaWXAgnh6ejJt2jSGDBlC/fr16d27t3UfOb1eT4UKFQq9+MCYMWP47bffmD9/PjExMbRu3dq6l192djbff//9A9GLe/jwYZ5++mkaNmxIrVq18Pf35+LFiyxbtgytVmuzcXHuRuDvvfcex48fx8PDAw8PD4YOHVqoul599VXmzp3LoEGDWL9+PeXLl2fbtm0kJSVRp06dPKupDh06lK1bt/Lzzz8TFhbGU089hbu7O/Hx8axdu5Y5c+ZY/048+eST/Prrrzz77LN06tTJuuhKfhtFA3z66ads3bqVf/3rX+zYsYPHHnvMuo+cs7Mzc+fOvefFJnKHKxd2oZzhw4eTlpbGu+++y6OPPkrjxo1p0KABbm5u3Lhxg5iYGLZt24Zery/S/pjNmzdnxIgRfPnll9SqVYsePXqgqipLlizh/PnzjBw5kubNm1vzBwYG0rt3bxYuXMijjz5Khw4duHr1KkuXLqVDhw42+6XdLiIigi1bttC4cWOefPJJLl26xMKFC3F0dGT27NlFup9t2rShcePGPPfcc3h7e7Nq1SqOHj1K+/btefHFF/Pkd3BwYMCAAdaVLYtzWCXkLHjUqFEjfv75ZxISEmjcuDHx8fGsWLGCzp0751kd+ccff+Sbb76hZcuWhIaG4u7uzrFjx1i1ahXlypUrcNpAnTp12LRpE61bt6ZDhw6sWbOmSEGwEEI8lO7jVgf/k5KSktQJEyaoderUUV1cXFRHR0c1KChIff755232hrtdfvu52XPnPnLvvPOO3X3c7pSdna36+fmpDg4O6uXLl4tcr6rm7MPUsWNH1dPTU9XpdGrVqlXVcePG2ezhdmc7i7qPHP/dhym/x+37dxW0b1ZB+3VhZz+t3Pw3btxQBw0apPr6+qp6vV5t0KCBunz58nzbvGjRIrVNmzaql5eX6ujoqAYGBqotW7ZUp06dmmefpBs3bqhDhgxRy5cvrzo5OamPPvqoumTJkgL3q5o9e7YaHh6u6nQ6tWLFiuro0aPV9PT0Iu8jd+d+V6pa8L5tv/zyi1qvXj1Vr9ervr6+6qBBg9QbN26orq6uap06dfK9H3dKTU1Vx40bp1atWlXV6XSqp6en2rFjR5v9ugrTnvwUtI9cYfdrO3/+vPrOO++ojRs3Vn19fVWdTqdWqlRJ7dmzp7p79+4850dHR6u1a9dW9Xp9nn31Cqo318aNG9XHHntM1ev1qo+Pj9q3b1/18uXL+b6eLRaL+t1336mNGzdWXVxcVGdnZzUsLEx95ZVX1Pj4eGu+7OxsdcyYMWqlSpVUBweHPPfS3mtGVXP28xo5cqRauXJl1dHRUS1Xrpzas2dP9ciRI3nyFvR7bW+PvZs3b6oajUZt1qxZgffEnhMnTqgjR45Ua9eurbq7u6sODg5quXLl1CeeeEL94IMPbK79btd4u6ioKLVhw4aqs7Oz6uzsrDZs2DDffevS0tLUESNGqH5+fqper1cjIiLUH3/8scB95Fq0aKGeO3dOffbZZ1UvLy/VYDCozZs3t9lf7W5uv5fffvutGh4erur1erVixYrqO++8Y3ePwlwnTpxQAbVSpUqF2rMuP/ntI3f16lV1wIABaoUKFVQnJye1du3a6tdff62eOXMmT/5du3apL7/8slqrVi3V09NTNRgMalhYmDpy5Mg8P7/8Xv+HDx9Wy5Urp7q6utr9uyGEEP9LFFUtYH1nIYTIx+nTpwkLC6NXr14sWrSotJsjyoCVK1fStWtXfvvtNzp16lTazSlxiqLQokULu/Pw7peff/6Z3r17M2nSpDxDuYUQQpRtMkdOCFGgmzdv2iwhDpCRkcEbb7wB8EDMaxNlw7Zt26hTp87/RBD3IFBVlWnTpuHg4MDAgQNLuzlCCCGKmfTICSEKtGzZMgYOHEi7du2oVKkS169fZ9OmTZw9e5Ynn3yS9evXyya9QthRWj1yR44cYeXKlezYsYOVK1fyyiuvMGvWrPvaBiGEECVPFjsRQhSoZs2atG3blu3bt7Ns2TIgZ/GVDz/8kNGjR0sQJ8QDZv/+/bz77rt4enrSr18/Pvvss9JukhBCiBIgPXJCCCGEEEIIUcbIV+lCCCGEEEIIUcZIICeEEEIIIYQQZYzMkcuHxWLh0qVLuLm5oShKaTdHCCGEEPdIVVVu3bpFhQoVZH6vEKLMk0AuH5cuXSIoKKi0myGEEEKIYnb+/HkqVqxY2s0QQoh7IoFcPtzc3ICcP/bu7u6l3BohhBBC3KuUlBSCgoKs7/FCCFGWSSCXj9zhlO7u7hLICSGEEA8RmTIhhHgYyABxIYQQQgghhChjJJATQgghhBBCiDJGAjkhhBBCCCGEKGMkkBNCCCGEEEKIMkYCOSGEEEIIIYQoYySQE0IIIYQQQogyRgI5IYQQQgghhChjJJATQgghhBBCiDJGAjkhhBBCCCGEKGMkkBNCCCGEEEKIMkYCOSGEEEIIIYQoYySQE0IIIYQQQogyRgI5IYQQQgghhChjJJATQgghhBBCiDJGAjkhhBBCCCGEKGMkkBNCCCGEEEKIMkYCOSGEEEIIIYQoYySQE0IIIYQQQogyRgI5IYQQQgghhChjJJATQgghhBBCiDKm1AO5WbNmERERgbu7O+7u7jRp0oTVq1dbj6uqysSJE6lQoQIGg4GWLVvy119/3bXcxYsXEx4ejl6vJzw8nKVLl5bkZQghhBBCCCHEfVPqgVzFihX55JNP2LdvH/v27ePJJ5+kW7du1mBtypQpTJs2ja+++oq9e/fi7+9P27ZtuXXrVr5l7ty5k969e9O3b18OHz5M37596dWrF7t3775flyWEEEIIIYQQJUZRVVUt7Ubcydvbm88++4wBAwZQoUIFXn/9dd5++20AsrKy8PPz49NPP+Xll1+2e37v3r1JSUmx6dnr0KEDXl5eLFiwoFBtSElJwcPDg+TkZNzd3e/9ooQQQghRquS9XQjxMCn1Hrnbmc1mFi5cSFpaGk2aNCEuLo7Lly/Trl07ax69Xk+LFi3YsWNHvuXs3LnT5hyA9u3bF3iOEEIIIYQQQpQVDqXdAIAjR47QpEkTMjMzcXV1ZenSpYSHh1sDLz8/P5v8fn5+nDt3Lt/yLl++bPecy5cv53tOVlYWWVlZ1ucpKSn/5FKEEEIIIYQQosQ9ED1y1apV49ChQ+zatYuhQ4fSv39/jh07Zj2uKIpNflVV86TdqajnTJ48GQ8PD+sjKCjoH1yJEEIIIYQQQpS8ByKQ0+l0hIaG0qBBAyZPnkydOnWYMWMG/v7+AHl60q5evZqnx+12/v7+RT5n7NixJCcnWx/nz5+/hysSQgghhBBCiJLzQARyd1JVlaysLKpUqYK/vz/r16+3HjMajWzdupWmTZvme36TJk1szgFYt25dgefo9XrrFgi5DyGEEEIIIYR4EJX6HLl3332Xjh07EhQUxK1bt1i4cCFbtmxhzZo1KIrC66+/zscff0xYWBhhYWF8/PHHODs78/zzz1vL6NevH4GBgUyePBmA1157jebNm/Ppp5/SrVs3li9fzoYNG/jjjz9K6zKFEEIIIYQQotiUeiB35coV+vbtS0JCAh4eHkRERLBmzRratm0LwJgxY8jIyGDYsGHcvHmTxx57jHXr1uHm5mYtIz4+Ho3m787Fpk2bsnDhQt5//33GjRvHI488wqJFi3jsscfu+/UJIYQQQgghRHF7IPeRexDIXjNCCCHEw0Xe24UQD5MHco6cEEIIIYQQQoj8SSAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcaUeiA3efJkGjZsiJubG76+vnTv3p0TJ07Y5FEUxe7js88+y7fc6Ohou+dkZmaW9CUJIYQQQgghRIkq9UBu69atDB8+nF27drF+/XpMJhPt2rUjLS3NmichIcHmERUVhaIo9OjRo8Cy3d3d85zr5ORU0pckhBBCCCGEECXKobQbsGbNGpvnc+fOxdfXl/3799O8eXMA/P39bfIsX76cVq1aERISUmDZiqLkOVcIIYQQQgghyrpS75G7U3JyMgDe3t52j1+5coXffvuNgQMH3rWs1NRUKleuTMWKFenSpQsHDx4s1rYKIYQQQgghRGl4oAI5VVUZNWoUjz/+OLVq1bKbZ968ebi5ufHMM88UWFb16tWJjo5mxYoVLFiwACcnJ5o1a8apU6fs5s/KyiIlJcXmIYQQQgghhBAPIkVVVbW0G5Fr+PDh/Pbbb/zxxx9UrFjRbp7q1avTtm1bvvzyyyKVbbFYqF+/Ps2bN2fmzJl5jk+cOJFJkyblSU9OTsbd3b1IdQkhhBDiwZOSkoKHh4e8twshHgoPTI/ciBEjWLFiBZs3b843iNu2bRsnTpxg0KBBRS5fo9HQsGHDfHvkxo4dS3JysvVx/vz5ItchhBBCCCGEEPdDqS92oqoqI0aMYOnSpWzZsoUqVarkm3fOnDk8+uij1KlT5x/Vc+jQIWrXrm33uF6vR6/XF7lcIYQQQgghhLjfSj2QGz58OD/99BPLly/Hzc2Ny5cvA+Dh4YHBYLDmS0lJ4ZdffmHq1Kl2y+nXrx+BgYFMnjwZgEmTJtG4cWPCwsJISUlh5syZHDp0iK+//rrkL0oIIYQQQgghSlCpB3KzZs0CoGXLljbpc+fOJTIy0vp84cKFqKpKnz597JYTHx+PRvP3SNGkpCSGDBnC5cuX8fDwoF69evz+++80atSo2K9BCCGEEEIIIe6nB2qxkweJTIgWQgghHi7y3i6EeJg8MIudCCGEEEIIIYQoHAnkhBBCCCGEEKKMkUBOCCGEEEIIIcoYCeSEEEIIIYQQooyRQE4IIYQQQgghyhgJ5IQQQgghhBCijJFATgghhBBCCCHKGAnkhBBCCCGEEKKMkUBOCCGEEEIIIcoYCeSEEEIIIYQQooyRQE4IIYQQQgghyhgJ5IQQQgghhBCijJFATgghhBBCCCHKGAnkhBBCCCGEEKKMkUBOCCGEEEIIIcoYCeSEEEIIIYQQooyRQE4IIYQQQgghyhgJ5IQQQgghxEMjMjKS7t27l3YzHhhyPx5eEsgJIYQQQoj7SlGUAh+RkZH3vU2bN2+mU6dO+Pj44OzsTHh4OG+++SYXL14sdBktW7bk9ddfL7lG/gMzZswgOjq6tJshSoAEckIIIYQQ4r5KSEiwPqZPn467u7tN2owZM+5re7799lvatGmDv78/ixcv5tixY/z73/8mOTmZqVOn3te2FBez2YzFYsHDwwNPT8/Sbo4oARLICSGEEEKI+8rf39/68PDwQFEU63NHR0deeeUVKlasiLOzM7Vr12bBggU25//666/Url0bg8GAj48Pbdq0IS0tzW5d+/fvx9fXl48++sju8QsXLjBy5EhGjhxJVFQULVu2JDg4mObNm/Pdd98xfvx4AG7cuEGfPn3ybVdkZCRbt25lxowZ1p7Fs2fPAnDs2DE6deqEq6srfn5+9O3bl+vXr1vPvXXrFi+88AIuLi4EBATwxRdf5Ondu3nzJv369cPLywtnZ2c6duzIqVOnrMejo6Px9PRk5cqVhIeHo9frOXfuXJ6hlaqqMmXKFEJCQjAYDNSpU4dff/3Vpp4XXniB8uXLYzAYCAsLY+7cuQX/QEWpkEBOCCGEEEI8MDIzM3n00UdZuXIlR48eZciQIfTt25fdu3cDOb15ffr0YcCAAcTExLBlyxaeeeYZVFXNU9aWLVto3bo1kyZN4r333rNb3y+//ILRaGTMmDF2j+f2Zt2tXTNmzKBJkyYMHjzY2rMYFBREQkICLVq0oG7duuzbt481a9Zw5coVevXqZa1j1KhRbN++nRUrVrB+/Xq2bdvGgQMHbNoRGRnJvn37WLFiBTt37kRVVTp16kR2drY1T3p6OpMnT+a7777jr7/+wtfXN8/1vP/++8ydO5dZs2bx119/8cYbb/Diiy+ydetWAMaNG8exY8dYvXo1MTExzJo1i3LlyuX34xKlyKG0GyCEEEIIIUSuwMBARo8ebX0+YsQI1qxZwy+//MJjjz1GQkICJpOJZ555hsqVKwNQu3btPOUsX76cvn378u2339KnT5986zt16hTu7u4EBATcU7s8PDzQ6XQ4Ozvj7+9vzTdr1izq16/Pxx9/bE2LiooiKCiIkydPEhAQwLx58/jpp59o3bo1AHPnzqVChQo2bVyxYgXbt2+nadOmAPz4448EBQWxbNkynn32WQCys7P55ptvqFOnjt1rSEtLY9q0aWzatIkmTZoAEBISwh9//MG3335LixYtiI+Pp169ejRo0ACA4ODgAu+LKD0SyAkhhBBCiAeG2Wzmk08+YdGiRVy8eJGsrCyysrJwcXEBoE6dOrRu3ZratWvTvn172rVrR8+ePfHy8rKWsXv3blauXMkvv/zC008/XWB9qqqiKMo9tys/+/fvZ/Pmzbi6uuY5FhsbS0ZGBtnZ2TRq1Mia7uHhQbVq1azPY2JicHBw4LHHHrOm+fj4UK1aNWJiYqxpOp2OiIiIfNty7NgxMjMzadu2rU260WikXr16AAwdOpQePXpw4MAB2rVrR/fu3a3Bo3iwSCAnhBBCCCEeGFOnTuWLL75g+vTp1K5dGxcXF15//XWMRiMAWq2W9evXs2PHDtatW8eXX37Je++9x+7du6lSpQoAjzzyCD4+PkRFRdG5c2d0Ol2+9VWtWpXk5GQSEhIK7JW7W7vyY7FY6Nq1K59++mmeYwEBAdZ5bncGk7cPFbU3bDQ3/fbzDAZDgUGpxWIB4LfffiMwMNDmmF6vB6Bjx46cO3eO3377jQ0bNtC6dWuGDx/O559/XtBlilIgc+SEEEIIIcQDY9u2bXTr1o0XX3yROnXqEBISYrOoB+QEPc2aNWPSpEkcPHgQnU7H0qVLrcfLlSvHpk2biI2NpXfv3jbzyO7Us2dPdDodU6ZMsXs8KSmp0O3S6XSYzWabtPr16/PXX38RHBxMaGiozcPFxYVHHnkER0dH9uzZYz0nJSXFpuzw8HBMJpN1Ph7kLL5y8uRJatSoke+13Sl3EZT4+Pg8bQkKCrLmK1++PJGRkfzwww9Mnz6d//u//yt0HeL+kUBOCCGEEEI8MEJDQ609bjExMbz88stcvnzZenz37t18/PHH7Nu3j/j4eJYsWcK1a9fyBDS+vr5s2rSJ48eP06dPH0wmk936goKC+OKLL5gxYwYDBw5k69atnDt3ju3bt/Pyyy/z4YcfFqpdkDOfbPfu3Zw9e5br169jsVgYPnw4iYmJ9OnThz179nDmzBnWrVvHgAEDMJvNuLm50b9/f9566y02b97MX3/9xYABA9BoNNbetbCwMLp168bgwYP5448/OHz4MC+++CKBgYF069at0PfWzc2N0aNH88YbbzBv3jxiY2M5ePAgX3/9NfPmzQNg/PjxLF++nNOnT/PXX3+xcuXKIgWL4v6RQE4IIYQQQjwwxo0bR/369Wnfvj0tW7bE39/fZvl8d3d3fv/9dzp16kTVqlV5//33mTp1Kh07dsxTlr+/P5s2beLIkSO88MILeXrLcg0bNox169Zx8eJFnn76aapXr86gQYNwd3e3LnByt3YBjB49Gq1WS3h4OOXLlyc+Pp4KFSqwfft2zGYz7du3p1atWrz22mt4eHig0eR8FJ82bRpNmjShS5cutGnThmbNmlGjRg2cnJysZc+dO5dHH32ULl260KRJE1RVZdWqVTg6Ohbp/n744YeMHz+eyZMnU6NGDdq3b89//vMf67BUnU7H2LFjiYiIoHnz5mi1WhYuXFikOsT9oaj5Dbr9H5eSkoKHhwfJycm4u7uXdnOEEEIIcY/kvV2UFWlpaQQGBjJ16lQGDhxY2s0RDyhZ7EQIIYQQQohSdPDgQY4fP06jRo1ITk7mgw8+ACjSsEnxv0cCOSGEEEIIIUrZ559/zokTJ9DpdDz66KNs27ZNNuIWBZJATgghhBBCiFJUr1499u/fX9rNEGWMLHYihBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCEeKJGRkXTv3r20myHEA00COSGEEEKIYqJaLCSfPM+1vcdJPnke1WK5L/Xu2LEDrVZLhw4d7kt9d9qyZQuKopCUlFSk886ePYuiKBw6dMgmfcaMGURHRxdb+4R4GMmG4EIIIYQQxeDGwVOc/n4dScfOYc40onXS4RlemdB+7fCpF1aidUdFRTFixAi+++474uPjqVSpUonWV9I8PDxKuwlCPPCkR04IIYQQ4h7dOHiKQx/M5/q+k+i83XELC0Tn7c71fSc59MF8bhw8VWJ1p6Wl8fPPPzN06FC6dOli05OV21O2ceNGGjRogLOzM02bNuXEiRPWPBMnTqRu3brMnz+f4OBgPDw8eO6557h165Y1T1ZWFiNHjsTX1xcnJycef/xx9u7dC+T0qrVq1QoALy8vFEUhMjISgDVr1vD444/j6emJj48PXbp0ITY21lpulSpVAKhXrx6KotCyZUsg79DKguov7HUK8bCRQE4IIYQQ4h6oFgunv19H5vVkPGpUQufujEarRefujEeNSmReT+b0/PUlNsxy0aJFVKtWjWrVqvHiiy8yd+5cVFW1yfPee+8xdepU9u3bh4ODAwMGDLA5Hhsby7Jly1i5ciUrV65k69atfPLJJ9bjY8aMYfHixcybN48DBw4QGhpK+/btSUxMJCgoiMWLFwNw4sQJEhISmDFjBpATZI4aNYq9e/eyceNGNBoNTz/9NJb/3os9e/YAsGHDBhISEliyZIndayyo/qJcpxAPEwnkhBBCCCHuQcrpiyQdO4dzxfIoimJzTFEUnCuWJ+mvs6Scvlgi9c+ZM4cXX3wRgA4dOpCamsrGjRtt8nz00Ue0aNGC8PBw3nnnHXbs2EFmZqb1uMViITo6mlq1avHEE0/Qt29faxlpaWnMmjWLzz77jI4dOxIeHs7s2bMxGAzMmTMHrVaLt7c3AL6+vvj7+1uHRvbo0YNnnnmGsLAw6taty5w5czhy5AjHjh0DoHz58gD4+Pjg7+9vLed2d6u/KNcpxMNEAjkhhBBCiHtgTE7DnGnEwUVv97iDsx5zphFjclqx133ixAn27NnDc889l1OXgwO9e/cmKirKJl9ERIT1/wEBAQBcvXrVmhYcHIybm5tNntzjsbGxZGdn06xZM+txR0dHGjVqRExMTIHti42N5fnnnyckJAR3d3frUMr4+PhCX2NR6r/bdQrxMJHFToQQQggh7oHOwwWtkw5TWhY6d+c8x03pWWiddOg8XIq97jlz5mAymQgMDLSmqaqKo6MjN2/etKY5Ojpa/5/ba2i5bajn7cdz8+Qezx2meWdvo6qqedLu1LVrV4KCgpg9ezYVKlTAYrFQq1YtjEZjoa+xKPXf7TqFeJhIj5wQQgghxD1wDw3EM7wy6Reu5Zmbpqoq6Reu4VkzGPfQwHxK+GdMJhPff/89U6dO5dChQ9bH4cOHqVy5Mj/++GOx1BMaGopOp+OPP/6wpmVnZ7Nv3z5q1KgBgE6nA8BsNlvz3Lhxg5iYGN5//31at25NjRo1bILL/M77J/UL8b9IeuSEEEIIIe6BotEQ2q8dqWevkBwTj3PF8jg46zGlZ5F+4RpO5T0I7dsWRVO835+vXLmSmzdvMnDgwDzL9ffs2ZM5c+bwxRdf3HM9Li4uDB06lLfeegtvb28qVarElClTSE9PZ+DAgQBUrlwZRVFYuXIlnTp1wmAw4OXlhY+PD//3f/9HQEAA8fHxvPPOOzZl+/r6YjAYWLNmDRUrVsTJySnPtRSmfiH+F0mPnBBCCCHEPfKpF0bd8X0p16AqxsQUbp2+iDExhXINq1F3XN8S2Uduzpw5tGnTxu6eaz169ODQoUMcOHCgWOr65JNP6NGjB3379qV+/fqcPn2atWvX4uXlBUBgYCCTJk3inXfewc/Pj1dffRWNRsPChQvZv38/tWrV4o033uCzzz6zKdfBwYGZM2fy7bffUqFCBbp16/aP6hfif5Gi3jkGQACQkpKCh4cHycnJuLu7l3ZzhBBCCHGP7sd7u2qxkHL6IsbkNHQeLriHBhZ7T5wQQoAMrRRCCCGEKDaKRoNH1aDSboYQ4n+AfEUkhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGVMqQdykydPpmHDhri5ueHr60v37t05ceKETZ7IyEgURbF5NG7c+K5lL168mPDwcPR6PeHh4SxdurSkLkMIIYQQQggh7ptSD+S2bt3K8OHD2bVrF+vXr8dkMtGuXTvS0tJs8nXo0IGEhATrY9WqVQWWu3PnTnr37k3fvn05fPgwffv2pVevXuzevbskL0cIIYQQosyZOHEidevWtT6PjIyke/fuJVpny5Ytef31163Pg4ODmT59eonWWdYoisKyZctKtI7o6Gg8PT1LtA5RMko9kFuzZg2RkZHUrFmTOnXqMHfuXOLj49m/f79NPr1ej7+/v/Xh7e1dYLnTp0+nbdu2jB07lurVqzN27Fhat24tfyCEEEIIUWJUi4Ws+IukHztFVvxFVIulxOu8fPkyI0aMICQkBL1eT1BQEF27dmXjxo0lXrcoWQkJCXTs2LHYyrMXLPfu3ZuTJ08WWx3i/nEo7QbcKTk5GSBPoLZlyxZ8fX3x9PSkRYsWfPTRR/j6+uZbzs6dO3njjTds0tq3b59vIJeVlUVWVpb1eUpKyj+8AiGEEEL8L0o/cYabqzeRGXce1WhE0elwqhKEV8cnca4WUiJ1nj17lmbNmuHp6cmUKVOIiIggOzubtWvXMnz4cI4fP14i9Yr7w9/fv8TrMBgMGAyGEq9HFL9S75G7naqqjBo1iscff5xatWpZ0zt27MiPP/7Ipk2bmDp1Knv37uXJJ5+0CbzudPnyZfz8/GzS/Pz8uHz5st38kydPxsPDw/oICgoqnosSQgghxEMv/cQZrkQtJOP4abTubugqBqJ1dyPj+GmuRC0k/cSZEql32LBhKIrCnj176NmzJ1WrVqVmzZqMGjWKXbt2WfMlJyczZMgQfH19cXd358knn+Tw4cP3VHdWVhZjxowhKCgIvV5PWFgYc+bMsR4/duwYnTp1wtXVFT8/P/r27cv169fvqc6oqChq1qyJXq8nICCAV1991XosPj6ebt264erqiru7O7169eLKlSvW47nDR6OioqhUqRKurq4MHToUs9nMlClT8Pf3x9fXl48++simTkVR+Pbbb+nSpQvOzs7UqFGDnTt3cvr0aVq2bImLiwtNmjQhNjbWeo69oamvv/46LVu2tD5v2bIlI0eOZMyYMXh7e+Pv78/EiRPz1H370MoLFy7w3HPP4e3tjYuLCw0aNLBOG4qNjaVbt274+fnh6upKw4YN2bBhg019586d44033rCuOQH2h1bOmjWLRx55BJ1OR7Vq1Zg/f36edn333Xc8/fTTODs7ExYWxooVK+z/0ESJeaACuVdffZU///yTBQsW2KT37t2bzp07U6tWLbp27crq1as5efIkv/32W4Hl5b5Ac6mqmict19ixY0lOTrY+zp8/f28XI4QQQoj/CarFws3VmzAnp6CrHITWxRlFq0Hr4oyuchDm5BRurt5c7MMsExMTWbNmDcOHD8fFxSXP8dwP56qq0rlzZy5fvsyqVavYv38/9evXp3Xr1iQmJv7j+vv168fChQuZOXMmMTEx/Pvf/8bV1RXIGRLYokUL6taty759+1izZg1XrlyhV69e/7i+WbNmMXz4cIYMGcKRI0dYsWIFoaGh1mvs3r07iYmJbN26lfXr1xMbG0vv3r1tyoiNjWX16tWsWbOGBQsWEBUVRefOnblw4QJbt27l008/5f3337cJggE+/PBD+vXrx6FDh6hevTrPP/88L7/8MmPHjmXfvn0ANkFlYc2bNw8XFxd2797NlClT+OCDD1i/fr3dvKmpqbRo0YJLly6xYsUKDh8+zJgxY7D893WVmppKp06d2LBhAwcPHqR9+/Z07dqV+Ph4AJYsWULFihX54IMPrGtO2LN06VJee+013nzzTY4ePcrLL7/MSy+9xObNm23yTZo0iV69evHnn3/SqVMnXnjhhXt6PYmie2CGVo4YMYIVK1bw+++/U7FixQLzBgQEULlyZU6dOpVvHn9//zy9b1evXs3TS5dLr9ej1+uL3nAhhBBC/E8zXkggM+48DuXL5fnCWFEUHMqXIzMuHuOFBPSVAout3tOnT6OqKtWrVy8w3+bNmzly5AhXr161ftb5/PPPWbZsGb/++itDhgwpct0nT57k559/Zv369bRp0waAkJC/h4/OmjWL+vXr8/HHH1vToqKiCAoK4uTJk1StWrXIdf7rX//izTff5LXXXrOmNWzYEIANGzbw559/EhcXZx1VNX/+fGrWrMnevXut+SwWC1FRUbi5uREeHk6rVq04ceIEq1atQqPRUK1aNT799FO2bNlis0L6Sy+9ZA1C3377bZo0acK4ceNo3749AK+99hovvfRSka8pIiKCCRMmABAWFsZXX33Fxo0badu2bZ68P/30E9euXWPv3r3WKUi5gSxAnTp1qFOnjs39Wrp0KStWrODVV1/F29sbrVaLm5tbgUM2P//8cyIjIxk2bBiAtXf3888/p1WrVtZ8kZGR9OnTB4CPP/6YL7/8kj179tChQ4ci3wfxz5R6j5yqqrz66qssWbKETZs2UaVKlbuec+PGDc6fP09AQEC+eZo0aZLnG41169bRtGnTe26zEEIIIUQuc2o6qtGIxsnJ7nGNkx7VaMScml6s9aqqCuQdgXSn/fv3k5qaio+PD66urtZHXFyczXDA/Pz44482523bto1Dhw6h1Wpp0aJFvnVu3rzZ5rzcgLMwdd7p6tWrXLp0idatW9s9HhMTQ1BQkM3UmPDwcDw9PYmJibGmBQcH4+bmZn3u5+dHeHg4Go3GJu3q1as25UdERNgcB6hdu7ZNWmZmZpHXWLi9XMjprLiz7lyHDh2iXr16+S74l5aWxpgxY6zX7erqyvHjx609coUVExNDs2bNbNKaNWtmcx/vbLuLiwtubm75tl2UjFLvkRs+fDg//fQTy5cvx83NzdqL5uHhgcFgIDU1lYkTJ9KjRw8CAgI4e/Ys7777LuXKlePpp5+2ltOvXz8CAwOZPHkykPPNSPPmzfn000/p1q0by5cvZ8OGDfzxxx+lcp1CCCGEeDhpXZ1RdDosmZloXZzzHLdkZqHodGhd8x67F2FhYSiKQkxMTIFbBVgsFgICAtiyZUueY4VZdv6pp57isccesz4PDAy0mXuVX51du3bl008/zXOsoC/i83O3xTjymz5zZ7qjo6PNcUVR7KZZ7hgGe3ue3PLspeWep9ForIF2ruzs7DztK0zdue52D9566y3Wrl3L559/TmhoKAaDgZ49e2I0Ggs8z57CTE8qSttFySj1HrlZs2aRnJxMy5YtCQgIsD4WLVoEgFar5ciRI3Tr1o2qVavSv39/qlatys6dO22+UYmPj7cZ69u0aVMWLlzI3LlziYiIIDo6mkWLFtn8IRJCCCGEuFe6igE4VQnCdO16ng/vqqpiunYdpyqV0FUsegBTEG9vb9q3b8/XX3+dZ/9dgKSkJADq16/P5cuXcXBwIDQ01OZRrly5u9bj5uZmc47BYKB27dpYLBa2bt1q95z69evz119/ERwcnKdOe/P5CtOG4ODgfLdUCA8PJz4+3maNg2PHjpGcnEyNGjWKXN+9Kl++fJ45aIcOHbqnMiMiIjh06FC+89C2bdtGZGQkTz/9NLVr18bf35+zZ8/a5NHpdJjN5gLrqVGjRp6Ojx07dpTKfRQFK/VATlVVu4/IyEgg59uHtWvXcvXqVYxGI+fOnSM6OjrPqpJbtmwhOjraJq1nz54cP34co9FITEwMzzzzzH26KiGEEEL8r1A0Grw6PonWwx3jufOY09JRzWbMaekYz51H6+GOV8dWKJri/9j1zTffYDabadSoEYsXL+bUqVPExMQwc+ZMmjRpAkCbNm1o0qQJ3bt3Z+3atZw9e5YdO3bw/vvvWxfqKKrg4GD69+/PgAEDWLZsGXFxcWzZsoWff/4ZyBlxlZiYSJ8+fdizZw9nzpxh3bp1DBgw4K6BRH4mTpzI1KlTmTlzJqdOneLAgQN8+eWX1muMiIjghRde4MCBA+zZs4d+/frRokULGjRo8I/quxdPPvkk+/bt4/vvv+fUqVNMmDCBo0eP3lOZffr0wd/fn+7du7N9+3bOnDnD4sWL2blzJ5AzX27JkiUcOnSIw4cP8/zzz+fpIQsODub333/n4sWL+a4g+tZbbxEdHc2///1vTp06xbRp01iyZAmjR4++p/aL4lfqgZwQQgghRFnnXC0EvwHPYageijnlFsYLlzCn3MJQPQy/Ac+V2D5yVapU4cCBA7Rq1Yo333yTWrVq0bZtWzZu3MisWbOAnCFvq1atonnz5gwYMICqVavy3HPPcfbs2XwXgSuMWbNm0bNnT4YNG0b16tUZPHiwtWewQoUKbN++HbPZTPv27alVqxavvfYaHh4eNvPRiqJ///5Mnz6db775hpo1a9KlSxfrwne5y/R7eXnRvHlz2rRpQ0hIiHWE1/3Wvn17xo0bx5gxY2jYsCG3bt2iX79+91SmTqdj3bp1+Pr60qlTJ2rXrs0nn3yCVqsF4IsvvsDLy4umTZvStWtX2rdvT/369W3K+OCDDzh79iyPPPII5cuXt1tP9+7dmTFjBp999hk1a9bk22+/Ze7cuTZbJ4gHg6LeOQZAADkbgnt4eJCcnIy7u3tpN0cIIYQQ9+h+vLerFgvGCwmYU9PRujqjqxhQIj1xQghR6oudCCGEEEI8LBSNpli3GBBCiPzIV0RCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQDzlFUVi2bFlpN8OuiRMnUrdu3dJuhhBljgRyQgghhBDFRLVYMF++gCnuBObLF1AtlvtS7+XLlxkxYgQhISHo9XqCgoLo2rUrGzduLPa6tmzZgqIoJCUlFUt5o0ePLpF2CvGwcyjtBgghhBBCPAxM8acx7liPOeEcZGeDoyPagMromrbFoVJoidV79uxZmjVrhqenJ1OmTCEiIoLs7GzWrl3L8OHDOX78eInVfS9UVcVsNuPq6oqrq2tpN0eIMkd65IQQQggh7pEp/jSZ//kB07lTKC7uKH4VUFzcMZ07lZMef7rE6h42bBiKorBnzx569uxJ1apVqVmzJqNGjWLXrl158tvrUTt06BCKonD27FkAzp07R9euXfHy8sLFxYWaNWuyatUqzp49S6tWrQDw8vJCURQiIyOBnMBsypQphISEYDAYqFOnDr/++mueeteuXUuDBg3Q6/Vs27Ytz9DKyMhIunfvzueff05AQAA+Pj4MHz6c7Oxsa56EhAQ6d+6MwWCgSpUq/PTTTwQHBzN9+vRiu69CPOikR04IIYQQ4h6oFgvGHeuxpKagCQhCUZScAwZnNE5BWBLOY9y5Hm3FEBRN8X6HnpiYyJo1a/joo49wcXHJc9zT0/MflTt8+HCMRiO///47Li4uHDt2DFdXV4KCgli8eDE9evTgxIkTuLu7YzAYAHj//fdZsmQJs2bNIiwsjN9//50XX3yR8uXL06JFC2vZY8aM4fPPPyckJARPT0+2bt2ap/7NmzcTEBDA5s2bOX36NL1796Zu3boMHjwYgH79+nH9+nW2bNmCo6Mjo0aN4urVq//oWoUoqySQE0IIIYS4B5arlzAnnEPjVe7vIO6/FEVB41UO86VzWK5eQutfsVjrPn36NKqqUr169WItNz4+nh49elC7dm0AQkJCrMe8vb0B8PX1tQaKaWlpTJs2jU2bNtGkSRPrOX/88QfffvutTSD3wQcf0LZt2wLr9/Ly4quvvkKr1VK9enU6d+7Mxo0bGTx4MMePH2fDhg3s3buXBg0aAPDdd98RFhZWbNcvRFkggZwQQgghxD1QM9Jy5sTp9fYz6PWQlJ2Tr7jrVlWAPAHkvRo5ciRDhw5l3bp1tGnThh49ehAREZFv/mPHjpGZmZknQDMajdSrV88mLTf4KkjNmjXRarXW5wEBARw5cgSAEydO4ODgQP369a3HQ0ND8fLyKtS1CfGwkDlyQgghhBD3QDG4gKMjZGXZz5CVBY6OOfmKWVhYGIqiEBMTU+hzNP8d3pkbBAI2888ABg0axJkzZ+jbty9HjhyhQYMGfPnll/mWafnv6py//fYbhw4dsj6OHTtmM08OsDsE9E6Ojo42zxVFsdZxe7tvl1+6EA8rCeSEEEIIIe6BxrcC2oDKWG5ezxNMqKqK5eZ1tBUqo/GtUOx1e3t70759e77++mvS0vL2+NnbIqB8+fJAzoIhuQ4dOpQnX1BQEK+88gpLlizhzTffZPbs2QDodDoAzGazNW94eDh6vZ74+HhCQ0NtHkFBQfdyiXlUr14dk8nEwYMHrWmnT58utu0QhCgrJJATQgghhLgHikaDrmlbNK7uWBLOo2ako1rMqBnpWBLOo3F1R9ekbbEvdJLrm2++wWw206hRIxYvXsypU6eIiYlh5syZ1vlqt8sNriZOnMjJkyf57bffmDp1qk2e119/nbVr1xIXF8eBAwfYtGkTNWrUAKBy5cooisLKlSu5du0aqampuLm5MXr0aN544w3mzZtHbGwsBw8e5Ouvv2bevHnFer3Vq1enTZs2DBkyhD179nDw4EGGDBmCwWAo9iGmQjzIJJATQgghhLhHDpVCcer6Ig6Vw1DTUlCvXEJNS8EhOCwnvQT3katSpQoHDhygVatWvPnmm9SqVYu2bduyceNGZs2alSe/o6MjCxYs4Pjx49SpU4dPP/2Uf/3rXzZ5zGYzw4cPp0aNGnTo0IFq1arxzTffABAYGMikSZN455138PPz49VXXwXgww8/ZPz48UyePJkaNWrQvn17/vOf/1ClSpViv+bvv/8ePz8/mjdvztNPP83gwYNxc3PDycmp2OsS4kGlqDKg2K6UlBQ8PDxITk7G3d29tJsjhBBCiHt0P97bVYsFy9VLqBlpKAYXNL4VSqwnTvztwoULBAUFsWHDBlq3bl3azRHivpBVK4UQQgghiomi0RT7FgMir02bNpGamkrt2rVJSEhgzJgxBAcH07x589JumhD3jQRyQgghhBCiTMnOzubdd9/lzJkzuLm50bRpU3788cc8q10K8TCToZX5kKGVQgghxMNF3tuFEA8TGbQthBBCCCGEEGWMBHJCCCGEEEIIUcZIICeEEEIIIYQQZYwEckIIIYQQQghRxkggJ4QQQgghhBBljARyQgghhBBCCFHGSCAnhBBCCCEeSBMnTqRu3boF5jl79iyKonDo0KH70qZc0dHReHp63tc67Snu6w8ODmb69OnFUpYoWRLIPYTMxmzMxuzSboYQQgjxP0dVLajJCajXYnP+VS0lXufly5cZMWIEISEh6PV6goKC6Nq1Kxs3bizxukva6NGjba4jMjKS7t272+QJCgoiISGBWrVq3efWPZz27t3LkCFDSrsZohAcSrsBoniYjdlc2nCAc0v/IDnmHADuoYFUevpxAts3xMFJV8otFEIIIR5u6vWzWE7/jnrzIliyQeOI4hWIJrQ5SrngEqnz7NmzNGvWDE9PT6ZMmUJERATZ2dmsXbuW4cOHc/z48RKp935xdXXF1dW1wDxarRZ/f//71KKHX/ny5Uu7CaKQpEfuIWBKz2TvmP9j39v/x9Wdf6FaLKgWC9f2nWD/u3PYM+objClppd1MIYQQ4qGlXj+L5eBi1OtnQO8Kbv6gd0W9fua/6WdLpN5hw4ahKAp79uyhZ8+eVK1alZo1azJq1Ch27dplzRcfH0+3bt1wdXXF3d2dXr16ceXKFevx3CGMUVFRVKpUCVdXV4YOHYrZbGbKlCn4+/vj6+vLRx99ZFO/oijMmjWLjh07YjAYqFKlCr/88otNniNHjvDkk09iMBjw8fFhyJAhpKamWo9v2bKFRo0a4eLigqenJ82aNePcuXM27cr9/7x581i+fDmKoqAoClu2bLE7tHDr1q00atQIvV5PQEAA77zzDiaTyXq8ZcuWjBw5kjFjxuDt7Y2/vz8TJ060afe0adOoXbs2Li4uBAUFMWzYMJt2F8bbb79N1apVcXZ2JiQkhHHjxpGd/feoqdzrmz9/PsHBwXh4ePDcc89x69Yta541a9bw+OOP4+npiY+PD126dCE2NtZufaqqEhoayueff26TfvToUTQajfW8iRMnUqlSJfR6PRUqVGDkyJHWvHcOrSworyhdEsg9BP6avpiLa/fiVN4D95AAnMp54FQu5//OAV4kbDrIn58sKJnK1WzIPg8Z+yF9F2QeAtNlUM0lU58QQgjxgFFVS05PXNYt8AhE0RlQNBoUnQE8AlGzbuUcL+ZhlomJiaxZs4bhw4fj4uKS53ju/C1VVenevTuJiYls3bqV9evXExsbS+/evW3yx8bGsnr1atasWcOCBQuIioqic+fOXLhwga1bt/Lpp5/y/vvv2wSIAOPGjaNHjx4cPnyYF198kT59+hATEwNAeno6HTp0wMvLi7179/LLL7+wYcMGXn31VQBMJhPdu3enRYsW/Pnnn+zcuZMhQ4agKEqe6xk9ejS9evWiQ4cOJCQkkJCQQNOmTfPku3jxIp06daJhw4YcPnyYWbNmMWfOHP71r3/Z5Js3bx4uLi7s3r2bKVOm8MEHH7B+/XrrcY1Gw8yZMzl69Cjz5s1j06ZNjBkzphA/mb+5ubkRHR3NsWPHmDFjBrNnz+aLL77Ic9+XLVvGypUrWblyJVu3buWTTz6xHk9LS2PUqFHs3buXjRs3otFoePrpp7FY8r6eFEVhwIABzJ071yY9KiqKJ554gkceeYRff/2VL774gm+//ZZTp06xbNkyateubbf9Rckr7j8ZWllKVFXl5pEzJB4+gyXbhHMFH/ybR+Dg7FSkctITbnB+5U70ni44uhryHHdwdsKpnDsJGw9wKy4BtyoBxXUJYL4JmYdBze3t04DZAqZLoPEEpzqgyfvGIoQQQjxUUq7kDKd09skTgCiKgursk3M85Qp4FN/78OnTp1FVlerVqxeYb8OGDfz555/ExcURFBQEwPz586lZsyZ79+6lYcOGAFgsFqKionBzcyM8PJxWrVpx4sQJVq1ahUajoVq1anz66ads2bKFxo0bW8t/9tlnGTRoEAAffvgh69ev58svv+Sbb77hxx9/JCMjg++//94abH711Vd07dqVTz/9FEdHR5KTk+nSpQuPPPIIADVq1LB7Ha6urhgMBrKysgocSvnNN98QFBTEV199haIoVK9enUuXLvH2228zfvx4NJqcfoyIiAgmTJgAQFhYGF999RUbN26kbdu2ALz++uvWMqtUqcKHH37I0KFD+eabbwq837d7//33rf8PDg7mzTffZNGiRTYBocViITo6Gjc3NwD69u3Lxo0brb2fPXr0sClzzpw5+Pr6cuzYMbvzAl966SXGjx/Pnj17aNSoEdnZ2fzwww989tlnQE7vrL+/P23atMHR0ZFKlSrRqFEju+0vSl5x/0mPXClIPnWB7YM+Z1vkFP78+EeOfv4ze96cxcanxxP38xZUVS10WVe2HcGYlIbexz3fPDpPN4wp6Vzeerg4mp/DkgqZB0FNB8UFNG45QZvWDRQDWG7m9M6psuiKEEKIh5wxPWdOnIPe/nEHXc5xY3qxVpv7ecFe79XtYmJiCAoKsgZxAOHh4Xh6elp7ziAn0MgNJgD8/PwIDw+3Bj65aVevXrUpv0mTJnme55YbExNDnTp1bHoMmzVrhsVi4cSJE3h7exMZGUn79u3p2rUrM2bMICEhobC3IN/rbdKkic19adasGampqVy4cMGaFhERYXNeQECAzbVt3ryZtm3bEhgYiJubG/369ePGjRukpRV+usqvv/7K448/jr+/P66urowbN474+HibPHfe9zvbERsby/PPP09ISAju7u5UqVIFIE85t5/fuXNnoqKiAFi5ciWZmZk8++yzQE7gnZGRQUhICIMHD2bp0qU2w05vV5S84v6TQO4+u3XmErtGfMmV7UdxdHfGLbQC7qEVcAkqT8aVmxz+6EdOf78u3/NVsxE1KQb13BLUU1EYT60Hcxqo2eQX/ymanHHkxqRinCdnjAc1IyeIU+54GSnanHRzEpju7Y+xEEII8cDTOYPGEUxZ9o+bjDnHdc7FWm1YWBiKotgEY/aoqmo32Lsz3dHR0ea4oih20+wN6btTbrn51X17nrlz57Jz506aNm3KokWLqFq1ap7hm0Vhr057QW9B13bu3Dk6depErVq1WLx4Mfv37+frr78GsJnjVpBdu3bx3HPP0bFjR1auXMnBgwd57733MBqNNvnudo+7du3KjRs3mD17Nrt372b37t0Aecq53aBBg1i4cCEZGRnMnTuX3r174+yc8/oLCgrixIkTfP311xgMBoYNG0bz5s3tXldR8or7TwK5+yxm1gpS4y7j/kgFdO7O1j8oWp0jrpV80ThqOfHtf0i7eC3PuWr2LTi/Ai6th/QLYDbi4GRBNWWhpiVAdpLdYE5VVVSLiqNb3qGX/4iaDeYEwAHy+xZQ0eQcy75YPHUKIYQQDyp3PxSvQEi/kWdUjaqqkH4j57i7X7FW6+3tTfv27fn666/t9hIlJSUBOb1v8fHxnD9/3nrs2LFjJCcn5zuMsSjuDLp27dplHe4ZHh7OoUOHbNq3fft2NBoNVatWtabVq1ePsWPHsmPHDmrVqsVPP/1kty6dTofZXPA8/PDwcHbs2GHzs9ixYwdubm4EBgYW6pr27duHyWRi6tSpNG7cmKpVq3Lp0qVCnZtr+/btVK5cmffee48GDRoQFhZmXcSlsG7cuEFMTAzvv/8+rVu3pkaNGty8efOu53Xq1AkXFxdmzZrF6tWrGTBggM1xg8HAU089xcyZM9myZQs7d+7kyJEjdssqSl5xf0kgdx+lnr/Kld+P4FTOHUVr/9Yb/L3ISrzFxdV7bNJV1fx3AOfoDjpvcHTFt1FldG5OGFOMkJUEpryrKRmT03B0M+D3eDFNTlWzQDWB4niXjA458+eKMFRUCCGEKGsURZOzxYDeDZIvohozUC1mVGMGJF9EcXLPOX7nCJZi8M0332A2m2nUqBGLFy/m1KlTxMTEMHPmTOuQxzZt2hAREcELL7zAgQMH2LNnD/369aNFixY0aNDgntvwyy+/EBUVxcmTJ5kwYQJ79uyxLmbywgsv4OTkRP/+/Tl69CibN29mxIgR9O3bFz8/P+Li4hg7diw7d+7k3LlzrFu3jpMnT+YbYAYHB/Pnn39y4sQJrl+/brdnaNiwYZw/f54RI0Zw/Phxli9fzoQJExg1apTNMNGCPPLII5hMJr788kvOnDnD/Pnz+fe//12k+xIaGkp8fDwLFy4kNjaWmTNnsnTp0iKV4eXlhY+PD//3f//H6dOn2bRpE6NGjbrreVqtlsjISMaOHUtoaKjN8Nfo6GjmzJnD0aNHrddmMBioXLlynnKKklfcfxLI3UcpJy+QnZKGziv//VAUjQaNg5abR8/aHkg7D+mXcoI4zd9r1LhWcCWgaQWykkyYskxgTLb5BsqcaSTzahK+zWrhXrViMV2JFlCAuw2tUHPy3mXsvhBCCFHWKeWC0dTrgVIuBLJS4dYVyEpFKReCpu4zJbaPXJUqVThw4ACtWrXizTffpFatWrRt25aNGzcya9asnLYpCsuWLcPLy4vmzZvTpk0bQkJCWLRoUbG0YdKkSSxcuJCIiAjmzZvHjz/+SHh4OADOzs6sXbuWxMREGjZsSM+ePWndujVfffWV9fjx48fp0aMHVatWZciQIbz66qu8/PLLdusaPHgw1apVo0GDBpQvX57t27fnyRMYGMiqVavYs2cPderU4ZVXXmHgwIE2C4/cTd26dZk2bRqffvoptWrV4scff2Ty5MlFui/dunXjjTfe4NVXX6Vu3brs2LGDcePGFakMjUbDwoUL2b9/P7Vq1eKNN96wLlpyNwMHDsRoNObpjfP09GT27Nk0a9aMiIgINm7cyH/+8x98fHzylFGUvOL+U9SirKzxPyQlJQUPDw+Sk5Nxd89/IZGiuLh+H7tGfoX7IxVQNPkHN7fOJBDQuj6NZ7xqTVMTNsLNo6DP+4tjvGVk7+Q9XD1wBUVR0fkGoGh0ZCWloWabKNeoOo2mDsWpnEexXAeqChm7chY00eQTlKoqqKngUBmc8q6oJIQQQtxvJfHefidVteSsTmlMz5kT5+5XIj1xDwpFUVi6dCndu3cv7aaIO2zfvp2WLVty4cIF/PyKd1iveDDI9gP3kfsjFXB0M2BMTkXv5WY3j2pRsZjMeFQLsj2QnZp3UZH/0rnpaPT+Y5zfGM+51SdJu5oCqgk3Pz3B7atQ6ZnHcHQrxn3dFAUcgyDrZs58OXtDLNUswAEcCzcWXQghhHgYKIqmWLcYEKKosrKyOH/+POPGjaNXr14SxD3EJJC7j9xCKuDbOJyLa/eh83C12yuXeT0JnYcrFTvesUeHVg8FbCTq6OxISEd/qrSArFQH0BjQeRrQaNWcvd7OnUb1b4niHmZznmrMwnTiMNmxMWDMRPHwxrFGfbRBIQUvZ+xQAcyJYLrw32BOT85IXTOomYACuqo5+8kJIYQQQoj7YsGCBQwcOJC6desyf/780m6OKEEytDIfJTX8IunYWXYOn0nGlUScA8vhYMjZc8ZiNpN5NQlzhpFqr3Ql/NWnbc5Tk2Pg4nrQeeQs738n1QSp5wEVXCrZzKNDVSE7CTR6qPQUiiFnE01T/GkylkVjuX4ZLBbQaHL+1elxrFobw1P9UZwL2NBbtUD2OciOz9lPDhXQgMYdHCvnBHsyP04IIcQD4n4MrRRCiPtFeuTuM8/wYBp9MYzD//qBlJMXMGeb/tvzpaL38aDa4C6EDeyIJekG2Uf3Yjp1FNWYhcbbBwefLBzKJ6IYyuUNkLKScoI5fTnbIA7+OxTSE4w3cubZGfwxX75Axs/fYklORONdHsUhZ3ikqqqQmU720X1gNmPoPRTFIZ+XiaIBXRVwrASW5P+uZKkDjYcEcEIIIYQQQpQgCeRKgU/dUFouHMfVHX+R+OcZVJMZQ4A3FdrUx8nHg+yje8lctQBLShJotSgaLeYLZ8jGgra8HkOzbDQu7qDV5fS2mdPBdAu0Tjk9dvYoCmgNkBqHas4ka9cGLDevo/GrYDMJW1EUMLig0WjJPnkEx9i/cKxWp+ALUrSg9S6+GySEEEIIIYQokARypUTjoMW/eQT+zSNs0k1nYshY/j1qdhaa8gEot+13ohqzMF++Qsaeazi39EAx/3cumr48qGbQOOW7IAoAigNYjFiSrmKKOYji4pbvSlqK3gmSE8k+vOvugZwQQgghhBDivpJA7gGiqipZO9ahZqSh8a2QZ7ERRadH4+OH+fJNTGo9HENCcwI3Bzc4Mx9M6XepwAyKBvVWCmpWJhp3zwKzKzo9lquX7vGqhBBCCCGEEMWtSBubXLx4kd27d3Pz5k27x69fv873339fLA37X2S5fhnz2ZMobh75rhip6PRgsWD66zCK3htF54mi0YJbKFiMOUMt82POAOdAFEfXnKGWd1vnxmKB/ObHCSGEEEIIIUpNoQI5i8XCoEGDqFSpEk2bNsXf35+33nqL7Oxsm3yxsbG89NJLJdLQh5FqNqImH0e9uhP16k4sCUdQs7NyhjXmIzPbTJJR5eDhk7y+4CAfrzzG1hNXyXKuClpnyE62H6Blp+YsguJVC41vBTTuXqhpKfm3TVVRs41oQ2oUx6UKIYQQQgghilGhArk5c+YQHR3NsGHDiI6Opk+fPnzxxRe0bt2aW7dulXQbH0pq8gmI+wkuroPre3IeN3dDdgqqMdXuOTdSjZy4fIuUdCMpZoWb6UYOX0jim02n+df6KyS5NwWNIxhvkHLmCn9FHWbj4LWsffE//P7OLs7ucMJoLofiqEP36OOo2UbUbKP99qXcRDE4o4t4rCRvgxBCCCHuE0VRWLZsWWk3o0BbtmxBURSSkpLua70TJ06kbt261ueRkZF07979vrahqO5XG4ODg5k+fXqJ1yOKrlCB3Lfffsvrr7/Ol19+Sd++fYmOjmbVqlUcOXKEtm3bkpycXNLtfKioyScgYVNOL5nOA/Q+oPdB6xuIxsURNekyqinN5pzkjGzOJ6ZjsVhw0li4Vj6Mcq56Kno54+vuxPHLKUzbpZJdoTPndjqwbcxuTiw4SerlNIwZGhJPmjn4+R/8MeAzbsUloGv0JA6P1MSSeA1Lyk1UizmnbcYszNcvo5pN6Ft0QesfVBq3SAghhCiTVNWCmnkdNe1Czr+qpUTri4yMRFEUFEXB0dERPz8/2rZtS1RUFBaLbd0JCQl07NixRNtzr5o2bUpCQgIeHvmswn2fzJgxg+joaOvzli1b8vrrr5dae+6H6OhoPD0986Tv3buXIUOG3P8GibsqVCB36tSpPL/47dq1Y+PGjZw+fZo2bdrkO29O2FIt2XBtV86eazpPm829FZ0Ox2qPgMmEmnqN3L3aVeBqShZms4Xypluk61yJLVfNep7OQUOAh4GTV26xbc05/vz2L0yqO241quISUhXnSiG4hVbCtZIvSX+dZe9b32IygfNzQ9E1a4/iqMNy/QrmKxex3EpC6xuI4al+6Jq1v893RwghhCi71PRLcHENXFhp86+aXrILh3Xo0IGEhATOnj3L6tWradWqFa+99hpdunTBZDJZ8/n7+6PX60u0LfdKp9Ph7++f71oB94uHh4fdoKa4GY32R0Y9SMqXL4+zs3NpN0PYUahATqfT5ZkPB1C/fn3Wr19PXFwcrVu35saNG8XewIdO6tmceWwObnYP62o/gkNwBdRbqZivX8SUlUlmZhbmtFv4m25i0jryR3BrUvXutuc55PwoT3y/AdOtDFyCfNFodTkLofyXRueAa7A/STHnuLRhP4qTM4aOvXEZPhHnPsMxPDsEl35v4PLKOHT1mpX6H1EhhBCirFDTL8HlzZB+MWfOut4n59/0i3B5c4kGc3q9Hn9/fwIDA6lfvz7vvvsuy5cvZ/Xq1Ta9SrcPrTx79iyKovDzzz/zxBNPYDAYaNiwISdPnmTv3r00aNAAV1dXOnTowLVr12zqmzt3LjVq1MDJyYnq1avzzTffWI/llrtkyRJatWqFs7MzderUYefOndY8586do2vXrnh5eeHi4kLNmjVZtWoVYH9o5eLFi6lZsyZ6vZ7g4GCmTp1q057g4GA+/vhjBgwYgJubG5UqVeL//u//bPK8/fbbVK1aFWdnZ0JCQhg3bpzdz7a5bh+2GBkZydatW5kxY4a19zMuLo7Q0FA+//xzm/OOHj2KRqMhNja2wHInT55MhQoVqFq1KpCzoGDv3r3x8vLCx8eHbt26cfbsWet5ZrOZUaNG4enpiY+PD2PGjLF+4X/7fbhzCGTdunWZOHGi9XlSUhJDhgzBz88PJycnatWqxcqVK9myZQsvvfQSycnJ1mvMPe/OcuPj4+nWrRuurq64u7vTq1cvrly5Yj2eO0x1/vz5BAcH4+HhwXPPPSfTsUpAoQK5sLAwm1/A29WrV4/169cTHx9P//79i9yAyZMn07BhQ9zc3PD19aV79+6cOHHCejw7O5u3336b2rVr4+LiQoUKFejXrx+XLhX8BzE6Otr6Qrz9kZmZWeQ2FqusxJwuNo391SBVrYasx2tjqlORG1qVq1cvkXL1Ag4YifN6hLXVunHmtt6427klp8DxOPTl8l/1UqtzAEXhwuo91jSNizuO4fXR1WmMQ0gNFK3W7rlCCCHysqgWEtKvcjL5DKeS47iWcSPPByzxcFNVCyQeytkGSF8etPqc7YG0+pznpnRIPFTiwyxv9+STT1KnTh2WLFlSYL4JEybw/vvvc+DAARwcHOjTpw9jxoxhxowZbNu2jdjYWMaPH2/NP3v2bN577z0++ugjYmJi+Pjjjxk3bhzz5s2zKfe9995j9OjRHDp0iKpVq9KnTx9r7+Dw4cPJysri999/58iRI3z66ae4urrabd/+/fvp1asXzz33HEeOHGHixImMGzfOJkAFmDp1Kg0aNODgwYMMGzaMoUOHcvz4cetxNzc3oqOjOXbsGDNmzGD27Nl88cUXhbqXM2bMoEmTJgwePJiEhAQSEhKoVKkSAwYMYO7cuTZ5o6KieOKJJ3jkkUfyLW/jxo3ExMSwfv16Vq5cSXp6Oq1atcLV1ZXff/+dP/74wxpE5/bYTZ06laioKObMmcMff/xBYmIiS5cuLVT7c1ksFjp27MiOHTv44YcfOHbsGJ988glarZamTZsyffp03N3drdc4evToPGWoqkr37t1JTExk69atrF+/ntjYWHr37m2TLzY2lmXLlrFy5UpWrlzJ1q1b+eSTT4rUXnF3hVpb/sknn2TevHmMHz8erZ0P+fXq1WPDhg20adOmyA3YunUrw4cPp2HDhphMJt577z3atWvHsWPHcHFxIT09nQMHDjBu3Djq1KnDzZs3ef3113nqqafYt29fgWW7u7vbBIUATk75rwh5fyjkRHJ5mVULF1NTSMs24lMtgJgqj3I+UUd6ZgbxJi2pOl/cs/3yjb41t9LRmsw4GHQFtsDBSUfG5cR7uwwhhBBcTLvM0cQTJBtvof73b7tW0eKl96CuTzg+Tl6l3EJxX2QlQuZVcPTI2d7ndoqSk555NSefU7n71qzq1avz559/Fphn9OjRtG+fM5Xitddeo0+fPmzcuJFmzZoBMHDgQJug6cMPP2Tq1Kk888wzAFSpUoVjx47x7bff2nyhP3r0aDp37gzApEmTqFmzJqdPn6Z69erEx8fTo0cPateuDUBISEi+7Zs2bRqtW7dm3LhxAFStWpVjx47x2WefERkZac3XqVMnhg0bBuT0vn3xxRds2bKF6tWrA/D+++9b8wYHB/Pmm2+yaNEixowZU+D9gZxhljqdDmdnZ/z9/a3pL730EuPHj2fPnj00atSI7OxsfvjhBz777LMCy3NxceG7775Dp8v5vBYVFYVGo+G7776zfhE/d+5cPD092bJlC+3atWP69OmMHTuWHj16APDvf/+btWvX3rXtt9uwYQN79uwhJibG2hN4+7338MjpCLj9Gu2V8eeffxIXF0dQUM46CvPnz6dmzZrs3buXhg0bAjlBY3R0NG5uOSPQ+vbty8aNG/noo4+K1GZRsEIFcq+++ipNmjQhNTU138mndevWZceOHfn23OVnzZo1Ns/nzp2Lr68v+/fvp3nz5nh4eLB+/XqbPF9++SWNGjUiPj6eSpUq5Vv23V6MpcLJJ+ePusWUp1fuSloqqdlGnDWgUTSk6b3ICPAGFVITkjCSSbomEdes8jbnaTUqQd4Z6Kpl4+EKpuwMNHpHu71yqkXFmJSKOSub7S9Pw9HdGb/Ha1OhdX0cXQ0leulCCPEwuZCawJ5rh8i2mHB2MOCgcUBVVUyqiWuZiWy/so9mfg0kmPtfYM7MeV93dLR/XOMI2aacfPeRqqp3nSYRERFh/b+fnx+ANcDKTbt69SoA165d4/z58wwcOJDBgwdb85hMpjyfD28vNyAgAICrV69SvXp1Ro4cydChQ1m3bh1t2rShR48eNvlvFxMTQ7du3WzSmjVrxvTp0zGbzdYOhtvPz/38l9tugF9//ZXp06dz+vRpUlNTMZlMuLvbTlMpqoCAADp37kxUVBSNGjVi5cqVZGZm8uyzzxZ4Xu3ata1BHOT0Op4+fdoa9OTKzMwkNjaW5ORkEhISaNKkifWYg4MDDRo0KFLv/6FDh6hYsaI1iPsnYmJiCAoKsgZxAOHh4Xh6ehITE2MN5IKDg22uJyAgwObnIYpHoQI5f39/67cqBalateo9vTgA6wqY3t7eBeZRFOWuk1BTU1OpXLkyZrOZunXr8uGHH1KvXj27ebOyssjKyrI+T0nJf4+1f0pVVUzXjVhOJoHxPIq3Lw5BvihaLUaziVvZWTgoCm5KFjdxJgFPICfuK+9m4HySmSxtOgYlG62a82ZR2SeDZlWTcNNno62l58wKN1JiruHgYkbBGQUn65eD5iwjScfiybyWjM7blas7/8JiMnNh1W5OBPvz6L8G4FMvrNivWwghHjbZFhOHbhzDZDHj5uhq/bCsKAqOiiPujg6kZN/i8I1jtKrQVOYcP+y0Tjlfzlqyc4ZT3smSnXNce39HBcXExFClSpUC8zjeFnzmvk7vTMtd/TL339mzZ/PYY7bbE905YsteubnnDxo0iPbt2/Pbb7+xbt06Jk+ezNSpUxkxYkSe9tkLRu0FL453BNG3t3vXrl0899xzTJo0ifbt2+Ph4cHChQvzzLX7JwYNGkTfvn354osvmDt3Lr17977rwiAuLi42zy0WC48++ig//vhjnrzly5fPk5YfjUaT597cPg/QYLj3L+zz+3LgzvSCfh6i+BRqjtz9oqoqo0aN4vHHH6dWrVp282RmZvLOO+/w/PPPF/hNSvXq1YmOjmbFihUsWLAAJycnmjVrxqlTp+zmnzx5Mh4eHtbH7d80FAfzxTjSoqaQPn8GmXtOk7nvDBnrd5K2ZDPZsRdJzTaiVU34aDLJwoH9ajCW23485V31eDvpMalmrhqTSTeaCPBI48nw67g5GbmVqcXF4EXlnnUBDcakdFQ1FZWcb/8sZjM3j50j81oSWoMO71pVcAv2xyM0ENcgX1LjLrP3rW9JiS3ZlbWEEOJhcDHtMmmmdJwdna0fXkwWE0azkWxLzgcnZwcDiVnJXM+UVZ0fenpvcPLNWczsziBDVXPSnXxz8t0nmzZt4siRI9aheMXBz8+PwMBAzpw5Q2hoqM3jbgHjnYKCgnjllVdYsmQJb775JrNnz7abLzw8nD/++MMmbceOHVStWtXudB97tm/fTuXKlXnvvfdo0KABYWFhnDt3rkjt1el0mM3mPOmdOnXCxcWFWbNmsXr1agYMGFCkciFn8cBTp07h6+ub577mfi4NCAhg165d1nNMJhP79++3Kad8+fIkJCRYn6ekpBAXF2d9HhERwYULFzh58mSRrvF24eHhxMfHc/78eWvasWPHSE5OpkaNGkW6bnHvHqhA7tVXX+XPP/9kwYIFdo9nZ2fz3HPPYbFYbFZIsqdx48a8+OKL1KlThyeeeIKff/6ZqlWr8uWXX9rNP3bsWJKTk62P21+g98p88SzpC77GHHcCjcEVjV8lNP5VUNzcsCTdImPrXhxPnsJFMXEDV35Xq3EJ26E4igLB5VzxcnbEx9WRDKOJhiHJODmqmC0Ggsu54ensSMXu4VR+thbZyUZS425iTL5Jdlomt2ITyLqajNagw7NGJbROf3fpa3QOuIX4k3b+KnELNxXbdQshxMMq2ZiCiopW0ZBlNpKYmcT1zERuZN3keuZNrmcmkmUyYrKYSDYW/wgP8WBRFA141wUHZ8i6BuYsUC05/2ZdAwcX8K6bk68EZGVlcfnyZS5evMiBAwf4+OOP6datG126dKFfv37FWtfEiROZPHkyM2bM4OTJkxw5coS5c+cybdq0Qpfx+uuvs3btWuLi4jhw4ACbNm3KNwh488032bhxIx9++CEnT55k3rx5fPXVV3YX4shPaGgo8fHxLFy4kNjYWGbOnFnkhUKCg4PZvXs3Z8+e5fr169beJa1WS2RkJGPHjiU0NNRm+GNhvfDCC5QrV45u3bqxbds24uLi2Lp1K6+99hoXLlwAcuYvfvLJJyxdupTjx48zbNiwPJumP/nkk8yfP59t27Zx9OhR+vfvbxPstmjRgubNm9OjRw/rqvOrV6+2TnMKDg4mNTWVjRs3cv36ddLT0/O0tU2bNkRERPDCCy9w4MAB9uzZQ79+/WjRogUNGjQo8rWLe/PABHIjRoxgxYoVbN68mYoVK+Y5np2dTa9evYiLi2P9+vVFHtes0Who2LBhvj1yer0ed3d3m0dxUFWVzA2LsSTdQONbAcXJkLOCpoMBxT0IjV8V0DiTcSSR325V5DdLbeuQSjul4ax3oFMdf0Z19qZmRQ2+bh5UKeeKiz5nlKxGqyH8nRZETGyNV91ATOmZZCUmknXzFg7uznjXCcWpXN55jopGg87TlQtr92JMTi2WaxdCiIeVCigoZJgyuZmVRKY5E1DQoEVBIdtiIiU7lUxzFmY15xvubIuJ1Ow00k0ZsqrlQ0hxrgD+rcA5EMzpkHUj51/nQPBvmXO8hKxZs4aAgACCg4Pp0KEDmzdvZubMmSxfvrzQvVaFNWjQIL777juio6OpXbs2LVq0IDo6ukg9cmazmeHDh1OjRg06dOhAtWrV8v2Cvn79+vz8888sXLiQWrVqMX78eD744AObhU7uplu3brzxxhu8+uqr1jUdchdPKazRo0ej1WoJDw+nfPnyxMfHW48NHDgQo9H4j3rjAJydnfn999+pVKkSzzzzDDVq1GDAgAFkZGRYP4+++eab9OvXj8jISJo0aYKbmxtPP/20TTljx46lefPmdOnShU6dOtG9e/c8q2cuXryYhg0b0qdPH8LDwxkzZoy1F65p06a88sor9O7dm/LlyzNlypQ8bc3dwsLLy4vmzZvTpk0bQkJCWLRo0T+6dnFvFLWU301UVWXEiBEsXbqULVu2EBaWd45WbhB36tQpNm/eXKTxwrfX06hRI2rXrk1UVNRd86ekpODh4UFycvI9BXU5Qyo/Q9EbUAz2x0yrZjPZ1xL4reajxIXUwNXR/qqT19JTuZVtxEPnRE0fE80CjaRn6/B0csJT52R3DHnquUsoRn/2DF2MOdOIwS//SfemtEyybqbSctE4PMLyBtNCCCFyxKacY8/VQ2Sas1BVFY2izfM32GwxY1bN1PAKxdXBmfi0S2RbTCiAl86DKu6VqORaAU0J9dKIvIrrvb0gqmrJWZ3SnJkzJ07vXWI9ceLBsH37dlq2bMmFCxesC8YIcT8UarGTkjR8+HB++uknli9fjpubG5cvXwZylkA1GAyYTCZ69uzJgQMHWLlyJWaz2ZrH29vbuupPv379CAwMZPLkyUDOUreNGzcmLCyMlJQUZs6cyaFDh/j666/v6/WZr1xENWaheOQ/Ll7RatEqUD3byBFjFjqNFt0d36Bdz0jjSnoaTloHnBwcMDiCQjZZ5mwup2WTaTLh7+xq80FCURRcKnvgqPHHwdWAKS2jwLZazBYUjYLGsdRfFkII8UCr6OLPXkXBpJpxVHJXCVb/nvCvgoqKRlE4nXwWvUaHzkGHo8YBi6pyNTORa1mJXMu8waPlahcYzF1JuMXhfReIPXUds8mCfwV36jQI5JGwcmi0EiA8aBRFc1+3GBClJysri/PnzzNu3Dh69eolQZy47wr9iT0jI4Nly5Zx7tw5ypcvz1NPPfWPesbuNGvWLABatmxpkz537lwiIyO5cOECK1asAHK2OLjd5s2brefFx8ej0fz9hpa7c/3ly5fx8PCgXr16/P777zRq1Oie21wkhd78U6GaZzmqe5fn5M0bKIDhvyv+3MrKIjErHb3WgWAPLxw0GlKNCiqZODtqyDKpJGVlYHBwxFP/94pYORuPKiiKK37NanL6+/UFLkWclZiCR7VKuFS895/rg86smklIv0ZiVhKqasHFwZlAF38MDqW9z6AQoizQa/UYtE7cyk7DrJpQ1ZyNwW+nVTTWYZaeeg/0Wt1t5+swmo3EpcTjqXMnzCPvsDRVVdm++QwbV58gPc2Io6MWRaNw4VwSh/ZeoEaEPz2er4uTIZ8l74UQJWrBggUMHDiQunXrMn/+/NJujvgfVKihlZcuXaJ58+bExcVZx/V7eHiwevVqGjduXOKNLA3FNfzCdO4kafOmoTG4ojjZX/ZVtViwXEvAqcsLKA1bcPjaZfZducDV9Jy5amaLhSRjJkGuHmitwapK66AUfAxmUowassxmHBQtHg6uOUGgToOzLhtF0eOkbcLNI+f4Y+DnKBrF7vDK7NQMMq4mUff9Fwl57sl/fL1lwZWM6xy8fpQUYyoW1P9u0a5i0DpR1aMK1T1DZalwIcRdrTv/O5czrpJpNlo3A8+lkDNjLmdBFC0+Tl42gVyu1Ow0XB1daFfxCbSK7UiMA7vPs2TBYbRaBQ9P2+HzGRnZJCdlUr9REL361ZO/WYV0P4ZWCiHE/VKoMRnvv/8+Fy9e5P333+e3335j+vTp6HQ6hg4dWtLtK/O0QaFoAypjSUnMd3K7mnITxdUDx5oN0GkdaOhfkVciHuPthi14u2ELvA3OuOv0twVxAAp/Xncm06zg4mBGNVtIzcrm/M1ULiWlkZiWREJyFtdTKqIoDnhHPEL1l7tgzsrm1pkETGmZqBYVc1Y2aRevk3E5kcB2Daj8zBP358aUkuuZiey6coBk4y2cHQx46Nxw17nh5uiGSTVzJPEEx27aXxBHCCFup9M6YrSY0CpaHDWOOCgOOCgO6DSO6P+7l1hOgJezuqU9Tlo9qdlp3MxKtkk3mSxs23gai8WCp5chT6BmMDji7q7nr8MJXLpge25JM5stXIhP4syp61xJSJGFW4QQopQUamjl+vXreffdd60r/HTs2JFHHnmEp556iitXrsiY4AIoGg1OrbuTvujfWK5fRuNZDuW/QyZVsxk15SaqxYJT605o3P5eTVJRFJwccvIZzWa78yeuZTiy/owTtXzS8HUFdw1oDSqqCjdSHVn1pxOXk24wqn05wvzcCBvYCSc/L2Lnryf55AUsWdkoDlqcA7yp3KM5of3bo9U9vEN0VFXlaOIJMsyZuDu62Xww0igKzg4G0k0ZnEw+Q2W3irg6FryhpxDif5uDxgFVVXHQaHPmReXpFPs7wLmzty2XRtGgqhayLSab9LjT17l6ORUPz/w38DU4O5KcnMlfhxIIDPIsUtvVzBTUC3+iXjsN5mxwLY+mYh3wCc63d89strB3xzn2bo/n6pVbmM0qjo4agoK9aNYyhBq1/YvUBiGEEPemUIHc5cuXad68uU1ay5YtUVVVArlCcAipgXPPwWSsXojlxhUwm3M2hgM07p7on+iI7rHW+Z7v5WTgZmbehUosqsrei0Z2ntMQWh78XVScje4kp+k5d90JkwWup6Xz485zTOhWE0VRqNS1KRU7PcbNP8+QlXgLrUGPd50QHF3y/7BQkAyjmcPnk7iRmoWDVkP1ADcqeTs/kMN8kowp3Mi6iUGb99vtXAatEynZtzifeokaXqH3uYVCiLLkljENB40Ws2rJ2XTgtr8rqvr3YEuVnHm5Dkret1yzakajaNFrbIdd3krOwmy2oNPlv3S8oihoFIWbiQUvZHUny4XDWA6vgIwkQAFFAwnHMcftRqlQC039HiiOtvOFLWYLK345wt7t51AUcHN3QuugwWg0cfr4NeLjEun0dE0aP1G0TaGFEEL8c4UK5MxmMwaD7Qd9J6ecP/Imk8neKeIODmG1cA0ej+nUEcyX4sFiRuPji0ON+igGZyzqDVQ1A9CgUTzQKK7Wc+uWr8DppBuYLBYsFkgzmnP2p8s2k2WyoHPQEJds4uJ1V1yzPK3nKQqUc9UTey2Vk1duUc0/Zz6ARqvFp17ebR6KQlVV1v11hZWHL3H1VpY1zeCopWagB/2bBuPn8WAtHJJiTCXbYsbgmH/QqigKCgrJxlv3sWVCiLLIpJpwdXQh05yF0ZwNas7ecn8Pp9Ri+e8eciaLGQdN3rfcTFMW3npPvPS2+3s6OGpQAYtFRaPJ/4sxVQW9vvD7hKlXTmLZ/yuYjeDqi/LfIfuqqkJ2Bur5A1gATaPnbQLTQ/susndHPC6uOpxd/g46HRx0ODvruJmYztoVMVQK9qZCUN69SoUQQhS/Qq9aeeLECRwc/s6eu3ng8ePH8+StX79+MTTt4aM46nAMfxTH8EetaSZLAibzn1jUdPjvm7+CAxrFG0dtGBrFhVrl/Pj9fBwnbiSSmalgtuRkNasqZosFrSM4qFqcsvNO3DY4armRauTs9XRrIFcclh+8xM9749FoFPzc9ThqNTn71mWZ2BuXyNWUTMZ0rI6v+4MTzOV8Jrn7XA7VmlcIIfJn0DqRYcrER+9FpjmLDFMmZtWMQk5PmcliJlM1gWomMfMmBkcnXByc0Wl1OV/GmbNQFIUwjyp5RgkEh3jj6qoj9VYW7vl8KWbKNqPRKIRULdxS96qqYjm5GbLTwc0vz3Y16Jxz3oUuHYUbZ6FcFet5e3ecQ1VVmyDudp5eBq4k3OLg3vMSyAkhxH1S6EAuMjLSbnrfvn2t/89d2j43yBMFM1kukm0+jooFBScURfvfSeMmzOpVVHM6Om1d0jO1XDpnIBUFB70JR0cFRdVgNJvBYiErS4sm0wMHR32eOhRFQVFyhmEWl4s301l+6CI6Bw0+rn/XqSgKbk6OOOscOHsjjRWHLjKo+SPFVu+98tS5o9M4YrRk2109Dv6+T156z/vYMiFEWVTZLZDrWYk5q946OOVsX6L+P3v3HR9FtTZw/Dfbd9N7gUAIoXdEpUkTBFQuKAo2JIpYUVER9SqIomIDpbyWi0IAvYLligoqKgpKEQFBQXrokBAgvW2bef9YWVlSSCAhLDzfz2ev7MyZc86c3bubZ0/TyHUWUOQqQlM1dOg8K+MqGoXOIopdJdgMnuHdRp2RFmGNqRcYXyrv4FArrdrXYfXyPVhtRoxG3143TdM4fryI6Jggmras5Ny03MNox/eBJaT84e9GK5TkoR7ciP7vQC77eBGHD+YSGFj25yZ4Pv/NFgNbNx3hmutbVq4+QgghzkqlArnZs2fXdD0uOppWgtO9Cw3QKQHe454vVyNoBlQtH6c7jU/X2diXaSc+NB5VKcFhKETVudC79ORk6SjOM6HXNILjVYynbA5rd7rRKwpxIWc2B64sK3ceo6DERUJ42XnqdQrBFiO/7s5i8CUJhJXzC+65FmwKIsYaxYHCw5h0xjL/kCl2FWPVm0kIiKuFGgoh/ElCQDw7c/eS58wnyBiITtFR7C6hyFUEmgKKZy+5MFMwbk2lyFWMU3VS7LbTJCSJJqFJRJjDyg2q+lzThCOH89i98xhmi4HAQDOKDoqLXBQW2AkNs3HdzW0qnEd3Mq0oB1wOsJbfY6YoCppODwXHvMecThVN09CfZvNxvU6HwyE/5F5IJkyYwMKFC9m4cWOl0iuKwueff86gQYNqtF4XgpSUFHJycli4cOFZ55Wamsro0aPJyck567yEf6lUIDd8+PCarsdFx6UdQcOOQkCZ5xVFAc2Mw53JjiM2QqxGjHoDuAIxuzzz51xulfS8PDS3hlNTyS50Eh38Tw+ZpmkcK7CTEB5Ay7rlD6vMdeSzP/8Qx+3ZaGiEm0KpF1SHUFNwmX9g7MwswKhXyjxnd7rJLnJS7HBR6HDz+YaDDOlQj0BLpTt/a4xbcxNlDedgYTpHi49jNVgJMFjR6XS4VTdF7mJ06GgZ3kQ2BhdCnJbFYKZjTDt+PbKBPGcBehQKXcWoqvr3QiQ6gk1BWP7+PAkw2lBVlTxnPsHGQCIt4RXmHxBoZtjdl7HixzQ2/HaQ3NwS0MBsMdChUz269GxIfN3KD2NUdHrPuHFNK2OFzZOoKuj/WcE4KNiM0WTAbndhruCz3O5wEVNH9mbz7A6fDzgBI+iCanS8fkpKCnPmzCl1vG/fvnz77bdnlfeYMWN48MEHzyqP80VVg1J/MnToUK6++uraroaoBdX217WqqixevJhZs2bx+eefV1e2FyxVzQMof3gLAEac7mKCrDpUNbTUWYNeR2ywhYPZxbhUjfwShzeQc7pVjuXbsRgNDL00AYOu9C+pmqaxJXsn23PTcKhOdIoOBYUjRcfYlb+PpKB6tIloVmrrA7eqlaq3pkFGbjGZeXacqufXW7cKC38/xPo92dzaqT6dkys3j6MmHC46wp/Ht5LnKMCtuVFRyXPkke/Mx6w3Y9IZCTQG0DysEfUD69RaPYUQ/iXcHErP+E7sLzhMWt5e8pyF6HV6bAYrVoMFo853SxedTodBZ+RQUQatIpqeNn9bgImrBjSjW+9kjqTn43K6CQw2ExUTVOEiKGUKSwBLENjzwRpaZhJN9UzCVqL+WbXXFmCiZds4fv15L0HB5jI+/zXy8+wU5NkxmfT8+fshmrSIwWyu/R/wzjl3Njj2gJoHuAE96ILB1AD0YTVWbL9+/UqNnjKbS0+3qCxN03C73QQGBhIYGHj6C0StslqtpRYlFBeHSm0IXpEdO3bw5JNPUrduXQYOHMj3339fHfW6CKhU/JPoPxTK/zEvKshM3TArOkWhwO5mf1YRB7KKOJJbQniAmXt7NKRDg7J/9d2Zt4e/sncACsHGIIKMgQQaAwg2BaFTdOzI3c1fWTtKXdcgKgCHS/XZBDYjt5jDOSWeuSJGPUaDDotRT50wK9lFDv6zPI11e7Iqdb/V7XDREdYc2UCuIx+rwUqYOZRYaxQRlnDMOhO6vxcb6Fu3G4lBdc/LrROEEOcvq8FCk9AkOsVc4u1pCzYFlQriTvAshFK1FZ9zsorZvDGdT+Zt4N03V/LqhB/4ZuFfHEmv/Aq7ijkAJaEdOErQ3M5S5zVNg6IssIag1G3tc67jFYkEh1o4mlmIqv7z2V9U6GDPruPs35NNSYmL7X8d4aNZ65k2aRnrf91/cW0W7s4G+2ZwZ4FiAiXI81931t/Hs2usaLPZTGxsrM8jLOyfwFFRFN577z2uu+46bDYbjRo14ssvv/SeX7ZsGYqisGTJEjp06IDZbOaXX35hwoQJtG3b1qesWbNm0aJFC8xmM3FxcYwaNcrn/LFjxypVTrt27bBarfTq1YvMzEy++eYbmjVrRnBwMDfffDNFRUXe6zRN49VXXyUpKQmr1UqbNm349NNPS+W7dOlSOnTogM1mo3Pnzmzfvh3wDD187rnn+OOPP/5eO0AhNTW1zLZcu3Ytffr0ITIykpCQELp3787vv//uk+Z07el2uxkxYgQNGjTAarXSpEkTpk6dWu7rN3fuXCIiIrDb7T7HBw8ezO233w7AH3/8Qc+ePQkKCiI4OJhLLrmEdevWee8vNDTUe11FacWF5YwCuaKiIlJTU+nWrRvNmjXjtddeIyIigmnTpnHo0KHqruMFybO9gHaaLzkXOsVAQYmRonLmHSiKQlSQmchAM1c0imRAm3gGtIln1JWNeG1IGzo2jCjzOqfqZHvObnSKDqvBUmr1MovejEFnIC1/H8WuEp9rOzeMxGrSk1fi+UPE4VLJzLej14HRoENDw+XWCLF6Fj6JDbFQ4nTz6fqDuFS1Su10ttyayqbj23CoToKMgRh0+hM3icVgJtwShkExcLjoSCXWsxRCiPJ5Pjf1uNSK54m5VTc2g63S+W7acJiZ01fx8w87KS5yogBFBXaWfbeLmdNWsnVTRqXz0jXphRLVEAqOoRXnoKme7Ww0RzHkHwG9CV2bgSgW3yGScXVCGHJ7O0LCrBw9UsDRIwVkHM5jb1oWhQUOrDYDSY0iiY0PJjzCSk52MQsX/Mm61fsrXTe/pml/98TZPT1witHzC6xi9DxX7Z7ztRjYPvfccwwZMoQ///yTq6++mltvvZWsLN8fWMeOHcukSZPYunUrrVu3LpXH22+/zQMPPMDdd9/Npk2b+PLLL0lO9t1ztTLlTJgwgRkzZrBq1SoOHDjAkCFDePPNN/nvf//L4sWL+f7775k+fbo3/TPPPMPs2bN5++23+euvv3jkkUe47bbbWL58uU++Tz/9NJMnT2bdunUYDAbuvPNOwDP08LHHHqNFixakp6eTnp7O0KFDy2yn/Px8hg8fzi+//MKvv/5Ko0aNuPrqq8nP9/3RpKL7VFWVunXr8vHHH7NlyxbGjx/Pv//9bz7++OMyy7zxxhtxu90+weCxY8dYtGgRd9xxBwC33nordevWZe3ataxfv54nn3wSo7HsH4uqklb4tyqNe/j111+ZNWsWCxYsID8/n6CgIG655Rb++9//8n//93+lNg0X5dPrYnGp+wEHUHr4g2czWTsWYwRBFjP7judhM+nL7C3KK3ERZDGQ0rUB9SPKnnN3qvSiTIpdxQQYy09v1VvIdxZwsDCdRiH/bPKaFBVAr2bRfLMpHbeqYXe5cblVLEY9blXD6VIxG/TE/r1ktqIoRASa2Z9VyNbD+bSqwpyOs3Wk+Ci5jnxshrI3KVcUBZvRSoGzkMNFR8pcPU4IISrDarAQZ4thb/4BzHpTmZ85qqaioVE/qHJDuA8fyGXh/D8oKXYSExdUatPxY0cL+ey/G7nrwc7Exp9+fppiDkDXaTjq9qVoBzZC4XFPcGEwocQ0Rte4J0pM4zKvbdQ0mgfGXMGmDYf5a2M6WzdnYDDqiE8IITjE4h3qaTDqiYwKIOtYId8v3kbz1rEEBJ75MD+/oOZ7hlPqbKWH0CiK57ia50mnr/55hIsWLSo1BPKJJ55g3Lhx3ucpKSncfPPNALz00ktMnz6d3377jX79+nnTPP/88/Tp06fccl544QUee+wxHn74Ye+xSy+91CdNZcp54YUX6NKlCwAjRozgqaeeIi0tjaSkJABuuOEGfvrpJ5544gkKCwuZMmUKP/74I506dQIgKSmJFStW8O6779K9e3dvvi+++KL3+ZNPPsk111xDSUkJVquVwMBADAYDsbEVr/Laq1cvn+fvvvsuYWFhLF++nGuvvbZS92k0Gnnuuee8aRs0aMCqVav4+OOPGTJkSKkyrVYrt9xyC7Nnz+bGG28E4MMPP6Ru3br06NEDgP379/P444/TtKlnSHajRuXvB1yVtMK/VSqQmzJlCu+//z7btm1D0zS6du3KiBEjuPHGG3E4HHz44Yc1Xc8Ljk4JQK+ri0vd+/cvdP/MO9A0NxolKJgx6htyXXvYe2wH6bklRAebvfPdNE0jr8RFXrGTvi1jqRde+V94i1wlf6+YWX6n7In6nNojpygKt3Wsj9mg44ctmRzLd+B0a6C4MZrcBFn1xAdZsZ60kprFqMft1jiSV0Irzl0gl+fIR0P7pyeuDHpF700rhBBnIzmkPunFmRQ6CwkwBvgEXqqmku8sIMwUQt1Kroy77td9FOQ7iIkLLBUYKopCZFQAR9Lz+f23A1w9qEWl8lTMAehb/wutyZWQfQBNdaPYQiEk/rRDy4OCLXTunkTdeqHs35tNZHQgVmvZv/SHhFk5llnI5o3pXN41sVJ1819OPHPiyvuzSv/3+dJDWqtDz549efvtt32OhYf7Tqs4uYctICCAoKAgMjMzfdJ06NCh3DIyMzM5fPgwV155ZYV1qUw5J6eJiYnBZrN5g7gTx3777TcAtmzZQklJSakA0+Fw0K5du3LzjYuL89a7Xr16Fdb5ZJmZmYwfP54ff/yRI0eO4Ha7KSoqYv9+397l093nO++8w3vvvce+ffsoLi7G4XCUGqZ6spEjR3LppZdy6NAh6tSpw+zZs0lJSfH+f/LRRx/lrrvuYt68efTu3Zsbb7yRhg3L3uKpKmmFf6tUIDdmzBgUReGaa67hjTfe8HkzOJ0186F0MTDqGgIKbvUAGoUnjbhQ0CkBGHVN0CuhtKoL9/VMJnXlHtJzPKuWKTrPwmIBZj19W8YyrFNileZ26RXPEMgTe/+Vp7xgz6DXcfPl9enTPJbXvtvC9twMAsOc6AwqegVcFFLoDMDiDEKvGf/uYfRsTeAvHLkFFB06hqLXE1A/BoPl/NhGQQhx/oq0hHNpZGvWH9tEnjMfvaJHp+hwqy40IMwUQseY9uXuZXkyp9PNpg3pWG2Gcj+nFUXBYjHy57pD9B3Q7LRbBPhcaw6A2KaVnK3t60h6Pk6Hm7BwKxqgFhajFnt+9NNZzOgCbej1OjRNq9I8Pv9lxBOsuf7+96n+XvikzHNnLyAgoNQQx1OdOrROURTUU6Y7BARUMEqnkotpVKack9MoilLhNSf+u3jxYurU8e3JPnVBl1PzPfn6ykpJSeHo0aO8+eab1K9fH7PZTKdOnXA4HOWWdWqdP/74Yx555BEmT55Mp06dCAoK4rXXXmPNmjXlltuuXTvatGnD3Llz6du3L5s2beKrr77ynp8wYQK33HILixcv5ptvvuHZZ59l/vz5XHfddaXyqkpa4d8qFci1bduWjRs3snjxYo4cOcKdd97JzTffTHCwLDN8NhRFh0mfjKqr49kAXCsGdOiUYPRKFIryTy/SpQ3CaRYXzJo9x9mRkY/TrRIXYuXyhhFV6ok7IdISjlFnwKk6MZXzB4VLdaFXdBUukW0xQ0BMHoHGYow6HTpNDxpoiordlIvTUERQSTTFRTqsRj1JUZUb+lldgo1BKCi4VHe5vXJuzTOfJdgUBEDBgUzSPvyBQ9/8hiOvCEUBa0w49a/rSoObemIKPrf3IITwL3UD4wg1B7O/4DCHCjNwqi6sZgv1A+tQNyAOk75yf8zbi124nO5Sm4Gfymjy7N/msLuw2s7tD07uwmKc6UdwFxZ5fl1EAZ2CzmrFFB9zTutSq3RBnrlw7qy/58idFB5rGqhFoA/3pPNTQUFBJCYmsnTpUnr27HnOym3evDlms5n9+/f7DKOsKpPJhNt9+n0Of/nlF9566y3vcv4HDhzg2LFjp7mqdB6dO3fm/vvv9x5LS0s77XV33XUXb7zxBocOHaJ3794kJCT4nG/cuDGNGzfmkUce4eabb2b27NnlBmdVSSv8V6UCud9//50//viDmTNn8tFHH3Hffffx6KOPMnjwYAYPHlzTdbzg6RQrOqX+adMFWgxc2SyGK5ud/ZdjqCmYKEsEhwuPYNAZSvW6aZpGkauYcHMo0dayF0wB+HbvDvLchRgx4nRomI06zyqbmueXWFXnpMB0jJyjIVzaIPyMgs6zEWOLItgURK4jjyBj6aFJmqZR5Cwm0BhAvC2G3B0HWDP6/8jfnY4pJABLZDCoGsVHstj8xqccWbmZy6eOwhwqyzELIcp3YjuT5mFnPjfFZDGgN+hwOSvuUXA6VcwWA0bTuVvuPzo2CL3mJj8tHaPbjmI0opg8QaTmVlELCynevR8tJIro2Ivg81JRPFsM2Av/mSt3YjilWgQ6s+d8Da2KbLfbycjwXfTGYDAQGVm92/5MmDCBe++9l+joaPr3709+fj4rV66s0b3mgoKCGDNmDI888giqqtK1a1fy8vJYtWoVgYGBld7rODExkT179rBx40bq1q1LUFBQmVs0JCcnM2/ePDp06EBeXh6PP/54lZf2T05OZu7cuSxZsoQGDRowb9481q5dS4MGDSq87tZbb2XMmDHMnDmTuXPneo8XFxfz+OOPc8MNN9CgQQMOHjzI2rVry/wbvCpphf+r9BiMNm3aMGPGDA4fPswHH3xAx44d+fDDD7n++utRFIXPPvus1IeIOH8dzbdTmBXHvkOBbNpXRG5xCaqmomoqJS47eY58Agw2OkS1LnceXVZJEX8dP0KQyURCWAB6nUKJ041b1f5eAVJBdekopoToSLj18vrnfGl/vaKjdURTjDoj+c4C3CetJudW3RT8vedTq/Cm6FT4fdxs8vdkENQwHmtMGAarGUOAhYC6UQQmRHF0zVb+euPTCkoUQojqYTLpad46jqIiR7krHGuaRkmxi1Zt4zEYznpHoUpLqB9CRNEx8p06FLMFRf9Pr6Gi16GzWMh36bEU5tK8RdQ5q1et0oeBuaWn501zgFbg+a8+/O/jNbeP3LfffktcXJzPo2vXrtVezvDhw3nzzTd56623aNGiBddeey07d+6s9nJONXHiRMaPH8+kSZNo1qwZffv25auvvjptYHSywYMH069fP3r27ElUVBQfffRRmelmzZpFdnY27dq1Y9iwYTz00ENER0dXqb733nsv119/PUOHDuXyyy/n+PHjPr1z5QkODmbw4MEEBgYyaNAg73G9Xs/x48e5/fbbady4MUOGDKF///4+C6qcSVrh/xTtLDZ52bt3L++//z5z5szh4MGDGI1GBgwY4LO3h7/Ky8sjJCSE3NzcC2oIaV6xk//+uo/f9mRRYHehaRoO1YHZ7KJJfRctGrgxGwzEB8TQNDSZEFP5w0DWHTnIJzs2EWUNQKco5BU7Sc8tocjh4sQ2Q3pFwWR106deMkObtzxHd1na4cIj/JG1lXxHAf+EmQqBxgBahTchITCejF/+ZPUDU70BXFlKjuaCotDrswnY4srvqRRCiOqwf282s/9vNS6nm7AI39V3NU0j61gRJrOBEQ92ok5C6DmrV9FfO/jtxdn84EqkWDMQpHdj1Hk+W90aFLj1oGl0VA7Q5/HBBF3a5pzVrSLn5Ltd0zyrU+IEjJ7hlLI/qaikPn360KxZM6ZNm1bbVRF+4KzGYSQmJjJx4kSef/55vv32W95//30WL15cXXUT1azA7uKN77az6VAuIVYjdf7eSNzltnK8wM7WXS5iTWEM79yAQNPp54E5/h5rrvv7CyrYaiTIaqTQ7sL+91Agm1lPnqMYm7niOR41LT4ghhhbJOlFR8lz5AGe+XOxtmjv3LnsP7ZQp2c0oU2j0dwa+fuKyd1ZhHbSqCZzRBD5aekcW7udev/qXBu3IoS4iNRLDOOawS1Z/NlmjqQXYAswYjB6hlsWFTmxBZj4142tzmkQB1Cy9wCxah59o3NZkR1MltOI6lbg7x/KggxuOoQU0DD7OPa9B8sN5PJyiknbcQyHw01gkJnkJpGYLX6+35Wi1MgWA+LClpWVxXfffcePP/7IjBkzars6wk9Uy4B6RVHo378//fv35/jx49WRpagB3/+VwV+H84gLsWI6aQiOQa8jJsRKbrGTX3cW0Kuxm2aV2E4t0Gj6eyER1bslggIEmg2c2DJI0zRyHRBoqv0VH/WKnroBsRBwyh4ymgqOXST0VlG7NURnNKAAqlujKN3OnoVHKTjgWY1N0elAUXAV28/9DQghLkodOtYjKjqQdav3sXXzEZwOFYNRx+VdE+nQqR71EmtuyF55NLcKCtS1OLgx9hgHSswcdRjRNAgxumlgLcGk07DngqaVnuNXXOTgu6+28efvhygscICioABhETY6dWtAlx4N0FVhBU4h/F379u3Jzs7mlVdeoUmTJrVdHeEnqn1mdESEDDc7H9ldbpZtP4rJoPMJ4k4WbDFwoMjJLzuP0qwSG8s2Dosk1Gwh32EnzFL2ROAilxOrwUDz8KqNLz9nNA0c28G5B51JT/7uIgw2z0bmOpNCYF0LjW+LY1vqIYrSHbgdTlAULBHya6sQ4typnxRO/aRw7CVO7CUuzBZDrfZcmWKjQFFQHU70JiOJVjuJVt8fuDSXCzQwxfjOkSspdvLh++vYuTUTW4CJqJhAdDoFl0slL7eErz//i7zcYq6+rsU5n1ctRG3Zu3dvbVdB+KFK/dyl1+sr/TAYzt2qWaLyjuTZySqwE1zBF7+iKFhNeram51UqT4vBSKe4ejhVlQJH6cn4JS4X+Q47LSNjiQk4T5dcVvPBuR8wYgwMRdHrUJ0uzymHRtERB+ZwA3V6erZgKMnMIaBuJNGdK7fxrhBCVCezxUhwqLXWhx8GtG2BKTYa1/GsctM4s3IwRIQR2KG1z/HfVu5j1/ajhEcGEBxiQff3/qIGg47wCBu2ACO//rKXPbtkhI8QQlSkUlGXpmmEhITwr3/964Ja+ONiciLIOt2Pm4oCqlr59W+uqJtIgdPOr+kHOFJkx6w3oChgd7vRKwqto+IYkNQUVVPJKD5KVkkOGhqBhgDiA2IqtSlujXIdBs0FukAMAQqWyFCKM7JQdIp3FTZHnpvQxgGglKA6XDS8rY+3104IIS5GOpORyMFXkzHzvzgyjmKMDEcxeD4zNbcb1/Ec0DQiruuLPuCfbWdcTjfrf92PwaDDZCp77nRgkJmM9Hw2/HaApEbVu3y+EEJcSCoVyKWkpPDpp5/y2Wefcf311zNixIiz2pRRnHuRgWYCzQYK7C4sFWwuW+xw07YKk+Z1io6rGzSleUQMGzIPsycvG1XTaBIQTPvoeBqHRZJlz+b3jM3k2vO9K0YCWLMtNA1tSKPgBrU3fEbNB0XnnZ8R3KgOmlul5HguaE4UvQ63HWyxFqwRBmJ7Xk3DW6+snboKIUQt0TQNV3Yumt2OPjAAfVAgwV06oLndHF3wJY6MzL/nwikoChjCQokY1JfQKz1L4GcdLyL9YC5Zx4vIzCggKKT0ysAaoGpONM2N0ayRtiPz3N6kEEL4mUpvP1BYWMj8+fOZPXs2q1atIikpiTvvvJOUlBTi4yuxMoafuRC3H/hg9V6+2HCYOmFW9LrSgVORw0VOkZPH+jahQ2J4tZR5vCSbFRnrKHGXYDNYMeg8vx2omkqxqwQNjVbhTWka2rBayquy4rXgPgq6fzas1VQNe3YexRnZuAqKwaAQWDcIh70pwU3aypwNIcRFQ9M0Ctb+Qc5PqyjesRvcKorRQGCH1oRe2RVrciLugkLyf9uI/UA6aBqmOjEEXdYOQ0gQx48WsvSb7WzdfITiQgdOp5vcnBICAoxExwVjs3mGiLo0O053AW7NCWgU5GpYbHrue6YxEZbmGHVV25C5PBfid7sQ4uJ1RvvIbd++nffff5958+Zx7NgxrrrqKh599FGuvPLC6am4ED/sM/NKmPT1Vg5mFRMdbPb2zGmaRn6JJ4jr2DCCh/s08q5CeTY0TePn9DWkF2cSbAwqMwAqchWjV3T0qdONAGP1fFFXiWMn2Hd4ArnyAjT17wn8tq5QTX9MCCHE+U7TNI59soisr35Ac7rQhwSh6PVoDgeu/EIMwYHEjrwFrXETiotdWK0GgkP/+Yw8frSQOe+uIeNQHoFBZgICTaiqStqO4zgdbsxmA/EJIZhtKnZ3LhoqiqJHQSHriJv6zfVcdZsOiz6MOoGdMepsFdS2ci7E73YhxMXrrDYEd7vdvPjiizz//PP861//4n//+1911q1WXagf9oeyi3hnWRq7jxbidHuWhNY0CDDruaxBOCldG2AzVc+CNdn2XH48tBKDzohJX/bEfE3TyHPm0y6iBU1qo1dOLYSilYBWdpCmqaAVgKEeWFqXPi+EEBeovF9/J/3/5qAzmzGE+n4PaprGnkNFbNNFczS6IaqiQ2/QkdQoksu7JtK4eTQfzPyNPzccJjomEP1JWwlkZuRz7GghABargZj6KhoqOuXvecl2jcI8lX63BpHUwohDzSfE1IC4gA5nf08X6He7KN/evXtp0KABGzZsoG3btjVWTkpKCjk5OSxcuLDGyhDiVGf0F3tubi7//e9/mTVrFuvXr6dJkybccMMN1V03UQPqhNmYMLAlWw7nsvlQLnanSrDVSIcG4dQLP/tfO0+W68jHqbqwGsrvxfL00inkOPKrtexK0wWAqSE4doBaBIrFM2dO0wCnpzdOF+RJI4QQFwlN08hZugLN5cJwyvYBAJsLAlilxmEvcRNcWIAtJhKn081ff6Szc9tROnVLZOe2owQHW3yCOIDQcBsF+XZKSlxoigO3Cga9EU3TsBdrFORpJDU3Ub+pCUVR0CtmClyHcapF1dIrd6FJSUlhzpw5pY7v3LmT5OTks84/NTWV0aNHk5OTc9Z5XQjKCwynTp1aavVuIWpalQK5pUuXMmvWLBYuXIher+fGG29k6tSpdO7cuabqJ2qAXqfQqm4oreqG1mg5ivd/Kpm2thiTQDGAYzdoRaACigYYwBAN5maegE8IIS4SrqNZlKTtw1BGr1W63cjqHM/xcIMDXVEBZkssZouBwCAzOdnFLP1mB6qqUbdeSKnrTSY9dRJCST+Uh0srJusI6HVuzzmLQpN2ZroPDMBg8Hwz6BUzDjWfElcWRtP5H8hpmoZGAZrmRFGMKATW+Nzqfv36MXv2bJ9jUVGlA/Da5nQ6MRprd+uMmhISUvq9LkRNq9REqOeff56kpCSuuuoq9u/fz4wZM0hPT+f999+XIE6UK8QUjFFnxKE6y01z4terUHMtDnFRFDDWB9sVYGnnCdzMLcDaESwdfBZCEUKIi4FaUoLmdqMYS//eu7XAhl3VEah3g6Kgud0+50PDrDgdbkqKneUGMGaLgfpJYbTorNCio0KrThYuv8rG9feEcNVNgVhspf88OXnV4/OVW8vB4f4Tu/sPHOomz3/df+LWcmq0XLPZTGxsrM9D//cWOl999RWXXHIJFouFpKQknnvuOVwul/faKVOm0KpVKwICAkhISOD++++noKAAgGXLlnHHHXeQm5uLoigoisKECRMAz4iaU4cRhoaGkpqaCnh6rhRF4eOPP6ZHjx5YLBY++OADAGbPnk2zZs2wWCw0bdqUt956q8L7+/TTT2nVqhVWq5WIiAh69+5NYWGh93xV89uyZQtXX301gYGBxMTEMGzYMI4dO+Y9r6oqr7zyCsnJyZjNZurVq8eLL74IQIMGDQBo164diqLQo0cPwNMzOmjQIG8edrudhx56iOjoaCwWC127dmXt2rXe88uWLUNRFJYuXUqHDh2w2Wx07tyZ7du3V1h3IU5WqR65CRMmEBwczF133UWTJk3Izs7m3XffLTOtoig88sgj1VpJ4Z9CTEFEWyI4VJiBSWcsd7ETi95EQsB5sPKpYgBDXG3XQgghap0+KBCd0Yhqd6Cz/LNVgEuDvcUWzDo3iuJZ5VexlP5TwhpgojCzgJISF5YyzoNnBLvZbKB9b5WQwIBygz4NFzpFf94Pq3RrOTjdW9FwoGABrIALlRyc7iLQN0OvhJ7TOi1ZsoTbbruNadOmccUVV5CWlsbdd98NwLPPPguATqdj2rRpJCYmsmfPHu6//37Gjh3LW2+9RefOnXnzzTcZP368N8AIDKzaj5tPPPEEkydPZvbs2ZjNZmbOnMmzzz7LjBkzaNeuHRs2bGDkyJEEBAQwfPjwUtenp6dz88038+qrr3LdddeRn5/PL7/84v0h+Ezy6969OyNHjmTKlCkUFxfzxBNPMGTIEH788UcAnnrqKWbOnMkbb7xB165dSU9PZ9u2bQD89ttvXHbZZfzwww+0aNECk6ns/XDHjh3LZ599xpw5c6hfvz6vvvoqffv2ZdeuXYSH/7My+NNPP83kyZOJiori3nvv5c4772TlypVVamNx8ar00Mq8vDxmzpx52nQSyIkTFEWhZXgTchx55DnzPdsPKAYURcGtuSl2laCg0DKsCVaDbLAthBDnC0NYCAGtm5G7ci364H+GBro0BbemoEfzRGKaijE8tNT1gYEmcrN15GYVYYkve8RFQb4dDgdgsThxaw4MShl7y2kaTrUYmyESi756tsWpCZqm4XLv+zuIO3kopRE0AxoFuNz70OlDamSY5aJFi3wCrP79+/PJJ5/w4osv8uSTT3oDmqSkJCZOnMjYsWO9gdzo0aO91zVo0ICJEydy33338dZbb2EymQgJ8dQ5Njb2jOo2evRorr/+eu/ziRMnMnnyZO+xBg0asGXLFt59991yAy+Xy8X1119P/fr1AWjVqtUZ5/f222/Tvn17XnrpJe+xWbNmkZCQwI4dO4iLi2Pq1KnMmDHDe33Dhg3p2tWzJ+KJIasRERHltklhYSFvv/02qamp9O/fH/AEnN9//z3vv/8+jz/+uDftiy++6N2b+cknn+Saa66hpKQEi0X+LhKnV6lAbs+ePTVdD3GBCjOH0CWmAxuO/0W2PYdircR7LtBgo1lYIxoEJdRiDYUQQpQltPcVFGzcgjPzGMboSBRFwahomHQqJS4F1Wn/e0XL0nODnE434ZEB6HSQdayI0HArur/3L9U0zbPYSbGLy1skE261k2NPw6WCXjF5Ax1NU3GqRegVI+GWJuf1Hp4aBajko2ApVU9FUUCzoJKPRgEKQdVefs+ePXn77be9zwMCPPO6169fz9q1a73DAsGz4nhJSQlFRUXYbDZ++uknXnrpJbZs2UJeXh4ul4uSkhIKCwu9+ZyNDh3+WW306NGjHDhwgBEjRjBy5EjvcZfLVe4cszZt2nDllVfSqlUr+vbty1VXXcUNN9xAWFjYGeW3fv16fvrppzJ7FtPS0sjJycFut5/VllppaWk4nU66dOniPWY0GrnsssvYunWrT9rWrf9ZETsuzjMqKDMzk3r16p1x+eLiUalA7sQvIEKciXBLKL3iO3OsJIssew6qphFgtBFvi/ZuEC6EEOL8YmveiJg7biRzzqc4DhxGZ7WgGAw0dCv87owi0GLG3CCh1Dw6TdMoKXbS46pGREYHsuSrrRw9UuDZqlNRUN0qVpuJK65sSJ9rmqLoNEAj17Efh5qHgg4NDQUw6KxEWVsTaDy/h71rmhNw4xlOWRY9UOJJVwPxaEBAQJkrVKqqynPPPefTI3aCxWJh3759XH311dx7771MnDiR8PBwVqxYwYgRI3A6y5/fDp4A9dRVGsu65uRgUFU92x7NnDmTyy+/3CfdiTl9p9Lr9Xz//fesWrWK7777junTp/P000+zZs0abDZblfNTVZUBAwbwyiuvlDoXFxfH7t27y7yuKk60y6lBvaZppY6dvPjLiXMn2kmI05G/osU5oSgKUdYIoqwRtV0VIYQQlRRyxeWUhESStexXHJv/wqi5aZlgYm9RCAWYsJ4y/EtTNY4eLSQkzEr7y+sRExdEk+bR/LnhMAf2ZuN2a8TEBdG6fTwxcf8MuYy2tiPEnES+4yAONR8FHVZDJEHGuhh0pYdcnm8UxYgnWHMBZa3K6Ab0f6c7d9q3b8/27dvL3YZg3bp1uFwuJk+ejE7nWWDm448/9kljMplwn7KgDXiGGKanp3uf79y5k6KiogrrExMTQ506ddi9eze33nprpe9DURS6dOlCly5dGD9+PPXr1+fzzz/n0UcfrXJ+7du357PPPiMxMRGDofSfwY0aNcJqtbJ06VLuuuuuUudPzIkrq01OSE5OxmQysWLFCm655RbAE+SuW7fOZyirEGerUoFcr169yj1nNBqJiYmhT58+3HzzzWX+n0IIIYQQ/uXg/hxWLN3F9i2Z2EuCUKydCAoy0e6yelxXL5RFn20iM6MAo0mH0aDH6XLjdLgJDbcx+Na2xMR5hhAGh1rp2rPivTgVRcGiD8ViDT0Hd1b9FALREYRKDmgGn14Xz3YEJegIReHcroI8fvx4rr32WhISErjxxhvR6XT8+eefbNq0iRdeeIGGDRvicrmYPn06AwYMYOXKlbzzzjs+eSQmJlJQUMDSpUtp06YNNpsNm81Gr169mDFjBh07dkRVVZ544olKbS0wYcIEHnroIYKDg+nfvz92u51169aRnZ3No48+Wir9mjVrWLp0KVdddRXR0dGsWbOGo0eP0qxZszPK74EHHmDmzJncfPPNPP7440RGRrJr1y7mz5/PzJkzsVgsPPHEE4wdOxaTyUSXLl04evQof/31FyNGjCA6Ohqr1cq3335L3bp1sVgspYZxBgQEcN999/H4448THh5OvXr1ePXVVykqKmLEiBFVeQmFqFClth/IzMzk6NGjZT7279/Pl19+yfDhw+natetpf40RQgghLkaapqJqhahaAZrmOv0FtWj3zmPMfXcNv689iE6nEBZhIyTUQnGRix+/3c6Kn9IYdvdl9L+uOVExgRhMeiKjA+k3sDn3jO5C4yYhoNpBuziGiCmKgkFfHwWTdw85TVM9/6UABZPn/Dme59e3b18WLVrE999/z6WXXkrHjh2ZMmWKd8pM27ZtmTJlCq+88gotW7bkww8/ZNKkST55dO7cmXvvvZehQ4cSFRXFq6++CsDkyZNJSEigW7du3HLLLYwZM8Y71LEid911F++99x6pqam0atWK7t27k5qa6l3W/1TBwcH8/PPPXH311TRu3JhnnnmGyZMnexcRqWp+8fHxrFy5ErfbTd++fWnZsiUPP/wwISEh3l7JcePG8dhjjzF+/HiaNWvG0KFDyczMBMBgMDBt2jTeffdd4uPjGThwYJnlvPzyywwePJhhw4bRvn17du3axZIlSwgLCzttGwlRWYpWTdvQf/PNNwwdOpQHH3zQZ1Ktv8rLyyMkJITc3FyCy9gQVQghhKgMTXPi0g7iVg+jnVjwSTFhUOIx6OqilLFa49lw5eSRt3o9+Ws24M4rQB8SRHDHdgR1bI8h5PTfZ/YSJ9Nf/ZljmYVEx5TeFsDpdHP8aCGduicxaOg/CzWgucB1CJwHQf17jy+dDYx1wVAHzvGwwrLU9He7W8vB5d6HSj4nhlPqCMKgr3/Otx4QQlz4qi2QA88SqvPmzfPuteHPJJATQghxtjTNgT17GRxN8/xdbwlAi4lDM+kANzolGJO+NTqlevZIK9q6k/S35+E8cgzFYEAxGtCcTjSXG2NcNPH33461cVKFeWxce5AFc34nPNKGwVD2wJ283BJ0eh0PP9WD4BALaA4o2Qiuo6Do/gnaNKenV04fAdZ2UM1Ba1Wdi+92z1BKT6+cohhP2Y5ACCGqT6WGVlbWZZddxr59+6ozSyGEEMIvaUXZOFfNQPfzV+j+3IR+y2b0v69B/9MS9Dt3o6gWVC3Ps4l0Nfym6kg/wuH/m4Mz8zim+FhMcdEYI8MxxcVgio/BmZHpPV+R3buOoapquUEcQGCQmcJ8O/t2Z3kO2Ld5gjidDXQBoJg8D12A55j7ONj/Out79AeKoqBTgtDrwtEpQRLECSFqTLUGcsXFxbLYiRBCiIueVpyHe9X7KId3gN4AQSFoQSFoAUHgdKLbshHd1j9RNBNuLQeV3DMuy5WbT/Z3P7N/0v9RvHMP6PVoLt9l4BW9HlN8LI6MTHJX/FZhfg6727vnW3lOnHc53Z5hlK4M0JlAKWPJd0UPOrMn0HPnV+3mhBBClKtao66vv/6a5s2bV2eWQgghhN9Rd/0C2fvRAgM9Ac4JOh3YAtDsJej27ESNS4BwK6qahV4fWuVycpf/ytH5X+LMysZ55BioGs7DGZ5NvKMiMMVFe3uEFJ0OndlM7s9riBh4FUo5+2yFhdtQ1bL3vDrBYXehN+gIDrWA+yjghApXZDSCZvek1Vf/hthCCHExqlQgl5WVVe45t9vNoUOHWLBgAe+99x5vvfVWtVVOCCGE8Deaoxht/+9oRrMncCuLyQz5uegO7sUd3gzP/mNVk7f6d47M/hjN7cYYFYErKxdFp4BOh+Zy4czIBEXBHBftvUZnMaMWFKKW2NEHlD0vr0XbOFYuS6O4yIktwFRmmtycEuLqhJDYMALcf/cmVjSEUFE8G2Gf56t1CiGEP6lUIBcZGVmpMd6jRo3i7rvvPutKCSGEEH5FU0HN9iy5X3gcXMUoJisaJYCGJ4o5iaKg6Q0oWUeBZp75ZFUpzuXi+MJvUR0OzPGxaG43KJ4NuXV6BcVoRNWcuI4ewxgRhs5k/Ps6N4rRgGIqfwXJOgkhtGgTz++/HUBRFKy2f9JqmkZudgl6vY4rrmyIXq8D9cTCJlr5wdyJOYDnwcqVQghxoahUIDd+/PhyAzmDwUBMTAy9evUiKanilbCEEEKIC4qmgeswOPeAmg+ooHOia9MALdcOmYdB9SxDX4oCoKJgRK9EVanYwr924DiYjjEi3JOVXo8hKBBXVi4YPV/titGAVuLAlZOLKToSALWwiJBendFVsHGzoigMHNIKVVX56490cnNLMBl1qBq4HG5CHAVckmQj9MAejhtLCG8Vj4LJs3JluatSOgAjGKp2n0IIIcpXqUBuwoQJNVwNIYQQwg8594JjO6CCYgH0oHOCIRtdtBFsCbj37gNVxWd9MU1DcTlxhwSj18WgUwKqVuyRY2iqis78T0+eISIMV24eqtOFzmhAURRUNDSHw3PN8Wx0VgshV1x+2vwtViNDb2/PnrTjbFx3iCPp+eiPHyds10b0hw5TuMrBesBgMRHaqgHNR7Ymqh2g6UE55U8LzQWqw7OfnK6ieXRCCCGqQpaYFEIIIc6EOw8cOwHFJ0BR9CbQB6AVZ6ELCkCLjEHNzMAzD+7v0S1OO5peh1KvLUZd4yoXreh0oGk+C5LogwMxxUbjSM9ELbGjGA2gaagOJ/aD6SgmI1FDrsXWvFGlytDpdTRsHEXDxlHkbN3Hrw99QeGBo1ijQzGGeHr4XIUlHF+7nTVph7j0+R7EXK54RpJiRHWpHF27j+y/jqGqNgISA4i7sgBTiARzQghRHap1+wEhhBDiouE67NnwWrGUPhcQjqIzo7mc6MKC0SkhKFhAU6DEjlLiQpfQAWN0b5RTe7AqwZJcH53FglpY5D2moGCMjcLSIAF9UCCqyw2qhqLXEdCmGfEP3kHYNVdWuSxVVflz0n8p2JdJcHI8plDPBteKomAMtBKUHI8jp4g/31iPS2sG+giO/5HBsju+YPUjy9jy7ha2vbeR9c/M4YdB40n74Ptq2TdPiMpYtmwZiqKQk5NT21W5qCiKwsKFC2u7GlWSmppKaGhoteS1d+9eFEVh48aN1ZJfeaRHTgghhDgT7mOePdLKmEOuGExo4XVR8tLB6AJ3IUqhHQUNTMEojTuga9kfRXdmX8Pm+nWxNkumcP0mdDarp4cOTzBnCA1BFxKEsv8whvBQEv79IOZ68VXemNqelcf+r1aza84Sjv62DZ1Bj+ZyYY0NxxwR/M+2BopCQN1ICvYe4cjq41hjI1nz9HyKjxQREB+PIcAzb051uSnOyGbTawtwO1w0vrP/Gd27KC0jI4NJkyaxePFiDh48SEhICI0aNeK2227j9ttvx2Yre4VSUbYePXrQtm1b3nzzzdquyhnbu3cvDRo0YMOGDbRt27a2q3PRSUhIID09ncjIyBotRwI5IYQQ4kxoKqVWozyJYjCjhdYFVz665DgoUcESjBLfEiUg7KyKVhSF6Fuu42B6JvaD6RhCQ9AHeebZqUXFOLNyMIQGE3f/7Vjq16ly/gX7j/Dbo2+T89denEUloGqgUyg5mov9eB7WuAiCk+t4tjsA9GYjqlsla9Nu8j47SHFGls95AJ1BT0DdSIrSj7Pj/a+p2/8ybHERZ9UO5yNN03Coebg1B3rFhEkXXOUguip2795Nly5dCA0N5aWXXqJVq1a4XC527NjBrFmziI+P51//+leNlX8mHA4HJlPVVmr1RxfLfZ7sYrznsuj1emJjY2u8HBlaKYQQQpwJfeBp90VTcKMYbOga9kTX6hp0ja446yDuBHNCHHUfv5fgTpeguZw4DqbjOJiOu6iIgJZNqPPISALbNK9yvm6Hk3VPzSR78x4CE2OwRASj6HXoTUYMARYUg56iw8cpOnzM914VKE7P4vj6HVhjwnyCuJNZY8KwH8/j0Ldrz+i+z2fFrmMcLlrDocJVHC78zfPfojUUu46d/uIzdP/992MwGFi3bh1DhgyhWbNmtGrVisGDB7N48WIGDBjgTZubm8vdd99NdHQ0wcHB9OrViz/++MN7fsKECbRt25Z58+aRmJhISEgIN910E/n5+d40mqbx6quvkpSUhNVqpU2bNnz66acV1jExMZEXXniBlJQUQkJCGDlyJACrVq2iW7duWK1WEhISeOihhygsLPRel56ezjXXXIPVaqVBgwb897//JTEx0dtTVtbwtZycHBRFYdmyZWXW5fjx49x8883UrVsXm81Gq1at+Oijj7znU1JSWL58OVOnTvUOId67dy8Ay5cv57LLLsNsNhMXF8eTTz6Jy/XPZ0CPHj0YNWoUjz76KJGRkfTp04c777yTa6+91qcOLpeL2NhYZs2aVWYdTwzxW7JkCc2aNSMwMJB+/fqRnp7uk2727Nk0a9YMi8VC06ZNffZybtCgAQDt2rVDURR69OjBpk2b0Ol0HDvmeT9mZ2ej0+m48cYbvddNmjSJTp06eZ+fyT2X5fnnnycmJsb7Wr311ls0atQIi8VCTEwMN9xwQ5nXwelfsxP1eOihhxg7dizh4eHExsaWWqxxypQptGrVioCAABISErj//vspKCgos8y9e/ei0+lYt26dz/Hp06dTv359NE0jOzubW2+9laioKKxWK40aNWL27Nne609+b1aU9mxIj5wQQghxJgx1wJXhCebKmuemaaDZwVgPdGXMo6sG5jqx1HnkLuyHMrDvO4jmVjHFRWNpWP+Me4EyV24m+889BCREoTMa0FvNnn3vVBVFp0NnNKC63BQdOo4tLgJFr0N1uwFQ9AquwhKsceHl5q/odCg6hdwdB86ofuerYtcxjhRtwKXZMepsGBUDKi7PcXc+MbZ2WA3VO8zq+PHjfPfdd7z00ksEBJS98umJ94GmaVxzzTWEh4fz9ddfExISwrvvvsuVV17Jjh07CA/3vGZpaWksXLiQRYsWkZ2dzZAhQ3j55Zd58cUXAXjmmWf43//+x9tvv02jRo34+eefue2224iKiqJ79+7l1vW1115j3LhxPPPMMwBs2rSJvn37MnHiRN5//32OHj3KqFGjGDVqlPcP3Ntvv51jx46xbNkyjEYjjz76KJmZmWfVZiUlJVxyySU88cQTBAcHs3jxYoYNG0ZSUhKXX345U6dOZceOHbRs2ZLnn38egKioKA4dOsTVV19NSkoKc+fOZdu2bYwcORKLxeITMMyZM4f77ruPlStXomkaWVlZdOvWjfT0dOLi4gD4+uuvKSgoYMiQIeXWs6ioiNdff5158+ah0+m47bbbGDNmDB9++CEAM2fO5Nlnn2XGjBm0a9eODRs2MHLkSAICAhg+fDi//fYbl112GT/88AMtWrTAZDIRFhZGREQEy5cvZ/Dgwfz8889ERETw888/e8tdtmyZ93U803s+maZpjB49moULF7JixQoaNWrEunXreOihh5g3bx6dO3cmKyuLX3755Yxfs5Pr8eijj7JmzRpWr15NSkoKXbp08QaXOp2OadOmkZiYyJ49e7j//vsZO3asTwB8QmJiIr1792b27Nl06NDBe3z27NmkpKSgKArjxo1jy5YtfPPNN0RGRrJr1y6Ki4vLvIeqpK0KCeSEEEKIM6GPBH0UuDL/DtQM/8yX01TQCkFnBWNijVfFXCcWc53qGcaTsfxPVJcLg9Uzt80cFoTBZsFVVILB5jmmNxtxldhx5BZiDg+i5EgO5sgQwtsmc+DL1ZUrqAaHG55rmqaRZd+FS7Nj1oX8s5IoJnQ6I3Y1l2z7Liz6iGodZrlr1y40TaNJkyY+xyMjIykpKQHggQce4JVXXuGnn35i06ZNZGZmYjZ7XsfXX3+dhQsX8umnn3L33XcDnsVtUlNTCQoKAmDYsGEsXbqUF198kcLCQqZMmcKPP/7o7bVJSkpixYoVvPvuuxUGcr169WLMmDHe57fffju33HILo0ePBqBRo0ZMmzaN7t278/bbb7N3715++OEH1q5d6/1D+r333qNRo8qtulqeOnXq+NTjwQcf5Ntvv+WTTz7h8ssvJyQkBJPJhM1m8xka99Zbb5GQkMCMGTNQFIWmTZty+PBhnnjiCcaPH4/u73mqycnJvPrqqz5lNmnShHnz5jF27FjAEwzceOONBAaWv4Kr0+nknXfeoWHDhgCMGjXKG1gCTJw4kcmTJ3P99dcDnh64LVu28O677zJ8+HCiojx7NkZERPjcR7du3Vi2bBmDBw9m2bJlDB8+nDlz5rBlyxYaN27MqlWreOSRR876nsHT83j77bezbt06Vq5cSd26dQHYv38/AQEBXHvttQQFBVG/fn3atWt3xq/ZCa1bt+bZZ58FPO+nGTNmsHTpUm8gd+K9dqK9Jk6cyH333VdmIAdw1113ce+99zJlyhTMZjN//PEHGzdu5H//+5/3Ptq1a+d9fyYmJpZ7D1VJWxUytFIIIYQ4E4oeLG3AEONZvVIrALXAszG4VgRKIFjagT64tmtaJfasPHSGfzYwV3QKgYkxKDod7mKHp6dRUUD1bG1QlJ6F2+Gk8R39ibykMYZAK87cwnLz19wqmgqhTeudi9s5JxxqHnZ3NkadrVSgpigKRp2NEnc2DjWvRso/tczffvuNjRs30qJFC+x2OwDr16+noKCAiIgIAgMDvY89e/aQlpbmvTYxMdEbxAHExcV5e8G2bNlCSUkJffr08clj7ty5PnmU5eRejRP1SU1N9cmnb9++qKrKnj172L59OwaDgfbt23uvSU5OJizs7IYmu91uXnzxRVq3bu1ti++++479+/dXeN3WrVvp1KmTT1t36dKFgoICDh48WO59gicgONHLmJmZyeLFi7nzzjsrLM9ms3mDOPB9HY4ePcqBAwcYMWKET/u98MILp30devTo4R12unz5cnr27Em3bt1Yvnw5a9eupbi4mC5dupz1PQM88sgjrF69ml9++cUbxAH06dOH+vXrk5SUxLBhw/jwww8pKioqMw+o/GvWunVrn+cntxnATz/9RJ8+fahTpw5BQUHcfvvtHD9+3Gc478kGDRqEwWDg888/B2DWrFn07NnTG4Tdd999zJ8/n7Zt2zJ27FhWrVpV7j1UJW1VSI+cEEIIcaYUE1guATUbnBmgFXuGWeojPQHeGWwtUNtMYUGoLtXnmDU6FE1VKdidgavIjqZpaC43RUdyCKgbSdN7riU5pS8AUZc15fAPv2MMtnlX0zxZcUYWlshg6vS79Jzcz7ng1hyomoqxnNdbhwGnVoxbc1RrucnJySiKwrZt23yOJyUlAWC1Wr3HVFUlLi6uzLljJy+5bjQafc4pioKqqt48ABYvXkydOr6L6Jzo5SvPqUM/VVXlnnvu4aGHHiqVtl69emzfvr3MfE4euneiR+jkY06ns8J6TJ48mTfeeIM333zTO19q9OjROBwVvzYn79l4al1OPl7WENfbb7+dJ598ktWrV7N69WoSExO54oorKiyvrNfhRHknXoeZM2f69EiBZ5GNivTo0YOHH36YXbt2sXnzZq644grS0tJYvnw5OTk5XHLJJd5A/mzuGTwB20cffcSSJUu49dZbvceDgoL4/fffWbZsGd999x3jx49nwoQJrF27tszl/yv7mlX03t23bx9XX3019957LxMnTiQ8PJwVK1YwYsSIct8zJpOJYcOGMXv2bK6//nr++9//+qxk2r9/f/bt28fixYv54YcfuPLKK3nggQd4/fXXS+VVlbRV4X/fMEIIIcT5RFFAH+55XABiu7dh3/9+wVXswGD9Z/U5W2w45ogQ7EdzKDx8DGOAlVZP3ES9AZ2wxvxz781GDSJn237y0g5ji4vwLJCiKLgdLoozsgBocd+/sEZXz6Iv5wO9YkKn6FBxoaf0in0qLnSKDr1Svav5RURE0KdPH2bMmMGDDz5Y7h/UAO3btycjIwODwXDGw7qaN2+O2Wxm//79FQ6jrIz27dvz119/kZycXOb5pk2b4nK52LBhA5dccgngGUp68n5wJ4YPpqene4fmnW7frl9++YWBAwdy2223AZ6gaOfOnTRr1sybxmQy4f573ucJzZs357PPPvMJblatWkVQUFCpoPZUERERDBo0iNmzZ7N69WruuOOOCtOfTkxMDHXq1GH37t0+AdLJTqwceep9tGzZkoiICF544QXatGlDcHAw3bt3Z9KkSWRnZ/u8rmdzzwD/+te/GDBgALfccgt6vZ6bbrrJe85gMNC7d2969+7Ns88+S2hoKD/++KN3qOjJKvOanc66detwuVxMnjzZ+wPAxx9/fNrr7rrrLlq2bMlbb72F0+ksVb+oqChSUlJISUnhiiuu4PHHHy83OKtK2sqSoZVCCCGE8Irp2pLQ5vUpPJDp2VT8JHqjHoPNjDksiDb/vpUmd13jE8QBhDarT8epo4i8pAn27ALy09LJ23WYwgOZWGPCaPvMbSTdUvWNyc9nJl0wZn0YTrWozMUenGoRFn0YJl31D7N96623cLlcdOjQgQULFrB161a2b9/OBx98wLZt27w9NL1796ZTp04MGjSIJUuWsHfvXlatWsUzzzxTamW+8gQFBTFmzBgeeeQR5syZQ1paGhs2bOD//u//mDNnTpXq/cQTT7B69WoeeOABNm7cyM6dO/nyyy958MEHAU8g17t3b+6++25+++03NmzYwN13343VavUGFVarlY4dO/Lyyy+zZcsWfv75Z+9iKuVJTk7m+++/Z9WqVWzdupV77rmHjIwMnzSJiYmsWbOGvXv3cuzYMVRV5f777+fAgQM8+OCDbNu2jS+++IJnn32WRx991BsYVOSuu+5izpw5bN26leHDh1eprcoyYcIEJk2a5F2cZdOmTcyePZspU6YAEB0djdVq5dtvv+XIkSPk5uYCnl6qbt268cEHH9CjRw/AMyTR4XCwdOlS7zHgrO8Z4LrrrmPevHnccccd3tVNFy1axLRp09i4cSP79u1j7ty5qKpaaq7nCZV5zU6nYcOGuFwupk+fzu7du5k3bx7vvPPOaa9r1qwZHTt25IknnuDmm2/26eUeP348X3zxBbt27eKvv/5i0aJF5QaXVUlbFbUeyE2aNIlLL72UoKAgoqOjGTRoUKnudE3TmDBhAvHx8VitVnr06MFff/112rw/++wz769HzZs3945xFUIIIUTZ9CYjHSbdRWiz+uTvTqfwQCb2nAJKjueRl3YYR04hDW7sTsNhZS8zDhDWsgHd5j5Jl/88SssxQ2jxyA1c9vq99PrfcyTe0L1G91WrDYqiEG5OxqCYsau5uDUHmqbi1hzY1VwMioUwc3KN3HfDhg3ZsGEDvXv35qmnnqJNmzZ06NCB6dOnM2bMGCZOnOit49dff023bt248847ady4MTfddBN79+4lJiam0uVNnDiR8ePHM2nSJJo1a0bfvn356quvvMvdV1br1q1Zvnw5O3fu5IorrqBdu3aMGzfOu7IjwNy5c4mJiaFbt25cd911jBw5kqCgICyWf1aBnTVrFk6nkw4dOvDwww/zwgsvVFjuuHHjaN++PX379qVHjx7ExsYyaNAgnzRjxoxBr9fTvHlzoqKi2L9/P3Xq1OHrr7/mt99+o02bNtx7772MGDHitIHjCb179yYuLo6+ffsSHx9f+YYqx1133cV7771HamoqrVq1onv37qSmpnpfB4PBwLRp03j33XeJj49n4MCB3mt79uyJ2+32Bm2KoniHenbt2tWb7mzv+YQbbriBOXPmMGzYMP73v/8RGhrK//73P3r16kWzZs145513+Oijj2jRokWZ11fmNTudtm3bMmXKFF555RVatmzJhx9+yKRJkyp17YgRI3A4HKXmNZpMJp566ilat25Nt27d0Ov1zJ8/v8w8qpK2KhTt1J+OzrF+/fpx0003cemll+JyuXj66afZtGkTW7Zs8Q4ReOWVV3jxxRdJTU2lcePGvPDCC/z8889s377dZ0LuyVavXs0VV1zBxIkTue666/j8888ZP348K1asKDWeuCx5eXmEhISQm5tLcLB/TVQXQgghzlbJsVz2L1zBvoUrKDmWh6JTCG+dRL1BXYnvcwm608zFOR/V9Hd7sesYWfZd2N3ZqJqKTtFh0YcRZk6u9q0HLkYHDx4kISHBO8fInxQVFREfH8+sWbPKHD4ozl8vvvgi8+fPZ9OmTbVdlVJqPZA71dGjR4mOjmb58uV069YNTdOIj49n9OjRPPHEEwDY7XZiYmJ45ZVXuOeee8rMZ+jQoeTl5fHNN994j/Xr14+wsLBSmwiWRQI5IYQQAlSXG2deITqjAUOg1a97087Fd7umaTjUPNyaA71iwqQL9us2q00//vgjBQUFtGrVivT0dMaOHcuhQ4fYsWNHqYUtzleqqpKRkcHkyZP59NNPSUtLw2CQJSr8QUFBAVu3bmXAgAFMnDjRu5H9+aTWh1ae6sQY3hMbU+7Zs4eMjAyuuuoqbxqz2Uz37t0rXLpz9erVPtcA9O3bt9xr7HY7eXl5Pg8hhBDiYqcz6DGHB2MMKr20vihNURTM+hBshijM+hBps7PgdDr597//TYsWLbjuuuuIiorybg7uL04My/z444+ZNWuWBHF+ZNSoUXTt2pXu3bufdruI2nJevZs0TePRRx+la9eutGzZEsA7mfHU8dsxMTHs27ev3LwyMjLKvKa8yZGTJk3iueeeO5vqCyGEEEKIatK3b1/69u1b29U4K4mJiaUWwBH+ITU1ldTU1NquRoXOqx65UaNG8eeff5Y59LGsfSxO9ytXVa556qmnyM3N9T4OHDhQxdoLIYQQQgghxLlx3vTIPfjgg3z55Zf8/PPPPru/x8bGAp4etpNXMsrMzKxwlaXY2NhSvW8VXWM2m0+7maUQQgghhBBCnA9qvUdO0zRGjRrF//73P3788cdSy9c2aNCA2NhYvv/+e+8xh8PB8uXL6dy5c7n5durUyecagO+++67Ca4QQQgghhBDCH9R6j9wDDzzAf//7X7744guCgoK8vWghISHeTR9Hjx7NSy+9RKNGjWjUqBEvvfQSNpuNW265xZvP7bffTp06dbx7Qjz88MN069aNV155hYEDB/LFF1/www8/sGLFilq5TyGEEEIIIYSoLrUeyL399tsAPjvJA8yePZuUlBQAxo4dS3FxMffffz/Z2dlcfvnlfPfddz57yO3fv99nl/nOnTszf/58nnnmGcaNG0fDhg1ZsGBBpfaQE0IIIYQQQojz2Xm3j9z5QvaRE0IIIS4s8t0uhLiQ1PocOSGEEEIIIU62bNkyFEUhJyentqsixHlLAjkhhBBCCD+VkpLCoEGDKp1eURQWLlxYY/U5Ez169GD06NE+xzp37kx6ejohISG1Uykh/ECtz5ETQgghhLhQaJpGriMfh+rEpDMSYgo67b63Fyqn04nRaDyja00mk3cLKiFE2aRHTgghhBCiGhwryWLVkfWsPLKW1Ud+Z+WRtaw6sp5jJVnnpPwePXrw0EMPMXbsWMLDw4mNjWXChAne84mJiQBcd911KIrifQ7w1Vdfcckll2CxWEhKSuK5557D5XJ5z2/bto2uXbtisVho3rw5P/zwg0/v3t69e1EUhY8//pgePXpgsVj44IMPOH78ODfffDN169bFZrPRqlUrPvroI2++KSkpLF++nKlTp6IoCoqisHfv3jKHVn722We0aNECs9lMYmIikydP9rn/xMREXnrpJe68806CgoKoV68e//nPf6qtfYU430ggJ4QQQghxlo6VZLH+6CaOlWRh1pkJMQVh1pl9jp8Lc+bMISAggDVr1vDqq6/y/PPPe/fVXbt2LeBZGTw9Pd37fMmSJdx222089NBDbNmyhXfffZfU1FRefPFFAFRVZdCgQdhsNtasWcN//vMfnn766TLLf+KJJ3jooYfYunUrffv2paSkhEsuuYRFixaxefNm7r77boYNG8aaNWsAmDp1Kp06dWLkyJGkp6eTnp5OQkJCqXzXr1/PkCFDuOmmm9i0aRMTJkxg3LhxpKam+qSbPHkyHTp0YMOGDdx///3cd999bNu2rVraVojzjQytFEIIIYQ4C5qmsT1nNyVuO6GmYO9QSpPeiFEXTI4jjx05u4mICavxYZatW7fm2WefBaBRo0bMmDGDpUuX0qdPH6KiogAIDQ31Gbb44osv8uSTTzJ8+HAAkpKSmDhxImPHjuXZZ5/lu+++Iy0tjWXLlnmve/HFF+nTp0+p8kePHs3111/vc2zMmDHefz/44IN8++23fPLJJ1x++eWEhIRgMpmw2WwVDqWcMmUKV155JePGjQOgcePGbNmyhddee827XRXA1Vdfzf333w94gso33niDZcuW0bRp00q3oRD+QgI5IYQQQoizkOvIJ8eRS4DBVipQUxSFAIONbEcuuY58Qs01u+1B69atfZ7HxcWRmZlZ4TXr169n7dq13h44ALfbTUlJCUVFRWzfvp2EhASfQOuyyy4rM68OHTr4PHe73bz88sssWLCAQ4cOYbfbsdvtBAQEVOm+tm7dysCBA32OdenShTfffBO3241erwd8719RFGJjY097/0L4KwnkhBBCCCHOgkN14lJVAgxl/1ll0OkpdKk4VGeN1+XUxUUURUFV1QqvUVWV5557rlRPGoDFYkHTtEr3JJ4aoE2ePJk33niDN998k1atWhEQEMDo0aNxOByVyu+EsupQ1lbIZ3L/QvgrCeSEEEIIIc6CSWfEoNPhUl2Y9KVXaXSpbgw6HSbdma3gWJ2MRiNut9vnWPv27dm+fTvJycllXtO0aVP279/PkSNHiImJAf6Zb3c6v/zyCwMHDuS2224DPEHjzp07adasmTeNyWQqVadTNW/enBUrVvgcW7VqFY0bN/b2xglxsZHFToQQQgghzkKIKYhQUwiFrqJSvUSaplHoKiLMFEKIKaiWaviPxMREli5dSkZGBtnZ2QCMHz+euXPnMmHCBP766y+2bt3KggULeOaZZwDo06cPDRs2ZPjw4fz555+sXLnSu9jJ6XrqkpOT+f7771m1ahVbt27lnnvuISMjo1Sd1qxZw969ezl27FiZPWiPPfYYS5cuZeLEiezYsYM5c+YwY8YMn/l3QlxsJJATQgghhDgLiqLQJDQJi95MjiMPh9uJqqk43E5yHHlY9GYahyadF/vJTZ48me+//56EhATatWsHQN++fVm0aBHff/89l156KR07dmTKlCnUr18fAL1ez8KFCykoKODSSy/lrrvu8gZ5FoulwvLGjRtH+/bt6du3Lz169CA2NrbUBuZjxoxBr9fTvHlzoqKi2L9/f6l82rdvz8cff8z8+fNp2bIl48eP5/nnn/dZ6ESIi42ilTXAWJCXl0dISAi5ubkEB9fsxGQhhBBC1Lya/m4/VpLF9pzd5DhycakqBp2OMFMIjUOTiLSEV3t5tWnlypV07dqVXbt20bBhw9qujhAXJZkjJ4QQQghRDSIt4UTEhJHryMehOjHpjISYgs6Lnriz9fnnnxMYGEijRo3YtWsXDz/8MF26dJEgTohaJIGcEEIIIUQ1URSlxrcYqA35+fmMHTuWAwcOEBkZSe/evZk8eXJtV0uIi5oMrSyHDK0UQgghLizy3S6EuJDIYidCCCGEEEII4WckkBNCCCGEEEIIPyOBnBBCCCGEEEL4GQnkhBBCCCGEEMLPSCAnhBBCCCGEEH5GAjkhhBBCCCGE8DMSyAkhhBBCiHNm2bJlKIpCTk4OAKmpqYSGhtZ4uYqisHDhwhovR4hzRQI5IYQQQohqomoa6YX57M7NIr0wH7WGt+tNSUlBURRefvlln+MLFy5EUZQaLbu6DB06lB07dlRbfhMmTKBt27aljqenp9O/f/9qK0eI2mao7QoIIYQQQlwI9uZls+LQXg4X5OFU3Rh1euIDg+laJ5HE4LAaK9disfDKK69wzz33EBZWPeU4HA5MJlO15HU6VqsVq9Va4+XExsbWeBlCnEvSIyeEEEIIcZb25mWzcNdf7M3NJtBoItYWRKDRxN7cv4/nZddY2b179yY2NpZJkyaVm+azzz6jRYsWmM1mEhMTmTx5ss/5xMREXnjhBVJSUggJCWHkyJHeIY+LFi2iSZMm2Gw2brjhBgoLC5kzZw6JiYmEhYXx4IMP4na7vXl98MEHdOjQgaCgIGJjY7nlllvIzMwst26nDq1MTExEUZRSjxOeeOIJGjdujM1mIykpiXHjxuF0Or15Pffcc/zxxx/e61JTU4HSQys3bdpEr169sFqtREREcPfdd1NQUOA9n5KSwqBBg3j99deJi4sjIiKCBx54wFuWELVNAjkhhBBCiLOgahorDu2lwOEgPiAIq8GITlGwGozEBwRR4HCw8tDeGhtmqdfreemll5g+fToHDx4sdX79+vUMGTKEm266iU2bNjFhwgTGjRvnDXBOeO2112jZsiXr169n3LhxABQVFTFt2jTmz5/Pt99+y7Jly7j++uv5+uuv+frrr5k3bx7/+c9/+PTTT735OBwOJk6cyB9//MHChQvZs2cPKSkplb6ftWvXkp6eTnp6OgcPHqRjx45cccUV3vNBQUGkpqayZcsWpk6dysyZM3njjTcAzzDNxx57jBYtWnjzGDp0aKkyioqK6NevH2FhYaxdu5ZPPvmEH374gVGjRvmk++mnn0hLS+Onn35izpw5pKamlmo3IWqLDK0UQgghhDgLR4oKOFyQR7jFWmpemqIohFusHCrI40hRAXEBQTVSh+uuu462bdvy7LPP8v777/ucmzJlCldeeaU3OGvcuDFbtmzhtdde8wmwevXqxZgxY7zPV6xYgdPp5O2336Zhw4YA3HDDDcybN48jR44QGBhI8+bN6dmzJz/99JM3YLrzzju9eSQlJTFt2jQuu+wyCgoKCAwMPO29REVFef/98MMPk56eztq1a73HnnnmGe+/ExMTeeyxx1iwYAFjx47FarUSGBiIwWCocCjlhx9+SHFxMXPnziUgIACAGTNmMGDAAF555RViYmIACAsLY8aMGej1epo2bco111zD0qVLGTly5GnvQ4iaJj1yQgghhBBnodjlxKm6MevL/n3cpNfjVN0Uu2p2SN4rr7zCnDlz2LJli8/xrVu30qVLF59jXbp0YefOnT5DIjt06FAqT5vN5g3iAGJiYkhMTPQJyGJiYnyGTm7YsIGBAwdSv359goKC6NGjBwD79++v0v385z//4f333+eLL77wCe4+/fRTunbtSmxsLIGBgYwbN67KeW/dupU2bdp4gzjwtImqqmzfvt17rEWLFuj1eu/zuLi4CoeJCnEuSSAnhBBCCHEWrAYjRp0eu9tV5nmH27PwidVgrNF6dOvWjb59+/Lvf//b57imaaV6CrUyhnmeHNScYDT61llRlDKPqaoKQGFhIVdddRWBgYF88MEHrF27ls8//xzwDLmsrGXLlvHggw8yd+5c2rRp4z3+66+/ctNNN9G/f38WLVrEhg0bePrpp6uUN5TdJiffzwkV3asQtU2GVgohhBBCnIUYWyDxgcHszc0mPsDgEwhomkZWSTENQsKIsZ1+WOHZevnll2nbti2NGzf2HmvevDkrVqzwSbdq1SoaN27s09tUHbZt28axY8d4+eWXSUhIAGDdunVVymPXrl0MHjyYf//731x//fU+51auXEn9+vV5+umnvcf27dvnk8ZkMvn0NJalefPmzJkzh8LCQm8Au3LlSnQ6nU/bCXE+kx45IYQQQoizoFMUutZJJNBk4nBhPsUuJ25Npdjl5HBhPkEmM13qJKI7B/u6tWrViltvvZXp06d7jz322GMsXbqUiRMnsmPHDubMmcOMGTN85sNVl3r16mEymZg+fTq7d+/myy+/ZOLEiZW+vri4mAEDBtC2bVvuvvtuMjIyvA+A5ORk9u/fz/z580lLS2PatGneHr8TEhMT2bNnDxs3buTYsWPY7fZS5dx6661YLBaGDx/O5s2b+emnn3jwwQcZNmyYd36cEOc7CeSEEEIIIc5SYnAYg5JbkBgSRoHTwZGiAgqcDhqEhDEwuXmN7iN3qokTJ/oMnWzfvj0ff/wx8+fPp2XLlowfP57nn3++SitJVlZUVBSpqal88sknNG/enJdffpnXX3+90tcfOXKEbdu28eOPPxIfH09cXJz3ATBw4EAeeeQRRo0aRdu2bVm1apV3EZcTBg8eTL9+/ejZsydRUVF89NFHpcqx2WwsWbKErKwsLr30Um644QauvPJKZsyYcXYNIMQ5pGhlDZIW5OXlERISQm5uLsHBwbVdHSGEEEKcpXPx3a5qGkeKCih2ObEajMTYAs9JT5wQ4uIjc+SEEEIIIaqJTlFqbIsBIYQ4mQytFEIIIYQQQgg/I4GcEEIIIYQQQvgZCeSEEEIIIYQQws9IICeEEEIIIYQQfkYCOSGEEEIIIYTwMxLICSGEEEIIIYSfkUBOCCGEEEIIIfyMBHJCCCGEEEII4WckkBNCCCGEEJUyYcIE2rZt632ekpLCoEGDaq0+54tly5ahKAo5OTm1XRVxEZFATgghhBCimqiaxv7jhWw5nMv+44WomlbjZR44cIARI0YQHx+PyWSifv36PPzwwxw/frzGy546dSqpqane5z169GD06NE1Xm518Ke6ClEWQ21XQAghhBDiQrA9I4/Ff6az91ghdpeK2aAjMTKAa1rH0SQ2uEbK3L17N506daJx48Z89NFHNGjQgL/++ovHH3+cb775hl9//ZXw8PAaKRsgJCSkxvI+H2iahtvtxmCQP5nF+Ud65IQQQgghztL2jDze/3kP29LzCLYYSQizEWwxsi3dc3x7Rl6NlPvAAw9gMpn47rvv6N69O/Xq1aN///788MMPHDp0iKefftqbVlEUFi5c6HN9aGioT4/aE088QePGjbHZbCQlJTFu3DicTme55Z88tDIlJYXly5czdepUFEVBURT27NlDcnIyr7/+us91mzdvRqfTkZaWVm7es2bNokWLFpjNZuLi4hg1apT33P79+xk4cCCBgYEEBwczZMgQjhw54j1/YgjovHnzSExMJCQkhJtuuon8/Pxy67p3717vEMklS5bQoUMHzGYzv/zyC3a7nYceeojo6GgsFgtdu3Zl7dq15dZdiHNBAjkhhBBCiLOgahqL/0wnp9hBYkQAAWYDep1CgNlAYkQAOcUOvv4zvdqHWWZlZbFkyRLuv/9+rFarz7nY2FhuvfVWFixYgFaFcoOCgkhNTWXLli1MnTqVmTNn8sYbb1Tq2qlTp9KpUydGjhxJeno66enp1KtXjzvvvJPZs2f7pJ01axZXXHEFDRs2LDOvt99+mwceeIC7776bTZs28eWXX5KcnAx4eskGDRpEVlYWy5cv5/vvvyctLY2hQ4f65JGWlsbChQtZtGgRixYtYvny5bz88svl1jUhIcF77dixY5k0aRJbt26ldevWjB07ls8++4w5c+bw+++/k5ycTN++fcnKyqp02wpR3aSfWAghhBDiLBzMKmLvsUKigywoiuJzTlEUooMs7DlWyMGsIupFBFRbuTt37kTTNJo1a1bm+WbNmpGdnc3Ro0eJjo6uVJ7PPPOM99+JiYk89thjLFiwgLFjx5722pCQEEwmEzabjdjYWO/xO+64g/Hjx/Pbb79x2WWX4XQ6+eCDD3jttdfKzeuFF17gscce4+GHH/Yeu/TSSwH44Ycf+PPPP9mzZ483+Jo3bx4tWrRg7dq13nSqqpKamkpQUBAAw4YNY+nSpbz44ovl1vWE559/nj59+gBQWFjI22+/TWpqKv379wdg5syZfP/997z//vs8/vjjp20bIWqC9MgJIYQQQpyFArsLu0vFYtSXed5i1GN3qRTYXee0Xid64kwmU6Wv+fTTT+natSuxsbEEBgYybtw49u/ff1b1iIuL45prrmHWrFkALFq0iJKSEm688cYy02dmZnL48GGuvPLKMs9v3bqVhIQEnx605s2bExoaytatW73HEhMTvUHciXpkZmZWqs4dOnTw/jstLQ2n00mXLl28x4xGI5dddplPeUKcaxLICSGEEEKchUCzAbNBR4nTXeb5Eqcbs0FHoLl6B0IlJyejKApbtmwp8/y2bduIiooiNDQU8PQOnjrM8uT5b7/++is33XQT/fv3Z9GiRWzYsIGnn34ah8Nx1nW96667mD9/PsXFxcyePZuhQ4dis9nKTHvqMNFTaZpWquezrONGo9HnvKIoqKpaqfoGBPzTc3qizU4ts7x6CHGuSCAnhBBCCHEW6obbSIwMIDO/pFSgpGkamfklNIgMoG542YHLmYqIiKBPnz689dZbFBcX+5zLyMjgww8/JCUlxXssKiqK9PR07/OdO3dSVFTkfb5y5Urq16/P008/TYcOHWjUqBH79u2rUp1MJhNud+mA9uqrryYgIIC3336bb775hjvvvLPcPIKCgkhMTGTp0qVlnm/evDn79+/nwIED3mNbtmwhNze33GGmVanrqZKTkzGZTKxYscJ7zOl0sm7duiqVJ0R1k0BOCCGEEOIs6BSFa1rHEWo1sfd4IYV2F25Vo9DuYu/xQkKtJq5uHYeuBnpvZsyYgd1up2/fvvz8888cOHCAb7/9lj59+tC4cWPGjx/vTdurVy9mzJjB77//zrp167j33nt9eq2Sk5PZv38/8+fPJy0tjWnTpvH5559XqT6JiYmsWbOGvXv3cuzYMW8PmF6vJyUlhaeeeork5GQ6depUYT4TJkxg8uTJTJs2jZ07d/L7778zffp0AHr37k3r1q259dZb+f333/ntt9+4/fbb6d69u8+QyDOt66kCAgK47777ePzxx/n222/ZsmULI0eOpKioiBEjRlS6PCGqmwRyQgghhBBnqUlsMCO6NaBpXDB5JU4OZBeRV+KkWZzneE3tI9eoUSPWrl1LUlISQ4YMoX79+vTv35/GjRuzcuVKAgMDvWknT55MQkIC3bp145ZbbmHMmDE+wxsHDhzII488wqhRo2jbti2rVq1i3LhxVarPmDFj0Ov1NG/enKioKJ/5dSNGjMDhcFTYG3fC8OHDefPNN3nrrbdo0aIF1157LTt37gT+2UYhLCyMbt260bt3b5KSkliwYEG11fVUL7/8MoMHD2bYsGG0b9+eXbt2sWTJEsLCwqpUphDVSdGqsibtRSQvL4+QkBByc3MJDq6ZD18hhBBCnDvn4rtd1TQOZhVRYHcRaDZQN9xWIz1xFXn22WeZMmUK33333Wl7vs6llStX0qNHDw4ePEhMTExtV0cIvyfbDwghhBBCVBOdolTrFgNn4rnnnvMOG7z88svR6Wp3AJbdbufAgQOMGzeOIUOGSBAnRDWRQE4IIYQQ4gJzxx131HYVvD766CNGjBhB27ZtmTdvXm1XR4gLhgytLIcMrRRCCCEuLPLdLoS4kNT6Yic///wzAwYMID4+3jt59WSKopT5eO2118rNMzU1tcxrSkpKavhuhBBCCCGEEKLm1XogV1hYSJs2bZgxY0aZ59PT030es2bNQlEUBg8eXGG+wcHBpa61WCw1cQtCCCGEEEIIcU7V+hy5/v37079//3LPx8bG+jz/4osv6NmzJ0lJSRXmqyhKqWuFEEIIIYQQ4kJQ6z1yVXHkyBEWL15cqc0XCwoKqF+/PnXr1uXaa69lw4YNFaa32+3k5eX5PIQQQgghhBDifORXgdycOXMICgri+uuvrzBd06ZNSU1N5csvv+Sjjz7CYrHQpUsX70aSZZk0aRIhISHeR0JCQnVXXwghhBBCCCGqxXm1aqWiKHz++ecMGjSozPNNmzalT58+TJ8+vUr5qqpK+/bt6datG9OmTSszjd1ux263e5/n5eWRkJAgK1sJIYQQFwhZtVIIcSHxmx65X375he3bt3PXXXdV+VqdTsell15aYY+c2WwmODjY5yGEEEIIcbFJSUkp90d1f7Rs2TIURSEnJ6dK102YMIGYmJgyV1UX4nxQ64udVNb777/PJZdcQps2bap8raZpbNy4kVatWtVAzYQQQgghPFRVIzM9n6IiBzabiei4IHQ6pUbKUpSK8x0+fDipqalVznfq1KlU54CtlJQUcnJyai0Y6ty5M+np6YSEhFT6mq1bt/Lcc8/x+eef07FjR8LCwmqwhkKcmVoP5AoKCti1a5f3+Z49e9i4cSPh4eHUq1cP8AyF+OSTT5g8eXKZedx+++3UqVOHSZMmAfDcc8/RsWNHGjVqRF5eHtOmTWPjxo383//9X83fkBBCCCEuSvt2Z7HipzQOH8jF4XBjMumJTwiha8+G1E8Kr/by0tPTvf9esGAB48ePZ/v27d5jVqvVJ73T6cRoNJ4236oEPOeSw+HAZDJV+TqTyVTllczT0tIAGDhw4GkD5opUts2FOBO1PrRy3bp1tGvXjnbt2gHw6KOP0q5dO8aPH+9NM3/+fDRN4+abby4zj/379/t8mOXk5HD33XfTrFkzrrrqKg4dOsTPP//MZZddVrM3I4QQQoiL0r7dWSxc8Cd7dmUREGgmNj6IgEAze3Z5ju/bnVXtZcbGxnofISEh3q2XYmNjKSkpITQ0lI8//pgePXpgsVj44IMPOH78ODfffDN169bFZrPRqlUrPvroI598Tx1aqWkar776KklJSVitVtq0acOnn37qc81ff/3FNddcQ3BwMEFBQVxxxRWkpaUxYcIE5syZwxdffIGiKCiKwrJlywDYtGkTvXr1wmq1EhERwd13301BQUGpekyaNIn4+HgaN27M888/X+YIq0suucTnb8eTnTq0MjU1ldDQUJYsWUKzZs0IDAykX79+3r8lJ0yYwIABAwDP9JyTA7nZs2fTrFkzLBYLTZs25a233vKe27t3L4qilGrzyl73v//9j549e2Kz2WjTpg2rV6/2uY+VK1fSvXt3bDYbYWFh9O3bl+zs7Eq/RuICpIky5ebmaoCWm5tb21URQgghRDWoqe92t1vVPnjvN23ik99q789Ypc36v9Xex/szVmkTn/xW++C93zS3W63Wck82e/ZsLSQkxPt8z549GqAlJiZqn332mbZ7927t0KFD2sGDB7XXXntN27Bhg5aWlqZNmzZN0+v12q+//uq9dvjw4drAgQO9z//9739rTZs21b799lstLS1Nmz17tmY2m7Vly5ZpmqZpBw8e1MLDw7Xrr79eW7t2rbZ9+3Zt1qxZ2rZt27T8/HxtyJAhWr9+/bT09HQtPT1ds9vtWmFhoRYfH69df/312qZNm7SlS5dqDRo00IYPH+5Tj8DAQG3YsGHa5s2btU2bNmkHDhzQdDqd9ttvv3nT/fHHH5qiKFpaWlqZbfPTTz9pgJadne1tK6PRqPXu3Vtbu3attn79eq1Zs2baLbfcommapuXn52uzZ8/WAG+dNU3T/vOf/2hxcXHe9vzss8+08PBwLTU1tcI2r+x1TZs21RYtWqRt375du+GGG7T69etrTqdT0zRN27Bhg2Y2m7X77rtP27hxo7Z582Zt+vTp2tGjRyv1GokLU60PrRRCCCGE8GeZ6fkcPpBLeISt1DA8RVEIj7Bx+EAumen5xNY5t4upjR49utS2TWPGjPH++8EHH+Tbb7/lk08+4fLLLy91fWFhIVOmTOHHH3+kU6dOACQlJbFixQreffddunfvzv/93/8REhLC/PnzvcMIGzdu7M3DarVit9t9hjfOmTOH4uJi5s6dS0BAAAAzZsxgwIABvPLKK8TExAAQEBDAe++95zOksm/fvsyePZtLL70U8PR2de/enaSkpEq3i9Pp5J133qFhw4YAjBo1iueffx6AwMBAQkNDAXzqPHHiRCZPnuxtzwYNGrBlyxbeffddhg8f7k13aptX9roxY8ZwzTXXAJ5pQi1atGDXrl00bdqUV199lQ4dOvj05LVo0QKo3GskLkwSyAkhhBBCnIWiIgcOh5twi77M8yaznuwsN0VFjnNcM+jQoYPPc7fbzcsvv8yCBQs4dOiQd/ulE8HUqbZs2UJJSQl9+vTxOe5wOLzTYjZu3MgVV1xRpblgW7dupU2bNj7ldunSBVVV2b59uzeQa9WqVal5cSNHjuTOO+9kypQp6PV6Pvzww3LXUSiPzWbzBnEAcXFxZGZmlpv+6NGjHDhwgBEjRjBy5EjvcZfLVWpO4cltXpXrWrdu7VMfgMzMTJo2bcrGjRu58cYby6xbZV4jcWGSQE4IIYQQ4izYbCZMJj32EjdWW+nlBxx2z8InNlvVF+o4W6cGaJMnT+aNN97gzTffpFWrVgQEBDB69GgcjrKDTFVVAVi8eDF16tTxOWc2m4HSi6pUhqZp5S4icvLxsgLMAQMGYDab+fzzzzGbzdjtdgYPHlyl8k8NOhVFqXClzhPtMHPmzFI9l3q9bwB/cp2rct3JdTrRBieur6iNK/MaiQuTBHJCCCGEEGchOi6I+IQQ9uzKIt4a7BOIaJpG1vEiGiSHEx0XVIu19Pjll18YOHAgt912G+AJAnbu3EmzZs3KTN+8eXPMZjP79+8vd4he69atmTNnTrkrNJpMJtxud6l858yZQ2FhoTfwWblyJTqdzmdYZlkMBgPDhw9n9uzZmM1mbrrpJmw222nv/WzExMRQp04ddu/eza233lrj152qdevWLF26lOeee67Uucq8RuLCJIGcEEIIIcRZ0OkUuvZsyLHMQg4fzCM8wobJrMdhd5N1vIigYDNdezassf3kqiI5OZnPPvuMVatWERYWxpQpU8jIyCg3kAsKCmLMmDE88sgjqKpK165dycvLY9WqVQQGBjJ8+HBGjRrF9OnTuemmm3jqqacICQnh119/5bLLLqNJkyYkJiayZMkStm/fTkREBCEhIdx66608++yzDB8+nAkTJnD06FEefPBBhg0b5h1WWZG77rrLW+eVK1dWaxuVZ8KECTz00EMEBwfTv39/7HY769atIzs7m0cffbTarzvZU089RatWrbj//vu59957MZlM/PTTT9x4441ERkae9jUSF6Za335ACCGEEMLf1U8KZ9DQ1jRIDqewwM6R9HwKC+w0SPYcr4l95M7EuHHjaN++PX379qVHjx7Exsb6bDVQlokTJzJ+/HgmTZpEs2bN6Nu3L1999RUNGjQAICIigh9//JGCggK6d+/OJZdcwsyZM729cyNHjqRJkyZ06NCBqKgoVq5cic1mY8mSJWRlZXHppZdyww03cOWVVzJjxoxK3UejRo3o3LkzTZo0KXORlppw11138d5775GamkqrVq3o3r07qamp3nao7utO1rhxY7777jv++OMPLrvsMjp16sQXX3yBweDpkzndayQuTIpW0YDgi1heXh4hISHk5uYSHHxuV5gSQgghRPU7F9/tqqqRmZ5PUZEDm81EdFzQedETVxU333wzer3euwfa+UjTNJo2bco999xT6V4tIS40MrRSCCGEEKKa6HTKOd9ioLq4XC527NjB6tWrueeee2q7OuXKzMxk3rx5HDp0iDvuuKO2qyNErZFATgghhBBCsHnzZjp37kzPnj259957a7s65YqJiSEyMpL//Oc/hIWF1XZ1hKg1EsgJIYQQQgjatm1LUVFRbVfjtGRWkBAestiJEEIIIYQQQvgZCeSEEEIIIYQQws9IICeEEEIIIYQQfkYCOSGEEEIIIYTwMxLICSGEEEIIIYSfkUBOCCGEEEIIIfyMBHJCCCGEEEII4WckkBNCCCGEEEIIPyOBnBBCCCGEEEL4GQnkhBBCCCGEEMLPSCAnhBBCCCGEEH5GAjkhhBBCCCGE8DMSyAkhhBBCCCGEn5FATgghhBBCCCH8jARyQgghhBBCCOFnJJATQgghhBBCCD8jgZwQQgghhBBC+BkJ5IQQQgghhBDCz0ggJ4QQQgghhBB+RgI5IYQQQgghhPAzEsgJIYQQQgghhJ+RQE4IIYQQQggh/IwEckIIIYQQQgjhZySQE0IIIYQQQgg/I4GcEEIIIYQQQvgZCeSEEEIIIYQQws9IICeEEEIIIYQQfkYCOSGEEEIIIYTwMxLICSGEEEIIIYSfkUBOCCGEEEIIIfyMBHJCCCGEEEII4WckkBNCCCGEEEIIPyOBnBBCCCGEEEL4GQnkhBBCCCGEEMLPSCAnhBBCCCGEEH5GAjkhhBBCCCGE8DMSyAkhhBBCCCGEn5FATgghhBBCCCH8jARyQgghhBBCCOFnJJATQgghhBBCCD8jgZwQQgghhBBC+BkJ5IQQQgghhBDCz0ggJ4QQQgghhBB+RgI5IYQQQgghhPAzEsgJIYQQQgghhJ+RQE4IIYQQQggh/IwEckIIIYQQQgjhZ2o9kPv5558ZMGAA8fHxKIrCwoULfc6npKSgKIrPo2PHjqfN97PPPqN58+aYzWaaN2/O559/XkN3IIQQQgghhBDnVq0HcoWFhbRp04YZM2aUm6Zfv36kp6d7H19//XWFea5evZqhQ4cybNgw/vjjD4YNG8aQIUNYs2ZNdVdfCCGEEEIIIc45RdM0rbYrcYKiKHz++ecMGjTIeywlJYWcnJxSPXUVGTp0KHl5eXzzzTfeY/369SMsLIyPPvqoUnnk5eUREhJCbm4uwcHBlS5bCCGEEOcn+W4XQlxIar1HrjKWLVtGdHQ0jRs3ZuTIkWRmZlaYfvXq1Vx11VU+x/r27cuqVavKvcZut5OXl+fzEEIIIYQQQojz0XkfyPXv358PP/yQH3/8kcmTJ7N27Vp69eqF3W4v95qMjAxiYmJ8jsXExJCRkVHuNZMmTSIkJMT7SEhIqLZ7EEIIIYQQQojqZKjtCpzO0KFDvf9u2bIlHTp0oH79+ixevJjrr7++3OsURfF5rmlaqWMne+qpp3j00Ue9z/Py8iSYE0IIIYQQQpyXzvtA7lRxcXHUr1+fnTt3lpsmNja2VO9bZmZmqV66k5nNZsxmc7XVUwghhBBCCCFqynk/tPJUx48f58CBA8TFxZWbplOnTnz//fc+x7777js6d+5c09UTQgghhBBCiBpX6z1yBQUF7Nq1y/t8z549bNy4kfDwcMLDw5kwYQKDBw8mLi6OvXv38u9//5vIyEiuu+467zW33347derUYdKkSQA8/PDDdOvWjVdeeYWBAwfyxRdf8MMPP7BixYpzfn9CCCGEEEIIUd1qPZBbt24dPXv29D4/MU9t+PDhvP3222zatIm5c+eSk5NDXFwcPXv2ZMGCBQQFBXmv2b9/PzrdP52LnTt3Zv78+TzzzDOMGzeOhg0bsmDBAi6//PJzd2NCCCGEEEIIUUPOq33kziey14wQQghxYZHvdiHEhcTv5sgJIYQQQgghxMVOAjkhhBBCCCGE8DMSyAkhhBBCCCGEn5FATgghhBBCCCH8jARyQgghhBBCCOFnJJATQgghhBBCCD8jgZwQQgghhBBC+BkJ5IQQQgghhBDCz0ggJ4QQQgghhBB+RgI5IYQQQgghhPAzEsgJIYQQQgghhJ+RQE4IIYQQQggh/IwEckIIIYQQQgjhZySQE0IIIYQQQgg/I4GcEEIIIYQQQvgZCeSEEEIIIYQQws9IICeEEEIIIYQQfkYCOSGEEEIIIYTwMxLICSGEEEIIIYSfkUBOCCGEEEIIIfyMBHJCCCGEEEII4WckkBNCCCGEEEIIPyOBnBBCCCGEEEL4GQnkhBBCCCGEEMLPSCAnhBBCCCGEEH5GAjkhhBBCCCGE8DMSyAkhhBBCCCGEn5FATgghhBBCCCH8jARyQgghhBBCCOFnJJATQgghhBBCCD8jgZwQQgghhBBC+BkJ5IQQQgghhBDCz0ggJ4QQQgghhBB+RgI5IYQQQgghhPAzEsgJIYQQQgghhJ+RQE4IIYQQQggh/IwEckIIIYQQQgjhZySQE0IIIYQQQgg/I4GcEEIIIYQQQvgZCeSEEEIIIYQQws9IICeEEEIIIYQQfkYCOSGEEEIIIYTwMxLICSGEEEIIIYSfkUBOCCGEEEIIIfyMBHJCCCGEEEII4WckkBNCCCGEEEIIPyOBnBBCCCGEEEL4GQnkhBBCCCGEEMLPSCAnhBBCCCGEEH5GAjkhhBBCCCGE8DMSyAkhhBBCCCGEn5FATgghhBBCCCH8jARyQgghhBBCCOFnJJATQgghhBBCCD8jgZwQQgghhBBC+BkJ5IQQQgghhBDCz0ggJ4QQQgghhBB+RgI5IYQQQgghhPAzEsgJIYQQQgghhJ8x1HYFzleapgGQl5dXyzURQgghRHU48Z1+4jteCCH8mQRy5cjPzwcgISGhlmsihBBCiOqUn59PSEhIbVdDCCHOiqLJz1JlUlWVw4cPExQUhKIoZabJy8sjISGBAwcOEBwcfI5reP6R9ihN2sSXtIcvaQ9f0h6+pD18VUd7aJpGfn4+8fHx6HQyu0QI4d+kR64cOp2OunXrViptcHCwfMmeRNqjNGkTX9IevqQ9fEl7+JL28HW27SE9cUKIC4X8HCWEEEIIIYQQfkYCOSGEEEIIIYTwMxLInQWz2cyzzz6L2Wyu7aqcF6Q9SpM28SXt4Uvaw5e0hy9pD1/SHkII4UsWOxFCCCGEEEIIPyM9ckIIIYQQQgjhZySQE0IIIYQQQgg/I4GcEEIIIYQQQvgZCeTKsXfvXkaMGEGDBg2wWq00bNiQZ599FofD4ZNOUZRSj3feeafCvO12Ow8++CCRkZEEBATwr3/9i4MHD9bk7VSLyrTJH3/8wc0330xCQgJWq5VmzZoxderU0+bdo0ePUu1400031eTtnLXKvkf279/PgAEDCAgIIDIykoceeqhUmlP563vkxRdfpHPnzthsNkJDQ0udT01NLfP/M4qikJmZWW6+/vj+gNO3B1xcnyGna4+L6fMDKvf+uJg+P061bNmycj8v1q5dW+51KSkppdJ37NjxHNZcCCHODdkQvBzbtm1DVVXeffddkpOT2bx5MyNHjqSwsJDXX3/dJ+3s2bPp16+f9/npNhsdPXo0X331FfPnzyciIoLHHnuMa6+9lvXr16PX62vkfqpDZdpk/fr1REVF8cEHH5CQkMCqVau4++670ev1jBo1qsL8R44cyfPPP+99brVaa/R+zlZl2sPtdnPNNdcQFRXFihUrOH78OMOHD0fTNKZPn15u3v76HnE4HNx444106tSJ999/v9T5oUOH+vx/BTx/dJWUlBAdHV1h3v72/oDTt8cJF8tnyOna42L6/IDTt8fF9vlxqs6dO5Oenu5zbNy4cfzwww906NChwmv79evH7Nmzvc9NJlON1FEIIWqVJirt1Vdf1Ro0aOBzDNA+//zzSueRk5OjGY1Gbf78+d5jhw4d0nQ6nfbtt99WV1XPmbLa5FT333+/1rNnzwrTdO/eXXv44YersWa149T2+PrrrzWdTqcdOnTIe+yjjz7SzGazlpubW2YeF8J7ZPbs2VpISMhp02VmZmpGo1GbO3duhen8/f1RUXtcjJ8hlX1/aNrF8flRXntcrJ8f5XE4HFp0dLT2/PPPV5hu+PDh2sCBA89NpYQQohbJ0MoqyM3NJTw8vNTxUaNGERkZyaWXXso777yDqqrl5rF+/XqcTidXXXWV91h8fDwtW7Zk1apVNVLvmlRem1Q1DcCHH35IZGQkLVq0YMyYMeTn51dXNc+ZU+919erVtGzZkvj4eO+xvn37YrfbWb9+fZl5XGjvkYrMnTsXm83GDTfccNq0F8L7ozwX82fI6VxMnx+nks8PX19++SXHjh0jJSXltGmXLVtGdHQ0jRs3ZuTIkRUO3RZCCH8lQysrKS0tjenTpzN58mSf4xMnTuTKK6/EarWydOlSHnvsMY4dO8YzzzxTZj4ZGRmYTCbCwsJ8jsfExJCRkVFj9a8J5bXJyVavXs3HH3/M4sWLK8zr1ltvpUGDBsTGxrJ582aeeuop/vjjD77//vvqrnaNKas9MjIyiImJ8UkXFhaGyWQq9/W+kN4jpzNr1ixuueWW0w6DuxDeH+W5mD9DTudi+vwoi3x++Hr//ffp27cvCQkJFabr378/N954I/Xr12fPnj2MGzeOXr16sX79etlMXAhxYantLsFz7dlnn9WACh9r1671uebQoUNacnKyNmLEiNPm//rrr2vBwcHlnv/www81k8lU6njv3r21e+65p+o3VA1qqk02b96sRUVFaRMnTqxyndatW6cB2vr166t87dmqzvYYOXKkdtVVV5Uqw2g0ah999FGZ5Z9v75EzaY/KDJ1btWqVBmjr1q2rcp387f1RlaGE/vYZUlPtcTF9fpTXHhfC50dZzqSNDhw4oOl0Ou3TTz+tcnmHDx/WjEaj9tlnn1XXLQghxHnhouuRGzVq1GlXM0tMTPT++/Dhw/Ts2ZNOnTrxn//857T5d+zYkby8PI4cOVLql1SA2NhYHA4H2dnZPr+YZmZm0rlz58rfSDWqiTbZsmULvXr1YuTIkeX2LFSkffv2GI1Gdu7cSfv27at8/dmozvaIjY1lzZo1Pseys7NxOp1lvj9OXHM+vUeq2h6V9d5779G2bVsuueSSKl/rT++PqvK3z5CaaI+L6fOjIhfC50dZzqSNZs+eTUREBP/617+qXF5cXBz169dn586dVb5WCCHOa7UdSZ7PDh48qDVq1Ei76aabNJfLValrpk+frlksFq2kpKTM8ycmoi9YsMB77PDhw34zEb0ybbJ582YtOjpae/zxx8+4nE2bNmmAtnz58jPO41w4XXucWKzg8OHD3mPz58+v1GIF/voe0bTT97jk5+drgYGB2vTp088of395f5xQlR65C/0zRNMqbo+L6fPj0I1qbgAAB35JREFUhNMtdnKxfX6cSlVVrUGDBtpjjz12RtcfO3ZMM5vN2pw5c6q5ZkIIUbskkCvHiaFyvf6/vXsLiar7wzj+TIyRoRVCdLITEqMidCTMDpQdLsoKRDtYYSVR4OBN0EUEWVREBdVNF9WMNlFhRUQhvpIxd0ZEGkRSIEyKUnSUsgM09nsvouFvzei8/07u/H7Ai9nrt/beazEu5nHv2ebmWltbmz158iTy89W1a9fs5MmTdv/+fWtubrZTp07ZkCFDrKysLFLT1tZmHo/Hbt++Hdm2bds2S01Ntbq6OmtoaLDc3FybPHly3GHxT4lnTr7eDrVu3bpu7c+ePYvUfDsnzc3NtmfPHrtz546FQiGrrq629PR0mzp1ap+ek3jmIxwOW1ZWli1cuNAaGhqsrq7OUlNTzev1Rmr+pvdIS0uLNTY22p49eywpKckaGxutsbHR3r59263u9OnTNmjQIHv16tV3+/hb3h9mvc9Hf1tDepuP/rR+mPU+H/1t/Yilrq7OJFlTU1PUdo/HY1euXDGzL38k2r59u9XX11soFLJgMGizZs2yMWPG2Js3b37naQPAL0eQi6GioiLmvftf1dTU2JQpUywpKckGDx5sWVlZduzYMfv06VOkJhQKmSQLBoORbR8+fDCv12spKSmWmJhoeXl51tra+juH93+JZ05iffdh/PjxkZpv56S1tdXmzZtnKSkpNnDgQEtLS7OysjJ7+fLlbx7hfxPPfJh9+bC2bNkyS0xMtJSUFPN6vd2utvxN75Hi4uKo8/G/YzMzmzVrlhUVFUXdx9/y/jDrfT762xrS23z0p/XDLL7fl/60fsSydu1ay8nJidkuySoqKszM7P3797ZkyRIbPny4JSQk2Lhx46y4uNjR4weAWFxmZj/nJk0AAAAAwO/A/5EDAAAAAIchyAEAAACAwxDkAAAAAMBhCHIAAAAA4DAEOQAAAABwGIIcAAAAADgMQQ4AAAAAHIYgBwAAAAAOQ5AD0CeUl5fL5XLpxYsXUduzsrI0f/78yOvHjx/L5XLJ5XKpvLw8ap/NmzdHamKZNm2aXC6Xjhw5ErW9srIysg+XyyW3263U1FRt2rRJ7e3tvY4rEAhozZo18ng8GjBggCZMmNBrHwAAgN4Q5AA4WnJysiorK/X58+du2zs7O3Xp0iUNGTIkZt979+6psbFRkuTz+Xo8TkVFhW7duqUbN25oy5YtunDhgubOnat379712O/s2bN68OCBZs6cqbS0tDhHBQAA0DOCHABHW716tVpaWnTz5s1u26uqqtTV1aUVK1bE7Hv69GlJ0rJly/Tw4UPV19fHrM3KylJ2drYWLFig3bt3a8eOHQqFQrp69WqP51dbW6v79+/r7Nmz8ng88Q8MAACgBwQ5AI7m8XiUk5Mjv9/fbbvf71d+fr6GDh0atd/Hjx91/vx5TZ8+XUePHo30iVd2drYkqaWlpce6AQNYZgEAwM/HJwwAjldSUqKrV6/q9evXkqRHjx6pvr5eJSUlMftcuXJFr1+/1ubNmzVp0iTNmTNHVVVV6uzsjOuYzc3NkqThw4f/+AAAAAD+I4IcAMdbtWqV3G63zp8/L+nL990mTpzY7eEo3/L5fBo0aJCKiookfQmDnZ2dunjxYtT6rq4uhcNhdXZ2qrq6Wvv27VNycnKPt24CAAD8KgQ5AI6XlJSkwsJC+f1+hcNhBQIBbdq0KebTKkOhkILBoPLz8zVs2DBJUmFhoZKTk2PeXpmdna2EhAQlJycrLy9PI0eOVE1NjUaMGPGrhgUAABCT+0+fAABIktv9ZTnq6uqK2h4Oh5WQkBCzf0lJiebMmaP9+/fr+fPn2rhxY8xav98vM1NBQYE6Ojoi21esWKFz587p4cOHSk9P79YnEAgoIyNDbrdbI0aM0KhRo+IfHAAAwE9GkAPQJ3y9stXe3v7dVS4z05MnTzRjxoyY/WfPni2Px6O9e/dq8eLFGjt2bNS6z58/q7KyUpKUn58ftcbv9+vQoUPdtmVkZPR4fAAAgN+JIAegT8jNzZXL5VJVVZWmTZvWre2ff/7RmzdvtGjRoh73sWvXLl2+fFmlpaUxa2pra9XW1qbS0lIVFBR81+71ehUIBHTgwIHIVUIAAIC+hk8pAPqEtLQ0eb1eHT58WB0dHVq6dKkSExN1584dHTx4UDNmzIg8mCSW9evXa/369T3W+Hw+ud1u7dy5U6NHj/6ufevWrSorK1N1dbVWrlz5Q2OSpKamJjU1NUmSnj59qvfv3+vy5cuSpMzMTGVmZv7wMQAAQP/Dw04A9BnHjx/XiRMn1NDQoKKiIi1fvlxnzpxRaWmpgsGgBg4c+EP7f/Hiha5fv668vLyoIU6SNmzYoMTERPl8vh861lcXL15UYWGhCgsLdffuXT1//jzyOtYTMgEAAHrjMjP70ycBAAAAAIgfV+QAAAAAwGEIcgAAAADgMAQ5AAAAAHAYghwAAAAAOAxBDgAAAAAchiAHAAAAAA5DkAMAAAAAhyHIAQAAAIDDEOQAAAAAwGEIcgAAAADgMAQ5AAAAAHAYghwAAAAAOMy/WbRPbq531VoAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qixin/miniconda3/envs/BIA/lib/python3.11/site-packages/umap/umap_.py:1945: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtsAAAJOCAYAAABm28tYAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAACXBElEQVR4nOzdd3hUZcLG4d+kTXpCgBR6b1IEASlKUWmiYgVxpQiiq4AF21pQ0F2xIdb1c1WKrgIWEFYUAREUQQQERECkg5DQk0B6Muf74+RMMslMSCCHUJ77uuaCnPqemeTMM++8xWEYhoGIiIiIiJQ7v4ougIiIiIjI+UphW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYfsC99///peLL76Y8PBwHA4H48aNq+gi+VSnTh3q1KlTIeeeOnUqDoeDqVOnlnofh8NBt27dPJYNHToUh8PBrl27yrV8Z9L+/fu5/fbbqV69On5+fjgcjoou0nnD2++MnJ3GjRuHw+FgyZIlFV0UETnLXVBhe8mSJTgcDoYOHVrmbbp164bD4cDhcDB//nyf+7dp08a93c8//+xzu6effhqHw4HT6eTIkSM+tyt8XofDgZ+fH5UqVaJLly5MnToVwzB87nsyy5cvZ9CgQaSnpzNy5EieeeaZk77R16lTx6M83h7JycmnXCY5+w0dOpRPPvmE7t27M3bsWJ555pkSt7c+qLzwwgtnpHxncwiy/p7PF9u3b2fMmDFcfPHFREdHExgYSNWqVenatSv/+te/+Ouvvyq6iOedM/33JCKnL6CiC3CuCQgIYPLkyfTu3bvYuvXr17N27VoCAgLIzc31eQyXy8W0adNwOBxkZ2fz3//+l/vvv7/E8z700EOEh4eTl5fHjh07mDVrFj/++CNr1qzhzTffPKVr+frrrwH48MMP6dChQ6n38/f356mnnvK5Pjg4+JTKcyGYMGEC//jHP6hevXpFF+WUZGdn891339GzZ0/++9//VnRxzjubN28mNDS0ootRKq+++iqPPfYYeXl5dOjQgUGDBhEZGcnRo0f55ZdfGDt2LM8++yyrVq2iZcuWFV1cEZEKo7BdRn369GHOnDkcOXKEypUre6z74IMPCAgIoEePHnzzzTc+j7Fw4UL27NnDPffcw4cffsgHH3xw0rD98MMPEx8f7/55w4YNXHrppbz99tuMGTOGunXrlvla9u/fD+Bx3NIICAg4q5ubnM0SEhJISEio6GKcsqSkJFwuV5l/Z6R0mjRpUtFFKJV3332Xhx56iPr16/Ppp5/Spk2bYtts3ryZsWPHkpqaWgElFBE5e1xQzUjKwx133EF2djYff/yxx/Ls7Gw++eQT+vbtS2xsbInH+OCDDwC49957ueGGG9iwYQOrVq0qUzlatGhB165dMQyDNWvWuJcvX76cvn37EhMTQ3BwME2aNGHcuHGkp6e7t7GaykyZMgWAunXrupuAlDfra/OsrCyeeOIJatWqRUhICJdccgmLFi0C4Pjx49x3331Ur16d4OBgOnbsyOrVq30e89ixY4wYMYK4uDhCQkJo3749c+fO9bqtYRhMnjyZzp07ExkZSWhoKG3btmXy5Mletz969Ch///vfiYuLIzQ0lHbt2jF79uwSr/H999+nefPmBAcHU7NmTR599FEyMzO9buutzbb1eowbN45ff/2VXr16ERERQVRUFDfccIPP9t2zZs2ibdu2hISEEBcXx4gRIzh27FiZ27anp6czbtw4mjRpQnBwMDExMfTt25fly5d7bNetWzdq164N4P5m5nTa+VvlTEtLY8yYMVSvXh2n00nLli35/PPPi22fkpLC008/TbNmzQgPDycqKoomTZpwxx13sHfvXncZx48fD0D37t3dZSz8fFjnTU5O5r777qNmzZoEBAS42+OX9Pz5agZiGAbTpk2jS5cuREdHExoaSsOGDfn73//Onj17ALM99tKlS93/tx6Fm6z5arN95MgRHnzwQerWrYvT6SQ2NpYBAwawadOmYtsW/h3797//TdOmTQkODqZ27dqMHz8el8vl9dpK69ixYzz66KMEBwfzzTffeA3aAE2bNuXzzz8v9q2ZdY379u1j6NChxMfH4+fn59HsZ9q0aXTo0IHw8HDCw8Pp0KED06ZNK3aOkvpSFP678nb+vXv3MmDAACpXrkxYWBjdunUr9jtfWu+99x4XXXQRwcHB1KpVi8cff9zjHrBjxw78/Pzo27ev1/2PHTtGcHAwrVq1OqXz+7JmzRpGjRpF8+bNiYqKIiQkhBYtWvDCCy+Qk5NTbPuy/k0C7Nq1iwEDBhATE0N4eDhdu3blhx9+8Nqc61Rer9mzZzNw4EAaNGhAaGgoUVFRXH755XzxxRc+r/vdd991vx6F78m+/r6OHz/OM888w0UXXURISAjR0dH07t2bZcuW+TyHSFmoZruMOnbsSNOmTZkyZQr33Xefe/mXX37JkSNHGDZsGLNmzfK5/5EjR5gzZw6tW7emefPmDB48mP/+97988MEHtGvX7rTK9sUXX3DrrbcSFBTEgAEDiI2NZdGiRYwfP54FCxbw/fff43Q6qVOnDs888wxffvkl69ev5/777yc6Ovq0zn0yAwYMYMOGDVx33XVkZGTw8ccfc80117B8+XLuvvtuMjMzufnmmzl06BAzZ86kV69e7Ny5k8jISI/jZGdnc9VVV5GRkcGQIUNITk5mxowZXH/99Xz00Uf87W9/c29rGAa33347n3zyCY0aNeK2224jKCiIhQsXMnz4cDZt2sQrr7zi3j49PZ1u3bqxYcMGOnbsSNeuXd1vyD179vR6Xc899xxPP/20O+wGBgYyc+ZMNm/eXObnaPXq1bz88st069aNu+++m7Vr1/Lll1+yYcMGfv/9d4/mOZMnT2b48OFER0czePBgoqKi+Prrr+nRowc5OTkEBgaW6pxZWVlceeWV/Pzzz7Rp04YHHniAgwcPMnPmTBYsWMDMmTO58cYbATPEXXzxxbz++uu0atWK66+/HuC0OvTl5OTQs2dPjh49yo033kh6ejozZsygf//+zJ8/3/28G4ZBr169WLlyJZ07d6Z37974+fmxa9cuZs+ezZAhQ6hZs6Y7uC5dupQhQ4a4Q3PR3++srCyuuOIKjh8/zrXXXktQUBBxcXGndA2GYTBw4EBmzpxJ9erVGThwIJGRkezatYuZM2fSu3dvatWqxTPPPMPUqVPZvXu3Rzv3iy++uMTjHzlyhA4dOrBt2za6devGrbfeyq5du/j888+ZN28eCxcupGPHjsX2e+SRR1iyZAnXXHMNPXv25Msvv2TcuHFkZ2fzr3/965SuFeDzzz8nNTWVQYMG0bBhw5NuHxBQ/G3myJEjdOzYkZiYGAYMGEB2drb7b/3BBx/ktddeo3r16gwfPhyHw8EXX3zB0KFDWb9+Pa+++uopl91y7NgxOnfuTEJCAnfddRf79u1j5syZdO/enW+//bZMv9MTJ05kyZIlDBgwgGuuuYavv/6aF154gbVr1/LNN9/gcDioV68eV111FfPnz+evv/6iRo0aHsf46KOPyMrKYsSIEad9bYW99957/O9//6NLly5cffXVpKens2TJEh5//HFWrVrlNbCW9m8SYN++fXTq1InExESuvvpqWrVqxZYtW+jZsyfdu3cvl2t4/PHHCQoK4rLLLiMhIYFDhw4xd+5cbr75Zt544w1Gjx7tsf3TTz/Nc889535tAwIC+Oyzz/jjjz+8Hv/o0aN06dKFjRs3cvnll9OrVy9SUlKYM2cO3bt357PPPnPf60ROmXEB+f777w3AGDJkSJm36dq1qwEYiYmJxksvvWQAxq+//upe37NnTyMuLs7IyckxhgwZYgDGihUrih1/0qRJBmC8+uqrhmEYRl5enlGjRg0jMjLSSEtLK7Z94fMW9ttvvxkhISGGw+Ewdu7caaSmphrR0dGG0+k01q9f797O5XIZt912mwEYzz33nMcxrHLu3LnT5/NRVO3atQ1/f3/jmWee8fp45513vJa/c+fOxokTJ9zLZ8yYYQBGdHS0ccsttxg5OTnudS+++KLHc1T43IBxxRVXGNnZ2e7lmzdvNkJCQozo6GgjNTXVvfw///mPARjDhw/3OH5WVpZx7bXXGoCxevVq9/JnnnnGAIwRI0Z4nPfbb781AAMwpkyZ4l6+detWIyAgwKhevbpx4MAB9/KUlBSjcePGBmB07drV41jennPrdw4wZsyY4bH9oEGDDMCYPn26e9mxY8eM8PBwIyIiwti+fbt7eU5OjnHVVVcZgFG7dm2jNJ599lkDMP72t78ZLpfLvXz9+vWG0+k0KlWq5PGc7ty586R/Q0VNmTLFAIwJEyZ4LLdez379+hlZWVnu5YsWLTIAo1evXu5lv/32mwEYN9xwQ7HjZ2ZmGsePH3f/bL2O33//vdfyWOft2bOnkZ6e7nW9r+fP+n0u7O233zYA48orryx2vPT0dOPIkSMl7l+Yt9+ZYcOGGYDx+OOPeyyfP3++ARgNGzY08vLy3Mut37G6desa+/fvdy8/dOiQER0dbURERHg832V1xx13GIAxefLkU9rf+l2/4447jNzcXI91P/zwgwEYTZs2NZKTk93Lk5OTjSZNmhiA8eOPP7qXW79bhf8uLdbf1TPPPOP1/IMGDfL4nV+yZInhcDiMBg0aeDyfvli/Z8HBwcbvv//uXp6Tk2P06NHDAIwPP/zQvfyzzz4zAGP8+PHFjtWyZUsjODjYOHbs2EnP6+vvyZtdu3YVe45dLpf7d2rZsmUe68ryN2kYhnH77bcbgPHyyy97LWPRv8NTeb0K3+Msx48fN1q0aGFERUV5vG9u2bLF8Pf3N2rVqmUcPnzYY/uLLrrI69+X9f5Y9Pc5KSnJqFmzplG1alUjIyOjWBlEykJhu5TbFA69SUlJRkBAgDFq1CjDMAxjz549hp+fn/HII48YhmGUGLZbtGhh+Pv7e4Tnxx57zACMadOmFdveOu9DDz1kPPPMM8ZTTz1l3HbbbUZwcLABGPfdd59hGIbx4YcfGoBxzz33FDvGnj17jICAAKN+/foey081bFs3UW+PVq1aeS3/kiVLPJbn5uYagYGBBmDs3r27WHm9vQbWuX/66adi5Ro5cqQBGB999JF7WcuWLY2wsDCvN0orvD300EPuZXXr1jWCgoKKfbAxDMO48sori71JjB8/3gCMiRMnFtv+o48+KnPY7tKlS7HjWOvGjBnjXjZ16lQDMB588MFi269YsaJMYbtevXpGYGCgsXfv3mLr7r777mLPqR1he8eOHcX2qV27thETE+P+2Xq9brvttpOer7Rhu/CH0qLryxK2mzVrZvj7+xt//vnnSctW1rCdlZVlhISEGJUrV/b6YbxXr17FAqj1O+YtDFvrfvvtt5OW1Zc+ffoYgDF//vxi69asWVPsA/j//ve/YtcYFBRkHDp0qNj+VgicOXNmsXXTp093f3i2nGrY9vf3N/bs2VNsn759+xZ7Pn3x9eHcMAxj1apV7g9gluzsbCMuLs6oU6eOR8j/5ZdfDMC4/fbbT3pOwyhb2PZlzZo1BmCMGzfOY3lZ/iYzMzMNp9NpxMXFFfvw5nK53B+OTjds+zJx4sRi7y3jxo0zAOP1118vtr31+1P47+vQoUOGv7+/x+tU2BtvvGEAxX6HRcpKzUhOQVxcHFdffTWffPIJr7zyClOnTsXlcnHHHXeUuN8vv/zChg0b6N27t0cHsyFDhvDiiy8yefJkBg8e7HXfiRMnAmZ7w8jISNq1a8fw4cPd269duxbw/pV+zZo1qV+/Plu2bOH48eNEREScymW7OZ1On22SfWndurXHz/7+/sTGxpKWlkatWrU81lkdCPft21fsOIGBgV5HTrn88st5++23WbduHbfffjvp6els2LCBatWqeR0iy2qvaH21ePz4cXbu3EmzZs28dv67/PLL+e677zyWrV+/3r3O2/Zl5a3tq/V1c+HhFK3zdurUqdj27du39/q1vTepqans2LGDpk2bFvtaG8zfpXfffdf9nNohOjraa+feGjVqsGLFCvfPTZs2pUWLFnzyySfs3buX66+/nssvv5w2bdrg7+9f5vMGBwfTokWL0yo7QFpaGps2baJBgwalalJRVn/88QcZGRl069bN6ygl3bp149tvv2XdunVcdtllHutK+/tUVkYJw43++uuv7jbzlrvvvptrrrnGY1ndunWpUqVKsf1Luo9Zy9atW1e2AntRu3ZtatasWWz55Zdfzrx587w+n754+1u3+lIULmtgYCDDhg1jwoQJLFy40N0cw+rDc+edd57ClZQsOzubt956ixkzZvDHH39w4sQJj9fP6iRfWGn/Jrds2UJWVhZt27YlKCjIY1uHw0HHjh19Nt0oi4MHD/LCCy/wzTffsHv3bjIyMjzWF76Gku6N3patWrWKvLw8MjMzvfY92bp1K2D+HRb9HRYpiwsqbPv5mf1BS+ogZK2ztvXljjvuYO7cucyePZupU6fSoUMHmjZtWuI+Vqe8QYMGeSxv2rQpbdu2ZenSpWzbto0GDRoU2zcxMbHEESCsHv++2p3Gx8ezZcsWUlNTTztsn4qiba/BbMsZFRXldTngtQNP5cqVvb421nWnpKQAZptMwzDYt29fsTf/wtLS0jz289W51dvzWtI+p9L+t6TnIi8vz73Meq2rVq1abHs/Pz+vIcab0vzOQMF12sHbNYN53YX/TgMCAli8eDHjxo1j1qxZPPTQQwBUqVKF0aNH8+STT5YpdMfGxpZLh2ArtNo1lOPpvEal/X0qK6ss3j4M33nnne7QuGTJEp/tdn1dT2pqKn5+fl5/t+Pi4vDz8yuX38eT/Z2X5Ry+jhUbG1vsORoxYgQvvPAC77//Pj179iQ9PZ3p06fTqFEjunbtWupzltbNN9/M//73Pxo1auTuxxMYGEhycjKvv/46WVlZxfYp7d9kSfchOLV7YFFHjx6lXbt27Nmzh86dO3PVVVcRHR2Nv78/69atY86cOR7XUFKZvJXn6NGjAPz000/89NNPPsthvU+InKoLajQS6yZS0iQyhw8f9tjWl2uuuYa4uDgefvhhduzYwbBhw0rc3rqpAvztb38rNhGMNfqGr1EyTsYKswcOHPC63lruLfSeS44cOeL1w5J1fdbrZl3nJZdcgmE2l/L6+P777z22P3jwoNfzenterXN528fX61AerLIeOnSo2DqXy+X+HS7tcc6V35kqVarw1ltvsW/fPjZt2sRbb71F5cqVeeaZZ3jppZfKdKySgrafn5/PcfKLhjDrd8Bb8CwPZ+NrZNUQWn87p8LX8x8ZGYnL5fL6u33w4EFcLpfHtVofvL29XiUF5pP9nZ/s/l+aYx08eLDYcerWrUuPHj2YM2cOhw8f5tNPPyU1NdWWWu1Vq1bxv//9j169erFp0ybee+89/vWvfzFu3DhuvfXW0z5+Sfch8P47W9bX64MPPmDPnj3885//ZNmyZbz55ps899xzjBs3zus3nCWVyVt5rO0feuihEt8nTjZxl8jJXFBhu3HjxgQFBbFq1Sqfb6bW12Qnm4QhICCA22+/nX379hEaGsqAAQNK3N7qwX/xxRczfPhwr4/AwECmTZt2SrVOVjMNb7Pm7du3j+3bt1OvXr0KqdUuTzk5OV5n5vzxxx+BgpEdIiIiaNq0KZs3by7VV+aRkZHUrVuXbdu2kZSU5PP4hVnDdHlb521ZebHO622Ysl9++aXECZUKi4yMpF69emzbts1rWLSGqTvZaBlnmsPhoGnTpowcOZKFCxcCeAz9aNVwn2rtbaVKlTh48GCx5zEtLc39tbIlPDycZs2asXPnzmLrvClr2azhGFetWuUxfKelIl6jm2++mYiICD777LNSXXNZlHQf83atlSpVArx/2LGapHize/du93CRhRW9j5SGt7/11atXk5GR4fU4d911F9nZ2e45FgIDAxkyZEipz1da27dvB6Bv377FvvUpj/tT48aNcTqdrFmzhuzsbI91hmF4vU+X9fWyruG6664rtq6ke7K3e6O3Ze3atcPhcHg0jxGxwwUVtoODg+nfvz+HDh3in//8Z7H1GzZs4P333yciIoIbbrjhpMd75JFHmD17Nt9+++1Ja5asdnmTJk3i/fff9/q45ppr2L9/f4kT4vjSr18/oqKimDJlChs3bnQvNwyDxx9/nJycnBKnqT+XjB071qOJyR9//MHkyZOJioqiX79+7uX33Xcf6enpjBgxwuvXgDt37vQYw3rQoEFkZ2fz9NNPe2y3YMGCYu21AW677Tb8/f159dVXPWq3UlNTvf5+lZd+/foRHh7O+++/z86dO93Lc3NzGTt2bJmONWTIEHJycnj88cc92nL+/vvvTJkyhaioqLNi2KudO3d6HVPaqq0KCQlxL4uJiQE45anC27ZtS05OjsdY+tbfkbffo5EjR5KXl8e9995brD1pZmam+6vqUylbUFAQAwcO5PDhw0yYMMFj3aJFi/jmm29o0KABnTt3LvX1eVOnTp1i47/7EhMTw4QJE8jKyqJPnz4+Q+2ptAu3Quf48eM9JsNJTU11NwcrHEzbtGmDw+FgxowZHv1Itm7dyuuvv+7zPHl5eTz55JMev/NLly7l66+/pkGDBl7b9/ry0Ucfedxzc3NzeeKJJ4qV1dKvXz/i4+OZOHEiy5Yt47rrrjvp3AynwhoTv+hY0Rs3biz2u3QqnE4nN998M0lJSbzxxhse6z788EOvw5+W9fXydQ2ffPKJewbkwm699Vb8/Px49dVXPb7BTktL8zrcZXx8PP3792f58uW8/PLLXvsjrFy50usHXZGyuKDabIPZ0XDlypWMHz+er776iq5duxIcHMyff/7J3LlzMQyDjz/+uFTjTsfFxZUqiGzbto0ffviBevXqldgu74477mD27Nl88MEHZe6MERkZyXvvvcfAgQO59NJLGTBgAFWrVuW7775j9erVtG/fnkceeaRMx/QlNze3xIlMhg4dWqZJVcoiISGB5ORkLr74Yvr27UtKSgrTp08nMzOT9957z6Pm/u677+bnn39m2rRp/PTTT1x11VVUq1aNAwcO8Mcff7By5Uo++eQTd1kfffRRZs2axXvvvcfGjRvp0qULe/fu5dNPP6Vv377MmzfPoywNGjTg6aef5plnnqFly5b079+fgIAAvvjiC1q0aMGWLVtseQ6io6N59dVXueuuu2jTpg0DBgxwj7PtdDqpVq3aSfscWB599FHmzZvHRx99xObNm7nyyivdY53n5OTw4YcfnhXfhqxfv54bbriBdu3a0bx5c+Lj49m3bx9ffvkl/v7+7jbcUDCZzZNPPskff/xBVFQUUVFR3HPPPaU616hRo5gyZQp33nknCxcupGrVqvz4448kJyfTqlUrdycsyz333MPSpUv59NNPadiwIddddx2RkZHs2bOHb7/9lg8++MB9n7jiiiv4/PPPueWWW7j66qvdHTV9TXYC8OKLL7J06VL++c9/snz5ci699FL3ONuhoaFMmTKl1K+3L1bTrNJ2rh05ciRpaWk88cQTXHLJJXTo0IG2bdsSERHBkSNH2Lx5Mz/++CNOp7NM8wd06dKF0aNH8+abb9K8eXNuuukmDMNg1qxZ7N27l/vuu48uXbq4t69evToDBgxgxowZXHLJJfTu3ZuDBw8ye/Zsevfu7XPik5YtW7JkyRI6dOjAFVdcwf79+5kxYwaBgYG89957ZXo+r7rqKjp06MCtt95KTEwMX3/9Nb///ju9evXy2rE4ICCAYcOG8fzzzwOn3jGypLGjb7vtNq688krat2/Pp59+SmJiIh06dGDPnj3MnTuXvn37+pykpiwmTJjAokWLeOSRR/j++++5+OKL2bJlC1999RW9e/dm/vz5Hs9lWV+vQYMG8eKLLzJ69Gi+//57ateuzW+//caiRYu48cYbi81p0bhxY/7xj3/w/PPP06JFC2655RYCAgKYNWsWLVq04Pfffy/22v773/9my5YtPProo3z00Ud07NiRqKgo9u7dy5o1a9i6dSuJiYleOyiLlNoZG/fkLJKcnGw888wzRqtWrYywsDAjMDDQqFmzpnHbbbd5jJ1dmK/xrr0pOvTfP/7xDwMv41wXlZOTY8TFxRkBAQFGUlJSmc9rGOY4tX369DGio6ONoKAgo1GjRsbYsWM9xrguWs7yHPqPIkM9lTTUWUlDrOFl2Dxr+yNHjhh33nmnERsbazidTqNt27bGnDlzfJZ55syZxlVXXWVUqlTJCAwMNKpXr25069bNmDhxYrHhx44cOWLcddddRtWqVY3g4GDjkksuMWbNmlXikFXvvfee0axZMyMoKMioUaOG8fDDDxvp6ellHvrP25BXJQ2199lnnxmtW7c2nE6nERsba9x5553GkSNHjPDw8GJDMJbkxIkTxtixY41GjRoZQUFBRnR0tNGnTx+vw5+V99B/pR1ib+/evcY//vEPo0OHDkZsbKwRFBRk1KpVy7j55puNlStXFtt/6tSpRosWLQyn01lsKMSSzmv57rvvjEsvvdRwOp1G5cqVjUGDBhlJSUk+f59dLpfx/vvvGx06dDDCwsKM0NBQo2HDhsbf//53jyHmcnJyjEcffdSoVauWERAQUOy59PY7YxjmEGX33XefUbt2bSMwMNCoUqWKcfPNNxsbNmwotm1Jf9fehkU8duyY4efnZ3Tu3LnE58SbLVu2GPfdd5/RokULIzIy0ggICDCqVKliXH755cazzz7rdXg9X9dY2OTJk4127doZoaGhRmhoqNGuXTuf43qnpaUZo0ePNuLi4gyn02m0bNnS+Pjjj0sc+q9r167G7t27jVtuucWoVKmSERISYnTp0qXYuNMlKfxcvvvuu0azZs0Mp9Np1KhRw/jHP/7hdQx3y5YtWwzAqFWrVqnG9C6s8BjWvh6TJk0yDMMwDh48aAwbNsyoVq2aERwcbLRo0cJ4++23jR07dvgcXrUsw14ahmHs2LHDuOWWW4yoqCgjNDTUuPzyy42lS5cao0aNMgBj7dq1HtuX9fVat26d0bNnT6NSpUpGRESE0bVrV2PRokUl3pP//e9/G02bNvW4J+/du9cgfwzxotLT042XXnrJuOSSS4ywsDAjJCTEqFu3rnH99dcbH374occ8DSKnwmEYJYzjJCLnlG3bttGwYUP69+/PzJkzK7o4cg746quvuPbaa5k3bx5XX311RRfHdg6Hg65du3ptF36mfPrppwwYMIDx48cXa7Z2vrjssstYsWIFKSkphIeHV3RxWLRoET169ODRRx/lxRdfrOjiyAXmgmqzLXK+OHbsWLFhuzIyMnjwwQcBzop21nJu+PHHH2nVqtUFEbTPBoZh8OqrrxIQEMDw4cMrujinLTExsdiyjz/+2N1070wH7UOHDhXrgJycnMzjjz8O6N4oFeOCa7Mtcj5YunQpw4cPp2fPntSqVYvDhw+zePFidu3axRVXXHHS0XFELC+++KJq+s6ADRs28NVXX7F8+XJWrlzJ3//+d9vGZz+TmjdvTuvWrWnWrJl7/OslS5YQERHBK6+8csbL8/HHH/PKK69wxRVXUK1aNRITE5k/fz4HDx5k6NChdOzY8YyXSURhW+QcdNFFF9GjRw9++uknvvzyS8DssPncc8/x8MMPn3aHOREpX2vWrOGJJ54gOjqawYMH8/LLL1d0kcrF3//+d/73v/+xevVq0tLSqFq1Krfddhtjx46lSZMmZ7w8nTp14pJLLmHRokUcPXoUf39/mjZtytixY7n33nvPeHlEANRmW0RERETEJqr+EhERERGxicK2iIiIiIhN1GbbB5fLxf79+4mIiMDhcFR0cUREROQ0GYbB8ePHyzT5l8jpUtj2Yf/+/dSsWbOiiyEiIiLlbO/evdSoUaOiiyEXCIVtH6wpqvfu3UtkZGQFl0ZEREROV2pqKjVr1nS/x4ucCQrbPlhNRyIjIxW2RUREziNqHipnkhosiYiIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxSYWH7XfeeYeWLVsSGRlJZGQkHTt25JtvvnGvNwyDcePGUa1aNUJCQujWrRsbN2486XG/+OILmjVrhtPppFmzZsyePdvOyxARERERKabCw3aNGjV44YUXWL16NatXr+aKK66gX79+7kD90ksv8eqrr/LWW2+xatUq4uPj6dGjB8ePH/d5zBUrVjBgwAAGDRrE+vXrGTRoEP3792flypVn6rJERERERHAYhmFUdCGKiomJ4eWXX2bYsGFUq1aNBx54gMceewyArKws4uLiePHFF7n77ru97j9gwABSU1M9ash79+5NpUqVmD59eqnKkJqaSlRUFCkpKURGRp7+RYmIiEiF0nu7VIQKr9kuLC8vjxkzZpCWlkbHjh3ZuXMnSUlJ9OzZ072N0+mka9euLF++3OdxVqxY4bEPQK9evUrcR0RERESkvAVUdAEANmzYQMeOHcnMzCQ8PJzZs2fTrFkzdziOi4vz2D4uLo7du3f7PF5SUpLXfZKSknzuk5WVRVZWlvvn1NTUU7kUERERERG3s6Jmu3Hjxqxbt46ff/6Ze+65hyFDhrBp0yb3eofD4bG9YRjFlhVV1n0mTJhAVFSU+1GzZs1TuBIRERERkQJnRdgOCgqiQYMGtG3blgkTJtCqVStef/114uPjAYrVSB88eLBYzXVh8fHxZd7n8ccfJyUlxf3Yu3fvaVyRiIiIiMhZEraLMgyDrKws6tatS3x8PAsXLnSvy87OZunSpXTq1Mnn/h07dvTYB2DBggUl7uN0Ot3DD1oPEREREZHTUeFttp944gn69OlDzZo1OX78ODNmzGDJkiXMnz8fh8PBAw88wPPPP0/Dhg1p2LAhzz//PKGhodx2223uYwwePJjq1aszYcIEAO6//366dOnCiy++SL9+/ZgzZw6LFi1i2bJlFXWZIiIiInIBqvCwfeDAAQYNGkRiYiJRUVG0bNmS+fPn06NHDwAeffRRMjIyuPfeezl27BiXXnopCxYsICIiwn2MPXv24OdXUEnfqVMnZsyYwVNPPcXYsWOpX78+M2fO5NJLLz3j1yciIiIiF66zcpzts4HG4hQRETm/6L1dKsJZ2WZbREREROR8oLAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbFLhYXvChAm0a9eOiIgIYmNjuf7669myZYvHNg6Hw+vj5Zdf9nncqVOnet0nMzPT7ksSEREREQHOgrC9dOlSRo4cyc8//8zChQvJzc2lZ8+epKWlubdJTEz0eEyePBmHw8FNN91U4rEjIyOL7RscHGz3JYmIiIiIABBQ0QWYP3++x89TpkwhNjaWNWvW0KVLFwDi4+M9tpkzZw7du3enXr16JR7b4XAU21dERERE5Eyp8JrtolJSUgCIiYnxuv7AgQPMmzeP4cOHn/RYJ06coHbt2tSoUYNrrrmGtWvXlmtZRURERERKclaFbcMwGDNmDJdddhnNmzf3us20adOIiIjgxhtvLPFYTZo0YerUqcydO5fp06cTHBxM586d2bp1q9fts7KySE1N9XiIiIiIiJwOh2EYRkUXwjJy5EjmzZvHsmXLqFGjhtdtmjRpQo8ePXjzzTfLdGyXy0WbNm3o0qULb7zxRrH148aNY/z48cWWp6SkEBkZWaZziYiIyNknNTWVqKgovbfLGXXW1GyPHj2auXPn8v333/sM2j/++CNbtmzhzjvvLPPx/fz8aNeunc+a7ccff5yUlBT3Y+/evWU+h4iIiIhIYRXeQdIwDEaPHs3s2bNZsmQJdevW9bntBx98wCWXXEKrVq1O6Tzr1q2jRYsWXtc7nU6cTmeZjysiIiIi4kuFh+2RI0fyySefMGfOHCIiIkhKSgIgKiqKkJAQ93apqal89tlnTJw40etxBg8eTPXq1ZkwYQIA48ePp0OHDjRs2JDU1FTeeOMN1q1bx9tvv23/RYmIiIiIcBaE7XfeeQeAbt26eSyfMmUKQ4cOdf88Y8YMDMNg4MCBXo+zZ88e/PwKWsUkJydz1113kZSURFRUFK1bt+aHH36gffv25X4NIiIiIiLenFUdJM8m6kQhIiJyftF7u1SEs6aDpIiIiIjI+UZhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYVHrYnTJhAu3btiIiIIDY2luuvv54tW7Z4bDN06FAcDofHo0OHDic99hdffEGzZs1wOp00a9aM2bNn23UZIiIiIiLFVHjYXrp0KSNHjuTnn39m4cKF5Obm0rNnT9LS0jy26927N4mJie7H119/XeJxV6xYwYABAxg0aBDr169n0KBB9O/fn5UrV9p5OSIiIiIibg7DMIyKLkRhhw4dIjY2lqVLl9KlSxfArNlOTk7myy+/LPVxBgwYQGpqKt988417We/evalUqRLTp08/6f6pqalERUWRkpJCZGRkma9DREREzi56b5eKUOE120WlpKQAEBMT47F8yZIlxMbG0qhRI0aMGMHBgwdLPM6KFSvo2bOnx7JevXqxfPlyr9tnZWWRmprq8RAREREROR1nVdg2DIMxY8Zw2WWX0bx5c/fyPn368PHHH7N48WImTpzIqlWruOKKK8jKyvJ5rKSkJOLi4jyWxcXFkZSU5HX7CRMmEBUV5X7UrFmzfC5KRERERC5YARVdgMJGjRrFb7/9xrJlyzyWDxgwwP3/5s2b07ZtW2rXrs28efO48cYbfR7P4XB4/GwYRrFllscff5wxY8a4f05NTVXgFhEREZHTctaE7dGjRzN37lx++OEHatSoUeK2CQkJ1K5dm61bt/rcJj4+vlgt9sGDB4vVdlucTidOp7PsBRcRERER8aHCm5EYhsGoUaOYNWsWixcvpm7duifd58iRI+zdu5eEhASf23Ts2JGFCxd6LFuwYAGdOnU67TKLiIiIiJRGhddsjxw5kk8++YQ5c+YQERHhro2OiooiJCSEEydOMG7cOG666SYSEhLYtWsXTzzxBFWqVOGGG25wH2fw4MFUr16dCRMmAHD//ffTpUsXXnzxRfr168ecOXNYtGhRsSYqIiIiIiJ2qfCa7XfeeYeUlBS6detGQkKC+zFz5kwA/P392bBhA/369aNRo0YMGTKERo0asWLFCiIiItzH2bNnD4mJie6fO3XqxIwZM5gyZQotW7Zk6tSpzJw5k0svvfSMX6OIiIiIXJjOunG2zxYai1NEROT8ovd2qQgVXrMtIiIiInK+UtgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYPi+58h8iIiIiUpECKroAUl5cwDHgEJCevywEqApUAvwrqFwiIiIiFy6F7fNCHrADM2xDwct6PP8RBdRHL7eIiIjImaX0dV7YCxwFnHjWYAdiBvFkYDdm4C5vuZghPzn//0GYNenRqJWSiIiIXOgUtiuIYRgc27CDo+t34MrJJbRaZeK7tCQgNLiMR8rCDNoBeG8q4o8ZupOBDMymJeXlOLATyMz/2YHZnOUIEA7UA8p6PSIiIiLnD4XtCpCy9S82vDCdI2u3kZeRBQ4HOCCsRlUaDb+aOrd0xeFwlPZomDXKJYXaAMygnUz5he0MYDuQjVmjXrgW24UZxLcDjdGvmYiIiFyolILOsOM79vPz6Dc5sSuJkPgYQqtXxuFwkJedQ0bSMdb/62NyM7JoOKSXjyPkUdARMgsz7GZj1l77Y9YuF2Utyy3HKzmYf/4QL+f0wwz/aZi17rHleF4RERGRc4ca1Z5hm9+Zy4mdSUTWr0ZQZKi7Bts/KJDwWrH4Bfqz5d3/kbbvkJe9s4AtmJ0hj2MGb+txPH+94WU/a1l5fbbKpaDpiq8aeOtX63A5nVNERETk3KOwfQad2HuQAz9sILhKJA5/7099SHwlso4eZ983vxRZ46IgZDsxa5SD8v+1jpUJ5Hg5ah5mrXfU6V8E5J/DOmZJ/PH9AUBERETk/KewfQal/vkXOalpBFUK97mNw88PvwB/jv2+q+jewAmKt4+2OkCCGcgz8Qy3LsxmJlGUX3ttP8wa7ZOFaCN/u9K2PxcREZGihg4dyvXXX1/RxThrnGvPh8L2GWS4XPnx9OTh0zCKBtljmMHZW21yCGbgdmA28cjK/zcj//+RQO1Snbd0rBp1b7XoFgOz9ru8atNFRETOXg6Ho8TH0KFDz3iZvv/+e66++moqV65MaGgozZo146GHHmLfvn2lPka3bt144IEH7CvkKXj99deZOnVqRRej1BS2z6DI+tUIjAghO+WEz20Ml4ErN4+oxjWLrMnG98vlB4RR0FkxG7MWPDt/fShm8C0vDsyZKcF3p8sczA8GVcrxvCIiImenxMRE9+O1114jMjLSY9nrr79+Rsvz7rvvctVVVxEfH88XX3zBpk2b+L//+z9SUlKYOHHiGS1LecnLy8PlchEVFUV0dHRFF6fUFLbPoIh61Yjt0IzMgykYLu9NMDIPJxMUFU6NPu2LrPHHrNn2xYH5clrHdWKOCOIHJAGbMTs1FpWHOS72DmAbsAezXfjJmohUxgzSOZhNV/IoqM3OyC9rNczxtkVERM5v8fHx7kdUVBQOh8P9c2BgIH//+9+pUaMGoaGhtGjRgunTp3vs//nnn9OiRQtCQkKoXLkyV111FWlpaV7PtWbNGmJjY/nXv/7ldf1ff/3Ffffdx3333cfkyZPp1q0bderUoUuXLrz//vs8/fTTABw5coSBAwf6LNfQoUNZunQpr7/+uruGfteuXQBs2rSJq6++mvDwcOLi4hg0aBCHDxcMinD8+HH+9re/ERYWRkJCApMmTSpWS37s2DEGDx5MpUqVCA0NpU+fPmzdutW9furUqURHR/PVV1/RrFkznE4nu3fvLtaMxDAMXnrpJerVq0dISAitWrXi888/9zjP3/72N6pWrUpISAgNGzZkypQpJb+g5Uhh+wxr8vdrCa1WmeM79pObkeVe7srLIz3xCLknMqk/6Coi6iYU2TOagkljvHEB6fnbhGPWcjsp6EyZhzkBTeFa9ePARszxsA9hhvFE4A/M4F3SUIF+QB2gVv45rNCdg1nLXhdIQO21RUTkQpeZmckll1zCV199xe+//85dd93FoEGDWLlyJWDWig8cOJBhw4axefNmlixZwo033uilSSksWbKEK6+8kvHjx/Pkk096Pd9nn31GdnY2jz76qNf1Vq3wycr1+uuv07FjR0aMGOGuoa9ZsyaJiYl07dqViy++mNWrVzN//nwOHDhA//793ecYM2YMP/30E3PnzmXhwoX8+OOP/Prrrx7lGDp0KKtXr2bu3LmsWLECwzC4+uqryckpaKaanp7OhAkTeP/999m4cSOxscWHE37qqaeYMmUK77zzDhs3buTBBx/k9ttvZ+nSpQCMHTuWTZs28c0337B582beeecdqlQ5c9+8a5ztMyy6WR3aT7qX9f/8L6l//kVeTm7+8H8GzspRNB5xDQ2H96FgZsgUzKAchFm7nYHZLKRoiM3EDNyFRyexODADcQbm+NjhmMF8e/55ggvtY9VOH83/fwMvx7P4AfGY42in5e8XgBm2FbJFREQAqlevzsMPP+z+efTo0cyfP5/PPvuMSy+9lMTERHJzc7nxxhupXbs2AC1atCh2nDlz5jBo0CDeffddBg4c6PN8W7duJTIykoSEohV3ZStXVFQUQUFBhIaGEh8f797unXfeoU2bNjz//PPuZZMnT6ZmzZr8+eefJCQkMG3aND755BOuvPJKAKZMmUK1atU8yjh37lx++uknOnXqBMDHH39MzZo1+fLLL7nlllsAyMnJ4d///jetWrXyeg1paWm8+uqrLF68mI4dOwJQr149li1bxrvvvkvXrl3Zs2cPrVu3pm3btgDUqVOnxOelvClsV4DKFzeg24yxHFy+kaO/7cDIzSMkIYZqV7UhuHIUZrOOPXi20z5BQRBOp2ASGwOzNjkr/2dfM0k6MF/uZMwa6yS8T0pjbefI3zYFqHSSK/IDIkp17SIiIheavLw8XnjhBWbOnMm+ffvIysoiKyuLsLAwAFq1asWVV15JixYt6NWrFz179uTmm2+mUqWC99+VK1fy1Vdf8dlnn3HDDTeUeD7DMEo1E/XJyuXLmjVr+P777wkPL95UdPv27WRkZJCTk0P79gVNYqOiomjcuLH7582bNxMQEMCll17qXla5cmUaN27M5s2b3cuCgoJo2bKlz7Js2rSJzMxMevTo4bE8Ozub1q1bA3DPPfdw00038euvv9KzZ0+uv/56d8A/ExS2K4hfgD/xXVoS36XoL1AqsIuCWurCfyx5mDXYVhi2mnmEYYZuXzNIus9KQZvqZEqelMYa9eQIJw/bIiIi4svEiROZNGkSr732Gi1atCAsLIwHHniA7GxzIAN/f38WLlzI8uXLWbBgAW+++SZPPvkkK1eupG7dugDUr1+fypUrM3nyZPr27UtQUJDP8zVq1IiUlBQSExNLrN0+Wbl8cblcXHvttbz44ovF1iUkJLjbXRcN/IWbxXhrImMtL7xfSEhIiR8cXC6zee28efOoXr26xzqn0wlAnz592L17N/PmzWPRokVceeWVjBw5kldeeaWkyyw3arN9VjEw20znYjb7KPrLZdVcuzDbSl8ENAea4rtGu+jxrZBuNfkoiR9mMBcREZFT9eOPP9KvXz9uv/12WrVqRb169Tw6AoIZTDt37sz48eNZu3YtQUFBzJ49272+SpUqLF68mO3btzNgwACPds1F3XzzzQQFBfHSSy95XZ+cnFzqcgUFBZGX5zmiWZs2bdi4cSN16tShQYMGHo+wsDDq169PYGAgv/xSMEFfamqqx7GbNWtGbm6uu304mB02//zzT5o2berz2oqyOk7u2bOnWFlq1iwY2a1q1aoMHTqU//73v7z22mv85z//KfU5TleZwva+fftYuXIlx44d87r+8OHDfPjhh+VSsAtTJmanxSBKrnE2MJt3hFDQ3roSBSOC+JKD2dzD+jRc2klpRERE5FQ1aNDAXXO9efNm7r77bpKSktzrV65cyfPPP8/q1avZs2cPs2bN4tChQ8VCZ2xsLIsXL+aPP/5g4MCB5OZ6H8igZs2aTJo0iddff53hw4ezdOlSdu/ezU8//cTdd9/Nc889V6pygdm+eeXKlezatYvDhw/jcrkYOXIkR48eZeDAgfzyyy/s2LGDBQsWMGzYMPLy8oiIiGDIkCE88sgjfP/992zcuJFhw4bh5+fnrqVu2LAh/fr1Y8SIESxbtoz169dz++23U716dfr161fq5zYiIoKHH36YBx98kGnTprF9+3bWrl3L22+/zbRp0wB4+umnmTNnDtu2bWPjxo189dVXZQr0p6tUYdvlcnHnnXdSq1YtOnXqRHx8PI888kixT1Xbt2/njjvusKWg56OctAz2/G85G1//go2vf8GBn9ZhuPIo+WXJy38cAH4DtgCHgRjMdty+pke32n/HYgb0IEoebcTArEGPLNtFiYiIiIexY8fSpk0bevXqRbdu3YiPj/cYui4yMpIffviBq6++mkaNGvHUU08xceJE+vTpU+xY8fHxLF68mA0bNvC3v/2tWK2z5d5772XBggXs27ePG264gSZNmnDnnXcSGRnp7hR5snIBPPzww/j7+9OsWTOqVq3Knj17qFatGj/99BN5eXn06tWL5s2bc//99xMVFYWfn5lhXn31VTp27Mg111zDVVddRefOnWnatCnBwQXfxE+ZMoVLLrmEa665ho4dO2IYBl9//TWBgYGUxXPPPcfTTz/NhAkTaNq0Kb169eJ///ufuwlOUFAQjz/+OC1btqRLly74+/szY8aMMp3jdDgMX41mCnnvvfe45557uOeee2jfvj3fffcd//3vf+nUqRPz5s0jIsLsHLdy5Uo6derk84U/l6SmphIVFUVKSgqRkeUfOPd+tYJNb8wmbd8hdzaOalqJ9i93IyAyipAqMV72yqJgTOvCTUrAbLcdA+zHDNH++f/mYJ7AD3MimnqYoXw/sBezuYq3WSmt9lpNMUc/ERERObfZ/d4uvqWlpVG9enUmTpzI8OHDK7o4Z1SpOki+++67PPDAA+6G5IMGDeK2225jwIAB9OjRg2+//ZaoKE3LXVp75/3M2nHTcOXkElYzFv8g82VwZeeRtv84wTm54PIjJDa60F45mO2nraYd1hjaYAZua/zsBpgh2hq6D8ww7Y/Z9GQLUB+zhjs1f1kABdO951EwDXsNFLRFRESkrNauXcsff/xB+/btSUlJ4dlnnwUoUxOR80WpmpFs3bq12FcZPXv25LvvvmPbtm1cddVVPttxi6fcjCw2vTmLvOwcwmvHuYM2gMPhz9G1yfgH+pOeeBAjz6q1NjBrta2f/Shod239HIw51nUKBWNnR2JOhhOFOba2M3+b7fn7NcCceMYPs8Y8HTNoB2NOWFMwpqaIiIhIWbzyyiu0atXKPRvmjz/+eEYnkzlblKpmOygoyGuv1zZt2rBw4UJ69OjBlVdeyT//+c9yL+D5JmnpetL3Hia0pvdftqQlSYTVDCOmZSVy0pIJcn/NZT3/DrxPXGP9fADvwwZa2wRjhupjmM1KamEG7uP5+wVhBnMNVCMiIiKnpnXr1qxZs6aii3FWKFWiatiwIStWrPC6rnXr1ixcuJA9e/YwZMiQMhdgwoQJtGvXjoiICGJjY7n++uvZsmWLe31OTg6PPfaYe/zHatWqMXjwYPbv31/icadOnYrD4Sj2yMzMLHMZy9Px7fsxMPAP8t7435WTx/aPNrN16nqyU1IxQ/BxzNrtQMy22b7H1jRrqK0mId5YL/nRQssCMdt7V8GsDVfQFhEpPRfm3AVJmBUe1j1bRKSUqeqKK65g2rRpPjs+tm7dmkWLFvkcoLwkS5cuZeTIkfz8888sXLiQ3NxcevbsSVpaGgDp6en8+uuvjB07ll9//ZVZs2bx559/ct1115302JGRkSQmJno8CveCrQgOP78S7sEGkIYr5wS7Pt9E4qJ9mDXUAfnrrIlrfLGamZS0DZgve8kD1ouISGkcAzYCWzFn/t0F/AFspqAvjYhcyErVjGTUqFF07NiREydO+OwIefHFF7N8+XKfNeC+zJ8/3+PnKVOmEBsby5o1a+jSpQtRUVEsXLjQY5s333yT9u3bs2fPHmrVquXz2A6Hg/j4s6vdcWSjGjj8HORlZuMfXLSGOgPIJTcjD0eAPxH14zGbfTgxOzPm5m9TdBpVa8r2XMwabavNtrfabSN/OxdmZ0l/zHbd0WhCURGRsjgK7MTsWB5EwTwILsza7a1AQ8ymeSJyoSpVuoqPj6dv374n3a5Ro0Y0atTotAqUkpICQEyMt6HvCrZxOBxER0eXeKwTJ05Qu3Zt8vLyuPjii3nuuedo3bq1122zsrLIyspy/5yamlr2wp+EYRgEhAXjFxTIsY27CKsVS3DlSLO2O38UEMOA9P2pRDWqSmzHOvl7WqOPpGPWSAdTUHttjVJijZltzfqYk79P4QlyXJgdJHMwm46k5i87mn/MupiT3oiISMnyMEd+ysOzcsOBeX8OwbwX7wWaoAnCRC5cZ1VVpmEYjBkzhssuu4zmzZt73SYzM5N//OMf3HbbbSWOkdmkSROmTp1KixYtSE1N5fXXX6dz586sX7+ehg0bFtt+woQJjB8/vtyupahjv+9kw0szOLphJ9lHj5N97ARZR1IJjAglon4CIbGh5KZnkZ54gqCoUJo/fAV+gYWbgzgxA3U25g08GPMmn07BONphmOE5PX9dRqF9zSYqZtC2trVaEbkw23pvBxpjvkmIiIhvxzC/RXRSEKQLz+Lrj1nZcSL/oYoMkQtVqSa1OVNGjhzJvHnzWLZsGTVq1Ci2Picnh1tuuYU9e/awZMmSMg1I73K5aNOmDV26dOGNN94ott5bzXbNmjXLZeD7Yxt38fPoN0hPPEJofGUCwoPJOnqc1K37yDmeDkBwlXCclYKIbBzPRQ92I65TXS9HMjBv2lbnykwKOk4GU9C2OwMzlLsoCNY5+dtbPxftoGntF4c57J+IiPi2B0jEnIsgFzN4Fx61yxqi1YV5T407w+UTbzSpjVSEs6Zme/To0cydO5cffvjBZ9Du378/O3fuZPHixWX+I/Hz86Ndu3Zs3brV63qn04nT6fS67nQYhsHGSZ+Rvv8IkQ2q4/Aza0CCK0firBRBdvIJTuw5SGB4AG1f7kW1K5vjF1DSy+LPib0h5GVkE17LgV9QKA6/wm2/raEBAzDDtVUbbs06GYb3l92Rv/wY5mQ2Z82vhojIWSybgm8YHXg228vI/9mq07ImDXPg2cRPRM5nFT7Gm2EYjBo1ilmzZrF48WL3PPaFWUF769atLFq0iMqVK5/SedatW0dCQkJ5FLvUkjfu4sjabYTEV3IHbYvDz4EzJoJKzesA/uQez8UvwPcXDQdXbGPF6P+x+Ka32Dr1G9L3H+HIrztJ338Ew1V4P+tGHpH/b1XM2hcrhPvij/lmoJFKRERKZo1sZc3s65f/sAK39X8XZv+YPcBvwO/5jz+Aw2iIwPOP4XKR8udeDq36g5Q/92K4XCff6TQMHTrUPbxxQEAAtWrV4p577tFkg2eRCq++HDlyJJ988glz5swhIiKCpKQkAKKioggJCSE3N5ebb76ZX3/9la+++oq8vDz3NjExMQQFmbW6gwcPpnr16kyYMAGA8ePH06FDBxo2bEhqaipvvPEG69at4+233z6j15fy51/kpWcRWs33BwS/AH8MF6RsOU5Bm2rP4ft2z1nPb/9aQM7xPJxVKhMUGYbD34+ctExS/txLzvF0IhvWKBLorRu+1Uve+9CNBaybvmpbRERKVomCkUj8KbhvGl7+fxDz7TaQgnv78UKP2pRU95W6fT975/3M4ZWbceXkEdmoOjWu7kDVS5vg53+yoV7lTDqydivbPlxA8qbd7lHHopvVpsHgnlRuXby/WHnp3bs3U6ZMITc3l02bNjFs2DCSk5OZPn26beeU0it12M7IyODLL79k9+7dVK1aleuuu46qVauedgHeeecdALp16+axfMqUKQwdOpS//vqLuXPnAubwgoV9//337v327NmDn1/BzSo5OZm77rqLpKQkoqKiaN26NT/88APt27c/7TKXSRk+0RpGFOYQfCn5S8yX5/iOg2x4cRGuHIhoUAuHw4/MQ1kYLnBGh5KTlk164lECI0MJTSgc6l0UNCtxYU64UPiNoKjc/G3LvznN2ceF+TyfwHxOnJhvniVNGCQiYgnMf2RSMMeBNR+CxXpPysP8drFwX5kAzHvuofx1xdt0G4bBtg8X8Mc7c8lOPoF/cBAOPwfHft/B3v/9TMKVrWnz7B0EhqtT+9ngyNqtrHv2IzIPpxBaoyoBYU5y07I4vPpPTuw6wMVPD7ItcDudTvdQxzVq1GDAgAFMnTrVlnNJ2ZUqbO/fv58uXbqwc+dO98Q1Dz/8MN988w0dOnQ4rQKcrH9mnTp1SjVZzpIlSzx+njRpEpMmTTqdopWL8LoJ+AUHkZuW6fOG6MpzgcNBRN3qQAPMofgOYd7EYe+8HWQdzSWyYU0cDvPmnbotlbS9JwivG0FeVh6u7BzS9x8mJD4Mh8Oq0c6lIEQG5x8zB++B0upFX5WzoHWRzVKB3RR8/WvZB8RjTl+v2n0ROZkgzLfRHB/rrfBtNSkpygrcB/F2790zdzkbJ32OX2AAkQ2r59/bTTnH0/nr61/wCwyg7QsjPNbJmWe4XGz7cAGZh1OIalrL/XoERYYS2LQWKZv3sO2jhcS0qp8/3K99duzYwfz58wkM9D5TtZx5pXrFn3rqKfbt28dTTz3FvHnzeO211wgKCuKee+6xu3znvMqtG1DpotpkJB31+aEh88BRgitHUr1XO8yvGKsCTYGWQEsSFycSEBLqDtoAGLDnyz3kHs8hJCGYgIhActPSyUtPxRzi7zhm2+vK+ccMB6pR0GnHCtcuzF70WZhTtp/+txVnt+OYQxxmYH4QCaWgPbsL+AvYX2GlE5FzSQDmvdRq+udX6P9Fg46vMGzVjqd5LHXl5LJt6nwMl4vQhJhiYTowIpTg2CgSv/uVlM27T/tKysaasyGVgs6hF7bUbftI3rSb0BpVi71WDoeD0BpVSd64i9Rt+2w5/1dffUV4eDghISHUr1+fTZs28dhjj9lyLim7UtVsL1y4kCeeeIKxY8cC0KdPH+rXr891113HgQMHiIvTkEa+OPz8aDb6RlaO+TfHdyQSVqMK/k6zZtmVm0dG0jGMvDya3XcNwVUKz85pjQ4CuelZOAKLt8tL3ZbKnx9sovYNtQipFo6jkhNHQOHOOX6YtdnRmGE7AbMmJgkzbFrNTJyYQTyO87tW28Csvc7GDNdF27c789clYX5ICS56ABGRQvww7yuF22wX5q1Jia9jePapObRqC8e3JxISV8nn2YOiwkg9mMy+hWuIblanDOUG8153jILJzZyY971wfH8wcGG+pxyiYB4HR/4+cZjfol6YslPSyMvMJiDMezPMgFAnGfuzyU5J87r+dHXv3p133nmH9PR03n//ff78809Gjx5ty7mk7EoVtpOSkujSpYvHsm7dumEYhsJ2KVS9tCntXryL316czomdSRh5Lve9LDi2Eo3vvJp6f7vK5/6h1auQvu+wlzUGqVuT+P2VRIJjwwirHcVF93XB3xlKwUtbdAazKpg31BOYX1/6Yd4oT62TTU5aBgeX/U560lH8gwKp3KahOSX9WfmVZjrmdZc05FYg5nN2FPMDiIiIL9bcBdacBoXvK1b/GCtwu/B+n7UqPTzfjjMPJePKzSUgxHcfGofDgcPPQUbikTKW+yhmU7psCkZPsYJ0JcxxwYvGAyN/n0P52wcW2s/q/1KTC3U88aCoMPyDg8hNyyIoMrTY+tz0LPyDgwiKCrPl/GFhYTRo0ACAN954g+7duzN+/Hiee+45W84nZVOqsJ2Xl0dIiGd74+Bgs9YvNzfX2y5SRNxlLbiiXWOSfviN5E27MfJchNeKJeGqSwgIdXLgxw2k7z+MX4A/lVrVJ6phwVjjNa/pyKGfN5GXnYN/kDWqiNUExIUrz0HSD7updX1LAiOiipzZ2wxmDk53NjPDMNg543u2Tp1P2l+HzIUuczr6qh2a0uKxgYTXjD2tc5S/TMznrqROkNYbT0YJ24iIgHkPDsZss51HQXCG0oftbMz5DzxDmL8zEMMAI8+Fw9/3N46GyyAgtCzfwqVgjqLiwvMbPqt23Qru9fH88HAEM2gH4hkd/PJ/zsJshhde7FouBJENqhPdrDaHV/9JYKE222C+X6b/dYgq7RoT2aD6GSnPM888Q58+fbjnnnuoVk0VRxWt1KORbNmyhYBCk63k5Zlfef3xxx/Ftm3Tpk05FO384+8MonqPtlTv0da9bO+8n9nyn684viMRI8+FYRgEhodQtUNTmj/Un4g68VTvcQk7Zizk2G/bCK8TmR+4AQwMl4sTu1NxVg6j/m2XeDsr5htBOuU5XfCf73/N5jdn4/D3I6xmLP5BARiGQU5KGvsXriFt7yE6/vt+wqqfq23Az8aaeRE5uwRi3l/D8/+1Qrf1od0K4C7M/iJOzA/7gZjh1prgJp6i95zKbRrijIkg80gqIbHRXs+el5WDw9+PKu2blLK8Buasl9bIU0Wb0lnv8cfwrKAxMIM2+I4NQZiVFEe4EMO2w8+PBoN7cmLXAVI27zFHIwl1kpueRfpfhwiuGkWDQT1s7xxp6datGxdddBHPP/88b7311hk5p/hW6rA9dOhQr8sHDRrk/r9hGDgcDncQl5Lt+nwp6yd8gis7l5D4GAJCgjAMg+z8wHpiZxId/30/4bUq0f7lq1j1SArHNh4EHAQEB+DKzcWVnUtwXDgXP92TmJbePr0WrrUoH8d37OfP97/GPyTIoz2hw+EgKDqcgPAQUrbs5c8PvqH104PL7bynz2pek0vxzksWa/SA4l8Dioh4qgLsyv9/UP7DwAydWRQfajULsybbasrmD1TH7JzuKSS2EjV6t2f7x4vMJgpOz3uW4TJI++sQEfUSiO92cSnLW5qmdP75ZTxCQdjOwuwQWdLoFtb1HANqlbI855fKrRty8dOD3ONsZ+w3x9mu0q4xDQb1sHWcbW/GjBnDHXfcwWOPPUbNmjXP6LnFk8Moxbh606ZNK9NBhwwZcsoFOlukpqYSFRVFSkpKmaeGL42Mg8dYfNM4ctMyCKtRvPbXlZfH8W37qXX9ZbR7sQdwkJw0B4mLt/PX15tI359KYGQACd1qUuPqhoQmRGHeGIt+aramB26I2VHy9G16azab3/qy2FBUnteXjF+AP1fOfpbgquVz3vKxFbO9YtFaHUsW5nPYHI25LSIlywE2Y4Zr655iTd9eWChm8M7G/LBv1WbHU1KHxKxjx1l5/1sc+mULgeHBOGMiwM9BTmo6WUdSCalWhfav/J0qlzQqZXmPYd4Dg32e05QJRAKN83/OADZiVlaU1L8nG/P+2bqU5Tnz7H5vB3MYwNRt+8hOSSMoKozIBtXPWI22nJ1KVbN9PoTns82++avIPJRCZH3v08f7+fvjrBJF0tK1nNjbgPCaUQSGBVHr2ubUurZ5/lbWV5NW++0cPCeksW7uIZg3Tu9St+1j79crObZ+O4bLRfRFdah59aVENa3tNUwfXbcd/6AAH0HbLIezkj9pe4+RsvV3gqt24CyYrBSzbBGYbzjHMWtpnBR0cLI6C9VAQVtETi4Qs22zNZyoFbatttt+mPdf637ipGD41RBO1rTPWSmCDm+OZuu0Beydu5yMA8cwDIOAsBBq39SFBoN7Et2kLLXI3ma59MYaG7zwdfpRMFumL3mY13Vhc/j5EdVINclSoNwSkMvlYt68eUyePJnZs2eX12HPW0c37MDh5yix44uzUjip2/8i5Y/9hNes4mULP8waCmtylsJh2wqP/pjhsfh5DMPgj3fmsm3at2Qnp+HvDACHg4MrNrFzxvfU6d+V5mP64xfgeXM1XOYkPMVl4v7q1OECcsHYD/yO2Uvd95T19kvGHJXFGvLQesPLpGAK5WDMEUgqspwicm4Jwxzt6Sjm5DTWCCVBmPfjouHU6lB4DPPeXLKgqHAuuu9GGg3rQ+q2feRl5eCsEklk3YRTqC0Nw7zf5eK7QsFqSlf4g0AA5iglBykYhaToPjn5x/XHbIISzamOciVyvjntsP3nn38yefJkPvzwQ5KSkggNVVvX0jBy83D4naQTnhVoXSW19LFumOmYtQrW15eO/HW18DX26fb/LuKPd+YSEOokslFBkxDDMMg6epxt0xYQEOyk2X03euwX3bQ2B5dvdLfRN2XmP8zanOyUTALCgwmvE495E96F+SZTEeOwJmPWPOVRUJNtvTlkUfCVrvcPJSIiJQvCvIdEYDa3CKLkoGl1niy99P2H2b9wDX/N/4W8jCwCQoOpcfWl1OrXmcj6pR1tIhCzMiER8+2/6P3OwLwnBlG8HXkc5r00E89mKLmY7ztW85hkzBFPrPkbqlByLbrI+e+UkkV6ejpTp06lS5cuNG3alJdffpnKlSvzxhtvsG+fPbMjnW8iG1THlZtX4lT0OanpBIaFEFYrBt83ZitUB2HeRBPyH/WAFnjreAOQcyKDbdO+xS/Qn5C4Sh5NQhwOB8GVIwkMD2HHzO/JOHjMY98afdoTGB5C1tHj+UusWSjNoG24DLIOpxHftQFh1Sth3nTzMCeUcZX0tNjAhVmjnYf5BmH9ylvPW3j+smNoFjQROT3Wt2Qnu89ZH/xLZ9+C1Swb9jJ/fvA1OSlpOPwcZCefYMt/vuLHO14k8fu1ZShjAmazwkwKmrwYmGE5A/N+WIviNd+hmO8rQfn7ZmCG7OP5+wbkHzeUggnCrHG5RS5sZQrbP//8M3fddRcJCQkMGzaM9evXc9tttwHw9ttvM2rUKKKiio7zLN7U6NOeoKgwsg6neF1vGAYZB45R+ZLGRDWpi3nj8hUGrRtdbcybZC3M2gTfNStJP/xGRuLREmcnC64SRfax4+xf9KvH8uiL6lD7pi5kJ58w2xDmFfS6z03P5viOI4TVjKbRnR3z97CCbQbmjflMSs0/r6/e99bMkVmYNTIiIqcqCLP5RA6+79dWEC9dc7XkzbtZO34a2alpRDasTmi1yjhjIgmtVpnIhtXJPnaCX5+eSsrWv0pZxkCgAWbo9sO892VgfgCIyl/nq2xRQDPM95pwCtqmh2HW6lvvOVYTR2vW3pxSlk3k/FSqsP3qq69y0UUX0blzZ95//31atWrFlClTSExM5M033yyxdla8i6hXjboDupOTmmEGVldBTUheVg7HdyQSXDWaJn+/DoejOmaYzsTzBl64nVwVytIxJSPpKIZh4BfouyWR1Z48I+mo53KHgxYP96fxiL44/P1J3XGA1G2HObH7GP4hflTr2ZBL37iRqIaFR1nxp+AryjPJas9e0le6foW2FRE5HbEUTPJS9L3RhXkfD8HXt45F7Z71I1lHUgmvHVesU7rD4SC8ThyZh5LZM3d5GcoYiFkp0xxolP9oijn6yMkqzIIwm5TUwrzOcMwKC2+VGU7M96hjXtaJXDhK1Wb74YcfxuFw0LdvXyZNmkT9+vXd63Jy9In1VDW770b8Av3Z8clijm9PNG/LhoFfgD8RdRNo9eTfCo3LWQ/zKzkrEFozk/lj3txrUZZ2cf7OQDCMIu2uvTAoNr4rgF9gABc9cDN1B3QncelCnDGZRDWqTHCVUALCAnE4/DC/YrQ6CFlvOudS271cCprHeOvoJCJSVARQl4L7tTWNu1WhEoY5gsnJ337zsrLZt2A1QVFhPu/TDoeDwPAQ/vp6JRfdf1OxDu0lC+TUh4RNp2BWTGv2Sau5o3/+w3qfUkWGXNhKFbYvvvhi1q1bx7x58zhw4ADDhg1j4MCBto1ReaHwC/Cn2egbqXNTF/YtXEP6vsP4BQZQqUVd4rtdTEBw4TZzlTBv4kcxJyWwpgmO4VQmYKncpiEBYcHkpKYTFOV9tq/ctEz8ggKo3Mb3QPyhCZHUv7UxZnMN600FCmqxczHfXKyhpM70zGLW2LclDVnlKrQtmDVPBzCf69z8ZUFAVQpqrUREfLHuy0cxa3XzMINtFcx7eenuITknMsjLyvFa4VGYf3AQeRlZ5KZnEhR5pu+xuZj3zDw8K1WsEZ5EpFR/8b/++ivr16/nvffeY/r06dxzzz2MGTOGm266iZtuusnuMp73QqtVoeGQXqXYMgAz7MWe9jmjmtSiSrvGJH6/joCw4OLD++W5SN9/hEqt6p9kKuC9mKHa6hRUuAOiVduRnr88hjM/Bmtk/jkz8D6Rg/WhwIlZw5OOOelDJubzbb3J5WBeawpmm8aS3/xE5EJnDSVa2pFCigsIC8EvMIC8rJK/Qc7LziEgNJiAkNJ3ujx91r08jYLKFOv+b3W4TMO8jyp0y4Wt1B0kW7VqxVtvvcX+/fv573//S4cOHfj444+58cYbcTgcfPHFFyQlJdlZVilHDkc2rcdfQ6M72xJSzUH28eO4cnJx5eaReTiF1G37Ca1ZldbPDMbP31eNcBZmrU0QZi2OVYNcuHbDgRlUAzHH2j7TzUj88s/rjxmgC48SYLWf9MMc9s8B7KSgTaU1fJc/ZhgPxqzBL21HJBGRUxcQHES1K1uTnZLms2+UYRjkpKZTvVfbEvvglL8wzHtoHp7fakJBzba1PvoMlkvKw65du3A4HKxbt66ii3JeKPPQf06nk9tuu43vvvuO7du388QTT1C9enXefPNNateuzc0332xHOaXc5AA7gN8JiT3KRfdfQtsJ3Wn38mVUaRtF2l+HMAyofdPldPr3A0Q3rV3CsVIpGAklEPPmG0Dx9nvW+NoVNbNYNGab9xAKet5n5P8/OH9dZczrsdqZe/tQUHgyijPd0VNELkS1b7ycoKhw0v86XCxwG4ZB2t5DOCtFULtf5zNcshN4huqinfcLz6JZdPr6852Bec3We8rZPYjE0KFDuf766yu6GOe10/oYXKdOHZ577jmeffZZ5s+fzwcffMC8efPKq2xS7nKBbZg3gEAgBL9AByFVw3BGRxHdLIGGw6/CP6guodWqlnwooKCW2AqmgZi/UlbQtm7EVjOTilQJs5d9CgU3/pD8ZVbZjuI5xGJA/qNw8A6kYAjDM/mVrYhciGJa1qflPwby2wvTSd22n6CoMPydgeRlZZOdkkZQdAStnvwb0c3qnOGSpVHQD6fwzLxWE0JrqnqrOYmvYWazMd+TXJj328L35HPRcSCJgg6kfpjf/FqTHsmFqFymy3M4HPTp04fPP/+cv/7SV+xnr4OYN7VgPKfc9cMvMJTA0DAi6vgRWq20IdIKooWbZjjylzsxm2FYv2JnQxtnq4a9ev4jhoJamb8wb5BWZ59MzDcIa8IGi9U05kxPziMiF6raN1xGp3cfpM7NXcw23JnZ+AUGUm9Adzq/+yA1r+l48oPYJhAzRIZhhuvg/P9H4vtbQjDvq7uA3zFn+N2FWRn0O+YMl2d3bbB3xzGv4wTm+2BI/r8n8pfbM89Et27dGDVqFKNGjSI6OprKlSvz1FNPYRgGzz77LC1atCi2zyWXXMLTTz/NuHHjmDZtGnPmzMHhcOBwOFiyZIl7ux07dtC9e3dCQ0Np1aoVK1as8DjOF198wUUXXYTT6aROnTpMnDjRY32dOnV4/vnnGTZsGBEREdSqVYv//Oc/tjwPZ7Nyb+BVuXLpBuqXMy0PcyYvf3x/xgrArKE4gnmjPJkozECdg+9a3tz8c0aXoaxnkkFB0LbeFPwLrcvDDN3heM4MpxFJROTMqXxxAypf3ICctAxy0zIJCAsmMKyimuZBwT3fqr0NpHililXTXfT9wfqWNYWCUGpVYlid0XOomH4+p8qgoMLGuh4wr88f8701CfO9pPyvadq0aQwfPpyVK1eyevVq7rrrLmrXrs2wYcMYP348q1atol27dgD89ttvrF27ls8++4zY2Fg2b95MamoqU6ZMASAmJob9+/cD8OSTT/LKK6/QsGFDnnzySQYOHMi2bdsICAhgzZo19O/fn3HjxjFgwACWL1/OvffeS+XKlRk6dKi7bBMnTuS5557jiSee4PPPP+eee+6hS5cuNGlS0uAL55dSJQZ/nx3kinM4HOTm5p58QznDsjBvXiW95Fazj9J++rZGR/mr0LEL30Ty8pdX5VSGJzwzMjBr/AMwPzjkYs2GWdDe0OpIGUbBBwvNlCoiZ15gWEgFh2xLNGYtttX3xZtszPtq0SYkhzC/ZS06f4EfBRPhHMg/x7kyxLA1fb232YqtWZTT87cr//fDmjVrMmnSJBwOB40bN2bDhg1MmjSJESNG0KtXL6ZMmeIO21OmTKFr167Uq1cPgJCQELKysoiPjy923Icffpi+ffsCMH78eC666CK2bdtGkyZNePXVV7nyyisZO3YsAI0aNWLTpk28/PLLHmH76quv5t577wXgscceY9KkSSxZskRhuyjDMIiKiuK6667T2NrntNJ+LVeWr+8SMAPqQcybiHXjtGo0YjAn3HFh3lytYaKclGW8WfscwfxQYI1pHoj5BmH1ri88okoW5nXEcW63KRQROV1+mM3xrBGcCjcZMTDvo2AOfVj4Pu/CDNtW5Y43gZj33MOcO2E7l4JJfrzxx3xO7KmM7NChg8fERx07dmTixInk5eUxYsQIhg0bxquvvoq/vz8ff/xxseYevrRs2dL9/4SEBAAOHjxIkyZN2Lx5M/369fPYvnPnzrz22mvk5eW5K2oLH8PhcBAfH8/BgwdP+VrPRaVKOkOHDuXzzz/niy++4MYbb2T48OF07drV7rJJuQrCfLmtZh2+lHWYJgfmV32VMG+MVq14KOYIH1GYAduaTa1wp8q/MMN6HBX3VaE1Brh1fqvGKAfPzj5Wk5JqmOUVEbmQGJj3xTwKRqCqTEFTvKKzRAZh3i+tzvZZmO8FWZjhvHg/HsMwyMzIJScnj8AAFwGBKQQGFdvsLBWA+V5iPT9FWUMknvkKpmuvvRan08ns2bNxOp1kZWWVeo6UwMCC18kK8y6X+T7ubQZqb0NUFj6GdRzrGBeKUr3qkydP5s0332TGjBlMmTKF7t27U69ePYYNG8bQoUOpVu3UB+2XMyUA88ZodTzxFm6tm0FZ2907MDvIeOtpfQKzbV4OBeNWQ0HNx978/yeU8ZzlqfDNweo5notZPmvccOtDRQLnThtCEZHTZWAOd3oI835ujTQSjdmMsEr+/49SELit2Y0DMYP1PiCZgtrfHAraNpsx5MSJLI4cTCMjIweXyyAiwkFOTjprV63jqqsbExV9NjSdKUkI5nvHCQqmqrdY73fh2DUE7s8//1zs54YNG7prl4cMGcKUKVNwOp3ceuuthIYWNGUJCgoiLy+PsmrWrBnLli3zWLZ8+XIaNWpUpubHF4JSj0YSFhbG8OHDWbZsGZs3b+bGG290j63dt29fvvvuOzvLKeUijoLZFAv/YVk1FlmYN83yao9s1XhkY958C//xWZ1mHJgfACpq3OpwCmquLQ4Kxg2PxLyBhmB+CFHQFpELhYEZlLdjdma0RpsyMMP3n5gh24VZ2ZIA1MZ8r7GC9lbMbz0dFARSq2leGpDD8ZRM9u1JIT0tG39/B05nAKFh/hw+6GLlsl18+H+/kHysaM352caBObyfNdCA1f8nN//ngPz19ryH7N27lzFjxrBlyxamT5/Om2++yf333+9ef+edd7J48WK++eYbhg0b5rFvnTp1+O2339iyZQuHDx8mJ6fkGUstDz30EN999x3PPfccf/75J9OmTeOtt97i4YcfLtdrOx+c0tB/jRs35qWXXuKvv/5i7NixfPvtt7z99tvlXTYpd06gIeZNMQezCYXVYcOFWUNRl3IaETL/2Cfw3mHEYo1mcrSczllWlTFvgr5uLtYHkWg0rraIXFiOYVaG+GMG5cD8/wfl/5wN/AGsBzYCGzADeHL+/nsx3wesfa1O59Z7Qh55uekcSDyOy2UQ5AzA38+PwEAwDNi7y5/Y2HD+2nOM777+48xc8mmJAOpgVuJYITs3/+c62DnO9uDBg8nIyKB9+/aMHDmS0aNHc9ddd7nXN2zYkE6dOtG4cWMuvfRSj31HjBhB48aNadu2LVWrVuWnn34q1TnbtGnDp59+yowZM2jevDlPP/00zz77rEfnSDGdUuOhlJQUPvnkEyZPnsyaNWto3LixZo48Z4QATTE7K1oTCQRihsny7iFt1aCX1OjO6oBYUbUWwZjtCv/CrIWxxga32mhnYz5naiolIhcSA7Pju4H3eRKyKWjD7Yd5L3VhBvRUzCYmKXjO6WBxYobQPNJO5JCTk0tQUIA5ZocTQsNhz074azf4B/gRFh7ExvVJXNknneiYs3VkK0sEZri2gnbhoQ3tExgYyGuvvcY777zjdb1hGBw4cIC777672LqqVauyYMECr/sUFh0dXWzZTTfdVGL77127dhVbdiFOAV+msP3dd98xefJkvvzyS/z9/bnlllt4/fXX6dSpk13lE1s4MJuKnC3D11X05AXxmG8WSXg2Z7HaJdbC99BWIiLno2zMZh7egrZVawtmTXde/r/++dtnAfsx7+1hXvb3z1+ejsvIoXJVPwICzDCakwM7/oSVywysPnRh4U4OHTjB3t3J50DYBvM99uwp58GDB/noo4/Yt28fd9xxR0UX54JUqrD97LPPMnXqVHbv3k2nTp1466236N+/P2Fh3v6IRCyhFNyIff2qGYW2rSgOzDaGVTBrYrLzl4VT0L5QRORCkofvoeyyKegoWbTPC5g119kUjOjkjT8QzqrlyeRkG0THhJCZDnt3wdEjnltaA1648iq6YubcFBcXR5UqVfjPf/5DpUpFxzyXM6FUYXvcuHFERkZy55130rhxY44dO8a7777rdVuHw8GDDz5YroWUc1UIZgfDYxTvnW3JxqwJiTmD5fLFn7OjHCIiFc0ayq5o4Lb6sVhNAK3Q7W3/TAqaUvhg+LF0YRpxCcHFhpGzZGXlEhjoT3TM2T4iScUoPL26N96G45Mzq9TNSFJTU3nvvfdOup3CthRwYE56YHXCdFIwprWLgkkPqlNyu24RETmzgjCbGh7Bc3bgojXZvtp0W/tk4ztq5NCidSTLvs8l7UQ24RHFO6EbhkHKsUzq1I+hVl3Vysq5qVRhe+fOnXaXQ85bYUADYA9m+7/CXys6MTseVqmYoomISAmsTo5ZFAzVWrhG22qr7a2yxEVBX5csPEelsobEyyOuWg1atw/n5x/MnBEWHuSu4c7Lc3HsaAbBIYF069XIZ823yNmuVGG7du3adpdDzmvhmCOgHMdzuvZoNO25iMjZKhJz3Ow9mN9OWs0BzWH7zJ9DKd6MxAriCZiB25phsnDYNseddjhq0PfG6hiGwbpVf3Eg8Th+/n4YLrP2PDI6hL43XESTizRzr5y7zvy8oXKBcmDeuCMruiAiIlJqVTCjwiHMIf2szuOZ+euLVpgYFAyjWhWz70405lwK1gyUIZj9Y8yO8YGBcP2AlnS4vA7r1+zj8ME0AgL8qF0vhhatq3ltXiJyLilV2L7iiit8rgsMDCQuLo4ePXowcOBAAgKU30VERM59aZiT2qRg1lRbM0jGYDYRtGq8/SjoTOnC/OayLgVTkwdhDrHqm8PhIKF6FAnVz5YhaUXKj8MoRTfV5s2b+2wrlZubS2JiIqmpqbRv357FixcTGnr2jC95qlJTU4mKiiIlJYXISNXGiojI6XJRMJZ/EGd3M7pUzGnarRGjrLJaE9lEYM5BkIrZidIadaQyZhj3p6Dz5NnT1lrv7VIRSlUN/fvvv590m2+++YYBAwbwr3/9i3/961+nXTAREZHzQy7mbIyHKRiFKRAzmMZS/qMxZWM22ziaf25reNUYvI8cUlQesAszWBed/dCJ+aHhOOb11KFght28/GXbKWhm4sRsTlIZtVw9t+zatYu6deuydu1aLr744goty9SpU3nggQdITk627RzdunXj4osv5rXXXiv3Y3sbHPOU9OnTh8cee4wvvviivA4pIiJyjssBNgM7KWiz7J+/fB/wJwXBtDyk5p9vN2YzkGzMYLwL2JT//5M5ll8mawSSovwwg/NRCj485AJbC53XalqSkX/urZjXfP5zuQyS9qWyY+thkval4nKd/eNcDx06lOuvv76ii3HeKtePme3bt+ef//xneR5SRETkHJUFbMAMn1ZozcUMoUGYYTYNM4w25vSbW2QAOzADcNEaaaPQ+ib55/blOL4nq7EEYgbyE5g15nsw23Y78WweE4BZE56KGcQblPpqzkW7dxxl2ffb2b83hezsPIKC/KlWM4rLutendj1NmnahKreabYCMjAx1kBQRESEb+APPWl5/CqY4z8p/BGKG2xOnca4c4ACwhYKaa1eRbRyYATwTs6lHSUqaZr3w8axtMzFrwwu37S7ML39dMuYkZ+en3TuO8uXM39i57Shh4U7iq0UQFu5k5zZz+e4dR205b7du3Rg1ahSjRo0iOjqaypUr89RTT7lnjnz22Wdp0aJFsf0uueQSnn76acaNG8e0adOYM2cODocDh8PhMSvljh076N69O6GhobRq1YoVK1Z4HOeLL77goosuwul0UqdOHSZOnOixvk6dOjz//PMMGzaMiIgIatWqxX/+859TutYvv/ySRo0aERwcTI8ePdi7d6973fbt2+nXrx9xcXGEh4fTrl07Fi1a5LH/v//9bxo2bEhwcDBxcXHcfPPNHutdLhePPvooMTExxMfHM27cuFMqZ1HlGra//vprmjVrVp6HFBEROQclYQZtKJg5l/x/rbdeqwmGVfN7Kg4Bv2PWjqdSEH6PY9ZkF27C4MAMw4eLLC8qiOIzRRZljU4ShFmjfZJp2QnI3yelhG3OXS6XwbLvt3M8NYtqNSIJCQ3Ez8+PkNBAqtWI5HhqFsu+325bk5Jp06YREBDAypUreeONN5g0aRLvv/8+AMOGDWPTpk2sWrXKvf1vv/3G2rVrGTp0KA8//DD9+/end+/eJCYmkpiYSKdOndzbPvnkkzz88MOsW7eORo0aMXDgQHJzcwFYs2YN/fv359Zbb2XDhg2MGzeOsWPHMnXqVI/yTZw4kbZt27J27Vruvfde7rnnHv74448yXWN6ejr/+te/mDZtGj/99BOpqanceuut7vUnTpzg6quvZtGiRaxdu5ZevXpx7bXXsmfPHgBWr17Nfffdx7PPPsuWLVuYP38+Xbp0KfY8hoWFsXLlSl566SWeffZZFi5cWKZyelOqauijR31/GsvLy2Pfvn3MnDmT999/n3//+9+nXSgREZFzVy7mCB3WBDDeODCDsdWOOe8UznMEs2mGNVFYNgVB3qo9h4Ih+Mhfn5f/8BUBYjBrykvaJhtznOxwzFr5kq6VQutO5TrPfgcTj7N/bwoxlUOLjd7mcDiIqRzK/r0pHEw8Tnz18h8FpWbNmkyaNAmHw0Hjxo3ZsGEDkyZNYsSIEdSoUYNevXoxZcoU2rVrB8CUKVPo2rUr9erVAyAkJISsrCzi44sP0fjwww/Tt29fAMaPH89FF13Etm3baNKkCa+++ipXXnklY8eOBaBRo0Zs2rSJl19+maFDh7qPcfXVV3PvvfcC8NhjjzFp0iSWLFlCkyZNSn2NOTk5vPXWW1x66aWAGYybNm3KL7/8Qvv27WnVqhWtWrVyb//Pf/6T2bNnM3fuXEaNGsWePXsICwvjmmuuISIigtq1a9O6dWuPc7Rs2ZJnnnkGgIYNG/LWW2/x3Xff0aNHj1KX05tS1WxXqVKFqlWren3Ex8dzySWX8NJLLzFy5Ejuuuuu0yqQiIjIuceqnT6COfJIDiXXZ1nhNDf/59KMElL0fPspmBa9cNgtWnteuEmJVbNa0tt/KFApf9/cIuusEO/AHDvbah5zsppwa9352dQ0PT2b7Ow8nMHeh3MMcvqTnZ1Henq21/Wnq0OHDh4hv2PHjmzdupW8PPPDzYgRI5g+fTqZmZnk5OTw8ccfM2zYsFIdu2XLlu7/JyQkAHDw4EEANm/eTOfOnT2279y5s8e5ix7D4XAQHx/vPkZpBQQE0LZtW/fPTZo0ITo6ms2bNwOQlpbGo48+SrNmzYiOjiY8PJw//vjDXbPdo0cPateuTb169Rg0aBAff/wx6emezZoKl9O63rKW02vZS7PR008/7XOc7YCAAOLi4rjiiivcn5BEREQuDAZmwE7CbLbhyl9mjU9t1WD7CrfW1OXRZTxvKgUzNULBCCHZFLSbLlx7bnWIzMUciq+ksO3AnKbdwGyLXVBjbhh5pGw5RsoWF7gyCa+bQEyrGjgcARQMM+iN1czk/Jy0JjQ0iKAgf7Iy8wgJLf7cZmeZnSVDQ8t7mMfSufbaa3E6ncyePRun00lWVhY33XRTqfYNDCx4Ta0s6HKZH+AMwyiWD71N31L4GNZxrGOUhbcsai175JFH+Pbbb3nllVdo0KABISEh3HzzzWRnmx9wIiIi+PXXX1myZAkLFixwt1dftWoV0dHR5VrOokoVtsurgbiIiMj55QCwFzOYBlEwk2I2BSHV6nBYdHQQI39ZDJ5NPUoji4JhBC1BmMHaCveFOzFCQRCvUorjBwD1Mdt+HwEySNlykI2vLePwqj3kpmWbHxOCg4huUZdmoy6lavswCmq6C8vLL1eVU7jOc0NsQgTVakaxc9tRqoVEeoRCwzA4eiSdug1iiE2IsOX8P//8c7GfGzZsiL+/+VoEBAQwZMgQpkyZgtPp5NZbb/WYgDAoKMijJrq0mjVrxrJlyzyWLV++nEaNGrnPXV5yc3NZvXo17du3B2DLli0kJye7m6L8+OOPDB06lBtuuAEw23Dv2rXL4xgBAQFcddVVXHXVVTzzzDNER0ezePFibrzxxnIta1Hn5/c5IiIitkvHHCvbgdmUw+KPWZOcSUHNtTWVedEAXAlzJsaycuAZ2Mk/T3D+efMoqL12Yda6O4AaQGnbDDvyt40kefNufr7vK9L2HiIkNpqQBDOw56ZlcmTVFlY+uI92L/Yl7rLKhcpi1fC7MEN2GCfvSHlu8vNzcFn3+hw+mMb+v1KJqRxqNh3JyuPokXQiIp1c1r0+fn72zKa5d+9exowZw913382vv/7Km2++WWxUkDvvvJOmTZsC8NNPP3msq1OnDt9++y1btmyhcuXKREWV7huIhx56iHbt2vHcc88xYMAAVqxYwVtvvWVL/73AwEBGjx7NG2+8QWBgIKNGjaJDhw7u8N2gQQNmzZrFtddei8PhYOzYsR610l999RU7duygS5cuVKpUia+//hqXy0Xjxo3LvaxFletoJCIiIhcOa5pyb00DnBQETgMzbAZREJL9MZtzNOHUpm0Py9+vcG2kI/+8oZjNOawmLWCG5gaY7azLxuVy8duETzix+yCRDaoRFB3uHiIuMDyEiAbVyE5O57cXfiA30wrzeZgBPze/XDmYnTk3YDa5Ofsneimr2vViuH5AS+o2iCHtRBYHEo+TdiKLug3M5XaOsz148GAyMjJo3749I0eOZPTo0cX60DVs2JBOnTrRuHFjdydDy4gRI2jcuDFt27alatWqxcK4L23atOHTTz9lxowZNG/enKeffppnn33Wo3NkaXTr1u2k+4SGhvLYY49x22230bFjR0JCQpgxY4Z7/aRJk6hUqRKdOnXi2muvpVevXrRp08a9Pjo6mlmzZnHFFVfQtGlT/u///o/p06dz0UUXlamsp8JheGtcI6SmphIVFUVKSgqRkeXfc1hERM51v2MGymAf6/Mwh/+zArnVkTAAs0lFTU69zsvAnH0ymeIT2EBBbXYQZqD3tk3Jso6msud/K9g27VsO/fIHfgH+OGMiCImPwVnZs6lEXlYO6fsO0/7Ve6neszGwDbNWO4iCmmyrptvArGFPKFN5ysOZeG93uQwOJh4nPT2b0NAgYhMibKvRhtJPM24YBk2aNOHuu+9mzJgxtpXnVNSpU4dx48aVOaSfK86/73JERETOiMJNOLzxx6yBzgRiMWubAzGbjpQ0g2NpODDDeiZmqA6k4C09j4KgWw+zprtsTuw5wC9j3iF54y5y0jPBZYCfg8xDKWQdSSUkoTKRDarjyA+R/s5AXHkujv2+k+o9I/A+i6VV854FJGK2VT/d5+Hs4+fnsGV4v9Nx8OBBPvroI/bt28cdd9xR0cXx8McffxAREcHgwYMruii2UdgWERE5JSGcfEZEF2borYH35ianIxRoBPyFOTpJRv5yf8ymHNUpffvsAnnZOax+/D2O/b6TiLrxZBw4RvbR4/gHBUJQIK6cXNL3HyEg1ElYjaru/RwOCIpyYHaqtJrMeBOUX9ajVETt9oUoLi6OKlWq8J///IdKlSpVdHE8NGnShA0bNlR0MWylsC0iInJKKmMGxjy8t7s2MNsqV6X8g7YlBGiIGV7T888ZjFmjfmpNFw7+9DvHfttJWM2q+AUG4B/iBIcDw+XC4eeHX2AArtw80vcdITShMg5/P1z5I1lENIjBfD5Kul5rZJbzd+r2M6nw1Oq+qMVwxVIHSRERkVMSlf/IwmyXXTjQWNOmB3EqnRLLLgQz/FfBnNXx1NsIJy39DVduLgEhZhMPZ6UIAkKDycvMcW/j7wwkNzOL7BRzSvrMA8k4q0RRuXWDMpzJvnbMImcThW0REZFT4ofZJroSZtjOoKCGOQuzhrkBp9JmuiJlHU3FL6Cgpt7h5yC8ThwOPz/yMrLBMMw2Iy4DV3YO6YlHycvOodEdfQiKrELxUVKKskZoObeeF5FTpWYkIiIipywQsxnHCcwmJdbEMZGYIbx8J/Y4E4IqReDK9Zw1LyQ2GsPl4sSOJHLTszAMAyM3j/QDyYTVqEKTu6+hwdBe+VtHYs486Y/32mtrdk37hsITOZsobIuIiJwWBxCR/zj3xXdtxe5ZP5KbkU1ASEHb69D4GJyVo8g6lEza/sMEhoXQ4rFbqXVtR0LiCgfn6pi1+xmYo41Ys1laM2ta21TM1OUiZ5qakYiIiIhb3GXNiW5Wm7S9B3HlejYH8Q/0JyDUibNSBK2e+BuN7+xbJGiD2TmzAeaHjxwKhifMwgzYtTGHQhS5MFR42J4wYQLt2rUjIiKC2NhYrr/+erZs2eKxjWEYjBs3jmrVqhESEkK3bt3YuHHjSY/9xRdf0KxZM5xOJ82aNWP27Nl2XYaIiMh5wT8okLYT7iS6aW2O70gkbe9BspJPkHkkldTt+8lOTqPuLV2pP6hHCUcJB5oCjTHHA68B1AeaYwZtdY6UC0eFh+2lS5cycuRIfv75ZxYuXEhubi49e/YkLS3Nvc1LL73Eq6++yltvvcWqVauIj4+nR48eHD9+3OdxV6xYwYABAxg0aBDr169n0KBB9O/fn5UrV56JyxIRETlnRdSrRuf3HqLFw/0JrV4FV3YuGAbxl7eg7Ut30WrsII9OlN45MNtvJwDVMEdLOffasIucrrNuuvZDhw4RGxvL0qVL6dKlC4ZhUK1aNR544AEee+wxALKysoiLi+PFF1/k7rvv9nqcAQMGkJqayjfffONe1rt3bypVqsT06dNPWg5N1y4iIgKu3DxyUtPwCwwgIDzEY5r2c43e26UiVHjNdlEpKSkAxMSYbcB27txJUlISPXv2dG/jdDrp2rUry5cv93mcFStWeOwD0KtXL5/7ZGVlkZqa6vEQERG50PkF+OOMiSQwIvScDtoiFeWsCtuGYTBmzBguu+wymjdvDkBSUhJgTjVaWFxcnHudN0lJSWXaZ8KECURFRbkfNWvWPJ1LERERERE5u8L2qFGj+O2337w28yj6adowjJN+wi7LPo8//jgpKSnux969e8tYehERERERT2fNONujR49m7ty5/PDDD9SoUcO9PD7enOY2KSmJhIQE9/KDBw8Wq7kuLD4+vlgtdkn7OJ1OnE7n6VyCiIiIiIiHCq/ZNgyDUaNGMWvWLBYvXkzdunU91tetW5f4+HgWLlzoXpadnc3SpUvp1KmTz+N27NjRYx+ABQsWlLiPiIiIiEh5qvCa7ZEjR/LJJ58wZ84cIiIi3LXRUVFRhISYvZ4feOABnn/+eRo2bEjDhg15/vnnCQ0N5bbbbnMfZ/DgwVSvXp0JEyYAcP/999OlSxdefPFF+vXrx5w5c1i0aBHLli2rkOsUERERkQtPhYftd955B4Bu3bp5LJ8yZQpDhw4F4NFHHyUjI4N7772XY8eOcemll7JgwQIiIgqmxt2zZw9+fgUV9Z06dWLGjBk89dRTjB07lvr16zNz5kwuvfRS269JRERERATOwnG2zxYai1NEROT8ovd2qQgV3mZbREREROR8pbAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbFLhYfuHH37g2muvpVq1ajgcDr788kuP9Q6Hw+vj5Zdf9nnMqVOnet0nMzPT5qsRERERESlQ4WE7LS2NVq1a8dZbb3ldn5iY6PGYPHkyDoeDm266qcTjRkZGFts3ODjYjksQEREREfEqoKIL0KdPH/r06eNzfXx8vMfPc+bMoXv37tSrV6/E4zocjmL7ioiIiIicSRVes10WBw4cYN68eQwfPvyk2544cYLatWtTo0YNrrnmGtauXVvi9llZWaSmpno8REREREROxzkVtqdNm0ZERAQ33nhjids1adKEqVOnMnfuXKZPn05wcDCdO3dm69atPveZMGECUVFR7kfNmjXLu/giIiIicoFxGIZhVHQhLA6Hg9mzZ3P99dd7Xd+kSRN69OjBm2++Wabjulwu2rRpQ5cuXXjjjTe8bpOVlUVWVpb759TUVGrWrElKSgqRkZFlOp+IiIicfVJTU4mKitJ7u5xRFd5mu7R+/PFHtmzZwsyZM8u8r5+fH+3atSuxZtvpdOJ0Ok+niCIiIiIiHs6ZZiQffPABl1xyCa1atSrzvoZhsG7dOhISEmwomYiIiIiIdxVes33ixAm2bdvm/nnnzp2sW7eOmJgYatWqBZhf+3z22WdMnDjR6zEGDx5M9erVmTBhAgDjx4+nQ4cONGzYkNTUVN544w3WrVvH22+/bf8FiYiIiIjkq/CwvXr1arp37+7+ecyYMQAMGTKEqVOnAjBjxgwMw2DgwIFej7Fnzx78/Aoq6ZOTk7nrrrtISkoiKiqK1q1b88MPP9C+fXv7LkREREREpIizqoPk2USdKERERM4vem+XinDOtNkWERERETnXKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERm1R42P7hhx+49tprqVatGg6Hgy+//NJj/dChQ3E4HB6PDh06nPS4X3zxBc2aNcPpdNKsWTNmz55t0xWIiIiIiHhX4WE7LS2NVq1a8dZbb/ncpnfv3iQmJrofX3/9dYnHXLFiBQMGDGDQoEGsX7+eQYMG0b9/f1auXFnexRcRERER8clhGIZR0YWwOBwOZs+ezfXXX+9eNnToUJKTk4vVeJdkwIABpKam8s0337iX9e7dm0qVKjF9+vRSHSM1NZWoqChSUlKIjIws9blFRETk7KT3dqkIFV6zXRpLliwhNjaWRo0aMWLECA4ePFji9itWrKBnz54ey3r16sXy5ct97pOVlUVqaqrHQ0RERETkdJz1YbtPnz58/PHHLF68mIkTJ7Jq1SquuOIKsrKyfO6TlJREXFycx7K4uDiSkpJ87jNhwgSioqLcj5o1a5bbNYiIiIjIhSmgogtwMgMGDHD/v3nz5rRt25batWszb948brzxRp/7ORwOj58Nwyi2rLDHH3+cMWPGuH9OTU1V4BYRERGR03LWh+2iEhISqF27Nlu3bvW5TXx8fLFa7IMHDxar7S7M6XTidDrLrZwiIiIiImd9M5Kijhw5wt69e0lISPC5TceOHVm4cKHHsgULFtCpUye7iyciIiIi4lbhNdsnTpxg27Zt7p937tzJunXriImJISYmhnHjxnHTTTeRkJDArl27eOKJJ6hSpQo33HCDe5/BgwdTvXp1JkyYAMD9999Ply5dePHFF+nXrx9z5sxh0aJFLFu27Ixfn4iIiIhcuCo8bK9evZru3bu7f7baTQ8ZMoR33nmHDRs28OGHH5KcnExCQgLdu3dn5syZREREuPfZs2cPfn4FlfSdOnVixowZPPXUU4wdO5b69eszc+ZMLr300jN3YSIiIiJywTurxtk+m2gsThERkfOL3tulIpxzbbZFRERERM4VCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIjZR2BYRERERsYnCtoiIiIiITRS2RURERERsorAtIiIiImIThW0REREREZsobIuIiIiI2ERhW0RERETEJgrbIiIiIiI2UdgWEREREbGJwraIiIiIiE0UtkVEREREbKKwLSIiIiJiE4VtERERERGbKGyLiIiIiNhEYVtERERExCYK2yIiIiIiNlHYFhERERGxicK2iIiIiIhNFLZFRERERGyisC0iIiIiYhOFbRERERERmyhsi4iIiIjYRGFbRERERMQmCtsiIiIiIv/f3v3HRFnHcQB/H3LgKaAivyQIGRpqbBlaA7JSNLVE3Jykog2VkS0vaqPaajr8ka2ZrRxbMxMOMX9gZk5H6sRhW8M5PMlJzKYNfwASYiigKQKf/nDceuCeu0N4gLt7vzY27/t8v9/n+/345XsfHp7n0AiTbSIiIiIijTDZJiIiIiLSCJNtIiIiIiKNMNkmIiIiItIIk20iIiIiIo0w2SYiIiIi0giTbSIiIiIijTDZJiIiIiLSCJNtIiIiIiKNeA70AAYrEQEANDU1DfBIiIiIqC90vqd3vscT9Qcm2yqam5sBAOHh4QM8EiIiIupLzc3NGDFixEAPg9yETvjjnVUdHR2ora2Fr68vdDqd1TpNTU0IDw/HjRs34Ofn188jHHwYj+4YEyXGQ4nxUGI8lBgPpb6Ih4igubkZoaGh8PDgnbTUP3hlW4WHhwfCwsIcquvn58eN8H8Yj+4YEyXGQ4nxUGI8lBgPpd7Gg1e0qb/xxzoiIiIiIo0w2SYiIiIi0giT7V7w9vZGdnY2vL29B3oogwLj0R1josR4KDEeSoyHEuOhxHiQs+IDkkREREREGuGVbSIiIiIijTDZJiIiIiLSCJNtIiIiIiKNMNlWcfXqVaSnpyMyMhIGgwFRUVHIzs5Ga2urop5Op+v2tX37dpt9P3z4EO+99x4CAgIwfPhwJCcno7q6Wsvp9AlHYnLhwgUsXboU4eHhMBgMmDhxIrZt22a37+nTp3eL45IlS7ScTq85ukauX7+O+fPnY/jw4QgICEBmZma3Ol056xrZvHkzEhISMGzYMIwcObLb8fz8fKvfMzqdDvX19ar9OuP6AOzHA3CvPcRePNxp/wAcWx/utH90dfr0adX9oqysTLXdihUrutWPi4vrx5ETKfGP2qi4dOkSOjo68N1332HcuHGoqKhARkYG7t27h61btyrqmkwmzJ071/La3gfmf/DBBzh69Cj279+P0aNHIysrC0lJSTCbzRgyZIgm8+kLjsTEbDYjMDAQP/zwA8LDw1FaWoq3334bQ4YMgdFotNl/RkYGNm7caHltMBg0nU9vORKP9vZ2zJs3D4GBgfjtt99w+/ZtpKWlQUSQk5Oj2rezrpHW1lakpKQgPj4eubm53Y4vXrxY8b0CPH5jfPDgAYKCgmz27WzrA7Afj07usofYi4c77R+A/Xi42/7RVUJCAm7evKkoW7duHYqLizF16lSbbefOnQuTyWR57eXlpckYiRwi5LAtW7ZIZGSkogyA/Pzzzw73cefOHdHr9bJ//35LWU1NjXh4eMjx48f7aqj9xlpMunr33XdlxowZNuu8+uqr8v777/fhyAZG13j88ssv4uHhITU1NZayffv2ibe3t9y9e9dqH66wRkwmk4wYMcJuvfr6etHr9VJQUGCznrOvD1vxcMc9xNH1IeIe+4daPNx1/1DT2toqQUFBsnHjRpv10tLSZMGCBf0zKCIH8DaSHrh79y78/f27lRuNRgQEBOCFF17A9u3b0dHRodqH2WzGo0ePMHv2bEtZaGgoYmJiUFpaqsm4taQWk57WAYA9e/YgICAAzz77LD788EM0Nzf31TD7Tde5njlzBjExMQgNDbWUzZkzBw8fPoTZbLbah6utEVsKCgowbNgwLFq0yG5dV1gfatx5D7HHnfaPrrh/KB05cgQNDQ1YsWKF3bqnT59GUFAQnnnmGWRkZNi8TY1Ia7yNxEF//fUXcnJy8NVXXynKN23ahJkzZ8JgMODUqVPIyspCQ0MD1q5da7Wfuro6eHl5YdSoUYry4OBg1NXVaTZ+LajF5P/OnDmDAwcOoKioyGZfy5YtQ2RkJEJCQlBRUYFPPvkEFy5cwMmTJ/t62JqxFo+6ujoEBwcr6o0aNQpeXl6q/9+utEbsycvLQ2pqqt1f+bvC+lDjznuIPe60f1jD/UMpNzcXc+bMQXh4uM16r7/+OlJSUhAREYGqqiqsW7cOiYmJMJvN/IM4NDAG+tJ6f8vOzhYANr/KysoUbWpqamTcuHGSnp5ut/+tW7eKn5+f6vE9e/aIl5dXt/JZs2bJ6tWrez6hPqBVTCoqKiQwMFA2bdrU4zGdO3dOAIjZbO5x297qy3hkZGTI7Nmzu51Dr9fLvn37rJ5/sK2RJ4mHI7cJlJaWCgA5d+5cj8fkbOujJ7dNONseolU83Gn/UIuHK+wf1jxJjG7cuCEeHh5y8ODBHp+vtrZW9Hq9/PTTT301BaIecbsr20aj0e5T6mPHjrX8u7a2FjNmzEB8fDx27Nhht/+4uDg0NTXh77//7nZFAgBCQkLQ2tqKxsZGxZWH+vp6JCQkOD6RPqRFTCorK5GYmIiMjAzVK3S2xMbGQq/X4/Lly4iNje1x+97oy3iEhITg7NmzirLGxkY8evTI6vrobDOY1khP4+GonTt3YvLkyZgyZUqP2zrT+ugpZ9tDtIiHO+0ftrjC/mHNk8TIZDJh9OjRSE5O7vH5xowZg4iICFy+fLnHbYn6xEBn+4NZdXW1jB8/XpYsWSJtbW0OtcnJyZGhQ4fKgwcPrB7vfHilsLDQUlZbW+s0D684EpOKigoJCgqSjz766InPc/HiRQEgv/766xP30R/sxaPzAafa2lpL2f79+x16wMlZ14iI/SuXzc3N4uPjIzk5OU/Uv7Osj049ubLt6nuIiO14uNP+0cneA5Lutn901dHRIZGRkZKVlfVE7RsaGsTb21t27drVxyMjcgyTbRWdtwUkJiZKdXW13Lx50/LV6ciRI7Jjxw65ePGiXLlyRb7//nvx8/OTzMxMS53q6mqJjo6Ws2fPWsreeecdCQsLk+LiYjl//rwkJibKc88953BCP1AciUnnr36XLVumOF5fX2+p0zUmV65ckQ0bNkhZWZlUVVVJUVGRTJgwQZ5//vlBHRNH4tHW1iYxMTEyc+ZMOX/+vBQXF0tYWJgYjUZLHVdaI9euXZPy8nLZsGGD+Pj4SHl5uZSXl0tzc7Oi3s6dO2Xo0KHyzz//dOvDVdaHiP14uNseYi8e7rR/iNiPh7vtH2qKi4sFgFRWVlo9Hh0dLYcOHRKRxz/IZ2VlSWlpqVRVVUlJSYnEx8fLU089JU1NTf05bCILJtsqTCaT6r1knY4dOyaTJ08WHx8fGTZsmMTExMg333wjjx49stSpqqoSAFJSUmIp+/fff8VoNIq/v78YDAZJSkqS69ev9+f0nogjMVG7Fy8iIsJSp2tMrl+/Lq+88or4+/uLl5eXREVFSWZmpty+fbufZ9gzjsRD5PEb6rx588RgMIi/v78YjUbFVUtXWiNpaWlW4/H/uYmIxMfHS2pqqtU+XGV9iNiPh7vtIfbi4U77h4hj3y/utH+oWbp0qSQkJKgeByAmk0lERO7fvy+zZ8+WwMBA0ev18vTTT0taWppTz5+cn05EpG9uSCEiIiIiov/j52wTEREREWmEyTYRERERkUaYbBMRERERaYTJNhERERGRRphsExERERFphMk2EREREZFGmGwTEREREWmEyTYRERERkUaYbBPRoLB+/XrodDo0NDRYPR4TE4Pp06dbXl+9ehU6nQ46nQ7r16+32mbVqlWWOmpiY2Oh0+mwdetWq8fz8/Mtfeh0Onh6eiIsLAwrV65ETU2N3XkVFBRgyZIliI6OhoeHB8aOHWu3DRERuQ4m20Tk1Hx9fZGfn4+Ojg5FeUtLC3788Uf4+fmptv39999RXl4OAMjNzbV5HpPJhDNnzuDkyZPIyMjAvn378PLLL+PevXs22+3evRt//PEHXnzxRURFRTk4KyIichVMtonIqS1evBjXrl3DqVOnFOWFhYVob29HcnKyatudO3cCAObNm4dLly6htLRUtW5MTAzi4uIwY8YMZGdn4+OPP0ZVVRUOHz5sc3wnTpzAxYsXsXv3bkRHRzs+MSIicglMtonIqUVHRyMhIQF5eXmK8ry8PCxcuBAjRoyw2u7BgwfYu3cvpkyZgq+//trSxlFxcXEAgGvXrtms5+HBbZaIyJ3xXYCInF56ejoOHz6MxsZGAMCff/6J0tJSpKenq7Y5dOgQGhsbsWrVKowfPx7Tpk1DYWEhWlpaHDrnlStXAACBgYG9nwAREbksJttE5PTefPNNeHp6Yu/evQAe338dGRmpeKCyq9zcXAwdOhSpqakAHifsLS0tOHDggNX67e3taGtrQ0tLC4qKivDZZ5/B19fX5m0qRERETLaJyOn5+PggJSUFeXl5aGtrQ0FBAVauXKn6KSRVVVUoKSnBwoULMXLkSABASkoKfH19VW8liYuLg16vh6+vL5KSkhASEoJjx44hODhYq2kREZEL8BzoARARAYCn5+PtqL293erxtrY26PV61fbp6emYNm0aNm/ejFu3bmHFihWqdfPy8iAiWLRoEe7cuWMpT05Oxp49e3Dp0iVMmDBB0aagoAATJ06Ep6cngoODMWbMGMcnR0REbovJNhENCp1XiGtqarpdLRYR3Lx5E1OnTlVt/9JLLyE6OhobN27Ea6+9hvDwcKv1Ojo6kJ+fDwBYuHCh1Tp5eXnYsmWLomzixIk2z09ERGQNk20iGhQSExOh0+lQWFiI2NhYxbHjx4+jqakJs2bNstnH2rVrcfDgQaxZs0a1zokTJ1BdXY01a9Zg0aJF3Y4bjUYUFBTg888/t1xtJyIielJ8JyGiQSEqKgpGoxFffvkl7ty5gzfeeAMGgwFlZWX44osvMHXqVMvDjGqWL1+O5cuX26yTm5sLT09PfPrppwgNDe12fPXq1cjMzERRUREWLFjQqzkBQGVlJSorKwEAdXV1uH//Pg4ePAgAmDRpEiZNmtTrcxAR0eDFBySJaNDYtm0bvv32W5w/fx6pqamYP38+du3ahTVr1qCkpAReXl696r+hoQFHjx5FUlKS1UQbAN566y0YDAa7f1HSUQcOHEBKSgpSUlJgNptx69Yty2u1Tz4hIiLXoRMRGehBEBERERG5Il7ZJiIiIiLSCJNtIiIiIiKNMNkmIiIiItIIk20iIiIiIo0w2SYiIiIi0giTbSIiIiIijTDZJiIiIiLSCJNtIiIiIiKNMNkmIiIiItIIk20iIiIiIo0w2SYiIiIi0giTbSIiIiIijfwH95MvtageuO4AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABKwAAAVKCAYAAADEzeNiAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdd3gU5drH8d+mN9IooYVQRToCooBIESKIKHI8YqEKWBAbKkcUBdQDCoqoKIpSxAb2ihRpFuBQFREbAqFI6CSUkJDd5/2Dd9csKSRhYGeT7+e65oKdnbnnntnZ2d07z/OMwxhjBAAAAAAAANhEgK8TAAAAAAAAAHKiYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBfixmTNnyuFwqHr16r5Oxcvo0aPlcDjUvn17X6dSIrRv314Oh0OjR4/2dSp5Wrp0qRwOhxwOR7HWL+g8Ls3nUvXq1eVwODRz5szzvu2CjrsdXhNfHhvASt999526deum8uXLKzAwUA6HQz169PB1WigEu34HA4CShIIVYBH3j7jTp7CwMFWtWlXXXHON3n//fRljfJ0qTuMuCBVmKo2FExTfyZMnNX36dF111VWqUqWKQkNDFRMTowsuuEAdOnTQiBEj9PXXX+vEiRO+TrXEGD16tEaPHq1t27b5ZPs5C7hLly494/Lu4lv//v3PeW5FMWnSJI0ePVo//vijr1MpsVauXKmOHTtq7ty5OnDggOLj45WQkKC4uLhCrV/Uc+188PX7zwrbtm3z7AcAwLeCfJ0AUBIlJCR4/p+WlqZdu3Zp165d+uKLLzRz5kx98sknCg0NPevtxMTEqG7duqpSpcpZx7JSuXLlVLduXVWrVs3XqRRJcHCw4uPjC1zmTM/DWv56LknSjh07dNVVV2njxo2eeSEhIQoMDNRff/2lP//8U0uXLtXTTz+tJUuW5CqG1qpVS2FhYYqJiTnPmdv/uBd0bMaMGSPpVCGalg/FN2nSJKWkpKh69epq2rSpr9MpkSZNmqTs7Gy1adNGn3/+eYn4fCkJ779t27Z59qOgopVdv4MBQElCwQo4B1JTUz3/d7lc+vXXX3X//fdr4cKF+vrrrzVy5EhNmDDhrLdz3XXX6brrrjvrOFYbOnSohg4d6us0iqx169a2+Ss1TvHXc8npdOraa6/Vxo0bFRERoREjRqhfv36qWrWqHA6HMjMztWHDBs2dO1dvvfVWnjEWLVp0nrP+h92Puy+PDWCVn3/+WZJ04403lohiVWlj1+9gAFCS0CUQOMcCAgLUoEEDff7556pdu7Yk6bXXXlN2draPMwNwrixevFjr16+XJE2bNk0jR45UYmKiZ5yv0NBQXXzxxRo1apT+/PNPtWrVypfpAvCB48ePS5KioqJ8nAkAAPZEwQo4T8LCwvTvf/9bknTkyBH99ttvkk51GZw9e7ZuueUWNWrUSPHx8QoLC1NSUpJuvvlmrVy5Mt+YRRms+qOPPlJycrIqVKiggIAAjR49WmvWrJHD4VBQUJDS0tJyxbj99ts942N8+eWXuZ5/77335HA4cnUbOtOgzPPnz1fPnj1VtWpVhYSEKDo6WjVr1lRycrKeffZZHTx4MM/1Tpw4oRdffFHt2rVTuXLlFBISoooVK6pHjx6aN29evsfpXNq2bZvnGG3btk0pKSkaPHiwqlWrprCwMNWqVUsjR47UsWPHPOts3LhRvXv3VmJiosLCwlSnTh099dRTOnny5Bm3l5WVpaefflqNGzdWZGSk4uLi1LlzZ3399ddnXPevv/7S3XffrXr16ikqKkoRERGqV6+e7rvvPm3fvr3AdX/77TfdcsstqlixosLCwlSzZk3dfffd2rNnz5kPkk6N1dKjRw+VK1dO4eHhqlu3rh599FEdPXq0wPUKOpf69+/vNfbPhx9+qPbt2ys+Pl4RERFq2rSpXnjhBblcrnzjG2M0Y8YMtWrVSmXKlFFMTIwuueQSTZ06VcaYXNsorJzj/lx77bUFLutwOPLsIlzQwOI5x605cOCAhg0bplq1aik8PFxJSUkaOnSo9u3b51k+JSVFd955p2rUqKGwsDBVq1ZNDzzwgI4cOZJnTsUdWD0jI0Off/65Bg8erKZNm6p8+fIKDQ1V5cqV1aNHjwLP09OvZ0uWLFGPHj1UqVIlBQYGer0GeR0b92vl1qFDB6/x59xxH374YTkcDjVo0KDAfUlPT1dUVJQtBndPS0vTf//7X11yySWKi4tTaGioEhMTddNNNxX4GfH7779rwoQJ6tSpk+f8iI6O1kUXXaSRI0dq//79udZxv/YpKSmSpAEDBuQay8/tfFz/zuYz8vTz+P3331e7du0UHx+vyMhINW/eXJMnT5bT6Szw+J/J+vXr1bdvXyUlJSksLExxcXFq3bq1Jk2apMzMzFzL5zxmUu5jbNX4TzmvE0eOHNHIkSN14YUXKjw8XGXLltXVV1+t//3vf/muf+jQIT3++ONq1qyZoqOjPZ+5jRs31h133OHV0rGw7z8p94061q9fr1tuuUVVq1ZVcHCw13WnMGN0FebGJP/73/80YMAA1a5dW5GRkYqOjlb9+vV16623asGCBZ7lqlevrg4dOuTavnvKeR0qzKDrf/31l+68807VqVPH8/5r1qyZnnjiCaWnp+e5zunHZ/Pmzbr11luVmJio0NBQVa1aVYMHD9auXbvy3S4AlBgGgCVGjRplJJmC3lYvv/yyZ5kffvgh13qSTFRUlAkNDfU8djgc5oUXXsgz3owZM4wkk5SUlG8+7dq1M8OGDfPEiouLM4GBgWbUqFHG6XSa2NhYI8l89tlnuWLUrl3bk8f999+f6/lBgwYZSaZv3775bvt0Y8aM8drfiIgIExUV5TVvyZIludb7448/TJ06dbyOS0xMjNd6d955Z57H6UzatWuXb75nsnXrVs/2P/roI8/xjI6ONoGBgZ7n2rZta7KyssyXX35pIiIijCQTExNjHA6HZ5levXoVmN+IESNM27ZtjSQTFBTk2ZZ7GjVqVL55Tp061QQHB3uWDQ0NNeHh4Z7H0dHRZsGCBXmu+/XXX3udk1FRUSYsLMxIMpUqVTLTp08v8NyfNm2aCQgI8CwTExNjQkJCjCRz4YUXmokTJxbqPD5dv379jCTTr18/c9dddxlJJiAgINdxOf38dMvOzja9evXyOqfi4uI8ud50001e2yiK8ePHe+L+8ccfRVrXLSkpyUgyM2bMyPWcO/abb75pqlataiSZyMhIz3GVZOrVq2cOHTpkVq1aZcqVK+d5nYOCgjzLtGnTxmRnZ+eKX9BxL+g59zXJPYWHh3vOd/f0wAMP5Lm/Oa9nL7zwgue9ERMTY4KDg71eg7yOzT333GMSEhI824mLizMJCQmeqUWLFsYYY7Zs2eKJ/d133+V7/KdMmeLZ/rFjx/Jd7nRLliwp8Fp2Ove+5HeOrVy50mu/AgMDTZkyZbzO27FjxxYY271cbGys1zWnSpUq5rfffvNaZ8KECSYhIcHzPoiOjvY6jgkJCZ5lz8f172w+I3Oeq8OHD8/zfS7JXHnllebEiRNnfK3y8vzzz3vth/t8dT9u3Lix+fvvv73WcR/H/I7x9u3bC7XtM51r7ufeffddz+d5WFiY13syODjYzJs3L9e6O3bsMNWqVfMsFxAQ4Pn+4J6X8xpQ2Pff6Xl/+OGHnuMVHR1twsLCvOIW5r3k/ozM6zMwOzvb3HPPPV7nUGRkpNcxiImJ8SzfokULExcX53nu9HP/nnvu8Sxb0HcwY4yZM2eO17lapkwZr8eJiYlm06ZNudbLeXwWL17s+Y5UpkwZr+t35cqVzc6dO/M9LgBQElCwAixSmILVQw895Fnm119/Ncac+lF0//33m5UrV5pDhw4ZY4xxuVxmy5Yt5t577zUOh8MEBgaadevW5YpXmIKV+4vO8OHDzd69e40xxpw4ccJs27bNGGPMNddcYySZe++912v9HTt2eL5ASjJNmjTJtY1atWrl+YM6vx+027Zt83xBHzZsmNm1a5fnucOHD5vvvvvODBkyxKxZs8ZrvUOHDpnq1asbSaZjx47m22+/9fy4OHz4sJk4caJnPydNmpQrzzOxqmAVGxtrrrjiCvPLL78YY4w5fvy4efHFFz1f8EeOHGliYmJMr169PMf/yJEj5tFHH/XEWLhwYb75xcTEmNDQUPPqq6+ajIwMY4wx27dvN9dff71n/bwKj5988onnh8nDDz9stm3bZlwul3G5XOa3334z//73vz2vdUpKite6O3bs8JwDjRs3Nv/73/+MMcY4nU7z9ddfm6pVq3oViE63du1azxfs9u3be877rKws895775nY2FjP+sUtWMXFxZmQkBAzceJEk5aWZowxZv/+/Z6CqiSzaNGiXOuPGzfO8/ywYcPM/v37jTHGpKWlmbFjx3p+2BZUTMjP0qVLPbE7duxYrB8VhSlYxcbGmqZNm5qVK1caY/45ru4fY0OHDjVJSUmmY8eOZuPGjcYYYzIyMsxLL73kOS9ff/31XPGLW7D65JNPzG233WaWLFniOZ7GGPP333+bMWPGeH6Y5nWeuq9nYWFhJjAw0PTv39/zwz07O9ts3ry5SMemoB+4Xbp0MVL+xUxjjGnWrJnnGBaFlQWrrVu3et4f119/vVm7dq05efKkMcaYPXv2mMcee8zz/vrkk09yrd+rVy/z0ksvmc2bN5vMzExjjDGZmZnmm2++MS1btjSSTLNmzQrMK69jnDO/c339O5vPSPe56v7jxtChQz2fg2lpaebJJ5/0FJvy+qPMmXzxxRee3K+99lqzZcsWY8ypYzxr1ixPYbF169Z5FoYLc4wLUtiCVVxcnKlfv75ZvHixcTqdxuVymVWrVpm6det6rr1Op9Nr3YEDBxpJpnr16uabb77x5J+dnW22bdtmpkyZYv7zn//ku82Czv2ceUdFRZmrrrrK89lgjPEq8p9twcpdqJRkbr31VvP77797ntuzZ4/59NNPcxVLc+ZXkIK+g61du9ZzvWvTpo356aefjDGnPjs///xzU6lSJSPJ1KpVyxw5ciTf7cfFxZlrrrnGc3wyMzPNnDlzPOdWnz59CswRAPwdBSvAImcqWKWlpZnKlSsbSSY+Pj7Xl8P8uFuODBw4MNdzhSlYuX+M52fSpElGkmnUqJHX/DfffNOz3SpVqhiHw+H1A3T79u2e+O4fH6dv+/QftHPmzDGSzAUXXFCIPf/Hgw8+6Pnh7/6xdrqPP/7YSDLlypXLd5n8uL/sBgcH5/pr6unT7NmzvdbN+YOtQYMGef6Vvk+fPp5lOnfubFwuV65l3C2n8nqd3flJMtOmTcv1vNPpNJdffrmRZOrXr+/1XGZmpqlSpUq+67rlV7i88847jSRTtmxZs2fPnlzr/fzzz16tCU7XtWtXz2t+/PjxXM/PmzfPs25xC1YF/eBr3ry5kWQGDRrkNf/YsWOeQlxexzzntotTsDLGmM6dO3vWDwwMNK1atTL33XefeeuttwrV6qowRZmEhASv96XbY489Vujz8oorrsj1XHELVmcyYcKEfLeZs3VWz549C4xztgWrTz/91EinWoC5iyA5rV271hNnw4YNZ9otL6f/2DzTNcVdxM/rHHMXowv6UepuoZjXHxUKcuTIEU+LmLxamhW1YHWurn9nUtBnZM73cH7HcOTIkUY61Wo15x9RCqN+/fpGkrnsssvyLEh9/vnnnu1/8MEHuZ4/XwWr8uXL53n93rBhg2eZ77//3uu5evXqGelU66yiKGrBqmXLlnkeu6LEy69g9fvvv3veX8OHDy/0PlhRsHIXxWvXrp1nC81169Z5is0TJkzId/sdOnTI8/viiy++6LmGFfU7DwD4E8awAs6xw4cPa9GiRerYsaP+/vtvSdK9996rgIDCvf26desmSfr++++Ltf2AgAD95z//yfd591gNGzdu9BrzZsmSJZKkjh07qkOHDjLGeOZJpwaVlqQaNWooKSmpULnExsZKOjWGV84xTQpijNH06dMlSQ888ICCgvK+uWmPHj0UHR2t/fv3a+3atYWKfbqTJ09qz549BU4ZGRn5rn///ffnORbRlVde6fm/e/yc/JbZsGFDvvETExM1YMCAXPMDAgI0cuRISdKmTZs8d56SpK+//lq7du1SQkJCnuu69e3bV9Kp8cXcjDGaM2eOJOmOO+5QhQoVcq3XsGFDXX/99XnGPHz4sCfeQw89pPDw8FzLXHnllWc94HhiYqIn/9Ndc801knIf1/nz53vGD3n00UfzXPeBBx5QREREsfP65JNPNGTIEAUHB8vpdGrFihWaNGmS+vTpowsuuEDVq1fXmDFj8h3HpDAGDx6ssmXL5pqf85wbNmxYgedlQeec1dzXsxUrVhQ4btCIESPOaR5XX321EhMTlZGRkeddGqdOnSpJatWqlRo1alTs7Rw6dOiM15T8xlg7ePCgPv74Y0mnrhv5cZ/7P/30U6HHlJNODfTdrl07ScX/fMnpXF//8lPYz8jHH388z/nua1N2drY++uijQm93w4YN2rRpkyTpscceU2BgYK5lunfvrpYtW0o6Neajr9x22215Xr8bNWqkGjVqSMp97N2f17t37z6nuT300EN5HjsrvPnmm3K5XCpbtqzGjBlzTraRl9M/+/L6HLnooovUs2dPSQWfG4888kie3xfdYyNmZGTozz//tCJtALAlClbAOZBzgM64uDh16tTJU0Tp3bt3rh/IW7Zs0YMPPqjmzZsrNjZWgYGBnvWvuuoqSdLOnTuLlUvt2rXz/KLq1qhRI5UrVy5XQer0gpX0T5Eq5/9zDk56Ji1btlS5cuW0e/duXXLJJZo8ebJ+++03GWPyXWfTpk2eQdj79++vihUr5jlVqlTJM4C3e7DgomrXrp3MqZan+U4FDb7t/mFyuoSEBM//L7744gKXOXToUL7x3QPL5uXyyy/3FPPWrFnjme/+EXfo0CFVqlQp3+M3ePBgSd7HbuvWrZ5j37Fjx3zzyu+5devWeX6MF2f9wrr44ovzLQBXrlxZknIN5L9u3TpJUrVq1Tw/2E5XpkwZNW/evNh5RUZG6uWXX9bOnTs1depU9enTR/Xq1fP8OEtJSdHo0aPVtGlT/fXXX8Xaxrk+54pjz549GjVqlFq1aqWyZcsqKCjIcz2rX7++pFN3R8tvu+Hh4WrWrJmlOZ0uMDBQgwYNkiS9/vrrXs8dO3bM8wPytttuO6vtLFmy5IzXlPwK/itWrPB6/+T33s05eHxe174vv/xSvXr1Us2aNRUZGen1+fT+++9LKv7nS07n8lw828/IxMREz116TxcdHe15n+e8dp6Je9mgoCBP4S8vnTt3LnJsq11yySX5PpffNfLqq6+WdKrIeNttt2nevHlnVVzPT5s2bSyP6bZ8+XJJp16DsLCwc7ad061bt87zvaZTp075Luc+NzZs2JDvTQfye+3cr5uU+7UDgJIk76YKAM5Kzi/ooaGhKleunC666CLdcsstuQo8n3zyiW666SavOwlFR0crLCxMDodDWVlZOnToUKFbJJ2uoGKVJM8dlD788EMtXrxYN9xwg7Zs2aKUlBTVq1dPFStW9BQUchas3AWtohSsYmNj9d577+nmm2/WL7/8orvvvluSFBMTo8svv1w33HCDevXqpeDgYM867lZpkrxagBXEfavw861MmTJ5zs/ZKuxMyxR0p8AqVark+1xoaKjKli2rPXv2aO/evZ757uOXlZVVqNYXOVuQ5YxT0LarVq2a5/yzXb+w8jumUv7H1X0u5fzSn5eC8i6sChUqaPDgwZ6i4NGjR7V48WJNmDBB33//vbZu3aobb7xRq1evLnJsK8657OzsIm83PytWrNBVV12lw4cPe+a570jpcDjkdDo9d6Y7duyYypUrlytG2bJlC90C9WwMGjRITz75pH7++WetXLlSl156qSRp9uzZSk9PV2xsrG644YZznkd+cl77CttyKue1z+VyqXfv3l6tN4KCghQXF6eQkBBJp+7Ad+LEiWJ/vuR0rq5/VnxGnul97H4+5zXrTNzLlitXLs+WZW7u61tRYlutONfIhx56SD/99JPef/99vf7663r99dc9d9fs0qWLBg8erAsuuOCsczvTd5SzkZqaKkmFbgVulaJ+9mVnZ+vgwYNe3x3dCvO+KswdhgHAX9HCCjgHUlNTPVNKSorWrl2rN954I1dx58CBA+rfv78yMzPVsWNHLV26VMePH1daWpr27Nmj1NRUffDBB2eVS2Ga2p/egsr9r7tQVb16ddWoUUO///67/v77b23evFk7duzwWrewOnXqpK1bt2rWrFnq16+f6tSpo7S0NH3xxRfq06ePLrroIq9bNefsNpSamnrG1gpnagXlz/JrXVUQ9/Hr0qVLoY5dfq3dirNtO3Pv55n2q6DWf8UVFRWla665RsuWLfO8f9asWaMff/zR8m2dT9nZ2brpppt0+PBhNW3aVHPnzlV6erqOHDniuZ6tXLnSs3x+x/ZcdQ86XeXKlT1dRt1dAKV/Wlz17t37rLqEni33ezc8PLzQ79327dt71p82bZree+89BQYG6vHHH9eff/6pzMxMHTx40PP55O7Oey7OcytY9Rl5Lq9fhY3tb9fQ4OBgzZkzRz/++KMef/xxdezYUREREdq4caOeffZZ1a9fX88999xZb+d8vN/94dj7Q44A4AsUrAAfcv+gi4uL0xdffKF27drlGufH/RfCc8ldmPrzzz+1c+fOXAUrybuo5X7+ggsuKFYLlMjISPXp00czZ87UH3/8oZ07d+qZZ55RWFiYV8srSapYsaLn/znHZiqNCurykpmZqQMHDkjy/ou1+/gV59jljFPQtnMWGPNbP79lzvTcueLOLWcrlryc6fmzERAQ4OmWJkm///77OdvW+bBixQqlpKQoMDBQX375pbp27ZqrdcD5uJ4VxR133CFJmjNnjtLT0/Xzzz/rf//7n6Sz7w54ttzv3YyMDG3evLnI68+ePVvSqZZkY8aMUe3atXO1XLPb63E6qz4jz9Tl0X0NKkprH/ey+/bt82r9ld+2y5cvX+jYdtKkSRONGTNGixYt0uHDh/XNN9/o8ssvl9Pp9LTCOpfcBa0TJ07ku0xaWlqe8ytVqiRJ2rZtm+V5FaSwn53u59wtHwEAuVGwAnzI3Uqpbt26+f4l/5tvvjnneVx44YWeL3aLFy/W0qVLFRAQ4PXX+pzdAovTHbAgVapU0fDhw/XAAw9IkhYuXOh5rmHDhoqOjpb0zw+w0mrZsmX5toT47rvvPF27WrRo4ZnvHh9k165dRR5YuUaNGoqPj5ckr/HNTpezq2hOzZo18/xALs7655J7jKSUlJR8f8wcPXq02AP4F1ZUVJTn/wV1K/IH7utZ+fLl8y1kn4/rmbulQmFaDXXq1Em1a9fW8ePH9c4773haV53tYOtWaN26tWdfinPtc78eF110UZ7PHz161FOcy4v7vevL1ldWfUbu2LEj33Hijhw54nmf57x2nol72ezsbC1btizf5dz55Td+lz8JCgrSFVdcoa+++kqhoaEyxuQ6/kV5/xWGu5DjPhdOd+TIEf366695Pte6dWtJp75TFFTwOl3Owm5x9iPnZ9+iRYvyXc597Jo0aeI1FAIA4B8UrAAfiomJkST98ccfeX6Z+vHHH/Xuu++el1zcxadXXnlFu3fvVpMmTTzFipzPuwtaOecVVkF/hZbk+ct5zi4CQUFBuvXWWyWduuPPmYouJXnw0e3bt+vNN9/MNd/lcmns2LGSpHr16nn90O7evbunGHnvvfeecXyvnMfP4XB4xvB59dVXPWMP5bRp0yZ9+OGHecaKjY1VcnKyJOnZZ5/N8xz/5ptvPAPjnk/JycmeQqj72J3u+eefL/Z4aBs3bixUy7FZs2Z5/p9fYcFfuK9n7rvfnW7nzp168cUXz3ke7tc15zha+XE4HLr99tslnbr2vf3225J837pKOtVKw30nsAkTJuiPP/4ocPnTr33u1yO/FjBPPvmkjhw5km+8ohzHc8XKz8gnn3wyz/nPPfecMjIyFBQU5LlrW2E0btzYcxOBp556Ks+7Xs6dO9dTFLzpppsKHdsOCvq8Dg0N9XxOn96lz+rzpkmTJpKU7x0cn3322Xxz7d+/vwIDA3XgwAGNGjWq0Nt074NUvP2IjY313PlywoQJeX6O/PTTT5598rdzAwDOJwpWgA8lJycrICBABw8e1C233OL5gZuVlaX3339fycnJBQ6WaiV38cn95fr0O7dVrlxZdevWVUpKiqcLRs4WWIXxzDPPqGvXrnrrrbe8mslnZmbq/fff14QJEyTJc9cnt8cee0y1atVSdna2unTpookTJ3oNwJ6WlqZ58+apX79+atu2bZFy8icxMTG688479frrr3t+vO3YsUM33XSTpwXTf//7X691wsLC9Morr8jhcGjdunVq06aN5s+fr6ysLM8yW7du1WuvvaaWLVvqlVde8Vp/xIgRKlOmjPbv36/OnTt77nRljNGCBQvUtWvXAsf5efLJJxUYGKjffvtN3bp183R7y87O1vvvv68bbrjBc/v08ykyMlL/+c9/JJ0as2j48OGeH/xHjhzRM888o9GjRxe7m8bSpUtVs2ZN9erVSx988IHXreFPnDih77//Xtdcc43nB8v1119/3gcGttpll12myMhIGWN0ww03eAosTqdT8+fPL/Aul1Zq2LChJOmdd94pVMFxwIABCg0N1caNG3Xo0CHFxsaqV69e5zrNQnnuuedUtmxZpaen67LLLtP06dO9uj/t379fH3/8sXr27JnrR2+XLl0knTq/p06d6nnPp6am6v7779f48eNVtmzZfLftPo4ffvih5XeSLCyrPiNjYmL05ptv6t577/UU3o8cOaKxY8d6Cll33XVXkbu4P/PMM5JOtXC9/vrrtXXrVkmnBsF+5513PK9J69at1aNHjyLF9rWkpCSNGDFCK1eu9CoIbd68WbfccouOHz+ugIAAT2HGrajvvzNxH8P58+dr1KhRnjsV7t+/X4888oieeuqpfD9DateurYceekiSNH78eA0aNEh//vmn5/l9+/Zpzpw5uu6667zWu+CCCzw3JnjjjTeK1crqv//9r4KDg7V582ZdeeWVnm75LpdLc+fO1VVXXaXs7GzVqlXLUzQHAOTBALDEqFGjjCRT1LfVf/7zH896kkxMTIwJDg42kkyNGjXMO++8k2/cGTNmGEkmKSkp33zatWtXqDz++usvrzy++uqrXMvccccdnufr16+fb6z8tp3zGEky4eHhJj4+3jgcDs+8evXqmd27d+eKuWXLFtOkSROv9WNjY010dLTXvNq1axdqf3Nq166dkWSCg4NNQkLCGaectm7d6tn21q1b84y/ZMmSM54bBb2W7vxGjBhhLrvsMk+ucXFxXvs+cuTIfOO//fbbJiIiwrNsUFCQKVu2rAkNDfWK8dRTT+Va98svv/RarkyZMiY8PNxIMpUqVTLTp08vcP9ee+01r9c4JibGE+/CCy80EydOLNZ53K9fPyPJ9OvXL9/9Lui4njx50lx//fWevAICAkxcXJwJDAw0kkyfPn1M3759jSRz++2357uNvLz66qtex1WSCQsLy/WaSTLJyckmLS0tV4ykpCQjycyYMSPXc+51lyxZkuf2z/a8LOi4F/TclClTvPYtKirKhIWFGUmmXLly5vPPP883r4Jeq9MVdGzeeustzzaCg4NNlSpVTFJSkmnTpk2+8Xr37u1ZZ+jQoWfcfkFyHtf8Xp+c3PuS33m8bt06U716dU9Mh8Nh4uLiTFRUlNex7tSpk9d6hw4dMhdeeKHX+R0bG+t5L95+++0FvoeWLVvmWTYwMNBUqlTJJCUleb0+5+P6dzafkTnP1eHDh3uOQ3x8vOd97j52GRkZ+eZXkIkTJ3pd32JjY01ISIjncaNGjcyuXbvyXLeg87gwznSuFeY8dH++jBo1Ks91c14b3e9l93n4/PPP54pXmPdfYc4Jt+zsbNOhQ4dc57/D4TAOh8NMmDAh331wr3/XXXflui7l/DyMiYnJtd7AgQM9z0dERJhq1aqZpKQk88ADD3iWOdM1a/bs2V7nQnR0tNcxTExMNJs2bcq1XmGPT1GuMwDgr2hhBfjY008/rVmzZqlly5YKDw/XyZMnVbt2bT3yyCNav369KleufF7yqFmzpqeFR1BQUJ4tlfIahL0obrvtNk2dOlU33XSTGjZsqIiICM+Aum3bttWkSZO0bt06r4HW3WrUqKE1a9Zo1qxZuvrqq1WpUiUdO3ZMWVlZqlGjhq677jpNnz5dK1asKHJebidPnvR0Zypo8pWQkBAtWrRIY8eOVd26dZWZmamYmBjPmCL5dXmRpFtuuUWbN2/WyJEj1aJFC0VFRenw4cMKCwtT06ZNNXToUH3zzTeeVkc5devWTevWrdONN96oChUqKCsrSwkJCRo6dKjWr1+vGjVqFJj3bbfdph9++EHdu3dXfHy8MjMzPX+9X7Vqlc8Gmw0KCtL777+vN954w/P+y87OVosWLfTGG29o1qxZnu4gRW0Fdvvtt+unn37SM888o2uvvVa1a9dWYGCg0tLSVKZMGdWvX199+/bV3LlzNX/+fK8uKP7sjjvu0FdffaX27dsrKipK2dnZqlKliu6++2799NNP52VcqN69e+utt97SZZddpoiICO3evVspKSkFDn7873//2/N/O3QHzOmiiy7Spk2bNHnyZHXq1EnlypXTkSNH5HK5VKdOHd18882aPXu2Pv74Y6/1YmNjtXz5ct13332qXr26AgMDFRQUpPbt2+u9997Tq6++WuB2L7/8cn311Vfq1KmTYmJitGfPHqWkpCglJeVc7m4uVn1GPvPMM5o9e7batGkjl8ulkJAQNW3aVC+88ILmzZunsLCwYuV3//33a82aNerdu7cSExN1/PhxhYeH69JLL9XEiRO1atWq8/Y5bqUFCxZoxIgRatu2rRITE5WRkSHpVKulAQMGaPXq1brvvvtyrVec919BAgMD9dVXX2nMmDG68MILFRISIofDoeTkZC1cuFAPPvjgGdefPHmyvv/+e91yyy2qVq2aTp48qZCQEDVo0EADBw7Ms7vhyy+/rNGjR3tajG3fvl0pKSl5do3PT69evfTLL7/o9ttvV61atZSZmamgoCA1bdpUY8aM0caNG1WvXr2iHRAAKGUcxtj0XsYAAPiQMUbVqlXTzp07NWvWLPXp08fXKeEcufvuuzV58mS1atXKJ2Oq4dwYPXq0xowZo3bt2nnGXgQAAP6DFlYAAOTBPdaa+85YKJnS09M9g9/feeedPs4GAAAAbhSsAACl1k033aQPP/zQq5vHnj179PTTT2vw4MGSpL59+/pllx6cWWZmpu69916lp6crMTHRNoOtAwAAQArydQIAAPjK119/rdmzZ0uSIiIiFBwc7HUXtrZt2+r555/3VXo4RyZNmqRJkyZp7969nrF5Jk6c6LkzGAAAAHyPFlYAgFLrxRdf1I033qi6desqNDRUx48fV/ny5dW5c2dNmzZNixYtKjEDouMfhw8fVkpKiowxatq0qebMmaPrr7/e12kBAAAgBwZdBwAAAAAAgK3QwgoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtkLBCgAAAAAAALZCwQoAgHNs5syZcjgcWrNmja9TKbH27NmjRx55RE2bNlV0dLRCQkJUtWpV9ezZU59//rmcTqevUwQAAEARBPk6AQAAgLOxcuVKXXPNNTLG6M4779Sll16qqKgobd++XV988YV69uyp1157TQMHDvR1qgAAACgkClYAAMBvHT58WD169FBUVJR++OEHVapUyev53r17a8OGDTpw4ECBcTIyMhQWFiaHw3Eu0wUAAEAh0SUQAAAbOHHihB544AE1bdpUMTExio+PV6tWrfTZZ5/lWtbhcGjo0KF66623VK9ePUVERKhJkyb68ssvcy372WefqXHjxgoNDVXNmjX1wgsvaPTo0V6FmW3btsnhcGjmzJl5bmv06NGex5s3b9aAAQNUp04dRUREqEqVKurevbt+/vnnXOv+8ssvSk5OVkREhMqXL6+77rpLX331lRwOh5YuXeq17DfffKMrrrhC0dHRioiIUJs2bbRo0aIzHrfXX39de/bs0fjx43MVq9waN26sDh06eB67u2guWLBAt956q8qXL6+IiAhlZmbK5XJp/PjxuvDCCxUaGqoKFSqob9++2rlzp1fM6tWrq3///rm21b59e7Vv397zeOnSpXI4HHr77bc1bNgwVaxYUeHh4WrXrp3Wr1/vte6WLVt04403qnLlygoNDVVCQoKuuOIK/fjjj2c8DgAAACUNLawAALCBzMxMHTx4UA8++KCqVKmirKwsffPNN+rZs6dmzJihvn37ei3/1VdfafXq1XriiScUFRWl8ePH67rrrtPvv/+umjVrSpLmzZunnj176vLLL9ecOXOUnZ2tZ599Vnv27Cl2nn///bfKli2rp59+WuXLl9fBgwf15ptv6pJLLtH69etVt25dSdLu3bvVrl07RUZGasqUKapQoYLee+89DR06NFfMt99+W3379tW1116rN998U8HBwXrttdd05ZVXav78+briiivyzWfhwoUKDAzUVVddVeR9ufXWW9WtWze99dZbOnbsmIKDg3XnnXdq6tSpGjp0qK6++mpt27ZNjz32mJYuXap169apXLlyRd6OJD3yyCNq1qyZ3njjDaWlpWn06NFq37691q9f73m9rrrqKjmdTo0fP17VqlXT/v37tXz5ch0+fLhY2wQAAPBnFKwAALCBmJgYzZgxw/PY6XTqiiuu0KFDhzRp0qRcBauMjAx98803KlOmjCSpWbNmqly5st5//309/PDDkqTHH39cVapU0fz58xUSEiJJ6tKli6pXr17sPC+//HJdfvnlXnl269ZNDRo00GuvvaaJEydKkp5//nkdPHhQ3377rerXry9J6tq1q7p06aJt27Z51j9+/LjuvfdeXX311frkk08886+66io1a9ZMjzzyiP73v//lm8+OHTs8LaRycrlccrlcnscBAQEKCPBuWH7FFVfotdde8zz+7bffNHXqVA0ZMkQvvfSSZ/5FF12kSy65RM8//7z++9//FuYw5VK+fHl98sknnpZtl112merUqaNx48bp9ddf14EDB/T7779r0qRJ6t27t2e9nj17Fmt7AAAA/o4ugQAA2MQHH3ygNm3aKCoqSkFBQQoODta0adP066+/5lq2Q4cOnmKVJCUkJKhChQpKSUmRJB07dkxr1qxRjx49PMUqSYqKilL37t2LnWN2drbGjh2r+vXrKyQkREFBQQoJCdGff/7pleeyZcvUsGFDT7HK7aabbvJ6vHz5ch08eFD9+vVTdna2Z3K5XOrSpYtWr16tY8eOFTnPYcOGKTg42DNdc801uZb517/+5fV4yZIlkpSrq1/Lli1Vr169QnVRzM/NN9/s1Q0zKSlJrVu39mwzPj5etWrV0oQJEzRx4kStX7/eq+AGAABQ2lCwAgDABj7++GPdcMMNqlKlit5++22tWLFCq1ev1q233qoTJ07kWr5s2bK55oWGhiojI0OSdOjQIRljlJCQkGu5vOYV1rBhw/TYY4+pR48e+uKLL/S///1Pq1evVpMmTTzblqQDBw4Uatvu7onXX3+9V4EpODhYzzzzjIwxOnjwYL75VKtWTfv27dPx48e95j/wwANavXq1Vq9ene/YVqfPdw/MntfylStXPuPA7QWpWLFinvPcMR0OhxYtWqQrr7xS48ePV7NmzVS+fHndc889OnLkSLG3CwAA4K/oEggAgA28/fbbqlGjhubMmePVEiczM7NY8eLi4uRwOPIcryo1NdXrcVhYWJ7byqtA4x5vauzYsV7z9+/fr9jYWM/jsmXLFmrb7jGhXnrpJV166aV57ktBBbbOnTtrwYIFmjt3rq6//nrP/MTERCUmJkqSVwuznE6/I6C7CLh7925VrVrV67m///7ba/yqsLCwPF+b/fv35znO1en77Z6Xs/CYlJSkadOmSZL++OMPvf/++xo9erSysrL06quv5rkPAAAAJRUtrAAAsAGHw6GQkBCvIkpqamqedwksjMjISLVo0UKffvqpsrKyPPOPHj2a626CCQkJCgsL04YNG7zm53eHwtDQUK95X331lXbt2uU1r127dtq4caM2bdrkNX/27Nlej9u0aaPY2Fht2rRJLVq0yHPKr+AkSYMGDVJCQoKGDx+u3bt357tcYXTs2FHSqaJcTqtXr9avv/7qNfh79erVcx2vP/74Q7///nuesd977z0ZYzyPU1JStHz5cq87CuZ0wQUXaOTIkWrUqJHWrVtXnN0BAADwa7SwAgDgPFm8eLHXgONuV111la6++mp9/PHHGjJkiK6//nrt2LFDTz75pCpVqqQ///yzWNt74okn1K1bN1155ZW699575XQ6NWHCBEVFRXl1s3M4HOrdu7emT5+uWrVqqUmTJlq1apXefffdXDGvvvpqzZw5UxdeeKEaN26stWvXasKECblaJN13332aPn26unbtqieeeEIJCQl699139dtvv0mSZwD0qKgovfTSS+rXr58OHjyo66+/XhUqVNC+ffv0008/ad++fZoyZUq++xgbG6tPP/1U3bt3V5MmTXTnnXfq0ksvVVRUlA4cOKBvv/1Wqampat269RmPV926dXXbbbfppZdeUkBAgLp27eq5S2BiYqLuv/9+z7J9+vRR7969NWTIEP3rX/9SSkqKxo8fr/Lly+cZe+/evbruuus0ePBgpaWladSoUQoLC9OIESMkSRs2bNDQoUP173//W3Xq1FFISIgWL16sDRs2eAbRBwAAKFUMAAA4p2bMmGEk5Ttt3brVGGPM008/bapXr25CQ0NNvXr1zOuvv25GjRplTv+4lmTuuuuuXNtJSkoy/fr185r3ySefmEaNGpmQkBBTrVo18/TTT5t77rnHxMXFeS2XlpZmBg0aZBISEkxkZKTp3r272bZtm5FkRo0a5Vnu0KFDZuDAgaZChQomIiLCXHbZZea7774z7dq1M+3atfOKuXHjRtOpUycTFhZm4uPjzcCBA82bb75pJJmffvrJa9lly5aZbt26mfj4eBMcHGyqVKliunXrZj744INCHePU1FQzYsQI07hxYxMZGWmCg4NN5cqVTffu3c2sWbPMyZMnc70eq1evzhXH6XSaZ555xlxwwQUmODjYlCtXzvTu3dvs2LHDazmXy2XGjx9vatasacLCwkyLFi3M4sWLcx2HJUuWGEnmrbfeMvfcc48pX768CQ0NNW3btjVr1qzxLLdnzx7Tv39/c+GFF5rIyEgTFRVlGjdubJ5//nmTnZ1dqGMAAABQkjiMydE+HQAAlGgnT55U06ZNVaVKFS1YsOC8b/+2227Te++9pwMHDhTY1a+kWLp0qTp06KAPPvjAa4wtAAAAFIwugQAAlGADBw5U586dValSJaWmpurVV1/Vr7/+qhdeeOGcb/uJJ55Q5cqVVbNmTc/YWW+88YZGjhxZKopVAAAAKD4KVgAAlGBHjhzRgw8+qH379ik4OFjNmjXT3Llz1alTp3O+7eDgYE2YMEE7d+5Udna26tSpo4kTJ+ree+8959sGAACAf6NLIAAAAAAAAGwlwNcJAAAAAAAAADlRsAIAAAAAAICtMIZVEbhcLv39998qU6aMHA6Hr9MBAAAAAMCLMUZHjhxR5cqVFRBAGxX4LwpWRfD3338rMTHR12kAAAAAAFCgHTt2qGrVqr5OAyg2ClZFUKZMGUnSjh0zFB0d4eNs7GXyprJnHWNo/QMWZGINK/ZHsmaf7JSLVThfcitp+yPZa5+AwuL8P7esOr5W4DWCP/p2cR1L4myoeNCSOFbgvWi99PTjSkwc4Pn9CvgrClZF4O4GGB0dQcHqNGFRkWcdIzo6w4JMrGHF/kjW7JOdcrEK50tuJW1/JHvtE1BYnP/nllXH1wq8RvBHkZFRlsQJi8q0JI4VeC+eOwxjA39Hh1YAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2EqQrxMAAADF43B0L9RyS5aMVfv2jSzZ3l13ddPkyXcUa/3MzJN6440Fev/977Rx43alpx9XdHSEGjVK0k03Xa6bb26nMmUizjpPAAAA+D8KVgAA+KkVKyZ4PX7yyTlasmSDFi/+r9f8+vWrnc+08rRvX5q6dBmljRtT1K/fFbrnnu6qUCFWBw6ka/HiDRo+fKa+/36T3nrrAV+nCgAAABugYAUAgJ+69NILvR6XLx+tgICAXPPtoHfv5/Tzzyn65psndfnlDb2e69GjlUaNuklff722wBhOp1PZ2S6Fhgafy1QBAABgA4xhBQBACfbyy1/p8ssfVoUKvRUZeb0aNRqq8eM/0smT2V7LrV//l66+eowqVOit0NDrVLlyP3XrNkY7d+7PN7YxRo88MkvBwT30+uvz811u9eo/tGDBet1225W5ilVuZctGq3fvDp7H27btkcPRXePHf6SnnpqjGjUGKjS0p5Ys2SBJ+vzz/6lVqwcVEfEvlSlzgzp3fkwrVvzmFbN//+dVvfrAXNsaPfrdXN0pHY7uGjr0Va14/xONu/oGDW92ucZfe5PWf73Qa7msjBP6/NkX9d8uPfWf5u00sk2ynu81QOvmLsh3/wEAAFB0tLACAKAE++uv3br55naqUSNBISFB+umnrfrvf9/Xb7/t1PTp90qSjh07oc6dH1ONGgl6+eU7lJAQq9TUQ1qy5GcdOZKRZ9zMzJPq33+Svvpqtb744jF16dI83xwWLvxRknTNNS2LnP+LL36hCy6oomefvVXR0RGqU6ey3n13qW655TklJ1+k9957SJmZJzV+/Mdq336EFi16Spdd1qDI25FOFcGyQjapy12DFRIepuWzP9bbwx9XQGCgmiR3PLXMhBe09st56jL0NlWpV1dZGRlK/XOLjh9OK9Y2AQAAkDcKVgAAlGATJw7y/N/lcqlt2wYqW7aMBgx4Qc89N1BxcVH67bedOnDgiKZNu0fXXnupZ/kbbmibZ8yDB4/o2muf0tate/Tdd8+oSZMaBeawY8epVlpJSRW85htj5HS6PI8dDikwMNBrmbCwEM2fP0bBwUGefWjTZrgaNaqur78erYCAU43Fr7qqhWrVuk3/+c+b+uGH8Wc6LHnavz9d/5n7psqUi5ck1WvbWhOuu0VzX5jiKVhtXb9BF7RqqXZ9b/KsV//yNsXaHgAAAPJHl0AAAEqw9ev/0jXXPKmyZW9WYOC1Cg7uob59n5fT6dIff+ySJNWuXUlxcVH6z39m6tVXv9amTdvzjbd16x61avWQ0tOPa+XKZ89YrCrIZ5+tVHBwD88UE3NjrmWuuaalp1glSb//vkt//31Qffp08BSrJCkqKlz/+lcrrVz5u44fP1GsfK64oomnWCVJAYGBatqlk/Zv36nDqXslSdUa1ddv36/Ul8+/os2r1+nkieJtCwAAAAWjYAUAQAm1fftetW37sHbtOqAXXhis7757WqtXT9TLL98hScrIyJIkxcREatmycWratKYeeWSWGjS4S5Ur99OoUe/kGutq1ao/9Mcfu9SrV1tVrVquUHlUq1ZekpSSstdrfvv2jbR69UStXj1RV199cZ7rVqoU7/X4wIH0/58fl2vZypXLyuVy6dChY4XK63QVK+aO6S5gHU871eWvx8PD1OHW3tq4eJmm3HqXRl52pabf8x/tS9lRrG0CAAAgb3QJBACghPr005U6duyEPv74Ea/ueD/+uCXXso0aVdfs2cNljNGGDds0c+Y3euKJ2QoPD9HDD//bs1yvXm1VsWKcHn30LblcRiNH9jpjHp07N9Ujj8zS55+vUnJyM8/82NgotWhRR5JUtmyZPNd1OLwfly0bLUnavftQrmX//vuAAgICFBcXKelUd8LMzJO5ltu/Pz3PbaWmHlK90+Yd2X9QkhQREyNJCo0IV5e7BqvLXYN1ZP9B/fr9Cs2d9IqmDX1QD38xJ8+4AAAAKDpaWAEAUEI5/r/aExoa7JlnjNHrr+d/RzuHw6EmTWro+ecHKzY2UuvW/ZVrmZEje2nSpMF6/PF3NGLEm2fMo0WLOkpOvkivvz5f3333SzH25B9161ZRlSpl9e67y2SM8cw/duyEPvpohVq1qquIiDBJUvXqFbR3b5r27PmnuJWVdVLz56/LM/aiRT95ClSS5HI69eO8b1Q2sYpiK1bItXyZcvFq2aObLuraWfu2bVdWBt0DAQAArEILKwAASqjOnZsqJCRIN900QcOH99SJEyc1ZcpcHTp01Gu5L79cpVdemasePS5VzZoVZYzRxx+v0OHDx9S580V5xr733msUFRWm2257WUePntCLL97mKZDl5e23H9CVVz6uTp1Gqn//K3Tllc1UoUKM0tMztGHDVn3zzU+Kjo444z4FBARo/Pj+uuWW53T11U/o9tu7KDPzpCZM+FiHDx/V00/38yzbq1dbPf74u7rxxgl66KGeOnEiSy+++IXXQO85lSsXrSmDhqrz7QM8dwncuzVFvSc86VnmhZsHqv7lbVTpgtoKjy6jvVu2ac2X85TUpKFCwsPOmD8AAAAKh4IVAAAl1IUXJuqjj0Zo5Mi31bPnOJUtW0Y339xOw4b1UNeuoz3L1alTWbGxkRo//mP9/fcBhYQEq27dKpo58z7163dFvvEHDkxWZGSY+vSZqGPHTuiNN+72Ggg9p/LlY7RixbN6/fX5mjPnO73//vc6evSEoqMj1KBBNd1zT3cNGpRcqP26+eb2iowM07hxH6pXr/EKDAzQpZfW1ZIlY9W69T+d+mrUqKjPPntUjzzylq6/fpwqVYrXsGE9tG9fmsaMeS9X3GuuuUSpcfX19Uuv6fDuPSqbWEW3PD1aF3Xp5Fmmdsvm+mXp91r21mydPJGpmArl1aJ7V3W6rX+hcgcAAEDhOEzO9vQoUHp6umJiYpSWNqdQfwUuTSZuLNzAuwUZ1nC/BZlYw4r9kazZJzvlYhXOl9xK2v5I9ton4Ewcju66665uqnnHSEvicf7nzarrixV4jeCPliysa0mc9ZUOWBLHCrwXrZeeflwxMb2Ulpam6OhoX6cDFBtjWAEAAAAAAMBWKFgBAAAAAADAVhjDCgAAlHrGfCFJmrjRx4kAAABAEi2sAAAAAAAAYDMUrAAAAAAAAGArdAksAqfTKUnauXM/dwk8zeE9rrOOsTP2oAWZWMOK/ZGs2Sc75WIVzpfcStr+SPbaJ6CwOP/PLauOrxV4jeCP9u2LsSTO4YBDlsSxAu9F66WnH5f0z+9XwF85jDHG10n4i9WrV6tly5a+TgMAAAAAgAKtWrVKF198sa/TAIqNglURHDp0SPHx8Xps4pcKC4/0dTolzp0Hp/k6BcvtbdHE1yl4fF/2e1+nYKk9q+/zdQqW+rP6Zl+nAMBCj6xOPOsYH5QPtSAT6V8115x1jID/NbIgE+QnoduDZx3DOa/p2SdSAgV2+dHXKVhq3ubHLInzdXS6JXHswoprbkly9MQxNRv7Lx08eFBxcXG+TgcoNroEFkFgYKAkKSw8UmHhUT7OpuSJDgvxdQqWy4gM93UKHhFRwb5OwVIl7T0YEkk3Y6AkKRN29n/YCgsPsyATqYwFn0UBFuwP8hcdFXjWMZwl8HuUFQItOLZ2EhFhzfeFkMhsS+LYhRXX3JLI/fsV8FcMug4AAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAFBCfPvtt+revbsqV64sh8OhTz/99IzrLFu2TM2bN1dYWJhq1qypV1999dwnegYUrAAAAAAAAEqIY8eOqUmTJpo8eXKhlt+6dauuuuoqtW3bVuvXr9cjjzyie+65Rx999NE5zrRgQT7dOgAAAAAAACzTtWtXde3atdDLv/rqq6pWrZomTZokSapXr57WrFmjZ599Vv/617/OUZZnRgsrAAAAAACAUmrFihVKTk72mnfllVdqzZo1OnnypI+yooVVkTidTknSiYxjPs6kZEo/keXrFCx35FiGr1PwOB7muwvNuXAi46ivU7BU1rHjvk4BgIWOnDj77wonMrItyMSaz6IAC/YH+Ys46jzrGM4S+D3KCoEWHFs7OX7cmu8LWUEl63uHFdfckuTo/x8P9+9X/OPEiRPKyvLP66UxRg6Hw2teaGioQkNDzzp2amqqEhISvOYlJCQoOztb+/fvV6VKlc56G8VBwaoINm/eLEl6ctjVPs6kZHrU1wmcE2/7OoES7DNfJwAA+Zrp6wRyKJmfryXM41YE2WhFkJKnxL0Bevk6AVua6esEbGrz5s26+OKLfZ2GbZw4cUI1kqKUutc/C3lRUVE6etT7j/ajRo3S6NGjLYl/ejHMGJPn/POJglUR1K5dW5L087JERZcJtCRm62U3WhJHkqq/Y+0bb9st1uyj2/J2sy2NZyUrXwcA/iN2dbCl8Q5fbG1LxtovH7Es1t7W8ZbFslr5NemWxjtRIcLSeGF7rWuJ4IwIsSyWJLmCrR3dYXdfa1p1SVLV1y0LJUna1S7S0njVPttnWSyTal0sSQqIibY0ngk/+7++uzmOWtt6/GjjypbGCzuYaVksZ6i1P5UONLTudag8b7dlsSRJ///D1CrOsmUsi7XjyijLYklS9Zk7LI23v32iZbHiNlnTg+Bk9gmt/G2K5/crTsnKylLqXqdS1lZXdBn/Gh0p/YhLSc23aceOHYqO/uczworWVZJUsWJFpaames3bu3evgoKCVLZsWUu2URwUrIogMPBUAadqpWDLTvCg+BhL4khSWJi1BaugeGsLVlUr2fd0s/J1AOA/QqKsLR4ExVvbxDwsxLq/aIVExVoWy2ph1r4MUri1P27CQqz7/HKGWfeDVbK+YBVU1rqClcW7quAYi1/XYOsKGybA2u5VAUHW/diXJBMcZlksR5C13w+zI2ItjRd27IRlsZxh1n53DSlj3esQFmTx0AhWF6xCrPtuHRRj7fshLOiQpfGs/HwNC7HoN2bgqQuw+/crvEWXCbCsAcr5Fh0d7VWwskqrVq30xRdfeM1bsGCBWrRooeBga//AWxT2rSAAAAAAAABYyCUjl1y+TqNIXCpaUfno0aOeIY0kaevWrfrxxx8VHx+vatWqacSIEdq1a5dmzZolSbrjjjs0efJkDRs2TIMHD9aKFSs0bdo0vffee5buR1FRsAIAAAAAACgh1qxZow4dOngeDxs2TJLUr18/zZw5U7t379b27ds9z9eoUUNz587V/fffr5dfflmVK1fWiy++qH/961/nPfecSl3B6pVXXtGECRO0e/duNWjQQJMmTVLbtm19nRYAAAAAAMBZa9++vWfQ9LzMnDkz17x27dpp3bp15zCrovOvkcbO0pw5c3Tffffp0Ucf1fr169W2bVt17drVq7IIAAAAAAAA3ypVBauJEydq4MCBGjRokOrVq6dJkyYpMTFRU6ZM8XVqAAAAAADgHHMal19OpVGpKVhlZWVp7dq1Sk5O9pqfnJys5cuX57lOZmam0tPTvSYAAAAAAACcW6WmYLV//345nU4lJCR4zU9ISFBqamqe64wbN04xMTGeKTEx8XykCgAAAAAAUKqVmoKVm8Ph8HpsjMk1z23EiBFKS0vzTDt27DgfKQIAAAAAAJRqpeYugeXKlVNgYGCu1lR79+7N1erKLTQ0VKGhoecjPQAAAAAAcI65ZORS/nfQsyN/y9cqpaaFVUhIiJo3b66FCxd6zV+4cKFat27to6wAAAAAAABwulLTwkqShg0bpj59+qhFixZq1aqVpk6dqu3bt+uOO+7wdWoAAAAAAAD4f6WqYNWrVy8dOHBATzzxhHbv3q2GDRtq7ty5SkpK8nVqAAAAAADgHHPJJZevkygi/8vYGqWqYCVJQ4YM0ZAhQ3ydBgAAAAAAAPJRasawAgAAAAAAgH+gYAUAAAAAAABbKXVdAgEAAAAAQOnkNEZOY3ydRpH4W75WoWBVDI0X91FAeJglsbYkT7MkjiQ1+MPasbm2JL9iabyaCwZaGs9KVr4OAPxHy6V3WhrP6mtJt5HdLIsVlGFZqFMs/N4UsPeQdcEkhTmt/VIXcOCwZbEcJ6MtiyVJJsjaxvIOR7B1sTKtHSDWYfF4s2ZXqrUBLZS9c5el8QLLlbUslgkMtCyWJIUezLI0XmD6CctiBWRYu6+SNb8fJEnZTutiSTLHjlsaT+VjLAsVdMxhWSxJMidPWhov/KB1r0VAVrY1cZzWxAF8jS6BAAAAAAAAsBUKVgAAAAAAALAVugQCAAAAAIBSwSUjl5VjG5wH/pavVWhhBQAAAAAAAFuhYAUAAAAAAABboUsgAAAAAAAoFVwycvpZFzu6BAIAAAAAAAA2QMEKAAAAAAAAtkLBCgAAAAAAALbCGFYAAAAAAKBUcMn43ZhQ/pavVWhhBQAAAAAAAFuhYAUAAAAAAABboWAFAAAAAAAAW2EMKwAAAAAAUCo4jZHT+NeYUP6Wr1VoYQUAAAAAAABboWAFAAAAAAAAW6FLIAAAAAAAKBVc/z/5E3/L1yq0sAIAAAAAAICt0MKqGGrOdCooyGlJrAZ/DLEkjiT9MvQVy2JJUoPJ1uUmSbWXHbc0npWsfB0A+I9qPx62NJ7V183KtTIsixWdkmlZLEkKzMi2LNaJ+lUsiyVJR6qGWBqvzM5Iy2I5Q639W6HD4j+5VpuUZV2wQGsHiK0+5XdL4xmnNd/lJEkXVLculqSgI9a99yXJBAVaFiurSqxlsSQp6Ii116bsmHDrYkVa+1Op0rdplsU61qSyZbEkKeSQhe99SSfKh1oWq+qSo5bFkqSTDapZGi/ib+ver9kxYdbEse4jGvApWlgBAAAAAADAVmhhBQAAAAAASgWnjJyythXwueZv+VqFFlYAAAAAAACwFQpWAAAAAAAAsBUKVgAAAAAAALAVxrACAAAAAAClgtOcmvyJv+VrFVpYAQAAAAAAwFYoWAEAAAAAAMBW6BIIAAAAAABKBdf/T/7E3/K1Ci2sAAAAAAAAYCsUrAAAAAAAAGArFKwAAAAAAABgK4xhBQAAAAAASgWXHHLK4es0isTlZ/lahRZWAAAAAAAAsBUKVgAAAAAAALAVClYAAAAAAACwFcawAgAAAAAApYLLnJr8ib/laxVaWAEAAAAAAMBWKFgBAAAAAADAVugSWAxb+gcqIDzQmljJr1gSR5IaTB5iWSxJ+mWodblJUs0LBloaz0pWvg4A/EfLv++0NJ7V181us7pZFmtfp2qWxZIkmVDLQpVfsNWyWJIUsifW0ngBBw5bFsvERVsWS5JMkLV/e9w6MtiyWNXHWdt/YduddS2NlzTxJ+uC/bXDuliSso8dszReYLmylsUKST9qWSxJyq5d2dJ4QYeOWxYr8Lg13/fd/u4YZ1msqh+kWBZLkswx646bJEVkVbEs1o7O1l43k17/3dJ4R9vUsixW1OY0awI5M62JA/gYBSsAAAAAAFAqOOWQUw5fp1Ek/pavVegSCAAAAAAAAFuhYAUAAAAAAABboUsgAAAAAAAoFegS6D9oYQUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFthDCsAAAAAAFAquIxDLuNfY0L5W75WoYUVAAAAAAAAbIWCFQAAAAAAAGyFghUAAAAAAABshTGsAAAAAABAqeCUQ07515hQ/pavVWhhBQAAAAAAAFuhYAUAAAAAAABboUsgAAAAAAAoFZwKkNPP2u44fZ2Aj/jXqwQAAAAAAIASj4IVAAAAAAAAbIWCFQAAAAAAAGyFMawAAAAAAECpYIxDLuPwdRpFYvwsX6tQsCqGDR3fUnQZaxqn1Vww0JI4klR72XHLYklSzQusy02StiRPszSelax8HQD4j/hga+NZfS25sEy6ZbGywy0LZTlXhThL452oFGVpvLBA674kOqNCLYslSa5gaxvLG5NtXazQQMtiSZKxuF+Ao0pFy2KZXamWxZKkoKpVLI1nIsIsi+U4csyyWJKUGR9iaTyH02VZLGeYjX8qBVn7/nJEW3vdtFJ2pLE0niPY2g//jHjrXouIEGvOOZeztA7RjZKGLoEAAAAAAACwFQpWAAAAAAAAsBUbt3MFAAAAAACwjlMOOeVfY0L5W75WoYUVAAAAAAAAbIWCFQAAAAAAAGyFLoEAAAAAAKBUcJoAOa2+9ew55rT2Zpl+w79eJQAAAAAAAJR4FKwAAAAAAABgKxSsAAAAAAAAYCuMYQUAAAAAAEoFlxxy+VnbHZdK5yBW/vUqAQAAAAAAoMSjYAUAAAAAAABboWAFAAAAAAAAW2EMKwAAAAAAUCo45ZBTDl+nUST+lq9VaGEFAAAAAAAAW6FgBQAAAAAAAFuhSyAAAAAAACgVnCZATuNfbXecxvg6BZ/wr1cJAAAAAAAAJR4FKwAAAAAAANgKBSsAAAAAAADYCmNYAQAAAACAUsElh1xy+DqNIvG3fK1CCysAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKY1gBAAAAAIBSwaUAOf2s7Y5Lxtcp+IR/vUoAAAAAAAAo8ShYAQAAAAAAwFYoWAEAAAAAgFLBaQL8ciqOV155RTVq1FBYWJiaN2+u7777rsDlX375ZdWrV0/h4eGqW7euZs2aVaztWoUxrAAAAAAAAEqQOXPm6L777tMrr7yiNm3a6LXXXlPXrl21adMmVatWLdfyU6ZM0YgRI/T666/r4osv1qpVqzR48GDFxcWpe/fuPtgDWlgBAAAAAACUKBMnTtTAgQM1aNAg1atXT5MmTVJiYqKmTJmS5/JvvfWWbr/9dvXq1Us1a9bUjTfeqIEDB+qZZ545z5n/g4IVAAAAAABACZGVlaW1a9cqOTnZa35ycrKWL1+e5zqZmZkKCwvzmhceHq5Vq1bp5MmT5yzXglCwAgAAAAAApYJLAX45SVJ6errXlJmZmec+7t+/X06nUwkJCV7zExISlJqamuc6V155pd544w2tXbtWxhitWbNG06dP18mTJ7V//35rX4RComAFAAAAAABgc4mJiYqJifFM48aNK3B5h8Ph9dgYk2ue22OPPaauXbvq0ksvVXBwsK699lr1799fkhQYGGhJ/kVFwQoAAAAAAMDmduzYobS0NM80YsSIPJcrV66cAgMDc7Wm2rt3b65WV27h4eGaPn26jh8/rm3btmn79u2qXr26ypQpo3Llylm+L4VBwQoAAAAAAMDmoqOjvabQ0NA8lwsJCVHz5s21cOFCr/kLFy5U69atC9xGcHCwqlatqsDAQM2ePVtXX321AgJ8UzoK8slWAQAAAAAAzjOncchp8u4WZ1fFyXfYsGHq06ePWrRooVatWmnq1Knavn277rjjDknSiBEjtGvXLs2aNUuS9Mcff2jVqlW65JJLdOjQIU2cOFEbN27Um2++aem+FAUFKwAAAAAAgBKkV69eOnDggJ544gnt3r1bDRs21Ny5c5WUlCRJ2r17t7Zv3+5Z3ul06rnnntPvv/+u4OBgdejQQcuXL1f16tV9tAcUrAAAAAAAAEqcIUOGaMiQIXk+N3PmTK/H9erV0/r1689DVoVHwaoYGi/uo4DwMEtibUmeZkkcSWrwR94nYnFtSX7F0ng1Fwy0NJ6VrHwdAPiPlkvvtDSe1deSbiO7WRYrKMOyUKcY60IF7D1kXTBJYU4Lk5MUcOCwZbEcJ6MtiyVJJsjaMSUcjmDrYmW6LIslSQ5rw8nsyvu23naQvXOXpfECy5W1LJax+E5RoQezLI0XmH7CslgBGVbfFcua3w+SpGyndbEkmWPHLY2n8jGWhQo6Zm3XLXPypKXxwg9a91oEZGVbE8dpTZySyqkAOf1sOG+nlV+8/Ih/vUoAAAAAAAAo8ShYAQAAAAAAwFYoWAEAAAAAAMBWGMMKAAAAAACUCi4TIJfxr7Y7LsMYVgAAAAAAAIDPUbACAAAAAACArVCwAgAAAAAAgK0whhUAAAAAACgVnAqQ08/a7jjFGFYAAAAAAACAz1GwAgAAAAAAgK3QJRAAAAAAAJQKLklO4/B1GkXi8nUCPkILKwAAAAAAANgKBSsAAAAAAADYCgUrAAAAAAAA2ApjWAEAAAAAgFLBpQC5/Kztjr/la5XSudcAAAAAAACwLQpWAAAAAAAAsBUKVgAAAAAAALAVxrACAAAAAAClgtMEyGn8q+2Ov+VrldK51wAAAAAAALAtClYAAAAAAACwFboEAgAAAACAUsElh1xy+DqNIvG3fK1CCysAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKY1gBAAAAAIBSwWkC5DT+1XbH3/K1SuncawAAAAAAANgWBSsAAAAAAADYCgUrAAAAAAAA2ApjWAEAAAAAgFLBqQA5/aztjr/la5XSudcAAAAAAACwLQpWAAAAAAAAsBUKVgAAAAAAALAVxrACAAAAAAClgss45DIOX6dRJP6Wr1VoYQUAAAAAAABboWAFAAAAAAAAW6FLIAAAAAAAKBVcCpDTz9ruuPwsX6uUzr0GAAAAAACAbVGwAgAAAAAAgK1QsAIAAAAAAICtMIYVAAAAAAAoFVwmQC7jX213/C1fq5TOvQYAAAAAAIBtUbACAAAAAACArVCwAgAAAAAAgK0whhUAAAAAACgVnHLIKYev0ygSf8vXKrSwAgAAAAAAgK1QsAIAAAAAAICt0CUQAAAAAACUCi4TIJfxr7Y7/pavVUrnXgMAAAAAAMC2KFgBAAAAAADAVihYAQAAAAAAwFYYwwoAAAAAAJQKTklOOXydRpE4fZ2Aj9DCCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtsIYVgAAAAAAoFRwmQC5jH+13fG3fK1SOvcaAAAAAAAAtkXBCgAAAAAAALZCl0AAAAAAAFAqOE2AnH7Wxc7f8rVK6dxrAAAAAAAA2BYFKwAAAAAAANgKBSsAAAAAAADYCmNYAQAAAACAUsHIIZccvk6jSIyf5WsVWlgBAAAAAADAVihYAQAAAAAAwFYoWAEAAAAAAMBWGMMKAAAAAACUCk4TIKfxr7Y7/pavVUrnXgMAAAAAAMC2KFgBAAAAAADAVugSCAAAAAAASgWXcchlHL5Oo0j8LV+r0MIKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whhWAAAAAACgVHAqQE4/a7vjb/lapXTuNQAAAAAAAGyLghUAAAAAAABshYIVAAAAAAAAbIUxrAAAAAAAQKngMg65jMPXaRSJv+VrFVpYAQAAAAAAwFYoWAEAAAAAAMBW6BIIAAAAAABKBZcC5PKztjv+lq9VSudeAwAAAAAAwLYoWAEAAAAAAMBWSkzB6ttvv1X37t1VuXJlORwOffrpp17PG2M0evRoVa5cWeHh4Wrfvr1++eUX3yQLAAAAAACAfJWYgtWxY8fUpEkTTZ48Oc/nx48fr4kTJ2ry5MlavXq1KlasqM6dO+vIkSPnOVMAAAAAAOALTuPwy6k0KjGDrnft2lVdu3bN8zljjCZNmqRHH31UPXv2lCS9+eabSkhI0Lvvvqvbb7/9fKYKAAAAAACAApSYFlYF2bp1q1JTU5WcnOyZFxoaqnbt2mn58uU+zAwAAAAAAACnKzEtrAqSmpoqSUpISPCan5CQoJSUlHzXy8zMVGZmpudxenr6uUkQAAAAAAAAHqWiYOXmcHj3+zTG5JqX07hx4zRmzJhznRYAAAAAADgPXMYhl5+NCeVv+VqlVHQJrFixoqR/Wlq57d27N1erq5xGjBihtLQ0z7Rjx45zmicAAAAAAABKScGqRo0aqlixohYuXOiZl5WVpWXLlql169b5rhcaGqro6GivCQAAAAAAAOdWiekSePToUW3evNnzeOvWrfrxxx8VHx+vatWq6b777tPYsWNVp04d1alTR2PHjlVERIRuvvlmH2YNAAAAAADOF2MC5DL+1XbH+Fm+VikxBas1a9aoQ4cOnsfDhg2TJPXr108zZ87U8OHDlZGRoSFDhujQoUO65JJLtGDBApUpU8ZXKQMAAAAAACAPJaZg1b59exlj8n3e4XBo9OjRGj169PlLCgAAAAAAAEVWOtuVAQAAAAAAwLZKTAsrAAAAAACAgjjlkFMOX6dRJP6Wr1VoYQUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAEoFl5FcxuFnU/H29ZVXXlGNGjUUFham5s2b67vvvitw+XfeeUdNmjRRRESEKlWqpAEDBujAgQPF27gFKFgBAAAAAACUIHPmzNF9992nRx99VOvXr1fbtm3VtWtXbd++Pc/lv//+e/Xt21cDBw7UL7/8og8++ECrV6/WoEGDznPm/6BgBQAAAAAAUIJMnDhRAwcO1KBBg1SvXj1NmjRJiYmJmjJlSp7Lr1y5UtWrV9c999yjGjVq6LLLLtPtt9+uNWvWnOfM/0HBCgAAAAAAlAouE+CXU1FkZWVp7dq1Sk5O9pqfnJys5cuX57lO69attXPnTs2dO1fGGO3Zs0cffvihunXrVuxjfbYoWAEAAAAAANhcenq615SZmZnncvv375fT6VRCQoLX/ISEBKWmpua5TuvWrfXOO++oV69eCgkJUcWKFRUbG6uXXnrJ8v0oLApWAAAAAAAANpeYmKiYmBjPNG7cuAKXdzgcXo+NMbnmuW3atEn33HOPHn/8ca1du1bz5s3T1q1bdccdd1iWf1EF+WzLAAAAAAAAKJQdO3YoOjra8zg0NDTP5cqVK6fAwMBcran27t2bq9WV27hx49SmTRs99NBDkqTGjRsrMjJSbdu21VNPPaVKlSpZtBeFRwsrAAAAAABQKrjk8MtJkqKjo72m/ApWISEhat68uRYuXOg1f+HChWrdunWe6xw/flwBAd4losDAQEmnWmb5AgUrAAAAAACAEmTYsGF64403NH36dP3666+6//77tX37dk8XvxEjRqhv376e5bt3766PP/5YU6ZM0ZYtW/TDDz/onnvuUcuWLVW5cmWf7ANdAgEAAAAAAEqQXr166cCBA3riiSe0e/duNWzYUHPnzlVSUpIkaffu3dq+fbtn+f79++vIkSOaPHmyHnjgAcXGxqpjx4565plnfLULFKwAAAAAAABKmiFDhmjIkCF5Pjdz5sxc8+6++27dfffd5zirwqNgBQAAAAAASgWncchp8r5Tnl35W75WYQwrAAAAAAAA2AoFKwAAAAAAANgKXQIBAAAAAECp4DIBchn/arvjb/lapXTuNQAAAAAAAGyLghUAAAAAAABshYIVAAAAAAAAbIUxrAAAAAAAQKngkkMu4/B1GkXikn/laxVaWAEAAAAAAMBWKFgBAAAAAADAVihYAQAAAAAAwFYYwwoAAAAAAJQKRg6/GxPK+Fm+VqGFFQAAAAAAAGyFghUAAAAAAABshYIVAAAAAAAAbIUxrAAAAAAAQKngMg65jH+NCeVv+VqFFlYAAAAAAACwFQpWAAAAAAAAsBW6BAIAAAAAgFLBZQLkMv7Vdsff8rVK6dxrAAAAAAAA2BYFKwAAAAAAANgKBSsAAAAAAADYCmNYAQAAAACAUsFlHHIZh6/TKBJ/y9cqtLACAAAAAACArVCwAgAAAAAAgK1QsAIAAAAAAICtMIYVAAAAAAAoFVxyyCX/GhPK3/K1Ci2sAAAAAAAAYCsUrAAAAAAAAGArdAkEAAAAAAClgss45DL+1cXO3/K1Ci2sAAAAAAAAYCsUrAAAAAAAAGArFKwAAAAAAABgK4xhBQAAAAAASgXGsPIftLACAAAAAACArVCwAgAAAAAAgK1QsAIAAAAAAICtMIYVAAAAAAAoFRjDyn/QwgoAAAAAAAC2QsEKAAAAAAAAtkKXQAAAAAAAUCrQJdB/0MIKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whhWAAAAAACgVDCSXPKvMaGMrxPwEVpYAQAAAAAAwFYoWAEAAAAAAMBWKFgBAAAAAADAVhjDCgAAAAAAlAou45DL+NcYVv6Wr1VoYQUAAAAAAABboWAFAAAAAAAAW6FLIAAAAAAAKBXoEug/aGEFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABbYQwrAAAAAABQKjCGlf+ghRUAAAAAAABshYIVAAAAAAAAbIWCFQAAAAAAAGyFMawAAAAAAECpwBhW/oMWVgAAAAAAALAVClYAAAAAAACwFboEAgAAAACAUsEYh4yfdbHzt3ytQgsrAAAAAAAA2AoFKwAAAAAAANgKBSsAAAAAAADYCmNYAQAAAACAUsElh1zyrzGh/C1fq9DCCgAAAAAAALZCwQoAAAAAAAC2QsEKAAAAAAAAtsIYVgAAAAAAoFRwGYdcxr/GhPK3fK1CCysAAAAAAADYCgUrAAAAAAAA2ApdAgEAAAAAQKlgjEPGz7rY+Vu+VqGFFQAAAAAAAGyFghUAAAAAAABshYIVAAAAAAAAbIUxrAAAAAAAQKngMg65/GxMKH/L1yq0sAIAAAAAAICtULACAAAAAACArVCwAgAAAAAAgK0whhUAAAAAACgVjHHI+NmYUP6Wr1VoYQUAAAAAAABboWAFAAAAAAAAW6FLIAAAAAAAKBWMccjlZ13s6BIIAAAAAAAA2AAFKwAAAAAAANgKBSsAAAAAAADYCmNYAQAAAACAUsFIMsbXWRSNn6VrGVpYAQAAAAAAwFYoWAEAAAAAAMBWKFgBAAAAAADAVhjDCgAAAAAAlAouOeSQw9dpFInLz/K1Ci2sAAAAAAAAYCsUrAAAAAAAAGArdAkEAAAAAAClgjEOGeNfXez8LV+r0MIKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whhWAAAAAACgVHAZhxx+NiaUy8/ytQotrAAAAAAAAGArFKwAAAAAAABKmFdeeUU1atRQWFiYmjdvru+++y7fZfv37y+Hw5FratCgwXnM2BsFKwAAAAAAgBJkzpw5uu+++/Too49q/fr1atu2rbp27art27fnufwLL7yg3bt3e6YdO3YoPj5e//73v89z5v+gYAUAAAAAAEoFY/xzKqqJEydq4MCBGjRokOrVq6dJkyYpMTFRU6ZMyXP5mJgYVaxY0TOtWbNGhw4d0oABA87yiBcfBSsAAAAAAACbS09P95oyMzPzXC4rK0tr165VcnKy1/zk5GQtX768UNuaNm2aOnXqpKSkpLPOu7goWAEAAAAAANhcYmKiYmJiPNO4cePyXG7//v1yOp1KSEjwmp+QkKDU1NQzbmf37t36+uuvNWjQIEvyLq4gn24dAAAAAAAAZ7Rjxw5FR0d7HoeGhha4vMPh8HpsjMk1Ly8zZ85UbGysevToUaw8rULBCgAAAAAAlArGOGTMmYs2duLONzo62qtglZ9y5copMDAwV2uqvXv35mp1lXtbRtOnT1efPn0UEhJS/KQtQJdAAAAAAACAEiIkJETNmzfXwoULveYvXLhQrVu3LnDdZcuWafPmzRo4cOC5TLFQaGEFAAAAAABQggwbNkx9+vRRixYt1KpVK02dOlXbt2/XHXfcIUkaMWKEdu3apVmzZnmtN23aNF1yySVq2LChL9L2QsEKAAAAAACUCv7cJbAoevXqpQMHDuiJJ57Q7t271bBhQ82dO9dz17/du3dr+/btXuukpaXpo48+0gsvvGBJ3meLghUAAAAAAEAJM2TIEA0ZMiTP52bOnJlrXkxMjI4fP36Osyo8xrACAAAAAACArVCwAgAAAAAAgK3QJRAAAAAAAJQKLuOQw8/GsHL5Wb5WoYUVAAAAAAAAbIWCFQAAAAAAAGyFghUAAAAAAABshTGsAAAAAABAqWDMqcmf+Fu+VqGFFQAAAAAAAGyFghUAAAAAAABshS6BAAAAAACgVDjVJdDh6zSKhC6BAAAAAAAAgA1QsAIAAAAAAICtULACAAAAAACArTCGFQAAAAAAKBWMcfjhGFb+la9VaGEFAAAAAAAAW6FgBQAAAAAAAFuhYAUAAAAAAABbYQwrAAAAAABQKpj/n/yJv+VrFVpYAQAAAAAAwFYoWAEAAAAAAMBW6BIIAAAAAABKBWMcMsbh6zSKxN/ytQotrAAAAAAAAGArFKwAAAAAAABgKxSsAAAAAAAAYCuMYQUAAAAAAEoH8/+TP/G3fC1CCysAAAAAAADYCgUrAAAAAAAA2AoFKwAAAAAAANgKY1gBAAAAAIDSwThkjMPXWRSNv+VrEVpYAQAAAAAAwFYoWAEAAAAAAMBW6BIIAAAAAABKBWNOTf7E3/K1CgWrYohbGaLAkBBLYrVceqclcSSp2o+HLYslSS3/ti43SYoPtjScpax8HQD4j1Vjp1gar+Uj1l5L4ioftSxWWi3LQkmSIpsctCyWc2NZy2JJkgm0tgG5CbUwP5d1oSTJGW3N9xGP38IsC5Udk2lZLEkK22/tt3VH5QTLYmWXK2NZLEkK2pduaTxHttO6YCesfV33Nw63NF65n6w7T7LirH1/Hatq3QXAREdaFkuSTtaoYGm8TAu/+J9sfMyyWJLkCLL4J7CFQws5y4RaEye7lFY3UOLQJRAAAAAAAAC2QsEKAAAAAAAAtkKXQAAAAAAAUCoY45AxFvblPA/8LV+r0MIKAAAAAAAAtkLBCgAAAAAAALZCwQoAAAAAAAC2whhWAAAAAACgdDCOU5M/8bd8LUILKwAAAAAAANgKBSsAAAAAAADYCl0CAQAAAABAqWDMqcmf+Fu+VqGFFQAAAAAAAGyFghUAAAAAAABshYIVAAAAAAAAbIUxrAAAAAAAQOlg/n/yJ/6Wr0VoYQUAAAAAAABboWAFAAAAAAAAW6FgBQAAAAAAAFthDCsAAAAAAFAqGOOQMQ5fp1Ek/pavVWhhBQAAAAAAAFuhhVUxHLo0SwHh1tT6tiRPsySOJDWYPMSyWJL0y9BXLI1Xc8FAS+NZycrXAYD/aPnInZbGWzV2iqXxurXsZlmsmL+iLIslSdocb1mowL+3WhZLklxlYy2NF3DgsGWxTFy0ZbEkyXHIaWk8XZhtWaigT1yWxZKkE+XCLI1n/t5jWaxAC2NJkvPYMUvjBZYra2GwQOtiSSq3IcPSeEGHrYsXeCzLsliSFLkzzrJYjnRrz5Hg3fssjRdYq4plsYI3WHvdNNnWXedOBbQuVOCRTEviGKe15y7gKxSsAAAAAABA6WFhoRHnDl0CAQAAAAAAYCsUrAAAAAAAAGArFKwAAAAAAABQbO3bt9esWbOUkWHdWIIUrAAAAAAAQKlgjMMvJ7tr3ry5hg8frooVK2rw4MFauXLlWccsMQWrcePG6eKLL1aZMmVUoUIF9ejRQ7///rvXMsYYjR49WpUrV1Z4eLjat2+vX375xUcZAwAAAAAA+L/nnntOu3bt0qxZs7Rv3z5dfvnlql+/vp599lnt2VO8u+mWmILVsmXLdNddd2nlypVauHChsrOzlZycrGM5bgs8fvx4TZw4UZMnT9bq1atVsWJFde7cWUeOHPFh5gAAAAAAAP4tMDBQ1157rT799FPt2rVLN998sx577DElJiaqR48eWrx4cZHiBZ2jPM+7efPmeT2eMWOGKlSooLVr1+ryyy+XMUaTJk3So48+qp49e0qS3nzzTSUkJOjdd9/V7bff7ou0AQAAAAAASoxVq1ZpxowZeu+991ShQgX1799fu3fvVvfu3XXnnXfq2WefLVScEtPC6nRpaWmSpPj4eEnS1q1blZqaquTkZM8yoaGhateunZYvX55njMzMTKWnp3tNAAAAAADATxk/nWxu7969eu6559SwYUO1bdtW+/bt0+zZs7Vt2zaNGTNGU6dO1WeffaZXX3210DFLTAurnIwxGjZsmC677DI1bNhQkpSamipJSkhI8Fo2ISFBKSkpecYZN26cxowZc26TBQAAAAAA8GNVq1ZVrVq1dOutt6p///4qX758rmVatmypiy++uNAxS2TBaujQodqwYYO+//77XM85HN6j6xtjcs1zGzFihIYNG+Z5nJ6ersTERGuTBQAAAAAA8GOLFi1S27ZtC1wmOjpaS5YsKXTMEtcl8O6779bnn3+uJUuWqGrVqp75FStWlPRPSyu3vXv35mp15RYaGqro6GivCQAAAAAA+CuHn072NmrUKB0+fDjX/PT0dHXs2LFYMUtMwcoYo6FDh+rjjz/W4sWLVaNGDa/na9SooYoVK2rhwoWeeVlZWVq2bJlat259vtMFAAAAAAAoEZYtW6asrKxc80+cOKHvvvuuWDFLTJfAu+66S++++64+++wzlSlTxtOSKiYmRuHh4XI4HLrvvvs0duxY1alTR3Xq1NHYsWMVERGhm2++2cfZAwAAAAAA+JcNGzZIOtWIaNOmTV692pxOp+bNm6cqVaoUK3aJKVhNmTJFktS+fXuv+TNmzFD//v0lScOHD1dGRoaGDBmiQ4cO6ZJLLtGCBQtUpkyZ85wtAAAAAACAf2vatKkcDoccDkeeXf/Cw8P10ksvFSt2iSlYGXPm+zw6HA6NHj1ao0ePPvcJAQAAAAAAezH/P/kTG+e7detWGWNUs2ZNrVq1yuvugCEhIapQoYICAwOLFbvEFKwAAAAAAABw/iQlJUmSXC6X5bEpWAEAAAAAAKBIPv/8c3Xt2lXBwcH6/PPPC1z2mmuuKXJ8ClYAAAAAAAAokh49eig1NVUVKlRQjx498l3O4XDI6XQWOb7DFGbwJ0iS0tPTFRMToyvqDlNQYKglMR1HjlsSR5KyalWwLJYkhfy119J4pkyEpfGsZOXrAMB/ZFeOtzRe0N8HLY331aqvLIuV/O/+lsWSpODt+y2LtWVinGWxJKnCrHBL4+3tm2FZrNiPIy2LJUlyWBsucnfu21EXV8jeY5bFkqSTcda+rkGHT1gW63j1aMtiSbL8dQ1JP2lZrJMR1v69O+CktV1IgjKK/oMoP5nxIZbFkqTQA5mWxXKFFm88mPycjLT2dbXydQjMyLYsliTLxwIKyLIuv6yy1lznsrNP6IdFo5WWlqboaIuvT37M/Xs+8ZXRCggP83U6ReLKOKEdQ0rfaxrg6wQAAAAAAADgn06ePKkOHTrojz/+sDQuBSsAAAAAAAAUS3BwsDZu3CiHw9omwhSsAAAAAABA6WAc/jnZXN++fTVt2jRLYzLoOgAAAAAAAIotKytLb7zxhhYuXKgWLVooMtJ73M6JEycWOSYFKwAAAAAAABTbxo0b1axZM0nKNZZVcbsKUrACAAAAAABAsS1ZssTymBSsAAAAAABAqWDMqcmf+Fu+VvFZwWrPnj1KSEjI9/k1a9aoRYsW5zEjAAAAAAAAFMfq1av1wQcfaPv27crKyvJ67uOPPy5yPJ/dJbBp06ZavHhxns+98MILuuyyy85zRgAAAAAAACiq2bNnq02bNtq0aZM++eQTnTx5Ups2bdLixYsVExNTrJg+K1jVr19fV155pUaPHi3z/+3b0tLS1LNnT91///0aPHiwr1IDAAAAAABAIY0dO1bPP/+8vvzyS4WEhOiFF17Qr7/+qhtuuEHVqlUrVkyfFay++eYbPfLII3rqqafUqVMnffnll2ratKmWLFmiDz/8UC+99JKvUgMAAAAAACWR8dPJ5v766y9169ZNkhQaGqpjx47J4XDo/vvv19SpU4sV02cFK4fDoTFjxmjBggVasWKFrr32WpUpU0br169Xz549fZUWAAAAAAAAiiA+Pl5HjhyRJFWpUkUbN26UJB0+fFjHjx8vVkyfFawkKT09XS+//LJOnDihChUqaOvWrfr22299mRIAAAAAAACKoG3btlq4cKEk6YYbbtC9996rwYMH66abbtIVV1xRrJg+u0vg+vXrdcMNN2jfvn16//33ddVVV+n222/XgAEDtGzZMr388ssKCwvzVXoAAAAAAAAohMmTJ+vEiROSpBEjRig4OFjff/+9evbsqccee6xYMX1WsGrdurXq1aunefPmqVatWpKkt956S+3atdO9996r1atXa8OGDb5KDwAAAAAAlDTGcWryJ36Qb3x8vOf/AQEBGj58uIYPH35WMX1WsOrfv78mTZqk0NBQr/mDBg3SJZdcol69evkoMwAAAAAAABQkPT290MtGR0cXOb7PClZTpkzJ97lGjRppzZo15zEbAAAAAAAAFFZsbKwcjoJbfxlj5HA45HQ6ixzfZwUrt/nz52vp0qXav3+/HnvsMVWrVk2rV69W9erVFRER4ev0AAAAAABACeEwpyZ/Ytd8lyxZck7j+6xgdfz4cV177bVatGiRpyJ35513qlq1anr22WeVmJioZ5991lfpFWhP27IKDLFmQPigDEvCSJKiUzKtCyZpX6dqlsbLDrc0nKWsfB0A+I+0WtbGi/krytJ4yf/ub1msBR/MtCyWJF31+1WWxSr/grUfEA6Xtd/qyr9jXX7OEMtCnWKs3deULqFnXqiQ6sw6ZlksSdrbwto/ZFb+5oRlscL2WxdLklxB1t7IO+Tvw9bFyjppWSxJ2tY3ydJ4ifPTLIvlcAVbFkuStl1t3Tlc+819lsWSpIBy1n5+ZSRYdy05eqF1sSSp4syfLI13vEMDy2JllA20JI4zq+gtWYDiaNeu3TmN77OC1aOPPqo1a9boo48+UufOnb36MyYnJ+ull17yVWoAAAAAAAAowIYNG9SwYUMFBASc8aZ5jRs3LnJ8nxWsPvjgAz355JO67rrrcvVlrFatmrZv3+6jzAAAAAAAAFCQpk2bKjU1VRUqVFDTpk3lcDhk8mj97XdjWO3bt08NGuTdfDIgIEAZGfTRAgAAAAAAFjL/P/kTm+a7detWlS9f3vN/q/msYFWlShX9/PPP6tChQ67nNmzYoBo1avggKwAAAAAAAJxJUlJSnv+3is8KVj179tR///tftW3b1tOX0eFwKCUlRc8//7wGDBjgq9QAAAAAAABQBLt27dIPP/ygvXv3yuVyeT13zz33FDmezwpWo0aN0qJFi9SyZUs1bNhQDodDAwYM0F9//aW6devq4Ycf9lVqAAAAAAAAKKQZM2bojjvuUEhIiMqWLSuHw+F5zuFwFKtgZe09c4ugTJkyWr58uZ588klFRUWpVq1aioiI0IgRI/Ttt98qPNzaW1wDAAAAAIBSzjj8c7K5xx9/XI8//rjS0tK0bds2bd261TNt2bKlWDF91sJKksLDw/Xwww/TmgoAAAAAAMBPHT9+XDfeeKMCAqxrF+WzFlYAAAAAAAA4N1555RXVqFFDYWFhat68ub777rsCl8/MzNSjjz6qpKQkhYaGqlatWpo+fXqhtjVw4EB98MEHVqTtcV5bWHXs2LHQyzocDi1atOgcZgMAAAAAAEoV8/+TPylGvnPmzNF9992nV155RW3atNFrr72mrl27atOmTapWrVqe69xwww3as2ePpk2bptq1a2vv3r3Kzs4u1PbGjRunq6++WvPmzVOjRo0UHBzs9fzEiROLvA/ntWDlcrm8Bt76/ffflZqaqqSkJFWsWFGpqalKSUlRpUqVVLdu3fOZGgAAAAAAQIkwceJEDRw4UIMGDZIkTZo0SfPnz9eUKVM0bty4XMvPmzdPy5Yt05YtWxQfHy9Jql69eqG3N3bsWM2fP99Tyzl90PXiOK8Fq6VLl3r+P2/ePA0ePFg//PCDWrVq5Zm/fPly9erVSw8++OD5TA0AAAAAAMC20tPTvR6HhoYqNDQ013JZWVlau3ZtrvHCk5OTtXz58jxjf/7552rRooXGjx+vt956S5GRkbrmmmv05JNPFuqmeBMnTtT06dPVv3//wu/QGfhsDKuRI0dq9OjRXsUqSWrdurVGjRqlRx991EeZAQAAAAAA2EtiYqJiYmI8U14tpSRp//79cjqdSkhI8JqfkJCg1NTUPNfZsmWLvv/+e23cuFGffPKJJk2apA8//FB33XVXoXILDQ1VmzZtirZDZ+CzuwT+8ssvSkxMzPO5atWq6bfffjvPGQEAAAAAgBLNj8ew2rFjh6Kjoz2z82pdldPpXfGMMfl2z3MP4fTOO+8oJiZG0qlWU9dff71efvnlM7ayuvfee/XSSy/pxRdfPNPeFJrPClYJCQn66KOPlJycnOu5Dz74IFclEAAAAAAAoLSKjo72Kljlp1y5cgoMDMzVmmrv3r351loqVaqkKlWqeIpVklSvXj0ZY7Rz507VqVOnwG2uWrVKixcv1pdffqkGDRrkGnT9448/PmPep/NZwWrIkCF6+OGHdfDgQd18882eQdffeecdffLJJ/k2bQMAAAAAAEDeQkJC1Lx5cy1cuFDXXXedZ/7ChQt17bXX5rlOmzZt9MEHH+jo0aOKioqSJP3xxx8KCAhQ1apVz7jN2NhY9ezZ05od+H8+K1gNHz5cx48f1/jx4z2VNmOMwsLC9Oijj2r48OG+Sg0AAAAAAMBvDRs2TH369FGLFi3UqlUrTZ06Vdu3b9cdd9whSRoxYoR27dqlWbNmSZJuvvlmPfnkkxowYIDGjBmj/fv366GHHtKtt95aqEHXZ8yYYfk++KxgJUmjR4/W/fffrxUrVujAgQMqW7asLr30UsXGxvoyLQAAAAAAUBL58RhWRdGrVy8dOHBATzzxhHbv3q2GDRtq7ty5SkpKkiTt3r1b27dv9ywfFRWlhQsX6u6771aLFi1UtmxZ3XDDDXrqqaes2osi82nBSpJiYmLUpUsXX6cBAAAAAABQYgwZMkRDhgzJ87mZM2fmmnfhhRdq4cKFhY7frFkzLVq0SHFxcbrooovyHdBdktatW1fouG4+LVgZY7R69WqlpKQoIyMj1/N9+/b1QVbnmYWV3cCMbOuCSZIp+I4DJYq/VdgBWCKyyUFrA26OtzRc8Pb9lsX6P/buPU7Guv/j+Pva83mx2CUb63xW0WGpnOVM0kElilJ0cKsUHRwqSm7R7aZUSCH9SmeUYt0KhZSidKIVlpzXsqeZ6/fHZmraJTu+dq6xr+fjcT0e9ppr3vOZmb2uufbje32n85bOxrIkaVGdRcayOm+52liWJMlt+KAedOKTr+LKLxtlLEuS8qPMnsqFHDP32Z9VPf6fNyqGkGNm31fL7TaWFZRt9hws+GiO0TzlmavPzjpqLEuSQszGyRUZ+s8bnaKgHHO/I5JkuQxmZRX+2+m0lIs2Ghd6xNyTPXie2fehUliY0TyTovaY2VfzDe7zwMn06NHD802FPXv2NJ7vt4bVDz/8oO7du+vHH3+UbRc+AbEsq3Q0rAAAAAAAQMmwrYIlkDi03lGjRhX5b1P81rAaMmSIsrOztWDBAjVu3NjTlQMAAAAAAEBgOt7rycrKUvv27VWrVi2fcvzWsPriiy/0wgsvqHfv3v4qAQAAAAAAAD66//77lZubqylTpkiScnNzdckll2jz5s2KiorS8OHD9dFHH6l58+bFzg4yXeypiomJUVxcnL8eHgAAAAAAAKdh8eLFatu2refnuXPnKj09XT/++KMOHDigq6++Wk888YRP2X5rWN18882aN2+evx4eAAAAAACUMpYdmItTpaenq379+p6fP/roI/Xu3VtVq1aVZVm65557tGHDBp+y/XZJYMOGDTV//nx1795d3bp1U0JCQqFtevXq5YfKAAAAAAAA8E+CgoK8vkhvzZo1euSRRzw/lylTRgcOHPAp228Nq+uvv16StHXrVr3//vuFbrcsSy6Xwe9+BQAAAAAAgDF169bVe++9p2HDhmnTpk1KT09X69atPbf/+uuvSkxM9Cnbbw2r5cuX++uhAQAAAAAAcJruv/9+9enTRx988IE2bdqkzp07KyUlxXP7okWLdNFFF/mU7beGVcuWLf310AAAAAAAoDSy/1gCiYPrveqqq7Ro0SJ98MEH6tChg+666y6v26OiojR48GCfsv3WsAIAAAAAAEBga9eundq1a1fkbaNGjfI5t0QbVnffffcpb2tZlqZMmXIGqwEAAAAAAIATlWjDaurUqae8LQ0rAAAAAACA0qlEG1Zut7skHw4AAAAAAAABKMjfBQAAAAAAAAB/RcMKAAAAAAAAPmvTpo0OHjxYaP3hw4fVpk0bnzL5lkAAAAAAAFAqWJIs299VFI/l7wJOQVpamnJzcwutz87O1sqVK33KpGEFAAAAAACAYtu4caPn35s3b1ZGRobnZ5fLpSVLluicc87xKZuGFQAAAAAAAIrtvPPOk2VZsiyryEv/IiMj9Z///MenbL80rFwul37++WdVrFhRZcqU8UcJp6Xi54cUEpxtJCtozwEjOZKUXd+3ruWJVPhoq9E8d8WyRvNMMvk+AAgcrm8TjOYF7zR73Pzl2QrGsipMiTSWJUmdt1xtLGvRJ/9nLEuS2tw0wGjesjkvGctq8kUfY1mSdDgj3GhejQVmzm8kKfyX341lSVJwTqLRPDvM3GlwdmKUsSxJyk6IM5oXuTffWFZwjstYliSFHzR7XU7wsTxjWe7wYGNZkpTy9mFjWa7EMsayzgQr39z7Wm/SIWNZkqQK5YzGxXyb8c8bnaKsBoaOc8xUjRK0detW2bat6tWr64svvlCFCn+ev4aFhalixYoKDvbteOqXhpVt26pfv77ee+89derUyR8lAAAAAACA0sa2CpZA4uB6q1atKklyu93Gs/3Sew0JCVFSUtIZeUIAAAAAAAAoOePHj9fMmTMLrZ85c6aeeuopnzL9Nljwuuuu05w5c/z18AAAAAAAADDg+eefV926dQutb9CggZ577jmfMv026fp5552nBQsWqE2bNurVq5cqVaoky/Ie5tarVy8/VQcAAAAAAM469h9LIAmAejMyMlSpUqVC6ytUqKBdu3b5lOm3htVNN90kSdqxY4fS0tIK3W5Zllwus5M6AgAAAAAAwKzk5GR99tlnSklJ8Vr/2WefqXLlyj5l+q1htXz5cn89NAAAAAAAAAwZOHCghg4dqry8PLVp00aS9Mknn2j48OG69957fcr0W8OqZcuW/npoAAAAAAAAGDJ8+HDt379fgwcPVm5uriQpIiJCDzzwgEaMGOFTpt8aVsdlZmZq9erV2rdvn8qXL69LLrlEsbGx/i4LAAAAAACcbZjD6oywLEtPPfWUHnnkEX333XeKjIxUrVq1FB4e7nOmXxtWEydO1JgxY3T06FHZdsE7EB0drTFjxmjYsGH+LA0AAAAAAADFEBMTowsvvNBIlt8aVnPmzNHw4cPVqVMn9e/fX5UrV9bOnTv18ssv6/7771eFChXUt29ff5UHAAAAAACAU9C6dWtZlnXC25ctW1bsTL81rJ555hldf/31evXVV73WX3311brxxhv1zDPP0LACAAAAAABwuPPOO8/r57y8PH311Vf69ttv1a9fP58y/daw+v777zV+/Pgib7vxxht15ZVXlnBFAAAAAADgbGbZBUsgCYR6n3nmmSLXjx49WkeOHPEpM+h0CjodkZGR2r9/f5G37d+/X5GRkSVcEQAAAAAAAEy58cYbNXPmTJ/u67eG1WWXXabRo0dr586dXuszMjI0duxYXX755X6qDAAAAAAAAKdr9erVioiI8Om+frskcNy4cWrevLlq1qyptm3bqlKlStq1a5eWLVum0NBQLVy40F+lAQAAAACAs5H9xxJIAqDeXr16ef1s27Z27dqldevW6ZFHHvEp028NqwYNGmjt2rUaNWqUli9frn379ikhIUE9e/bUqFGjVLt2bX+VBgAAAAAAgFMUHx/v9XNQUJDq1KmjsWPHqkOHDj5l+q1hJUm1a9fW/Pnz/VkCAAAAAAAATsOsWbOMZ/ptDqs2bdro+++/L/K2H374QW3atCnhigAAAAAAAOAEfhthlZaWpsOHDxd5W2ZmplasWFHCFQEAAAAAgLMac1gZU7ZsWVmWdUrb7t+/v9j5fr0k8ER27dqlqKgof5dxQtmJ0QoJ9W2W+7+LcJn7zcusEmYsS5LCdpcxmpddKcZonkkm3wcAgcMONjvQ2J1QxmhexTmRxrIst+HjnMG8NjcNMJYlScvmvGQ0z2R95Q2/DfGGP1rzI4ONZYUHm8uSpKBct9m8Q1nGssIizJ5Shx3IMZoXnJltLCuncpyxLEkq93XR/4HtK7fB9yIo3+zvnB1qbp+wDJ+72mFmPw9d4ebyspPj/3mjYojcsttoXu655Y1lhRx1mQnKN5QD/IPJkyd7/r1v3z49/vjjuuKKK5Samiqp4BsCP/zww8CYdP2dd97RO++84/n5scceU4UKFby2OXbsmNLS0nT++eeXZGkAAAAAAAA4Rf369fP8+6qrrtLYsWN15513etbdfffdmjp1qj7++GP961//KnZ+iTasNm/erP/7v/+TJFmWpWXLlikoyLv7Hh4erkaNGmnKlCklWRoAAAAAAAB88OGHH+qpp54qtP6KK67Qgw8+6FNmiU66PmLECGVmZiozM1O2bWv58uWen48ve/fu1fLly9W4ceOSLA0AAAAAAJzlLDswF6dLSEjQW2+9VWj922+/rYSEBJ8y/TaHldtt9ppwAAAAAAAAlLwxY8ZowIABSktL88xhtWbNGi1ZskQvvviiT5klOsLqr9asWaPXX3+9yNtef/11ff755yVcEQAAAAAAAIqrf//+WrVqlcqUKaOFCxfqzTffVHx8vD777DP179/fp0y/jbAaOXKkWrRooWuuuabQbZs3b9YLL7ygpUuX+qEyAAAAAABwVrKtgiWQBEi9F198sebOnWssz28Nq40bN+ree+8t8raLL75Y06ZNK+GKAAAAAAAA4AuXy6W3335b3333nSzLUv369dW9e3cFBwf7lOe3hlVWVpZCQop++KCgIGVmZpZwRQAAAAAAACiun376SZ07d9aOHTtUp04d2batH374QcnJyfrggw9Uo0aNYmf6bQ6rlJQULV++vMjbli9frqpVq5ZwRQAAAAAAACiuu+++WzVq1ND27dv15ZdfasOGDUpPT1dKSoruvvtunzL91rC67rrr9Mwzz2jWrFle62fPnq3JkyerT58+fqoMAAAAAACclewAXRxuxYoVmjBhgsqVK+dZl5CQoCeffFIrVqzwKdNvDasHH3xQLVq00IABAxQdHa1atWopOjpaAwYMUIsWLTRixAh/lQYAAAAAAIBTFB4eXuTUTkeOHFFYWJhPmX5rWIWFhWnp0qV6+eWX1aNHD6WkpKhnz556+eWX9dFHH/n8hAAAAAAAAFByunbtqttuu02ff/65bNuWbdtas2aNbr/9dnXv3t2nTL9Nui5JwcHB6tu3r/r27evPMgAAAAAAAOCjZ599Vv369VNqaqpCQ0MlSfn5+erevbumTJniU6ZfG1aS9P3332vFihXau3evBgwYoKSkJO3cuVNly5ZVZGSkv8sDAAAAAABnCcsuWAKJ0+u1bVuHDh3S/PnztXPnTn333XeybVv169dXzZo1fc71W8PK5XLptttu0+zZs2XbtizLUqdOnZSUlKRBgwbp/PPP19ixY/1VHgAAAAAAAP6BbduqVauWNm3apFq1ap1Wk+qv/DaH1RNPPKF58+bp6aef1rfffivb/rNl2KlTJy1ZssRfpQEAAAAAAOAUBAUFqVatWtq3b5/ZXKNpxTB79mw98sgjGjZsmOrUqeN1W0pKirZu3eqnygAAAAAAAHCqJkyYoPvvv1/ffvutsUy/XRK4Y8cOpaamFnlbREREkV+HCAAAAAAA4DP7jyWQBEC9N954o44ePaomTZooLCys0Jzk+/fvL3am3xpWFStW1C+//KLWrVsXum3Lli2qUqWKH6oCAAAAAABAcTzzzDOyLMtopt8aVp07d9YTTzyhjh07KikpSZJkWZYOHTqkZ599Vt26dfNXaf8oYneWQoLzjWQF7TtoJEeSYn+LNpYlma1NkiKCzf7ymmT6uQIIDHZ4gtE808eSPaNCjWVVmGv4m3eDzB3Tl815yViWJLW5aYDRPJP1Nfmij7EsSTqcEWs0r8YCl7kwl8EsSe4wszNZuOPNnTflxocZy5Kk7ASzp+iRe8ONZQXnmH1f9zeJM5pX7pvDxrLyoiKMZUlS8JFco3kmWbluo3nBwebywneZe08lyQ43u7+GbTc3Z09Wg0QjOfl5wUZygOLo06eP8vPzFR1t7vPVb3NYjR07Vvn5+apfv76uuuoqWZalkSNHqmHDhsrOztYjjzzir9IAAAAAAMDZyJasAFucfEng3r171aVLF8XExCguLk7NmzfXL7/8YiS7RBtWv//+u+ffiYmJWrt2rfr06aP169crODhYX3/9tTp16qRVq1apXLlyJVkaAAAAAAAAimHEiBFav369xowZo6efflp79+7VoEGDjGSX6CWB55xzjrp3764BAwaoY8eOSkxM1HPPPVeSJQAAAAAAAMCADz/8UDNnzlTnzp0lFUz/1LBhQ+Xl5Sk09PSmtyjREVZXX321Fi9erK5duyo5OVkPP/ywsaFiAAAAAAAAKDk7d+7U+eef7/m5bt26CgsL086dO087u0QbVnPnztWuXbv03//+V+ecc47GjRunWrVqqVWrVnr11VeVnZ1dkuUAAAAAAIDSxA7QxaFs21ZIiPfFeyEhIXK7T//LF0p80vW4uDjdfvvt+vzzz7Vp0yb961//0pYtW3TTTTcpKSlJd9xxh9auXVvSZQEAAAAAAKAYbNtW27ZtdcEFF3iWo0ePqlu3bl7rfFGic1j9Xb169TRx4kQ99dRT+uCDDzRz5kzNnDlTM2bMUIMGDbRx40Z/lgcAAAAAAIATGDVqVKF1PXr0MJLt14bVccHBwerevbsuvvhiTZgwQZMnT9amTZv8XRYAAAAAAABOoKiGlSl+b1i5XC69++67mjVrlpYsWaL8/Hw1btxYAwYM8HdpAAAAAADgbOLwOaGKFGj1GuK3htWmTZs0c+ZMvfrqq9q7d6/i4uI0cOBADRgwQE2bNvVXWQAAAAAAAPCzEm1YHT58WPPmzdPMmTO1fv16SdLll1+uAQMGqHfv3oqIiCjJcgAAAAAAAOBAJdqwSkpKUk5OjipVqqQHH3xQt9xyi2rUqFGSJQAAAAAAgFLKsguWQBJo9ZoSVJIP1rFjR7377rtKT0/XE088QbMKAAAAAAAgwM2ZM0c5OTmF1ufm5mrOnDk+ZZZow2rhwoXq0qWLgoJK9GEBAAAAAABwhtx88806dOhQofWZmZm6+eabfcqkcwQAAAAAAACf2bYty7IKrf/tt98UHx/vU6bfviUQAAAAAAAAgev888+XZVmyLEtt27ZVSMifbSaXy6WtW7eqY8eOPmXTsAIAAAAAAECx9ezZU5L01Vdf6YorrlBMTIzntrCwMFWrVk1XXXWVT9k0rAAAAAAAAFBso0aNkiRVq1ZN1157rSIiIoxl07DygSs6TFZIuJEsKy/OSI4kucLNTklmlzVXmyS5Ysy8ZmeCyfcBQABxm40zfdwsszDaWJYrzFiUJCm/bJSxrCZf9DGWJUnlDX/1s8n6vr5ovrEsSaq/6kajeblx5t7X0LKxxrIkKSsp1Ghe5MZ9xrLcVWL+eaNiCM0ye3AKyjOXF3Ko8DdAnY5jFSKN5lkucwcAy232YJJT3txzNfk8JSnkaL7RvJyy5v7MDP/N7P6QX97s/iqDb0VubLCRnPw8MzlAcfTr10+StH79en333XeyLEv169fX+eef73MmDSsAAAAAAFA62DLaaCwRAVDvnj17dN111yktLU1lypSRbds6dOiQWrdurddee00VKlQodibfEggAAAAAAACf3XXXXTp8+LA2bdqk/fv368CBA/r22291+PBh3X333T5lMsIKAAAAAAAAPluyZIk+/vhj1atXz7Oufv36+u9//6sOHTr4lEnDCgAAAAAAlAqWXbAEkkCo1+12KzS08HyToaGhcrt9m4uOSwIBAAAAAADgszZt2uiee+7Rzp07Pet27Nihf/3rX2rbtq1PmTSsAAAAAAAA4LOpU6cqMzNT1apVU40aNVSzZk2lpKQoMzNT//nPf3zK5JJAAAAAAAAA+Cw5OVlffvmlli5dqu+//162bat+/fpq166dz5k0rAAAAAAAQOkRAHNCBar27durffv2RrJoWAEAAAAAAMAnbrdbs2fP1sKFC7Vt2zZZlqWUlBT17t1bffv2lWVZPuUyhxUAAAAAAACKzbZtde/eXQMHDtSOHTvUqFEjNWjQQL/++qv69++vK6+80udsRlgBAAAAAACg2GbPnq3//e9/+uSTT9S6dWuv25YtW6aePXtqzpw5uummm4qdzQgrAAAAAABQOtgBujjU/PnzNXLkyELNKklq06aNHnzwQc2dO9enbBpWAAAAAAAAKLaNGzeqY8eOJ7y9U6dO+vrrr33KpmEFAAAAAACAYtu/f78SExNPeHtiYqIOHDjgUzZzWAEAAAAAgFLBsguWQOLkel0ul0JCTtxaCg4OVn5+vk/ZNKwAAAAAAABQbLZtq3///goPDy/y9pycHJ+zaVgBAAAAAACg2Pr16/eP2/jyDYESDSsAAAAAAAD4YNasWWcsm4YVAAAAAAAoHew/lkASaPUaQsPKB+7QILlDzHzBom0oR5Ist7EoSWZrkwpeN6cy/VwBBAZXXJjRPOuAy2ieLINZttkznfwoc6cQhzOKnvPAV/ExRuN0OCPWWFb9VTcay5Kkzc1fNZp36fxBxrKsnDxjWZIUlG/4bD3c3P4fesS3yWRPJCjH7LEk5HC2saz8+AhjWZIUZPbXRPmx5o4nrvBgY1mSlB9l7nwzaucxY1mSZFsmP3DM7q/uWLO/c6Hb9xnNy69U1lhW6BEzf9BZeYb/MAT8hL/SAQAAAAAA4Cg0rAAAAAAAAM4y06ZNU0pKiiIiItS0aVOtXLnyhNumpaXJsqxCy/fff1+CFXvjkkAAAAAAAFAqWHbBEkh8qXfBggUaOnSopk2bphYtWuj5559Xp06dtHnzZp177rknvN+WLVsUFxfn+blChQq+lGwEI6wAAAAAAADOIpMmTdKAAQM0cOBA1atXT5MnT1ZycrKmT59+0vtVrFhRSUlJniU42Ox8fsVBwwoAAAAAAMDhDh8+7LXk5OQUuV1ubq7Wr1+vDh06eK3v0KGDVq1addLHOP/881WpUiW1bdtWy5cvN1a7L2hYAQAAAACA0sEO0EVScnKy4uPjPcv48eOLfIp79+6Vy+VSYmKi1/rExERlZGQUeZ9KlSppxowZevPNN7Vw4ULVqVNHbdu21f/+979/ekXPGOawAgAAAAAAcLjt27d7zS8VHh5+0u0ty/L62bbtQuuOq1OnjurUqeP5OTU1Vdu3b9fEiRN1+eWXn0bVvmOEFQAAAAAAgMPFxcV5LSdqWJUvX17BwcGFRlPt2bOn0Kirk7nkkkv0448/nlbNp4OGFQAAAAAAwFkiLCxMTZs21dKlS73WL126VM2bNz/lnA0bNqhSpUqmyztlXBIIAAAAAABKh7/MCRUwfKh32LBh6tu3r5o1a6bU1FTNmDFD6enpuv322yVJI0aM0I4dOzRnzhxJ0uTJk1WtWjU1aNBAubm5evXVV/Xmm2/qzTffNPlMioWGFQAAAAAAwFnk2muv1b59+zR27Fjt2rVLDRs21KJFi1S1alVJ0q5du5Senu7ZPjc3V/fdd5927NihyMhINWjQQB988IE6d+7sr6dAwwoAAAAAAOBsM3jwYA0ePLjI22bPnu318/DhwzV8+PASqOrUMYcVAAAAAAAAHIURVgAAAAAAoFSw7IIlkARavaYwwgoAAAAAAACOQsMKAAAAAAAAjsIlgQAAAAAAoHSw/1gCSaDVawgjrAAAAAAAAOAoNKwAAAAAAADgKFwS6IMdA/IVFJVvJMuyQo3kSNK5k3ONZUnS1ofN1SZJtm3mNTsTTL4PAALI9xFm8+qaPc5VmW7uuP5rx3BjWZIUcsxcXo0F2cayJCk/MthoXo0FLmNZuXFRxrIk6dL5g4zmffqf541ltblpgLEsSdrdwuz1EGXXm/sdDspzG8uSpKBcs8cSV6y55xqy94ixLEmK2G92nwg+Yu64mXmu2ePmgXqWsawaGzONZUmSO97w+5ATZizrp+vijGVJUq3R24zm7e2cbCwrZoe5zxvgbEDDCgAAAAAAlA7MYRUwuCQQAAAAAAAAjkLDCgAAAAAAAI5CwwoAAAAAAACOwhxWAAAAAACgVLDsgiWQBFq9pjDCCgAAAAAAAI5CwwoAAAAAAACOwiWBAAAAAACgdLD/WAJJoNVrCCOsAAAAAAAA4Cg0rAAAAAAAAOAoNKwAAAAAAADgKMxhBQAAAAAASgXLLlgCSaDVawojrAAAAAAAAOAoNKwAAAAAAADgKDSsAAAAAAAA4CjMYQUAAAAAAEoH+48lkARavYYwwgoAAAAAAACOQsMKAAAAAAAAjsIlgT44979SiKFXzspxmwmSpGCz4wSrjTebZ4cHG80zyej7ACBg5MfnGM0LecvssSQoN8tYVq055rIkKat6vLGs8F9+N5YlSeHBhj9vXC5jUaFlY41lSZKVk2c0r81NA4xlLZvzkrEsSerc7hqjeVauudduW7dIY1mSFL09ymheTjlzWdVeP2ouTFL8z8eM5uWXCTeWdbCOZSxLktFLevIrxpkLk5STYO51k6Q9F5j7M7P622Z/5/LPr2U073B1c1l2kJnPL1euc//ucgQuCQwYjLACAAAAAACAo9CwAgAAAAAAgKPQsAIAAAAAAICjMIcVAAAAAAAoFaw/lkASaPWawggrAAAAAAAAOMpZ07CaPn26GjdurLi4OMXFxSk1NVWLFy/23G7btkaPHq3KlSsrMjJSrVq10qZNm/xYMQAAAAAAAIpy1jSsqlSpoieffFLr1q3TunXr1KZNG/Xo0cPTlJowYYImTZqkqVOnau3atUpKSlL79u2VmZnp58oBAAAAAADwV2dNw6pbt27q3Lmzateurdq1a+uJJ55QTEyM1qxZI9u2NXnyZD300EPq1auXGjZsqJdffllHjx7VvHnz/F06AAAAAAAoCXaALqXQWdOw+iuXy6XXXntNWVlZSk1N1datW5WRkaEOHTp4tgkPD1fLli21atWqE+bk5OTo8OHDXgsAAAAAAADOrLOqYfXNN98oJiZG4eHhuv322/XWW2+pfv36ysjIkCQlJiZ6bZ+YmOi5rSjjx49XfHy8Z0lOTj6j9QMAAAAAAOAsa1jVqVNHX331ldasWaM77rhD/fr10+bNmz23W5b3l0Hatl1o3V+NGDFChw4d8izbt28/Y7UDAAAAAACgQIi/CzApLCxMNWvWlCQ1a9ZMa9eu1ZQpU/TAAw9IkjIyMlSpUiXP9nv27Ck06uqvwsPDFR4efmaLBgAAAAAAJcKyC5ZAEmj1mnJWjbD6O9u2lZOTo5SUFCUlJWnp0qWe23Jzc7VixQo1b97cjxUCAAAAAADg786aEVYjR45Up06dlJycrMzMTL322mtKS0vTkiVLZFmWhg4dqnHjxqlWrVqqVauWxo0bp6ioKF1//fX+Lh0AAAAAAAB/cdY0rHbv3q2+fftq165dio+PV+PGjbVkyRK1b99ekjR8+HAdO3ZMgwcP1oEDB3TxxRfro48+UmxsrJ8rBwAAAAAAJcL+YwkkgVavIWdNw+qll1466e2WZWn06NEaPXp0yRQEAAAAAAAAn5zVc1gBAAAAAAAg8NCwAgAAAAAAgKOcNZcEAgAAAAAA/KNSOidUoKFh5YPtbWMUHBFhJMtyG4mRJFWbvsVcmKRtd9Qxmmc7eDyfyfcBQOCI2Gv2bCW7vJnPhuPOWX7UWNaeZlHGsiQp5Ji51y44J9FYliQF5Zo9qLvDzH2AZSWFGsuSpKB8s7/Du1uYy+vc7hpjWZK06OPXjeZ1uaiLsayar+w3liVJrthwo3nB+7OMZbl++NlYliT9dnVzo3lV3ztgLKvGS78by5KknV2TjWWFbN5mLEuSQsuWMZoX8Xu8sayDtaKNZUlS+bTtRvNqvFHWXJjbzOdXvivHSA7gbw5uIQAAAAAAAKA0omEFAAAAAAAAR+GSQAAAAAAAUCpYdsESSAKtXlMYYQUAAAAAAABHoWEFAAAAAAAAR+GSQAAAAAAAUDrYfyyBJNDqNYQRVgAAAAAAAHAUGlYAAAAAAABwFBpWAAAAAAAAcBTmsAIAAAAAAKWCZRcsgSTQ6jWFEVYAAAAAAABwFBpWAAAAAAAAcBQaVgAAAAAAAHAU5rACAAAAAAClg/3HEkgCrV5DGGEFAAAAAAAAR6FhBQAAAAAAAEfhkkAAAAAAAFAqWHbBEkgCrV5TGGEFAAAAAAAAR6FhBQAAAAAAAEfhkkAfVHtjj0KCw41k2TsyjORIku1yGcuSpKqTvjaaZ52TZDTPJJPvA4DAYVVONJpn79xtNE8pycaiKn+cbSxLkiy321iWHWb2dCToUJbRPHd8tLGsyI37jGVJksLDjMaVXW/m/EaSrNw8Y1mS1OWiLkbzPvjiA2NZnRu0NpYlSUG5uUbzFBlhLOqXCanGsiSp9rO/Gs1LeuOQsaxdt1Q2liVJ5b8+aizLKlvGWJYk/d7S7HMt+72547DlNnwtVJ7ZY1NenLnjcPjOw0ZyglyGjyGAn9CwAgAAAAAApYP9xxJIAq1eQ7gkEAAAAAAAAI5CwwoAAAAAAACOQsMKAAAAAAAAjsIcVgAAAAAAoHRgDquAwQgrAAAAAAAAOAoNKwAAAAAAADgKlwQCAAAAAIBSwbILlkASaPWawggrAAAAAAAAOAoNKwAAAAAAADgKDSsAAAAAAAA4CnNYAQAAAACA0sH+YwkkgVavIYywAgAAAAAAgKPQsAIAAAAAAICj0LACAAAAAACAozCHFQAAAAAAKBUs25ZlB9akUIFWrymMsAIAAAAAAICj0LACAAAAAACAo3BJoA/sXbtlW2H+LqOw2tXM5v283WicvSPDaB4AnK788rFG84J37jaad7RanLGsiL3ZxrIkKSg731hWdmKUsSxJCoswe3qTG2/uM99dJcZYliSFHjH3PkhSUJ7bWNa2bpHGsiSp5iv7jeZ1btDaWNaiTcuNZUlS/WmDjebF/GbuUpLaz6Yby5Kk33pXNZpn3/y7saxdbcsby5Jk9Gvp42NDzYVJyo2zjOb90D/CWFb9MduMZUmSHWP2Mycrydx7kX6Fmd85d3a29IiRqLOTLaP7Y4kItHoNYYQVAAAAAAAAHIWGFQAAAAAAAByFhhUAAAAAAAAchTmsAAAAAABAqWDZBUsgCbR6TWGEFQAAAAAAAByFhhUAAAAAAAAchYYVAAAAAAAAHIU5rAAAAAAAQOlg/7EEkkCr1xBGWAEAAAAAAMBRaFgBAAAAAADAUbgkEAAAAAAAlAqWXbAEkkCr1xRGWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAgNLBDtDFB9OmTVNKSooiIiLUtGlTrVy58pTu99lnnykkJETnnXeebw9sCA0rAAAAAACAs8iCBQs0dOhQPfTQQ9qwYYMuu+wyderUSenp6Se936FDh3TTTTepbdu2JVTpidGwAgAAAAAAOItMmjRJAwYM0MCBA1WvXj1NnjxZycnJmj59+knvN2jQIF1//fVKTU0toUpPjIYVAAAAAADAWSI3N1fr169Xhw4dvNZ36NBBq1atOuH9Zs2apZ9//lmjRo060yWekhB/FwAAAAAAAFASLLtgCSTH6z18+LDX+vDwcIWHhxfafu/evXK5XEpMTPRan5iYqIyMjCIf48cff9SDDz6olStXKiTEGa0iZ1QRYILKxCsoqPAvhS/yf9thJEeSQjKPGcuSpPysLKN5IVXOMZpnksn3AUDgCPn98D9vVAwuw8dNWeai3CFmB1UHH80xlpWdEGcsS5LCDpirTZKyE8ydLoVmuY1lSVJQjstsXm6+sazo7VHGsiTJFWvm3Ou4oNxcY1n1pw02liVJmwdPM5o3cndjY1nr54Qay5KkMj+ZPT+0Mo8ay0pK228sS5Ist7n93+TzLFDJaFrMTnOfOXaW4ecaHWk0LuGTrcaywrKqGsnJz3Npm5EkOE1ycrLXz6NGjdLo0aNPuL1leZ9M2rZdaJ0kuVwuXX/99RozZoxq165tpFYTaFgBAAAAAAA43Pbt2xUX9+d/9BU1ukqSypcvr+Dg4EKjqfbs2VNo1JUkZWZmat26ddqwYYPuvPNOSZLb7ZZt2woJCdFHH32kNm3aGHwmp4aGFQAAAAAAKB3sP5ZA8ke9cXFxXg2rEwkLC1PTpk21dOlSXXnllZ71S5cuVY8ePQptHxcXp2+++cZr3bRp07Rs2TK98cYbSklJOb36fUTDCgAAAAAA4CwybNgw9e3bV82aNVNqaqpmzJih9PR03X777ZKkESNGaMeOHZozZ46CgoLUsGFDr/tXrFhRERERhdaXJBpWAAAAAAAAZ5Frr71W+/bt09ixY7Vr1y41bNhQixYtUtWqBXOl7dq1S+np6X6u8uRoWAEAAAAAAJxlBg8erMGDi/5ikNmzZ5/0vqNHjz7phO4lgYYVAAAAAAAoNaxAm8OqlDL7HdcAAAAAAADAaaJhBQAAAAAAAEehYQUAAAAAAABHYQ4rAAAAAABQOth2wRJIAq1eQxhhBQAAAAAAAEehYQUAAAAAAABH4ZJAAAAAAABQKlh2wRJIAq1eUxhhBQAAAAAAAEehYQUAAAAAAABHoWEFAAAAAAAAR2EOKwAAAAAAUDrYfyyBJNDqNYQRVgAAAAAAAHAUGlYAAAAAAABwFC4J9IEdGSE7ONxIVnD5BCM5kmSHBBvLkszWJkl2VITRPJNMP1cAgcHKdxnNM30sCTucZy5r50FjWZKkvHxjUZF7zWVJUnBmttG8yL1mPvMlKSjPbSxLkkIOm32urlhzzzWnnLEoSVLw/iyzgZHmzktifjN7rcbI3Y2N5o1L3Ggsq3N8a2NZkhSVfthonsJCjUW5Y8KMZUmSlWPuMyco39y+Kknh+8weS4JzDb52QZa5LMno74gkuRPNHexCjhj6HTF8fgP4Cw0rAAAAAABQKljugiWQBFq9pnBJIAAAAAAAAByFhhUAAAAAAAAchYYVAAAAAAAAHIU5rAAAAAAAQOlg/7EEkkCr1xBGWAEAAAAAAMBRaFgBAAAAAADAUbgkEAAAAAAAlAqWXbAEkkCr1xRGWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFGYwwoAAAAAAJQOtl2wBJJAq9cQRlgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRmMMKAAAAAACUCpZdsASSQKvXFEZYAQAAAAAAwFFoWAEAAAAAAMBRuCTQB9aRo7KC8o1k2cHBRnIkKfecMsayJCns8BGjeVZmltE8k0y+DwACSHaO2TzDx5K8KHMf02G5ecayJMnOOmosKzjHZSxLknIqxxnNM1lfyCGzv3P58RFG80L2mvvsr/a6ud8RSXL98LPRvF8mpBrLqv1surEsSVo/J9RoXuf41sayFm1abixLkjo3aW8077cbahnLqrTikLEsSQrKNfP3gyQdrVnOWJYkHa1o9s/ChC8PGMs60KmesSxJiv31mNE8GbxUK2K7md+5fJfh85uzjS2j71uJCLR6DWGEFQAAAAAAAByFhhUAAAAAAAAchYYVAAAAAAAAHIU5rAAAAAAAQKlg2QVLIAm0ek1hhBUAAAAAAAAchYYVAAAAAAAAHIWGFQAAAAAAAByFOawAAAAAAEDpYNsFSyAJtHoNYYQVAAAAAAAAHIWGFQAAAAAAAByFSwIBAAAAAECpYNkFSyAJtHpNYYQVAAAAAAAAHIWGFQAAAAAAAByFhhUAAAAAAAAchTmsAAAAAABA6WD/sQSSQKvXEEZYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUZjDCgAAAAAAlAqWXbAEkkCr1xQaVj44cn4VhYRGGMkK359rJEeSQjJzjGVJUn7NykbzcsqFGc0zyeT7ACBw7G0caTSv/MZjRvOC8tzGsrbdVNVYliSFHDWXFX7Q7FlYua8PG83b3yTOWNaxCmZ/54LyjMYpYn+Usaz4n83uD79d3dxoXu1nfzWW9Vtvs/tXmZ/OMZoXlW5un+jcpL2xLEla9PVSo3kdeyQZy8pNMHO+f9zuC8ONZVVZfsRYliS5wyyjeb9fXNZYVuz2fGNZkhSUYzgv21ze/mbljeS4crOlH41EAX7FJYEAAAAAAABwFEZYAQAAAACA0sFtFyyBJNDqNYQRVgAAAAAAAHAUGlYAAAAAAABwFBpWAAAAAAAAcBTmsAIAAAAAAKWD/ccSSAKtXkMYYQUAAAAAAABHoWEFAAAAAAAAR6FhBQAAAAAAAEdhDisAAAAAAFAqWJKsAJsTyvJ3AX7CCCsAAAAAAAA4Cg0rAAAAAAAAOAqXBAIAAAAAgNLBtguWQBJo9RrCCCsAAAAAAAA4Cg0rAAAAAAAAOAoNKwAAAAAAADgKc1gBAAAAAIBSwbILlkASaPWawggrAAAAAAAAOAoNKwAAAAAAADgKlwT6IGJvtkIMvXLBh7PNBEnKj480liVJIQeOGs2zXG6jeSaZfB8ABI7yX5sdXx1y8JjRPCnKWFLyh4eMZUmSKzLUWFbwsTxjWZLkjjB7elPum8PGsiyX2d+5/Nhwo3nBR3KNZeWXMVtb1fcOGM1LesPcPmHf/LuxLEmyMs2egynM3P762w21jGVJUsceSUbzlrzzirGsVrfeaixLksr85DKWZQdZxrIk6VCK2eNm2e/NHdczq5itLfIXs+f9R+olGMuK22amtvx8/rbB2YGGFQAAAAAAKB3sP5ZAEmj1GsIlgQAAAAAAAHAUGlYAAAAAAABwFC4JBAAAAAAApYJl27LswLrGLtDqNYURVgAAAAAAAHAUGlYAAAAAAABwFBpWAAAAAAAAcBTmsAIAAAAAAKWD+48lkARavYYwwgoAAAAAAACOQsMKAAAAAAAAjkLDCgAAAAAAAI7CHFYAAAAAAKBUsGxblm37u4xiCbR6TWGEFQAAAAAAAByFhhUAAAAAAAAchUsCAQAAAABA6WD/sQSSQKvXEEZYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUZjDygeuiBBZIWZeuqBjwUZyJCk/2uzbGXzUXG1SwevmVCbfBwCBI7dsmNG84Kxco3k55czVZ7lDjWVJUlCO21iWO9zsMTgo31xtkpQXFWEsy3KbnYTCZfi1yzw33FjWwTqWsSxJqvHS70bzdt1S2VxW2/LGsiQpKW2/0Tx3jLljSaUVh4xlSVJugrn9S5Ja3Xqrsay0F14wliVJTddfYywr740EY1mSFHbI7LFp52XmzvurLjpmLEuSjtYqZzRv90XmjsNhh6KM5LhygqQ1RqLOTrZdsASSQKvXEEZYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUZw7qRAAAAAAAIBBll2wBJJAq9cURlgBAAAAAADAUWhYAQAAAAAAwFG4JBAAAAAAAJQOtl2wBJJAq9cQRlgBAAAAAACcZaZNm6aUlBRFRESoadOmWrly5Qm3/fTTT9WiRQslJCQoMjJSdevW1TPPPFOC1RbGCCsAAAAAAICzyIIFCzR06FBNmzZNLVq00PPPP69OnTpp8+bNOvfccwttHx0drTvvvFONGzdWdHS0Pv30Uw0aNEjR0dG67bbb/PAMGGEFAAAAAABwVpk0aZIGDBiggQMHql69epo8ebKSk5M1ffr0Irc///zz1adPHzVo0EDVqlXTjTfeqCuuuOKko7LONBpWAAAAAACgVLDcgbkUR25urtavX68OHTp4re/QoYNWrVp1ShkbNmzQqlWr1LJly+I9uEFcEggAAAAAAOBwhw8f9vo5PDxc4eHhhbbbu3evXC6XEhMTvdYnJiYqIyPjpI9RpUoV/f7778rPz9fo0aM1cODA0y/cR4ywAgAAAAAAcLjk5GTFx8d7lvHjx590e8uyvH62bbvQur9buXKl1q1bp+eee06TJ0/W/PnzT7tuXzHCCgAAAAAAwOG2b9+uuLg4z89Fja6SpPLlyys4OLjQaKo9e/YUGnX1dykpKZKkRo0aaffu3Ro9erT69OlzmpX7hhFWAAAAAACgdLDtwFwkxcXFeS0naliFhYWpadOmWrp0qdf6pUuXqnnz5sV4qWzl5OT4/lqfJkZYAQAAAAAAnEWGDRumvn37qlmzZkpNTdWMGTOUnp6u22+/XZI0YsQI7dixQ3PmzJEk/fe//9W5556runXrSpI+/fRTTZw4UXfddZffngMNKwAAAAAAgLPItddeq3379mns2LHatWuXGjZsqEWLFqlq1aqSpF27dik9Pd2zvdvt1ogRI7R161aFhISoRo0aevLJJzVo0CB/PQUaVgAAAAAAAGebwYMHa/DgwUXeNnv2bK+f77rrLr+OpioKDSsAAAAAAFA62H8sgSTQ6jWEhpUP9jaOUHB4hKE0UzlSpf8dMpYlSTvblDWa52zm3gcAgSOrittoXvRvZo+biZ9nGcva1jXKWJYkWS5zWSlvHzYXJskODTaaF3wk11hWTvlIY1mSlB9l9vtzDtQ7+VddF4vhk+udXZON5pX/+qi5MMPP1XKbPTZZOeZ22KDcfGNZkrT7wqInDPZVmZ/MPdem668xliVJ65u+biwrdf7txrIkSXlm41wGP3J+7Wz2uFntPYP7vqTcsuaOm8FHzfx57jJ7CAH8hm8JBAAAAAAAgKMwwgoAAAAAAJQKlm3LsgPrGrtAq9cURlgBAAAAAADAUWhYAQAAAAAAwFHOyobV+PHjZVmWhg4d6lln27ZGjx6typUrKzIyUq1atdKmTZv8VyQAAAAAAACKdNY1rNauXasZM2aocePGXusnTJigSZMmaerUqVq7dq2SkpLUvn17ZWZm+qlSAAAAAABQomw7MJdS6KxqWB05ckQ33HCDXnjhBZUt++dXi9u2rcmTJ+uhhx5Sr1691LBhQ7388ss6evSo5s2b58eKAQAAAAAA8HdnVcNqyJAh6tKli9q1a+e1fuvWrcrIyFCHDh0868LDw9WyZUutWrWqpMsEAAAAAADASYT4uwBTXnvtNX355Zdau3ZtodsyMjIkSYmJiV7rExMT9euvv54wMycnRzk5OZ6fDx8+bKhaAAAAAAAAnMhZMcJq+/btuueee/Tqq68qIiLihNtZluX1s23bhdb91fjx4xUfH+9ZkpOTjdUMAAAAAABKmC3JHWBL6ZzC6uxoWK1fv1579uxR06ZNFRISopCQEK1YsULPPvusQkJCPCOrjo+0Om7Pnj2FRl391YgRI3To0CHPsn379jP6PAAAAAAAAHCWXBLYtm1bffPNN17rbr75ZtWtW1cPPPCAqlevrqSkJC1dulTnn3++JCk3N1crVqzQU089dcLc8PBwhYeHn9HaAQAAAAAA4O2saFjFxsaqYcOGXuuio6OVkJDgWT906FCNGzdOtWrVUq1atTRu3DhFRUXp+uuv90fJAAAAAACghFm2LcsOrGvsAq1eU86KhtWpGD58uI4dO6bBgwfrwIEDuvjii/XRRx8pNjbW36UBAAAAAADgL87ahlVaWprXz5ZlafTo0Ro9erRf6gEAAAAAAMCpOSsmXQcAAAAAAMDZ46wdYQUAAAAAAODFlhRoc0IFWLmmMMIKAAAAAAAAjsIIKx+c8/4OhQSFmwnLd5nJkZTVpLKxLEmq8n+/Gs1TSLDZPJMMvg8AAocdF200zzqcZTQvp1aisayaL/9uLEuSrKxjxrJciWWMZUmS5XLuf0Oari1qp7n3QZJqbMw0lpVfMc5YliSFbN5mNM8qW8ZYVnxsqLEsSbIyjxrNC8o3dN4q6WjNcsayJKnK8iNG8+wgy1hW3hsJxrIkKXX+7cayVk98zliWJNWce4fRvIpfmDvWxf9g7rgkSfkxYUbzqr3jNpblisw3kpOfl6+fjCQB/sUIKwAAAAAAADgKI6wAAAAAAEDpYNsBOIdVgNVrCCOsAAAAAAAA4Cg0rAAAAAAAAOAoXBIIAAAAAABKB7ckc9/PUDLMze0fUBhhBQAAAAAAAEehYQUAAAAAAABHoWEFAAAAAAAAR2EOKwAAAAAAUCpYti3Ltv1dRrEEWr2mMMIKAAAAAAAAjkLDCgAAAAAAAI5CwwoAAAAAAACOwhxWAAAAAACgdLDtgiWQBFq9hjDCCgAAAAAAAI5CwwoAAAAAAACOwiWBAAAAAACgdOCSwIDBCCsAAAAAAAA4Cg0rAAAAAAAAOAoNKwAAAAAAADgKc1j5wuA1r3bWUSM5khR2INdYlmS2Nkmy4mKM5plk+rkCCAx5KRWN5oXu+t1oXl60uY/poPKGj8Hlos3mGWSHmf3/OCvXbSwr5Gi+sSxJsi3LaJ47PspYVk5CuLEsSQotW8Zo3u8tKxvLyo0z+z5IlYymhe/LNpZ1tKLZPx/cYWZfu0Mp5uoLO2R4zpg8c1E1595hLkzSTzdMN5p3wS/m6susbvbzK3ZrltG8/KgIY1m50WY+v1y5jEs5KeawChj8JgMAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFGYwwoAAAAAAJQObkmmpx4808xNpxlQGGEFAAAAAAAAR6FhBQAAAAAAAEfhkkAAAAAAAFAqWLYty7b9XUaxBFq9pjDCCgAAAAAAAI5CwwoAAAAAAACOQsMKAAAAAAAAjsIcVgAAAAAAoHSw7YIlkARavYYwwgoAAAAAAACOQsMKAAAAAAAAjkLDCgAAAAAAAI7CHFYAAAAAAKB0cNuSFWBzQrkDrF5DGGEFAAAAAAAAR6FhBQAAAAAAAEfhkkAAAAAAAFA62HbBEkgCrV5DGGEFAAAAAAAAR2GElQ9c5eNkBUeYCasQbyZHUnaFcGNZkhSVe47RPEcz+D4ACBw55UKN5gXXMHvcDDnmMpZ1LNHsZ0ToEXO1Wflm/9fQFW72/+OCg93GsnLKmj31CjL82gXnhBnL2nOB2eca8bvZz+qy32cZy/qhv6Hzwj/E7DT8O5xr7n1N+PKAsSxJ+v3iskbzyn6fZyxr52Vmf4ddUeayKn5hdt+/4Jc7jOZ9+ch0Y1kde/Q1liVJtmU0TumdDO6vQWbeV/cxW3rdSBTgV4ywAgAAAAAAgKMwwgoAAAAAAJQSATiHlQKtXjMYYQUAAAAAAABHoWEFAAAAAAAAR6FhBQAAAAAAAEdhDisAAAAAAFA62AE4h1Wg1WsII6wAAAAAAADgKDSsAAAAAAAA4ChcEggAAAAAAEoHty0pwC6xcwdYvYYwwgoAAAAAAACOQsMKAAAAAAAAjkLDCgAAAAAAAI7CHFYAAAAAAKB0sN0FSyAJtHoNYYQVAAAAAAAAHIWGFQAAAAAAAByFhhUAAAAAAAAchTmsAAAAAABA6WDbBUsgCbR6DWGEFQAAAAAAAByFEVY++LVLrIIiIoxkhWRZRnIkqcryI8ayJGl7+zijefnRzu0Km3wfAASOvMZZRvNCN5o9bp6zwlx9R+qGG8uSpIPnmfu2mnqTDhnLkqTs5HijeeG7DpvL+s3st/y4Y82cjxz303Xmfoerv33UWJYkHawVbTTPcps7L6k/ZpuxLEmys8y+dgoyd55zoFM9Y1mSFLs932heZhVzf95UXXTMWJYk/do50lhW/A+ZxrIkKbN6jNG8jj36Gsta8s4rxrIkqctFXYzmnbMsylhW7OZ9RnLyXTnabiQJ8C8aVgAAAAAAoHRw25KcO5iiSAb/kyWQcEkgAAAAAAAAHIWGFQAAAAAAAByFhhUAAAAAAAAchTmsAAAAAABA6WDbBUsgCbR6DWGEFQAAAAAAAByFhhUAAAAAAAAchYYVAAAAAAAAHIU5rAAAAAAAQOlgK/DmhAqwck1hhBUAAAAAAAAchYYVAAAAAAAAHIWGFQAAAAAAAByFOawAAAAAAEDpYNsBOIdVgNVrCCOsAAAAAAAA4Cg0rAAAAAAAAOAoXBIIAAAAAABKB7dbktvfVRSPO8DqNYQRVgAAAAAAAHAUGlYAAAAAAABwFBpWAAAAAAAAcBTmsPJB9Rd+VUhQmJEsOy/PSI4k5TU411iWJFV9YYvRPCs01GieSSbfBwCBwwox+zFo5+cbzcuvVcVYVtLsr41lSVKlMDOfg5KkCuXMZUmK3LLbaJ4dbu655pePMZYlSaHb9xnNqzV6m7Gs/PNrGcuSpPJp243myeBnvx0TZSxLkhQdaTYvzNw5WOyvx4xlSVJQjtnjZuQv2cayjtYye2yq9t5RY1n5MQaPwZJit2YZzbMtc1ldLupiLkzSB198YDTvwofvMJYVE27mvMR2uYzknLVsu2AJJIFWryGMsAIAAAAAADjLTJs2TSkpKYqIiFDTpk21cuXKE267cOFCtW/fXhUqVFBcXJxSU1P14YcflmC1hdGwAgAAAAAAOIssWLBAQ4cO1UMPPaQNGzbosssuU6dOnZSenl7k9v/73//Uvn17LVq0SOvXr1fr1q3VrVs3bdiwoYQr/xMNKwAAAAAAgLPIpEmTNGDAAA0cOFD16tXT5MmTlZycrOnTpxe5/eTJkzV8+HBdeOGFqlWrlsaNG6datWrpvffeK+HK/0TDCgAAAAAAlA7H57AKtKUYcnNztX79enXo0MFrfYcOHbRq1apTynC73crMzFS5cmbn8ysOJl0HAAAAAABwuMOHD3v9HB4ervDw8ELb7d27Vy6XS4mJiV7rExMTlZGRcUqP9e9//1tZWVm65pprfC/4NDHCCgAAAAAAwOGSk5MVHx/vWcaPH3/S7S3L+ys7bdsutK4o8+fP1+jRo7VgwQJVrFjxtGo+HYywAgAAAAAApYPbllS8S+z8zl1Q7/bt2xUXF+dZXdToKkkqX768goODC42m2rNnT6FRV3+3YMECDRgwQP/3f/+ndu3anWbhp4cRVgAAAAAAAA4XFxfntZyoYRUWFqamTZtq6dKlXuuXLl2q5s2bnzB//vz56t+/v+bNm6cuXboYrd0XjLACAAAAAAA4iwwbNkx9+/ZVs2bNlJqaqhkzZig9PV233367JGnEiBHasWOH5syZI6mgWXXTTTdpypQpuuSSSzyjsyIjIxUfH++X50DDCgAAAAAA4Cxy7bXXat++fRo7dqx27dqlhg0batGiRapataokadeuXUpPT/ds//zzzys/P19DhgzRkCFDPOv79eun2bNnl3T5kmhYAQAAAACAUsK23bJtt7/LKBZf6x08eLAGDx5c5G1/b0KlpaX59BhnEnNYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUZjDCgAAAAAAlA62Lbltf1dRPHaA1WsII6wAAAAAAADgKDSsAAAAAAAA4ChcEggAAAAAAEoH25YUYJfYcUkgAAAAAAAA4H+MsPLB7+2qKjgswkhW5H6XkRxJitp5zFiWJB1pUcNo3rFywUbzTDL5PgAIIJbhPMP/+RX1W5axrKOtGxjLMi3m2wyjebnnljeaF7Z9n7kww78j+ZXKGs3b2znZWNbh6saiJEk13jD7XPPiwoxlZSWFGsuSpIRPthrNcyeWMxdm+Hc4KDvfaN6RegnGsnZfZPbcNbesuQ+dau+4jWVJUn6Umb9tjkvvZG5cxDnLooxlSdKFD99hNG/t49ONZTWYOthIjisnW/rWSBTgV4ywAgAAAAAAgKMwwgoAAAAAAJQObrdkmR2leMbZAVavIYywAgAAAAAAgKPQsAIAAAAAAICj0LACAAAAAACAozCHFQAAAAAAKB1sW8a/8vRMswOsXkMYYQUAAAAAAABHoWEFAAAAAAAAR+GSQAAAAAAAUCrYbrdsy+3vMorFtgOrXlMYYQUAAAAAAABHoWEFAAAAAAAAR6FhBQAAAAAAAEdhDisAAAAAAFA62LYk299VFI8dYPUawggrAAAAAAAAOAoNKwAAAAAAADgKDSsAAAAAAAA4CnNYAQAAAACA0sFtS1aAzQnFHFYAAAAAAACA/zHCygflvslUSHCukayg3HwjOZKUHx9hLEuSYn46ZDQvKsy5v24m3wcAgcMVG240Lzgzx2he9jmxxrKOJQQby5KkqD3mjptZDRKNZUlSyFGX0TyT9eXGmn0fQo+4jebF7DD32tlBZp+r3Gafa/jOw8ay0q8obyxLksKyqhrNCzli7n2N2G72/HB/M7OvXdy2bGNZYYeijGVJUvBRc+fCrkiz56650YbHMQSZGw0Su3mfsSxJigk3+zdJg6mDjWVtunOakZzDmW6VnWgkCvAr53YQAAAAAAAATLJtSWb/E+SM45JAAAAAAAAAwP9oWAEAAAAAAMBRaFgBAAAAAADAUZjDCgAAAAAAlAq225ZtBdacUDZzWAEAAAAAAAD+R8MKAAAAAAAAjkLDCgAAAAAAAI7CHFYAAAAAAKB0sN2S3P6uonjsAKvXEEZYAQAAAAAAwFFoWAEAAAAAAMBRuCQQAAAAAACUCrbblm3Z/i6jWGw7sOo1hRFWAAAAAAAAcBQaVgAAAAAAAHAUGlYAAAAAAABwFOawAgAAAAAApYPtluT2dxXFYwdYvYYwwgoAAAAAAACOQsMKAAAAAAAAjsIlgcVw/Ksk8105xjKDXPnGsvLNRRUw+Dwlye1yGc0zyeT7ACBwuPLNfkWw7co1mpefH2osy5Vr9hicn2fwuGn6v8/yTT/XYEdmSZKV59xLBFy5Zp+ryfMvSQoyuL+6s7ONZUlSfp7hcyaD+4Tp98GVa/i1yzeX58oxe3ByGdxdjR6DJblyzT5X9zFzn6+mf+dsw3+TuHLM/c4dzjTzS3L4SEHO8b9f4S1feVKAvTT5yvN3CX5h2fwWn7LffvtNycnJ/i4DAAAAAICT2r59u6pUqeLvMhwjOztbKSkpysjI8HcpPklKStLWrVsVERHh71JKDA2rYnC73dq5c6diY2NlWdYJtzt8+LCSk5O1fft2xcXFlWCFgPOwPwB/Yn8AvLFPAH9if4Aptm0rMzNTlStXVlAQswD9VXZ2tnJzzY6ILylhYWGlqlklcUlgsQQFBRWrQx0XF8eHDfAH9gfgT+wPgDf2CeBP7A8wIT4+3t8lOFJERESpa/oEMtqtAAAAAAAAcBQaVgAAAAAAAHAUGlZnQHh4uEaNGqXw8HB/lwL4HfsD8Cf2B8Ab+wTwJ/YHAPDGpOsAAAAAAABwFEZYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWBk2bdo0paSkKCIiQk2bNtXKlSv9XRJQIv73v/+pW7duqly5sizL0ttvv+11u23bGj16tCpXrqzIyEi1atVKmzZt8k+xwBk0fvx4XXjhhYqNjVXFihXVs2dPbdmyxWsb9geUJtOnT1fjxo0VFxenuLg4paamavHixZ7b2R9Qmo0fP16WZWno0KGedewTAFCAhpVBCxYs0NChQ/XQQw9pw4YNuuyyy9SpUyelp6f7uzTgjMvKylKTJk00derUIm+fMGGCJk2apKlTp2rt2rVKSkpS+/btlZmZWcKVAmfWihUrNGTIEK1Zs0ZLly5Vfn6+OnTooKysLM827A8oTapUqaInn3xS69at07p169SmTRv16NHD8wc4+wNKq7Vr12rGjBlq3Lix13r2CQAoYNm2bfu7iLPFxRdfrAsuuEDTp0/3rKtXr5569uyp8ePH+7EyoGRZlqW33npLPXv2lFTwP4WVK1fW0KFD9cADD0iScnJylJiYqKeeekqDBg3yY7XAmfX777+rYsWKWrFihS6//HL2B0BSuXLl9PTTT+uWW25hf0CpdOTIEV1wwQWaNm2aHn/8cZ133nmaPHkynxEA8BeMsDIkNzdX69evV4cOHbzWd+jQQatWrfJTVYAzbN26VRkZGV77R3h4uFq2bMn+gbPeoUOHJBX8gS6xP6B0c7lceu2115SVlaXU1FT2B5RaQ4YMUZcuXdSuXTuv9ewTAPCnEH8XcLbYu3evXC6XEhMTvdYnJiYqIyPDT1UBznB8Hyhq//j111/9URJQImzb1rBhw3TppZeqYcOGktgfUDp98803Sk1NVXZ2tmJiYvTWW2+pfv36nj/A2R9Qmrz22mv68ssvtXbt2kK38RkBAH+iYWWYZVleP9u2XWgdUFqxf6C0ufPOO7Vx40Z9+umnhW5jf0BpUqdOHX311Vc6ePCg3nzzTfXr108rVqzw3M7+gNJi+/btuueee/TRRx8pIiLihNuxTwAAlwQaU758eQUHBxcaTbVnz55C/0MClDZJSUmSxP6BUuWuu+7Su+++q+XLl6tKlSqe9ewPKI3CwsJUs2ZNNWvWTOPHj1eTJk00ZcoU9geUOuvXr9eePXvUtGlThYSEKCQkRCtWrNCzzz6rkJAQz+89+wQA0LAyJiwsTE2bNtXSpUu91i9dulTNmzf3U1WAM6SkpCgpKclr/8jNzdWKFSvYP3DWsW1bd955pxYuXKhly5YpJSXF63b2B6BgP8nJyWF/QKnTtm1bffPNN/rqq688S7NmzXTDDTfoq6++UvXq1dknAOAPXBJo0LBhw9S3b181a9ZMqampmjFjhtLT03X77bf7uzTgjDty5Ih++uknz89bt27VV199pXLlyuncc8/V0KFDNW7cONWqVUu1atXSuHHjFBUVpeuvv96PVQPmDRkyRPPmzdM777yj2NhYz/+Sx8fHKzIyUpZlsT+gVBk5cqQ6deqk5ORkZWZm6rXXXlNaWpqWLFnC/oBSJzY21jOn4XHR0dFKSEjwrGefAIACNKwMuvbaa7Vv3z6NHTtWu3btUsOGDbVo0SJVrVrV36UBZ9y6devUunVrz8/Dhg2TJPXr10+zZ8/W8OHDdezYMQ0ePFgHDhzQxRdfrI8++kixsbH+Khk4I6ZPny5JatWqldf6WbNmqX///pLE/oBSZffu3erbt6927dql+Ph4NW7cWEuWLFH79u0lsT8Af8c+AQAFLNu2bX8XAQAAAAAAABzHHFYAAAAAAABwFBpWAAAAAAAAcBQaVgAAAAAAAHAUGlYAAAAAAABwFBpWAAAAAAAAcBQaVgAAAAAAAHAUGlYAAAAAAABwFBpWAAAEmCuvvFKRkZE6ePDgCbe54YYbFBoaqt27dxt97HHjxuntt982mvlPLMvSnXfeaTSzVatWatWqldFMAAAAmEPDCgCAADNgwABlZ2dr3rx5Rd5+6NAhvfXWW+ratasSExONPrY/GlYAAAAofWhYAQAQYDp16qTKlStr5syZRd4+f/58HTt2TAMGDCjhynyTl5en/Px8f5cBAAAAB6FhBQBAgAkODla/fv20fv16ffPNN4VunzVrlipVqqROnTpJkjIyMjRo0CBVqVJFYWFhSklJ0ZgxYwo1iXJycjR27FjVq1dPERERSkhIUOvWrbVq1SpJBZfmZWVl6eWXX5ZlWbIsy+uyum+//VY9evRQ2bJlFRERofPOO08vv/yy12OkpaXJsiy98soruvfee3XOOecoPDxcP/300yk//+MZ8+fP10MPPaTKlSsrLi5O7dq105YtW7y2tW1bEyZMUNWqVRUREaELLrhAixcvLjL38OHDuu+++5SSkqKwsDCdc845Gjp0qLKysjzb3H777YqIiND69es969xut9q2bavExETt2rXrlJ8HAAAATizE3wUAAIDiu+WWW/Tkk09q5syZeuaZZzzrN2/erC+++EIPPviggoODlZGRoYsuukhBQUF69NFHVaNGDa1evVqPP/64tm3bplmzZkmS8vPz1alTJ61cuVJDhw5VmzZtlJ+frzVr1ig9PV3NmzfX6tWr1aZNG7Vu3VqPPPKIJCkuLk6StGXLFjVv3lwVK1bUs88+q4SEBL366qvq37+/du/ereHDh3vVP2LECKWmpuq5555TUFCQKlasWOzXYOTIkWrRooVefPFFHT58WA888IC6deum7777TsHBwZKkMWPGaMyYMRowYIB69+6t7du369Zbb5XL5VKdOnU8WUePHlXLli3122+/aeTIkWrcuLE2bdqkRx99VN98840+/vhjWZalyZMn6/PPP9c111yj9evXq0yZMhozZozS0tK0ZMkSVapUqdjPAwAAAEWwAQBAQGrZsqVdvnx5Ozc317Pu3nvvtSXZP/zwg23btj1o0CA7JibG/vXXX73uO3HiRFuSvWnTJtu2bXvOnDm2JPuFF1446WNGR0fb/fr1K7T+uuuus8PDw+309HSv9Z06dbKjoqLsgwcP2rZt28uXL7cl2ZdffvkpP09J9pAhQzw/H8/o3Lmz13avv/66LclevXq1bdu2feDAATsiIsK+8sorvbb77LPPbEl2y5YtPevGjx9vBwUF2WvXrvXa9o033rAl2YsWLfKs+/HHH+24uDi7Z8+e9scff2wHBQXZDz/88Ck/HwAAAPwzLgkEACBADRgwQHv37tW7774rqWCU1KuvvqrLLrtMtWrVkiS9//77at26tSpXrqz8/HzPcvxywRUrVkiSFi9erIiICN1yyy0+1bJs2TK1bdtWycnJXuv79++vo0ePavXq1V7rr7rqKp8e56+6d+/u9XPjxo0lSb/++qskafXq1crOztYNN9zgtV3z5s1VtWpVr3Xvv/++GjZsqPPOO8/rdbriiitkWZbS0tI829asWVMvvPCC3n77bXXt2lWXXXaZRo8efdrPBwAAAH+iYQUAQIDq3bu34uPjPZf1LVq0SLt37/aabH337t167733FBoa6rU0aNBAkrR3715J0u+//67KlSsrKMi3U4N9+/YVeTlc5cqVPbf/lYlL5xISErx+Dg8PlyQdO3bM6zGTkpIK3ffv63bv3q2NGzcWep1iY2Nl27bndTquS5cuSkxMVHZ2toYNG+a5BBEAAABmMIcVAAABKjIyUn369NELL7ygXbt2aebMmYqNjdXVV1/t2aZ8+fJq3LixnnjiiSIzjjeUKlSooE8//VRut9unplVCQkKRE47v3LnTU8dfWZZV7MfwpSapYNL5v8vIyFC1atU8P5cvX16RkZEn/ObFv9d/++23KzMzUw0aNNDdd9+tyy67TGXLljVXPAAAQCnHCCsAAALYgAED5HK59PTTT2vRokW67rrrFBUV5bm9a9eu+vbbb1WjRg01a9as0HK8YdWpUydlZ2dr9uzZJ3288PBwzwimv2rbtq2WLVvmaVAdN2fOHEVFRemSSy45/SdbTJdccokiIiI0d+5cr/WrVq3yXDZ4XNeuXfXzzz8rISGhyNfpr82tF198Ua+++qqmTp2qd999VwcPHtTNN99cEk8JAACg1GCEFQAAAaxZs2Zq3LixJk+eLNu2vS4HlKSxY8dq6dKlat68ue6++27VqVNH2dnZ2rZtmxYtWqTnnntOVapUUZ8+fTRr1izdfvvt2rJli1q3bi23263PP/9c9erV03XXXSdJatSokdLS0vTee++pUqVKio2NVZ06dTRq1CjPfFmPPvqoypUrp7lz5+qDDz7QhAkTFB8fX+KvTdmyZXXffffp8ccf18CBA3X11Vdr+/btGj16dKFLAocOHao333xTl19+uf71r3+pcePGcrvdSk9P10cffaR7771XF198sb755hvdfffd6tevn6dJ9dJLL6l3796aPHmyhg4dWuLPEwAA4GxEwwoAgAA3YMAA3XPPPapfv74uvvhir9sqVaqkdevW6bHHHtPTTz+t3377TbGxsUpJSVHHjh09l7GFhIRo0aJFGj9+vObPn6/JkycrNjZWTZo0UceOHT15U6ZM0ZAhQ3Tdddfp6NGjatmypdLS0lSnTh2tWrVKI0eO1JAhQ3Ts2DHVq1dPs2bNUv/+/Uvy5fAyduxYRUdHa9q0aXrllVdUt25dPffcc5o4caLXdtHR0Vq5cqWefPJJzZgxQ1u3blVkZKTOPfdctWvXTtWqVVNWVpauueYapaSkaNq0aZ77XnXVVRoyZIiGDx+u5s2b66KLLirppwkAAHDWsWzbtv1dBAAAAAAAAHAcc1gBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAAAAAAwFFoWAEAAAAAAMBRaFgBAAAAAADAUWhYAQAA+KBatWqaPHmyv8sAAAA4K9GwAgDgLJCRkaF77rlHNWvWVEREhBITE3XppZfqueee09GjR/1d3gnR9AEAAEBRQvxdAAAAOD2//PKLWrRooTJlymjcuHFq1KiR8vPz9cMPP2jmzJmqXLmyunfvXuR98/LyFBoaWsIVAwAAACfHCCsAAALc4MGDFRISonXr1umaa65RvXr11KhRI1111VX64IMP1K1bN8+2lmXpueeeU48ePRQdHa3HH39ckjR9+nTVqFFDYWFhqlOnjl555RXPfbZt2ybLsvTVV1951h08eFCWZSktLU2SlJaWJsuy9MEHH6hJkyaKiIjQxRdfrG+++ea0ntt7772npk2bKiIiQtWrV9eYMWOUn5/vuf3777/XpZdeqoiICNWvX18ff/yxLMvS22+/7dlmx44duvbaa1W2bFklJCSoR48e2rZtm+f2/v37q2fPnpo4caIqVaqkhIQEDRkyRHl5eZ5t9uzZo27duikyMlIpKSmaO3fuaT0vAAAAnBwNKwAAAti+ffv00UcfaciQIYqOji5yG8uyvH4eNWqUevTooW+++Ua33HKL3nrrLd1zzz2699579e2332rQoEG6+eabtXz58mLXc//992vixIlau3atKlasqO7du3s1forjww8/1I033qi7775bmzdv1vPPP6/Zs2friSeekCS53W717NlTUVFR+vzzzzVjxgw99NBDXhlHjx5V69atFRMTo//973/69NNPFRMTo44dOyo3N9ez3fLly/Xzzz9r+fLlevnllzV79mzNnj3bc3v//v21bds2LVu2TG+88YamTZumPXv2+PS8AAAA8M9oWAEAEMB++ukn2batOnXqeK0vX768YmJiFBMTowceeMDrtuuvv1633HKLqlevrqpVq2rixInq37+/Bg8erNq1a2vYsGHq1auXJk6cWOx6Ro0apfbt26tRo0Z6+eWXtXv3br311ls+PbcnnnhCDz74oPr166fq1aurffv2euyxx/T8889Lkj766CP9/PPPmjNnjpo0aaJLL73U08w67rXXXlNQUJBefPFFNWrUSPXq1dOsWbOUnp7uGR0mSWXLltXUqVNVt25dde3aVV26dNEnn3wiSfrhhx+0ePFivfjii0pNTVXTpk310ksv6dixYz49LwAAAPwz5rACAOAs8PdRVF988YXcbrduuOEG5eTkeN3WrFkzr5+/++473XbbbV7rWrRooSlTphS7jtTUVM+/y5Urpzp16ui7774rdo4krV+/XmvXrvVqQrlcLmVnZ+vo0aPasmWLkpOTlZSU5Ln9oosuKpTx008/KTY21mt9dna2fv75Z8/PDRo0UHBwsOfnSpUqeS5n/O677xQSEuL1utWtW1dlypTx6XkBAADgn9GwAgAggNWsWVOWZen777/3Wl+9enVJUmRkZKH7FHXp4N8bXrZte9YFBQV51h1XnMv8/p59qtxut8aMGaNevXoVui0iIsKrxpNlNG3atMg5pypUqOD5998nnrcsS263W9Kfz9vX5wEAAIDi45JAAAACWEJCgtq3b6+pU6cqKyvLp4x69erp008/9Vq3atUq1atXT9KfjZ1du3Z5bv/rBOx/tWbNGs+/Dxw4oB9++EF169b1qa4LLrhAW7ZsUc2aNQstQUFBqlu3rtLT07V7927PfdauXVso48cff1TFihULZcTHx59SHfXq1VN+fr7WrVvnWbdlyxYdPHjQp+cFAACAf8YIKwAAAty0adPUokULNWvWTKNHj1bjxo0VFBSktWvX6vvvv1fTpk1Pev/7779f11xzjS644AK1bdtW7733nhYuXKiPP/5YUsEorUsuuURPPvmkqlWrpr179+rhhx8uMmvs2LFKSEhQYmKiHnroIZUvX149e/Y86ePv2LGjUAPs3HPP1aOPPqquXbsqOTlZV199tYKCgrRx40Z98803evzxx9W+fXvVqFFD/fr104QJE5SZmemZdP34aKgbbrhBTz/9tHr06KGxY8eqSpUqSk9P18KFC3X//ferSpUq//j61qlTRx07dtStt96qGTNmKCQkREOHDi1y9BoAAADMYIQVAAABrkaNGtqwYYPatWunESNGqEmTJmrWrJn+85//6L777tNjjz120vv37NlTU6ZM0dNPP60GDRro+eef16xZs9SqVSvPNjNnzlReXp6aNWume+65R48//niRWU8++aTuueceNW3aVLt27dK7776rsLCwkz7+xIkTdf7553st7777rq644gq9//77Wrp0qS688EJdcsklmjRpkqpWrSpJCg4O1ttvv60jR47owgsv1MCBAz2NtIiICElSVFSU/ve//+ncc89Vr169VK9ePd1yyy06duyY4uLiTvUl1qxZs5ScnKyWLVuqV69euu2221SxYsVTvj8AAACKx7L/OiEFAACAD9LS0tS6dWsdOHDAr5ORf/bZZ7r00kv1008/qUaNGn6rAwAAAKeHSwIBAEDAeuuttxQTE6NatWrpp59+0j333KMWLVrQrAIAAAhwNKwAAEDAyszM1PDhw7V9+3aVL19e7dq107///W9/lwUAAIDTxCWBAAAAAAAAcBQmXQcAAAAAAICj0LACAAAAAACAo9CwAgAAAAAAgKPQsAIAAAAAAICj0LACAAAAAACAo9CwAgAAAAAAgKPQsAIAAAAAAICjhJgIOZybo2OufBNRPosMDlFcWLhP9509e7Zuvvlmz8/BwcGqUKGCWrZsqccee0y1atUyVSbOEvbhfdKxI/4tIjJGVlyCz3ffuHGjnnnmGaWlpWnXrl0KCQlR7dq1dd1112ngwIEqV65csfJGjx6tMWPGyLZtz7pWrVpJktLS0nyu0xdOquVsZFmWRo0apdGjR5+R/HHjxql+/frq2bOn1/q0tDS1bt1ay5cv97yfpVbeLsl10L81BJeRQisV+25//8z9O9Pvb1HHg1PRqlUr7d27V99++62xWkrKzp07NWPGDPXs2VPnnXeev8s5qTO9X1erVk2tWrXS7NmzjWcDpyPQzyWPH8vDw8O1ZcsWVa1a1ev2QD6G9u/fX2lpadq2bZtn3Zk8lhw9elQTJkxQq1atCh0Hj7/OW7duVbVq1Yw/NuB0p92wOpybo1k/fiVXMU8ETQu2LN1c6zyfm1aSNGvWLNWtW1fZ2dn67LPP9MQTT2j58uX6/vvvVbZsWYPVIpDZh/fJPeshyZXn30KCQxV08xM+nWi88MILGjx4sOrUqaP7779f9evXV15entatW6fnnntOq1ev1ltvvXUGisbZYPXq1apSpcoZyx83bpx69+5dqGF1wQUXaPXq1apfv/4Ze+yAkLdL2tZDsnP9W4cVJlV7x6emlfTnZ+7fmX5/Bw4cqI4dOxrNdLqdO3dqzJgxqlatmuMbVmfaW2+9pbi4OH+XAXg5G84lj8vJydHDDz+sV155xWBhznMmjyVHjx7VmDFjJKlQw6pLly5avXq1KlXy7bMWCHSn3bA65sr3e7NKkly2rWOufMXJ94ZVw4YN1axZM0kFBwuXy6VRo0bp7bffPun/BqOUOXbE/ycYUkENx45IxTzJWL16te644w61b99eb7/9tsLD/9xn2rdvr3vvvVdLliwxXS3OIpdccolfHjcuLs5vj+0oroP+b1ZJBTW4DvrcsPrrZ+6ZcPToUUVFRalKlSpntMEKZzv//PP9XQJQWICfS/5Vx44dNW/ePN13331q0qSJweL+dOzYMUVGRp6R7FPlr2NJhQoVVKFCBb88NuAEzGF1EsdPpHfv3u3nSgBzxo0bJ8uyNGPGDK9m1XFhYWHq3r2717oFCxYoNTVV0dHRiomJ0RVXXKENGzYYrcvtdus///mPzjvvPEVGRqpMmTK65JJL9O6775Z4LfPmzVNqaqpiYmIUExOj8847Ty+99JLXNjNnzlSTJk0UERGhcuXK6corr9R3333ntU3//v0VExOj77//XldccYWio6NVqVIlPfnkk5KkNWvW6NJLL1V0dLRq166tl19+2ev+s2fPlmVZWrZsmW699VYlJCQoLi5ON910k7KyspSRkaFrrrlGZcqUUaVKlXTfffcpL+/PE+C0tDRZllXoMsht27bJsiyvYe3Ha/3pp5/UuXNnxcTEKDk5Wffee69ycnK87m9ZVqHLAXfs2KHbbrtNycnJCgsLU+XKldW7d2/P8TM7O1v33nuvzjvvPMXHx6tcuXJKTU3VO++8Uyg7KytLL7/8sizLkmVZXpdzFvV83n33XaWmpioqKkqxsbFq3769Vq9e7bXN6NGjZVmWNm3apD59+ig+Pl6JiYm65ZZbdOjQIcGZLMvSnXfeqVdeeUX16tVTVFSUmjRpovfff99ru+Pv75dffqnevXurbNmyqlGjhtdtf3cq+7kkrV27VpdddpmioqJUvXp1Pfnkk3K73Z7bj/9ezps3Tw888IAqVaqkmJgYdevWTbt371ZmZqZuu+02lS9fXuXLl9fNN9+sI0e8LwOybVvTpk3zHP/Kli2r3r1765dffvHarlWrVmrYsOFJa0pLS9OFF14oSbr55ps9+9HJLuH9/fffNXjwYNWvX18xMTGqWLGi2rRpo5UrV3ptd/zYMXHiRE2aNEkpKSmKiYlRamqq1qxZ47XtunXrdN1116latWqKjIxUtWrV1KdPH/36668nrEOSXnnlFVmWVWgflqSxY8cqNDRUO3fulCRt2LBBXbt2VcWKFRUeHq7KlSurS5cu+u233zz3qVatmvr37+/52e126/HHH1edOnU8nzWNGzfWlClTTloXgKINHz5cCQkJeuCBB066XXZ2tkaMGKGUlBSFhYXpnHPO0ZAhQ3Tw4EGv7apVq6auXbtq4cKFOv/88xUREaExY8YYOdb+97//1eWXX66KFSsqOjpajRo10oQJE7zOnU7k78eSVq1aeY6vf1+On1+dyrF127ZtnobUmDFjPBnHH+v4ueBfL0+UinceeirndoBT0bA6ia1bt0qSateu7edKADNcLpeWLVumpk2bKjk5+ZTuM27cOPXp00f169fX66+/rldeeUWZmZm67LLLtHnzZmO19e/fX/fcc48uvPBCLViwQK+99pq6d+/u9QFdErU8+uijuuGGG1S5cmXNnj1bb731lvr16+f1R9b48eM1YMAANWjQQAsXLtSUKVO0ceNGpaam6scff/TKy8vLU69evdSlSxe988476tSpk0aMGKGRI0eqX79+uuWWW/TWW2+pTp066t+/v9avX1+opoEDByo+Pl6vvfaaHn74Yc2bN0+33nqrunTpoiZNmuiNN95Qv3799O9//1v/+c9/fH7ueXl56t69u9q2bat33nlHt9xyi5555hk99dRTJ73fjh07dOGFF+qtt97SsGHDtHjxYk2ePFnx8fE6cOCApIJLBvbv36/77rtPb7/9tubPn69LL71UvXr10pw5czxZq1evVmRkpDp37qzVq1dr9erVmjZt2gkfe968eerRo4fi4uI0f/58vfTSSzpw4IBatWqlTz/9tND2V111lWrXrq0333xTDz74oObNm6d//etfPr5iOB0ul0v5+flei8vlKrTdBx98oKlTp2rs2LF68803PSfmf2/mSFKvXr1Us2ZN/d///Z+ee+65Ez72qeznkpSRkaEbbrhBN954o959913P/vvqq68Wyhw5cqT27Nmj2bNn69///rfS0tLUp08fXXXVVYqPj9f8+fM1fPhwvfLKKxo5cqTXfQcNGqShQ4eqXbt2evvttzVt2jRt2rRJzZs3L/SfZv9U0wUXXKBZs2ZJkh5++GHPfjRw4MATvh779++XJI0aNUoffPCBZs2aperVq6tVq1ZFzvv33//+V0uXLtXkyZM1d+5cZWVlqXPnzl7N323btqlOnTqaPHmyPvzwQz311FPatWuXLrzwQu3du/eEtVx77bVKSkrSf//7X6/1+fn5ev7553XllVeqcuXKysrKUvv27bV7926ves4991xlZmaeMH/ChAkaPXq0+vTpow8++EALFizQgAEDCv3RDODUxMbG6uGHH9aHH36oZcuWFbmNbdvq2bOnJk6cqL59++qDDz7QsGHD9PLLL6tNmzaFmidffvml7r//ft19991asmSJrrrqKs9tp3Os/fnnn3X99dfrlVde0fvvv68BAwbo6aef1qBBg4r9vKdNm+Y5vh5f2rVrp+DgYNWpU0fSqR1bK1Wq5LmyYcCAAZ6sRx555ISPXdzzUF/O7QCnMDLp+tni+Mnz8TmsHn/8cV1++eWFRpsAgWrv3r06evSoUlJSTmn77du3a9SoUbrzzjv17LPPeta3b99etWrV0pgxY7RgwYLTrmvlypV65ZVX9NBDD+nxxx/3rP/rvDMlUcvWrVs1btw43XDDDV5/kLZv397z74MHD+qxxx5T586dNW/ePM/6Vq1aqVatWho9erTmzp3rWZ+bm6vHH39cvXr18mz3/vvva/z48fryyy89Q8ybNWumihUrat68eWratKlXXV27dtXEiRM9taxevVrz58/XpEmTPM2Wdu3a6cMPP9TcuXM1bNgwn55/bm6uxowZo6uvvlqS1LZtW61bt07z5s3To48+esL7Pfroo9q7d6++/vpr1atXz7P+mmuu8fw7Pj7e80e0VHC8bdu2rQ4cOKDJkyfrpptuklRwuWFQUJAqVKjwj5f/ud1u3X///WrUqJEWL16soKCC/4Pp3LmzatSooQceeECfffaZ130GDBig+++/X1LBa/bTTz9p5syZeumll4ochYMzp6j3Nzg4WPn53l/icuzYMX388ceKjY2VVNCQqVy5sl5//XU9+OCDXtv269fPMw/IiZzKfn7cvn37tGjRIl100UWSCn5n0tLSNG/ePM/v7HGNGzf2+h3//vvvNXnyZN199916+umnPY+xevVqzZ0713McW7NmjV544QX9+9//9tp3L7vsMtWuXVuTJk3y+sPin2qKi4tTw4YNJUk1atQ4pcto69Sp49UYdrlcuuKKK7Rt2zY9++yzheZUiY2N1fvvv6/g4GBJUuXKlXXRRRdp8eLFuu666yRJvXv3Vu/evb0yu3btqsTERM2bN0933313kbWEhYVp0KBBGj9+vCZNmqSKFStKkhYuXKidO3fqzjvv9Ly++/bt00svvaQePXp47v/X405RPvvsMzVq1MhrxNkVV1zxD68QgJO5/fbbNWXKFD3wwAP64osvCn2efvTRR/rwww81YcIEz2dw+/btlZycrGuvvVZz5szRrbfe6tl+z5492rx5s9eggeMNHl+PtZI0adIkz7/dbrcuu+wyJSQk6Oabb9a///3vYs1Z/Pf5FidOnKiPP/5YM2bMUGpqqqRTO7aGh4d7zvuqVKnyj8dsX85DfTm3A5yCEVZ/cckllyg0NFSxsbHq2LGjypYtq3feeUchIfT1UDp9+OGHys/P10033eQ1CiIiIkItW7Ys9jfu/X1ExfFLWBYvXixJGjJkSInVUpSlS5fK5XKdtI7Vq1fr2LFjXsPCJSk5OVlt2rTRJ5984rXesix17tzZ83NISIhq1qypSpUqec2HUK5cOVWsWLHIy2W6du3q9fPxplCXLl0Krf+ny21OxrIsdevWzWtd48aN/zFz8eLFat26tVezqij/93//pxYtWigmJkYhISEKDQ3VSy+9VGgI+6nasmWLdu7cqb59+3qaVZIUExOjq666SmvWrNHRo0e97vP3/4Bo3LixsrOztWfPHp9qgO/mzJmjtWvXei2ff/55oe1at27taVZJUmJi4gn3lb/+L/yJnMp+flxSUpKnMXTcifaJ4uyn+/fv91yq8v7778uyLN14441ex7akpCQ1adKk0LGtODUVx3PPPacLLrhAERERnv3zk08+KXL/7NKli6dZdfzxJXnVcOTIET3wwAOqWbOmQkJCFBISopiYGGVlZf3jPn/HHXdIKviCkOOmTp2qRo0a6fLLL5ck1axZU2XLltUDDzyg55577pRH2V500UX6+uuvNXjwYH344Yc6fPjwKd0PwImFhYXp8ccf17p16/T6668Xuv34yKu/nztdffXVio6OLnTu1Lhx4xNe4eLrsVYquIy4e/fuSkhIUHBwsEJDQ3XTTTfJ5XLphx9+OLUnW4Tjo7oefvhhr8abVLxj66nw5TzUl3M7wCloWP3F8ZPnZcuWadCgQfruu+/Up08ff5cFGFO+fHlFRUV5Lnf9J8cvRbnwwgsVGhrqtSxYsOCkl3UUpW3btl4Zt9xyi6SCa/yDg4OVlJRUYrUU5ffff5ekk07QvG/fPkkq8ttaKleu7Ln9uKioKEVERHitCwsLU7ly5QrdPywsTNnZ2YXW/33bsLCwE64v6v6nqqhaw8PD/zHz999//8dJrRcuXKhrrrlG55xzjl599VWtXr1aa9eu1S233OJzzf/0Xrjdbs8licclJHhPLHt8Hrdjx475VAN8V69ePTVr1sxr+fvoQqnweyYVvG9FvWen8i1Kp7Kf+/LYxdlPJXl+73fv3i3btpWYmFjo2LZmzZpCx7bi1HSqJk2apDvuuEMXX3yx3nzzTa1Zs0Zr165Vx44di8w9lf3o+uuv19SpUzVw4EB9+OGH+uKLL7R27VpVqFDhH2tNTEzUtddeq+eff14ul0sbN27UypUrPaOrpIJRmytWrNB5552nkSNHqkGDBqpcubJGjRp10vloRowYoYkTJ2rNmjXq1KmTEhISPCMOAPjuuuuu0wUXXKCHHnqo0D64b98+hYSEFJo83LIsJSUlFTp3Otmx3NdjbXp6ui677DLt2LFDU6ZM0cqVK7V27VrP5ce+HkOXL1+u/v3766abbtJjjz3mdVtxj62nwsR56Kmc2wFOwdChvzh+8iwV/I+uy+XSiy++qDfeeMNrWDsQqIKDg9W2bVstXrxYv/322z/+wVa+fHlJ0htvvKGqVaue9uM///zzXnOLHM+vUKGCXC6XMjIyTniSYrqWohw/kfrtt99OOMfX8T/Udu3aVei2nTt3eup0guMnKH+fG8JEc++vKlSo4DXJcVFeffVVpaSkaMGCBV6XCpzOpJ//9F4EBQUVa3g/At+pXNZ5Kvt5SSpfvrwsy9LKlSuL/CKMotaZ9uqrr6pVq1aaPn261/qTzQV1MocOHdL777+vUaNGeV22eXwuu1Nxzz336JVXXtE777yjJUuWqEyZMrrhhhu8tmnUqJFee+012batjRs3avbs2Ro7dqwiIyMLXS56XEhIiIYNG6Zhw4bp4MGD+vjjjzVy5EhdccUV2r59u6Kionx6zkBpZ1mWnnrqKbVv314zZszwui0hIUH5+fn6/fffvZpWtm0rIyPD80URf80y7e2331ZWVpYWLlzodR751Vdf+Zy5ceNG9ezZUy1btvQaEXqc6WOrFFjnoYAJjLA6iQkTJqhs2bJ69NFHvb4RCAhkI0aMkG3buvXWW5Wbm1vo9ry8PL333nuSCub1CAkJ0c8//1xoJMTxpTjq1Knjdd9q1apJkjp16iRJhT7Q/8p0LUXp0KGDgoODT1pHamqqIiMjC026/Ntvv2nZsmVq27btaddhyvHXd+PGjV7r//7Ni6erU6dOWr58ubZs2XLCbSzLUlhYmNdJaEZGRqFvCZROfbRInTp1dM4552jevHmybduzPisrS2+++abnmwOBvzqV/bwkde3aVbZta8eOHUUe1xo1alTszOKOHLQsq1BjbOPGjUV+U9+p5tm2XSjzxRdfLHJi/aI0bdpUzZs311NPPaW5c+eqf//+io6OPuHjNWnSRM8884zKlCmjL7/88pQeo0yZMurdu7eGDBmi/fv3F/oWLgDF065dO7Vv315jx471uhTv+LnR38+d3nzzTWVlZZXIudPx84+/Hpds2y6y0XQq0tPT1alTJ1WvXl1vvvmmQkNDi3zMUzm2FueYHUjnoYAJjLA6ibJly2rEiBEaPny45s2bpxtvvNHfJQGnLTU1VdOnT9fgwYPVtGlT3XHHHWrQoIHy8vK0YcMGzZgxQw0bNlS3bt1UrVo1jR07Vg899JB++eUXz9xuu3fv1hdffKHo6Oh/nOD4VFx22WXq27evHn/8ce3evVtdu3ZVeHi4NmzYoKioKN11110lUku1atU0cuRIPfbYYzp27Jj69Omj+Ph4bd68WXv37tWYMWNUpkwZPfLIIxo5cqRuuukm9enTR/v27dOYMWMUERGhUaNGnfbrYUpSUpLatWun8ePHq2zZsqpatao++eQTLVy40OjjjB07VosXL9bll1+ukSNHqlGjRjp48KCWLFmiYcOGqW7dup6vqB48eLB69+6t7du367HHHlOlSpUKfaNNo0aNlJaWpvfee0+VKlVSbGys5xt3/iooKEgTJkzQDTfcoK5du2rQoEHKycnR008/rYMHD+rJJ580+jxh1rfffltognWpYKLwv182YtKp7OclqUWLFrrtttt08803a926dbr88ssVHR2tXbt26dNPP1WjRo08czqdqho1aigyMlJz585VvXr1FBMTo8qVK6ty5cpFbt+1a1c99thjGjVqlFq2bKktW7Zo7NixSklJKfI9+idxcXG6/PLL9fTTT6t8+fKqVq2aVqxYoZdeekllypQ55Zx77rlH1157rSzL0uDBg71ue//99zVt2jT17NlT1atXl23bWrhwoQ4ePFjkBPrHdevWTQ0bNlSzZs1UoUIF/frrr5o8ebKqVq2qWrVqFfu5AvD21FNPqWnTptqzZ48aNGggqWAS9CuuuEIPPPCADh8+rBYtWmjjxo0aNWqUzj//fPXt2/eM19W+fXuFhYWpT58+Gj58uLKzszV9+vRCUwecqk6dOungwYOaOnWqNm3a5HXb8c+xUz22xsbGqmrVqnrnnXfUtm1blStXznPs/LtAOg8FTKBh9Q/uuusuz9dp9+nTx2uSUSBQ3Xrrrbrooos8X2ubkZGh0NBQ1a5dW9dff73XPCEjRoxQ/fr1NWXKFM2fP185OTlKSkrShRdeqNtvv91YTbNnz9YFF1ygl156SbNnz1ZkZKTq16///+3deXwN1//48dfNfpPYEiL2kNpii30JlUUsiZ0WscVesZa2RJGEloraKwlFLEUpsaaoJaHaIEr5lZZuoUpIoypEIsv9/eFz5+u6WW42ifb9fDzyeOTOPXPmzNw5Z2bOnEVnSuKXkZZ58+ZRu3ZtVq1axeDBgzExMaF27do6M1r5+/tjZ2fHypUr2bFjB2q1GldXVxYsWFDiHni2bNnCpEmTmDFjBhkZGfTo0YPt27cXSos0rSpVqnDu3DkCAgL46KOPSExMpEKFCrRv314ZU2LEiBHcu3ePsLAwNmzYQK1atZg5cya3bt3SqyBYsWIFEyZMYODAgSQnJ+c4qL6Pjw9WVlYsXLiQAQMGYGxsTJs2bYiKiqJdu3aFto+i8I0YMSLL5Z9++imjR48u0m0bks9fpjVr1tCmTRvWrFlDSEgImZmZVK5cGRcXF70B1g1haWnJhg0bCAoKonPnzqSlpREQEKAzM97z3n//fZKTk1m/fj3BwcE4OTkRFhbGnj178j2hxbZt25gyZQrvvfce6enpuLi4cPToUb2BkXPSu3dvzM3NcXNz0ytba9euTdmyZQkODub27duYmZlRt25dNm7cyPDhw7ON083Njd27d7Nu3ToePnyIvb09np6ezJkzJ8sWEkKIvGnatCmDBg3SmcFOpVKxd+9eAgMDCQ8P58MPP6R8+fIMHTqUBQsWvJSuz/Xq1WP37t3Mnj2bvn37Ymtri4+PD9OmTVNa+ueFdqIH7SzQzwsPD8fX1zdPZev69et599136dmzJ6mpqQwfPpyNGzdmue1X6T5UiIJSaZ7vR5EPD5+mEv7z92QULJoCM1apGFHbmdJmRV/gif82zcNEMsPfh4zsB3V9KYxNMRrxIarS+gPwCiH+xdLuQFwv0Oh36X2pVGbgsA9Mcx/oXIj8OHDgAD179iQyMlJntlUhXnVyLymEEIYpcIUVPKu0epKR9ybjhUltbCKVVeKl0TxMhCePcg9YlNTWcoMhxH9V2h3IeFC8aTAuK5VVokhcvXqVGzduMGXKFKysrLhw4UKRDMIsRHGSe0khhMhdoVRYCSGEEEIIURhcXV355ptvaNasGZs2baJevXrFnSQhhBBCFAOpsBJCCCGEEEIIIYQQJYpRcSdACCGEEEIIIYQQQojnSYWVEEIIIYQQQgghhChRpMJKCCGEEEIIIYQQQpQoUmElhBBCCCGEEEIIIUoUqbASQgghhBBCCCGEECWKVFgJIYQQQgghhBBCiBJFKqyEEEIIIYQQQgghRIliUjjRPAGeFk5U+WYGqIs5DeK/4uGjVJ6kphdrGtTmJpS2Ni/WNAghikd6ahKZaSnFmgYjUwtMzEsVaxqEEOJVJfeSQgiRu0KosHoCRAGZBY+qQIwAN0pCpdWCBQtwcnKid+/eOss3btzIiBEjiI2NpUWLFsWTOFFgDx+lEr73BzIyNMWaDmNjFSN6NyzQjcbly5dZtmwZ0dHR3LlzBxMTE+rUqcPAgQMZPXo0NjY2uLq6AhAdHV04CX9BSEgIlpaW+Pr6Fkn8KpWKgIAAAgMDiyR+IV629NQk7n23FTQZxZsQlTF2zQfnudJKey3MyvTp0/n4448LI3U6rl69ys6dO/H19cXBwaHQ4y8p4uLiqFmzJuHh4UVSpiYnJxMcHIyrq6tybdDS/q6///77v/oYC1EY/k33kmfOnGHJkiWcPn2axMREbG1t6dChA++++y4tW7YsxNRmzdfXl+joaOLi4pRlDg4OuLq6snHjRgBu377N2rVr6d27N87OzkWepqK2bds27t27x9SpUws9binLRUlTCBVWTyn+yip4loanlJQKq/79++tVWIl/hyep6cV+gwGQkaHhSWp6vm8yPv30U/z8/Khbty7vvvsuTk5OpKWlcf78ecLCwoiJiWHPnj2FnGp9ISEhlC9fvsgqrGJiYqhatWqRxC1EcchMSyn+yioATcaztOSzlVV4eDj16tXTWVa5cuXCSJmeq1evEhQUhKurq9yAF0BycjJBQUEAehVW3t7exMTEUKlSpWJImRCvln/LveSqVauYOnUqrVq1Ijg4mBo1anDz5k1Wr15NmzZtCA0NZezYsYWc6tzt2bOH0qVLK59v375NUFAQDg4O/5oKqx9++KFIKqyEKGkKqUugEOJVEhMTw/jx4/H09GTv3r2Ym//fjYqnpyfTp0/n8OHDxZjCgtFoNKSkpKBWq2nTpk1xJ0cIkYWGDRu+8q2N09LSUKlUmJjI7VSFChWoUKFCcSdDCPGSfPPNN0ydOhUvLy/27NmjUw4OHDiQPn364OfnR9OmTV9KS6vnNW3a9KVuryTLyMggPT1d515fiFeJDLoOBAYGolKpuHjxIn379qV06dKUKVOGIUOGkJCQAMCoUaOwsbEhOTlZb313d3caNGgAPOt+9PjxYzZt2oRKpUKlUum9hUxKSmL8+PGUL18eW1tb+vbty+3bt3XCZGZmEhwcTL169TA3N8fOzo5hw4Zx69YtnXCurq40bNiQ2NhYOnTogKWlJbVq1eKjjz4iM7MktHwTJdGCBQtQqVSsXbs2ywuYmZkZPXv2zHLd6OhoVCqVXhfBuLg4VCqV0vwa4LfffmPgwIFUrlwZc3NzKlasiIeHB99//z3wrMn2lStXOHnypJJfnm/98PDhQ9555x1q1qyJmZkZVapUYerUqTx+/Fhn2yqViokTJxIWFkb9+vUxNzdn06ZNynfPdwfcuHEjKpWKqKioXPNhamoq06dPx97eHktLS15//XW+++47HBwciqxFmBDimR07dtC2bVusrKywtramS5cuXLx4USfM+fPnGThwIA4ODqjVahwcHBg0aBA3btxQwmzcuJE33ngDADc3N6Ws0ZZV2eXnF7u9acu+LVu2MH36dKpUqYK5uTm//PILAMeOHcPDw4PSpUtjaWmJi4sLx48fz3U/MzMz+eCDD6hbty5qtZqyZcvSuHFjVqxYoRPu559/xsfHBzs7O8zNzalfvz6rV6825FAavO6DBw+YPn06tWrVUu49vLy8+Omnn4iLi1MqpIKCgpTjqD122rL1+W45ABs2bKBJkyZYWFhgY2NDnz59+PHHH3XC+Pr6Ym1tzS+//IKXlxfW1tZUq1aN6dOnk5qaatA+CiFeroULF6JSqQgNDdWrtDcxMSEkJEQJp5Vdt2zts9jzVq9ezeuvv46dnR1WVlY0atSI4OBg0tLSck3b8+V6dHS0UmE2YsQIpewKDAxky5YtqFQqYmJi9OKYN28epqameveGL/rpp58YNGgQFStWxNzcnOrVqzNs2DCdsuuHH36gV69elCtXDgsLC5ydnZX7VC3tNWb79u28//77VK5cmdKlS9OpUyeuXbumhHN1dSUyMpIbN24o+6I9dtp78eDgYD744ANq1qyJubk5UVFRAOzfv5+2bdtiaWlJqVKl8PT0zHLfhShJpMLqOX369OG1115j165dBAYGsnfvXrp06UJaWhpTpkzh77//Ztu2bTrrXL16laioKCZMmAA8a7miVqvx8vIiJiaGmJgYpcDWGj16NKampmzbto3g4GCio6MZMmSITpjx48czY8YMPD092b9/P/Pnz+fw4cO0a9eOv/76SydsfHw8gwcPZsiQIezfv59u3brh7+/PZ599VgRHSbzqMjIyOHHiBM2bN6datWpFui0vLy++++47goODOXr0KKGhoTRt2pQHDx4Az5ps16pVi6ZNmyr5RdsNMTk5mY4dO7Jp0yYmT57MoUOHmDFjBhs3bqRnz55oNLpN6ffu3UtoaChz587lyJEjdOjQIce0GZIPR4wYwfLlyxkxYgT79u2jX79+9OnTR0m/ECL/tG99n//TWrBgAYMGDcLJyYmdO3eyZcsWkpKS6NChA1evXlXCxcXFUbduXZYvX86RI0dYtGgRd+7coWXLlsq10tvbmwULFgDPHoC0ZY23t3e+0u3v78/NmzcJCwvjwIED2NnZ8dlnn9G5c2dKly7Npk2b2LlzJzY2NnTp0iXXSqvg4GACAwMZNGgQkZGR7Nixg1GjRumUM1evXqVly5b88MMPLFmyhIMHD+Lt7c3kyZOVLnrZMXTdpKQk2rdvz5o1axgxYgQHDhwgLCyMOnXqcOfOHSpVqqS0vB01apRyHOfMmZPtthcuXMioUaNo0KABERERrFixgsuXL9O2bVt+/vlnnbBpaWn07NkTDw8P9u3bx8iRI1m2bBmLFi3Kcf+EEC9fRkYGUVFRtGjRItthF6pVq0bz5s05duxYvl6i//rrr/j4+LBlyxYOHjzIqFGjWLx4MePGjctTPM2aNSM8PByA2bNnK2XX6NGjGTBgAPb29noV+Onp6axZs4Y+ffrk2FX90qVLtGzZkjNnzjBv3jwOHTrEwoULSU1N5enTZxOSXbt2jXbt2nHlyhVWrlxJREQETk5O+Pr6EhwcrBfnrFmzuHHjBuvWrWPt2rX8/PPP9OjRg4yMZ0MBhISE4OLigr29vbIvL1Y6rVy5khMnTvDxxx9z6NAh6tWrx7Zt2+jVqxelS5dm+/btrF+/nr///htXV1dOnz6dp2MqxMskbdif07dvX6Xg6Ny5MxUrVmTw4MHs3LmTwYMH07FjR1avXs3o0aOVdT755BNKly7NsGHDAGjTpg1GTidwHQAAUPlJREFURkZUqFAh265IXbt2ZeXKlcrn+/fv89577xEfH4+9vT0//fQTa9euxc/Pj1WrVinhmjZtSuvWrVm2bBkffvihsjwxMZEvv/ySVq1aAdCpUyeio6PZtm2bki4htP766y+Sk5OpWbNmkW4nMTGRa9eusXz5cp2KoL59+yr/N23aFLVaTenSpfXyy8qVK7l8+TJnz55Vug15eHhQpUoV+vfvz+HDh+nWrZsS/tGjR/y///f/KFeunEHpyy0fXr16le3btzNjxgzl7aCnpycVK1Zk0KBBeT8gQggdWV0j09LSuHPnDgEBAUycOFEnj3p6elK7dm2CgoLYsWMHAP3796d///5KmIyMDLp3707FihXZtm0bkydPpkKFCtSuXRsAJyenAncTdnR05IsvvlA+JycnM2XKFLp3764z7p+XlxfNmjVj1qxZnD17Ntv4vvnmGxo1aqTTErRLly46YaZNm0apUqU4ffq0Mi6Lp6cnqampfPTRR0yePDnbss/QdZcvX86VK1c4evQonTp1UtZ/vsxu3rw5AFWrVs31OD548ID58+fj5eWl87LP1dWV2rVrExgYyNatW5XlT58+JSgoSGkN5+Hhwfnz59m2bRtz587NcVtCiJfL0HvJmjVrcu7cOe7fv0/58uXztI2lS5cq/2dmZtKhQwdsbW0ZMWIES5YsMfh+r3Tp0jRs2BB4Vn6/WHaNGzeOhQsXsnTpUuzs7ACIiIjg9u3bTJw4Mce4p02bhomJCefOndPpEj148GDl/8DAQJ4+fUpUVJTyotjLy4sHDx4QFBTEuHHjKFOmjBLeyclJp9GBsbExb775JrGxsbRp0wYnJyfKli2Lubl5tuWwhYUFR44cwdTUFHh2/FxcXGjUqBGHDh3CyMhISYejoyMzZszgm2++yflAClFMpIXVc54vXADefPNNTExMlGaUU6ZM4fvvv1cy9MOHD9myZQvDhw/H2tra4O282NWqcePGAEoXBu32Xuyi0KpVK+rXr6/3ttbe3l6prHo+zue7RAjxstnY2ODo6MjixYtZunQpFy9ezNMbtoMHD9KwYUOcnZ11WmB06dIlyy6J7u7uBt+8QO758OTJk8CzcuB5/fv3l/FqhCgEmzdvJjY2VufPxMSEI0eOkJ6ezrBhw3TyvoWFBR07dtTJ+48ePWLGjBm89tprmJiYYGJigrW1NY8fP9brdlZY+vXrp/P522+/5f79+wwfPlwnvZmZmXTt2pXY2Fi9bszPa9WqFZcuXcLPz48jR47w8OFDne9TUlI4fvw4ffr0wdLSUmcbXl5epKSkcObMmSzjzsu6hw4dok6dOjqVVQURExPDkydP9O5lqlWrhru7u969jEqlokePHjrL5F5GiFebtjX8i939DHHx4kV69uyJra0txsbGmJqaMmzYMDIyMrh+/XqhpXH8+PHAs8mItD755BMaNWrE66+/nu16ycnJnDx5kjfffDPH8ftOnDiBh4eHXq8GX19fkpOT9VpH5XZ/aoiePXsqlVXwrJXX7du3GTp0qFJZBWBtbU2/fv04c+ZMlsPeCFESSIXVc+zt7XU+m5iYYGtrS2JiIgC9evXCwcFBaTa6ceNGHj9+rHQHNJStra3OZ+0YQk+ePAFQtpfVTDuVK1dWvs8uPm2c2viEeF758uWxtLTk999/L9LtqFQqjh8/TpcuXQgODqZZs2ZUqFCByZMnk5SUlOv6d+/e5fLly5iamur8lSpVCo1Go9c1Nq8zUxmaDytWrKgTTlsuCCEKpn79+rRo0ULnD57lfYCWLVvq5f8dO3bo5H0fHx8++eQTRo8ezZEjRzh37hyxsbFUqFChyK6BL5Y12vT2799fL72LFi1Co9Fw//79bOPz9/fn448/5syZM3Tr1g1bW1uldRE8K4vS09NZtWqVXvxeXl4AeuWhVl7WTUhIKNQZVfN6L2NpaYmFhYXOMnNzc1JSUgotTUKIwmHovWRcXBxqtTrP9003b96kQ4cO/Pnnn6xYsYKvv/6a2NhY5RmsMMv3ihUrMmDAANasWUNGRgaXL1/m66+/zrV11d9//01GRkau5WZiYmK25aD2++fldn9qiBe3l1t5nJmZyd9//21w/EK8TNJM4Dnx8fFUqVJF+Zyenk5iYqJScBgZGTFhwgRmzZrFkiVLCAkJwcPDg7p16xZqOrTbu3Pnjl4hePv27Tw3qRXiecbGxnh4eHDo0CFu3bqV5wcU7QPFiwPhZvXAVKNGDdavXw/A9evX2blzp9I0OiwsLMftlC9fHrVazYYNG7L9/nn5eXuXE20+vHv3bpblghCiaGjz9q5du6hRo0a24f755x8OHjxIQEAAM2fOVJanpqbmWEH0IgsLiywH9v7rr7+yvN6+WNZow6xatSrb7hkvVnw/z8TEhGnTpjFt2jQePHjAsWPHmDVrFl26dOGPP/6gXLlyGBsbM3To0GxfkGXXLScv61aoUEFvYpeCeP5e5kVyLyPEq83Y2Bh3d/cc7yVv3brFd999R9euXZVlOZW3z9u7dy+PHz8mIiJC5zqgnbSnsE2ZMoUtW7awb98+Dh8+TNmyZfV63rzIxsYGY2PjXMtNW1vbbMtB0L+fLQwvXqdyK4+NjIzy1EtBiJdJWlg95/mxFAB27txJenq6zixBo0ePxszMjMGDB3Pt2rUsa98L2rrJ3d0dQG/Q9NjYWH788Uc8PDzyHbcQ8OyNvkajYcyYMcqgkM9LS0vjwIEDWa6rnd3l8uXLOsv379+f4zbr1KnD7NmzadSoERcuXFCWZ5dfunfvzq+//oqtra1eK4wWLVpkOctMYdI2A9eOlaO1a9cuncGhhRCFq0uXLpiYmPDrr79mmfe1LbFUKhUajUZvptN169Ypg9Nq5fSG2sHBQa88u379us6sTDlxcXGhbNmyXL16Ndv0mpmZGRRX2bJl6d+/PxMmTOD+/fvExcVhaWmJm5sbFy9epHHjxlnGn13rhbys261bN65fv86JEyeyTV9e3vS3bdsWtVqtdy9z69YtpYuMEOLVNXPmTDQaDX5+fnplbkZGBuPHjycjI4MpU6Yoyx0cHLh3757SMhWejV935MgRnfW1FS7Pl+8ajUan215e5FZ2NW/enHbt2rFo0SK2bt2Kr68vVlZWOcapVqvp2LEjX3zxRbatXOHZeHwnTpzQm21w8+bNWFpa5mtcxbw+a9atW5cqVaqwbds2nUmLHj9+zO7du5WZA4UoiaSF1XMiIiIwMTHB09OTK1euMGfOHJo0aaIzhk3ZsmUZNmwYoaGh1KhRQ2+8BYBGjRoRHR3NgQMHqFSpEqVKlcpTK6y6desyduxYVq1ahZGREd26dSMuLo45c+ZQrVo13n777ULZX/Hf1bZtW0JDQ/Hz86N58+aMHz+eBg0akJaWxsWLF1m7di0NGzbM8vy2t7enU6dOLFy4kHLlylGjRg2OHz9ORESETrjLly8zceJE3njjDWrXro2ZmRknTpzg8uXLOq0hGjVqxOeff86OHTuoVasWFhYWNGrUiKlTp7J7925ef/113n77bRo3bkxmZiY3b97kq6++Yvr06bRu3brIjlGDBg0YNGgQS5YsUd4kXrlyhSVLllCmTBmdMQCEEIXHwcGBefPm8f777/Pbb7/RtWtXypUrx927dzl37hxWVlYEBQVRunRpXn/9dRYvXkz58uVxcHDg5MmTrF+/nrJly+rEqR1wd+3atZQqVQoLCwtq1qyJra0tQ4cOZciQIfj5+dGvXz9u3LhBcHBwjmOSPM/a2ppVq1YxfPhw7t+/T//+/bGzsyMhIYFLly6RkJBAaGhotuv36NGDhg0b0qJFCypUqMCNGzdYvnw5NWrUUAaLX7FiBe3bt6dDhw6MHz8eBwcHkpKS+OWXXzhw4ECOlUyGrjt16lR27NhBr169mDlzJq1ateLJkyecPHmS7t274+bmRqlSpahRowb79u3Dw8MDGxsb5di/qGzZssyZM4dZs2YxbNgwBg0aRGJiIkFBQVhYWBAQEGDQ8RVClEwuLi4sX76cKVOm0L59eyZOnEj16tW5efOmMiNrYGAgnp6eyjoDBgxg7ty5DBw4kHfffZeUlBRWrlypV+Hl6emJmZkZgwYN4r333iMlJYXQ0NB8d1tzdHRErVazdetW6tevj7W1NZUrV9aZAXDKlCkMGDAAlUqFn5+fQfEuXbqU9u3b07p1a2bOnMlrr73G3bt32b9/P2vWrKFUqVIEBARw8OBB3NzcmDt3LjY2NmzdupXIyEiCg4N1Blw3VKNGjYiIiCA0NJTmzZtjZGSkvMzJipGREcHBwQwePJju3bszbtw4UlNTWbx4MQ8ePOCjjz7KcxqEeFmkwuo5ERERBAYGEhoaqgz+uXz5cr03owMGDCA0NJTx48dn+dC6YsUKJkyYwMCBA0lOTtYbJNYQoaGhODo6sn79elavXk2ZMmXo2rUrCxculPFzRKEYM2YMrVq1UqYNj4+Px9TUlDp16uDj45Nj3/0tW7YwadIkZsyYQUZGBj169GD79u06F0t7e3scHR0JCQnhjz/+QKVSUatWLZYsWcKkSZOUcEFBQdy5c4cxY8aQlJREjRo1iIuLw8rKiq+//pqPPvqItWvX8vvvv6NWq6levTqdOnUq8hZWAOHh4VSqVIn169ezbNkynJ2d2blzJ127dtV7IBZCFB5/f3+cnJxYsWIF27dvJzU1FXt7e1q2bMlbb72lhNu2bRtTpkzhvffeIz09HRcXF44ePYq3t7dOfDVr1mT58uWsWLECV1dXMjIyCA8Px9fXFx8fH27fvk1YWBjh4eE0bNiQ0NBQgoKCDE7vkCFDqF69OsHBwYwbN46kpCTs7OxwdnbWG3T8RW5ubuzevZt169bx8OFD7O3t8fT0ZM6cOcqguU5OTly4cIH58+cze/Zs7t27R9myZaldu7YyFlV2DF1XO5NgYGAga9euJSgoiHLlytGyZUvGjh2rhFu/fj3vvvsuPXv2JDU1leHDh7Nx48Yst+3v74+dnR0rV65kx44dqNVqXF1dWbBggVIZJ4R4dU2aNIkWLVqwZMkSpk+fTkJCApmZmVhYWBAZGalXPtWsWZN9+/Yxa9Ys+vfvT6VKlZg2bRoJCQk6ZW69evXYvXs3s2fPpm/fvtja2uLj48O0adN0Zog2lKWlJRs2bCAoKIjOnTuTlpZGQECAzuysvXv3xtzcHDc3N4PLpyZNmnDu3DkCAgLw9/cnKSkJe3t73N3dlefHunXr8u233zJr1iwmTJjAkydPqF+/vnINyo8pU6Zw5coVZs2axT///INGo9FpOZUVHx8frKysWLhwIQMGDMDY2Jg2bdoQFRVFu3bt8pUOIV4GlSa3sztXT4AowPDZv4qGEeAGqPO8ZmBgIEFBQSQkJBjUj3j69OmEhobyxx9/SOXRf9DDR6mE7/2BjIwCZp0CMjZWMaJ3Q0pbm+ceWBSab7/9FhcXF7Zu3YqPj09xJ0f8B6WnJnHvu62gycg9cFFSGWPXfDAm5qWKNx1CCPGK+TffS27evJnhw4fz3nvvsWjRokKLt6gdOHCAnj17ZlnRJoQoPoXQwkrNs4oi/XFwXi4z8lNZlRdnzpzh+vXrhISEMG7cOKms+o8qbW3OiN4NeZJavOMYqc1NpLKqiB09epSYmBiaN2+OWq3m0qVLfPTRR9SuXZu+ffsWd/LEf5SJeSnsmg8mM614Z08zMrWQyiohhMiHf/O95LBhw7hz5w4zZ87EysqKuXPnFmr8he3q1avcuHGD6dOn4+zsnK8WXEKIolMILaxefYa2sFKpVFhaWuLl5UV4eDjW1tYvMZVCiJft7NmzTJ8+natXr5KUlET58uXp0qULCxcuzHJqYCGEEEII8epwdXXlm2++oVmzZmzatIl69eoVd5KEEM+RCishhBBCCCGEEEIIUaLINFdCCCGEEEIIIYQQokSRCishhBBCCCGEEEIIUaJIhZUQQgghhBBCCCGEKFGkwkoIIYQQQgghhBBClChSYSWEEEIIIYQQQgghShSpsBJCCCGEEEIIIYQQJYpJYUSS8iSNtLSMwogq30xNjbFQmxZrGsR/R8bDFDKfpBdrGozUJhiXtijWNAghiseTJ094+vRpsabBzMwMtVpdrGkQQohXldxLCiFE7gpcYZXyJI2z394gM1NTGOnJNyMjFa3b1SgRlVYLFizAycmJ3r176yzfuHEjI0aMIDY2lhYtWhRP4gzg6+vLrl27ePTo0UvdrkqlYsKECXzyyScvdbt5lfEwhYR130FG8Z7zGKuoMLp5vm80Ll++zLJly4iOjubOnTuYmJhQp04dBg4cyOjRo7GxsSnkBJdMDg4OuLq6snHjRgBu377N2rVr6d27N87OzjphAwMDCQoKQqMp5t/+f15Me0lUlGksab/Hy/LkyROioqLIzMws1nQYGRnh5uaW70qry5cvs2LFCqKjo7l9+zYAVatWxd3dnTFjxpTo62RJVZA88eWXX3Lu3DkCAwMLP2ElTHJyMsHBwbi6uuLq6lrcyVE4ODjQsGFDDh48WNxJKXQqlYqAgIAiOb98fX2Jjo4mLi6u0OMuKq/6vaRKpTIoXFRUVIHyWFFf57/99lu++uorpk6dStmyZYtkG/lV0PN6x44dzJs3j99++42UlBQuXryod18rxKugwF0C09Iyir2yCiAzU1Psrby0FixYwN69e4s7GaKIZD5JL/4bDIAMTb7fzH366ac0b96c2NhY3n33XQ4fPsyePXt44403CAsLY9SoUYWc2JJrz549zJkzR/l8+/ZtgoKC+P777/XCjh49mpiYmJeYOpGT/+rv8fTp02KvrALIzMzMdyuvNWvW0Lx5c86ePcuUKVM4ePAgkZGRTJ06lStXrtCyZUt+/fXXQk6xyMmXX35JUFBQcSfjpUhOTiYoKIjo6OjiToooBHPmzGHPnj3FnYw8edXvJWNiYnT+vLy8UKvVesubNWtWoOQV9XX+22+/JSgoiAcPHhTZNvKrIOd1QkICQ4cOxdHRkcOHDxMTE0OdOnUKOYVCvByF0iVQCPHqiImJYfz48Xh6erJ3717Mzc2V7zw9PZk+fTqHDx8uxhS+XE2bNjU4bNWqValatWoRpkbkhfwer6ZvvvkGPz8/vL292bVrF2ZmZsp37u7uTJgwgS+++OKV62745MkTLCwsDG558F/x5MmTV+63fNmSk5OxtLQs7mS8shwdHYs7Cf85bdq00flcoUIFjIyM9Ja/KK/n+qt6nS+MPF2Q8/r69eukpaUxZMgQOnbsWKB0aMk1ThSX//yg64GBgahUKi5evEjfvn0pXbo0ZcqUYciQISQkJCjhRo0ahY2NDcnJyXpxuLu706BBA+BZE9nHjx+zadMmVCoVKpVKrylsUlIS48ePp3z58tja2tK3b1+lO4RWZmYmwcHB1KtXD3Nzc+zs7Bg2bBi3bt3SCefq6krDhg2JjY2lQ4cOWFpaUqtWLT766KMCv4G/cuUKHh4eWFlZUaFCBSZOnKi3/6tXr+b111/Hzs4OKysrGjVqRHBwMGlpaTrhLl68SPfu3bGzs8Pc3JzKlSvj7e2ttz8AW7ZsoX79+lhaWtKkSZN/ZdP44rRgwQJUKhVr167VqazSMjMzo2fPnsrnvJ6LMTExtGvXDrVajYODA+Hh4QBERkbSrFkzLC0tadSokV6lmKF5MS9pMuS8c3BwwNfXF4Do6GhatmwJwIgRI5Q8rO3CoE1jftJiaF5NSUlh+vTpODs7U6ZMGWxsbGjbti379u3T+60MZWg+Leo0Pnr0iLJlyzJu3Di97+Li4jA2Nmbx4sXAs5u9d955h5o1a2JhYYGNjQ0tWrRg+/btyjpZ/R4nTpzA1dUVW1tb1Go11atXp1+/flmW3aJ4LFiwAGNjY9asWaNTWfW8N954g8qVK+ssO3/+PD179sTGxgYLCwuaNm3Kzp07dcJs3LgRlUpFVFRUrtdZeNZlom3btlhZWWFtbU2XLl24ePFirvug3c5XX33FyJEjqVChApaWlqSmpuYp3k8//ZQ6depgbm6Ok5MT27Ztw9fXFwcHByVMdHQ0KpVKrzVQXFwcKpUq1+62O3bsoHPnzlSqVAm1Wk39+vWZOXMmjx8/VsL4+vqyevVqAKXcU6lUSleUlJQU/P39qVmzJmZmZlSpUoUJEybotUpwcHCge/fuRERE0LRpUywsLAgKCsLDw4N69erpdevRaDS89tpreHt757gP2ngPHz5Ms2bNUKvV1KtXjw0bNuiFjY+PZ9y4cVStWhUzMzNq1qxJUFAQ6enpynGrUKECAEFBQcq++vr6cuXKFVQqFV988YUS33fffYdKpVLu87R69uxJ8+bNlc95vRacOnWKdu3aYWlpyciRI7Pd95CQEExMTAgICFCWhYaG0qRJE6ytrSlVqhT16tVj1qxZOR5D7f62bt0aGxsbSpcuTbNmzVi/fr3e72Lo8U5ISMDPzw8nJyesra2xs7PD3d2dr7/+Osd0xMXFYWJiwsKFC/W+O3XqlM5vkJCQwNixY6lWrRrm5uZUqFABFxcXjh07pqzzYp4B+OKLL2jdujVlypRRrmU5HWdR+HI61w0plyDr67x2fUPK2LNnz9KjRw9sbW2xsLDA0dGRqVOnKnG/++67ANSsWVMpC7RlbUHztKHPjtnJ6rxWqVRMnDgxx2clX19f2rdvD8CAAQP0nkXzci0tyDXO19cXa2trfvnlF7y8vLC2tqZatWpMnz5diUcrNTWVefPmUb9+fSwsLLC1tcXNzY1vv/1WCaPRaAgJCcHZ2Rm1Wk25cuXo378/v/32W47HUbz6/vMVVlp9+vThtddeY9euXQQGBrJ37166dOmiPNBNmTKFv//+m23btumsd/XqVaKiopgwYQLwrPWKWq3Gy8tLaQ4bEhKis87o0aMxNTVl27ZtBAcHEx0dzZAhQ3TCjB8/nhkzZuDp6cn+/fuZP38+hw8fpl27dvz11186YePj4xk8eDBDhgxh//79dOvWDX9/fz777LN8H4+0tDS8vLzw8PBg7969TJw4kTVr1jBgwACdcL/++is+Pj5s2bKFgwcPMmrUKBYvXqzzQPr48WM8PT25e/cuq1ev5ujRoyxfvpzq1auTlJSkE19kZCSffPIJ8+bNY/fu3djY2NCnTx8pjApJRkYGJ06coHnz5lSrVs2gdfJ6Lo4YMYLRo0ezb98+GjVqxMiRI5k3bx7+/v6899577N69G2tra3r37p3lA2RuedHQNOXlvNNq1qyZUsE2e/ZsJQ+PHj260I5Pbnk1NTWV+/fv884777B37162b99O+/bt6du3L5s3b87l18qaIfn0ZaTR2tqakSNHsnXrVv755x+d70JCQjAzM1NuZqdNm0ZoaCiTJ0/m8OHDbNmyhTfeeIPExMRs44+Li8Pb2xszMzM2bNjA4cOH+eijj7Cysir2AcrFMxkZGURFRdGiRQsqVapk8HpRUVG4uLjw4MEDwsLC2LdvH87OzgwYMCDLChtDrrMLFixg0KBBODk5sXPnTrZs2UJSUhIdOnTg6tWrBqVr5MiRmJqasmXLFnbt2oWpqanB8a5du5axY8fSuHFjIiIimD17dpF0U/v555/x8vJi/fr1HD58mKlTp7Jz50569OihhJkzZw79+/cHdLv6VKpUCY1GQ+/evfn4448ZOnQokZGRTJs2jU2bNuHu7q734HHhwgXeffddJe/269ePKVOmcO3aNY4fP64T9tChQ/z666/KPVROLl26xPTp03n77bfZt28fjRs3ZtSoUZw6dUoJEx8fT6tWrThy5Ahz587l0KFDjBo1ioULFzJmzBgAKlWqpLwwGTVqlLKvc+bMoUGDBlSqVEmnMuTYsWOo1WquXr2qXLPS09M5efIknTp1UsLl5Vpw584dhgwZgo+PD19++SV+fn56+6vRaHjnnXeYOnUq69atU7prfv755/j5+dGxY0f27NnD3r17efvtt/Ue9LMSFxfHuHHj2LlzJxEREfTt25dJkyYxf/78fB3v+/fvAxAQEEBkZCTh4eHUqlULV1fXHM9jBwcHevbsSVhYGBkZusN5fPLJJ1SuXJk+ffoAMHToUPbu3cvcuXP56quvWLduHZ06dcrxWhATE8OAAQOoVasWn3/+OZGRkcydO1eptBQvT3bnuiHlUnYMLWOPHDlChw4duHnzJkuXLuXQoUPMnj2bu3fvAs+uE5MmTQIgIiJCrxtjQfO0oc+OeZXbs9KcOXOUFxALFizQeRbN67W0INc4ePY82bNnTzw8PNi3bx8jR45k2bJlLFq0SAmTnp5Ot27dmD9/Pt27d2fPnj1s3LiRdu3acfPmTSXcuHHjmDp1Kp06dWLv3r2EhIRw5coV2rVrp/ym4l9KU0AP/3miOfHV9RLx9/CfJ3lOf0BAgAbQvP322zrLt27dqgE0n332mbKsY8eOGmdnZ51w48eP15QuXVqTlJSkLLOystIMHz5cb1vh4eEaQOPn56ezPDg4WANo7ty5o9FoNJoff/wxy3Bnz57VAJpZs2bppAnQnD17Viesk5OTpkuXLgYcAX3Dhw/XAJoVK1boLP/www81gOb06dNZrpeRkaFJS0vTbN68WWNsbKy5f/++RqPRaM6fP68BNHv37s1xu4CmYsWKmocPHyrL4uPjNUZGRpqFCxfma1+KwtP4JM2d4K9LxN/T+KTcE/yc+Ph4DaAZOHCgQeHzcy6eP39eWZaYmKgxNjbWqNVqzZ9//qks//777zWAZuXKlcoyQ/OioWky9LyrUaOGTn6NjY3VAJrw8HC9sNo0FuT45DWvpqena9LS0jSjRo3SNG3aNMe0GyK7fPqy0vjrr79qjIyMNMuWLVOWPXnyRGNra6sZMWKEsqxhw4aa3r1757gvL/4eu3bt0gCa77//Psf1XnUPHjzQHDhwoET8PXjwIE9pz6kM0p5H2r/MzEzlu3r16mmaNm2qSUtL01mne/fumkqVKmkyMjI0Go3h19mbN29qTExMNJMmTdIJl5SUpLG3t9e8+eabOe6HdjvDhg3TWW5ovBkZGRp7e3tN69atdcLduHFDY2pqqqlRo4ayLCoqSgNooqKidML+/vvvemXVi3niRZmZmZq0tDTNyZMnNYDm0qVLyncTJkzIct3Dhw9rAE1wcLDO8h07dmgAzdq1a5VlNWrU0BgbG2uuXbumEzYjI0NTq1YtTa9evXSWd+vWTePo6KjzW2elRo0aGgsLC82NGzeUZU+ePNHY2Nhoxo0bpywbN26cxtraWiecRqPRfPzxxxpAc+XKFY1Go9EkJCRoAE1AQIDetoYMGaKpVauW8rlTp06aMWPGaMqVK6fZtGmTRqPRaL755hsNoPnqq680Gk3+rgXHjx/Pcj+9vb01ycnJmn79+mnKlCmjOXbsmE6YiRMnasqWLZvj8TKE9lowb948ja2trc5vYOjxfpE2D3t4eGj69Omj892Lx1t7Xu/Zs0dZ9ueff2pMTEw0QUFByjJra2vN1KlTc9yX4cOH6+QZ7e+d1/LpZXqV7yWzMnz4cI2VlZXOspzO9eflVC69WKblpex2dHTUODo6ap48yf75cPHixRpA8/vvv+ssL6w8beizY1ZePK81GsOflbT564svvtBZP6/X0vxe47TpBzQ7d+7UCevl5aWpW7eu8nnz5s0aQPPpp59meyxiYmI0gGbJkiU6y//44w+NWq3WvPfee9muK1590sLqfwYPHqzz+c0338TExISoqChl2ZQpU/j+++/55ptvAHj48CFbtmxh+PDhWFtbG7yt57tbATRu3BiAGzduACjb1HZT0mrVqhX169fXe0Npb29Pq1at9OLUxpdfLx4THx8fnfTBsy5XPXv2xNbWFmNjY0xNTRk2bBgZGRlcv34dgNdee41y5coxY8YMwsLCcnxz7ebmRqlSpZTPFStWxM7OrsD7IvInr+dipUqVdLpI2NjYYGdnh7Ozs073nvr16wNk+bvmlhcNTVNezrv8Kqq8+sUXX+Di4oK1tTUmJiaYmpqyfv16fvzxx3yl05B8+rLSWKtWLbp3705ISIjSDWXbtm0kJiYyceJEJVyrVq04dOgQM2fOJDo6midPnuS6n87OzpiZmTF27Fg2bdokLTNfMc2bN8fU1FT5W7JkCQC//PILP/30k1I2pKenK39eXl7cuXOHa9eu6cSV23X2yJEjpKenM2zYMJ34LCws6Nixo8GtnPr166fz2dB4r127Rnx8PG+++abO+tWrV8fFxcWgbRvqt99+w8fHB3t7eyX/a8c0MaRMOXHiBKBfzr3xxhtYWVnplXONGzfWG9zXyMiIiRMncvDgQeWN+a+//srhw4fx8/MzaEwUZ2dnqlevrny2sLCgTp06OmXTwYMHcXNzo3LlyjrHv1u3bgCcPHky1+14eHjw22+/8fvvv5OSksLp06fp2rUrbm5uHD16FHjW6src3FzpdpPXa0G5cuVwd3fPcvuJiYm4u7tz7tw5Tp8+jYeHh16cDx48YNCgQezbt0+vpUdOTpw4QadOnShTpoxyLsydO5fExETu3bunE9aQ4w0QFhZGs2bNsLCwUK4Fx48fz/XccnV1pUmTJkpLEG1cKpWKsWPH6uzvxo0b+eCDDzhz5oxeV/asaLv2v/nmm+zcuZM///wz13VE0cjuXM9vuWRoGXv9+nV+/fVXRo0ahYVF3mfTLqw8XVjPjs/L77NSfq6l+b3GaalUKr1Wcy/eUx46dAgLC4scu+wePHgQlUrFkCFDdLZrb29PkyZNZAKNfzmpsPofe3t7nc8mJibY2trqNDnu1asXDg4OysV148aNPH78OM9NOm1tbXU+a8cR0j6QabeZVXeJypUr6zWDfjE+bZyGPOBlR7v/z9MeI+32b968SYcOHfjzzz9ZsWIFX3/9NbGxscrx0W6/TJkynDx5EmdnZ2bNmkWDBg2oXLkyAQEBejceRbEv4v+UL18eS0tLfv/9d4PC5/VctLGx0QtnZmamt1w7bk1KSope+NzyoqFpyst5l19FkVcjIiJ48803qVKlCp999hkxMTHExsYycuTILI9XbgzNpy8zjVOmTOHnn39WHv5Wr15N27ZtdWYTWrlyJTNmzGDv3r24ublhY2ND7969+fnnn7ON19HRkWPHjmFnZ8eECRNwdHTE0dGRFStW5Jom8XKUL18etVqd5Y31tm3biI2NZf/+/TrLtU3933nnHZ0KLVNTU6V7yYsP7bldZ7VxtmzZUi/OHTt2GFwJ8GLeNzRebdlQsWJFvTizWpZfjx49okOHDpw9e5YPPviA6OhoYmNjiYiIAPTzf1YSExMxMTFRxn3SUqlU2Nvb65Vz2XX1HDlyJGq1mrCwMOBZvler1QaPK2RI2XT37l0OHDigd+y1Y8UY8rtqu/kdO3aM06dPk5aWhru7O506dVIeUo8dO4aLi4symHxerwU5dYe9fv06Z8+epVu3bjRs2FDv+6FDh7JhwwZu3LhBv379sLOzo3Xr1kp5mp1z587RuXNn4NnYad988w2xsbG8//77QP6uBUuXLmX8+PG0bt2a3bt3c+bMGWJjY+natatB59bkyZM5fvw4165dIy0tjU8//ZT+/fvr3Afs2LGD4cOHs27dOtq2bYuNjQ3Dhg0jPj4+23hff/119u7dqzxYV61alYYNG+qMgShejqzO9YKUS4aWsdqxT/M7YHth5enCenZ8Xn6flfJzLc3vNU7L0tJSr8LQ3Nxc514xISGBypUrY2SUfbXE3bt30Wg0VKxYUW+7Z86cyVPFvXj1yCyB/xMfH0+VKlWUz+np6SQmJuoUCkZGRkyYMIFZs2axZMkSQkJC8PDwoG7duoWaFu0279y5o1fQ3r59m/Llyxfq9rKS1f5rbw60y/bu3cvjx4+JiIigRo0aSrjvv/9eL75GjRrx+eefo9FouHz5Mhs3bmTevHmo1WpmzpxZtDsjFMbGxnh4eHDo0CFu3bqV64W8OM7F3PJiXtJU1OddURyfzz77jJo1a7Jjxw6dVgcvjhNjqLzk05eVRnd3dxo2bMgnn3yCtbU1Fy5c0Btzz8rKiqCgIIKCgrh7967S2qpHjx789NNP2cbdoUMHOnToQEZGBufPn2fVqlVMnTqVihUrMnDgwPztsCg0xsbGuLu789VXX3Hnzh2dm2EnJycAZaBvLW0+8vf3p2/fvlnGm9frsDbOXbt26eSLvHqxZZCh8WrLjqzG3XjxQVx7s/9i/jLkBv3EiRPcvn2b6OhonZmi8jKFu62tLenp6SQkJOhUWmk0GuLj45XWLFrZtZYqU6aMUvHwzjvvEB4ejo+PD2XLljU4LbkpX748jRs35sMPP8zy+xcH8s9K1apVqVOnDseOHcPBwYEWLVpQtmxZPDw88PPz4+zZs5w5c0YZUwryfi3IqUVZ27ZteeONNxg1ahTwbID1Fx/kRowYwYgRI3j8+DGnTp0iICCA7t27c/369WzPu88//xxTU1MOHjyo8wC5d+/eXI9Jdj777DNcXV0JDQ3VWZ7dOJEv8vHxYcaMGaxevZo2bdoQHx+v9yBfvnx5li9fzvLly7l58yb79+9n5syZ3Lt3L8cZjXv16kWvXr1ITU3lzJkzLFy4EB8fHxwcHGjbtm3ed1bkS1bnekHKJUPLWG1ZldXkToYorDz9sp4dDZGfa2l+r3F5UaFCBU6fPk1mZma2lVbly5dHpVLx9ddfZzlhVFbLxL+HtLD6n61bt+p83rlzJ+np6Xoz/I0ePRozMzMGDx7MtWvXdLqwaBW0RZC2SemLD3CxsbH8+OOPes3Di8qLx0Q7aKD2mGgLsecLCY1Gw6effpptnCqViiZNmrBs2TLKli3LhQsXCjnVIjf+/v5oNBrGjBmT5UDUaWlpHDhwACieczG3vJifNOXlvHuxJUZOiuL4qFQqzMzMdG4S4uPj8z1LYH7y6ctI4+TJk4mMjMTf35+KFSvyxhtvZBu2YsWK+Pr6MmjQIK5du2bQjH/Gxsa0bt1aeaspZU3J4e/vT0ZGBm+99ZZBrR3r1q1L7dq1uXTpEi1atMjy7/nuEYbo0qULJiYm/Prrr9nGmR+Gxlu3bl3s7e31Zma6efOmzqxIgDJL1OXLl3WWv9gSLStZ5X+ANWvW6IXNruzTlmMvlnO7d+/m8ePHeSrnJk+ezF9//UX//v158OBBlvdQBdG9e3d++OEHHB0dszz22gqr3Mr5Tp06ceLECY4ePYqnpycAderUoXr16sydO5e0tDSdAdcL+1owfPhwPv/8c8LDw5Xu21mxsrKiW7duvP/++zx9+pQrV65kG6dKpcLExARjY2Nl2ZMnT9iyZUue0vZinC+eW5cvXyYmJsag9S0sLJQu3EuXLsXZ2TnHLrHVq1dn4sSJeHp6Glymm5ub07FjR2WQZ0NmARVFKy/l0osMLWPr1KmDo6MjGzZsyPFlWnZlQWHmaUOeHV+GwriWFsW1s1u3bqSkpOQ442337t3RaDT8+eefWW6zUaNGed6ueHVIC6v/iYiIwMTEBE9PT65cucKcOXNo0qSJ3vgSZcuWZdiwYYSGhlKjRo0sZ7No1KgR0dHRHDhwgEqVKlGqVKk81aTXrVuXsWPHsmrVKoyMjOjWrRtxcXHMmTOHatWq8fbbb+drH11dXTl58qTe9MVZMTMzY8mSJTx69IiWLVvy7bff8sEHH9CtWzdlzAZPT0/MzMwYNGgQ7733HikpKYSGhvL333/rxHXw4EFCQkLo3bs3tWrVQqPREBERwYMHD5QbQfHytG3bltDQUPz8/GjevDnjx4+nQYMGpKWlcfHiRdauXUvDhg3p0aNHkZ2LOcktLxqapvyed46OjqjVarZu3Ur9+vWxtramcuXKWb6ZL4rjo50S3s/Pj/79+/PHH38wf/58KlWqlGN3uOwYmk9fdhqHDBmCv78/p06dYvbs2Uo3Ua3WrVvTvXt3GjduTLly5fjxxx/ZsmULbdu2xdLSMss4w8LCOHHiBN7e3lSvXp2UlBRlGvbnHy5F8XJxcWH16tVMmjSJZs2aMXbsWBo0aICRkRF37txh9+7dAJQuXVpZZ82aNXTr1o0uXbrg6+tLlSpVuH//Pj/++CMXLlzgiy++yFMaHBwcmDdvHu+//z6//fYbXbt2pVy5cty9e5dz584pLfzyytB4jYyMCAoKYty4cfTv35+RI0fy4MEDgoKCqFSpks5bZnt7ezp16sTChQspV64cNWrU4Pjx40r3mZy0a9eOcuXK8dZbbxEQEICpqSlbt27l0qVLemG1N/yLFi2iW7duGBsb07hxYzw9PenSpQszZszg4cOHuLi4cPnyZQICAmjatClDhw41+PjUqVOHrl27cujQIdq3b0+TJk0MXtcQ8+bN4+jRo7Rr147JkydTt25dUlJSiIuL48svvyQsLIyqVatSqlQpatSowb59+/Dw8MDGxoby5csrlYMeHh6EhITw119/sXz5ciV+Dw8PwsPDKVeunM54jUVxLejfvz+Wlpb079+fJ0+esH37dszMzBgzZgxqtRoXFxcqVapEfHw8CxcupEyZMnqt3Z7n7e3N0qVL8fHxYezYsSQmJvLxxx8XqGVC9+7dmT9/PgEBAXTs2JFr164xb948atasafCMfH5+fgQHB/Pdd9+xbt06ne/++ecf3Nzc8PHxoV69epQqVYrY2FgOHz6cbQsRgLlz53Lr1i08PDyoWrUqDx48YMWKFTrjJInik5dy6UV5KbtXr15Njx49aNOmDW+//TbVq1fn5s2bHDlyRHk5qi33VqxYwfDhwzE1NaVu3bqFmqcNeXZ8WQp6LS2Ka+egQYMIDw/nrbfe4tq1a7i5uZGZmcnZs2epX78+AwcOxMXFhbFjxzJixAjOnz/P66+/jpWVFXfu3OH06dM0atSI8ePHF+TQiBJMKqz+JyIigsDAQEJDQ5UB4pYvX673EAUwYMAAQkNDGT9+fJZNF1esWMGECRMYOHAgycnJeRrAVSs0NBRHR0fWr1/P6tWrKVOmDF27dmXhwoVZ9l02xKNHj/TGB8qOttn45MmT+eCDD1Cr1YwZM4bFixcrYerVq8fu3buZPXs2ffv2xdbWFh8fH6ZNm6YMcApQu3ZtypYtS3BwMLdv38bMzIy6deuyceNGhg8fnq99EQUzZswYWrVqpUwtGx8fj6mpKXXq1MHHx0fn7U9RnIs5MSQvGpKm/J53lpaWbNiwgaCgIDp37kxaWhoBAQEEBgZmGb6wj8+IESO4d+8eYWFhbNiwgVq1ajFz5kxu3bqVrwdoQ/Ppy06jWq2mR48efPbZZ7z11lt637u7u7N//36WLVtGcnIyVapUYdiwYcp4K1lxdnbmq6++IiAggPj4eKytrWnYsCH79+9Xxm4RJcNbb71F27ZtWbFiBcuWLeP27duoVCqqVq1Ku3btOH78uM4Atm5ubpw7d44PP/yQqVOn8vfff2Nra4uTk5PeiyVD+fv74+TkxIoVK9i+fTupqanY29vTsmXLLM/Jwo537NixqFQqgoOD6dOnDw4ODsycOZN9+/bpTOUNsGXLFiZNmsSMGTPIyMigR48ebN++Pde32ba2tkRGRjJ9+nSGDBmClZUVvXr1YseOHTpjxsGz7lnffPMNISEhzJs3D41Gw++//46DgwN79+4lMDCQ8PBwPvzwQ8qXL8/QoUNZsGBBnis8BgwYwKFDh4qklUGlSpU4f/488+fPZ/Hixdy6dYtSpUpRs2ZN5cFKa/369bz77rv07NmT1NRUhg8frrzhd3d3x8jICLVardN9rFOnToSHh+Pm5qZ3/1cU10ovLy++/PJLevToQa9evYiIiKBDhw5s3LiRnTt38vfff1O+fHnat2/P5s2b9cYZe567uzsbNmxg0aJF9OjRgypVqjBmzBjs7OyU7od59f7775OcnMz69esJDg7GycmJsLAw9uzZY/B9b5UqVWjfvj2XL19WJvfRsrCwoHXr1mzZsoW4uDjS0tKoXr06M2bM4L333ss2ztatW3P+/HlmzJhBQkICZcuWpUWLFpw4cUIZz0wUn7yUS1kxtIzt0qULp06dYt68eUyePJmUlBSqVq2qMymHq6sr/v7+bNq0iU8//ZTMzEyioqKUrq6Fladze3Z8WQrjWlrY104TExO+/PJLFi5cyPbt21m+fDmlSpWiSZMmdO3aVQm3Zs0a2rRpw5o1awgJCSEzM5PKlSvj4uKiN1mQ+HdRaQxpbpODlCdpnP32BpmZBYqmwIyMVLRuVwMLtWme1gsMDCQoKIiEhASDx5uZPn06oaGh/PHHH0XywF4UkpKSsLGxYfny5QUa6E9AxsMUEtZ9BxnFe85jrKLC6OYYl8777CclUX7yonh1PX36FAcHB9q3b6/XLUrk7MmTJ0RFRZGZmVms6TAyMsLNzU0ZeFoU3IMHD6hTpw69e/dm7dq1xZ2cItGvXz/OnDlDXFwcpqZ5u2cT/z737t2jRo0aTJo0ieDg4OJOzksj95KGefvtt9myZcsrP6j2q/jsKERJUeAWVhZqU1q3q0FaWtb9618WU1PjPFdW5dWZM2e4fv06ISEhjBs37pUqcE6dOqW8TRMFY1zaggqjm5P5xLDm7kXFSG1SYm8whMhOQkIC165dIzw8nLt378qkC/mgVqtxc3PLcgy6l8nMzEwqqwogPj6eDz/8EDc3N2xtbblx4wbLli0jKSmJKVOmFHfyClVqaioXLlzg3Llz7Nmzh6VLl0pl1X/crVu3+O2331i8eDFGRkb/unM+N3IvmbN79+4RExNDRETEKz1I/qv87ChESVEoXQIt1KZFXllUEmjHTunevTsffPBBcScnT7y9vfH29i7uZPxrGJe2wLh07uGEELoiIyMZMWIElSpVIiQkxKDm/0KfWq2WyqJXnLm5OXFxcfj5+XH//n0sLS1p06YNYWFh/7puS3fu3KFdu3aULl2acePGMWnSpOJOkihm69atY968eTg4OLB161ad2YH/K+ReMntffvklEydOpE2bNqxYsaK4k5Nvr/KzoxAlRYG7BAohhBBCCCGEEEIIUZiKb9Q3IYQQQgghhBBCCCGyIBVWQgghhBBCCCGEEKJEkQorIYQQQgghhBBCCFGiSIWVEEIIIYQQQgghhChRpMJKCCGEEEIIIYQQQpQoUmElhBBCCCGEEEIIIUoUqbASQgghhBBCCCGEECWKSWFEcj/lMY/SUwsjqnyzNjHHxsKqQHGsXLmSKVOm0KBBA3744YdCSlnhWLBgAU5OTvTu3Ttf6ycnJxMcHIyrqyuurq46323cuJERI0bw+++/4+DgUOC0/hc8TksgNSOpWNNgblwKK9MKeV5P+3sr8ZibU7ZsWerXr0/nzp0ZNWoUdnZ2OusEBgYSFBSERqMpcLqLm0qlIiAggMDAwOJOSpZ5z9fXl+joaOLi4pRwBc3/L1tRny9Xr15l586d+Pr66pVZWR2/f6NX/bqrPfdjY2Np0aKFwevldC0rqb788kvOnTuXZZnj4OCAq6srGzdufOnpEkIIIYQo6QpcYXU/5TFzzh8gXZNZGOnJNxOVEfNb9ChQpdWGDRsAuHLlCmfPnqV169aFlbwCW7BgAf379y9QhVVQUBCA3k2+t7c3MTExVKpUqYCp/G94nJbAwd+nkKlJK9Z0GKlM6V5zRb4qrQDCw8OpV68eaWlp3Lt3j9OnT7No0SI+/vhjduzYQadOnZSwo0ePpmvXroWV9GIVExND1apVizsZ2ZozZw5TpkzRWVbQ/P9vc/XqVYKCgnB1ddWrsMrq+P3b/Juuu3mV07WspPryyy9ZvXp1lhVWe/bsoXTp0i8/UUIIIYQQr4ACV1g9Sk8t9ptmgHRNJo/SU7EhfzfO58+f59KlS3h7exMZGcn69etLVIVVUapQoQIVKuSv0uO/KDUjqdgrqwAyNWmkZiTlu8KqYcOGOi0b+vXrx9tvv0379u3p27cvP//8MxUrVgSgatWqJbqSJy/atGlT3EnIkaOjY7FtOy0tDZVKhYlJoTS+LRbFefxeln/LdfdVlZycjKWlZaHE1bRp00KJRwghhBDi30jGsPqf9evXA/DRRx/Rrl07Pv/8c5KTk5Xv4+LiUKlUfPzxxyxdupSaNWtibW1N27ZtOXPmjE5cvr6+WFtb88svv+Dl5YW1tTXVqlVj+vTppKbqduG4f/8+fn5+VKlSBTMzM2rVqsX777+vE06lUvH48WM2bdqESqVCpVIpb5YTEhLw8/PDyckJa2tr7OzscHd35+uvv9ZJu7ZCKigoSInD19cXeNY1Q6VS6XWh2bBhA02aNMHCwgIbGxv69OnDjz/+mO99FSVf9erVWbJkCUlJSaxZs0ZZHhgYiEql0gnr4OBA9+7dOXjwIE2bNkWtVlO/fn0OHjwIPDuv6tevj5WVFa1ateL8+fN62zt//jw9e/bExsYGCwsLmjZtys6dO3XCaM/PqKgoxo8fT/ny5bG1taVv377cvn1bJ+yJEydwdXXF1tYWtVpN9erV6devn05eVqlUei0dfvjhB3r16kW5cuWwsLDA2dmZTZs26YSJjo5GpVKxfft23n//fSpXrkzp0qXp1KkT165d0wl79OhRevXqRdWqVbGwsOC1115j3Lhx/PXXX7n8Auh1c8su/8fFxWFiYsLChQv14jh16hQqlYovvvgi2+1o92fLli1Mnz6dKlWqYG5uzi+//ALAsWPH8PDwoHTp0lhaWuLi4sLx48f14omMjMTZ2Rlzc3Nq1qzJxx9/rBdGW35m1e0pq9/jp59+YtCgQVSsWBFzc3OqV6/OsGHDSE1NZePGjbzxxhsAuLm5KcdEG3dW3QRTUlLw9/enZs2amJmZUaVKFSZMmMCDBw90wmnP6cOHD9OsWTPUajX16tVTWt+KomPItSS3axnAzz//jI+PD3Z2dpibm1O/fn1Wr16tt70rV67QuXNnLC0tqVChAhMmTCAyMhKVSkV0dLQSztXVlYYNG3Lq1CnatWuHpaUlI0eOBGDHjh107tyZSpUqKeXfzJkzefz4sc5+abevTe/z11sHBwed9APcvHmTIUOG6OzDkiVLyMz8v0rKvNyTCCGEEEK8qqTCCnjy5Anbt2+nZcuWNGzYkJEjR5KUlJTlw97q1as5evQoy5cvZ+vWrTx+/BgvLy/++ecfnXBpaWn07NkTDw8P9u3bx8iRI1m2bBmLFi1SwqSkpODm5sbmzZuZNm0akZGRDBkyhODgYPr27auEi4mJQa1W4+XlRUxMDDExMYSEhADPKrwAAgICiIyMJDw8nFq1auHq6qrcdFeqVInDhw8DMGrUKCWOOXPmZHtMFi5cyKhRo2jQoAERERGsWLGCy5cv07ZtW37++ec876t4dXh5eWFsbMypU6dyDXvp0iX8/f2ZMWMGERERlClThr59+xIQEMC6detYsGABW7du5Z9//qF79+48efJEWTcqKgoXFxcePHhAWFgY+/btw9nZmQEDBmRZsTF69GhMTU3Ztm0bwcHBREdHM2TIEOX7uLg4vL29MTMzY8OGDRw+fJiPPvoIKysrnj59mu0+XLt2jXbt2nHlyhVWrlxJREQETk5O+Pr6EhwcrBd+1qxZ3Lhxg3Xr1rF27Vp+/vlnevToQUZGhhLm119/pW3btoSGhvLVV18xd+5czp49S/v27UlLy1vrvOzyv4ODAz179iQsLExn2wCffPIJlStXpk+fPrnG7+/vz82bNwkLC+PAgQPY2dnx2Wef0blzZ0qXLs2mTZvYuXMnNjY2dOnSRafS6vjx4/Tq1YtSpUrx+eefs3jxYnbu3El4eHie9vF5ly5domXLlpw5c4Z58+Zx6NAhFi5cSGpqKk+fPsXb25sFCxYAz8pj7THx9vbOMj6NRkPv3r35+OOPGTp0KJGRkUybNo1Nmzbh7u6uV7F+6dIlpk+fzttvv82+ffto3Lgxo0aNMig/iILJ7VqS27Xs6tWrtGzZkh9++IElS5Zw8OBBvL29mTx5stKNEODOnTt07NiRa9euERoayubNm0lKSmLixIlZpuvOnTsMGTIEHx8fvvzyS/z8/IBnlWNeXl6sX7+ew4cPM3XqVHbu3EmPHj2UdefMmUP//v0BlPTm1AU/ISGBdu3a8dVXXzF//nz2799Pp06deOedd7JMn6H3JEIIIYQQr6JXt99HIdq1axf//PMPo0aNAmDAgAFMnTqV9evXM3z4cJ2wpUqV4uDBgxgbGwNQuXJlWrVqxaFDhxg4cKAS7unTpwQFBSktATw8PDh//jzbtm1j7ty5AGzatInLly+zc+dOJZynpyfW1tbMmDGDo0eP4unpSZs2bTAyMqJChQp63Znq1q2rVF4BZGRk0KVLF+Li4li5ciWurq6Ym5vTvHlz4FnXrty6RD148ID58+fj5eXFtm3blOWurq7Url2bwMBAtm7dmqd9Fa8OKysrypcvr9d6KSuJiYmcOXOGKlWqAM/yg7OzM59++im//PKL0m1GpVLRu3dvjh07pjzM+fn50aBBA06cOKF0QevSpQt//fUXs2bNYtiwYRgZ/V+deteuXVm5cqXy+f79+7z33nvEx8djb2/Pd999R0pKCosXL6ZJkyZKOB8fnxz3ITAwkKdPnxIVFUW1atWAZ5V2Dx48ICgoiHHjxlGmTBklvJOTE5999pny2djYmDfffJPY2Fglb7311lvK9xqNhnbt2uHq6kqNGjU4dOgQPXv2zPXYauWU/ydPnoybmxsHDhxQxre6ffs2e/bsYc6cOQZ17XN0dNSpnE9OTmbKlCl0796dPXv2KMu9vLxo1qwZs2bN4uzZswC8//77VKxYkaNHj2JhYQE8+w0LMnnDtGnTMDEx4dy5czpdlQcPHgw8K4Nr164NPPstcivPvvrqK44cOUJwcDDvvvsu8KycrVatGgMGDGDz5s2MGTNGCf/XX3/xzTffUL16dQBef/11jh8/zrZt23j99dfzvV8id7ldS3K7lk2bNo1SpUpx+vRpZVwoT09PUlNT+eijj5g8eTLlypVj2bJl3L9/n1OnTuHk5ARAt27d6Nq1a5aD9d+/f58vvvgCd3d3neWzZ89W/tdoNLi4uFC/fn06duzI5cuXady4MY6OjkrXakO6Iy9dupQ///yTs2fP0qpVK+BZnsrIyCAsLIypU6dSp04dJbyh9yRCCCGEEK8iaWHFs+6AarVaubmztrbmjTfe4Ouvv9ZrTeTt7a3cGAI0btwYgBs3buiEU6lUOm9ZtWGfD3fixAmsrKyUt69a2u4BWXW/yUpYWBjNmjXDwsICExMTTE1NOX78uF73PUPFxMTw5MkTvW4K1apVw93dXS9dhuyreLUYOrubs7OzUlkFUL9+feBZ5ebzY7xol2vPiV9++YWffvpJqYRIT09X/ry8vLhz545eN7sXK3lezHvOzs6YmZkxduxYNm3axG+//WbQPpw4cQIPDw+lskrL19eX5ORkYmJi8pQOgHv37vHWW29RrVo1JU/WqFEDIN/5Miuurq40adJEp8tTWFgYKpWKsWPHGhRHv379dD5/++233L9/n+HDh+v8LpmZmXTt2pXY2FgeP37M48ePiY2NpW/fvkplFTx7gH6xPDBUcnIyJ0+e5M033yy0cfVOnDgBoFeevfHGG1hZWemVZ87OzkplFYCFhQV16tSR8uwlKMi1JCUlhePHj9OnTx8sLS31ypSUlBSlq9zJkydp2LChUlmlNWjQoCzjLleunF5lFcBvv/2Gj48P9vb2GBsbY2pqSseOHYH85/MTJ07g5OSkVFZp+fr6otFolPNZy9B7EiGEEEKIV9F/vsLql19+4dSpU3h7e6PRaHjw4AEPHjxQKpFeHLvE1tZW57O5uTmATlcnAEtLS52HOG3YlJQU5XNiYiL29vZ6YwPZ2dlhYmJCYmJirulfunQp48ePp3Xr1uzevZszZ84QGxtL165d9dJkKO12s+qyULlyZb10GbKv4tXx+PFjEhMTqVy5cq5hbWxsdD6bmZnluFx7Tty9exeAd955B1NTU50/bXebF8d7yi3vOTo6cuzYMezs7JgwYQKOjo44OjqyYsWKHPchMTEx23Nd+31e0pGZmUnnzp2JiIjgvffe4/jx45w7d055WM5vvszO5MmTOX78ONeuXSMtLY1PP/2U/v37Y29vb9D6L+679rfp37+/3m+zaNEiNBoN9+/f5++//yYzMzPL7Ri67Rf9/fffZGRkFOoA/4mJiZiYmOhVgKlUKuzt7XP9feHZb1zYv5vQV5BrSWJiIunp6axatUrvvPXy8gL+r0xJTExUWj09L6tlkPW18NGjR3To0IGzZ8/ywQcfEB0dTWxsLBEREUD+83lhl0dCCCGEEK+y/3yXwA0bNqDRaNi1axe7du3S+37Tpk188MEHRbJtW1tbzp49i0aj0am0unfvHunp6ZQvXz7XOD777DNcXV0JDQ3VWZ6UlFSgdMGzcTtedPv2bYPSJV5dkZGRZGRkFOmU8dpzyN/fX2e8tufVrVs3z/F26NCBDh06kJGRwfnz51m1ahVTp06lYsWK2XaPsbW1zfZcfz6thvrhhx+4dOkSGzdu1OlSrB3MvLD5+PgwY8YMVq9eTZs2bYiPj2fChAkGr/9ihbl2f1etWpVtF6aKFSsqMwrGx8frff/iMm0lxIvjRb348G1jY4OxsTG3bt0yOP25sbW1JT09nYSEBJ1KK41GQ3x8PC1btiy0bYniU65cOYyNjRk6dGi253/NmjWBZ+eEtmL2eVmdy6CfR+BZS6jbt28THR2ttKoC9Abyz6vCLo+EEEIIIV5l/+kWVhkZGWzatAlHR0eioqL0/qZPn86dO3c4dOhQkWzfw8ODR48esXfvXp3lmzdvVr7Xyu4Nv0qlUt6oal2+fFmvG1Ne3rq2bdsWtVqtM04PwK1bt5TuU+Lf6ebNm7zzzjuUKVOGcePGFdl26tatS+3atbl06RItWrTI8q9UqVL5jt/Y2JjWrVsrXeUuXLiQbVgPDw/l4fN5mzdvxtLS0qBxZ56nfbh9MV8+P+tiXuXUwsfCwkLpBrl06VKcnZ1xcXHJ97ZcXFwoW7YsV69ezfa3MTMzU2Z/jIiI0GkBk5SUxIEDB3TirFixIhYWFly+fFln+b59+3Q+q9VqOnbsyBdffJHjjIp5Kc+05dWL5dnu3bt5/PixlGevmJxaNbu5uXHx4kUaN26c5XmrfRnTsWNHfvjhB65evaoTx+eff25wOvKSz/N6vl69elWvzNq8eTMqlQo3NzeD0yiEEEII8ar7T7ewOnToELdv32bRokVZtiZp2LAhn3zyCevXr2fZsmWFvv1hw4axevVqhg8fTlxcHI0aNeL06dMsWLAALy8vOnXqpIRt1KgR0dHRHDhwgEqVKlGqVCnq1q1L9+7dmT9/PgEBAcqsR/PmzaNmzZqkp6cr65cqVYoaNWqwb98+PDw8sLGxoXz58lkOjly2bFnmzJmjDHw9aNAgEhMTCQoKwsLCgoCAgEI/FuLl++GHH5QxXu7du8fXX39NeHg4xsbG7Nmzp9DGEMrOmjVr6NatG126dMHX15cqVapw//59fvzxRy5cuJDlLJ05CQsL48SJE3h7e1O9enVSUlKULr3P56UXBQQEcPDgQdzc3Jg7dy42NjZs3bqVyMhIgoODdQZcN0S9evVwdHRk5syZaDQabGxsOHDgAEePHs1TPM/LLv9r+fn5ERwczHfffce6devyvR14NobfqlWrGD58OPfv36d///7Y2dmRkJDApUuXSEhIUFp0zp8/n65du+Lp6cn06dPJyMhg0aJFWFlZKTOYwrOH+yFDhrBhwwYcHR1p0qQJ586d05nUQWvp0qW0b9+e1q1bM3PmTF577TXu3r3L/v37WbNmDaVKlaJhw4YArF27llKlSmFhYUHNmjWz7M7n6elJly5dmDFjBg8fPsTFxYXLly8TEBBA06ZNGTp0aIGOl3i5crqWrVixgvbt29OhQwfGjx+Pg4MDSUlJ/PLLLxw4cEAZ/2nq1Kls2LCBbt26MW/ePCpWrMi2bdv46aefAHQme8hOu3btKFeuHG+99RYBAQGYmpqydetWLl26pBe2UaNGACxatIhu3bphbGxM48aNla7Sz3v77bfZvHkz3t7ezJs3jxo1ahAZGUlISAjjx4/XGXBdCCGEEOLf7j/dwmr9+vWYmZkxYsSILL8vX748ffr04eDBg1l2HygoCwsLoqKiGDx4MIsXL6Zbt25s3LiRd955RxkHQ2vFihXUrl2bgQMH0rJlS6X1y/vvv8/06dNZv3493t7erFu3jrCwMNq3b5/l/lpaWtKzZ09atmxJYGBgtmnz9/dn3bp1XLp0id69ezNx4kQaNGjAt99+q8zQJV5tI0aMoG3btnh4eDB+/HguXrzIjBkz+Omnn17KW3w3NzfOnTtH2bJlmTp1Kp06dWL8+PEcO3Ysxwqm7Dg7O5Oenk5AQADdunVj6NChJCQksH//fjp37pztenXr1uXbb7+lbt26TJgwgd69e/PDDz8QHh6uzCqXF6amphw4cIA6deowbtw4Bg0axL179zh27Fie49LKLv9rValShfbt22NjY5PrrIiGGDJkCFFRUTx69Ihx48bRqVMnpkyZwoULF3RaJHl6erJ3714ePnzIgAEDmDZtGv369WPkyJF6cS5ZsoQhQ4YQHBxMr169iImJ4eDBg3rhtJVZzZs3x9/fn65duzJjxgzMzc2VB/yaNWuyfPlyLl26hKurKy1bttRr1aWlUqnYu3cv06ZNIzw8HC8vLz7++GOGDh3KiRMn9FrIiJIvu2uZk5MTFy5coGHDhsyePZvOnTszatQodu3apXPeVq5cmZMnT1KnTh3eeustBg8ejJmZGfPmzQOevbTJja2tLZGRkVhaWjJkyBBGjhyJtbU1O3bs0Avr4+PD6NGjCQkJoW3btrRs2TLbWVgrVKjAt99+i7u7O/7+/nTv3l2Z5XLVqlV5P1hCCCGEEK8wlcbQ6cCycT/lMXPOHyBdk1lYacoXE5UR81v0wMbCqljTIf79HqclcPD3KWRq0oo1HUYqU7rXXIGVadG2hBIl371796hRowaTJk0iODi4uJMjiphcd4vG2LFj2b59O4mJiVm2fhJCCCGEEC9XgSus4NnN86P01NwDFiFrE/N/zU2zKPkepyWQmpH/ge0Lg7lxKams+o+7desWv/32G4sXL+bEiRNcv36dKlWqFHeyxEsg192CmTdvHpUrV6ZWrVo8evSIgwcPsm7dOmbPnq20tBJCCCGEEMWrUMawsrGwwoZX86ZViPywMq0glUWi2K1bt4558+bh4ODA1q1bpbLqP0SuuwVjamrK4sWLuXXrFunp6dSuXZulS5cyZcqU4k6aEEIIIYT4n0JpYSWEEEIIIYQQQgghRGH5Tw+6LoQQQgghhBBCCCFKHqmwEkIIIYQQQgghhBAlilRYCSGEEEIIIYQQQogSRSqshBBCCCGEEEIIIUSJIhVWQgghhBBCCCGEEKJEkQorIYQQQgghhBBCCFGiSIWVEEIIIYQQQgghhChRpMJKCCGEEEIIIYQQQpQoUmElhBBCCCGEEEIIIUoUqbASQgghhBBCCCGEECWKVFgJIYQQQgghhBBCiBJFKqyEEEIIIYQQQgghRIkiFVZCCCGEEEIIIYQQokSRCishhBBCCCGEEEIIUaJIhZUQQgghhBBCCCGEKFGkwkoIIYQQQgghhBBClCj/HxMKqwsUQUVwAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 1500x1400 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vecs_Inst =  cal_EmbVec(RefcodeDB, \"Instruction\")   \n",
    "plot_cluster_umap(vecs_Inst, \"Task\", \"UMAPof Embedding of Instruction, Group by Task\")\n",
    "plot_cluster_umap(vecs_Inst, \"Language\", \"UMAPof Embedding of Instruction, Group by Language\")\n",
    "embedding_heatmap(vecs_Inst, RefcodeDB, \"Pairwise Embedding Similarity Heatmap of Instruction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "af6eec97",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qixin/miniconda3/envs/BIA/lib/python3.11/site-packages/umap/umap_.py:1945: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA2kAAAJOCAYAAAAtVk+cAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdeXxM5/4H8M+ZNZN9kV1kQUhssRe1UwSX1lYtFVQXWi1VpaW0va1yKdpfq6UkdL+1VVXt+76rEltIQiQi+57Znt8fuZkamURCkgk+79drXu2c85zzfM/JMTPf8yxHEkIIEBERERERUY0gs3YARERERERE9A8maURERERERDUIkzQiIiIiIqIahEkaERERERFRDcIkjYiIiIiIqAZhkkZERERERFSDMEkjIiIiIiKqQZikERERERER1SBM0oiIiIiIiGoQJmlUKb7//nuEhYXB3t4ekiRh9uzZ1g6pVAEBAQgICLBK3VFRUZAkCVFRUeXeRpIkdOnSxWxZREQEJElCbGxspcZXnW7evIkRI0bA19cXMpkMkiRZO6R76tKli1XijI2NhSRJiIiIqLY6LV13VLms+VlkDda4jomIHlZM0u7D7t277/lFU1qZ4h95kiRh8+bNpW7fokULU7nDhw+XWu7999+HJElQq9VITU0ttdyd9UqSBJlMBhcXF3Tq1AlRUVEQQpS67b0cPHgQI0eORF5eHiZMmIBZs2bd88ddQECAWTyWXhkZGfcdE9V8ERER+PHHH9G1a1fMnDkTs2bNKtd2QgisXbsWzzzzDGrXrg21Wg0HBwc0a9YMkyZNwvnz56s48gdTnKjf/e/R2dkZHTt2RGRkpLVDrDZCCGzYsAFDhw6Fv78/NBoNNBoNgoKCMGTIEPz444/Q6XTWDvORUXxzp7yvitxMIiKiyqWwdgCPK4VCgRUrVqB3794l1p05cwanTp2CQqGAXq8vdR9GoxErV66EJEnQarX4/vvv8cYbb5RZ71tvvQV7e3sYDAZcvXoVa9euxb59+3DixAl88cUX93UsmzZtAgCsWrUKTzzxRLm3k8vlmDFjRqnrbWxs7iuex8GcOXMwbdo0+Pr6WjuU+6LVarFjxw489dRT+P7778u9XVpaGoYMGYKdO3fC2dkZPXv2RFBQELRaLc6dO4evvvoKn3/+OXbs2FHjW4G6d++OJ598EgCg1+tx/fp1/PbbbxgzZgwuXLiAuXPnWjnCqpWWloZhw4Zh+/btcHR0RPfu3VG3bl3IZDJcv34du3fvxurVq/HFF1/g0KFD1g73kTBw4MASLXfr16/HmTNnMGrUqBLrwsLCqi02IiIyxyTNSvr06YPffvsNqampcHNzM1u3fPlyKBQK9OzZE3/++Wep+9i2bRvi4+Px6quvYtWqVVi+fPk9k7QpU6bAy8vL9P7s2bNo27YtvvzyS0yePBmBgYEVPpabN28CgNl+y0OhUNTobpE1mbe3N7y9va0dxn1LSkqC0Wis0DWj1+vx9NNPY+/evRgxYgS+/PJLODo6mpVJTEzEe++9h8zMzMoOudL16NED06ZNM1t27do1NGnSBIsXL8bs2bOh0WisFF3V0uv1GDhwIPbt24eIiAgsWrQITk5OZmWMRiPWrVuHr7/+2kpRPnoGDhyIgQMHmi2LjY3FmTNnEBERUeNvbBARPU7Y3dFKRo8eDa1Wix9++MFsuVarxY8//oi+ffvCw8OjzH0sX74cADB+/Hg8/fTTOHv2LI4dO1ahOJo0aYLOnTtDCIETJ06Ylh88eBB9+/aFq6srbGxs0LBhQ8yePRt5eXmmMsVdOou7ZwUGBpq6yVS24u6ahYWFePfdd1GnTh1oNBq0bNkS27dvBwBkZ2dj4sSJ8PX1hY2NDdq1a4fjx4+Xus/09HSMGzcOnp6e0Gg0aNOmDTZs2GCxrBACK1asQIcOHeDo6AhbW1u0atUKK1assFg+LS0Nr7zyCjw9PWFra4vWrVtj3bp1ZR7jt99+i8aNG8PGxgZ+fn6YOnUqCgoKLJa1NCat+O8xe/ZsnDx5Er169YKDgwOcnJzw9NNPlzp+be3atWjVqhU0Gg08PT0xbtw4pKenV3i8TF5eHmbPno2GDRvCxsYGrq6u6Nu3Lw4ePGhWrkuXLvD39wcAU0twecYxfvfdd9i7dy86deqElStXlkjQgKLk1VIL9blz5zBs2DB4eHhArVYjMDAQkyZNQlpamsW69u/fj86dO8POzg5ubm4YNmwYrl+/XmpsFb0+ShMYGIgGDRqgsLAQ2dnZ5domPj4eY8eOha+vL1QqFWrXro2xY8eWGm92djY+/PBDNG3aFHZ2dnByckLz5s0xc+bMe3YtFEJg4sSJkCQJo0ePLrOlvywrV67Evn370L17d6xYsaJEggYAMpkMgwYNKnGj6s5xnX/88Qc6duwIBwcHs2s1NTUVkyZNQmBgINRqNTw8PDBs2DCLXWHLGmdo6d/ZnfWvW7cOrVu3hq2tLby8vPDqq68iPT29wuejPJ9Fs2bNgiRJ+PXXXy3u46uvvoIkSVi4cGGF67dkxYoVGDBgAAICAkz/nnv16oVdu3ZZLL9mzRp07twZHh4eps+w3r17Y/369fesq7CwEIMGDYIkSZg5c2alxE9E9NATVGG7du0SAMSoUaMqXKZz584CgEhMTBQhISEiLCzMbP0vv/wiAIjffvtNjBo1SgAQhw4dKrH/lJQUoVKpRPPmzYUQQmzdulUAEC+//LLFeO6s9269e/cWAMSvv/4qhBBi9erVQqFQCFtbWzF69GjxzjvviJYtWwoAol27dqKgoEAIIcS1a9fErFmzRLNmzQQA8cYbb4hZs2aJWbNmlXpeivn7+wu1Wn3PcnfHP2DAABEUFCQmTJggxowZI9RqtVCr1eLEiROiVatWonHjxmLixIli+PDhQiaTCVdXV5GZmVmibm9vb9GiRQsREhIi3n77bTFu3Djh4OAgJEkS33//vVl5o9EonnvuOQFABAcHi5dfflm8/vrromHDhgKAeOutt8zK5+bmiiZNmpjO17Rp08Tzzz8vlEql6Nu3rwAgIiMjzbb58MMPBQDh6ekpXnvtNTFp0iRRp04d0a9fPwFAdO7c2ax88bVx7do107Lia65v377C1tZWhIeHi7feekt069ZNABB169YV+fn5ZvtZvny5ACCcnZ3FSy+9JN5++23RqFEj0bJlS+Hj4yP8/f3L9fcpKCgQTzzxhAAgWrRoId555x0xevRoYWtrKxQKhVizZo2pbGRkpHjjjTcEANGsWTPTNbNr164y6+jQoYMAILZu3VqumIodOHBA2NnZCYVCIZ599lkxbdo00aVLFwFA1K9fX6SkpJiV3759u1AqlUKtVosXXnhBTJs2TbRu3Vr4+fmJpk2birs/Nit6fURGRgoAYs6cOSVijYuLE3Z2dqJ27dpmy69du2bx8+TSpUvCw8NDABD9+/cX06ZNE/379xcAhIeHh7h8+bJZ+du3b4vQ0FABQISFhYnJkyeLN998U/Tu3VsolUqRnp5uKnv3dVdYWCieffZZAUC8/fbb9zrtZWrfvr0AIHbs2FHhbYvPX3h4uFAoFGLgwIFi6tSp4tVXXxVCFH021qtXTwAQXbp0EdOmTRPPPvusUCgUws7OThw8eNBsf8WfLZZY+ndWXH/fvn2FSqUSzz//vJg2bZpo166d6ZrOy8sr17FU5LMoPj5eyOVy0bNnT4v7at68uVCpVCWu53spPsa7//3Z2NiItm3birFjx4pp06aJkSNHCgcHByGTycT69evNyn711VcCgPD29hYvvfSSmD59uoiIiBChoaFm16yl6zgzM1N06dJFSJIkPv/88wrFTkT0KGOSdh8qK0mbN2+eACBOnjxpWv/UU08JT09PodPpykzSFi5cKACIzz77TAghhMFgELVr1xaOjo4iNze3RPnSkrS//vpLaDQaIUmSuHbtmsjKyhLOzs5CrVaLM2fOmMrd+UP0o48+MtuHpR8y9+Lv7y/kcrnpB/rdryVLlliMv0OHDiInJ8e0/OeffzYlGUOGDBE6nc60bu7cuWbn6M66AYhu3boJrVZrWh4dHS00Go1wdnYWWVlZpuVLly4VAMTYsWPN9l9YWGj6QXz8+HHT8lmzZgkAYty4cWb1btmyRQAokaRdvnxZKBQK4evrK27dumVanpmZKRo0aFDhJA2A+Pnnn83Kjxw5UgAQP/30k2lZenq6sLe3Fw4ODiImJsa0XKfTiR49eggA5U7SipPM559/XhiNRtPyM2fOCLVaLVxcXMzOaWlJR2l0Op1QKpVCoVCUSDTLYjAYRP369QUAsXnzZrN106dPN/1d7ywfFBQkJEkS+/btMy2/8/q/+wd9Ra+P4h/53bt3N13v7733nhg1apRwcXERHh4eYvv27WZ1lHa+ihPwb775xmz5N998Y6rjTkOGDBEAxLvvvlviXCUlJZnFf+d1l52dLXr27CkkSRLz588vsW1FFP8tlUql6YZPRRSfP0mSxLZt20qsHzNmjAAgpk+fbrZ88+bNpsTcYDCYlt9vkgagxN9p9OjRAoD48MMPy3UsFf0s6tu3r+mz+k6nTp0SAMSwYcPKVa+lY7w7Sbt69WqJsjdv3hQ+Pj6ifv36ZstbtGghVCqVSE5OLrHNnUnj3ddxUlKSCAsLEyqVyuyziYiImKTdl8pK0pKSkoRCoRCvvfaaEKLoTqlMJjPdpS4rSWvSpImQy+VmSdc777wjAIiVK1eWKF9c71tvvSVmzZolZsyYIZ577jlhY2MjAIiJEycKIYRYtWqVAGC6K32n+Ph4oVAoRN26dc2W32+SVvxDx9KrWbNmFuPfvXu32XK9Xi+USqUAIOLi4krEa+lvUFz3gQMHSsQ1YcIEAUB89913pmVNmzYVdnZ2FpODv/76q0RrSWBgoFCpVBZbLbt3714iSfvggw8EALFgwYIS5b/77rsKJ2mdOnUqsZ/idZMnTzYti4qKEgDEpEmTSpQ/dOhQhZK0oKAgoVQqxfXr10use/nll0uc04omaUlJSQKA8PLyKlf5Ynv37hUARJ8+fUqsy8nJEW5ubkKj0YjCwkIhhBB79uwxtUrdLTY2Vsjl8hI/6Ct6fdz5I//ul0KhEG+88YbIyMgw24+l81V8fYeGhpolxkIUJZUhISECgIiPjxdCFJ1DSZJE3bp1zRKC0hRfd8nJyaJVq1ZCoVCIVatW3XO7e7l161aZf8vly5eXuGlz53VVfP6efvrpEtsWFhYKjUYj3NzcLN6s6tWrlwBgloDfb5JmqUUrISFBKJXKEp+RpanoZ9GGDRsEADFz5kyzsuPHj7eYNJZHaUlaaV5//XUBQMTGxpqWtWjRQtjZ2Zm1xFpy53V85coVERQUJOzt7SvcOk5E9DjgxCFW5OnpifDwcPz444+YP38+oqKiYDQaMXr06DK3O3r0KM6ePYvevXubTbwwatQozJ07FytWrMALL7xgcdsFCxYAKHoGkqOjI1q3bo2xY8eayp86dQoALA4g9/PzQ926dXHx4kVkZ2fDwcHhfg7bRK1WlzrmqjTNmzc3ey+Xy+Hh4YHc3FzUqVPHbF3xxBoJCQkl9qNUKi3ORNmxY0d8+eWXOH36NEaMGIG8vDycPXsWPj4++PTTT0uULx7Dc+HCBQBF432uXbuG0NBQi5NidOzYETt27DBbdubMGdM6S+UrqkWLFiWW1a5dGwDMHmtQXG/79u1LlG/Tpg0UivJ9PGRlZeHq1asICQkx1XOnLl264JtvvjGd0+pU1vVsZ2eHVq1aYcuWLbh06RIaN25c5t/C398ffn5+ZuOTKnp93Kl4hk4AMBgMSEhIwHfffYeZM2di9+7dOHbsGJRK5T2PrXPnziXGVEmShE6dOiE6OhpnzpyBn58fjh8/DiEEunbtWuZ+73Tr1i08+eSTuHHjBn777TeEh4eXa7uyiHs87mPFihU4cOCA2bLevXuXuLbatGlTYtsLFy4gPz8fXbp0ga2tbYn1Xbp0wZYtW3D69GnTzJr3y9I14uPjg7p16+LChQvl/ows72cRAISHh6N27dqIjIzE7NmzIZPJUFBQgB9//BFBQUHo1q3bAx3Tna5evYo5c+Zg586dSEhIQGFhodn6mzdvmsaWDh06FNOmTUPjxo3x7LPPokuXLnjyySfh7Oxscd8XLlxAhw4dYDAYsHPnTrRu3brS4iYielQwSbsPMlnRfCtGo7HUMsXrisuWZvTo0diwYQPWrVuHqKgoPPHEEwgJCSlzm+LJCEaOHGm2PCQkBK1atcKePXtw5coV1KtXr8S2iYmJZc6ol5WVBaAogbTEy8sLFy9eRFZW1gMnaffD0mQRCoXC4sQDxUmGpckQ3NzcLP5tio+7eHbA9PR0CCGQkJCADz74oNS4cnNzzbYrbdIXS+e1rG1K+zuUpaxzYTAYTMuK/9bu7u4lystkMtSqVatc9ZXnmgHwQDMuurm5QalUIjU1FYWFhVCr1VUSW3n+fncmaRW9Pkojl8tRp04dvPfee7h48SK+++47/PTTT6XebAEqfmzFCXpFHtuQmJiIrKwsBAcHV9oPaTc3NygUCqSkpFj8W+7fv9/0/xEREVi5cqXF/Vg67uq4FouVdY1cuHCh3J+R5f0sAoquk7Fjx+KDDz7A5s2bER4ejtWrVyMjIwNvv/12pU3adOXKFbRp0wZZWVno2rUr+vfvD0dHR8hkMuzevRt79uwxS9qmTp0KNzc3fP311/jss8+wYMECKBQKhIeHY9GiRSVmDb506RLS09Px5JNPolGjRpUSMxHRo4azO96H4h/BZT08OiUlxaxsafr16wdPT09MmTIFV69exZgxY8osn5eXh59++gkA8Pzzz5d4+GjxbIYVnVWuWHESdOvWLYvri5dbSpYeJqmpqRaT7OLjK/67FR9ny5YtIYq6B1t8Fc94Vlw+OTnZYr2WzmtxXZa2Ke3vUBmKY719+3aJdUaj0XQNl3c/VXnNKBQKtGnTBjqdDnv37i33dhWNray/haX9VPT6KI/iFqKTJ0+WWa6ix1bcqmGpZbk0YWFhWL58OS5fvoxu3bpZvFYqSqFQoHXr1tDr9WYJWUVZSkju51osTpAszVRZVjJ3r2ukvNd7eT+Lir344ouQy+X49ttvARTNCqtQKBAREVGu+spj4cKFSE9Px8qVK7Ft2zYsWrQIH374oWn21rtJkoQXX3wRx48fx+3bt7Fu3To888wz2LBhA/r27Wt2cwgA/vWvf2H27NnYv38/+vXrZzZrMBERFWGSdh8aNGgAlUqFY8eOlToFdfHDV5s2bVrmvhQKBUaMGIGEhATY2tpi2LBhZZZfvXo1srKyEBYWhrFjx1p8KZVKrFy5ssQXY3kUdyfcvXt3iXUJCQmIiYlBUFCQVVrRKpNOp8Phw4dLLN+3bx+Afx7i6uDggJCQEERHR5t1FSyNo6MjAgMDceXKFSQlJZW6/zs1a9as1HWWllWW4nrvniIfKOpSW97p1R0dHREUFIQrV65YTAD27NkD4MEfjDt27FgAwCeffHLPLnPFd/nLup7z8vJw/PhxaDQaNGjQAEDZf4u4uLgS09pX9Pooj+LHApTVUg/8cz737t1b4nwIIUpcy61atYJMJsOuXbvuOdX+nUaPHo0VK1bg/Pnz6Nq1a6nJSUUUd+meM2fOPf+WFVH8+Idjx45Z/OFv6Vp0cXEBUDJ5NRqNpu6vlli6Rm7evImYmBjUrVu33J+R5f0sKla7dm306dMHGzduxIEDB7B3716Eh4fDx8enXPWVR0xMDICiZOpORqOxRFfUu7m5uWHgwIH45Zdf0K1bN0RHR+PKlSslys2aNQsffvghdu3ahb59+zJRIyK6C5O0+2BjY4OhQ4fi9u3b+Pe//11i/dmzZ/Htt9/CwcEBTz/99D339/bbb2PdunXYsmXLPe++Fj8bbeHChfj2228tvvr164ebN2+W+SDs0gwYMABOTk6IjIzEuXPnTMuFEJg+fTp0Ol2l3rG1prufC3XhwgXTM5sGDBhgWj5x4kTk5eVh3LhxFrutXbt2zawL3MiRI6HVavH++++bldu6dWuJ8WgA8Nxzz0Eul+Ozzz4z+wGclZVl8fqqLAMGDIC9vT2+/fZbXLt2zbRcr9dX+FlFo0aNgk6nw/Tp081+dP/999+IjIyEk5NTiYfoVtTIkSPRsWNH7N69G6NHj7b4HLFbt25h3Lhx2Lx5MwCgQ4cOqFu3Lv7880/T8/SKzZkzBykpKRg+fDhUKhUA4Mknn0RgYCA2btxo1sojhMC7775r8cZHRa+PsmRmZiIqKgrAvccj1qlTB127dsW5c+dKtJyvWLEC586dQ7du3eDn5wegqPvcoEGDEBMTY7FrZnJycqmJ+ahRoxAZGYno6Gh069atRKJW/Iy+8j4MOSIiAu3bt8eOHTswZswYUzfFOwkhLC4vi0qlwvDhw5GSkoI5c+aYrdu+fTv+/PNP1KtXDx06dDAtb9WqFQCYznuxzz77zOzfxd22bdtW4t/zjBkzoNPpMGrUqArFXd7PomIvv/wydDodhg4dCiEExo0bV6H67qV4rNndLZ1z587F33//XaL8li1bSlw7Op3OdMOhtIeyz5w5Ex999BF2796N8PDwe3YLJiJ6nHBM2n1asGABjhw5gg8++AAbN25E586dYWNjg0uXLmHDhg0QQuCHH34odeD0nTw9Pcv1A/bKlSvYu3cvgoKC0Llz51LLjR49GuvWrcPy5cvRr1+/ChxVUavIsmXLMHz4cLRt2xbDhg2Du7s7duzYgePHj6NNmzZ4++23K7TP0uj1+jIfYBwREVGhhylXhLe3NzIyMhAWFoa+ffsiMzMTP/30EwoKCrBs2TKzu+Avv/wyDh8+jJUrV+LAgQPo0aMHfHx8cOvWLVy4cAFHjhzBjz/+aIp16tSpWLt2LZYtW4Zz586hU6dOuH79Ov773/+ib9+++OOPP8xiqVevHt5//33MmjULTZs2xdChQ6FQKLBmzRo0adIEFy9erJJz4OzsjM8++wwvvfQSWrRogWHDhsHJyQmbNm2CWq2Gj4/PPcdUFps6dSr++OMPfPfdd4iOjkb37t1x+/Zt/PLLL9DpdFi1atUDt74qFAqsX78eQ4YMwcqVK7FhwwY89dRTCAwMhFarxfnz57F7927odDrTRAsymQxRUVHo1asXwsPDMWTIEPj7++PIkSPYuXMn6tatazbhh0wmw9KlSxEeHo4ePXpg2LBh8PHxwc6dO5GYmIimTZvir7/+MourotdHse3bt5smzjEajUhISMCGDRuQkpKCnj17YtCgQfc8J0uWLMGTTz6JcePG4ffff0doaCjOnz+PDRs2wN3dHUuWLDEr/9VXX+Hvv//Gxx9/jE2bNqFbt24QQuDSpUvYunUrbt26Vepn1gsvvABJkhAREYEuXbpg165dpnFTxa1+5Z1sRqlU4rfffsPQoUMRFRWFtWvXonv37qhXrx4kSUJSUhL27NmDuLg4BAUFVaiVaO7cudizZw/+/e9/4+DBg2jbti1iY2OxevVq2NraIjIy0uy6Hj16NObNm4fZs2fj9OnTqFu3Lo4fP46///4bnTt3NrW+3a1v376ma8rPzw979uzBoUOH0KxZM0yZMqXc8Vbks6hYeHg4/Pz8cP36dfj6+qJPnz7lrq88XnnlFURGRuKZZ57BsGHD4ObmhsOHD+PkyZMWP8OGDRsGW1tbPPnkk/D394dOp8O2bdtw/vx5DBs2rMSkTneaMWMGZDIZ3nvvPYSHh2PTpk2ws7Or1OMhInooVcscko+ojIwM08Oc7ezshFKpFH5+fuK5554ze/bZncp6qPTd7p6Cf9q0aQIWnlN2N51OJzw9PYVCoRBJSUkVrleIoqnL+/TpI5ydnYVKpRLBwcFi5syZZs8ouzvOypyCH3dNCV3WNNn+/v6lThUPC9PXF5dPTU0VL774ovDw8BBqtVq0atVK/Pbbb6XG/Msvv4gePXoIFxcXoVQqha+vr+jSpYtYsGCBuH37tlnZ1NRU8dJLLwl3d3dhY2MjWrZsKdauXWuavvvuh1kLIcSyZctEaGioUKlUonbt2mLKlCkiLy+vwlPwW3qYeFlT3v/666+iefPmQq1WCw8PD/Hiiy+K1NRUYW9vX+JRCGXJyckRM2fOFMHBwUKlUglnZ2fRp08fs+nOyxPPvRiNRrF69WoxcOBA4ePjI1QqlbC1tTU9yPz8+fMltvnrr7/E4MGDRa1atYRSqRT+/v5i4sSJJf5uxfbu3Ss6deokNBqNcHV1FUOGDBFxcXFlXoflvT5Km4Lf3t5etGnTRixcuLDEFPllna/Y2FgxevRo4e3tLRQKhfD29hajR482myL9TpmZmWLmzJmiYcOGQq1WCycnJxEWFibef/99s3otXXdCFD0WQi6Xi5CQENPnyeLFiwUAsWzZMot1lsZoNIq1a9eKZ555Rvj5+Qm1Wi1sbGxEQECAeOaZZ8QPP/xgejzC3efP0r+hYrdv3xYTJ04U/v7+QqlUilq1aonBgweLs2fPWix/8uRJ0b17d2FrayscHR3FgAEDxOXLl8ucgj8yMlKsXbtWtGzZUtjY2AgPDw/x8ssvi9TU1HIf//1+Fgnxz3P+ZsyYUe76LCltCv5du3aJDh06CAcHB+Hs7CzCw8PFiRMnTM+BvLP8V199Jf71r38Jf39/YWNjI9zc3ETbtm3FN998Y/bsvbKu408++UQAEB07dhTZ2dkPdExERI8CSYhKHBBARI+EK1euoH79+hg6dCh++eUXa4dDNdzgwYNx+PBhXL161dR19FEVFRWF0aNHIzIy0qpdv8PDw7F582ZcvXq1ynocEBGR9XBMGtFjLD09vcTzj/Lz8zFp0iQAeOBxZPR42L9/P6ZMmfLIJ2g1xblz57B582b07t2bCRoR0SOKY9KIHmN79uzB2LFj8dRTT6FOnTpISUnBzp07ERsbi27dut1ztlEiABZnMqXK9+OPP+LixYtYtWoVAFR4gh8iInp4MEkjeow1atQIPXv2xIEDB7B+/XoARROZfPTRR5gyZUq5Jw4hoqq3dOlS7Nu3D/7+/li+fDnatWtn7ZCIiKiKcEwaERERERFRDcLb5ERERERERDUIkzQiIiIiIqIa5LEck2Y0GnHz5k04ODhAkiRrh0NEREQPSAiB7Oxs+Pj4cDwtET30Hssk7ebNm/Dz87N2GERERFTJrl+/jtq1a1s7DCKiB/JYJmkODg4Aij7IHR0drRwNERERPaisrCz4+fmZvuOJiB5mj2WSVtzF0dHRkUkaERHRI4TDGIjoUcBO20RERERERDUIkzQiIiIiIqIahEkaERERERFRDcIkjYiIiIiIqAZhkkZERERERFSDWD1JCwgIgCRJJV4TJkywWH737t0Wy1+4cKGaIyciIiIiIqp8Vp+C/9ixYzAYDKb3f//9N3r27IkhQ4aUud3FixfNps93d3evshiJiIiIiIiqi9WTtLuTq08//RR169ZF586dy9zOw8MDzs7OVRgZERERERFR9bN6d8c7abVafP/99xgzZsw9H0bZvHlzeHt7o3v37ti1a1eZZQsLC5GVlWX2IiIiIiIiqolqVJK2fv16ZGRkICIiotQy3t7eWLp0KdasWYO1a9eiQYMG6N69O/bu3VvqNnPmzIGTk5Pp5efnVwXRExERERERPThJCCGsHUSxXr16QaVS4ffff6/Qdv3794ckSdiwYYPF9YWFhSgsLDS9z8rKgp+fHzIzM83GtRERlUWXnIqC2OsQRiNUXu5Q+9e+Z6s/EVWPrKwsODk58budiB4JVh+TViwuLg7bt2/H2rVrK7ztE088ge+//77U9Wq1Gmq1+kHCI6LHmDY5BSlrNiH3xFkYcnIBAcg0NtA0CILb071h27CetUMkIiKiR0iNSdIiIyPh4eGBvn37VnjbU6dOwdvbuwqiIqLHnfbWbSTM/wYFcTegcHSEyscLkCQYc/ORe/o8Cq5dh8+ECNg1C7F2qERERPSIqBFJmtFoRGRkJEaNGgWFwjyk6dOnIyEhAatWrQIALFq0CAEBAWjUqJFpopE1a9ZgzZo11gidiB5xt3/6DQVxCVD7ekOSy03L5fa2kNlpoE28hVtRvyBgznTIbNhiT0RERA+uRiRp27dvR3x8PMaMGVNiXWJiIuLj403vtVotpkyZgoSEBGg0GjRq1Ah//PEHwsPDqzNkInoMFCYkIfevaCicncwStGKSJEHp7gZt0m3knDwLx/atrBAlERERPWpq1MQh1YWDi4moPDL3HEbiklVQ+XqXOUFI4fWbcAnvBs9Rg6sxOiK6E7/biehRUqOm4CciqkmE3gBAKscMjgJCr6+OkIiIiOgxwCSNiKgUylqukBRyGO94hMfdhNFYVNbdrbrCIiIiokcckzQiolLYNgqGytcLutT0UssYMrMhs7eHwxPNqzEyIiIiepQxSSMiKoWkkMNtYG9IKiW0ySmmVjMAEEJAn5EFQ14+nLt3gMqjlhUjJSIiokdJjZjdkYiopnJs1wLGwkLc/uk3aBOSAEkCJAnCaITcTgPX8K5wH9rP2mESERHRI4RJGhHRPTh3aQf7ZqHIOnwSBVdiIfR6qGp7w/GJllD7eVs7PCIiInrEMEkjIioHhYsTXPt0tXYYRERE9BjgmDQiIiIiIqIahEkaERERERFRDcIkjYiIiIiIqAZhkkZERERERFSDMEkjIiIiIiKqQZikERERERER1SBM0oiIiIiIiGoQJmlEREREREQ1CJM0IiIiIiKiGoRJGhERERERUQ3CJI2IiIiIiKgGYZJGRERERERUgzBJIyIiIiIiqkGYpBEREREREdUgTNKIiIiIiIhqECZpRERERERENQiTNCIiIiIiohqESRoREREREVENwiSNiIiIiIioBmGSRkREREREVIMwSSMiIiIiIqpBmKQRERERERHVIEzSiIiIiIiIahAmaURERERERDUIkzQiIiIiIqIahEkaERERERFRDcIkjYiIiIiIqAZhkkZERERERFSDMEkjIiIiIiKqQZikERERERER1SBM0oiIiIiIiGoQJmlEREREREQ1CJM0IiIiIiKiGoRJGhERERERUQ3CJI2IiIiIiKgGYZJGRERERERUgzBJIyIiIiIiqkGYpBEREREREdUgTNKIiIiIiIhqEIW1AyAiepglZxXg8NVU3MzIh0ySEFjLDm2D3OCoUVo7NCIiInpIMUkjIroPRiGw7sQN/PFXIrIL9ZAACAFIErD2RAKGP1EHnYLdrR0mERERPYSYpBFVo9u3cpCclA1JAjy9HeHmbmftkOg+/XYyAauP34BaKUNtFw1kkgQAMBgFbmcXYvneq1DJZXiirpuVIyUiIqKHDZM0omqQmJCJnZsv4XL0bRTk6wAAGlslGoR6oGvvBvD0drByhFQRabla/HE2ESqlDG72arN1cpkELycbJKTnY82JG2gV4AKFnMN/iYiIqPz4y4Goit2IS8eqr4/irxMJUCgkuHvaw93THjKZhJPHbmDVN0eQmJBp7TCpAo5cTUVWvg6utqpSy9SyVyEhPQ9/J2RVY2RERET0KGCSRlSFjAYjfvvvWaSl5cHDywF29mrIZBJkMgn2Dmp4ejkgJTkXG349CyGEtcOlcrqVVQAAkMmkUsuolXIYxD9lqZoJI2BIBfRJgD4VEAZrR0RERFRu7O5IVIWuXklFYkImXFw0Fn/Qy2QSnF1scCMuA3FX0xDA8UsPBblMwr1SaiEEhBCQl5HIURUQAtDHA7o4wJgLwAhABki2gLIOoPQHJN6fJCKimo3fVERVKP5aOvQ6I9Q2pd8PUdsooNMaEB+bXo2R0YMIcreHTAJ0BmOpZXILDbBRyhHEyWGqjxCA9iJQeK4oQZNsAMm+6L8iH9BGA4Xni1rZiIiIajCrJ2kBAQGQJKnEa8KECaVus2fPHrRs2RI2NjYICgrC119/XY0RE5WfwWAsmpO9DFLxrIB6/nB8WLT0d4GXkwbJWQUWu6kajQJpuVo09HZAYC0madXGkALoYgFJCcjsAEle9O9PkgMyW0BSAfrrRV0giYiIajCrJ2nHjh1DYmKi6bVt2zYAwJAhQyyWv3btGsLDw9GxY0ecOnUK7777LiZOnIg1a9ZUZ9hE5eLsogGEKErWSmHQGyFJElxcbasxMnoQNko5ItoHwF6txI30fORp9abujdkFOlxPz4e3sw1GtAswJeFUDfQ3isaeSWrL6yUVAPG/chwDSkRENZfVx6S5u5s/7PXTTz9F3bp10blzZ4vlv/76a9SpUweLFi0CAISEhOD48eOYP38+Bg0aVNXhElVIaFMvbPvDBlmZBaUmYZmZBXB20aBBY89qjo4eRHN/F7z5VDB+PXYdV2/nIDVHCwHATiXHE0GuGNamDvyYeFcfYSyaIERSll1OUgKGTAA6AKXPzklERGRNVk/S7qTVavH9999j8uTJpd59PnToEJ566imzZb169cLy5cuh0+mgVN7jC5pqLO2t28g+cgrapNuQFHJogvxh37oZ5HYP7w9dO3s1nuxaF5t/O4/MjAI4OqlN17YQAlkZBTAaBDr2qAuNhtfuw6axrxMa+Tji8q0cJGbmQyZJ8K9lhzpMzqzACEDcs3sxIBWVFcai/yUiIqqBalSStn79emRkZCAiIqLUMklJSfD0NG9x8PT0hF6vR0pKCry9vUtsU1hYiMLCQtP7rCw+t6gmEXoDbv+6ERnb98OQkwMJEgQEMiQZlGs2weOFwXBo3czaYd63J7vVhU5nwP6dMbiVmAO5vGhmQKPBCDt7NXr2Dka7ToHWDpPukyRJCPZyQLAXH0huXXJAZgMYc0rv7ggAQv+/CUV4U4SIiGquGpWkLV++HH369IGPj0+Z5e5uZSseuF9a69ucOXPwwQcfVE6QVOlu/7oRaRu2Qm5rC3Vtn39amvR6aG+nIvHr7yFTqWDXLMTKkd4fmUxC9z4N0LSFL/46mYCb1zMBCahdxxlNW/jCjbP/ET04SQIUtQHt/2ZvtDTNvjAWjVlT+RZNJkJERFRD1ZgkLS4uDtu3b8fatWvLLOfl5YWkJPOZuZKTk6FQKODmZvkZU9OnT8fkyZNN77OysuDn5/fgQdMD0ybeQsb2fZDb2kLh4mS2TlIooPLygDYhESnr/oRt04YP9SQM7p726N6ngbXDIHp0KX2KJgUxZP4zu2MxYQCMeYDMviiZIyIiqsFqTJIWGRkJDw8P9O3bt8xy7dq1w++//262bOvWrWjVqlWp49HUajXU6jK6v5DVZB8+BUNOLtS1LbeeSpIEhasLCq7Go+DyNWiCg6o5QiJ6aEhqwKY5UHAGMGYCRlHUoib+9ygMuSOgbgbINNaOlIiIqExWn4IfAIxGIyIjIzFq1CgoFOZ54/Tp0/HCCy+Y3r/yyiuIi4vD5MmTER0djRUrVmD58uWYMmVKdYdNlaAwMRmSJCuzhUymsYHQaqG9lVKNkRHRQ0lmD2ieAGxaAMragLwWoPQtSt407YsSNSIiohquRrSkbd++HfHx8RgzZkyJdYmJiYiPjze9DwwMxKZNmzBp0iR8+eWX8PHxweeff87p9x9Sklxe7ucVSfIacU+BqNJdT8vDsWtpyMzXQa2QIdTHCY1rO0Ih4zV/XyQ5oPAqehERET2EJCEevyd6ZmVlwcnJCZmZmXB05F1Va8rYvh9J3/4IlbdnUcJmgT4rG0JvQMDHU6Hy9qjmCImqTr7WgJUHruHw1VTkFhoACAASlHIJgbXs8XKXINRx48QyROXB73YiepTwNi1ZlUPbMCjcXKBLToWl+wXCYIA+Iwv2zRsxQaNHit5oxDd7YrAjOhlKuQx+rhrUcbNDHTdbuNipcDk5G59tvYRbmQXWDpWIiIiqGZM0siq5gz08RjwDSa2E9mYSjPlFP0iFENBnZaMwIQlqP2/UGtLPypESVa6zNzJx9Goqatmr4KhRmo3LtFHK4eOswY30PPz5d6IVoyQiIiJrqBFj0ujx5tiuJWQqJVLWbUZhXAJEahogAJmdBo7tW8J92L+g8nK3dphElWrfpdvQGwVs1ZY/huUyCfZqBQ5dScWglrXhYMOHLxMRET0umKRRjWDfsinswhoh/+JV6JJTALkcNnX9ofbxtHZoRFUiLjUPNsqyH6hsb6NAeq4OyVmFTNKIiIgeI0zSqMaQ5HLYhtYHQutbOxQqp5sZ+ThyNRW3sgqglMtQ39MBrQNcoVGVnXxQ0TMA7zVvkxBFj/eSyR7eh7gTERFRxTFJI6IK0xmM+PlIPHZeSEZOgR6SVJRQbP07CR5ONhjTIRDN/V2sHWaNFuLtgLiUXAghSn1OYFa+Di62Kng72VRzdERERGRNnDiEiCrsx8Nx2PjXTcgkwM9VAz9XW9Rxs4WXkw1uZxXgy51XcP5mprXDrNE61neHRiVHZr7O4nqt3ogCnRGdG7jfs1skERERPVqYpBFRhcSn5WHXhWTYqxVwtlWZtQIp5DL4OGuQWaDD2hMJ9+zO9zir72mPPk28kFtoQHJWAXQGIwDAKAQy8rRIyixAqI8jejf2tnKkREREVN3Y3ZGIKuTQlRTkFhrg56qxuF6SJLjaqXDpVjZiU/MQWIsPY7ZEkiQMa1MHThol/jybhOSsQgghIADYqxXoFuKB557wh70NP6aJiIgeN/z2J6IKSUjPh1wmlTqOCgDsVHKk5WpxK7OASVoZZJKE8KY+6NrQE3/dyEBGng5qhQyhPo7wcOQ4NCIioscVkzQiqhCFvJyzEgLgpITlo1HJ0TbIzdphEBERUQ3BMWlEVCH1PR0gABiMpSdqWQU62KsVCHK3r77AiIiIiB4RTNKIqELa1XWDi60Kt7MLLbao6Y1GZObr0SbQFbUc1FaIkIiIiOjhxiSNiCrE2VaF556oA4VMws2MfBToDAAAIQQy83VISM9HgJstnmlZ28qREhERET2cOCaNiCqscwMP2CjlWHcyAfFpudAbBIQA7G0U6BTsjmfb1OHEF0RERET3iUkaEd2XtkFuaBnggguJ2UjJLoRcJqGehz28nS1PzU9ERERE5cMkjYjum0ImQ2NfJ2uHQURERPRI4Zg0IiIiIiKiGoRJGhERERERUQ3CJI2IiIiIiKgGYZJGRERERERUgzBJIyIiIiIiqkGYpBEREREREdUgTNKIiIiIiIhqECZpRERERERENQiTNCIiIiIiohqESRoREREREVENwiSNiIiIiIioBmGSRkREREREVIMwSSMiIiIiIqpBmKQRERERERHVIEzSiIiIiIiIahAmaURERERERDUIkzQiIiIiIqIahEkaERERERFRDcIkjYiIiIgeChERERg4cKC1w6gxeD4eXUzSiIiIiKjSSJJU5isiIqLaY9q1axfCw8Ph5uYGW1tbhIaG4q233kJCQkK599GlSxe8+eabVRfkfVi8eDGioqKsHQZVAYW1A6CapTAtCzd3nEJ+YipkSgVcw+qiVpuGkMnl1g6NiIiIHgKJiYmm///ll1/w/vvv4+LFi6ZlGo2mWuP55ptvMH78eIwaNQpr1qxBQEAA4uPjsWrVKixYsACfffZZtcZTGQwGAyRJgpOTk7VDoSrCljQCABgNBlxc9gd2PP0+Ts2KwoWvf8f5z9fi4KuLsOf5T5B+LtbaIRIREdFDwMvLy/RycnKCJEmm90qlEq+88gpq164NW1tbNGnSBD/99JPZ9qtXr0aTJk2g0Wjg5uaGHj16IDc312JdJ06cgIeHBz7++GOL62/cuIGJEydi4sSJWLFiBbp06YKAgAB06tQJ3377Ld5//30AQGpqKoYPH15qXBEREdizZw8WL15sahGMjY0FAJw/fx7h4eGwt7eHp6cnRo4ciZSUFNO22dnZeP7552FnZwdvb28sXLiwRKtceno6XnjhBbi4uMDW1hZ9+vTB5cuXTeujoqLg7OyMjRs3IjQ0FGq1GnFxcSW6OwohMG/ePAQFBUGj0aBZs2ZYvXq1WT3PP/883N3dodFoUL9+fURGRpb9ByWrYJJGAIALX/2G84vXQJ9bAPsATzjW84FjfV/YuDki/UwMjrzxf8i8dN3aYRIREdFDrKCgAC1btsTGjRvx999/46WXXsLIkSNx5MgRAEWtcMOHD8eYMWMQHR2N3bt345lnnoEQosS+du/eje7du+ODDz7Ae++9Z7G+X3/9FVqtFlOnTrW43tnZuVxxLV68GO3atcO4ceOQmJiIxMRE+Pn5ITExEZ07d0ZYWBiOHz+OzZs349atWxg6dKipjsmTJ+PAgQPYsGEDtm3bhn379uHkyZNmcUREROD48ePYsGEDDh06BCEEwsPDodPpTGXy8vIwZ84cfPvttzh37hw8PDxKHM+MGTMQGRmJJUuW4Ny5c5g0aRJGjBiBPXv2AABmzpyJ8+fP488//0R0dDSWLFmCWrVqlfbnIitid0dCVsxNxHy/HQp7DTQezmbrFHY2cKjrjazLCbi47A+0+c8r1gmSiIiIHnq+vr6YMmWK6f3rr7+OzZs349dff0Xbtm2RmJgIvV6PZ555Bv7+/gCAJk2alNjPb7/9hpEjR+Kbb77B8OHDS63v8uXLcHR0hLe39wPF5eTkBJVKBVtbW3h5eZnKLVmyBC1atMAnn3xiWrZixQr4+fnh0qVL8Pb2xsqVK/Hjjz+ie/fuAIDIyEj4+PiYxbhhwwYcOHAA7du3BwD88MMP8PPzw/r16zFkyBAAgE6nw1dffYVmzZpZPIbc3Fx89tln2LlzJ9q1awcACAoKwv79+/HNN9+gc+fOiI+PR/PmzdGqVSsAQEBAQJnnhayHSRrhxp9HoM3MhWN9X4vrJZkMNu5OuLX3L+TE34J9Hc9qjpCIiIgeBQaDAZ9++il++eUXJCQkoLCwEIWFhbCzswMANGvWDN27d0eTJk3Qq1cvPPXUUxg8eDBcXFxM+zhy5Ag2btyIX3/9FU8//XSZ9QkhIEnSA8dVmhMnTmDXrl2wt7cvsS4mJgb5+fnQ6XRo06aNabmTkxMaNGhgeh8dHQ2FQoG2bdualrm5uaFBgwaIjo42LVOpVGjatGmpsZw/fx4FBQXo2bOn2XKtVovmzZsDAF599VUMGjQIJ0+exFNPPYWBAweaEkOqWZikETKi4yFTKsr8EFM52SP7WiKyY24ySSMiIqL7smDBAixcuBCLFi1CkyZNYGdnhzfffBNarRYAIJfLsW3bNhw8eBBbt27FF198gffeew9HjhxBYGAgAKBu3bpwc3PDihUr0LdvX6hUqlLrCw4ORmZmJhITE8tsTbtXXKUxGo3o378/5s6dW2Kdt7e3aVzZ3b+x7uy+aakrZ/HyO7fTaDRl/lYzGo0AgD/++AO+vuY33tVqNQCgT58+iIuLwx9//IHt27eje/fumDBhAubPn1/WYZIVcEwa/Y/lDwhzElCOu1FEREREluzbtw8DBgzAiBEj0KxZMwQFBZlNkAEUJTQdOnTABx98gFOnTkGlUmHdunWm9bVq1cLOnTsRExODYcOGmY3butvgwYOhUqkwb948i+szMjLKHZdKpYLBYDBb1qJFC5w7dw4BAQGoV6+e2cvOzg5169aFUqnE0aNHTdtkZWWZ7Ts0NBR6vd40/g0omsjk0qVLCAkJKfXY7lY8oUh8fHyJWPz8/Ezl3N3dERERge+//x6LFi3C0qVLy10HVR8maQTXxoEw6gwQxtITtcL0bCgdbUvtEklERER0L/Xq1TO1lEVHR+Pll19GUlKSaf2RI0fwySef4Pjx44iPj8fatWtx+/btEsmKh4cHdu7ciQsXLmD48OHQ6/UW6/Pz88PChQuxePFijB07Fnv27EFcXBwOHDiAl19+GR999FG54gKKxm8dOXIEsbGxSElJgdFoxIQJE5CWlobhw4fj6NGjuHr1KrZu3YoxY8bAYDDAwcEBo0aNwttvv41du3bh3LlzGDNmDGQymalVrH79+hgwYADGjRuH/fv348yZMxgxYgR8fX0xYMCAcp9bBwcHTJkyBZMmTcLKlSsRExODU6dO4csvv8TKlSsBAO+//z5+++03XLlyBefOncPGjRsrlAhS9WGSRvDt0wYqZ3sUJKdbXC8MRhSmZsOrczPY+bpXc3RERET0qJg5cyZatGiBXr16oUuXLvDy8jKbQt7R0RF79+5FeHg4goODMWPGDCxYsAB9+vQpsS8vLy/s3LkTZ8+exfPPP1+ilavY+PHjsXXrViQkJODpp59Gw4YN8eKLL8LR0dE0Wci94gKAKVOmQC6XIzQ0FO7u7oiPj4ePjw8OHDgAg8GAXr16oXHjxnjjjTfg5OQEmazoZ/Znn32Gdu3aoV+/fujRowc6dOiAkJAQ2NjYmPYdGRmJli1bol+/fmjXrh2EENi0aROUSmWFzu9HH32E999/H3PmzEFISAh69eqF33//3dRVVKVSYfr06WjatCk6deoEuVyOn3/+uUJ1UPWQRGkdYR9hWVlZcHJyQmZmJhwdHa0dTo1wcelGRP/fekgKGTSeLpApFRBCQJedj/ykNNgHeKH9kjfhEFj27Eh0b0IIGGGEBAkyifdJiIgqA7/b6WGRm5sLX19fLFiwAGPHjrV2OFRDceIQAgAEj+sLha0alyI3Iyc+uWiImhCQa1TwaBeKpu8+zwTtAWkNOlzPvYlrWfHI0ecBkOBh44YAh9rwtvUo1+xTRERE9HA5deoULly4gDZt2iAzMxMffvghAFSoKyM9fpikEYCiQbp1R/REnQEdkLTnDPIS0yBTyuEaVg+uzeoygXhAefp8HLp1EikFaZAgQSlTQsCI67k3kZCXhLqOdRDm1ogta0RERI+g+fPn4+LFi1CpVGjZsiX27dvHh0hTmdjdkV0iqIoJIbAn8TCS8m7DXmUHuSQ3W19o0KLQUIhmbqFo6FzXSlESET3c+N1ORI8S3rYnqmLJBam4XZAGW6VtiQQNANRyFeSSHDGZsdAbLc9ORURERESPD3Z3JKpiN3NvwSAMUMpK/+dmo1AjV5eP5PxU+NjxYeFEdxNCIOF6Js4cv4G4q2kQAvDxc0JYq9oIqOvKLtlERPRIYZJGVMUKDYWQUPYPSLkkhxFGaI3aaoqK6OFhNAps/+MCDuy+ioJ8PZQqGSRIiI9Nx8kj19G8jR/+NbgxFMqSLdVEREQPIyZpRFVMJVPhXkM/jcJomlCEiMwd3H0Vu7deho2NAp7e9qZWMyEE8vN0OLo/FiqVHP0GNbZypERERJWDY9KIqpiXnTtkkqzM8WYFhkLYKjRwt3GrxsiIar6CfB0O7LoKuUIGBycbs26NkiTB1k4FW1slThy5jrTUPCtGSkREVHlqRJKWkJCAESNGwM3NDba2tggLC8OJEydKLb97925IklTideHChWqMmqh8vDTucLVxRq4+D0ZhLLFeZ9RBb9Qj0MEPKjlb0ojudPF8MjIy8uHoZFNqGTsHNfJztTh/JrEaIyMiIqo6Vu/umJ6ejg4dOqBr1674888/4eHhgZiYGDg7O99z24sXL5pNs+vu7l6FkRLdH5kkQxv3MBy8dQIZ2kwoJAWUMgUE/hmvVsfeFyHO9awdKlGNk51ZACEEFIrS7ynKZEWta1mZBdUVFhE9gIiICGRkZGD9+vXWDoWoxrJ6kjZ37lz4+fkhMjLStCwgIKBc23p4eJQrmSOyNkeVPTp7t8W17Ou4ln0dhYZCABLcbVwR6FgHdex9LE7PT/S4UyhlEKJo/Nm9ZnBUcuIQqmLCaETWlQRoM3OhcrKDYz1fSLKq75R08OBBdOzYET179sTmzZurvL677d69G127dkV6enqFfnfFxsYiMDAQp06dQlhYmGn54sWL7zlWm+hxZ/UkbcOGDejVqxeGDBmCPXv2wNfXF+PHj8e4cePuuW3z5s1RUFCA0NBQzJgxA127dq2GiInuj0Zhg1CX+mjgHIRCgxYSJNjI1Zw6nKgMgfXcYGOjQF6uDnb2KotlCgv1UChkCKzPMZ1UdVJPXcaVVVuRcT4OhgIt5DYqOIf6o94LT8Gtef0qrXvFihV4/fXX8e233yI+Ph516tSp0vqqmpOTk7VDIKrxrD4m7erVq1iyZAnq16+PLVu24JVXXsHEiROxatWqUrfx9vbG0qVLsWbNGqxduxYNGjRA9+7dsXfvXovlCwsLkZWVZfYisha5JIetQgONwoYJGtE9eHo7IjjEA9lZBdDrS47pNBoF0tPy4ePnjKB6TNKoaqSeuozTH36HlOOXoHJ1hEN9X6hcHZFy/BJOf/gdUk9drrK6c3Nz8d///hevvvoq+vXrh6ioKNO64jH6O3bsQKtWrWBra4v27dvj4sWLpjKzZ89GWFgYvvvuOwQEBMDJyQnPPvsssrOzTWUKCwsxceJEeHh4wMbGBk8++SSOHTsGoKg1rPgmuIuLCyRJQkREBABg8+bNePLJJ+Hs7Aw3Nzf069cPMTExpv0GBgYCKLqpLkkSunTpAqCou+PAgQPLVX95j5PoUWP1JM1oNKJFixb45JNP0Lx5c7z88ssYN24clixZUuo2DRo0wLhx49CiRQu0a9cOX331Ffr27Yv58+dbLD9nzhw4OTmZXn5+flV1OEREVMn6PtMItf1dcDs5Bxnp+dDpDNDrDMjKLMCtxGy4e9jj6eFNIZNb/SuNHkHCaMSVVVtRkJIJp5A6UDnaQiaXQ+VoC6eQOihIycSV77ZBGEveRKgMv/zyCxo0aIAGDRpgxIgRiIyMLNFV8L333sOCBQtw/PhxKBQKjBkzxmx9TEwM1q9fj40bN2Ljxo3Ys2cPPv30U9P6qVOnYs2aNVi5ciVOnjyJevXqoVevXkhLS4Ofnx/WrFkDoGgugMTERCxevBhAUQI5efJkHDt2DDt27IBMJsPTTz8N4//OxdGjRwEA27dvR2JiItauXWvxGMuqvyLHSfQosfo3mre3N0JDQ82WhYSEID4+vkL7eeKJJ3D5suU7WdOnT0dmZqbpdf369fuOl4iIqpezqy1GvdwGXXrWh9pGgcyMAmSkF0Aml6Fd50CMeqUtvH3ZfYqqRtaVBGScj4NtbfcSvR8kSYJtbXdknItF1pWEKql/+fLlGDFiBACgd+/eyMnJwY4dO8zKfPzxx+jcuTNCQ0Mxbdo0HDx4EAUF/0ykYzQaERUVhcaNG6Njx44YOXKkaR+5ublYsmQJ/vOf/6BPnz4IDQ3FsmXLoNFosHz5csjlcri6ugIomgvAy8vL1F1x0KBBeOaZZ1C/fn2EhYVh+fLlOHv2LM6fPw/gnwnd3Nzc4OXlZdrPne5Vf0WOk+hRYvUxaR06dCjRXH3p0iX4+/tXaD+nTp2Ct7e3xXVqtRpqtfq+YyQiul8GYcSt/NvI1uZCAuCkdoSHjRu7ulaQo7MG4U83Qpen6uP2rRwYjQK1POzg4Fj61PxElUGbmQtDgRYKO8u/IxS2auTf1EKbmVvpdV+8eBFHjx41tUApFAoMGzYMK1asQI8ePUzlmjZtavr/4t9CycnJprFrAQEBcHBwMCuTnJwMoKiVTafToUOHDqb1SqUSbdq0QXR0dJnxxcTEYObMmTh8+DBSUlJMLWjx8fFo3Lh8D5evSP33Ok6iR4nVk7RJkyahffv2+OSTTzB06FAcPXoUS5cuxdKlS01lpk+fjoSEBNM4tUWLFiEgIACNGjWCVqvF999/jzVr1pia44mIaoIbuUn4O+0isrTZECjqniSTZHBROaGZWwjcNRxDVVG2dir4B5W8G09UVVROdpDbqKDPLYTK0bbEen1eIeQ2Kqic7Cq97uXLl0Ov18PX19e0TAgBpVKJ9PR00zKl8p9nbBbfADLe0f3yzvXFZYrXF3edvPvGUXlmVO3fvz/8/PywbNky+Pj4wGg0onHjxtBqteU+xorUf6/jJHqUWL27Y+vWrbFu3Tr89NNPaNy4MT766CMsWrQIzz//vKlMYmKiWfdHrVaLKVOmoGnTpujYsSP279+PP/74A88884w1DoGI7mAUAudvZmLDqQSsPXEDey/dRm6h3tphVbsbOYk4mnwKWdosaBQaOKoc4KhygFquRkpBGg7eOoHb+anWDpOI7sGxni+cQ/2Rd+N2ibFgQgjk3bgN50YBcKznW8oe7o9er8eqVauwYMECnD592vQ6c+YM/P398cMPP1RKPfXq1YNKpcL+/ftNy3Q6HY4fP46QkBAAgEpVNLOqwWAwlUlNTUV0dDRmzJiB7t27IyQkxCxxLG27+6mf6HFk9ZY0AOjXrx/69etX6vo7ZzICigaYTp06tYqjIqKKupaSi6j91xBzOwdavRHF90Dd7NXo18wHfZp4PRbd/HRGPc6knYfeaIC90t7smJUyBRxVDsjS5eBMWjS6+3R4LM4J0cNKkslQ74WnkBN7C5nR8bCt7Q6FrRr6vELk3bgNG3cn1BvZs9Kfl7Zx40akp6dj7NixJaasHzx4MJYvX46FCxc+cD12dnZ49dVX8fbbb8PV1RV16tTBvHnzkJeXh7FjxwIA/P39IUkSNm7ciPDwcGg0Gri4uMDNzQ1Lly6Ft7c34uPjMW3aNLN9e3h4QKPRYPPmzahduzZsbGxKHEt56id6HFm9JY2IHg3xaXlYsPkCzidmwcFGgdouGtR2tYWXkw2yC3T4/lAsNpy+ae0wq8XN3CTk6PJgq7S1mIBJkgRbhQ3SCzNxuyDNwh6IqCZxa14fYe+PRK1WwdCmZSH7SgK0aVmo1boBwmaOrJLnpC1fvhw9evSw+EyxQYMG4fTp0zh58mSl1PXpp59i0KBBGDlyJFq0aIErV65gy5YtcHFxAQD4+vrigw8+wLRp0+Dp6YnXXnsNMpkMP//8M06cOIHGjRtj0qRJ+M9//mO2X4VCgc8//xzffPMNfHx8MGDAgPuqn+hxJInH8JHvWVlZcHJyQmZmJhwdHa0dDtEjYdG2S9h/+Tb8XG0hs5CYpOYUQiZJmDOoKTydHu3JHv5KjUZ0xhU4qhxKLSOEQLYuBy1qNUZ9p8BqjI7o0VQd3+3CaETWlQRoM3OhcrKDYz3fSm9BIyICakh3RyJ6uN3MyMfp+Ay42KosJmgA4GqnwvW0fByMScHTLWpXc4TVS/zvVZbiFjZxz5JEVFNIMhmcgvmsVSKqerz9Q0QPLCE9H3laPextSr/vI0kS5DIJV29X/jTVNY2D0g4SAKMofdYxvVEPSZLgoLSvvsCIiIjoocAkjYiokvnaeUGjsEGePt/ieiEE8vT5cFI5wlNTq5qjIyIiopqOSRoRPTBfFw1sVQrkFJQ+1b4QAgajQGCtyn+WUE2jlqvQ2KUBJEjI0eWatagZhAE5ulwoZUo0cWkAmcSPYSIiIjLHMWlE9MB8nDVo5ueEg1dSYW+jsDguLS1XCwcbBdrXezxajgIdisatnEu/hBydeRdPB6UdmrqFwsfO0xqhERERUQ3HJI2IKsUzLWvjSnIOEtLzUcteDY1KDgDQG41Iz9VBZzBiSKva8HrEZ3YsJkkSghzroLadNxJyk5CtywEkCc4qR/jYekIhk1s7RCIiIqqhmKQRUaXwd7PDW70aIHL/NVy9nYuUnELTw6xd7FTo38wP4U29rRqjNajkSgQ6cjY4IiIiKj8maURUaYLc7fHBwMY4fzMLl29lQ28UcLdXo1WAa5kzPxIRERHRP/iriYgqlUyS0NjXCY19nawdChEREdFDidOKERERET3CZs+ejbCwMNP7iIgIDBw4sErr7NKlC958803T+4CAACxatKhK63zYSJKE9evXV2kdUVFRcHZ2rtI6qGowSSMiIiIqB2E0ojA+AXnnL6MwPgHCWPoD6ytLUlISXn/9dQQFBUGtVsPPzw/9+/fHjh07qrxuqlqJiYno06dPpe3PUiI8bNgwXLp0qdLqoOrD7o5ERI+xwgIdUm/nwSgE3GrZQmOrsnZIRDVS3sWrSP9zJwquXYfQaiGpVLAJ9INLn26wbRBUJXXGxsaiQ4cOcHZ2xrx589C0aVPodDps2bIFEyZMwIULF6qkXqoeXl5eVV6HRqOBRqOp8nqo8rEljYjoMZSdVYCtv0dj0ce7sWTBPnz92X4s+ng3Nq0/h4y0PGuHR1Sj5F28ilsrfkb+hSuQOzpAVdsXckcH5F+4glsrfkbexatVUu/48eMhSRKOHj2KwYMHIzg4GI0aNcLkyZNx+PBhU7nMzEy89NJL8PDwgKOjI7p164YzZ848UN2FhYWYOnUq/Pz8oFarUb9+fSxfvty0/vz58wgPD4e9vT08PT0xcuRIpKSkPFCdK1asQKNGjaBWq+Ht7Y3XXnvNtC4+Ph4DBgyAvb09HB0dMXToUNy6dcu0vrhL54oVK1CnTh3Y29vj1VdfhcFgwLx58+Dl5QUPDw98/PHHZnVKkoRvvvkG/fr1g62tLUJCQnDo0CFcuXIFXbp0gZ2dHdq1a4eYmBjTNpa6i7755pvo0qWL6X2XLl0wceJETJ06Fa6urvDy8sLs2bNL1H1nd8cbN27g2WefhaurK+zs7NCqVSscOXIEABATE4MBAwbA09MT9vb2aN26NbZv325WX1xcHCZNmgRJkiD973mllro7LlmyBHXr1oVKpUKDBg3w3XfflYjr22+/xdNPPw1bW1vUr18fGzZssPxHoyrDJI2I6DGTkZ6PqK+PYMefF5GbWwh7BxUcHFUoKNBhz9bLiFpyBCnJOdYOk6hGEEYj0v/cCUNmFlT+fpDb2UKSyyC3s4XK3w+GzCyk/7mr0rs+pqWlYfPmzZgwYQLs7OxKrC/+4S2EQN++fZGUlIRNmzbhxIkTaNGiBbp37460tLT7rv+FF17Azz//jM8//xzR0dH4+uuvYW9vD6Com17nzp0RFhaG48ePY/Pmzbh16xaGDh163/UtWbIEEyZMwEsvvYSzZ89iw4YNqFevnukYBw4ciLS0NOzZswfbtm1DTEwMhg0bZraPmJgY/Pnnn9i8eTN++uknrFixAn379sWNGzewZ88ezJ07FzNmzDBLcAHgo48+wgsvvIDTp0+jYcOGeO655/Dyyy9j+vTpOH78OACYJYzltXLlStjZ2eHIkSOYN28ePvzwQ2zbts1i2ZycHHTu3Bk3b97Ehg0bcObMGUydOhXG/11XOTk5CA8Px/bt23Hq1Cn06tUL/fv3R3x8PABg7dq1qF27Nj788EMkJiYiMTHRYj3r1q3DG2+8gbfeegt///03Xn75ZYwePRq7du0yK/fBBx9g6NCh+OuvvxAeHo7nn3/+ga4nqjh2dyQiesxsXH0W12PT4e5hD4Xin3t1KpUCBkc1EhOysP6XvzD2tXamu7FEjyvtjUQUXLsOhXutEv8eJEmCwr0WCq7FQ3sjEeo6vpVW75UrVyCEQMOGDcsst2vXLpw9exbJyclQq9UAgPnz52P9+vVYvXo1XnrppQrXfenSJfz3v//Ftm3b0KNHDwBAUNA/XTqXLFmCFi1a4JNPPjEtW7FiBfz8/HDp0iUEBwdXuM5///vfeOutt/DGG2+YlrVu3RoAsH37dvz111+4du0a/PyKnjv53XffoVGjRjh27JipnNFoxIoVK+Dg4IDQ0FB07doVFy9exKZNmyCTydCgQQPMnTsXu3fvxhNPPGGqZ/To0aYE85133kG7du0wc+ZM9OrVCwDwxhtvYPTo0RU+pqZNm2LWrFkAgPr16+P//u//sGPHDvTs2bNE2R9//BG3b9/GsWPH4OrqCgCmJBUAmjVrhmbNmpmdr3Xr1mHDhg147bXX4OrqCrlcDgcHhzK7Uc6fPx8REREYP348AJhaZefPn4+uXbuaykVERGD48OEAgE8++QRffPEFjh49it69e1f4PND9YUsaEdFjJOlmFi5fuA1HRxuzBK2YXC6Dk4sN4q6mIf5auhUiJKpZDDl5EFotZDY2FtfLbNQQWi0MOZXbTVgIAQD3vFFy4sQJ5OTkwM3NDfb29qbXtWvXzLroleaHH34w227fvn04ffo05HI5OnfuXGqdu3btMtuuOJksT513S05Oxs2bN9G9e3eL66Ojo+Hn52dK0AAgNDQUzs7OiI6ONi0LCAiAg4OD6b2npydCQ0Mhk8nMliUnJ5vtv2nTpmbrAaBJkyZmywoKCpCVlVWh47pzvwDg7e1dou5ip0+fRvPmzU0J2t1yc3MxdepU03Hb29vjwoULppa08oqOjkaHDh3MlnXo0MHsPN4du52dHRwcHEqNnaoGW9KIiB4jVy+noLBADydnyz84AcDGRoGsjAJcvZwC/yDLPxiIHhdye1tIKhWMBQWQ29mWWG8sKISkUkFuX3Ldg6hfvz4kSUJ0dHSZ0+UbjUZ4e3tj9+7dJdaVZ+r1f/3rX2jbtq3pva+vr9lYp9Lq7N+/P+bOnVtinbe39z3rvNu9JrYQQlhMVu9erlQqzdZLkmRxmfGurql3linen6VlxdvJZDJTEl1Mp9OViK88dRe71zl4++23sWXLFsyfPx/16tWDRqPB4MGDodVqy9zOkrvPpaXzW5HYqWqwJY2I6DGi1RoAqey788XrtIWG6gqLqMZS1faGTaAf9LdTSvwwF0JAfzsFNoF1oKpd8eSkLK6urujVqxe+/PJL5ObmllifkZEBAGjRogWSkpKgUChQr149s1etWrXuWY+Dg4PZNhqNBk2aNIHRaMSePXssbtOiRQucO3cOAQEBJeq0NH6uPDEEBASU+liB0NBQxMfH4/r166Zl58+fR2ZmJkJCQipc34Nyd3cvMebr9OnTD7TPpk2b4vTp06WO+9q3bx8iIiLw9NNPo0mTJvDy8kJsbKxZGZVKBYOh7M/tkJAQ7N+/32zZwYMHrXIeqWxM0oiIHiMODkVjVgyG0u+IGo1FP0TtHdXVEhNRTSbJZHDp0w1yJ0do467DkJsHYTDAkJsHbdx1yJ0c4dKnKyRZ5f+k+uqrr2AwGNCmTRusWbMGly9fRnR0ND7//HO0a9cOANCjRw+0a9cOAwcOxJYtWxAbG4uDBw9ixowZpkkvKiogIACjRo3CmDFjsH79ely7dg27d+/Gf//7XwDAhAkTkJaWhuHDh+Po0aO4evUqtm7dijFjxtwzSSjN7NmzsWDBAnz++ee4fPkyTp48iS+++MJ0jE2bNsXzzz+PkydP4ujRo3jhhRfQuXNntGrV6r7qexDdunXD8ePHsWrVKly+fBmzZs3C33///UD7HD58OLy8vDBw4EAcOHAAV69exZo1a3Do0CEARePT1q5di9OnT+PMmTN47rnnSrRsBQQEYO/evUhISCh1ps23334bUVFR+Prrr3H58mV89tlnWLt2LaZMmfJA8VPlY5JGRPQYadDIEw6ONsjOKiy1TG52IWztVGjUtOqf4UP0MLBtEATPMc9C07AeDFnZ0N64CUNWNjQN68NzzLNV9py0wMBAnDx5El27dsVbb72Fxo0bo2fPntixYweWLFkCoKjle9OmTejUqRPGjBmD4OBgPPvss4iNjTWNr7ofS5YsweDBgzF+/Hg0bNgQ48aNM7Xo+fj44MCBAzAYDOjVqxcaN26MN954A05OTmbjvypi1KhRWLRoEb766is0atQI/fr1w+XLl03HuH79eri4uKBTp07o0aMHgoKC8Msvv9z38T2IXr16YebMmZg6dSpat26N7OxsvPDCCw+0T5VKha1bt8LDwwPh4eFo0qQJPv30U8jlcgDAwoUL4eLigvbt26N///7o1asXWrRoYbaPDz/8ELGxsahbty7c3d0t1jNw4EAsXrwY//nPf9CoUSN88803iIyMNHt8ANUMkri77f4xkJWVBScnJ2RmZsLR0dHa4RARVavtmy5g+6aLsLdXwdZOZdb1MS9Pi+zMQjzZrS76DWpsxSipsul1BuRkF0KSSXBwtIFM9mjN3Fkd3+3CaIT2RiIMOXmQ29tCVdu7SlrQiIg4cQgR0WOma69g5OVqcexgPLKzsqG2Kfoq0BYaoFDK0OqJOniqP8cnPCpysgpx7FAcTh65bmpBdfe0R6t2ddC8jR9UKrmVI3x4SDJZpU6zT0RUGraksSWNiB5DQghcu5KK08duIP5aOoQQqO3vjOat/RAUXOuRa2V5XKWl5OKH5cdxIy4DKpUcGlslhBDIzdVBGAUaNvHEs6NaQG2jvPfOajh+txPRo4QtaUREjyFJkhBUvxaC6t979jd6OBmNAmt+PI3rcRlw97Azey6erZ0KhYV6nP8rCdv+uIB+g5qUsSciIqpu7EhNRET0CIqNSUXc1XS4uFp+cLlarYCtRolTxxKQnVVghQiJiKg0bEkjeggIIZChzcLNvFvQGXRQypXw1njARe1U5vOuiOjxdTn6NnQ6A9Tq0h+Sa++oxu1bObhy4Taat/GrxuiIiKgsTNKIargCfSGOp/yFW/m3oTP+7/kzQuCC7Ao8NLXQyr0pNAob6wZJRDVOQYEOEGU/uFwmkyBJEgoK9NUYGRER3QuTtIdcYXo2tOnZkGvU0Hi5slXlEaMz6nHo1kncKrgNjdwGGqUGkiRBCAGdUYeE3CRojTp09GoDlfzhH/hPRJXH1k4FQEAIUep3Q9FDzcX/yhIRUU3BJO0hlfZXDK79dw8Sd56EoVAHmUIOl6Z1ETikM3x6tmSy9oiIy7mB5IIU2CvsIJf9M022JElQyVWQS3KkFKQhLucG6jsFWjFSIqppQhp7Yt+OK8jP18HW1nISlp1ZCEdnDYJDLD/4loiIrIMThzyEErYdx8FXFiF29R5AAConO8hUSiQf/BvH3v4G5xevwWP4ZIVHjhAC17KuQ4JklqDdSS6TQ5IkXM2+zr85EZnxreOM4BAPZGUUQKs1lFifl6dFoVaP1u3rQFNKEkdERNbBJO0hkxN3C6c/+g66nHw41veFjbsTlPYaqF3s4VjXBwo7NS5HbkbClmPWDpUekF7okaPLg/Ie3RhVMhXydHnQGnXVFBkRPQwkScLAZ5uhfogH0tPykJyUg+ysAmRlFuBWYjbycnRo3a4OuvSsb+1Q6QFJkoT169dbOwyLZs+ejbCwMGuHQfTQYZL2kLn+x2EUJGfAvo6HxS6NNrWcYNQbEPvrHrasPPQklK/TKv/ORGSZvYMaI19qgyEjmiOwnhtkMhnkChkah3nj+Rdb4elnm0GhtNxSTyUJoxGGpBvQX7sIQ9INCKOxWupNSkrC66+/jqCgIKjVavj5+aF///7YsWNHpde1e/duSJKEjIyMStnflClTqiROokcdx6Q9ZG5uPQ6FRg1JVvrPdxs3R6SdvYq8hBTY1eY4g4eVQpLDSe2I2/kpsJGrSy1XaNDC3cYVKhknDiGiktRqBVo+UQctn6gDg8EICYBMznu0FaWPvwLtwW0wJMYBOh2gVELu7Q9V+55Q1KlXZfXGxsaiQ4cOcHZ2xrx589C0aVPodDps2bIFEyZMwIULF6qs7gchhIDBYIC9vT3s7e2tHQ7RQ4ef0g8ZXU4+ZKqyc2uZSgGhN0Cfy4eTPswkSUKggx8ACbpSujLqjHpIkBDoUIeTxRDRPcnlMiZo90EffwUFv38PfdxlSHaOkDx9INk5Qh93uWh5/JUqq3v8+PGQJAlHjx7F4MGDERwcjEaNGmHy5Mk4fPhwifKWWsJOnz4NSZIQGxsLAIiLi0P//v3h4uICOzs7NGrUCJs2bUJsbCy6du0KAHBxcYEkSYiIiABQlHTNmzcPQUFB0Gg0aNasGVavXl2i3i1btqBVq1ZQq9XYt29fie6OERERGDhwIObPnw9vb2+4ublhwoQJ0On++Z5LTExE3759odFoEBgYiB9//BEBAQFYtGhRpZ1XopqOLWkPGRsPZ+Qnp5dZRp+vhUylhMqFd64edn723riZewvxuQlQyQxQy9WQ/W8K/gJDIbRGLWrbeaOOvY+1QyUieiQJoxHag9tgzMmCzNvvnxtiGlvIbPxgTLwO7aFtkNcOgiSr3AQ4LS0Nmzdvxscffww7O7sS652dne9rvxMmTIBWq8XevXthZ2eH8+fPw97eHn5+flizZg0GDRqEixcvwtHRERpN0cPQZ8yYgbVr12LJkiWoX78+9u7dixEjRsDd3R2dO3c27Xvq1KmYP38+goKC4OzsjD179pSof9euXfD29sauXbtw5coVDBs2DGFhYRg3bhwA4IUXXkBKSgp2794NpVKJyZMnIzk5+b6OlehhxSTtIePX9wmknroCo94AmaLkOAIhBApTs+DXvx00Hi5WiJAqk1ySo41HM9im2SA25wZydDmmdWq5CsFOQWji2rDU2R+JiOjBGJNvwpAYB5lLrRI9FiRJgsylFgw342BMvgm5V+1KrfvKlSsQQqBhw4aVut/4+HgMGjQITZo0AQAEBQWZ1rm6ugIAPDw8TElgbm4uPvvsM+zcuRPt2rUzbbN//3588803Zknahx9+iJ49e5ZZv4uLC/7v//4PcrkcDRs2RN++fbFjxw6MGzcOFy5cwPbt23Hs2DG0atUKAPDtt9+ifn1OcEOPFyZpD5nafdrg6k87kRVzEw6BXmaJmhACuddvQ+Vkh8ChXawXJFUqhUyBsFqN0MC5LpLyb0Nn0EEpU8LL1h0ahY21wyMieqSJ/NyiMWjqUsYGq9VAhq6oXGXX/b8JwCq7O/vEiRPx6quvYuvWrejRowcGDRqEpk2bllr+/PnzKCgoKJF8abVaNG/e3GxZcWJVlkaNGkEu/+f3i7e3N86ePQsAuHjxIhQKBVq0aGFaX69ePbi48MYzPV7YMf0ho3Z1ROt5L8GxrjeyryUhJ+4W8pMzkJuQgqzLCVDYqtFsxgi4t67cu25kfRqFDQId/BDsHIRARz8maERE1UDS2AFKJVBYaLlAYSGgVBaVq2T169eHJEmIjo4u9zay/3W5vHOG5zvHewHAiy++iKtXr2LkyJE4e/YsWrVqhS+++KLUfRr/N4vlH3/8gdOnT5te58+fNxuXBsBit8y7KZXmE11JkmSqo7SZqTljNT1umKQ9hJxDA9Axahqavfs8HBv4Qa5WQuPligav9EfHldNQp397a4dIRET0SJB5+EDu7Q9jekqJREEIAWN6CuQ+/pB5VP7YYFdXV/Tq1QtffvklcnNLttRZmibf3b1oVufExETTstOnT5co5+fnh1deeQVr167FW2+9hWXLlgEAVKqiB5sbDP88AD00NBRqtRrx8fGoV6+e2cvPz+9BDrGEhg0bQq/X49SpU6ZlV65cqbRHAhA9LNjd8SFlU8sJ9Ub2RL2RPSGE4Mx+REREVUCSyaBq3xPG1FswJl6HzKVWURfHwkIY01Mgs3eEql3PSp80pNhXX32F9u3bo02bNvjwww/RtGlT6PV6bNu2DUuWLCnRylacOM2ePRv//ve/cfnyZSxYsMCszJtvvok+ffogODgY6enp2LlzJ0JCQgAA/v7+kCQJGzduRHh4ODQaDRwcHDBlyhRMmjQJRqMRTz75JLKysnDw4EHY29tj1KhRlXa8DRs2RI8ePfDSSy9hyZIlUCqVeOutt6DRaPhbhx4rbEl7BPBDi4iIqOoo6tSDTf8RUPjXh8jNgrh1EyI3C4qA+kXLq/A5aYGBgTh58iS6du2Kt956C40bN0bPnj2xY8cOLFmypER5pVKJn376CRcuXECzZs0wd+5c/Pvf/zYrYzAYMGHCBISEhKB3795o0KABvvrqKwCAr68vPvjgA0ybNg2enp547bXXAAAfffQR3n//fcyZMwchISHo1asXfv/9dwQGBlb6Ma9atQqenp7o1KkTnn76aYwbNw4ODg6wsWE3f3p8SOIx7OSblZUFJycnZGZmwtHR0drhEBER0QOqju92YTTCmHwTIj8XksYOMg+fKmtBo3/cuHEDfn5+2L59O7p3727tcIiqBbs7EhEREZWDJJNV+jT7VNLOnTuRk5ODJk2aIDExEVOnTkVAQAA6depk7dCIqg2TNKJqZBRGZBRmQWvUQSVTwlntCJnEu7BERETFdDod3n33XVy9ehUODg5o3749fvjhhxKzQhI9ytjdkd0dqRoIIRCbfQNXsmKRqc2GURghk2RwUjmgvlMA/O1rc2whEdED4Hc7ET1K2JJGVMWEEPgr7QIuZsQAAGwUNpBLMhiEAemFmTiW/Beytblo7NqAiRoRERERcXZHoqqWmJeMS5lXoZQr4aCyh1KmgEySQSkreq+QKXAx8yqS8m9bO1QiIiIiqgGYpBFVsWvZ12EURtjI1RbX2yjUMAgDYrOvV3NkRERERFQTMUkjqkI6ox6381OhkqnKLKeUKXErPwV6o6GaIiMiIiKimopJGlEVMggDjBD3HGsmkyQIIWAUxmqKjIiIiIhqKiZpRFVIJVNCLVdCb9SXWU5nNEAtV0Mhk1dTZERERERUUzFJI6pCMkmGAAc/6I16GEt52oVRGGEQegQ41OYz04iI6LExe/ZshIWFlVkmNjYWkiTh9OnT1RJTsaioKDg7O1drnZZU9vEHBARg0aJFlbIvqlr8RUhUxQLs/eCgske2LrtEd0aDMCJblwtHpQMCHGpbKUIiIioPIYwQmYkQt2OK/lsNXdSTkpLw+uuvIygoCGq1Gn5+fujfvz927NhR5XVXtSlTppgdR0REBAYOHGhWxs/PD4mJiWjcuHE1R/doOnbsGF566SVrh0HlUCOStISEBIwYMQJubm6wtbVFWFgYTpw4UeY2e/bsQcuWLWFjY4OgoCB8/fXX1RQtUcXYKTVo79kSTipH5OhykaXNNv03V5cLZ5UD2nm2gK1CY+1QiYioFCIlFsYj38NwMAqGoz/AcDAKxiPfQ6TEVlmdsbGxaNmyJXbu3Il58+bh7Nmz2Lx5M7p27YoJEyZUWb3Vxd7eHm5ubmWWkcvl8PLygkLBR/tWBnd3d9ja2lo7DCoHqydp6enp6NChA5RKJf7880+cP38eCxYsKLOJ+dq1awgPD0fHjh1x6tQpvPvuu5g4cSLWrFlTfYETVYCL2gndfTqgjUcYfG294KJ2gq+dF9p6hKG7bwe4qJ2sHSLRY0sIASEKYRT5EIIzrFJJIiUWxlNrIFKuAmp7wMELUNtDpFz93/LYKql3/PjxkCQJR48exeDBgxEcHIxGjRph8uTJOHz4sKlcfHw8BgwYAHt7ezg6OmLo0KG4deuWaX1xt8IVK1agTp06sLe3x6uvvgqDwYB58+bBy8sLHh4e+Pjjj83qlyQJS5YsQZ8+faDRaBAYGIhff/3VrMzZs2fRrVs3aDQauLm54aWXXkJOTo5p/e7du9GmTRvY2dnB2dkZHTp0QFxcnFlcxf+/cuVK/Pbbb5AkCZIkYffu3Ra7++3Zswdt2rSBWq2Gt7c3pk2bBr3+n7HfXbp0wcSJEzF16lS4urrCy8sLs2fPNov7s88+Q5MmTWBnZwc/Pz+MHz/eLO7yeOeddxAcHAxbW1sEBQVh5syZ0Ol0Jc77d999h4CAADg5OeHZZ59Fdna2qczmzZvx5JNPwtnZGW5ubujXrx9iYmIs1ieEQL169TB//nyz5X///TdkMplpu9mzZ6NOnTpQq9Xw8fHBxIkTTWXv7u5YVlmyLqvflpg7dy78/PwQGRlpWhYQEFDmNl9//TXq1KljushCQkJw/PhxzJ8/H4MGDarCaInun0quRKCDHwId/KwdCt2DMOqA/ETAoAXkNoCtNySJk7o8aoQQMIgkGIw3YRRZAAQgKSCXvKGQ+UIm8W4zFXVxNF7ZC1GYDTj5/jNbr0oDofSFyEyA8cpeyNzqQKrEccVpaWnYvHkzPv74Y9jZ2ZVYX3wzWwiBgQMHws7ODnv27IFer8f48eMxbNgw7N6921Q+JiYGf/75JzZv3oyYmBgMHjwY165dQ3BwMPbs2YODBw9izJgx6N69O5544gnTdjNnzsSnn36KxYsX47vvvsPw4cPRuHFjhISEIC8vD71798YTTzyBY8eOITk5GS+++CJee+01REVFQa/XY+DAgRg3bhx++uknaLVaHD161OKMx1OmTEF0dDSysrJMvwldXV1x8+ZNs3IJCQkIDw9HREQEVq1ahQsXLmDcuHGwsbExS8RWrlyJyZMn48iRIzh06BAiIiLQoUMH9OzZEwAgk8nw+eefIyAgANeuXcP48eMxdepUfPXVV+X+Gzk4OCAqKgo+Pj44e/Ysxo0bBwcHB0ydOtXsvK9fvx4bN25Eeno6hg4dik8//dSUEOfm5mLy5Mlo0qQJcnNz8f777+Ppp5/G6dOnIZOZX0+SJGHMmDGIjIzElClTTMtXrFiBjh07om7duli9ejUWLlyIn3/+GY0aNUJSUhLOnDljMf6KlCUrEFYWEhIi3nzzTTF48GDh7u4uwsLCxNKlS8vcpmPHjmLixIlmy9auXSsUCoXQarX3rDMzM1MAEJmZmQ8UOxHVHLdvZYttf1wQn320Q3z87hbxxae7xb4dV0RmRn6592E06oUx5bgwXo4SxvOfC+O5RcJ4/gthjPleGNP+EkajsQqPgKqT0WgQhfrzIle7VeRqt4o87S6Rp90t8rQ7RK52i8jT7RMGY4a1w6QKqKrvdmPGTaHb9InQ7fhc6PctK/HS7fhc6DZ9IowZNyu13iNHjggAYu3atWWW27p1q5DL5SI+Pt607Ny5cwKAOHr0qBBCiFmzZglbW1uRlZVlKtOrVy8REBAgDAaDaVmDBg3EnDlzTO8BiFdeecWsvrZt24pXX31VCCHE0qVLhYuLi8jJyTGt/+OPP4RMJhNJSUkiNTVVABC7d++2GPusWbNEs2bNTO9HjRolBgwYYFbm2rVrAoA4deqUEEKId999VzRo0MDs8/jLL78U9vb2pmPp3LmzePLJJ83207p1a/HOO+9YjEMIIf773/8KNzc30/vIyEjh5ORUanlL5s2bJ1q2bGl2fHef97ffflu0bdu21H0kJycLAOLs2bNCiJLHf/PmTSGXy8WRI0eEEEJotVrh7u4uoqKihBBCLFiwQAQHB5f6e9jf318sXLiwXGXJuqze3fHq1atYsmQJ6tevjy1btuCVV17BxIkTsWrVqlK3SUpKgqenp9kyT09P6PV6pKSklChfWFiIrKwssxcRPTrOnUnENwv3Y9vGC0hLyYOuUI/Em1n4fc3fWLroAOJj0++5DyEMQOJO4NYBQJcDKBwAlSugsAMK04Gk3UDyQYhSZumkh4tB3IDeeAMSVJBJdpAkFSRJCUmygQQ7CJEHreEchNDde2f0aNPmAUYdoFBbXq9QFa3X5lVqtcWfNfd6zmZ0dDT8/Pzg5/dPL43Q0FA4OzsjOjratCwgIAAODg6m956enggNDTVrrfH09ERycrLZ/tu1a1fiffF+o6Oj0axZM7OWvg4dOsBoNOLixYtwdXVFREQEevXqhf79+2Px4sVITEws7yko9XjbtWtndl46dOiAnJwc3Lhxw7SsadOmZtt5e3ubHduuXbvQs2dP+Pr6wsHBAS+88AJSU1ORm5tb7lhWr16NJ598El5eXrC3t8fMmTMRHx9vVubu8353HDExMXjuuecQFBQER0dHBAYGAkCJ/dy5fd++fbFixQoAwMaNG1FQUIAhQ4YAAIYMGYL8/HwEBQVh3LhxWLdunVlX0DtVpCxVP6snaUajES1atMAnn3yC5s2b4+WXX8a4ceOwZMmSMre7+0OrrA+zOXPmwMnJyfS684OMiB5uN+LSsfbH08jN1cHT2x6utWzh6GyDWu528PC0R0pyDn6JOoHMjPyyd5R1Gci8AChsAZUTIFMAkgTIlIDKGZCpgfTTQO716jgsqkJCGKE33gAgQZKUJdZLkgQJtjCKXBjE7eoPkGoWlW3R54C+0PJ6vfZ/nxOV2z22fv36kCTJLNGyRAhh8bfP3cuVSvNrXZIki8uMxnvPWFm839LqvrNMZGQkDh06hPbt2+OXX35BcHCw2Xi6irJUp6XfgGUdW1xcHMLDw9G4cWOsWbMGJ06cwJdffgkAZmPKynL48GE8++yz6NOnDzZu3IhTp07hvffeg1arNSt3r3Pcv39/pKamYtmyZThy5AiOHDkCACX2c6cXX3wRP//8M/Lz8xEZGYlhw4aZJgPx8/PDxYsX8eWXX0Kj0WD8+PHo1KmTxeOqSFmqflZP0ry9vREaGmq2LCQkpNQ7CADg5eWFpKQks2XJyclQKBQWZwmaPn06MjMzTa/r1/kji+hRcfRgPHKyC1HL3bbEF7dMJsHd0x4pt3Nx5nhCqfsQQgAZ5wCIojFolihsAaMeyCz7BxPVfEZkwijyIKGUlhHANLbIYGSS9thz9ITk4gvkpZZoSRdCAHmpResdPUvZwf1xdXVFr1698OWXX1ps3cnIyABQ1GoWHx9v9tvm/PnzyMzMREhIyAPHcXdCdfjwYTRs2NBU9+nTp83iO3DgAGQyGYKDg03LmjdvjunTp+PgwYNo3LgxfvzxR4t1qVQqGAxlT94TGhqKgwfNezUcPHgQDg4O8PX1LdcxHT9+HHq9HgsWLMATTzyB4ODgEmPf7uXAgQPw9/fHe++9h1atWqF+/fqmCVHKKzU1FdHR0ZgxYwa6d++OkJAQpKffu+dHeHg47OzssGTJEvz5558YM2aM2XqNRoN//etf+Pzzz7F7924cOnQIZ8+etbivipSl6mX1JK1Dhw64ePGi2bJLly7B39+/1G3atWuHbdu2mS3bunUrWrVqVeKOBQCo1Wo4OjqavYjo4Zefp8X5M4mwtVOVejdXJpOgVMhw6mgZN2cMeUDBbUB2j8cgyNRA7vVqeTYSVSFhACBw769ACQKl382mx4MkySCr1wmS2gHITIDQ5kMYDRDafCAzAZKNY9H6Spw0pNhXX30Fg8GANm3aYM2aNbh8+TKio6Px+eefm7oh9ujRA02bNsXzzz+PkydP4ujRo3jhhRfQuXNntGrV6oFj+PXXX7FixQpcunQJs2bNwtGjR/Haa68BAJ5//nnY2Nhg1KhR+Pvvv7Fr1y68/vrrGDlyJDw9PXHt2jVMnz4dhw4dQlxcHLZu3YpLly6VmjwGBATgr7/+wsWLF5GSkmKxRWf8+PG4fv06Xn/9dVy4cAG//fYbZs2ahcmTJ5eYaKM0devWhV6vxxdffIGrV6/iu+++q/CjnOrVq4f4+Hj8/PPPiImJweeff45169ZVaB8uLi5wc3PD0qVLceXKFezcuROTJ0++53ZyuRwRERGYPn066tWrZ9YlNSoqCsuXL8fff/9tOjaNRmPxd3VFylL1s3qSNmnSJBw+fBiffPIJrly5gh9//BFLly41e/7H9OnT8cILL5jev/LKK4iLi8PkyZMRHR2NFStWYPny5WYz3RDRoy8vVwedzgCVquyZF1UqObKzC2E0lJJcCSMgRFH3xrJIMgDGovL00Crq4igDcK/p9gUklNKySo8VqVYAZM0HQaoVBBTmANm3gMIcSLWCIAt7BlKtgCqpNzAwECdPnkTXrl3x1ltvoXHjxujZsyd27NhhGhYiSRLWr18PFxcXdOrUCT169EBQUBB++eWXSonhgw8+wM8//4ymTZti5cqV+OGHH0w9oGxtbbFlyxakpaWhdevWGDx4MLp3747/+7//M62/cOECBg0ahODgYLz00kt47bXX8PLLL1usa9y4cWjQoAFatWoFd3d3HDhwoEQZX19fbNq0CUePHkWzZs3wyiuvYOzYsZgxY0a5jyksLAyfffYZ5s6di8aNG+OHH37AnDlzKnReBgwYgEmTJuG1115DWFgYDh48iJkzZ1ZoHzKZDD///DNOnDiBxo0bY9KkSfjPf/5Trm3Hjh0LrVZbohXN2dkZy5YtQ4cOHdC0aVPs2LEDv//+u8WeZhUpS9VPEjVgFPzGjRsxffp0XL58GYGBgZg8eTLGjRtnWh8REYHY2FizqWT37NmDSZMm4dy5c/Dx8cE777yDV155pVz1ZWVlwcnJCZmZmWxVI3qIZWcV4LOPdkIul2BnX3rXtfS0PNg7qDH1g54W1wujDoj5HjDkA8oyPhO06YCNO6TAZx80dLIiIQQKDUdhFFmQSSWnNi8qY4BAAVTyplDIKrcbG1WN6vhuF8IIZN0qmiREZVvUFbIKWtBqCkmSsG7dOgwcONDaodBdDhw4gC5duuDGjRslJtOjR4PVn5MGAP369UO/fv1KXR8VFVViWefOnXHy5MkqjIqIajoHRxsE1HND9F9JpSZpQggUFujRrlNgqfuRZEoIp4ZAytGiVjJLP7qEoWidU2jJdWSREAYgNwHIugRoMwCZHLCrAzgGQ1I63HP7qiJJEhQyP2gN5yFEISTJ/NoRwgiBfMgkJ8ilWlaKkmoiSZIBTt7WDoMeY4WFhbh+/TpmzpyJoUOHMkF7hD26t3+I6LHQpr0/lEo5sjILSqwTQiAtNQ929mqEta5d9o6cGwEql6LWMuNdUxAbdUXLNZ6AU7Dl7cmMMBQAN/4Erm8ompQl/xaQewO4tR+49jNE5sV776QKFT2wOhCAgFHkQIgCCFEIo8iDQB5kkiNU8sZ8iDkR1Sg//fQTGjRogMzMTMybN8/a4VAVqhHdHasbuzsSPTqEENi5+RJ2bbkMvd4Ae3s1FAoZdDoDcrK10Ngp8a8hTdCizb0fvSEKkoGb24CCVACiqEWtuGXNxhPw7QVJ5VT1B/WQE8II3NgEZF8pet6cXH3nSkCXWdSq5hsOyd56A9SFEDCKdOiNiTAiFRBGSJINFDIfyCUvSJLKarFRxfG7nYgeJTWiuyMR0f2SJAndegfDy8cRRw/EIe5qGrSFeiiUcjRvUxttnwxAYL3yDYKWbDwgAoYCObFA9rWiMWoKO8AhCLD3Z6tKeeUlFJ3DuxM0oGhyFqUToE0DUk9A2NW558N6q4okSZBLrpDLXP83nbd4pMcXERHRw4NJGhE99CRJQqNm3ght6oXM9HwUFOhhZ6+Cg2PFZ+aTZErAsX7Ri+5P1mVA6EsmaMUkqSj5zU8CCm8DNh7VG5/FkCQA1kkWiYiI7sYkjYgeGZIkwdnV1tphkDYTkO7x9SJTAfpcQJcLznJPRERkjv06iIiocsmU936WnPjfmD92ISUiIiqhQklaQkICjhw5gvT0dIvrU1JSsGrVqkoJjIiIHlJ2fgBE2YmaIa+oy6OG00cTERHdrVxJmtFoxIsvvog6deqgffv28PLywttvvw2dTmdWLiYmBqNHj66SQImI6CHhWB9QOhTN4mhpAmGjDjAWAk4NIZU2bo2IiOgxVq4kbfny5YiKisL48eMRFRWF4cOHY+HChejevTuys7OrOkYiInqISApbwKtz0bgzbWrRLJnCWPT8OV1W0cvOH3Brae1QiR4ZkiRh/fr11g6jTLt374YkScjIyKjWemfPno2wsDDT+4iICAwcOLBaY6io6ooxICAAixYtqvJ6qOLKlaR98803ePPNN/HFF19g5MiRiIqKwqZNm3D27Fn07NkTmZmZVR0nERE9RCSHukDtvkWPLzDqAV0GoM8B5BqgVlugdjhb0eihI4QRoiAFIvdG0X/vNfbyAUVERECSJEiSBKVSCU9PT/Ts2RMrVqyA0Whed2JiIvr06VOl8Tyo9u3bIzExEU5O1n3e5OLFixEVFWV636VLF7z55ptWi6c6REVFwdnZucTyY8eO4aWXXqr+gOieyjW74+XLlzF37lyzZU899RR27NiBp556Cj169MDWrVurJEAiIno4SXa1IWx9AW06oMv530PBPSDJ+ZBoeviIvJtA2mmgILnoxoNMAdh4QLiGQbL1qbJ6e/fujcjISBgMBty6dQubN2/GG2+8gdWrV2PDhg1QKIp+ynl5eVVZDJVFpVLViDirK0nUarVQqWr25527u7u1Q6BSlKslTaVSlRh/BgAtWrTAtm3bcO3aNXTv3h2pqamVHiARET28JEmCpHaFZF8Hkl1tJmj0UBJ5N4GkXUUPapfbAmq3ov/mJQBJu4rWVxG1Wg0vLy/4+vqiRYsWePfdd/Hbb7/hzz//NGsNurO7Y2xsLCRJwn//+1907NgRGo0GrVu3xqVLl3Ds2DG0atUK9vb26N27N27fvm1WX2RkJEJCQmBjY4OGDRviq6++Mq0r3u/atWvRtWtX2NraolmzZjh06JCpTFxcHPr37w8XFxfY2dmhUaNG2LRpEwDL3R3XrFmDRo0aQa1WIyAgAAsWLDCLJyAgAJ988gnGjBkDBwcH1KlTB0uXLjUr88477yA4OBi2trYICgrCzJkzLf5uLXZnV8KIiAjs2bMHixcvNrVaXrt2DfXq1cP8+fPNtvv7778hk8kQExNT5n7nzJkDHx8fBAcHAyiaeG/YsGFwcXGBm5sbBgwYgNjYWNN2BoMBkydPhrOzM9zc3DB16lSIu8bzWuqWGBYWhtmzZ5veZ2Rk4KWXXoKnpydsbGzQuHFjbNy4Ebt378bo0aORmZlpOsbi7e7eb3x8PAYMGAB7e3s4Ojpi6NChuHXrlml9cdfR7777DgEBAXBycsKzzz7L4U9VoFxJWv369c3+Ad6pefPm2LZtG+Lj4zFq1KhKDY6IiIjImoQwFrWg6fMAtXvRQ9olWdF/1e5Fy9NOV3nXxzt169YNzZo1w9q1a8ssN2vWLMyYMQMnT56EQqHA8OHDMXXqVCxevBj79u1DTEwM3n//fVP5ZcuW4b333sPHH3+M6OhofPLJJ5g5cyZWrlxptt/33nsPU6ZMwenTpxEcHIzhw4dDr9cDACZMmIDCwkLs3bsXZ8+exdy5c2Fvb28xvhMnTmDo0KF49tlncfbsWcyePRszZ840Sz4BYMGCBWjVqhVOnTqF8ePH49VXX8WFCxdM6x0cHBAVFYXz589j8eLFWLZsGRYuXFiuc7l48WK0a9cO48aNQ2JiIhITE1GnTh2MGTMGkZGRZmVXrFiBjh07om7duqXub8eOHYiOjsa2bduwceNG5OXloWvXrrC3t8fevXuxf/9+U4Ks1WpNx7dixQosX74c+/fvR1paGtatW1eu+IsZjUb06dMHBw8exPfff4/z58/j008/hVwuR/v27bFo0SI4OjqajnHKlCkl9iGEwMCBA5GWloY9e/Zg27ZtiImJwbBhw8zKxcTEYP369di4cSM2btyIPXv24NNPP61QvFQOohzee+894e/vL/R6fallTp06Jdzc3IRMJivPLq0qMzNTABCZmZnWDoWIiIgqQVV9txvzbwvj5UhhvPqLMMatL/m6+kvR+vzblVqvEEKMGjVKDBgwwOK6YcOGiZCQENN7AGLdunVCCCGuXbsmAIhvv/3WtP6nn34SAMSOHTtMy+bMmSMaNGhgeu/n5yd+/PFHs3o++ugj0a5du1L3e+7cOQFAREdHCyGEaNKkiZg9e7bFmHft2iUAiPT0dCGEEM8995zo2bOnWZm3335bhIaGmt77+/uLESNGmN4bjUbh4eEhlixZYrEOIYSYN2+eaNmypen9rFmzRLNmzUzv7z6vnTt3Fm+88YbZPm7evCnkcrk4cuSIEEIIrVYr3N3dRVRUVKn1jho1Snh6eorCwkLTsuXLl4sGDRoIo9FoWlZYWCg0Go3YsmWLEEIIb29v8emnn5rW63Q6Ubt2bbMY/f39xcKFC83qa9asmZg1a5YQQogtW7YImUwmLl68aDG2yMhI4eTkVGL5nfvdunWrkMvlIj4+3rS++O979OhRIUTRubS1tRVZWVmmMm+//bZo27ZtqeeF7k+5xqS99tpraNeuHXJyckrtxxsWFoaDBw+W2uJGRERE9NAxFBSNQVMqLa+XKQGdvqhcNRJCQJKkMss0bdrU9P+enkXPJGzSpInZsuTkZADA7du3cf36dYwdOxbjxo0zldHr9SV++925X29vbwBAcnIyGjZsiIkTJ+LVV1/F1q1b0aNHDwwaNMis/J2io6MxYMAAs2UdOnTAokWLYDAYIJfLS9QnSRK8vLxMcQPA6tWrsWjRIly5cgU5OTnQ6/VwdHQs89zci7e3N/r27YsVK1agTZs22LhxIwoKCjBkyJAyt2vSpInZOLQTJ07gypUrcHBwMCtXUFCAmJgYZGZmIjExEe3atTOtUygUaNWqVYkuj2U5ffo0ateubepieT+io6Ph5+cHPz8/07LQ0FA4OzsjOjoarVu3BlDURfLO4/H29jb7e1DlKFeS5uXlhb59+96zXHBw8ANdHEREREQ1itymaJIQo66oi+PdjLqi9XKbag0rOjoagYGBZZZR3pFYFid0dy8rniWy+L/Lli1D27ZtzfZTnCyVtd/i7V988UX06tULf/zxB7Zu3Yo5c+ZgwYIFeP3110vEZynRtJSYKO9KkO+M+/Dhw3j22WfxwQcfoFevXnBycsLPP/9cYmzb/XjxxRcxcuRILFy4EJGRkRg2bBhsbW3L3MbOzs7svdFoRMuWLfHDDz+UKFuRSTtkMlmJc3PnuDuNRlPufZWmtMT/7uVl/T2o8pRrTBoRERHRY0ntCth4WH44uxBFy208ispVk507d+Ls2bMYNGhQpe3T09MTvr6+uHr1KurVq2f2ulcyeDc/Pz+88sorWLt2Ld566y0sW7bMYrnQ0FDs37/fbNnBgwcRHBxcIjEszYEDB+Dv74/33nsPrVq1Qv369REXF1eheFUqFQwGQ4nl4eHhsLOzw5IlS/Dnn39izJgxFdovUDTJ3uXLl+Hh4VHivDo5OcHJyQne3t44fPiwaRu9Xo8TJ06Y7cfd3R2JiYmm91lZWbh27ZrpfdOmTXHjxg1cunSpQsd4p9DQUMTHx+P69eumZefPn0dmZiZCQkIqdNz04JikEREREZVCkmSAaxigsAUKbwOGwqKHsxsKi94r7ADXsKJyVaCwsBBJSUlISEjAyZMn8cknn2DAgAH4f/buOzyqKn3g+PdOn0lPCEmoIYTei0ovAiK4CooNEUER114R1kJRVGwgKj91ZZWmK7qirIqKijQBFRQU6S2EEgghvU275/dHZJYhCQRJMinv53nmgbn33HvfmUky88455z1/+9vfuOWWW8r1WtOmTWPGjBm8+uqr7N69m61btzJv3jxmzZpV5nM8+OCDLF++nAMHDvDrr7/y/fffl/oB/5FHHmHFihVMnz6d3bt3s2DBAubMmVNiUYvSJCYmkpyczOLFi9m3bx+vvfbaeRfdiI+P56effiIpKYm0tDRfr5DRaGTs2LE89thjJCYm+g1JLKtRo0ZRp04dhg0bxtq1azlw4ACrV6/mgQce4PDhwwA88MADPP/883z66afs3LmTu+++u9iC35deeimLFi1i7dq1/PHHH4wZM8Yvke3bty99+vRhxIgRvsrrX331FV9//bXvMebm5rJixQrS0tLIz88vFuvAgQNp3749o0aN4tdff+Xnn3/mlltuoW/fvnTt2vW8H7u4MJKkCSGEEEKcheaoB7H9wVEfvPngPFn0r6M+xPar0HXSvv76a+Li4oiPj+fyyy9n5cqVvPbaa/z3v/8tc29TWd1+++3861//Yv78+bRr146+ffsyf/788+pJ83q93HPPPbRq1YrLL7+cFi1a+JXxP13nzp356KOPWLx4MW3btmXKlCk8/fTTjB07tszXGzZsGA899BD33nuvrz7C5MmTy3w8wIQJEzAajbRu3Zro6GiSk5N9+8aNG4fL5fpLvWgADoeDNWvW0KhRI6655hpatWrFbbfdRkFBgW/e3COPPMItt9zC2LFj6d69OyEhIVx99dV+53nsscfo06cPf/vb3xg6dCjDhw8vVmVyyZIlXHTRRYwcOZLWrVszceJEX+9Zjx49uPPOO7nhhhuIjo7mxRdfLBbrqWUcIiIi6NOnDwMHDiQhIYEPP/zwLz12cWE0dT6zEmuI7OxswsLCyMrKuuCJpUIIIYQIvMp4b1dKB2d6UZEQow2skRXWgyaqhnXr1tGvXz8OHz7sK74iRGUoU+EQIUTN43W5Ob52K1k7k1EKQprEEtuvA+agC598LIQQNZGmGcBWJ9BhiErgdDo5dOgQkydP5vrrr5cETVS6MidpBQUFLF26lIMHDxIdHc1VV111XlVpRO108GQeP+xJ448jWei6omGkg57N6tCuQRgmg3z7GCjH1vzO1pc+JDfpGMr7Z0UmDRz169DqnuE0uqrHOUs7CyGEEDXVBx98wLhx4+jYsSOLFi0KdDiiFirTcMejR4/Sp08fDhw44Cv/GRYWxldffUW3bt0qPMjyJsMdK55Sis+2HOXTX4+Q5/RgNRvQAKdHx2TQ6BIfyZ39mhJklc7cynb8h638/Og/cefkE1QvCqOtaD0Xr8tDQcpJNJORjpNH03h4rwBHKoQQZSfv7UKImqRMXRlPPvkkR44c4cknn2TZsmXMnj0bi8XCXXfdVdHxiWpq1a4TfLgxGVA0jLQTE2qjbqiNhpEOwhxmftx3knfW7j+vhRrFhdM9Xra9ugR3dh4hTWJ9CRqA0WIiuHEMyquzY85S3DnFKz+Jv04pF179BB79GF49vWhuixBCCCFECcrUjfHtt9/y+OOP+6rlDBkyhKZNm3LVVVdx/PhxGacr/Li9Ost+OwoKooKLL/zpsJiIDIJNSRkcSMsjITo4AFHWTid+3kn27sM44iJLHc7oiIskNzmVo99vpvGwnpUcYc2jlBu3fgCvfgyFE1CAAYMWjMnQGKMWK0NLhRBCCOGnTD1px44do0+fPn7b+vXrh1KK48ePV0hg1ZLSQc8FbzYoV6CjCZgdR7M5mllAZLCl1DZBViMFLi8/7U+vxMhEzr6j6G4vJoet1DYGswmUImff0UqMrGZSyoPL+zsePQnwoGFHIwgNC7rKweXdjkcdOtdphKgZ9ALwpBXddGegoxFCiCqtTD1pXq8Xu92/4pvNVvQhz+PxlH9U1Y3ygucIuA8VJWko0ExgjAVzYzCGBDrCSpWR78KjK6ym0tdv0TQNgwYn8+SNujJJh03l8qhDeNXJouRMO/33wYSGCaUK8Xj3YdQiMWjSo1zd5bnzOZh7hCN5x3DrHuwmG42D69MgKA6L0Rzo8AJHzwHXfvCkAu6ibZql6D3SkgAGR0DDE0KIqqjMVRt27dqFyfS/5qcWx9u5c2extp07dy6H0KoJ5QXnVvAcBbSiNx40UB7wJIE3FWwdwRgZ2DgrkcVU1EGr6wqDofSsQFdgP0siJ8pfaLMGGCwm3HmFmINK7k3zujygaYQ2q1/J0dUsSnnx6kcA4xkJ2umsKPLw6ClYjM0qMzxRzg7nHeOXE79T4C3EqBkxaAZy3XmkFqSxJ+sA3WI6EWaphcUsvFlQ+CvoeaBZQfszIVMucB8EPR1sXcAQFNg4hRCiiilzklba6u+jR4/2/V8phaZpvgSuVnAngfsIGGygnfZNqWYCZQWVC4W/g6On//4arGVsKGF2M1kFbiKCSh7y6PLoGA0abeqHVXJ0tVtU1+aEtWxIxu/7MTWtV2wulFKK/KNpBNWPJu7STgGKsmZQ5KKrQjSKz8s8RdM0lDKgKxn2W52dLMxg04nfcHndhJpD/H6vdKWT4cpmw/HN9K/XHaux9GHgNY7SwfkH6PlgCPHvytdsoCxFvWzO7WDrKl39QghxmjIlafPmzavoOKon5S0a4qgZS07ANA0IApUPnmNgbljpIQZCRJCFbk2j+HrrMYKsJl/P2im6UhzPLqRhpINOjcMDE2QtZTAaafvw9fz88Bvk7E/BERfpm5/mLXSRn5KO0Wah9YMjZFHrC6Qoa+VSDZBKj9XZ3uwkCr3OYgkagEEzEGIOIsuVzaHcoySGxQcmyEDwpoOeDQZ7yQmYZgCs/2tnlC/tKtO0adNYunQpW7ZsKVN7TdP49NNPGT58eIXGVROMHTuWzMxMli5desHnmj9/Pg8++CCZmZkXfC5RvZQpSRszZkxFx1E96ZmgCoq+ESyNZgClisbi15IkDeD6ixpyOD2fP45kYzMbCLOb0TSNfJeHrAI3MaE2/t636VnnrYmKEX1xSy6eeRdbX/6Q7F2H0T1FvTiaQSO4SSyt772a+oMvCnCU1Z8BGxpGwAOc7efci4YM9aquCj1OjualYjVYS63SadAMaGgczD1Su5I0PQPQi0aWlMoMOIvaVockTami3j/cgLl4D2E5Gzt2LAsWLCi2ffDgwXz99dcXdO4JEyZw3333XdA5qorzTTirkxtuuIGhQ4cGOgwRAOW2krCu6yxbtox3332XTz/9tLxOW7UpL6fKaZ+dgaIParVHiM3MI5e35OutKazedYL0PBcKhcNsYmDrGIa2r0ejSJksHijRl7Si3weTOfHTDrJ2HQJdJ7hJHDG92mK01qLhWBVI02wYtGi86igoS4kf4JXyAAZMhrjKD1CUi0KvE6/ynnMYo9FgJN9TUElRVRHKyzk7lDUNlFY0NLKq82aA60BRrx9ewAiGULA0AWNEhV328ssvLzaiyWotfRj1uSil8Hq9BAcHExwsBYuqOrvdXqx4n6gdylSC/2x2797NP/7xDxo0aMCwYcP49ttvyyOu6kGzUPQUnmMOntLP3ttWQwVbTVzbtSEvXd+Bp4a3ZdpVbXnx+g7c2S9RErQqwGAyEtOzLc1vG0Lz26+g3oDOkqCVM7OxMRo2FHl+i1crpVDKjaIQoxaFQYsKYJTiQpgMRjRNQz9HkqGUwmQot+9FqweDrWg0rzpLpnbqeavq75HejKL5dd70ovd+LaToX2/6n9szKuzSVquV2NhYv1tExP+SQk3T+Ne//sXVV1+Nw+GgWbNmfPbZZ779q1atQtM0li9fTteuXbFaraxdu5Zp06bRsWNHv2u9++67tGnTBqvVSlxcHPfee6/f/rS0tDJdp1OnTtjtdi699FJSU1P56quvaNWqFaGhoYwcOZL8/HzfcUopXnzxRRISErDb7XTo0IGPP/642HlXrFhB165dcTgc9OjRg127dgFFwwGfeuopfvvtNzRNQ9M05s+fX+JzuXHjRgYNGkSdOnUICwujb9++/Prrr35tzvV8er1exo0bR5MmTbDb7bRo0YJXX3211Ndv4cKFREVF4XT6V7MeMWIEt9xyCwC//fYb/fv3JyQkhNDQULp06cKmTZt8jy88PNx33NnaiprlLyVp+fn5zJ8/nz59+tCqVSteeukloqKieO211zhy5Eh5x1h1GcKKvkVThaW3UUWV8jDFVl5cVYzNbCSxbjDNY0OILKWQiBA1kUELwWJsi0ELQlGArnLRVV5R0oYHoxaDxdgWTbvg78tEgASZHERYwij0lr42plIKj/LQwFHL3geMsUWJjDrLUiuqsChBM0VXXlznS6k/e9CcRe/5mrnofV0zF93XnUX7z5aMVrCnnnqK66+/nt9//52hQ4cyatQo0tP9CxJNnDiRGTNmsGPHDtq3b1/sHG+++Sb33HMPd9xxB1u3buWzzz4jMTHxvK8zbdo05syZw/r16zl06BDXX389s2fP5t///jfLli3j22+/5fXXX/e1f/LJJ5k3bx5vvvkm27Zt46GHHuLmm29m9erVfud94oknmDlzJps2bcJkMnHbbbcBRcMBH3nkEdq0aUNKSgopKSnccMMNJT5POTk5jBkzhrVr1/Ljjz/SrFkzhg4dSk5OTpkfp67rNGjQgI8++ojt27czZcoUHn/8cT766KMSr3ndddfh9Xr9Er20tDS++OILbr31VgBGjRpFgwYN2LhxI7/88gv/+Mc/MJtLLjZ3Pm1FNafOw4YNG9T48eNVaGio0jRNhYaGqptvvlkZDAa1evXq8zlVQGVlZSlAZWVlXfjJXEeUyvlKqZyvlcpdpVTe6v/dcr9XKudzpfI2KKV7L/xaQohqSdfdyu09qgrdW1Whe4tyenYqj56udF0vx2vo6nj+CbUjY4/alr5bJWUfUi6vq9zOL0p3IPuQ+nDv5+rTA1+rL5NX+t2WHfxefbTvC/XJ/q9VpjM70KFWvsKdSuUsUyr3G//3yNxVSuUsVyrnS6Wc+8vlUuX63n46T5ZSuSuVylunVP7PxW9564r2e8r5ukqpMWPGKKPRqIKCgvxuTz/9tK8NoJ588knf/dzcXKVpmvrqq6+UUkqtXLlSAWrp0qV+5546darq0KGD7369evXUE088UWosZb3Od99952szY8YMBah9+/b5tv39739XgwcP9p3DZrOp9evX+11r3LhxauTIkaWed9myZQpQBQUFJT6WsvJ4PCokJER9/vnnZX6cJbn77rvViBEjfPfHjBmjhg0b5rt/1113qSFDhvjuz549WyUkJPjeA0JCQtT8+fNLPPe8efNUWFiY7/7Z2oqapUxjL2bNmsU777zDzp07UUrRq1cvxo0bx3XXXYfL5eL999+vmAyyOjDFFX1L6NpTVG5fmSga3/HnHDRDFNg6/FnFSghRG2maCZMWV2Fzz9ILM/n15B9kOLP8ht05THZahSfSNLRxqUUtxIVrHFyfdGcme7OScHndWI1WDJoBr/JQ6HViMZjpVKcNYZaQQIda+SzNiuameQ4BuaD+fC9UelFPlDkRzPGBjLAM3BRNayjtI5Pxz/3uCrl6//79efPNN/22RUb6r716es9YUFAQISEhpKam+rXp2rVrqddITU3l6NGjDBgw4KyxlOU6p7eJiYnB4XCQkJDgt+3nn38GYPv27RQWFjJo0CC/c7hcLjp18l8G5vTzxsXF+eJu1KjRWWM+XWpqKlOmTOH777/n+PHjeL1e8vPzSU5OPq/H+dZbb/Gvf/2LgwcPUlBQgMvlKjZ09HTjx4/noosu4siRI9SvX5958+YxduxY39/lhx9+mNtvv51FixYxcOBArrvuOpo2bVriuc6nrajeypSkTZgwAU3TuOKKK3jllVf8fhjc7or5o1RtaNqfk4Yji9ZL854AvEVDIU31i4Y5nrWylRBC/HUZzizWHd9Enicfh8mB+c95T7rSyfcUsPnkNnSl0zw84RxnEn+Vpml0impDhDWMfdkHyXJlo+sKo2akYVA9EsPiibHXCXSYgaEZwNoKzPXBc7RocWtNA0M4mOqBsTokrmbwVWotaVjZn0VEStx34YKCgooNOzzTmcPdNE1D1/3nSQYFlV5FtqyFKcpyndPbaJp21mNO/bts2TLq16/v1+7M4ihnnvf048tq7NixnDhxgtmzZ9O4cWOsVivdu3fH5fIfrny2mD/66CMeeughZs6cSffu3QkJCeGll17ip59+KvW6nTp1okOHDixcuJDBgwezdetWPv/8c9/+adOmcdNNN7Fs2TK++uorpk6dyuLFi7n66quLnet82orqrUzZQ8eOHdmyZQvLli3j+PHj3HbbbYwcOZLQ0NCKjq/6MIZVj/LBQogaZVv6bvLc+YRa/NfoMmgGgs1B5HsK2J65h/pBcQSZpUJYRdE0jSYhDYkPbkCOOxePXlTxMcgsRZLQtOr9HmkIKZp75k3/c07aab3SShUt1m2MLGpXTYWEhBAfH8+KFSvo379/pV23devWWK1WkpOT6du3718+j8Viwes9RxE3YO3atbzxxhu+kvaHDh0iLS3tvK61du1aevTowd133+3btm/fvnMed/vtt/PKK69w5MgRBg4cSMOG/ssyNW/enObNm/PQQw8xcuRI5s2bV2ridT5tRfVVpiTt119/5bfffmPu3Ll88MEH3HXXXTz88MOMGDGCESNGVHSMQogKppwFuHduwf37z+gZJ9AsNswtO2BudwmGqLqBDk+UIsuVzfHCNGwmW6nDGe1GG9nuHA7lHaVluAyJqWiaphFaG4c11mSnRsw48/5cnNuBb4ijng8Ga9H+ChpS7HQ6OXbsmN82k8lEnTrl2zs7bdo07rzzTurWrcuQIUPIyclh3bp1FbqWWkhICBMmTOChhx5C13V69epFdnY269evJzg4uMzr9MbHx3PgwAG2bNlCgwYNCAkJKXGZgsTERBYtWkTXrl3Jzs7m0UcfPe/y9omJiSxcuJDly5fTpEkTFi1axMaNG2nSpMlZjxs1ahQTJkxg7ty5LFy40Le9oKCARx99lGuvvZYmTZpw+PBhNm7cWOLn6/NpK6q/Mk+U6tChA3PmzOHo0aO89957dOvWjffff59rrrkGTdNYsmRJsT8iQoiqT09PJW/+LAo+fgfPnq2orHT0Y4co/O5TcufOwL3150CHKEqR5crBrbuxGEofZqVpGhoamc7sSoxMiBrGGAHWtkU9Zsr15xx0V9F9a9sKXSft66+/Ji4uzu/Wq1evcr/OmDFjmD17Nm+88QZt2rThb3/7G3v27Cn365xp+vTpTJkyhRkzZtCqVSsGDx7M559/fs6k53QjRozg8ssvp3///kRHR/PBBx+U2O7dd98lIyODTp06MXr0aO6//37q1j2/LyLvvPNOrrnmGm644QYuueQSTp486derVprQ0FBGjBhBcHAww4cP9203Go2cPHmSW265hebNm3P99dczZMgQnnrqqWLnOJ+2ovrTlPrrNWOTkpJ45513WLBgAYcPH8ZsNnPllVf6rW9RFWVnZxMWFkZWVpYM2RS1mnIWkjd/Jt7kvRjqxKCZ/vdhXymFSj8BZguOkXdjSmgVwEhFSZJzj7D++C+EmkPOWhgk25VD4+AGdIvpVGobIaq7SnlvVwr0HIqKhJiLhjhKUR5RRoMGDaJVq1a89tprgQ5FVAMXVHIwPj6e6dOnc/DgQZYtW8aVV17JsmXLyis2IUQFc+/cgvfIgWIJGvzZAxMZjSrIw/XTygBFKM4m1ByC2WDGrZdewEkphQLCrfKFlBAXTNPAGArGqKJ/JUETZZCens7ixYv5/vvvueeeewIdjqgmyqXsoKZpDBkyhCFDhnDy5MnyOKUQohJ4tm4EKJagnaJpGlpwKJ7929HTUzFEyvy0qiTMEkK0LYqjeccwG8wl9qYVeAuxGiw0DKqY8v9CCCHOrnPnzmRkZPDCCy/QokWLQIcjqolyrw0fFRVV3qcUQlQQPfMEmtly1jaa1YaelYGemyNJWhWjaRptI5uT6coix52Lw2THdFoJ/gJPIQpFm8jmUmVQCCECJCkpKdAhiGqoTEma0Wgs8wk1TcPj8fzlgIQQlchqR3nP8fvq9YLBgGaumDWAxIWJtIbTM6Yrv6b9QaYrG6+ngFP9aQ6TnRbhTWkWGh/IEIUQQghxnsqUpCmlCAsL46qrrpJCG0LUIOaWHfEe3IPSdTRDyVNU9dwsjHUbYKhbv8T9IvCibBEMqN+T1IKTpDsz0ZVOsNlBPUcsFqMk10IIIUR1U6YkbezYsXz88ccsWbKEa665hnHjxl3QooNCiKrB3PYiXOu/Rc84gSGybrE5TaogH3QdS5feaOfRoy4qn0EzEOuIJtYRHehQhBBCCHGBylTd8d133yUlJYVXX32V/fv3079/fxITE3nuuec4evRoRccohKgghog62P52E5rVjp56FD0vB+V2oQoL8KYdQ8/NxtyxB+aufQIdqhBCCCFErfGX1knbtWsX77zzDosWLSItLY3LLruMhx9+mAEDBlREjOVO1kkTwp/n4G5cP36PZ+82lMeFphkxRMdi6dIbc5c+0osmhKjy5L1dCFGTXNBi1l6vl2effZann36aq666ik8++aQ8Y6sw8odciJLpmSfRc7PQTBYM0XGSnAkhqg15b699kpKSaNKkCZs3b6Zjx44Vdp2xY8eSmZnJ0qVLK+waQpzpLy1mnZWVxZtvvkm3bt2YNm0azZo149prry3v2IQQlcwQHoWpQQLG2AaSoAkhRICNHTu2aL3KM2579+4tl/PPnz+f8PDwcjlXTZCUlISmaWzZssVv+6uvvsr8+fMDEpOovc5rnbQVK1bw7rvvsnTpUoxGI9dddx2vvvoqPXr0qKj4hBBCCCGqBKUUilyUcqNpZjSCS1xEvjxdfvnlzJs3z29bdHTVKxDkdrsx19ClWsLCwgIdgqiFytST9vTTT5OQkMBll11GcnIyc+bMISUlhXfeeUcSNCGEEELUeF6Vicv7O07vb7j0rUX/en/HqzIr9LpWq5XY2Fi/26n1az///HO6dOmCzWYjISGBp556ym+t2lmzZtGuXTuCgoJo2LAhd999N7m5uQCsWrWKW2+9laysLF8P3bRp04CiNW/PHNoXHh7u60061eP00Ucf0a9fP2w2G++99x4A8+bNo1WrVthsNlq2bMkbb7xx1sf38ccf065dO+x2O1FRUQwcOJC8vDzf/vM93/bt2xk6dCjBwcHExMQwevRo0tLSfPt1XeeFF14gMTERq9VKo0aNePbZZwFo0qQJAJ06dULTNPr16wcU9WgOHz7cdw6n08n9999P3bp1sdls9OrVi40bN/r2r1q1Ck3TWLFiBV27dsXhcNCjRw927dp11tiFOF2ZetKmTZtGaGgot99+Oy1atCAjI4N//vOfJbbVNI2HHnqoXIMUQgghhAgUr8rE7d2BwoWGDbADHnQycXvzwdgKoxZeqTEtX76cm2++mddee43evXuzb98+7rjjDgCmTp0KgMFg4LXXXiM+Pp4DBw5w9913M3HiRN544w169OjB7NmzmTJlii95CA4OPq8YJk2axMyZM5k3bx5Wq5W5c+cydepU5syZQ6dOndi8eTPjx48nKCiIMWPGFDs+JSWFkSNH8uKLL3L11VeTk5PD2rVrOVUu4a+cr2/fvowfP55Zs2ZRUFDApEmTuP766/n+++8BeOyxx5g7dy6vvPIKvXr1IiUlhZ07dwLw888/c/HFF/Pdd9/Rpk0bLBZLiY974sSJLFmyhAULFtC4cWNefPFFBg8ezN69e4mMjPS1e+KJJ5g5cybR0dHceeed3Hbbbaxbt+68nmNRi6ky0DStzDeDwVCWUwZUVlaWAlRWVlagQxFCCCFEOaio93Zd11Whe4vKd69SBe5NqtDzi+9W4N6k8t2rVKF7i9J1vVyvq5RSY8aMUUajUQUFBflu1157rVJKqd69e6vnnnvOr/2iRYtUXFxcqef76KOPVFRUlO/+vHnzVFhYWLF2gPr000/9toWFhal58+YppZQ6cOCAAtTs2bP92jRs2FD9+9//9ts2ffp01b179xLj+eWXXxSgkpKSStx/rvOdimPz5s1KKaUmT56sLrvsMr/2hw4dUoDatWuXys7OVlarVc2dO7fE6515vlPGjBmjhg0bppRSKjc3V5nNZvX+++/79rtcLlWvXj314osvKqWUWrlypQLUd99952uzbNkyBaiCgoISry3EmcrUk3bgwIGKyRAp6qV76qmn/LbFxMRw7NixEtuvWrWK/v37F9u+Y8cOWrZsWSExCiGEEKJ2UuSik4OGrdj8M03TQNnQyUGRi0ZIuV+/f//+vPnmm777QUFBAPzyyy9s3LjRN1QPiqpuFxYWkp+fj8PhYOXKlTz33HNs376d7OxsPB4PhYWF5OXl+c5zIbp27er7/4kTJzh06BDjxo1j/Pjxvu0ej6fUOV0dOnRgwIABtGvXjsGDB3PZZZdx7bXXEhER8ZfO98svv7By5coSewT37dtHZmYmTqfzgpaM2rdvH263m549e/q2mc1mLr74Ynbs2OHXtn379r7/x8XFAZCamkqjRo3+8vVF7VGmJK1x48YVGkSbNm347rvvfPeNZagqt2vXLr8Su1VxEq0QQgghqjel3ICXoiGOJTEChUXtKqCGSFBQEImJicW267rOU089xTXXXFNsn81m4+DBgwwdOpQ777yT6dOnExkZyQ8//MC4ceNwu91nvaamab4hh6eUdMzpiZ6u60DREMVLLrnEr11pn+uMRiPffvst69ev55tvvuH111/niSee4KeffsLhcJz3+XRd58orr+SFF14oti8uLo79+/eXeNz5OPW8nJmwK6WKbTu9kMqpfaeeJyHO5byqO1YUk8lEbGzseR1Tt25dKRsrhBBCiAqlaWaKEjEPUFL1Qi9g/LNd5encuTO7du0qMYED2LRpEx6Ph5kzZ2IwFNWJ++ijj/zaWCwWvF5vsWOjo6NJSUnx3d+zZw/5+flnjScmJob69euzf/9+Ro0aVebHoWkaPXv2pGfPnkyZMoXGjRvz6aef8vDDD5/3+Tp37sySJUuIj4/HZCr+EbdZs2bY7XZWrFjB7bffXmz/qTloJT0npyQmJmKxWPjhhx+46aabgKIEdtOmTTz44INlilOIsihTknbppZeWus9sNhMTE8OgQYMYOXJkib8U57Jnzx7q1auH1Wrlkksu4bnnniMhIeGsx3Tq1InCwkJat27Nk08+WeIQSCGEEEKIC6ERjIEQdDJBmfx6S5RSKAoxEI7G+RXduFBTpkzhb3/7Gw0bNuS6667DYDDw+++/s3XrVp555hmaNm2Kx+Ph9ddf58orr2TdunW89dZbfueIj48nNzeXFStW0KFDBxwOBw6Hg0svvZQ5c+bQrVs3dF1n0qRJZSqvP23aNO6//35CQ0MZMmQITqeTTZs2kZGRwcMPP1ys/U8//cSKFSu47LLLqFu3Lj/99BMnTpygVatWf+l899xzD3PnzmXkyJE8+uij1KlTh71797J48WLmzp2LzWZj0qRJTJw4EYvFQs+ePTlx4gTbtm1j3Lhx1K1bF7vdztdff02DBg2w2WzFhlYGBQVx11138eijjxIZGUmjRo148cUXyc/PZ9y4cefzEgpxVmUqwZ+amsqJEydKvCUnJ/PZZ58xZswYevXqdc5vWs50ySWXsHDhQpYvX87cuXM5duwYPXr04OTJkyW2j4uL4+2332bJkiV88skntGjRggEDBrBmzZpSr+F0OsnOzva7CSGEEEKci6ZpmIyN0bD41khTSi/6l1w0LEX7K3i9tDMNHjyYL774gm+//ZaLLrqIbt26MWvWLN8UlY4dOzJr1ixeeOEF2rZty/vvv8+MGTP8ztGjRw/uvPNObrjhBqKjo3nxxRcBmDlzJg0bNqRPnz7cdNNNTJgwwTf88Gxuv/12/vWvfzF//nzatWtH3759mT9/vq+0/ZlCQ0NZs2YNQ4cOpXnz5jz55JPMnDmTIUOG/KXz1atXj3Xr1uH1ehk8eDBt27blgQceICwszNebOHnyZB555BGmTJlCq1atuOGGG0hNTQWKRna99tpr/POf/6RevXoMGzasxOs8//zzjBgxgtGjR9O5c2f27t3L8uXLiYiIOOdzJERZaerMQcd/0VdffcUNN9zAfffd5zeJ9Xzl5eXRtGlTJk6cWOK3JCW58sor0TSNzz77rMT9JRUnAcjKyvKb1yZEeVO6F1xOMJnRTDVzkU8hhKgKsrOzCQsLq7D3dq/KxOM9iE4Op4Y4GgjBZGxc6eX3hRA1X7klaQDPPvssixYt8q038VcNGjSIxMREv2pG57rue++9V6yqzilOpxOn0+m7n52dTcOGDSVJExVGT0/F9duPuLesRxUWgMGIqUV7LJ16YGrcPNDhCSFEjVPRSRqcGt5Y1JumaWY0giu9B00IUTuUabhjWV188cUcPHjwgs7hdDrZsWOHr1RpWWzevPms7a1WK6GhoX43ISqK5+Bu8ua9jHPFUlROFmgaeFy4N60hf9GrODd8d+6T1CJe3YvT68KrpOKVEKJq0zQNgxaC0RCJQQuRBE0IUWHKtbpjQUHBeRcOmTBhAldeeSWNGjUiNTWVZ555huzsbN9K8o899hhHjhxh4cKFAMyePZv4+HjatGmDy+XivffeY8mSJSxZsqQ8H4oQf4menUHBknfRM09iiI5DM5z2PUhoBHpWOs7vPsEQVRdz8/aln6gWSCtM50DOIY7mHUdXOiaDiUbB9WgS0pBQS/mvNSSEEEIIUV2Ua5L25Zdf0rp16/M65vDhw4wcOZK0tDSio6Pp1q0bP/74o2/ia0pKCsnJyb72LpeLCRMmcOTIEex2O23atGHZsmUMHTq0PB+KEH+J+4+N6BmpGOqckaD9yRAWiTf1KK6NqzA1a1drv4Xdm5XE7+k7cOluLAYLBs2AS3ezI2MvB3OPcHF0B2IddQMdphBCCCFEQJRpTlp6enqp+7xeL0eOHOHDDz/kpZde4o033uCOO+4o1yDLW2WMWxe1U+4/n8WbchBjVEypbfT8XPC4Cb57GoaIOpUYXdVwvCCNH45tRCmFw2QvVs46x52H3WTl0no9CTafu5qYEEKAvLcLIWqWMvWk1alTp0zf+N97771VPkEToiKpvOxzVnHUTGZ0ZwGq8PyWq6gp9mUfxKN7CDEXn3CvaRoh5iCy3Tkk5x6mdYQUWRFCCCFE7VOmJG3KlCmlJmkmk4mYmBguvfTScy5ALURNpzmCUdmZZ22jPG40ownNVvt6iQo8hRzPP4HFaCn1b4qmaRg1Iwdzj0qSJoQQQohaqUxJ2rRp0yo4DCFqBnPbiyg8fACl6yXOSQNQudmYmrdHC4+q5OgCz6170NExa2fvbTRqRlxeF0qpWjtvTwghhBC1V7mW4BeitjO3uxhDeB30k8cpabqnnp0BZguWrn1qZfJhNpgwYDhnuX2v8p61t00IIYQQoiaTJE3UWGkFeaw8tI+Pdv3Ox7u38mNKMrku57kPvACGsEjs19yKISQc/fgR9MyTqPw89JwsvKlHQdexXToMU8uOFRpHVWU32YixR/t6yUqilMKrvDQKrlfJ0QkhhKhOVq1ahaZpZGZmBjqUWkXTNJYuXRroMM7L/PnzCQ8PL5dzJSUloWkaW7ZsKZfzlUaSNFHjeHWdb5J2M2fLBr46sItfUo+w8dhhPt27jVc3r2dz6tEKvb4poRWOsY9g6XsFmj0I5XGDwYC5Q3ccN92LtdfltbqHqGlYY0wGE/megmKJWlF1x1wcJjuNg+sHKEIhhKg6jh07xgMPPEBiYiI2m42YmBh69erFW2+9RX5+7SxAdSH69evHgw8+GOgwLkhlJQmiZA0bNiQlJYW2bdtW6HXKdZ00IaqC7w/t4/tD+7EZjcQ4/ldB0Kt0MgsL+XTvNswGA23rxFZYDMboOOyXXYu6dBiqIB/NYkGz2ivsetVJjL0OHSJb8Xv6DrLdOb510rzKi9vrxmG2c1F0B4LNQYEOVVQwpRSHcrLYmXGCQo8bu8lMi8hoGgaH1eovMkTVpZTCpWfjVS6MmgWLIbRCf1b3799Pz549CQ8P57nnnqNdu3Z4PB52797Nu+++S7169bjqqqsq7Pp/hcvlwmKxBDqMCldbHufpauNjLonRaCQ2tuI+Q54iPWmiRsksLGD90YNYjUZCrTa/N0+jZiDSZsft9fL9oX149bPPiyoPmsmMISRMErQzJIbF0zv2EhJCGmHUDCilYzGYaRWRSL+4bsTJQtY1XrazkPnbfuHtrT/z3cE9/HAkiW8P7mHu7z+zYPuvZLsKAx2iEH4KPGkczf+JI3nrOZr3c9G/+T9R4EmrsGvefffdmEwmNm3axPXXX0+rVq1o164dI0aMYNmyZVx55ZW+tllZWdxxxx3UrVuX0NBQLr30Un777Tff/mnTptGxY0cWLVpEfHw8YWFh3HjjjeTk5PjaKKV48cUXSUhIwG6306FDBz7++OOzxhgfH88zzzzD2LFjCQsLY/z48QCsX7+ePn36YLfbadiwIffffz95eXm+41JSUrjiiiuw2+00adKEf//738THxzN79myg5N6izMxMNE1j1apVJcZy8uRJRo4cSYMGDXA4HLRr144PPvjAt3/s2LGsXr2aV199FU3T0DSNpKQkAFavXs3FF1+M1WolLi6Of/zjH3g8Ht+x/fr149577+Xhhx+mTp06DBo0iNtuu42//e1vfjF4PB5iY2N59913S4zx1LC75cuX06pVK4KDg7n88stJSUnxazdv3jxatWqFzWajZcuWvPHGG759TZo0AaBTp05omka/fv3YunUrBoOBtLSin8eMjAwMBgPXXXed77gZM2bQvXt33/2/8phL8vTTTxMTE+N7rd544w2aNWvm6/m99tprSzwOzv2anYrj/vvvZ+LEiURGRhIbG1usqOGsWbNo164dQUFBNGzYkLvvvpvc3NwSr5mUlITBYGDTpk1+219//XUaN26MUoqMjAxGjRpFdHQ0drudZs2aMW/ePN/xp/9snq3thZCeNFGjbD15nDy3m7qOknthNE0jzGrlWF4O+7PSaVYLF5OuKqLtkUTbI/HoXrzKg8lgwqgZAx2WqAQFHjfv79zCgax0wqw2Iv78QkUpRaHXw7aTxyn0ehjTujP2c6w7KERlKPCkcTx/Mx7lxGxwYNZM6HiKtntziHF0wm4q3/eTkydP8s033/Dcc88RFFT6exoUJVdXXHEFkZGRfPnll4SFhfHPf/6TAQMGsHv3biIjIwHYt28fS5cu5YsvviAjI4Prr7+e559/nmeffRaAJ598kk8++YQ333yTZs2asWbNGm6++Waio6Pp27dvqbG+9NJLTJ48mSeffBKArVu3MnjwYKZPn84777zDiRMnuPfee7n33nt9H15vueUW0tLSWLVqFWazmYcffpjU1NQLes4KCwvp0qULkyZNIjQ0lGXLljF69GgSEhK45JJLePXVV9m9ezdt27bl6aefBiA6OpojR44wdOhQxo4dy8KFC9m5cyfjx4/HZrP5JQMLFizgrrvuYt26dSilSE9Pp0+fPqSkpBAXFwfAl19+SW5uLtdff32pcebn5/Pyyy+zaNEiDAYDN998MxMmTOD9998HYO7cuUydOpU5c+bQqVMnNm/ezPjx4wkKCmLMmDH8/PPPXHzxxXz33Xe0adMGi8VCREQEUVFRrF69mhEjRrBmzRqioqJYs2aN77qrVq3yvY5/9TGfTinFgw8+yNKlS/nhhx9o1qwZmzZt4v7772fRokX06NGD9PR01q5d+5dfs9PjePjhh/npp5/YsGEDY8eOpWfPnr7E0WAw8NprrxEfH8+BAwe4++67mThxol9ye0p8fDwDBw5k3rx5dO3a1bd93rx5jB07Fk3TmDx5Mtu3b+err76iTp067N27l4KCghIfw/m0PS+qFsrKylKAysrKCnQoopwt3btNPbrmSzVz05qz3iau+VL9eDQ50OEKUSutO5KkJq75Ur3w88oSfz+f/3mlmrjmS7Xh6MFAhyqqkYp6b9d1XR3O3aD2ZH6ukrPXqEM5a3235Ow1ak/m5+pI7gal63q5XvfHH39UgPrkk0/8tkdFRamgoCAVFBSkJk6cqJRSasWKFSo0NFQVFhb6tW3atKn65z//qZRSaurUqcrhcKjs7Gzf/kcffVRdcsklSimlcnNzlc1mU+vXr/c7x7hx49TIkSNLjbNx48Zq+PDhfttGjx6t7rjjDr9ta9euVQaDQRUUFKgdO3YoQG3cuNG3f8+ePQpQr7zyilJKqQMHDihAbd682dcmIyNDAWrlypVKKaVWrlypAJWRkVFqfEOHDlWPPPKI737fvn3VAw884Nfm8ccfVy1atPB7Df/v//5PBQcHK6/X6zuuY8eOxc7funVr9cILL/juDx8+XI0dO7bUeObNm6cAtXfvXr9rxcTE+O43bNhQ/fvf//Y7bvr06ap79+5KqZKfG6WUuuaaa9S9996rlFLqwQcfVI888oiqU6eO2rZtm3K73So4OFh99dVXF/yYAfWf//xH3Xzzzaply5bq0KFDvn1LlixRoaGhfj9n56uk16xXr15+bS666CI1adKkUs/x0UcfqaioKN/9efPmqbCwMN/9Dz/8UEVERPh+Z7Zs2aI0TVMHDhxQSil15ZVXqltvvbXEc5/5/J+t7YWQ4Y6iRjFoGpRSNfAUpRSoP9sKISqVrhSbjh/GqGmYDCX3nJoNRgyaxsZjh0utAipEZXHp2Ti9GZgNjmLzzzRNw2xwUOjNwKVnV8j1z7zmzz//zJYtW2jTpg1OZ1HF4l9++YXc3FyioqIIDg723Q4cOMC+fft8x8bHxxMSEuK7HxcX5+u92r59O4WFhQwaNMjvHAsXLvQ7R0lO7404Fc/8+fP9zjN48GB0XefAgQPs2rULk8lE586dfcckJiYSERHx156kP3m9Xp599lnat2/vey6++eYbkpOTz3rcjh076N69u99z3bNnT3Jzczl8+HCpjxPg9ttv9/UOpqamsmzZMm677bazXs/hcNC0aVPf/dNfhxMnTnDo0CHGjRvn9/w988wz53wd+vXr5xsKunr1avr370+fPn1YvXo1GzdupKCggJ49e17wYwZ46KGH2LBhA2vXrqVBgwa+7YMGDaJx48YkJCQwevRo3n///bMWuCnra9a+fXu/+6c/ZwArV65k0KBB1K9fn5CQEG655RZOnjzpN8T2dMOHD8dkMvHpp58C8O6779K/f3/i4+MBuOuuu1i8eDEdO3Zk4sSJrF+/vtTHcD5tz4cMdxQ1SoPgMAyahkfXMZWymHSBx43VZKJBSFglRyeEKPS4ySgsOOcwRofJTHphPoVejwx5FAHlVS50pWPWSv7IZMCEWxXgVa5yvW5iYiKaprFz506/7QkJCQDY7f+b66zrOnFxcSXO1Tq97LjZ7P+7pGka+p/zs0/9u2zZMurX96+ua7VazxrrmcMxdV3n73//O/fff3+xto0aNWLXrl0lnuf0L2UMf76Hn77N7XafNY6ZM2fyyiuvMHv2bN/8pAcffBCX6+yvjVKqWDJ86rqnby9p2Oktt9zCP/7xDzZs2MCGDRuIj4+nd+/eZ71eSa/Dqeudeh3mzp3rN9wPigpWnE2/fv144IEH2Lt3L3/88Qe9e/dm3759rF69mszMTLp06eJL0i/kMUNRMvbBBx+wfPlyRo0a5dseEhLCr7/+yqpVq/jmm2+YMmUK06ZNY+PGjSWWwC/ra3a2n92DBw8ydOhQ7rzzTqZPn05kZCQ//PAD48aNK/VnxmKxMHr0aObNm8c111zDv//9b998SIAhQ4Zw8OBBli1bxnfffceAAQO45557ePnll4ud63zang9J0kSN0jqyLpE2BxmFBdSxF//WU1eKbJeTNnViiAsKKeUsQoiK4ptDc452RR8WNAxIj7cILKNWVIFWx4OR4pXtdDwYNANGrXyr3kVFRTFo0CDmzJnDfffdV+qHZYDOnTtz7NgxTCaTryfgfLVu3Rqr1UpycvJZ55+VRefOndm2bRuJiYkl7m/ZsiUej4fNmzfTpUsXAPbu3eu33ll0dDRQVGCkU6dOAOcsOb927VqGDRvGzTffDBQlPHv27KFVq1a+NhaLBa/X63dc69atWbJkiV/isn79ekJCQoolrGeKiopi+PDhzJs3jw0bNnDrrbeetf25xMTEUL9+ffbv3++X/JzuVIXFMx9H27ZtiYqK4plnnqFDhw6EhobSt29fZsyYQUZGht/reiGPGeCqq67iyiuv5KabbsJoNHLjjTf69plMJgYOHMjAgQOZOnUq4eHhfP/991xzzTXFzlOW1+xcNm3ahMfjYebMmb7k/qOPPjrncbfffjtt27bljTfewO12F4svOjqasWPHMnbsWHr37s2jjz5aauJ1Pm3LSoY7ihrFajJxVdPW2M1mUgvyKPC4UUqhlCLP7eJ4fi7RjmCGxLcIdKhClDulFMpZgCosvgZdVWEzmogLCiX/HN+I53vc1AsOwXKOb46FqGgWQyhWYwRuPb/EwgluPR+bMQKLIbTcr/3GG2/g8Xjo2rUrH374ITt27GDXrl2899577Ny509ezMnDgQLp3787w4cNZvnw5SUlJrF+/nieffLJYBbvShISEMGHCBB566CEWLFjAvn372Lx5M//3f//HggULzivuSZMmsWHDBu655x62bNnCnj17+Oyzz7jvvvuAoiRt4MCB3HHHHfz8889s3ryZO+64A7vd7ksY7HY73bp14/nnn2f79u2sWbPGV5ikNImJiXz77besX7+eHTt28Pe//51jx475tYmPj+enn34iKSmJtLQ0dF3n7rvv5tChQ9x3333s3LmT//73v0ydOpWHH37Y96H/bG6//XYWLFjAjh07GDNmzHk9VyWZNm0aM2bM8BU62bp1K/PmzWPWrFkA1K1bF7vdztdff83x48fJysoCir4E69OnD++99x79+vUDioYJulwuVqxY4dsGXPBjBrj66qtZtGgRt956q68K6BdffMFrr73Gli1bOHjwIAsXLkTXdVq0KPlzV1les3Np2rQpHo+H119/nf3797No0SLeeuutcx7XqlUrunXrxqRJkxg5cqRf7/SUKVP473//y969e9m2bRtffPFFqYnj+bQ9H5KkiRqnZWQ0N7XsSEJYJAUeD8cL8jien4dH1+kYHcfY1p2p6wgOdJhClBvlduHavI68d18kZ9Y/yHnlH+T963lcv6xBOatWKXtN0+gSUx8NKDyt1PPpCj0eNE3jopgGsl6aCDhN04i0JmLSrDj1LLzKhVI6XuXCqWdh0mxEWBMr5Ge1adOmbN68mYEDB/LYY4/RoUMHunbtyuuvv86ECROYPn26L8Yvv/ySPn36cNttt9G8eXNuvPFGkpKSiImJKfP1pk+fzpQpU5gxYwatWrVi8ODBfP75576S72XVvn17Vq9ezZ49e+jduzedOnVi8uTJvgqIAAsXLiQmJoY+ffpw9dVXM378eEJCQrDZbL427777Lm63m65du/LAAw/wzDPPnPW6kydPpnPnzgwePJh+/foRGxvL8OHD/dpMmDABo9FI69atiY6OJjk5mfr16/Pll1/y888/06FDB+68807GjRt3zqTwlIEDBxIXF8fgwYOpV69e2Z+oUtx+++3861//Yv78+bRr146+ffsyf/583+tgMpl47bXX+Oc//0m9evUYNmyY79j+/fvj9Xp9CZmmab7hl7169fK1u9DHfMq1117LggULGD16NJ988gnh4eF88sknXHrppbRq1Yq33nqLDz74gDZt2pR4fFles3Pp2LEjs2bN4oUXXqBt27a8//77zJgxo0zHjhs3DpfLVWweocVi4bHHHqN9+/b06dMHo9HI4sWLSzzH+bQ9H5qqql+3VqDs7GzCwsLIysoiNLT8v/kSVYP6c6Hck4X5GDSNekEhREtyJmoYVVhAwZJ/4d5VtB6S5ggCNFRBHiiFKbENjmvv+HN71eDRvXy463d+TzuGzWgixGLFoGl4lU6uy0WBx03HuvW4vnm7UouLCHGmin5vL/Ckke7ci9Obga50DJoBmzGCCGtiuZffr40OHz5Mw4YNfXN6qpP8/Hzq1avHu+++W+KQPlF1PfvssyxevJitW7cGOpRiZE6aqLE0TaNRaDiNQsMDHYoQFabwm49x79iMITwKzfq/b6AJCkG5nHh2/07Bl//Gce34wAV5BpPByHXN2xFutfNr6lFOFOT5Zp4FW6x0i2vIgEbNJEETVYrdVId6xihcejZe5cKoWbAYQqW39y/6/vvvyc3NpV27dqSkpDBx4kTi4+Pp06dPoEMrM13XOXbsGDNnziQsLIyrrroq0CGJMsrNzWXHjh28/vrrvh7pqkaSNCGEqKb0jDTcf2xECwr2T9D+pFmsaMFheHb9hvdECsbouBLOEhgWo4krElrSp0ET9maepMDjxm4y0yw8imDL2SvJCREomqZhNUpl4PLgdrt5/PHH2b9/PyEhIfTo0YP333+/WBW/qiw5OZkmTZrQoEED5s+fj8kkH6uri3vvvZcPPviA4cOHn3PJhECR4Y4y3FGIGkH3eMnecxhPgRNbVBhBjerW+G+4XT+vpODz9zDUiUUrZaK3Ugr9xFFsl9+ItedllRxh9ZXtLGR7eipZzkJMBgPxoRE0CYuU9RWrMHlvF0LUJJLyCyGqNd3jJWnJGpI+Xk3OvqMoj47BZqZOlxY0HT2ImB4lT1auCVRh0QKhpSVocKrkvYYqLHlBT+HPq+usSN7Hj8eSyXP/uU6PApPRQIPgMIY3bU1csCQAQgghKpYkaUKIakv3eNkyfREHl6wBg4atThgGswlPgZOU1VtI+2UXHR4fRePhvc59smpIszoAUEpH00rvSStqay9xv/gfpRRfJe3ihyNJWI0mou1Bvp4zp8fDgewMFu3YzK1tukgRIiGEEBVKSvBXEKV0nN4sCj3pePSqVQJbiJoi+b/rOPjJWqxRoYTEx2IOtmO0mrGGBxPatB7K7WXri4vJ2X800KFWCFOzNmh2Byovt9Q2qiAPzWrH1LxdJUZWPR3JzebnY4dxmMyEWW1+QxutJhN17UGcKMhj9eEDAYxSCCFEbSBJWjnTlZcM514O5qzgYM5KknNXcyDnG1LyNlHoyQh0eELUGErXSfp4NQCWsOLl5TVNw9GgDs70HJK/+LGyw6sUhsi6mFt3QeXloFzOYvuV24XKzsTUvB3GuvUDEGH1suVECoUeN0FmS4n7DZpGsNnim6smhBBCVBRJ0sqRrjwcy99Iav5vOL05GDUzRs0KCrJcBzict4489/FAhylEjZB78DjZe45gjSp9fpCmaZgcVlK+31yJkVUu2+DrMLXogJ6VjjftGHpeDnpeDt604+gZaZgS22C7YmSgw6wWjuRmYTYYz1pwxmEyU+B2c6JA5vgJIYSoODInrRxlOPeS4zqMyWDHoP2vhKxBM2FUVtx6Lsfzf6VRyKWYDFJiWogL4S10oXt1DOazr6WlmYx4C4r3MtUUmj0Ixw134t76E65ffkBPOwaAsV4jLJ17YW5/icxHKyMNOFe541P7pcajEEKIiiRJWjnRlYcsVxKaZvRL0E7RNA2zIQi3nkuu+wjh1oQARClEzWGrE4bJZsGTV4jJXvqXHt58J462TSoxssqnWaxYuvTB3Lk3FOSjUGj2oBq/BEF5axQSzt7MkyilSn3u8j0uHGYzdaVwiBAXZNWqVfTv35+MjAzCw8MDHY4QVY4Mdywnhd4MPHo+Rq34grKnFFVf08hzH6u8wISooWzR4cT07YAzPYfSlnv0ujwoXafh37pXcnSBoWkamiMIgyNYErS/oEN0HHaTmZxTpffP4FU6eW43bevEEiILbotKMnbsWIYPH17m9pqmsXTp0gqL56/o168fDz74oN+2Hj16kJKSQliYLA4uREkkSSsnuvIUfft6rkEwmoYXd+UEJUQN1/SmS7FGhZKbdByl6377vC43uUnHCG/bhPqDugQoQlGdxAaF0LNeY5weDxmFBXj//JlSSpHvcXMiP4+4oBD6NqjZPbOidEopMp3ZpBacJNOZXeoXRLWF2/3XP89YLBZiY2PlCyUhSiFJWjkxaTY0zYCO9+wNlY5Zc1ROUELUcJHtm9LlmduwRoWSvfcouUnHyDuSRva+o+QdOkFE+wQufulOzCHyOyfOTdM0BjZuxuVNmmM3mzlZmE9qfi6p+Xk4vR5aRkZzS+vORNpq58+Tys9A37cefecK9L0/oHJOBDqkSpVWmM7647+w7vhGNhz/lXXHN7L++C+kFaZXWgz9+vXj/vvvZ+LEiURGRhIbG8u0adN8++Pj4wG4+uqr0TTNdx/g888/p0uXLthsNhISEnjqqafweDy+/Tt37qRXr17YbDZat27Nd99959crl5SUhKZpfPTRR/Tr1w+bzcZ7773HyZMnGTlyJA0aNMDhcNCuXTs++OAD33nHjh3L6tWrefXVV4t6+zWNpKQkVq1ahaZpZGZm+touWbKENm3aYLVaiY+PZ+bMmX6PPz4+nueee47bbruNkJAQGjVqxNtvv11uz68QVYnMSSsnVmM4NmME+Z6TGI3F56RBUW8bmoEQS4NKjk6ImiuufyfCWjXmyFc/c/T7zXhy83HUi6L+5RcTd2knzEFSNEOUnUHT6NsggYtjG7Iz/QRZzkJMBgPxoRHUDw6tld/6K3ch+ravUMmbwZV/aiuY7Wj12mBofyWatWbP0UsrTOeXE1sp9DoJMjkIMpnw6B7SCtPJPZFHl+h21LFFVkosCxYs4OGHH+ann35iw4YNjB07lp49ezJo0CA2btxI3bp1mTdvHpdffjlGY1FhpeXLl3PzzTfz2muv0bt3b/bt28cdd9wBwNSpU9F1neHDh9OoUSN++ukncnJyeOSRR0q8/qRJk5g5cybz5s3DarVSWFhIly5dmDRpEqGhoSxbtozRo0eTkJDAJZdcwquvvsru3btp27YtTz/9NADR0dEkJSX5nfeXX37h+uuvZ9q0adxwww2sX7+eu+++m6ioKMaOHetrN3PmTKZPn87jjz/Oxx9/zF133UWfPn1o2bJl+T/ZQgSQJGnlRNM0IqzNKPRm4tbzMWl2vzdzXXlx63kEmeoSZIoJYKRC1DyO2Eia3Xo5zW69PNChiBrCbjLTqW69QIcRcMrrQd/0EerIb2BxQEg0mmYoGubnykMlbUQvyMbQfQyaufQ52dWZUopdmfsp9DoJt/wvUbcYzZgNoWS6stmduZ+omIhKSeLbt2/P1KlTAWjWrBlz5sxhxYoVDBo0iOjoaADCw8OJjY31HfPss8/yj3/8gzFjxgCQkJDA9OnTmThxIlOnTuWbb75h3759rFq1ynfcs88+y6BBg4pd/8EHH+Saa67x2zZhwgTf/++77z6+/vpr/vOf/3DJJZcQFhaGxWLB4XD4xXSmWbNmMWDAACZPngxA8+bN2b59Oy+99JJfkjZ06FDuvvtuoChhfOWVV1i1apUkaaLGkSStHIVY6uNVhZwo2IZLz8agmQANpbyggcNUl9igi/8sICKEEEJUberoH6ijW8Ee7peEaZoG1mCU0YJK3YM6uAktsVcAI604Wa4cMl1ZBJkcxZIwTdMIMjnIcGWR5coh3Fr6uo3lpX379n734+LiSE1NPesxv/zyCxs3buTZZ5/1bfN6vRQWFpKfn8+uXbto2LChXxJ18cUXl3iurl27+t33er08//zzfPjhhxw5cgSn04nT6SQoKOi8HteOHTsYNmyY37aePXsye/ZsvF6vr1fw9MevaRqxsbHnfPxCVEeSpJWzcGtT7KY6ZLsOkec+hsKL2RBMqKURwea4PxM3IYQQoupTBzcCqtReMs1kQRkMqKSNqIQeaIaa9yWkS3fj0XWCTCW/f5sMRvI8Oi69coqCmc3+Uyo0TUM/o3DSmXRd56mnnirWAwZgs9nOuuzEmc5MvmbOnMkrr7zC7NmzadeuHUFBQTz44IO4XCVXSS1NSTGUVJjlrzx+IaojyRgqgNUYRrQ9jGh720CHIoQQQvwlSveiMo6A+RzzOi1BqLyT4MwBe80rp24xmDEZDHh0D5YS5px7dC8mgwGLoeT56JXNbDbj9foXMevcuTO7du0iMTGxxGNatmxJcnIyx48fJyamaErGxo0by3S9tWvXMmzYMG6++WagKCHcs2cPrVq18rWxWCzFYjpT69at+eGHH/y2rV+/nubNm/t60YSoTWreV15CCCGEKAenejFqX7GU04VZQgi3hJHnyS/Ws6OUIs+TT4QljDBLSIAi9BcfH8+KFSs4duwYGRkZAEyZMoWFCxcybdo0tm3bxo4dO/jwww958sknARg0aBBNmzZlzJgx/P7776xbt44nnngC4Jw9bImJiXz77besX7+eHTt28Pe//51jx/zXg42Pj+enn34iKSmJtLS0Enu+HnnkEVasWMH06dPZvXs3CxYsYM6cOX7z3YSoTSRJE0IIIUQxmsGEFh4HroKzN3TlozkioIZWeNQ0jRbhCdiMVjJd2bi8bnSl4/K6yXRlYzNaaR6eUGUqf86cOZNvv/2Whg0b0qlTJwAGDx7MF198wbfffstFF11Et27dmDVrFo0bNwbAaDSydOlScnNzueiii7j99tt9CZzNdvaCMJMnT6Zz584MHjyYfv36ERsbW2zx7QkTJmA0GmndujXR0dEkJycXO0/nzp356KOPWLx4MW3btmXKlCk8/fTTfkVDhKhNNFULV2LMzs4mLCyMrKwsQkMrfpKvEEIIUR3pyb+ib/wA7GFoJmux/crrhryTGDoMx9CsdwAi/J+Kfm9PK0xnV+Z+Ml1ZeHQdk8FAhCWM5uEJlVZ+vzKtW7eOXr16sXfvXpo2bRrocISodWROmhBCCCFKpNVvh3boN9TRrShbCFiC0DTtzxL8+VCYjRadiNa467lPVs3VsUUSFRNBlisHl+7GYjATZgmpMj1oF+rTTz8lODiYZs2asXfvXh544AF69uwpCZoQASJJmhBCCCFKpBnNGC66Ef13B+rIH5BzHIVG0WLWNrSGHTF0HI5mqR2LxmuaVill9gMhJyeHiRMncujQIerUqcPAgQOZOXNmoMMSotaS4Y4y3FEIIYQ4J5VzAnVsR1EPmtmGFtMCLSwu0GH5yHu7EKImkZ40IYQQQpyTFhKNFhId6DCEEKJWkOqOQgghhBBCCFGFSJImhBBCCCGEEFWIJGlCCCGEEEIIUYVIkiaEEEIIIYQQVYgkaUIIIYQQQghRhUiSJoQQQgghys2qVavQNI3MzEwA5s+fT3h4eIVfV9M0li5dWuHXEaIySJImhBBCCFEGulKk5OWwPyudlLwc9EpYanbs2LFomsbzzz/vt33p0qVomlbh1y8PN9xwA7t37y63802bNo2OHTsW256SksKQIUPK7TpCBJKskyaEEEIIcQ5J2Rn8cCSJo7nZuHUvZoOResGh9KofT3xoRIVe22az8cILL/D3v/+diIjyuZbL5cJisZTLuc7Fbrdjt9sr/DqxsbEVfg0hKov0pAkhhBBCnEVSdgZL924jKSuDYLOFWEcIwWYLSVl/bs/OqNDrDxw4kNjYWGbMmFFqmyVLltCmTRusVivx8fHMnDnTb398fDzPPPMMY8eOJSwsjPHjx/uGIX7xxRe0aNECh8PBtddeS15eHgsWLCA+Pp6IiAjuu+8+vF6v71zvvfceXbt2JSQkhNjYWG666SZSU1NLje3M4Y7x8fFomlbsdsqkSZNo3rw5DoeDhIQEJk+ejNvt9p3rqaee4rfffvMdN3/+fKD4cMetW7dy6aWXYrfbiYqK4o477iA3N9e3f+zYsQwfPpyXX36ZuLg4oqKiuOeee3zXEiKQJEkTQgghhCiFrhQ/HEki1+WiXlAIdpMZg6ZhN5mpFxRCrsvFuiNJFTr00Wg08txzz/H6669z+PDhYvt/+eUXrr/+em688Ua2bt3KtGnTmDx5si95OeWll16ibdu2/PLLL0yePBmA/Px8XnvtNRYvXszXX3/NqlWruOaaa/jyyy/58ssvWbRoEW+//TYff/yx7zwul4vp06fz22+/sXTpUg4cOMDYsWPL/Hg2btxISkoKKSkpHD58mG7dutG7d2/f/pCQEObPn8/27dt59dVXmTt3Lq+88gpQNHTykUceoU2bNr5z3HDDDcWukZ+fz+WXX05ERAQbN27kP//5D9999x333nuvX7uVK1eyb98+Vq5cyYIFC5g/f36x502IQJDhjkIIIYQQpTien8vR3GwibfZic8A0TSPSZudIbjbH83OJCwqpsDiuvvpqOnbsyNSpU3nnnXf89s2aNYsBAwb4Eq/mzZuzfft2XnrpJb/k6dJLL2XChAm++z/88ANut5s333yTpk2bAnDttdeyaNEijh8/TnBwMK1bt6Z///6sXLnSlwzddtttvnMkJCTw2muvcfHFF5Obm0twcPA5H0t0dLTv/w888AApKSls3LjRt+3JJ5/0/T8+Pp5HHnmEDz/8kIkTJ2K32wkODsZkMp11eOP7779PQUEBCxcuJCgoCIA5c+Zw5ZVX8sILLxATEwNAREQEc+bMwWg00rJlS6644gpWrFjB+PHjz/k4hKhI0pMmhBBCCFGKAo8bt+7Faiz5e22L0Yhb91Lgqfghci+88AILFixg+/btftt37NhBz549/bb17NmTPXv2+A1T7Nq1a7FzOhwOX4IGEBMTQ3x8vF+yFRMT4zeccfPmzQwbNozGjRsTEhJCv379AEhOTj6vx/P222/zzjvv8N///tcvcfv444/p1asXsbGxBAcHM3ny5PM+944dO+jQoYMvQYOi50TXdXbt2uXb1qZNG4xGo+9+XFzcWYduClFZJEkTQgghhCiF3WTGbDDi9HpK3O/yFhURsZvMFR5Lnz59GDx4MI8//rjfdqVUsV4+VcLwy9MTllPMZv+4NU0rcZuu6wDk5eVx2WWXERwczHvvvcfGjRv59NNPgaJhkGW1atUq7rvvPhYuXEiHDh1823/88UduvPFGhgwZwhdffMHmzZt54oknzuvcUPJzcvrjOeVsj1WIQJLhjkIIIYQQpYhxBFMvOJSkrAzqBZn8PuArpUgvLKBJWAQxjnMP8ysPzz//PB07dqR58+a+ba1bt+aHH37wa7d+/XqaN2/u10tUHnbu3ElaWhrPP/88DRs2BGDTpk3ndY69e/cyYsQIHn/8ca655hq/fevWraNx48Y88cQTvm0HDx70a2OxWPx6CEvSunVrFixYQF5eni85XbduHQaDwe+5E6KqCnhP2rRp04pV9zlXCdXVq1fTpUsXbDYbCQkJvPXWW5UUrRBCCCFqE4Om0at+PMEWC0fzcijwuPEqnQKPm6N5OYRYrPSsH4+hktYsa9euHaNGjeL111/3bXvkkUdYsWIF06dPZ/fu3SxYsIA5c+b4zT8rL40aNcJisfD666+zf/9+PvvsM6ZPn17m4wsKCrjyyivp2LEjd9xxB8eOHfPdABITE0lOTmbx4sXs27eP1157zddTd0p8fDwHDhxgy5YtpKWl4XQ6i11n1KhR2Gw2xowZwx9//MHKlSu57777GD16tG8+mhBVWcCTNMCvQk9KSgpbt24tte2BAwcYOnQovXv3ZvPmzTz++OPcf//9LFmypBIjFkIIIURtER8awfDENsSHRZDrdnE8P5dct4smYREMS2xd4euknWn69Ol+wxk7d+7MRx99xOLFi2nbti1Tpkzh6aefPq+Ki2UVHR3N/Pnz+c9//kPr1q15/vnnefnll8t8/PHjx9m5cyfff/899erVIy4uzncDGDZsGA899BD33nsvHTt2ZP369b6CKKeMGDGCyy+/nP79+xMdHc0HH3xQ7DoOh4Ply5eTnp7ORRddxLXXXsuAAQOYM2fOhT0BQlQSTZU0aLkSTZs2jaVLl7Jly5YytZ80aRKfffYZO3bs8G278847+e2339iwYUOZzpGdnU1YWBhZWVmEhob+lbCFEEIIUYVUxnu7rhTH83Mp8Lixm8zEOIIrrQdNCFG7VImetD179lCvXj2aNGnCjTfeyP79+0ttu2HDBi677DK/bYMHD2bTpk2lLj7odDrJzs72uwlxIZTyoKt8dFVY4uRsIYQQNY9B04gLCiEhLJK4oBBJ0IQQFSbgSdoll1zCwoULWb58OXPnzuXYsWP06NGDkydPltj+2LFjxcYSx8TE4PF4SEtLK/GYGTNmEBYW5rudmugqxPnSVR4u7y4KPetwen7E6dmAy/sLHj1FkjUhhBBCCFEuAp6kDRkyhBEjRtCuXTsGDhzIsmXLAFiwYEGpx5RWZra0UquPPfYYWVlZvtuhQ4fKKXpRm3hVBk7vr3j0g4CXouKoBrwqA5d3G259J0pJ2V4hhBBCCHFhqlwJ/qCgINq1a8eePXtK3B8bG+urAHRKamoqJpOJqKioEo+xWq1YrdZyj1XUHkq5cHm3oVQBGkF+XwhomFHKjUc/jEELxqRJT60QQgghhPjrAt6Tdian08mOHTt8VX7O1L17d7799lu/bd988w1du3YttiChEOXFq1JRKh8NR4k9tppmBjQ8+mHpTRNCCCGEEBck4D1pEyZM4Morr6RRo0akpqbyzDPPkJ2dzZgxY4CioYpHjhxh4cKFQFElxzlz5vDwww8zfvx4NmzYwDvvvFNi+VVRMZRSOA8eJuenLbhT09CsFhwtEwnu2h6jwx7o8CqEVz8BaGha6d9raFjQVR462RgJr7TYhBBCCCFEzRLwJO3w4cOMHDmStLQ0oqOj6datGz/++CONGzcGICUlheTkZF/7Jk2a8OWXX/LQQw/xf//3f9SrV4/XXnuNESNGBOoh1Cp6oZPj8/9Dzo+/4s0vQDMYUEona+V6zDHRxN52A0HtWwU6zHKncAPnquJlABQo77mbClECr9NF6obtFBzPwGgxE9kpkZD42ECHJYQQQohKFvB10gJB1kn7a5RSpLyxgOw1P2MMD8UY/L+5WcrjwZWahjEkmAaP3IG9eUKAoy1fTs9veNVxDFpwqW2U8gBurKaLMGjycyXKTilF8tJ17H7nS3KTjqF0BSjMIQ5ierej7SPX44grec6tEKKIvLcLIWqSKjcnTVRdBbv2kfPTFkwRYZhCgv2LZ5hMWOJi8GRkcfLz7wIYZcUwGmIADaW8pbZRONG0UDRCKi8wUSPsW/Qtm59aQN6hVBz1oghNrEdIQj0MZhOHvviRDfe8SsHx9ECHKYQQQohKIkmaKLOc9b+gO10Yghwl7tc0DVNEGPnbduE8erySo6tYRi0agxaKoqBYYRClFEoVAhomQ8NSl4IQoiS5B4+z883PMJiMBDeOwWizAKAZNKyRIYQ0iSVzWxK7//VlgCMVQtR006ZNo2PHjr77Y8eOZfjw4QGLp6pYtWoVmqaRmZkZ6FBELSJJmigz55FjGMymsyYhRocdvcCJ+0TJi5FXV5pmxGJs+2eilo+u8lDKiVKFKPIAMBsTMWox5zhT1ZV7KJWTW/aStesQuqf0HkNRvg5/+RPOjBzssZEl7jeYTVjCgzn89UYKT2ZVcnRCiNPpSpF8Mo/tR7NIPpmHXgkzRg4dOsS4ceOoV68eFouFxo0b88ADD3DyZMW/z7766qvMnz/fd79fv348+OCDFX7d8lCdYhWiJAEvHCKqD81kBP0cb0i6Ag00Q83L/w1aEFZjZ7zqOB79aFHvmaZh0uIwGuIwauGBDvEvSf1xO/ve/460n3bidbrQjEZCmtajyXV9aXxNbwwmY6BDrNFObt6DwWJCM5T+5Yc1MoS8QyfI2nUYW4+wSoxOCHHKrmPZLPs9haS0PJweHavJQHydIK5oH0eL2IqZA7d//366d+9O8+bN+eCDD2jSpAnbtm3j0Ucf5auvvuLHH38kMrLkL3jKQ1hYzf57o5TC6/ViMsnHYVH11LxP0qLCONo0R3k9KL30dcA82TmYwkOxxTeoxMgqj6ZZMBkaYjNdgs3UB5uxNxZjq2qboCV/tp6fHphDyne/YrCasMdGYAkPImtXMlueXsiW6YukV62C6R7vuYfIahpKAWf53RNCVJxdx7J5Z80BdqZkE2oz0zDCQajNzM6Uou27jmVXyHXvueceLBYL33zzDX379qVRo0YMGTKE7777jiNHjvDEE0/42mqaxtKlS/2ODw8P9+sJmzRpEs2bN8fhcJCQkMDkyZNxu92lXv/04Y5jx45l9erVvPrqq2iahqZpHDhwgMTERF5++WW/4/744w8MBgP79u0r9dzvvvsubdq0wWq1EhcXx7333uvbl5yczLBhwwgODiY0NJTrr7+e48f/N43i1LDMRYsWER8fT1hYGDfeeCM5OTmlxpqUlOQbtrh8+XK6du2K1Wpl7dq1OJ1O7r//furWrYvNZqNXr15s3Lix1NiFqAySpIkyC+3WBWNoKO60kgsY6G43el4+Id27YAwpvQpiTXHqD391lbP/KL+/8AFel5uQxHpYI0IwWi2Yg+2ExMdijQzh4CdrSf7vukCHWqOFtWyI1+nmbIV23dl5mINtBDWuvsNphaiudKVY9nsKmQUu4qOCCLKaMBo0gqwm4qOCyCxw8eXvKeU+9DE9PZ3ly5dz9913Y7f7r0EaGxvLqFGj+PDDD8/6t+NMISEhzJ8/n+3bt/Pqq68yd+5cXnnllTId++qrr9K9e3fGjx9PSkoKKSkpNGrUiNtuu4158+b5tX333Xfp3bs3TZs2LfFcb775Jvfccw933HEHW7du5bPPPiMxMREo6t0aPnw46enprF69mm+//ZZ9+/Zxww03+J1j3759LF26lC+++IIvvviC1atX8/zzz5caa8OGDX3HTpw4kRkzZrBjxw7at2/PxIkTWbJkCQsWLODXX38lMTGRwYMHk54uBZtE4EiSJsrMXDeK6BuvAk3DeeQYemEhAErX8WRk4UpJxd6iKVFXXRbgSEVZHFr2I86T2QQ1iC4x2bSEFyXaSUvWnLX3VFyYBpdfjCnIhisjt8T9SlcUnsgipnc7ghvWreTohBCH0/NJSsujboit2N9KTdOoG2LjQFoeh9Pzy/W6e/bsQSlFq1Ylrz3aqlUrMjIyOHHiRJnP+eSTT9KjRw/i4+O58soreeSRR/joo4/KdGxYWBgWiwWHw0FsbCyxsbEYjUZuvfVWdu3axc8//wyA2+3mvffe47bbbiv1XM888wyPPPIIDzzwAM2bN+eiiy7yzR/77rvv+P333/n3v/9Nly5duOSSS1i0aBGrV6/2693SdZ358+fTtm1bevfuzejRo1mxYsVZYz3l6aefZtCgQTRt2hSbzcabb77JSy+9xJAhQ2jdujVz587FbrfzzjvvlPm5FaK8SZImzkv4gJ7E3TUaW5OGeLJycB46iuvoMTBoRFzWh/oT7sAUJiXoq4OUlVsw2a1n7Q20RoWSvfswuUnHKjGy2iWiXQKNhvXEmZ5DYVrWn2ukFfEWusjZn4K9Xh2a3zYkgFEKUXvlOj04PTo2c8nzc21mI06PTq7TU6lxnepBs1gsZT7m448/plevXsTGxhIcHMzkyZNJTk6+oDji4uK44oorePfddwH44osvKCws5LrrriuxfWpqKkePHmXAgAEl7t+xYwcNGzb06/lq3bo14eHh7Nixw7ctPj6ekJD/fd6Ii4sjNTW1TDF37drV9/99+/bhdrvp2bOnb5vZbObiiy/2u54QlU1mSorzFtqtMyFdO1Cwez/ukxloZhP25gmYI8MDHZo4D568QrRSPnScYjAb0b06XmfpcxbEhdE0jfb/GInRYubgpz+QvffInzvAYDIR2qw+HafeQnjr+IDGKURtFWw1YTUZKHR7CbIW/9hU6PZiNRkILmHfhUhMTETTNLZv315iGfydO3cSHR1NeHg4UPS35Myhj6fPN/vxxx+58cYbeeqppxg8eDBhYWEsXryYmTNnXnCst99+O6NHj+aVV15h3rx53HDDDTgcJS/Xc+bQzTMppUr88vDM7Waz2W+/pmnoZRz1ERQU5HfeU8eXJQ4hKoskaeIv0UxGHK2bBToMcQEc9euQfyTtrG08eYUYbRasURVTuUwUMVrMtP/HSBJuupSj32wi/1gGRouJqM7NiOnTHqPFfO6TCCEqRINIB/F1gtiZkk28Jcjvg7tSitScQlrFhdIgsuSk5K+Kiopi0KBBvPHGGzz00EN+yc2xY8d4//33ueeee3zboqOjSUlJ8d3fs2cP+fn/G4K5bt06Gjdu7Fds5ODBg+cVk8ViwestXkxq6NChBAUF8eabb/LVV1+xZs2aUs8REhJCfHw8K1asoH///sX2t27dmuTkZA4dOuTrTdu+fTtZWVmlDv08n1jPlJiYiMVi4YcffuCmm24CipLbTZs2SQl/EVCSpAlRSzX8W3dO/Lgdr8uD0VL8T4FSCmdGDo2u6om9bkQAIqx9ghvF0Pz2KwIdhhDiNAZN44r2cRzLLCTpZNHcNJvZSKHbS2pOIeF2C0Pbx2GogF6XOXPm0KNHDwYPHswzzzzjV4K/efPmTJkyxdf20ksvZc6cOXTr1g1d15k0aZJfb1NiYiLJycksXryYiy66iGXLlvHpp5+eVzzx8fH89NNPJCUlERwcTGRkJAaDAaPRyNixY3nsscdITEyke/fuZz3PtGnTuPPOO6lbty5DhgwhJyeHdevWcd999zFw4EDat2/PqFGjmD17Nh6Ph7vvvpu+ffv6DVP8K7GWJCgoiLvuuotHH32UyMhIGjVqxIsvvkh+fj7jxo07r+dHiPIkc9KEqKXqD+pCeJsm5CYdw+vyH86odJ3cg8exRoTS9KaS5w0IIURt0SI2lHF9mtAyLpTsQjeHMvLJLnTTKq5oe0Wtk9asWTM2btxIQkIC119/PY0bN2bIkCE0b96cdevWERz8v0rKM2fOpGHDhvTp04ebbrqJCRMm+A05HDZsGA899BD33nsvHTt2ZP369UyePPm84pkwYQJGo5HWrVsTHR3tN59t3LhxuFyusxYMOWXMmDHMnj2bN954gzZt2vC3v/2NPXv2AP9bSiAiIoI+ffowcOBAEhIS+PDDD8st1jM9//zzjBgxgtGjR9O5c2f27t3L8uXLiYiQLyhF4GjqfGq31hDZ2dmEhYWRlZVFaKgM4xK1V07SMTZO/CeZfySBBia7Bd3txet0Y4+NpOPk0dQb0DnQYQohxDlVxnu7rhSH0/PJdXoItppoEOmokB60s5k6dSqzZs3im2++OWePVWVat24d/fr14/Dhw8TEyHIhQlwoSdIkSRO1nDuvgJQVmzn05U8UpJzEFGynXv+ONBh6CY56dQIdnhBClEltem+fN28eWVlZ3H///RgMgR0U5XQ6OXToEHfccQdxcXG8//77AY1HiJpCkrQa/odcCCGEqA3kvT0w5s+fz7hx4+jYsSOfffYZ9evXD3RIQtQIkqTJH3IhqjXlTIfCVFA6WMLAHoemyXRbIWobeW8XQtQkUt1RCFEtKWcGpK6HvGTQXUUbNSPY6qKiL0YLjg9ofEIIIYQQf5V83SyEqHaUMwMOfQ45e0AzgSUSrFFgdEBBChz5GpW9N9BhCiGEEEL8JZKkCSGqn9T14EovSs5MdjhVXc1oKdqmu+H4WpTXGdg4hRBCCCH+AknShBDVinKmFw1xNAZBSXPPNA3MoeDOhpx9lR+gEEIIIcQFkiRNCFG9FBwvmoNmtJXeRjMCqqitEEIIIUQ1I0maEKKa0Yv+OecCslpRxUchhBBCiGpGkjQhRPViDivqKTtV0bEkSgEKrOGVFZUQQlQLY8eOZfjw4YEOo9ysWrUKTdPIzMw8r+OmTZtGTEwMmqaxdOnSColNiAsh66TJWipCVCtK6ZD0Hyg4VlQkpKQeNXdu0fYmI9Es8jsuRG1QGe/tuq5ITckhP9+Fw2GhblwIBsO5evX/Gu0cowXGjBnD/Pnzz/u8WVlZKKUIDw//a4GdYezYsWRmZgYs0XG5XKSnp/sSrrLYsWMHrVu35tNPP6Vbt25ERERgtVorOFIhzo+skyaEqFY0zYCqczEcXQ7uzKIiIZqxaKdS4MkD5YLIiyVBE0KUm4P70/lh5T6OHsrC5fJisRip1zCMXv2b0jghstyvl5KS4vv/hx9+yJQpU9i1a5dvm91u92vvdrsxm83nPG9YWFj5BVmOXC4XFovlvI+zWCzExsae1zH79hUVlRo2bFiZE7uSlPU5F+KvkOGOQohqRwtpAnEDwBQErkwoTAPnyaKy/JoGURdB9EWBDlMIUUMc3J/O0g9/58DedIKCrcTWCyEo2MqBvUXbD+5PL/drxsbG+m5hYWFomua7X1hYSHh4OB999BH9+vXDZrPx3nvvcfLkSUaOHEmDBg1wOBy0a9eODz74wO+8Zw53VErx4osvkpCQgN1up0OHDnz88cd+x2zbto0rrriC0NBQQkJC6N27N/v27WPatGksWLCA//73v2iahqZprFq1CoCtW7dy6aWXYrfbiYqK4o477iA3N7dYHDNmzKBevXo0b96cp59+mnbt2hV7Lrp06cKUKVNKfJ7OHO44f/58wsPDWb58Oa1atSI4OJjLL7/cl/ROmzaNK6+8EgCDweCXpM2bN49WrVphs9lo2bIlb7zxhm9fUlISmqYVe87Letwnn3xC//79cTgcdOjQgQ0bNvg9jnXr1tG3b18cDgcREREMHjyYjIyMMr9GouaRnjQhRLWkhTZDBTUsKrNfcLyoSIg1AkIS0SxV85tiIUT1o+uKH1buIyfbSb0Gob4P9XaHgXr2UI4ezuaHlftoGB9RYUMfSzNp0iRmzpzJvHnzsFqtFBYW0qVLFyZNmkRoaCjLli1j9OjRJCQkcMkll5R4jieffJJPPvmEN998k2bNmrFmzRpuvvlmoqOj6du3L0eOHKFPnz7069eP77//ntDQUNatW4fH42HChAns2LGD7Oxs5s2bB0BkZCT5+flcfvnldOvWjY0bN5Kamsrtt9/Ovffe6zdEc8WKFYSGhvLtt9/6hmA+9dRTbNy4kYsuKvqi7ffff2fz5s385z//KfPzkp+fz8svv8yiRYswGAzcfPPNTJgwgffff58JEyYQHx/Prbfe6tdbOXfuXKZOncqcOXPo1KkTmzdvZvz48QQFBTFmzJhSn/OyHvfEE0/w8ssv06xZM5544glGjhzJ3r17MZlMbNmyhQEDBnDbbbfx2muvYTKZWLlyJV6vt0yvkaihVC2UlZWlAJWVlRXoUIQQQghRDirqvT3lcJZ6ceq36v9eWqPe/b8NxW7/99Ia9eLUb1XK4Yr7TDFv3jwVFhbmu3/gwAEFqNmzZ5/z2KFDh6pHHnnEd3/MmDFq2LBhSimlcnNzlc1mU+vXr/c7Zty4cWrkyJFKKaUee+wx1aRJE+VyuUo8/+nnO+Xtt99WERERKjc317dt2bJlymAwqGPHjvmOi4mJUU6n0+/YIUOGqLvuust3/8EHH1T9+vUr9fGtXLlSASojI0MpVfRcAWrv3r2+Nv/3f/+nYmJifPc//fRTdeZH4IYNG6p///vfftumT5+uunfvrpQq/Tkv63H/+te/fPu3bdumALVjxw6llFIjR45UPXv2LPHxleU1EjWT9KQJIYQQQpQiP9+Fy+Ul0mYscb/FaiQj3Ut+/lkqzlaQrl27+t33er08//zzfPjhhxw5cgSn04nT6SQoKKjE47dv305hYSGDBg3y2+5yuejUqRMAW7ZsoXfv3uc192rHjh106NDB77o9e/ZE13V27dpFTEwMAO3atSs2D238+PHcdtttzJo1C6PRyPvvv8/MmTPLfG0Ah8NB06ZNfffj4uJITU0ttf2JEyc4dOgQ48aNY/z48b7tHo+n2By+05/z8zmuffv2fvEApKam0rJlS7Zs2cJ1111XYmxleY1EzSRJmhBCCCFEKRwOCxaLEWehF7uj+FR+l7OoiIjDcf5FLy7UmcnXzJkzeeWVV5g9ezbt2rUjKCiIBx98EJer5ARS14vWkly2bBn169f323eq2uGZBUrKQilVakGO07eXlDxeeeWVWK1WPv30U6xWK06nkxEjRpzX9c9MKDVNQ52lmPmp52Hu3LnFhoUajf7J+ekxn89xp8d06jk4dfzZnuOyvEaiZpIkTQghRI3hys7j6He/cmzN77iz87DHRlJvYBdierXFaJEqbOL81Y0LoV7DMA7sTaeePdQvyVBKkX4ynyaJkdSNCwlglEXWrl3LsGHDuPnmm4GiD/h79uyhVatWJbZv3bo1VquV5OTkUuc2tW/fngULFpRaydBisfjmTp1+3gULFpCXl+dLatatW4fBYKB58+ZnfQwmk4kxY8b45nzdeOONOByOcz72CxETE0P9+vXZv38/o0aNqvDjztS+fXtWrFjBU089VWxfWV4jUTNJkiaEEKJGSPtlN78++Q65B4+DpmEwGfG6PRz6fANRHRPp+sJ4HPXqBDpMUc0YDBq9+jclLTWPo4eziYxyYLEacTm9pJ/MJyTUSq/+TSu9aEhJEhMTWbJkCevXryciIoJZs2Zx7NixUpO0kJAQJkyYwEMPPYSu6/Tq1Yvs7GzWr19PcHAwY8aM4d577+X111/nxhtv5LHHHiMsLIwff/yRiy++mBYtWhAfH8/y5cvZtWsXUVFRhIWFMWrUKKZOncqYMWOYNm0aJ06c4L777mP06NG+oY5nc/vtt/tiXrduXbk+R6WZNm0a999/P6GhoQwZMgSn08mmTZvIyMjg4YcfLvfjTvfYY4/Rrl077r77bu68804sFgsrV67kuuuuo06dOud8jUTNJCX4hRBCVAilFCrjMOrYTlTafpTXXWHXyt57hI2P/pPcg6kEN44htGk9ghvHEJZYH3tsBCc27uTnR9/CnVtQYTGImqtxQiTDb2hPk8RI8nKdHE/JIS/XSZPEou0VsU7aXzF58mQ6d+7M4MGD6devH7GxsX7l9ksyffp0pkyZwowZM2jVqhWDBw/m888/p0mTJgBERUXx/fffk5ubS9++fenSpQtz58719aqNHz+eFi1a0LVrV6Kjo1m3bh0Oh4Ply5eTnp7ORRddxLXXXsuAAQOYM2dOmR5Hs2bN6NGjBy1atCi1KmV5u/322/nXv/7F/PnzadeuHX379mX+/Pm+56G8jztd8+bN+eabb/jtt9+4+OKL6d69O//9738xmYr6Us71GomaSVNnG6RbQ2VnZxMWFkZWVhahobLYrRBClCelFOrIVtTetaiMw6B7QDOgBUWhNbkYrWkvNGP5DuTY8sx77HvvW0IT66EZin//6HW6yTt0gi7PjqPx1b3K9dqiaqiM93ZdV6Sm5JCf78LhsFA3LqRK9KCdj5EjR2I0Gn1rfFVFSilatmzJ3//+9zL3RglR00hPmhBCiHKl9v6AvvED1MkksDggKApsoaj8dPTfv0D/5SOU7im367mycjnyzUYsYUElJmgARqsZDBrJn60vt+uK2sdg0IitH0pCszrE1g+tVgmax+Nh+/btbNiwgTZt2gQ6nFKlpqYya9Ysjhw5wq233hrocIQIGJmTJoQQotyojMPo25cX9ZyF1P3fDoMRTFEoVwHq0GZUnQS0hG7lcs3CE1l48gqxhgeftZ05yEbuwWPlck0hqps//viDHj160L9/f+68885Ah1OqmJgY6tSpw9tvv01ERESgwxEiYCRJE0IIUW70Q7+CqwBCSy4OoFnsKFcuKulnVPzFpfZ8nQ+DxYRmMKCfUWHuTMrrxWit/DLpQlQFHTt2JD8/P9BhnFMtnIUjRIlkuKMQQojyk7ITzNZS10gCwBKMyj4OeSfL5ZKO+nUIaRqH82ROqW2UUrhzC4np277UNkIIIURVIUmaEEKIcqO8btDO8dZiMIDSiwqKlAOD0Uj8NX1AV7hzSu4pKDyegTnUQaO/dS+XawohhBAVSZI0IYQQ5UYLjgKP8+yN3E4wWcBWfhX4Gl/diwZ/60bB8UzyDp3AW+hCeXXcuQXk7E/B6/bS8s6riGgrJauFEEJUfTInTQghRLnRGnVBpe5Fed1oRnOx/UopcOWhNe2JZg0qt+sazCY6T7+VsBYNSVqyhvxDJ9C9XowWMxHtE2h68yAaDLm43K4nhBBCVCRJ0oQQQpQbrX47tAM/otIOoIKi0Ez/K9ShdG/RPDRHBIamPcr92kaLmea3DSHhpgFkbN2PN9+JJSKYiLZNyqVAiRBCCFFZJEkTQghRbjSzDcPFN6Nv/DfqZBJK6WAwg/KCUhAUhbHr9WhhcRUWg8lmIfqilhV2fiGEEKKiSZImhBCiXGlBERh6/x11fCfq8O+QnwkWO1pcK7R67cp1mKMQQghRE0mSJoQQotxpRhNavbZQr22gQxFCCCGqHRmkL4QQQgghhBBViPSkCSGEEDVAXq6TA3tP4nJ6cQRbSGhWB4vFGOiwhBBC/AWSpAkhhBDVmLPQzfdf7+bXnw+Tk1UIgMGgERUdRPe+TejWuwkGgxbgKIUQQpwPSdKEEEKIasrl9LB4/q9s33oMu81EnbpBGI0G3G4vGen5fPHxH2RnFjL4qlZomiRqQghRXcicNCGEEKKa2rghmR1/HCci0k5YhB2jseht3Ww2ElUnCJvdzPpV+0nalx7gSIUQQpwPSdKEEEKIasjr1dm0IRmjUcNqLXlgTHCIBafTw+afD1VydEIIIS6EDHcUopIppcgqcOPVFaF2M2ajfFcihDh/6Wn5pKflERRsKbWNpmnY7Gb27kqrxMiEEEJcKEnShKgkHq/Our1prNyZysGT+SgFITYT/VrWpX/LukQGlf5BSwghzqTrCqUUhnPMNdM00L16JUUlhBCiPEiSJkQlcHt1/rlqHz/sSQOKetA0TSMz38Xin5LZsDeNhy5rTv0IR6BDFUJUE2HhNux2CwUFbqy20t/OnYVeGjWJrMTIhBBCXCgZZyVEJfjv5iOs2X2CcIeZ+hEOQmxmgq0m6obaqB9h5+DJfP7v+7245dtuIUQZ2exmOl5Un4J8N7quSmzjcnkA6HJJo8oMTQghxAWqUknajBkz0DSNBx98sNQ2q1atQtO0YredO3dWXqDVkDsnn0Nf/sTehd9w4D+ryT14PNAh1Rr5Lg8rd6ZiMxsJKmFyv9GgERNm5UBaHr8dyqz8AIUQ1dYlveKpGxdC6rFc3C6vb7tSioICN+lp+SS2jKZV+9gARimEEOJ8VZnhjhs3buTtt9+mffv2ZWq/a9cuQkNDffejo6MrKrRqTfd62TNvOfs/WEHBsaISzEpXWEIdxPbtQLtHb8AWHR7YIGu4HUezOZnrIibUWmobq8mIx6vYkpxJ13gZliSEKJvIOkGMGteVj9/bQsrhLLxeVTQHTSmsVhOtO8Qx4qaOWCzGQIcqhBDiPFSJJC03N5dRo0Yxd+5cnnnmmTIdU7duXcLDwys2sGpOKcW2WR+zZ8FyTFYzwY3qYjCbULrClZlD8n/Xk3foBN3/736skaHnPqH4SwrcXry6wnSOKo5Gg0ZuoaeSoqpaXF432e4clFIEm4Owm2yBDkmIaiOufhh3PdyLPTtPsHtHKoUFHkLDbLRuH0ujJhGyiLUQQlRDVSJJu+eee7jiiisYOHBgmZO0Tp06UVhYSOvWrXnyySfp379/qW2dTidOp9N3Pzs7+4Jjrg7St+xl/+LvsYQFYYv6XxKmGTSskaGYgu2c3LyHff/+ntb3Dg9coDVcsNWMyaDh8uhYTKUnal5dEeYwV2JkgVfodbI7cz8Hc49Q6HWilMJsMFM/KIYW4QmEWeTLAyHKwmQ20qpdLK3aybBGIYSoCQI+J23x4sX8+uuvzJgxo0zt4+LiePvtt1myZAmffPIJLVq0YMCAAaxZs6bUY2bMmEFYWJjv1rBhw/IKv0pL/nwDnvxCrJEhJe43WsyYg2wk/3cd7ryCSo6u9mhVL4S6YTYy8l2ltilweTGbNLo0jqjEyAKrwFPIDykb2Z65B5fuxm604TDbAcX+7GTWpPzMycKMQIcphBBCCFHpAtqTdujQIR544AG++eYbbLayDW9q0aIFLVq08N3v3r07hw4d4uWXX6ZPnz4lHvPYY4/x8MMP++5nZ2fXikTt5K97MDlsZx3qYokIpjAti7zkVMJbNa7E6GoPq8nI4DYxLFx/kOwCN6F2/94yt1cnNcdJ+wZhtKkfFqAoK9/v6TtJK0wn2BKEUfvffBmjyYjVaCXHncumE78zsEEvv/1CCCGEEDVdQJO0X375hdTUVLp06eLb5vV6WbNmDXPmzMHpdGI0nvvDWbdu3XjvvfdK3W+1WrFaSy/aUGOpognkZ6VpoBRKSr9XqMFt40jNcfLttuNkFbgJtprQNMh3Fc1XaxUXwt39EzEaasfckVx3PkfzjmE1WktMwDRNw2FykOXK4Vj+CeoHyRAuIYQQQtQeAU3SBgwYwNatW/223XrrrbRs2ZJJkyaVKUED2Lx5M3FxcRURYrUW1rIRWXsOn7WNOysPS1gQQQ2lOmZFMho0xvSIp0ODcNbsPsG2o1noChLqBNOvZTTdE+sQXEJ5/poqrfAkLt1NiDm41DYmgxFd6ZwoPClJGkWFgI4ezmLLxiMkJ6WDgvqNwunYtT4N46U4hBBCCFGTBPRTYUhICG3btvXbFhQURFRUlG/7Y489xpEjR1i4cCEAs2fPJj4+njZt2uByuXjvvfdYsmQJS5YsqfT4q7pGV3bnyNc/48rOxxLqKLZf93hxZeXTbEQfLGGlf1gW5UPTNDo1jqBT4wi8ukJXCvM5Kj7WVF5V1HN7zsRCA4/uPXubWkD36nyzbCcbVh+goMCNxWwEDZL2pbNpw0G6dm/E0KvbYjpLYRohhBBCVB9V/qv7lJQUkpOTffddLhcTJkzgyJEj2O122rRpw7Jlyxg6dGgAo6yaoru3pv7gi0n+fD2624M1IgTtz+F0nrxC8o6kEdqsPk1HDQxwpLWP0aBhpPb2fNiMVjQ0vLoXo6HkHnOlFAAOKcfP2u/3sfrbvdjtJmLjQnzJrVKK/DwX61cnYbGYuHxY6wBHKoQQQojyoKlTn4RqkezsbMLCwsjKyvJbELsm8hQ42friYg5/+RPu7HxOvdhGq5nwNvF0fvpWQpvWC2iMovbx6F6+ObyGPE8+weagEtsUleTXGVC/F2GWkiuU1gYF+S5mP7uKggIXEZHFe8QBsjILMRo1HnisH6Hh9kqOUIiqoTa9twshar4q35MmLozJbqXT1DEkjhlMyvebcaZlYbRZiL6kFVFdm2Mo47w/IcqTyWCkRXgCv6RtpcBTgM3oX4XU5XXh9DppGtq4VidoADv/OE5WVgF1oktOZgFCQq2cOJ7Ltt+O0b1vk0qMTgghhBAVQZK0WiIkPpaQ24YEOgwhfBJCGuH0utiZuZdsd86fVR41vMqDUTMSH9yQTlFtAh1mwGVnFQJgPMv8RcOfw5hPtRVCCCFE9SZJmhAiIDRNo3VEM+o5YkjOPcKJwnR0pRNuDaVxcH2ibVFSsRAwmYygiuaflfZ8nBq1bjJL4RAhhBCiJpAkTQgRUOHWUMKtMn+kNE0So7BYTRTku3EEWUps43R6MJmNNEmMquTohBBCCFER5GtXIYSowuIahJLQLIqsrEK8JSw6r+uKzIxCGjQKI76pJGlCCCFETSA9aUIIUYVpmsaV17YjM72Ao0eycDjMBAVZQIP8PDd5uS7qxoYw/MYOvrlpQgghhKjeJEkTQogqLio6iLF3XcKaFXv5/dejZGYUogC73UT3vk3ofWlT6tSVBemFEEKImkLWSZO1VIQQ1UhujpMTx3NRSlGnbjChYbLYtxAg7+1CiJpFetKEEKIaCQ6xEhxiDXQYQgghhKhAUjhECCGEEEIIIaoQSdKEEEIIIYQQogqRJE0IIYQQQgghqhBJ0oQQQgghhBCiCpEkTQghhBBCCCGqEEnShBBCCCGEEKIKkSRNCCGEEEIIIaoQSdKEEEIIIYQQogqRJE0IIYQQQgghqhBJ0oQQQgghhBCiCpEkTQghhBBCCCGqEEnShBBCCCGEEKIKkSRNCCGEEEIIIaoQSdKEEEIIIYQQogqRJE0IIYQQQgghqhBJ0oQQQgghhBCiCpEkTQghhBBCCCGqEEnShBBCCCGEEKIKkSRNCCGEEEIIIaoQSdKEEEIIIYQQogqRJE0IIYQQQgghqhBJ0oQQQgghhBCiCpEkTQghhBBCCCGqEEnShBBCCCGEEKIKMQU6ACGEqE08uoej+cfJdGYDEGIOpn5QLBajOcCRCSGEEKKqkCRNCCEqyeG8Y/x2cju57jy/7Y4MO60jmpEQ0ghN0wIUnRBCCCGqCknShBCiEhzNO87PqVvw6B6CzA6MmhEAXenkewrYnLYNgKahjQMZphBCCCGqAJmTJoQQFUxXOlszduLW3QSbg3wJGoBBMxBsDkKh2JaxG5fXHcBIhRBCCFEVSJImhBAV7HhBGlmuHBwme6nDGR0mOwWeQg7npVRydEIIIYSoaiRJE0KICpbjzkUphclQ+ghzg2b4s21eqW2EEEIIUTtIkiaEEBWuqPdMKXXWVsrXUgghhBC1mSRpQghRwcItoRg0Ax7lKbWNV3nRgHBraOUFJoQQQogqSZI0IYSoYNG2SCKsYeR7CkvsTVNKke8uINgcRJwjJgARCiGEEKIqkSRNCCEqmKZpdIhsjd1oJdudi1t3+5I1j+4hx52LyWCiQ2QrzGeZtyaEEEKI2kE+DQghRCWItkfSI6Yrv6VvJ8OZRYEqBEBDI9wSStvIFtQPig1wlEIIIYSoCiRJE0KIShJtj2RAvZ6cKEwn05UNKELMwcTY6/iqOwohhBBCSJImhBCVSNM06tqjqGuPCnQoQgghhKiiJEkTogzST+bz26bD7PzjOE6nh8goBx26NKBVuxgsVvk1mdkzSQAAE/ZJREFUEkIIIYQQ5Uc+XQpxDr//eoTP/rOVnKxCTGYjRqPG8aM57Nh6nEZNIrhhbBcioxyBDlMIIYQQQtQQMglCiLPYvyeNTz/4jYI8NzGxIdSJDiIi0kHd2GAiouwc2HeSD+f/gstZ+vpXQgghhBBCnA9J0oQ4i/Wr9pOf5yYq2oFm0Pz2mc1G6kQHkXwggx1bjwcoQiGEEEIIUdNIkiZEKdLT8ti7K42QUAuappXYxmw2AvDbL4crMzQhhBBCCFGDSZImRClysp243d5zFgYxWwykn8yvpKiEEEIIIURNJ0maEKUwm40YDBq6V521ndersEqFRyGEEEIIUU4kSROiFDH1QqgTHUxOjrPUNkpXeNw6rdrFVmJkQgghhBCiJpMkTYhSGI0GLurZCF1XFBS4i+1XSnHyZD4hYVbad6kfgAiFEEIIIURNVKWStBkzZqBpGg8++OBZ261evZouXbpgs9lISEjgrbfeqpwARa1zcc94Ol7UgJxsJ2mpeTgLPbjdXvJynRw/lovFYuSq69rJOmlCCCGEEKLcVJmJNBs3buTtt9+mffv2Z2134MABhg4dyvjx43nvvfdYt24dd999N9HR0YwYMaKSohW1hclkYMRNHWnYOJyN65NJS81F1xVms5F2HePo0S+BhGZ1Ah2mEEIIIYSoQapEkpabm8uoUaOYO3cuzzzzzFnbvvXWWzRq1IjZs2cD0KpVKzZt2sTLL78sSZqoECaTgR59E7ikVzzHj+bgdnsJCbUSWSco0KEJIYQQ/9/e3QdHVd97HP+czW4SCEmQgEkWwkMVE1I6OECdCPZBRXAUwi2XIGUMLXK9di6OdupYa6cd6IOirTNUmvJHNTGNVgoqYoG2GBw6yCQDSkjFclW4TcNjbohANiHmcX/3jzV7u2Q3hJDdPbt5v2YWZn/n9zv7PV++Q/Ldc/YsgDhki8sd16xZo3vvvVfz5s274tzq6mrNnz8/YGzBggV6//331dXV93NDktTR0SGPxxPwAK5WQoJD7px0TfrCGBo0AAAAhE3Um7Q//OEPqqmp0fr16wc0v6GhQZmZmQFjmZmZ6u7uVlNTU9A169evV3p6uv+Rk5NzzXEDAAAAQDhEtUk7efKkHn30Ub3yyitKTk4e8DrLsgKeG2OCjvd68skn1dzc7H+cPHly8EEDAAAAQBhF9TNphw4dUmNjo2bNmuUf6+np0b59+1RSUqKOjg4lJCQErMnKylJDQ0PAWGNjo5xOpzIyMoK+TlJSkpKSkob+AAAE1dlwTp6q99V29JhMd7cSx2cpbc4sjZw2VZYj6ifwAQAAbC2qTdqdd96pI0eOBIytWrVKeXl5euKJJ/o0aJJ06623aseOHQFjb7/9tmbPni2XyxXWeAFc2YXKfWraulPdnhY5nC7JYemz/z4uz74DSr3lZmX+xzeVMGLgZ84BAACGm6g2aampqZo+fXrAWEpKijIyMvzjTz75pE6fPq2KigpJ0ne+8x2VlJToe9/7nh588EFVV1ertLRUmzdvjnj8AAJ5qg+p8ZVtsmQpaYI74BLknkttat7/nuRwKPu/Voa8PBkAAGC4s/11R2fPntWJEyf8z6dMmaI//elP+utf/6qbb75ZP/vZz7Rx40Zuvw9Emenu0fkdlTKd3XKNy+jThCWkjJRrdJpaDtaq/R8nQuwFccl7Ser8H6n9A6n9Q6nrlGSC340XAABIlum968Yw4vF4lJ6erubmZqWlpUU7HCAutB09ppNPl8g5Ok2O5OCfATXGqOPUGWX82926/puLIxwhIs54pc5PpK4TvqbMkmTk+9saKSVNk5xZUQ4S8YKf7QDiiS2+zBpA7Os+f1Gmqytkgyb57sBqORLU3XQ+gpEhajo/kbr+IckpOUZJvWdXjVcybb4za8lOyTk2qmECAGA3tr/cEUBssBJdkiUZr7f/iV6vrMTEyASF6PG2+s6gySk5kv+/QZMkyyFZKZK6pM5j0vC7oAMAgH7RpAEYEiOmTlFCWqp6mltCzjHd3ZLD0sj8GyMYGaKiu+HzSxxDnFm1LMlKlrzNkvdiREMDAMDuaNIADAnndelKK5ipnpZWeTv73hTCGKPO/22SK3OcRs2eEYUIEVHe1s8/e9bfXTydkumRvG2RigoAgJjAZ9IADJmMf79H7SdOq+3oMSWMHKGEtFRZDod62trUfdEj15jRyuJ70oaJAb4HaPn/AAAAn6NJAzBknOmpmvDYf+rTXe/I8+5BdX96XsYYOZKTlTZntsYsvFMjbpwc7TARCQmjpa6TvpuEWKEatk5JLikhPYKBAQBgfzRpAIZUQuooXb98sTIW3aWO+tMyPT1yXZ+hxMxx0Q4NkeTMkhzHJfOZpJF9L3s0Xsl0SM4JkiMlKiECAGBXNGkAwiIhZaRG5k+NdhiIFitRSsyTOj6UTKukEZLl/PxOjp2+Bs2RJiVSIwAAXI4mDQAQHi63ZCVIncclr8d3Vs1IslyS0y0l5nIWDQCAIGjSAADh48yUEsZJ3guS9zNJlu8zaI5R0Y4MAADbokkDAISX5ZASMqSEaAcCAEBs4HvSAAAAAMBGaNIAAAAAwEZo0gAAAADARmjSAAAAAMBGaNIAAAAAwEZo0gAAAADARmjSAAAAAMBGaNIAAAAAwEZo0gAAAADARmjSAAAAAMBGaNIAAAAAwEac0Q4gGowxkiSPxxPlSAAAwFDo/Zne+zMeAGLZsGzSWlpaJEk5OTlRjgQAAAyllpYWpaenRzsMALgmlhmGbzl5vV6dOXNGqampsiwr6ByPx6OcnBydPHlSaWlpEY7QPsgDOehFHnzIgw958CEPPnbIgzFGLS0tcrvdcjj4NAeA2DYsz6Q5HA5NmDBhQHPT0tKG9Q/eXuSBHPQiDz7kwYc8+JAHn2jngTNoAOIFbzUBAAAAgI3QpAEAAACAjdCkhZCUlKS1a9cqKSkp2qFEFXkgB73Igw958CEPPuTBhzwAwNAaljcOAQAAAAC74kwaAAAAANgITRoAAAAA2AhNGgAAAADYyLBo0vbt26dFixbJ7XbLsixt3749YPu2bdu0YMECjR07VpZlqba2dkD7feONN5Sfn6+kpCTl5+frzTffHPrgh1A48lBeXi7Lsvo82tvbw3MQQ6C/PHR1demJJ57Ql770JaWkpMjtdmvlypU6c+bMFfcbS/UQjhzEWy1I0rp165SXl6eUlBRdd911mjdvng4cOHDF/cZSLUjhyUM81sO/euihh2RZln71q19dcb/xVg//aqB5iMV6AIBoGhZN2qVLlzRjxgyVlJSE3D537lw988wzA95ndXW17rvvPhUXF+tvf/ubiouLtWzZsgH9Ahct4ciD5Pvy0rNnzwY8kpOThyLksOgvD21tbaqpqdGPf/xj1dTUaNu2bfrkk09UWFjY7z5jrR7CkQMpvmpBkm666SaVlJToyJEj2r9/vyZPnqz58+fr3LlzIfcZa7UghScPUvzVQ6/t27frwIEDcrvdV9xnPNZDr6vJgxR79QAAUWWGGUnmzTffDLqtrq7OSDKHDx++4n6WLVtm7r777oCxBQsWmOXLlw9BlOE3VHl46aWXTHp6+pDGFkn95aHXwYMHjSRTX18fck4s18NQ5WA41EJzc7ORZPbs2RNyTizXgjFDl4d4rYdTp06Z8ePHmw8//NBMmjTJbNiwod/9xGs9XG0eYr0eACDShsWZtHCorq7W/PnzA8YWLFigqqqqKEUUPa2trZo0aZImTJighQsX6vDhw9EOaUg1NzfLsiyNHj065Jx4r4eB5ECK71ro7OzUb3/7W6Wnp2vGjBkh58V7LQw0D1L81YPX61VxcbEef/xxffGLXxzQmnish8HkQYq/egCAcKJJG6SGhgZlZmYGjGVmZqqhoSFKEUVHXl6eysvL9cc//lGbN29WcnKy5s6dq2PHjkU7tCHR3t6uH/zgB1qxYoXS0tJCzovnehhoDuK1Fnbu3KlRo0YpOTlZGzZsUGVlpcaOHRtyfrzWwtXmIR7r4dlnn5XT6dQjjzwy4DXxWA+DyUM81gMAhJMz2gHEMsuyAp4bY/qMxbuCggIVFBT4n8+dO1czZ87Ur3/9a23cuDGKkV27rq4uLV++XF6vV5s2bbri/Hish6vJQbzWwu23367a2lo1NTXphRde8H+e6Prrrw+5Jh5r4WrzEG/1cOjQIT3//POqqam56n/LeKqHweYh3uoBAMKNM2mDlJWV1eed0MbGxj7vmA43DodDX/7yl2P+3dGuri4tW7ZMdXV1qqys7PcMkhSf9XC1ObhcvNRCSkqKbrzxRhUUFKi0tFROp1OlpaUh58djLUhXn4fLxXo9vPvuu2psbNTEiRPldDrldDpVX1+vxx57TJMnTw65Lt7qYbB5uFys1wMAhBtN2iDdeuutqqysDBh7++23NWfOnChFZA/GGNXW1io7OzvaoQxab3Ny7Ngx7dmzRxkZGVdcE2/1MJgcXC4eaiEYY4w6OjpCbo+3WgjlSnkINj+W66G4uFgffPCBamtr/Q+3263HH39cu3fvDrku3uphsHm4XKzXAwCE27C43LG1tVXHjx/3P6+rq1Ntba3GjBmjiRMn6vz58zpx4oT/e6A+/vhjSb53QLOysiRJK1eu1Pjx47V+/XpJ0qOPPqqvfvWrevbZZ7V48WK99dZb2rNnj/bv3x/hoxu4cOThJz/5iQoKCjR16lR5PB5t3LhRtbW1+s1vfhPhoxu4/vLgdru1dOlS1dTUaOfOnerp6fG/Cz5mzBglJiZKiv16CEcO4q0WMjIy9NRTT6mwsFDZ2dn69NNPtWnTJp06dUpFRUX+NbFeC1J48hBv9TBx4sQ+b1a4XC5lZWUpNzfXPxbv9TDYPMRiPQBAVEXrtpKRtHfvXiOpz+Nb3/qWMcZ3a+Bg29euXevfx9e+9jX//F6vvfaayc3NNS6Xy+Tl5Zk33ngjcgc1COHIw3e/+10zceJEk5iYaMaNG2fmz59vqqqqIntgV6m/PPR+/UCwx969e/37iPV6CEcO4q0WPvvsM/ONb3zDuN1uk5iYaLKzs01hYaE5ePBgwD5ivRaMCU8e4q0eggl26/l4r4dgBpKHWKwHAIgmyxhjhqLZAwAAAABcOz6TBgAAAAA2QpMGAAAAADZCkwYAAAAANkKTBgAAAAA2QpMGAAAAADZCkwYAAAAANkKTBgAAAAA2QpMGAAAAADZCkwYgItatWyfLstTU1BR0+/Tp0/X1r3/d//yf//ynLMuSZVlat25d0DUPPPCAf04oM2fOlGVZeu6554JuLy8v9+/Dsiw5nU5NmDBBq1at0unTp694XBUVFVq+fLlyc3PlcDg0efLkK64BAADoD00aAFtLTU1VeXm5vF5vwHhra6tee+01paWlhVxbW1urw4cPS5JKS0v7fZ2XXnpJ1dXVqqys1IMPPqjNmzfrK1/5ii5dutTvupdffll///vfdcstt+iGG24Y4FEBAACERpMGwNbuu+8+1dfX65133gkY37Jli3p6elRYWBhy7YsvvihJuvfee/XRRx+pqqoq5Nzp06eroKBAt99+u9auXavvf//7qqur0/bt2/uNb/fu3Tpy5Ihefvll5ebmDvzAAAAAQqBJA2Brubm5mjNnjsrKygLGy8rKtGTJEqWnpwdd197erldffVWzZs3Shg0b/GsGqqCgQJJUX1/f7zyHg/9GAQDA0OK3CwC2t3r1am3fvl0XLlyQJH388ceqqqrS6tWrQ67Ztm2bLly4oAceeEBTp07Vbbfdpi1btqi1tXVAr3n8+HFJ0rhx4679AAAAAK4CTRoA21u2bJmcTqdeffVVSb7Pl02ZMiXgRiOXKy0tVXJyslasWCHJ1+i1trZq69atQef39PSou7tbra2t2rVrl37+858rNTW138spAQAAwoEmDYDtjRo1SkVFRSorK1N3d7cqKiq0atWqkHd1rKur0969e7VkyRKNHj1aklRUVKTU1NSQlzwWFBTI5XIpNTVVCxcuVFZWlv785z8rMzMzXIcFAAAQlDPaAQAYHpxO3383PT09Qbd3d3fL5XKFXL969Wrddttteuqpp3Tu3Dl9+9vfDjm3rKxMxhgtXbpUFy9e9I8XFhbq97//vT766CPl5eUFrKmoqNC0adPkdDqVmZmp7OzsgR8cAADAEKJJAxARvWekTp8+3efslDFGZ8+e1ezZs0Ounzt3rnJzc/XTn/5Ud911l3JycoLO83q9Ki8vlyQtWbIk6JyysjL94he/CBibNm1av68PAAAQKTRpACLijjvukGVZ2rJli2bOnBmw7S9/+Ys8Ho/mzZvX7z5+9KMf6fXXX9eaNWtCztm9e7dOnTqlNWvWaOnSpX22P/zww6qoqNDTTz/tP7sHAABgJ/yGAiAibrjhBj388MP65S9/qYsXL+qee+7RiBEj9N577+mZZ57R7Nmz/Tf5COX+++/X/fff3++c0tJSOZ1O/fCHP5Tb7e6z/aGHHtIjjzyiXbt2afHixdd0TJJ09OhRHT16VJLU0NCgtrY2vf7665Kk/Px85efnX/NrAACA4YUbhwCImOeff16bNm1STU2NVqxYoUWLFul3v/ud1qxZo7179yoxMfGa9t/U1KQdO3Zo4cKFQRs0SSouLtaIESNUWlp6Ta/Va+vWrSoqKlJRUZEOHTqkc+fO+Z+HupMkAABAfyxjjIl2EAAAAAAAH86kAQAAAICN0KQBAAAAgI3QpAEAAACAjdCkAQAAAICN0KQBAAAAgI3QpAEAAACAjdCkAQAAAICN0KQBAAAAgI3QpAEAAACAjdCkAQAAAICN0KQBAAAAgI3QpAEAAACAjfwfPuez4PQGjFwAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/qixin/miniconda3/envs/BIA/lib/python3.11/site-packages/umap/umap_.py:1945: UserWarning: n_jobs value 1 overridden to 1 by setting random_state. Use no seed for parallelism.\n",
      "  warn(f\"n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.\")\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtIAAAJOCAYAAACazuCSAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAACdwElEQVR4nOzdd3xUVfrH8c9NJpkUUiAkECAQepMmiIKroKgoyNpF10Kz19XFtjbUXf1tcUVdV9dCsbsKKoq9gA1RFBtNOhhCCSUhpGfO74+TmTBkJiQhyaR836/XKLn3zswz7d7nnvuccxxjjEFERERERKolLNQBiIiIiIg0RkqkRURERERqQIm0iIiIiEgNKJEWEREREakBJdIiIiIiIjWgRFpEREREpAaUSIuIiIiI1IASaRERERGRGlAiLSIiIiJSA0qkm7jnn3+egQMH0qJFCxzHYdq0aaEOKaj09HTS09ND8tyzZs3CcRxmzZpV5fs4jsPIkSP9lk2cOBHHcdiwYUOtxleftmzZwoUXXkj79u0JCwvDcZxQh3RQI0eODEmcGzZswHEcJk6cWG/PGeh7J7UrlPsiEWlcmlQivWDBgoMe1IJt4z0QO47De++9F/T+hx9+uG+7r7/+Ouh2d911F47j4Ha72blzZ9Dt9n9ex3EICwujZcuWHHvsscyaNYtDmcH9q6++4qKLLiIvL4+rr76au++++6AH4PT0dL94At327NlT45ik4Zs4cSIvvvgixx13HHfeeSd33313le5njGHu3LmceeaZdOjQAbfbTVxcHAMGDOCGG25g+fLldRz5ofGeTB34e0xMTOSYY45h5syZoQ6x3hhjmDdvHueeey6dOnUiOjqa6OhounTpwjnnnMOLL75IcXFxqMNscrzHg61bt4Y6FBGpIleoA2hoXC4XM2bM4OSTT66w7scff2Tp0qW4XC5KSkqCPobH42H27Nk4jkNRURHPP/88119/faXP+6c//YkWLVpQWlrKunXrmDt3Lp9//jnfffcdjz76aI1eyzvvvAPAs88+y1FHHVXl+4WHh3PHHXcEXR8VFVWjeJqDBx54gFtvvZX27duHOpQaKSoq4uOPP+akk07i+eefr/L9du3axTnnnMMnn3xCYmIiJ554Il26dKGoqIhly5bxn//8h0ceeYSPP/64wbemjho1it/97ncAlJSUsHnzZt58800mT57MypUr+dvf/hbiCOvWrl27GD9+PB999BHx8fGMGjWKrl27EhYWxubNm1mwYAGvvfYajz76KIsWLQp1uCIiIaVE+gCnnHIKb775Jjt37iQpKclv3TPPPIPL5eLEE0/k3XffDfoYH374IZs2beLKK6/k2Wef5ZlnnjloIj116lTatm3r+/vnn3/myCOP5LHHHuPGG2+kc+fO1X4tW7ZsAfB73KpwuVwNugSkIUtNTSU1NTXUYdTY1q1b8Xg81frOlJSUcMYZZ/DZZ59x4YUX8thjjxEfH++3TWZmJrfffjvZ2dm1HXKtO+GEE7j11lv9lq1fv55+/frx8MMPM23aNKKjo0MUXd0qKSnh9NNP5/PPP2fixIlMnz6dhIQEv208Hg+vv/46TzzxRIiiFBFpOJpUaUdtmDRpEkVFRbzwwgt+y4uKinjxxRcZO3YsKSkplT7GM888A8BVV13FGWecwc8//8y3335brTj69evHiBEjMMbw3Xff+ZZ/9dVXjB07llatWhEVFUWvXr2YNm0aeXl5vm285SveS9GdO3f2Xaqubd5LkYWFhfz5z3+mY8eOREdHM3jwYD766CMA9u7dy3XXXUf79u2Jiopi2LBhLFmyJOhj7t69m0svvZQ2bdoQHR3N0KFDmTdvXsBtjTHMmDGDo48+mvj4eGJiYhgyZAgzZswIuP2uXbu44ooraNOmDTExMRxxxBG8/vrrlb7Gp59+msMOO4yoqCjS0tK4+eabKSgoCLhtoBpp7+cxbdo0vv/+e0aPHk1cXBwJCQmcccYZQeup586dy5AhQ4iOjqZNmzZceuml7N69u9r1m3l5eUybNo1evXoRFRVFq1atGDt2LF999ZXfdiNHjqRTp04AvisqVamrf+655/jss8849thjmT17doUkGuwJRqArPcuWLWP8+PGkpKTgdrvp3LkzN9xwA7t27Qr4XF988QUjRowgNjaWpKQkxo8fz+bNm4PGVt3vRzCdO3emZ8+eFBYWsnfv3irdZ9OmTUyZMoX27dsTGRlJhw4dmDJlStB49+7dy7333kv//v2JjY0lISGBQYMGceeddx60jMIYw3XXXYfjOEyaNKnSK2aVmT17Np9//jmjRo1ixowZFZJogLCwMM4666wKjQn79zOYP38+xxxzDHFxcX7f1Z07d3LDDTfQuXNn3G43KSkpjB8/PmDZT2V174F+Z/s//+uvv84RRxxBTEwMbdu25corr2T37t3Vfj+qsi+6++67cRyHV199NeBj/Oc//8FxHB566KFqP39lZsyYwWmnnUZ6errvdz169Gg+/fTTCtvWxz6oup9XdnY2f/vb3xgxYgTt2rUjMjKSdu3acfHFF7N27dqAj5OVlcVll11GSkqK3/67sj4uP/30E+eddx6pqalERkbSqVMnrr322kpLLkWqxTQhn376qQHMhAkTqr3NiBEjDGAyMzNN7969zcCBA/3Wv/LKKwYwb775ppkwYYIBzKJFiyo8flZWlomMjDSDBg0yxhjzwQcfGMBcfvnlAePZ/3kPdPLJJxvAvPrqq8YYY1577TXjcrlMTEyMmTRpkrnlllvM4MGDDWCGDRtmCgoKjDHGrF+/3tx9991mwIABBjDXX3+9ufvuu83dd98d9H3x6tSpk3G73Qfd7sD4TzvtNNOlSxdz9dVXm8mTJxu3223cbrf57rvvzJAhQ8xhhx1mrrvuOnP++eebsLAw06pVK5OdnV3huVNTU83hhx9uevfubW666SZz6aWXmri4OOM4jnn++ef9tvd4POYPf/iDAUyPHj3M5Zdfbq699lrTq1cvA5g//elPftvv27fP9OvXz/d+3XrrreaCCy4wERERZuzYsQYwM2fO9LvPvffeawDTpk0bc80115gbbrjBdOzY0Zx66qkGMCNGjPDb3vvdWL9+vW+Z9zs3duxYExMTY8aMGWP+9Kc/meOPP94ApmvXriY/P9/vcZ555hkDmMTERHPZZZeZm266yfTt29cMHjzYtGvXznTq1KlKn09BQYE56qijDGAOP/xwc8stt5hJkyaZmJgY43K5zJw5c3zbzpw501x//fUGMAMGDPB9Zz799NNKn+Poo482gPnggw+qFJPXl19+aWJjY43L5TLnnXeeufXWW83IkSMNYLp3726ysrL8tv/oo49MRESEcbvd5uKLLza33nqrOeKII0xaWprp37+/OXB3Vt3vx8yZMw1gHnjggQqxbty40cTGxpoOHTr4LV+/fn3A/cmvv/5qUlJSDGDGjRtnbr31VjNu3DgDmJSUFLN69Wq/7Xfs2GH69OljADNw4EBz4403mj/+8Y/m5JNPNhEREWb37t2+bQ/83hUWFprzzjvPAOamm2462NteqeHDhxvAfPzxx9W+r/f9GzNmjHG5XOb00083N998s7nyyiuNMXbf2K1bNwOYkSNHmltvvdWcd955xuVymdjYWPPVV1/5PZ533xJIoN+Z9/nHjh1rIiMjzQUXXGBuvfVWM2zYMN93Oi8vr0qvpTr7ok2bNpnw8HBz4oknBnysQYMGmcjIyArf50AqOx4cKCoqyhx55JFmypQp5tZbbzUXXXSRiYuLM2FhYeaNN97w27Y+9kHV/bwWLVpkIiMjzejRo81VV11lbrrpJjNu3DgTHh5uWrVqZTZs2OD3GHv37vX9Rn73u9/5XrPb7fb9tg7cf7/55pvG7XabmJgYc95555mbbrrJt6/v3r272bVr10HfZ5GDUSJdZv8d2N///ncDmO+//963/qSTTjJt2rQxxcXFlSbSDz30kAHMv/71L2OMMaWlpaZDhw4mPj7e7Nu3r8L2wXacP/30k4mOjjaO45j169ebnJwck5iYaNxut/nxxx992+2fLNx3331+jxFo53UwnTp1MuHh4b4k6sDb448/HjD+o48+2uTm5vqWv/zyy76d8DnnnGOKi4t96/72t7/5vUf7Pzdgjj/+eFNUVORbvmLFChMdHW0SExNNTk6Ob/mTTz5pADNlyhS/xy8sLPTtWJcsWeJbfvfddxvAXHrppX7P+/777xugwo549erVxuVymfbt25tt27b5lmdnZ5uePXtWO5EGzMsvv+y3/UUXXWQA89JLL/mW7d6927Ro0cLExcWZtWvX+pYXFxebE044wQBVTqS9JwIXXHCB8Xg8vuU//vijcbvdpmXLln7vabDEMJji4mITERFhXC5XhQNxZUpLS0337t0NYN577z2/dbfddpvvc91/+y5duhjHccznn3/uW77/9//Ag3h1vx/eRGzUqFG+7/vtt99uJkyYYFq2bGlSUlLMRx995Pccwd4vb4Ly3//+12/5f//7X99z7O+cc84xgPnzn/9c4b3aunWrX/z7f+/27t1rTjzxROM4jvnnP/9Z4b7V4f0sIyIifCfl1eF9/xzHMR9++GGF9ZMnTzaAue222/yWv/fee77EprS01Le8pok0UOFzmjRpkgHMvffeW6XXUt190dixY3376v0tXbrUAGb8+PFVet7qJNLr1q2rsGzLli2mXbt2pnv37n7L62MfVN3Pa8+ePWbnzp0Vtv3kk09MWFiYueSSS/yW33HHHQYwV199ddDXtv/+Oysry8THx5sOHTqYjRs3+t3nxRdfNIC55pprAsYrUh1KpMvsvwPbunWrcblcvh/Zpk2bTFhYmK+1p7JEul+/fiY8PNxvR3jLLbcYwMyePbvC9t7n/dOf/mTuvvtuc8cdd5g//OEPJioqygDmuuuuM8YY8+yzzxrA17qzv02bNhmXy2W6du3qt7ymibR3pxToNmDAgIDxL1iwwG95SUmJiYiIMECFndimTZsCfgbe5/7yyy8rxHX11VcbwDz33HO+Zf379zexsbEBE7iffvqpQqtj586dTWRkZMCD1KhRoyrsiO+55x4DmAcffLDC9s8991y1E+ljjz22wuN41914442+ZbNmzTKAueGGGypsv2jRomol0l26dDERERFm8+bNFdZdfvnlFd7T6ibSW7duNYBp27Ztlbb3+uyzzwxgTjnllArrcnNzTVJSkomOjjaFhYXGGGMWLlzoa9090IYNG0x4eHiFg3h1vx/7J2IH3lwul7n++uvNnj17/B4n0Pvl/X736dPH7+TFGJv49+7d2wBm06ZNxhj7HjqOY7p27eqXtAXj/d5t377dDBkyxLhcLvPss88e9H4Hs23btko/y2eeeabCifX+3yvv+3fGGWdUuG9hYaGJjo42SUlJARsURo8ebQC/k6SaJtKBWoYzMjJMREREhX1kMNXdF82bN88A5s477/Tb9qqrrgqY2AdTnUQ6mGuvvdYAfi269bEPqu7nVZl+/fqZ9PR0v2Xp6enG7Xab7du3V9je+/3Zf//9r3/9q8LntL/DDz/ctG7dukrxiFRGnQ0DaNOmDWPGjOHFF1/kn//8J7NmzcLj8TBp0qRK7/fNN9/w888/c/LJJ/t11powYQJ/+9vfmDFjBhdffHHA+z744IOAHSM2Pj6eI444gilTpvi2X7p0KUDAEQ/S0tLo2rUrq1atYu/evcTFxdXkZfu43e6gNcDBDBo0yO/v8PBwUlJS2LdvHx07dvRb5+2Ml5GRUeFxIiIiAo4wcswxx/DYY4/xww8/cOGFF5KXl8fPP/9Mu3bt+L//+78K23trSleuXAnY+tP169fTp0+fgB3pjjnmGD7++GO/ZT/++KNvXaDtq+vwww+vsKxDhw4AfkMKep93+PDhFbYfOnQoLlfVfrY5OTmsW7eO3r17+55nfyNHjuS///2v7z2tT5V9n2NjYxkyZAjvv/8+v/76K4cddliln0WnTp1IS0vzq7+s7vdjf96RVwBKS0vJyMjgueee484772TBggV8++23REREHPS1jRgxokLNqOM4HHvssaxYsYIff/yRtLQ0lixZgjGG4447rtLH3d+2bdv43e9+x2+//cabb77JmDFjqnS/ypiDDLU5Y8YMvvzyS79lJ598coXv1tChQyvcd+XKleTn5zNy5EhiYmIqrB85ciTvv/8+P/zwg2/ElJoK9B1p164dXbt2ZeXKlVXeR1Z1XwQwZswYOnTowMyZM5k2bRphYWEUFBTw4osv0qVLF44//vhDek2BrFu3jgceeIBPPvmEjIwMCgsL/dZv2bLF1+fBq773QQezYMECpk+fzuLFi8nKyvKr7Y+MjPT9Oycnhw0bNtC3b1+Sk5MrPM7w4cN5//33/ZZ5h6f9+uuvWbNmTYX7FBQUkJWVRVZWFq1bt66V1yPNU5NKpMPCbN9Jj8cTdBvvOu+2wUyaNIl58+b5OjIcddRR9O7du9L7eDswXXTRRX7Le/fuzZAhQ1i4cCFr1qyhW7duFe6bmZlZ6UgJOTk5gE3yA2nbti2rVq0iJyfnkBPpmgjUwczlcgXsrOTdCQfqQJWUlBTws/G+bu+oD7t378YYQ0ZGBvfcc0/QuPbt2+d3v2AdRQO9r5XdJ9jnUJnK3ovS0lLfMu9nHeiAERYWVuWdflW+M8AhjaSRlJREREQEO3fupLCwELfbXSexVeXz2z+Rru73I5jw8HA6duzI7bffzqpVq3juued46aWXgp4QQ/VfmzeBqc6QiZmZmeTk5NCjRw+OOOKIKt+vMklJSbhcLrKysgJ+ll988YXv3xMnTmT27NkBHyfQ666P76JXZd+RlStXVnkfWdV9EdjvyZQpU7jnnnt47733GDNmDK+99hp79uzhpptuqvWO3mvWrGHo0KHk5ORw3HHHMW7cOOLj4wkLC2PBggUsXLiwQmIN9b8Pqsyrr77K+PHjadGiBaNHjyY9PZ2YmBhfp8GNGzdWKR4I/L3ydlh+7LHHKo1j3759SqTlkDSpUTu8O4nKeuNmZWX5bRvMqaeeSps2bZg6dSrr1q1j8uTJlW6fl5fHSy+9BMAFF1xQYWIH7ygV1R0twMubqG7bti3geu/yQAltY7Jz586AJ0Le1+f93Lyvc/DgwRhbohTw5u3B7t1++/btAZ830Pvqfa5A9wn2OdQGb6w7duyosM7j8fi+w1V9nLr8zrhcLoYOHUpxcTGfffZZle9X3dgq+ywCPU51vx9V4W1p/f777yvdrrqvLTExEQh8hSaYgQMH8swzz7B69WqOP/74gN+V6nK5XBxxxBGUlJT4Jc3VFShprMl30ZvEBhqBpLKE+2Dfkap+36u6L/K65JJLCA8P5+mnnwbsaD8ul6tOZr186KGH2L17N7Nnz+bDDz9k+vTp3Hvvvb7ReQ5VTfZB1f28pk2bRlRUFN999x2vvvoq//jHP7jnnnt8y6saDwT+Xnnv8/PPP1e6Dziw1V6kuppUIt2zZ08iIyP59ttvgw7/5J1AoH///pU+lsvl4sILLyQjI4OYmBjGjx9f6favvfYaOTk5DBw4kClTpgS8RUREMHv2bL8z/6rylk4sWLCgwrqMjAzWrl1Lly5dQtIaXZuKi4sDzhj5+eefAzaBAIiLi6N3796sWLGiSjMtxsfH07lzZ9asWRNw1jDv4+9vwIABQdcFWlZbvM974PB0YMuHqjq0WXx8PF26dGHNmjUBk7SFCxcC5e9pTU2ZMgWA+++//6DlAd5Wssq+z3l5eSxZsoTo6Gh69uwJVP5ZbNy4scKQctX9flSFt4WrsiteUP5+fvbZZxXeD2NMhe/ykCFDCAsL49NPP63WbIGTJk1ixowZLF++nOOOOy5oAlkd3vK1Bx544JBmVT2Qd+jFb7/91m+oTq9A38WWLVsCFU8wPB6Pr/QgkEDfkS1btrB27Vq6du1a5X1kVfdFXh06dOCUU07h7bff5ssvv+Szzz5jzJgxtGvXrkrPVx3e4eF+//vf+y33eDwVym9qoib7oOp+XmvXrqV37950797db7n3s9pffHw86enprFmzJmAyHSjOI488EkCTBkndq5dK7Hp04YUXGiDgUG8//fSTryfy/sNJGRO4k8fWrVvN66+/7tcBxpjAnQ2PPfZYA1Q6VNgZZ5xhAPPWW29V+ryBZGdnm4SEBBMVFWV++eUX33KPx+PrdX1gj/SadjasyfB3wR4rWKc4AnTUO1hP+YSEBL+e8o8//rgBzNlnn+03YojXunXr/F77XXfdZajmqB3h4eG1NmpHoO9koM5q+/eY379nfk1G7fB2mLzooov8Or79/PPPJioqqsJ7Wt3Oht64jjnmGN/99n88r61bt5pLLrnENyxXaWmp6dq1qwEqjPDg7Z0/efJk37LS0lLTuXPnao3aUd3vR2XD3+3Zs8d07tzZAOZ///ufb3mw9+u4444zgHn66af9lj/99NO+7/j+vKN23H777RWee9u2bUFH7TDGmNmzZ5uwsDDTt29fv++pMeXfvQO/p8EUFRX5hsCbOHFihSEqjbHvuXdftv8+0Pv+HTgEmZd35Iw77rjDb/mHH35oANOtWze/UTvuv/9+A5hp06b5bf+Pf/zD93mHatSOA383Xm+99ZYBTLt27Srs66uiqseDyy67zADmnXfe8Vvufc8OPBbVxz6oup9Xjx49THx8vNm6datvWX5+vvn9738f8Pd8++23G6j6qB3bt283cXFxJjk52e+Y6bVv376AAwaIVFeTqpEG22lv8eLF3HPPPbz99tuMGDGCqKgofv31V+bNm4cxhhdeeMF3ObUybdq04fTTTz/odmvWrOGzzz6jS5cujBgxIuh2kyZN4vXXX+eZZ57h1FNPrcarsmfkTz31FOeffz5HHnkk48ePJzk5mY8//pglS5YwdOhQbrrppmo9ZjAlJSWVTsIxceLEak0IUh2pqans2bOHgQMHMnbsWLKzs3nppZcoKCjgqaee8mtNuvzyy/n666+ZPXs2X375JSeccALt2rVj27ZtrFy5ksWLF/Piiy/6Yr355puZO3cuTz31FMuWLePYY49l8+bN/O9//2Ps2LHMnz/fL5Zu3bpx1113cffdd9O/f3/OPfdcXC4Xc+bMoV+/fqxatapO3oPExET+9a9/cdlll3H44Yczfvx4EhISeOedd3C73bRr1+6gNf5eN998M/Pnz+e5555jxYoVjBo1ih07dvDKK69QXFzMs88+e8hXMVwuF2+88QbnnHMOs2fPZt68eZx00kl07tyZoqIili9fzoIFCyguLvZ1zgoLC2PWrFmMHj2aMWPGcM4559CpUycWL17MJ598QteuXf06CYaFhfHkk08yZswYTjjhBMaPH0+7du345JNPyMzMpH///vz0009+cVX3++H10Ucf+TrbejweMjIymDdvHllZWZx44omcddZZB31PHn/8cX73u99x6aWX8tZbb9GnTx+WL1/OvHnzSE5O5vHHH/fb/j//+Q+//PILf/3rX3nnnXc4/vjjMcbw66+/8sEHH7Bt27ag+6yLL74Yx3GYOHEiI0eO5NNPP/XVjHpbz6vaOSwiIoI333yTc889l1mzZjF37lxGjRpFt27dcByHrVu3snDhQjZu3EiXLl2q1dr6t7/9jYULF/KXv/yFr776iiOPPJINGzbw2muvERMTw8yZM/2+15MmTeLvf/8706ZN44cffqBr164sWbKEX375hREjRvhasQ80duxY33cqLS2NhQsXsmjRIgYMGMDUqVOrHG919kVeY8aMIS0tjc2bN9O+fXtOOeWUKj/f/q6//vqgs2f+5z//4YorrmDmzJmceeaZjB8/nqSkJL7++mu+//77gPuy6qrJPqi6n9e1117Ltddey6BBgzj77LMpKSnhww8/xBjDgAEDKrRi33LLLcyZM4fHHnuMn376ydfZ9n//+x/jxo3jrbfe8ospOTmZl156iXPOOYcBAwZw8skn06tXLwoKCti4cSMLFy5k+PDhvPfee4f0Xok0uRZpY2zrkXdCktjYWBMREWHS0tLMH/7wB7+xofdXnWGHDmyRvvXWWw1UHMf5QMXFxaZNmzbG5XL5zsKrO9zRZ599Zk455RSTmJhoIiMjTY8ePcydd94ZsMWtLoa/44CWjtpuke7UqZPZuXOnueSSS0xKSopxu91myJAh5s033wwa8yuvvGJOOOEE07JlSxMREWHat29vRo4caR588EGzY8cOv2137txpLrvsMpOcnGyioqLM4MGDzdy5cyttTXvqqadMnz59TGRkpOnQoYOZOnWqycvLq7MWaa9XX33VDBo0yLjdbpOSkmIuueQSs3PnTtOiRYsKwxBWJjc319x5552mR48eJjIy0iQmJppTTjmlwpWWg8VzMB6Px7z22mvm9NNPN+3atTORkZEmJibGNxnP8uXLK9znp59+MmeffbZp3bq1iYiIMJ06dTLXXXddhc/N67PPPjPHHnusiY6ONq1atTLnnHOO2bhxY6Xfw6p+P4INf9eiRQszdOhQ89BDD1UYnq6y92vDhg1m0qRJJjU11bhcLpOammomTZpUYaIJr+zsbHPnnXeaXr16GbfbbRISEszAgQPNXXfd5fe8gb53xtghGcPDw03v3r19+5OHH37YAOapp54K+JzBeDweM3fuXHPmmWeatLQ043a7TVRUlElPTzdnnnmmeeGFF3xDEx74/gVrkTbGTjxz3XXXmU6dOpmIiAjTunVrc/bZZ5uff/454Pbff/+9GTVqlImJiTHx8fHmtNNOM6tXr650+LuZM2eauXPnmsGDB5uoqCiTkpJiLr/88oBjFgdT032RMeXjoB/Y8l4V3u9xZTfv1dRPP/3UHH300SYuLs4kJiaaMWPGmO+++843Xv6htEh7VXcfVJ3Py+PxmCeeeML07dvXREVFmbZt25opU6aYbdu2Bf09b9++3UyZMsW0bt3ab//9z3/+0wDm9ddfr3CflStXmilTpphOnTqZyMhI07JlS9OvXz9z3XXXmW+++eYgn4jIwTnG1GIhnIjUqTVr1tC9e3fOPfdcXnnllVCHIw3c2Wefzddff826dev8hhNrimbNmsWkSZOYOXNmnXTwq6oxY8bw3nvvsW7dujq7chdKDXEfdOGFF/LCCy+wfPnyg46uJVLbmlRnQ5GmYvfu3RWGr8rPz+eGG24AqFLJkcgXX3zB1KlTm3wS3VAsW7aM9957j5NPPrnRJ9ENcR+UmZlZYdnChQt5+eWX6dmzp5JoCYkmVyMt0hQsXLiQKVOmcNJJJ9GxY0eysrL45JNP2LBhA8cff/xBR5ERAQKOUCO178UXX2TVqlU8++yzANx5550hjujQNcR90JgxY4iOjmbgwIHExsayfPly3nvvPcLDw3n00UfrPR4RUCIt0iD17duXE088kS+//JI33ngDsJ0f77vvPqZOnVrlzoYiUveefPJJPv/8czp16sQzzzzDsGHDQh3SIWuI+6AJEybwwgsv8PLLL7N3714SExMZN24ct912m2+4O5H6phppEREREZEaULOWiIiIiEgNKJEWEREREamBZlkj7fF42LJlC3FxcTiOE+pwRERE5BAZY9i7d2+1Jq0SOVTNMpHesmULaWlpoQ5DREREatnmzZvp0KFDqMOQZqJZJtLeqV03b95MfHx8iKMRERGRQ5WTk0NaWlrA6dtF6kqzTKS95Rzx8fFKpEVERJoQlWxKfVIRkYiIiIhIDSiRFhERERGpASXSIiIiIiI1oERaRERERKQGlEiLiIiIiNRAyBPp9PR0HMepcLv66qsDbr9gwYKA269cubKeIxcRERGR5izkw999++23lJaW+v7+5ZdfOPHEEznnnHMqvd+qVav8hq5LTk6usxhFRERERA4U8kT6wAT4//7v/+jatSsjRoyo9H4pKSkkJibWYWQiIiIiIsGFvLRjf0VFRTz//PNMnjz5oAOqDxo0iNTUVEaNGsWnn35a6baFhYXk5OT43UREREREDkWDSqTfeOMN9uzZw8SJE4Nuk5qaypNPPsmcOXOYO3cuPXv2ZNSoUXz22WdB7/PAAw+QkJDgu6WlpdVB9CIiIiLSnDjGGBPqILxGjx5NZGQkb731VrXuN27cOBzHYd68eQHXFxYWUlhY6Ps7JyeHtLQ0srOzNUW4iFRDIbCv7N9uIAbQdMQiDUFOTg4JCQk6tku9CnmNtNfGjRv56KOPmDt3brXve9RRR/H8888HXe92u3G73YcSnog0awXAFmA3UFK2LBxoAbQH4kIUl4iIhFKDSaRnzpxJSkoKY8eOrfZ9ly5dSmpqah1EJSJSAKwG8oAIILpseSmQXba8K5AQkuhERCR0GkQi7fF4mDlzJhMmTMDl8g/ptttuIyMjg2effRaA6dOnk56eTt++fX2dE+fMmcOcOXNCEbqINHm/YZPlaPzLOFzYVukCYCPQt+xvERFpLhpEIv3RRx+xadMmJk+eXGFdZmYmmzZt8v1dVFTE1KlTycjIIDo6mr59+zJ//nzGjBlTnyGLSLOQj211jiBwLbSDrZUuAPYASfUWmYiIhF6D6mxYX9QhQUSqZgewjoqt0QfKA9oCneojKBEJQMd2CYUGNfydiEjDYrAJdFVG5vDUcSwiItLQKJEWEQnKjU2iSyvZxntRL6ruwxERkQZFibSISFBx2AS5qJJtirHdTVrWS0QiItJwKJEWEQkqDGhX9v8CylufKft3Mba1OgW1SIuIND8NYtQOEZGGKwlb/7wZO4qHt17aYHehbbGTsoiISHOjRFpE5KCSsROu7AJysUl0NNAKO024iIg0R0qkRUSqJBLb+iwiImKpRlpEREREpAaUSIuIiIiI1IASaRERERGRGlAiLSIiIiJSA0qkRURERERqQIm0iIiIiEgNKJEWEREREakBJdIiIiIiIjWgRFpEREREpAaUSIuIiIiI1IASaRERERGRGlAiLSIiIiJSA0qkRURERERqQIm0iIiIiEgNKJEWEREREakBJdIiIiIiIjWgRFpEREREpAaUSIuIiIiI1IASaRERERGRGlAiLSIiIiJSA0qkRURERERqQIm0iIiIiEgNKJEWEREREakBJdIiIiIiIjWgRFpEREREpAaUSIuIiIiI1IASaRERERGRGlAiLSIiIiJSA0qkRURERERqQIm0iIiIiEgNKJEWEREREakBJdIiIiIiIjWgRFpEREREpAaUSIuIiIiI1IASaRERERGRGlAiLSIiIiJSA0qkRURERERqQIm0iIiIiEgNKJEWEREREakBV6gDEBFpzPZl7CDj/SXkrt8KYQ6JfTrR/qQhuFvGhTo0ERGpY0qkRURqwHg8rPzv26x59gOK9uTiOGCMwcFh1RNv0feGs+n4++GhDlNEROqQEmmRepUPFJT9OxqICmEscihWPTWflf95E1eMm/iuqThhtlLOU1JK3pad/HDvs4RHRdL+pCEhjlREROqKaqRF6kUesAZYDvxadltWtiw/hHFJTeRv383a5z4kPCqS6DYtfUk0QJgrnBYdUygtKGLl4/PwFJeEMFIREalLSqRF6lwuNnHeCTjYlujosn/vLFuXF7LopPq2fLCEgp05RKckBt0mOrUVOWsz2LF4Rf0FJiIi9UqJtEidMsBGoBCbPEdgE2in7N/R2FKPjWXbSmOwb/MOAJzw4LtQV7QbU+rxbSv1zQPkALvL/u8JbTgi0iSpRlqkTuVgW5vd2OT5QA4QiW21zgU00kNj4LjCwVR+4mOMAY+pNNmWumCA7WW3AmwCHYb9DaaU3fSZiEjt0N5EpE7lYg/k4ZVsE162TW69RCSHLrFvOk54GKVFxUG3Kc7JwxUbRcvD0usvsGbPAJuxV3gKKL/qE4m9KrQJXf0RkdoU8kQ6PT0dx3Eq3K6++uqg91m4cCGDBw8mKiqKLl268MQTT9RjxCLVYQjcEr0/b6mHDu6NRepxA2nRqQ37fsuyLc8HMKUe8rftJmlwDxJ6dwpBhM1VNrANe7E1CnuS6mAPdVHYxDoL2BWqAEWkiQl5Iv3tt9+SmZnpu3344YcAnHPOOQG3X79+PWPGjOGYY45h6dKl/PnPf+a6665jzpw59Rm2SBW5y/5fWZLsrd2MrONYpLa4ot30v/UPRCbEsnfNFor35mOMwRhD4e695KzJoEWnNhw29Vwc52AnUlJ7srC/p4gg613Y32IWOnEVkdrgmEDNKSH0xz/+kbfffpvVq1cHPADdcsstzJs3jxUrynvCX3HFFfz4448sWrSoSs+Rk5NDQkIC2dnZxMfH11rsIhUVA78ApZQn1QcqxB7gD0PdFhqXHYtXsPyxN9jzy3pK84sAcLWIJmVYH/pcdybx3dqHOMLmxAP8gE2QKzspLS7bpj/BE25pjHRsl1BoUEftoqIinn/+eW688cagrTiLFi3ipJNO8ls2evRonnnmGYqLi4mI0I6x8SrAXnItxF6OjQVa0sC+ptUUAbTF1m0WUT5qB9iDufegnkrjfp3NU/KRvTl2aC92/biW3I3bcMIcEnp1JKF7h1CH1gx524SqUkplUIu0iNSGBnXkfuONN9izZw8TJ04Mus3WrVtp06aN37I2bdpQUlJCVlYWqampFe5TWFhIYWGh7++cnJxai1lqgwfIwPay33/yCqdseSdsQt1YtcW+xq3YyVf2T6RdQHvsSALSGDmOQ9LAbiQN7BbqUJq5MOyJqreTYTClZesb1OFPRBqpkNdI7++ZZ57hlFNOoV27dpVud2Brtbc6JVgr9gMPPEBCQoLvlpaWVjsBSy3JADLL/h0NxJTd3NhW3HXYTkSNlYNNlvsAHbAnBS2BNKAv0I6Dt6KJSOUcIBl70hpszGiDTaRb08AOfyLSSDWYPcnGjRv56KOPuOSSSyrdrm3btmzdutVv2fbt23G5XCQlJQW8z2233UZ2drbvtnnz5lqLWw5VPrYlOhxb17h/QuntaV8CbKHxX4qNxibU3ctu7bCvT0RqRxK2JMw7fvT+PNj9TTQ2kRYROXQN5trWzJkzSUlJYezYsZVuN2zYMN566y2/ZR988AFDhgwJWh/tdrtxu4N19JLQ2o1NlKODrNeEJSJSVRFAV+xVrH3Yk+8wbBLt7XfRheAdf0VEqqdBtEh7PB5mzpzJhAkTcLn8c/vbbruNiy++2Pf3FVdcwcaNG7nxxhtZsWIFM2bM4JlnnmHq1Kn1HbbUCm/NcGWlDeHYA2JhJduIiIA9Ke+NveqTDMRjW6C7li2PCV1oItLkNIgW6Y8++ohNmzYxefLkCusyMzPZtGmT7+/OnTvzzjvvcMMNN/DYY4/Rrl07HnnkEc4666z6DFlqTVUnIqnKxCYijVUesAc7iksY9spLPA2kraMRCqO8L4KISN1pcONI1weNNdmQbAc2YGuFgyXK3iHi+qKaYmlaSrFTVu8q+7dXGLbltDNqQRWpGh3bJRTU3CEh1hJbA11I4JZp71jLCSiJlqbFA6wHdmB3xfuPWBOB7ROwBttxTkREGiIl0hJiEdhh4MKwCYO3Vc6bQHt72WuCC2lqcrCdbSPxn6gHbL+AaOz3f1v9hyYiIlXSIGqkpblLwibSW7C1okWUT1aShE2i1RotTc1ObKt0sN2wU7ZuF3aoRM3aKiLS0CiRlgaiJZAI7MV/ivBgw+KJNHZ52JbnyoRjr8wUokRaRKThUSItDYiDHalAGo98bHlCAfaqQgvsSdHBEkSpHo1YIyLSECmRFpEa8AC/YTvKlVA+jOF27GQXnbBXGCS4OGyrdGVDO5Zga6hV2iQi0hCps6GI1MBmYGvZv6MpH3HCjS1DWIftTCfBJWFb7kuCrPdgO9+2Ri38IiINkxJpEammPGxLtAvbWrp/a2oYtvW0GNt5tNkNU18NLYA22ES6EJs4Q/mINQXYVus2IYlOREQOTqUdIlJN3slDIoOsd8rW5WKT7th6iquxcbAj0kRgh7grLFvuHbEmGTs0pHbTIiINlfbQIlJN+dgksLIOcOHYYQwLUSJdGQdoi02asymfIjweWyYjIiINmRJpEakmjSBR+8KBVqEOQkREqkk10iJSTS2w5QeV1T+XYM/T1RotIiJNlxJpEammVtga6EICJ9PeznItUXmCiIg0ZUqkRaSaIrGd4BzsyBKlZcu9CXQ+dii89iGJTkREpL6oRlpEaqA19jx8CzZxLqJ8tIkkbKKt1mgREWnalEiLSA21ws5emIst83Cw9dOahU9ERJoHJdIicgi8Q7WJiIg0P6qRFhERERGpASXSIiIiIiI1oERaRERERKQGlEiLiIiIiNSAEmkRERERkRpQIi0iIiIiUgNKpEVEREREakCJtIiIiIhIDSiRFhERERGpASXSIiIiIiI1oERaRERERKQGlEiLiIiIiNSAEmkRERERkRpQIi0iIiIiUgNKpEVEREREakCJtIiIiIhIDSiRFhERERGpASXSIiIiIiI1oERaRERERKQGlEiLiIiIiNSAK9QBSMNSuCuHLR8vJT9zJ2ERLloN7Errob0ICw8PdWgiIiIiDYoSaQHAU1rK6hnvsfb5DynYkW0XGkOYO4KEXh0ZeOdFtOybHtIYRURERBoSlXYIACv/8ybLH55Dyb4CWqS3Ib5bO+K7tycqKZ7dP65l8fX/JvvXzaEOU0RERKTBUCIt5KzdwtrnP8LVIpqYdkmEucrLOFyxUcR1TWXf5u2semp+CKMUERERaViUSAu/vbuYoux9RCUnBFzvhIURlZzAts9+InfTtnqOTkRERKRhUiIt7FmxibAIF47jBN0mMqEFxXvz2Lt2Sz1GJiIiItJwKZGWMqYK2zhQSbItIiIi0pwokRZaHdYZT3EpxhM8mS7cvZeI+Bjiu7evx8hEREREGi4l0kL7U4YSmdiCgu27A643pR4Kd+6l7YgBxLZPrufoRERERBomJdJCXHpbekw+hdKCYvb9tgNPcQkAxhiKcvLIWbuFFp3b0uuKcSGOtKkwgKfsJiIiIo2VJmQRAHpcOhZXjJtfZ75H7qbtNtczhvDoSFKG9aH/ny8grnNqqMNs5EqAXcAOoLBsWTzQGkgAVH8uIiLSmDjGmKr0MmtScnJySEhIIDs7m/j4+FCH06AU781j68IfycvcRVhEOK0GdqPVgK6VjughVVEIrAVyy/52Yc9WSrEJdDLQEV0kEhGpGR3bJRTUIi1+IuJiSDt1WKjDaGIMsB7YC0RRMVkuBrYDbkCt/iIiIo2Fmr9E6tzespubwD+5iLLl27Et1CIiItIYqEVapM7txnYsDK9kmwhs+cdeILEeYhJpbAyQB+zE/k4AYoAkIA71MRCRUFAiLVLnSjj4QT4MmyiU1H04Io2OAX4DtmGv2oRhf1O5QBa2w24ndJFVROqbEmmROhfOwWeO9GATg8parUWaq21AJvb3EU35iam3w+4ObBLdKSTRiUjzpdN3kTqXiP2pVVb/XAxEYi9Ri0i5EmAr9jcUif/VHQfbHhSObZkurHBvEZG61CAS6YyMDC688EKSkpKIiYlh4MCBfPfdd0G3X7BgAY7jVLitXLmyHqMWqap4bC1nIYFbpkuwSXZrdJFI5EDZQBG2H0EwEdjfUeDZWUVE6krIj9q7d+/m6KOP5rjjjuPdd98lJSWFtWvXkpiYeND7rlq1ym+syORkTV8tDVEY0AVYg+0sFU55uYe3JjoJaBeS6EQatuKy/1fW7uNtpS6q41hERPyFPJH+29/+RlpaGjNnzvQtS09Pr9J9U1JSqpRwi4ReNNATe/l5B+UJdBy2JTqJBnKBSKSBcSibapWqddoVEak/Id/rzJs3jyFDhnDOOeeQkpLCoEGDeOqpp6p030GDBpGamsqoUaP49NNP6zhSkUMViW117gccVvb/XthZDUP+UxRpoOKwV3AqG9HGO5KH+hiISP0K+dF73bp1PP7443Tv3p3333+fK664guuuu45nn3026H1SU1N58sknmTNnDnPnzqVnz56MGjWKzz77LOD2hYWF5OTk+N1EQicMOznLgR2nRKSiGCABW+LhCbDeYPsfxGD7I4iI1B/HGHOwcbnqVGRkJEOGDOGrr77yLbvuuuv49ttvWbRoUZUfZ9y4cTiOw7x58yqsmzZtGvfcc0+F5dnZ2X411iIi0hAVAquBfdjWaW/Hw5KyWxTQHZtMS3OVk5NDQkKCju1Sr0LeIp2amkqfPn38lvXu3ZtNmzZV63GOOuooVq9eHXDdbbfdRnZ2tu+2efPmGscrIiL1zQ30AFKxiXRR2c0BUsrWKYkWkfoX8s6GRx99NKtWrfJb9uuvv9KpU/UG1l+6dCmpqakB17ndbtxud41jFBGpOQ+QAxSU/R2DprSuiUigI7afQX7ZMm+JlIhIaIQ8kb7hhhsYPnw4999/P+eeey7ffPMNTz75JE8++aRvm9tuu42MjAxf3fT06dNJT0+nb9++FBUV8fzzzzNnzhzmzJkTqpchIhLAbuzU1vn7LXOwyXRH1DmuJlzofRORhiLkifQRRxzB66+/zm233ca9995L586dmT59OhdccIFvm8zMTL9Sj6KiIqZOnUpGRgbR0dH07duX+fPnM2bMmFC8BBHxY4C92HpWD7bVMJEGsLupZ7uA9dgRJSIpn/69FMjFjiveDSWFIiKNV8g7G4aCOiSI1JV9wEbKk2hv+UIk0BZoQ/MoaSgFfsF2koui4ms22FbqFkCfAOtFpLp0bJdQaG5NRCJSZ/KwIysUYluhw7AJogc7dNnmsn83hxkcd1P+PgRKkp2ydXnY1nsd9EVEGqOQj9ohIk3FFmzyGI0tY/AmkN5xs8OATMo73TVl+dhW58p2sWHYE4v8SrYREZGGTIm0iNSCfCAbO75vsDKFCOyYv7vqK6gGzqF8+msREWmMlEiLSC0owCbJlVWLeRPHffUSUWh566IrS5JLy7aJqpeIRESk9imRFhGpdS2xHSwLg6w32AlFolF9tIhI46VEWkRqQRS2Nbqkkm1M2a05zEDnorxTZQG2FtrLU7YsHOiAdsMiIo2XRu0QkVoQDSQAO7G7lUB10sVl65LqMa5QSi77v7cTJtgTCW85Rxp2fG0REWmslEiLSC1ph51oJB87Sod3AhJvGYMHaE/zqQl2gBSgFXY4vP2nCE+k/P0REZHGSom0iNSSGKA7sAE7PnIR5S2wEdgyhjahCi6EXJS3TouISFOiRFpEalEsdqa+vdjWaYPtdNcS7W5ERKSp0ZFNRGqZgx2JQqNRiIhI06bu4iIiIiIiNaBEWkRERESkBlTaISLSjBXvy2ffpu0YjyE2LZnI+NhQhyQi0mgokRYRaYYKsrJZ++LHbJ73FQU7cwCITGxB2qlH0fUPo4hJbS7jfYuI1JwSaRGRZiZv6y6+vu5Rdv+0jogW0UQlxYMDRXv28euT89n22U8c9ci1tOjUHIcrFBGpOtVIi4g0Mz898CK7f1pHi/S2xLRLwhXjxhXtJia1FXFdU8letZkf7n0WY0yoQxURadCUSIuINCPZq39j+1fLiEpOJDyy4kXJMFc4Mamt2Ll0Nbt+XBuCCEVEGg8l0iIizUjWNyspyc0nMjF4p0JXi2hK84vYsXhFPUYmItL4KJEWEWlGSvMLwXFwHCfoNo7jgONQml9Uj5GJiDQ+SqRFRJoRd+sEADwlpUG3MR4PGENUa81OKSJSGSXSIiLNSNtj+xOVnEDBjuyg2xTuzCEisQWpow6vx8hERBofJdIiIs2Iu1U8ncePpCSvgMLdeyuMzFGUvY+iPfvodPrRGku6yfEAhUARoBFZRGqDxpEWEWlmel42jqLduWyY8xkFO7KJiI0CB0r2FRLmjqDTmcfQ5/qzQh2m1JpiYAeQVfZvgCggGUgCwkMUl0jjp0RaRKSZCXOF0//PF9DupCFsfvtrdv2wBowh8bDOdBw3jOQje+OE6YJl01AArAH2YRNmb9K8r+y2B+iKkmmRmlEiLSLSDDmOQ/IRvUg+oleoQ5E6Y4D12IQ5Cv9qThdQCuwGfgM61Xt0Ik2BmhxERESapL1ALuAm8OE+HJtQ78TWTYtIdalFWqQRMMaQvWIjWxf+RFHOPiLjY2lzTD8S+6ZXOh6wiDRn2dgOhpW1mUUA+UAO0Lo+ghJpUpRIizRwBTuz+eGeZ9n21TJKcgvAAYxh1dPzSRnWh0F3TyAqOTHUYYpIg+MdK7yyk22n7BZ8XHERCU6JdCNXuHsvRbv3Eh7tJrptK7VONjHF+/L5duoTbP9qOdFtWxKTaj9jYwzFOXls+fA7irPzGPaf64mIiwl1uCLSoHgP8YbgybQ5YFsRqQ79chqpXT+tZf3/FpL5yfeUFhYT5gqnZf+udD5nBO1OHKyEuonY/PbX7Fi8khadUgiPivQtdxyHyIRYwqMiyfpuFZvmfUXXC04IYaQi0vAkAluxrc3BDvdF2PKOhHqKSaRpUWfDRijjwyV8dcV0Nry2EAxEJsQSFhnB9q9+4dub/svyh+dUmGRBGh9jDBtf/xzHFeaXRO8v3B1BmCucja9/bqd1FhHxicUmyEUELt0owdZQJ6N2NZGa0S+nkcnduI0f7nuO4tx84ru392t5drdsQUFWNqtnvkdCr450OHloCCOVQ1Wyr4B9m7YTGV95yUZEQiz7MnZSlJOHO7FFPUUnIg2fA6Rjk+icsmUubDlHCbYtrTXQLhTBiTQJapFuZDbP/5qC7Xto0TElYPlGVOsEPCWlbHh1oVqlGzvH/uegH6Mp6y6kah4RqSAC6A50BuLKljlAK6Bb2XKlAiI1pV9PI7PlgyW4ot04YcGzpqikeHb9vI68jKx6jExqmysmioReaRRn51a6XeGeXOK7tyciPraeIhORxiUcW77RGxgIDMIm1y2pfEQPETkYJdKNTHFuPmGRlVfkhEW6MCWllOwrqKeopC44jkOnM34HOBTn5gfcpnhfAQ7Q6Yxj1MFURKogDCXPIrVHiXQjE5WSSEl+YaXblOQXERYZQWRL1cs2du1HH0G7k4aQt2UX+dt24ym1HQpNqYf87XvIy8gi9fhBdBijengREZH6ps6GjUza2KPYuXQNnpJSwlzhFdYbYyjcmUPauGFEp7QMQYRSm8IjIxj81ylEt23F5nlfkbs+E29rUmTLFnSbMJo+151BuDvwqB4iIiJSd5RINzIdThnKupc+IWftFuI6t/VLpo0x7Nu8g8iEWDqfOzJ0QUqtckW76X/zefSYfArbvviZ4pw8IuJiSDm6r06WREREQsgxzXBoh5ycHBISEsjOziY+Pj7U4VTbnuUb+Pbm/5KzNpPwSBfh0W48xSWU7CvA3SqO/rf9gY7jhoc6TBERkXrT2I/t0jipRboRSuyTzjGzbuW3d79h01tfUbBtN+6keNqdOJi0sUeR0L1DqEMUERERafLUIt0EzlqNMRqxQUREmrWmdmyXxkGjdjQBSqJFRERE6p8SaRERERGRGlAiLSIiIiJSA+psKFKvDLAPKMVO2xuLZhkTERFpnJRIi9QLA2QB24G8sr8dIAZoAyShhFpERKRxUSItUucM8BuQWfZ3JLaqyoNtnV4P5AMdUDItIiLSeKhGWqTO7QG2Ys9bo7ElHU7Z/71/bwWyQxSfiIiI1IQSaZE6l4VtfY4Isj6ibH1WvUUkIiIih06JtEidKgX2EjyJ9nIBOWXbi4iISGOgRFqkTnko71hYGadsu2Y30aiIiEijpURapE6FY1ubD9bS7C39CK/ziERERKR2KJEWqVNhQGtsIh2stdmUrW+NRu0QERFpPJRIi9S51kAUdog7zwHrPGXLo8u2ExERkcaiQSTSGRkZXHjhhSQlJRETE8PAgQP57rvvKr3PwoULGTx4MFFRUXTp0oUnnniinqIVqS430A07+UohNnEuKPt/YdnyrtjxpUVERKSxCPmELLt37+boo4/muOOO49133yUlJYW1a9eSmJgY9D7r169nzJgxXHrppTz//PN8+eWXXHXVVSQnJ3PWWWfVX/AiVRYL9AZ2l91KsD+/VkAiDeCnKNJsGWMo3JlDaWER7lbxuKLdoQ5JRBoJxxgT0mECbr31Vr788ks+//zzKt/nlltuYd68eaxYscK37IorruDHH39k0aJFB71/Tk4OCQkJZGdnEx8fX6O4RaQpKwVyy/7vAlrQQC7gSS3ylJaS8d63bJz7ObuXbcCUeoiIiyFt3FGkn3UsLTq2CXWIUg06tksohPzIMG/ePIYMGcI555xDSkoKgwYN4qmnnqr0PosWLeKkk07yWzZ69GiWLFlCcXFxXYYrIg1WAZAB/AQsBX7BzhhZVI3H8GCncv8F+BVYDawClgHb0fCETYenpJQf//oCS259iu1fLycsIhxXbBTFOftY9d+3+fLSB9n109pQhykiDVzIE+l169bx+OOP0717d95//32uuOIKrrvuOp599tmg99m6dStt2vi3FLRp04aSkhKysirODldYWEhOTo7fTUSakt3ACuA3bOLs7cS5sWx5bhUewwNsADYDxdia9WjssIQFZet+Q8l007D+lU/Z8L8FuFu2IL5rO9wt44iMjyGmXRLxXduxb9N2vvvzMxTvzQt1qCLSgIU8kfZ4PBx++OHcf//9DBo0iMsvv5xLL72Uxx9/vNL7OY7/MGHeCpUDlwM88MADJCQk+G5paWm19wJEJMRygfXYuvNobOfOSOxIKdHYJHgtB2+Z3oWdpj2i7DHCsMMRhpc9lgvbwq0T8cautKiY9a8swHGFE5nYosJ6JzyMFult2Ls+ky0ffx+CCEWksQh5Ip2amkqfPn38lvXu3ZtNmzYFvU/btm3ZunWr37Lt27fjcrlISkqqsP1tt91Gdna277Z58+baCV5EGoAd2BZkNxXH4XYoT6Z3VvIYpuxxIHjHzwhsq3XFq17SuOz+ZT25m7YRlZwQdJuwCPs9yPz0h3qKSkQao5APFXD00UezatUqv2W//vornTp1CnqfYcOG8dZbb/kt++CDDxgyZAgREREVtne73bjd6oUt0vSUYMs6XASfzMbBthlkAalBtikG8jj4LtGFbZGuyrTv0lCV5BbgKS4hLLLyzzsswkXRnqqUBYlIcxXyFukbbriBr7/+mvvvv581a9bw4osv8uSTT3L11Vf7trntttu4+OKLfX9fccUVbNy4kRtvvJEVK1YwY8YMnnnmGaZOnRqKlyAiIVOCbSU+2K4sDJssVza7JBw8OXbKtlWddGMWmRhLWGQEpfmVl/t4ikqISkmsn6Ck2Zo4cSKnn356qMNoMBrb+xHyRPqII47g9ddf56WXXuKwww7jvvvuY/r06VxwwQW+bTIzM/1KPTp37sw777zDggULGDhwIPfddx+PPPKIxpAWaXa8dcwHS2wNttY5WKLsKltfepDHKcXWX4d81ymHILFvOgk9OlCwIzvoNqUFRThhDu1PGFyPkUlD5ThOpbeJEyfWe0yffvopY8aM8U1m16dPH/70pz+RkZFR5ccYOXIkf/zjH+suyBp4+OGHmTVrVqjDqLKQl3YAnHrqqZx66qlB1wd6Q0eMGMH336sTiEjzFgnEYcs7KpZ1WQabAKdU8jjhQBKwheBlG56yW3JNg22GPMBebH16IfZ9TcBORBS6cruw8HC6/GEUS++aSf72PUQf0OrsKS4hd9N2Wg3oSpsRA0ITpDQomZmZvn+/8sor3HXXXX5lqdHR0fUaz3//+1+uuuoqJkyYwJw5c0hPT2fTpk08++yzPPjgg/zrX/+q13hqQ2lpKY7jkJAQvO9CQ6RmFRFp5JKxu7JAl+kNNoGLAFof5HFSsKNz5GMTwP2VYjsstsAm3HJwJcAa7HjcO7Cjq+QAm7Djcoe202bH3w+n5+Xj8JSUkrM6g/xtuynIyiZ3w1b2bthGy8M6M+Rvl+GKigxpnNIwtG3b1ndLSEjAcRzf3xEREVxxxRV06NCBmJgY+vXrx0svveR3/9dee41+/foRHR1NUlISJ5xwAvv27Qv4XN999x0pKSn89a9/Dbj+t99+47rrruO6665jxowZjBw5kvT0dI499liefvpp7rrrLgB27tzJ+eefHzSuiRMnsnDhQh5++GFfy/qGDRsAWL58OWPGjKFFixa0adOGiy66yG944b1793LBBRcQGxtLamoqDz30UIXW7d27d3PxxRfTsmVLYmJiOOWUU1i9erVv/axZs0hMTOTtt9+mT58+uN1uNm7cWKG0wxjD3//+d7p06UJ0dDQDBgzgtdde83ueCy64gOTkZKKjo+nevTszZ86s/AOtRUqkRaSRSwTaUT52dDE28S0u+zsM6AjEHORx3EC3su0KsZ0PC8oeowjb8t2VBnIhr4EzwDrKrxTEUD4cYTT289kI7AlRfPZSfa8rf8/wx/9Ip7OPJTzajRMWRnzPNAbecSFHP/Un4tLbhiw+aTwKCgoYPHgwb7/9Nr/88guXXXYZF110EYsXLwZsa/b555/P5MmTWbFiBQsWLODMM88k0MTSCxYsYNSoUdxzzz3cfvvtAZ/v1VdfpaioiJtvvjng+sTExCrF9fDDDzNs2DAuvfRSMjMzyczMJC0tjczMTEaMGMHAgQNZsmQJ7733Htu2bePcc8/1PceNN97Il19+ybx58/jwww/5/PPPK1QJTJw4kSVLljBv3jwWLVqEMYYxY8b4TZyXl5fHAw88wNNPP82yZctISal45fCOO+5g5syZPP744yxbtowbbriBCy+8kIULFwJw5513snz5ct59911WrFjB448/TuvWB2s4qUWmGcrOzjaAyc7ODnUoIlIrPMaYXcaYlcaYJcaYb8v+v8YYU93feYkxJqvsviuNMWvLHru0toJtBrKNMd8YY743xvwU4PajMeZrY8wKYz+70PN4PKa0qDjUYcghqK9j+8yZM01CQkKl24wZM8b86U9/MsYY89133xnAbNiwIeC2EyZMMKeddpp54403TFxcnHnxxRcrfewrr7zSxMfH1yj2/eMyxpgRI0aY66+/3m+bO++805x00kl+yzZv3mwAs2rVKpOTk2MiIiLMq6++6lu/Z88eExMT43usX3/91QDmyy+/9G2TlZVloqOjzf/+9z9jjH0fAfPDDz/4PZf3/TDGmNzcXBMVFWW++uorv22mTJlizj//fGOMMePGjTOTJk2q/ptRS9S0IiJNgAO0xLZOF2FbPF3YGurq8tZLq4Sj5nZirxAEO8Q42M8mF9vyH1tPcQXnOA5OhA6JUn2lpaX83//9H6+88goZGRkUFhZSWFhIbKz9Xg8YMIBRo0bRr18/Ro8ezUknncTZZ59Ny5YtfY+xePFi3n77bV599VXOOOOMSp/PGBNw8rnqxhXMd999x6effkqLFhUnK1q7di35+fkUFxczdOhQ3/KEhAR69uzp+3vFihW4XC6OPPJI37KkpCR69uzJihUrfMsiIyPp379/0FiWL19OQUEBJ554ot/yoqIiBg0aBMCVV17JWWedxffff89JJ53E6aefzvDhwyt9jbVJew0RaUIcQtmJTbwKOXjlYDi2/KaIhpBIi9TUgw8+yEMPPcT06dPp168fsbGx/PGPf6SoyPbbCA8P58MPP+Srr77igw8+4NFHH+X2229n8eLFdO7cGYCuXbuSlJTEjBkzGDt2LJGRwRsBevToQXZ2NpmZmaSmBhsb/+BxBePxeBg3bhx/+9vfKqxLTU311TkHm2H6wH8fuM3+94uOjq70pMDjsf1V5s+fT/v27f3WeecHOeWUU9i4cSPz58/no48+YtSoUVx99dX885//rOxl1hrVSIuISC0Lo2pDEnq3FWm8Pv/8c0477TQuvPBCBgwYQJcuXfw61YFNOo8++mjuueceli5dSmRkJK+//rpvfevWrfnkk09Yu3Yt48eP96sjPtDZZ59NZGQkf//73wOu37NnT5XjioyMpLTUf9jPww8/nGXLlpGenk63bt38brGxsXTt2pWIiAi++eYb331ycnL8HrtPnz6UlJT46rHBdn789ddf6d27d9DXdiBvJ8RNmzZViCUtLc23XXJyMhMnTuT5559n+vTpPPnkk1V+jkNVrT1YRkYGixcvZvfu3QHXZ2Vl8eyzz9ZKYCIi0lglcPCJa4qx5R1qjZbGrVu3br4W5xUrVnD55ZezdetW3/rFixdz//33s2TJEjZt2sTcuXPZsWNHhYQyJSWFTz75hJUrV3L++edTUlIS8PnS0tJ46KGHePjhh5kyZQoLFy5k48aNfPnll1x++eXcd999VYoLID09ncWLF7NhwwaysrLweDxcffXV7Nq1i/PPP59vvvmGdevW8cEHHzB58mRKS0uJi4tjwoQJ3HTTTXz66acsW7aMyZMnExYW5mtd7t69O6eddhqXXnopX3zxBT/++CMXXngh7du357TTTqvyexsXF8fUqVO54YYbmD17NmvXrmXp0qU89thjzJ49G4C77rqLN998kzVr1rBs2TLefvvtaiXrh6pKibTH4+GSSy6hY8eODB8+nLZt23LTTTdVOGNau3YtkyZNqpNARUSksfCOE11A4GS6tOyWhCoMpbG78847Ofzwwxk9ejQjR46kbdu2fsO3xcfH89lnnzFmzBh69OjBHXfcwYMPPsgpp5xS4bHatm3LJ598ws8//8wFF1xQobXY66qrruKDDz4gIyODM844g169enHJJZcQHx/vm+X5YHEBTJ06lfDwcPr06UNycjKbNm2iXbt2fPnll5SWljJ69GgOO+wwrr/+ehISEggLs2njv/71L4YNG8app57KCSecwNFHH03v3r2JioryPfbMmTMZPHgwp556KsOGDcMYwzvvvENERLAx/wO77777uOuuu3jggQfo3bs3o0eP5q233vKVxURGRnLbbbfRv39/jj32WMLDw3n55Zer9RyHwjHBCln289RTT3HllVdy5ZVXMnToUD7++GOef/55hg8fzvz584mLiwPsWdfw4cODfvANRU5ODgkJCWRnZxMfHx/qcEREmqDd2CHwSrAtz+HYpNo7PGECdrhBJdJSO3RsD519+/bRvn17HnzwQaZMmRLqcOpVlVqk//vf//LHP/6RRx99lIsuuohZs2bxzjvv8PPPP3PiiSeSnR18mlUREWmOWgLdy/7vndCmCJs4t0dJtEjjtXTpUl566SXWrl3L999/zwUXXABQrbKNpqJKifTq1asrXII46aST+Pjjj1mzZg0nnHBC0LppERFpruKxyXRfoGfZrR/QASXRIo3bP//5TwYMGOCbpfHzzz+v34lQGogq7ckiIyMD9iA9/PDD+fDDDznxxBMZNWoUf/nLX2o9QBERacwcymc0FJGmYNCgQXz33XehDqNBqFKLdPfu3Vm0aFHAdYMGDeLDDz9k06ZNTJgwoVaDExERERFpqKqUSB9//PHMnj07aCfCQYMG8dFHHwUdgFtEREREpKmpUmnHNddcw7Bhw8jNzSUhISHgNgMHDuSrr74K2nItIiIiItKUVGn4u6ZGQ+SIiIg0LTq2SyhoblYRERERkRpQIi0iIiIiUgNKpEVEREREakCJtIiIiEgVGI+H7F83s+PblWT/uhnj8dT5c06cOBHHcXAcB5fLRceOHbnyyis1EV4DoamlRJotD5AN5AEGO2FGIhAewphERBqmnUtXs+bZD9izfCOlBUWER0WS2KcT3S4+iaRB3ev0uU8++WRmzpxJSUkJy5cvZ/LkyezZs4eXXnqpTp9XDq7KiXR+fj5vvPEGGzduJDk5md///vckJyfXZWzSJOQBWcBeypO1JOzUwbogEjp7gM1APvZzccqWu4H22M/ICXhPEZHmZufS1fxw73MUZGUT0yEZV6ybkn2FZC35ldwN2xh410V1mky73W7atm0LQIcOHRg/fjyzZs2qs+eTqqtSIr1lyxaOPfZY1q9f75t0ZerUqbz77rscddRRdRqgNFYGyCy7lWCTZgfYB+zCtnx2RhdFQmEPsBYoxSbO3hMaD1AEbMB+fjpRFhExHg9rnv2AgqxsEnp3xHFsI0NkfAwRvTuSvWITa577kFYDuuKE1X0D0bp163jvvfeIiIio8+eSg6vSJ37HHXeQkZHBHXfcwfz585k+fTqRkZFceeWVdR2fNFpZQEbZv6OBKGzSFgNEYJPpDdiETeqPB/gNm0RH4b8LCCtbZrCfXUm9R9e0FQO7gZ1ADvazEJGGLmdNBnuWbySmQ7IvifZyHIeYDsnsWbaBnDUZQR7h0L399tu0aNGC6OhounbtyvLly7nlllvq7Pmk6qrUHPjhhx/y5z//mTvvvBOAU045ha5du/L73/+ebdu20aZNmzoNUhobD7C17N+RAdaHly3fgy39iK2fsARbYpOPff+DlW5EAoXYz6d1/YTVpJUAW7AJdDH2RCUMe4KZCrRCZTQiDVdR9j5KC4pwxboDrnfFuMnfUkRR9r46i+G4447j8ccfJy8vj6effppff/2Va6+9ts6eT6quSi3SW7du5dhjj/VbNnLkSIwxbNu2rU4Ca5wMNknJwx4wm6u9QAG25TmYcGyr6K56iUi8vDXRlXUoDKP8uyyHpgRYgy1x8mBb/KOxv408YB2gfag0F4XYDs7Z2DKyxiEyIZbwqEhK9hUGXF+SV0h4VCSRCXXXKBQbG0u3bt3o378/jzzyCIWFhdxzzz119nxSdVVqkS4tLSU6OtpvWVRUFAAlJbr8aw+QWcAO/BOVlkAbbDlDc1KMfU8qS9acslvj2ZmKVN92bNLgxv/3EI5NqAuxZTTxNL/9RFNUiL3ysBvbUBCB7bjbiubdHyQfezLpfV/Avh8tgXbY30fDFd+tPYl9OpG15Fci9quRBjDGkPfbDlof0ZP4bu3rLaa7776bU045hSuvvJJ27drV2/NKRVX+Za9atQqXq3zz0lL7Y1i5cmWFbQ8//PBaCK2x8GBblbwtqxHYBLEUexDdA3TFHiibC2+SvP9oEIEcrGVUal8M5d/PYO+9p2yb6CDrpWpKsSfX4QR/ryOxScZOlEg3drux/T6KKO9cXYCth9+GPQ40x894H/aqjPcqpTdpLsEeI3OB7tirNQ2TExZGt4tPInfDNrJXbLKjdsS4KckrJO+3HUQlJ9DtohPrpaOh18iRI+nbty/3338///73v+vteaUix3iH4ahEWFhYhQJ7sGdiB56ZOY7jS7IbqpycHBISEsjOziY+/lAT3EzsMGIRVDwvMdidhxvoG2B9U1UE/IJ9/YFqpMEma4VAN2xrjdQPAyzHHtyiqHiiY7CfSyRwGDrRORS5wArsvqGy97EAm2D1rY+gpE7kAr9ik8MDf1ceyj/jXlRe8tbUePc3udgT80D7m3zsKE49Aqyvnto9tlcUcBzpvul0u+jEOh36buLEiezZs4c33njDb/mLL77IpEmTWLNmDWlpaXX2/FK5KiXSs2fPrtaDTpgwocYB1Yfa+7GVYhPGYoJfmvImjJ1pXsOJbcC2whw4MgSU7zxjgD4oWatvOdgWohJswux9/73f1TDs9zUpJNE1HXuBldj3uLKWqkLs/qNffQQldWId9upDoGQRypPpdGy5X3ORjT3BqOxksgR7LO3NoXY8r+tEGuxQeDlrMijK3kdkQizx3drXa0u0NDxVaiJt6Ilx6OzDtr4Ga3WF8gPoHppXIt0emyznYHegLuwBpqTs5sYeVJRE17947JWATdjPqJDyg38U0AFdJagN3jG6S6k8kfagMprGrBi7f/eW9QXiLfXYSfNKpHM5eH+ZcOx7mEtjGMHJCQsjoYdaf6VcrZ1GeTwe3nrrLc4444zaeshGwMPB64ChvCa1OYnA1r11KPt3Mfakw8GeUPQA4kIWncRjrwb0ADpiP6fu2HIOJdG1IxLbmco75F0gpdjfhFr/G69iDn6yBDZhbG6dq6syVrr3+Klx1aVxOuSi3V9//ZUZM2bw7LPPsnXrVmJimlNnCm8r68HOuD1U3mrdVLmwLdNtKR/NxE3zfC8aojAgoewmdaMN9vK2t6+EN9ky2OSrCJts6zNovLytzQerkvSOH96cePf1lTU4eQ7YVqRxqdGvOi8vj1mzZnHsscfSu3dv/vGPf5CUlMQjjzxCRkbdzezT8MRi63wra2Xwtji1rJeIGqZwoAW2BVo7S2lOYrGjNURhS2jysCeV+dgSp1ZAF5pfgtWUuLGfc2VzB3hPnJrbcaAl5VckgynGHhd0MimNU7VapL/++mtmzJjBK6+8wt69e4mLi+MPf/gDL774Io899liFSVuaPgfb2roOm0wfWCPn7bwVj3YSIs1VPHZEjt3YPgOl2MShFfYEs7ZmNTTYDo77KB8xJ5HmM1pQqHjL1fZik8IDR+XwjoQTQfObKTQS+95kYt8b71VcsO+Ld86Btuh7Ko1Vlb65//rXv3jmmWdYuXIlxhh+97vfMWXKFM455xyKiop44YUX6jrOBqwVdmeQgW1lCse/JjoOtTiJNHfh2CSqrhKpXGzn0X2UjwMONpFJBVKovYRdKkrCvvfbsFcavI0qHsoTyI40z06l7bHvg3fCMu+x0EN5+V9z6oApTU2VEumpU6fiOA5jx47loYceomvXrr51xcXNeSpsKG+VjsPObpiN3UHEYg+aLdHIFCJSd7wTXniH0dt/OMMibIJtsPspqRsONlGOwSaMeZTXRLfCJorNaVKu/YVh35vW2FFL9mHfr1jsCUhz6lclTVGVEumBAwfyww8/MH/+fLZt28bkyZM5//zz62ycxsYplsYwdI+INDUZ2CT6wDGMwyivzd6CPalv2FMxN27eEo/W2M6l3inC9Z6XJ846RkrTU6V6g++//56lS5dy5ZVXsmbNGq688kpSU1O5+OKLWbBgQR2HKCJ1rxR7RWUV8BOwDJugFYQyKDmoPGzddSTBSzciseUFu+orqGbOwZ7UtEBJtEjTV+XC3QEDBvDvf/+bLVu28Pzzz3PUUUfxwgsvcOaZZ+I4DnPmzGHr1q11GauI1IkC7Ax867ClSUXYBO037PS+O0MXmhxEPvYkqLLyMafsllcvEYlIw7ZhwwYcx+GHH34IdShNQrV7wLndbv7whz/w8ccfs3btWv785z/Tvn17Hn30UTp16sTZZ59dF3GKSJ0oBdZiO6u5sS1pbmxJQHTZ+g3YVk8RkebOUH4lyFsL33BNnDiR008/PdRhNGmHNJREeno69913Hxs3bmT+/PmMGzeO+fPn11ZsIlLndmM7/0RRcXfgYJPqEuxoBNLwRGNboyubOdWU3dSpS+TQ7MV27F2NbYBYXfb33lAGJSFWK2OyOY7DKaecwmuvvcZvv/1WGw8pIvXCWzcbbFfgYDtM5aB66YYoGjtiUBHBW8a8Y9xr6neRmtuLvTqXix2nIbrs/7lly+smmR45ciTXXHMN11xzDYmJiSQlJXHHHXdgjOHee++lX79+Fe4zePBg7rrrLqZNm8bs2bN58803cRwHx3H8+rWtW7eO4447jpiYGAYMGMCiRYv8HmfOnDn07dsXt9tNeno6Dz74oN/69PR07r//fiZPnkxcXBwdO3bkySefrJP3oSGr9cGNk5KSavshRaTOFHLw3UA45ePhSsPiAB2wHQq99dJe3gmhDHYsaXV8E6kZA2zFXp3zJtAO5Ql1Sdn6uinzmD17Ni6Xi8WLF/PII4/w0EMP8fTTTzN58mSWL1/Ot99+69v2p59+YunSpUycOJGpU6dy7rnncvLJJ5OZmUlmZibDhw/3bXv77bczdepUfvjhB3r06MH5559PSUkJAN999x3nnnsu5513Hj///DPTpk3jzjvvZNasWX6xPfjggwwZMoSlS5dy1VVXceWVV7Jy5co6eR8aqioNfxceXvVxkB3H8X0QItLQhXPwnb93vSYVaphige7ARmzNZiHlI3hEYJNoTXghUnP52N9WoNFxnLLleWXb1X4JVVpaGg899BCO49CzZ09+/vlnHnroIS699FJGjx7NzJkzOeKIIwCYOXMmI0aMoEuXLgBER0dTWFhI27YVx5GfOnUqY8eOBeCee+6hb9++rFmzhl69evGvf/2LUaNGceeddwLQo0cPli9fzj/+8Q8mTpzoe4wxY8Zw1VVXAXDLLbfw0EMPsWDBAnr16lXr70NDVaVE2hhDQkICv//97zV2tEiTkoi9JGkIPnxaMbbVpTnOytZYtAB64z9FuBtNES5SG0qwV3iCNSqGY0uo6qYR8aijjsJxyvfPw4YN48EHH6S0tJRLL72UyZMn869//Yvw8HBeeOGFCiUYwfTv39/379TUVAC2b99Or169WLFiBaeddprf9kcffTTTp0+ntLTU18C6/2M4jkPbtm3Zvn17jV9rY1SlPezEiRN57bXXmDNnDmeeeSZTpkxhxIgRdR2biNS5JGxHQu+seAcm0yXYpCwZtUg3dGFAQtlNRGqPC/v7KiVw2lRatr7+T1rHjRuH2+3m9ddfx+12U1hYyFlnnVWl+0ZERPj+7U3UPR4PYBtQ90/evcsqewzv43gfo7mo0pFxxowZZGZm8vDDD/uK07t168b999/Pli1b6jpGEakzbqATdleQj2199mAPDAVlf7fGJtIiIs1RNLZkI1CnXlO2PIa6umr39ddfV/i7e/fuhIeH43K5mDBhAjNnzmTmzJmcd955xMSUl5dERkZSWlrZqD6B9enThy+++MJv2VdffUWPHj2qVe7bHFS5iSk2NpYpU6bwxRdfsGLFCs4880zf2NFjx47l448/rss4RaTOtAJ6YFunvQcFb6eaTkA6ao0WkebLAdpiW5zzKb9SV1L2t6tsfbDyuEOzefNmbrzxRlatWsVLL73Eo48+yvXXX+9bf8kll/DJJ5/w7rvvMnnyZL/7pqen89NPP7Fq1SqysrIoLq5ap/E//elPfPzxx9x33338+uuvzJ49m3//+99MnTq1Vl9bU1Cj6xA9e/bk73//Ow888AB//etfuffee3G73YwaNaq24xORehFXdivEtkKHEXhsaRGR5igO26iwFduxsAi7f2yBTaLj6uyZL774YvLz8xk6dCjh4eFce+21XHbZZb713bt3Z/jw4ezcuZMjjzzS776XXnopCxYsYMiQIeTm5vLpp5+Snp5+0Oc8/PDD+d///sddd93FfffdR2pqKvfee69fR0OxapRIZ2dn8+KLLzJjxgy+++47evbsqRkNRZoENxomTUQkkDhs4uxtlfYOf1c3LdFeERERTJ8+nccffzzgemMM27Zt4/LLL6+wLjk5mQ8++CDgffaXmJhYYdlZZ51Vab31hg0bKixrjtOOVyuR/vjjj5kxYwZvvPEG4eHhnHPOOTz88MN+4xKKiIiINE0ODWmW0O3bt/Pcc8+RkZHBpEmTQh1Os1SlRPree+9l1qxZbNy4keHDh/Pvf/+bc889l9jY2LqOT0REREQCaNOmDa1bt+bJJ5+kZcuWoQ6nWXJMoPFMDhAWFkZ8fDzjx4+nZ8+elT+g43DDDTfUWoB1IScnh4SEBLKzszUutoiISBOgY7uEQpUT6So/oOPUaKiV+qQfm4iISNOiY7uEQpVKO9avX19nAUybNo177rnHb1mbNm3YunVrwO0XLFjAcccdV2H5ihUrmtWUlCIiIiISWlVKpDt16lSnQfTt25ePPvrI93dVBvtetWqV3xlncrImjBARERGR+lP/81kG4HK5aNu2bbXuk5KSQmJiYt0EJCIiIiJyEFVKpI8//vig6yIiImjTpg0nnngi559/Pi5X9XPz1atX065dO9xuN0ceeST3338/Xbp0qfQ+gwYNoqCggD59+nDHHXcELPcQEREREakrVepseNhhh+E4gQccLykpITMzk5ycHIYOHconn3ziN8/7wbz77rvk5eXRo0cPtm3bxl/+8hdWrlzJsmXLSEpKqrD9qlWr+Oyzzxg8eDCFhYU899xzPPHEEyxYsIBjjz024HMUFhZSWFjo+zsnJ4e0tDR1SBAREWki1NlQQqFKiXRVvPvuu4wfP55rr72Wv/71rzV+nH379tG1a1duvvlmbrzxxirdZ9y4cTiOw7x58wKuD9ShEdCPTeqBAUqxU8lqum0RkbqiRLrqNmzYQOfOnVm6dCkDBw4MaSyzZs3ij3/8I3v27Kmz5xg5ciQDBw5k+vTptf7YtXZkP+WUU7jllluYM2fOIT1ObGws/fr1Y/Xq1VW+z1FHHVXp9rfddhvZ2dm+2+bNmw8pRpGDKwAygJ/Kbj8C64C9oQxKREQOgcdj2JqRw7rVWWzNyMHjqZW2yDo1ceJETj/99FCH0WTVamfDoUOH8pe//OWQHqOwsJAVK1ZwzDHHVPk+S5cuJTU1Neh6t9uN2+0+pLhEqm4vsBYoBMKx56seYAewC+gAVK9zbdPmwbbae98rEZGGZ+O6XXzx6Vq2bM6mqKiUyMhw2qUl8LvjutKpS6tQhychUqtHrfz8/Gp3Npw6dSoLFy5k/fr1LF68mLPPPpucnBwmTJgA2Nbkiy++2Lf99OnTeeONN1i9ejXLli3jtttuY86cOVxzzTW1+VJEaqgI2/JcBEQDbiACiAS8fQd+A/aEIrgGZi+wHviB8pb7TUB+CGMSEalo47pdvPHKT6xfs4vYFm7atosjtoWb9Wvs8o3rdtXJ844cOZJrrrmGa665hsTERJKSkrjjjjvwVuXee++99OvXr8L9Bg8ezF133cW0adOYPXs2b775Jo7j4DgOCxYs8G23bt06jjvuOGJiYhgwYACLFi3ye5w5c+bQt29f3G436enpPPjgg37r09PTuf/++5k8eTJxcXF07NiRJ598skav9Y033qBHjx5ERUVx4okn+lUPrF27ltNOO402bdrQokULjjjiCL9hkwH+85//0L17d6KiomjTpg1nn32233qPx8PNN99Mq1ataNu2LdOmTatRnAeq1UT6nXfeoU+fPtW6z2+//cb5559Pz549OfPMM4mMjOTrr7/2jV2dmZnJpk2bfNsXFRUxdepU+vfvzzHHHMMXX3zB/PnzOfPMM2vzpYjU0C5sWUcUEKiDrhvb+rodWz/dXG0DfqX8fXCAEiATWAlkhy40EZH9eDyGLz5dy96cQtp1iCc6JoKwsDCiYyJo1yGevTmFfPHp2jor85g9ezYul4vFixfzyCOP8NBDD/H0008DMHnyZJYvX863337r2/6nn35i6dKlTJw4kalTp3Luuedy8sknk5mZSWZmJsOHD/dte/vttzN16lR++OEHevTowfnnn09JSQkA3333Heeeey7nnXceP//8M9OmTePOO+9k1qxZfvE9+OCDDBkyhKVLl3LVVVdx5ZVXsnLlymq9xry8PP76178ye/ZsvvzyS3JycjjvvPN863NzcxkzZgwfffQRS5cuZfTo0YwbN86XHy5ZsoTrrruOe++9l1WrVvHee+9VGIBi9uzZxMbGsnjxYv7+979z77338uGHH1YrzkCq1Nlw167gZ1qlpaVkZGTwyiuv8I9//IP//Oc/XHbZZYccWF1ShwSpO8uAPGwiHUwxNnk8DJtYNzfZgLdPQyT+JxwGeyISAfSm8vdRRKRcXR3bt2bk8NxT3xDbwk10TESF9fl5xezLLeSiS4fStn3t5hQjR45k+/btLFu2zDd62q233sq8efNYvnw5AGPGjCE9PZ3//Oc/ANxwww388MMPfPrpp4Ctkd6zZw9vvPGG73G9nQ2ffvpppkyZAsDy5cvp27evb6boCy64gB07dvDBBx/47nfzzTczf/58li1bBtgW6WOOOYbnnnsOAGMMbdu25Z577uGKK66o0mucNWsWkyZN4uuvv+bII48EYOXKlfTu3ZvFixczdOjQgPfr27cvV155Jddccw1z585l0qRJ/Pbbb8TFxQV8H0tLS/n88899y4YOHcrxxx/P//3f/1UpzmCq1CLdunVrkpOTA97atm3L4MGD+fvf/87VV1/d4JNokbpVwsF/Vt6a6ZK6D6dB2oFtlT8wiabs7yhsaczOeo5LRKSivLwiiopKcUcFnnU50h1OUVEpeXlFdfL8Rx11lN8QxMOGDWP16tWUlpYCcOmll/LSSy9RUFBAcXExL7zwApMnT67SY/fv39/3b29fs+3btwOwYsUKjj76aL/tjz76aL/nPvAxHMehbdu2vseoKpfLxZAhQ3x/9+rVi8TERFasWAHYEd1uvvlm+vTpQ2JiIi1atGDlypW+FukTTzyRTp060aVLFy666CJeeOEF8vLygr5W7+utbpwBY6/KRnfddVfQcaRdLhdt2rTh+OOPP+gkKiJNnwubBFbGYJPpBjGxaD0rwrZIRxC49IWy5WHYRLp9PcUlIhJYTEwkkZHhFBaUEh1TsaGkqNB2PIyJiQxBdHYIYLfbzeuvv47b7aawsJCzzjqrSveNiChvYffmeR6PB7CtywfmfoGKGPZ/DO/jeB+jOgLlmd5lN910E++//z7//Oc/6datG9HR0Zx99tkUFdnjbVxcHN9//z0LFizggw8+8NWHf/vtt75ZsGsrzgNV6UheWwXZIk1fKyCX8rrfQIqBBGyLbHNTin1vArfslAvDtthX9j6KiNS9lNQ42qUlsH7NLtpFx/slfMYYdu3Mo3O3VqSkViwpqA1ff/11hb+7d+9OeLjdj7pcLiZMmMDMmTNxu92cd955fhPjRUZG+rUgV1WfPn344osv/JZ99dVX9OjRw/fctaWkpIQlS5b4yjhWrVrFnj176NWrFwCff/45EydO5IwzzgBszfSGDRv8HsPlcnHCCSdwwgkncPfdd5OYmMgnn3xS533ommOTmEgdaoXtSBesw2FR2bKUAOuag3Ds6/ZQeTLtwdaPN8f3SEQakrAwh98d15Ws7fvY8lsOrZJibDlHYSm7duYRF+/md8d1JSysbvZXmzdv5sYbb+Tyyy/n+++/59FHH60wesYll1xC7969Afjyyy/91qWnp/P++++zatUqkpKSSEhIqNLz/ulPf+KII47gvvvuY/z48SxatIh///vfvlrs2hQREcG1117LI488QkREBNdccw1HHXWUL7Hu1q0bc+fO9U3Ad+edd/q1Jr/99tusW7eOY489lpYtW/LOO+/g8Xjo2bNnrcd6IA3aKk1YAbAFO6bzOmyCW1zHz+kGumBLF/KxY0mXYBPoPGwLawcgsY7jaKgisa3x3g6XgRhsIp1UX0GJiFSqU5dWnD6+P527tWJfbiHbMveyL7eQzt3s8rocR/riiy8mPz+foUOHcvXVV3PttddW6I/WvXt3hg8fTs+ePX0d9rwuvfRSevbsyZAhQ0hOTq6QaAdz+OGH87///Y+XX36Zww47jLvuuot7772XiRMnViv+kSNHHvQ+MTEx3HLLLfzhD39g2LBhREdH8/LLL/vWP/TQQ7Rs2ZLhw4czbtw4Ro8ezeGHH+5bn5iYyNy5czn++OPp3bs3TzzxBC+99BJ9+/atVqw1UWtThDcmGrWjqfNgE+ht2CTWoTxpiwTSgNZ1HEM+5ROwlJbFkFD2vFVrDWi6NGqHiNS++ji2ezyG7Zl7ycsrIiYmkpTUuDpriYaqT21tjKFXr15cfvnl3HjjjXUWT02kp6czbdq0aifgjYVKO6QJ2lJ2C8dOiuLdyRlsC/EG7MWYupyJKhroiG19LimLpXZryhqvBOzJzG/YEw4X5eUe3tE8OqMkuukzxrD7p3Vs/ewnivfmEREfS9tj+9GyX5egHdxFQikszKn1Ie4O1fbt23nuuefIyMhg0qRJoQ7Hz8qVK4mLi/ObWK+pUSItTUwhdpKPcCp25nOwpRfeko9E6r66KSxAHAJtsDM97sDO8miwu6MUIBl7IiJNWf723Sy9exY7Fq+gJK/Qd+Fo9Yx3ST6qN4ffM5Go5MRQhynS4LVp04bWrVvz5JNP0rJly1CH46dXr178/PPPoQ6jTimRliZmN7b+Nlgi5mAT23zsFNXNvcwilOLKbqWUdz5Ut43moChnH4tv+A87l/xKdGorYtq3xnEcjDEU780j8+Pv+SY3n2GPXU9EXMzBH1Ckidp/Ou9gmmGFboOio5Y0MQVl/6/ssnA4NnErrPtwpArCsTXR2h01F5vfWsSu71fTIr0NkfExvjIOx3GIjI+lRXobspb8yub5Xx/kkUREQktHLmliqlJX6T17Vw2mSH0zHg8bX/+csIhwwt0Vp1sGCHdH4oSHsXHu52ptE5EGTYm0NDGxlHdcC6YU2woaWy8RiUi54r157MvYSURC5b+/yIRY9v22g5Lc/HqKTESk+pRISxPTEtuhsJDA4xQb7JjO8djObiJSrxwHxwEO0tLsa4muw6HFREQOlRJpaWLCgU7YfrT5lE8zbbCdEPOxHRHTQhWgSJ0xxlC8L5/i3PwGWxIRERdDQo80Cvfsq3S7oj25JPbuhCtGwyCKSMOlUTukCUoEumKHuMvFtkCD/bonAe3R8GrSlJQUFJHx3jdsnPs5OWsyAGjRuS3pZxxDhzFHNqhk1HEcOp7xO7KWrKI4N5+IFhV/i8W5+ThhYXQ68xiNJy0iDZoSaWmiErDlG/uwI3k42FIOJdDStBTn5rPk1ifZ+umPGAfcCbHgwO4f17Hrh7VkfLCEI/5xOZEJLUIdqk+HU4aydeGP/PbuN0TGx+BuHU9YeDieklIKd+ZQlL2PtFOH0e7EwaEOVUSkUkqkpQlzgBZlN5Gm6ZcH/8eWj74ntn1rXLHlLc/uVvGU5BeydeFP/Hj/ixzxt8tCGKW/8MgIBv91CjGpSWya9xW567f51kW1TqDHpSPpffXphEcGHtVDRKShUCItItJI7cvYwW/vfoO7VZxfEu3linYTlZxA5qdL2btuC3Fd2oUgysBc0W763TSe7pNPZvtXyyjOySMiPoY2Rx+Gu1XDmoJZRCQYJdIi0iR4SkrJWf0bJfmFRCUlENsxpcnX1277/GeKs/cR1zU16DaRLVuwd80Wtn72U4NKpL2ikhLoOG54qMMIoIjymVIdymfibNrfKRGpHiXSItKoeUpK2TDnMza8tpC9a7dgSjyERUXQenBPul50Im2G9w11iHWmOCfPDicXFnwAJsdxIMyx20oVeLAdlbdjR/0BO+pPGHbs+XQ0dKaIeCmRFpFGy1NSyg/3PcfGOZ9BmENU6wTCIlyU5BeSufAHsr5bxYA/X0Cn038X6lDrRERcDMYYjMcTNJk2xoDH4AowOoYcyACbgW3YoTSjKG+BLgX2AquBHqjjsoiAxpGuM6WlHrZuyeG3jbvZm1MQ6nBEmqRNb37Jxrmf406KJy69LREtogl3R+BObEF813aY4lJ+/vvL7F23JdSh1omU3x1GZEIshbv2Bt2maE8urrho2h7bvx4ja6z2ATuwbUyR+JdxhGOT5wIgs/5DE5EGSS3Stay4uJRvvtzIkq82sXNHLh4DkZHh9OnflmEjOtM+LTHUIYo0CcbjYcNrCwE7nfSBHMchpkNrclZnsOntr+l73Zn1HWKda5GWQvuThrD+fwtwxUThinH7rS/JL6JgezZpvx9GfLf2IYqyMdmFbXmODLLewR4292BrqINtJyLNhRLpWlRUVMqrz33PL0szCQt3aBEXSVhYGIUFxXzz1UZWLd/O+IsPp1uv5FCHKtLo5W7cRs7qDNxJwUd4cBwHV4ybzE+WNslEGuCwqedSsH03Wz//GScszJ5UOFCUvQ9T4iHl6L4MuO0PoQ6zkdiHvVBbWYdCF1CInSVVibRIc6dEuhZ9tWAdP3+/hYSWUURFlY9/GhkZTos4Nzu25TL3xR+4+uZjiW3hruSRRORgSguK8JR6CIsIr3Q7xxVOaX5hPUVV/yLjYxn60NX8Nv9rNsz9nL3rtwKQ2LsTnc48hrRTjyIiVvW8tU+jd4iIEulaU1RYwrdfbSQiMtwvifZyHIek5Fiytu/jlx8yOfJ36fUfpEgTEtU6AVdUJCX7CnBFBz8xLc0rJOawzvUYWf1zRbtJP3sEnc46luKcfRhjy12a+vB/ta8FkIPtdBjsvSvBHjobzrTrIhI66mxYS37btIc9u/KJiwt+QA8PD8NxHFYt2xZ0GxGpmqjkRNqMGEDhrr12ZIoASotKMB4PaacOq+foQsNxHCITWuBObKEkukaSsElycZD1BptIt0RlHSICSqRrTVFRKR6Ph3BX5W9peLhDQUFJpduISNV0/cPxuJPiyd2wDePx+K0rLSomd8NWEg/rTPsTB4coQmlcooE22A6HhdgxpaE8gc7HjiHd8Ca2EZHQUGlHLWkR58YVEU5RYQlR0RVLO7xKSjy0bKV6RZHa0Kp/Vwb/ZTJL73mWnDVbCI904US4KC0oAqBl/y4M/ccVRMRpAg2pCgdojx3qbis2mQabSIcDiUAnoLn2cSnEjlhSin0/4tF42tLcKZGuJe3TEujQMZH1a3cGTaQLC0sICw+j3yANQyVSW1KPG0RC705kvPsNWz5ZSkluHjHtkmh/8lBSjx+kjnZSTQ6QCiRTPsxdGHZ68BiaZyfDUuxENTuxLfMO9uTChT256AgEb0ASacqUSNcSx3H43fFdydi0h9278khsGe1Xo1hcVMqunXl075ms4e9EallM21Z0n3Qy3SedHOpQpMlwAa1DHUQD4AHWYcfYdmFboL2JdAmQha0p74ZSCmmOVCNdi/oOSGXMmX0JDw9jW+ZedmXlsWd3Ptu35rJ7Vz5de7Tm3AmH4zpIHbWIiEjDsLvs5sZ/tkcH2wrtBrKxrdUizY9OH2vZUcd0Jr1rEj8s+Y1Vy7ZTXFxK65RYBg1Jo3f/tkRGVj7mrYiISMORVfb/YMeucGyb3A4gheZZ+iLNmWOCjRvVhOXk5JCQkEB2djbx8cFnRRMREWm+PMAP2DKOyob7Kynbtt9BtqtbOrZLKKjGQERERCqhVmaRYJRIi4iISABh2JFKDjb3QQm2JVrVotL8KJEWERGRILwjl5QGWe8puyWjlEKaI33rRUREJIhW2LGiC7HD3Hm7VXmHvyvATsyioQKleVIiLSIiIkGEAV2wLc4GO016Xtn/S4EkoCsq65DmSt98ERERqYQLm0wXYGd7LMEOe5eAraEWab6USIuIiEgVRAFtQx2ESIOi0g4RERERkRpQIi0iIiIiUgNKpEVEREREakCJtIiIiIhIDSiRFhERERGpASXSIiIiIiI1oERaRERERKQGlEiLiIiIiNSAEmkRERERkRpQIi0iIiIiUgNKpEVEREREakCJtIiIiIhIDSiRFhERERGpASXSIiIiIiI1EPJEetq0aTiO43dr27ZtpfdZuHAhgwcPJioqii5duvDEE0/UU7QiIiIiIpYr1AEA9O3bl48++sj3d3h4eNBt169fz5gxY7j00kt5/vnn+fLLL7nqqqtITk7mrLPOqo9wRUREREQaRiLtcrkO2grt9cQTT9CxY0emT58OQO/evVmyZAn//Oc/lUiLiIiISL0JeWkHwOrVq2nXrh2dO3fmvPPOY926dUG3XbRoESeddJLfstGjR7NkyRKKi4sD3qewsJCcnBy/m8ihKQUKgELAhDgWERERCYWQJ9JHHnkkzz77LO+//z5PPfUUW7duZfjw4ezcuTPg9lu3bqVNmzZ+y9q0aUNJSQlZWVkB7/PAAw+QkJDgu6WlpdX665DmIh/YCPwE/FJ2WwlkoYRaRESkeQl5In3KKadw1lln0a9fP0444QTmz58PwOzZs4Pex3Ecv7+NMQGXe912221kZ2f7bps3b66l6KV52QusArYCHiAc+xPaC6wHNpQtFxERkeagQdRI7y82NpZ+/fqxevXqgOvbtm3L1q1b/ZZt374dl8tFUlJSwPu43W7cbnetxyrNSTGwDigCooH9T9pcQAmwA4gB2lS4t4iIiDQ9IW+RPlBhYSErVqwgNTU14Pphw4bx4Ycf+i374IMPGDJkCBEREfURojRLu7H10FH4J9FerrLl21GrtIiISPMQ8kR66tSpLFy4kPXr17N48WLOPvtscnJymDBhAmDLMi6++GLf9ldccQUbN27kxhtvZMWKFcyYMYNnnnmGqVOnhuolNEMG2Af8BqzBttRmYVtlm6o9Zf8PXD5kRWA7IObVeTQiIiISeiEv7fjtt984//zzycrKIjk5maOOOoqvv/6aTp06AZCZmcmmTZt823fu3Jl33nmHG264gccee4x27drxyCOPaOi7elOK7Wy3q+zfDjaxzgLcQDqQEKrg6lAJlSfRYM9LDfZ9EakJD5CNLSVygBbYUiIREWmIHOPtqdeM5OTkkJCQQHZ2NvHx8aEOpxExlLc+R1BezgA2ASgsW9YdiAtFgHVoDfbkobKkpgSbRPcGYusjKGkyvCejmdirGt5lLuyJaRr2RFVEgtGxXUIh5KUd0pjkYpPJSGwivX8LbRi2frgYmww0NS3L/l9Z/XMxtrNhTN2HI03MNuyoL4XYhDma8k6tO4HV2I6uIiLSkCiRlmrYSfmwb4E42AR7L3a85aYkEdvKXEDFZNpgkxwHO2LHwUpARPZXAGRgvzdRlO+Wvb+nKGyfhC0hiU5ERIJTIi3VkI/9ylSWKLqw5Q2F9RJR/QkHulCeTBdgk+dC7PtigA5Aq1AFWAsKsCdBeWjkkfq0E1sWFBlkfRj2d7ULe9VDREQaipB3NpTGxNuZrjLe9U2xVTYa6IlNaLKwSXQ4kAS0pvHWhWdjSwv2YhNob8toCvZ16Xy7bu2laieohdiTnKbYmVdEpHFSIi3VEIcdBs4Q/KBfgr0c3VTrhCOw5RttaBonDVnYUVi8n1sk9nXlY2t29wGdUDIdat7vWLPrGy4i0qDp6CjVkIRNtoJ1evJgEzLvdk2dQ+NOovOBTdjPLRr7mYVhW9mjyv7egS09kLoTg/0MKkuSSyj/XEREpKFQIi3V4MbWAYNNwrzjJXs72xVgx70NPCulNDQ7sTW3boLP1gh2tka1hNadVtgkOdiERgb7OSWgRFpEpGFRaYdUUwr2a7MFm0wXYpMwV9m6DjSP1uimYA8Hr82NwH7OBWhikLoSi72Ks6Ps70Djs0eiE1QRkYZHibTUQCvscHC52JboMGxLdLBRB6RhKuXgF6W8M1dqFI+641Beh56F/9CRDvYEJh1N8iMi0vAokZYaCgM0c1TjFsnBhyn0YD9rXWWoW2HYZLoNdlSY/U9QE1EVnohIw6REWqTZak35kHeBEjVvbW5rdLWhvkQB7UIdhIiIVJGaOUSarZbYESOCzdZYQPlwfyIiInIgtUiLNFsuoCuwDjteNJRPuuPBtkKnY8sLRERE5EBKpEWatWigF7AbOxxeEXYotkTsSBLukEUmIiLS0CmRFmn2wrF10K1DHYiIiEijokRaRBq1veu2sHvZBkyJh9iOKSQN6oYTpu4fIiJS95RIi0ijtHfDVpY99BrbFy2jZG8+Bgh3R5DYpxO9rvw9bY/pH+oQRUSkiVOzjYg0OnvXZ7LoqofJeO9bwt0RxHVrR0L39rhbxbHrhzV8e9N/yfhwSajDFBGRJk6JtIg0Osumz2Hv2i3EdWuHu2UcjmOn1I5oEU1c13aU5BXw899fpnhvXogjFRGRpkyJtIg0Kjlrt7B90TKiUhIIC6+4C3Mch9j2rcnL2MmWj74PQYQiItJcKJEWkUZl9y/rKd6bT2Ri8PGtwyJcGONh97L19RiZiIg0N0qkRaRRMaUeHMfxlXME4+DgKS6tp6hERKQ5UiItIo1KbFoyYZEuivcVBN3GeDwYY2jRMaUeIxMRkeZGibSINCpJh3cnoVca+Zm7MMYE3KZgRzaRLeNof/IR9RydiIg0J0qkRaRRCQsPp/eVpxERF03uxu14ikt864zHQ/623RTvK6DrBaOIbZ8cwkhFRKSpUyItIo1O2xEDOPzeSUSnJJK7cTvZv/5GzuoMctZmQlgYvS4/lV5X/D7UYYqISBOnmQ1FpFFqP/oIkof1YctH37P7l/WYklJadGpD+9FHENtBLdEiIlL3HBOsyLAJy8nJISEhgezsbOLj40MdjoiIiBwiHdslFFTaISIiIiJSA0qkRURERERqQDXSIiLShJQAu4E9QCkQCbQEElDbkYjUNiXSIiLSROwF1gGFZX87gAfIAloAXQF3aEITkSZJibSIiNQRA+QBxUA4EEvdtQrnAWuBImyyvP/zlGKT7LVAD3ToE5Haor2JiIjUMoMtr9iKTXA92NZhN5AMtKH2E+rt2Jbo6LLn2l84EAXklsWl4RFFpHaoYExERGrZNmzrby7lSWwENtHdDKzHJte1xVsX7aJiEu3lPdxl1eLzikhzp0RaRERq0T4gA5vQRlOe3O6fUO+kdhPaImz5RvhBtgunvH5aROTQKZEWEZFatBPbQhwZZL23onAHtgSkNngPZQd7PEPwFmsRkepTIi0iIrVoD7blt7KENQLIBwpq6Tnd2Nbv4kq2MdhW68Raek4RESXSIiJSq7wdCyvjYBPb2mqRdijvQFgSZJtibGt4Ui09p4iIEmkREalVbmzLb2W89cwRtfi8rbFJcjG2pdtDeSt0ftnf7bDjSYuI1A4NfyciIrWoNXbMZg+B22oMttU4hdpNpMOAzkAMtv66kPKa6FigLdCqFp9PRESJtIiI1KpW2DGdc7Gt0/uPpGGwrcWR2LGka1sYkIpN0vdhk3kXNpFWJ0MRqX1KpEVEpBaFA90oH0faYBNcbz20m/KW47qMIb4OH19ExFIiLSIitcwN9AKygV3YcZ7DsSNmtKR2SzpEREJHibSIiNSBMGzS3DLUgYiI1BmN2iEiIiIiUgNqkRYREWkSiikfMcUFxHHwadNF5FAokRYREWnUSoEM7PTsxZQP+xeFHR0lBY1aIlI3lEiLiIg0WqXYEVJ2Yw/pUdik2YMdS3sjtrNnB5RMi9Q+1UiLiIg0WjuAPdiRUiIpT5bDsEm1C9iGLfkQkdqmRFpERKRR8mATaYfgtdAubKv1zvoKSqRZUWmHSL0zlNcxRqDzWRGpmcKyW2WHcm+SnVMvEYk0N0qkReqNBzs5xQ4gr2yZC2gNJGMvy4qIVNfBap8dymeWFJHapERapF54gPWUX151YQ9uxcBv2AS7GxAdkuhEpDGKwLY2l1D5MHelQIt6iUikudE1ZZF6kYlNoiOwyXIENpl2l/2dB6zDJtwiIlXhApKwiXKwFufSsv+3rpeIRJqbBpVIP/DAAziOwx//+Meg2yxYsADHcSrcVq5cWX+BNkol2ERuK7AdKAhtOM1KCbacI4zAF4EcbEK9D8iux7hEpPFLwY7OkY//ibjB7nsKgXggsd4jE2kOGkxpx7fffsuTTz5J//79q7T9qlWriI+P9/2dnJxcV6E1cgbbGrodO5aod5kLu2NNQ7W5dW0v9r13V7JNOPZzyQZa1kdQItIkRAHdsVe08iifjMVgT95bAp3RDIcidaNBtEjn5uZywQUX8NRTT9GyZdWSiJSUFNq2beu7hYdrJ1GRATZja3BLKC8jiMbuaLOA1dg6Xak73suuB/u5OdjPqTkqwZ5w5FB+wiciVRMD9MEm1G2w5R7tgF5lyyJCF5pIE9cgEumrr76asWPHcsIJJ1T5PoMGDSI1NZVRo0bx6aefVrptYWEhOTk5frfmIRfbEu2txfV+3A52xxpVts22kETXfHg7Fh6s/tk7HF5zUow92fsFWAmsKvu3t3VNRKrG2/qcDnTFXm2MQ7MZitStkCfSL7/8Mt9//z0PPPBAlbZPTU3lySefZM6cOcydO5eePXsyatQoPvvss6D3eeCBB0hISPDd0tLSaiv8Bi4L2xoarIInDHu5byflHVKk9sVhT2Qqa2ktxX4eifURUANRBPwKbMG2SEdSXma0o2xdbmhCExERqQLHGBOywSU3b97MkCFD+OCDDxgwYAAAI0eOZODAgUyfPr3KjzNu3Dgcx2HevHkB1xcWFlJYWOj7Oycnh7S0NLKzs/3qrJuen7EdTSqrzS3Ftgr2AWLrI6hmaiuwCXtSc2Crswfb+TMeeym2ubQgrcWe7EVR8ZzeYDtPxWK/myE/5xeRBi4nJ4eEhIRmcGyXhiSkR6fvvvuO7du3M3jwYFwuFy6Xi4ULF/LII4/gcrkoLa1aK+lRRx3F6tWrg653u93Ex8f73eRAGqy/brUpu5ViSxaKsCcw+diTnTigC80niS4A9hB8ZkfvSCZ5aCQTERFpqEI6aseoUaP4+eef/ZZNmjSJXr16ccstt1S5A+HSpUtJTU2tixAbuRgOXmfqLf2IqvtwmjUH6AgkYFthvXX6sdhZDVvRgAbRqQe52O9eZd8770gme9FIJmCMIXvFRjbPX8yuH9ZggJZ900kbexQt+3fBcZrLSZiISMMR0iN3XFwchx12mN+y2NhYkpKSfMtvu+02MjIyePbZZwGYPn066enp9O3bl6KiIp5//nnmzJnDnDlz6j3+hi8JO2NeCYE/au84o22DrJfa5WBroBOx731VRvJoqrwdL6uS/Kl+31NayvJHX2fdCx9TvDeP8ChbS77zu1/ZMOcz0s88hn43n0dYhH7HIiL1qcHvdTMzM9m0aZPv76KiIqZOnUpGRgbR0dH07duX+fPnM2bMmBBG2VAlYFs6sygfO9qbuJRiSwqisSUHUr8cmk8ZRyDeOnEPwU8mvOVGGud8zaz3Wf30O0TExxDfvb2v9dkYQ9HuXNa+8DHhMVEcdsPZIY5URKR5CWlnw1BpXh0SSrGd3Lwt015h2LKCzthkWqQ+lQLLsCdzwco7irHJdB+a83e0KGcfH512JyU5+4hpH3ia5/xtu3Fc4Yx6/V6iU1QGI81T8zq2S0PRXK8rNyPh2GS5D7ZGNxVoD/QEetOcExQJpXBsSRHYjpcHns+XYBPpVjT37+jWhT9SsG0XUSmJQbeJap1A0a69bPno+/oLTEREGn5ph9QW74yGIg1FMjZZ3oodvSQMW+7iHVO7Nfbkr3kr2L4HcCqtf3bCw8BxyrYVEZH6okRaRELEwV4daYmdFGgvtmU6BttRVrOyAYRFRmCMwRgTdGQO7/pwd3ObGVNEJLSUSItIiMWU3SSQ1kN6EBEbRXFOHpEJgSdNKtlXQLg7gqQhPeo5OhGR5k010iIiDVhCr460PqIn+Vt34SmpOBSgp9RDXuZOWh7WmdaDlUiLiNQntUiLiDRgjuPQ/88XkLdlJ9krNxOZGEtkyxYAFO3ZR9HuvcR1acfAuy/GCVPbiIhIfdJeV0SkgWuRlsLwJ26g24QTCXNHkLdlJ3lbduKEh9HlD6MY/sQNJHTvEOowRUSaHY0jrbEmRaQRKdyVw951mRhjiEtvS1RyYqhDEmkQdGyXUFBph4hII+JuFY+7lZIEEZGGQKUdIiIiIiI1oERaRERERKQGlEiLiIiIiNSAEmkRERERkRpQIi0iIiIiUgNKpEVEREREakCJtIiIiIj8f3v3HhxVef9x/HPC5kJDEiBikoUEEDEhpXUGsBNFba0IHcU4IkEGjS0yjs7g4G3U6rQD7aho6wxKkXG0iTFaKSqI1xbBYbwMGS+EtLSOVH+N4ZqGICQESMjl/P54spGF3c1yyO5mz75fMzuw5zzn5NkvX3a/efY5z4EDFNIAAACAAxTSAAAAgAMU0gAAAIADFNIAAACAAxTSAAAAgAMU0gAAAIADFNIAAACAAxTSAAAAgAMU0gAAAIADFNIAAACAAxTSAAAAgAMU0gAAAIADFNIAAACAAxTSAAAAgAMU0gAAAIADFNIAAACAAxTSAAAAgAOeWHcAABJLt6TDko71Pk+TNEK8HQNA/OGdGwCi5pCkXZI6JNm92yxJeyV5JY3qfQ4AiAcU0gAQFYck/VdmRDpV38+s65F0QlJD7/Nzo981AIAjzJEGgIjrkRl17paZynHyW29S7zZJ2iepK7pdAwA4RiENABHXKum4pBQFn7qRIjMyfShanQIAnCUKaQCIuHaZOdFDQrRJ6m3THpUeAQDOHoU0AESFre8vMAyGCw0BIJ5QSANAxP1A5u22J0SbnpPaAgDiAYU0AERchkyBfPKydyeze/elShoevW4BAM4KhTQARJwlqUDmgsLjMqt3+Arqbpl50UMk5Sv0PGoAwGDCOtIAEBUZks6XuSHLMZmpHFbvY6ikMTJ3OAQAxAsKaQCImgxJxZKOyP8W4ZniC0IAiD8U0gAQVZZM4ZwZ644AAM4ShTQQlg5JByUd1ve3eM6WuTCMOa0AACQiCmmgXwclNUjqlPn63ZK5YOywpGGSJsgU1gAAIJEwKQ8IqVXStzKj0ENl5rOm9v49VWau6//17gcAAImEQhoI6X+SumSK5lPvOpckU1i3yYxOAwCAREIhDQTVLjMinazgt272/Rc6GJUeAQCAwYNCGgiqU2at3/4uJkySuRgRAAAkEgppICjff49At3Q+mS1W7gAAIPFQSANB+S4u7AzRxpYZtR4ejQ4BAIBBhEIaCCpJ0rkyxXJXgP22zJSOZJk1pQEAQCIZVIX08uXLZVmW7r777pDtPvzwQ02dOlVpaWk677zz9Oyzz0ang0hAo2SK5E6Ziw+7ZUagO2XWkk6SNFasIw0AQOIZNIX0559/rueee04//vGPQ7arr6/X1Vdfrcsuu0zbt2/Xww8/rCVLlmjdunVR6ikSS5Kk8TLFsm+axwmZ0eiRkiaK0WgAABLToLizYVtbm2666SY9//zzeuSRR0K2ffbZZ1VQUKCnnnpKkjRp0iR98cUXevLJJ3XDDTdEobdIPEmScmRGp4/LjEgnyxTWAAAgUQ2KEenFixfrmmuu0YwZM/ptW1NTo5kzZ/ptmzVrlr744gt1dga+KKyjo0Otra1+D+DMJUlKl5QhimgAABDzQvqvf/2ramtrtXz58rDaNzY2Kicnx29bTk6Ourq61NzcHPCY5cuXKysrq++Rn59/1v0GAABAYotpIb17927dddddevnll5WWFv4In2X532XOtu2A230eeughtbS09D12797tvNMAAACAYjxHetu2bWpqatLUqVP7tnV3d+ujjz7SqlWr1NHRoSFD/G90kZubq8bGRr9tTU1N8ng8ys4OfNFXamqqUlNZVQGInnaZ26YfkZlTPlTmoswMBb/dOgAA8SWmhfSVV16pHTt2+G1buHChioqK9OCDD55WREvSxRdfrLfffttv2/vvv69p06YpOTk5ov0FEI7/Sdors8KJ70uvI5KaJY2QWQWFO0ECAOJfTAvpjIwMTZ482W9benq6srOz+7Y/9NBD2rt3r6qrqyVJd9xxh1atWqV7771Xt912m2pqalRRUaE1a9ZEvf8ATnVQkm/q1FD5jz539e63JJ0nRqYBAPFuUCx/F8r+/fu1a9euvufjx4/Xe++9p3vuuUfPPPOMvF6vVq5cydJ3QMz1SNqv76dynMr3dnNI0lFJw6LUL8Reu6Tvev+0ZP7tRygOPoIAICTL9l2pl0BaW1uVlZWllpYWZWZmxro7gEu0Stops8Z2sKkbtsxa3F5JrJ7jfj0y03yaZL6R8LEkpcjc6GhEDPoFN+KzHbEQ8+XvALjFCZnCKdT8Z6v30RGVHiHW9sh8SyGZbyl+0PtIlcmX/0pqiU3XAGAAUEgDGCC+t5P+vuSyxVtPIjgu6YDML1Yp8p8TnyRzU6MuSfvUf84AwODEpxmAATJMZlpH4DuMGj0yBRVfu7rfdzKFcrDVlHzTO9pk5swDQPyhkAYwQFIkjZQpnnoC7LdlpnSkinmxicB3YWGo1VmGyORKe1R6BAADjUumAQyg0ZKOyawbPUTfj0Z2y4xUp0gaJ9aRTgSWwp+ywVKIAOITI9IABlCypIkyq3IMkbmgrEOmoMru3ZcVs94hmoap/2K6S2Y8Jz0qPQKAgcaINIABliyztF2ezOi0LTOdIy2WnULUjZC5kNA3nefUUWdb5luKbJEbAOIVI9IAIsQjc1FhliiUEpHvF6okmTnQ3b3bfQX0cZkl8cbEpHcAMBAYkQYAREi2TCG9T+bbiRMyhbSnd98Y8UsWgHhGIQ0AiKARkobLXIDaITPFI12BbyMPAPGFQhoAEGGsHQ7AnZgjDQAAADhAIQ0AAAA4QCENAAAAOEAhDQAAADhAIQ0AAAA4QCENAAAAOEAhDQAAADhAIQ0AAAA4QCENAAAAOEAhDQAAADhAIQ0AAAA44Il1B2LBtm1JUmtra4x7AgAABoLvM933GQ9EQ0IW0keOHJEk5efnx7gnAABgIB05ckRZWVmx7gYShGUn4K9uPT092rdvnzIyMmRZVsA2ra2tys/P1+7du5WZmRnlHg4exIEY+BAHgzgYxMEgDsZgiINt2zpy5Ii8Xq+Skpi5iuhIyBHppKQkjRkzJqy2mZmZCf3m6EMciIEPcTCIg0EcDOJgxDoOjEQj2viVDQAAAHCAQhoAAABwgEI6iNTUVC1dulSpqamx7kpMEQdi4EMcDOJgEAeDOBjEAYkqIS82BAAAAM4WI9IAAACAAxTSAAAAgAMU0gAAAIADCVFIf/TRR7r22mvl9XplWZY2bNjgt3/9+vWaNWuWzjnnHFmWpbq6urDOu27dOhUXFys1NVXFxcV64403Br7zAygScaiqqpJlWac92tvbI/MiBkCoOHR2durBBx/Uj370I6Wnp8vr9eqWW27Rvn37+j1vPOVDJGLgtlyQpGXLlqmoqEjp6ekaMWKEZsyYoU8//bTf88ZTLkiRiYMb8+Fkt99+uyzL0lNPPdXved2WDycLNw7xmA9AOBKikD569KguvPBCrVq1Kuj+6dOn6/HHHw/7nDU1NbrxxhtVXl6uf/zjHyovL9e8efPC+pCNlUjEQTIL8O/fv9/vkZaWNhBdjohQcTh27Jhqa2v129/+VrW1tVq/fr3+85//qLS0NOQ54y0fIhEDyV25IEkXXHCBVq1apR07duiTTz7RuHHjNHPmTB04cCDoOeMtF6TIxEFyXz74bNiwQZ9++qm8Xm+/53RjPvicSRyk+MsHICx2gpFkv/HGGwH31dfX25Ls7du393ueefPm2b/4xS/8ts2aNcueP3/+APQy8gYqDi+88IKdlZU1oH2LplBx8Pnss89sSXZDQ0PQNvGcDwMVg0TIhZaWFluSvXnz5qBt4jkXbHvg4uDWfNizZ489evRo+1//+pc9duxYe8WKFSHP49Z8ONM4xHs+AMEkxIh0JNTU1GjmzJl+22bNmqWtW7fGqEex09bWprFjx2rMmDGaPXu2tm/fHusuDaiWlhZZlqXhw4cHbeP2fAgnBpK7c+HEiRN67rnnlJWVpQsvvDBoO7fnQrhxkNyXDz09PSovL9f999+vH/7wh2Ed48Z8cBIHyX35AEgJMrUjEhobG5WTk+O3LScnR42NjTHqUWwUFRWpqqpKb731ltasWaO0tDRNnz5dX3/9day7NiDa29v161//WgsWLFBmZmbQdm7Oh3Bj4NZceOeddzRs2DClpaVpxYoV2rRpk84555yg7d2aC2caBzfmwxNPPCGPx6MlS5aEfYwb88FJHNyYD4AkeWLdgXhmWZbfc9u2T9vmdiUlJSopKel7Pn36dE2ZMkV/+tOftHLlyhj27Ox1dnZq/vz56unp0erVq/tt78Z8OJMYuDUXrrjiCtXV1am5uVnPP/983/zWc889N+gxbsyFM42D2/Jh27Ztevrpp1VbW3vG/5ZuygencXBbPgA+jEg7lJube9qIQlNT02kjD4kmKSlJF110UdyPMnR2dmrevHmqr6/Xpk2bQo7ESu7MhzONwanckgvp6ek6//zzVVJSooqKCnk8HlVUVARt78ZckM48DqeK93z4+OOP1dTUpIKCAnk8Hnk8HjU0NOi+++7TuHHjgh7ntnxwGodTxXs+AD4U0g5dfPHF2rRpk9+2999/X5dcckmMejQ42Laturo65eXlxborjvkKyK+//lqbN29WdnZ2v8e4LR+cxOBUbsiFQGzbVkdHR9D9bsuFYPqLQ6D28ZwP5eXl+uc//6m6urq+h9fr1f3336+NGzcGPc5t+eA0DqeK93wAfBJiakdbW5u++eabvuf19fWqq6vTyJEjVVBQoO+++067du3qWyd3586dksxIQm5uriTplltu0ejRo7V8+XJJ0l133aXLL79cTzzxhK677jq9+eab2rx5sz755JMov7rwRSIOv/vd71RSUqKJEyeqtbVVK1euVF1dnZ555pkov7rwhYqD1+vV3LlzVVtbq3feeUfd3d19o0kjR45USkqKpPjPh0jEwG25kJ2drUcffVSlpaXKy8vTwYMHtXr1au3Zs0dlZWV9x8R7LkiRiYPb8qGgoOC0XyiTk5OVm5urwsLCvm1uzwencYjHfADCEqvlQqJpy5YttqTTHr/85S9t2zbL8gTav3Tp0r5z/PSnP+1r7/Paa6/ZhYWFdnJysl1UVGSvW7cuei/KgUjE4e6777YLCgrslJQUe9SoUfbMmTPtrVu3RveFnaFQcfAt/RfosWXLlr5zxHs+RCIGbsuF48eP29dff73t9XrtlJQUOy8vzy4tLbU/++wzv3PEey7YdmTi4LZ8CCTQsm9uz4dAwolDPOYDEA7Ltm17IApyAAAAIJEwRxoAAABwgEIaAAAAcIBCGgAAAHCAQhoAAABwgEIaAAAAcIBCGgAAAHCAQhoAAABwgEIaAAAAcIBCGkBULFu2TJZlqbm5OeD+yZMn62c/+1nf82+//VaWZcmyLC1btizgMbfeemtfm2CmTJkiy7L05JNPBtxfVVXVdw7LsuTxeDRmzBgtXLhQe/fu7fd1VVdXa/78+SosLFRSUpLGjRvX7zEAAHegkAYwqGVkZKiqqko9PT1+29va2vTaa68pMzMz6LF1dXXavn27JKmioiLkz3nhhRdUU1OjTZs26bbbbtOaNWt02WWX6ejRoyGPe+mll/Tvf/9bP/nJTzRhwoQwXxUAwA0opAEMajfeeKMaGhr0wQcf+G1fu3aturu7VVpaGvTYP//5z5Kka665Rl999ZW2bt0atO3kyZNVUlKiK664QkuXLtUDDzyg+vp6bdiwIWT/Nm7cqB07duill15SYWFh+C8MABD3KKQBDGqFhYW65JJLVFlZ6be9srJSc+bMUVZWVsDj2tvb9corr2jq1KlasWJF3zHhKikpkSQ1NDSEbJeUxNsoACQqPgEADHqLFi3Shg0bdOjQIUnSzp07tXXrVi1atCjoMevXr9ehQ4d06623auLEibr00ku1du1atbW1hfUzv/nmG0nSqFGjzv4FAABciUIawKA3b948eTwevfLKK5LMfOfx48f7XZx4qoqKCqWlpWnBggWSTDHe1tamV199NWD77u5udXV1qa2tTe+++64eeeQRZWRkhJw6AgBIbBTSAAa9YcOGqaysTJWVlerq6lJ1dbUWLlwYdLWO+vp6bdmyRXPmzNHw4cMlSWVlZcrIyAg6vaOkpETJycnKyMjQ7NmzlZubq7/97W/KycmJ1MsCAMQ5T6w7ACAxeDzm7aa7uzvg/q6uLiUnJwc9ftGiRbr00kv16KOP6sCBA/rVr34VtG1lZaVs29bcuXN1+PDhvu2lpaX6y1/+oq+++kpFRUV+x1RXV2vSpEnyeDzKyclRXl5e+C8OAJCQKKQBRIVvZHfv3r2njfLatq39+/dr2rRpQY+fPn26CgsL9fvf/15XXXWV8vPzA7br6elRVVWVJGnOnDkB21RWVuoPf/iD37ZJkyaF/PkAAJyKQhpAVPz85z+XZVlau3atpkyZ4rfv73//u1pbWzVjxoyQ5/jNb36j119/XYsXLw7aZuPGjdqzZ48WL16suXPnnrb/zjvvVHV1tR577LG+UXIAAJzgUwRAVEyYMEF33nmn/vjHP+rw4cO6+uqrNXToUH3++ed6/PHHNW3atL4LA4O5+eabdfPNN4dsU1FRIY/Ho4cfflher/e0/bfffruWLFmid999V9ddd91ZvSZJ+vLLL/Xll19KkhobG3Xs2DG9/vrrkqTi4mIVFxef9c8AAAxOXGwIIGqefvpprV69WrW1tVqwYIGuvfZavfjii1q8eLG2bNmilJSUszp/c3Oz3n77bc2ePTtgES1J5eXlGjp0aL93OgzXq6++qrKyMpWVlWnbtm06cOBA3/NgK4QAANzBsm3bjnUnAAAAgHjDiDQAAADgAIU0AAAA4ACFNAAAAOAAhTQAAADgAIU0AAAA4ACFNAAAAOAAhTQAAADgAIU0AAAA4ACFNAAAAOAAhTQAAADgAIU0AAAA4ACFNAAAAODA/wMDTWf81XtkmgAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABKwAAAVKCAYAAADEzeNiAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAAEAAElEQVR4nOzdeZyN5f/H8feZfcaYxTa2wYhEtlBCso8kJSklW6FsUSqllGmjVKKUUiRtaF8oZGtBlpQkaRlrxm5GlhlzzvX7w/fcvzlmMTMO5z4zr+fjcT8493Ldn/s++3uu+zoOY4wRAAAAAAAAYBMBvi4AAAAAAAAAyIrACgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArRBYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFQIrAAAAAAAA2AqBFQAAAAAAAGyFwArwczNnzpTD4VC1atV8XYqHpKQkORwOtW7d2telFAmtW7eWw+FQUlKSr0vJ0bJly+RwOORwOAq1fV6P4+L8WKpWrZocDodmzpx53ved13m3w33iy3MDeNN3332nzp07q2zZsgoMDJTD4VDXrl19XZZX2fk9bOvWrdb719atW31dzlm/nwJAUUJgBXiR+0vc6VNYWJgqV66sa6+9VnPnzpUxxtel4jTuD9P5mYpjcILCO3nypGbMmKGrr75alSpVUmhoqKKjo3XhhReqTZs2Gj16tL766iudOHHC16UWGUlJSUpKSvLZl8+sXziXLVt2xvXd4Vu/fv3OeW0FMWnSJCUlJennn3/2dSlF1qpVq9S2bVvNnz9fBw4cUKlSpRQXF6fY2NhCtffDDz9o+PDhatiwocqWLavg4GDFxsaqQYMGuvPOO7V48eIi9Rkk63Pt9Ck0NFSVK1fWNddco9mzZxep4waA4iLI1wUARVVcXJz1/9TUVO3atUu7du3SF198oZkzZ+qTTz5RaGjoWe8nOjpatWrVUqVKlc66LW8qU6aMatWqpSpVqvi6lAIJDg5WqVKl8lznTMvhXf76WJKkHTt26Oqrr9bGjRuteSEhIQoMDNTff/+tP//8U8uWLdPTTz+tpUuXZgtDL7jgAoWFhSk6Ovo8V27/857XuXnssccknQqi7db71J9MmjRJ27ZtU7Vq1dSwYUNfl1MkTZo0SZmZmWrRooU+//zzQr+/pKSkqF+/flqwYIE1LyAgQNHR0frvv/+0YcMGbdiwQdOmTdMll1yiuXPnqkaNGt46DFuIjY1VSEiIdfvQoUPWZ6958+Zp5syZ+uyzz7zy2QsAcH7Qwwo4R1JSUqzp6NGj2rhxozp06CBJ+uqrrzRmzBiv7Of666/X5s2btXjxYq+05y3Dhg3T5s2bNWvWLF+XUiDNmzf3uO9ymj7++GNfl1ms+Otjyel06rrrrtPGjRsVERGhJ554Qtu3b9eJEyd08OBBHTt2TKtXr1ZSUpIuuOCCHNtYvHixNm/erOuvv/48V2//8+7LcwN4y6+//ipJuvnmmwsdViUnJ6tJkyZasGCBgoODNXjwYP3444/KyMjQwYMHlZ6err///luTJ09WQkKC1q9f7xGiFxUff/yxx3v1iRMn9Oeff6pnz56SpAULFmjixIk+rhIAUBAEVsB5EBAQoIsvvliff/659RfN1157TZmZmT6uDMC5smTJEq1fv16SNH36dI0ZM0bx8fHWuCShoaG69NJLNXbsWP35559q1qyZL8sF4APHjh2TJEVGRhZq+/T0dHXr1k27du1SyZIltWjRIr3yyiu67LLLFBgYKElyOByqXr26hg8fri1btmjs2LHWsqLM4XCoRo0amjVrli688EJJ0qeffurbogAABUJgBZxHYWFhuvHGGyVJR44c0ebNmyWdumRw9uzZuvXWW1WvXj2VKlVKYWFhqlq1qnr27KlVq1bl2mZBBqv+6KOPlJiYqHLlyikgIEBJSUlau3atHA6HgoKClJqamq2NO++80xoP4ssvv8y2/P3335fD4ch22dCZBmVesGCBunXrpsqVKyskJERRUVGqXr26EhMT9dxzz+ngwYM5bnfixAm9+OKLatWqlcqUKaOQkBCVL19eXbt21ddff53reTqXTh+wddu2bRo4cKCqVKmisLAwXXDBBRozZoyOHj1qbbNx40b16tVL8fHxCgsLU82aNfXkk0/q5MmTZ9xfRkaGnn76adWvX18lSpRQbGysOnTooK+++uqM2/7999+66667VLt2bUVGRioiIkK1a9fW3Xffre3bt+e57ebNm3XrrbeqfPnyCgsLU/Xq1XXXXXdpz549Zz5JOjVWS9euXVWmTBmFh4erVq1aevjhh/Xff//luV1ej6V+/fp5jP3z4YcfqnXr1ipVqpQiIiLUsGFDTZ48WS6XK9f2jTF688031axZM5UsWVLR0dFq2rSppk2bJmNMtn3kV9Zxf6677ro813WPt3K6vAYWzzpG0oEDBzRy5EhdcMEFCg8PV9WqVTVs2DDt27fPWn/btm0aPHiwEhISFBYWpipVqujee+/VkSNHcqypsAOrHz9+XJ9//rkGDhxojaMTGhqqihUrqmvXrnk+Tk9/PVu6dKm6du2qChUqKDAw0OM+yOncuO8rtzZt2niMaeNu98EHH5TD4dDFF1+c57GkpaUpMjLSFoO7p6am6qmnnlLTpk0VGxur0NBQxcfH65ZbbsnzPeKPP/7Qs88+q/bt21uPj6ioKF1yySUaM2aM9u/fn20b932/bds2SdJtt92WbXwgt/Px+nc275GnP47nzp2rVq1aqVSpUipRooQaN26sKVOmyOl05nn+z2T9+vXq06ePqlatqrCwMMXGxqp58+aaNGmS0tPTs61/+iDfp5/j/I6/NmPGDOu15uWXX1arVq3yXD8oKEhJSUnq3LlztmUnTpzQpEmT1Lx5c8XGxlrnuE+fPmccx8zpdGrKlClq1KiRSpQooVKlSql169b68MMP83Uc0qlzePvtt+uCCy5QRESEIiMj1aBBg1wfp/kVGBioevXqSdIZ329yk5qaqscff1yNGjVSVFSUwsPDVbNmTQ0ePFj//PPPGbf/8ccfddttt6lGjRoqUaKEoqKiVKdOHd1+++1auHBhgWo5cOCAmjVrJofDoYSEBG3ZsqVQxwQAfsEA8JqxY8caSSavp9bLL79srfPDDz9k206SiYyMNKGhodZth8NhJk+enGN7b775ppFkqlatmms9rVq1MiNHjrTaio2NNYGBgWbs2LHG6XSamJgYI8l89tln2dqoUaOGVcc999yTbfmAAQOMJNOnT59c9326xx57zON4IyIiTGRkpMe8pUuXZttuy5YtpmbNmh7nJTo62mO7wYMH53iezqRVq1a51nsmycnJ1v4/+ugj63xGRUWZwMBAa1nLli1NRkaG+fLLL01ERISRZKKjo43D4bDW6dGjR571jR492rRs2dJIMkFBQda+3NPYsWNzrXPatGkmODjYWjc0NNSEh4dbt6OioszChQtz3Parr77yeExGRkaasLAwI8lUqFDBzJgxI8/H/vTp001AQIC1TnR0tAkJCTGSzEUXXWQmTpyYr8fx6fr27Wskmb59+5qhQ4caSSYgICDbeTn98emWmZlpevTo4fGYio2NtWq95ZZbPPZREBMmTLDa3bJlS4G2datataqRZN58881sy9xtv/XWW6Zy5cpGkilRooR1XiWZ2rVrm0OHDpnVq1ebMmXKWPdzUFCQtU6LFi1MZmZmtvbzOu95LXO/Jrmn8PBw6/Hunu69994cjzfr69nkyZOt50Z0dLQJDg72uA9yOjfDhw83cXFx1n5iY2NNXFycNTVp0sQYY8w///xjtf3dd9/lev6nTp1q7f/o0aO5rne6pUuX5vladjr3seT2GFu1apXHcQUGBpqSJUt6PG7HjRuXZ9vu9WJiYjxecypVqmQ2b97ssc2zzz5r4uLirOdBVFSUx3mMi4uz1j0fr39n8x6Z9bE6atSoHJ/nkkzHjh3NiRMnznhf5eSFF17wOA7349V9u379+ubff//12MZ9HnM7x9u3b8/XvuvUqWMkmRo1ahiXy1Wo+o0xZufOnaZu3bpWzcHBwR7vrwEBAebFF1/McdsTJ06Yjh07eqyb9XH2wAMPWO9hub1HPfroox7nMCIiwuO1rEKFCuann37Ktl1+nmtOp9PUqlXLSDLXX399tuVZH8PJycnZlm/cuNF6jZVkwsLCPJ5/oaGh5sMPP8xx35mZmWb48OEej98SJUp4vCZGR0fnekyn27p1q7nooouMJNOgQYNsjysAKGoIrAAvyk9gdf/991vr/P7778aYU1+K7rnnHrNq1Spz6NAhY4wxLpfL/PPPP2bEiBHG4XCYwMDAHD+s5SewcodBo0aNMnv37jXGnPqAuXXrVmOMMddee62RZEaMGOGx/Y4dO6wP0u4PR6e74IILcvxCndsX2q1bt1of0EeOHGl27dplLTt8+LD57rvvzJAhQ8zatWs9tjt06JCpVq2akWTatm1rvv32W+vLxeHDh83EiROt45w0aVK2Os/EW4FVTEyMadeunfntt9+MMcYcO3bMvPjii9YXtzFjxpjo6GjTo0cP6/wfOXLEPPzww1YbixYtyrW+6OhoExoaal599VVz/PhxY4wx27dvN927d7e2zyl4/OSTT6wvIQ8++KDZunWrcblcxuVymc2bN5sbb7zRuq+3bdvmse2OHTusx0D9+vXNjz/+aIw59SXgq6++MpUrV/YIiE63bt06KyBp3bq19bjPyMgw77//vomJibG2L2xgFRsba0JCQszEiRNNamqqMcaY/fv3W4GqJLN48eJs248fP95aPnLkSLN//35jjDGpqalm3Lhx1hfbwgRWy5Yts9pu27at2blzZ4G2NyZ/gVVMTIxp2LChWbVqlTHm/8+r+wvRsGHDTNWqVU3btm3Nxo0bjTHGHD9+3Lz00kvW4/L111/P1n5hA6tPPvnE3HHHHWbp0qXW+TTGmH///dc89thj1hf5nB6n7tezsLAwExgYaPr162d9cc/MzDR//fVXgc5NXmHRVVddZaTcw0xjjGnUqJF1DgvCm4FVcnKy9fzo3r27WbdunTl58qQxxpg9e/aYRx55xHp+ffLJJ9m279Gjh3nppZfMX3/9ZdLT040xxqSnp5tvvvnGXHbZZUaSadSoUZ515XSOs9Z3rl//zuY90v1YdYcvw4YNs94HU1NTzRNPPGEFJTn9UeZMvvjiC6v26667zvzzzz/GmFPneNasWVaw0bx58xyD4fyc49zs3r3b2vd9991X4O3dMjMzTdOmTa3z9M4771iPlb///ttcc8011n7mz5+fbft77rnHCgKffPJJ6zV4z549ZvDgwR7nP6fA6oUXXjCSTMmSJc348ePN7t27rbrWrl1r2rZtaySZypUrmyNHjnhse6bn2t9//2169+5tpFNB77fffpttnbwCq7S0NJOQkGCkU+HuvHnzjNPpNMYY8/PPP5vLL7/cCq1+/vnnbG27Q1JJ5vbbbzd//PGHtWzPnj3m008/zRbU5hZYbdiwwVSsWNFIMm3atLHOMwAUZQRWgBedKbBKTU21PmyUKlXK+tBzJu6eI/3798+2LD+BlfvLeG4mTZpkJJl69ep5zH/rrbes/VaqVMk4HA6PL6Dbt2+32nd/+Th936d/oZ0zZ46RZC688MJ8HPn/u++++6wv/u4va6f7+OOPjSRTpkyZXNfJjTsQCg4OztaT4PRp9uzZHttm/bB78cUX5/hXevcHZkmmQ4cOOf4l3N1zKqf72V2fJDN9+vRsy51Op7nyyiuNJFOnTh2PZenp6aZSpUq5buuWW3Dp/sJRunRps2fPnmzb/frrrx69CU7XqVMn6z4/duxYtuVff/21tW1hA6u8vvA1btzYSDIDBgzwmH/06FEriMvpnGfdd2ECK2OM6dChg7V9YGCgadasmbn77rvN22+/na9eV/kJZeLi4jyel26PPPJIvh+X7dq1y7assIHVmTz77LO57jNr76xu3brl2c7ZBlaffvqpkU71AHOHIFmtW7fOamfDhg1nOiwPWb9wnt7LK6fJHeLn9Bhzh9G9e/fOdX/uHoo5/VEhL0eOHLF6buXU06yggdW5ev07k7zeI7M+h3M7h2PGjDHSqV6rWf+Ikh/uHk5XXHFFjoHU559/bu3/gw8+yLb8bAKrb775xmr73XffLfD2brNnz7ba+frrr7MtP3nypBVo1a1b12PZrl27rMD0kUceybH9W265xWr/9MBq3759JiIiwjgcDvPNN9/kuP3Jkyet1/EXXnjBY1lezzV3L7zAwECTmJiY6+tBXoHV008/bX02+PXXX7Ntm5aWZv0xrXPnzh7L/vjjD+u5PWrUqBz3nZOcAqtly5ZZod+NN95Y6N6AAOBvGMMKOA8OHz6sxYsXq23btvr3338lSSNGjFBAQP6egu6xJr7//vtC7T8gIEAPPPBArsvbtGkj6dS4IlnHvFm6dKkkqW3btmrTpo2MMdY86dSg0pKUkJCgqlWr5quWmJgYSafG8Mo6pklejDGaMWOGJOnee+9VUFBQjut17dpVUVFR2r9/v9atW5evtk938uRJ7dmzJ8/p+PHjuW5/zz335DgWUceOHa3/u8fPyW2dDRs25Np+fHy8brvttmzzAwICrF+e3LRpk/XLU9KpX6XctWuX4uLictzWrU+fPpLk8bPoxhjNmTNHkjRo0CCVK1cu23Z169ZV9+7dc2zz8OHDVnv333+/wsPDs63TsWPHsx5wPD4+3qr/dNdee62k7Od1wYIFSktLkyQ9/PDDOW577733KiIiotB1ffLJJxoyZIiCg4PldDq1cuVKTZo0Sb1799aFF16oatWq6bHHHrPqKIyBAweqdOnS2eZnfcyNHDkyz8dlXo85b3O/nq1cuTLPcYNGjx59Tuu45pprFB8fr+PHj+vtt9/OtnzatGmSpGbNmlnj3xTGoUOHzviaktsYawcPHrR+lfTBBx/MdR/ux/4vv/yS7zHlpFMDfbvHPCrs+0tW5/r1Lzf5fY989NFHc5zvfm3KzMzURx99lO/9btiwQZs2bZIkPfLIIzkOZN6lSxdddtllkk6N+ehNBw4csP5f2F8YlGS9xjdr1szjvnILCgrS2LFjJZ36nJD1/eXDDz9UZmamwsPDdd999+XYflJSUq77fvfdd3Xs2DE1adJE7dq1y3GdoKAg3XLLLZI8359Od/pzzT12mNPp1MGDB7Vr165ct82N+9x0795ddevWzba8ZMmSGjVqlKRT77VZxwJ966235HK5VLp0aT322GMF3rfbRx99pI4dOyo1NVXDhg3T7Nmzc3yeAUBRRGAFnCNZB0+NjY1V+/btrRClV69e2b4g//PPP7rvvvvUuHFjxcTEKDAw0Nr+6quvliTt3LmzULXUqFEjx6DBrV69eipTpky2QOr0wEr6/5Aq6//dy/LjsssuU5kyZbR79241bdpUU6ZM0ebNm2WMyXWbTZs2WYOw9+vXT+XLl89xqlChgjWgqnuw4IJq1aqVzKnep7lOeQ2+7f5icrq4uDjr/5deemme6xw6dCjX9lu3bp3jlz1JuvLKK60wb+3atdZ895e4Q4cOqUKFCrmev4EDB0ryPHfJycnWuW/btm2udeW27KeffrK+jBdm+/y69NJLcw2AK1asKEnZBvL/6aefJElVqlRRQkJCjtuWLFlSjRs3LnRdJUqU0Msvv6ydO3dq2rRp6t27t2rXrm19sd22bZuSkpLUsGFD/f3334Xax7l+zBXGnj17NHbsWDVr1kylS5dWUFCQ9XpWp04dSad+HS23/YaHh6tRo0Zerel0gYGBGjBggCTp9ddf91h29OhRK1y44447zmo/S5cuPeNrSm6B/8qVKz2eP7k9d7MOHp/Ta9+XX36pHj16qHr16ipRooTH+9PcuXMlFf79Jatz+Vg82/fI+Ph461d6TxcVFWU9z7O+dp6Je92goKA8Bzvv0KFDgdvOj6zvm7m9L+SHu6727dvnuk6bNm2s162sx+H+f5MmTRQVFZXjthdeeKEqVaqU4zL3+9PGjRtzfXyXL19ejz/+uKS839tPf66dPHlS//zzj55++mlroP+8gt/TZWRkWAFqXufGff+6XC7rfUWSVqxYYS0PCwvL936zevnll3XTTTcpPT1dTz31lF566aV8/7ETAIqCnLspADhrWT+gh4aGqkyZMrrkkkt06623Zgt4PvnkE91yyy0evyQUFRWlsLAwORwOZWRk6NChQ/nukXS6vMIqSdYvKH344YdasmSJbrrpJv3zzz/atm2bateurfLly1uBQtbAyh1oFSSwiomJ0fvvv6+ePXvqt99+01133SVJio6O1pVXXqmbbrpJPXr0UHBwsLWNu1eaJI8eYHlx/1T4+VayZMkc52ftFXamdfL6pcDcPvRLpx5npUuX1p49e7R3715rvvv8ZWRk5Kv3RdYeZFnbyWvflStXznH+2W6fX7mdUyn38+p+LLkDrdzkVXd+lStXTgMHDrRCwf/++09LlizRs88+q++//17Jycm6+eabtWbNmgK37Y3HXGZmZoH3m5uVK1fq6quv1uHDh6157l+kdDgccjqd1i9+HT16VGXKlMnWRunSpc/Ll7IBAwboiSee0K+//qpVq1bp8ssvlyTNnj1baWlpiomJ0U033XTO68hN1te+/Pacyvra53K51KtXL4+ePUFBQYqNjVVISIikU79+duLEiUK/v2R1rl7/vPEeeabnsXt51tesM3GvW6ZMmTx7vLhf3wrSdn5kfe5k7W1VUO668jpHYWFhKlOmTLb3l/xsK506Bzn1cHI/xo8fP55n72W3gry3BwUFKSEhQQ888IAiIiI0fPhwPfPMM+rUqdMZf01ROvVHDncv0Py+f2U9NykpKZKU7x7oORk2bJgkaejQoXrooYcK3Q4A+CsieuAcSUlJsaZt27Zp3bp1euONN7KFOwcOHFC/fv2Unp6utm3batmyZTp27JhSU1O1Z88epaSk6IMPPjirWnK6TOF0p/egcv/rDqqqVaumhIQE/fHHH/r333/1119/aceOHR7b5lf79u2VnJysWbNmqW/fvqpZs6ZSU1P1xRdfqHfv3rrkkks8PthmvWwoJSXljL0VztQLyp8V5q/o7vN31VVX5evc5dbb7Wz+gm9H7uM803Hl1fuvsCIjI3Xttddq+fLl1vNn7dq1Z/zpeLvLzMzULbfcosOHD6thw4aaP3++0tLSdOTIEev1bNWqVdb6uZ3b/LxmeUPFihWtS0bdlwBK/9/jqlevXmd1SejZcj93w8PD8/3cbd26tbX99OnT9f777yswMFCPPvqo/vzzT6Wnp+vgwYPW+5P7ct5z8Tj3Bm+9R57L16/8tu3tGrL2rFu/fv1Zt3c2x1HYY3M/xgcNGpSvx/fWrVsLtZ/+/ftbNRbm0sy8ji/rMm+eG+nUa5B06jXpiy++KHQ7AOCvCKwAH3N/oYuNjdUXX3yhVq1aZRvnx/1XunPJHUz9+eef2rlzZ7bASvIMtdzL8+rqn5cSJUqod+/emjlzprZs2aKdO3fqmWeeUVhYmEfPK0kqX7689f+sY2cUR3ld8pKenm79lT1rrzr3+SvMucvaTl77zm1skKzb5zV+SGHGFjlb7tqy9mLJyZmWn42AgADrsjRJ+uOPP87Zvs6HlStXatu2bQoMDNSXX36pTp06ZetRcz5ezwpi0KBBkk6NVZOWlqZff/1VP/74o6SzvxzwbLmfu8ePH9dff/1V4O1nz54t6VRPsscee0w1atTI1nPNbvfH6bz1HnmmSx7dr0Fn6pGclXvdffv2efT+ym3fZcuWzXfb+VG+fHnrEtvPPvus0KGj+zjcf4TKyYkTJ6z3l6zH4d42v+f3dGfz/lQQERERVo+05OTkfG1TqlQpKzzP69xkXZb13FSoUEGSCh2ySafGwerbt68yMjJ0ww036NNPPy10WwDgjwisAB9zf9CpVatWrn/J/+abb855HRdddJH14WrJkiVatmyZAgICPP5an/WywMJcDpiXSpUqadSoUbr33nslSYsWLbKW1a1b1xobw/0FrLhavnx5rl9KvvvuO+vSriZNmljzW7RoIenUF4aCDqyckJBgDeabdXyz02W9VDSrRo0aWV+QC7P9ueQeI2nbtm25fqH477//Cj2Af35FRkZa//f3gXTdr2dly5bNNcg+H69n7h4N+fkC3759e9WoUUPHjh3Tu+++a/WuOtvB1r2hefPm1rEU5rXPfX9ccsklOS7/77//rHAuJ+7nri97X3nrPXLHjh25jhN35MgR63me9bXzTNzrZmZmavny5bmu564vt/G7zsbQoUMlnfpj0zvvvJPv7bIO9O8+jsWLF+e6/rJly6z3l6zH4d527dq1OnLkSI7buv8QlhP3+9OqVasKPfZkfmQN3EqUKJGvbUJCQlS/fn1JeZ8b9/0bEBDgMfZe8+bNJZ36PHPixIlC1R0QEKAZM2aof//+OnnypG666aYC/TAAAPg7AivAx6KjoyVJW7ZsyfEDzc8//6z33nvvvNTiDp9eeeUV7d69Ww0aNPD45aGsPayWLVvmMS+/8vortCTrL+dZLwkKCgrS7bffLunUXxvPFLqcPrh2UbJ9+3a99dZb2ea7XC6NGzdOklS7dm2PL9pdunSxwsgRI0accQyQrOfP4XBYY/i8+uqr1thDWW3atEkffvhhjm3FxMQoMTFRkvTcc8/l+Bj/5ptvrMFpz6fExEQrCHWfu9O98MILhR4PbePGjfnqOTZr1izr/7kFC/7C/Xrm/pWu0+3cuVMvvvjiOa/Dfb9mHUcrNw6HQ3feeaekU6997i/9vu5dJZ3qvXLddddJkp599llt2bIlz/VPf+1z3x+//PJLjus/8cQTuYYMUsHO47nizffIJ554Isf5zz//vI4fP66goCB169Yt37XVr1/f6uH05JNP5virl/Pnz7dCQfcv3XlT//79rdf7oUOH6ttvv81zfafTqccff1zz5s2z5t18882STvWQXLhwYbZtMjMzrUHP69at6/FreTfccIOCgoJ0/PhxPf/88znu071tTnr37q3w8HA5nU4NHTo0z18OdblchX4szp492wrpChJKus/Nhx9+qI0bN2Zb/t9//2nChAmSpKuvvtp6vEqnfiQmMDBQBw4csH5lsTACAgL0+uuv64477tDJkyd18803Wz+WAABFHYEV4GOJiYkKCAjQwYMHdeutt1pfcDMyMjR37lwlJibmOaC0N7nDJ/eH69N/ua1ixYqqVauWtm3bZl2CkbUHVn64Bzx9++23Pf7imp6errlz5+rZZ5+VJOtXn9weeeQRXXDBBcrMzNRVV12liRMnegzAnpqaqq+//lp9+/ZVy5YtC1STP4mOjtbgwYP1+uuvW1/eduzYoVtuucXqwfTUU095bBMWFqZXXnlFDodDP/30k1q0aKEFCxYoIyPDWic5OVmvvfaaLrvsMr3yyise248ePVolS5bU/v371aFDB+tXoYwxWrhwoTp16pTnOD9PPPGEAgMDtXnzZnXu3Nm67C0zM1Nz587VTTfdpJiYmLM+NwVVokQJPfDAA5JOjQ8yatQo6wv/kSNH9MwzzygpKUmxsbGFan/ZsmWqXr26evTooQ8++EC7d++2lp04cULff/+9rr32Wuuv5d27dz+rwXnt4IorrlCJEiVkjNFNN91kBSxOp1MLFizI81cuvcn9hfrdd9/NV+B42223KTQ0VBs3btShQ4cUExOjHj16nOsy8+X5559X6dKllZaWpiuuuEIzZsxQamqqtXz//v36+OOP1a1bt2yByFVXXSXp1ON72rRp1nM+JSVF99xzjyZMmKDSpUvnum/3efzwww+9/kuS+eWt98jo6Gi99dZbGjFihBW8HzlyROPGjbOCrKFDhxb4EvdnnnlG0qkert27d7cuNzt58qTeffdd6z5p3ry5unbtWqC28yM0NFSffPKJKlSooCNHjqh9+/YaOnSo1qxZ4xH+bN26Va+88oouuugijR071mPZDTfcoKZNm0qSbrrpJr333nvW4PfJycm64YYbtHLlSkmywhm3SpUqaciQIZJOvdaPHz/eCkH37dunYcOG6Z133vEIcrIqX768nn76aUnSvHnz1KFDB/3www9WfcYYbd68WRMnTlTdunX15ZdfFuj8HD161LrfpVMhbP/+/fO9/eDBg5WQkKCTJ0+qU6dO+uqrr6zg69dff1XHjh2VnJyskJAQPfnkkx7b1qhRQ/fff7+kU+dtwIAB+vPPP63l+/bt05w5c3T99defsQ6Hw6FXX31VQ4YMUWZmpnr27FmosbgAwO8YAF4zduxYI8kU9Kn1wAMPWNtJMtHR0SY4ONhIMgkJCebdd9/Ntd0333zTSDJVq1bNtZ5WrVrlq46///7bo4558+ZlW2fQoEHW8jp16uTaVm77znqOJJnw8HBTqlQp43A4rHm1a9c2u3fvztbmP//8Yxo0aOCxfUxMjImKivKYV6NGjXwdb1atWrUykkxwcLCJi4s745RVcnKyte/k5OQc21+6dOkZHxt53Zfu+kaPHm2uuOIKq9bY2FiPYx8zZkyu7b/zzjsmIiLCWjcoKMiULl3ahIaGerTx5JNPZtv2yy+/9FivZMmSJjw83EgyFSpUMDNmzMjz+F577TWP+zg6Otpq76KLLjITJ04s1OO4b9++RpLp27dvrsed13k9efKk6d69u1VXQECAiY2NNYGBgUaS6d27t+nTp4+RZO68885c95GTV1991eO8SjJhYWHZ7jNJJjEx0aSmpmZro2rVqkaSefPNN7Mtc2+7dOnSHPd/to/LvM57XsumTp3qcWyRkZEmLCzMSDJlypQxn3/+ea515XVfnS6vc/P2229b+wgODjaVKlUyVatWNS1atMi1vV69elnbDBs27Iz7z0vW85rb/ZOV+1hyexz/9NNPplq1alabDofDxMbGmsjISI9z3b59e4/tDh06ZC666CKPx3dMTIz1XLzzzjvzfA4tX77cWjcwMNBUqFDBVK1a1eP+OR+vf2fzHpn1sTpq1CjrPJQqVcp6nrvP3fHjx3OtLy8TJ070eH2LiYkxISEh1u169eqZXbt25bhtXo/jgti1a5dp166dx3lyH2fWWiSZpk2bmr///ttj+507d5qLL77YWickJMTExMR4tDV58uQc9338+HHTvn17a93AwEATGxtrnZMHHnjAeg8bO3Zsjm1MmDDB4/4ICQkxpUuXtu5n9/TOO+94bJf1sRUbG+vxPl26dGkTEBBgLS9VqlSOz8czPYZ//fVXU6lSJY/X8ayfO0JDQ80HH3yQ43FlZmaaoUOHZntNzPpeHB0dnesx5WT48OHWeZ41a1aO6wBAUUEPK8AGnn76ac2aNUuXXXaZwsPDdfLkSdWoUUMPPfSQ1q9fr4oVK56XOqpXr2718AgKCsqxp1JOg7AXxB133KFp06bplltuUd26dRUREWENqNuyZUtNmjRJP/30k8dA624JCQlau3atZs2apWuuuUYVKlTQ0aNHlZGRoYSEBF1//fWaMWOG9Zfgwjh58qR1OVNek6+EhIRo8eLFGjdunGrVqqX09HRFR0erXbt2mjdvXq6XvEjSrbfeqr/++ktjxoxRkyZNFBkZqcOHDyssLEwNGzbUsGHD9M0331i9jrLq3LmzfvrpJ918880qV66cMjIyFBcXp2HDhmn9+vVKSEjIs+477rhDP/zwg7p06aJSpUopPT1dVatW1ejRo7V69epC92I6W0FBQZo7d67eeOMN6/mXmZmpJk2a6I033tCsWbOsS1AK2gvszjvv1C+//KJnnnlG1113nWrUqKHAwEClpqaqZMmSqlOnjvr06aP58+drwYIF1uVX/m7QoEGaN2+eWrdurcjISGVmZqpSpUq666679Msvv5yXcaF69eqlt99+W1dccYUiIiK0e/dubdu2Lc+BoW+88Ubr/3a4HDCrSy65RJs2bdKUKVPUvn17lSlTRkeOHJHL5VLNmjXVs2dPzZ49Wx9//LHHdjExMVqxYoXuvvtuVatWTYGBgQoKClLr1q31/vvv69VXX81zv1deeaXmzZun9u3bKzo6Wnv27NG2bdvO6VhDOfHWe+Qzzzyj2bNnq0WLFnK5XAoJCVHDhg01efJkff311woLCytUfffcc4/Wrl2rXr16KT4+XseOHVN4eLguv/xyTZw4UatXrz7n7+MVK1bUN998o2+//VZDhw5V/fr1FRMTo7S0NIWHh6tBgwYaNGiQli1bplWrVql69eoe21eqVElr167VxIkTdfnllys8PFzHjh1TfHy8evfurXXr1mn48OE57jssLExfffWVJk+erIYNGyokJETGGLVs2VJz5861elDl5f7779fmzZt1zz33qH79+goLC9Phw4cVGRmpSy+9VKNGjdKKFSvUs2fPXNs4dOiQx/v04cOHFR0drWbNmunxxx/X5s2bC9wrXDrV0/C3335TUlKSGjZsqKCgIKWnp+uCCy7QoEGD9Ntvv1m/tnm6wMBATZkyRd9//71uvfVWValSRSdPnlRISIguvvhi9e/fv8BjUk2ePFkjR46U0+lUv379NHPmzAIfEwD4C4cxNv0dYwAAfMwYoypVqmjnzp2aNWuWevfu7euScI7cddddmjJlipo1a+aTMdVwbiQlJemxxx5Tq1atrLEXAQCAf6CHFQAAuXCPtRYUFKR27dr5uhycI2lpadbg94MHD/ZxNQAAAJAIrAAAxdwtt9yiDz/80OMXEPfs2aOnn35aAwcOlCT16dPnvF2ai/MrPT1dI0aMUFpamuLj420z2DoAAEBxF+TrAgAA8KWvvvpKs2fPliRFREQoODjY41fYWrZsqRdeeMFX5eEcmTRpkiZNmqS9e/fq+PHjkqSJEycqJCTEx5UBAABAoocVAKCYe/HFF3XzzTerVq1aCg0N1bFjx1S2bFl16NBB06dP1+LFi4vMgOj4f4cPH9a2bdtkjFHDhg01Z86cXAdOBgAAwPnHoOsAAAAAAACwFXpYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFQIrAAAAAAAA2AqBFQAAAAAAAGyFwAoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0QWAEAAAAAAMBWCKwAAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANgKgRUAAAAAAABshcAKAAAAAAAAtkJgBQAAAAAAAFshsAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCsEVgAAAAAAALAVAisAAAAAAADYCoEVAAAAAAAAbIXACgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArRBYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFQIrAAAAAAAA2AqBFQAAAAAAAGyFwAoAAAAAAAC2QmAFAMA5NnPmTDkcDq1du9bXpRRZe/bs0UMPPaSGDRsqKipKISEhqly5srp166bPP/9cTqfT1yUCAACgAIJ8XQAAAMDZWLVqla699loZYzR48GBdfvnlioyM1Pbt2/XFF1+oW7dueu2119S/f39flwoAAIB8IrACAAB+6/Dhw+ratasiIyP1ww8/qEKFCh7Le/XqpQ0bNujAgQN5tnP8+HGFhYXJ4XCcy3IBAACQT1wSCACADZw4cUL33nuvGjZsqOjoaJUqVUrNmjXTZ599lm1dh8OhYcOG6e2331bt2rUVERGhBg0a6Msvv8y27meffab69esrNDRU1atX1+TJk5WUlOQRzGzdulUOh0MzZ87McV9JSUnW7b/++ku33XabatasqYiICFWqVEldunTRr7/+mm3b3377TYmJiYqIiFDZsmU1dOhQzZs3Tw6HQ8uWLfNY95tvvlG7du0UFRWliIgItWjRQosXLz7jeXv99de1Z88eTZgwIVtY5Va/fn21adPGuu2+RHPhwoW6/fbbVbZsWUVERCg9PV0ul0sTJkzQRRddpNDQUJUrV059+vTRzp07PdqsVq2a+vXrl21frVu3VuvWra3by5Ytk8Ph0DvvvKORI0eqfPnyCg8PV6tWrbR+/XqPbf/55x/dfPPNqlixokJDQxUXF6d27drp559/PuN5AAAAKGroYQUAgA2kp6fr4MGDuu+++1SpUiVlZGTom2++Ubdu3fTmm2+qT58+HuvPmzdPa9as0eOPP67IyEhNmDBB119/vf744w9Vr15dkvT111+rW7duuvLKKzVnzhxlZmbqueee0549ewpd57///qvSpUvr6aefVtmyZXXw4EG99dZbatq0qdavX69atWpJknbv3q1WrVqpRIkSmjp1qsqVK6f3339fw4YNy9bmO++8oz59+ui6667TW2+9peDgYL322mvq2LGjFixYoHbt2uVaz6JFixQYGKirr766wMdy++23q3Pnznr77bd19OhRBQcHa/DgwZo2bZqGDRuma665Rlu3btUjjzyiZcuW6aefflKZMmUKvB9Jeuihh9SoUSO98cYbSk1NVVJSklq3bq3169db99fVV18tp9OpCRMmqEqVKtq/f79WrFihw4cPF2qfAAAA/ozACgAAG4iOjtabb75p3XY6nWrXrp0OHTqkSZMmZQusjh8/rm+++UYlS5aUJDVq1EgVK1bU3Llz9eCDD0qSHn30UVWqVEkLFixQSEiIJOmqq65StWrVCl3nlVdeqSuvvNKjzs6dO+viiy/Wa6+9pokTJ0qSXnjhBR08eFDffvut6tSpI0nq1KmTrrrqKm3dutXa/tixYxoxYoSuueYaffLJJ9b8q6++Wo0aNdJDDz2kH3/8Mdd6duzYYfWQysrlcsnlclm3AwICFBDg2bG8Xbt2eu2116zbmzdv1rRp0zRkyBC99NJL1vxLLrlETZs21QsvvKCnnnoqP6cpm7Jly+qTTz6xerZdccUVqlmzpsaPH6/XX39dBw4c0B9//KFJkyapV69e1nbdunUr1P4AAAD8HZcEAgBgEx988IFatGihyMhIBQUFKTg4WNOnT9fvv/+ebd02bdpYYZUkxcXFqVy5ctq2bZsk6ejRo1q7dq26du1qhVWSFBkZqS5duhS6xszMTI0bN0516tRRSEiIgoKCFBISoj///NOjzuXLl6tu3bpWWOV2yy23eNxesWKFDh48qL59+yozM9OaXC6XrrrqKq1Zs0ZHjx4tcJ0jR45UcHCwNV177bXZ1rnhhhs8bi9dulSSsl3qd9lll6l27dr5ukQxNz179vS4DLNq1apq3ry5tc9SpUrpggsu0LPPPquJEydq/fr1HoEbAABAcUNgBQCADXz88ce66aabVKlSJb3zzjtauXKl1qxZo9tvv10nTpzItn7p0qWzzQsNDdXx48clSYcOHZIxRnFxcdnWy2lefo0cOVKPPPKIunbtqi+++EI//vij1qxZowYNGlj7lqQDBw7ka9/uyxO7d+/uETAFBwfrmWeekTFGBw8ezLWeKlWqaN++fTp27JjH/HvvvVdr1qzRmjVrch3b6vT57oHZc1q/YsWKZxy4PS/ly5fPcZ67TYfDocWLF6tjx46aMGGCGjVqpLJly2r48OE6cuRIofcLAADgr7gkEAAAG3jnnXeUkJCgOXPmePTESU9PL1R7sbGxcjgcOY5XlZKS4nE7LCwsx33lFNC4x5saN26cx/z9+/crJibGul26dOl87ds9JtRLL72kyy+/PMdjyStg69ChgxYuXKj58+ere/fu1vz4+HjFx8dLkkcPs6xO/0VAdwi4e/duVa5c2WPZv//+6zF+VVhYWI73zf79+3Mc5+r043bPyxo8Vq1aVdOnT5ckbdmyRXPnzlVSUpIyMjL06quv5ngMAAAARRU9rAAAsAGHw6GQkBCPECUlJSXHXwnMjxIlSqhJkyb69NNPlZGRYc3/77//sv2aYFxcnMLCwrRhwwaP+bn9QmFoaKjHvHnz5mnXrl0e81q1aqWNGzdq06ZNHvNnz57tcbtFixaKiYnRpk2b1KRJkxyn3AInSRowYIDi4uI0atQo7d69O9f18qNt27aSToVyWa1Zs0a///67x+Dv1apVy3a+tmzZoj/++CPHtt9//30ZY6zb27Zt04oVKzx+UTCrCy+8UGPGjFG9evX0008/FeZwAAAA/Bo9rAAAOE+WLFniMeC429VXX61rrrlGH3/8sYYMGaLu3btrx44deuKJJ1ShQgX9+eefhdrf448/rs6dO6tjx44aMWKEnE6nnn32WUVGRnpcZudwONSrVy/NmDFDF1xwgRo0aKDVq1frvffey9bmNddco5kzZ+qiiy5S/fr1tW7dOj377LPZeiTdfffdmjFjhjp16qTHH39ccXFxeu+997R582ZJsgZAj4yM1EsvvaS+ffvq4MGD6t69u8qVK6d9+/bpl19+0b59+zR16tRcjzEmJkaffvqpunTpogYNGmjw4MG6/PLLFRkZqQMHDujbb79VSkqKmjdvfsbzVatWLd1xxx166aWXFBAQoE6dOlm/EhgfH6977rnHWrd3797q1auXhgwZohtuuEHbtm3ThAkTVLZs2Rzb3rt3r66//noNHDhQqampGjt2rMLCwjR69GhJ0oYNGzRs2DDdeOONqlmzpkJCQrRkyRJt2LDBGkQfAACgWDEAAOCcevPNN42kXKfk5GRjjDFPP/20qVatmgkNDTW1a9c2r7/+uhk7dqw5/e1akhk6dGi2/VStWtX07dvXY94nn3xi6tWrZ0JCQkyVKlXM008/bYYPH25iY2M91ktNTTUDBgwwcXFxpkSJEqZLly5m69atRpIZO3astd6hQ4dM//79Tbly5UxERIS54oorzHfffWdatWplWrVq5dHmxo0bTfv27U1YWJgpVaqU6d+/v3nrrbeMJPPLL794rLt8+XLTuXNnU6pUKRMcHGwqVapkOnfubD744IN8neOUlBQzevRoU79+fVOiRAkTHBxsKlasaLp06WJmzZplTp48me3+WLNmTbZ2nE6neeaZZ8yFF15ogoODTZkyZUyvXr3Mjh07PNZzuVxmwoQJpnr16iYsLMw0adLELFmyJNt5WLp0qZFk3n77bTN8+HBTtmxZExoaalq2bGnWrl1rrbdnzx7Tr18/c9FFF5kSJUqYyMhIU79+ffPCCy+YzMzMfJ0DAACAosRhTJb+6QAAoEg7efKkGjZsqEqVKmnhwoXnff933HGH3n//fR04cCDPS/2KimXLlqlNmzb64IMPPMbYAgAAQN64JBAAgCKsf//+6tChgypUqKCUlBS9+uqr+v333zV58uRzvu/HH39cFStWVPXq1a2xs9544w2NGTOmWIRVAAAAKDwCKwAAirAjR47ovvvu0759+xQcHKxGjRpp/vz5at++/Tnfd3BwsJ599lnt3LlTmZmZqlmzpiZOnKgRI0ac830DAADAv3FJIAAAAAAAAGwlwNcFAAAAAAAAAFkRWAEAAAAAAMBWGMOqAFwul/7991+VLFlSDofD1+UAAAAAAODBGKMjR46oYsWKCgigjwr8F4FVAfz777+Kj4/3dRkAAAAAAORpx44dqly5sq/LAAqNwKoASpYsKUnaseNNRUVF+Lgae5myqfRZtzGszgEvVOId3jgeyTvHZKdavIXHS3ZF7Xgkex0TkF88/s8tb51fb+A+gj/6dklNr7SzofxBr7TjDTwXvS8t7Zji42+zvr8C/orAqgDclwFGRUUQWJ0mLLLEWbcRFXXcC5V4hzeOR/LOMdmpFm/h8ZJdUTseyV7HBOQXj/9zy1vn1xu4j+CPSpSI9Eo7YZHpXmnHG3gunjsMYwN/xwWtAAAAAAAAsBUCKwAAAAAAANgKgRUAAAAAAABshcAKAAAAAAAAtkJgBQAAAAAAAFshsAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCsEVgAAAAAAALAVAisAAAAAAADYCoEVAAAAAAAAbIXACgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArRBYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFQIrAAAAAAAA2AqBFQAAAAAAAGyFwAoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0QWAEAAAAAAMBWCKwAAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANhKkK8LAAAAheNwdMnXekuXjlPr1vW8sr+hQztrypRBhdo+Pf2k3nhjoebO/U4bN25XWtoxRUVFqF69qrrllivVs2crlSwZcdZ1AgAAwP8RWAEA4KdWrnzW4/YTT8zR0qUbtGTJUx7z69Spcj7LytG+fam66qqx2rhxm/r2bafhw7uoXLkYHTiQpiVLNmjUqJn6/vtNevvte31dKgAAAGyAwAoAAD91+eUXedwuWzZKAQEB2ebbQa9ez+vXX7fpm2+e0JVX1vVY1rVrM40de4u++mpdnm04nU5lZroUGhp8LksFAACADTCGFQAARdjLL8/TlVc+qHLleqlEie6qV2+YJkz4SCdPZnqst37937rmmsdUrlwvhYZer4oV+6pz58e0c+f+XNs2xuihh2YpOLirXn99Qa7rrVmzRQsXrtcdd3TMFla5lS4dpV692li3t27dI4ejiyZM+EhPPjlHCQn9FRraTUuXbpAkff75j2rW7D5FRNygkiVvUocOj2jlys0ebfbr94KqVeufbV9JSe9lu5zS4eiiYcNe1cq5n2j8NTdpVKMrNeG6W7T+q0Ue62UcP6HPn3tRT13VTQ80bqUxLRL1Qo/b9NP8hbkePwAAAAqOHlYAABRhf/+9Wz17tlJCQpxCQoL0yy/Jeuqpudq8eadmzBghSTp69IQ6dHhECQlxevnlQYqLi1FKyiEtXfqrjhw5nmO76ekn1a/fJM2bt0ZffPGIrrqqca41LFr0syTp2msvK3D9L774hS68sJKee+52RUVFqGbNinrvvWW69dbnlZh4id5//36lp5/UhAkfq3Xr0Vq8+EldccXFBd6PdCoEywjZpKuGDlRIeJhWzP5Y74x6VAGBgWqQ2PbUOs9O1rovv9ZVw+5Qpdq1lHH8uFL+/EfHDqcWap8AAADIGYEVAABF2MSJA6z/u1wutWx5sUqXLqnbbpus55/vr9jYSG3evFMHDhzR9OnDdd11l1vr33RTyxzbPHjwiK677kklJ+/Rd989owYNEvKsYceOU720qlYt5zHfGCOn02XddjikwMBAj3XCwkK0YMFjCg4Oso6hRYtRqlevmr76KkkBAac6i199dRNdcMEdeuCBt/TDDxPOdFpytH9/mh6Y/5ZKliklSardsrmevf5WzZ881Qqsktdv0IXNLlOrPrdY29W5skWh9gcAAIDccUkgAABF2Pr1f+vaa59Q6dI9FRh4nYKDu6pPnxfkdLq0ZcsuSVKNGhUUGxupBx6YqVdf/UqbNm3Ptb3k5D1q1ux+paUd06pVz50xrMrLZ5+tUnBwV2uKjr452zrXXnuZFVZJ0h9/7NK//x5U795trLBKkiIjw3XDDc20atUfOnbsRKHqadeugRVWSVJAYKAaXtVe+7fv1OGUvZKkKvXqaPP3q/TlC6/orzU/6eSJwu0LAAAAeSOwAgCgiNq+fa9atnxQu3Yd0OTJA/Xdd09rzZqJevnlQZKk48czJEnR0SW0fPl4NWxYXQ89NEsXXzxUFSv21dix72Yb62r16i3asmWXevRoqcqVy+SrjipVykqStm3b6zG/det6WrNmotasmahrrrk0x20rVCjlcfvAgbT/zY/Ntm7FiqXlcrl06NDRfNV1uvLls7fpDrCOpZ665K/rgyPV5vZe2rhkuabePlRjruioGcMf0L5tOwq1TwAAAOSMSwIBACiiPv10lY4ePaGPP37I43K8n3/+J9u69epV0+zZo2SM0YYNWzVz5jd6/PHZCg8P0YMP3mit16NHS5UvH6uHH35bLpfRmDE9zlhHhw4N9dBDs/T556uVmNjImh8TE6kmTWpKkkqXLpnjtg6H5+3SpaMkSbt3H8q27r//HlBAQIBiY0tIOnU5YXr6yWzr7d+fluO+UlIOqfZp847sPyhJioiOliSFRoTrqqEDddXQgTqy/6B+/36l5k96RdOH3acHv5iTY7sAAAAoOHpYAQBQRDn+l/aEhgZb84wxev313H/RzuFwqEGDBL3wwkDFxJTQTz/9nW2dMWN6aNKkgXr00Xc1evRbZ6yjSZOaSky8RK+/vkDfffdbIY7k/9WqVUmVKpXWe+8tlzHGmn/06Al99NFKNWtWSxERYZKkatXKae/eVO3Z8//hVkbGSS1Y8FOObS9e/IsVUEmSy+nUz19/o9LxlRRTvly29UuWKaXLunbWJZ06aN/W7co4zuWBAAAA3kIPKwAAiqgOHRoqJCRIt9zyrEaN6qYTJ05q6tT5OnToP4/1vvxytV55Zb66dr1c1auXlzFGH3+8UocPH1WHDpfk2PaIEdcqMjJMd9zxsv7774RefPEOKyDLyTvv3KuOHR9V+/Zj1K9fO3Xs2EjlykUrLe24NmxI1jff/KKoqIgzHlNAQIAmTOinW299Xtdc87juvPMqpaef1LPPfqzDh//T00/3tdbt0aOlHn30Pd1887O6//5uOnEiQy+++IXHQO9ZlSkTpakDhqnDnbdZvxK4N3mbej37hLXO5J79VefKFqpwYQ2FR5XU3n+2au2XX6tqg7oKCQ87Y/0AAADIHwIrAACKqIsuitdHH43WmDHvqFu38SpduqR69mylkSO7qlOnJGu9mjUrKiamhCZM+Fj//ntAISHBqlWrkmbOvFt9+7bLtf3+/RNVokSYeveeqKNHT+iNN+7yGAg9q7Jlo7Vy5XN6/fUFmjPnO82d+73++++EoqIidPHFVTR8eBcNGJCYr+Pq2bO1SpQI0/jxH6pHjwkKDAzQ5ZfX0tKl49S8+f9f1JeQUF6fffawHnrobXXvPl4VKpTSyJFdtW9fqh577P1s7V57bVOlxNbRVy+9psO796h0fCXd+nSSLrmqvbVOjcsa67dl32v527N18kS6osuVVZMundT+jn75qh0AAAD54zBZ+9MjT2lpaYqOjlZq6px8/RW4OJm4MX8D7+ZlZN39XqjEO7xxPJJ3jslOtXgLj5fsitrxSPY6JuBMHI4uGjq0s6oPGuOV9nj858xbry/ewH0Ef7R0US2vtLO+wgGvtOMNPBe9Ly3tmKKjeyg1NVVRUVG+LgcoNMawAgAAAAAAgK0QWAEAAAAAAMBWGMMKAAAUe8Z8IUmauNHHhQAAAEASPawAAAAAAABgMwRWAAAAAAAAsBUuCSwAp9MpSdq5cz+/Eniaw3tcZ93GzpiDXqjEO7xxPJJ3jslOtXgLj5fsitrxSPY6JiC/ePyfW946v97AfQR/tG9ftFfaORxwyCvteAPPRe9LSzsm6f+/vwL+ymGMMb4uwl+sWbNGl112ma/LAAAAAAAgT6tXr9all17q6zKAQiOwKoBDhw6pVKlSemTilwoLL+HrcoqcwQen+7oEr9vbpIGvS7B8X/p7X5fgVXvW3O3rErzqz2p/+boEAF700Jr4s27jg7KhXqhEuqH62rNuI+DHel6oBLmJ63zfWbfh/Lrh2RdSBAVe9bOvS/Cqr/96xCvtfBWV5pV27MIbr7lFyX8njqrRuBt08OBBxcbG+rocoNC4JLAAAgMDJUlh4SUUFh7p42qKnqiwEF+X4HXHS4T7ugRLRGSwr0vwqqL2HAwpwWXGQFFSMuzs/7AVFh7mhUqkkl54LwrwwvEgd1GRgWfdhrMIfo7yhkAvnFs7iYjwzueFkBKZXmnHLrzxmlsUub+/Av6KQdcBAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANgKgRUAAAAAAABshcAKAAAAAAAAtkJgBQAAAAAAAFshsAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCsEVgAAAAAAALAVAisAAAAAAADYCoEVAAAAAAAAbIXACgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArRBYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFQIrAAAAAAAA2AqBFQAAAAAAAGyFwAoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0QWAEAAAAAAMBWCKwAAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANgKgRUAAAAAAABshcAKAAAAAAAAtkJgBQAAAAAAAFshsAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCsEVgAAAAAAALAVAisAAAAAAADYCoEVAAAAAAAAbIXACgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArRBYAQAAAAAAFBHffvutunTpoooVK8rhcOjTTz894zbLly9X48aNFRYWpurVq+vVV18994WeAYEVAAAAAABAEXH06FE1aNBAU6ZMydf6ycnJuvrqq9WyZUutX79eDz30kIYPH66PPvroHFeatyCf7h0AAAAAAABe06lTJ3Xq1Cnf67/66quqUqWKJk2aJEmqXbu21q5dq+eee0433HDDOaryzOhhBQAAAAAAUEytXLlSiYmJHvM6duyotWvX6uTJkz6qih5WBeJ0OiVJJ44f9XElRVPaiQxfl+B1R44e93UJlmNhvnuhORdOHP/P1yV4VcbRY74uAYAXHTlx9p8VThzP9EIl3nkvCvDC8SB3Ef85z7oNZxH8HOUNgV44t3Zy7Jh3Pi9kBBWtzx3eeM0tSv773/lwf3/F/ztx4oQyMvzz9dIYI4fD4TEvNDRUoaGhZ912SkqK4uLiPObFxcUpMzNT+/fvV4UKFc56H4VBYFUAf/31lyTpiZHX+LiSoulhXxdwTrzj6wKKsM98XQAA5GqmrwvIomi+vxYxj3qjkY3eaKToKXJPgB6+LsCWZvq6AJv666+/dOmll/q6DNs4ceKEEqpGKmWvfwZ5kZGR+u8/zz/ajx07VklJSV5p//QwzBiT4/zzicCqAGrUqCFJ+nV5vKJKBnqlzVtv7+aVdiQpaG+q19o6JzK9+MIQ5J3zDwAAgKIj89/dXmsrqKJvehQAZ+ukM10/7HnH+v6KUzIyMpSy16lt66opqqR/jY6UdsSlqo23aseOHYqKirLme6N3lSSVL19eKSkpHvP27t2roKAglS5d2iv7KAwCqwIIDDwVklSuEOy1B3hYaLRX2pGkoCCX19o6N7xzaYMkKYiHLgAAADxlOrz3B9ygoJJeaws4n4ICQiT9//dXeIoqGeC1DijnW1RUlEdg5S3NmjXTF1984TFv4cKFatKkiYKDg72+v/ziWz8AAAAAACgWXDJyye6dPTy5ZAq0/n///WcNaSRJycnJ+vnnn1WqVClVqVJFo0eP1q5duzRr1ixJ0qBBgzRlyhSNHDlSAwcO1MqVKzV9+nS9//77Xj2OgiKwAgAAAAAAKCLWrl2rNm3aWLdHjhwpSerbt69mzpyp3bt3a/v27dbyhIQEzZ8/X/fcc49efvllVaxYUS+++KJuuOGG8157VsUusHrllVf07LPPavfu3br44os1adIktWzZ0tdlAQAAAAAAnLXWrVtbg6bnZObMmdnmtWrVSj/99NM5rKrg/GuksbM0Z84c3X333Xr44Ye1fv16tWzZUp06dfJIFgEAAAAAAOBbxSqwmjhxovr3768BAwaodu3amjRpkuLj4zV16lRflwYAAAAAAM4xp3H55VQcFZvAKiMjQ+vWrVNiYqLH/MTERK1YsSLHbdLT05WWluYxAQAAAAAA4NwqNoHV/v375XQ6FRcX5zE/Li5OKSkpOW4zfvx4RUdHW1N8fPz5KBUAAAAAAKBYKzaBlZvD4fC4bYzJNs9t9OjRSk1NtaYdO3acjxIBAAAAAACKtWLzK4FlypRRYGBgtt5Ue/fuzdbryi00NFShoaHnozwAAAAAAHCOuWTkUu6/oGdH/lavtxSbHlYhISFq3LixFi1a5DF/0aJFat68uY+qAgAAAAAAwOmKTQ8rSRo5cqR69+6tJk2aqFmzZpo2bZq2b9+uQYMG+bo0AAAAAAAA/E+xCqx69OihAwcO6PHHH9fu3btVt25dzZ8/X1WrVvV1aQAAAAAA4BxzySWXr4soIP+r2DuKVWAlSUOGDNGQIUN8XQYAAAAAAAByUWzGsAIAAAAAAIB/ILACAAAAAACArRS7SwIBAAAAAEDx5DRGTmN8XUaB+Fu93kJgVQjX39JDQUFhXmlrwadve6UdSWo9cKDX2joXticGeq2tKgudXmsLAAAARUPEsWNea+to/Qpeaws4nzJPnpB2+roK4OxxSSAAAAAAAABshcAKAAAAAAAAtsIlgQAAAAAAoFhwycgl/xoTyt/q9RZ6WAEAAAAAAMBWCKwAAAAAAABgK1wSCAAAAAAAigWXjJx+dokdlwQCAAAAAAAANkBgBQAAAAAAAFshsAIAAAAAAICtMIYVAAAAAAAoFlwyfjcmlL/V6y30sAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCuMYQUAAAAAAIoFpzFyGv8aE8rf6vUWelgBAAAAAADAVgisAAAAAAAAYCtcEggAAAAAAIoF1/8mf+Jv9XoLPawAAAAAAABgK/SwKoSgPYcVFBDqlbZaDxzolXYkadnrr3utrXOh6QODvdaW3Y8VAAAA51/ijf281hafN+Gv0o64FHuhr6sAzh49rAAAAAAAAGAr9LACAAAAAADFglNGThlfl1Eg/lavt9DDCgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArTCGFQAAAAAAKBac5tTkT/ytXm+hhxUAAAAAAABshcAKAAAAAAAAtsIlgQAAAAAAoFhw/W/yJ/5Wr7fQwwoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0whhUAAAAAACgWXHLIKYevyygQl5/V6y30sAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCuMYQUAAAAAAIoFlzk1+RN/q9db6GEFAAAAAAAAWyGwAgAAAAAAgK0QWAEAAAAAAMBWGMMKAAAAAAAUC0455JTD12UUiL/V6y30sAIAAAAAAICtEFgBAAAAAADAVrgkEAAAAAAAFAtcEug/6GEFAAAAAAAAWyGwAgAAAAAAgK0QWAEAAAAAAMBWGMMKAAAAAAAUCy7jkMv415hQ/lavt9DDCgAAAAAAALZCYAUAAAAAAABbIbACAAAAAACArTCGFQAAAAAAKBaccsgp/xoTyt/q9RZ6WAEAAAAAAMBWCKwAAAAAAABgK1wSCAAAAAAAigWnAuT0s747Tl8X4CP+dS8BAAAAAACgyCOwAgAAAAAAgK0QWAEAAAAAAMBWGMMKAAAAAAAUC8Y45DIOX5dRIMbP6vUWAqvCyHRKAZleaWp7YqBX2pGkpg8M9lpb58KPz0z1Wlt2P1YAAACcf6X3pHitLT5vwl85M05IetjXZQBnjUsCAQAAAAAAYCsEVgAAAAAAALAVLgkEAAAAAADFglMOOeVfY0L5W73eQg8rAAAAAAAA2AqBFQAAAAAAAGyFSwIBAAAAAECx4DQBchr/6rvjNL6uwDf8614CAAAAAABAkUdgBQAAAAAAAFshsAIAAAAAAICtMIYVAAAAAAAoFlxyyOVnfXdcKp6DWPnXvQQAAAAAAIAij8AKAAAAAAAAtkJgBQAAAAAAAFthDCsAAAAAAFAsOOWQUw5fl1Eg/lavt9DDCgAAAAAAALZCYAUAAAAAAABb4ZJAAAAAAABQLDhNgJzGv/ruOI3xdQk+4V/3EgAAAAAAAIo8elgVRlCgFOCdU1dlodMr7UjSstdf91pb50LTBwZ7ra0fn5nqtbYAAABQNCTe2M9rbfF5E/4q7YhLsbN9XQVw9uhhBQAAAAAAAFuhhxUAAAAAACgWXHLIJYevyygQf6vXW+hhBQAAAAAAAFshsAIAAAAAAICtEFgBAAAAAADAVhjDCgAAAAAAFAsuBcjpZ313XDK+LsEn/OteAgAAAAAAQJFHYAUAAAAAAABb4ZJAAAAAAABQLDhNgJzGv/ruOA2XBAIAAAAAAAA+R2AFAAAAAAAAWyGwAgAAAAAAgK0whhUAAAAAACgWXAqQy8/67rjEGFYAAAAAAACAzxFYAQAAAAAAwFYIrAAAAAAAAGArjGEFAAAAAACKBadxyGkcvi6jQPytXm+hhxUAAAAAAABshcAKAAAAAAAAtsIlgQAAAAAAoFhwKkBOP+u745TxdQk+4V/3EgAAAAAAAIo8AisAAAAAAADYCoEVAAAAAAAAbIUxrAAAAAAAQLHgMgFyGf/qu+MyjGEFAAAAAAAA+ByBFQAAAAAAAGyFwAoAAAAAAAC2whhWAAAAAACgWHAqQE4/67vjFGNYAQAAAAAAAD5HYAUAAAAAAABb4ZJAAAAAAABQLLgkOY3D12UUiMvXBfgIPawAAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANgKY1gBAAAAAIBiwaUAufys746/1estxfOoAQAAAAAAYFsEVgAAAAAAALAVAisAAAAAAADYCmNYAQAAAACAYsFpAuQ0/tV3x9/q9ZbiedQAAAAAAABF2CuvvKKEhASFhYWpcePG+u677/Jc/+WXX1bt2rUVHh6uWrVqadasWeep0pzRwwoAAAAAAKAImTNnju6++2698soratGihV577TV16tRJmzZtUpUqVbKtP3XqVI0ePVqvv/66Lr30Uq1evVoDBw5UbGysunTp4oMjoIcVAAAAAAAoJlxy+OVUUBMnTlT//v01YMAA1a5dW5MmTVJ8fLymTp2a4/pvv/227rzzTvXo0UPVq1fXzTffrP79++uZZ54521NeaARWAAAAAAAANpeWluYxpaen57heRkaG1q1bp8TERI/5iYmJWrFiRY7bpKenKywszGNeeHi4Vq9erZMnT3rnAAqIwAoAAAAAAMDm4uPjFR0dbU3jx4/Pcb39+/fL6XQqLi7OY35cXJxSUlJy3KZjx4564403tG7dOhljtHbtWs2YMUMnT57U/v37vX4s+cEYVgAAAAAAADa3Y8cORUVFWbdDQ0PzXN/h8LyU0BiTbZ7bI488opSUFF1++eUyxiguLk79+vXThAkTFBgYePbFFwI9rAAAAAAAQLHgNAF+OUlSVFSUx5RbYFWmTBkFBgZm6021d+/ebL2u3MLDwzVjxgwdO3ZMW7du1fbt21WtWjWVLFlSZcqU8e6dkE8EVgAAAAAAAEVESEiIGjdurEWLFnnMX7RokZo3b57ntsHBwapcubICAwM1e/ZsXXPNNQoI8E10xCWBAAAAAAAARcjIkSPVu3dvNWnSRM2aNdO0adO0fft2DRo0SJI0evRo7dq1S7NmzZIkbdmyRatXr1bTpk116NAhTZw4URs3btRbb73ls2MgsAIAAAAAAChCevTooQMHDujxxx/X7t27VbduXc2fP19Vq1aVJO3evVvbt2+31nc6nXr++ef1xx9/KDg4WG3atNGKFStUrVo1Hx0BgRUAAAAAACgmnAqQ089GRypsvUOGDNGQIUNyXDZz5kyP27Vr19b69esLtZ9zxb/uJQAAAAAAABR5BFYAAAAAAACwFQIrAAAAAAAA2ApjWAEAAAAAgGLBZRxyGYevyygQf6vXW+hhBQAAAAAAAFshsAIAAAAAAICtcEkgAAAAAAAoFlwKkNPP+u64/KxebymeRw0AAAAAAADbIrACAAAAAACArRBYAQAAAAAAwFYYwwoAAAAAABQLLhMgl/Gvvjv+Vq+3FM+jBgAAAAAAgG0RWAEAAAAAAMBWCKwAAAAAAABgK4xhBQAAAAAAigWnHHLK4esyCsTf6vUWelgBAAAAAADAVgisAAAAAAAAYCtcEggAAAAAAIoFlwmQy/hX3x1/q9dbiudRAwAAAAAAwLYIrAAAAAAAAGArBFYAAAAAAACwFcawAgAAAAAAxYJTklMOX5dRIE5fF+Aj9LACAAAAAACArRBYAQAAAAAAwFYIrAAAAAAAAGArjGEFAAAAAACKBZcJkMv4V98df6vXW4rnUQMAAAAAAMC2CKwAAAAAAABgK1wSCAAAAAAAigWnCZDTzy6x87d6vaV4HjUAAAAAAABsi8AKAAAAAAAAtkJgBQAAAAAAAFthDCsAAAAAAFAsGDnkksPXZRSI8bN6vYUeVgAAAAAAALAVAisAAAAAAADYCoEVAAAAAAAAbIUxrAAAAAAAQLHgNAFyGv/qu+Nv9XpL8TxqAAAAAAAA2BaBFQAAAAAAAGyFSwIBAAAAAECx4DIOuYzD12UUiL/V6y30sAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCuMYQUAAAAAAIoFpwLk9LO+O/5Wr7cUz6MGAAAAAACAbRFYAQAAAAAAwFYIrAAAAAAAAGArjGEFAAAAAACKBZdxyGUcvi6jQPytXm+hhxUAAAAAAABshcAKAAAAAAAAtsIlgQAAAAAAoFhwKUAuP+u742/1ekvxPGoAAAAAAADYFoEVAAAAAAAAbKXIBFbffvutunTpoooVK8rhcOjTTz/1WG6MUVJSkipWrKjw8HC1bt1av/32m2+KBQAAAAAAQK6KTGB19OhRNWjQQFOmTMlx+YQJEzRx4kRNmTJFa9asUfny5dWhQwcdOXLkPFcKAAAAAAB8wWkcfjkVR0Vm0PVOnTqpU6dOOS4zxmjSpEl6+OGH1a1bN0nSW2+9pbi4OL333nu68847z2epAAAAAAAAyEOR6WGVl+TkZKWkpCgxMdGaFxoaqlatWmnFihU+rAwAAAAAAACnKzI9rPKSkpIiSYqLi/OYHxcXp23btuW6XXp6utLT063baWlp56ZAAAAAAAAAWIpFYOXmcHhe92mMyTYvq/Hjx+uxxx4712UBAAAAAIDzwGUccvnZmFD+Vq+3FItLAsuXLy/p/3taue3duzdbr6usRo8erdTUVGvasWPHOa0TAAAAAAAAxSSwSkhIUPny5bVo0SJrXkZGhpYvX67mzZvnul1oaKiioqI8JgAAAAAAAJxbReaSwP/++09//fWXdTs5OVk///yzSpUqpSpVqujuu+/WuHHjVLNmTdWsWVPjxo1TRESEevbs6cOqAQAAAADA+WJMgFzGv/ruGD+r11uKTGC1du1atWnTxro9cuRISVLfvn01c+ZMjRo1SsePH9eQIUN06NAhNW3aVAsXLlTJkiV9VTIAAAAAAAByUGQCq9atW8sYk+tyh8OhpKQkJSUlnb+iAAAAAAAAUGDFs18ZAAAAAAAAbKvI9LACAAAAAADIi1MOOeXwdRkF4m/1egs9rAAAAAAAAGArBFYAAAAAAACwFQIrAAAAAAAA2ApjWAEAAAAAgGLBZSSX8a8xoVzG1xX4Bj2sAAAAAAAAYCsEVgAAAAAAALAVLgkEAAAAAADFgssEyGX8q++Ov9XrLcXzqAEAAAAAAGBbBFYAAAAAAACwFQIrAAAAAAAA2ApjWAEAAAAAgGLBJYdccvi6jALxt3q9hR5WAAAAAAAAsBUCKwAAAAAAANgKgRUAAAAAAABshTGsAAAAAABAseA0DjmNf40J5W/1egs9rAAAAAAAAGArBFYAAAAAAACwFS4JBAAAAAAAxYLLBMhl/Kvvjr/V6y3F86gBAAAAAABgWwRWAAAAAAAAsBUCKwAAAAAAANgKY1gBAAAAAIBiwSWHXMbh6zIKxCX/qtdb6GEFAAAAAAAAWyGwAgAAAAAAgK0QWAEAAAAAAMBWGMMKAAAAAAAUC0YOvxsTyvhZvd5CDysAAAAAAADYCoEVAAAAAAAAbIXACgAAAAAAALbCGFYAAAAAAKBYcBmHXMa/xoTyt3q9hR5WAAAAAAAAsBUCKwAAAAAAANgKlwQCAAAAAIBiwWUC5DL+1XfH3+r1luJ51AAAAAAAALAtAisAAAAAAADYCoEVAAAAAAAAbIUxrAAAAAAAQLHgMg65jMPXZRSIv9XrLfSwAgAAAAAAgK0QWAEAAAAAAMBWCKwAAAAAAABgK4xhBQAAAAAAigWXHHLJv8aE8rd6vYUeVgAAAAAAALAVAisAAAAAAADYCpcEAgAAAACAYsFlHHIZ/7rEzt/q9RZ6WAEAAAAAAMBWCKwAAAAAAABgKwRWAAAAAAAAsBXGsAIAAAAAAMUCY1j5D3pYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFcawAgAAAAAAxQJjWPkPelgBAAAAAADAVgisAAAAAAAAYCtcEggAAAAAAIoFLgn0H/SwAgAAAAAAgK0QWAEAAAAAAMBWCKwAAAAAAABgK4xhBQAAAAAAigUjySX/GhPK+LoAH6GHFQAAAAAAAGyFwAoAAAAAAAC2QmAFAAAAAAAAW2EMKwAAAAAAUCy4jEMu419jWPlbvd5CDysAAAAAAADYCoEVAAAAAABAEfPKK68oISFBYWFhaty4sb777rs813/33XfVoEEDRUREqEKFCrrtttt04MCB81RtdgRWAAAAAACgWHBfEuhvU0HNmTNHd999tx5++GGtX79eLVu2VKdOnbR9+/Yc1//+++/Vp08f9e/fX7/99ps++OADrVmzRgMGDDjbU15oBFYAAAAAAABFyMSJE9W/f38NGDBAtWvX1qRJkxQfH6+pU6fmuP6qVatUrVo1DR8+XAkJCbriiit05513au3atee58v9HYAUAAAAAAGBzaWlpHlN6enqO62VkZGjdunVKTEz0mJ+YmKgVK1bkuE3z5s21c+dOzZ8/X8YY7dmzRx9++KE6d+7s9ePILwIrAAAAAAAAm4uPj1d0dLQ1jR8/Psf19u/fL6fTqbi4OI/5cXFxSklJyXGb5s2b691331WPHj0UEhKi8uXLKyYmRi+99JLXjyO/gny2ZwAAAAAAgPOosGNC+ZK73h07digqKsqaHxoamud2DofncRpjss1z27Rpk4YPH65HH31UHTt21O7du3X//fdr0KBBmj59+lkeQeEQWAEAAAAAANhcVFSUR2CVmzJlyigwMDBbb6q9e/dm63XlNn78eLVo0UL333+/JKl+/foqUaKEWrZsqSeffFIVKlQ4+wMoIC4JBAAAAAAAKCJCQkLUuHFjLVq0yGP+okWL1Lx58xy3OXbsmAICPCOiwMBASad6ZvkCgRUAAAAAAEARMnLkSL3xxhuaMWOGfv/9d91zzz3avn27Bg0aJEkaPXq0+vTpY63fpUsXffzxx5o6dar++ecf/fDDDxo+fLguu+wyVaxY0SfHwCWBAAAAAACgWPDnMawKokePHjpw4IAef/xx7d69W3Xr1tX8+fNVtWpVSdLu3bu1fft2a/1+/frpyJEjmjJliu69917FxMSobdu2euaZZ7x2HAVFYAUAAAAAAFDEDBkyREOGDMlx2cyZM7PNu+uuu3TXXXed46ryj0sCAQAAAAAAYCv0sAIAAAAAAMWCMQ4ZP7sk0N/q9RZ6WAEAAAAAAMBWCKwAAAAAAABgKwRWAAAAAAAAsBXGsAIAAAAAAMWCSw655F9jQvlbvd5CDysAAAAAAADYCoEVAAAAAAAAbIXACgAAAAAAALbCGFYAAAAAAKBYcBmHXMa/xoTyt3q9hR5WAAAAAAAAsBUCKwAAAAAAANgKlwQCAAAAAIBiwRiHjJ9dYudv9XoLPawAAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANgKY1gBAAAAAIBiwWUccvnZmFD+Vq+30MMKAAAAAAAAtkJgBQAAAAAAAFshsAIAAAAAAICtMIYVAAAAAAAoFoxxyPjZmFD+Vq+30MMKAAAAAAAAtkJgBQAAAAAAAFvhkkAAAAAAAFAsGOOQy88useOSQAAAAAAAAMAGCKwAAAAAAABgKwRWAAAAAAAAsBXGsAIAAAAAAMWCkWSMr6soGD8r12voYQUAAAAAAABbIbACAAAAAACArRBYAQAAAAAAwFYYwwoAAAAAABQLLjnkkMPXZRSIy8/q9RZ6WAEAAAAAAMBWCKwAAAAAAABgK1wSCAAAAAAAigVjHDLGvy6x87d6vYUeVgAAAAAAALAVAisAAAAAAADYCoEVAAAAAAAAbIUxrAAAAAAAQLHgMg45/GxMKJef1est9LACAAAAAACArRBYAQAAAAAAwFYIrAAAAAAAAGArjGEFAAAAAACKBWNOTf7E3+r1FnpYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFcawAgAAAAAAxYIxDhnj8HUZBeJv9XoLPawAAAAAAABgKwRWAAAAAAAAsBUuCQQAAAAAAMUClwT6D3pYAQAAAAAAwFYIrAAAAAAAAGArBFYAAAAAAACwFcawAgAAAAAAxYLLOOTwszGhXH5Wr7fQwwoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0whhUAAAAAACgWjDk1+RN/q9db6GEFAAAAAAAAWyGwAgAAAAAAgK1wSSAAAAAAACgWTl0S6PB1GQXCJYEAAAAAAACADRBYAQAAAAAAwFYIrAAAAAAAAGArjGEFAAAAAACKBWMcfjiGlX/V6y30sAIAAAAAAICtEFgBAAAAAADAVgisAAAAAAAAYCuMYQUAAAAAAIoF87/Jn/hbvd5CDysAAAAAAADYCoEVAAAAAAAAbIVLAgEAAAAAQLFgjEPGOHxdRoH4W73eQg8rAAAAAAAA2AqBFQAAAAAAAGyFwAoAAAAAAAC2whhWAAAAAACgeDD/m/yJv9XrJfSwAgAAAAAAgK0QWAEAAAAAAMBWCKwAAAAAAABgK4xhBQAAAAAAigfjkDEOX1dRMP5Wr5fQwwoAAAAAAAC2QmAFAAAAAAAAW+GSQAAAAAAAUCwYc2ryJ/5Wr7fQwwoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0whhUAAAAAACgWjHHIGIevyygQf6vXW+hhBQAAAAAAAFshsAIAAAAAAICtEFgBAAAAAADAVhjDCgAAAAAAFA/GcWryJ/5Wr5fQwwoAAAAAAAC2QmAFAAAAAAAAW+GSQAAAAAAAUCwYc2ryJ/5Wr7fQwwoAAAAAAAC2QmAFAAAAAAAAWyGwAgAAAAAAgK0whhUAAAAAACgezP8mf+Jv9XoJPawAAAAAAABgKwRWAAAAAAAAsBUCKwAAAAAAANgKY1gBAAAAAIBiwRiHjHH4uowC8bd6vYUeVgAAAAAAALAVAisAAAAAAADYCpcEAgAAAACA4sP4ugDkBz2sAAAAAAAAYCsEVgAAAAAAALAVAisAAAAAAAAUWuvWrTVr1iwdP37ca20SWAEAAAAAgGLBGIdfTnbXuHFjjRo1SuXLl9fAgQO1atWqs26zyARW48eP16WXXqqSJUuqXLly6tq1q/744w+PdYwxSkpKUsWKFRUeHq7WrVvrt99+81HFAAAAAAAA/u/555/Xrl27NGvWLO3bt09XXnml6tSpo+eee0579uwpVJtFJrBavny5hg4dqlWrVmnRokXKzMxUYmKijh49aq0zYcIETZw4UVOmTNGaNWtUvnx5dejQQUeOHPFh5QAAAAAAAP4tMDBQ1113nT799FPt2rVLPXv21COPPKL4+Hh17dpVS5YsKVB7QeeozvPu66+/9rj95ptvqly5clq3bp2uvPJKGWM0adIkPfzww+rWrZsk6a233lJcXJzee+893Xnnnb4oGwAAAAAAoMhYvXq13nzzTb3//vsqV66c+vXrp927d6tLly4aPHiwnnvuuXy1U2R6WJ0uNTVVklSqVClJUnJyslJSUpSYmGitExoaqlatWmnFihU5tpGenq60tDSPCQAAAAAA+Cnjp5PN7d27V88//7zq1q2rli1bat++fZo9e7a2bt2qxx57TNOmTdNnn32mV199Nd9tFpkeVlkZYzRy5EhdccUVqlu3riQpJSVFkhQXF+exblxcnLZt25ZjO+PHj9djjz12bosFAAAAAADwY5UrV9YFF1yg22+/Xf369VPZsmWzrXPZZZfp0ksvzXebRTKwGjZsmDZs2KDvv/8+2zKHw3N0fWNMtnluo0eP1siRI63baWlpio+P926xAAAAAAAAfmzx4sVq2bJlnutERUVp6dKl+W6zyF0SeNddd+nzzz/X0qVLVblyZWt++fLlJf1/Tyu3vXv3Zut15RYaGqqoqCiPCQAAAAAA+CuHn072NnbsWB0+fDjb/LS0NLVt27ZQbRaZwMoYo2HDhunjjz/WkiVLlJCQ4LE8ISFB5cuX16JFi6x5GRkZWr58uZo3b36+ywUAAAAAACgSli9froyMjGzzT5w4oe+++65QbRaZSwKHDh2q9957T5999plKlixp9aSKjo5WeHi4HA6H7r77bo0bN041a9ZUzZo1NW7cOEVERKhnz54+rh4AAAAAAMC/bNiwQdKpTkSbNm3yuKrN6XTq66+/VqVKlQrVdpEJrKZOnSpJat26tcf8N998U/369ZMkjRo1SsePH9eQIUN06NAhNW3aVAsXLlTJkiXPc7UAAAAAAAD+rWHDhnI4HHI4HDle+hceHq6XXnqpUG0XmcDKmDP/zqPD4VBSUpKSkpLOfUEAAAAAAMBezP8mf2LjepOTk2WMUfXq1bV69WqPXwcMCQlRuXLlFBgYWKi2i0xgBQAAAAAAgPOnatWqkiSXy+X1tgmsAAAAAAAAUCCff/65OnXqpODgYH3++ed5rnvttdcWuH0CKwAAAAAAABRI165dlZKSonLlyqlr1665rudwOOR0OgvcPoEVAAAAAAAoHhjDymuyXgZ4Li4JDPB6iwAAAAAAACgWTp48qTZt2mjLli1ebZfACgAAAAAAAIUSHBysjRs3yuFweLVdAisAAAAAAFA8GId/TjbXp08fTZ8+3attMoYVAAAAAAAACi0jI0NvvPGGFi1apCZNmqhEiRIeyydOnFjgNgmsAAAAAAAAUGgbN25Uo0aNJCnbWFaFvVSQwAoAAAAAAACFtnTpUq+3yRhWAAAAAACgWDDGP6fCeOWVV5SQkKCwsDA1btxY3333Xa7r9uvXTw6HI9t08cUXF/JMnz2f9bDas2eP4uLicl2+du1aNWnS5DxWBAAAAAAA4P/mzJmju+++W6+88opatGih1157TZ06ddKmTZtUpUqVbOtPnjxZTz/9tHU7MzNTDRo00I033pjvfa5Zs0YffPCBtm/froyMDI9lH3/8cYGPwWc9rBo2bKglS5bkuGzy5Mm64oorznNFAAAAAAAA/m/ixInq37+/BgwYoNq1a2vSpEmKj4/X1KlTc1w/Ojpa5cuXt6a1a9fq0KFDuu222/K1v9mzZ6tFixbatGmTPvnkE508eVKbNm3SkiVLFB0dXahj8FlgVadOHXXs2FFJSUky/+vflpqaqm7duumee+7RwIEDfVUaAAAAAACAX8rIyNC6deuUmJjoMT8xMVErVqzIVxvTp09X+/btVbVq1XytP27cOL3wwgv68ssvFRISosmTJ+v333/XTTfdlGOPrvzwWWD1zTff6KGHHtKTTz6p9u3b68svv1TDhg21dOlSffjhh3rppZd8VRoAAAAAACiKjJ9OktLS0jym9PT0HA9x//79cjqd2YZhiouLU0pKyhlP0e7du/XVV19pwIABZ1zX7e+//1bnzp0lSaGhoTp69KgcDofuueceTZs2Ld/tZOWzwMrhcOixxx7TwoULtXLlSl133XUqWbKk1q9fr27duvmqLAAAAAAAANuJj49XdHS0NY0fPz7P9R0Oh8dtY0y2eTmZOXOmYmJi1LVr13zXVqpUKR05ckSSVKlSJW3cuFGSdPjwYR07dizf7WTls0HXpVPp4Msvv6wTJ04oLi5OycnJ+vbbb1WtWjVflgUAAAAAAGArO3bsUFRUlHU7NDQ0x/XKlCmjwMDAbL2p9u7dm+eP30mnQq0ZM2aod+/eCgkJyXdtLVu21KJFi1SvXj3ddNNNGjFihJYsWaJFixapXbt2+W4nK5/1sFq/fr0aN26sxYsXa+7cufr777/VtWtX3Xbbberfv79OnDjhq9IAAAAAAABsJSoqymPKLbAKCQlR48aNtWjRIo/5ixYtUvPmzfPcx/Lly/XXX3+pf//+BaptypQpuvnmmyVJo0eP1n333ac9e/aoW7dumj59eoHacvNZD6vmzZurdu3a+vrrr3XBBRdIkt5++221atVKI0aM0Jo1a7RhwwZflQcAAAAAAIoa4zg1+ZNC1Dty5Ej17t1bTZo0UbNmzTRt2jRt375dgwYNknQqVNq1a5dmzZrlsd306dPVtGlT1a1bt0D7K1WqlPX/gIAAjRo1SqNGjSpw3Vn5LLDq16+fJk2alC0RHDBggJo2baoePXr4qDIAAAAAAAD/1aNHDx04cECPP/64du/erbp162r+/PnWr/7t3r1b27dv99gmNTVVH330kSZPnpyvfaSlpeW7nqyXMuaXzwKrqVOn5rqsXr16Wrt27XmsBgAAAAAAoOgYMmSIhgwZkuOymTNnZpsXHR1doAHSY2JizjiIu3ugd6fTme923Xw66LokLViwQMuWLdP+/fv1yCOPqEqVKlqzZo2qVaumiIgIX5cHAAAAAACKCIc5NfkTu9a7dOnSc9q+zwKrY8eO6brrrtPixYutRG7w4MGqUqWKnnvuOcXHx+u5557zVXkAAAAAAADIRatWrc5p+z4LrB5++GGtXbtWH330kTp06OBxPWNiYqJeeuklX5UGAAAAAACAPGzYsEF169ZVQEDAGX80r379+gVu32eB1QcffKAnnnhC119/fbZrGatUqZJt8C8AAAAAAADYQ8OGDZWSkqJy5cqpYcOGcjgcMib79Yt+N4bVvn37dPHFF+e4LCAgQMePHz/PFQEAAAAAgCLN/G/yJzatNzk5WWXLlrX+720+C6wqVaqkX3/9VW3atMm2bMOGDUpISPBBVQAAAAAAADiTqlWr5vh/b/FZYNWtWzc99dRTatmypXUto8Ph0LZt2/TCCy/otttu81VpAAAAAAAAKIBdu3bphx9+0N69e+VyuTyWDR8+vMDt+SywGjt2rBYvXqzLLrtMdevWlcPh0G233aa///5btWrV0oMPPuir0gAAAAAAAJBPb775pgYNGqSQkBCVLl1aDofDWuZwOAoVWAV4s8CCKFmypFasWKEnnnhCkZGRuuCCCxQREaHRo0fr22+/VXh4uK9KAwAAAAAARZFx+Odkc48++qgeffRRpaamauvWrUpOTramf/75p1Bt+qyHlSSFh4frwQcfpDcVAAAAAACAnzp27JhuvvlmBQR4r1+Uz3pYAQAAAAAAwP/1799fH3zwgVfbPK89rNq2bZvvdR0OhxYvXnwOqwEAAAAAAMWK+d/kT/yg3vHjx+uaa67R119/rXr16ik4ONhj+cSJEwvc5nkNrFwul8fAW3/88YdSUlJUtWpVlS9fXikpKdq2bZsqVKigWrVqnc/SAAAAAAAAUAjjxo3TggULrCzn9EHXC+O8BlbLli2z/v/1119r4MCB+uGHH9SsWTNr/ooVK9SjRw/dd99957M0AAAAAAAAFMLEiRM1Y8YM9evXz2tt+mwMqzFjxigpKckjrJKk5s2ba+zYsXr44Yd9VBkAAAAAAADyKzQ0VC1atPBqmz4LrH777TfFx8fnuKxKlSravHnzea4IAAAAAAAUacZPJ5sbMWKEXnrpJa+2eV4vCcwqLi5OH330kRITE7Mt++CDDxQXF+eDqgAAAAAAAFAQq1ev1pIlS/Tll1/q4osvzjbo+scff1zgNn0WWA0ZMkQPPvigDh48qJ49e1qDrr/77rv65JNPNH78eF+VBgAAAAAAgHyKiYlRt27dvNqmzwKrUaNG6dixY5owYYKVtBljFBYWpocfflijRo3yVWkAAAAAAADIpzfffNPrbfossJKkpKQk3XPPPVq5cqUOHDig0qVL6/LLL1dMTIwvywIAAAAAAEWRn4wJ5cHf6vUSnwZWkhQdHa2rrrrK12UAAAAAAAAgnxo1aqTFixcrNjZWl1xyiRwOR67r/vTTTwVu36eBlTFGa9as0bZt23T8+PFsy/v06eODqgAAAAAAAJCX6667TqGhoZKkrl27er19nwVWW7Zs0bXXXqs///xTxmTv3+ZwOAisAAAAAACA9xjHqcmf2LTesWPH5vh/b/FZYDV06FCdOHFCc+bMUf369a1UDgAAAAAAAP7JnfUcPXpUHTp0UM2aNQvVjs8Cq9WrV+v1119X9+7dfVUCAAAAAAAACun+++9XRkaGJk+eLEnKyMjQ5Zdfrk2bNikiIkKjRo3SwoUL1bx58wK3HeDtYvMrMjJSUVFRvto9AAAAAAAAzsJXX32ldu3aWbffffddbd++XX/++acOHTqkG2+8UU899VSh2vZZYHXbbbfpvffe89XuAQAAAABAMeMw/jnZ1fbt21WnTh3r9sKFC9W9e3dVrVpVDodDI0aM0Pr16wvVts8uCaxbt67ef/99XXvtterSpYtKly6dbZ1u3br5oDIAAAAAAACcSUBAgMcP6a1atUqPPPKIdTsmJkaHDh0qVNs+C6x69uwpSUpOTtaXX36ZbbnD4ZDT6TzfZQEAAAAAACAfLrroIn3xxRcaOXKkfvvtN23fvl1t2rSxlm/btk1xcXGFattngdXSpUt9tWsAAAAAAACcpfvvv1+33HKL5s2bp99++01XX321EhISrOXz58/XZZddVqi2fRZYtWrVyle7BgAAAAAAxZH53+RPbFzvDTfcoPnz52vevHlKTEzUXXfd5bE8IiJCQ4YMKVTbPgusAAAAAAAA4N/at2+v9u3b57hs7NixhW73vAZWw4cPz/e6DodDkydPPofVAAAAAAAAwI7Oa2A1ZcqUfK9LYAUAAAAAAFA8ndfAyuVync/dAQAAAAAAwA8F+LoAAAAAAAAAICsCKwAAAAAAABRa27Ztdfjw4Wzz09LS1LZt20K1ya8EAgAAAACAYsEhyWF8XUXBOHxdQD4sW7ZMGRkZ2eafOHFC3333XaHaJLACAAAAAABAgW3YsMH6/6ZNm5SSkmLddjqd+vrrr1WpUqVCtU1gBQAAAAAAgAJr2LChHA6HHA5Hjpf+hYeH66WXXipU2z4JrJxOp/7++2+VK1dOMTExvigBAAAAAAAAZyE5OVnGGFWvXl2rV69W2bJlrWUhISEqV66cAgMDC9W2TwIrY4zq1KmjL774Qp06dfJFCQAAAAAAoLgxjlOTP7FxvVWrVpUkuVwur7ftk18JDAoKUvny5c/JAQEAAAAAAOD8GT9+vGbMmJFt/owZM/TMM88Uqk2fBFaSdPPNN2vWrFm+2j0AAAAAAAC84LXXXtNFF12Ubf7FF1+sV199tVBt+mzQ9YYNG2rOnDlq27atunXrpgoVKsjh8Ozm1q1bNx9VBwAAAAAAihzzf+zde3zO9f/H8ednmx2YOQybsWxOcz4rI4yZ8ympHHJWir74UoTKJlFLooQUJqf0yyEppJivamqkryI6ihhyihzGts/vj7Xr67Jhm8uuz2WP++32vt1cn8/nel+v6+Lz2cdz7/f7+qe5Eheo9+jRoypdunSm7SVLllRSUlKu+nRaYNW3b19J0uHDhxUfH59pv2EYSk1NzeOqAAAAAAAAkBPBwcH64osvFBoaarf9iy++UFBQUK76dFpgtWXLFme9NAAAAAAAABxk8ODBGjlypK5cuaKWLVtKkj777DONGTNGo0ePzlWfTgusmjdv7qyXBgAAAAAAgIOMGTNGp06d0tChQ3X58mVJkre3t8aOHatx48blqk+nBVYZzp07p4SEBJ08eVIlSpRQo0aNVLhwYWeXBQAAAAAA7jSsYXVbGIahl156Sc8++6x++OEH+fj4qFKlSvLy8sp1n04NrKZNm6aYmBhduHBBppn+N1CoUCHFxMRo1KhRziwNAAAAAAAAOeDr66uGDRs6pC+nBVbvvPOOxowZo3bt2ql///4KCgrSkSNHtGjRIj311FMqWbKk+vTp46zyAAAAAAAAkA0tWrSQYRjX3b958+Yc9+m0wOrVV19Vr169tGTJErvtDzzwgB5++GG9+uqrBFYAAAAAAAAWV6dOHbvHV65c0bfffqvvv/9e/fr1y1WfTgus9u3bp6lTp2a57+GHH9Z9992XxxUBAAAAAIA7mWGmN1fiCvW++uqrWW6Pjo7W33//nas+3W6loFvh4+OjU6dOZbnv1KlT8vHxyeOKAAAAAAAA4CgPP/ywFixYkKvnOi2watq0qaKjo3XkyBG77UePHtWkSZPUrFkzJ1UGAAAAAACAW5WQkCBvb+9cPddpUwKnTJmixo0bq2LFioqMjFTp0qWVlJSkzZs3q0CBAlq1apWzSgMAAAAAAHci85/mSlyg3m7dutk9Nk1TSUlJ2rFjh5599tlc9em0wKp69epKTEzUxIkTtWXLFp08eVL+/v7q2rWrJk6cqMqVKzurNAAAAAAAAGRTkSJF7B67ubkpLCxMkyZNUuvWrXPVp9MCK0mqXLmyli9f7swSAAAAAAAAcAsWLlzo8D6dtoZVy5YttW/fviz3/fjjj2rZsmUeVwQAAAAAAAArcNoIq/j4eJ09ezbLfefOndPWrVvzuCIAAAAAAHBHYw0rhylWrJgMw8jWsadOncpx/06dEng9SUlJKliwoLPLAAAAAAAAQBZmzJhh+/PJkyc1efJktWnTRuHh4ZLSvyFw48aNrrHo+gcffKAPPvjA9vj5559XyZIl7Y65ePGi4uPjVbdu3bwsDQAAAAAAANnUr18/25/vv/9+TZo0SU888YRt2/DhwzVr1ix9+umn+ve//53j/vM0sNq7d6/+7//+T5JkGIY2b94sNzf7ZbS8vLxUs2ZNzZw5My9LAwAAAAAAQC5s3LhRL730Uqbtbdq00dNPP52rPvN00fVx48bp3LlzOnfunEzT1JYtW2yPM9qJEye0ZcsW1apVKy9LAwAAAAAAdzjDdM1mdf7+/lq9enWm7WvWrJG/v3+u+nTaGlZpaWnOemkAAAAAAAA4SExMjAYNGqT4+HjbGlbbt2/Xhg0b9Pbbb+eqzzwdYXW17du367333sty33vvvaevvvoqjysCAAAAAABATvXv319ffvmlihYtqlWrVmnlypUqUqSIvvjiC/Xv3z9XfTpthNX48ePVpEkTPfjgg5n27d27V2+99ZY2bdrkhMoAAAAAAMAdyTTSmytxkXrvueceLV261GH9OS2w2r17t0aPHp3lvnvuuUezZ8/O44oAAAAAAACQG6mpqVqzZo1++OEHGYahatWqqXPnznJ3d89Vf04LrM6fPy8Pj6xf3s3NTefOncvjigAAAAAAAJBTP//8s9q3b6/Dhw8rLCxMpmnqxx9/VHBwsD766CNVqFAhx306bQ2r0NBQbdmyJct9W7ZsUbly5fK4IgAAAAAAAOTU8OHDVaFCBR06dEjffPONdu3apYMHDyo0NFTDhw/PVZ9OC6x69OihV199VQsXLrTbHhcXpxkzZqhnz55OqgwAAAAAANyRTBdtFrd161bFxsaqePHitm3+/v568cUXtXXr1lz16bTA6umnn1aTJk00aNAgFSpUSJUqVVKhQoU0aNAgNWnSROPGjXNWaQAAAAAAAMgmLy+vLJd2+vvvv+Xp6ZmrPp0WWHl6emrTpk1atGiRunTpotDQUHXt2lWLFi3SJ598kus3BAAAAAAAgLzTsWNHPfroo/rqq69kmqZM09T27dv12GOPqXPnzrnq02mLrkuSu7u7+vTpoz59+jizDAAAAAAAAOTSa6+9pn79+ik8PFwFChSQJKWkpKhz586aOXNmrvp0amAlSfv27dPWrVt14sQJDRo0SIGBgTpy5IiKFSsmHx8fZ5cHAAAAAADuEIaZ3lyJ1es1TVN//fWXli9friNHjuiHH36QaZqqVq2aKlasmOt+nRZYpaam6tFHH1VcXJxM05RhGGrXrp0CAwM1ZMgQ1a1bV5MmTXJWeQAAAAAAALgJ0zRVqVIl7dmzR5UqVbqlkOpqTlvD6oUXXtCyZcv08ssv6/vvv5dp/i8ybNeunTZs2OCs0gAAAAAAAJANbm5uqlSpkk6ePOnYfh3aWw7ExcXp2Wef1ahRoxQWFma3LzQ0VL/99puTKgMAAAAAAEB2xcbG6qmnntL333/vsD6dNiXw8OHDCg8Pz3Kft7d3ll+HCAAAAAAAkGvmP82VuEC9Dz/8sC5cuKDatWvL09Mz05rkp06dynGfTgusSpUqpV9//VUtWrTItG///v0qW7asE6oCAAAAAABATrz66qsyDMOhfTotsGrfvr1eeOEFtW3bVoGBgZIkwzD0119/6bXXXlOnTp2cVRoAAAAAAACyqWfPnkpJSVGhQoUc1qfT1rCaNGmSUlJSVK1aNd1///0yDEPjx49XjRo1dOnSJT377LPOKg0AAAAAANyJTMlwsWblKYEnTpxQhw4d5OvrKz8/PzVu3Fi//vqrQ/rO08Dqzz//tP05ICBAiYmJ6tmzp3bu3Cl3d3f997//Vbt27fTll1+qePHieVkaAAAAAAAAcmDcuHHauXOnYmJi9PLLL+vEiRMaMmSIQ/rO0ymBZcqUUefOnTVo0CC1bdtWAQEBmjt3bl6WAAAAAAAAAAfYuHGjFixYoPbt20tKX/6pRo0aunLligoUKHBLfefpCKsHHnhA69evV8eOHRUcHKxnnnnGYUPFAAAAAAAAkHeOHDmiunXr2h5XqVJFnp6eOnLkyC33naeB1dKlS5WUlKQ33nhDZcqU0ZQpU1SpUiVFRERoyZIlunTpUl6WAwAAAAAA8hPTRZtFmaYpDw/7yXseHh5KS0u75b7zfNF1Pz8/PfbYY/rqq6+0Z88e/fvf/9b+/fvVt29fBQYG6vHHH1diYmJelwUAAAAAAIAcME1TkZGRqlevnq1duHBBnTp1stuWG3m6htW1qlatqmnTpumll17SRx99pAULFmjBggWaN2+eqlevrt27dzuzPAAAAAAAAFzHxIkTM23r0qWLQ/p2amCVwd3dXZ07d9Y999yj2NhYzZgxQ3v27HF2WQAAAAAAALiOrAIrR3F6YJWamqq1a9dq4cKF2rBhg1JSUlSrVi0NGjTI2aUBAAAAAIA7icXXhMqSq9XrIE4LrPbs2aMFCxZoyZIlOnHihPz8/DR48GANGjRI9evXd1ZZAAAAAAAAcLI8DazOnj2rZcuWacGCBdq5c6ckqVmzZho0aJC6d+8ub2/vvCwHAAAAAAAAFpSngVVgYKCSk5NVunRpPf300xo4cKAqVKiQlyUAAAAAAIB8yjDTmytxtXodxS0vX6xt27Zau3atDh48qBdeeIGwCgAAAAAAwMW98847Sk5OzrT98uXLeuedd3LVZ54GVqtWrVKHDh3k5panLwsAAAAAAIDbZMCAAfrrr78ybT937pwGDBiQqz5JjgAAAAAAAJBrpmnKMIxM2//44w8VKVIkV3067VsCAQAAAAAA4Lrq1q0rwzBkGIYiIyPl4fG/mCk1NVW//fab2rZtm6u+CawAAAAAAACQY127dpUkffvtt2rTpo18fX1t+zw9PRUSEqL7778/V30TWAEAAAAAACDHJk6cKEkKCQnRQw89JG9vb4f1TWAFAAAAAACAXOvXr58kaefOnfrhhx9kGIaqVaumunXr5rpPAisAAAAAAJA/mP80V+IC9R4/flw9evRQfHy8ihYtKtM09ddff6lFixZ69913VbJkyRz3ybcEAgAAAAAAINf+9a9/6ezZs9qzZ49OnTql06dP6/vvv9fZs2c1fPjwXPXJCCsAAAAAAADk2oYNG/Tpp5+qatWqtm3VqlXTG2+8odatW+eqTwIrAAAAAACQLxhmenMlrlBvWlqaChQokGl7gQIFlJaWlqs+mRIIAAAAAACAXGvZsqVGjBihI0eO2LYdPnxY//73vxUZGZmrPgmsAAAAAAAAkGuzZs3SuXPnFBISogoVKqhixYoKDQ3VuXPn9Prrr+eqT6YEAgAAAAAAINeCg4P1zTffaNOmTdq3b59M01S1atXUqlWrXPdJYAUAAAAAAPIPF1gTylVFRUUpKirKIX0RWAEAAAAAACBX0tLSFBcXp1WrVunAgQMyDEOhoaHq3r27+vTpI8MwctUva1gBAAAAAAAgx0zTVOfOnTV48GAdPnxYNWvWVPXq1fX777+rf//+uu+++3LdNyOsAAAAAAAAkGNxcXH6z3/+o88++0wtWrSw27d582Z17dpV77zzjvr27ZvjvhlhBQAAAAAA8gfTRZtFLV++XOPHj88UVklSy5Yt9fTTT2vp0qW56pvACgAAAAAAADm2e/dutW3b9rr727Vrp//+97+56pvACgAAAAAAADl26tQpBQQEXHd/QECATp8+nau+WcMKAAAAAADkC4aZ3lyJletNTU2Vh8f1oyV3d3elpKTkqm8CKwAAAAAAAOSYaZrq37+/vLy8styfnJyc674JrAAAAAAAAJBj/fr1u+kxufmGQInACgAAAAAAALmwcOHC29Y3gRUAAAAAAMgfzH+aK3G1eh2EbwkEAAAAAACApRBYAQAAAAAAwFIIrAAAAAAAAGAprGEFAAAAAADyBcNMb67E1ep1FEZYAQAAAAAAwFIIrAAAAAAAAGApTAkEAAAAAAD5g/lPcyWuVq+DMMIKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClsIYVAAAAAADIH1jDymUwwgoAAAAAAACWQmAFAAAAAABwh5k9e7ZCQ0Pl7e2t+vXra9u2bTc8Pjk5WRMmTFC5cuXk5eWlChUqaMGCBXlUbWZMCQQAAAAAALiDrFixQiNHjtTs2bPVpEkTvfnmm2rXrp327t2ru+66K8vnPPjggzp27Jjmz5+vihUr6vjx40pJScnjyv+HwAoAAAAAAOQLhpneXElu6p0+fboGDRqkwYMHS5JmzJihjRs3as6cOZo6dWqm4zds2KCtW7fq119/VfHixSVJISEht1L2LWNKIAAAAAAAgMWdPXvWriUnJ2d53OXLl7Vz5061bt3abnvr1q315ZdfZvmctWvXqkGDBoqNjVWZMmVUuXJlPfnkk7p48aLD30d2McIKAAAAAADA4oKDg+0eT5w4UdHR0ZmOO3HihFJTUxUQEGC3PSAgQEePHs2y719//VWff/65vL29tXr1ap04cUJDhw7VqVOnnLaOFYEVAAAAAADIH8x/miv5p95Dhw7Jz8/PttnLy+uGTzMMw74b08y0LUNaWpoMw9DSpUtVpEgRSenTCrt376433nhDPj4+t/AGcocpgQAAAAAAABbn5+dn164XWJUoUULu7u6ZRlMdP34806irDKVLl1aZMmVsYZUkVa1aVaZp6o8//nDcm8gBAisAAAAAAIA7hKenp+rXr69NmzbZbd+0aZMaN26c5XOaNGmiI0eO6O+//7Zt+/HHH+Xm5qayZcve1nqvh8AKAAAAAADgDjJq1Ci9/fbbWrBggX744Qf9+9//1sGDB/XYY49JksaNG6e+ffvaju/Vq5f8/f01YMAA7d27V//5z3/01FNPaeDAgU6ZDiixhhUAAAAAAMgvXHgNq5x46KGHdPLkSU2aNElJSUmqUaOGPv74Y5UrV06SlJSUpIMHD9qO9/X11aZNm/Svf/1LDRo0kL+/vx588EFNnjzZUe8ixwisAAAAAAAA7jBDhw7V0KFDs9wXFxeXaVuVKlUyTSN0JqYEAgAAAAAAwFIIrAAAAAAAAGApTAkEAAAAAAD5gmGmN1fiavU6CiOsAAAAAAAAYCkEVgAAAAAAALAUpgQCAAAAAID8wfynuRJXq9dBGGEFAAAAAAAASyGwAgAAAAAAgKUQWAEAAAAAAMBSWMMKAAAAAADkC4aZ3lyJq9XrKIywAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKaxhBQAAAAAA8gfzn+ZKXK1eB2GEFQAAAAAAACyFwAoAAAAAAACWwpRAAAAAAACQPzAl0GUwwgoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKWwhhUAAAAAAMgXjH+aK3G1eh2FEVYAAAAAAACwlDsmsJozZ45q1aolPz8/+fn5KTw8XOvXr7ftN01T0dHRCgoKko+PjyIiIrRnzx4nVgwAAAAAAICs3DGBVdmyZfXiiy9qx44d2rFjh1q2bKkuXbrYQqnY2FhNnz5ds2bNUmJiogIDAxUVFaVz5845uXIAAAAAAABc7Y4JrDp16qT27durcuXKqly5sl544QX5+vpq+/btMk1TM2bM0IQJE9StWzfVqFFDixYt0oULF7Rs2TJnlw4AAAAAAPKC6aItH7pjAqurpaam6t1339X58+cVHh6u3377TUePHlXr1q1tx3h5eal58+b68ssvr9tPcnKyzp49a9cAAAAAAABwe91RgdV3330nX19feXl56bHHHtPq1atVrVo1HT16VJIUEBBgd3xAQIBtX1amTp2qIkWK2FpwcPBtrR8AAAAAAAB3WGAVFhamb7/9Vtu3b9fjjz+ufv36ae/evbb9hmH/ZZCmaWbadrVx48bpr7/+srVDhw7dttoBAAAAAACQzsPZBTiSp6enKlasKElq0KCBEhMTNXPmTI0dO1aSdPToUZUuXdp2/PHjxzONurqal5eXvLy8bm/RAAAAAAAgTxhmenMlrlavo9xRI6yuZZqmkpOTFRoaqsDAQG3atMm27/Lly9q6dasaN27sxAoBAAAAAABwrTtmhNX48ePVrl07BQcH69y5c3r33XcVHx+vDRs2yDAMjRw5UlOmTFGlSpVUqVIlTZkyRQULFlSvXr2cXToAAAAAAACucscEVseOHVOfPn2UlJSkIkWKqFatWtqwYYOioqIkSWPGjNHFixc1dOhQnT59Wvfcc48++eQTFS5c2MmVAwAAAACAPGH+01yJq9XrIHdMYDV//vwb7jcMQ9HR0YqOjs6bggAAAAAAAJArd/QaVgAAAAAAAHA9BFYAAAAAAACwlDtmSiAAAAAAAMBN5dM1oVwNI6wAAAAAAABgKQRWAAAAAAAAsBQCKwAAAAAAAFgKa1gBAAAAAIB8wTDTmytxtXodhRFWAAAAAAAAsBQCKwAAAAAAAFgKUwIBAAAAAED+YP7TXImr1esgjLACAAAAAACApRBYAQAAAAAAwFIIrAAAAAAAAGAprGEFAAAAAADyBcNMb67E1ep1FEZYAQAAAAAAwFIIrAAAAAAAAGApBFYAAAAAAACwFNawAgAAAAAA+YP5T3MlrlavgzDCCgAAAAAAAJZCYAUAAAAAAABLYUogAAAAAADIFwwzvbkSV6vXURhhBQAAAAAAAEshsAIAAAAAAIClEFgBAAAAAADAUljDCgAAAAAA5A/mP82VuFq9DsIIKwAAAAAAAFgKgRUAAAAAAAAshcAKAAAAAAAAlsIaVgAAAAAAIH9gDSuXwQgrAAAAAAAAWAqBFQAAAAAAACyFKYEAAAAAACBfMMz05kpcrV5HYYQVAAAAAAAALIXACgAAAAAAAJZCYAUAAAAAAABLYQ0rAAAAAACQP5j/NFfiavU6CCOsAAAAAAAAYCkEVgAAAAAAALAUAisAAAAAAABYCmtYAQAAAACAfMEwTRmmay0K5Wr1OgojrAAAAAAAAGApBFYAAAAAAACwFKYEAgAAAACA/MH8p7kSV6vXQRhhBQAAAAAAAEshsAIAAAAAAIClEFgBAAAAAADAUljDCgAAAAAA5AuGmd5ciavV6yiMsAIAAAAAAIClEFgBAAAAAADAUgisAAAAAAAAYCmsYQUAAAAAAPIH85/mSlytXgdhhBUAAAAAAAAshcAKAAAAAAAAlsKUQAAAAAAAkC8YZnpzJa5Wr6MwwgoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKWwhhUAAAAAAMgfzH+aK3G1eh2EEVYAAAAAAACwFAIrAAAAAAAAWAqBFQAAAAAAACyFNawAAAAAAEC+YJjpzZW4Wr2OwggrAAAAAAAAWAqBFQAAAAAAACyFKYEAAAAAACB/MP9prsTV6nUQRlgBAAAAAADAUgisAAAAAAAAYCkEVgAAAAAAALAU1rACAAAAAAD5hpFP14RyNYywAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKaxhBQAAAAAA8gfTTG+uxNXqdRBGWAEAAAAAAMBSCKwAAAAAAABgKUwJBAAAAAAA+YJhpjdX4mr1OgojrAAAAAAAAGApBFYAAAAAAACwFAIrAAAAAAAAWAprWAEAAAAAgPzB/Ke5Eler10EYYQUAAAAAAABLIbACAAAAAACApRBYAQAAAAAAwFJYwwoAAAAAAOQLRlp6cyWuVq+jMMIKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClsIYVAAAAAADIH8x/mitxtXodhBFWAAAAAAAAsBQCKwAAAAAAAFgKUwIBAAAAAEC+YJjpzZW4Wr2OwggrAAAAAAAAWAqBFQAAAAAAACyFwAoAAAAAAACWwhpWAAAAAAAgfzDN9OZKXK1eB2GEFQAAAAAAACyFwAoAAAAAAACWQmAFAAAAAABwh5k9e7ZCQ0Pl7e2t+vXra9u2bdc9Nj4+XoZhZGr79u3Lw4rtsYYVAAAAAADIFwwzvbmS3NS7YsUKjRw5UrNnz1aTJk305ptvql27dtq7d6/uuuuu6z5v//798vPzsz0uWbJkbkp2CEZYAQAAAAAA3EGmT5+uQYMGafDgwapatapmzJih4OBgzZkz54bPK1WqlAIDA23N3d09jyrOjMAKAAAAAADA4s6ePWvXkpOTszzu8uXL2rlzp1q3bm23vXXr1vryyy9v+Bp169ZV6dKlFRkZqS1btjis9twgsAIAAAAAAPmD6aJNUnBwsIoUKWJrU6dOzfItnjhxQqmpqQoICLDbHhAQoKNHj2b5nNKlS2vevHlauXKlVq1apbCwMEVGRuo///nPzT7R24Y1rAAAAAAAACzu0KFDdutLeXl53fB4wzDsHpummWlbhrCwMIWFhdkeh4eH69ChQ5o2bZqaNWt2C1XnHiOsAAAAAAAALM7Pz8+uXS+wKlGihNzd3TONpjp+/HimUVc30qhRI/3000+3VPOtILACAAAAAAC4Q3h6eqp+/fratGmT3fZNmzapcePG2e5n165dKl26tKPLyzamBAIAAAAAgHzBMNObK8lNvaNGjVKfPn3UoEEDhYeHa968eTp48KAee+wxSdK4ceN0+PBhvfPOO5KkGTNmKCQkRNWrV9fly5e1ZMkSrVy5UitXrnTkW8kRAisAAAAAAIA7yEMPPaSTJ09q0qRJSkpKUo0aNfTxxx+rXLlykqSkpCQdPHjQdvzly5f15JNP6vDhw/Lx8VH16tX10UcfqX379s56CwRWAAAAAAAAd5qhQ4dq6NChWe6Li4uzezxmzBiNGTMmD6rKPtawAgAAAAAAgKUwwgoAAAAAAOQPppneXImr1esgjLACAAAAAACApRBYAQAAAAAAwFKYEggAAAAAAPIFw0xvrsTV6nUURlgBAAAAAADAUgisAAAAAAAAYCkEVgAAAAAAALAU1rACAAAAAAD5g/lPcyWuVq+DMMIKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClsIYVAAAAAADIFwwzvbkSV6vXURhhBQAAAAAAAEshsAIAAAAAAIClMCUQAAAAAADkD2lmenMlrlavgzDCCgAAAAAAAJZCYAUAAAAAAABLIbACAAAAAACApbCGFQAAAAAAyB/Mf5orcbV6HYQRVgAAAAAAALAUAisAAAAAAABYCoEVAAAAAAAALIU1rAAAAAAAQL5gSDJcbE0ow9kFOAkjrAAAAAAAAGApBFYAAAAAAACwFKYEAgAAAACA/ME005srcbV6HYQRVgAAAAAAALAUAisAAAAAAABYCoEVAAAAAAAALIU1rAAAAAAAQL5gmOnNlbhavY7CCCsAAAAAAABYCoEVAAAAAAAALIUpgbmQ8scRySjgkL4KXrjgkH4kqfUD/R3W1+3gf+yow/qy+nsFAABA3vvk/+Ic1hf3m3BVKSmXJE12dhnALSOwAgAAAAAA+YP5T3MlrlavgzAlEAAAAAAAAJZCYAUAAAAAAABLYUogAAAAAADIFwzTlGG61hw7V6vXURhhBQAAAAAAAEshsAIAAAAAAIClEFgBAAAAAADAUljDCgAAAAAA5A9p/zRX4mr1OggjrAAAAAAAAGApBFYAAAAAAACwFAIrAAAAAAAAWAprWAEAAAAAgHzBME0ZpunsMnLE1ep1FEZYAQAAAAAAwFIIrAAAAAAAAGApTAkEAAAAAAD5g/lPcyWuVq+DMMIKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClsIZVLniUDZKHm5dD+jpfq7RD+pGk+Lfeclhft8M9Yx93WF9fvTTHYX0BAADgztD6gf4O6+uT/4tzWF9AXjp7Lk3FKju7CgszzfTmSlytXgdhhBUAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEthDSsAAAAAAJAvGGZ6cyWuVq+jMMIKAAAAAAAAlkJgBQAAAAAAAEthSiAAAAAAAMgfTDO9uRJXq9dBGGEFAAAAAAAASyGwAgAAAAAAgKUQWAEAAAAAAMBSWMMKAAAAAADkC0ZaenMlrlavozDCCgAAAAAAAJZCYAUAAAAAAABLIbACAAAAAACApbCGFQAAAAAAyB9MM725Eler10EYYQUAAAAAAABLIbACAAAAAACApRBYAQAAAAAAwFJYwwoAAAAAAOQP5j/NlbhavQ7CCCsAAAAAAABYCoEVAAAAAAAALIUpgQAAAAAAIF8wTFOG6Vpz7FytXkdhhBUAAAAAAAAshcAKAAAAAAAAlnJHBlZTp06VYRgaOXKkbZtpmoqOjlZQUJB8fHwUERGhPXv2OK9IAAAAAAAAZOmOC6wSExM1b9481apVy257bGyspk+frlmzZikxMVGBgYGKiorSuXPnnFQpAAAAAADIU6bpmi0fuqMCq7///lu9e/fWW2+9pWLFitm2m6apGTNmaMKECerWrZtq1KihRYsW6cKFC1q2bJkTKwYAAAAAAMC17qjAatiwYerQoYNatWplt/23337T0aNH1bp1a9s2Ly8vNW/eXF9++WVelwkAAAAAAIAb8HB2AY7y7rvv6ptvvlFiYmKmfUePHpUkBQQE2G0PCAjQ77//ft0+k5OTlZycbHt89uxZB1ULAAAAAACA67kjRlgdOnRII0aM0JIlS+Tt7X3d4wzDsHtsmmambVebOnWqihQpYmvBwcEOqxkAAAAAAOQxU1Kai7X8uYTVnRFY7dy5U8ePH1f9+vXl4eEhDw8Pbd26Va+99po8PDxsI6syRlplOH78eKZRV1cbN26c/vrrL1s7dOjQbX0fAAAAAAAAuEOmBEZGRuq7776z2zZgwABVqVJFY8eOVfny5RUYGKhNmzapbt26kqTLly9r69ateumll67br5eXl7y8vG5r7QAAAAAAALB3RwRWhQsXVo0aNey2FSpUSP7+/rbtI0eO1JQpU1SpUiVVqlRJU6ZMUcGCBdWrVy9nlAwAAAAAAPKYYZoyTNeaY+dq9TrKHRFYZceYMWN08eJFDR06VKdPn9Y999yjTz75RIULF3Z2aQAAAAAAALjKHRtYxcfH2z02DEPR0dGKjo52Sj0AAAAAAADInjti0XUAAAAAAADcOe7YEVYAAAAAAAB2TEmutiaUi5XrKIywAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKaxhBQAAAAAA8gfTdME1rFysXgdhhBUAAAAAAAAshcAKAAAAAAAAlsKUQAAAAAAAkD+kSTKcXUQOpTm7AOdghBUAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEthDSsAAAAAAJAvGKYpwzSdXUaOuFq9jsIIKwAAAAAAAFgKgRUAAAAAAAAshcAKAAAAAAAAlsIaVgAAAAAAIH8wzfTmSlytXgdhhBUAAAAAAAAshcAKAAAAAAAAlsKUQAAAAAAAkD8wJdBlMMIKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClsIYVAAAAAADIH1jDymUwwgoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKWwhhUAAAAAAMgf0iQZzi4ih9KcXYBzMMIKAAAAAAAAlkJgBQAAAAAAAEthSiAAAAAAAMgXDNOUYZrOLiNHXK1eR2GEFQAAAAAAACyFwAoAAAAAAACWQmAFAAAAAAAAS2ENKwAAAAAAkD+YZnpzJa5Wr4MwwgoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKWwhhUAAAAAAMgf0kzJcLE1odJcrF4HYYQVAAAAAAAALIXACgAAAAAAAJbClEAAAAAAAJA/mGZ6cyWuVq+DMMIKAAAAAADgDjN79myFhobK29tb9evX17Zt27L1vC+++EIeHh6qU6fO7S3wJgisAAAAAAAA7iArVqzQyJEjNWHCBO3atUtNmzZVu3btdPDgwRs+76+//lLfvn0VGRmZR5VeH4EVAAAAAADAHWT69OkaNGiQBg8erKpVq2rGjBkKDg7WnDlzbvi8IUOGqFevXgoPD8+jSq+PwAoAAAAAAOQT5v/WsXKVppytYXX58mXt3LlTrVu3ttveunVrffnll9d93sKFC/XLL79o4sSJuflgHY5F1wEAAAAAACzu7Nmzdo+9vLzk5eWV6bgTJ04oNTVVAQEBdtsDAgJ09OjRLPv+6aef9PTTT2vbtm3y8LBGVMQIKwAAAAAAAIsLDg5WkSJFbG3q1Kk3PN4wDLvHpmlm2iZJqamp6tWrl2JiYlS5cmWH1nwrrBGbAQAAAAAA4LoOHTokPz8/2+OsRldJUokSJeTu7p5pNNXx48czjbqSpHPnzmnHjh3atWuXnnjiCUlSWlqaTNOUh4eHPvnkE7Vs2dKB7yR7CKwAAAAAAED+YFsXyoX8U6+fn59dYHU9np6eql+/vjZt2qT77rvPtn3Tpk3q0qVLpuP9/Pz03Xff2W2bPXu2Nm/erPfff1+hoaG3+AZyh8AKAAAAAADgDjJq1Cj16dNHDRo0UHh4uObNm6eDBw/qsccekySNGzdOhw8f1jvvvCM3NzfVqFHD7vmlSpWSt7d3pu15icAKAAAAAADgDvLQQw/p5MmTmjRpkpKSklSjRg19/PHHKleunCQpKSlJBw8edHKVN0ZgBQAAAAAA8oc0U5KLTQlMy129Q4cO1dChQ7PcFxcXd8PnRkdHKzo6Olev6yh8SyAAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEthDSsAAAAAAJA/mGnpzZW4Wr0OwggrAAAAAAAAWAqBFQAAAAAAACyFwAoAAAAAAACWwhpWAAAAAAAgfzDN9OZKXK1eB2GEFQAAAAAAACyFwAoAAAAAAACWwpRAAAAAAACQP6SZklxsil2ai9XrIIywAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKaxhBQAAAAAA8gfTTG+uxNXqdRBGWAEAAAAAAMBSCKwAAAAAAABgKQRWAAAAAAAAsBTWsAIAAAAAAPmDKddbE8rFynUURlgBAAAAAADAUgisAAAAAAAAYCkEVgAAAAAAALAU1rACAAAAAAD5g2m64BpWLlavgzDCCgAAAAAAAJZCYAUAAAAAAABLYUogAAAAAADIH9LSJKU5u4qcSXOxeh2EEVYAAAAAAACwFAIrAAAAAAAAWAqBFQAAAAAAACyFNawAAAAAAED+YJrpzZW4Wr0OwggrAAAAAAAAWAqBFQAAAAAAACyFwAoAAAAAAACWwhpWAAAAAAAgf2ANK5fBCCsAAAAAAABYCoEVAAAAAAAALIUpgQAAAAAAIH9IMyW52BS7NBer10EYYQUAAAAAAABLIbACAAAAAACApRBYAQAAAAAAwFJYwwoAAAAAAOQLppkm00xzdhk54mr1OgojrAAAAAAAAGApBFYAAAAAAACwFAIrAAAAAAAAWAprWAEAAAAAgPzBNKU009lV5IzpYvU6CCOsAAAAAAAAYCkEVgAAAAAAALAUpgQCAAAAAID8wTQludgUO6YEAgAAAAAAAM5HYAUAAAAAAABLIbACAAAAAACApbCGFQAAAAAAyB/S0iQjzdlV5IzpYvU6CCOsAAAAAAAAYCkEVgAAAAAAALAUAisAAAAAAABYCmtYAQAAAACA/ME0JZnOriJnTBer10EYYQUAAAAAAABLIbACAAAAAACApTAlEAAAAAAA5AtmWppMI83ZZeSIabpWvY7CCCsAAAAAAABYCoEVAAAAAAAALIXACgAAAAAAAJbCGlYAAAAAACB/ME1JprOryBnTxep1EEZYAQAAAAAAwFIIrAAAAAAAAGApBFYAAAAAAACwFNawAgAAAAAA+UOaKRkutiYUa1gBAAAAAAAAzkdgBQAAAAAAAEthSiAAAAAAAMgfTFNSmrOryBmmBAIAAAAAAADOR2AFAAAAAAAASyGwAgAAAAAAgKWwhhUAAAAAAMgXzDRTpuFaa0KZrGEFAAAAAAAAOB+BFQAAAAAAACyFwAoAAAAAAACWwhpWAAAAAAAgfzDTJKU5u4qcMV2sXgdhhBUAAAAAAAAshcAKAAAAAAAAlsKUQAAAAAAAkC+YaaZMw3R2GTlimq5Vr6MwwgoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKWwhhUAAAAAAMgfzDRJac6uImdMF6vXQRhhBQAAAAAAAEshsAIAAAAAAIClMCUwBzK+SjIl7bLD+ky5cslhfZ09Z+1hgqmX8897BQAAQN5LSeF+Ezj7d/q/3Yz/v8Jeiq5ILvbRpOiKs0twCsPkX3G2/fHHHwoODnZ2GQAAAAAA3NChQ4dUtmxZZ5dhGZcuXVJoaKiOHj3q7FJyJTAwUL/99pu8vb2dXUqeIbDKgbS0NB05ckSFCxeWYRjXPe7s2bMKDg7WoUOH5Ofnl4cVAtbD+QD8D+cDYI9zAvgfzgc4immaOnfunIKCguTmxipAV7t06ZIuX3bcjKm85Onpma/CKokpgTni5uaWo4Taz8+PHzbAPzgfgP/hfADscU4A/8P5AEcoUqSIs0uwJG9v73wX+rgy4lYAAAAAAABYCoEVAAAAAAAALIXA6jbw8vLSxIkT5eXl5exSAKfjfAD+h/MBsMc5AfwP5wMA2GPRdQAAAAAAAFgKI6wAAAAAAABgKQRWAAAAAAAAsBQCKwAAAAAAAFgKgRUAAAAAAAAshcDKwWbPnq3Q0FB5e3urfv362rZtm7NLAvLEf/7zH3Xq1ElBQUEyDENr1qyx22+apqKjoxUUFCQfHx9FRERoz549zikWuI2mTp2qhg0bqnDhwipVqpS6du2q/fv32x3D+YD8ZM6cOapVq5b8/Pzk5+en8PBwrV+/3raf8wH52dSpU2UYhkaOHGnbxjkBAOkIrBxoxYoVGjlypCZMmKBdu3apadOmateunQ4ePOjs0oDb7vz586pdu7ZmzZqV5f7Y2FhNnz5ds2bNUmJiogIDAxUVFaVz587lcaXA7bV161YNGzZM27dv16ZNm5SSkqLWrVvr/PnztmM4H5CflC1bVi+++KJ27NihHTt2qGXLlurSpYvtP+CcD8ivEhMTNW/ePNWqVctuO+cEAKQzTNM0nV3EneKee+5RvXr1NGfOHNu2qlWrqmvXrpo6daoTKwPylmEYWr16tbp27Sop/TeFQUFBGjlypMaOHStJSk5OVkBAgF566SUNGTLEidUCt9eff/6pUqVKaevWrWrWrBnnAyCpePHievnllzVw4EDOB+RLf//9t+rVq6fZs2dr8uTJqlOnjmbMmMHPCAC4CiOsHOTy5cvauXOnWrdubbe9devW+vLLL51UFWANv/32m44ePWp3fnh5eal58+acH7jj/fXXX5LS/4MucT4gf0tNTdW7776r8+fPKzw8nPMB+dawYcPUoUMHtWrVym475wQA/I+Hswu4U5w4cUKpqakKCAiw2x4QEKCjR486qSrAGjLOgazOj99//90ZJQF5wjRNjRo1Svfee69q1KghifMB+dN3332n8PBwXbp0Sb6+vlq9erWqVatm+w845wPyk3fffVfffPONEhMTM+3jZwQA/A+BlYMZhmH32DTNTNuA/IrzA/nNE088od27d+vzzz/PtI/zAflJWFiYvv32W505c0YrV65Uv379tHXrVtt+zgfkF4cOHdKIESP0ySefyNvb+7rHcU4AAFMCHaZEiRJyd3fPNJrq+PHjmX5DAuQ3gYGBksT5gXzlX//6l9auXastW7aobNmytu2cD8iPPD09VbFiRTVo0EBTp05V7dq1NXPmTM4H5Ds7d+7U8ePHVb9+fXl4eMjDw0Nbt27Va6+9Jg8PD9u/e84JACCwchhPT0/Vr19fmzZtstu+adMmNW7c2ElVAdYQGhqqwMBAu/Pj8uXL2rp1K+cH7jimaeqJJ57QqlWrtHnzZoWGhtrt53wA0s+T5ORkzgfkO5GRkfruu+/07bff2lqDBg3Uu3dvffvttypfvjznBAD8gymBDjRq1Cj16dNHDRo0UHh4uObNm6eDBw/qsccec3ZpwG33999/6+eff7Y9/u233/Ttt9+qePHiuuuuuzRy5EhNmTJFlSpVUqVKlTRlyhQVLFhQvXr1cmLVgOMNGzZMy5Yt0wcffKDChQvbfktepEgR+fj4yDAMzgfkK+PHj1e7du0UHBysc+fO6d1331V8fLw2bNjA+YB8p3DhwrY1DTMUKlRI/v7+tu2cEwCQjsDKgR566CGdPHlSkyZNUlJSkmrUqKGPP/5Y5cqVc3ZpwG23Y8cOtWjRwvZ41KhRkqR+/fopLi5OY8aM0cWLFzV06FCdPn1a99xzjz755BMVLlzYWSUDt8WcOXMkSREREXbbFy5cqP79+0sS5wPylWPHjqlPnz5KSkpSkSJFVKtWLW3YsEFRUVGSOB+Aa3FOAEA6wzRN09lFAAAAAAAAABlYwwoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKQRWAAAAAAAAsBQCKwAAAAAAAFgKgRUAAC7mvvvuk4+Pj86cOXPdY3r37q0CBQro2LFjDn3tKVOmaM2aNQ7t82YMw9ATTzzh0D4jIiIUERHh0D4BAADgOARWAAC4mEGDBunSpUtatmxZlvv/+usvrV69Wh07dlRAQIBDX9sZgRUAAADyHwIrAABcTLt27RQUFKQFCxZkuX/58uW6ePGiBg0alMeV5c6VK1eUkpLi7DIAAABgIQRWAAC4GHd3d/Xr1087d+7Ud999l2n/woULVbp0abVr106SdPToUQ0ZMkRly5aVp6enQkNDFRMTkykkSk5O1qRJk1S1alV5e3vL399fLVq00JdffikpfWre+fPntWjRIhmGIcMw7KbVff/99+rSpYuKFSsmb29v1alTR4sWLbJ7jfj4eBmGocWLF2v06NEqU6aMvLy89PPPP2f7/Wf0sXz5ck2YMEFBQUHy8/NTq1attH//frtjTdNUbGysypUrJ29vb9WrV0/r16/Pst+zZ8/qySefVGhoqDw9PVWmTBmNHDlS58+ftx3z2GOPydvbWzt37rRtS0tLU2RkpAICApSUlJTt9wEAAIDr83B2AQAAIOcGDhyoF198UQsWLNCrr75q27537159/fXXevrpp+Xu7q6jR4/q7rvvlpubm5577jlVqFBBCQkJmjx5sg4cOKCFCxdKklJSUtSuXTtt27ZNI0eOVMuWLZWSkqLt27fr4MGDaty4sRISEtSyZUu1aNFCzz77rCTJz89PkrR//341btxYpUqV0muvvSZ/f38tWbJE/fv317FjxzRmzBi7+seNG6fw8HDNnTtXbm5uKlWqVI4/g/Hjx6tJkyZ6++23dfbsWY0dO1adOnXSDz/8IHd3d0lSTEyMYmJiNGjQIHXv3l2HDh3SI488otTUVIWFhdn6unDhgpo3b64//vhD48ePV61atbRnzx4999xz+u677/Tpp5/KMAzNmDFDX331lR588EHt3LlTRYsWVUxMjOLj47VhwwaVLl06x+8DAAAAWTABAIBLat68uVmiRAnz8uXLtm2jR482JZk//vijaZqmOWTIENPX19f8/fff7Z47bdo0U5K5Z88e0zRN85133jElmW+99dYNX7NQoUJmv379Mm3v0aOH6eXlZR48eNBue7t27cyCBQuaZ86cMU3TNLds2WJKMps1a5bt9ynJHDZsmO1xRh/t27e3O+69994zJZkJCQmmaZrm6dOnTW9vb/O+++6zO+6LL74wJZnNmze3bZs6darp5uZmJiYm2h37/vvvm5LMjz/+2Lbtp59+Mv38/MyuXbuan376qenm5mY+88wz2X4/AAAAuDmmBAIA4KIGDRqkEydOaO3atZLSR0ktWbJETZs2VaVKlSRJ69atU4sWLRQUFKSUlBRby5guuHXrVknS+vXr5e3trYEDB+aqls2bNysyMlLBwcF22/v3768LFy4oISHBbvv999+fq9e5WufOne0e16pVS5L0+++/S5ISEhJ06dIl9e7d2+64xo0bq1y5cnbb1q1bpxo1aqhOnTp2n1ObNm1kGIbi4+Ntx1asWFFvvfWW1qxZo44dO6pp06aKjo6+5fcDAACA/yGwAgDARXXv3l1FihSxTev7+OOPdezYMbvF1o8dO6YPP/xQBQoUsGvVq1eXJJ04cUKS9OeffyooKEhubrm7NTh58mSW0+GCgoJs+6/miKlz/v7+do+9vLwkSRcvXrR7zcDAwEzPvXbbsWPHtHv37kyfU+HChWWapu1zytChQwcFBATo0qVLGjVqlG0KIgAAAByDNawAAHBRPj4+6tmzp9566y0lJSVpwYIFKly4sB544AHbMSVKlFCtWrX0wgsvZNlHRqBUsmRJff7550pLS8tVaOXv75/lguNHjhyx1XE1wzBy/Bq5qUlKX3T+WkePHlVISIjtcYkSJeTj43Pdb168tv7HHntM586dU/Xq1TV8+HA1bdpUxYoVc1zxAAAA+RwjrAAAcGGDBg1SamqqXn75ZX388cfq0aOHChYsaNvfsWNHff/996pQoYIaNGiQqWUEVu3atdOlS5cUFxd3w9fz8vKyjWC6WmRkpDZv3mwLqDK88847KliwoBo1anTrbzaHGjVqJG9vby1dutRu+5dffmmbNpihY8eO+uWXX+Tv75/l53R1uPX2229ryZIlmjVrltauXaszZ85owIABefGWAAAA8g1GWAEA4MIaNGigWrVqacaMGTJN0246oCRNmjRJmzZtUuPGjTV8+HCFhYXp0qVLOnDggD7++GPNnTtXZcuWVc+ePbVw4UI99thj2r9/v1q0aKG0tDR99dVXqlq1qnr06CFJqlmzpuLj4/Xhhx+qdOnSKly4sMLCwjRx4kTbelnPPfecihcvrqVLl+qjjz5SbGysihQpkuefTbFixfTkk09q8uTJGjx4sB544AEdOnRI0dHRmaYEjhw5UitXrlSzZs3073//W7Vq1VJaWpoOHjyoTz75RKNHj9Y999yj7777TsOHD1e/fv1sIdX8+fPVvXt3zZgxQyNHjszz9wkAAHAnIrACAMDFDRo0SCNGjFC1atV0zz332O0rXbq0duzYoeeff14vv/yy/vjjDxUuXFihoaFq27atbRqbh4eHPv74Y02dOlXLly/XjBkzVLhwYdWuXVtt27a19Tdz5kwNGzZMPXr00IULF9S8eXPFx8crLCxMX375pcaPH69hw4bp4sWLqlq1qhYuXKj+/fvn5cdhZ9KkSSpUqJBmz56txYsXq0qVKpo7d66mTZtmd1yhQoW0bds2vfjii5o3b55+++03+fj46K677lKrVq0UEhKi8+fP68EHH1RoaKhmz55te+7999+vYcOGacyYMWrcuLHuvvvuvH6bAAAAdxzDNE3T2UUAAAAAAAAAGVjDCgAAAAAAAJZCYAUAAAAAAABLIbACAAAAAACApRBYAQAAAAAAwFIIrAAAAAAAAGApBFYAAAAAAACwFAIrAAAAAAAAWAqBFQAAAAAAACyFwAoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKQRWAAAAAAAAsBQCKwAAAAAAAFgKgRUAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClEFgBAAAAAADAUgisAAAAAAAAYCkEVgAAAAAAALAUAisAAAAAAABYCoEVAAAAAAAALIXACgAAAAAAAJZCYAUAAAAAAABLIbACAAAAAACApRBYAQAAAAAAwFIIrAAAAAAAAGApBFYAAAAAAACwFAIrAAAAAAAAWAqBFQAAAAAAACyFwAoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKQRWAAAAAAAAsBQCKwAAAAAAAFgKgRUAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClEFgBAAAAAADAUgisAAAAAAAAYCkEVgAAAAAAALAUAisAAAAAAABYCoEVAAAAAAAALIXACgAAAAAAAJZCYAUAAAAAAABLIbACAAAAAACApRBYAQAAAAAAwFIIrAAAAAAAAGApBFYAAAAAAACwFAIrAAAAAAAAWAqBFQAAAAAAACyFwAoAAAAAAACWQmAFAAAAAAAASyGwAgAAAAAAgKUQWAEAAAAAAMBSCKwAAAAAAABgKQRWAAAAAAAAsBQCKwAAAAAAAFgKgRUAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClEFgBAADkQkhIiGbMmOHsMgAAAO5IBFYAANwBjh49qhEjRqhixYry9vZWQECA7r33Xs2dO1cXLlxwdnnXRegDAACArHg4uwAAAHBrfv31VzVp0kRFixbVlClTVLNmTaWkpOjHH3/UggULFBQUpM6dO2f53CtXrqhAgQJ5XDEAAABwY4ywAgDAxQ0dOlQeHh7asWOHHnzwQVWtWlU1a9bU/fffr48++kidOnWyHWsYhubOnasuXbqoUKFCmjx5siRpzpw5qlChgjw9PRUWFqbFixfbnnPgwAEZhqFvv/3Wtu3MmTMyDEPx8fGSpPj4eBmGoY8++ki1a9eWt7e37rnnHn333Xe39N4+/PBD1a9fX97e3ipfvrxiYmKUkpJi279v3z7de++98vb2VrVq1fTpp5/KMAytWbPGdszhw4f10EMPqVixYvL391eXLl104MAB2/7+/fura9eumjZtmkqXLi1/f38NGzZMV65csR1z/PhxderUST4+PgoNDdXSpUtv6X0BAADgxgisAABwYSdPntQnn3yiYcOGqVChQlkeYxiG3eOJEyeqS5cu+u677zRw4ECtXr1aI0aM0OjRo/X9999ryJAhGjBggLZs2ZLjep566ilNmzZNiYmJKlWqlDp37mwX/OTExo0b9fDDD2v48OHau3ev3nzzTcXFxemFF16QJKWlpalr164qWLCgvvrqK82bN08TJkyw6+PChQtq0aKFfH199Z///Eeff/65fH191bZtW12+fNl23JYtW/TLL79oy5YtWrRokeLi4hQXF2fb379/fx04cECbN2/W+++/r9mzZ+v48eO5el8AAAC4OQIrAABc2M8//yzTNBUWFma3vUSJEvL19ZWvr6/Gjh1rt69Xr14aOHCgypcvr3LlymnatGnq37+/hg4dqsqVK2vUqFHq1q2bpk2bluN6Jk6cqKioKNWsWVOLFi3SsWPHtHr16ly9txdeeEFPP/20+vXrp/LlyysqKkrPP/+83nzzTUnSJ598ol9++UXvvPOOateurXvvvdcWZmV499135ebmprfffls1a9ZU1apVtXDhQh08eNA2OkySihUrplmzZqlKlSrq2LGjOnTooM8++0yS9OOPP2r9+vV6++23FR4ervr162v+/Pm6ePFirt4XAAAAbo41rAAAuANcO4rq66+/Vlpamnr37q3k5GS7fQ0aNLB7/MMPP+jRRx+129akSRPNnDkzx3WEh4fb/ly8eHGFhYXphx9+yHE/krRz504lJibahVCpqam6dOmSLly4oP379ys4OFiBgYG2/XfffXemPn7++WcVLlzYbvulS5f0yy+/2B5Xr15d7u7utselS5e2TWf84Ycf5OHhYfe5ValSRUWLFs3V+wIAAMDNEVgBAODCKlasKMMwtG/fPrvt5cuXlyT5+Phkek5WUwevDbxM07Rtc3Nzs23LkJNpftf2nV1paWmKiYlRt27dMu3z9va2q/FGfdSvXz/LNadKlixp+/O1C88bhqG0tDRJ/3vfuX0fAAAAyDmmBAIA4ML8/f0VFRWlWbNm6fz587nqo2rVqvr888/ttn355ZeqWrWqpP8FO0lJSbb9Vy/AfrXt27fb/nz69Gn9+OOPqlKlSq7qqlevnvbv36+KFStmam5ubqpSpYoOHjyoY8eO2Z6TmJiYqY+ffvpJpUqVytRHkSJFslVH1apVlZKSoh07dti27d+/X2fOnMnV+wIAAMDNMcIKAAAXN3v2bDVp0kQNGjRQdHS0atWqJTc3NyUmJmrfvn2qX7/+DZ//1FNP6cEHH1S9evUUGRmpDz/8UKtWrdKnn34qKX2UVqNGjfTiiy8qJCREJ06c0DPPPJNlX5MmTZK/v78CAgI0YcIElShRQl27dr3h6x8+fDhTAHbXXXfpueeeU8eOHRUcHKwHHnhAbm5u2r17t7777jtNnjxZUVFRqlChgvr166fY2FidO3fOtuh6xmio3r176+WXX1aXLl00adIklS1bVgcPHtSqVav01FNPqWzZsjf9fMPCwtS2bVs98sgjmjdvnjw8PDRy5MgsR68BAADAMRhhBQCAi6tQoYJ27dqlVq1aady4capdu7YaNGig119/XU8++aSef/75Gz6/a9eumjlzpl5++WVVr15db775phYuXKiIiAjbMQsWLNCVK1fUoEEDjRgxQpMnT86yrxdffFEjRoxQ/fr1lZSUpLVr18rT0/OGrz9t2jTVrVvXrq1du1Zt2rTRunXrtGnTJjVs2FCNGjXS9OnTVa5cOUmSu7u71qxZo7///lsNGzbU4MGDbUGat7e3JKlgwYL6z3/+o7vuukvdunVT1apVNXDgQF28eFF+fn7Z/Yi1cOFCBQcHq3nz5urWrZseffRRlSpVKtvPBwAAQM4Y5tULUgAAAORCfHy8WrRoodOnTzt1MfIvvvhC9957r37++WdVqFDBaXUAAADg1jAlEAAAuKzVq1fL19dXlSpV0s8//6wRI0aoSZMmhFUAAAAujsAKAAC4rHPnzmnMmDE6dOiQSpQooVatWumVV15xdlkAAAC4RUwJBAAAAAAAgKWw6DoAAAAAAAAshcAKAAAAAAAAlkJgBQAAAAAAAEshsAIAAAAAAIClEFgBAAAAAADAUgisAAAAAAAAYCkEVgAAAAAAALAUD0d0cvZysi6mpjiiq1zzcfeQn6dXrp4bFxenAQMG2B67u7urZMmSat68uZ5//nlVqlTJUWXiDmGePSld/Nu5Rfj4yvDzz/XTd+/erVdffVXx8fFKSkqSh4eHKleurB49emjw4MEqXrx4jvqLjo5WTEyMTNO0bYuIiJAkxcfH57rO3LBSLXciwzA0ceJERUdH35b+p0yZomrVqqlr16522+Pj49WiRQtt2bLF9veZb11JklLPOLcG96JSgdI5ftq1P3Ov5ei/36yuB9kRERGhEydO6Pvvv3dYLXnlyJEjmjdvnrp27ao6deo4u5wbut3ndUhIiCIiIhQXF+fwvoFb4er3khnXci8vL+3fv1/lypWz2+/K19D+/fsrPj5eBw4csG27ndeSCxcuKDY2VhEREZmugxmf82+//aaQkBCHvzZgdbccWJ29nKyFP32r1BzeCDqau2FoQKU6uQ6tJGnhwoWqUqWKLl26pC+++EIvvPCCtmzZon379qlYsWIOrBauzDx7UmkLJ0ipV5xbiHsBuQ14IVc3Gm+99ZaGDh2qsLAwPfXUU6pWrZquXLmiHTt2aO7cuUpISNDq1atvQ9G4EyQkJKhs2bK3rf8pU6aoe/fumQKrevXqKSEhQdWqVbttr+0SriRJB7pI5mXn1mF4SiEf5Cq0kv73M/dajv77HTx4sNq2bevQPq3uyJEjiomJUUhIiOUDq9tt9erV8vPzc3YZgJ074V4yQ3Jysp555hktXrzYgYVZz+28lly4cEExMTGSlCmw6tChgxISElS6dO5+1gKu7pYDq4upKU4PqyQp1TR1MTVFfsp9YFWjRg01aNBAUvrFIjU1VRMnTtSaNWtu+Ntg5DMX/3b+DYaUXsPFv6Uc3mQkJCTo8ccfV1RUlNasWSMvr/+dM1FRURo9erQ2bNjg6GpxB2nUqJFTXtfPz89pr20pqWecH1ZJ6TWknsl1YHX1z9zb4cKFCypYsKDKli17WwNWWFvdunWdXQKQmYvfS16tbdu2WrZsmZ588knVrl3bgcX9z8WLF+Xj43Nb+s4uZ11LSpYsqZIlSzrltQErYA2rG8i4kT527JiTKwEcZ8qUKTIMQ/PmzbMLqzJ4enqqc+fOdttWrFih8PBwFSpUSL6+vmrTpo127drl0LrS0tL0+uuvq06dOvLx8VHRokXVqFEjrV27Ns9rWbZsmcLDw+Xr6ytfX1/VqVNH8+fPtztmwYIFql27try9vVW8eHHdd999+uGHH+yO6d+/v3x9fbVv3z61adNGhQoVUunSpfXiiy9KkrZv3657771XhQoVUuXKlbVo0SK758fFxckwDG3evFmPPPKI/P395efnp759++r8+fM6evSoHnzwQRUtWlSlS5fWk08+qStX/ncDHB8fL8MwMk2DPHDggAzDsBvWnlHrzz//rPbt28vX11fBwcEaPXq0kpOT7Z5vGEam6YCHDx/Wo48+quDgYHl6eiooKEjdu3e3XT8vXbqk0aNHq06dOipSpIiKFy+u8PBwffDBB5n6Pn/+vBYtWiTDMGQYht10zqzez9q1axUeHq6CBQuqcOHCioqKUkJCgt0x0dHRMgxDe/bsUc+ePVWkSBEFBARo4MCB+uuvvwRrMgxDTzzxhBYvXqyqVauqYMGCql27ttatW2d3XMbf7zfffKPu3burWLFiqlChgt2+a2XnPJekxMRENW3aVAULFlT58uX14osvKi0tzbY/49/lsmXLNHbsWJUuXVq+vr7q1KmTjh07pnPnzunRRx9ViRIlVKJECQ0YMEB//20/Dcg0Tc2ePdt2/StWrJi6d++uX3/91e64iIgI1ahR44Y1xcfHq2HDhpKkAQMG2M6jG03h/fPPPzV06FBVq1ZNvr6+KlWqlFq2bKlt27bZHZdx7Zg2bZqmT5+u0NBQ+fr6Kjw8XNu3b7c7dseOHerRo4dCQkLk4+OjkJAQ9ezZU7///vt165CkxYsXyzCMTOewJE2aNEkFChTQkSNHJEm7du1Sx44dVapUKXl5eSkoKEgdOnTQH3/8YXtOSEiI+vfvb3uclpamyZMnKywszPazplatWpo5c+YN6wKQtTFjxsjf319jx4694XGXLl3SuHHjFBoaKk9PT5UpU0bDhg3TmTNn7I4LCQlRx44dtWrVKtWtW1fe3t6KiYlxyLX2jTfeULNmzVSqVCkVKlRINWvWVGxsrN290/Vcey2JiIiwXV+vbRn3V9m5th44cMAWSMXExNj6yHitjHvBq6cnSjm7D83OvR1gVQRWN/Dbb79JkipXruzkSgDHSE1N1ebNm1W/fn0FBwdn6zlTpkxRz549Va1aNb333ntavHixzp07p6ZNm2rv3r0Oq61///4aMWKEGjZsqBUrVujdd99V586d7X5A50Utzz33nHr37q2goCDFxcVp9erV6tevn91/sqZOnapBgwapevXqWrVqlWbOnKndu3crPDxcP/30k11/V65cUbdu3dShQwd98MEHateuncaNG6fx48erX79+GjhwoFavXq2wsDD1799fO3fuzFTT4MGDVaRIEb377rt65plntGzZMj3yyCPq0KGDateurffff1/9+vXTK6+8otdffz3X7/3KlSvq3LmzIiMj9cEHH2jgwIF69dVX9dJLL93weYcPH1bDhg21evVqjRo1SuvXr9eMGTNUpEgRnT59WlL6lIFTp07pySef1Jo1a7R8+XLde++96tatm9555x1bXwkJCfLx8VH79u2VkJCghIQEzZ49+7qvvWzZMnXp0kV+fn5avny55s+fr9OnTysiIkKff/55puPvv/9+Va5cWStXrtTTTz+tZcuW6d///ncuPzHcitTUVKWkpNi11NTUTMd99NFHmjVrliZNmqSVK1fabsyvDXMkqVu3bqpYsaL+7//+T3Pnzr3ua2fnPJeko0ePqnfv3nr44Ye1du1a2/m7ZMmSTH2OHz9ex48fV1xcnF555RXFx8erZ8+euv/++1WkSBEtX75cY8aM0eLFizV+/Hi75w4ZMkQjR45Uq1attGbNGs2ePVt79uxR48aNM/3S7GY11atXTwsXLpQkPfPMM7bzaPDgwdf9PE6dOiVJmjhxoj766CMtXLhQ5cuXV0RERJbr/r3xxhvatGmTZsyYoaVLl+r8+fNq3769Xfh74MABhYWFacaMGdq4caNeeuklJSUlqWHDhjpx4sR1a3nooYcUGBioN954w257SkqK3nzzTd13330KCgrS+fPnFRUVpWPHjtnVc9ddd+ncuXPX7T82NlbR0dHq2bOnPvroI61YsUKDBg3K9J9mANlTuHBhPfPMM9q4caM2b96c5TGmaapr166aNm2a+vTpo48++kijRo3SokWL1LJly0zhyTfffKOnnnpKw4cP14YNG3T//ffb9t3KtfaXX35Rr169tHjxYq1bt06DBg3Syy+/rCFDhuT4fc+ePdt2fc1orVq1kru7u8LCwiRl79paunRp28yGQYMG2fp69tlnr/vaOb0Pzc29HWAVDll0/U6RcfOcsYbV5MmT1axZs0yjTQBXdeLECV24cEGhoaHZOv7QoUOaOHGinnjiCb322mu27VFRUapUqZJiYmK0YsWKW65r27ZtWrx4sSZMmKDJkyfbtl+97kxe1PLbb79pypQp6t27t91/SKOiomx/PnPmjJ5//nm1b99ey5Yts22PiIhQpUqVFB0draVLl9q2X758WZMnT1a3bt1sx61bt05Tp07VN998Yxti3qBBA5UqVUrLli1T/fr17erq2LGjpk2bZqslISFBy5cv1/Tp021hS6tWrbRx40YtXbpUo0aNytX7v3z5smJiYvTAAw9IkiIjI7Vjxw4tW7ZMzz333HWf99xzz+nEiRP673//q6pVq9q2P/jgg7Y/FylSxPafaCn9ehsZGanTp09rxowZ6tu3r6T06YZubm4qWbLkTaf/paWl6amnnlLNmjW1fv16ubml/w6mffv2qlChgsaOHasvvvjC7jmDBg3SU089JSn9M/v555+1YMECzZ8/P8tROLh9svr7dXd3V0qK/Ze4XLx4UZ9++qkKFy4sKT2QCQoK0nvvvaenn37a7th+/frZ1gG5nuyc5xlOnjypjz/+WHfffbek9H8z8fHxWrZsme3fbIZatWrZ/Rvft2+fZsyYoeHDh+vll1+2vUZCQoKWLl1qu45t375db731ll555RW7c7dp06aqXLmypk+fbvcfi5vV5Ofnpxo1akiSKlSokK1ptGFhYXbBcGpqqtq0aaMDBw7otddey7SmSuHChbVu3Tq5u7tLkoKCgnT33Xdr/fr16tGjhySpe/fu6t69u12fHTt2VEBAgJYtW6bhw4dnWYunp6eGDBmiqVOnavr06SpVqpQkadWqVTpy5IieeOIJ2+d78uRJzZ8/X126dLE9/+rrTla++OIL1axZ027EWZs2bW7yCQG4kccee0wzZ87U2LFj9fXXX2f6efrJJ59o48aNio2Ntf0MjoqKUnBwsB566CG98847euSRR2zHHz9+XHv37rUbNJAR8OT2WitJ06dPt/05LS1NTZs2lb+/vwYMGKBXXnklR2sWX7ve4rRp0/Tpp59q3rx5Cg8Pl5S9a6uXl5ftvq9s2bI3vWbn5j40N/d2gFUwwuoqjRo1UoECBVS4cGG1bdtWxYoV0wcffCAPD3I95E8bN25USkqK+vbtazcKwtvbW82bN8/xN+5dO6IiYwrL+vXrJUnDhg3Ls1qysmnTJqWmpt6wjoSEBF28eNFuWLgkBQcHq2XLlvrss8/sthuGofbt29see3h4qGLFiipdurTdegjFixdXqVKlspwu07FjR7vHGaFQhw4dMm2/2XSbGzEMQ506dbLbVqtWrZv2uX79erVo0cIurMrK//3f/6lJkyby9fWVh4eHChQooPnz52cawp5d+/fv15EjR9SnTx9bWCVJvr6+uv/++7V9+3ZduHDB7jnX/gKiVq1aunTpko4fP56rGpB777zzjhITE+3aV199lem4Fi1a2MIqSQoICLjuuXL1b+GvJzvneYbAwEBbMJTheudETs7TU6dO2aaqrFu3ToZh6OGHH7a7tgUGBqp27dqZrm05qSkn5s6dq3r16snb29t2fn722WdZnp8dOnSwhVUZry/Jroa///5bY8eOVcWKFeXh4SEPDw/5+vrq/PnzNz3nH3/8cUnpXxCSYdasWapZs6aaNWsmSapYsaKKFSumsWPHau7cudkeZXv33Xfrv//9r4YOHaqNGzfq7Nmz2XoegOvz9PTU5MmTtWPHDr333nuZ9meMvLr23umBBx5QoUKFMt071apV67ozXHJ7rZXSpxF37txZ/v7+cnd3V4ECBdS3b1+lpqbqxx9/zN6bzULGqK5nnnnGLniTcnZtzY7c3Ifm5t4OsAoCq6tk3Dxv3rxZQ4YM0Q8//KCePXs6uyzAYUqUKKGCBQvaprveTMZUlIYNG6pAgQJ2bcWKFTec1pGVyMhIuz4GDhwoKX2Ov7u7uwIDA/Oslqz8+eefknTDBZpPnjwpSVl+W0tQUJBtf4aCBQvK29vbbpunp6eKFy+e6fmenp66dOlSpu3XHuvp6Xnd7Vk9P7uyqtXLy+umff755583XdR61apVevDBB1WmTBktWbJECQkJSkxM1MCBA3Nd883+LtLS0mxTEjP4+9svLJuxjtvFixdzVQNyr2rVqmrQoIFdu3Z0oZT570xK/3vL6u8sO9+ilJ3zPDevnZPzVJLt3/2xY8dkmqYCAgIyXdu2b9+e6dqWk5qya/r06Xr88cd1zz33aOXKldq+fbsSExPVtm3bLPvNznnUq1cvzZo1S4MHD9bGjRv19ddfKzExUSVLlrxprQEBAXrooYf05ptvKjU1Vbt379a2bdtso6uk9FGbW7duVZ06dTR+/HhVr15dQUFBmjhx4g3Xoxk3bpymTZum7du3q127dvL397eNOACQez169FC9evU0YcKETOfgyZMn5eHhkWnxcMMwFBgYmOne6UbX8txeaw8ePKimTZvq8OHDmjlzprZt26bExETb9OPcXkO3bNmi/v37q2/fvnr++eft9uX02podjrgPzc69HWAVDB26SsbNs5T+G93U1FS9/fbbev/99+2GtQOuyt3dXZGRkVq/fr3++OOPm/6HrUSJEpKk999/X+XKlbvl13/zzTft1hbJ6L9kyZJKTU3V0aNHr3uT4uhaspJxI/XHH39cd42vjP+oJSUlZdp35MgRW51WkHGDcu3aEI4I965WsmRJu0WOs7JkyRKFhoZqxYoVdlMFbmXRz5v9Xbi5ueVoeD9cX3amdWbnPM9LJUqUkGEY2rZtW5ZfhJHVNkdbsmSJIiIiNGfOHLvtN1oL6kb++usvrVu3ThMnTrSbtpmxll12jBgxQosXL9YHH3ygDRs2qGjRourdu7fdMTVr1tS7774r0zS1e/duxcXFadKkSfLx8ck0XTSDh4eHRo0apVGjRunMmTP69NNPNX78eLVp00aHDh1SwYIFc/WegfzOMAy99NJLioqK0rx58+z2+fv7KyUlRX/++addaGWapo4ePWr7ooir+3K0NWvW6Pz581q1apXdfeS3336b6z53796trl27qnnz5nYjQjM4+toqudZ9KOAIjLC6gdjYWBUrVkzPPfec3TcCAa5s3LhxMk1TjzzyiC5fvpxp/5UrV/Thhx9KSl/Xw8PDQ7/88kumkRAZLSfCwsLsnhsSEiJJateunSRl+oF+NUfXkpXWrVvL3d39hnWEh4fLx8cn06LLf/zxhzZv3qzIyMhbrsNRMj7f3bt3222/9psXb1W7du20ZcsW7d+//7rHGIYhT09Pu5vQo0ePZvqWQCn7o0XCwsJUpkwZLVu2TKZp2rafP39eK1eutH1zIHC17Jznealjx44yTVOHDx/O8rpWs2bNHPeZ05GDhmFkCsZ2796d5Tf1Zbc/0zQz9fn2229nubB+VurXr6/GjRvrpZde0tKlS9W/f38VKlTouq9Xu3ZtvfrqqypatKi++eabbL1G0aJF1b17dw0bNkynTp3K9C1cAHKmVatWioqK0qRJk+ym4mXcG11777Ry5UqdP38+T+6dMu4/rr4umaaZZdCUHQcPHlS7du1Uvnx5rVy5UgUKFMjyNbNzbc3JNduV7kMBR2CE1Q0UK1ZM48aN05gxY7Rs2TI9/PDDzi4JuGXh4eGaM2eOhg4dqvr16+vxxx9X9erVdeXKFe3atUvz5s1TjRo11KlTJ4WEhGjSpEmaMGGCfv31V9vabseOHdPXX3+tQoUK3XSB4+xo2rSp+vTpo8mTJ+vYsWPq2LGjvLy8tGvXLhUsWFD/+te/8qSWkJAQjR8/Xs8//7wuXryonj17qkiRItq7d69OnDihmJgYFS1aVM8++6zGjx+vvn37qmfPnjp58qRiYmLk7e2tiRMn3vLn4SiBgYFq1aqVpk6dqmLFiqlcuXL67LPPtGrVKoe+zqRJk7R+/Xo1a9ZM48ePV82aNXXmzBlt2LBBo0aNUpUqVWxfUT106FB1795dhw4d0vPPP6/SpUtn+kabmjVrKj4+Xh9++KFKly6twoUL275x52pubm6KjY1V79691bFjRw0ZMkTJycl6+eWXdebMGb344osOfZ9wrO+//z7TAutS+kLh104bcaTsnOd5qUmTJnr00Uc1YMAA7dixQ82aNVOhQoWUlJSkzz//XDVr1rSt6ZRdFSpUkI+Pj5YuXaqqVavK19dXQUFBCgoKyvL4jh076vnnn9fEiRPVvHlz7d+/X5MmTVJoaGiWf0c34+fnp2bNmunll19WiRIlFBISoq1bt2r+/PkqWrRotvsZMWKEHnroIRmGoaFDh9rtW7dunWbPnq2uXbuqfPnyMk1Tq1at0pkzZ7JcQD9Dp06dVKNGDTVo0EAlS5bU77//rhkzZqhcuXKqVKlSjt8rAHsvvfSS6tevr+PHj6t69eqS0hdBb9OmjcaOHauzZ8+qSZMm2r17tyZOnKi6deuqT58+t72uqKgoeXp6qmfPnhozZowuXbqkOXPmZFo6ILvatWunM2fOaNasWdqzZ4/dvoyfY9m9thYuXFjlypXTBx98oMjISBUvXtx27byWK92HAo5AYHUT//rXv2xfp92zZ0+7RUYBV/XII4/o7rvvtn2t7dGjR1WgQAFVrlxZvXr1slsnZNy4capWrZpmzpyp5cuXKzk5WYGBgWrYsKEee+wxh9UUFxenevXqaf78+YqLi5OPj4+qVatm95XEeVHLpEmTVKlSJb3++uvq3bu3PDw8VKlSJbtvtBo3bpxKlSql1157TStWrJCPj48iIiI0ZcoUy/2HZ/HixfrXv/6lsWPHKjU1VZ06ddLy5csdMiItQ5kyZfT1119r4sSJevHFF3Xy5EmVLFlS9957r21NiQEDBuj48eOaO3euFixYoPLly+vpp5/WH3/8kSkgmDlzpoYNG6YePXrowoULN1xUv1evXipUqJCmTp2qhx56SO7u7mrUqJG2bNmixo0bO+w9wvEGDBiQ5fa33npLgwcPvq2vnZ3zPC+9+eabatSokd58803Nnj1baWlpCgoKUpMmTTItsJ4dBQsW1IIFCxQTE6PWrVvrypUrmjhxot03411twoQJunDhgubPn6/Y2FhVq1ZNc+fO1erVq3P9hRbLli3TiBEjNGbMGKWkpKhJkybatGlTpoWRb6Rr167y8vJSixYtMl1bK1WqpKJFiyo2NlZHjhyRp6enwsLCFBcXp379+l23zxYtWmjlypV6++23dfbsWQUGBioqKkrPPvtsliMkAORM3bp11bNnT7tvsDMMQ2vWrFF0dLQWLlyoF154QSVKlFCfPn00ZcqUPJn6XKVKFa1cuVLPPPOMunXrJn9/f/Xq1UujRo2yjfTPiYwvesj4FuirLVy4UP3798/RtXX+/Pl66qmn1LlzZyUnJ6tfv36Ki4vL8rVd6T4UuFWGefU8ilw4ezlZC3/6Vqm31s0tczcMDahUR36et/+Ch/zNPHtSaQsnSKnXX9Q1T7gXkNuAF2T4ZV6AF8Ad7EqSdKCLZGae0punDE8p5AOpwM0XOgdy48MPP1Tnzp310Ucf2X3bKuDquJcEgOy55cBKSg+tLqbmfMi4I/m4exBWIc+YZ09KF/+++YG3k48vNxhAfnUlSUo949wa3IsSVuG22Lt3r37//XeNGDFChQoV0jfffHNbFmEGnIl7SQC4OYcEVgAAAIAjRERE6IsvvlC9evW0aNEiValSxdklAQAAJyCwAgAAAAAAgKW4ObsAAAAAAAAA4GoEVgAAAAAAALAUAisAAAAAAABYCoEVAAAAAAAALIXACgAAAAAAAJZCYAUAAAAAAABLIbACAAAAAACApXg4ppuLki47pqtc85Tk4+QakF+c/TtZF5NTnFqDj5eH/Hy9nFoDAOdIST6ntCuXnFqDWwFveXgVdmoNAOCquJcEgJtzQGB1UdIWSWm33tUtcZPUQlYIraZMmaJq1aqpa9eudtvj4uI0YMAAJSYmqkGDBs4pDrfs7N/JWrjme6Wmmk6tw93d0ICuNW7pRmP37t169dVXFR8fr6SkJHl4eKhy5crq0aOHBg8erOLFiysiIkKSFB8f75jCrzF79mwVLFhQ/fv3vy39G4ahiRMnKjo6+rb0D+S1lORzOr5zqWSmOrcQw12l6vfOcWiV8bMwK6NHj9a0adMcUZ2dvXv36r333lP//v0VEhLi8P6t4sCBAwoNDdXChQtvyzX1woULio2NVUREhO1nQ4aMv9fffvvtjv6MAUe4k+4lt2/frldeeUWff/65Tp48KX9/fzVt2lRPPfWUGjZs6MBqs9a/f3/Fx8frwIEDtm0hISGKiIhQXFycJOnIkSOaN2+eunbtqjp16tz2mm63ZcuW6fjx4xo5cqTD++ZaDqtxQGB1Wc4Pq6T0Gi7LKoFV9+7dMwVWuDNcTE5x+g2GJKWmmrqYnJLrm4y33npLQ4cOVVhYmJ566ilVq1ZNV65c0Y4dOzR37lwlJCRo9erVDq46s9mzZ6tEiRK3LbBKSEhQ2bJlb0vfgDOkXbnk/LBKkszU9FpyOcpq4cKFqlKlit22oKAgR1SWyd69exUTE6OIiAhuwG/BhQsXFBMTI0mZAqsOHTooISFBpUuXdkJlgGu5U+4lX3/9dY0cOVJ33323YmNjVa5cOR08eFBvvPGGGjVqpDlz5ujRRx91cNU3t3r1avn5+dkeHzlyRDExMQoJCbljAqvvv//+tgRWgNU4aEogAFeSkJCgxx9/XFFRUVqzZo28vP53oxIVFaXRo0drw4YNTqzw1pimqUuXLsnHx0eNGjVydjkAslCjRg2XH2185coVGYYhDw9up0qWLKmSJUs6uwwAeeSLL77QyJEj1b59e61evdruOtijRw/dd999Gjp0qOrWrZsnI62uVrdu3Tx9PStLTU1VSkqK3b0+4EpYdF1SdHS0DMPQrl271K1bN/n5+alIkSJ6+OGH9eeff0qSBg0apOLFi+vChQuZnt+yZUtVr15dUvr0o/Pnz2vRokUyDEOGYWT6LeS5c+f0+OOPq0SJEvL391e3bt105MgRu2PS0tIUGxurKlWqyMvLS6VKlVLfvn31xx9/2B0XERGhGjVqKDExUU2bNlXBggVVvnx5vfjii0pLs8LIN1jRlClTZBiG5s2bl+UPME9PT3Xu3DnL58bHx8swjExTBA8cOCDDMGzDryXp119/VY8ePRQUFCQvLy8FBAQoMjJS3377raT0Idt79uzR1q1bbefL1aMfzp49qyeffFKhoaHy9PRUmTJlNHLkSJ0/f97utQ3D0BNPPKG5c+eqatWq8vLy0qJFi2z7rp4OGBcXJ8MwtGXLlpueh8nJyRo9erQCAwNVsGBBNWvWTDt37lRISMhtGxEGIN2KFSsUHh6uQoUKydfXV23atNGuXbvsjtmxY4d69OihkJAQ+fj4KCQkRD179tTvv/9uOyYuLk4PPPCAJKlFixa2a03Gtep65/O1094yrn2LFy/W6NGjVaZMGXl5eennn3+WJH366aeKjIyUn5+fChYsqCZNmuizzz676ftMS0vT5MmTFRYWJh8fHxUtWlS1atXSzJkz7Y776aef1KtXL5UqVUpeXl6qWrWq3njjjex8lNl+7pkzZzR69GiVL1/edu/Rvn177du3TwcOHLAFUjExMbbPMeOzy7i2Xj0tR5IWLFig2rVry9vbW8WLF9d9992nH374we6Y/v37y9fXVz///LPat28vX19fBQcHa/To0UpOTs7WewSQt6ZOnSrDMDRnzpxMob2Hh4dmz55tOy7D9aZlZ/xf7GpvvPGGmjVrplKlSqlQoUKqWbOmYmNjdeXKlZvWdvV1PT4+3haYDRgwwHbtio6O1uLFi2UYhhISEjL1MWnSJBUoUCDTveG19u3bp549eyogIEBeXl6666671LdvX7tr1/fff68uXbqoWLFi8vb2Vp06dWz3qRkyfsYsX75cEyZMUFBQkPz8/NSqVSvt37/fdlxERIQ++ugj/f7777b3kvHZZdyLx8bGavLkyQoNDZWXl5e2bNkiSVq7dq3Cw8NVsGBBFS5cWFFRUVm+d8BKCKyuct9996lixYp6//33FR0drTVr1qhNmza6cuWKRowYodOnT2vZsmV2z9m7d6+2bNmiYcOGSUofueLj46P27dsrISFBCQkJtgt2hsGDB6tAgQJatmyZYmNjFR8fr4cfftjumMcff1xjx45VVFSU1q5dq+eff14bNmxQ48aNdeLECbtjjx49qt69e+vhhx/W2rVr1a5dO40bN05Lliy5DZ8SXF1qaqo2b96s+vXrKzg4+La+Vvv27bVz507FxsZq06ZNmjNnjurWraszZ85ISh+yXb58edWtW9d2vmRMQ7xw4YKaN2+uRYsWafjw4Vq/fr3Gjh2ruLg4de7cWaZpP5R+zZo1mjNnjp577jlt3LhRTZs2vWFt2TkPBwwYoBkzZmjAgAH64IMPdP/99+u+++6z1Q8g9zJ+63t1yzBlyhT17NlT1apV03vvvafFixfr3Llzatq0qfbu3Ws77sCBAwoLC9OMGTO0ceNGvfTSS0pKSlLDhg1tPys7dOigKVOmSEr/D1DGtaZDhw65qnvcuHE6ePCg5s6dqw8//FClSpXSkiVL1Lp1a/n5+WnRokV67733VLx4cbVp0+amoVVsbKyio6PVs2dPffTRR1qxYoUGDRpkd53Zu3evGjZsqO+//16vvPKK1q1bpw4dOmj48OG2KXrXk93nnjt3Tvfee6/efPNNDRgwQB9++KHmzp2rypUrKykpSaVLl7aNvB00aJDtc3z22Wev+9pTp07VoEGDVL16da1atUozZ87U7t27FR4erp9++snu2CtXrqhz586KjIzUBx98oIEDB+rVV1/VSy+9dMP3ByDvpaamasuWLWrQoMF1l10IDg5W/fr19emnn+bql+i//PKLevXqpcWLF2vdunUaNGiQXn75ZQ0ZMiRH/dSrV08LFy6UJD3zzDO2a9fgwYP10EMPKTAwMFOAn5KSojfffFP33XffDaeq//e//1XDhg21fft2TZo0SevXr9fUqVOVnJysy5fTv5Bs//79aty4sfbs2aPXXntNq1atUrVq1dS/f3/FxsZm6nP8+PH6/fff9fbbb2vevHn66aef1KlTJ6Wmpi8FMHv2bDVp0kSBgYG293Jt6PTaa69p8+bNmjZtmtavX68qVapo2bJl6tKli/z8/LR8+XLNnz9fp0+fVkREhD7//PMcfaZAXmIM+1W6detmu3C0bt1aAQEB6t27t9577z317t1bzZs31xtvvKHBgwfbnjNr1iz5+fmpb9++kqRGjRrJzc1NJUuWvO5UpLZt2+q1116zPT516pTGjBmjo0ePKjAwUPv27dO8efM0dOhQvf7667bj6tatq3vuuUevvvqqXnjhBdv2kydP6uOPP9bdd98tSWrVqpXi4+O1bNkyW11AhhMnTujChQsKDQ29ra9z8uRJ7d+/XzNmzLALgrp162b7c926deXj4yM/P79M58trr72m3bt366uvvrJNG4qMjFSZMmXUvXt3bdiwQe3atbMd//fff+u7775TsWLFslXfzc7DvXv3avny5Ro7dqztt4NRUVEKCAhQz549c/6BALCT1c/IK1euKCkpSRMnTtQTTzxhd45GRUWpUqVKiomJ0YoVKyRJ3bt3V/fu3W3HpKamqmPHjgoICNCyZcs0fPhwlSxZUpUqVZIkVatW7ZanCVeoUEH/93//Z3t84cIFjRgxQh07drRb9699+/aqV6+exo8fr6+++uq6/X3xxReqWbOm3UjQNm3a2B0zatQoFS5cWJ9//rltXZaoqCglJyfrxRdf1PDhw6977cvuc2fMmKE9e/Zo06ZNatWqle35V1+z69evL0kqW7bsTT/HM2fO6Pnnn1f79u3tftkXERGhSpUqKTo6WkuXLrVtv3z5smJiYmyj4SIjI7Vjxw4tW7ZMzz333A1fC0Deyu69ZGhoqL7++mudOnVKJUqUyNFrTJ8+3fbntLQ0NW3aVP7+/howYIBeeeWVbN/v+fn5qUaNGpLSr9/XXruGDBmiqVOnavr06SpVqpQkadWqVTpy5IieeOKJG/Y9atQoeXh46Ouvv7abEt27d2/bn6Ojo3X58mVt2bLF9ovi9u3b68yZM4qJidGQIUNUpEgR2/HVqlWzG3Tg7u6uBx98UImJiWrUqJGqVaumokWLysvL67rXYW9vb23cuFEFChSQlP75NWnSRDVr1tT69evl5uZmq6NChQoaO3asvvjiixt/kICTMMLqKldfXCTpwQcflIeHh20Y5YgRI/Ttt9/aTuizZ89q8eLF6tevn3x9fbP9OtdOtapVq5Yk2aYwZLzetVMU7r77blWtWjXTb2sDAwNtYdXVfV49JQLIa8WLF1eFChX08ssva/r06dq1a1eOfsO2bt061ahRQ3Xq1LEbgdGmTZsspyS2bNky2zcv0s3Pw61bt0pKvw5crXv37qxXAzjAO++8o8TERLvm4eGhjRs3KiUlRX379rU79729vdW8eXO7c//vv//W2LFjVbFiRXl4eMjDw0O+vr46f/58pmlnjnL//ffbPf7yyy916tQp9evXz67etLQ0tW3bVomJiZmmMV/t7rvv1n//+18NHTpUGzdu1NmzZ+32X7p0SZ999pnuu+8+FSxY0O412rdvr0uXLmn79u1Z9p2T565fv16VK1e2C6tuRUJCgi5evJjpXiY4OFgtW7bMdC9jGIY6depkt417GcC1ZYyGv3a6X3bs2rVLnTt3lr+/v9zd3VWgQAH17dtXqamp+vHHHx1W4+OPPy4p/cuIMsyaNUs1a9ZUs2bNrvu8CxcuaOvWrXrwwQdvuH7f5s2bFRkZmWlWQ//+/XXhwoVMo6Nudn+aHZ07d7aFVVL6KK8jR46oT58+trBKknx9fXX//fdr+/btWS57A1gBgdVVAgMD7R57eHjI399fJ0+elCR16dJFISEhtmGjcXFxOn/+vG06YHb5+/vbPc5YQ+jixYuSZHu9rL5pJygoyLb/ev1l9JnRH3C1EiVKqGDBgvrtt99u6+sYhqHPPvtMbdq0UWxsrOrVq6eSJUtq+PDhOnfu3E2ff+zYMe3evVsFChSwa4ULF5Zpmpmmxub0m6myex4GBATYHZdxXQBwa6pWraoGDRrYNSn93Jekhg0bZjr/V6xYYXfu9+rVS7NmzdLgwYO1ceNGff3110pMTFTJkiVv28/Aa681GfV27949U70vvfSSTNPUqVOnrtvfuHHjNG3aNG3fvl3t2rWTv7+/bXSRlH4tSklJ0euvv56p//bt20tSputhhpw8988//3ToN6rm9F6mYMGC8vb2ttvm5eWlS5cuOawmAI6R3XvJAwcOyMfHJ8f3TQcPHlTTpk11+PBhzZw5U9u2bVNiYqLt/2COvL4HBATooYce0ptvvqnU1FTt3r1b27Ztu+noqtOnTys1NfWm182TJ09e9zqYsf9qN7s/zY5rX+9m1+O0tDSdPn062/0DeYlhAlc5evSoypQpY3uckpKikydP2i4cbm5uGjZsmMaPH69XXnlFs2fPVmRkpMLCwhxaR8brJSUlZboIHjlyJMdDaoGrubu7KzIyUuvXr9cff/yR4/+gZPyH4tqFcLP6D1O5cuU0f/58SdKPP/6o9957zzY0eu7cuTd8nRIlSsjHx0cLFiy47v6r5ea3dzeScR4eO3Ysy+sCgNsj49x+//33Va5cuese99dff2ndunWaOHGinn76adv25OTkGwZE1/L29s5yYe8TJ05k+fP22mtNxjGvv/76dadnXBt8X83Dw0OjRo3SqFGjdObMGX366acaP3682rRpo0OHDqlYsWJyd3dXnz59rvsLsutNy8nJc0uWLJnpi11uxdX3MtfiXgZwbe7u7mrZsuUN7yX/+OMP7dy5U23btrVtu9H19mpr1qzR+fPntWrVKrufAxlf2uNoI0aM0OLFi/XBBx9ow4YNKlq0aKaZN9cqXry43N3db3rd9Pf3v+51UMp8P+sI1/6cutn12M3NLUezFIC8xAirq1y9loIkvffee0pJSbH7lqDBgwfL09NTvXv31v79+7NM3291dFPLli0lKdOi6YmJifrhhx8UGRmZ674BKf03+qZp6pFHHrEtCnm1K1eu6MMPP8zyuRnf7rJ792677WvXrr3ha1auXFnPPPOMatasqW+++ca2/XrnS8eOHfXLL7/I398/0yiMBg0aZPktM46UMQw8Y62cDO+//77d4tAAHKtNmzby8PDQL7/8kuW5nzESyzAMmaaZ6ZtO3377bdvitBlu9BvqkJCQTNezH3/80e5bmW6kSZMmKlq0qPbu3Xvdej09PbPVV9GiRdW9e3cNGzZMp06d0oEDB1SwYEG1aNFCu3btUq3/b+/Ow6qo/geOvy/76gKK4AZK7hvuC5osogJuuWQSKu6Ku1aKqQiWJuZWKWgqLqm54YqaC6KVuGXpk5ZWhmaKEmahgrHM7w+/d35eL8sFUbA+r+fhebhzz5w5M3fOOTNnzjnTsGGO8efWe6Eg6/r6+nLlyhXi4uJyTV9BnvS3bt0aS0tLvWuZGzduqENkhBAvr6lTp6IoCsHBwXplblZWFqNGjSIrK4vx48ery11cXLhz547aMxUez1/3xRdf6KyvbXB5snxXFEVn2F5B5Fd2NW3alDZt2jBv3jw2bNhAUFAQ1tbWecZpaWlJ+/bt2bp1a669XOHxfHxxcXF6bxtct24dVlZWhZpXsaD3mrVq1aJSpUps3LhR56VFDx48YPv27eqbA4UoiaSH1RNiYmIwMTHBx8eHixcvMmPGDBo1aqQzh02ZMmUYMGAAkZGRODs76823ANCgQQPi4+PZs2cPTk5O2NraFqgXVq1atRg+fDgff/wxRkZG+Pr6kpiYyIwZM6hSpQoTJ04skv0V/12tW7cmMjKS4OBgmjZtyqhRo6hXrx4ZGRl8++23rFixgvr16+d4fjs6OtKhQwfmzp1L2bJlcXZ25siRI8TExOiEu3DhAmPGjKFPnz7UqFEDMzMz4uLiuHDhgk5viAYNGvD555+zefNmqlevjoWFBQ0aNGDChAls376dV199lYkTJ9KwYUOys7O5fv06Bw8eZPLkybRs2fK5HaN69erRr18/FixYoD5JvHjxIgsWLKB06dI6cwAIIYqOi4sL4eHhvPvuu1y9epXOnTtTtmxZbt++zenTp7G2tiYsLIxSpUrx6quvMn/+fMqVK4eLiwvHjh1j1apVlClTRidO7YS7K1aswNbWFgsLC6pVq4a9vT39+/cnMDCQ4OBgevXqxbVr14iIiMhzTpIn2djY8PHHHzNw4EDu3r1L7969cXBwIDk5mfPnz5OcnExkZGSu63ft2pX69evTrFkzypcvz7Vr11i8eDHOzs7qZPFLliyhbdu2tGvXjlGjRuHi4kJqaio///wze/bsybORydB1J0yYwObNm+nevTtTp06lRYsWpKWlcezYMbp06YKnpye2trY4Ozuza9cuvL29sbOzU4/908qUKcOMGTOYNm0aAwYMoF+/fqSkpBAWFoaFhQWhoaEGHV8hRMnk7u7O4sWLGT9+PG3btmXMmDFUrVqV69evq29knTVrFj4+Puo6ffv2ZebMmbzxxhu8/fbbpKen89FHH+k1ePn4+GBmZka/fv145513SE9PJzIystDD1lxdXbG0tGTDhg3UqVMHGxsbKlasqPMGwPHjx9O3b180Gg3BwcEGxbtw4ULatm1Ly5YtmTp1Kq+88gq3b99m9+7dLF++HFtbW0JDQ9m7dy+enp7MnDkTOzs7NmzYQGxsLBEREToTrhuqQYMGxMTEEBkZSdOmTTEyMlIf5uTEyMiIiIgI3nzzTbp06cKIESN49OgR8+fP5969e3zwwQcFToMQL4o0WD0hJiaGWbNmERkZqU7+uXjxYr0no3379iUyMpJRo0bleNO6ZMkSRo8ezRtvvMHDhw/1Jok1RGRkJK6urqxatYqlS5dSunRpOnfuzNy5c2X+HFEkhg0bRosWLdTXhiclJWFqakrNmjUJCAjIc+z++vXrGTt2LFOmTCErK4uuXbuyadMmncrS0dERV1dXli1bxm+//YZGo6F69eosWLCAsWPHquHCwsK4desWw4YNIzU1FWdnZxITE7G2tubLL7/kgw8+YMWKFfz6669YWlpStWpVOnTo8Nx7WAFER0fj5OTEqlWrWLRoEW5ubmzZsoXOnTvr3RALIYpOSEgIdevWZcmSJWzatIlHjx7h6OhI8+bNGTlypBpu48aNjB8/nnfeeYfMzEzc3d05dOgQ/v7+OvFVq1aNxYsXs2TJEjw8PMjKyiI6OpqgoCACAgK4efMmUVFRREdHU79+fSIjIwkLCzM4vYGBgVStWpWIiAhGjBhBamoqDg4OuLm56U06/jRPT0+2b9/OypUr+fvvv3F0dMTHx4cZM2aok+bWrVuXc+fOMXv2bKZPn86dO3coU6YMNWrUUOeiyo2h62rfJDhr1ixWrFhBWFgYZcuWpXnz5gwfPlwNt2rVKt5++226devGo0ePGDhwIGvWrMlx2yEhITg4OPDRRx+xefNmLC0t8fDwYM6cOWpjnBDi5TV27FiaNWvGggULmDx5MsnJyWRnZ2NhYUFsbKxe+VStWjV27drFtGnT6N27N05OTkyaNInk5GSdMrd27dps376d6dOn07NnT+zt7QkICGDSpEk6b4g2lJWVFatXryYsLIyOHTuSkZFBaGiozttZe/Togbm5OZ6engaXT40aNeL06dOEhoYSEhJCamoqjo6OeHl5qfePtWrV4sSJE0ybNo3Ro0eTlpZGnTp11DqoMMaPH8/FixeZNm0af/31F4qi6PScyklAQADW1tbMnTuXvn37YmxsTKtWrTh69Cht2rQpVDqEeBE0Sn5nd77SgKOA4W//ej6MAE/AssBrzpo1i7CwMJKTkw0aRzx58mQiIyP57bffpPHoP+jv+4+I3vk9WVnPmHWekbGxhkE96lPKxjz/wKLInDhxAnd3dzZs2EBAQEBxJ0f8B2U+SuXONxtAyco/8POkMcah6ZuYmNsWbzqEEOIl82++lly3bh0DBw7knXfeYd68eUUW7/O2Z88eunXrlmNDmxCi+BRBDytLHjcU6c+D82KZUZjGqoI4efIkV65cYdmyZYwYMUIaq/6jStmYM6hHfdIeFe88RpbmJtJY9ZwdOnSIhIQEmjZtiqWlJefPn+eDDz6gRo0a9OzZs7iTJ/6jTMxtcWj6JtkZxfv2NCNTC2msEkKIQvg3X0sOGDCAW7duMXXqVKytrZk5c2aRxl/ULl26xLVr15g8eTJubm6F6sElhHh+iqCH1cvP0B5WGo0GKysr/Pz8iI6OxsbG5gWmUgjxop06dYrJkydz6dIlUlNTKVeuHJ06dWLu3Lk5vhpYCCGEEEK8PDw8PPj6669p0qQJa9eupXbt2sWdJCHEE6TBSgghhBBCCCGEEEKUKPKaKyGEEEIIIYQQQghRokiDlRBCCCGEEEIIIYQoUaTBSgghhBBCCCGEEEKUKNJgJYQQQgghhBBCCCFKFGmwEkIIIYQQQgghhBAlijRYCSGEEEIIIYQQQogSxaQoIklPyyAjI6sooio0U1NjLCxNizUN4r8j6+90stMyizUNRpYmGJeyKNY0CCGKR1paGv/880+xpsHMzAxLS8tiTYMQQrys5FpSCCHy98wNVulpGZw6cY3sbKUo0lNoRkYaWrZxLhGNVnPmzKFu3br06NFDZ/maNWsYNGgQZ86coVmzZsWTOAMEBQWxbds27t+//0K3q9FoGD16NJ988skL3W5BZf2dTvLKbyCreM95jDWUH9q00BcaFy5cYNGiRcTHx3Pr1i1MTEyoWbMmb7zxBkOHDsXOzq6IE1wyubi44OHhwZo1awC4efMmK1asoEePHri5uemEnTVrFmFhYShKMf/2//N02kui55nGkvZ7vChpaWkcPXqU7OzsYk2HkZERnp6ehW60unDhAkuWLCE+Pp6bN28CULlyZby8vBg2bFiJridLqmfJE/v27eP06dPMmjWr6BNWwjx8+JCIiAg8PDzw8PAo7uSoXFxcqF+/Pnv37i3upBQ5jUZDaGjoczm/goKCiI+PJzExscjjfl5e9mtJjUZjULijR48+Ux573vX8iRMnOHjwIBMmTKBMmTLPZRuF9azn9ebNmwkPD+fq1aukp6fz7bff6l3XCvEyeOYhgRkZWcXeWAWQna0Uey8vrTlz5rBz587iToZ4TrLTMov/AgMgSyn0k7lPP/2Upk2bcubMGd5++20OHDjAjh076NOnD1FRUQwZMqSIE1ty7dixgxkzZqifb968SVhYGN99951e2KFDh5KQkPACUyfy8l/9Pf75559ib6wCyM7OLnQvr+XLl9O0aVNOnTrF+PHj2bt3L7GxsUyYMIGLFy/SvHlzfvnllyJOscjLvn37CAsLK+5kvBAPHz4kLCyM+Pj44k6KKAIzZsxgx44dxZ2MAnnZryUTEhJ0/vz8/LC0tNRb3qRJk2dK3vOu50+cOEFYWBj37t17btsorGc5r5OTk+nfvz+urq4cOHCAhIQEatasWcQpFOLFKJIhgUKIl0dCQgKjRo3Cx8eHnTt3Ym5urn7n4+PD5MmTOXDgQDGm8MVq3LixwWErV65M5cqVn2NqREHI7/Fy+vrrrwkODsbf359t27ZhZmamfufl5cXo0aPZunXrSzfcMC0tDQsLC4N7HvxXpKWlvXS/5Yv28OFDrKysijsZLy1XV9fiTsJ/TqtWrXQ+ly9fHiMjI73lTyvouf6y1vNFkaef5by+cuUKGRkZBAYG0r59+2dKh5bUcaK4/OcnXZ81axYajYZvv/2Wnj17UqpUKUqXLk1gYCDJyclquCFDhmBnZ8fDhw/14vDy8qJevXrA4y6yDx48YO3atWg0GjQajV5X2NTUVEaNGkW5cuWwt7enZ8+e6nAIrezsbCIiIqhduzbm5uY4ODgwYMAAbty4oRPOw8OD+vXrc+bMGdq1a4eVlRXVq1fngw8+eOYn8BcvXsTb2xtra2vKly/PmDFj9PZ/6dKlvPrqqzg4OGBtbU2DBg2IiIggIyNDJ9y3335Lly5dcHBwwNzcnIoVK+Lv76+3PwDr16+nTp06WFlZ0ahRo39l1/jiNGfOHDQaDStWrNBprNIyMzOjW7du6ueCnosJCQm0adMGS0tLXFxciI6OBiA2NpYmTZpgZWVFgwYN9BrFDM2LBUmTIeedi4sLQUFBAMTHx9O8eXMABg0apOZh7RAGbRoLkxZD82p6ejqTJ0/Gzc2N0qVLY2dnR+vWrdm1a5feb2UoQ/Pp807j/fv3KVOmDCNGjND7LjExEWNjY+bPnw88vth76623qFatGhYWFtjZ2dGsWTM2bdqkrpPT7xEXF4eHhwf29vZYWlpStWpVevXqlWPZLYrHnDlzMDY2Zvny5TqNVU/q06cPFStW1Fl29uxZunXrhp2dHRYWFjRu3JgtW7bohFmzZg0ajYajR4/mW8/C4yETrVu3xtraGhsbGzp16sS3336b7z5ot3Pw4EEGDx5M+fLlsbKy4tGjRwWK99NPP6VmzZqYm5tTt25dNm7cSFBQEC4uLmqY+Ph4NBqNXm+gxMRENBpNvsNtN2/eTMeOHXFycsLS0pI6deowdepUHjx4oIYJCgpi6dKlAGq5p9Fo1KEo6enphISEUK1aNczMzKhUqRKjR4/W65Xg4uJCly5diImJoXHjxlhYWBAWFoa3tze1a9fWG9ajKAqvvPIK/v7+ee6DNt4DBw7QpEkTLC0tqV27NqtXr9YLm5SUxIgRI6hcuTJmZmZUq1aNsLAwMjMz1eNWvnx5AMLCwtR9DQoK4uLFi2g0GrZu3arG980336DRaNTrPK1u3brRtGlT9XNB64Ljx4/Tpk0brKysGDx4cK77vmzZMkxMTAgNDVWXRUZG0qhRI2xsbLC1taV27dpMmzYtz2Oo3d+WLVtiZ2dHqVKlaNKkCatWrdL7XQw93snJyQQHB1O3bl1sbGxwcHDAy8uLL7/8Ms90JCYmYmJiwty5c/W+O378uM5vkJyczPDhw6lSpQrm5uaUL18ed3d3Dh8+rK7zdJ4B2Lp1Ky1btqR06dJqXZbXcRZFL69z3ZByCXKu57XrG1LGnjp1iq5du2Jvb4+FhQWurq5MmDBBjfvtt98GoFq1ampZoC1rnzVPG3rvmJuczmuNRsOYMWPyvFcKCgqibdu2APTt21fvXrQgdemz1HFBQUHY2Njw888/4+fnh42NDVWqVGHy5MlqPFqPHj0iPDycOnXqYGFhgb29PZ6enpw4cUINoygKy5Ytw83NDUtLS8qWLUvv3r25evVqnsdRvPz+8w1WWq+99hqvvPIK27ZtY9asWezcuZNOnTqpN3Tjx4/nzz//ZOPGjTrrXbp0iaNHjzJ69Gjgce8VS0tL/Pz81O6wy5Yt01ln6NChmJqasnHjRiIiIoiPjycwMFAnzKhRo5gyZQo+Pj7s3r2b2bNnc+DAAdq0acMff/yhEzYpKYk333yTwMBAdu/eja+vLyEhIXz22WeFPh4ZGRn4+fnh7e3Nzp07GTNmDMuXL6dv37464X755RcCAgJYv349e/fuZciQIcyfP1/nhvTBgwf4+Phw+/Ztli5dyqFDh1i8RUsspwAANkNJREFUeDFVq1YlNTVVJ77Y2Fg++eQTwsPD2b59O3Z2drz22mtSGBWRrKws4uLiaNq0KVWqVDFonYKei4MGDWLo0KHs2rWLBg0aMHjwYMLDwwkJCeGdd95h+/bt2NjY0KNHjxxvIPPLi4amqSDnnVaTJk3UBrbp06ereXjo0KFFdnzyy6uPHj3i7t27vPXWW+zcuZNNmzbRtm1bevbsybp16/L5tXJmSD59EWm0sbFh8ODBbNiwgb/++kvnu2XLlmFmZqZezE6aNInIyEjGjRvHgQMHWL9+PX369CElJSXX+BMTE/H398fMzIzVq1dz4MABPvjgA6ytrYt9gnLxWFZWFkePHqVZs2Y4OTkZvN7Ro0dxd3fn3r17REVFsWvXLtzc3Ojbt2+ODTaG1LNz5syhX79+1K1bly1btrB+/XpSU1Np164dly5dMihdgwcPxtTUlPXr17Nt2zZMTU0NjnfFihUMHz6chg0bEhMTw/Tp05/LMLWffvoJPz8/Vq1axYEDB5gwYQJbtmyha9euapgZM2bQu3dvQHeoj5OTE4qi0KNHDz788EP69+9PbGwskyZNYu3atXh5eendeJw7d463335bzbu9evVi/PjxXL58mSNHjuiE3b9/P7/88ot6DZWX8+fPM3nyZCZOnMiuXbto2LAhQ4YM4fjx42qYpKQkWrRowRdffMHMmTPZv38/Q4YMYe7cuQwbNgwAJycn9YHJkCFD1H2dMWMG9erVw8nJSacx5PDhw1haWnLp0iW1zsrMzOTYsWN06NBBDVeQuuDWrVsEBgYSEBDAvn37CA4O1ttfRVF46623mDBhAitXrlSHa37++ecEBwfTvn17duzYwc6dO5k4caLejX5OEhMTGTFiBFu2bCEmJoaePXsyduxYZs+eXajjfffuXQBCQ0OJjY0lOjqa6tWr4+Hhked57OLiQrdu3YiKiiIrS3c6j08++YSKFSvy2muvAdC/f3927tzJzJkzOXjwICtXrqRDhw551gUJCQn07duX6tWr8/nnnxMbG8vMmTPVRkvx4uR2rhtSLuXG0DL2iy++oF27dly/fp2FCxeyf/9+pk+fzu3bt4HH9cTYsWMBiImJ0RvG+Kx52tB7x4LK715pxowZ6gOIOXPm6NyLFrQufZY6Dh7fT3br1g1vb2927drF4MGDWbRoEfPmzVPDZGZm4uvry+zZs+nSpQs7duxgzZo1tGnThuvXr6vhRowYwYQJE+jQoQM7d+5k2bJlXLx4kTZt2qi/qfiXUp7R33+lKXEHr5SIv7//Sitw+kNDQxVAmThxos7yDRs2KIDy2Wefqcvat2+vuLm56YQbNWqUUqpUKSU1NVVdZm1trQwcOFBvW9HR0QqgBAcH6yyPiIhQAOXWrVuKoijKDz/8kGO4U6dOKYAybdo0nTQByqlTp3TC1q1bV+nUqZMBR0DfwIEDFUBZsmSJzvL3339fAZSvvvoqx/WysrKUjIwMZd26dYqxsbFy9+5dRVEU5ezZswqg7Ny5M8/tAkqFChWUv//+W12WlJSkGBkZKXPnzi3UvjwP/ySlKrciviwRf/8kpeaf4CckJSUpgPLGG28YFL4w5+LZs2fVZSkpKYqxsbFiaWmp/P777+ry7777TgGUjz76SF1maF40NE2GnnfOzs46+fXMmTMKoERHR+uF1abxWY5PQfNqZmamkpGRoQwZMkRp3Lhxnmk3RG759EWl8ZdfflGMjIyURYsWqcvS0tIUe3t7ZdCgQeqy+vXrKz169MhzX57+PbZt26YAynfffZfnei+7e/fuKXv27CkRf/fu3StQ2vMqg7TnkfYvOztb/a527dpK48aNlYyMDJ11unTpojg5OSlZWVmKohhez16/fl0xMTFRxo4dqxMuNTVVcXR0VF5//fU890O7nQEDBugsNzTerKwsxdHRUWnZsqVOuGvXrimmpqaKs7Ozuuzo0aMKoBw9elQn7K+//qpXVj2dJ56WnZ2tZGRkKMeOHVMA5fz58+p3o0ePznHdAwcOKIASERGhs3zz5s0KoKxYsUJd5uzsrBgbGyuXL1/WCZuVlaVUr15d6d69u85yX19fxdXVVee3zomzs7NiYWGhXLt2TV2Wlpam2NnZKSNGjFCXjRgxQrGxsdEJpyiK8uGHHyqAcvHiRUVRFCU5OVkBlNDQUL1tBQYGKtWrV1c/d+jQQRk2bJhStmxZZe3atYqiKMrXX3+tAMrBgwcVRSlcXXDkyJEc99Pf3195+PCh0qtXL6V06dLK4cOHdcKMGTNGKVOmTJ7HyxDauiA8PFyxt7fX+Q0MPd5P0+Zhb29v5bXXXtP57unjrT2vd+zYoS77/fffFRMTEyUsLExdZmNjo0yYMCHPfRk4cKBOntH+3gUtn16kl/laMicDBw5UrK2tdZblda4/Ka9y6ekyrSBlt6urq+Lq6qqkpeV+fzh//nwFUH799Ved5UWVpw29d8zJ0+e1ohh+r6TNX1u3btVZv6B1aWHrOG36AWXLli06Yf38/JRatWqpn9etW6cAyqeffprrsUhISFAAZcGCBTrLf/vtN8XS0lJ55513cl1XvPykh9X/vPnmmzqfX3/9dUxMTDh69Ki6bPz48Xz33Xd8/fXXAPz999+sX7+egQMHYmNjY/C2nhxuBdCwYUMArl27BqBuUztMSatFixbUqVNH7wmlo6MjLVq00ItTG19hPX1MAgICdNIHj4dcdevWDXt7e4yNjTE1NWXAgAFkZWVx5coVAF555RXKli3LlClTiIqKyvPJtaenJ7a2turnChUq4ODg8Mz7IgqnoOeik5OTzhAJOzs7HBwccHNz0xneU6dOHYAcf9f88qKhaSrIeVdYzyuvbt26FXd3d2xsbDAxMcHU1JRVq1bxww8/FCqdhuTTF5XG6tWr06VLF5YtW6YOQ9m4cSMpKSmMGTNGDdeiRQv279/P1KlTiY+PJy0tLd/9dHNzw8zMjOHDh7N27VrpmfmSadq0KaampurfggULAPj555/58ccf1bIhMzNT/fPz8+PWrVtcvnxZJ6786tkvvviCzMxMBgwYoBOfhYUF7du3N7iXU69evXQ+Gxrv5cuXSUpK4vXXX9dZv2rVqri7uxu0bUNdvXqVgIAAHB0d1fyvndPEkDIlLi4O0C/n+vTpg7W1tV4517BhQ73JfY2MjBgzZgx79+5Vn5j/8ssvHDhwgODgYIPmRHFzc6Nq1arqZwsLC2rWrKlTNu3duxdPT08qVqyoc/x9fX0BOHbsWL7b8fb25urVq/z666+kp6fz1Vdf0blzZzw9PTl06BDwuNeVubm5OuymoHVB2bJl8fLyynH7KSkpeHl5cfr0ab766iu8vb314rx37x79+vVj165dej098hIXF0eHDh0oXbq0ei7MnDmTlJQU7ty5oxPWkOMNEBUVRZMmTbCwsFDrgiNHjuR7bnl4eNCoUSO1J4g2Lo1Gw/Dhw3X2d82aNbz33nucPHlSbyh7TrRD+19//XW2bNnC77//nu864vnI7VwvbLlkaBl75coVfvnlF4YMGYKFRcHfpl1Uebqo7h2fVNh7pcLUpYWt47Q0Go1er7mnryn379+PhYVFnkN29+7di0ajITAwUGe7jo6ONGrUSF6g8S8nDVb/4+joqPPZxMQEe3t7nS7H3bt3x8XFRa1c16xZw4MHDwrcpdPe3l7ns3YeIe0NmXabOQ2XqFixol436Kfj08ZpyA1ebrT7/yTtMdJu//r167Rr147ff/+dJUuW8OWXX3LmzBn1+Gi3X7p0aY4dO4abmxvTpk2jXr16VKxYkdDQUL0Lj+exL+L/lStXDisrK3799VeDwhf0XLSzs9MLZ2ZmprdcO29Nenq6Xvj88qKhaSrIeVdYzyOvxsTE8Prrr1OpUiU+++wzEhISOHPmDIMHD87xeOXH0Hz6ItM4fvx4fvrpJ/Xmb+nSpbRu3VrnbUIfffQRU6ZMYefOnXh6emJnZ0ePHj346aefco3X1dWVw4cP4+DgwOjRo3F1dcXV1ZUlS5bkmybxYpQrVw5LS8scL6w3btzImTNn2L17t85ybVf/t956S6dBy9TUVB1e8vRNe371rDbO5s2b68W5efNmgxsBns77hsarLRsqVKigF2dOywrr/v37tGvXjlOnTvHee+8RHx/PmTNniImJAfTzf05SUlIwMTFR533S0mg0ODo66pVzuQ31HDx4MJaWlkRFRQGP872lpaXB8woZUjbdvn2bPXv26B177Vwxhvyu2mF+hw8f5quvviIjIwMvLy86dOig3qQePnwYd3d3dTL5gtYFeQ2HvXLlCqdOncLX15f69evrfd+/f39Wr17NtWvX6NWrFw4ODrRs2VItT3Nz+vRpOnbsCDyeO+3rr7/mzJkzvPvuu0Dh6oKFCxcyatQoWrZsyfbt2zl58iRnzpyhc+fOBp1b48aN48iRI1y+fJmMjAw+/fRTevfurXMdsHnzZgYOHMjKlStp3bo1dnZ2DBgwgKSkpFzjffXVV9m5c6d6Y125cmXq16+vMweieDFyOtefpVwytIzVzn1a2AnbiypPF9W945MKe69UmLq0sHWclpWVlV6Dobm5uc61YnJyMhUrVsTIKPdmidu3b6MoChUqVNDb7smTJwvUcC9ePvKWwP9JSkqiUqVK6ufMzExSUlJ0CgUjIyNGjx7NtGnTWLBgAcuWLcPb25tatWoVaVq027x165ZeQXvz5k3KlStXpNvLSU77r7040C7buXMnDx48ICYmBmdnZzXcd999pxdfgwYN+Pzzz1EUhQsXLrBmzRrCw8OxtLRk6tSpz3dnhMrY2Bhvb2/279/PjRs38q3Ii+NczC8vFiRNz/u8ex7H57PPPqNatWps3rxZp9fB0/PEGKog+fRFpdHLy4v69evzySefYGNjw7lz5/Tm3LO2tiYsLIywsDBu376t9rbq2rUrP/74Y65xt2vXjnbt2pGVlcXZs2f5+OOPmTBhAhUqVOCNN94o3A6LImNsbIyXlxcHDx7k1q1bOhfDdevWBVAn+tbS5qOQkBB69uyZY7wFrYe1cW7btk0nXxTU0z2DDI1XW3bkNO/G0zfi2ov9p/OXIRfocXFx3Lx5k/j4eJ03RRXkFe729vZkZmaSnJys02ilKApJSUlqbxat3HpLlS5dWm14eOutt4iOjiYgIIAyZcoYnJb8lCtXjoYNG/L+++/n+P3TE/nnpHLlytSsWZPDhw/j4uJCs2bNKFOmDN7e3gQHB3Pq1ClOnjypzikFBa8L8upR1rp1a/r06cOQIUOAxxOsP30jN2jQIAYNGsSDBw84fvw4oaGhdOnShStXruR63n3++eeYmpqyd+9enRvInTt35ntMcvPZZ5/h4eFBZGSkzvLc5ol8WkBAAFOmTGHp0qW0atWKpKQkvRv5cuXKsXjxYhYvXsz169fZvXs3U6dO5c6dO3m+0bh79+50796dR48ecfLkSebOnUtAQAAuLi60bt264DsrCiWnc/1ZyiVDy1htWZXTy50MUVR5+kXdOxqiMHVpYeu4gihfvjxfffUV2dnZuTZalStXDo1Gw5dffpnjC6NyWib+PaSH1f9s2LBB5/OWLVvIzMzUe8Pf0KFDMTMz48033+Ty5cs6Q1i0nrVHkLZL6dM3cGfOnOGHH37Q6x7+vDx9TLSTBmqPibYQe7KQUBSFTz/9NNc4NRoNjRo1YtGiRZQpU4Zz584VcapFfkJCQlAUhWHDhuU4EXVGRgZ79uwBiudczC8vFiZNBTnvnu6JkZfncXw0Gg1mZmY6FwlJSUmFfktgYfLpi0jjuHHjiI2NJSQkhAoVKtCnT59cw1aoUIGgoCD69evH5cuXDXrjn7GxMS1btlSfakpZU3KEhISQlZXFyJEjDertWKtWLWrUqMH58+dp1qxZjn9PDo8wRKdOnTAxMeGXX37JNc7CMDTeWrVq4ejoqPdmpuvXr+u8FQlQ3xJ14cIFneVP90TLSU75H2D58uV6YXMr+7Tl2NPl3Pbt23nw4EGByrlx48bxxx9/0Lt3b+7du5fjNdSz6NKlC99//z2urq45Hnttg1V+5XyHDh2Ii4vj0KFD+Pj4AFCzZk2qVq3KzJkzycjI0JlwvajrgoEDB/L5558THR2tDt/OibW1Nb6+vrz77rv8888/XLx4Mdc4NRoNJiYmGBsbq8vS0tJYv359gdL2dJxPn1sXLlwgISHBoPUtLCzUIdwLFy7Ezc0tzyGxVatWZcyYMfj4+Bhcppubm9O+fXt1kmdD3gIqnq+ClEtPM7SMrVmzJq6urqxevTrPh2m5lQVFmacNuXd8EYqiLn0edaevry/p6el5vvG2S5cuKIrC77//nuM2GzRoUODtipeH9LD6n5iYGExMTPDx8eHixYvMmDGDRo0a6c0vUaZMGQYMGEBkZCTOzs45vs2iQYMGxMfHs2fPHpycnLC1tS1QS3qtWrUYPnw4H3/8MUZGRvj6+pKYmMiMGTOoUqUKEydOLNQ+enh4cOzYMb3XF+fEzMyMBQsWcP/+fZo3b86JEyd477338PX1Veds8PHxwczMjH79+vHOO++Qnp5OZGQkf/75p05ce/fuZdmyZfTo0YPq1aujKAoxMTHcu3dPvRAUL07r1q2JjIwkODiYpk2bMmrUKOrVq0dGRgbffvstK1asoH79+nTt2vW5nYt5yS8vGpqmwp53rq6uWFpasmHDBurUqYONjQ0VK1bM8cn88zg+2lfCBwcH07t3b3777Tdmz56Nk5NTnsPhcmNoPn3RaQwMDCQkJITjx48zffp0dZioVsuWLenSpQsNGzakbNmy/PDDD6xfv57WrVtjZWWVY5xRUVHExcXh7+9P1apVSU9PV1/D/uTNpShe7u7uLF26lLFjx9KkSROGDx9OvXr1MDIy4tatW2zfvh2AUqVKqessX74cX19fOnXqRFBQEJUqVeLu3bv88MMPnDt3jq1btxYoDS4uLoSHh/Puu+9y9epVOnfuTNmyZbl9+zanT59We/gVlKHxGhkZERYWxogRI+jduzeDBw/m3r17hIWF4eTkpPOU2dHRkQ4dOjB37lzKli2Ls7MzR44cUYfP5KVNmzaULVuWkSNHEhoaiqmpKRs2bOD8+fN6YbUX/PPmzcPX1xdjY2MaNmyIj48PnTp1YsqUKfz999+4u7tz4cIFQkNDady4Mf379zf4+NSsWZPOnTuzf/9+2rZtS6NGjQxe1xDh4eEcOnSINm3aMG7cOGrVqkV6ejqJiYns27ePqKgoKleujK2tLc7OzuzatQtvb2/s7OwoV66c2jjo7e3NsmXL+OOPP1i8eLEav7e3N9HR0ZQtW1ZnvsbnURf07t0bKysrevfuTVpaGps2bcLMzIxhw4ZhaWmJu7s7Tk5OJCUlMXfuXEqXLq3X2+1J/v7+LFy4kICAAIYPH05KSgoffvjhM/VM6NKlC7NnzyY0NJT27dtz+fJlwsPDqVatmsFv5AsODiYiIoJvvvmGlStX6nz3119/4enpSUBAALVr18bW1pYzZ85w4MCBXHuIAMycOZMbN27g7e1N5cqVuXfvHkuWLNGZJ0kUn4KUS08rSNm9dOlSunbtSqtWrZg4cSJVq1bl+vXrfPHFF+rDUW25t2TJEgYOHIipqSm1atUq0jxtyL3ji/KsdenzqDv79etHdHQ0I0eO5PLly3h6epKdnc2pU6eoU6cOb7zxBu7u7gwfPpxBgwZx9uxZXn31Vaytrbl16xZfffUVDRo0YNSoUc9yaEQJJg1W/xMTE8OsWbOIjIxUJ4hbvHix3k0UQN++fYmMjGTUqFE5dl1csmQJo0eP5o033uDhw4cFmsBVKzIyEldXV1atWsXSpUspXbo0nTt3Zu7cuTmOXTbE/fv39eYHyo222/i4ceN47733sLS0ZNiwYcyfP18NU7t2bbZv38706dPp2bMn9vb2BAQEMGnSJHWCU4AaNWpQpkwZIiIiuHnzJmZmZtSqVYs1a9YwcODAQu2LeDbDhg2jRYsW6qtlk5KSMDU1pWbNmgQEBOg8/Xke52JeDMmLhqSpsOedlZUVq1evJiwsjI4dO5KRkUFoaCizZs3KMXxRH59BgwZx584doqKiWL16NdWrV2fq1KncuHGjUDfQhubTF51GS0tLunbtymeffcbIkSP1vvfy8mL37t0sWrSIhw8fUqlSJQYMGKDOt5ITNzc3Dh48SGhoKElJSdjY2FC/fn12796tzt0iSoaRI0fSunVrlixZwqJFi7h58yYajYbKlSvTpk0bjhw5ojOBraenJ6dPn+b9999nwoQJ/Pnnn9jb21O3bl29B0uGCgkJoW7duixZsoRNmzbx6NEjHB0dad68eY7nZFHHO3z4cDQaDREREbz22mu4uLgwdepUdu3apfMqb4D169czduxYpkyZQlZWFl27dmXTpk35Ps22t7cnNjaWyZMnExgYiLW1Nd27d2fz5s06c8bB4+FZX3/9NcuWLSM8PBxFUfj1119xcXFh586dzJo1i+joaN5//33KlStH//79mTNnToEbPPr27cv+/fufSy8DJycnzp49y+zZs5k/fz43btzA1taWatWqqTdWWqtWreLtt9+mW7duPHr0iIEDB6pP+L28vDAyMsLS0lJn+FiHDh2Ijo7G09NT7/rvedSVfn5+7Nu3j65du9K9e3diYmJo164da9asYcuWLfz555+UK1eOtm3bsm7dOr15xp7k5eXF6tWrmTdvHl27dqVSpUoMGzYMBwcHdfhhQb377rs8fPiQVatWERERQd26dYmKimLHjh0GX/dWqlSJtm3bcuHCBfXlPloWFha0bNmS9evXk5iYSEZGBlWrVmXKlCm88847ucbZsmVLzp49y5QpU0hOTqZMmTI0a9aMuLg4dT4zUXwKUi7lxNAytlOnThw/fpzw8HDGjRtHeno6lStX1nkph4eHByEhIaxdu5ZPP/2U7Oxsjh49qg51Lao8nd+944tSFHVpUdedJiYm7Nu3j7lz57Jp0yYWL16Mra0tjRo1onPnzmq45cuX06pVK5YvX86yZcvIzs6mYsWKuLu7670sSPy7aBRDutvkIT0tg1MnrpGd/UzRPDMjIw0t2zhjYWlaoPVmzZpFWFgYycnJBs83M3nyZCIjI/ntt9+eyw3785CamoqdnR2LFy9+pon+BGT9nU7yym8gq3jPeYw1lB/aFONSBX/7SUlUmLwoXl7//PMPLi4utG3bVm9YlMhbWloaR48eJTs7u1jTYWRkhKenpzrxtHh29+7do2bNmvTo0YMVK1YUd3Kei169enHy5EkSExMxNS3YNZv497lz5w7Ozs6MHTuWiIiI4k7OCyPXkoaZOHEi69evf+kn1X4Z7x2FKCmeuYeVhaUpLds4k5GR8/j6F8XU1LjAjVUFdfLkSa5cucKyZcsYMWLES1XgHD9+XH2aJp6NcSkLyg9tSnaaYd3dnxcjS5MSe4EhRG6Sk5O5fPky0dHR3L59W166UAiWlpZ4enrmOAfdi2RmZiaNVc8gKSmJ999/H09PT+zt7bl27RqLFi0iNTWV8ePHF3fyitSjR484d+4cp0+fZseOHSxcuFAaq/7jbty4wdWrV5k/fz5GRkb/unM+P3Itmbc7d+6QkJBATEzMSz1J/st87yhESVEkQwItLE2fe2NRSaCdO6VLly689957xZ2cAvH398ff37+4k/GvYVzKAuNS+YcTQuiKjY1l0KBBODk5sWzZMoO6/wt9lpaW0lj0kjM3NycxMZHg4GDu3r2LlZUVrVq1Iioq6l83bOnWrVu0adOGUqVKMWLECMaOHVvcSRLFbOXKlYSHh+Pi4sKGDRt03g78XyHXkrnbt28fY8aMoVWrVixZsqS4k1NoL/O9oxAlxTMPCRRCCCGEEEIIIYQQoigV36xvQgghhBBCCCGEEELkQBqshBBCCCGEEEIIIUSJIg1WQgghhBBCCCGEEKJEkQYrIYQQQgghhBBCCFGiSIOVEEIIIYQQQgghhChRpMFKCCGEEEIIIYQQQpQo0mAlhBBCCCGEEEIIIUoUk6KI5G76A+5nPiqKqArNxsQcOwvrZ4rjo48+Yvz48dSrV4/vv/++iFJWNObMmUPdunXp0aNHodZ/+PAhEREReHh44OHhofPdmjVrGDRoEL/++isuLi7PnNb/ggcZyTzKSi3WNJgb22JtWr7A62l/bzUec3PKlClDnTp16NixI0OGDMHBwUFnnVmzZhEWFoaiKM+c7uKm0WgIDQ1l1qxZxZ2UHPNeUFAQ8fHxJCYmquGeNf+/aM/7fLl06RJbtmwhKChIr8zK6fj9G73s9a723D9z5gzNmjUzeL286rKSat++fZw+fTrHMsfFxQUPDw/WrFnzwtMlhBBCCFHSPXOD1d30B8w4u4dMJbso0lNoJhojZjfr+kyNVqtXrwbg4sWLnDp1ipYtWxZV8p7ZnDlz6N279zM1WIWFhQHoXeT7+/uTkJCAk5PTM6byv+FBRjJ7fx1PtpJRrOkw0pjSpdqSQjVaAURHR1O7dm0yMjK4c+cOX331FfPmzePDDz9k8+bNdOjQQQ07dOhQOnfuXFRJL1YJCQlUrly5uJORqxkzZjB+/HidZc+a//9tLl26RFhYGB4eHnoNVjkdv3+bf1O9W1B51WUl1b59+1i6dGmODVY7duygVKlSLz5RQgghhBAvgWdusLqf+ajYL5oBMpVs7mc+wo7CXTifPXuW8+fP4+/vT2xsLKtWrSpRDVbPU/ny5SlfvnCNHv9Fj7JSi72xCiBbyeBRVmqhG6zq16+v07OhV69eTJw4kbZt29KzZ09++uknKlSoAEDlypVLdCNPQbRq1aq4k5AnV1fXYtt2RkYGGo0GE5Mi6XxbLIrz+L0o/5Z692X18OFDrKysiiSuxo0bF0k8QgghhBD/RjKH1f+sWrUKgA8++IA2bdrw+eef8/DhQ/X7xMRENBoNH374IQsXLqRatWrY2NjQunVrTp48qRNXUFAQNjY2/Pzzz/j5+WFjY0OVKlWYPHkyjx7pDuG4e/cuwcHBVKpUCTMzM6pXr867776rE06j0fDgwQPWrl2LRqNBo9GoT5aTk5MJDg6mbt262NjY4ODggJeXF19++aVO2rUNUmFhYWocQUFBwOOhGRqNRm8IzerVq2nUqBEWFhbY2dnx2muv8cMPPxR6X0XJV7VqVRYsWEBqairLly9Xl8+aNQuNRqMT1sXFhS5durB3714aN26MpaUlderUYe/evcDj86pOnTpYW1vTokULzp49q7e9s2fP0q1bN+zs7LCwsKBx48Zs2bJFJ4z2/Dx69CijRo2iXLly2Nvb07NnT27evKkTNi4uDg8PD+zt7bG0tKRq1ar06tVLJy9rNBq9ng7ff/893bt3p2zZslhYWODm5sbatWt1wsTHx6PRaNi0aRPvvvsuFStWpFSpUnTo0IHLly/rhD106BDdu3encuXKWFhY8MorrzBixAj++OOPfH4B9Ia55Zb/ExMTMTExYe7cuXpxHD9+HI1Gw9atW3PdjnZ/1q9fz+TJk6lUqRLm5ub8/PPPABw+fBhvb29KlSqFlZUV7u7uHDlyRC+e2NhY3NzcMDc3p1q1anz44Yd6YbTlZ07DnnL6PX788Uf69etHhQoVMDc3p2rVqgwYMIBHjx6xZs0a+vTpA4Cnp6d6TLRx5zRMMD09nZCQEKpVq4aZmRmVKlVi9OjR3Lt3Tyec9pw+cOAATZo0wdLSktq1a6u9b8XzY0hdkl9dBvDTTz8REBCAg4MD5ubm1KlTh6VLl+pt7+LFi3Ts2BErKyvKly/P6NGjiY2NRaPREB8fr4bz8PCgfv36HD9+nDZt2mBlZcXgwYMB2Lx5Mx07dsTJyUkt/6ZOncqDBw909ku7fW16n6xvXVxcdNIPcP36dQIDA3X2YcGCBWRn/38jZUGuSYQQQgghXlbSYAWkpaWxadMmmjdvTv369Rk8eDCpqak53uwtXbqUQ4cOsXjxYjZs2MCDBw/w8/Pjr7/+0gmXkZFBt27d8Pb2ZteuXQwePJhFixYxb948NUx6ejqenp6sW7eOSZMmERsbS2BgIBEREfTs2VMNl5CQgKWlJX5+fiQkJJCQkMCyZcuAxw1eAKGhocTGxhIdHU316tXx8PBQL7qdnJw4cOAAAEOGDFHjmDFjRq7HZO7cuQwZMoR69eoRExPDkiVLuHDhAq1bt+ann34q8L6Kl4efnx/GxsYcP34837Dnz58nJCSEKVOmEBMTQ+nSpenZsyehoaGsXLmSOXPmsGHDBv766y+6dOlCWlqauu7Ro0dxd3fn3r17REVFsWvXLtzc3Ojbt2+ODRtDhw7F1NSUjRs3EhERQXx8PIGBger3iYmJ+Pv7Y2ZmxurVqzlw4AAffPAB1tbW/PPPP7nuw+XLl2nTpg0XL17ko48+IiYmhrp16xIUFERERIRe+GnTpnHt2jVWrlzJihUr+Omnn+jatStZWVlqmF9++YXWrVsTGRnJwYMHmTlzJqdOnaJt27ZkZBSsd15u+d/FxYVu3boRFRWls22ATz75hIoVK/Laa6/lG39ISAjXr18nKiqKPXv24ODgwGeffUbHjh0pVaoUa9euZcuWLdjZ2dGpUyedRqsjR47QvXt3bG1t+fzzz5k/fz5btmwhOjq6QPv4pPPnz9O8eXNOnjxJeHg4+/fvZ+7cuTx69Ih//vkHf39/5syZAzwuj7XHxN/fP8f4FEWhR48efPjhh/Tv35/Y2FgmTZrE2rVr8fLy0mtYP3/+PJMnT2bixIns2rWLhg0bMmTIEIPyg3g2+dUl+dVlly5donnz5nz//fcsWLCAvXv34u/vz7hx49RhhAC3bt2iffv2XL58mcjISNatW0dqaipjxozJMV23bt0iMDCQgIAA9u3bR3BwMPC4cczPz49Vq1Zx4MABJkyYwJYtW+jatau67owZM+jduzeAmt68huAnJyfTpk0bDh48yOzZs9m9ezcdOnTgrbfeyjF9hl6TCCGEEEK8jF7ecR9FaNu2bfz1118MGTIEgL59+zJhwgRWrVrFwIEDdcLa2tqyd+9ejI2NAahYsSItWrRg//79vPHGG2q4f/75h7CwMLUngLe3N2fPnmXjxo3MnDkTgLVr13LhwgW2bNmihvPx8cHGxoYpU6Zw6NAhfHx8aNWqFUZGRpQvX15vOFOtWrXUxiuArKwsOnXqRGJiIh999BEeHh6Ym5vTtGlT4PHQrvyGRN27d4/Zs2fj5+fHxo0b1eUeHh7UqFGDWbNmsWHDhgLtq3h5WFtbU65cOb3eSzlJSUnh5MmTVKpUCXicH9zc3Pj000/5+eef1WEzGo2GHj16cPjwYfVmLjg4mHr16hEXF6cOQevUqRN//PEH06ZNY8CAARgZ/X+beufOnfnoo4/Uz3fv3uWdd94hKSkJR0dHvvnmG9LT05k/fz6NGjVSwwUEBOS5D7NmzeKff/7h6NGjVKlSBXjcaHfv3j3CwsIYMWIEpUuXVsPXrVuXzz77TP1sbGzM66+/zpkzZ9S8NXLkSPV7RVFo06YNHh4eODs7s3//frp165bvsdXKK/+PGzcOT09P9uzZo85vdfPmTXbs2MGMGTMMGtrn6uqq0zj/8OFDxo8fT5cuXdixY4e63M/PjyZNmjBt2jROnToFwLvvvkuFChU4dOgQFhYWwOPf8Fle3jBp0iRMTEw4ffq0zlDlN998E3hcBteoUQN4/FvkV54dPHiQL774goiICN5++23gcTlbpUoV+vbty7p16xg2bJga/o8//uDrr7+matWqALz66qscOXKEjRs38uqrrxZ6v0T+8qtL8qvLJk2ahK2tLV999ZU6L5SPjw+PHj3igw8+YNy4cZQtW5ZFixZx9+5djh8/Tt26dQHw9fWlc+fOOU7Wf/fuXbZu3YqXl5fO8unTp6v/K4qCu7s7derUoX379ly4cIGGDRvi6uqqDq02ZDjywoUL+f333zl16hQtWrQAHueprKwsoqKimDBhAjVr1lTDG3pNIoQQQgjxMpIeVjweDmhpaale3NnY2NCnTx++/PJLvd5E/v7+6oUhQMOGDQG4du2aTjiNRqPzlFUb9slwcXFxWFtbq09ftbTDA3IafpOTqKgomjRpgoWFBSYmJpiamnLkyBG94XuGSkhIIC0tTW+YQpUqVfDy8tJLlyH7Kl4uhr7dzc3NTW2sAqhTpw7wuHHzyTletMu158TPP//Mjz/+qDZCZGZmqn9+fn7cunVLb5jd0408T+c9Nzc3zMzMGD58OGvXruXq1asG7UNcXBze3t5qY5VWUFAQDx8+JCEhoUDpALhz5w4jR46kSpUqap50dnYGKHS+zImHhweNGjXSGfIUFRWFRqNh+PDhBsXRq1cvnc8nTpzg7t27DBw4UOd3yc7OpnPnzpw5c4YHDx7w4MEDzpw5Q8+ePdXGKnh8A/10eWCohw8fcuzYMV5//fUim1cvLi4OQK8869OnD9bW1nrlmZubm9pYBWBhYUHNmjWlPHsBnqUuSU9P58iRI7z22mtYWVnplSnp6enqULljx45Rv359tbFKq1+/fjnGXbZsWb3GKoCrV68SEBCAo6MjxsbGmJqa0r59e6Dw+TwuLo66deuqjVVaQUFBKIqins9ahl6TCCGEEEK8jP7zDVY///wzx48fx9/fH0VRuHfvHvfu3VMbkZ6eu8Te3l7ns7m5OYDOUCcAKysrnZs4bdj09HT1c0pKCo6OjnpzAzk4OGBiYkJKSkq+6V+4cCGjRo2iZcuWbN++nZMnT3LmzBk6d+6slyZDabeb05CFihUr6qXLkH0VL48HDx6QkpJCxYoV8w1rZ2en89nMzCzP5dpz4vbt2wC89dZbmJqa6vxph9s8Pd9TfnnP1dWVw4cP4+DgwOjRo3F1dcXV1ZUlS5bkuQ8pKSm5nuva7wuSjuzsbDp27EhMTAzvvPMOR44c4fTp0+rNcmHzZW7GjRvHkSNHuHz5MhkZGXz66af07t0bR0dHg9Z/et+1v03v3r31fpt58+ahKAp3797lzz//JDs7O8ftGLrtp/35559kZWUV6QT/KSkpmJiY6DWAaTQaHB0d8/194fFvXNS/m9D3LHVJSkoKmZmZfPzxx3rnrZ+fH/D/ZUpKSora6+lJOS2DnOvC+/fv065dO06dOsV7771HfHw8Z86cISYmBih8Pi/q8kgIIYQQ4mX2nx8SuHr1ahRFYdu2bWzbtk3v+7Vr1/Lee+89l23b29tz6tQpFEXRabS6c+cOmZmZlCtXLt84PvvsMzw8PIiMjNRZnpqa+kzpgsfzdjzt5s2bBqVLvLxiY2PJysp6rq+M155DISEhOvO1PalWrVoFjrddu3a0a9eOrKwszp49y8cff8yECROoUKFCrsNj7O3tcz3Xn0yrob7//nvOnz/PmjVrdIYUayczL2oBAQFMmTKFpUuX0qpVK5KSkhg9erTB6z/dYK7d348//jjXIUwVKlRQ3yiYlJSk9/3Ty7SNEE/PF/X0zbednR3GxsbcuHHD4PTnx97enszMTJKTk3UarRRFISkpiebNmxfZtkTxKVu2LMbGxvTv3z/X879atWrA43NC2zD7pJzOZdDPI/C4J9TNmzeJj49Xe1UBehP5F1RRl0dCCCGEEC+z/3QPq6ysLNauXYurqytHjx7V+5s8eTK3bt1i//79z2X73t7e3L9/n507d+osX7dunfq9Vm5P+DUajfpEVevChQt6w5gK8tS1devWWFpa6szTA3Djxg11+JT4d7p+/TpvvfUWpUuXZsSIEc9tO7Vq1aJGjRqcP3+eZs2a5fhna2tb6PiNjY1p2bKlOlTu3LlzuYb19vZWbz6ftG7dOqysrAyad+ZJ2pvbp/Plk29dLKi8evhYWFiowyAXLlyIm5sb7u7uhd6Wu7s7ZcqU4dKlS7n+NmZmZurbH2NiYnR6wKSmprJnzx6dOCtUqICFhQUXLlzQWb5r1y6dz5aWlrRv356tW7fm+UbFgpRn2vLq6fJs+/btPHjwQMqzl0xevZo9PT359ttvadiwYY7nrfZhTPv27fn++++5dOmSThyff/65wekoSD4v6Pl66dIlvTJr3bp1aDQaPD09DU6jEEIIIcTL7j/dw2r//v3cvHmTefPm5dibpH79+nzyySesWrWKRYsWFfn2BwwYwNKlSxk4cCCJiYk0aNCAr776ijlz5uDn50eHDh3UsA0aNCA+Pp49e/bg5OSEra0ttWrVokuXLsyePZvQ0FD1rUfh4eFUq1aNzMxMdX1bW1ucnZ3ZtWsX3t7e2NnZUa5cuRwnRy5TpgwzZsxQJ77u168fKSkphIWFYWFhQWhoaJEfC/Hiff/99+ocL3fu3OHLL78kOjoaY2NjduzYUWRzCOVm+fLl+Pr60qlTJ4KCgqhUqRJ3797lhx9+4Ny5czm+pTMvUVFRxMXF4e/vT9WqVUlPT1eH9D6Zl54WGhrK3r178fT0ZObMmdjZ2bFhwwZiY2OJiIjQmXDdELVr18bV1ZWpU6eiKAp2dnbs2bOHQ4cOFSieJ+WW/7WCg4OJiIjgm2++YeXKlYXeDjyew+/jjz9m4MCB3L17l969e+Pg4EBycjLnz58nOTlZ7dE5e/ZsOnfujI+PD5MnTyYrK4t58+ZhbW2tvsEUHt/cBwYGsnr1alxdXWnUqBGnT5/WeamD1sKFC2nbti0tW7Zk6tSpvPLKK9y+fZvdu3ezfPlybG1tqV+/PgArVqzA1tYWCwsLqlWrluNwPh8fHzp16sSUKVP4+++/cXd358KFC4SGhtK4cWP69+//TMdLvFh51WVLliyhbdu2tGvXjlGjRuHi4kJqaio///wze/bsUed/mjBhAqtXr8bX15fw8HAqVKjAxo0b+fHHHwF0XvaQmzZt2lC2bFlGjhxJaGgopqambNiwgfPnz+uFbdCgAQDz5s3D19cXY2NjGjZsqA6VftLEiRNZt24d/v7+hIeH4+zsTGxsLMuWLWPUqFE6E64LIYQQQvzb/ad7WK1atQozMzMGDRqU4/flypXjtddeY+/evTkOH3hWFhYWHD16lDfffJP58+fj6+vLmjVreOutt9R5MLSWLFlCjRo1eOONN2jevLna++Xdd99l8uTJrFq1Cn9/f1auXElUVBRt27bNcX+trKzo1q0bzZs3Z9asWbmmLSQkhJUrV3L+/Hl69OjBmDFjqFevHidOnFDf0CVeboMGDaJ169Z4e3szatQovv32W6ZMmcKPP/74Qp7ie3p6cvr0acqUKcOECRPo0KEDo0aN4vDhw3k2MOXGzc2NzMxMQkND8fX1pX///iQnJ7N79246duyY63q1atXixIkT1KpVi9GjR9OjRw++//57oqOj1bfKFYSpqSl79uyhZs2ajBgxgn79+nHnzh0OHz5c4Li0csv/WpUqVaJt27bY2dnl+1ZEQwQGBnL06FHu37/PiBEj6NChA+PHj+fcuXM6PZJ8fHzYuXMnf//9N3379mXSpEn06tWLwYMH68W5YMECAgMDiYiIoHv37iQkJLB37169cNrGrKZNmxISEkLnzp2ZMmUK5ubm6g1+tWrVWLx4MefPn8fDw4PmzZvr9erS0mg07Ny5k0mTJhEdHY2fnx8ffvgh/fv3Jy4uTq+HjCj5cqvL6taty7lz56hfvz7Tp0+nY8eODBkyhG3btumctxUrVuTYsWPUrFmTkSNH8uabb2JmZkZ4eDjw+KFNfuzt7YmNjcXKyorAwEAGDx6MjY0Nmzdv1gsbEBDA0KFDWbZsGa1bt6Z58+a5voW1fPnynDhxAi8vL0JCQujSpYv6lsuPP/644AdLCCGEEOIlplEMfR1YLu6mP2DG2T1kKtlFlaZCMdEYMbtZV+wsrIs1HeLf70FGMnt/HU+2klGs6TDSmNKl2hKsTZ9vTyhR8t25cwdnZ2fGjh1LREREcSdHPGdS7z4fw4cPZ9OmTaSkpOTY+0kIIYQQQrxYz9xgBY8vnu9nPso/4HNkY2L+r7loFiXfg4xkHmUVfmL7omBubCuNVf9xN27c4OrVq8yfP5+4uDiuXLlCpUqVijtZ4gWQevfZhIeHU7FiRapXr879+/fZu3cvK1euZPr06WpPKyGEEEIIUbyKZA4rOwtr7Hg5L1qFKAxr0/LSWCSK3cqVKwkPD8fFxYUNGzZIY9V/iNS7z8bU1JT58+dz48YNMjMzqVGjBgsXLmT8+PHFnTQhhBBCCPE/RdLDSgghhBBCCCGEEEKIovKfnnRdCCGEEEIIIYQQQpQ80mAlhBBCCCGEEEIIIUoUabASQgghhBBCCCGEECWKNFgJIYQQQgghhBBCiBJFGqyEEEIIIYQQQgghRIkiDVZCCCGEEEIIIYQQokSRBishhBBCCCGEEEIIUaJIg5UQQgghhBBCCCGEKFGkwUoIIYQQQgghhBBClCjSYCWEEEIIIYQQQgghShRpsBJCCCGEEEIIIYQQJYo0WAkhhBBCCCGEEEKIEkUarIQQQgghhBBCCCFEiSINVkIIIYQQQgghhBCiRJEGKyGEEEIIIYQQQghRokiDlRBCCCGEEEIIIYQoUf4PZoofVMmJwRUAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1500x1400 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#vecs_Code = await acal_EmbVec(RefcodeDB, \"CodeBlock\")\n",
    "plot_cluster_umap(vecs_Code, \"Task\", \"UMAPof Embedding of CodeBlock, Group by Task\")\n",
    "plot_cluster_umap(vecs_Code, \"Language\", \"UMAPof Embedding of CodeBlock, Group by Language\")\n",
    "embedding_heatmap(vecs_Code, RefcodeDB, \"Pairwise Embedding Similarity Heatmap of CodeBlock\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "47b9e680",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'#----------------\\r\\r\\n# Import necessary modules\\r\\r\\n#----------------\\r\\r\\nimport scanpy as sc\\r\\r\\nimport anndata as ad\\r\\r\\n\\r\\r\\n# Set figure parameters for decent plotting\\r\\r\\nsc.settings.set_figure_params(dpi=80, facecolor=\\'white\\')\\r\\r\\n\\r\\r\\n#----------------\\r\\r\\n# Data Input: load a sample dataset\\r\\r\\n#----------------\\r\\r\\nadata = sc.datasets.pbmc3k()\\r\\r\\n# Data format: \\r\\r\\n# adata is an AnnData object containing the following core structure \\r\\r\\n# x: core data matrix (cells × genes), usually a sparse matrix or numpy array\\r\\r\\n# obs: cell-level metadata (DataFrame), stores properties for each cell\\r\\r\\n# var: gene-level metadata (DataFrame), stores attributes for each gene\\r\\r\\n\\r\\r\\n\\r\\r\\n#----------------\\r\\r\\n# Annotate genes for QC metrics and calculate metrics\\r\\r\\n#----------------\\r\\r\\n\\r\\r\\n# Mark mitochondrial genes\\r\\r\\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\\r\\r\\n\\r\\r\\n# Mark ribosomal genes\\r\\r\\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\\r\\r\\n\\r\\r\\n# Mark hemoglobin genes\\r\\r\\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\\r\\r\\n\\r\\r\\n# Calculate QC metrics\\r\\r\\nsc.pp.calculate_qc_metrics(\\r\\r\\n    adata, \\r\\r\\n    qc_vars=[\"mt\", \"ribo\", \"hb\"],\\r\\r\\n    inplace=True,\\r\\r\\n    log1p=True\\r\\r\\n)\\r\\r\\n\\r\\r\\n#----------------\\r\\r\\n# Create QC visualization plots\\r\\r\\n#----------------\\r\\r\\n# Violin plots for key metrics\\r\\r\\nsc.pl.violin(\\r\\r\\n    adata,\\r\\r\\n    keys=[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\\r\\r\\n    jitter=0.4,\\r\\r\\n    multi_panel=True\\r\\r\\n)\\r\\r\\n\\r\\r\\n# Scatter plot showing counts vs genes with mitochondrial percentage\\r\\r\\nsc.pl.scatter(\\r\\r\\n    adata, \\r\\r\\n    x=\"total_counts\",\\r\\r\\n    y=\"n_genes_by_counts\",\\r\\r\\n    color=\"pct_counts_mt\"\\r\\r\\n)\\r\\r\\n\\r\\r\\n#----------------\\r\\r\\n# Data output \\r\\r\\n#----------------\\r\\r\\nadata.write(\\'pbmc3k_qc.h5ad\\')'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RefcodeDB['CodeBlock'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "cd3282e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#vecs = []\n",
    "\n",
    "for i in range(11,len(RefcodeDB['CodeBlock'])): \n",
    "    text = RefcodeDB['CodeBlock'][i]\n",
    "    vec = embedding.embed_query(text)\n",
    "    vecs.append(vec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1347b67d",
   "metadata": {},
   "outputs": [],
   "source": [
    "vecs_Code = vecs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "4548bab8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n"
     ]
    }
   ],
   "source": [
    "for i in range(4,len(RefcodeDB['CodeBlock'])): \n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "55f36102",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1024"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(vecs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ff76dbc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1024"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(vec)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8908ecd7",
   "metadata": {},
   "source": [
    "## Build RefCodeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "9b9b7a98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert table to documents \n",
    "Docs = []\n",
    "IDs = []\n",
    "for idx, row in RefcodeDB.iterrows():\n",
    "    doc = Document(\n",
    "        page_content=row['Description'], # Set description as content, will be embedded \n",
    "        metadata = {\n",
    "            \"CodeBlock\": row[\"CodeBlock\"], # Reference code block will be stored in metadata\n",
    "            \"Omics\": row[\"Omics\"],\n",
    "            \"Task\": row[\"Task\"],\n",
    "            \"Tools\": row[\"Tools\"],\n",
    "            \"Source\":  row[\"Source\"],\n",
    "        }\n",
    "        ) \n",
    "    # Append \n",
    "    Docs.append(doc)\n",
    "    IDs.append(idx)\n",
    "    \n",
    "# Add to vector store in batch = 10\n",
    "for star in range(0,len(Docs),10):\n",
    "    end = min(star+10, len(Docs))\n",
    "    vector_store.add_documents(documents=Docs[star:end], ids=IDs[star:end])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "f92ef6c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rank #1 queried doc in task Quality control [0] of similarity score 0.9919554600768803\n",
      "Rank #2 queried doc in task Quality control [1] of similarity score 0.869372800663382\n",
      "Rank #3 queried doc in task Quality control [32] of similarity score 0.8335572679570653\n",
      "Rank #4 queried doc in task Quality control [4] of similarity score 0.8310848675865152\n",
      "Rank #5 queried doc in task Quality control [3] of similarity score 0.8168269614187583\n",
      "Rank #6 queried doc in task Quality control [2] of similarity score 0.8123400658750377\n",
      "Rank #7 queried doc in task Feature selection [33] of similarity score 0.7969401124787253\n",
      "Rank #8 queried doc in task Normalization [5] of similarity score 0.7067833296318904\n",
      "Rank #9 queried doc in task Feature selection [8] of similarity score 0.6891115863719738\n",
      "Rank #10 queried doc in task Normalization [6] of similarity score 0.676247794851803\n",
      "Rank #11 queried doc in task Integration [23] of similarity score 0.6681691405470728\n",
      "Rank #12 queried doc in task Feature selection [9] of similarity score 0.6610245512694702\n",
      "Rank #13 queried doc in task Feature selection [10] of similarity score 0.6548049265378848\n",
      "Rank #14 queried doc in task Gene regulatory networks analysis [47] of similarity score 0.652043503279273\n",
      "Rank #15 queried doc in task Integration [28] of similarity score 0.6502933114592351\n",
      "Rank #16 queried doc in task Dimensionality reduction [12] of similarity score 0.6443403191160821\n",
      "Rank #17 queried doc in task Dimensionality reduction [11] of similarity score 0.6390441920651931\n",
      "Rank #18 queried doc in task Integration [22] of similarity score 0.6374797873235141\n",
      "Rank #19 queried doc in task Integration [26] of similarity score 0.6326932590739109\n",
      "Rank #20 queried doc in task Enrichment analysis [39] of similarity score 0.630528098846419\n",
      "Rank #21 queried doc in task Integration [27] of similarity score 0.6295937741287069\n",
      "Rank #22 queried doc in task Integration [30] of similarity score 0.626303993087779\n",
      "Rank #23 queried doc in task Integration [25] of similarity score 0.6239528360479268\n",
      "Rank #24 queried doc in task Dimensionality reduction [13] of similarity score 0.6164597913807885\n",
      "Rank #25 queried doc in task Clustering [14] of similarity score 0.6163338284094582\n",
      "Rank #26 queried doc in task Integration [24] of similarity score 0.6152290564502454\n",
      "Rank #27 queried doc in task Annotation [21] of similarity score 0.6125379897586136\n",
      "Rank #28 queried doc in task Enrichment analysis [44] of similarity score 0.6124887198236657\n",
      "Rank #29 queried doc in task Normalization [7] of similarity score 0.6119046673888408\n",
      "Rank #30 queried doc in task Annotation [17] of similarity score 0.6050746260561292\n",
      "Rank #31 queried doc in task Annotation [19] of similarity score 0.6023522730435847\n",
      "Rank #32 queried doc in task Integration [29] of similarity score 0.6019419547423216\n",
      "Rank #33 queried doc in task Cell-cell communication [41] of similarity score 0.6011336442429899\n",
      "Rank #34 queried doc in task Clustering [15] of similarity score 0.5997423589459999\n",
      "Rank #35 queried doc in task Annotation [18] of similarity score 0.5973335032462781\n",
      "Rank #36 queried doc in task Trajectory inference [36] of similarity score 0.5961831578771691\n",
      "Rank #37 queried doc in task Enrichment analysis [45] of similarity score 0.5960205082868966\n",
      "Rank #38 queried doc in task Compositional analysis [38] of similarity score 0.5955028420087329\n",
      "Rank #39 queried doc in task Trajectory inference [34] of similarity score 0.5661059707904547\n",
      "Rank #40 queried doc in task Compositional analysis [37] of similarity score 0.555541932582871\n",
      "Rank #41 queried doc in task Trajectory inference [35] of similarity score 0.5554924474972611\n",
      "Rank #42 queried doc in task Annotation [20] of similarity score 0.5548077691059663\n",
      "Rank #43 queried doc in task Clustering [16] of similarity score 0.5444717950258539\n",
      "Rank #44 queried doc in task Trajectory inference [31] of similarity score 0.5261118945808662\n",
      "Rank #45 queried doc in task Enrichment analysis [46] of similarity score 0.5061706002849063\n",
      "Rank #46 queried doc in task Cell-cell communication [43] of similarity score 0.4986752123035536\n",
      "Rank #47 queried doc in task Cell-cell communication [40] of similarity score 0.486608030008316\n",
      "Rank #48 queried doc in task Cell-cell communication [42] of similarity score 0.4691095985043561\n"
     ]
    }
   ],
   "source": [
    "# Test vectordb query \n",
    "# Following query is index 0, Task in QC\n",
    "q = \"\"\"\n",
    "This script performs an integrated quality control (QC) analysis on single-cell RNA sequencing data to ensure that only high-quality cells are used for downstream analyses. The code starts by importing necessary libraries, such as Scanpy and anndata, which are essential tools for handling and visualizing high-dimensional transcriptomic data. It then loads a sample dataset??here, a built-in PBMC dataset??to simulate how experimental data might be processed. The first step in the QC workflow involves annotating genes to identify key categories that may indicate cell quality. Genes are flagged as mitochondrial, ribosomal, or hemoglobin based on specific naming conventions; for example, mitochondrial genes usually have names beginning with \"MT-\". These annotations allow the subsequent calculation of QC metrics, including the number of genes expressed per cell, the total counts per cell, and the percentage of transcripts derived from mitochondrial genes. High mitochondrial content, for instance, can serve as a marker for cellular stress or apoptosis. Finally, the script creates violin plots and scatter plots to visualize these metrics, enabling researchers to quickly identify outliers or low-quality cells by observing distributions and relationships between total counts, gene numbers, and mitochondrial percentages. This comprehensive QC process is crucial as it ensures that the data is reliable, allowing for accurate normalization, clustering, and further biological interpretation in subsequent analyses.\n",
    "\"\"\"\n",
    "\n",
    "# Test query\n",
    "results = vector_store.similarity_search_with_score(query=q,k=48)\n",
    "\n",
    "# Plot ranked queries\n",
    "rank = 0\n",
    "for doc, score in results:\n",
    "    rank+=1\n",
    "    print(\"Rank #\"+str(rank)+\" queried doc in task\",doc.metadata['Task'], \"[\"+doc.id+\"] of similarity score\", 1-score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ab03010",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "10188107",
   "metadata": {},
   "source": [
    "# Test retriever"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "4a3545da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Quality control 0\n",
      "Quality control 1\n",
      "Quality control 32\n",
      "Quality control 4\n",
      "Quality control 3\n",
      "Feature selection 33\n",
      "Normalization 5\n",
      "Integration 23\n",
      "Gene regulatory networks analysis 47\n",
      "Dimensionality reduction 11\n"
     ]
    }
   ],
   "source": [
    "q = \"\"\"\n",
    "This script performs an integrated quality control (QC) analysis on single-cell RNA sequencing data to ensure that only high-quality cells are used for downstream analyses. The code starts by importing necessary libraries, such as Scanpy and anndata, which are essential tools for handling and visualizing high-dimensional transcriptomic data. It then loads a sample dataset??here, a built-in PBMC dataset??to simulate how experimental data might be processed. The first step in the QC workflow involves annotating genes to identify key categories that may indicate cell quality. Genes are flagged as mitochondrial, ribosomal, or hemoglobin based on specific naming conventions; for example, mitochondrial genes usually have names beginning with \"MT-\". These annotations allow the subsequent calculation of QC metrics, including the number of genes expressed per cell, the total counts per cell, and the percentage of transcripts derived from mitochondrial genes. High mitochondrial content, for instance, can serve as a marker for cellular stress or apoptosis. Finally, the script creates violin plots and scatter plots to visualize these metrics, enabling researchers to quickly identify outliers or low-quality cells by observing distributions and relationships between total counts, gene numbers, and mitochondrial percentages. This comprehensive QC process is crucial as it ensures that the data is reliable, allowing for accurate normalization, clustering, and further biological interpretation in subsequent analyses.\n",
    "\"\"\"\n",
    "\n",
    "# Build retriever\n",
    "retriever = vector_store.as_retriever(search_type=\"mmr\", search_kwargs={\"k\": 10})\n",
    "\n",
    "# Retrieve \n",
    "docs = retriever.invoke(q)\n",
    "for doc in docs:\n",
    "    print(doc.metadata[\"Task\"],doc.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "73a958cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Quality control 0\n",
      "Quality control 4\n",
      "Quality control 32\n",
      "Normalization 6\n",
      "Integration 30\n",
      "Gene regulatory networks analysis 47\n",
      "Normalization 7\n",
      "Enrichment analysis 39\n",
      "Enrichment analysis 46\n",
      "Compositional analysis 38\n"
     ]
    }
   ],
   "source": [
    "# Build retriever\n",
    "retriever = vector_store.as_retriever(search_type=\"mmr\", search_kwargs={\"k\": 10})\n",
    "\n",
    "# Retrieve \n",
    "docs = retriever.invoke(\"quality control\")\n",
    "for doc in docs:\n",
    "    print(doc.metadata[\"Task\"],doc.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3079b46d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "84bd31bf",
   "metadata": {},
   "source": [
    "# RefCodeDB data processing\n",
    "\n",
    "In this part, we will reconstruct the RefCodeDB data from collected code blocks, which is:  \n",
    "code block -> task instruction -> task description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8330d456",
   "metadata": {},
   "outputs": [],
   "source": [
    "api = 'sk-NYPE5WNGTK781Nox801a5934A9F84a03BfC525359eBaE7B6'\n",
    "url = 'http://10.224.28.80:3000/v1'\n",
    "model = \"deepseek-v3-hs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8b32d3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from langchain_openai import ChatOpenAI # we use openai as example here "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd0d6396",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AIMessage(content=\"It seems like you're testing or checking something. Could you clarify what you'd like to test or what you need assistance with? I'm here to help! 😊\", additional_kwargs={'refusal': None}, response_metadata={'token_usage': {'completion_tokens': 34, 'prompt_tokens': 4, 'total_tokens': 38, 'completion_tokens_details': {'audio_tokens': None, 'reasoning_tokens': 0}, 'prompt_tokens_details': {'audio_tokens': None, 'cached_tokens': 0}}, 'model_name': 'deepseek-v3-241226', 'system_fingerprint': None, 'finish_reason': 'stop', 'logprobs': None}, id='run-7a48aff8-5d2e-4a89-82f0-989508e32a3a-0', usage_metadata={'input_tokens': 4, 'output_tokens': 34, 'total_tokens': 38, 'input_token_details': {'cache_read': 0}, 'output_token_details': {'reasoning': 0}})"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set up LLM \n",
    "llm = ChatOpenAI(api_key=api,base_url=url,model=model)\n",
    "llm.invoke(\"test\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f4156f52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load raw table\n",
    "raw_table = pd.read_csv('rawtable/RefCodeDB.csv')\n",
    "\n",
    "# Load workflow guideline\n",
    "with open('Single-Cell RNA Sequencing Data Analysis Workflow Guide.markdown', 'r', encoding='utf-8') as f:\n",
    "    gline = f.read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b4d86392",
   "metadata": {},
   "outputs": [],
   "source": [
    "prompt_code2instruction = \"\"\"\"\n",
    "Act as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n",
    "\n",
    "1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n",
    "\n",
    "2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n",
    "   - The purpose of this code block\n",
    "   - Packages/tools used in this code block\n",
    "   - Performing the main analysis or computations\n",
    "   - Creating visualizations or plots\n",
    "   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n",
    "\n",
    "3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n",
    "   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n",
    "   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n",
    "   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n",
    "\n",
    "4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n",
    "\n",
    "5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n",
    "   - Data format and any preprocessing steps\n",
    "   - Specific requirements or assumptions (e.g., file paths, data structure)\n",
    "   - Any custom functions or scripts used\n",
    "\n",
    "6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n",
    "\n",
    "7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n",
    "\n",
    "8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n",
    "\n",
    "## Code to be re-formatted:\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "prompt_instruct2descript = \"\"\"\n",
    "This prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n",
    "\n",
    " - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n",
    " - The programming language and main libraries or tools used.\n",
    " - Key steps such as quality control, data processing, computation, and visualization.\n",
    " - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n",
    " - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n",
    " - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n",
    " - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n",
    " - Do not start with `we performed`.\n",
    "\n",
    "\n",
    "## Example\n",
    "\n",
    "For an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n",
    "\n",
    "## Workflow Guideline to be aligned:\n",
    "\"\"\" + gline + \"\\n\\n## Instruction to be transformed:\\n\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "36ccfcb7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Instr_input",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Instructions",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Dscpt_input",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Description",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "2edb53f7-3ab5-48ed-a43a-d71e2d834940",
       "rows": [
        [
         "0",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#----------------\n\n# Import necessary modules\n\n#----------------\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n# Set figure parameters for decent plotting\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#----------------\n\n# Data Input: load a sample dataset\n\n#----------------\n\nadata = sc.datasets.pbmc3k()\n\n# Data format: \n\n# adata is an AnnData object containing the following core structure \n\n# x: core data matrix (cells x genes), usually a sparse matrix or numpy array\n\n# obs: cell-level metadata (DataFrame), stores properties for each cell\n\n# var: gene-level metadata (DataFrame), stores attributes for each gene\n\n\n\n\n\n#----------------\n\n# Annotate genes for QC metrics and calculate metrics\n\n#----------------\n\n\n\n# Mark mitochondrial genes\n\nadata.var[\"t\"] = adata.var_names.str.startswith(\"MT-\")\n\n\n\n# Mark ribosomal genes\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\n\n\n\n# Mark hemoglobin genes\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\n\n\n\n# Calculate QC metrics\n\nsc.pp.calculate_qc_metrics(\n\nadata, \n\nqc_vars=[\"mt\", \"ribo\", \"hb\"],\n\ninplace=True,\n\nlog1p=True\n\n)\n\n\n\n#----------------\n\n# Create QC visualization plots\n\n#----------------\n\n# Violin plots for key metrics\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n\njitter=0.4,\n\nmulti_panel=True\n\n)\n\n\n\n# Scatter plot showing counts vs genes with mitochondrial percentage\n\nsc.pl.scatter(\n\nadata, \n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ncolor=\"pct_counts_mt\"\n\n)\n\n\n\n#----------------\n\n# Data output \n\n#----------------\n\nadata.write('pbmc3k_qc.h5ad'))",
         "The analysis begins by importing the necessary Python modules. The `scanpy` module is imported as `sc`, which is a widely used library for single-cell RNA sequencing (scRNA-seq) data analysis. Additionally, the `anndata` module is imported as `ad`, which provides a data structure for handling annotated data matrices, a common format for scRNA-seq data. The figure parameters are set using `sc.settings.set_figure_params` to ensure that plots are generated with a resolution of 80 DPI and a white background, which improves the clarity and visual quality of the resulting figures.\n\nThe dataset used in this analysis is the `pbmc3k` dataset, a publicly available scRNA-seq dataset of peripheral blood mononuclear cells (PBMCs). This dataset is loaded into an `AnnData` object named `adata`. The `AnnData` object is a structured format that includes a core data matrix (`x`), cell-level metadata (`obs`), and gene-level metadata (`var`). The core data matrix represents the gene expression counts for each cell, while the metadata stores additional information about the cells and genes.\n\nNext, the genes in the dataset are annotated to identify specific gene types that are relevant for quality control (QC) metrics. Mitochondrial genes are identified by checking if the gene names start with \"MT-\". Ribosomal genes are identified by checking if the gene names start with \"RPS\" or \"RPL\". Hemoglobin genes are identified by checking if the gene names match the pattern \"^HB[^(P)]\". These annotations are stored in the `var` attribute of the `adata` object.\n\nQC metrics are then calculated using the `sc.pp.calculate_qc_metrics` function. This function computes various QC metrics, including the number of genes detected per cell (`n_genes_by_counts`), the total counts per cell (`total_counts`), and the percentage of counts from mitochondrial genes (`pct_counts_mt`). The `qc_vars` parameter specifies the gene types to include in the QC calculations, and the `log1p` parameter applies a log transformation to the data to stabilize variance. The results are stored in the `adata` object for further analysis.\n\nTo visualize the QC metrics, two types of plots are generated. First, violin plots are created using the `sc.pl.violin` function to display the distribution of key QC metrics, including the number of genes detected per cell, the total counts per cell, and the percentage of mitochondrial counts. The `jitter` parameter is set to 0.4 to add a small amount of random noise to the data points, improving the visibility of overlapping points. The `multi_panel` parameter ensures that the plots are displayed in a multi-panel layout for easier comparison. Second, a scatter plot is generated using the `sc.pl.scatter` function to visualize the relationship between the total counts per cell and the number of genes detected per cell, with the points colored by the percentage of mitochondrial counts. This plot helps identify potential outliers or low-quality cells based on their QC metrics.\n\nFinally, the annotated and QC-processed dataset is saved to a file named `pbmc3k_qc.h5ad` using the `adata.write` function. This file can be used for downstream analysis or shared with collaborators. The `h5ad` format is a standard file format for storing `AnnData` objects, ensuring compatibility with other tools and workflows in the scRNA-seq analysis ecosystem.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing the necessary Python modules. The `scanpy` module is imported as `sc`, which is a widely used library for single-cell RNA sequencing (scRNA-seq) data analysis. Additionally, the `anndata` module is imported as `ad`, which provides a data structure for handling annotated data matrices, a common format for scRNA-seq data. The figure parameters are set using `sc.settings.set_figure_params` to ensure that plots are generated with a resolution of 80 DPI and a white background, which improves the clarity and visual quality of the resulting figures.\n\nThe dataset used in this analysis is the `pbmc3k` dataset, a publicly available scRNA-seq dataset of peripheral blood mononuclear cells (PBMCs). This dataset is loaded into an `AnnData` object named `adata`. The `AnnData` object is a structured format that includes a core data matrix (`x`), cell-level metadata (`obs`), and gene-level metadata (`var`). The core data matrix represents the gene expression counts for each cell, while the metadata stores additional information about the cells and genes.\n\nNext, the genes in the dataset are annotated to identify specific gene types that are relevant for quality control (QC) metrics. Mitochondrial genes are identified by checking if the gene names start with \"MT-\". Ribosomal genes are identified by checking if the gene names start with \"RPS\" or \"RPL\". Hemoglobin genes are identified by checking if the gene names match the pattern \"^HB[^(P)]\". These annotations are stored in the `var` attribute of the `adata` object.\n\nQC metrics are then calculated using the `sc.pp.calculate_qc_metrics` function. This function computes various QC metrics, including the number of genes detected per cell (`n_genes_by_counts`), the total counts per cell (`total_counts`), and the percentage of counts from mitochondrial genes (`pct_counts_mt`). The `qc_vars` parameter specifies the gene types to include in the QC calculations, and the `log1p` parameter applies a log transformation to the data to stabilize variance. The results are stored in the `adata` object for further analysis.\n\nTo visualize the QC metrics, two types of plots are generated. First, violin plots are created using the `sc.pl.violin` function to display the distribution of key QC metrics, including the number of genes detected per cell, the total counts per cell, and the percentage of mitochondrial counts. The `jitter` parameter is set to 0.4 to add a small amount of random noise to the data points, improving the visibility of overlapping points. The `multi_panel` parameter ensures that the plots are displayed in a multi-panel layout for easier comparison. Second, a scatter plot is generated using the `sc.pl.scatter` function to visualize the relationship between the total counts per cell and the number of genes detected per cell, with the points colored by the percentage of mitochondrial counts. This plot helps identify potential outliers or low-quality cells based on their QC metrics.\n\nFinally, the annotated and QC-processed dataset is saved to a file named `pbmc3k_qc.h5ad` using the `adata.write` function. This file can be used for downstream analysis or shared with collaborators. The `h5ad` format is a standard file format for storing `AnnData` objects, ensuring compatibility with other tools and workflows in the scRNA-seq analysis ecosystem.",
         "The analysis was conducted in Python using the Scanpy library, which is specialized for single-cell RNA sequencing (scRNA-seq) data. The `pbmc3k` dataset, comprising peripheral blood mononuclear cells, was loaded into an `AnnData` object for structured data handling. Gene annotations were performed to identify mitochondrial, ribosomal, and hemoglobin genes, which are critical for quality control (QC) metrics. QC metrics, including gene counts, total counts, and mitochondrial percentages, were calculated and visualized using violin and scatter plots to identify potential outliers. The processed dataset was saved in the `h5ad` format for downstream analysis."
        ],
        [
         "1",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#----------------\n\n# Package Load\n\n#----------------\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#----------------\n\n# Data Input\n\n#----------------\n\n# Load built-in PBMC 3k dataset\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Add dummy sample identifier required for batch-level analysis\n\nadata.obs[\"sample\"] = \"sample1\"\n\n\n\n#----------------\n\n# Preprocessing & Quality Control\n\n#----------------\n\n# Flag mitochondrial, ribosomal and hemoglobin genes\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\") # Mitochondrial genes\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\")) # Ribosomal genes\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\") # Hemoglobin genes\n\n\n\n# Calculate quality control metrics\n\nsc.pp.calculate_qc_metrics(\n\nadata, \n\nqc_vars=[\"mt\", \"ribo\", \"hb\"],\n\ninplace=True,\n\nlog1p=True\n\n)\n\n\n\n# Detect potential doublets using Scrublet\n\nsc.pp.scrublet(adata, batch_key=\"sample\")\n\n\n\n#----------------\n\n# Visualization\n\n#----------------\n\n# Configure plotting parameters\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\")\n\n\n\n# QC distributions\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\ntitle=\"Quality Control Distributions\"\n\n)\n\n\n\n# Relationship between read depth and mitochondrial content\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ncolor=\"pct_counts_mt\",\n\ntitle=\"Gene Counts vs. Total Reads (MT% Colored)\"\n\n)\n\n\n\n# Doublet visualization\n\nfor color_field in [\"predicted_doublet\", \"doublet_score\"]:\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ncolor=color_field,\n\ntitle=f\"Doublet Detection: {color_field.replace('_', ' ').title()}\"\n\n)\n\n\n\n#----------------\n\n# Data Output\n\n#----------------\n\n# Save processed data for future analysis\n\nadata.write(\"processed_pbmc3k.h5ad\", compression=\"gzip\")",
         "The analysis begins by loading the necessary Python packages, Scanpy and AnnData, which are essential for single-cell RNA sequencing data analysis. Scanpy provides tools for preprocessing, visualization, and analysis, while AnnData is used for storing and manipulating single-cell data in a structured format.\n\nThe dataset used in this analysis is the built-in PBMC 3k dataset, which contains single-cell RNA sequencing data from peripheral blood mononuclear cells. A dummy sample identifier, \"sample1,\" is added to the dataset to facilitate batch-level analysis, ensuring compatibility with downstream functions that require sample-level information.\n\nPreprocessing and quality control steps are then performed. Mitochondrial, ribosomal, and hemoglobin genes are flagged based on their gene names. Mitochondrial genes are identified by the prefix \"MT-,\" ribosomal genes by the prefixes \"RPS\" or \"RPL,\" and hemoglobin genes by the pattern \"^HB[^(P)].\" Quality control metrics are calculated using the `calculate_qc_metrics` function, which computes metrics such as the number of genes detected per cell, total counts per cell, and the percentage of counts from mitochondrial genes. The `log1p` parameter is set to `True` to apply a log transformation to the data, which is useful for normalizing highly skewed distributions. Potential doublets, which are artifacts resulting from two cells being captured in the same droplet, are detected using the Scrublet algorithm, with the `batch_key` parameter set to \"sample\" to ensure batch-specific detection.\n\nVisualization of the data is conducted to assess quality control metrics and doublet detection. The figure parameters are configured with a DPI of 80 and a white background for clarity. Violin plots are generated to visualize the distributions of key quality control metrics, including the number of genes detected per cell, total counts per cell, and the percentage of mitochondrial counts. Scatter plots are used to explore the relationship between read depth (total counts) and the number of genes detected, with mitochondrial content (percentage of mitochondrial counts) represented by color. Additionally, scatter plots are created to visualize the results of doublet detection, showing the predicted doublet status and doublet score for each cell.\n\nFinally, the processed dataset, including all annotations and quality control metrics, is saved in the H5AD format using the `write` function with GZIP compression. This file can be loaded in future analyses to avoid reprocessing the data. The saved file is named \"processed_pbmc3k.h5ad.\" This step ensures that the preprocessed data is preserved for subsequent analyses or sharing with collaborators.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by loading the necessary Python packages, Scanpy and AnnData, which are essential for single-cell RNA sequencing data analysis. Scanpy provides tools for preprocessing, visualization, and analysis, while AnnData is used for storing and manipulating single-cell data in a structured format.\n\nThe dataset used in this analysis is the built-in PBMC 3k dataset, which contains single-cell RNA sequencing data from peripheral blood mononuclear cells. A dummy sample identifier, \"sample1,\" is added to the dataset to facilitate batch-level analysis, ensuring compatibility with downstream functions that require sample-level information.\n\nPreprocessing and quality control steps are then performed. Mitochondrial, ribosomal, and hemoglobin genes are flagged based on their gene names. Mitochondrial genes are identified by the prefix \"MT-,\" ribosomal genes by the prefixes \"RPS\" or \"RPL,\" and hemoglobin genes by the pattern \"^HB[^(P)].\" Quality control metrics are calculated using the `calculate_qc_metrics` function, which computes metrics such as the number of genes detected per cell, total counts per cell, and the percentage of counts from mitochondrial genes. The `log1p` parameter is set to `True` to apply a log transformation to the data, which is useful for normalizing highly skewed distributions. Potential doublets, which are artifacts resulting from two cells being captured in the same droplet, are detected using the Scrublet algorithm, with the `batch_key` parameter set to \"sample\" to ensure batch-specific detection.\n\nVisualization of the data is conducted to assess quality control metrics and doublet detection. The figure parameters are configured with a DPI of 80 and a white background for clarity. Violin plots are generated to visualize the distributions of key quality control metrics, including the number of genes detected per cell, total counts per cell, and the percentage of mitochondrial counts. Scatter plots are used to explore the relationship between read depth (total counts) and the number of genes detected, with mitochondrial content (percentage of mitochondrial counts) represented by color. Additionally, scatter plots are created to visualize the results of doublet detection, showing the predicted doublet status and doublet score for each cell.\n\nFinally, the processed dataset, including all annotations and quality control metrics, is saved in the H5AD format using the `write` function with GZIP compression. This file can be loaded in future analyses to avoid reprocessing the data. The saved file is named \"processed_pbmc3k.h5ad.\" This step ensures that the preprocessed data is preserved for subsequent analyses or sharing with collaborators.",
         "The analysis was conducted in Python using Scanpy and AnnData for single-cell RNA sequencing data processing and visualization. The PBMC 3k dataset was preprocessed by flagging mitochondrial, ribosomal, and hemoglobin genes based on specific naming patterns. Quality control metrics, including gene counts, total counts, and mitochondrial percentages, were calculated and log-transformed. Doublets were detected using the Scrublet algorithm, with batch-specific adjustments. Data quality was assessed through violin and scatter plots, visualizing metrics such as read depth, gene counts, and mitochondrial content, alongside doublet scores. The processed dataset was saved in H5AD format for future use."
        ],
        [
         "2",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, data handling, and plotting\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport seaborn as sns\n\nfrom scipy.stats import median_abs_deviation\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\n# Verbosity 0 supresses most of the Scanpy output messages.\n\nsc.settings.verbosity = 0\n\n# Set figure parameters for high-quality plots.\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the single-cell dataset from a 10x Genomics HDF5 file\n\nadata = sc.read_10x_h5(filename=\"filtered_feature_bc_matrix.h5\")\n\n\n\n# Ensure that all gene names (variable names) are unique to prevent issues in downstream analysis\n\nadata.var_names_make_unique()\n\n\n\n#--------------------------------\n\n# Quality Control (QC) and Filtering\n\n#--------------------------------\n\n\n\n# Identify different sets of genes for quality control metrics\n\n# MT- genes are mitochondrial genes, which can indicate cell stress or damage if highly expressed.\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n\n# RPS and RPL genes are ribosomal genes.\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\n\n# HB genes are hemoglobin genes, which can indicate contamination from red blood cells.\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\n\n\n\n# Calculate QC metrics for each cell.\n\n# This computes metrics like total counts, number of genes, and percentages of mt, ribo, and hb genes.\n\nsc.pp.calculate_qc_metrics(\n\nadata, qc_vars=[\"mt\", \"ribo\", \"hb\"], inplace=True, percent_top=[20], log1p=True\n\n)\n\n\n\n# Define a function to identify outlier cells based on the Median Absolute Deviation (MAD).\n\n# MAD is a robust measure of variability and is less sensitive to extreme outliers than standard deviation.\n\ndef is_outlier(adata, metric: str, nmads: int):\n\n\"\"\"\n\nIdentifies outliers for a given metric in the AnnData object.\n\n\n\nArgs:\n\nadata: AnnData object containing the data.\n\nmetric: The observation metric to check for outliers (e.g., 'log1p_total_counts').\n\nnmads: The number of median absolute deviations to use as the threshold.\n\n\n\nReturns:\n\nA boolean Series indicating whether each cell is an outlier.\n\n\"\"\"\n\nM = adata.obs[metric]\n\n# An observation is an outlier if it is nmads MADs away from the median.\n\noutlier = (M < np.median(M) - nmads * median_abs_deviation(M)) | \\\n\n(np.median(M) + nmads * median_abs_deviation(M) < M)\n\nreturn outlier\n\n\n\n# Identify outlier cells based on general QC metrics.\n\n# Here, we mark cells as outliers if they are outliers for total counts, number of genes, or\n\n# the percentage of counts in the top 20 genes.\n\nadata.obs[\"outlier\"] = (\n\nis_outlier(adata, \"log1p_total_counts\", 5) |\n\nis_outlier(adata, \"log1p_n_genes_by_counts\", 5) |\n\nis_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n\n)\n\n\n\n# Identify cells with high mitochondrial gene content.\n\n# Cells are marked as outliers if their mitochondrial content is more than 3 MADs from the median\n\n# or if it exceeds a hard threshold of 8%.\n\nadata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 3) | (adata.obs[\"pct_counts_mt\"] > 8)\n\n\n\n# Print the number of cells before filtering\n\nprint(f\"Total number of cells: {adata.n_obs}\")\n\n\n\n# Filter out the low-quality cells identified in the previous steps.\n\n# A .copy() is used to create a new, clean AnnData object.\n\nadata = adata[(~adata.obs.outlier) & (~adata.obs.mt_outlier)].copy()\n\n\n\n# Print the number of cells remaining after filtering\n\nprint(f\"Number of cells after filtering of low-quality cells: {adata.n_obs}\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the distribution of total counts per cell after filtering.\n\n# This helps confirm that cells with extremely low or high counts have been removed.\n\nsns.displot(adata.obs[\"total_counts\"], bins=100, kde=False)\n\n\n\n# Create a violin plot to show the distribution of mitochondrial gene percentages.\n\n# This plot helps to visually inspect the effectiveness of the mitochondrial content filtering.\n\nsc.pl.violin(adata, \"pct_counts_mt\")\n\n\n\n# Create a scatter plot to visualize the relationship between total counts,\n\n# number of genes, and mitochondrial content.\n\n# This plot is useful for identifying any remaining artifacts or relationships in the data.\n\nsc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the filtered and processed AnnData object to a file.\n\n# This file can be used for subsequent downstream analyses like normalization,\n\n# clustering, and differential expression.\n\nadata.write(\"filtered_quality_control.h5ad\")\n\n",
         "The following section provides a detailed description of the steps involved in processing and analyzing single-cell RNA sequencing (scRNA-seq) data using Python, specifically leveraging the Scanpy library. The workflow includes data loading, quality control, filtering, visualization, and saving the processed data for downstream analysis.\n\nFirst, the necessary Python libraries are imported. These include NumPy for numerical operations, Scanpy for single-cell data analysis, Seaborn for plotting, and the median_absolute_deviation function from SciPy for robust statistical calculations. Scanpy is configured to suppress verbose output messages and to generate high-quality plots with a white background and no frame.\n\nThe scRNA-seq dataset is loaded from a 10x Genomics HDF5 file using the `sc.read_10x_h5` function. The dataset is stored in an AnnData object, a data structure specifically designed for single-cell data. To ensure compatibility with downstream analysis, gene names are made unique using the `adata.var_names_make_unique` method.\n\nQuality control (QC) metrics are calculated to identify and filter low-quality cells. Mitochondrial (MT), ribosomal (RPS/RPL), and hemoglobin (HB) genes are annotated in the dataset, as their expression levels can indicate cell stress, contamination, or other artifacts. QC metrics, including total counts, number of genes, and the percentage of counts in the top 20 genes, are computed using the `sc.pp.calculate_qc_metrics` function. Additionally, the percentage of mitochondrial, ribosomal, and hemoglobin genes is calculated.\n\nOutlier cells are identified using a custom function, `is_outlier`, which applies the Median Absolute Deviation (MAD) method to detect cells with extreme values for total counts, number of genes, or the percentage of counts in the top 20 genes. Cells are marked as outliers if their values deviate by more than a specified number of MADs from the median. Mitochondrial outliers are also identified, with cells flagged if their mitochondrial content exceeds 8% or deviates by more than 3 MADs from the median. The total number of cells before and after filtering is printed to track the impact of the filtering steps.\n\nVisualizations are generated to assess the effectiveness of the filtering steps. A histogram of total counts per cell is plotted using Seaborn to confirm the removal of cells with extreme counts. A violin plot of mitochondrial gene percentages is created using Scanpy to inspect the distribution of mitochondrial content. A scatter plot is generated to visualize the relationship between total counts, number of genes, and mitochondrial content, helping to identify any remaining artifacts or trends in the data.\n\nFinally, the filtered and processed AnnData object is saved to an HDF5 file (`filtered_quality_control.h5ad`) for use in subsequent analyses, such as normalization, clustering, and differential expression. This file contains the high-quality cells retained after filtering and is ready for further exploration.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following section provides a detailed description of the steps involved in processing and analyzing single-cell RNA sequencing (scRNA-seq) data using Python, specifically leveraging the Scanpy library. The workflow includes data loading, quality control, filtering, visualization, and saving the processed data for downstream analysis.\n\nFirst, the necessary Python libraries are imported. These include NumPy for numerical operations, Scanpy for single-cell data analysis, Seaborn for plotting, and the median_absolute_deviation function from SciPy for robust statistical calculations. Scanpy is configured to suppress verbose output messages and to generate high-quality plots with a white background and no frame.\n\nThe scRNA-seq dataset is loaded from a 10x Genomics HDF5 file using the `sc.read_10x_h5` function. The dataset is stored in an AnnData object, a data structure specifically designed for single-cell data. To ensure compatibility with downstream analysis, gene names are made unique using the `adata.var_names_make_unique` method.\n\nQuality control (QC) metrics are calculated to identify and filter low-quality cells. Mitochondrial (MT), ribosomal (RPS/RPL), and hemoglobin (HB) genes are annotated in the dataset, as their expression levels can indicate cell stress, contamination, or other artifacts. QC metrics, including total counts, number of genes, and the percentage of counts in the top 20 genes, are computed using the `sc.pp.calculate_qc_metrics` function. Additionally, the percentage of mitochondrial, ribosomal, and hemoglobin genes is calculated.\n\nOutlier cells are identified using a custom function, `is_outlier`, which applies the Median Absolute Deviation (MAD) method to detect cells with extreme values for total counts, number of genes, or the percentage of counts in the top 20 genes. Cells are marked as outliers if their values deviate by more than a specified number of MADs from the median. Mitochondrial outliers are also identified, with cells flagged if their mitochondrial content exceeds 8% or deviates by more than 3 MADs from the median. The total number of cells before and after filtering is printed to track the impact of the filtering steps.\n\nVisualizations are generated to assess the effectiveness of the filtering steps. A histogram of total counts per cell is plotted using Seaborn to confirm the removal of cells with extreme counts. A violin plot of mitochondrial gene percentages is created using Scanpy to inspect the distribution of mitochondrial content. A scatter plot is generated to visualize the relationship between total counts, number of genes, and mitochondrial content, helping to identify any remaining artifacts or trends in the data.\n\nFinally, the filtered and processed AnnData object is saved to an HDF5 file (`filtered_quality_control.h5ad`) for use in subsequent analyses, such as normalization, clustering, and differential expression. This file contains the high-quality cells retained after filtering and is ready for further exploration.",
         "The analysis of single-cell RNA sequencing (scRNA-seq) data was conducted in Python using the Scanpy library, complemented by NumPy for numerical operations and Seaborn for visualization. The workflow began with loading the dataset from a 10x Genomics HDF5 file into an AnnData object, ensuring unique gene names for downstream compatibility. Quality control metrics, including total counts, gene numbers, and mitochondrial, ribosomal, and hemoglobin gene percentages, were calculated to identify low-quality cells. Outliers were detected using the Median Absolute Deviation (MAD) method, with cells exhibiting extreme values filtered out. Visualization tools, such as histograms, violin plots, and scatter plots, were employed to assess the effectiveness of filtering and to inspect data distributions. The processed dataset was saved in HDF5 format for subsequent analyses."
        ],
        [
         "3",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, data handling, and plotting\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport seaborn as sns\n\nimport matplotlib.pyplot as plt\n\nfrom scipy.stats import median_abs_deviation\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\n# Verbosity 0 supresses most of the Scanpy output messages.\n\nsc.settings.verbosity = 0\n\n# Set figure parameters for high-quality plots.\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the single-cell dataset from a 10x Genomics HDF5 file\n\nadata = sc.read_10x_h5(filename=\"filtered_feature_bc_matrix.h5\")\n\n\n\n# Ensure that all gene names (variable names) are unique\n\nadata.var_names_make_unique()\n\n\n\n#--------------------------------\n\n# Quality Control (QC) and Filtering\n\n#--------------------------------\n\n\n\n# Identify different sets of genes for quality control metrics\n\n# MT- genes are mitochondrial. High percentage can indicate cell stress.\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n\n# RPS and RPL genes are ribosomal.\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\n\n# HB genes are hemoglobin. High percentage can indicate red blood cell contamination.\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\n\n\n\n# Calculate QC metrics for each cell.\n\nsc.pp.calculate_qc_metrics(\n\nadata, qc_vars=[\"mt\", \"ribo\", \"hb\"], inplace=True, percent_top=[20], log1p=True\n\n)\n\n\n\n# Define a function to identify outlier cells based on Median Absolute Deviation (MAD)\n\ndef is_outlier(adata, metric: str, nmads: int):\n\n\"\"\"\n\nIdentifies outliers for a given metric in the AnnData object.\n\n\n\nArgs:\n\nadata: AnnData object containing the data.\n\nmetric: The observation metric to check for outliers (e.g., 'log1p_total_counts').\n\nnmads: The number of median absolute deviations to use as the threshold.\n\n\n\nReturns:\n\nA boolean Series indicating whether each cell is an outlier.\n\n\"\"\"\n\nM = adata.obs[metric]\n\nmad = median_abs_deviation(M)\n\nmedian_val = np.median(M)\n\n# An observation is an outlier if it is nmads MADs away from the median.\n\noutlier = (M < median_val - nmads * mad) | (M > median_val + nmads * mad)\n\nreturn outlier\n\n\n\n# Identify outlier cells based on general QC metrics\n\nadata.obs[\"outlier\"] = (\n\nis_outlier(adata, \"log1p_total_counts\", 5) |\n\nis_outlier(adata, \"log1p_n_genes_by_counts\", 5) |\n\nis_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n\n)\n\n\n\n# Identify cells with high mitochondrial gene content\n\nadata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 3) | (adata.obs[\"pct_counts_mt\"] > 8)\n\n\n\n# Print the number of cells before filtering\n\nprint(f\"Total number of cells before filtering: {adata.n_obs}\")\n\n\n\n# Filter out the low-quality cells\n\nadata = adata[(~adata.obs[\"outlier\"]) & (~adata.obs[\"mt_outlier\"])].copy()\n\n\n\n# Print the number of cells remaining after filtering\n\nprint(f\"Total number of cells after filtering: {adata.n_obs}\")\n\n\n\n#--------------------------------\n\n# Doublet Detection\n\n#--------------------------------\n\n\n\n# Ensure a 'sample' column exists for batch processing, as required by scrublet.\n\n# If not present, create a dummy sample column.\n\nif \"sample\" not in adata.obs:\n\nadata.obs[\"sample\"] = \"sample1\"\n\n\n\n# Run Scrublet for doublet detection. This simulates artificial doublets\n\n# from the data and calculates a score for each real cell based on its\n\n# similarity to the simulated doublets.\n\nsc.pp.scrublet(adata, batch_key=\"sample\")\n\n\n\n# Compute a UMAP embedding for visualization purposes. This is necessary\n\n# to visualize doublet scores on a 2D projection of the data.\n\n# Note: This is a preliminary UMAP for visualization, not the final one for analysis.\n\nsc.pp.neighbors(adata)\n\nsc.tl.umap(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- QC Plots ---\n\n# Visualize the distribution of total counts per cell after filtering.\n\nsns.displot(adata.obs[\"total_counts\"], bins=100, kde=False)\n\n\n\n# Create a violin plot to show the distribution of mitochondrial gene percentages.\n\nsc.pl.violin(adata, \"pct_counts_mt\")\n\n\n\n# Create a scatter plot to visualize total counts vs. number of genes.\n\nsc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")\n\n\n\n\n\n# --- Doublet Plots ---\n\n# Plot a histogram of the doublet scores to help determine a threshold for calling doublets.\n\nplt.figure(figsize=(8, 4))\n\nplt.hist(adata.obs[\"doublet_score\"], bins=50, color=\"skyblue\", edgecolor=\"black\")\n\nplt.xlabel(\"Doublet Score\")\n\nplt.ylabel(\"Frequency\")\n\nplt.title(\"Doublet Score Distribution\")\n\nplt.show()\n\n\n\n\n\n# Visualize the UMAP embedding colored by doublet score and predicted doublet status.\n\n# This helps to see if predicted doublets form distinct clusters or are located\n\n# between major cell type clusters.\n\nsc.pl.umap(adata, color=[\"doublet_score\", \"predicted_doublet\"], wspace=0.5)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the filtered and processed AnnData object. This file now includes\n\n# doublet scores and predictions, ready for downstream analysis.\n\nadata.write(\"filtered_quality_control_with_doublet_detection.h5ad\")",
         "The analysis begins by importing essential Python libraries for numerical operations, data handling, and visualization, including NumPy, Scanpy, Seaborn, and Matplotlib. These libraries are critical for performing single-cell RNA sequencing (scRNA-seq) analysis, generating plots, and calculating statistical metrics. Global parameters for Scanpy are configured to suppress verbose output messages and set high-quality figure parameters for plotting.\n\nThe single-cell dataset is loaded from a 10x Genomics HDF5 file using the `sc.read_10x_h5` function. The dataset is processed to ensure all gene names are unique, which is necessary to avoid conflicts during downstream analysis. Quality control (QC) metrics are then calculated to assess cell quality. Specific gene sets are identified for QC, including mitochondrial genes (indicative of cell stress), ribosomal genes, and hemoglobin genes (indicative of red blood cell contamination). The `sc.pp.calculate_qc_metrics` function computes these metrics, including the percentage of counts in the top 20 genes and log-transformed total counts.\n\nOutlier cells are identified using a custom function based on the Median Absolute Deviation (MAD). This function flags cells as outliers if their values for total counts, number of genes, or percentage of counts in the top 20 genes deviate significantly from the median. Additionally, cells with high mitochondrial gene content are flagged as outliers. The dataset is filtered to remove low-quality cells based on these outlier criteria, and the number of cells before and after filtering is reported.\n\nDoublet detection is performed using the Scrublet algorithm, which simulates artificial doublets and calculates a doublet score for each cell. A preliminary UMAP embedding is generated to visualize the doublet scores and predicted doublet status. This step helps identify potential doublets that may confound downstream analysis.\n\nSeveral visualizations are created to assess the quality of the filtered dataset. These include a histogram of total counts per cell, a violin plot of mitochondrial gene percentages, and a scatter plot of total counts versus the number of genes, colored by mitochondrial gene percentage. Additionally, a histogram of doublet scores and UMAP plots colored by doublet score and predicted doublet status are generated to evaluate the effectiveness of the doublet detection process.\n\nFinally, the filtered and processed dataset, including QC metrics and doublet scores, is saved as an AnnData object in HDF5 format for downstream analysis. This file contains all necessary information for subsequent steps in the scRNA-seq analysis pipeline.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for numerical operations, data handling, and visualization, including NumPy, Scanpy, Seaborn, and Matplotlib. These libraries are critical for performing single-cell RNA sequencing (scRNA-seq) analysis, generating plots, and calculating statistical metrics. Global parameters for Scanpy are configured to suppress verbose output messages and set high-quality figure parameters for plotting.\n\nThe single-cell dataset is loaded from a 10x Genomics HDF5 file using the `sc.read_10x_h5` function. The dataset is processed to ensure all gene names are unique, which is necessary to avoid conflicts during downstream analysis. Quality control (QC) metrics are then calculated to assess cell quality. Specific gene sets are identified for QC, including mitochondrial genes (indicative of cell stress), ribosomal genes, and hemoglobin genes (indicative of red blood cell contamination). The `sc.pp.calculate_qc_metrics` function computes these metrics, including the percentage of counts in the top 20 genes and log-transformed total counts.\n\nOutlier cells are identified using a custom function based on the Median Absolute Deviation (MAD). This function flags cells as outliers if their values for total counts, number of genes, or percentage of counts in the top 20 genes deviate significantly from the median. Additionally, cells with high mitochondrial gene content are flagged as outliers. The dataset is filtered to remove low-quality cells based on these outlier criteria, and the number of cells before and after filtering is reported.\n\nDoublet detection is performed using the Scrublet algorithm, which simulates artificial doublets and calculates a doublet score for each cell. A preliminary UMAP embedding is generated to visualize the doublet scores and predicted doublet status. This step helps identify potential doublets that may confound downstream analysis.\n\nSeveral visualizations are created to assess the quality of the filtered dataset. These include a histogram of total counts per cell, a violin plot of mitochondrial gene percentages, and a scatter plot of total counts versus the number of genes, colored by mitochondrial gene percentage. Additionally, a histogram of doublet scores and UMAP plots colored by doublet score and predicted doublet status are generated to evaluate the effectiveness of the doublet detection process.\n\nFinally, the filtered and processed dataset, including QC metrics and doublet scores, is saved as an AnnData object in HDF5 format for downstream analysis. This file contains all necessary information for subsequent steps in the scRNA-seq analysis pipeline.",
         "The analysis was conducted in Python using Scanpy, NumPy, Seaborn, and Matplotlib for single-cell RNA sequencing (scRNA-seq) data processing and visualization. The dataset, loaded from a 10x Genomics HDF5 file, underwent quality control (QC) to assess cell integrity. QC metrics, including mitochondrial, ribosomal, and hemoglobin gene percentages, were calculated, and outlier cells were identified using Median Absolute Deviation (MAD). Doublet detection was performed using Scrublet, with results visualized on UMAP embeddings. The dataset was filtered to remove low-quality cells and potential doublets, and QC metrics were visualized through histograms, violin plots, and scatter plots. The processed dataset was saved as an AnnData object for downstream analysis."
        ],
        [
         "4",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#----------------\n# Package Load\n#----------------\nlibrary(Seurat) # Core package for single-cell analysis\nlibrary(dplyr) # For data manipulation and piping\nlibrary(ggplot2) # For advanced plotting customization (optional)\n\n#----------------\n# Data Input & Requirements\n#----------------\n# Requirements: \n# - 10X Genomics data in filtered_gene_bc_matrices format\n# - Directory path must point to uncompressed matrix files (matrix.mtx, genes.tsv, barcodes.tsv)\n# - Mitochondrial genes prefixed with \"MT-\" (human genome hg19 convention)\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\" # SET YOUR DATA PATH HERE\npbmc.data <- Read10X(data.dir = data_dir) # Load raw count matrix\n\n#----------------\n# Major Analysis Tasks\n#----------------\n# Initialize Seurat object with quality filters:\n# - min.cells=3: Genes must appear in >3 cells\n# - min.features=200: Cells must express >200 genes\npbmc <- CreateSeuratObject(\ncounts = pbmc.data, \nproject = \"mc3k\", \nmin.cells = 3, \nmin.features = 200\n)\n\n# Calculate mitochondrial QC metrics:\n# - Identifies genes starting with \"MT-\" prefix\n# - Computes percentage of mitochondrial reads per cell\npbmc[[\"percent.mt\"]] <- PercentageFeatureSet(pbmc, pattern = \"^MT-\")\n\n# Apply QC filtering thresholds:\n# - Retain cells with 200-2500 detected genes (nFeature_RNA)\n# - Exclude cells with >5% mitochondrial reads (percent.mt)\npbmc <- subset(\npbmc, \nsubset = nFeature_RNA > 200 & \nnFeature_RNA < 2500 & \npercent.mt < 5\n)\n\n#----------------\n# Plotting\n#----------------\n# Violin plots for key QC metrics:\n# - nFeature_RNA: Genes per cell\n# - nCount_RNA: UMI counts per cell\n# - percent.mt: Mitochondrial percentage\nVlnPlot(\npbmc, \nfeatures = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), \nncol = 3\n)\n\n# Scatter plots for QC relationships:\n# - Plot 1: UMI counts vs. mitochondrial percentage\n# - Plot 2: UMI counts vs. detected genes\nplot1 <- FeatureScatter(pbmc, feature1 = \"nCount_RNA\", feature2 = \"percent.mt\")\nplot2 <- FeatureScatter(pbmc, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")\nplot1 + plot2 # Combine plots using patchwork\n\n#----------------\n# Data Output\n#----------------\n# Save QC-processed object in RDS format:\n# - Preserves all data, metadata, and calculations\n# - Compatible with Seurat::LoadRDS() for reloading\nsaveRDS(pbmc, file = \"pbmc_QC_filtered.rds\"))(",
         "The analysis begins by loading essential R packages for single-cell RNA sequencing data processing and visualization. The Seurat package is used as the core tool for single-cell analysis, providing functions for data normalization, quality control, and downstream analysis. The dplyr package is employed for data manipulation and piping operations, while ggplot2 is utilized for advanced plotting customization, though it is optional.\n\nThe input data consists of 10X Genomics single-cell RNA sequencing data in the filtered_gene_bc_matrices format. The directory path must point to uncompressed matrix files, including matrix.mtx, genes.tsv, and barcodes.tsv. Additionally, mitochondrial genes are identified using the \"MT-\" prefix, following the human genome hg19 convention. The raw count matrix is loaded using the Read10X function, which reads the data from the specified directory.\n\nA Seurat object is initialized with quality filters to ensure data integrity. The CreateSeuratObject function is used with parameters set to include genes detected in more than three cells (min.cells = 3) and cells expressing more than 200 genes (min.features = 200). This step ensures that only high-quality data is retained for further analysis.\n\nMitochondrial quality control metrics are calculated to assess the percentage of mitochondrial reads per cell. The PercentageFeatureSet function identifies genes with the \"MT-\" prefix and computes the proportion of mitochondrial reads. This metric is crucial for identifying cells with potential stress or damage.\n\nQuality control filtering is applied to retain cells with 200 to 2500 detected genes (nFeature_RNA) and exclude cells with more than 5% mitochondrial reads (percent.mt). The subset function is used to filter the Seurat object based on these criteria, ensuring that only high-quality cells are included in the analysis.\n\nVisualization of key quality control metrics is performed using violin plots. The VlnPlot function generates plots for the number of genes per cell (nFeature_RNA), UMI counts per cell (nCount_RNA), and mitochondrial percentage (percent.mt). These plots provide an overview of the data distribution and help identify outliers.\n\nScatter plots are created to explore relationships between quality control metrics. The FeatureScatter function generates plots comparing UMI counts to mitochondrial percentage and UMI counts to the number of detected genes. These plots are combined using the patchwork package to facilitate comparison and interpretation.\n\nFinally, the quality control-processed Seurat object is saved in RDS format using the saveRDS function. This preserves all data, metadata, and calculations, allowing for easy reloading and further analysis using the Seurat::LoadRDS function. The saved file, named \"pbmc_QC_filtered.rds,\" serves as the output of the quality control pipeline.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by loading essential R packages for single-cell RNA sequencing data processing and visualization. The Seurat package is used as the core tool for single-cell analysis, providing functions for data normalization, quality control, and downstream analysis. The dplyr package is employed for data manipulation and piping operations, while ggplot2 is utilized for advanced plotting customization, though it is optional.\n\nThe input data consists of 10X Genomics single-cell RNA sequencing data in the filtered_gene_bc_matrices format. The directory path must point to uncompressed matrix files, including matrix.mtx, genes.tsv, and barcodes.tsv. Additionally, mitochondrial genes are identified using the \"MT-\" prefix, following the human genome hg19 convention. The raw count matrix is loaded using the Read10X function, which reads the data from the specified directory.\n\nA Seurat object is initialized with quality filters to ensure data integrity. The CreateSeuratObject function is used with parameters set to include genes detected in more than three cells (min.cells = 3) and cells expressing more than 200 genes (min.features = 200). This step ensures that only high-quality data is retained for further analysis.\n\nMitochondrial quality control metrics are calculated to assess the percentage of mitochondrial reads per cell. The PercentageFeatureSet function identifies genes with the \"MT-\" prefix and computes the proportion of mitochondrial reads. This metric is crucial for identifying cells with potential stress or damage.\n\nQuality control filtering is applied to retain cells with 200 to 2500 detected genes (nFeature_RNA) and exclude cells with more than 5% mitochondrial reads (percent.mt). The subset function is used to filter the Seurat object based on these criteria, ensuring that only high-quality cells are included in the analysis.\n\nVisualization of key quality control metrics is performed using violin plots. The VlnPlot function generates plots for the number of genes per cell (nFeature_RNA), UMI counts per cell (nCount_RNA), and mitochondrial percentage (percent.mt). These plots provide an overview of the data distribution and help identify outliers.\n\nScatter plots are created to explore relationships between quality control metrics. The FeatureScatter function generates plots comparing UMI counts to mitochondrial percentage and UMI counts to the number of detected genes. These plots are combined using the patchwork package to facilitate comparison and interpretation.\n\nFinally, the quality control-processed Seurat object is saved in RDS format using the saveRDS function. This preserves all data, metadata, and calculations, allowing for easy reloading and further analysis using the Seurat::LoadRDS function. The saved file, named \"pbmc_QC_filtered.rds,\" serves as the output of the quality control pipeline.",
         "The analysis was conducted in R using the Seurat package for single-cell RNA sequencing data processing and visualization, supplemented by dplyr for data manipulation and ggplot2 for advanced plotting. The workflow began with loading 10X Genomics data in the filtered_gene_bc_matrices format, identifying mitochondrial genes via the \"MT-\" prefix. A Seurat object was initialized with quality filters, retaining genes detected in at least three cells and cells expressing over 200 genes. Mitochondrial read percentages were calculated to assess cell quality, and cells with 200–2500 detected genes and <5% mitochondrial reads were retained. Quality control metrics were visualized using violin and scatter plots to identify outliers and assess data distributions. The processed Seurat object was saved for downstream analysis."
        ],
        [
         "5",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example peripheral blood mononuclear cell (PBMC) dataset\n\n# This is a common benchmark dataset in single-cell analysis.\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Data Normalization and Preprocessing\n\n#--------------------------------\n\n\n\n# Save the raw, unnormalized count data into a separate layer.\n\n# This is a good practice as it preserves the original data for future reference\n\n# or for algorithms that require raw counts.\n\nadata.layers[\"counts\"] = adata.X.copy()\n\n\n\n# Normalize the data for sequencing depth.\n\n# Each cell's total counts are scaled to a target sum (e.g., 10,000).\n\n# This removes technical differences in library size across cells.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to the normalized data.\n\n# This helps to stabilize the variance and makes the data more suitable for\n\n# linear models used in downstream analysis (like PCA or differential expression).\n\nsc.pp.log1p(adata)\n\n\n\n# For verification, print a small slice of the transformed data matrix.\n\nprint(\"Normalized data sample (first 5 cells, first 5 genes):\")\n\nprint(adata.X[:5, :5])\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create violin plots to visualize the distributions of total counts and\n\n# the number of detected genes per cell after normalization. This helps to\n\n# confirm that the normalization has been applied correctly.\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"total_counts\", \"n_genes_by_counts\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\ntitle=\"QC Metrics after Normalization\"\n\n)\n\n\n\n# Create a scatter plot to examine the relationship between the total counts and\n\n# the number of genes detected in each cell. This can help identify if there\n\n# are any remaining technical artifacts or biases in the data.\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ntitle=\"Total Counts vs Number of Genes after Normalization\"\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the normalized and processed AnnData object to a file.\n\n# This file can be loaded for subsequent analysis steps, such as\n\n# identifying highly variable genes, dimensionality reduction, and clustering.\n\nadata.write(\"pbmc3k_normalized.h5ad\")\n\n",
         "The following is a step-by-step guide to the analysis performed in the provided Python code, which focuses on single-cell RNA sequencing data processing and visualization using the Scanpy and AnnData libraries.\n\nFirst, the necessary modules for single-cell analysis and data handling are imported. The Scanpy library is used for single-cell data analysis, while AnnData provides a data structure for storing and manipulating single-cell datasets. These libraries are essential for performing the subsequent steps in the analysis.\n\nNext, global parameters for plotting are configured using Scanpy. The settings ensure that visualizations are consistent and clear by specifying a resolution of 80 DPI and a white background for figures. This step is important for producing high-quality plots that accurately represent the data.\n\nThe analysis begins with the loading of an example peripheral blood mononuclear cell (PBMC) dataset, which is a common benchmark dataset in single-cell RNA sequencing studies. This dataset is stored in an AnnData object, which is the primary data structure used throughout the analysis.\n\nThe raw, unnormalized count data is then preserved by copying it into a separate layer within the AnnData object. This is a critical step, as it ensures that the original data remains accessible for future reference or for algorithms that require raw counts. Normalization is performed to account for differences in sequencing depth across cells. Each cell's total counts are scaled to a target sum of 10,000, which removes technical variations in library size. Following normalization, a log(1+x) transformation is applied to the data. This transformation stabilizes the variance and makes the data more suitable for downstream analyses, such as principal component analysis (PCA) or differential expression. A small sample of the transformed data matrix is printed to verify the normalization and transformation steps.\n\nTo assess the quality of the normalization, violin plots are generated to visualize the distributions of total counts and the number of detected genes per cell. These plots help confirm that the normalization has been applied correctly and that the data is ready for further analysis. Additionally, a scatter plot is created to examine the relationship between total counts and the number of genes detected in each cell. This plot can reveal any remaining technical artifacts or biases in the data.\n\nFinally, the normalized and processed AnnData object is saved to a file in H5AD format. This file can be loaded in subsequent analysis steps, such as identifying highly variable genes, performing dimensionality reduction, and clustering. Saving the processed data ensures that the results of this analysis are preserved and can be easily accessed for further exploration.\n\nThis guide provides a comprehensive overview of the steps involved in processing and visualizing single-cell RNA sequencing data using Scanpy and AnnData. Each step is explained in detail, highlighting its purpose, rationale, and key parameters, to enable replication of the analysis.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following is a step-by-step guide to the analysis performed in the provided Python code, which focuses on single-cell RNA sequencing data processing and visualization using the Scanpy and AnnData libraries.\n\nFirst, the necessary modules for single-cell analysis and data handling are imported. The Scanpy library is used for single-cell data analysis, while AnnData provides a data structure for storing and manipulating single-cell datasets. These libraries are essential for performing the subsequent steps in the analysis.\n\nNext, global parameters for plotting are configured using Scanpy. The settings ensure that visualizations are consistent and clear by specifying a resolution of 80 DPI and a white background for figures. This step is important for producing high-quality plots that accurately represent the data.\n\nThe analysis begins with the loading of an example peripheral blood mononuclear cell (PBMC) dataset, which is a common benchmark dataset in single-cell RNA sequencing studies. This dataset is stored in an AnnData object, which is the primary data structure used throughout the analysis.\n\nThe raw, unnormalized count data is then preserved by copying it into a separate layer within the AnnData object. This is a critical step, as it ensures that the original data remains accessible for future reference or for algorithms that require raw counts. Normalization is performed to account for differences in sequencing depth across cells. Each cell's total counts are scaled to a target sum of 10,000, which removes technical variations in library size. Following normalization, a log(1+x) transformation is applied to the data. This transformation stabilizes the variance and makes the data more suitable for downstream analyses, such as principal component analysis (PCA) or differential expression. A small sample of the transformed data matrix is printed to verify the normalization and transformation steps.\n\nTo assess the quality of the normalization, violin plots are generated to visualize the distributions of total counts and the number of detected genes per cell. These plots help confirm that the normalization has been applied correctly and that the data is ready for further analysis. Additionally, a scatter plot is created to examine the relationship between total counts and the number of genes detected in each cell. This plot can reveal any remaining technical artifacts or biases in the data.\n\nFinally, the normalized and processed AnnData object is saved to a file in H5AD format. This file can be loaded in subsequent analysis steps, such as identifying highly variable genes, performing dimensionality reduction, and clustering. Saving the processed data ensures that the results of this analysis are preserved and can be easily accessed for further exploration.\n\nThis guide provides a comprehensive overview of the steps involved in processing and visualizing single-cell RNA sequencing data using Scanpy and AnnData. Each step is explained in detail, highlighting its purpose, rationale, and key parameters, to enable replication of the analysis.",
         "The analysis was conducted in Python using the Scanpy and AnnData libraries for single-cell RNA sequencing data processing and visualization. A peripheral blood mononuclear cell (PBMC) dataset was loaded into an AnnData object, with raw counts preserved in a separate layer for downstream reference. Normalization was performed by scaling total counts to a target sum of 10,000, followed by a log(1+x) transformation to stabilize variance. Quality control was assessed through violin plots of total counts and detected genes per cell, alongside scatter plots to evaluate relationships between these metrics. The processed data was saved in H5AD format for subsequent analysis steps."
        ],
        [
         "6",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting\n\n# This ensures all plots have a consistent appearance.\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example peripheral blood mononuclear cell (PBMC) dataset\n\n# This is a common benchmark dataset in single-cell analysis.\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Data Normalization and Preprocessing\n\n#--------------------------------\n\n\n\n# Save the raw, unnormalized count data into a separate layer.\n\n# This is a good practice as it preserves the original data for future reference\n\n# or for algorithms that require raw counts.\n\nadata.layers[\"counts\"] = adata.X.copy()\n\n\n\n# Normalize the data for sequencing depth.\n\n# Each cell's total counts are scaled to a target sum (e.g., 10,000).\n\n# This removes technical differences in library size across cells.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to the normalized data.\n\n# This helps to stabilize the variance and makes the data more suitable for\n\n# linear models used in downstream analysis (like PCA or differential expression).\n\nsc.pp.log1p(adata)\n\n\n\n# For verification, print a small slice of the transformed data matrix.\n\nprint(\"Normalized data sample (first 5 cells, first 5 genes):\")\n\nprint(adata.X[:5, :5])\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create violin plots to visualize the distributions of total counts and\n\n# the number of detected genes per cell after normalization. This helps to\n\n# confirm that the normalization has been applied correctly.\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"total_counts\", \"n_genes_by_counts\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\ntitle=\"QC Metrics after Normalization\"\n\n)\n\n\n\n# Create a scatter plot to examine the relationship between the total counts and\n\n# the number of genes detected in each cell. This can help identify if there\n\n# are any remaining technical artifacts or biases in the data.\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ntitle=\"Total Counts vs Number of Genes after Normalization\"\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the normalized and processed AnnData object to a file.\n\n# This file can be loaded for subsequent analysis steps, such as\n\n# identifying highly variable genes, dimensionality reduction, and clustering.\n\nadata.write(\"pbmc3k_normalized.h5ad\")",
         "The analysis begins by importing essential modules for single-cell analysis and data handling. The `scanpy` library is used for single-cell RNA-seq data analysis, while `anndata` provides a data structure for storing and manipulating single-cell datasets. These tools are critical for performing the subsequent steps in the analysis.\n\nTo ensure consistency in the appearance of all plots generated during the analysis, global parameters for plotting are configured using `scanpy`. Specifically, the figure resolution (dpi) is set to 80, and the background color is set to white. This step ensures that all visualizations have a uniform and professional appearance.\n\nThe analysis utilizes a publicly available peripheral blood mononuclear cell (PBMC) dataset, which serves as a benchmark in single-cell RNA-seq studies. This dataset is loaded into an `AnnData` object, a specialized data structure for single-cell data. The dataset includes raw gene expression counts for each cell, which are stored in the `X` attribute of the `AnnData` object.\n\nTo preserve the original raw counts for downstream analysis or algorithms that require unnormalized data, the raw counts are copied into a separate layer within the `AnnData` object. This step ensures that the original data remains accessible even after normalization. Normalization is then performed to account for differences in sequencing depth across cells. Each cell’s total counts are scaled to a target sum of 10,000, effectively removing technical variations in library size. Following normalization, a log(1+x) transformation is applied to the data. This transformation stabilizes the variance and makes the data more suitable for linear models used in downstream analyses, such as principal component analysis (PCA) or differential expression testing. A small sample of the transformed data matrix is printed to verify the normalization and transformation steps.\n\nTo assess the quality of the normalization, violin plots are generated to visualize the distributions of total counts and the number of detected genes per cell. These plots help confirm that the normalization has been applied correctly and that the data is ready for further analysis. Additionally, a scatter plot is created to examine the relationship between total counts and the number of genes detected in each cell. This visualization helps identify any remaining technical artifacts or biases in the data.\n\nFinally, the normalized and processed `AnnData` object is saved to a file in H5AD format. This file can be loaded in subsequent analysis steps, such as identifying highly variable genes, performing dimensionality reduction, and clustering. Saving the processed data ensures that the results of this analysis can be easily accessed and used for further investigation.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential modules for single-cell analysis and data handling. The `scanpy` library is used for single-cell RNA-seq data analysis, while `anndata` provides a data structure for storing and manipulating single-cell datasets. These tools are critical for performing the subsequent steps in the analysis.\n\nTo ensure consistency in the appearance of all plots generated during the analysis, global parameters for plotting are configured using `scanpy`. Specifically, the figure resolution (dpi) is set to 80, and the background color is set to white. This step ensures that all visualizations have a uniform and professional appearance.\n\nThe analysis utilizes a publicly available peripheral blood mononuclear cell (PBMC) dataset, which serves as a benchmark in single-cell RNA-seq studies. This dataset is loaded into an `AnnData` object, a specialized data structure for single-cell data. The dataset includes raw gene expression counts for each cell, which are stored in the `X` attribute of the `AnnData` object.\n\nTo preserve the original raw counts for downstream analysis or algorithms that require unnormalized data, the raw counts are copied into a separate layer within the `AnnData` object. This step ensures that the original data remains accessible even after normalization. Normalization is then performed to account for differences in sequencing depth across cells. Each cell’s total counts are scaled to a target sum of 10,000, effectively removing technical variations in library size. Following normalization, a log(1+x) transformation is applied to the data. This transformation stabilizes the variance and makes the data more suitable for linear models used in downstream analyses, such as principal component analysis (PCA) or differential expression testing. A small sample of the transformed data matrix is printed to verify the normalization and transformation steps.\n\nTo assess the quality of the normalization, violin plots are generated to visualize the distributions of total counts and the number of detected genes per cell. These plots help confirm that the normalization has been applied correctly and that the data is ready for further analysis. Additionally, a scatter plot is created to examine the relationship between total counts and the number of genes detected in each cell. This visualization helps identify any remaining technical artifacts or biases in the data.\n\nFinally, the normalized and processed `AnnData` object is saved to a file in H5AD format. This file can be loaded in subsequent analysis steps, such as identifying highly variable genes, performing dimensionality reduction, and clustering. Saving the processed data ensures that the results of this analysis can be easily accessed and used for further investigation.",
         "The analysis was conducted using Python with the Scanpy and AnnData libraries for single-cell RNA-seq data processing and visualization. A publicly available PBMC dataset was loaded into an AnnData object, preserving raw counts in a separate layer for downstream use. Normalization was performed by scaling total counts to a target sum of 10,000, followed by log(1+x) transformation to stabilize variance. Quality control included generating violin and scatter plots to assess distributions of total counts and detected genes per cell, ensuring normalization efficacy. The processed data was saved in H5AD format for subsequent steps, such as feature selection and clustering."
        ],
        [
         "7",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Import necessary libraries for single-cell analysis and data manipulation\n\nlibrary(Seurat) # Main package for single-cell analysis\n\nlibrary(dplyr) # For data manipulation functions\n\nlibrary(ggplot2) # For advanced plotting capabilities\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix.\n\n# Basic filtering is applied at this stage:\n\n# - min.cells = 3: Keep genes that are expressed in at least 3 cells.\n\n# - min.features = 200: Keep cells that have at least 200 detected genes.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"DataNormalization\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Data Normalization\n\n#--------------------------------\n\n\n\n# Perform LogNormalization on the data. This is a standard method that:\n\n# 1. Normalizes gene counts for each cell by the total counts for that cell.\n\n# 2. Multiplies the result by a scaling factor (default is 10,000).\n\n# 3. Applies a natural-log transformation (log1p).\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a violin plot to visualize the distribution of RNA counts per cell.\n\n# This plot helps to assess the effect of normalization on the data distribution\n\n# across all cells. 'nCount_RNA' is automatically calculated by Seurat.\n\nVlnPlot(seurat_obj, features = \"nCount_RNA\", pt.size = 0.1) +\n\nggtitle(\"Distribution of Normalized RNA Counts\")\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the normalized Seurat object to an RDS file.\n\n# This file can be loaded later for downstream analysis, such as scaling,\n\n# dimensionality reduction, and clustering.\n\nsaveRDS(seurat_obj, file = \"Normalized_Data_Object.rds\")",
         "The analysis begins by loading the necessary libraries for single-cell RNA sequencing data analysis and visualization. The Seurat package is imported as the primary tool for single-cell analysis, providing functions for data normalization, dimensionality reduction, and clustering. The dplyr package is used for efficient data manipulation, and ggplot2 is employed for advanced plotting capabilities.\n\nThe input data consists of raw count matrices generated from 10X Genomics sequencing. The directory containing these matrices is specified, and the data is read into a raw count matrix using the `Read10X` function. A Seurat object is then created from this raw count matrix. During this step, basic quality control filtering is applied: genes detected in fewer than 3 cells are excluded, and cells with fewer than 200 detected genes are removed. These thresholds ensure that only reliable genes and cells are retained for downstream analysis.\n\nNext, the data is normalized using the `NormalizeData` function with the LogNormalize method. This normalization process adjusts gene expression counts for each cell by dividing them by the total counts for that cell, scaling the result by a factor of 10,000, and applying a natural-log transformation (log1p). This step is critical for reducing technical variability and enabling meaningful comparisons of gene expression across cells.\n\nTo assess the effect of normalization, a violin plot is generated to visualize the distribution of RNA counts per cell. The `VlnPlot` function is used to create this plot, with the `nCount_RNA` feature representing the total RNA counts per cell. The plot provides insights into the distribution of normalized RNA counts across all cells, helping to evaluate the success of the normalization process.\n\nFinally, the normalized Seurat object is saved to an RDS file using the `saveRDS` function. This file preserves the normalized data for subsequent analyses, such as scaling, dimensionality reduction, and clustering, ensuring reproducibility and facilitating further exploration of the single-cell dataset.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by loading the necessary libraries for single-cell RNA sequencing data analysis and visualization. The Seurat package is imported as the primary tool for single-cell analysis, providing functions for data normalization, dimensionality reduction, and clustering. The dplyr package is used for efficient data manipulation, and ggplot2 is employed for advanced plotting capabilities.\n\nThe input data consists of raw count matrices generated from 10X Genomics sequencing. The directory containing these matrices is specified, and the data is read into a raw count matrix using the `Read10X` function. A Seurat object is then created from this raw count matrix. During this step, basic quality control filtering is applied: genes detected in fewer than 3 cells are excluded, and cells with fewer than 200 detected genes are removed. These thresholds ensure that only reliable genes and cells are retained for downstream analysis.\n\nNext, the data is normalized using the `NormalizeData` function with the LogNormalize method. This normalization process adjusts gene expression counts for each cell by dividing them by the total counts for that cell, scaling the result by a factor of 10,000, and applying a natural-log transformation (log1p). This step is critical for reducing technical variability and enabling meaningful comparisons of gene expression across cells.\n\nTo assess the effect of normalization, a violin plot is generated to visualize the distribution of RNA counts per cell. The `VlnPlot` function is used to create this plot, with the `nCount_RNA` feature representing the total RNA counts per cell. The plot provides insights into the distribution of normalized RNA counts across all cells, helping to evaluate the success of the normalization process.\n\nFinally, the normalized Seurat object is saved to an RDS file using the `saveRDS` function. This file preserves the normalized data for subsequent analyses, such as scaling, dimensionality reduction, and clustering, ensuring reproducibility and facilitating further exploration of the single-cell dataset.",
         "The analysis was conducted in R using the Seurat package for single-cell RNA sequencing data processing, normalization, and visualization. Raw count matrices from 10X Genomics sequencing were loaded and filtered to exclude genes detected in fewer than 3 cells and cells with fewer than 200 detected genes, ensuring data quality. Normalization was performed using the LogNormalize method, scaling gene expression counts by total cellular counts and applying a log transformation to reduce technical variability. The distribution of normalized RNA counts was visualized using violin plots to assess normalization efficacy. This workflow provides a robust foundation for downstream analyses, including dimensionality reduction and clustering."
        ],
        [
         "8",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting\n\n# This ensures all plots have a consistent, clean appearance.\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example peripheral blood mononuclear cell (PBMC) dataset\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# It's good practice to keep a copy of the raw data.\n\n# The .raw attribute is the conventional place to store it in Scanpy.\n\nadata.raw = adata.copy()\n\n\n\n\n\n#--------------------------------\n\n# Data Normalization\n\n#--------------------------------\n\n\n\n# Normalize the data for sequencing depth by scaling total counts per cell to a target sum.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to the normalized data.\n\n# This stabilizes variance and makes the data more comparable across genes with different expression levels.\n\nsc.pp.log1p(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Feature Selection\n\n#--------------------------------\n\n\n\n# Identify highly variable genes (HVGs). These are genes that show high cell-to-cell\n\n# variation and are more likely to be biologically significant.\n\n# The 'seurat' flavor is a commonly used method for this.\n\n# `n_top_genes` specifies how many of the most variable genes to select.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# Print the number of genes identified as highly variable for confirmation.\n\nprint(f\"Number of highly variable genes found: {sum(adata.var.highly_variable)}\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a scatter plot to visualize the highly variable genes.\n\n# The plot shows each gene's mean expression versus its dispersion (a measure of variability).\n\n# The selected HVGs are highlighted, which is useful for a visual quality check of the selection process.\n\nsc.pl.highly_variable_genes(adata)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the normalized data and the list of\n\n# highly variable genes. This object is ready for downstream steps like scaling and\n\n# dimensionality reduction (PCA).\n\nadata.write(\"pbmc3k_hvg_selected.h5ad\")\n\n",
         "The analysis begins by importing the necessary modules for single-cell analysis and data handling. The `scanpy` library is used for single-cell RNA-seq data analysis, while `anndata` provides a data structure for storing and manipulating the dataset. These tools are essential for performing the subsequent steps in the pipeline.\n\nTo ensure consistency in the visual appearance of plots, global parameters for plotting are configured using `scanpy`. Specifically, the resolution (dpi) is set to 80, and the background color of plots is set to white. This step ensures that all visualizations generated during the analysis have a clean and uniform appearance.\n\nThe dataset used in this analysis is an example peripheral blood mononuclear cell (PBMC) dataset, which is loaded using the `scanpy` function `datasets.pbmc3k()`. This dataset is stored in an `AnnData` object, a standard format for single-cell data. To preserve the raw data for future reference, a copy of the dataset is stored in the `.raw` attribute of the `AnnData` object. This practice is important for maintaining the integrity of the original data while performing downstream transformations.\n\nNext, the data is normalized to account for differences in sequencing depth across cells. This is achieved by scaling the total counts per cell to a target sum of 10,000 using the `normalize_total` function. Following normalization, a log(1+x) transformation is applied to the data. This transformation stabilizes the variance across genes with different expression levels, making the data more comparable and suitable for downstream analysis.\n\nThe primary analysis task involves identifying highly variable genes (HVGs), which are genes that exhibit significant cell-to-cell variation and are more likely to be biologically relevant. The `highly_variable_genes` function is used for this purpose, with the `seurat` flavor selected as the method for identifying HVGs. The parameter `n_top_genes` is set to 2000, specifying the number of most variable genes to retain. After identifying HVGs, the number of genes selected is printed for confirmation, ensuring the process has been executed correctly.\n\nTo visualize the results of the HVG selection, a scatter plot is generated using the `highly_variable_genes` plotting function. This plot displays each gene’s mean expression against its dispersion, with the selected HVGs highlighted. This visualization serves as a quality check, allowing for a quick assessment of the gene selection process.\n\nFinally, the `AnnData` object, which now contains the normalized data and the list of highly variable genes, is saved to a file named `pbmc3k_hvg_selected.h5ad`. This file is ready for subsequent steps in the analysis pipeline, such as scaling and dimensionality reduction (e.g., PCA). Saving the processed data ensures that the results can be easily loaded and used in future analyses.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing the necessary modules for single-cell analysis and data handling. The `scanpy` library is used for single-cell RNA-seq data analysis, while `anndata` provides a data structure for storing and manipulating the dataset. These tools are essential for performing the subsequent steps in the pipeline.\n\nTo ensure consistency in the visual appearance of plots, global parameters for plotting are configured using `scanpy`. Specifically, the resolution (dpi) is set to 80, and the background color of plots is set to white. This step ensures that all visualizations generated during the analysis have a clean and uniform appearance.\n\nThe dataset used in this analysis is an example peripheral blood mononuclear cell (PBMC) dataset, which is loaded using the `scanpy` function `datasets.pbmc3k()`. This dataset is stored in an `AnnData` object, a standard format for single-cell data. To preserve the raw data for future reference, a copy of the dataset is stored in the `.raw` attribute of the `AnnData` object. This practice is important for maintaining the integrity of the original data while performing downstream transformations.\n\nNext, the data is normalized to account for differences in sequencing depth across cells. This is achieved by scaling the total counts per cell to a target sum of 10,000 using the `normalize_total` function. Following normalization, a log(1+x) transformation is applied to the data. This transformation stabilizes the variance across genes with different expression levels, making the data more comparable and suitable for downstream analysis.\n\nThe primary analysis task involves identifying highly variable genes (HVGs), which are genes that exhibit significant cell-to-cell variation and are more likely to be biologically relevant. The `highly_variable_genes` function is used for this purpose, with the `seurat` flavor selected as the method for identifying HVGs. The parameter `n_top_genes` is set to 2000, specifying the number of most variable genes to retain. After identifying HVGs, the number of genes selected is printed for confirmation, ensuring the process has been executed correctly.\n\nTo visualize the results of the HVG selection, a scatter plot is generated using the `highly_variable_genes` plotting function. This plot displays each gene’s mean expression against its dispersion, with the selected HVGs highlighted. This visualization serves as a quality check, allowing for a quick assessment of the gene selection process.\n\nFinally, the `AnnData` object, which now contains the normalized data and the list of highly variable genes, is saved to a file named `pbmc3k_hvg_selected.h5ad`. This file is ready for subsequent steps in the analysis pipeline, such as scaling and dimensionality reduction (e.g., PCA). Saving the processed data ensures that the results can be easily loaded and used in future analyses.",
         "The analysis was conducted using Python with the `scanpy` library for single-cell RNA-seq data processing and the `anndata` data structure for dataset management. The workflow began with the normalization of raw counts to account for sequencing depth variations, followed by a log(1+x) transformation to stabilize gene expression variance. Highly variable genes (HVGs) were identified using the `seurat` flavor of the `highly_variable_genes` function, retaining the top 2,000 genes for downstream analysis. The results were visualized through a scatter plot comparing mean expression and dispersion of genes, highlighting selected HVGs. The processed data, including normalized counts and HVGs, was stored in an `AnnData` object for subsequent analysis steps."
        ],
        [
         "9",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Import necessary libraries for single-cell analysis and data manipulation\n\nlibrary(Seurat) # Main package for single-cell analysis\n\nlibrary(dplyr) # For data manipulation functions\n\nlibrary(ggplot2) # For advanced plotting capabilities\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"FeatureSelectionDemo\",\n\nmin.cells = 3, # Keep genes expressed in at least 3 cells\n\nmin.features = 200 # Keep cells with at least 200 detected genes\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing and Normalization\n\n#--------------------------------\n\n\n\n# OPTIONAL QC: Calculate the percentage of mitochondrial genes for each cell.\n\n# This is a common QC metric; high percentages can indicate cell stress.\n\nseurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(seurat_obj, pattern = \"^MT-\")\n\n\n\n# Normalize the data using the \"LogNormalize\" method.\n\n# This scales the data by library size and applies a log transformation.\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Feature Selection\n\n#--------------------------------\n\n\n\n# Identify highly variable features (genes) using the variance-stabilizing transformation (vst) method.\n\n# These genes show high cell-to-cell variation and are used for downstream analyses like PCA.\n\nseurat_obj <- FindVariableFeatures(\n\nseurat_obj,\n\nselection.method = \"vst\",\n\nnfeatures = 2000 # Select the top 2000 most variable genes\n\n)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Identify the top 10 most highly variable genes for labeling purposes.\n\ntop10 <- head(VariableFeatures(seurat_obj), 10)\n\n\n\n# Create a variable feature plot to visualize gene variability.\n\n# The plot shows the relationship between average expression and a measure of dispersion for each gene.\n\nvariable_feature_plot <- VariableFeaturePlot(seurat_obj)\n\n\n\n# Add labels for the top 10 HVGs to the plot for better interpretation.\n\nlabeled_plot <- LabelPoints(plot = variable_feature_plot, points = top10, repel = TRUE)\n\n\n\n# Print the final plot to the console.\n\nprint(labeled_plot)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the Seurat object, which now includes the identified variable features.\n\n# This object can be used for the next steps in the analysis, such as scaling and clustering.\n\nsaveRDS(seurat_obj, file = \"FeatureSelection_SeuratObject.rds\")",
         "To perform single-cell RNA sequencing analysis, the following steps were executed using the Seurat package in R. The analysis focused on preprocessing, feature selection, and visualization of highly variable genes.\n\nFirst, the necessary libraries were loaded. The Seurat package was used for single-cell analysis, dplyr for data manipulation, and ggplot2 for advanced plotting capabilities. These libraries are essential for handling and analyzing single-cell data effectively.\n\nThe 10X Genomics count matrices were read into a raw count matrix using the `Read10X` function. The directory containing the data was specified, and the raw data was imported. A Seurat object was then created from the raw count matrix using the `CreateSeuratObject` function. Initial filtering was applied to retain genes expressed in at least three cells (`min.cells = 3`) and cells with at least 200 detected genes (`min.features = 200`). This step ensures that only high-quality data is included in the analysis.\n\nNext, preprocessing and normalization were performed. The percentage of mitochondrial genes was calculated for each cell using the `PercentageFeatureSet` function. This metric is commonly used for quality control, as high percentages can indicate cell stress. The data was normalized using the \"LogNormalize\" method, which scales the data by library size and applies a log transformation. The `NormalizeData` function was used with a scale factor of 10,000 to achieve this normalization.\n\nFor feature selection, highly variable genes were identified using the variance-stabilizing transformation (vst) method. The `FindVariableFeatures` function was employed to select the top 2,000 most variable genes (`nfeatures = 2000`). These genes exhibit high cell-to-cell variation and are crucial for downstream analyses such as principal component analysis (PCA).\n\nVisualization of the highly variable genes was performed to interpret the results. The top 10 most highly variable genes were identified for labeling purposes. A variable feature plot was generated using the `VariableFeaturePlot` function to visualize the relationship between average expression and a measure of dispersion for each gene. Labels for the top 10 highly variable genes were added to the plot using the `LabelPoints` function, and the final plot was printed to the console.\n\nFinally, the Seurat object, which now includes the identified variable features, was saved using the `saveRDS` function. This object can be used for subsequent steps in the analysis, such as scaling and clustering. The saved file is named \"FeatureSelection_SeuratObject.rds\".",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform single-cell RNA sequencing analysis, the following steps were executed using the Seurat package in R. The analysis focused on preprocessing, feature selection, and visualization of highly variable genes.\n\nFirst, the necessary libraries were loaded. The Seurat package was used for single-cell analysis, dplyr for data manipulation, and ggplot2 for advanced plotting capabilities. These libraries are essential for handling and analyzing single-cell data effectively.\n\nThe 10X Genomics count matrices were read into a raw count matrix using the `Read10X` function. The directory containing the data was specified, and the raw data was imported. A Seurat object was then created from the raw count matrix using the `CreateSeuratObject` function. Initial filtering was applied to retain genes expressed in at least three cells (`min.cells = 3`) and cells with at least 200 detected genes (`min.features = 200`). This step ensures that only high-quality data is included in the analysis.\n\nNext, preprocessing and normalization were performed. The percentage of mitochondrial genes was calculated for each cell using the `PercentageFeatureSet` function. This metric is commonly used for quality control, as high percentages can indicate cell stress. The data was normalized using the \"LogNormalize\" method, which scales the data by library size and applies a log transformation. The `NormalizeData` function was used with a scale factor of 10,000 to achieve this normalization.\n\nFor feature selection, highly variable genes were identified using the variance-stabilizing transformation (vst) method. The `FindVariableFeatures` function was employed to select the top 2,000 most variable genes (`nfeatures = 2000`). These genes exhibit high cell-to-cell variation and are crucial for downstream analyses such as principal component analysis (PCA).\n\nVisualization of the highly variable genes was performed to interpret the results. The top 10 most highly variable genes were identified for labeling purposes. A variable feature plot was generated using the `VariableFeaturePlot` function to visualize the relationship between average expression and a measure of dispersion for each gene. Labels for the top 10 highly variable genes were added to the plot using the `LabelPoints` function, and the final plot was printed to the console.\n\nFinally, the Seurat object, which now includes the identified variable features, was saved using the `saveRDS` function. This object can be used for subsequent steps in the analysis, such as scaling and clustering. The saved file is named \"FeatureSelection_SeuratObject.rds\".",
         "Single-cell RNA sequencing analysis was conducted using the Seurat package in R, focusing on preprocessing, feature selection, and visualization of highly variable genes. The workflow began with the creation of a Seurat object from 10X Genomics count matrices, followed by initial quality control to filter low-quality cells and genes. Normalization was performed using the \"LogNormalize\" method, scaling data by library size and applying a log transformation. Highly variable genes were identified using the variance-stabilizing transformation (vst) method, selecting the top 2,000 genes for downstream analysis. Visualization of these genes was achieved through a variable feature plot, highlighting the relationship between average expression and dispersion. The processed data was saved for subsequent analysis steps."
        ],
        [
         "10",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport scanpy as sc\n\nimport anndata as ad\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC 3k dataset from Scanpy as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization and Feature Selection ---\n\n# Normalize total counts per cell to account for differences in sequencing depth\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to stabilize the variance\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs), which are most informative for downstream analysis\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# --- Scaling and PCA ---\n\n# Subset the AnnData object to keep only the highly variable genes\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# Scale the data so that each gene has zero mean and unit variance.\n\n# This is crucial for PCA and other algorithms sensitive to feature scales.\n\nsc.pp.scale(adata)\n\n\n\n# Perform Principal Component Analysis (PCA) on the scaled data.\n\n# This reduces the dimensionality of the data while retaining most of the variation.\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# --- UMAP Embedding ---\n\n# Compute the neighborhood graph of cells. This is based on the PCA representation\n\n# and is a prerequisite for UMAP.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n\n\n\n# Run Uniform Manifold Approximation and Projection (UMAP) to generate a\n\n# non-linear 2D embedding for visualization.\n\nsc.tl.umap(adata)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the PCA variance ratio to see how much variance is captured by each principal component.\n\n# The log scale helps visualize the \"elbow\" where the explained variance plateaus.\n\nsc.pl.pca_variance_ratio(adata, log=True)\n\n\n\n# Create a scatter plot of the first two principal components.\n\n# Cells can be colored by marker gene expression to identify potential cell types.\n\nsc.pl.pca(adata, color=['CST3', 'NKG7'])\n\n\n\n# Visualize the UMAP embedding.\n\n# Coloring by marker genes helps to see how cell clusters are separated in the UMAP space.\n\nsc.pl.umap(adata, color='CST3')\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed AnnData object, which now contains PCA and UMAP coordinates.\n\n# This file is ready for downstream clustering and cell type annotation.\n\nadata.write(\"pbmc3k_pca_umap.h5ad\")\n\n",
         "To analyze single-cell RNA sequencing data, the following steps were performed using the Scanpy and AnnData libraries in Python. The analysis begins with data loading and preprocessing, followed by dimensionality reduction and visualization. The goal is to identify highly variable genes, reduce data dimensionality, and visualize cell clusters.\n\nFirst, the necessary libraries for single-cell analysis and plotting were imported. Scanpy was used for data processing and visualization, while AnnData was employed to handle the single-cell data structure. Matplotlib was used for additional plotting capabilities. Global parameters for Scanpy plotting were configured to ensure consistent and clear visuals, with a DPI of 80 and a white background.\n\nThe analysis used the built-in PBMC 3k dataset from Scanpy as an example. This dataset contains single-cell RNA sequencing data from peripheral blood mononuclear cells (PBMCs). The dataset was loaded into an AnnData object, which serves as the primary data structure for the analysis.\n\nPreprocessing began with normalization to account for differences in sequencing depth. Total counts per cell were normalized to a target sum of 10,000. A log(1+x) transformation was then applied to stabilize the variance across genes. Highly variable genes (HVGs) were identified using the Seurat flavor, with the top 2,000 genes selected based on their variability. These genes are considered the most informative for downstream analysis.\n\nThe dataset was subset to retain only the highly variable genes. The data were then scaled so that each gene had zero mean and unit variance, a step crucial for algorithms sensitive to feature scales, such as Principal Component Analysis (PCA). PCA was performed on the scaled data to reduce dimensionality while retaining most of the variation. The number of principal components was determined using the ARPACK solver.\n\nTo further reduce dimensionality and visualize the data, a neighborhood graph of cells was computed based on the PCA representation. This graph was used as input for Uniform Manifold Approximation and Projection (UMAP), which generated a non-linear 2D embedding for visualization. The UMAP embedding was computed with 10 nearest neighbors and 40 principal components.\n\nVisualization steps included plotting the PCA variance ratio on a log scale to identify the \"elbow\" point where the explained variance plateaus. Scatter plots of the first two principal components were created, with cells colored by the expression of marker genes (e.g., CST3 and NKG7) to identify potential cell types. The UMAP embedding was also visualized, with cells colored by the expression of the marker gene CST3 to assess cluster separation.\n\nFinally, the processed AnnData object, which now includes PCA and UMAP coordinates, was saved in H5AD format for downstream clustering and cell type annotation. This file is ready for further analysis, such as clustering and annotation of cell types based on gene expression patterns.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo analyze single-cell RNA sequencing data, the following steps were performed using the Scanpy and AnnData libraries in Python. The analysis begins with data loading and preprocessing, followed by dimensionality reduction and visualization. The goal is to identify highly variable genes, reduce data dimensionality, and visualize cell clusters.\n\nFirst, the necessary libraries for single-cell analysis and plotting were imported. Scanpy was used for data processing and visualization, while AnnData was employed to handle the single-cell data structure. Matplotlib was used for additional plotting capabilities. Global parameters for Scanpy plotting were configured to ensure consistent and clear visuals, with a DPI of 80 and a white background.\n\nThe analysis used the built-in PBMC 3k dataset from Scanpy as an example. This dataset contains single-cell RNA sequencing data from peripheral blood mononuclear cells (PBMCs). The dataset was loaded into an AnnData object, which serves as the primary data structure for the analysis.\n\nPreprocessing began with normalization to account for differences in sequencing depth. Total counts per cell were normalized to a target sum of 10,000. A log(1+x) transformation was then applied to stabilize the variance across genes. Highly variable genes (HVGs) were identified using the Seurat flavor, with the top 2,000 genes selected based on their variability. These genes are considered the most informative for downstream analysis.\n\nThe dataset was subset to retain only the highly variable genes. The data were then scaled so that each gene had zero mean and unit variance, a step crucial for algorithms sensitive to feature scales, such as Principal Component Analysis (PCA). PCA was performed on the scaled data to reduce dimensionality while retaining most of the variation. The number of principal components was determined using the ARPACK solver.\n\nTo further reduce dimensionality and visualize the data, a neighborhood graph of cells was computed based on the PCA representation. This graph was used as input for Uniform Manifold Approximation and Projection (UMAP), which generated a non-linear 2D embedding for visualization. The UMAP embedding was computed with 10 nearest neighbors and 40 principal components.\n\nVisualization steps included plotting the PCA variance ratio on a log scale to identify the \"elbow\" point where the explained variance plateaus. Scatter plots of the first two principal components were created, with cells colored by the expression of marker genes (e.g., CST3 and NKG7) to identify potential cell types. The UMAP embedding was also visualized, with cells colored by the expression of the marker gene CST3 to assess cluster separation.\n\nFinally, the processed AnnData object, which now includes PCA and UMAP coordinates, was saved in H5AD format for downstream clustering and cell type annotation. This file is ready for further analysis, such as clustering and annotation of cell types based on gene expression patterns.",
         "Single-cell RNA sequencing data analysis was conducted using Python with the Scanpy and AnnData libraries. The workflow included normalization of raw counts to account for sequencing depth, followed by log(1+x) transformation and identification of highly variable genes (HVGs) using the Seurat flavor. The dataset was subset to retain the top 2,000 HVGs and scaled for dimensionality reduction. Principal Component Analysis (PCA) was performed to reduce dimensionality, and a neighborhood graph was constructed for Uniform Manifold Approximation and Projection (UMAP) to generate a 2D embedding. Visualizations included PCA variance ratio plots and UMAP embeddings colored by marker gene expression, facilitating cluster identification and downstream analysis."
        ],
        [
         "11",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis and plotting\nimport scanpy as sc\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\nsc.settings.verbosity = 0\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the preprocessed AnnData object. This file should have undergone\n# prior QC, normalization, and feature selection.\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"s4d8_feature_selection.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/40016014\"\n)\n\n#--------------------------------\n# Major Analysis Tasks: Dimensionality Reduction\n#--------------------------------\n\n# Ensure the analysis uses the log-normalized data stored in a specific layer.\n# This step makes the choice of data layer explicit.\nadata.X = adata.layers[\"log1p_norm\"]\n\n# --- Principal Component Analysis (PCA) ---\n# Perform PCA on the highly variable genes to capture the main axes of variation.\nsc.pp.pca(adata, svd_solver=\"arpack\", use_highly_variable=True)\n\n# --- t-SNE & UMAP Prerequisite: Neighborhood Graph ---\n# Compute the neighborhood graph, which is required for both t-SNE and UMAP.\n# This is calculated on the PCA representation of the data.\nsc.pp.neighbors(adata, n_neighbors=15, n_pcs=40)\n\n# --- t-distributed Stochastic Neighbor Embedding (t-SNE) ---\n# Run t-SNE to generate a 2D embedding for visualization.\n# It uses the PCA representation (`use_rep=\"X_pca\"`) as input.\nsc.tl.tsne(adata, use_rep=\"X_pca\")\n\n# --- Uniform Manifold Approximation and Projection (UMAP) ---\n# Compute the UMAP embedding based on the neighborhood graph.\nsc.tl.umap(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the PCA result, coloring cells by total counts to check for technical artifacts.\nsc.pl.pca_scatter(adata, color=\"total_counts\", title=\"PCA: First 2 Components\")\n\n# Visualize the t-SNE embedding, also colored by total counts.\nsc.pl.tsne(adata, color=\"total_counts\", title=\"t-SNE Embedding\")\n\n# Visualize the UMAP embedding, colored by total counts.\nsc.pl.umap(adata, color=\"total_counts\", title=\"UMAP Embedding\")\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the final AnnData object, which now contains the results of PCA, t-SNE, and UMAP.\n# This object is ready for clustering and cell type annotation.\nadata.write(\"s4d8_dimensionality_reduction.h5ad\")\n",
         "The analysis begins by importing essential Python libraries for single-cell data analysis and visualization, including Scanpy for single-cell analysis, Seaborn for advanced plotting, and Matplotlib for generating figures. Scanpy’s global parameters are configured to suppress verbose output and set figure properties, such as resolution (dpi=80) and background color (facecolor=\"white\"), to ensure consistent and clear visualizations.\n\nThe preprocessed single-cell dataset is loaded as an AnnData object from a local file, with a backup URL provided for accessibility. This dataset is assumed to have undergone prior quality control, normalization, and feature selection steps. The analysis explicitly uses the log-normalized data stored in a specific layer of the AnnData object, ensuring consistency in downstream computations.\n\nPrincipal Component Analysis (PCA) is performed on the highly variable genes to capture the primary axes of variation in the dataset. The PCA is computed using the ARPACK solver, which is efficient for sparse matrices. Following PCA, a neighborhood graph is constructed using the first 40 principal components and 15 nearest neighbors. This graph serves as the foundation for both t-distributed Stochastic Neighbor Embedding (t-SNE) and Uniform Manifold Approximation and Projection (UMAP) algorithms.\n\nt-SNE is applied to generate a two-dimensional embedding of the data, using the PCA representation as input. This embedding is useful for visualizing high-dimensional data in a reduced space. Similarly, UMAP is computed based on the neighborhood graph, providing an alternative low-dimensional representation that preserves both local and global data structures.\n\nVisualizations are generated to inspect the results of each dimensionality reduction technique. PCA scatter plots, t-SNE embeddings, and UMAP embeddings are plotted, with cells colored by their total counts to identify potential technical artifacts or biases in the data. These visualizations help assess the quality and interpretability of the dimensionality reduction results.\n\nFinally, the AnnData object, now enriched with the results of PCA, t-SNE, and UMAP, is saved to a file. This object is ready for subsequent analyses, such as clustering and cell type annotation. The entire workflow ensures reproducibility and provides a comprehensive foundation for further exploration of the single-cell dataset.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for single-cell data analysis and visualization, including Scanpy for single-cell analysis, Seaborn for advanced plotting, and Matplotlib for generating figures. Scanpy’s global parameters are configured to suppress verbose output and set figure properties, such as resolution (dpi=80) and background color (facecolor=\"white\"), to ensure consistent and clear visualizations.\n\nThe preprocessed single-cell dataset is loaded as an AnnData object from a local file, with a backup URL provided for accessibility. This dataset is assumed to have undergone prior quality control, normalization, and feature selection steps. The analysis explicitly uses the log-normalized data stored in a specific layer of the AnnData object, ensuring consistency in downstream computations.\n\nPrincipal Component Analysis (PCA) is performed on the highly variable genes to capture the primary axes of variation in the dataset. The PCA is computed using the ARPACK solver, which is efficient for sparse matrices. Following PCA, a neighborhood graph is constructed using the first 40 principal components and 15 nearest neighbors. This graph serves as the foundation for both t-distributed Stochastic Neighbor Embedding (t-SNE) and Uniform Manifold Approximation and Projection (UMAP) algorithms.\n\nt-SNE is applied to generate a two-dimensional embedding of the data, using the PCA representation as input. This embedding is useful for visualizing high-dimensional data in a reduced space. Similarly, UMAP is computed based on the neighborhood graph, providing an alternative low-dimensional representation that preserves both local and global data structures.\n\nVisualizations are generated to inspect the results of each dimensionality reduction technique. PCA scatter plots, t-SNE embeddings, and UMAP embeddings are plotted, with cells colored by their total counts to identify potential technical artifacts or biases in the data. These visualizations help assess the quality and interpretability of the dimensionality reduction results.\n\nFinally, the AnnData object, now enriched with the results of PCA, t-SNE, and UMAP, is saved to a file. This object is ready for subsequent analyses, such as clustering and cell type annotation. The entire workflow ensures reproducibility and provides a comprehensive foundation for further exploration of the single-cell dataset.",
         "The analysis was conducted in Python using Scanpy for single-cell data processing and visualization, complemented by Seaborn and Matplotlib for advanced plotting. The workflow focused on dimensionality reduction and visualization of a preprocessed single-cell dataset, loaded as an AnnData object. Principal Component Analysis (PCA) was performed on highly variable genes, followed by the construction of a neighborhood graph using the first 40 principal components and 15 nearest neighbors. This graph served as the basis for generating t-SNE and UMAP embeddings, which were visualized to assess data structure and identify potential technical artifacts. The enriched AnnData object, containing PCA, t-SNE, and UMAP results, was prepared for downstream analyses such as clustering and annotation."
        ],
        [
         "12",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell analysis and plotting\n\nlibrary(Seurat) # Main package for single-cell workflows\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"DimRed_DimDetermination\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing and Normalization\n\n#--------------------------------\n\n\n\n# 1. Normalize the data using the \"LogNormalize\" method.\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n# 2. Identify highly variable features (HVGs).\n\nseurat_obj <- FindVariableFeatures(\n\nseurat_obj,\n\nselection.method = \"vst\",\n\nnfeatures = 2000\n\n)\n\n\n\n# 3. Scale the data. This shifts the expression of each gene so that the mean\n\n# expression across cells is 0 and the variance is 1.\n\nall_genes <- rownames(seurat_obj)\n\nseurat_obj <- ScaleData(seurat_obj, features = all_genes)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Principal Component Analysis (PCA) ---\n\n# Run PCA on the scaled data, using only the highly variable features.\n\nseurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj))\n\n\n\n# --- UMAP Embedding ---\n\n# Note: Choosing the number of PCs is a critical step. The Elbow Plot in the\n\n# visualization section helps with this decision. Here, we use 10 as an example.\n\n# First, compute the neighborhood graph based on the selected PCs.\n\nseurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)\n\n\n\n# Then, run UMAP to get a non-linear embedding for visualization.\n\nseurat_obj <- RunUMAP(seurat_obj, dims = 1:10)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- PCA Plot ---\n\n# Visualize the first two principal components. Coloring by 'nCount_RNA' can help\n\n# reveal if technical factors (like library size) are driving the main sources of variation.\n\npca_plot <- DimPlot(seurat_obj, reduction = \"pca\") +\n\nggtitle(\"PCA: PC1 vs PC2\")\n\n\n\n# --- Elbow Plot for Dimensionality Determination ---\n\n# An Elbow Plot helps to determine the number of significant PCs to use for downstream\n\n# analysis. The \"elbow\" of the plot represents the point of diminishing returns,\n\n# where adding more PCs explains little additional variance.\n\nelbow_plot <- ElbowPlot(seurat_obj, ndims = 50) +\n\nggtitle(\"Elbow Plot: Variance Explained vs. PC\")\n\n\n\n# --- UMAP Plot ---\n\n# Visualize the UMAP embedding, which is often used for identifying cell clusters.\n\numap_plot <- DimPlot(seurat_obj, reduction = \"umap\") +\n\nggtitle(\"UMAP Embedding (Using PCs 1-10)\")\n\n\n\n# --- Combined Plot ---\n\n# Use patchwork to arrange the plots for a comprehensive overview.\n\n(pca_plot | elbow_plot) / umap_plot\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed Seurat object. It now contains the scaled data,\n\n# PCA results, and UMAP embedding, ready for clustering and annotation.\n\nsaveRDS(seurat_obj, file = \"DimRed_DimDetermined_SeuratObject.rds\")\n\n",
         "To perform single-cell RNA-seq analysis, the following workflow was implemented using the Seurat package in R. The analysis begins with data loading and preprocessing, followed by dimensionality reduction and visualization. Below is a detailed description of each step.\n\nFirst, the necessary libraries were loaded. The Seurat package was used for single-cell analysis, while ggplot2 and patchwork were employed for creating and combining visualizations. These tools are essential for handling single-cell data and generating publication-quality plots.\n\nThe raw count matrix was obtained from 10X Genomics data stored in a specified directory. The Read10X function was used to read the data, and a Seurat object was created using the CreateSeuratObject function. During this step, initial filtering was applied to retain genes detected in at least 3 cells and cells expressing at least 200 genes. This filtering ensures that low-quality cells and genes with minimal expression are excluded from downstream analysis.\n\nNext, the data underwent preprocessing and normalization. The LogNormalize method was applied to normalize the data, scaling the counts by a factor of 10,000 to account for differences in sequencing depth. Highly variable genes (HVGs) were identified using the FindVariableFeatures function with the \"vst\" selection method, retaining the top 2,000 genes. These HVGs are critical for capturing biological variation in the dataset. The data were then scaled to standardize gene expression levels, setting the mean expression to 0 and the variance to 1 across cells. This step ensures that downstream analyses are not biased by differences in gene expression magnitude.\n\nDimensionality reduction was performed to reduce the complexity of the data. Principal Component Analysis (PCA) was conducted using the scaled data and the highly variable genes. The resulting principal components (PCs) were used to compute a neighborhood graph, which served as the basis for Uniform Manifold Approximation and Projection (UMAP). UMAP was run using the first 10 PCs to generate a low-dimensional embedding for visualization. The choice of PCs was informed by an Elbow Plot, which helps determine the optimal number of PCs by identifying the point of diminishing returns in explained variance.\n\nVisualizations were created to explore the data. A PCA plot was generated to visualize the first two PCs, with points colored by the total number of RNA molecules per cell to assess potential technical biases. An Elbow Plot was used to determine the number of significant PCs for downstream analysis. Additionally, a UMAP plot was created to visualize the low-dimensional embedding, which is useful for identifying potential cell clusters. These plots were combined using the patchwork package to provide a comprehensive overview of the data.\n\nFinally, the processed Seurat object, containing the scaled data, PCA results, and UMAP embedding, was saved as an RDS file for further analysis, such as clustering and cell type annotation. This file preserves the intermediate results and can be reloaded for subsequent steps in the analysis pipeline.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform single-cell RNA-seq analysis, the following workflow was implemented using the Seurat package in R. The analysis begins with data loading and preprocessing, followed by dimensionality reduction and visualization. Below is a detailed description of each step.\n\nFirst, the necessary libraries were loaded. The Seurat package was used for single-cell analysis, while ggplot2 and patchwork were employed for creating and combining visualizations. These tools are essential for handling single-cell data and generating publication-quality plots.\n\nThe raw count matrix was obtained from 10X Genomics data stored in a specified directory. The Read10X function was used to read the data, and a Seurat object was created using the CreateSeuratObject function. During this step, initial filtering was applied to retain genes detected in at least 3 cells and cells expressing at least 200 genes. This filtering ensures that low-quality cells and genes with minimal expression are excluded from downstream analysis.\n\nNext, the data underwent preprocessing and normalization. The LogNormalize method was applied to normalize the data, scaling the counts by a factor of 10,000 to account for differences in sequencing depth. Highly variable genes (HVGs) were identified using the FindVariableFeatures function with the \"vst\" selection method, retaining the top 2,000 genes. These HVGs are critical for capturing biological variation in the dataset. The data were then scaled to standardize gene expression levels, setting the mean expression to 0 and the variance to 1 across cells. This step ensures that downstream analyses are not biased by differences in gene expression magnitude.\n\nDimensionality reduction was performed to reduce the complexity of the data. Principal Component Analysis (PCA) was conducted using the scaled data and the highly variable genes. The resulting principal components (PCs) were used to compute a neighborhood graph, which served as the basis for Uniform Manifold Approximation and Projection (UMAP). UMAP was run using the first 10 PCs to generate a low-dimensional embedding for visualization. The choice of PCs was informed by an Elbow Plot, which helps determine the optimal number of PCs by identifying the point of diminishing returns in explained variance.\n\nVisualizations were created to explore the data. A PCA plot was generated to visualize the first two PCs, with points colored by the total number of RNA molecules per cell to assess potential technical biases. An Elbow Plot was used to determine the number of significant PCs for downstream analysis. Additionally, a UMAP plot was created to visualize the low-dimensional embedding, which is useful for identifying potential cell clusters. These plots were combined using the patchwork package to provide a comprehensive overview of the data.\n\nFinally, the processed Seurat object, containing the scaled data, PCA results, and UMAP embedding, was saved as an RDS file for further analysis, such as clustering and cell type annotation. This file preserves the intermediate results and can be reloaded for subsequent steps in the analysis pipeline.",
         "Single-cell RNA sequencing data were analyzed using the Seurat package in R, focusing on dimensionality reduction and visualization. The raw count matrix was filtered to retain genes detected in at least three cells and cells expressing a minimum of 200 genes. Data normalization was performed using the LogNormalize method, followed by the identification of highly variable genes (HVGs) to capture biological variation. Principal Component Analysis (PCA) was conducted on scaled data, and the resulting PCs were used to construct a neighborhood graph for Uniform Manifold Approximation and Projection (UMAP) embedding. Visualization included PCA, Elbow Plot, and UMAP plots to explore data structure and identify potential clusters."
        ],
        [
         "13",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import the Scanpy library for single-cell analysis\n\nimport scanpy as sc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC 3k dataset from Scanpy as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# It's good practice to keep a copy of the raw data.\n\n# The .raw attribute is the conventional place to store it in Scanpy.\n\nadata.raw = adata.copy()\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization, Feature Selection, and Scaling ---\n\n# Normalize total counts per cell to account for sequencing depth differences\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to stabilize the variance\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) for downstream analysis\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# Subset the AnnData object to keep only the highly variable genes\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# Scale the data so that each gene has zero mean and unit variance\n\nsc.pp.scale(adata)\n\n\n\n# --- PCA and Neighborhood Graph ---\n\n# Perform Principal Component Analysis (PCA) on the scaled data\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# Compute the neighborhood graph of cells based on the PCA representation.\n\n# This graph is the basis for both clustering and UMAP embedding.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n\n\n\n# --- UMAP for Visualization ---\n\n# Generate a UMAP embedding from the neighbors graph. This is for visualization purposes.\n\nsc.tl.umap(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Clustering\n\n#--------------------------------\n\n\n\n# Apply the Leiden algorithm to find communities (clusters) in the neighborhood graph.\n\n# The `flavor='igraph'` and a fixed number of iterations are used for performance and reproducibility.\n\nsc.tl.leiden(adata, flavor='igraph', n_iterations=2)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the UMAP embedding, with cells colored by their assigned Leiden cluster.\n\n# This plot allows for a visual assessment of the clustering results.\n\nsc.pl.umap(adata, color=['leiden'], title=\"Leiden Clustering on PBMC3k\", size=20)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed AnnData object. It now contains the clustering results\n\n# in `adata.obs['leiden']` and is ready for cell type annotation.\n\nadata.write(\"pbmc3k_clustered.h5ad\")\n\n",
         "To analyze single-cell RNA sequencing data using the Scanpy library in Python, the following steps were performed:\n\nFirst, the Scanpy library was imported to facilitate single-cell data analysis. This library provides essential tools for preprocessing, dimensionality reduction, clustering, and visualization of single-cell datasets. Global plotting parameters were configured to ensure consistent and clear visual outputs, with a resolution of 80 DPI and a white background for figures.\n\nThe analysis began with the loading of the PBMC 3k dataset, a publicly available single-cell dataset included in Scanpy for demonstration purposes. To preserve the raw data for downstream comparisons, a copy of the dataset was stored in the `.raw` attribute of the AnnData object, which is the standard data structure used by Scanpy.\n\nPreprocessing steps were then applied to prepare the data for analysis. The total counts per cell were normalized to a target sum of 10,000 to account for differences in sequencing depth. A log(1+x) transformation was applied to stabilize the variance across genes. Highly variable genes (HVGs) were identified using the Seurat flavor method, with the top 2,000 genes selected for downstream analysis. The dataset was subset to retain only these highly variable genes, and the data were scaled to ensure each gene had zero mean and unit variance.\n\nDimensionality reduction was performed using Principal Component Analysis (PCA) with the ARPACK solver. A neighborhood graph of cells was computed based on the PCA representation, using 10 nearest neighbors and 40 principal components. This graph served as the foundation for clustering and visualization. A UMAP embedding was generated from the neighborhood graph to enable two-dimensional visualization of the data.\n\nClustering was performed using the Leiden algorithm, which identifies communities (clusters) within the neighborhood graph. The algorithm was configured with the `igraph` flavor and a fixed number of iterations (2) to ensure reproducibility and performance. The resulting clusters were stored in the `leiden` column of the AnnData object’s observation metadata.\n\nTo visualize the clustering results, a UMAP plot was generated with cells colored by their assigned Leiden cluster. This plot provided a visual assessment of the clustering outcomes and the overall structure of the dataset.\n\nFinally, the processed AnnData object, which now included the clustering results, was saved to a file named `pbmc3k_clustered.h5ad`. This file is ready for further analysis, such as cell type annotation or additional downstream investigations.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo analyze single-cell RNA sequencing data using the Scanpy library in Python, the following steps were performed:\n\nFirst, the Scanpy library was imported to facilitate single-cell data analysis. This library provides essential tools for preprocessing, dimensionality reduction, clustering, and visualization of single-cell datasets. Global plotting parameters were configured to ensure consistent and clear visual outputs, with a resolution of 80 DPI and a white background for figures.\n\nThe analysis began with the loading of the PBMC 3k dataset, a publicly available single-cell dataset included in Scanpy for demonstration purposes. To preserve the raw data for downstream comparisons, a copy of the dataset was stored in the `.raw` attribute of the AnnData object, which is the standard data structure used by Scanpy.\n\nPreprocessing steps were then applied to prepare the data for analysis. The total counts per cell were normalized to a target sum of 10,000 to account for differences in sequencing depth. A log(1+x) transformation was applied to stabilize the variance across genes. Highly variable genes (HVGs) were identified using the Seurat flavor method, with the top 2,000 genes selected for downstream analysis. The dataset was subset to retain only these highly variable genes, and the data were scaled to ensure each gene had zero mean and unit variance.\n\nDimensionality reduction was performed using Principal Component Analysis (PCA) with the ARPACK solver. A neighborhood graph of cells was computed based on the PCA representation, using 10 nearest neighbors and 40 principal components. This graph served as the foundation for clustering and visualization. A UMAP embedding was generated from the neighborhood graph to enable two-dimensional visualization of the data.\n\nClustering was performed using the Leiden algorithm, which identifies communities (clusters) within the neighborhood graph. The algorithm was configured with the `igraph` flavor and a fixed number of iterations (2) to ensure reproducibility and performance. The resulting clusters were stored in the `leiden` column of the AnnData object’s observation metadata.\n\nTo visualize the clustering results, a UMAP plot was generated with cells colored by their assigned Leiden cluster. This plot provided a visual assessment of the clustering outcomes and the overall structure of the dataset.\n\nFinally, the processed AnnData object, which now included the clustering results, was saved to a file named `pbmc3k_clustered.h5ad`. This file is ready for further analysis, such as cell type annotation or additional downstream investigations.",
         "Single-cell RNA sequencing data analysis was conducted using the Scanpy library in Python. The workflow encompassed preprocessing, dimensionality reduction, clustering, and visualization. The PBMC 3k dataset was normalized to account for sequencing depth, log-transformed, and subset to retain highly variable genes identified via the Seurat flavor method. Principal Component Analysis (PCA) was employed for dimensionality reduction, followed by the construction of a neighborhood graph. Leiden clustering was applied to identify cell communities, and results were visualized using UMAP embeddings. This pipeline facilitated the exploration of cellular heterogeneity and prepared the dataset for downstream analyses such as cell type annotation."
        ],
        [
         "14",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis and plotting\nimport scanpy as sc\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\nsc.settings.verbosity = 0\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the preprocessed AnnData object. This file should contain\n# dimensionality reduction results (e.g., PCA).\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"s4d8_dimensionality_reduction.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/40016014\"\n)\n\n#--------------------------------\n# Major Analysis Task: Clustering\n#--------------------------------\n\n# --- Neighborhood Graph ---\n# Compute the neighborhood graph of cells based on the PCA representation.\n# This graph is the foundation for both clustering and UMAP.\nsc.pp.neighbors(adata, n_neighbors=15, n_pcs=30)\n\n# --- Leiden Clustering ---\n# Apply the Leiden algorithm to partition the neighborhood graph into clusters.\n# The 'resolution' parameter controls the number of clusters found; higher\n# values lead to more, smaller clusters.\nsc.tl.leiden(adata, resolution=1.0, key_added=\"leiden\")\n\n# --- UMAP for Visualization ---\n# Compute the UMAP embedding from the neighborhood graph. This is done\n# here to ensure it's available for plotting, in case it wasn't run previously.\nsc.tl.umap(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, with cells colored by their assigned Leiden cluster.\n# The `legend_loc='on data'` option places cluster labels directly on the plot.\nsc.pl.umap(adata, color=[\"leiden\"], title=\"Leiden Clustering\", legend_loc='on data')\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the updated AnnData object. It now contains the clustering results\n# stored in `adata.obs['leiden']` and is ready for marker gene analysis.\nadata.write(\"s4d8_clustering.h5ad\")\n",
         "The analysis begins by importing the necessary Python modules for single-cell data analysis and visualization. The `scanpy` library is used for handling single-cell data, while `matplotlib.pyplot` is employed for plotting. Global parameters for `scanpy` are configured to suppress verbose output and set plotting preferences, including a resolution of 80 DPI and a white background without frames.\n\nThe preprocessed single-cell data, stored in an AnnData object, is loaded from a local file named `s4d8_dimensionality_reduction.h5ad`. This file is expected to contain dimensionality reduction results, such as principal component analysis (PCA). If the local file is unavailable, a backup URL is provided to download the data. The AnnData object, `adata`, serves as the primary data structure for the analysis.\n\nThe core analysis involves clustering the cells based on their PCA representation. First, a neighborhood graph of cells is computed using the PCA results, with 15 neighbors and 30 principal components specified as parameters. This graph forms the basis for both clustering and subsequent visualization. The Leiden clustering algorithm is then applied to partition the neighborhood graph into distinct clusters. The `resolution` parameter is set to 1.0, which controls the granularity of the clustering; higher values result in more, smaller clusters. The clustering results are stored in `adata.obs['leiden']`. To facilitate visualization, a UMAP embedding is computed from the neighborhood graph, ensuring it is available for plotting even if it was not previously generated.\n\nThe UMAP embedding is visualized with cells colored by their assigned Leiden clusters. Cluster labels are directly placed on the plot using the `legend_loc='on data'` option, providing a clear representation of the clustering results.\n\nFinally, the updated AnnData object, now containing the clustering results, is saved to a new file named `s4d8_clustering.h5ad`. This file is ready for downstream analyses, such as marker gene identification. The entire process ensures a comprehensive and reproducible workflow for single-cell data clustering and visualization.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing the necessary Python modules for single-cell data analysis and visualization. The `scanpy` library is used for handling single-cell data, while `matplotlib.pyplot` is employed for plotting. Global parameters for `scanpy` are configured to suppress verbose output and set plotting preferences, including a resolution of 80 DPI and a white background without frames.\n\nThe preprocessed single-cell data, stored in an AnnData object, is loaded from a local file named `s4d8_dimensionality_reduction.h5ad`. This file is expected to contain dimensionality reduction results, such as principal component analysis (PCA). If the local file is unavailable, a backup URL is provided to download the data. The AnnData object, `adata`, serves as the primary data structure for the analysis.\n\nThe core analysis involves clustering the cells based on their PCA representation. First, a neighborhood graph of cells is computed using the PCA results, with 15 neighbors and 30 principal components specified as parameters. This graph forms the basis for both clustering and subsequent visualization. The Leiden clustering algorithm is then applied to partition the neighborhood graph into distinct clusters. The `resolution` parameter is set to 1.0, which controls the granularity of the clustering; higher values result in more, smaller clusters. The clustering results are stored in `adata.obs['leiden']`. To facilitate visualization, a UMAP embedding is computed from the neighborhood graph, ensuring it is available for plotting even if it was not previously generated.\n\nThe UMAP embedding is visualized with cells colored by their assigned Leiden clusters. Cluster labels are directly placed on the plot using the `legend_loc='on data'` option, providing a clear representation of the clustering results.\n\nFinally, the updated AnnData object, now containing the clustering results, is saved to a new file named `s4d8_clustering.h5ad`. This file is ready for downstream analyses, such as marker gene identification. The entire process ensures a comprehensive and reproducible workflow for single-cell data clustering and visualization.",
         "The analysis was conducted in Python using the Scanpy library for single-cell data processing and visualization, with Matplotlib employed for plotting. The workflow focused on clustering cells based on principal component analysis (PCA) results stored in an AnnData object. A neighborhood graph was constructed using 15 neighbors and 30 principal components, followed by Leiden clustering with a resolution parameter of 1.0 to define cell populations. A UMAP embedding was generated for visualization, with cells colored by their assigned Leiden clusters and labels overlaid directly on the plot. The updated AnnData object, including clustering results, was prepared for downstream analyses."
        ],
        [
         "15",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell analysis, data manipulation, and plotting\n\nlibrary(Seurat) # Main package for single-cell RNA-seq analysis\n\nlibrary(dplyr) # For data manipulation functions\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"ClusteringDemo\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing, QC, and Normalization\n\n#--------------------------------\n\n\n\n# Calculate the percentage of mitochondrial genes for each cell as a QC metric.\n\nseurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(seurat_obj, pattern = \"^MT-\")\n\n\n\n# Normalize the data to account for differences in library size.\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n# Identify highly variable features (genes) to be used for downstream analysis.\n\nseurat_obj <- FindVariableFeatures(\n\nseurat_obj,\n\nselection.method = \"vst\",\n\nnfeatures = 2000\n\n)\n\n\n\n# Scale the data, which is a standard step before running PCA.\n\nseurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj))\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Dimensionality Reduction & Clustering\n\n#--------------------------------\n\n\n\n# --- PCA ---\n\n# Run Principal Component Analysis on the highly variable features.\n\nseurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj))\n\n\n\n# --- Clustering ---\n\n# Build a K-nearest neighbor (KNN) graph based on the PCA space.\n\n# The number of PCs (dims) used is a critical parameter.\n\nseurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)\n\n\n\n# Apply graph-based clustering to partition the cells.\n\n# The `resolution` parameter controls the number of clusters identified.\n\nseurat_obj <- FindClusters(seurat_obj, resolution = 0.5)\n\n\n\n# --- UMAP for Visualization ---\n\n# Run UMAP to generate a non-linear 2D embedding for visualization.\n\nseurat_obj <- RunUMAP(seurat_obj, dims = 1:10)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize QC metrics with violin plots.\n\nVlnPlot(seurat_obj, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3)\n\n\n\n# Visualize the PCA results.\n\nDimPlot(seurat_obj, reduction = \"pca\") + ggtitle(\"PCA Plot\")\n\n\n\n# Visualize the UMAP embedding with cells colored by their assigned cluster.\n\nDimPlot(seurat_obj, reduction = \"umap\", label = TRUE, pt.size = 0.8) +\n\nggtitle(\"UMAP Plot with Cluster Labels\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed Seurat object, which now contains clustering results.\n\n# This object is ready for downstream marker analysis and cell type annotation.\n\nsaveRDS(seurat_obj, file = \"Clustering_SeuratObject.rds\")\n\n",
         "To perform single-cell RNA sequencing analysis, the following steps were executed using the Seurat package in R, along with additional libraries for data manipulation and visualization.\n\nFirst, the necessary libraries were loaded. Seurat was used for single-cell RNA-seq analysis, dplyr for data manipulation, ggplot2 for creating custom plots, and patchwork for combining multiple ggplot2 plots. These tools are essential for handling, analyzing, and visualizing single-cell data.\n\nThe raw data, consisting of 10X Genomics count matrices, was read into a raw count matrix. The data directory was specified, and the Read10X function was used to load the data. A Seurat object was then created from the raw count matrix with basic initial filtering. The CreateSeuratObject function was employed, with parameters set to include genes detected in at least 3 cells (min.cells = 3) and cells with at least 200 features (min.features = 200). This step ensures that only high-quality cells and genes are included in the analysis.\n\nNext, preprocessing, quality control (QC), and normalization were performed. The percentage of mitochondrial genes was calculated for each cell as a QC metric using the PercentageFeatureSet function. The data was normalized to account for differences in library size using the NormalizeData function with the LogNormalize method and a scale factor of 10,000. Highly variable features were identified using the FindVariableFeatures function with the vst selection method and a set of 2000 features. The data was then scaled using the ScaleData function, a standard step before running principal component analysis (PCA).\n\nFor dimensionality reduction and clustering, PCA was run on the highly variable features using the RunPCA function. A K-nearest neighbor (KNN) graph was built based on the PCA space using the FindNeighbors function, with the first 10 principal components (dims = 1:10) used as the critical parameter. Graph-based clustering was applied to partition the cells using the FindClusters function, with a resolution parameter of 0.5 to control the number of clusters identified. Uniform Manifold Approximation and Projection (UMAP) was run to generate a non-linear 2D embedding for visualization using the RunUMAP function, again using the first 10 principal components.\n\nVisualization of the results was performed using various plotting functions. QC metrics were visualized with violin plots using the VlnPlot function, showing the number of features, counts, and mitochondrial gene percentage. PCA results were visualized using the DimPlot function with the reduction parameter set to \"pca\". The UMAP embedding was visualized with cells colored by their assigned cluster using the DimPlot function with the reduction parameter set to \"umap\", and cluster labels were added for clarity.\n\nFinally, the processed Seurat object, which now contains clustering results, was saved for downstream marker analysis and cell type annotation using the saveRDS function. This object is ready for further analysis and interpretation.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform single-cell RNA sequencing analysis, the following steps were executed using the Seurat package in R, along with additional libraries for data manipulation and visualization.\n\nFirst, the necessary libraries were loaded. Seurat was used for single-cell RNA-seq analysis, dplyr for data manipulation, ggplot2 for creating custom plots, and patchwork for combining multiple ggplot2 plots. These tools are essential for handling, analyzing, and visualizing single-cell data.\n\nThe raw data, consisting of 10X Genomics count matrices, was read into a raw count matrix. The data directory was specified, and the Read10X function was used to load the data. A Seurat object was then created from the raw count matrix with basic initial filtering. The CreateSeuratObject function was employed, with parameters set to include genes detected in at least 3 cells (min.cells = 3) and cells with at least 200 features (min.features = 200). This step ensures that only high-quality cells and genes are included in the analysis.\n\nNext, preprocessing, quality control (QC), and normalization were performed. The percentage of mitochondrial genes was calculated for each cell as a QC metric using the PercentageFeatureSet function. The data was normalized to account for differences in library size using the NormalizeData function with the LogNormalize method and a scale factor of 10,000. Highly variable features were identified using the FindVariableFeatures function with the vst selection method and a set of 2000 features. The data was then scaled using the ScaleData function, a standard step before running principal component analysis (PCA).\n\nFor dimensionality reduction and clustering, PCA was run on the highly variable features using the RunPCA function. A K-nearest neighbor (KNN) graph was built based on the PCA space using the FindNeighbors function, with the first 10 principal components (dims = 1:10) used as the critical parameter. Graph-based clustering was applied to partition the cells using the FindClusters function, with a resolution parameter of 0.5 to control the number of clusters identified. Uniform Manifold Approximation and Projection (UMAP) was run to generate a non-linear 2D embedding for visualization using the RunUMAP function, again using the first 10 principal components.\n\nVisualization of the results was performed using various plotting functions. QC metrics were visualized with violin plots using the VlnPlot function, showing the number of features, counts, and mitochondrial gene percentage. PCA results were visualized using the DimPlot function with the reduction parameter set to \"pca\". The UMAP embedding was visualized with cells colored by their assigned cluster using the DimPlot function with the reduction parameter set to \"umap\", and cluster labels were added for clarity.\n\nFinally, the processed Seurat object, which now contains clustering results, was saved for downstream marker analysis and cell type annotation using the saveRDS function. This object is ready for further analysis and interpretation.",
         "Single-cell RNA sequencing data were analyzed using the Seurat package in R, supplemented by dplyr, ggplot2, and patchwork for data manipulation and visualization. The workflow began with the creation of a Seurat object from 10X Genomics count matrices, incorporating initial filtering to retain high-quality cells and genes. Quality control metrics, including mitochondrial gene percentages, were calculated, followed by normalization using the LogNormalize method. Highly variable features were identified and scaled, enabling dimensionality reduction via principal component analysis (PCA). A K-nearest neighbor graph was constructed, and graph-based clustering was performed, with results visualized using UMAP embeddings. The processed data were saved for downstream marker analysis and cell type annotation."
        ],
        [
         "16",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC 3k dataset from Scanpy as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Preprocessing & Clustering\n\n#--------------------------------\n\n\n\n# --- Normalization, Feature Selection, and Scaling ---\n\n# Normalize total counts per cell to account for sequencing depth differences\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to stabilize the variance\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) for downstream analysis\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# Subset the AnnData object to keep only the highly variable genes\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# Scale the data so that each gene has zero mean and unit variance\n\nsc.pp.scale(adata)\n\n\n\n# --- PCA, Neighborhood Graph, and UMAP ---\n\n# Perform Principal Component Analysis (PCA) on the scaled data\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# Compute the neighborhood graph, which is the basis for clustering and UMAP\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n\n\n\n# Generate a UMAP embedding for visualization\n\nsc.tl.umap(adata)\n\n\n\n# --- Leiden Clustering ---\n\n# Apply the Leiden algorithm to find communities (clusters) of cells\n\nsc.tl.leiden(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Cell-Type Annotation\n\n#--------------------------------\n\n\n\n# Define a dictionary of canonical marker genes for common PBMC types.\n\n# These genes are used to identify cell types based on their expression patterns.\n\nmarker_genes = {\n\n'B cells': ['MS4A1', 'CD79A'],\n\n'T cells': ['CD3D', 'CD3E'],\n\n'NK cells': ['NKG7', 'GNLY'],\n\n'Monocytes': ['CD14', 'LYZ']\n\n}\n\n\n\n# Manually define a mapping from cluster IDs to cell type names.\n\n# NOTE: This mapping is created after inspecting the dot plot below and should be\n\n# adjusted based on the specific results of your analysis.\n\ncluster_to_cell_type = {\n\n'0': 'T cells',\n\n'1': 'Monocytes',\n\n'2': 'NK cells',\n\n'3': 'B cells',\n\n'4': 'T cells' # Example of merging multiple clusters into one cell type\n\n}\n\n\n\n# Create a new column in the observation metadata to store the cell type annotations.\n\n# The .map() function applies the dictionary to the 'leiden' column.\n\nadata.obs['cell_type'] = adata.obs['leiden'].map(cluster_to_cell_type).astype('category')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a dot plot to visualize the expression of marker genes across clusters.\n\n# This plot is essential for assigning cell types to clusters.\n\nsc.pl.dotplot(adata, marker_genes, groupby='leiden', standard_scale='var',\n\ntitle='Marker Expression by Leiden Clusters')\n\n\n\n# Visualize the final cell type annotations on the UMAP embedding.\n\n# This plot shows the spatial distribution of the identified cell types.\n\nsc.pl.umap(adata, color='cell_type', title='Cell Type Annotation', legend_loc='on data')\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the fully processed and annotated AnnData object.\n\n# This file can be used for further downstream analysis, such as differential\n\n# expression between cell types or cell-cell interaction studies.\n\nadata.write(\"pbmc3k_annotated.h5ad\")\n\n",
         "The analysis was performed using Python, leveraging the Scanpy and AnnData libraries for single-cell RNA sequencing data processing and visualization. The workflow encompasses data loading, preprocessing, clustering, cell-type annotation, and visualization, as described below.\n\nFirst, the necessary libraries for single-cell analysis and data handling were imported. Scanpy was used for data processing and visualization, while AnnData provided the data structure for storing and manipulating the single-cell dataset. Global parameters for plotting were configured to ensure consistent and clear visuals, with a DPI of 80 and a white background.\n\nThe analysis began with the loading of the built-in PBMC 3k dataset from Scanpy, which serves as an example dataset for single-cell RNA sequencing analysis. This dataset was stored in an AnnData object, a specialized data structure for single-cell data.\n\nPreprocessing steps included normalization, feature selection, and scaling. The total counts per cell were normalized to a target sum of 10,000 to account for differences in sequencing depth. A log(1+x) transformation was applied to stabilize the variance. Highly variable genes (HVGs) were identified using the Seurat flavor, with the top 2,000 genes selected for downstream analysis. The dataset was subset to retain only these highly variable genes, and the data were scaled to ensure each gene had zero mean and unit variance.\n\nPrincipal Component Analysis (PCA) was performed on the scaled data using the ARPACK solver. A neighborhood graph was computed based on the top 40 principal components and 10 nearest neighbors, which served as the foundation for clustering and dimensionality reduction. A UMAP embedding was generated to visualize the data in two dimensions. The Leiden algorithm was applied to identify cell communities (clusters) based on the neighborhood graph.\n\nFor cell-type annotation, a dictionary of canonical marker genes for common PBMC cell types was defined. These marker genes included MS4A1 and CD79A for B cells, CD3D and CD3E for T cells, NKG7 and GNLY for NK cells, and CD14 and LYZ for monocytes. A manual mapping from cluster IDs to cell type names was created based on the inspection of marker gene expression patterns. This mapping was applied to the Leiden cluster labels to annotate cell types, and the results were stored in a new column in the observation metadata.\n\nVisualization included the generation of a dot plot to display the expression of marker genes across Leiden clusters, facilitating the assignment of cell types to clusters. The final cell type annotations were visualized on the UMAP embedding to show the spatial distribution of identified cell types.\n\nThe fully processed and annotated AnnData object was saved to a file named \"pbmc3k_annotated.h5ad\" for further downstream analysis, such as differential expression studies or cell-cell interaction analyses. This workflow provides a comprehensive pipeline for single-cell RNA sequencing data analysis, from raw data to annotated cell types and visualizations.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis was performed using Python, leveraging the Scanpy and AnnData libraries for single-cell RNA sequencing data processing and visualization. The workflow encompasses data loading, preprocessing, clustering, cell-type annotation, and visualization, as described below.\n\nFirst, the necessary libraries for single-cell analysis and data handling were imported. Scanpy was used for data processing and visualization, while AnnData provided the data structure for storing and manipulating the single-cell dataset. Global parameters for plotting were configured to ensure consistent and clear visuals, with a DPI of 80 and a white background.\n\nThe analysis began with the loading of the built-in PBMC 3k dataset from Scanpy, which serves as an example dataset for single-cell RNA sequencing analysis. This dataset was stored in an AnnData object, a specialized data structure for single-cell data.\n\nPreprocessing steps included normalization, feature selection, and scaling. The total counts per cell were normalized to a target sum of 10,000 to account for differences in sequencing depth. A log(1+x) transformation was applied to stabilize the variance. Highly variable genes (HVGs) were identified using the Seurat flavor, with the top 2,000 genes selected for downstream analysis. The dataset was subset to retain only these highly variable genes, and the data were scaled to ensure each gene had zero mean and unit variance.\n\nPrincipal Component Analysis (PCA) was performed on the scaled data using the ARPACK solver. A neighborhood graph was computed based on the top 40 principal components and 10 nearest neighbors, which served as the foundation for clustering and dimensionality reduction. A UMAP embedding was generated to visualize the data in two dimensions. The Leiden algorithm was applied to identify cell communities (clusters) based on the neighborhood graph.\n\nFor cell-type annotation, a dictionary of canonical marker genes for common PBMC cell types was defined. These marker genes included MS4A1 and CD79A for B cells, CD3D and CD3E for T cells, NKG7 and GNLY for NK cells, and CD14 and LYZ for monocytes. A manual mapping from cluster IDs to cell type names was created based on the inspection of marker gene expression patterns. This mapping was applied to the Leiden cluster labels to annotate cell types, and the results were stored in a new column in the observation metadata.\n\nVisualization included the generation of a dot plot to display the expression of marker genes across Leiden clusters, facilitating the assignment of cell types to clusters. The final cell type annotations were visualized on the UMAP embedding to show the spatial distribution of identified cell types.\n\nThe fully processed and annotated AnnData object was saved to a file named \"pbmc3k_annotated.h5ad\" for further downstream analysis, such as differential expression studies or cell-cell interaction analyses. This workflow provides a comprehensive pipeline for single-cell RNA sequencing data analysis, from raw data to annotated cell types and visualizations.",
         "The analysis was conducted in Python using the Scanpy and AnnData libraries for processing and visualizing single-cell RNA sequencing data. The workflow included normalization of total counts per cell, log transformation for variance stabilization, and identification of highly variable genes using the Seurat flavor. Principal Component Analysis (PCA) was performed, followed by UMAP for dimensionality reduction and Leiden clustering to identify cell communities. Cell-type annotation was achieved by mapping canonical marker genes to cluster IDs, enabling the identification of PBMC cell types such as B cells, T cells, NK cells, and monocytes. Visualization included dot plots and UMAP embeddings to illustrate marker gene expression and spatial distribution of annotated cell types."
        ],
        [
         "17",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis and plotting\nimport scanpy as sc\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\nsc.settings.verbosity = 0\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the preprocessed and clustered AnnData object.\n# This file is expected to contain clustering results (e.g., in `adata.obs['leiden']`).\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"s4d8_clustering.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/40016014\"\n)\n\n#--------------------------------\n# Major Analysis Task: Cell-Type Annotation\n#--------------------------------\n\n# Define a dictionary to manually map cluster IDs to cell type names.\n# This mapping is based on prior biological knowledge or marker gene analysis.\nmanual_labels = {\n\"0\": \"B cell\",\n\"1\": \"T cell\",\n\"2\": \"Myeloid cell\",\n\"3\": \"NK cell\",\n\"4\": \"Dendritic cell\"\n}\n\n# Create a new column 'manual_annotation' in the observation metadata.\n# The .map() function applies the dictionary. The lambda function with .get()\n# ensures that any cluster ID not in the dictionary is assigned \"Unknown\".\nadata.obs[\"manual_annotation\"] = adata.obs[\"leiden\"].map(\nlambda x: manual_labels.get(x, \"Unknown\")\n)\n\n# Print a count of cells assigned to each annotated cell type for a quick summary.\nprint(\"Manual annotation summary:\")\nprint(adata.obs[\"manual_annotation\"].value_counts())\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Ensure a UMAP embedding exists for visualization.\n# This check prevents errors if the loaded object doesn't have UMAP coordinates.\nif \"X_umap\" not in adata.obsm.keys():\nsc.pp.neighbors(adata) # UMAP requires a neighbor graph\nsc.tl.umap(adata)\n\n# Visualize the UMAP embedding, with cells colored by their manual annotation.\n# This plot provides a spatial view of the identified cell populations.\nsc.pl.umap(\nadata,\ncolor=\"manual_annotation\",\ntitle=\"Manual Cell Type Annotation\",\nlegend_loc=\"on data\"\n)\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the updated AnnData object, which now includes the manual cell type annotations.\n# This file is ready for downstream analyses focused on specific cell populations.\nadata.write(\"s4d8_manual_annotation.h5ad\")\n",
         "The analysis begins by importing essential Python libraries for single-cell RNA sequencing data analysis and visualization. The `scanpy` library is used for handling and analyzing single-cell data, while `matplotlib.pyplot` is employed for generating plots. These libraries are critical for performing the subsequent steps in the analysis.\n\nGlobal parameters for `scanpy` are configured to optimize the analysis environment. The verbosity level is set to 0 to suppress unnecessary output, and figure parameters are adjusted to ensure high-quality visualizations with a resolution of 80 DPI, a white background, and no frame around the plots. These settings streamline the analysis and improve the clarity of the resulting figures.\n\nThe preprocessed and clustered single-cell data is loaded from an AnnData object stored in an H5AD file. This file contains clustering results, such as Leiden cluster assignments, which are essential for downstream cell-type annotation. A backup URL is provided to ensure data accessibility in case the local file is unavailable. The AnnData object is read into the variable `adata` for further processing.\n\nCell-type annotation is performed by manually mapping cluster IDs to specific cell types based on prior biological knowledge or marker gene analysis. A dictionary named `manual_labels` is created to define this mapping, assigning cell types such as \"B cell,\" \"T cell,\" and \"Myeloid cell\" to their respective cluster IDs. This dictionary is applied to the Leiden cluster assignments in the observation metadata (`adata.obs['leiden']`) using the `.map()` function. A lambda function ensures that any cluster ID not present in the dictionary is labeled as \"Unknown.\" The resulting annotations are stored in a new column, `manual_annotation`, within the observation metadata. A summary of the annotated cell types is printed to provide an overview of the distribution of cell populations.\n\nTo visualize the annotated cell types, the analysis checks for the presence of UMAP coordinates in the AnnData object. If UMAP coordinates are not found, a neighbor graph is computed using `sc.pp.neighbors()`, and UMAP embedding is generated using `sc.tl.umap()`. This step ensures that the data is properly prepared for visualization. The UMAP plot is then created with cells colored according to their manual annotations. The plot includes a title and places the legend directly on the data points for clarity, providing a spatial representation of the identified cell populations.\n\nFinally, the updated AnnData object, now containing the manual cell-type annotations, is saved to a new H5AD file named `s4d8_manual_annotation.h5ad`. This file is ready for downstream analyses focused on specific cell populations, ensuring that the annotated data can be easily accessed and utilized in future studies.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for single-cell RNA sequencing data analysis and visualization. The `scanpy` library is used for handling and analyzing single-cell data, while `matplotlib.pyplot` is employed for generating plots. These libraries are critical for performing the subsequent steps in the analysis.\n\nGlobal parameters for `scanpy` are configured to optimize the analysis environment. The verbosity level is set to 0 to suppress unnecessary output, and figure parameters are adjusted to ensure high-quality visualizations with a resolution of 80 DPI, a white background, and no frame around the plots. These settings streamline the analysis and improve the clarity of the resulting figures.\n\nThe preprocessed and clustered single-cell data is loaded from an AnnData object stored in an H5AD file. This file contains clustering results, such as Leiden cluster assignments, which are essential for downstream cell-type annotation. A backup URL is provided to ensure data accessibility in case the local file is unavailable. The AnnData object is read into the variable `adata` for further processing.\n\nCell-type annotation is performed by manually mapping cluster IDs to specific cell types based on prior biological knowledge or marker gene analysis. A dictionary named `manual_labels` is created to define this mapping, assigning cell types such as \"B cell,\" \"T cell,\" and \"Myeloid cell\" to their respective cluster IDs. This dictionary is applied to the Leiden cluster assignments in the observation metadata (`adata.obs['leiden']`) using the `.map()` function. A lambda function ensures that any cluster ID not present in the dictionary is labeled as \"Unknown.\" The resulting annotations are stored in a new column, `manual_annotation`, within the observation metadata. A summary of the annotated cell types is printed to provide an overview of the distribution of cell populations.\n\nTo visualize the annotated cell types, the analysis checks for the presence of UMAP coordinates in the AnnData object. If UMAP coordinates are not found, a neighbor graph is computed using `sc.pp.neighbors()`, and UMAP embedding is generated using `sc.tl.umap()`. This step ensures that the data is properly prepared for visualization. The UMAP plot is then created with cells colored according to their manual annotations. The plot includes a title and places the legend directly on the data points for clarity, providing a spatial representation of the identified cell populations.\n\nFinally, the updated AnnData object, now containing the manual cell-type annotations, is saved to a new H5AD file named `s4d8_manual_annotation.h5ad`. This file is ready for downstream analyses focused on specific cell populations, ensuring that the annotated data can be easily accessed and utilized in future studies.",
         "The analysis was conducted in Python using the Scanpy library for single-cell RNA sequencing data processing and visualization, with Matplotlib employed for generating plots. Cell-type annotation was performed by manually mapping Leiden cluster IDs to specific cell types based on prior biological knowledge, creating a dictionary to define this mapping. The annotations were applied to the observation metadata, with unassigned clusters labeled as \"Unknown.\" UMAP coordinates were generated or verified for visualization, and a UMAP plot was created to spatially represent the annotated cell populations. The updated dataset, including manual annotations, was saved for downstream analyses."
        ],
        [
         "18",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell analysis, data manipulation, and plotting\n\nlibrary(Seurat) # Core package for single-cell RNA-seq analysis\n\nlibrary(dplyr) # For data manipulation, especially for handling marker gene tables\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"ClusterBiomarkers\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing and QC\n\n#--------------------------------\n\n\n\n# Calculate the percentage of mitochondrial genes for each cell as a QC metric.\n\nseurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(seurat_obj, pattern = \"^MT-\")\n\n\n\n# Normalize the data to account for differences in library size.\n\nseurat_obj <- NormalizeData(seurat_obj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n\n\n\n# Identify highly variable features (genes) for downstream analysis.\n\nseurat_obj <- FindVariableFeatures(seurat_obj, selection.method = \"vst\", nfeatures = 2000)\n\n\n\n# Scale the data, a standard step before running PCA.\n\nall_genes <- rownames(seurat_obj)\n\nseurat_obj <- ScaleData(seurat_obj, features = all_genes)\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# --- Dimensionality Reduction & Clustering ---\n\n# Run PCA on the highly variable features.\n\nseurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj))\n\n\n\n# Build a K-nearest neighbor (KNN) graph and perform clustering.\n\nseurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)\n\nseurat_obj <- FindClusters(seurat_obj, resolution = 0.5)\n\n\n\n# Run UMAP to generate a 2D embedding for visualization.\n\nseurat_obj <- RunUMAP(seurat_obj, dims = 1:10)\n\n\n\n\n\n# --- Finding Cluster Biomarkers ---\n\n# Find differentially expressed features (cluster biomarkers) for every cluster compared to all remaining cells.\n\n# `only.pos = TRUE` ensures we only get positive markers (genes upregulated in the cluster).\n\ncluster_markers <- FindAllMarkers(seurat_obj, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)\n\n\n\n\n\n# --- Cell Type Annotation ---\n\n# Manually define new cell type names based on canonical markers found in the literature\n\n# and observed in the `cluster_markers` table.\n\n# NOTE: The order and names must be adjusted based on your specific dataset and clustering results.\n\nnew.cluster.ids <- c(\"Naive CD4 T\", \"CD14+ Monocytes\", \"Memory CD4 T\", \"B cell\",\n\n\"CD8 T\", \"NK cell\", \"Dendritic cell\", \"Platelet\")\n\n\n\n# Assign the new names to the cluster levels in the Seurat object.\n\nnames(new.cluster.ids) <- levels(seurat_obj)\n\nseurat_obj <- RenameIdents(seurat_obj, new.cluster.ids)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the initial UMAP clusters before annotation.\n\nDimPlot(seurat_obj, reduction = \"umap\", label = TRUE, pt.size = 0.8) +\n\nggtitle(\"UMAP Plot: Clusters before Cell Type Assignment\")\n\n\n\n# --- Biomarker Visualization ---\n\n# Extract the top markers for each cluster to visualize on a heatmap.\n\n# Here, we group by cluster and select the top 2 genes based on log2 fold change.\n\ntop_markers <- cluster_markers %>%\n\ngroup_by(cluster) %>%\n\nslice_max(order_by = avg_log2FC, n = 2)\n\n\n\n# Create a heatmap of the top marker genes.\n\nDoHeatmap(seurat_obj, features = top_markers$gene) + NoLegend()\n\n\n\n\n\n# --- Final Annotation Visualization ---\n\n# Visualize the UMAP with the final, manually assigned cell type identities.\n\nDimPlot(seurat_obj, reduction = \"umap\", label = TRUE, pt.size = 0.8) +\n\nggtitle(\"UMAP Plot: Cell Type Identities\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the fully processed and annotated Seurat object.\n\nsaveRDS(seurat_obj, file = \"Annotated_SeuratObject.rds\")\n\n",
         "The analysis pipeline begins with loading essential R packages for single-cell RNA sequencing (scRNA-seq) analysis, data manipulation, and visualization. The Seurat package is used for core scRNA-seq analysis, dplyr for handling marker gene tables, ggplot2 for creating publication-quality plots, and patchwork for combining multiple ggplot2 plots. These packages are critical for performing the subsequent steps in the analysis.\n\nThe input data consists of 10X Genomics count matrices stored in a specified directory. The raw count matrix is read into R using the `Read10X` function, which processes the data into a format suitable for downstream analysis. A Seurat object is then created from the raw count matrix using the `CreateSeuratObject` function. Initial filtering is applied to retain genes detected in at least 3 cells (`min.cells = 3`) and cells with at least 200 detected genes (`min.features = 200`). This step ensures that low-quality cells and genes are excluded from the analysis.\n\nQuality control (QC) is performed by calculating the percentage of mitochondrial genes for each cell using the `PercentageFeatureSet` function. This metric helps identify cells with potential mitochondrial contamination or poor quality. The data is then normalized using the `NormalizeData` function with the \"LogNormalize\" method and a scale factor of 10,000 to account for differences in library size. Highly variable genes are identified using the `FindVariableFeatures` function with the \"vst\" selection method, retaining the top 2,000 genes for downstream analysis. The data is scaled using the `ScaleData` function, a necessary step before performing principal component analysis (PCA).\n\nDimensionality reduction and clustering are performed to identify distinct cell populations. PCA is run on the highly variable genes using the `RunPCA` function. A K-nearest neighbor (KNN) graph is constructed, and clustering is performed using the `FindNeighbors` and `FindClusters` functions, with a resolution parameter of 0.5 to control the granularity of clustering. Uniform Manifold Approximation and Projection (UMAP) is applied to generate a 2D embedding for visualization using the `RunUMAP` function.\n\nDifferentially expressed genes (cluster biomarkers) are identified for each cluster using the `FindAllMarkers` function. Only positive markers (genes upregulated in the cluster) are retained (`only.pos = TRUE`), with a minimum percentage of cells expressing the gene (`min.pct = 0.25`) and a log-fold change threshold of 0.25. Cell type annotation is performed by manually assigning new cell type identities based on canonical markers observed in the `cluster_markers` table. The `RenameIdents` function is used to update the cluster identities in the Seurat object.\n\nVisualizations are generated to interpret the results. The initial UMAP plot is created using the `DimPlot` function to visualize clusters before cell type assignment. A heatmap of the top marker genes for each cluster is generated using the `DoHeatmap` function, with the top 2 genes selected based on log2 fold change. Finally, the UMAP plot is updated to display the manually assigned cell type identities.\n\nThe fully processed and annotated Seurat object is saved as an RDS file using the `saveRDS` function, ensuring that the results can be easily reloaded for further analysis or sharing. This pipeline provides a comprehensive workflow for analyzing scRNA-seq data, from data preprocessing to cell type annotation and visualization.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis pipeline begins with loading essential R packages for single-cell RNA sequencing (scRNA-seq) analysis, data manipulation, and visualization. The Seurat package is used for core scRNA-seq analysis, dplyr for handling marker gene tables, ggplot2 for creating publication-quality plots, and patchwork for combining multiple ggplot2 plots. These packages are critical for performing the subsequent steps in the analysis.\n\nThe input data consists of 10X Genomics count matrices stored in a specified directory. The raw count matrix is read into R using the `Read10X` function, which processes the data into a format suitable for downstream analysis. A Seurat object is then created from the raw count matrix using the `CreateSeuratObject` function. Initial filtering is applied to retain genes detected in at least 3 cells (`min.cells = 3`) and cells with at least 200 detected genes (`min.features = 200`). This step ensures that low-quality cells and genes are excluded from the analysis.\n\nQuality control (QC) is performed by calculating the percentage of mitochondrial genes for each cell using the `PercentageFeatureSet` function. This metric helps identify cells with potential mitochondrial contamination or poor quality. The data is then normalized using the `NormalizeData` function with the \"LogNormalize\" method and a scale factor of 10,000 to account for differences in library size. Highly variable genes are identified using the `FindVariableFeatures` function with the \"vst\" selection method, retaining the top 2,000 genes for downstream analysis. The data is scaled using the `ScaleData` function, a necessary step before performing principal component analysis (PCA).\n\nDimensionality reduction and clustering are performed to identify distinct cell populations. PCA is run on the highly variable genes using the `RunPCA` function. A K-nearest neighbor (KNN) graph is constructed, and clustering is performed using the `FindNeighbors` and `FindClusters` functions, with a resolution parameter of 0.5 to control the granularity of clustering. Uniform Manifold Approximation and Projection (UMAP) is applied to generate a 2D embedding for visualization using the `RunUMAP` function.\n\nDifferentially expressed genes (cluster biomarkers) are identified for each cluster using the `FindAllMarkers` function. Only positive markers (genes upregulated in the cluster) are retained (`only.pos = TRUE`), with a minimum percentage of cells expressing the gene (`min.pct = 0.25`) and a log-fold change threshold of 0.25. Cell type annotation is performed by manually assigning new cell type identities based on canonical markers observed in the `cluster_markers` table. The `RenameIdents` function is used to update the cluster identities in the Seurat object.\n\nVisualizations are generated to interpret the results. The initial UMAP plot is created using the `DimPlot` function to visualize clusters before cell type assignment. A heatmap of the top marker genes for each cluster is generated using the `DoHeatmap` function, with the top 2 genes selected based on log2 fold change. Finally, the UMAP plot is updated to display the manually assigned cell type identities.\n\nThe fully processed and annotated Seurat object is saved as an RDS file using the `saveRDS` function, ensuring that the results can be easily reloaded for further analysis or sharing. This pipeline provides a comprehensive workflow for analyzing scRNA-seq data, from data preprocessing to cell type annotation and visualization.",
         "The analysis was conducted in R using the Seurat package for single-cell RNA sequencing (scRNA-seq) data processing, alongside dplyr for marker gene table manipulation and ggplot2 for visualization. The workflow began with the integration of 10X Genomics count matrices into a Seurat object, followed by initial filtering to retain high-quality cells and genes. Quality control metrics, including mitochondrial gene percentages, were calculated to identify and exclude low-quality cells. Data normalization was performed using the \"LogNormalize\" method, and highly variable genes were selected for downstream analysis. Dimensionality reduction was achieved through principal component analysis (PCA), followed by clustering using a K-nearest neighbor (KNN) graph and Uniform Manifold Approximation and Projection (UMAP) for visualization. Differentially expressed genes were identified for each cluster, and cell type annotation was performed based on canonical marker genes. Results were visualized through UMAP plots and heatmaps, providing insights into cellular heterogeneity and identity."
        ],
        [
         "19",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis, automatic annotation, and plotting\nimport scanpy as sc\nimport celltypist\nfrom celltypist import models\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\n# Verbosity 2 provides more detailed logging, which is useful for tracking progress.\nsc.settings.verbosity = 2\nsc.settings.set_figure_params(dpi=80, facecolor='white', frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load a preprocessed single-cell dataset (AnnData object).\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"demo_2000_cells.h5ad\",\nbackup_url=\"https://celltypist.cog.sanger.ac.uk/Notebook_demo_data/demo_2000_cells.h5ad\"\n)\nprint(\"Data shape:\", adata.shape)\n\n#--------------------------------\n# Major Analysis Task: Automatic Cell-Type Annotation\n#--------------------------------\n\n# --- Model Preparation ---\n# Download the latest CellTypist models to ensure access to the most recent annotations.\n# `force_update=True` will overwrite existing models if a newer version is available.\nprint(\"Downloading CellTypist models...\")\nmodels.download_models(force_update=True)\n\n# Load a specific pre-trained CellTypist model.\n# \"Immune_All_Low.pkl\" is designed for fine-grained annotation of immune cells.\nmodel = models.Model.load(model=\"Immune_All_Low.pkl\")\n\n# --- Annotation ---\n# Run CellTypist's automatic annotation on the dataset.\n# `majority_voting=True` assigns a final label to each cell based on a consensus\n# prediction from the underlying classification models.\nprint(\"Starting automatic annotation...\")\npredictions = celltypist.annotate(adata, model=model, majority_voting=True)\n\n# Store the prediction results in the AnnData object for easy access.\nadata.obs[\"celltypist_cell_label\"] = predictions.predicted_labels\nadata.obs[\"celltypist_confidence\"] = predictions.confidence\n\n# --- UMAP for Visualization ---\n# Compute the neighborhood graph and UMAP embedding. This is necessary for\n# visualizing the annotation results in a 2D space.\nsc.pp.neighbors(adata, n_neighbors=15, n_pcs=40)\nsc.tl.umap(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, creating two plots side-by-side:\n# 1. Cells colored by their predicted cell type label from CellTypist.\n# 2. Cells colored by the confidence score of the prediction.\nsc.pl.umap(\nadata,\ncolor=[\"celltypist_cell_label\", \"celltypist_confidence\"],\nwspace=0.4,\ntitle=[\"CellTypist: Predicted Labels\", \"Annotation Confidence\"]\n)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the updated AnnData object, which now includes the automatic cell type\n# annotations and confidence scores from CellTypist.\nadata.write(\"celltypist_annotated.h5ad\")\n",
         "To perform automatic cell-type annotation on a single-cell RNA-seq dataset using the CellTypist tool, the following steps were executed. The analysis was conducted in Python using the Scanpy and CellTypist libraries, which are widely used for single-cell data analysis and annotation.\n\nFirst, the necessary Python modules were imported. These included Scanpy for single-cell data manipulation and visualization, CellTypist for automatic cell-type annotation, and Matplotlib for plotting. These libraries are essential for handling single-cell datasets and performing downstream analyses.\n\nThe global parameters for Scanpy were configured to enhance the analysis process. The verbosity level was set to 2 to provide detailed logging, which aids in tracking the progress of the analysis. Additionally, the figure parameters were adjusted to ensure high-quality visualizations with a resolution of 80 DPI, a white background, and no frame around the plots.\n\nThe single-cell dataset was loaded as an AnnData object, a standard format for single-cell data in Scanpy. The dataset, named \"demo_2000_cells.h5ad,\" was either retrieved from a local file or downloaded from a backup URL if the local file was unavailable. The shape of the dataset, indicating the number of cells and genes, was printed to confirm successful loading.\n\nFor the automatic cell-type annotation, the latest CellTypist models were downloaded to ensure access to the most up-to-date annotations. The `force_update=True` parameter ensured that any existing models were overwritten if a newer version was available. A specific pre-trained model, \"Immune_All_Low.pkl,\" was loaded. This model is designed for fine-grained annotation of immune cells, making it suitable for the dataset in question.\n\nThe automatic annotation process was initiated using the `celltypist.annotate` function. The `majority_voting=True` parameter was applied to assign a final cell-type label to each cell based on a consensus prediction from the underlying classification models. The resulting predictions, including the predicted labels and confidence scores, were stored in the AnnData object for easy access.\n\nTo visualize the annotation results, a neighborhood graph was computed, and a UMAP embedding was generated. These steps are necessary for projecting the high-dimensional single-cell data into a two-dimensional space for visualization. The UMAP embedding was then used to create two side-by-side plots: one showing cells colored by their predicted cell-type labels and the other showing cells colored by the confidence scores of the predictions.\n\nFinally, the updated AnnData object, which now includes the automatic cell-type annotations and confidence scores, was saved to a file named \"celltypist_annotated.h5ad.\" This file can be used for further analysis or shared with collaborators.\n\nThis workflow provides a comprehensive approach to automatic cell-type annotation, from data loading and model preparation to annotation, visualization, and data output. The use of established tools like Scanpy and CellTypist ensures reproducibility and reliability in the analysis of single-cell RNA-seq data.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform automatic cell-type annotation on a single-cell RNA-seq dataset using the CellTypist tool, the following steps were executed. The analysis was conducted in Python using the Scanpy and CellTypist libraries, which are widely used for single-cell data analysis and annotation.\n\nFirst, the necessary Python modules were imported. These included Scanpy for single-cell data manipulation and visualization, CellTypist for automatic cell-type annotation, and Matplotlib for plotting. These libraries are essential for handling single-cell datasets and performing downstream analyses.\n\nThe global parameters for Scanpy were configured to enhance the analysis process. The verbosity level was set to 2 to provide detailed logging, which aids in tracking the progress of the analysis. Additionally, the figure parameters were adjusted to ensure high-quality visualizations with a resolution of 80 DPI, a white background, and no frame around the plots.\n\nThe single-cell dataset was loaded as an AnnData object, a standard format for single-cell data in Scanpy. The dataset, named \"demo_2000_cells.h5ad,\" was either retrieved from a local file or downloaded from a backup URL if the local file was unavailable. The shape of the dataset, indicating the number of cells and genes, was printed to confirm successful loading.\n\nFor the automatic cell-type annotation, the latest CellTypist models were downloaded to ensure access to the most up-to-date annotations. The `force_update=True` parameter ensured that any existing models were overwritten if a newer version was available. A specific pre-trained model, \"Immune_All_Low.pkl,\" was loaded. This model is designed for fine-grained annotation of immune cells, making it suitable for the dataset in question.\n\nThe automatic annotation process was initiated using the `celltypist.annotate` function. The `majority_voting=True` parameter was applied to assign a final cell-type label to each cell based on a consensus prediction from the underlying classification models. The resulting predictions, including the predicted labels and confidence scores, were stored in the AnnData object for easy access.\n\nTo visualize the annotation results, a neighborhood graph was computed, and a UMAP embedding was generated. These steps are necessary for projecting the high-dimensional single-cell data into a two-dimensional space for visualization. The UMAP embedding was then used to create two side-by-side plots: one showing cells colored by their predicted cell-type labels and the other showing cells colored by the confidence scores of the predictions.\n\nFinally, the updated AnnData object, which now includes the automatic cell-type annotations and confidence scores, was saved to a file named \"celltypist_annotated.h5ad.\" This file can be used for further analysis or shared with collaborators.\n\nThis workflow provides a comprehensive approach to automatic cell-type annotation, from data loading and model preparation to annotation, visualization, and data output. The use of established tools like Scanpy and CellTypist ensures reproducibility and reliability in the analysis of single-cell RNA-seq data.",
         "The analysis was conducted in Python using Scanpy and CellTypist for automatic cell-type annotation of single-cell RNA-seq data. The workflow involved loading the dataset as an AnnData object and downloading the latest pre-trained CellTypist models, including \"Immune_All_Low.pkl\" for fine-grained immune cell annotation. Automatic annotation was performed using majority voting to assign cell-type labels and confidence scores. Visualization was achieved by computing a neighborhood graph and generating a UMAP embedding, with cells colored by predicted labels and confidence scores. The annotated dataset was saved for downstream analysis, ensuring reproducibility and reliability."
        ],
        [
         "20",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, analysis, and plotting\nimport os\nimport tempfile\nimport matplotlib.pyplot as plt\nimport pandas as pd\nimport scanpy as sc\nimport scvi\nimport seaborn as sns\nimport torch\nfrom scvi.external import CellAssign\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set global parameters for plotting and computation\nsc.set_figure_params(figsize=(6, 6), frameon=False)\nsns.set_theme()\n# Set precision for matrix multiplication in PyTorch for performance.\ntorch.set_float32_matmul_precision(\"high\")\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Create a temporary directory to store downloaded data\nsave_dir = tempfile.TemporaryDirectory()\nprint(f\"Data will be downloaded to temporary directory: {save_dir.name}\")\n\n# Define file paths for the dataset and the marker gene matrix\nadata_path = os.path.join(save_dir.name, \"sce_follicular.h5ad\")\nmarker_path = os.path.join(save_dir.name, \"FL_celltype.csv\")\n\n# Download the follicular lymphoma dataset and its corresponding marker gene file\n# using wget. These files are stored in the temporary directory.\nprint(\"Downloading dataset and marker file...\")\nos.system(f\"wget -q https://ndownloader.figshare.com/files/27458798 -O {adata_path}\")\nos.system(f\"wget -q https://ndownloader.figshare.com/files/27458831 -O {marker_path}\")\n\n# Load the single-cell dataset as an AnnData object\nadata = sc.read(adata_path)\nadata.var_names_make_unique()\nadata.obs_names_make_unique()\n\n# Load the marker gene matrix. Rows should be genes and columns should be cell types.\nmarker_gene_mat = pd.read_csv(marker_path, index_col=0)\n\n#--------------------------------\n# Preprocessing for CellAssign\n#--------------------------------\n\n# Subset the dataset to include only the genes present in the marker matrix.\n# CellAssign only uses these genes for its predictions.\nadata_subset = adata[:, marker_gene_mat.index].copy()\n\n# Calculate per-cell library sizes and create a size factor.\n# This is used by the model to account for differences in sequencing depth.\nlibrary_sizes = adata_subset.X.sum(1)\nadata_subset.obs[\"size_factor\"] = library_sizes / library_sizes.mean()\n\n# Set up the AnnData object for the CellAssign model.\n# This function registers the necessary information for the model.\nscvi.external.CellAssign.setup_anndata(adata_subset, size_factor_key=\"size_factor\")\n\n\n#--------------------------------\n# Major Analysis Task: Annotation with CellAssign\n#--------------------------------\n\n# Initialize the CellAssign model with the subsetted data and the marker gene matrix.\nmodel = CellAssign(adata_subset, marker_gene_mat)\n\n# Train the model. The number of epochs can be adjusted based on convergence.\nprint(\"Training CellAssign model...\")\nmodel.train(max_epochs=400)\n\n# Predict the soft cell type assignment probabilities for each cell.\n# This returns a DataFrame where rows are cells and columns are cell types.\npredictions = model.predict()\n\n# Assign the most likely cell type to each cell in the original AnnData object.\n# The `.idxmax(axis=1)` function finds the cell type with the highest probability for each cell.\nadata.obs['cellassign_prediction'] = predictions.idxmax(axis=1)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# --- Model Convergence Plot ---\n# Plot the validation ELBO (Evidence Lower Bound) to check for model convergence.\n# A stable or plateauing curve indicates that the model has trained sufficiently.\nmodel.history[\"elbo_validation\"].plot()\nplt.xlabel(\"Epoch\")\nplt.ylabel(\"Validation ELBO\")\nplt.title(\"CellAssign Convergence Plot\")\nplt.show()\n\n# --- Prediction Heatmap ---\n# Visualize the raw probability matrix as a heatmap.\n# This shows the confidence of assignment for each cell to each possible cell type.\nplt.figure(figsize=(10, 10))\nsns.heatmap(predictions, cmap=\"viridis\")\nplt.title(\"CellAssign Prediction Probabilities\")\nplt.show()\n\n# --- UMAP Visualization of Final Annotations ---\n# Run standard dimensionality reduction to visualize the final annotations.\nprint(\"Computing UMAP for visualization...\")\nsc.pp.neighbors(adata)\nsc.tl.umap(adata)\nsc.pl.umap(adata, color='cellassign_prediction', title='Final CellAssign Annotations', legend_loc='on data')\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the final, annotated AnnData object.\nadata.write(\"cellassign_annotated.h5ad\")\n\n# Clean up the temporary directory\nsave_dir.cleanup()\n",
         "The analysis begins by importing necessary Python modules for system operations, data handling, analysis, and visualization. These include `os` for file operations, `matplotlib.pyplot` and `seaborn` for plotting, `pandas` for data manipulation, `scanpy` for single-cell data analysis, `scvi` for probabilistic modeling, and `torch` for tensor operations. The `CellAssign` module from `scvi.external` is specifically imported for cell type annotation.\n\nGlobal parameters are configured to optimize the analysis. Plotting settings are adjusted to produce figures of size 6x6 inches without frames, and the Seaborn theme is applied for consistent visualization aesthetics. Additionally, the precision for matrix multiplication in PyTorch is set to \"high\" to enhance computational performance.\n\nA temporary directory is created to store downloaded data. The dataset, consisting of a single-cell RNA-seq dataset in `.h5ad` format and a corresponding marker gene matrix in `.csv` format, is downloaded from a Figshare repository using the `wget` command. The single-cell dataset is loaded into an `AnnData` object, a data structure commonly used in single-cell analysis, and both the gene and cell identifiers are made unique to avoid conflicts. The marker gene matrix is loaded as a pandas DataFrame, with rows representing genes and columns representing cell types.\n\nPreprocessing steps are performed to prepare the data for the CellAssign model. The dataset is subset to include only the genes present in the marker gene matrix, as CellAssign relies on these genes for cell type annotation. Per-cell library sizes are calculated to account for differences in sequencing depth, and size factors are computed by normalizing the library sizes to their mean. The `AnnData` object is then set up for CellAssign by registering the size factor information.\n\nThe CellAssign model is initialized using the subsetted data and the marker gene matrix. The model is trained for a maximum of 400 epochs, with the number of epochs adjustable based on convergence. After training, the model predicts soft cell type assignment probabilities for each cell, resulting in a DataFrame where rows represent cells and columns represent cell types. The most likely cell type for each cell is assigned by selecting the cell type with the highest probability, and these annotations are added to the original `AnnData` object.\n\nVisualizations are generated to assess the model's performance and interpret the results. A convergence plot of the validation ELBO (Evidence Lower Bound) is created to confirm that the model has trained sufficiently. A heatmap of the raw probability matrix is generated to visualize the confidence of cell type assignments. Finally, a UMAP plot is computed to visualize the final cell type annotations in reduced dimensions.\n\nThe annotated `AnnData` object is saved to a file named `cellassign_annotated.h5ad` for further analysis or sharing. The temporary directory used for storing downloaded data is cleaned up to free system resources. This workflow provides a comprehensive pipeline for cell type annotation using the CellAssign model, from data preparation to visualization and output.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing necessary Python modules for system operations, data handling, analysis, and visualization. These include `os` for file operations, `matplotlib.pyplot` and `seaborn` for plotting, `pandas` for data manipulation, `scanpy` for single-cell data analysis, `scvi` for probabilistic modeling, and `torch` for tensor operations. The `CellAssign` module from `scvi.external` is specifically imported for cell type annotation.\n\nGlobal parameters are configured to optimize the analysis. Plotting settings are adjusted to produce figures of size 6x6 inches without frames, and the Seaborn theme is applied for consistent visualization aesthetics. Additionally, the precision for matrix multiplication in PyTorch is set to \"high\" to enhance computational performance.\n\nA temporary directory is created to store downloaded data. The dataset, consisting of a single-cell RNA-seq dataset in `.h5ad` format and a corresponding marker gene matrix in `.csv` format, is downloaded from a Figshare repository using the `wget` command. The single-cell dataset is loaded into an `AnnData` object, a data structure commonly used in single-cell analysis, and both the gene and cell identifiers are made unique to avoid conflicts. The marker gene matrix is loaded as a pandas DataFrame, with rows representing genes and columns representing cell types.\n\nPreprocessing steps are performed to prepare the data for the CellAssign model. The dataset is subset to include only the genes present in the marker gene matrix, as CellAssign relies on these genes for cell type annotation. Per-cell library sizes are calculated to account for differences in sequencing depth, and size factors are computed by normalizing the library sizes to their mean. The `AnnData` object is then set up for CellAssign by registering the size factor information.\n\nThe CellAssign model is initialized using the subsetted data and the marker gene matrix. The model is trained for a maximum of 400 epochs, with the number of epochs adjustable based on convergence. After training, the model predicts soft cell type assignment probabilities for each cell, resulting in a DataFrame where rows represent cells and columns represent cell types. The most likely cell type for each cell is assigned by selecting the cell type with the highest probability, and these annotations are added to the original `AnnData` object.\n\nVisualizations are generated to assess the model's performance and interpret the results. A convergence plot of the validation ELBO (Evidence Lower Bound) is created to confirm that the model has trained sufficiently. A heatmap of the raw probability matrix is generated to visualize the confidence of cell type assignments. Finally, a UMAP plot is computed to visualize the final cell type annotations in reduced dimensions.\n\nThe annotated `AnnData` object is saved to a file named `cellassign_annotated.h5ad` for further analysis or sharing. The temporary directory used for storing downloaded data is cleaned up to free system resources. This workflow provides a comprehensive pipeline for cell type annotation using the CellAssign model, from data preparation to visualization and output.",
         "The analysis was conducted in Python using Scanpy for single-cell RNA-seq data handling and visualization, and scVI for probabilistic modeling. Key steps included data preprocessing, where the dataset was subset to include only genes from a provided marker matrix, and library size normalization was applied. The CellAssign model from scVI was initialized and trained on the subsetted data to predict cell type probabilities. Model performance was assessed through convergence plots and heatmaps of assignment probabilities. Final cell type annotations were visualized using UMAP. The annotated dataset was saved for downstream analysis, providing a robust pipeline for automated cell type annotation."
        ],
        [
         "21",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport bbknn\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 3 # Set to a higher level for more detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load your AnnData object from a file.\n\n# NOTE: Replace \"your_data.h5ad\" with the path to your data.\n\n# This dataset should contain a 'batch' column in adata.obs identifying the batch for each cell.\n\nadata = sc.read(\"your_data.h5ad\")\n\n\n\n#--------------------------------\n\n# Preprocessing & Feature Selection\n\n#--------------------------------\n\n\n\n# --- Batch-Aware Feature Selection ---\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n\n# `batch_key=\"batch\"` ensures that HVGs are selected per batch, which can\n\n# improve the selection process in the presence of strong batch effects.\n\nsc.pp.highly_variable_genes(\n\nadata,\n\nn_top_genes=2000,\n\nflavor=\"cell_ranger\",\n\nbatch_key=\"batch\"\n\n)\n\n# Subset the data to contain only the selected highly variable genes.\n\nadata = adata[:, adata.var.highly_variable].copy()\n\n\n\n# --- Normalization and Scaling ---\n\n# Standard preprocessing steps: normalize, log-transform, and scale the data.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\nsc.pp.scale(adata)\n\n\n\n# --- PCA ---\n\n# Perform Principal Component Analysis (PCA) for dimensionality reduction.\n\n# The resulting PCs will be used as input for BBKNN.\n\nsc.tl.pca(adata, n_comps=50, svd_solver='arpack')\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with BBKNN\n\n#--------------------------------\n\n\n\n# Run Batch-Balanced K-Nearest Neighbors (BBKNN) to correct for batch effects.\n\n# BBKNN identifies nearest neighbors within each batch first before building the\n\n# final, integrated neighborhood graph.\n\n# `neighbors_within_batch` controls how many neighbors are identified within each batch.\n\nbbknn.bbknn(adata, batch_key='batch', neighbors_within_batch=3)\n\n\n\n# --- UMAP for Visualization ---\n\n# Compute the UMAP embedding. Importantly, this is calculated on the BBKNN-corrected\n\n# neighborhood graph, not the original PCA space.\n\nsc.tl.umap(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create a UMAP plot colored by the 'batch' variable.\n\n# Effective integration should result in cells from different batches mixing together,\n\n# rather than clustering separately by batch.\n\nsc.pl.umap(adata, color=['batch'], title=\"UMAP after BBKNN Integration\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object can be used for downstream analyses like clustering and cell type annotation.\n\nadata.write(\"bbknn_integrated_data.h5ad\")\n\n",
         "The analysis begins by importing essential Python libraries for single-cell RNA sequencing (scRNA-seq) data analysis, batch correction, and visualization. These include `numpy` for numerical operations, `scanpy` for single-cell analysis, `bbknn` for batch correction, and `matplotlib.pyplot` for plotting. The `scanpy` library is configured to provide detailed logging and to set plotting parameters, such as resolution (dpi) and background color, ensuring clarity in visual outputs.\n\nThe dataset is loaded as an AnnData object from a specified file path. The dataset must include a 'batch' column in the `adata.obs` metadata, which identifies the batch for each cell. This column is crucial for batch-aware analysis. The data undergoes preprocessing and feature selection to identify highly variable genes (HVGs) using a batch-aware method. The `sc.pp.highly_variable_genes` function is employed with parameters `n_top_genes=2000` to select the top 2000 HVGs, `flavor=\"cell_ranger\"` to use the Cell Ranger method, and `batch_key=\"batch\"` to ensure HVGs are selected per batch. This approach mitigates the impact of batch effects on gene selection. The dataset is then subset to retain only the selected HVGs.\n\nNormalization and scaling are performed to standardize the data. The `sc.pp.normalize_total` function normalizes the data to a total count of 10,000 per cell, followed by log transformation using `sc.pp.log1p` and scaling with `sc.pp.scale`. Principal Component Analysis (PCA) is then conducted for dimensionality reduction, generating 50 principal components (PCs) using the `sc.tl.pca` function with the `svd_solver='arpack'` parameter. These PCs serve as input for subsequent batch correction.\n\nBatch effects are addressed using Batch-Balanced K-Nearest Neighbors (BBKNN). The `bbknn.bbknn` function is applied with `batch_key='batch'` to specify the batch column and `neighbors_within_batch=3` to control the number of neighbors identified within each batch. This method constructs a neighborhood graph that balances batch effects, enabling more accurate downstream analysis. A UMAP embedding is computed on the BBKNN-corrected neighborhood graph using `sc.tl.umap` to visualize the integrated data.\n\nVisualization is performed by generating a UMAP plot colored by the 'batch' variable using `sc.pl.umap`. Effective batch integration is indicated by the mixing of cells from different batches rather than their separation into distinct clusters. Finally, the integrated AnnData object is saved to a file for downstream analyses, such as clustering and cell type annotation, using the `adata.write` function. This step ensures that the results of the batch correction and integration are preserved for further investigation.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for single-cell RNA sequencing (scRNA-seq) data analysis, batch correction, and visualization. These include `numpy` for numerical operations, `scanpy` for single-cell analysis, `bbknn` for batch correction, and `matplotlib.pyplot` for plotting. The `scanpy` library is configured to provide detailed logging and to set plotting parameters, such as resolution (dpi) and background color, ensuring clarity in visual outputs.\n\nThe dataset is loaded as an AnnData object from a specified file path. The dataset must include a 'batch' column in the `adata.obs` metadata, which identifies the batch for each cell. This column is crucial for batch-aware analysis. The data undergoes preprocessing and feature selection to identify highly variable genes (HVGs) using a batch-aware method. The `sc.pp.highly_variable_genes` function is employed with parameters `n_top_genes=2000` to select the top 2000 HVGs, `flavor=\"cell_ranger\"` to use the Cell Ranger method, and `batch_key=\"batch\"` to ensure HVGs are selected per batch. This approach mitigates the impact of batch effects on gene selection. The dataset is then subset to retain only the selected HVGs.\n\nNormalization and scaling are performed to standardize the data. The `sc.pp.normalize_total` function normalizes the data to a total count of 10,000 per cell, followed by log transformation using `sc.pp.log1p` and scaling with `sc.pp.scale`. Principal Component Analysis (PCA) is then conducted for dimensionality reduction, generating 50 principal components (PCs) using the `sc.tl.pca` function with the `svd_solver='arpack'` parameter. These PCs serve as input for subsequent batch correction.\n\nBatch effects are addressed using Batch-Balanced K-Nearest Neighbors (BBKNN). The `bbknn.bbknn` function is applied with `batch_key='batch'` to specify the batch column and `neighbors_within_batch=3` to control the number of neighbors identified within each batch. This method constructs a neighborhood graph that balances batch effects, enabling more accurate downstream analysis. A UMAP embedding is computed on the BBKNN-corrected neighborhood graph using `sc.tl.umap` to visualize the integrated data.\n\nVisualization is performed by generating a UMAP plot colored by the 'batch' variable using `sc.pl.umap`. Effective batch integration is indicated by the mixing of cells from different batches rather than their separation into distinct clusters. Finally, the integrated AnnData object is saved to a file for downstream analyses, such as clustering and cell type annotation, using the `adata.write` function. This step ensures that the results of the batch correction and integration are preserved for further investigation.",
         "The analysis was conducted in Python using the Scanpy library for single-cell RNA sequencing (scRNA-seq) data processing, complemented by BBKNN for batch correction and Matplotlib for visualization. The workflow began with preprocessing and batch-aware feature selection to identify highly variable genes (HVGs), followed by normalization, scaling, and dimensionality reduction via Principal Component Analysis (PCA). Batch effects were mitigated using Batch-Balanced K-Nearest Neighbors (BBKNN), which constructed a neighborhood graph balancing batch-specific effects. A UMAP embedding was computed to visualize the integrated data, ensuring effective batch mixing. The processed data was preserved for downstream clustering and cell type annotation."
        ],
        [
         "22",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport anndata\n\nimport scanpy as sc\n\nimport matplotlib.pyplot as plt\n\nimport numpy as np\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 2 # Set for detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor='white', figsize=(5, 5))\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed reference dataset (e.g., PBMC3k)\n\n# This dataset should ideally already have annotations and a UMAP embedding.\n\nprint(\"Loading reference dataset...\")\n\nadata_ref = sc.datasets.pbmc3k_processed()\n\n\n\n# Load the query dataset that you want to annotate.\n\nprint(\"Loading query dataset...\")\n\nadata = sc.datasets.pbmc68k_reduced()\n\n\n\n# Ensure both datasets share the same gene set by taking the intersection of their genes.\n\n# This is a critical step for mapping annotations accurately.\n\nvar_names = adata_ref.var_names.intersection(adata.var_names)\n\nadata_ref = adata_ref[:, var_names].copy()\n\nadata = adata[:, var_names].copy()\n\nprint(f\"Using {len(var_names)} common genes for integration.\")\n\n\n\n#--------------------------------\n\n# Preprocessing & Embedding on Reference Data\n\n#--------------------------------\n\n\n\n# The following steps ensure the reference data has the necessary components (PCA, neighbors, UMAP).\n\n# These might already be present in a pre-processed object but are run here for completeness.\n\nprint(\"Processing reference dataset...\")\n\nsc.pp.pca(adata_ref, n_comps=50)\n\nsc.pp.neighbors(adata_ref, n_neighbors=15, n_pcs=50)\n\nsc.tl.umap(adata_ref)\n\nsc.tl.louvain(adata_ref) # Example clustering on the reference\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Integrate Query Data via Ingest\n\n#--------------------------------\n\n\n\n# Use sc.tl.ingest to map annotations and embeddings from the reference to the query data.\n\n# 'louvain' labels from the reference will be transferred to the query.\n\n# The UMAP embedding will also be projected onto the query data.\n\nprint(\"Integrating query data onto reference using sc.tl.ingest...\")\n\nsc.tl.ingest(adata, adata_ref, obs='louvain', embedding_method='umap')\n\n\n\n# For consistent visualization, copy the color scheme for the clusters from the reference.\n\nadata.uns['louvain_colors'] = adata_ref.uns['louvain_colors']\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Reference Visualization ---\n\n# Plot the UMAP of the reference data to see the original clusters.\n\nsc.pl.umap(adata_ref, color='louvain', title='Reference UMAP with Louvain Clusters')\n\n\n\n# --- Query Visualization ---\n\n# Plot the UMAP of the query data. The coordinates are projected from the reference,\n\n# and the colors represent the newly ingested labels.\n\nsc.pl.umap(adata, color='louvain', title='Query UMAP with Ingested Annotations', size=30)\n\n\n\n# --- Combined Visualization (Optional) ---\n\n# Concatenate the reference and query objects to visualize them together.\n\n# This is useful for assessing the quality of the integration.\n\nprint(\"Creating a combined UMAP for visualization...\")\n\n# Note: The `ingest` function stores the projected UMAP in `adata.obsm['X_umap_ingest']`.\n\n# We need to rename it to `X_umap` in the query object before concatenating.\n\nadata.obsm['X_umap'] = adata.obsm.pop('X_umap_ingest')\n\nadata_concat = anndata.concat(\n\n[adata_ref, adata],\n\nlabel='batch',\n\nkeys=['reference', 'query'],\n\njoin='outer' # Keep all cells from both objects\n\n)\n\n\n\n# Plot the combined data, coloring by batch and by the transferred labels.\n\nsc.pl.umap(\n\nadata_concat,\n\ncolor=['batch', 'louvain'],\n\ntitle='Combined UMAP: Reference and Query',\n\nwspace=0.5,\n\nsize=20\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the annotated query dataset to a file.\n\nadata.write(\"ingest_annotated_query_data.h5ad\")\n\n\n\n# Save the combined dataset for further inspection.\n\nadata_concat.write(\"ingest_combined_data.h5ad\")\n\n",
         "The following section describes the computational workflow for integrating and annotating single-cell RNA sequencing data using the Scanpy toolkit in Python. The goal is to map annotations and embeddings from a pre-processed reference dataset to a query dataset.\n\nFirst, the necessary Python libraries are imported, including `anndata` for handling single-cell data, `scanpy` for analysis, `matplotlib` for plotting, and `numpy` for numerical operations. Scanpy’s global parameters are configured to enable detailed logging and set figure properties for consistent visualization, including a resolution of 80 DPI, a white background, and a figure size of 5x5 inches.\n\nThe reference dataset, PBMC3k, is loaded, which is pre-processed and includes annotations and a UMAP embedding. The query dataset, PBMC68k, is also loaded. To ensure compatibility, the intersection of genes shared between the two datasets is identified, and both datasets are subset to include only these common genes. This step is critical for accurate annotation transfer.\n\nThe reference dataset is further processed to ensure it contains the necessary components for integration. Principal Component Analysis (PCA) is performed with 50 components, followed by the computation of a nearest-neighbor graph using 15 neighbors and 50 principal components. A UMAP embedding is generated, and Louvain clustering is applied to the reference data to define cell populations.\n\nThe main analysis task involves integrating the query dataset with the reference dataset using the `sc.tl.ingest` function. This function transfers the Louvain cluster labels and UMAP embedding from the reference to the query data. The color scheme for the clusters is also copied from the reference to the query for consistent visualization.\n\nVisualization is performed to assess the results. The UMAP of the reference dataset is plotted to display the original Louvain clusters. The UMAP of the query dataset is plotted to show the newly ingested annotations. Optionally, the reference and query datasets are concatenated to visualize them together. The projected UMAP coordinates from the query dataset are renamed to match the reference format before concatenation. The combined UMAP is plotted, coloring cells by batch (reference or query) and by the transferred Louvain labels.\n\nFinally, the annotated query dataset is saved to a file named `ingest_annotated_query_data.h5ad`, and the combined dataset is saved to `ingest_combined_data.h5ad` for further inspection. This workflow provides a comprehensive approach to integrating and annotating single-cell data using Scanpy, ensuring reproducibility and clarity in the analysis.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following section describes the computational workflow for integrating and annotating single-cell RNA sequencing data using the Scanpy toolkit in Python. The goal is to map annotations and embeddings from a pre-processed reference dataset to a query dataset.\n\nFirst, the necessary Python libraries are imported, including `anndata` for handling single-cell data, `scanpy` for analysis, `matplotlib` for plotting, and `numpy` for numerical operations. Scanpy’s global parameters are configured to enable detailed logging and set figure properties for consistent visualization, including a resolution of 80 DPI, a white background, and a figure size of 5x5 inches.\n\nThe reference dataset, PBMC3k, is loaded, which is pre-processed and includes annotations and a UMAP embedding. The query dataset, PBMC68k, is also loaded. To ensure compatibility, the intersection of genes shared between the two datasets is identified, and both datasets are subset to include only these common genes. This step is critical for accurate annotation transfer.\n\nThe reference dataset is further processed to ensure it contains the necessary components for integration. Principal Component Analysis (PCA) is performed with 50 components, followed by the computation of a nearest-neighbor graph using 15 neighbors and 50 principal components. A UMAP embedding is generated, and Louvain clustering is applied to the reference data to define cell populations.\n\nThe main analysis task involves integrating the query dataset with the reference dataset using the `sc.tl.ingest` function. This function transfers the Louvain cluster labels and UMAP embedding from the reference to the query data. The color scheme for the clusters is also copied from the reference to the query for consistent visualization.\n\nVisualization is performed to assess the results. The UMAP of the reference dataset is plotted to display the original Louvain clusters. The UMAP of the query dataset is plotted to show the newly ingested annotations. Optionally, the reference and query datasets are concatenated to visualize them together. The projected UMAP coordinates from the query dataset are renamed to match the reference format before concatenation. The combined UMAP is plotted, coloring cells by batch (reference or query) and by the transferred Louvain labels.\n\nFinally, the annotated query dataset is saved to a file named `ingest_annotated_query_data.h5ad`, and the combined dataset is saved to `ingest_combined_data.h5ad` for further inspection. This workflow provides a comprehensive approach to integrating and annotating single-cell data using Scanpy, ensuring reproducibility and clarity in the analysis.",
         "The integration and annotation of single-cell RNA sequencing data were conducted using the Scanpy toolkit in Python. The workflow involved loading pre-processed reference (PBMC3k) and query (PBMC68k) datasets, identifying shared genes for compatibility, and performing PCA with 50 components on the reference data. A nearest-neighbor graph was constructed, followed by UMAP embedding and Louvain clustering to define cell populations. The `sc.tl.ingest` function was employed to transfer cluster labels and UMAP coordinates from the reference to the query dataset. Visualization included plotting UMAPs for both datasets, optionally concatenated to assess integration. The annotated query and combined datasets were saved for downstream analysis."
        ],
        [
         "23",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport anndata\n\nimport scanpy as sc\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 2 # Set for detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor='white', figsize=(4, 4))\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed reference dataset (e.g., PBMC3k)\n\n# This dataset should ideally already have annotations and a UMAP embedding.\n\nprint(\"Loading reference dataset...\")\n\nadata_ref = sc.datasets.pbmc3k_processed()\n\n\n\n# Load the query dataset that you want to annotate.\n\nprint(\"Loading query dataset...\")\n\nadata = sc.datasets.pbmc68k_reduced()\n\n\n\n# Ensure both datasets share the same gene set by taking the intersection of their genes.\n\n# This is a critical step for mapping annotations accurately.\n\nvar_names = adata_ref.var_names.intersection(adata.var_names)\n\nadata_ref = adata_ref[:, var_names].copy()\n\nadata = adata[:, var_names].copy()\n\nprint(f\"Using {len(var_names)} common genes for integration.\")\n\n\n\n#--------------------------------\n\n# Preprocessing & Embedding on Reference Data\n\n#--------------------------------\n\n\n\n# The following steps ensure the reference data has the necessary components (PCA, neighbors, UMAP).\n\n# These might already be present in a pre-processed object.\n\nprint(\"Processing reference dataset...\")\n\nsc.pp.pca(adata_ref, n_comps=50)\n\nsc.pp.neighbors(adata_ref, n_neighbors=15, n_pcs=50)\n\nsc.tl.umap(adata_ref)\n\nsc.tl.louvain(adata_ref) # Example clustering on the reference\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Integrate Query Data via Ingest\n\n#--------------------------------\n\n\n\n# Use sc.tl.ingest to map annotations and embeddings from the reference to the query data.\n\n# 'louvain' labels from the reference will be transferred to the query.\n\n# The UMAP embedding will also be projected onto the query data.\n\nprint(\"Integrating query data onto reference using sc.tl.ingest...\")\n\nsc.tl.ingest(adata, adata_ref, obs='louvain')\n\n\n\n# For consistent visualization, copy the color scheme for the clusters from the reference.\n\nadata.uns['louvain_colors'] = adata_ref.uns['louvain_colors']\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Reference Visualization ---\n\n# Plot the UMAP of the reference data to see the original clusters.\n\nsc.pl.umap(adata_ref, color='louvain', title='Reference UMAP with Louvain Clusters')\n\n\n\n# --- Query Visualization ---\n\n# Plot the UMAP of the query data. The coordinates are projected from the reference,\n\n# and the colors represent the newly ingested labels.\n\nsc.pl.umap(adata, color='louvain', title='Query UMAP with Ingested Annotations', size=30)\n\n\n\n# --- Combined Visualization (Optional) ---\n\n# Concatenate the reference and query objects to visualize them together.\n\n# This is useful for assessing the quality of the integration.\n\nprint(\"Creating a combined UMAP for visualization...\")\n\nadata_concat = anndata.concat(\n\n[adata_ref, adata],\n\nlabel='batch',\n\nkeys=['reference', 'query']\n\n)\n\n# The UMAP coordinates for the query part are from the ingest function.\n\n# We re-compute the reference UMAP to ensure it's in the same object for plotting.\n\nadata_concat.obsm['X_umap'] = np.concatenate(\n\n[adata_ref.obsm['X_umap'], adata.obsm['X_umap_ingest']]\n\n)\n\n\n\n# Plot the combined data, coloring by batch and by the transferred labels.\n\nsc.pl.umap(\n\nadata_concat,\n\ncolor=['batch', 'louvain'],\n\ntitle='Combined UMAP: Reference and Query',\n\nwspace=0.5,\n\nsize=20\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the annotated query dataset to a file.\n\nadata.write(\"ingest_annotated_query_data.h5ad\")\n\n",
         "The analysis begins by importing essential libraries for single-cell RNA sequencing (scRNA-seq) data analysis and visualization. The libraries include `anndata` for handling annotated data matrices, `scanpy` (abbreviated as `sc`) for single-cell analysis, and `matplotlib.pyplot` for generating plots. These tools are critical for processing, analyzing, and visualizing scRNA-seq datasets.\n\nGlobal parameters for `scanpy` are configured to enable detailed logging and set plotting preferences. The verbosity level is set to 2 to provide comprehensive logging during the analysis. Additionally, figure parameters are adjusted to ensure high-quality visualizations with a resolution of 80 DPI, a white background, and a default figure size of 4x4 inches.\n\nThe reference dataset, `pbmc3k_processed`, is loaded, which is a pre-processed dataset containing annotations and a UMAP embedding. This dataset serves as the basis for annotating the query dataset. The query dataset, `pbmc68k_reduced`, is also loaded. To ensure compatibility between the two datasets, the intersection of their gene sets is computed, and both datasets are subset to include only these common genes. This step is crucial for accurate annotation transfer and integration.\n\nThe reference dataset is further processed to ensure it contains the necessary components for analysis, including principal component analysis (PCA), nearest neighbors, and UMAP embeddings. PCA is performed with 50 components, followed by the computation of nearest neighbors using 15 neighbors and 50 principal components. UMAP is then applied to reduce the dimensionality of the data, and Louvain clustering is performed to identify cell clusters. These steps prepare the reference dataset for annotation transfer.\n\nThe major analysis task involves integrating the query dataset with the reference dataset using the `sc.tl.ingest` function. This function maps the Louvain cluster labels and UMAP embeddings from the reference dataset to the query dataset. The color scheme for the clusters is also copied from the reference to ensure consistent visualization.\n\nVisualizations are generated to assess the results of the integration. The UMAP plot of the reference dataset displays the original Louvain clusters. The UMAP plot of the query dataset shows the projected coordinates and the ingested cluster labels. Optionally, the reference and query datasets are concatenated to create a combined UMAP plot, which is useful for evaluating the quality of the integration. The combined plot is colored by batch (reference vs. query) and by the transferred Louvain labels.\n\nFinally, the annotated query dataset is saved to a file named `ingest_annotated_query_data.h5ad` for further analysis or sharing. This step ensures that the results of the integration process are preserved and can be reused in subsequent studies.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential libraries for single-cell RNA sequencing (scRNA-seq) data analysis and visualization. The libraries include `anndata` for handling annotated data matrices, `scanpy` (abbreviated as `sc`) for single-cell analysis, and `matplotlib.pyplot` for generating plots. These tools are critical for processing, analyzing, and visualizing scRNA-seq datasets.\n\nGlobal parameters for `scanpy` are configured to enable detailed logging and set plotting preferences. The verbosity level is set to 2 to provide comprehensive logging during the analysis. Additionally, figure parameters are adjusted to ensure high-quality visualizations with a resolution of 80 DPI, a white background, and a default figure size of 4x4 inches.\n\nThe reference dataset, `pbmc3k_processed`, is loaded, which is a pre-processed dataset containing annotations and a UMAP embedding. This dataset serves as the basis for annotating the query dataset. The query dataset, `pbmc68k_reduced`, is also loaded. To ensure compatibility between the two datasets, the intersection of their gene sets is computed, and both datasets are subset to include only these common genes. This step is crucial for accurate annotation transfer and integration.\n\nThe reference dataset is further processed to ensure it contains the necessary components for analysis, including principal component analysis (PCA), nearest neighbors, and UMAP embeddings. PCA is performed with 50 components, followed by the computation of nearest neighbors using 15 neighbors and 50 principal components. UMAP is then applied to reduce the dimensionality of the data, and Louvain clustering is performed to identify cell clusters. These steps prepare the reference dataset for annotation transfer.\n\nThe major analysis task involves integrating the query dataset with the reference dataset using the `sc.tl.ingest` function. This function maps the Louvain cluster labels and UMAP embeddings from the reference dataset to the query dataset. The color scheme for the clusters is also copied from the reference to ensure consistent visualization.\n\nVisualizations are generated to assess the results of the integration. The UMAP plot of the reference dataset displays the original Louvain clusters. The UMAP plot of the query dataset shows the projected coordinates and the ingested cluster labels. Optionally, the reference and query datasets are concatenated to create a combined UMAP plot, which is useful for evaluating the quality of the integration. The combined plot is colored by batch (reference vs. query) and by the transferred Louvain labels.\n\nFinally, the annotated query dataset is saved to a file named `ingest_annotated_query_data.h5ad` for further analysis or sharing. This step ensures that the results of the integration process are preserved and can be reused in subsequent studies.",
         "The analysis was conducted using Python, leveraging the `anndata` library for data handling and `scanpy` for single-cell RNA sequencing (scRNA-seq) analysis. Visualization was performed with `matplotlib.pyplot`. The workflow involved loading pre-processed reference (`pbmc3k_processed`) and query (`pbmc68k_reduced`) datasets, ensuring compatibility by subsetting to common genes. The reference dataset was processed with PCA, nearest neighbor computation, UMAP dimensionality reduction, and Louvain clustering to prepare for annotation transfer. Integration was achieved using `sc.tl.ingest`, mapping reference cluster labels and UMAP embeddings to the query dataset. Visualizations included UMAP plots of both datasets and a combined plot to assess integration quality. The annotated query dataset was saved for downstream use."
        ],
        [
         "24",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, analysis, and plotting\nimport os\nimport tempfile\nimport scanpy as sc\nimport scvi\nimport seaborn as sns\nimport torch\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set global parameters for plotting and computation\nsc.set_figure_params(figsize=(6, 6), frameon=False)\nsns.set_theme()\n# Set precision for matrix multiplication in PyTorch for performance.\ntorch.set_float32_matmul_precision(\"high\")\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Create a temporary directory to store downloaded data\nsave_dir = tempfile.TemporaryDirectory()\nprint(f\"Data will be downloaded to temporary directory: {save_dir.name}\")\n\n# Define file path for the dataset\nadata_path = os.path.join(save_dir.name, \"lung_atlas.h5ad\")\n\n# Download and load the lung atlas dataset.\n# The `backup_url` will automatically download the file if not found locally.\nprint(\"Downloading and loading dataset...\")\nadata = sc.read(adata_path, backup_url=\"https://figshare.com/ndownloader/files/24539942\")\nprint(adata)\n\n# Store the full, unprocessed dataset in the .raw attribute for reference.\nadata.raw = adata\n\n#--------------------------------\n# Preprocessing & Feature Selection\n#--------------------------------\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n# This approach finds genes that are variable across batches, which is robust\n# to batch-specific effects. The `subset=True` argument modifies adata in place.\nprint(\"Selecting highly variable genes...\")\nsc.pp.highly_variable_genes(\nadata,\nflavor=\"seurat_v3\",\nn_top_genes=2000,\nlayer=\"counts\",\nbatch_key=\"batch\",\nsubset=True,\n)\n\n#--------------------------------\n# Major Analysis Task: Batch Integration with scVI\n#--------------------------------\n\n# --- Model Setup and Training ---\n# Register the AnnData object for scVI analysis. This function specifies\n# the data layer, batch information, and other covariates for the model.\nscvi.model.SCVI.setup_anndata(adata, layer=\"counts\", batch_key=\"batch\")\n\n# Initialize the scVI model.\n# `n_layers`: number of hidden layers in the neural network.\n# `n_latent`: dimensionality of the latent space.\n# `gene_likelihood`: \"nb\" (Negative Binomial) is recommended for UMI-based count data.\nmodel = scvi.model.SCVI(adata, n_layers=2, n_latent=30, gene_likelihood=\"nb\")\n\n# Train the scVI model. This learns a batch-corrected latent representation of the data.\nprint(\"Training scVI model...\")\nmodel.train()\n\n# --- Post-Integration Steps ---\n# Retrieve the learned latent representation from the trained model.\nSCVI_LATENT_KEY = \"X_scVI\"\nadata.obsm[SCVI_LATENT_KEY] = model.get_latent_representation()\n\n# Compute the nearest-neighbor graph on the scVI latent space.\n# This graph will be used for downstream clustering and visualization.\nsc.pp.neighbors(adata, use_rep=SCVI_LATENT_KEY)\n\n# Compute the UMAP embedding based on the integrated neighbor graph.\nsc.tl.umap(adata)\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, coloring cells by their batch ID.\n# Effective integration should show cells from different batches mixing together,\n# rather than forming separate, batch-specific clusters.\nprint(\"Visualizing integrated data...\")\nsc.pl.umap(adata, color=\"batch\", title=\"UMAP after scVI Integration\")\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the integrated AnnData object to a file.\nprint(\"Saving integrated data object...\")\nadata.write(\"scvi_integrated_data.h5ad\")\n\n# Clean up the temporary directory\nsave_dir.cleanup()\n",
         "The analysis begins by importing necessary Python modules for system operations, data handling, analysis, and visualization. These include `os` for file path management, `tempfile` for creating temporary directories, `scanpy` (imported as `sc`) for single-cell RNA-seq data analysis, `scvi` for batch integration, `seaborn` (imported as `sns`) for enhanced plotting, and `torch` for tensor operations. Global parameters for plotting and computation are then configured. The figure size is set to 6x6 inches with no frame, and the Seaborn theme is applied for improved aesthetics. Additionally, the precision for matrix multiplication in PyTorch is set to \"high\" to optimize performance.\n\nA temporary directory is created to store the downloaded dataset. The dataset, a lung atlas in `.h5ad` format, is downloaded from a specified URL if not already present locally. The dataset is loaded into an `AnnData` object, a data structure commonly used in single-cell analysis. The unprocessed dataset is stored in the `.raw` attribute of the `AnnData` object for reference during later stages of the analysis.\n\nPreprocessing begins with the identification of highly variable genes (HVGs) using a batch-aware method. This step is crucial for reducing noise and focusing on genes that exhibit significant variability across batches. The `seurat_v3` flavor is used for HVG selection, with the top 2000 genes selected based on their variability. The `batch_key` parameter ensures that batch-specific effects are accounted for during this process. The `subset=True` argument modifies the `AnnData` object in place, retaining only the selected HVGs.\n\nThe core analysis involves batch integration using the scVI (single-cell Variational Inference) model. The `AnnData` object is registered for scVI analysis, specifying the data layer and batch information. The scVI model is initialized with two hidden layers and a 30-dimensional latent space. The gene likelihood is set to \"Negative Binomial\" (`nb`), which is recommended for UMI-based count data. The model is then trained to learn a batch-corrected latent representation of the data. After training, the latent representation is retrieved and stored in the `AnnData` object under the key `X_scVI`. A nearest-neighbor graph is computed on this latent representation, which is subsequently used for downstream clustering and visualization. A UMAP embedding is generated to visualize the integrated data in two dimensions.\n\nVisualization is performed by plotting the UMAP embedding, with cells colored by their batch ID. Effective batch integration is indicated by the mixing of cells from different batches rather than the formation of batch-specific clusters. Finally, the integrated `AnnData` object is saved to a file named `scvi_integrated_data.h5ad`, and the temporary directory is cleaned up to free up system resources. This completes the analysis, providing a batch-corrected representation of the lung atlas dataset ready for further downstream exploration.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing necessary Python modules for system operations, data handling, analysis, and visualization. These include `os` for file path management, `tempfile` for creating temporary directories, `scanpy` (imported as `sc`) for single-cell RNA-seq data analysis, `scvi` for batch integration, `seaborn` (imported as `sns`) for enhanced plotting, and `torch` for tensor operations. Global parameters for plotting and computation are then configured. The figure size is set to 6x6 inches with no frame, and the Seaborn theme is applied for improved aesthetics. Additionally, the precision for matrix multiplication in PyTorch is set to \"high\" to optimize performance.\n\nA temporary directory is created to store the downloaded dataset. The dataset, a lung atlas in `.h5ad` format, is downloaded from a specified URL if not already present locally. The dataset is loaded into an `AnnData` object, a data structure commonly used in single-cell analysis. The unprocessed dataset is stored in the `.raw` attribute of the `AnnData` object for reference during later stages of the analysis.\n\nPreprocessing begins with the identification of highly variable genes (HVGs) using a batch-aware method. This step is crucial for reducing noise and focusing on genes that exhibit significant variability across batches. The `seurat_v3` flavor is used for HVG selection, with the top 2000 genes selected based on their variability. The `batch_key` parameter ensures that batch-specific effects are accounted for during this process. The `subset=True` argument modifies the `AnnData` object in place, retaining only the selected HVGs.\n\nThe core analysis involves batch integration using the scVI (single-cell Variational Inference) model. The `AnnData` object is registered for scVI analysis, specifying the data layer and batch information. The scVI model is initialized with two hidden layers and a 30-dimensional latent space. The gene likelihood is set to \"Negative Binomial\" (`nb`), which is recommended for UMI-based count data. The model is then trained to learn a batch-corrected latent representation of the data. After training, the latent representation is retrieved and stored in the `AnnData` object under the key `X_scVI`. A nearest-neighbor graph is computed on this latent representation, which is subsequently used for downstream clustering and visualization. A UMAP embedding is generated to visualize the integrated data in two dimensions.\n\nVisualization is performed by plotting the UMAP embedding, with cells colored by their batch ID. Effective batch integration is indicated by the mixing of cells from different batches rather than the formation of batch-specific clusters. Finally, the integrated `AnnData` object is saved to a file named `scvi_integrated_data.h5ad`, and the temporary directory is cleaned up to free up system resources. This completes the analysis, providing a batch-corrected representation of the lung atlas dataset ready for further downstream exploration.",
         "The analysis was conducted in Python using Scanpy and scVI for single-cell RNA sequencing data processing and batch integration. Highly variable genes (HVGs) were identified using a batch-aware method with the Seurat v3 flavor, retaining the top 2000 genes. Batch integration was performed using the scVI model, which learned a 30-dimensional latent representation of the data with a Negative Binomial gene likelihood. A nearest-neighbor graph was constructed from the latent space, and a UMAP embedding was generated for visualization. Effective batch correction was confirmed by the mixing of cells from different batches in the UMAP plot. The integrated dataset was saved for downstream analysis."
        ],
        [
         "25",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, analysis, and plotting\nimport os\nimport tempfile\nimport scanpy as sc\nimport scvi\nimport seaborn as sns\nimport torch\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set global parameters for plotting and computation\nsc.set_figure_params(figsize=(6, 6), frameon=False)\nsns.set_theme()\n# Set precision for matrix multiplication in PyTorch for performance.\ntorch.set_float32_matmul_precision(\"high\")\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Create a temporary directory to store the downloaded dataset\nsave_dir = tempfile.TemporaryDirectory()\nprint(f\"Data will be downloaded to temporary directory: {save_dir.name}\")\n\n# Define file path for the dataset\nadata_path = os.path.join(save_dir.name, \"lung_atlas.h5ad\")\n\n# Download and load the lung atlas dataset.\n# The `backup_url` will automatically download the file if not found locally.\nprint(\"Downloading and loading dataset...\")\nadata = sc.read(adata_path, backup_url=\"https://figshare.com/ndownloader/files/24539942\")\nprint(adata)\n\n# Store the full, unprocessed dataset in the .raw attribute for reference.\nadata.raw = adata\n\n#--------------------------------\n# Preprocessing & Feature Selection\n#--------------------------------\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n# This helps focus on biological signal and reduces technical noise.\n# The `subset=True` argument modifies adata in place to keep only HVGs.\nprint(\"Selecting highly variable genes...\")\nsc.pp.highly_variable_genes(\nadata,\nflavor=\"seurat_v3\",\nn_top_genes=2000,\nlayer=\"counts\",\nbatch_key=\"batch\",\nsubset=True,\n)\n\n#--------------------------------\n# Major Analysis Task: Semi-Supervised Integration with scANVI\n#--------------------------------\n\n# --- Model Setup and Training ---\n# Register the AnnData object for scANVI analysis. This function specifies\n# the data layer, batch information, and the key for cell type labels.\nscvi.model.SCANVI.setup_anndata(adata, labels_key=\"cell_type\", batch_key=\"batch\")\n\n# Initialize the scANVI model. scANVI extends scVI for semi-supervised tasks.\n# It uses both labeled and unlabeled data for training.\n# `unlabeled_category=\"Unknown\"` tells the model which label to treat as unlabeled.\nscanvi_model = scvi.model.SCANVI(adata, unlabeled_category=\"Unknown\")\n\n# Train the scANVI model. This refines the latent space using the cell type labels.\nprint(\"Training scANVI model...\")\nscanvi_model.train(max_epochs=20, n_samples_per_label=100)\n\n# --- Post-Integration Steps ---\n# Retrieve the learned latent representation from the trained model.\nSCANVI_LATENT_KEY = \"X_scANVI\"\nadata.obsm[SCANVI_LATENT_KEY] = scanvi_model.get_latent_representation()\n\n# Compute the nearest-neighbor graph on the scANVI latent space.\n# This integrated graph is used for downstream visualization.\nsc.pp.neighbors(adata, use_rep=SCANVI_LATENT_KEY)\n\n# Compute the UMAP embedding based on the integrated neighbor graph.\nsc.tl.umap(adata, min_dist=0.3)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, coloring cells by their annotated cell type.\n# This plot shows how well scANVI has integrated the batches while preserving cell identities.\nprint(\"Visualizing integrated data...\")\nsc.pl.umap(adata, color=[\"cell_type\"], frameon=False, ncols=1)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the integrated and annotated AnnData object to a file.\nprint(\"Saving integrated data object...\")\nadata.write(\"scanvi_integrated_data.h5ad\")\n\n# Clean up the temporary directory\nsave_dir.cleanup()\n",
         "To perform semi-supervised integration of single-cell RNA sequencing data using the scANVI framework, the following steps were executed. The analysis was conducted in Python, utilizing the Scanpy and scvi-tools libraries for data handling, preprocessing, and integration.\n\nFirst, the necessary Python modules were imported. These included libraries for system operations (os, tempfile), single-cell data analysis (scanpy as sc, scvi), visualization (seaborn as sns, matplotlib.pyplot as plt), and numerical computation (torch). Global parameters were configured for plotting and computational performance, ensuring consistent figure sizes and optimized matrix multiplication precision in PyTorch.\n\nA temporary directory was created to store the downloaded dataset. The dataset, a lung atlas in the h5ad format, was downloaded from a specified URL if not already available locally. The dataset was loaded into an AnnData object, a standard data structure for single-cell analysis, and the unprocessed data was stored in the .raw attribute for reference.\n\nPreprocessing began with the identification of highly variable genes (HVGs) using a batch-aware method. This step focused on biological signals while reducing technical noise. The top 2000 HVGs were selected based on the Seurat v3 flavor, and the AnnData object was subset to retain only these genes.\n\nThe primary analysis task involved semi-supervised integration using the scANVI model. The AnnData object was registered for scANVI analysis, specifying the data layer, batch information, and cell type labels. The scANVI model was initialized, treating cells labeled as \"Unknown\" as unlabeled data. The model was trained for 20 epochs, with 100 samples per label, to refine the latent space representation.\n\nPost-integration, the learned latent representation was retrieved and stored in the AnnData object. A nearest-neighbor graph was computed on this latent space, and a UMAP embedding was generated with a minimum distance parameter of 0.3 to visualize the integrated data. The UMAP plot was colored by annotated cell types to assess the integration quality.\n\nFinally, the integrated and annotated AnnData object was saved to a file for downstream analysis. The temporary directory used for dataset storage was cleaned up to free system resources. This workflow provides a comprehensive approach to integrating single-cell datasets while leveraging both labeled and unlabeled data for improved biological insights.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform semi-supervised integration of single-cell RNA sequencing data using the scANVI framework, the following steps were executed. The analysis was conducted in Python, utilizing the Scanpy and scvi-tools libraries for data handling, preprocessing, and integration.\n\nFirst, the necessary Python modules were imported. These included libraries for system operations (os, tempfile), single-cell data analysis (scanpy as sc, scvi), visualization (seaborn as sns, matplotlib.pyplot as plt), and numerical computation (torch). Global parameters were configured for plotting and computational performance, ensuring consistent figure sizes and optimized matrix multiplication precision in PyTorch.\n\nA temporary directory was created to store the downloaded dataset. The dataset, a lung atlas in the h5ad format, was downloaded from a specified URL if not already available locally. The dataset was loaded into an AnnData object, a standard data structure for single-cell analysis, and the unprocessed data was stored in the .raw attribute for reference.\n\nPreprocessing began with the identification of highly variable genes (HVGs) using a batch-aware method. This step focused on biological signals while reducing technical noise. The top 2000 HVGs were selected based on the Seurat v3 flavor, and the AnnData object was subset to retain only these genes.\n\nThe primary analysis task involved semi-supervised integration using the scANVI model. The AnnData object was registered for scANVI analysis, specifying the data layer, batch information, and cell type labels. The scANVI model was initialized, treating cells labeled as \"Unknown\" as unlabeled data. The model was trained for 20 epochs, with 100 samples per label, to refine the latent space representation.\n\nPost-integration, the learned latent representation was retrieved and stored in the AnnData object. A nearest-neighbor graph was computed on this latent space, and a UMAP embedding was generated with a minimum distance parameter of 0.3 to visualize the integrated data. The UMAP plot was colored by annotated cell types to assess the integration quality.\n\nFinally, the integrated and annotated AnnData object was saved to a file for downstream analysis. The temporary directory used for dataset storage was cleaned up to free system resources. This workflow provides a comprehensive approach to integrating single-cell datasets while leveraging both labeled and unlabeled data for improved biological insights.",
         "The analysis employed Python with the Scanpy and scvi-tools libraries to perform semi-supervised integration of single-cell RNA sequencing data using the scANVI framework. Highly variable genes (HVGs) were identified batch-aware, retaining the top 2000 genes for downstream analysis. The scANVI model was initialized, integrating labeled and unlabeled cells by refining the latent space representation through training. Post-integration, a nearest-neighbor graph and UMAP embedding were generated to visualize the integrated data, colored by annotated cell types. This approach effectively combined labeled and unlabeled data, enhancing biological insights into cellular heterogeneity."
        ],
        [
         "26",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport scanpy as sc\n\nimport bbknn\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters and logging verbosity for clear output\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\nsc.settings.verbosity = 2\n\n\n\n#--------------------------------\n\n# Data Input & Simulation\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC3k dataset as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Simulate two batches by splitting the dataset in half.\n\n# This is for demonstration purposes; in a real scenario, the 'batch'\n\n# column would come from the experimental design.\n\nprint(\"Simulating two batches for demonstration...\")\n\nnum_cells = adata.n_obs\n\nbatch_labels = np.array([\"batch1\"] * num_cells)\n\nbatch_labels[int(num_cells / 2):] = \"batch2\"\n\nadata.obs[\"batch\"] = batch_labels\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization and Feature Selection ---\n\n# Normalize each cell to a target count of 10,000 and apply a log transformation.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) to focus on biological variation.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\nadata = adata[:, adata.var.highly_variable].copy()\n\n\n\n# --- Scaling and PCA ---\n\n# Scale the data to have zero mean and unit variance.\n\nsc.pp.scale(adata, max_value=10)\n\n\n\n# Perform Principal Component Analysis (PCA) for dimensionality reduction.\n\n# The PCs will be used as input for the BBKNN algorithm.\n\nsc.tl.pca(adata, svd_solver='arpack', n_comps=50)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with BBKNN\n\n#--------------------------------\n\n\n\n# --- Batch Integration ---\n\n# Run Batch-Balanced K-Nearest Neighbors (BBKNN) to correct for batch effects.\n\n# This function computes a neighborhood graph by finding neighbors within each\n\n# batch separately before merging them into a single, integrated graph.\n\nprint(\"\\nRunning BBKNN for batch integration...\")\n\nbbknn.bbknn(adata, batch_key='batch', n_pcs=50, neighbors_within_batch=3)\n\n\n\n# --- Post-Integration Embeddings ---\n\n# Compute UMAP and t-SNE embeddings based on the BBKNN-corrected neighborhood graph.\n\n# These will be used for visualizing the integrated data.\n\nsc.tl.umap(adata)\n\nsc.tl.tsne(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the UMAP embedding, coloring cells by their batch label.\n\n# Successful integration should show cells from 'batch1' and 'batch2' mixing well.\n\nsc.pl.umap(adata, color=[\"batch\"], title=\"UMAP: BBKNN Integrated\", size=30)\n\n\n\n# Plot the t-SNE embedding, also colored by batch, as an alternative visualization.\n\nsc.pl.tsne(adata, color=[\"batch\"], title=\"t-SNE: BBKNN Integrated\", size=30)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object can now be used for downstream analysis like clustering.\n\nprint(\"\\nSaving integrated data object...\")\n\nadata.write(\"pbmc3k_bbknn_integrated.h5ad\")\n\n",
         "To analyze single-cell RNA sequencing data and correct for batch effects, the following workflow was implemented using Python and the Scanpy library. The analysis begins with data loading and preprocessing, followed by dimensionality reduction, batch integration, and visualization.\n\nFirst, the necessary libraries were imported. Scanpy was used for single-cell analysis, BBKNN for batch correction, NumPy for numerical operations, and Matplotlib for plotting. Global plotting parameters were configured to ensure clear and consistent visual output, with a DPI of 80 and a white background. The verbosity level was set to 2 to provide detailed logging during the analysis.\n\nThe PBMC3k dataset, a publicly available single-cell RNA-seq dataset, was loaded as an example. To simulate batch effects for demonstration purposes, the dataset was split into two equal halves, labeled as \"batch1\" and \"batch2.\" This step mimics the presence of technical or experimental batches, which are common in real-world single-cell datasets.\n\nThe preprocessing steps included normalization and feature selection. Each cell was normalized to a target count of 10,000, followed by a log transformation to stabilize variance. Highly variable genes (HVGs) were identified using the Seurat flavor, with the top 2,000 genes selected to focus on biological variation. The data was then scaled to have zero mean and unit variance, ensuring that all genes contributed equally to downstream analyses. Principal Component Analysis (PCA) was performed for dimensionality reduction, retaining the top 50 principal components (PCs) for subsequent steps.\n\nBatch integration was achieved using the Batch-Balanced K-Nearest Neighbors (BBKNN) algorithm. BBKNN constructs a neighborhood graph by identifying nearest neighbors within each batch separately before merging them into a single, integrated graph. This approach corrects for batch effects while preserving biological variation. The algorithm was configured to use the 50 PCs computed earlier and to find 3 nearest neighbors within each batch.\n\nFollowing batch integration, Uniform Manifold Approximation and Projection (UMAP) and t-Distributed Stochastic Neighbor Embedding (t-SNE) embeddings were computed to visualize the integrated data. These embeddings provide low-dimensional representations of the dataset, facilitating the assessment of batch effect correction. Cells were colored by their batch labels to evaluate the success of integration, with well-mixed batches indicating effective correction.\n\nThe UMAP and t-SNE plots were generated to visualize the integrated data, with cells colored by their batch labels. Successful integration was confirmed by the intermingling of cells from both batches in the embeddings. Finally, the integrated AnnData object was saved to a file for downstream analyses, such as clustering or differential expression analysis. The file was saved in the H5AD format, which is commonly used for storing single-cell datasets.\n\nThis workflow provides a comprehensive approach to single-cell RNA-seq data analysis, including batch effect correction and visualization, and can be replicated using the described steps and parameters.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo analyze single-cell RNA sequencing data and correct for batch effects, the following workflow was implemented using Python and the Scanpy library. The analysis begins with data loading and preprocessing, followed by dimensionality reduction, batch integration, and visualization.\n\nFirst, the necessary libraries were imported. Scanpy was used for single-cell analysis, BBKNN for batch correction, NumPy for numerical operations, and Matplotlib for plotting. Global plotting parameters were configured to ensure clear and consistent visual output, with a DPI of 80 and a white background. The verbosity level was set to 2 to provide detailed logging during the analysis.\n\nThe PBMC3k dataset, a publicly available single-cell RNA-seq dataset, was loaded as an example. To simulate batch effects for demonstration purposes, the dataset was split into two equal halves, labeled as \"batch1\" and \"batch2.\" This step mimics the presence of technical or experimental batches, which are common in real-world single-cell datasets.\n\nThe preprocessing steps included normalization and feature selection. Each cell was normalized to a target count of 10,000, followed by a log transformation to stabilize variance. Highly variable genes (HVGs) were identified using the Seurat flavor, with the top 2,000 genes selected to focus on biological variation. The data was then scaled to have zero mean and unit variance, ensuring that all genes contributed equally to downstream analyses. Principal Component Analysis (PCA) was performed for dimensionality reduction, retaining the top 50 principal components (PCs) for subsequent steps.\n\nBatch integration was achieved using the Batch-Balanced K-Nearest Neighbors (BBKNN) algorithm. BBKNN constructs a neighborhood graph by identifying nearest neighbors within each batch separately before merging them into a single, integrated graph. This approach corrects for batch effects while preserving biological variation. The algorithm was configured to use the 50 PCs computed earlier and to find 3 nearest neighbors within each batch.\n\nFollowing batch integration, Uniform Manifold Approximation and Projection (UMAP) and t-Distributed Stochastic Neighbor Embedding (t-SNE) embeddings were computed to visualize the integrated data. These embeddings provide low-dimensional representations of the dataset, facilitating the assessment of batch effect correction. Cells were colored by their batch labels to evaluate the success of integration, with well-mixed batches indicating effective correction.\n\nThe UMAP and t-SNE plots were generated to visualize the integrated data, with cells colored by their batch labels. Successful integration was confirmed by the intermingling of cells from both batches in the embeddings. Finally, the integrated AnnData object was saved to a file for downstream analyses, such as clustering or differential expression analysis. The file was saved in the H5AD format, which is commonly used for storing single-cell datasets.\n\nThis workflow provides a comprehensive approach to single-cell RNA-seq data analysis, including batch effect correction and visualization, and can be replicated using the described steps and parameters.",
         "Single-cell RNA sequencing data analysis was conducted using Python with the Scanpy library, focusing on batch effect correction and visualization. The workflow involved preprocessing steps, including normalization to a target count of 10,000, log transformation, and selection of highly variable genes. Principal Component Analysis (PCA) reduced dimensionality, retaining the top 50 principal components. Batch effects were corrected using the Batch-Balanced K-Nearest Neighbors (BBKNN) algorithm, which constructs a neighborhood graph within each batch before integration. Visualization of the integrated data was achieved through Uniform Manifold Approximation and Projection (UMAP) and t-Distributed Stochastic Neighbor Embedding (t-SNE), confirming effective batch correction by intermingling of cells from different batches. The final integrated dataset was prepared for downstream analyses."
        ],
        [
         "27",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport scanpy as sc\n\nimport bbknn\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters and logging verbosity for clear output\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\nsc.settings.verbosity = 2\n\n\n\n#--------------------------------\n\n# Data Input & Simulation\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC3k dataset as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Simulate two batches by splitting the dataset in half.\n\n# This is for demonstration purposes; in a real scenario, the 'batch'\n\n# column would come from the experimental design.\n\nprint(\"Simulating two batches for demonstration...\")\n\nnum_cells = adata.n_obs\n\nbatch_labels = np.array([\"batch1\"] * num_cells)\n\nbatch_labels[int(num_cells / 2):] = \"batch2\"\n\nadata.obs[\"batch\"] = batch_labels\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization and Feature Selection ---\n\n# Normalize each cell to a target count of 10,000 and apply a log transformation.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) to focus on biological variation.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# --- Scaling and PCA ---\n\n# Scale the data to have zero mean and unit variance.\n\nsc.pp.scale(adata, max_value=10)\n\n\n\n# Perform Principal Component Analysis (PCA) for dimensionality reduction.\n\n# The PCs will be used as input for the BBKNN algorithm.\n\nsc.tl.pca(adata, svd_solver='arpack', n_comps=50)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with BBKNN\n\n#--------------------------------\n\n\n\n# --- Batch Integration ---\n\n# Run Batch-Balanced K-Nearest Neighbors (BBKNN) to correct for batch effects.\n\n# This function computes a neighborhood graph by finding neighbors within each\n\n# batch separately before merging them into a single, integrated graph.\n\nprint(\"Running BBKNN for batch integration...\")\n\nbbknn.bbknn(adata, batch_key='batch', n_pcs=50, neighbors_within_batch=3)\n\n\n\n# --- Post-Integration Embeddings ---\n\n# Compute UMAP and t-SNE embeddings based on the BBKNN-corrected neighborhood graph.\n\n# These will be used for visualizing the integrated data.\n\nsc.tl.umap(adata)\n\nsc.tl.tsne(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the UMAP embedding, coloring cells by their batch label.\n\n# Successful integration should show cells from 'batch1' and 'batch2' mixing well.\n\nsc.pl.umap(adata, color=[\"batch\"], title=\"UMAP: BBKNN Integrated\", size=30)\n\n\n\n# Plot the t-SNE embedding, also colored by batch, as an alternative visualization.\n\nsc.pl.tsne(adata, color=[\"batch\"], title=\"t-SNE: BBKNN Integrated\", size=30)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object can now be used for downstream analysis like clustering.\n\nprint(\"Saving integrated data object...\")\n\nadata.write(\"pbmc3k_bbknn_integrated.h5ad\")\n\n",
         "The following section describes the step-by-step workflow for performing single-cell RNA-seq data analysis, including batch integration and visualization, using Python and the Scanpy library. The analysis begins with data loading and preprocessing, followed by dimensionality reduction, batch correction, and visualization.\n\nFirst, the necessary Python libraries are imported. Scanpy is used for single-cell data analysis, BBKNN for batch correction, NumPy for numerical operations, and Matplotlib for plotting. These libraries are essential for handling single-cell datasets, performing batch integration, and generating visualizations.\n\nNext, global plotting parameters are configured to ensure high-quality figures with a resolution of 80 DPI and a white background. The logging verbosity is set to 2 to provide clear and informative output during the analysis.\n\nThe analysis begins with the loading of the PBMC3k dataset, a publicly available single-cell RNA-seq dataset, using the `sc.datasets.pbmc3k()` function. To demonstrate batch integration, the dataset is artificially split into two batches, labeled \"batch1\" and \"batch2.\" This step simulates a scenario where batch effects might arise due to experimental design.\n\nData preprocessing involves normalization and feature selection. The dataset is normalized to a target count of 10,000 reads per cell, followed by a log transformation to stabilize variance. Highly variable genes (HVGs) are identified using the `sc.pp.highly_variable_genes()` function, with the top 2000 genes selected based on the Seurat flavor. This step focuses the analysis on genes that exhibit biological variation rather than technical noise. The dataset is then scaled to have zero mean and unit variance, ensuring that all genes contribute equally to downstream analyses.\n\nPrincipal Component Analysis (PCA) is performed to reduce the dimensionality of the dataset. Fifty principal components (PCs) are computed using the ARPACK solver, which will serve as input for the batch integration step.\n\nThe core analysis involves batch integration using the Batch-Balanced K-Nearest Neighbors (BBKNN) algorithm. The `bbknn.bbknn()` function is applied to correct for batch effects by constructing a neighborhood graph that balances connections within and between batches. The `batch_key` parameter specifies the batch labels, while `n_pcs=50` ensures that the first 50 PCs are used for integration. The `neighbors_within_batch=3` parameter controls the number of neighbors considered within each batch.\n\nFollowing batch integration, Uniform Manifold Approximation and Projection (UMAP) and t-Distributed Stochastic Neighbor Embedding (t-SNE) are computed to visualize the integrated dataset in two dimensions. These embeddings provide insights into the structure of the data and the effectiveness of batch correction.\n\nVisualization is performed using UMAP and t-SNE plots, with cells colored by their batch labels. Successful integration is indicated by the mixing of cells from \"batch1\" and \"batch2\" in the embeddings. The `sc.pl.umap()` and `sc.pl.tsne()` functions generate these plots, with a cell size of 30 for clarity.\n\nFinally, the integrated dataset is saved as an AnnData object in the H5AD format using the `adata.write()` function. This file can be used for downstream analyses, such as clustering or differential expression testing.\n\nThis workflow provides a comprehensive approach to single-cell RNA-seq data analysis, including batch integration and visualization, using Python and the Scanpy library.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following section describes the step-by-step workflow for performing single-cell RNA-seq data analysis, including batch integration and visualization, using Python and the Scanpy library. The analysis begins with data loading and preprocessing, followed by dimensionality reduction, batch correction, and visualization.\n\nFirst, the necessary Python libraries are imported. Scanpy is used for single-cell data analysis, BBKNN for batch correction, NumPy for numerical operations, and Matplotlib for plotting. These libraries are essential for handling single-cell datasets, performing batch integration, and generating visualizations.\n\nNext, global plotting parameters are configured to ensure high-quality figures with a resolution of 80 DPI and a white background. The logging verbosity is set to 2 to provide clear and informative output during the analysis.\n\nThe analysis begins with the loading of the PBMC3k dataset, a publicly available single-cell RNA-seq dataset, using the `sc.datasets.pbmc3k()` function. To demonstrate batch integration, the dataset is artificially split into two batches, labeled \"batch1\" and \"batch2.\" This step simulates a scenario where batch effects might arise due to experimental design.\n\nData preprocessing involves normalization and feature selection. The dataset is normalized to a target count of 10,000 reads per cell, followed by a log transformation to stabilize variance. Highly variable genes (HVGs) are identified using the `sc.pp.highly_variable_genes()` function, with the top 2000 genes selected based on the Seurat flavor. This step focuses the analysis on genes that exhibit biological variation rather than technical noise. The dataset is then scaled to have zero mean and unit variance, ensuring that all genes contribute equally to downstream analyses.\n\nPrincipal Component Analysis (PCA) is performed to reduce the dimensionality of the dataset. Fifty principal components (PCs) are computed using the ARPACK solver, which will serve as input for the batch integration step.\n\nThe core analysis involves batch integration using the Batch-Balanced K-Nearest Neighbors (BBKNN) algorithm. The `bbknn.bbknn()` function is applied to correct for batch effects by constructing a neighborhood graph that balances connections within and between batches. The `batch_key` parameter specifies the batch labels, while `n_pcs=50` ensures that the first 50 PCs are used for integration. The `neighbors_within_batch=3` parameter controls the number of neighbors considered within each batch.\n\nFollowing batch integration, Uniform Manifold Approximation and Projection (UMAP) and t-Distributed Stochastic Neighbor Embedding (t-SNE) are computed to visualize the integrated dataset in two dimensions. These embeddings provide insights into the structure of the data and the effectiveness of batch correction.\n\nVisualization is performed using UMAP and t-SNE plots, with cells colored by their batch labels. Successful integration is indicated by the mixing of cells from \"batch1\" and \"batch2\" in the embeddings. The `sc.pl.umap()` and `sc.pl.tsne()` functions generate these plots, with a cell size of 30 for clarity.\n\nFinally, the integrated dataset is saved as an AnnData object in the H5AD format using the `adata.write()` function. This file can be used for downstream analyses, such as clustering or differential expression testing.\n\nThis workflow provides a comprehensive approach to single-cell RNA-seq data analysis, including batch integration and visualization, using Python and the Scanpy library.",
         "The analysis was conducted in Python using the Scanpy library, complemented by BBKNN for batch correction, NumPy for numerical operations, and Matplotlib for visualization. The workflow involved preprocessing a single-cell RNA-seq dataset, including normalization to 10,000 reads per cell, log transformation, and selection of highly variable genes. Principal Component Analysis (PCA) reduced dimensionality, followed by batch integration using the BBKNN algorithm to correct for batch effects. Uniform Manifold Approximation and Projection (UMAP) and t-Distributed Stochastic Neighbor Embedding (t-SNE) were employed for visualization, demonstrating effective batch correction through the mixing of cells from different batches. The integrated dataset was saved for downstream analyses."
        ],
        [
         "28",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport scanpy as sc\n\nimport scanpy.external as sce\n\nimport matplotlib.pyplot as plt\n\nimport numpy as np\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input & Simulation\n\n#--------------------------------\n\n\n\n# Load an example dataset (PBMC 3k) provided by Scanpy\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Simulate two batches by randomly assigning cells to one of two groups.\n\n# In a real-world scenario, this 'batch' information would come from the experiment metadata.\n\nprint(\"Simulating two batches for demonstration...\")\n\nnp.random.seed(42)\n\nadata.obs['batch'] = np.random.choice(['batch1', 'batch2'], size=adata.n_obs)\n\n\n\n# Split the main AnnData object into two separate objects, one for each batch.\n\n# MNN correction is typically performed on a list of batch-specific objects.\n\nadata_batch1 = adata[adata.obs['batch'] == 'batch1'].copy()\n\nadata_batch2 = adata[adata.obs['batch'] == 'batch2'].copy()\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing\n\n#--------------------------------\n\n\n\n# Preprocess each batch individually. This is a common strategy before integration.\n\nprint(\"Preprocessing each batch individually...\")\n\nfor ad in [adata_batch1, adata_batch2]:\n\n# 1. Normalize total counts per cell to account for library size differences.\n\nsc.pp.normalize_total(ad, target_sum=1e4)\n\n# 2. Log-transform the data to stabilize variance.\n\nsc.pp.log1p(ad)\n\n# 3. Identify highly variable genes to focus on biological signal.\n\nsc.pp.highly_variable_genes(ad, n_top_genes=2000, subset=True)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with MNN\n\n#--------------------------------\n\n\n\n# --- MNN Integration ---\n\n# Integrate the two batches using Mutual Nearest Neighbors (MNN) correction.\n\n# This method identifies shared cell populations across batches and aligns them.\n\n# The function returns a new, integrated AnnData object.\n\nprint(\"Running MNN integration...\")\n\nadata_corrected, _ = sce.pp.mnn_correct(adata_batch1, adata_batch2, batch_key='batch')\n\n\n\n# --- Post-Integration Steps ---\n\n# Perform PCA on the MNN-corrected data to find the main axes of variation.\n\nsc.tl.pca(adata_corrected, svd_solver='arpack')\n\n\n\n# Construct the neighborhood graph on the integrated PCA space.\n\nsc.pp.neighbors(adata_corrected, n_pcs=50)\n\n\n\n# Compute the UMAP embedding for visualization.\n\nsc.tl.umap(adata_corrected)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the UMAP embedding of the corrected data, coloring by batch.\n\n# Successful integration should result in cells from 'batch1' and 'batch2'\n\n# mixing together within cell type clusters.\n\nprint(\"Visualizing integrated data...\")\n\nsc.pl.umap(adata_corrected, color='batch', title='MNN Integrated UMAP')\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object is now ready for downstream analyses like clustering.\n\nprint(\"Saving integrated data object...\")\n\nadata_corrected.write(\"pbmc3k_mnn_integrated.h5ad\")\n\n",
         "The analysis begins by importing necessary Python libraries for single-cell RNA sequencing (scRNA-seq) data analysis, batch correction, and visualization. Specifically, the `scanpy` library is imported for single-cell analysis, `scanpy.external` for additional external tools, `matplotlib.pyplot` for plotting, and `numpy` for numerical operations. These libraries are essential for handling, processing, and visualizing scRNA-seq data.\n\nGlobal plotting parameters are configured to ensure clear and consistent visualizations. The resolution (`dpi=80`) and background color (`facecolor='white'`) are set to optimize the quality of the generated plots.\n\nAn example dataset, PBMC 3k, is loaded using the `scanpy` dataset module. This dataset represents peripheral blood mononuclear cells and serves as a benchmark for the analysis. To simulate a real-world scenario with batch effects, cells are randomly assigned to one of two batches (`batch1` or `batch2`) using `numpy`'s random choice function, with a fixed random seed (`seed=42`) for reproducibility. The dataset is then split into two separate objects, one for each batch, to prepare for batch correction.\n\nPreprocessing is performed individually on each batch to ensure data quality and comparability. First, total counts per cell are normalized to account for differences in library size, with a target sum of 10,000 counts per cell. The data is then log-transformed to stabilize variance and reduce the impact of extreme values. Highly variable genes are identified and subsetted to focus on genes with the most biological signal, retaining the top 2,000 genes for downstream analysis.\n\nThe primary analysis task involves integrating the two batches using Mutual Nearest Neighbors (MNN) correction, a method designed to align shared cell populations across batches. The `mnn_correct` function from `scanpy.external` is applied to the two batch-specific objects, producing a new integrated dataset. Post-integration, principal component analysis (PCA) is performed on the corrected data to identify the main axes of variation, using the `arpack` solver for efficiency. A neighborhood graph is constructed in the PCA space, and a UMAP embedding is computed to visualize the integrated data in two dimensions.\n\nThe UMAP embedding is plotted, with cells colored by their batch origin. Successful integration is indicated by the mixing of cells from both batches within cell type clusters, demonstrating the removal of batch effects. Finally, the integrated dataset is saved to a file (`pbmc3k_mnn_integrated.h5ad`) for downstream analyses, such as clustering or differential expression testing. This file preserves the corrected data and associated metadata, ensuring reproducibility and facilitating further exploration.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing necessary Python libraries for single-cell RNA sequencing (scRNA-seq) data analysis, batch correction, and visualization. Specifically, the `scanpy` library is imported for single-cell analysis, `scanpy.external` for additional external tools, `matplotlib.pyplot` for plotting, and `numpy` for numerical operations. These libraries are essential for handling, processing, and visualizing scRNA-seq data.\n\nGlobal plotting parameters are configured to ensure clear and consistent visualizations. The resolution (`dpi=80`) and background color (`facecolor='white'`) are set to optimize the quality of the generated plots.\n\nAn example dataset, PBMC 3k, is loaded using the `scanpy` dataset module. This dataset represents peripheral blood mononuclear cells and serves as a benchmark for the analysis. To simulate a real-world scenario with batch effects, cells are randomly assigned to one of two batches (`batch1` or `batch2`) using `numpy`'s random choice function, with a fixed random seed (`seed=42`) for reproducibility. The dataset is then split into two separate objects, one for each batch, to prepare for batch correction.\n\nPreprocessing is performed individually on each batch to ensure data quality and comparability. First, total counts per cell are normalized to account for differences in library size, with a target sum of 10,000 counts per cell. The data is then log-transformed to stabilize variance and reduce the impact of extreme values. Highly variable genes are identified and subsetted to focus on genes with the most biological signal, retaining the top 2,000 genes for downstream analysis.\n\nThe primary analysis task involves integrating the two batches using Mutual Nearest Neighbors (MNN) correction, a method designed to align shared cell populations across batches. The `mnn_correct` function from `scanpy.external` is applied to the two batch-specific objects, producing a new integrated dataset. Post-integration, principal component analysis (PCA) is performed on the corrected data to identify the main axes of variation, using the `arpack` solver for efficiency. A neighborhood graph is constructed in the PCA space, and a UMAP embedding is computed to visualize the integrated data in two dimensions.\n\nThe UMAP embedding is plotted, with cells colored by their batch origin. Successful integration is indicated by the mixing of cells from both batches within cell type clusters, demonstrating the removal of batch effects. Finally, the integrated dataset is saved to a file (`pbmc3k_mnn_integrated.h5ad`) for downstream analyses, such as clustering or differential expression testing. This file preserves the corrected data and associated metadata, ensuring reproducibility and facilitating further exploration.",
         "The analysis was conducted in Python using the Scanpy library for single-cell RNA sequencing (scRNA-seq) data processing, with additional tools from `scanpy.external` for batch correction and `matplotlib` for visualization. The PBMC 3k dataset was used, with cells randomly assigned to two batches to simulate batch effects. Preprocessing included library size normalization, log transformation, and selection of highly variable genes. Batch integration was performed using Mutual Nearest Neighbors (MNN) correction, followed by principal component analysis (PCA) and UMAP embedding to visualize the integrated data. Successful integration was confirmed by the mixing of cells from both batches in UMAP clusters."
        ],
        [
         "29",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load necessary libraries for single-cell analysis, data manipulation, and plotting\n\nlibrary(Seurat) # For single-cell RNA-seq analysis and integration\n\nlibrary(dplyr) # For data manipulation\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Define file paths for the Control and Stimulated datasets.\n\n# NOTE: Replace these with the actual paths to your data directories.\n\ndata_dir_control <- \"path/to/control_data/\"\n\ndata_dir_stim <- \"path/to/stimulated_data/\"\n\n\n\n# Read the raw 10X Genomics data for each condition.\n\ndata_control <- Read10X(data.dir = data_dir_control)\n\ndata_stim <- Read10X(data.dir = data_dir_stim)\n\n\n\n# Create individual Seurat objects for each dataset with basic filtering.\n\ncontrol <- CreateSeuratObject(counts = data_control, project = \"Control\", min.cells = 3, min.features = 200)\n\nstim <- CreateSeuratObject(counts = data_stim, project = \"Stimulated\", min.cells = 3, min.features = 200)\n\n\n\n#--------------------------------\n\n# Preprocessing of Individual Datasets\n\n#--------------------------------\n\n\n\n# It's a standard practice to preprocess each dataset individually before integration.\n\n# Here, we normalize and find variable features for both datasets.\n\nobject_list <- list(control, stim)\n\nfor (i in 1:length(object_list)) {\n\nobject_list[[i]] <- NormalizeData(object_list[[i]], normalization.method = \"LogNormalize\", scale.factor = 10000)\n\nobject_list[[i]] <- FindVariableFeatures(object_list[[i]], selection.method = \"vst\", nfeatures = 2000)\n\n}\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Data Integration\n\n#--------------------------------\n\n\n\n# --- Find Integration Anchors ---\n\n# Identify \"anchors\" or mutual nearest neighbors between the datasets. These anchors\n\n# represent shared cellular states across different conditions.\n\n# `dims = 1:20` specifies that the first 20 PCs are used for finding anchors.\n\nintegration_anchors <- FindIntegrationAnchors(object.list = object_list, dims = 1:20)\n\n\n\n# --- Integrate Data ---\n\n# Use the identified anchors to create an integrated (batch-corrected) data matrix.\n\nintegrated_data <- IntegrateData(anchorset = integration_anchors, dims = 1:20)\n\n\n\n# Set the default assay to \"integrated\" to ensure that downstream analysis\n\n# (like scaling, PCA, UMAP) is performed on the corrected data.\n\nDefaultAssay(integrated_data) <- \"integrated\"\n\n\n\n\n\n#--------------------------------\n\n# Post-Integration Analysis\n\n#--------------------------------\n\n\n\n# --- Scaling and Dimensionality Reduction ---\n\n# Scale the integrated data. This is necessary for PCA.\n\nintegrated_data <- ScaleData(integrated_data, verbose = FALSE)\n\n\n\n# Run PCA on the scaled, integrated data to reduce dimensionality.\n\nintegrated_data <- RunPCA(integrated_data, npcs = 30, verbose = FALSE)\n\n\n\n# Run UMAP for non-linear dimensionality reduction and visualization.\n\nintegrated_data <- RunUMAP(integrated_data, reduction = \"pca\", dims = 1:20)\n\n\n\n# --- Clustering (Optional) ---\n\n# Perform clustering on the integrated data to identify cell populations.\n\nintegrated_data <- FindNeighbors(integrated_data, reduction = \"pca\", dims = 1:20)\n\nintegrated_data <- FindClusters(integrated_data, resolution = 0.5)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a UMAP plot, coloring cells by their original condition ('orig.ident').\n\n# Successful integration should show cells from both conditions mixing well within clusters.\n\nDimPlot(integrated_data, reduction = \"umap\", group.by = \"orig.ident\", label = TRUE) +\n\nggtitle(\"UMAP Plot: Integrated Data by Condition\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final, integrated Seurat object for future analyses.\n\nsaveRDS(integrated_data, file = \"Integrated_Seurat_Object.rds\")\n\n",
         "The analysis begins with the loading of necessary libraries for single-cell RNA-seq analysis, data manipulation, and visualization. The Seurat package is used for single-cell RNA-seq analysis and integration, dplyr for data manipulation, ggplot2 for creating custom plots, and patchwork for combining multiple ggplot2 plots. These libraries are essential for performing the subsequent steps in the analysis.\n\nNext, the raw 10X Genomics data for both control and stimulated conditions are read into the environment. The file paths for the datasets are defined, and the data is loaded using the `Read10X` function. Individual Seurat objects are then created for each dataset using the `CreateSeuratObject` function. Basic filtering is applied during this step, with the parameters `min.cells = 3` and `min.features = 200` ensuring that only genes detected in at least 3 cells and cells expressing at least 200 genes are included in the analysis. This filtering step is crucial to remove low-quality cells and genes that could introduce noise into the dataset.\n\nFollowing data input, each dataset is preprocessed individually. This involves normalizing the data using the `NormalizeData` function with the `LogNormalize` method and a scale factor of 10,000. Variable features are then identified using the `FindVariableFeatures` function with the `vst` selection method and a limit of 2,000 features. Preprocessing each dataset separately is a standard practice that ensures the data is in a suitable state for downstream integration.\n\nThe major analysis task involves integrating the datasets to correct for batch effects. Integration is performed using the `FindIntegrationAnchors` function, which identifies mutual nearest neighbors (anchors) between the datasets. The first 20 principal components (PCs) are used for this step, as specified by the `dims = 1:20` parameter. The identified anchors are then used to create an integrated, batch-corrected data matrix using the `IntegrateData` function. The default assay is set to \"integrated\" to ensure that all subsequent analyses are performed on the corrected data.\n\nAfter integration, the data undergoes scaling and dimensionality reduction. The `ScaleData` function is used to scale the integrated data, which is necessary for principal component analysis (PCA). PCA is then performed using the `RunPCA` function with 30 principal components. Non-linear dimensionality reduction is achieved using UMAP, implemented via the `RunUMAP` function, with the first 20 PCs used as input. Clustering is optionally performed using the `FindNeighbors` and `FindClusters` functions, with a resolution parameter of 0.5 to identify cell populations.\n\nVisualization is carried out by generating a UMAP plot that colors cells by their original condition (`orig.ident`). The `DimPlot` function is used for this purpose, and the resulting plot is expected to show well-mixed cells from both conditions within clusters, indicating successful integration.\n\nFinally, the integrated Seurat object is saved to a file using the `saveRDS` function for future analyses. This step ensures that the results of the integration and downstream analyses are preserved and can be easily reloaded for further investigation.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins with the loading of necessary libraries for single-cell RNA-seq analysis, data manipulation, and visualization. The Seurat package is used for single-cell RNA-seq analysis and integration, dplyr for data manipulation, ggplot2 for creating custom plots, and patchwork for combining multiple ggplot2 plots. These libraries are essential for performing the subsequent steps in the analysis.\n\nNext, the raw 10X Genomics data for both control and stimulated conditions are read into the environment. The file paths for the datasets are defined, and the data is loaded using the `Read10X` function. Individual Seurat objects are then created for each dataset using the `CreateSeuratObject` function. Basic filtering is applied during this step, with the parameters `min.cells = 3` and `min.features = 200` ensuring that only genes detected in at least 3 cells and cells expressing at least 200 genes are included in the analysis. This filtering step is crucial to remove low-quality cells and genes that could introduce noise into the dataset.\n\nFollowing data input, each dataset is preprocessed individually. This involves normalizing the data using the `NormalizeData` function with the `LogNormalize` method and a scale factor of 10,000. Variable features are then identified using the `FindVariableFeatures` function with the `vst` selection method and a limit of 2,000 features. Preprocessing each dataset separately is a standard practice that ensures the data is in a suitable state for downstream integration.\n\nThe major analysis task involves integrating the datasets to correct for batch effects. Integration is performed using the `FindIntegrationAnchors` function, which identifies mutual nearest neighbors (anchors) between the datasets. The first 20 principal components (PCs) are used for this step, as specified by the `dims = 1:20` parameter. The identified anchors are then used to create an integrated, batch-corrected data matrix using the `IntegrateData` function. The default assay is set to \"integrated\" to ensure that all subsequent analyses are performed on the corrected data.\n\nAfter integration, the data undergoes scaling and dimensionality reduction. The `ScaleData` function is used to scale the integrated data, which is necessary for principal component analysis (PCA). PCA is then performed using the `RunPCA` function with 30 principal components. Non-linear dimensionality reduction is achieved using UMAP, implemented via the `RunUMAP` function, with the first 20 PCs used as input. Clustering is optionally performed using the `FindNeighbors` and `FindClusters` functions, with a resolution parameter of 0.5 to identify cell populations.\n\nVisualization is carried out by generating a UMAP plot that colors cells by their original condition (`orig.ident`). The `DimPlot` function is used for this purpose, and the resulting plot is expected to show well-mixed cells from both conditions within clusters, indicating successful integration.\n\nFinally, the integrated Seurat object is saved to a file using the `saveRDS` function for future analyses. This step ensures that the results of the integration and downstream analyses are preserved and can be easily reloaded for further investigation.",
         "The single-cell RNA sequencing analysis was conducted using the Seurat package in R, alongside dplyr for data manipulation and ggplot2 for visualization. Raw 10X Genomics data from control and stimulated conditions were processed individually, involving normalization, identification of variable features, and initial filtering to retain high-quality cells and genes. Integration of datasets was performed using mutual nearest neighbors to correct for batch effects, followed by scaling and dimensionality reduction via PCA and UMAP. Clustering was optionally applied to identify cell populations, and integration success was visualized using UMAP plots. The integrated dataset was saved for subsequent analyses."
        ],
        [
         "30",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport scanpy as sc\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global settings for clear visualization and detailed logging\n\nsc.settings.verbosity = 3\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the Paul et al. (2015) dataset, a common benchmark for trajectory inference.\n\nadata = sc.datasets.paul15()\n\n\n\n# Ensure the data matrix is in a float format for numerical stability.\n\nadata.X = adata.X.astype('float64')\n\n\n\n#--------------------------------\n\n# Preprocessing\n\n#--------------------------------\n\n\n\n# Apply a standard preprocessing recipe. This includes normalization per cell,\n\n# log transformation, and selection of highly variable genes.\n\nprint(\"Preprocessing data...\")\n\nsc.pp.recipe_zheng17(adata)\n\n\n\n# Perform PCA for dimensionality reduction.\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# Compute the k-nearest neighbor graph, which is required for clustering and PAGA.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=20)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Trajectory Inference\n\n#--------------------------------\n\n\n\n# --- Clustering ---\n\n# Cluster cells using the Louvain algorithm. The resulting clusters will be used\n\n# to build the abstracted graph in PAGA.\n\nprint(\"Clustering cells...\")\n\nsc.tl.louvain(adata, resolution=1.0)\n\n\n\n# --- PAGA (Partition-based Graph Abstraction) ---\n\n# Compute the PAGA graph, which represents the connectivity between cell clusters.\n\n# This provides a high-level, abstracted view of the developmental trajectory.\n\nprint(\"Computing PAGA graph...\")\n\nsc.tl.paga(adata, groups='louvain')\n\n\n\n# --- UMAP Initialization for Visualization ---\n\n# Compute a UMAP-like embedding, but initialize the positions using the PAGA graph.\n\n# This helps the final visualization better reflect the global topology inferred by PAGA.\n\nsc.tl.draw_graph(adata, init_pos='paga')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the PAGA abstracted graph. Edges represent the likelihood of a connection\n\n# between clusters. The `threshold` prunes weak connections for clarity.\n\nprint(\"Visualizing PAGA graph and trajectory...\")\n\nsc.pl.paga(\n\nadata,\n\ncolor=['louvain'],\n\nthreshold=0.03,\n\ntitle='PAGA Graph for Trajectory Inference'\n\n)\n\n\n\n# Plot the single-cell embedding initialized with PAGA.\n\n# This layout shows the fine-grained trajectory of individual cells, guided by the\n\n# coarse-grained PAGA structure.\n\nsc.pl.draw_graph(\n\nadata,\n\ncolor='louvain',\n\nlegend_loc='on data',\n\ntitle='Trajectory Inference via PAGA Initialization'\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the PAGA results and trajectory layout.\n\nprint(\"Saving trajectory data object...\")\n\nadata.write(\"paul15_paga_trajectory.h5ad\")\n\n",
         "The analysis begins by importing essential Python libraries for single-cell analysis and visualization. The `scanpy` library is used for single-cell RNA-seq data processing and analysis, while `matplotlib.pyplot` is employed for generating plots. Global settings are configured to ensure detailed logging and clear visualization, with a figure resolution of 80 DPI and a white background.\n\nThe dataset used in this analysis is the Paul et al. (2015) dataset, a widely recognized benchmark for trajectory inference in single-cell RNA-seq studies. The data matrix is converted to a float format to ensure numerical stability during computations.\n\nPreprocessing steps are applied to prepare the data for downstream analysis. This includes normalization per cell, log transformation, and the selection of highly variable genes using the `recipe_zheng17` function from Scanpy. Principal Component Analysis (PCA) is then performed for dimensionality reduction, followed by the computation of a k-nearest neighbor graph using 10 neighbors and 20 principal components. This graph is essential for subsequent clustering and trajectory inference.\n\nThe primary analysis task involves trajectory inference, which is achieved through a series of steps. First, cells are clustered using the Louvain algorithm with a resolution parameter of 1.0. These clusters are then used to construct a Partition-based Graph Abstraction (PAGA) graph, which provides a high-level representation of the developmental trajectory by abstracting the connectivity between cell clusters. To visualize the trajectory, a UMAP-like embedding is computed, with the initial positions of the cells initialized based on the PAGA graph. This ensures that the final visualization accurately reflects the global topology inferred by PAGA.\n\nVisualization of the results includes plotting the PAGA abstracted graph, where edges represent the likelihood of connections between clusters. A threshold of 0.03 is applied to prune weak connections for clarity. Additionally, the single-cell embedding initialized with PAGA is plotted to show the fine-grained trajectory of individual cells, guided by the coarse-grained PAGA structure.\n\nFinally, the processed data, including the PAGA results and trajectory layout, are saved as an AnnData object in the file `paul15_paga_trajectory.h5ad` for future use. This comprehensive workflow provides a robust framework for inferring and visualizing developmental trajectories in single-cell RNA-seq data.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for single-cell analysis and visualization. The `scanpy` library is used for single-cell RNA-seq data processing and analysis, while `matplotlib.pyplot` is employed for generating plots. Global settings are configured to ensure detailed logging and clear visualization, with a figure resolution of 80 DPI and a white background.\n\nThe dataset used in this analysis is the Paul et al. (2015) dataset, a widely recognized benchmark for trajectory inference in single-cell RNA-seq studies. The data matrix is converted to a float format to ensure numerical stability during computations.\n\nPreprocessing steps are applied to prepare the data for downstream analysis. This includes normalization per cell, log transformation, and the selection of highly variable genes using the `recipe_zheng17` function from Scanpy. Principal Component Analysis (PCA) is then performed for dimensionality reduction, followed by the computation of a k-nearest neighbor graph using 10 neighbors and 20 principal components. This graph is essential for subsequent clustering and trajectory inference.\n\nThe primary analysis task involves trajectory inference, which is achieved through a series of steps. First, cells are clustered using the Louvain algorithm with a resolution parameter of 1.0. These clusters are then used to construct a Partition-based Graph Abstraction (PAGA) graph, which provides a high-level representation of the developmental trajectory by abstracting the connectivity between cell clusters. To visualize the trajectory, a UMAP-like embedding is computed, with the initial positions of the cells initialized based on the PAGA graph. This ensures that the final visualization accurately reflects the global topology inferred by PAGA.\n\nVisualization of the results includes plotting the PAGA abstracted graph, where edges represent the likelihood of connections between clusters. A threshold of 0.03 is applied to prune weak connections for clarity. Additionally, the single-cell embedding initialized with PAGA is plotted to show the fine-grained trajectory of individual cells, guided by the coarse-grained PAGA structure.\n\nFinally, the processed data, including the PAGA results and trajectory layout, are saved as an AnnData object in the file `paul15_paga_trajectory.h5ad` for future use. This comprehensive workflow provides a robust framework for inferring and visualizing developmental trajectories in single-cell RNA-seq data.",
         "The analysis was conducted in Python using the `scanpy` library for single-cell RNA-seq data processing and `matplotlib.pyplot` for visualization. The Paul et al. (2015) dataset, a benchmark for trajectory inference, was preprocessed through normalization, log transformation, and selection of highly variable genes. Dimensionality reduction was performed using Principal Component Analysis (PCA), followed by the construction of a k-nearest neighbor graph. Clustering was achieved via the Louvain algorithm, and trajectory inference was implemented using Partition-based Graph Abstraction (PAGA). The PAGA graph abstracted cluster connectivity, while a UMAP-like embedding initialized with PAGA visualized the fine-grained trajectory. The results were saved as an AnnData object for further analysis."
        ],
        [
         "31",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport scanpy as sc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 3\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\")\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the dataset from a 10X Genomics matrix directory.\n\n# NOTE: Replace the path with the location of your pbmc3k_v1 data directory.\n\n# `cache=True` will save a processed version of the data for faster loading next time.\n\nadata = sc.read_10x_mtx(\"tutorial_data/pbmc3k_v1/\", cache=True)\n\n\n\n\n\n#--------------------------------\n\n# Quality Control (QC) and Filtering\n\n#--------------------------------\n\n\n\n# --- QC Metrics Calculation ---\n\n# Identify mitochondrial genes, which are often a sign of cell stress if highly expressed.\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n\n\n\n# Calculate standard quality control metrics.\n\nsc.pp.calculate_qc_metrics(\n\nadata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True\n\n)\n\n\n\n# --- Filtering ---\n\n# Define outlier cells based on QC metrics. These thresholds may need adjustment\n\n# depending on the specific dataset.\n\nadata.obs[\"outlier_mt\"] = adata.obs.pct_counts_mt > 5\n\nadata.obs[\"outlier_total\"] = adata.obs.total_counts > 5000\n\nadata.obs[\"outlier_ngenes\"] = adata.obs.n_genes_by_counts > 2500\n\n\n\n# Print the number of outliers detected for each metric.\n\nprint(f\"Found {sum(adata.obs['outlier_mt'])} cells with high mitochondrial content.\")\n\nprint(f\"Found {sum(adata.obs['outlier_total'])} cells with large total counts.\")\n\nprint(f\"Found {sum(adata.obs['outlier_ngenes'])} cells with a large number of genes.\")\n\n\n\n# Remove outlier cells from the dataset based on the flags defined above.\n\n# The `~` operator inverts the boolean selection.\n\nadata = adata[~adata.obs[\"outlier_mt\"], :]\n\nadata = adata[~adata.obs[\"outlier_total\"], :]\n\nadata = adata[~adata.obs[\"outlier_ngenes\"], :]\n\n\n\n# Filter out genes that are no longer expressed in any cells after cell filtering.\n\nsc.pp.filter_genes(adata, min_cells=1)\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing & Feature Selection\n\n#--------------------------------\n\n\n\n# Normalize each cell to a target count of 10,000 and apply a log transformation.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs). This step calculates mean expression\n\n# and residual variances, which are used in the plot below.\n\nsc.pp.highly_variable_genes(adata, flavor=\"seurat\", n_top_genes=2000)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- QC Plots ---\n\n# Plot violin plots of the main QC metrics to visualize their distributions.\n\nsc.pl.violin(\n\nadata,\n\n[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\n)\n\n\n\n# --- HVG Plot ---\n\n# Create a scatter plot to visualize highly variable genes.\n\n# This helps confirm that the feature selection has worked as expected.\n\nfig, ax = plt.subplots(figsize=(6, 6))\n\nhvgs = adata.var[\"highly_variable\"]\n\n\n\n# Plot all genes in grey\n\nax.scatter(\n\nadata.var[\"means\"], adata.var[\"dispersions_norm\"], s=3, edgecolor=\"none\"\n\n)\n\n# Highlight the selected highly variable genes in red\n\nax.scatter(\n\nadata.var[\"means\"][hvgs],\n\nadata.var[\"dispersions_norm\"][hvgs],\n\nc=\"tab:red\",\n\nlabel=\"selected genes\",\n\ns=3,\n\nedgecolor=\"none\",\n\n)\n\nax.set_xscale(\"log\")\n\nax.set_yscale(\"log\")\n\nax.set_xlabel(\"Mean Expression\")\n\nax.set_ylabel(\"Normalized Dispersion\")\n\nax.set_title(\"Highly Variable Genes\")\n\nplt.legend()\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the filtered and processed AnnData object to a file.\n\n# This object is ready for downstream analysis like dimensionality reduction and clustering.\n\nprint(\"Saving filtered data object...\")\n\nadata.write(\"pbmc3k_filtered_hvg.h5ad\")\n\n",
         "The following section describes the computational workflow for processing single-cell RNA sequencing data using the Scanpy library in Python. The analysis includes data loading, quality control, filtering, preprocessing, feature selection, visualization, and saving the processed data for downstream analysis.\n\nFirst, the necessary Python libraries were imported. NumPy was used for numerical operations, Matplotlib for plotting, and Scanpy for single-cell analysis. Scanpy’s global parameters were configured to set the verbosity level and figure parameters, ensuring consistent and detailed output during execution.\n\nThe dataset was loaded from a 10X Genomics matrix directory using the `sc.read_10x_mtx` function. The path to the data directory was specified, and the `cache=True` parameter was enabled to save a processed version of the data for faster loading in subsequent runs. This step is critical for efficient data handling and reproducibility.\n\nQuality control (QC) metrics were calculated to assess the quality of the dataset. Mitochondrial genes were identified by checking gene names for the prefix \"MT-\", as high expression of these genes can indicate cell stress. The `sc.pp.calculate_qc_metrics` function was used to compute standard QC metrics, including the percentage of mitochondrial gene expression. Outlier cells were flagged based on predefined thresholds: cells with mitochondrial gene expression exceeding 5%, total counts exceeding 5,000, and genes detected exceeding 2,500. These thresholds were chosen to exclude low-quality or damaged cells. The number of outliers for each metric was printed, and outlier cells were removed from the dataset. Additionally, genes no longer expressed in any cells after filtering were removed using the `sc.pp.filter_genes` function.\n\nThe dataset was preprocessed by normalizing the total counts per cell to 10,000 and applying a log transformation using the `sc.pp.normalize_total` and `sc.pp.log1p` functions. This step ensures comparability across cells and stabilizes variance. Highly variable genes (HVGs) were identified using the `sc.pp.highly_variable_genes` function with the \"seurat\" flavor, selecting the top 2,000 genes based on their normalized dispersion. HVGs are critical for capturing biological variability and reducing noise in downstream analyses.\n\nVisualizations were generated to inspect the data. Violin plots were created to visualize the distributions of key QC metrics, including the number of genes detected, total counts, and percentage of mitochondrial gene expression. A scatter plot was generated to visualize the selection of HVGs, with all genes plotted in grey and selected HVGs highlighted in red. The plot used logarithmic scales for mean expression and normalized dispersion, providing a clear representation of gene variability.\n\nFinally, the filtered and processed dataset was saved as an AnnData object using the `adata.write` function. This object is ready for downstream analyses such as dimensionality reduction and clustering. The workflow ensures robust preprocessing and quality control, enabling reliable interpretation of single-cell RNA sequencing data.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following section describes the computational workflow for processing single-cell RNA sequencing data using the Scanpy library in Python. The analysis includes data loading, quality control, filtering, preprocessing, feature selection, visualization, and saving the processed data for downstream analysis.\n\nFirst, the necessary Python libraries were imported. NumPy was used for numerical operations, Matplotlib for plotting, and Scanpy for single-cell analysis. Scanpy’s global parameters were configured to set the verbosity level and figure parameters, ensuring consistent and detailed output during execution.\n\nThe dataset was loaded from a 10X Genomics matrix directory using the `sc.read_10x_mtx` function. The path to the data directory was specified, and the `cache=True` parameter was enabled to save a processed version of the data for faster loading in subsequent runs. This step is critical for efficient data handling and reproducibility.\n\nQuality control (QC) metrics were calculated to assess the quality of the dataset. Mitochondrial genes were identified by checking gene names for the prefix \"MT-\", as high expression of these genes can indicate cell stress. The `sc.pp.calculate_qc_metrics` function was used to compute standard QC metrics, including the percentage of mitochondrial gene expression. Outlier cells were flagged based on predefined thresholds: cells with mitochondrial gene expression exceeding 5%, total counts exceeding 5,000, and genes detected exceeding 2,500. These thresholds were chosen to exclude low-quality or damaged cells. The number of outliers for each metric was printed, and outlier cells were removed from the dataset. Additionally, genes no longer expressed in any cells after filtering were removed using the `sc.pp.filter_genes` function.\n\nThe dataset was preprocessed by normalizing the total counts per cell to 10,000 and applying a log transformation using the `sc.pp.normalize_total` and `sc.pp.log1p` functions. This step ensures comparability across cells and stabilizes variance. Highly variable genes (HVGs) were identified using the `sc.pp.highly_variable_genes` function with the \"seurat\" flavor, selecting the top 2,000 genes based on their normalized dispersion. HVGs are critical for capturing biological variability and reducing noise in downstream analyses.\n\nVisualizations were generated to inspect the data. Violin plots were created to visualize the distributions of key QC metrics, including the number of genes detected, total counts, and percentage of mitochondrial gene expression. A scatter plot was generated to visualize the selection of HVGs, with all genes plotted in grey and selected HVGs highlighted in red. The plot used logarithmic scales for mean expression and normalized dispersion, providing a clear representation of gene variability.\n\nFinally, the filtered and processed dataset was saved as an AnnData object using the `adata.write` function. This object is ready for downstream analyses such as dimensionality reduction and clustering. The workflow ensures robust preprocessing and quality control, enabling reliable interpretation of single-cell RNA sequencing data.",
         "The analysis of single-cell RNA sequencing data was conducted in Python using the Scanpy library, focusing on quality control, preprocessing, and feature selection. The workflow began with the identification of mitochondrial genes and calculation of QC metrics, followed by the removal of outlier cells based on predefined thresholds for mitochondrial gene expression, total counts, and detected genes. Normalization was performed by scaling total counts per cell and applying a log transformation. Highly variable genes (HVGs) were identified using the \"seurat\" flavor, selecting the top 2,000 genes based on normalized dispersion. Visualization included violin plots for QC metrics and scatter plots for HVG selection. The processed data was saved as an AnnData object for downstream analysis."
        ],
        [
         "32",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport scanpy as sc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 3 # Show detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\")\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the dataset from a 10X Genomics formatted directory.\n\n# NOTE: Replace the path with the location of your data directory.\n\nadata = sc.read_10x_mtx(\"data/10x_genomics_directory/\", cache=True)\n\n\n\n\n\n#--------------------------------\n\n# Quality Control (QC)\n\n#--------------------------------\n\n\n\n# Identify mitochondrial genes for potential QC filtering later on.\n\nadata.var['mt'] = adata.var_names.str.startswith(\"MT-\")\n\n\n\n# Calculate basic quality control metrics for each cell.\n\nsc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], percent_top=None, inplace=True)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Feature Selection\n\n#--------------------------------\n\n\n\n# Identify highly variable genes using the Pearson residuals method.\n\n# This experimental function models technical variability and selects genes\n\n# that deviate significantly, which are likely biologically variable.\n\nsc.experimental.pp.highly_variable_genes(\n\nadata, flavor=\"pearson_residuals\", n_top_genes=2000\n\n)\n\n\n\n# Print the number of highly variable genes found for confirmation.\n\nprint(f\"Number of highly variable genes selected: {sum(adata.var['highly_variable'])}\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create a scatter plot to visualize the feature selection results.\n\n# The plot shows mean expression vs. the residual variance calculated by the model.\n\nfig, ax = plt.subplots(figsize=(6, 6))\n\nhvgs = adata.var[\"highly_variable\"]\n\n\n\n# Plot all genes in grey\n\nax.scatter(\n\nadata.var[\"mean_counts\"],\n\nadata.var[\"residual_variances\"],\n\ns=3,\n\nc='gray',\n\nlabel='All genes',\n\nalpha=0.5\n\n)\n\n# Highlight the selected highly variable genes in red\n\nax.scatter(\n\nadata.var[\"mean_counts\"][hvgs],\n\nadata.var[\"residual_variances\"][hvgs],\n\ns=3,\n\nc='red',\n\nlabel='Highly variable genes'\n\n)\n\n\n\nax.set_xscale(\"log\")\n\nax.set_yscale(\"log\")\n\nax.set_xlabel(\"Mean Expression\")\n\nax.set_ylabel(\"Residual Variance\")\n\nax.set_title(\"Feature Selection: Pearson Residuals\")\n\nplt.legend()\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Subset the AnnData object to keep only the highly variable genes.\n\nadata = adata[:, adata.var[\"highly_variable\"]]\n\n\n\n# Save the processed AnnData object to a file.\n\n# This object is ready for downstream analysis like dimensionality reduction.\n\nprint(\"Saving filtered data object...\")\n\nadata.write(\"pearson_residuals_hvg.h5ad\")\n\n",
         "The analysis was performed using Python, leveraging the Scanpy library for single-cell RNA-seq data processing and visualization. The workflow is divided into several key steps, described below.\n\nFirst, the necessary libraries were imported. NumPy was used for numerical operations, Matplotlib for plotting, and Scanpy for single-cell data analysis. Scanpy’s global parameters were configured to enable detailed logging and set figure parameters, including a resolution of 80 DPI and a white background for plots.\n\nThe dataset was loaded from a 10X Genomics formatted directory using the `sc.read_10x_mtx` function. This function reads the count matrix, gene, and barcode files into an AnnData object, which is the primary data structure in Scanpy. The `cache=True` parameter ensures that the data is cached for faster subsequent access.\n\nQuality control (QC) metrics were calculated to assess the data’s integrity. Mitochondrial genes were identified by filtering gene names starting with \"MT-\". The `sc.pp.calculate_qc_metrics` function computed basic QC metrics for each cell, including the percentage of mitochondrial gene expression, which is often used to filter out low-quality cells.\n\nNext, highly variable genes (HVGs) were identified using the Pearson residuals method, implemented in the experimental function `sc.experimental.pp.highly_variable_genes`. This method models technical variability and selects genes that deviate significantly from the model, likely representing biologically relevant variability. The `flavor=\"pearson_residuals\"` parameter specifies the method, and `n_top_genes=2000` limits the selection to the top 2000 genes. The number of HVGs selected was printed for confirmation.\n\nTo visualize the feature selection results, a scatter plot was generated using Matplotlib. The plot displayed mean expression versus residual variance for all genes, with highly variable genes highlighted in red. The axes were log-scaled to better visualize the distribution, and the plot included labels for mean expression, residual variance, and a title indicating the method used.\n\nFinally, the AnnData object was subset to retain only the highly variable genes, and the filtered object was saved to a file named `pearson_residuals_hvg.h5ad` using the `adata.write` function. This file is ready for downstream analyses, such as dimensionality reduction or clustering.\n\nThis workflow provides a comprehensive approach to preprocessing single-cell RNA-seq data, from data loading and quality control to feature selection and visualization, ensuring the data is prepared for further biological insights.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis was performed using Python, leveraging the Scanpy library for single-cell RNA-seq data processing and visualization. The workflow is divided into several key steps, described below.\n\nFirst, the necessary libraries were imported. NumPy was used for numerical operations, Matplotlib for plotting, and Scanpy for single-cell data analysis. Scanpy’s global parameters were configured to enable detailed logging and set figure parameters, including a resolution of 80 DPI and a white background for plots.\n\nThe dataset was loaded from a 10X Genomics formatted directory using the `sc.read_10x_mtx` function. This function reads the count matrix, gene, and barcode files into an AnnData object, which is the primary data structure in Scanpy. The `cache=True` parameter ensures that the data is cached for faster subsequent access.\n\nQuality control (QC) metrics were calculated to assess the data’s integrity. Mitochondrial genes were identified by filtering gene names starting with \"MT-\". The `sc.pp.calculate_qc_metrics` function computed basic QC metrics for each cell, including the percentage of mitochondrial gene expression, which is often used to filter out low-quality cells.\n\nNext, highly variable genes (HVGs) were identified using the Pearson residuals method, implemented in the experimental function `sc.experimental.pp.highly_variable_genes`. This method models technical variability and selects genes that deviate significantly from the model, likely representing biologically relevant variability. The `flavor=\"pearson_residuals\"` parameter specifies the method, and `n_top_genes=2000` limits the selection to the top 2000 genes. The number of HVGs selected was printed for confirmation.\n\nTo visualize the feature selection results, a scatter plot was generated using Matplotlib. The plot displayed mean expression versus residual variance for all genes, with highly variable genes highlighted in red. The axes were log-scaled to better visualize the distribution, and the plot included labels for mean expression, residual variance, and a title indicating the method used.\n\nFinally, the AnnData object was subset to retain only the highly variable genes, and the filtered object was saved to a file named `pearson_residuals_hvg.h5ad` using the `adata.write` function. This file is ready for downstream analyses, such as dimensionality reduction or clustering.\n\nThis workflow provides a comprehensive approach to preprocessing single-cell RNA-seq data, from data loading and quality control to feature selection and visualization, ensuring the data is prepared for further biological insights.",
         "The analysis was conducted in Python using the Scanpy library for processing and visualizing single-cell RNA-seq data. The workflow began with the calculation of quality control metrics, including the identification of mitochondrial genes and assessment of their expression percentages. Highly variable genes (HVGs) were then selected using the Pearson residuals method to capture biologically relevant variability. Visualization of the feature selection results was achieved through scatter plots, highlighting HVGs based on mean expression and residual variance. The dataset was subset to retain only HVGs, preparing it for downstream analyses such as dimensionality reduction and clustering. This approach ensures robust preprocessing for deriving biological insights."
        ],
        [
         "33",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, and single-cell analysis\nimport scanpy as sc\nfrom pathlib import Path\nimport numpy as np\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Define the directory to store the data and create it if it doesn't exist.\nDATA_DIR = Path(\"./data/\")\nDATA_DIR.mkdir(parents=True, exist_ok=True)\nFILE_NAME = DATA_DIR / \"bone_marrow.h5ad\"\n\n# Load the dataset. A backup URL is provided to download the data if not found locally.\nprint(\"Loading dataset...\")\nadata = sc.read(\nfilename=FILE_NAME,\nbackup_url=\"https://figshare.com/ndownloader/files/35826944\"\n)\n\n#--------------------------------\n# Preprocessing\n#--------------------------------\n\n# Apply a standard preprocessing workflow.\nprint(\"Preprocessing data...\")\n# Filter out genes that are not expressed in at least 20 cells.\nsc.pp.filter_genes(adata, min_counts=20)\n# Normalize each cell by total counts to account for library size differences.\nsc.pp.normalize_total(adata)\n# Apply a log transformation to the data to stabilize variance.\nsc.pp.log1p(adata)\n# Identify highly variable genes (HVGs) to focus on biological signal.\nsc.pp.highly_variable_genes(adata)\n# Perform PCA for dimensionality reduction.\nsc.tl.pca(adata)\n# Compute the nearest neighbor graph, which is required for clustering and trajectory analysis.\nsc.pp.neighbors(adata, n_pcs=10)\n\n#--------------------------------\n# Major Analysis Task: Trajectory Inference with DPT\n#--------------------------------\n\n# --- Clustering & Visualization Embedding ---\n# Cluster cells using the Louvain algorithm to identify cell groups.\nsc.tl.louvain(adata)\n# Compute a t-SNE embedding for visualization purposes.\nsc.tl.tsne(adata)\n\n# --- Diffusion Pseudotime (DPT) ---\n# Compute the diffusion map. This represents the data as a graph where edge\n# weights reflect the transition probability between cells.\nprint(\"Computing diffusion map and pseudotime...\")\nsc.tl.diffmap(adata)\n\n# Identify the root cell for the pseudotime trajectory. Here, we select the cell\n# with the minimum value in the 3rd diffusion component as the starting point.\n# This choice often corresponds to the earliest cell state in a differentiation process.\nroot_ixs = adata.obsm[\"X_diffmap\"][:, 3].argmin()\nadata.uns[\"iroot\"] = root_ixs\n\n# Compute Diffusion Pseudotime (DPT), which orders cells based on their distance\n# from the root cell along the diffusion map graph.\nsc.tl.dpt(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the t-SNE embedding, colored by the identified cell clusters.\nsc.pl.scatter(adata, basis=\"tsne\", color=\"louvain\", title=\"t-SNE by Cluster\")\n\n# Visualize the t-SNE embedding, colored by the calculated DPT pseudotime.\n# This plot shows the progression of cells along the inferred trajectory.\nsc.pl.scatter(\nadata, basis=\"tsne\", color=[\"dpt_pseudotime\"], color_map=\"gnuplot2\",\ntitle=\"t-SNE colored by Pseudotime\"\n)\n\n# Create a violin plot to compare the pseudotime distributions across the different cell clusters.\nsc.pl.violin(\nadata, keys=[\"dpt_pseudotime\"], groupby=\"louvain\", rotation=45,\ntitle=\"Pseudotime Distribution by Cluster\"\n)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the AnnData object, which now contains the DPT trajectory results.\nprint(\"Saving trajectory data object...\")\nadata.write(\"bone_marrow_dpt_trajectory.h5ad\")\n\nprint(\"\\nPseudotime analysis completed!\")\n",
         "To perform trajectory inference on single-cell RNA sequencing data using Diffusion Pseudotime (DPT), the following steps were executed using the Scanpy toolkit in Python. The analysis began with data loading and preprocessing, followed by dimensionality reduction, clustering, and trajectory inference. Finally, the results were visualized and saved.\n\nFirst, the necessary Python libraries were imported. Scanpy was used for single-cell analysis, Pathlib for directory management, NumPy for numerical operations, and Matplotlib for plotting. These libraries are essential for handling, processing, and visualizing single-cell data.\n\nThe dataset was loaded from a local directory. If the file was not found locally, it was downloaded from a backup URL. The data was stored in an AnnData object, a specialized data structure for single-cell analysis. The dataset used in this analysis was a bone marrow single-cell RNA-seq dataset, stored in the H5AD format.\n\nPreprocessing steps were applied to prepare the data for downstream analysis. Genes expressed in fewer than 20 cells were filtered out to remove noise. The data was normalized by total counts per cell to account for differences in library size. A log transformation was applied to stabilize variance across cells. Highly variable genes (HVGs) were identified to focus on biologically relevant signals. Principal Component Analysis (PCA) was performed to reduce dimensionality, and a nearest neighbor graph was computed to facilitate clustering and trajectory analysis.\n\nFor the major analysis task, trajectory inference was performed using Diffusion Pseudotime (DPT). Cells were clustered using the Louvain algorithm to identify distinct cell populations. A t-SNE embedding was computed for visualization purposes. A diffusion map was generated to represent the data as a graph, where edge weights reflected transition probabilities between cells. The root cell for the trajectory was identified as the cell with the minimum value in the third diffusion component, often corresponding to the earliest cell state in a differentiation process. Diffusion Pseudotime (DPT) was computed to order cells based on their distance from the root cell along the diffusion map graph.\n\nVisualizations were created to interpret the results. A t-SNE plot was generated, colored by the identified cell clusters, to visualize the cell population structure. Another t-SNE plot was created, colored by the computed pseudotime, to show the progression of cells along the inferred trajectory. A violin plot was used to compare the pseudotime distributions across the different cell clusters, providing insights into the temporal dynamics of each population.\n\nFinally, the AnnData object containing the trajectory analysis results was saved to a file for future use. This file includes all preprocessing, clustering, and trajectory inference results, ensuring the analysis can be reproduced or extended. The pseudotime analysis was thus completed, providing a comprehensive understanding of the cell differentiation process in the bone marrow dataset.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform trajectory inference on single-cell RNA sequencing data using Diffusion Pseudotime (DPT), the following steps were executed using the Scanpy toolkit in Python. The analysis began with data loading and preprocessing, followed by dimensionality reduction, clustering, and trajectory inference. Finally, the results were visualized and saved.\n\nFirst, the necessary Python libraries were imported. Scanpy was used for single-cell analysis, Pathlib for directory management, NumPy for numerical operations, and Matplotlib for plotting. These libraries are essential for handling, processing, and visualizing single-cell data.\n\nThe dataset was loaded from a local directory. If the file was not found locally, it was downloaded from a backup URL. The data was stored in an AnnData object, a specialized data structure for single-cell analysis. The dataset used in this analysis was a bone marrow single-cell RNA-seq dataset, stored in the H5AD format.\n\nPreprocessing steps were applied to prepare the data for downstream analysis. Genes expressed in fewer than 20 cells were filtered out to remove noise. The data was normalized by total counts per cell to account for differences in library size. A log transformation was applied to stabilize variance across cells. Highly variable genes (HVGs) were identified to focus on biologically relevant signals. Principal Component Analysis (PCA) was performed to reduce dimensionality, and a nearest neighbor graph was computed to facilitate clustering and trajectory analysis.\n\nFor the major analysis task, trajectory inference was performed using Diffusion Pseudotime (DPT). Cells were clustered using the Louvain algorithm to identify distinct cell populations. A t-SNE embedding was computed for visualization purposes. A diffusion map was generated to represent the data as a graph, where edge weights reflected transition probabilities between cells. The root cell for the trajectory was identified as the cell with the minimum value in the third diffusion component, often corresponding to the earliest cell state in a differentiation process. Diffusion Pseudotime (DPT) was computed to order cells based on their distance from the root cell along the diffusion map graph.\n\nVisualizations were created to interpret the results. A t-SNE plot was generated, colored by the identified cell clusters, to visualize the cell population structure. Another t-SNE plot was created, colored by the computed pseudotime, to show the progression of cells along the inferred trajectory. A violin plot was used to compare the pseudotime distributions across the different cell clusters, providing insights into the temporal dynamics of each population.\n\nFinally, the AnnData object containing the trajectory analysis results was saved to a file for future use. This file includes all preprocessing, clustering, and trajectory inference results, ensuring the analysis can be reproduced or extended. The pseudotime analysis was thus completed, providing a comprehensive understanding of the cell differentiation process in the bone marrow dataset.",
         "Trajectory inference on single-cell RNA sequencing data was performed using Diffusion Pseudotime (DPT) in Python with the Scanpy toolkit. The workflow included preprocessing steps such as gene filtering, normalization, and log transformation, followed by identification of highly variable genes (HVGs). Dimensionality reduction was achieved through Principal Component Analysis (PCA), and a nearest neighbor graph was constructed for clustering using the Louvain algorithm. A diffusion map was generated to model cell transitions, and the root cell was identified to compute DPT, ordering cells along the inferred trajectory. Results were visualized using t-SNE and violin plots, highlighting cell clusters and pseudotemporal dynamics. The processed data and trajectory analysis were stored in an AnnData object for reproducibility."
        ],
        [
         "34",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell trajectory analysis\n\nlibrary(monocle) # Core package for Monocle2 analysis\n\nlibrary(ggplot2) # For advanced plotting customization\n\nlibrary(Matrix) # For handling sparse matrices\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# NOTE: Replace file paths with the actual location of your data files.\n\n# Read the gene expression matrix (rows: genes, columns: cells).\n\nexpr_matrix <- read.table(\"fpkm_matrix.txt\", header = TRUE, row.names = 1)\n\n\n\n# Read the cell metadata sheet.\n\nsample_sheet <- read.delim(\"cell_sample_sheet.txt\", header = TRUE, row.names = 1)\n\n\n\n# Read the gene annotation file.\n\ngene_annotation <- read.delim(\"gene_annotations.txt\", header = TRUE, row.names = 1)\n\n\n\n\n\n#--------------------------------\n\n# Construct CellDataSet Object\n\n#--------------------------------\n\n\n\n# Create AnnotatedDataFrame objects to hold the metadata.\n\npd <- new(\"AnnotatedDataFrame\", data = sample_sheet)\n\nfd <- new(\"AnnotatedDataFrame\", data = gene_annotation)\n\n\n\n# Create the main CellDataSet (CDS) object, which is Monocle's primary data structure.\n\n# `negbinomial.size()` is a common choice for UMI-based scRNA-seq data.\n\ncds <- newCellDataSet(\n\nas.matrix(expr_matrix),\n\nphenoData = pd,\n\nfeatureData = fd,\n\nlowerDetectionLimit = 0.5,\n\nexpressionFamily = negbinomial.size()\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing and Normalization\n\n#--------------------------------\n\n\n\n# Estimate size factors for normalization and gene dispersions for downstream analysis.\n\ncds <- estimateSizeFactors(cds)\n\ncds <- estimateDispersions(cds)\n\n\n\n# Filter out low-quality genes. First, detect genes expressed above a minimum threshold.\n\ncds <- detectGenes(cds, min_expr = 0.1)\n\n\n\n# Then, subset the list to genes that are expressed in at least 10 cells.\n\nexpressed_genes <- row.names(subset(fData(cds), num_cells_expressed >= 10))\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Trajectory Inference\n\n#--------------------------------\n\n\n\n# --- Selecting Genes for Ordering ---\n\n# Identify genes that will be used to order cells in pseudotime.\n\n# Here, we perform differential expression between conditions (e.g., 'Media' column).\n\n# Genes that change significantly between conditions are good candidates for ordering.\n\ndiff_test_res <- differentialGeneTest(cds[expressed_genes, ], fullModelFormulaStr = \"~Media\")\n\nordering_genes <- row.names(subset(diff_test_res, qval < 0.01))\n\n\n\n# Set the identified ordering genes in the CDS object.\n\ncds <- setOrderingFilter(cds, ordering_genes)\n\n\n\n\n\n# --- Trajectory Construction ---\n\n# Reduce the dimensionality of the data using the DDRTree algorithm.\n\n# This method is well-suited for datasets with complex, branching trajectories.\n\ncds <- reduceDimension(cds, max_components = 2, method = \"DDRTree\")\n\n\n\n# Order the cells along the trajectory to calculate pseudotime.\n\ncds <- orderCells(cds)\n\n\n\n\n\n# --- Differential Expression Along Pseudotime ---\n\n# Identify genes whose expression changes as a function of pseudotime.\n\n# The `sm.ns` function models expression using a natural smoothing spline.\n\ndiff_genes_pseudotime <- differentialGeneTest(cds[expressed_genes, ], fullModelFormulaStr = \"~sm.ns(Pseudotime)\")\n\nsig_genes_pseudotime <- subset(diff_genes_pseudotime, qval < 0.1)\n\n\n\n# Print the top genes that vary along the trajectory.\n\nprint(\"Top genes changing as a function of pseudotime:\")\n\nprint(head(sig_genes_pseudotime[order(sig_genes_pseudotime$qval), ]))\n\n\n\n\n\n# --- Branch Analysis (BEAM) ---\n\n# Identify genes that are differentially expressed at a branch point in the trajectory.\n\n# NOTE: The `branch_point` parameter must be adjusted based on your specific trajectory.\n\nbranch_point <- 1\n\nbeam_res <- BEAM(cds, branch_point = branch_point, cores = 1)\n\nbeam_sig_genes <- row.names(subset(beam_res, qval < 1e-4))\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the dispersion of the ordering genes.\n\nplot_ordering_genes(cds)\n\n\n\n# Plot the main cell trajectory, colored by different variables.\n\n# Color by Pseudotime to see the progression.\n\nplot_cell_trajectory(cds, color_by = \"Pseudotime\")\n\n\n\n# Color by experimental condition to see how it maps onto the trajectory.\n\nplot_cell_trajectory(cds, color_by = \"Media\")\n\n\n\n# Plot the expression of branch-dependent genes along the branched trajectory.\n\nplot_genes_branched_pseudotime(\n\ncds[beam_sig_genes, ],\n\nbranch_point = branch_point,\n\ncolor_by = \"Pseudotime\"\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final Monocle object, which contains all analysis results.\n\nsaveRDS(cds, file = \"monocle_trajectory_object.rds\")\n\n",
         "To perform single-cell trajectory analysis using Monocle2, the following steps were executed. The analysis began by loading the necessary libraries: `monocle` for trajectory analysis, `ggplot2` for advanced plotting, and `Matrix` for handling sparse matrices. These libraries are essential for data manipulation, analysis, and visualization in the context of single-cell RNA sequencing.\n\nThe gene expression matrix, cell metadata, and gene annotation files were imported. The gene expression matrix contained rows representing genes and columns representing cells, while the cell metadata included information about each cell, such as experimental conditions. The gene annotation file provided additional details about the genes. These datasets were read into R using `read.table` and `read.delim` functions, ensuring proper formatting and header inclusion.\n\nNext, the `CellDataSet` (CDS) object, the primary data structure in Monocle, was constructed. The gene expression matrix, cell metadata, and gene annotation were combined into `AnnotatedDataFrame` objects and passed to the `newCellDataSet` function. The `negbinomial.size()` expression family was specified to model UMI-based single-cell RNA-seq data, and a lower detection limit of 0.5 was set to filter out lowly expressed genes.\n\nPreprocessing and normalization were performed to prepare the data for downstream analysis. Size factors were estimated to normalize the data, and gene dispersions were calculated to account for variability across genes. Low-quality genes were filtered out by retaining only those expressed above a minimum threshold of 0.1 and detected in at least 10 cells.\n\nFor trajectory inference, genes were selected for ordering based on differential expression between experimental conditions. Genes with a q-value less than 0.01 were identified as significant and used to order cells along the trajectory. The dimensionality of the data was reduced using the DDRTree algorithm, which is well-suited for datasets with complex, branching trajectories. Cells were then ordered along the trajectory to calculate pseudotime.\n\nDifferential expression analysis was conducted to identify genes whose expression changed as a function of pseudotime. A natural smoothing spline (`sm.ns`) was used to model gene expression, and genes with a q-value less than 0.1 were considered significant. Additionally, branch analysis (BEAM) was performed to identify genes differentially expressed at a specific branch point in the trajectory. Genes with a q-value less than 1e-4 were retained as significant branch-dependent genes.\n\nVisualizations were generated to explore the results. The dispersion of ordering genes was plotted to assess their variability. The main cell trajectory was visualized, with cells colored by pseudotime and experimental condition to examine their progression and distribution. The expression of branch-dependent genes was also plotted along the branched trajectory to investigate their dynamics.\n\nFinally, the Monocle object containing all analysis results was saved as an RDS file for future reference. This comprehensive workflow provides a detailed approach to single-cell trajectory analysis, enabling the exploration of cellular dynamics and gene expression changes over pseudotime.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform single-cell trajectory analysis using Monocle2, the following steps were executed. The analysis began by loading the necessary libraries: `monocle` for trajectory analysis, `ggplot2` for advanced plotting, and `Matrix` for handling sparse matrices. These libraries are essential for data manipulation, analysis, and visualization in the context of single-cell RNA sequencing.\n\nThe gene expression matrix, cell metadata, and gene annotation files were imported. The gene expression matrix contained rows representing genes and columns representing cells, while the cell metadata included information about each cell, such as experimental conditions. The gene annotation file provided additional details about the genes. These datasets were read into R using `read.table` and `read.delim` functions, ensuring proper formatting and header inclusion.\n\nNext, the `CellDataSet` (CDS) object, the primary data structure in Monocle, was constructed. The gene expression matrix, cell metadata, and gene annotation were combined into `AnnotatedDataFrame` objects and passed to the `newCellDataSet` function. The `negbinomial.size()` expression family was specified to model UMI-based single-cell RNA-seq data, and a lower detection limit of 0.5 was set to filter out lowly expressed genes.\n\nPreprocessing and normalization were performed to prepare the data for downstream analysis. Size factors were estimated to normalize the data, and gene dispersions were calculated to account for variability across genes. Low-quality genes were filtered out by retaining only those expressed above a minimum threshold of 0.1 and detected in at least 10 cells.\n\nFor trajectory inference, genes were selected for ordering based on differential expression between experimental conditions. Genes with a q-value less than 0.01 were identified as significant and used to order cells along the trajectory. The dimensionality of the data was reduced using the DDRTree algorithm, which is well-suited for datasets with complex, branching trajectories. Cells were then ordered along the trajectory to calculate pseudotime.\n\nDifferential expression analysis was conducted to identify genes whose expression changed as a function of pseudotime. A natural smoothing spline (`sm.ns`) was used to model gene expression, and genes with a q-value less than 0.1 were considered significant. Additionally, branch analysis (BEAM) was performed to identify genes differentially expressed at a specific branch point in the trajectory. Genes with a q-value less than 1e-4 were retained as significant branch-dependent genes.\n\nVisualizations were generated to explore the results. The dispersion of ordering genes was plotted to assess their variability. The main cell trajectory was visualized, with cells colored by pseudotime and experimental condition to examine their progression and distribution. The expression of branch-dependent genes was also plotted along the branched trajectory to investigate their dynamics.\n\nFinally, the Monocle object containing all analysis results was saved as an RDS file for future reference. This comprehensive workflow provides a detailed approach to single-cell trajectory analysis, enabling the exploration of cellular dynamics and gene expression changes over pseudotime.",
         "Single-cell trajectory analysis was conducted using Monocle2 in R, leveraging libraries such as `monocle`, `ggplot2`, and `Matrix`. The workflow involved constructing a `CellDataSet` object from gene expression matrices, cell metadata, and gene annotations, with UMI-based data modeled using the `negbinomial.size()` family. Preprocessing included size factor estimation, gene dispersion calculation, and filtering of low-quality genes. Trajectory inference was performed by selecting differentially expressed genes (q-value < 0.01) and reducing dimensionality using the DDRTree algorithm. Pseudotime ordering and branch analysis (BEAM) identified genes with dynamic expression patterns (q-value < 1e-4). Visualizations included trajectory plots colored by pseudotime and branch-dependent gene expression dynamics."
        ],
        [
         "35",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Package Load\n#--------------------------------\n\n# Load required libraries for single-cell trajectory analysis\nlibrary(monocle3) # Core package for Monocle3 analysis\nlibrary(ggplot2) # For advanced plotting customization\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the C. elegans embryo dataset as an example.\n# NOTE: Modify these paths if using your own data.\nexpression_matrix <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_expression.rds\"))\ncell_metadata <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_colData.rds\"))\ngene_annotation <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_rowData.rds\"))\n\n# Create the main cell_data_set (CDS) object, which is Monocle3's primary data structure.\ncds <- new_cell_data_set(\nexpression_matrix,\ncell_metadata = cell_metadata,\ngene_metadata = gene_annotation\n)\n\n\n#--------------------------------\n# Preprocessing and Dimensionality Reduction\n#--------------------------------\n\n# Apply a standard preprocessing workflow to the CDS object.\n# This function normalizes, performs PCA, and prepares the data for downstream analysis.\ncds <- preprocess_cds(cds, num_dim = 50)\n\n# If the data has multiple batches, correct for batch effects.\n# `alignment_group` should point to the column in cell_metadata that specifies the batch.\ncds <- align_cds(cds, alignment_group = \"batch\")\n\n# Reduce the dimensionality further using UMAP. The results are stored in the CDS object.\ncds <- reduce_dimension(cds)\n\n\n#--------------------------------\n# Major Analysis Tasks: Trajectory Inference\n#--------------------------------\n\n# --- Clustering & Trajectory Construction ---\n# Cluster cells to identify distinct cell states or types.\ncds <- cluster_cells(cds)\n\n# Learn the principal graph that represents the developmental trajectory.\ncds <- learn_graph(cds)\n\n# Order the cells along the trajectory to calculate pseudotime.\n# Monocle3 will prompt to select root node(s) interactively if not specified.\ncds <- order_cells(cds)\n\n\n# --- Differential Expression Analysis Along Trajectory ---\n# Identify genes whose expression changes significantly as cells traverse the learned trajectory.\n# `graph_test` is a powerful function for finding trajectory-dependent genes.\ngraph_test_res <- graph_test(cds, neighbor_graph = \"principal_graph\", cores = 4)\n\n# Filter the results to get genes with a q-value below a significance threshold.\nsignificant_genes <- subset(graph_test_res, q_value < 0.05)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Plot the trajectory, coloring cells by the calculated pseudotime.\n# This visualizes the developmental progression of the cells.\nplot_cells(\ncds,\ncolor_cells_by = \"pseudotime\",\nlabel_cell_groups = FALSE,\nlabel_leaves = FALSE,\nlabel_branch_points = FALSE\n)\n\n# Plot the expression trends of the top 6 most significant trajectory-dependent genes.\n# This helps to understand the molecular changes occurring along the trajectory.\nplot_genes_in_pseudotime(\ncds,\ngenes = rownames(significant_genes)[1:6],\nmin_expr = 0.5\n)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the differential expression results to a CSV file.\nwrite.csv(significant_genes, \"differential_expression_results.csv\")\n\n# Save the final Monocle3 object, which contains all analysis results.\nsaveRDS(cds, file = \"monocle3_trajectory_object.rds\")\n",
         "To perform single-cell trajectory analysis using Monocle3, the following steps were executed. The analysis was conducted in R, leveraging the monocle3 and ggplot2 packages for trajectory inference and visualization, respectively.\n\nFirst, the necessary libraries were loaded. The monocle3 package was used as the core tool for single-cell trajectory analysis, while ggplot2 facilitated advanced plotting customization. These libraries are essential for handling single-cell data and generating interpretable visualizations.\n\nThe C. elegans embryo dataset was used as an example for this analysis. The dataset consisted of an expression matrix, cell metadata, and gene annotation, which were loaded from specified URLs. These components were combined into a cell_data_set (CDS) object, the primary data structure in Monocle3. The CDS object integrates gene expression data with metadata, enabling downstream analysis.\n\nPreprocessing and dimensionality reduction were performed to prepare the data for trajectory inference. The preprocessing step normalized the data, performed principal component analysis (PCA), and reduced the dimensionality to 50 components. If the dataset contained multiple batches, batch effects were corrected using the align_cds function, with the alignment_group parameter specifying the batch column in the cell metadata. Further dimensionality reduction was achieved using Uniform Manifold Approximation and Projection (UMAP), which embedded the data into a lower-dimensional space for visualization and analysis.\n\nThe major analysis tasks involved clustering cells, constructing a developmental trajectory, and ordering cells along the trajectory. Cells were clustered to identify distinct cell states or types using the cluster_cells function. A principal graph representing the developmental trajectory was learned using the learn_graph function. Cells were then ordered along the trajectory to calculate pseudotime, a measure of developmental progression. The root node(s) for pseudotime calculation were selected interactively if not specified.\n\nDifferential expression analysis was conducted to identify genes whose expression changed significantly along the trajectory. The graph_test function was employed to test for trajectory-dependent genes, with the neighbor_graph parameter set to \"principal_graph\" and the cores parameter set to 4 for parallel processing. Genes with a q-value below 0.05 were considered significant and retained for further analysis.\n\nVisualization was performed to interpret the results. The trajectory was plotted with cells colored by pseudotime, providing a visual representation of developmental progression. Additionally, the expression trends of the top six most significant trajectory-dependent genes were plotted to elucidate molecular changes occurring along the trajectory.\n\nFinally, the results were saved for further use. The differential expression results were exported to a CSV file, and the final Monocle3 object, containing all analysis results, was saved as an RDS file. This ensures that the analysis can be reproduced and extended in future studies.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform single-cell trajectory analysis using Monocle3, the following steps were executed. The analysis was conducted in R, leveraging the monocle3 and ggplot2 packages for trajectory inference and visualization, respectively.\n\nFirst, the necessary libraries were loaded. The monocle3 package was used as the core tool for single-cell trajectory analysis, while ggplot2 facilitated advanced plotting customization. These libraries are essential for handling single-cell data and generating interpretable visualizations.\n\nThe C. elegans embryo dataset was used as an example for this analysis. The dataset consisted of an expression matrix, cell metadata, and gene annotation, which were loaded from specified URLs. These components were combined into a cell_data_set (CDS) object, the primary data structure in Monocle3. The CDS object integrates gene expression data with metadata, enabling downstream analysis.\n\nPreprocessing and dimensionality reduction were performed to prepare the data for trajectory inference. The preprocessing step normalized the data, performed principal component analysis (PCA), and reduced the dimensionality to 50 components. If the dataset contained multiple batches, batch effects were corrected using the align_cds function, with the alignment_group parameter specifying the batch column in the cell metadata. Further dimensionality reduction was achieved using Uniform Manifold Approximation and Projection (UMAP), which embedded the data into a lower-dimensional space for visualization and analysis.\n\nThe major analysis tasks involved clustering cells, constructing a developmental trajectory, and ordering cells along the trajectory. Cells were clustered to identify distinct cell states or types using the cluster_cells function. A principal graph representing the developmental trajectory was learned using the learn_graph function. Cells were then ordered along the trajectory to calculate pseudotime, a measure of developmental progression. The root node(s) for pseudotime calculation were selected interactively if not specified.\n\nDifferential expression analysis was conducted to identify genes whose expression changed significantly along the trajectory. The graph_test function was employed to test for trajectory-dependent genes, with the neighbor_graph parameter set to \"principal_graph\" and the cores parameter set to 4 for parallel processing. Genes with a q-value below 0.05 were considered significant and retained for further analysis.\n\nVisualization was performed to interpret the results. The trajectory was plotted with cells colored by pseudotime, providing a visual representation of developmental progression. Additionally, the expression trends of the top six most significant trajectory-dependent genes were plotted to elucidate molecular changes occurring along the trajectory.\n\nFinally, the results were saved for further use. The differential expression results were exported to a CSV file, and the final Monocle3 object, containing all analysis results, was saved as an RDS file. This ensures that the analysis can be reproduced and extended in future studies.",
         "Single-cell trajectory analysis was conducted in R using the monocle3 package, with ggplot2 employed for visualization. The C. elegans embryo dataset, comprising an expression matrix, cell metadata, and gene annotation, was integrated into a cell_data_set (CDS) object. Preprocessing included normalization, principal component analysis (PCA), and dimensionality reduction via UMAP, with batch correction applied if necessary. Cells were clustered to identify distinct states, and a principal graph was constructed to infer developmental trajectories. Pseudotime was calculated to order cells along the trajectory, and differential expression analysis identified genes significantly associated with developmental progression. Results were visualized through trajectory plots and gene expression trends, providing insights into molecular changes during development."
        ],
        [
         "36",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport pandas as pd\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport seaborn as sns\n\nimport scanpy as sc\n\nimport pertpy as pt # Pertpy includes tools for compositional analysis (scCODA)\n\n\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsns.set_theme()\n\n\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the Haber et al. (2017) dataset of small intestinal epithelial cells.\n\n# This dataset contains cell type labels and condition information suitable for this analysis.\n\nprint(\"Loading Haber et al. 2017 dataset...\")\n\nadata = pt.dt.haber_2017_regions()\n\nprint(\"Loaded data shape:\", adata.shape)\n\n\n\n\n\n#--------------------------------\n\n# Illustrative Example of Compositional Data\n\n#--------------------------------\n\n\n\n# This section demonstrates the core concept of compositional data analysis.\n\nprint(\"\\n--- Illustrating Compositional Data Concepts ---\")\n\n# Define example cell counts for three cell types in healthy and diseased tissue.\n\nhealthy_tissue = [2000, 2000, 2000]\n\ndiseased_tissue = [4000, 2000, 2000] # Note: Cell type A has doubled.\n\n\n\n# Create a DataFrame with the total cell counts.\n\nexample_data_global = pd.DataFrame(\n\ndata=np.array([healthy_tissue, diseased_tissue]),\n\nindex=[\"Healthy\", \"Diseased\"],\n\ncolumns=[\"A\", \"B\", \"C\"]\n\n)\n\n\n\n# Reshape the DataFrame for easy plotting with seaborn.\n\nplot_data_global = example_data_global.reset_index().melt(\n\nid_vars=\"index\", value_vars=[\"A\", \"B\", \"C\"],\n\nvar_name=\"Cell type\", value_name=\"count\"\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Compositional Analysis with scCODA\n\n#--------------------------------\n\n\n\n# --- Model and Data Initialization ---\n\n# Instantiate the scCODA model from pertpy.\n\nsccoda_model = pt.tl.Sccoda()\n\n\n\n# Load the data into the scCODA model structure.\n\n# This function automatically aggregates cell-level data to the sample level.\n\nprint(\"\\n--- Setting up scCODA Analysis ---\")\n\nsccoda_data = sccoda_model.load(\n\nadata,\n\ntype=\"cell_level\",\n\ngenerate_sample_level=True,\n\ncell_type_identifier=\"cell_label\",\n\nsample_identifier=\"batch\",\n\ncovariate_obs=[\"condition\"],\n\n)\n\nprint(\"scCODA data modalities:\", sccoda_data)\n\n\n\n# --- Model Configuration and Execution ---\n\n# Prepare the data for the model, specifying the statistical formula and a reference cell type.\n\n# The model will calculate changes relative to this reference.\n\nsccoda_data = sccoda_model.prepare(\n\nsccoda_data,\n\nmodality_key=\"coda\",\n\nformula=\"condition\",\n\nreference_cell_type=\"Endocrine\"\n\n)\n\n\n\n# Run the Bayesian inference sampler (NUTS) to fit the model.\n\n# `rng_key` ensures reproducibility.\n\nprint(\"\\nRunning Bayesian inference model...\")\n\nsccoda_model.run_nuts(sccoda_data, modality_key=\"coda\", rng_key=1234)\n\n\n\n# Set the False Discovery Rate (FDR) threshold for identifying credible effects.\n\nsccoda_model.set_fdr(sccoda_data, est_fdr=0.2)\n\n\n\n# Extract and print the credible effects (significant changes) for each cell type.\n\ncredible_effects = sccoda_model.credible_effects(sccoda_data, modality_key=\"coda\")\n\nprint(\"\\nCredible Effects (Significant Changes):\\n\", credible_effects)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Illustrative Example Plots ---\n\n# Plot barplots showing the global abundances from the illustrative example.\n\nfig, ax = plt.subplots(1, 2, figsize=(12, 6))\n\nfig.suptitle(\"Illustrative Example: Global Cell Abundances\")\n\nsns.barplot(data=plot_data_global, x=\"index\", y=\"count\", hue=\"Cell type\", ax=ax[0])\n\nax[0].set_title(\"Counts by Status\")\n\nsns.barplot(data=plot_data_global, x=\"Cell type\", y=\"count\", hue=\"index\", ax=ax[1])\n\nax[1].set_title(\"Counts by Cell Type\")\n\nplt.tight_layout(rect=[0, 0.03, 1, 0.95])\n\nplt.show()\n\n\n\n# --- scCODA Diagnostic Plots ---\n\n# Create boxplots of cell type counts across conditions to visualize distributions.\n\nprint(\"\\n--- Visualizing Compositional Data ---\")\n\nsccoda_model.plot_boxplots(\n\nsccoda_data,\n\nmodality_key=\"coda\",\n\nfeature_name=\"condition\",\n\nfigsize=(12, 5),\n\nadd_dots=True,\n\nargs_swarmplot={\"palette\": [\"red\"]},\n\n)\n\nplt.show()\n\n\n\n# Create a stacked barplot to show cell type proportions for each condition.\n\nsccoda_model.plot_stacked_barplot(\n\nsccoda_data,\n\nmodality_key=\"coda\",\n\nfeature_name=\"condition\",\n\nfigsize=(4, 2)\n\n)\n\nplt.show()\n\n\n\n# --- scCODA Final Effects Plot ---\n\n# Generate a bar plot of the final effects. This shows the log-fold changes of cell types\n\n# relative to the reference, highlighting significant changes.\n\nsccoda_model.plot_effects_barplot(sccoda_data, \"coda\", \"condition\")\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object and the scCODA results object.\n\nprint(\"\\nSaving results...\")\n\nadata.write(\"haber_2017_analyzed.h5ad\")\n\nsccoda_data.write(\"haber_2017_sccoda_results.h5ad\")\n\n\n\nprint(\"\\nCompositional analysis completed!\")\n\n",
         "The analysis begins by importing essential Python libraries for data handling, analysis, and visualization. These include `pandas` and `numpy` for data manipulation, `matplotlib` and `seaborn` for plotting, and `scanpy` and `pertpy` for single-cell data analysis, with `pertpy` specifically providing tools for compositional analysis (scCODA). Global plotting parameters are set using `seaborn` to ensure consistent and clear visualizations.\n\nThe dataset used in this analysis is the Haber et al. (2017) dataset, which contains small intestinal epithelial cells with cell type labels and condition information. This dataset is loaded using the `pertpy` function `haber_2017_regions()`, and its shape is printed to confirm successful loading.\n\nTo illustrate the core concept of compositional data analysis, an example dataset is created, representing cell counts for three cell types in healthy and diseased tissue. The cell counts are organized into a `pandas` DataFrame, which is then reshaped for plotting. This example demonstrates how changes in cell type proportions can be visualized and analyzed.\n\nThe primary analysis task involves compositional analysis using the scCODA model. The scCODA model is instantiated from `pertpy`, and the dataset is loaded into the model structure. The data is aggregated from the cell level to the sample level, with cell type and sample identifiers specified. Covariates such as condition are included to account for experimental variables. The model is prepared by specifying a statistical formula and a reference cell type (Endocrine), which serves as the baseline for calculating changes in other cell types.\n\nBayesian inference is performed using the No-U-Turn Sampler (NUTS) to fit the model, with a random number generator key (`rng_key=1234`) ensuring reproducibility. A False Discovery Rate (FDR) threshold of 0.2 is set to identify credible effects, and significant changes in cell type proportions are extracted and printed.\n\nVisualizations are generated to support the analysis. First, bar plots are created to illustrate the global cell abundances from the example dataset, showing counts by status and cell type. Next, diagnostic plots are produced, including boxplots of cell type counts across conditions and a stacked bar plot showing cell type proportions for each condition. Finally, a bar plot of the final effects is generated, displaying log-fold changes of cell types relative to the reference, with significant changes highlighted.\n\nThe results are saved in two formats: the analyzed dataset is saved as an AnnData object (`haber_2017_analyzed.h5ad`), and the scCODA results are saved as a separate AnnData object (`haber_2017_sccoda_results.h5ad`). The analysis concludes with a confirmation message indicating successful completion.\n\nThis workflow provides a comprehensive approach to compositional analysis of single-cell data, from data loading and preprocessing to model fitting, visualization, and result saving.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for data handling, analysis, and visualization. These include `pandas` and `numpy` for data manipulation, `matplotlib` and `seaborn` for plotting, and `scanpy` and `pertpy` for single-cell data analysis, with `pertpy` specifically providing tools for compositional analysis (scCODA). Global plotting parameters are set using `seaborn` to ensure consistent and clear visualizations.\n\nThe dataset used in this analysis is the Haber et al. (2017) dataset, which contains small intestinal epithelial cells with cell type labels and condition information. This dataset is loaded using the `pertpy` function `haber_2017_regions()`, and its shape is printed to confirm successful loading.\n\nTo illustrate the core concept of compositional data analysis, an example dataset is created, representing cell counts for three cell types in healthy and diseased tissue. The cell counts are organized into a `pandas` DataFrame, which is then reshaped for plotting. This example demonstrates how changes in cell type proportions can be visualized and analyzed.\n\nThe primary analysis task involves compositional analysis using the scCODA model. The scCODA model is instantiated from `pertpy`, and the dataset is loaded into the model structure. The data is aggregated from the cell level to the sample level, with cell type and sample identifiers specified. Covariates such as condition are included to account for experimental variables. The model is prepared by specifying a statistical formula and a reference cell type (Endocrine), which serves as the baseline for calculating changes in other cell types.\n\nBayesian inference is performed using the No-U-Turn Sampler (NUTS) to fit the model, with a random number generator key (`rng_key=1234`) ensuring reproducibility. A False Discovery Rate (FDR) threshold of 0.2 is set to identify credible effects, and significant changes in cell type proportions are extracted and printed.\n\nVisualizations are generated to support the analysis. First, bar plots are created to illustrate the global cell abundances from the example dataset, showing counts by status and cell type. Next, diagnostic plots are produced, including boxplots of cell type counts across conditions and a stacked bar plot showing cell type proportions for each condition. Finally, a bar plot of the final effects is generated, displaying log-fold changes of cell types relative to the reference, with significant changes highlighted.\n\nThe results are saved in two formats: the analyzed dataset is saved as an AnnData object (`haber_2017_analyzed.h5ad`), and the scCODA results are saved as a separate AnnData object (`haber_2017_sccoda_results.h5ad`). The analysis concludes with a confirmation message indicating successful completion.\n\nThis workflow provides a comprehensive approach to compositional analysis of single-cell data, from data loading and preprocessing to model fitting, visualization, and result saving.",
         "The analysis was conducted in Python using `pandas`, `numpy`, `matplotlib`, `seaborn`, `scanpy`, and `pertpy`, with a focus on compositional analysis of single-cell RNA sequencing data. The Haber et al. (2017) dataset, comprising small intestinal epithelial cells with annotated cell types and conditions, was utilized. The scCODA model from `pertpy` was employed to analyze changes in cell type proportions between healthy and diseased tissue. Bayesian inference was performed using the No-U-Turn Sampler (NUTS), with a False Discovery Rate (FDR) threshold of 0.2 to identify significant effects. Visualizations included bar plots of cell abundances, diagnostic boxplots, and stacked bar plots of cell type proportions. Results were saved as AnnData objects, providing a robust framework for compositional analysis in single-cell datasets."
        ],
        [
         "37",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport pandas as pd\n\nimport matplotlib.pyplot as plt\n\nimport scanpy as sc\n\nimport seaborn as sns\n\nimport pertpy as pt # Pertpy provides both scCODA and Milo\n\n\n\n#--------------------------------\n\n# Data Input & Preprocessing\n\n#--------------------------------\n\n\n\n# Load the Haber et al. (2017) dataset.\n\nprint(\"Loading Haber et al. 2017 dataset...\")\n\nadata = pt.dt.haber_2017_regions()\n\nprint(\"Loaded data shape:\", adata.shape)\n\n\n\n# --- Standard scRNA-seq preprocessing ---\n\n# Save raw counts into a new layer for reference.\n\nadata.layers[\"counts\"] = adata.X.copy()\n\n\n\n# Create log-transformed counts for analysis.\n\nadata.layers[\"logcounts\"] = np.log1p(adata.layers[\"counts\"])\n\nadata.X = adata.layers[\"logcounts\"].copy()\n\n\n\n# Identify highly variable genes to focus on biological signal.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=3000, subset=False)\n\n\n\n# Perform PCA on the highly variable genes.\n\nsc.pp.pca(adata)\n\n\n\n# Construct the k-nearest neighbor (KNN) graph. This is essential for defining neighborhoods.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=30)\n\n\n\n# Compute a UMAP embedding for visualization.\n\nsc.tl.umap(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Differential Abundance with Milo\n\n#--------------------------------\n\n\n\n# --- Model and Data Initialization ---\n\n# Instantiate the Milo model from pertpy.\n\nmilo = pt.tl.Milo()\n\n\n\n# Load the preprocessed AnnData object into the Milo framework.\n\n# This creates a MuData object, which can hold multiple data modalities.\n\nprint(\"\\n--- Setting up Milo Analysis ---\")\n\nmdata = milo.load(adata)\n\n\n\n# --- Define Cell Neighborhoods ---\n\n# Define cell neighborhoods based on the KNN graph. These are overlapping groups\n\n# of cells that represent a fine-grained partitioning of the data.\n\n# `prop=0.1` means 10% of cells are sampled as \"index\" cells to define the neighborhoods.\n\nmilo.make_nhoods(mdata, prop=0.1)\n\n\n\n# --- Count Cells in Neighborhoods ---\n\n# Count the number of cells from each sample ('batch') within each neighborhood.\n\n# This creates the count matrix that will be used for differential abundance testing.\n\nmilo.count_nhoods(mdata, sample_col=\"batch\")\n\n\n\n# --- Build Neighborhood Graph ---\n\n# Create a graph connecting overlapping neighborhoods for visualization and analysis.\n\nmilo.build_nhood_graph(mdata)\n\n\n\n# --- Differential Abundance Testing ---\n\n# Perform the DA test to compare conditions (e.g., Salmonella vs. Control).\n\n# The design formula `~condition` tells the model to test for changes related to the 'condition' variable.\n\nprint(\"\\nPerforming differential abundance testing...\")\n\nmilo.da_nhoods(mdata, design=\"~condition\", model_contrasts=\"conditionSalmonella-conditionControl\")\n\n\n\n# Retrieve the DA results from the Milo object.\n\nmilo_results = mdata[\"milo\"].obs.copy()\n\nprint(\"\\nDifferential Abundance (DA) results (first 5 neighborhoods):\")\n\nprint(milo_results.head())\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Raw Data ---\n\n# Visualize the initial UMAP colored by condition and batch to see the overall data structure.\n\nsc.pl.umap(adata, color=[\"condition\", \"batch\"], ncols=2, wspace=0.4, title=[\"Condition\", \"Batch\"])\n\n\n\n# --- Neighborhood Size Distribution ---\n\n# Plot a histogram of the number of cells in each neighborhood.\n\nprint(\"\\n--- Visualizing Milo Results ---\")\n\nnhood_size = adata.obsm[\"nhoods\"].toarray().sum(axis=0)\n\nplt.figure(figsize=(6, 4))\n\nplt.hist(nhood_size, bins=20, edgecolor=\"black\")\n\nplt.xlabel(\"# cells in each neighborhood\")\n\nplt.ylabel(\"Number of neighborhoods\")\n\nplt.title(\"Distribution of Neighborhood Sizes\")\n\nplt.show()\n\n\n\n# --- Neighborhood Graph ---\n\n# Visualize the graph of overlapping neighborhoods.\n\nwith plt.rc_context({\"figure.figsize\": [10, 10]}):\n\nmilo.plot_nhood_graph(mdata, alpha=0.1, min_size=5, plot_edges=False)\n\nplt.show()\n\n\n\n# --- DA Beeswarm Plot ---\n\n# Create a beeswarm plot to visualize the DA results across all neighborhoods.\n\n# This plot shows the log-fold change and significance for each neighborhood.\n\nmilo.plot_da_beeswarm(mdata)\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the MuData object, which contains the original data and all Milo results.\n\nprint(\"\\nSaving results...\")\n\nmdata.write(\"haber_2017_milo_results.h5mu\")\n\n\n\nprint(\"\\nDifferential abundance analysis with Milo completed!\")\n\n",
         "The analysis begins by importing essential Python libraries for data handling, analysis, and visualization, including NumPy, Pandas, Matplotlib, Scanpy, Seaborn, and Pertpy. Pertpy is particularly important as it provides tools for single-cell analysis, including the Milo framework for differential abundance testing.\n\nThe dataset used in this analysis is from Haber et al. (2017), which is loaded using the Pertpy library. The dataset is preprocessed to prepare it for downstream analysis. Raw counts are stored in a separate layer for reference, and log-transformed counts are generated to stabilize variance and improve interpretability. Highly variable genes are identified to focus on biologically relevant signals, and principal component analysis (PCA) is performed to reduce dimensionality. A k-nearest neighbor (KNN) graph is constructed to define cell neighborhoods, and a UMAP embedding is computed for visualization purposes.\n\nThe primary analysis task involves differential abundance testing using the Milo framework. The Milo model is initialized, and the preprocessed data is loaded into the Milo framework, creating a MuData object that can hold multiple data modalities. Cell neighborhoods are defined by sampling 10% of cells as index cells, and the number of cells from each sample within each neighborhood is counted to create a count matrix for differential abundance testing. A graph connecting overlapping neighborhoods is built for visualization and analysis. Differential abundance testing is performed using a design formula that compares conditions (e.g., Salmonella vs. Control), and the results are retrieved and examined.\n\nVisualizations are generated to interpret the results. The initial UMAP is plotted to show the overall data structure, colored by condition and batch. A histogram of neighborhood sizes is created to understand the distribution of cells across neighborhoods. The neighborhood graph is visualized to show the connections between overlapping neighborhoods. A beeswarm plot is generated to display the differential abundance results, showing the log-fold change and significance for each neighborhood.\n\nFinally, the MuData object, which contains the original data and all Milo results, is saved to a file. This ensures that the results can be revisited or shared for further analysis. The differential abundance analysis with Milo is then completed, providing insights into the changes in cell abundance across different conditions.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for data handling, analysis, and visualization, including NumPy, Pandas, Matplotlib, Scanpy, Seaborn, and Pertpy. Pertpy is particularly important as it provides tools for single-cell analysis, including the Milo framework for differential abundance testing.\n\nThe dataset used in this analysis is from Haber et al. (2017), which is loaded using the Pertpy library. The dataset is preprocessed to prepare it for downstream analysis. Raw counts are stored in a separate layer for reference, and log-transformed counts are generated to stabilize variance and improve interpretability. Highly variable genes are identified to focus on biologically relevant signals, and principal component analysis (PCA) is performed to reduce dimensionality. A k-nearest neighbor (KNN) graph is constructed to define cell neighborhoods, and a UMAP embedding is computed for visualization purposes.\n\nThe primary analysis task involves differential abundance testing using the Milo framework. The Milo model is initialized, and the preprocessed data is loaded into the Milo framework, creating a MuData object that can hold multiple data modalities. Cell neighborhoods are defined by sampling 10% of cells as index cells, and the number of cells from each sample within each neighborhood is counted to create a count matrix for differential abundance testing. A graph connecting overlapping neighborhoods is built for visualization and analysis. Differential abundance testing is performed using a design formula that compares conditions (e.g., Salmonella vs. Control), and the results are retrieved and examined.\n\nVisualizations are generated to interpret the results. The initial UMAP is plotted to show the overall data structure, colored by condition and batch. A histogram of neighborhood sizes is created to understand the distribution of cells across neighborhoods. The neighborhood graph is visualized to show the connections between overlapping neighborhoods. A beeswarm plot is generated to display the differential abundance results, showing the log-fold change and significance for each neighborhood.\n\nFinally, the MuData object, which contains the original data and all Milo results, is saved to a file. This ensures that the results can be revisited or shared for further analysis. The differential abundance analysis with Milo is then completed, providing insights into the changes in cell abundance across different conditions.",
         "The analysis was conducted in Python using libraries such as NumPy, Pandas, Matplotlib, Scanpy, Seaborn, and Pertpy, with a focus on the Milo framework for differential abundance testing. The dataset from Haber et al. (2017) was preprocessed, including log transformation of counts, identification of highly variable genes, and dimensionality reduction via PCA. A k-nearest neighbor graph and UMAP embedding were constructed for visualization. Differential abundance testing was performed using Milo, where cell neighborhoods were defined, and a count matrix was generated to compare conditions. Results were visualized through UMAP plots, neighborhood graphs, and beeswarm plots, providing insights into changes in cell abundance across conditions."
        ],
        [
         "38",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#!/usr/bin/env python\n\"\"\"\nExample script for gene set enrichment and pathway analysis for single\\cell RNA\\seq data.\nThis script reads an AnnData object, preprocesses the data, performs a t\\test to rank genes,\nextracts DE t-statistics for a selected cell type/condition, loads a Reactome GMT file, runs GSEA via decoupler,\nand plots the top enriched pathways.\n\"\n#------------------------------------------------------------------------------#\n# Import necessary modules #\n#------------------------------------------------------------------------------#\nfrom __future__ import annotations\nimport os\nfrom pathlib import Path\nimport numpy as np\nimport pandas as pd\nimport scanpy as sc\nimport decoupler\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n#------------------------------------------------------------------------------#\n# Script Configuration #\n#------------------------------------------------------------------------------#\n# Set global parameters for plotting and figure aesthetics\nsns.set(style=\"whitegrid\")\nsc.settings.set_figure_params(dpi=200, frameon=False, figsize=(4, 4))\n\n#------------------------------------------------------------------------------#\n# Data Input #\n#------------------------------------------------------------------------------#\n# Read the PBMC dataset from a .h5ad file.\n# A backup URL is provided for reproducibility if the file is not found locally.\nadata = sc.read(\n\"kang_counts_25k.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/34464122\"\n)\n\n# Save the raw counts in a separate layer for later use, a common best practice.\nadata.layers[\"counts\"] = adata.X.copy()\n\n# Rename the \"label\" column to \"condition\" for better clarity in downstream steps.\nadata.obs = adata.obs.rename(columns={\"label\": \"condition\"})\n\n#------------------------------------------------------------------------------#\n# Data Preprocessing #\n#------------------------------------------------------------------------------#\n# --- Normalization and Feature Selection ---\n# Normalize total counts per cell and apply a log transformation.\nsc.pp.normalize_total(adata)\nsc.pp.log1p(adata)\n\n# Identify the top 4000 highly variable genes using the \"seurat_v3\" method.\n# `layer=\"counts\"` ensures this is calculated on the untransformed data.\nsc.pp.highly_variable_genes(\nadata,\nn_top_genes=4000,\nflavor=\"seurat_v3\",\nsubset=False,\nlayer=\"counts\"\n)\n\n# --- Dimensionality Reduction for Visualization ---\n# Compute PCA, build a neighbor graph, and generate a UMAP representation.\nsc.pp.pca(adata)\nsc.pp.neighbors(adata)\nsc.tl.umap(adata)\n\n#------------------------------------------------------------------------------#\n# Major Analysis Task: Enrichment Analysis #\n#------------------------------------------------------------------------------#\n\n# --- 1. Differential Expression Analysis ---\n# Create a combined 'group' column (e.g., \"stim_FCGR3A+ Monocytes\") to perform\n# DE analysis on specific cell type/condition combinations.\nadata.obs[\"group\"] = adata.obs[\"condition\"].astype(str) + \"_\" + adata.obs[\"cell_type\"]\n\n# Run a t-test to rank genes based on differential expression across all groups.\nsc.tl.rank_genes_groups(adata, groupby=\"group\", method=\"t-test\", key_added=\"t-test\")\n\n# Define the target group for GSEA and extract its DE results.\ntarget_group = \"stim_FCGR3A+ Monocytes\"\nde_df = sc.get.rank_genes_groups_df(adata, group=target_group, key=\"t-test\")\n\n# Filter for highly variable genes and create a DataFrame of t-statistics.\nhvg_genes = adata.var_names[adata.var[\"highly_variable\"]]\nt_stats = de_df.set_index(\"names\").loc[hvg_genes].sort_values(\"scores\", key=np.abs, ascending=False)[[\"scores\"]]\nt_stats = t_stats.rename(columns={\"scores\": target_group})\n\n# --- 2. Prepare Pathway Gene Sets (Reactome GMT) ---\n# Define the path for the GMT file and download if it doesn't exist.\ngmt_path = Path(\"c2.cp.reactome.v7.5.1.symbols.gmt\")\nif not gmt_path.is_file():\nos.system(\"wget -O c2.cp.reactome.v7.5.1.symbols.gmt https://figshare.com/ndownloader/files/35233771\")\n\ndef gmt_to_decoupler(pth: Path) -> pd.DataFrame:\n\"\"\"Parse a GMT file into a long-format DataFrame compatible with decoupler.\"\"\"\npathways = {}\nwith pth.open(\"r\") as f:\nfor line in f:\nparts = line.strip().split(\"\\t\")\nname, _, *genes = parts\npathways[name] = genes\n# Convert dictionary to a long-format DataFrame\nreturn pd.DataFrame(\n[(gene_set, gene) for gene_set, genes in pathways.items() for gene in genes],\ncolumns=[\"geneset\", \"genesymbol\"]\n)\n\n# Parse the GMT file and filter gene sets by size for robustness.\nreactome = gmt_to_decoupler(gmt_path)\ngeneset_size = reactome.groupby(\"geneset\").size()\ngsea_genesets = geneset_size.index[(geneset_size > 15) & (geneset_size < 500)]\nreactome_filtered = reactome[reactome[\"geneset\"].isin(gsea_genesets)]\n\n# --- 3. Run GSEA with Decoupler ---\n# Run GSEA using the gene-level t-statistics and the filtered Reactome pathways.\nscores, norm, pvals = decoupler.run_gsea(\nmat=t_stats.T,\nnet=reactome_filtered,\nsource=\"geneset\",\ntarget=\"genesymbol\"\n)\n\n# Combine the GSEA results into a single, sorted DataFrame.\ngsea_results = pd.concat({\n\"score\": scores.T,\n\"norm_score\": norm.T,\n# \"pval\": pvals.T\n#}, axis=1).droplevel(level=1, axis=1).sort_values(\"pval\")\n}, axis=1).droplevel(level=1, axis=1).sort_values(\"norm_score\", ascending=False)\ngsea_results.index.name = \"Pathway\"\n\n#------------------------------------------------------------------------------#\n# Plotting / Visualization #\n#------------------------------------------------------------------------------#\n# Plot UMAP embeddings colored by condition and cell type to visualize the data structure.\nsc.pl.umap(adata, color=[\"condition\", \"cell_type\"], frameon=False, ncols=2)\n\n# Plot the top 20 enriched pathways based on their normalized enrichment score.\ntop20 = gsea_results.head(20).reset_index()\n\nplt.figure(figsize=(10, 8))\nsns.barplot(data=top20, x=\"norm_score\", y=\"Pathway\", palette=\"viridis\")\nplt.xlabel(\"Normalized Enrichment Score\")\nplt.ylabel(\"Pathway\")\nplt.title(f\"Top 20 Enriched Pathways in {target_group}\")\nplt.tight_layout()\nplt.show()\n\n#------------------------------------------------------------------------------#\n# Data Output #\n#------------------------------------------------------------------------------#\n# Save the final GSEA results to a CSV file.\nprint(\"\\nSaving GSEA results...\")\ngsea_results.to_csv(\"gsea_results.csv\")\n\n# Save the AnnData object containing the DE analysis results.\nadata.write(\"kang_processed_with_de.h5ad\")\n\nprint(\"\\nGene set enrichment analysis completed!\")\n\n",
         "The following is a step-by-step guide for performing gene set enrichment and pathway analysis on single-cell RNA-seq data using Python. The analysis involves preprocessing the data, identifying differentially expressed genes, and performing gene set enrichment analysis (GSEA) using the Reactome pathway database. The results are visualized and saved for further use.\n\nFirst, the necessary Python modules are imported, including `scanpy` for single-cell data analysis, `decoupler` for gene set enrichment analysis, and `matplotlib` and `seaborn` for visualization. Global parameters for plotting aesthetics are set to ensure consistent and high-quality figures.\n\nThe analysis begins by loading the single-cell RNA-seq dataset from a `.h5ad` file. If the file is not found locally, it is downloaded from a backup URL. The raw counts are saved in a separate layer for later use, and the column containing cell labels is renamed to \"condition\" for clarity. The data is then preprocessed by normalizing total counts per cell and applying a log transformation. Highly variable genes are identified using the \"seurat_v3\" method, and dimensionality reduction is performed using PCA, neighbor graph construction, and UMAP embedding for visualization.\n\nFor the main analysis, a combined column is created to represent specific cell type and condition combinations. A t-test is performed to rank genes based on differential expression across these groups. The results for a target group (e.g., \"stim_FCGR3A+ Monocytes\") are extracted, and t-statistics are computed for highly variable genes. These t-statistics are used as input for GSEA.\n\nThe Reactome pathway gene sets are prepared by downloading a GMT file if it is not already available. The GMT file is parsed into a long-format DataFrame compatible with `decoupler`, and gene sets are filtered to retain those with a size between 15 and 500 genes for robustness. GSEA is then performed using the t-statistics and the filtered Reactome pathways. The results are combined into a single DataFrame, sorted by normalized enrichment scores.\n\nVisualization includes generating UMAP plots colored by condition and cell type to explore the data structure. The top 20 enriched pathways are plotted as a bar chart based on their normalized enrichment scores. Finally, the GSEA results are saved to a CSV file, and the processed AnnData object, including differential expression analysis results, is saved for further use.\n\nThis workflow provides a comprehensive approach to analyzing single-cell RNA-seq data, identifying differentially expressed genes, and interpreting the results through pathway enrichment analysis.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following is a step-by-step guide for performing gene set enrichment and pathway analysis on single-cell RNA-seq data using Python. The analysis involves preprocessing the data, identifying differentially expressed genes, and performing gene set enrichment analysis (GSEA) using the Reactome pathway database. The results are visualized and saved for further use.\n\nFirst, the necessary Python modules are imported, including `scanpy` for single-cell data analysis, `decoupler` for gene set enrichment analysis, and `matplotlib` and `seaborn` for visualization. Global parameters for plotting aesthetics are set to ensure consistent and high-quality figures.\n\nThe analysis begins by loading the single-cell RNA-seq dataset from a `.h5ad` file. If the file is not found locally, it is downloaded from a backup URL. The raw counts are saved in a separate layer for later use, and the column containing cell labels is renamed to \"condition\" for clarity. The data is then preprocessed by normalizing total counts per cell and applying a log transformation. Highly variable genes are identified using the \"seurat_v3\" method, and dimensionality reduction is performed using PCA, neighbor graph construction, and UMAP embedding for visualization.\n\nFor the main analysis, a combined column is created to represent specific cell type and condition combinations. A t-test is performed to rank genes based on differential expression across these groups. The results for a target group (e.g., \"stim_FCGR3A+ Monocytes\") are extracted, and t-statistics are computed for highly variable genes. These t-statistics are used as input for GSEA.\n\nThe Reactome pathway gene sets are prepared by downloading a GMT file if it is not already available. The GMT file is parsed into a long-format DataFrame compatible with `decoupler`, and gene sets are filtered to retain those with a size between 15 and 500 genes for robustness. GSEA is then performed using the t-statistics and the filtered Reactome pathways. The results are combined into a single DataFrame, sorted by normalized enrichment scores.\n\nVisualization includes generating UMAP plots colored by condition and cell type to explore the data structure. The top 20 enriched pathways are plotted as a bar chart based on their normalized enrichment scores. Finally, the GSEA results are saved to a CSV file, and the processed AnnData object, including differential expression analysis results, is saved for further use.\n\nThis workflow provides a comprehensive approach to analyzing single-cell RNA-seq data, identifying differentially expressed genes, and interpreting the results through pathway enrichment analysis.",
         "Gene set enrichment and pathway analysis of single-cell RNA-seq data was conducted using Python with `scanpy` for data preprocessing and analysis, `decoupler` for gene set enrichment, and `matplotlib` and `seaborn` for visualization. The workflow involved normalization of raw counts, identification of highly variable genes, and dimensionality reduction via PCA and UMAP. Differential expression analysis was performed using a t-test to rank genes across cell type and condition combinations. Reactome pathway gene sets were filtered and used for gene set enrichment analysis (GSEA), with results visualized through UMAP plots and bar charts of enriched pathways. The final outputs included GSEA results and a processed dataset for further exploration."
        ],
        [
         "39",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\nimport pandas as pd\nimport numpy as np\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport scanpy as sc\nimport liana as li # For ligand-receptor interaction inference\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the single-cell transcriptomics dataset.\n# A backup URL is provided for reproducibility.\nprint(\"Loading Kang et al. PBMC dataset...\")\nadata = sc.read(\n\"kang_counts_25k.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/34464122\"\n)\n\n# Store the raw counts in a separate layer before normalization.\nadata.layers[\"counts\"] = adata.X.copy()\n\n#--------------------------------\n# Preprocessing & Quality Control\n#--------------------------------\n\n# Apply basic quality control filters to remove low-quality cells and genes.\nprint(\"Performing quality control...\")\nsc.pp.filter_cells(adata, min_genes=200)\nsc.pp.filter_genes(adata, min_cells=3)\n\n# Normalize each cell by total counts and apply a log transformation.\n# This is a standard preprocessing pipeline for many single-cell analyses.\nsc.pp.normalize_total(adata)\nsc.pp.log1p(adata)\n\n#--------------------------------\n# Major Analysis Task: Ligand-Receptor Interaction Analysis\n#--------------------------------\n\n# --- Define Cell Populations of Interest ---\n# Specify the cell types that will act as senders (source) and receivers (target)\n# in the communication analysis.\nsource_cell_types = [\"CD4 T cells\", \"B cells\", \"FCGR3A+ Monocytes\"]\ntarget_cell_types = [\"CD8 T cells\", \"NK cells\"]\n\n# --- Run LIANA with CellPhoneDB Method ---\n# Use the CellPhoneDB method within the LIANA framework to infer interactions.\n# `groupby=\"cell_type\"` specifies the column with cell annotations.\n# `use_raw=False` indicates that the normalized data in `adata.X` should be used.\nprint(\"\\nRunning ligand-receptor interaction analysis with CellPhoneDB...\")\nfrom liana.method import cellphonedb\ncellphonedb(\nadata,\ngroupby=\"cell_type\",\nuse_raw=False,\nreturn_all_lrs=True,\nverbose=True\n)\n\n# --- Filter and Extract Results ---\n# The results are stored in the `adata.uns['liana_res']` slot.\nresults = adata.uns[\"liana_res\"]\n\n# Filter for significant ligand-receptor interactions based on the p-value.\nsignificant_results = results[results[\"cellphone_pvals\"] <= 0.01]\nprint(f\"\\nFound {len(significant_results)} significant interactions (p-value <= 0.01).\")\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Generate a dot plot to visualize the top significant ligand-receptor interactions.\n# - `colour`: The color of the dots represents the mean expression of the LR pair.\n# - `size`: The size of the dots represents the significance (p-value).\n# - `source_labels` & `target_labels`: Filter the plot to show interactions between specific cell types.\n# - `top_n`: Display the top 20 interactions based on mean expression.\nprint(\"\\nGenerating dot plot of top interactions...\")\nli.pl.dotplot(\nadata=adata,\ncolour=\"lr_means\",\nsize=\"cellphone_pvals\",\ninverse_size=True, # Make smaller p-values correspond to larger dots\nsource_labels=source_cell_types,\ntarget_labels=target_cell_types,\nfilterby=\"cellphone_pvals\",\nfilter_lambda=lambda x: x <= 0.01,\norderby=\"lr_means\",\norderby_ascending=False,\ntop_n=20,\nfigure_size=(9, 5),\nsize_range=(1, 6)\n)\nplt.show()\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the significant interaction results to a CSV file for further inspection.\nprint(\"\\nSaving significant interaction results...\")\nsignificant_results.to_csv(\"liana_significant_interactions.csv\")\n\n# Save the AnnData object, which now contains the LIANA results in the .uns slot.\nadata.write(\"kang_processed_with_liana.h5ad\")\n\nprint(\"\\nLigand-receptor interaction analysis completed!\")\n",
         "The analysis begins by importing essential Python libraries for data handling, analysis, visualization, and single-cell transcriptomics. These include pandas and numpy for data manipulation, matplotlib and seaborn for plotting, scanpy for single-cell analysis, and liana for ligand-receptor interaction inference. These libraries are critical for performing the subsequent steps in the workflow.\n\nThe single-cell transcriptomics dataset from Kang et al. is loaded using the scanpy library. The dataset is stored in an AnnData object, a specialized data structure for single-cell data. A backup URL is provided to ensure reproducibility. The raw counts are preserved in a separate layer within the AnnData object to retain the original data before normalization.\n\nQuality control is performed to filter out low-quality cells and genes. Cells with fewer than 200 detected genes and genes detected in fewer than 3 cells are removed. This step ensures that only high-quality data is retained for downstream analysis. The data is then normalized by total counts per cell, followed by a log transformation. Normalization and log transformation are standard preprocessing steps in single-cell analysis to account for technical variability and stabilize variance.\n\nThe primary analysis task focuses on inferring ligand-receptor interactions using the LIANA framework with the CellPhoneDB method. Specific cell types are defined as senders (CD4 T cells, B cells, FCGR3A+ Monocytes) and receivers (CD8 T cells, NK cells) for the communication analysis. The CellPhoneDB method is applied to the normalized data, with cell type annotations specified in the `groupby` parameter. The results are stored in the AnnData object, and significant interactions are filtered based on a p-value threshold of 0.01.\n\nA dot plot is generated to visualize the top significant ligand-receptor interactions. The color of the dots represents the mean expression of the ligand-receptor pair, while the size of the dots corresponds to the significance of the interaction (smaller p-values are represented by larger dots). The plot is filtered to show interactions between the specified sender and receiver cell types, and the top 20 interactions based on mean expression are displayed.\n\nThe significant interaction results are saved to a CSV file for further inspection. The processed AnnData object, which now includes the LIANA results, is also saved to a file for future use. This completes the ligand-receptor interaction analysis.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for data handling, analysis, visualization, and single-cell transcriptomics. These include pandas and numpy for data manipulation, matplotlib and seaborn for plotting, scanpy for single-cell analysis, and liana for ligand-receptor interaction inference. These libraries are critical for performing the subsequent steps in the workflow.\n\nThe single-cell transcriptomics dataset from Kang et al. is loaded using the scanpy library. The dataset is stored in an AnnData object, a specialized data structure for single-cell data. A backup URL is provided to ensure reproducibility. The raw counts are preserved in a separate layer within the AnnData object to retain the original data before normalization.\n\nQuality control is performed to filter out low-quality cells and genes. Cells with fewer than 200 detected genes and genes detected in fewer than 3 cells are removed. This step ensures that only high-quality data is retained for downstream analysis. The data is then normalized by total counts per cell, followed by a log transformation. Normalization and log transformation are standard preprocessing steps in single-cell analysis to account for technical variability and stabilize variance.\n\nThe primary analysis task focuses on inferring ligand-receptor interactions using the LIANA framework with the CellPhoneDB method. Specific cell types are defined as senders (CD4 T cells, B cells, FCGR3A+ Monocytes) and receivers (CD8 T cells, NK cells) for the communication analysis. The CellPhoneDB method is applied to the normalized data, with cell type annotations specified in the `groupby` parameter. The results are stored in the AnnData object, and significant interactions are filtered based on a p-value threshold of 0.01.\n\nA dot plot is generated to visualize the top significant ligand-receptor interactions. The color of the dots represents the mean expression of the ligand-receptor pair, while the size of the dots corresponds to the significance of the interaction (smaller p-values are represented by larger dots). The plot is filtered to show interactions between the specified sender and receiver cell types, and the top 20 interactions based on mean expression are displayed.\n\nThe significant interaction results are saved to a CSV file for further inspection. The processed AnnData object, which now includes the LIANA results, is also saved to a file for future use. This completes the ligand-receptor interaction analysis.",
         "The analysis was conducted in Python using key libraries including Scanpy for single-cell transcriptomics and LIANA for ligand-receptor interaction inference. The dataset from Kang et al. was processed within an AnnData object, preserving raw counts for downstream steps. Quality control involved filtering low-quality cells and genes, followed by normalization and log transformation to account for technical variability. Ligand-receptor interactions were inferred using the CellPhoneDB method within LIANA, focusing on defined sender and receiver cell types. Significant interactions were visualized via a dot plot, highlighting mean expression and p-values. Results were stored for further analysis, completing the workflow."
        ],
        [
         "40",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport scanpy as sc\n\nimport liana as li\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the pbmc68k_reduced dataset, a built-in example from Scanpy.\n\nprint(\"Loading pbmc68k_reduced dataset...\")\n\nadata = sc.datasets.pbmc68k_reduced()\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Ligand-Receptor Interaction Analysis\n\n#--------------------------------\n\n\n\n# --- Run LIANA with CellPhoneDB Method ---\n\n# Use the CellPhoneDB method within the LIANA framework to infer interactions.\n\n# `groupby='bulk_labels'` specifies the column with cell annotations.\n\n# `resource_name='consensus'` uses a curated list of LR pairs.\n\n# `key_added='cpdb_res'` stores the results in the .uns slot of the AnnData object.\n\nprint(\"\\nRunning ligand-receptor interaction analysis with CellPhoneDB...\")\n\nli.cellphonedb(\n\nadata,\n\ngroupby='bulk_labels',\n\nresource_name='consensus',\n\nexpr_prop=0.1, # Minimum fraction of cells required to express ligand/receptor\n\nverbose=True,\n\nkey_added='cpdb_res'\n\n)\n\n\n\n# --- Inspect Results ---\n\n# Print the first few rows of the result DataFrame to check the output.\n\nprint(\"\\nCellPhoneDB Results (first 5 rows):\")\n\nprint(adata.uns['cpdb_res'].head())\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Cell Clusters ---\n\n# Visualize the cell clusters to get an overview of the data structure.\n\nsc.pl.umap(adata, color='bulk_labels', title='UMAP of PBMC68k Cell Types', frameon=False)\n\n\n\n# --- Dot Plot of Top Interactions ---\n\n# Generate a dot plot to visualize significant ligand-receptor interactions.\n\n# - Color represents the interaction score (lr_means).\n\n# - Size represents the statistical significance (p-value).\n\nprint(\"\\nGenerating dot plot of top interactions...\")\n\nli.pl.dotplot(\n\nadata=adata,\n\ncolour='lr_means',\n\nsize='cellphone_pvals',\n\ninverse_size=True, # Smaller p-values result in larger dots\n\nsource_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n\ntarget_labels=['CD34+', 'CD56+ NK'],\n\nfigure_size=(8, 7),\n\nfilter_fun=lambda x: x['cellphone_pvals'] <= 0.05, # Filter for significant interactions\n\nuns_key='cpdb_res'\n\n)\n\nplt.show()\n\n\n\n# --- Tile Plot of Expression Statistics ---\n\n# Generate a tile plot to show detailed expression statistics (mean expression and\n\n# proportion of expressing cells) for the top 10 interactions.\n\nprint(\"\\nGenerating tile plot of expression statistics...\")\n\nli.pl.tileplot(\n\nadata=adata,\n\nfill='means',\n\nlabel='props',\n\nlabel_fun=lambda x: f'{x:.2f}',\n\ntop_n=10,\n\norderby='cellphone_pvals',\n\norderby_ascending=True,\n\nsource_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n\ntarget_labels=['CD34+', 'CD56+ NK'],\n\nuns_key='cpdb_res',\n\nsource_title='Ligand',\n\ntarget_title='Receptor',\n\nfigure_size=(8, 7)\n\n)\n\nplt.show()\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the LIANA results in the .uns slot.\n\nprint(\"\\nSaving annotated data object...\")\n\nadata.write(\"pbmc68k_with_liana_results.h5ad\")\n\n\n\nprint(\"\\nLigand-receptor interaction analysis completed!\")\n\n",
         "To perform ligand-receptor interaction analysis on single-cell RNA-seq data, the following steps were executed using Python and the Scanpy and LIANA libraries. The analysis focused on the pbmc68k_reduced dataset, a built-in example from Scanpy, which represents peripheral blood mononuclear cells (PBMCs). The workflow included data loading, ligand-receptor interaction inference, result inspection, visualization, and data saving.\n\nFirst, the necessary Python libraries were imported. Scanpy was used for single-cell data handling and analysis, LIANA for ligand-receptor interaction inference, and Matplotlib for plotting. These libraries are essential for processing and visualizing single-cell data.\n\nThe pbmc68k_reduced dataset was loaded using the Scanpy dataset module. This dataset serves as a representative example of single-cell RNA-seq data and includes preprocessed gene expression profiles and cell annotations.\n\nThe primary analysis task involved inferring ligand-receptor interactions using the CellPhoneDB method within the LIANA framework. The `li.cellphonedb` function was applied to the dataset, with the `groupby` parameter set to 'bulk_labels' to specify the column containing cell annotations. The `resource_name` parameter was set to 'consensus' to use a curated list of ligand-receptor pairs. The `expr_prop` parameter was set to 0.1, ensuring that only ligand-receptor pairs expressed in at least 10% of the cells were considered. The results were stored in the `.uns` slot of the AnnData object under the key 'cpdb_res'. This step is critical for identifying potential cell-cell communication events based on ligand-receptor interactions.\n\nThe results of the CellPhoneDB analysis were inspected by printing the first few rows of the resulting DataFrame. This allowed for a quick verification of the output, ensuring that the analysis produced meaningful interactions.\n\nVisualizations were generated to interpret the results. A UMAP plot was created to visualize the cell clusters, providing an overview of the data structure. The `sc.pl.umap` function was used, with the `color` parameter set to 'bulk_labels' to color the cells by their annotated cell types.\n\nA dot plot was generated to visualize significant ligand-receptor interactions. The `li.pl.dotplot` function was used, with the `colour` parameter set to 'lr_means' to represent interaction scores and the `size` parameter set to 'cellphone_pvals' to represent statistical significance. The `inverse_size` parameter was set to True, ensuring that smaller p-values resulted in larger dots. The plot was filtered to include only interactions with p-values ≤ 0.05, focusing on statistically significant results.\n\nA tile plot was created to show detailed expression statistics for the top 10 interactions. The `li.pl.tileplot` function was used, with the `fill` parameter set to 'means' to represent mean expression and the `label` parameter set to 'props' to represent the proportion of expressing cells. The plot was ordered by p-values in ascending order, highlighting the most significant interactions.\n\nFinally, the annotated data object, which now included the LIANA results in the `.uns` slot, was saved to a file named \"pbmc68k_with_liana_results.h5ad\". This step ensures that the results can be reused for further analysis or shared with collaborators.\n\nThis workflow provides a comprehensive approach to inferring and visualizing ligand-receptor interactions in single-cell RNA-seq data, leveraging the capabilities of Scanpy and LIANA to generate biologically meaningful insights.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform ligand-receptor interaction analysis on single-cell RNA-seq data, the following steps were executed using Python and the Scanpy and LIANA libraries. The analysis focused on the pbmc68k_reduced dataset, a built-in example from Scanpy, which represents peripheral blood mononuclear cells (PBMCs). The workflow included data loading, ligand-receptor interaction inference, result inspection, visualization, and data saving.\n\nFirst, the necessary Python libraries were imported. Scanpy was used for single-cell data handling and analysis, LIANA for ligand-receptor interaction inference, and Matplotlib for plotting. These libraries are essential for processing and visualizing single-cell data.\n\nThe pbmc68k_reduced dataset was loaded using the Scanpy dataset module. This dataset serves as a representative example of single-cell RNA-seq data and includes preprocessed gene expression profiles and cell annotations.\n\nThe primary analysis task involved inferring ligand-receptor interactions using the CellPhoneDB method within the LIANA framework. The `li.cellphonedb` function was applied to the dataset, with the `groupby` parameter set to 'bulk_labels' to specify the column containing cell annotations. The `resource_name` parameter was set to 'consensus' to use a curated list of ligand-receptor pairs. The `expr_prop` parameter was set to 0.1, ensuring that only ligand-receptor pairs expressed in at least 10% of the cells were considered. The results were stored in the `.uns` slot of the AnnData object under the key 'cpdb_res'. This step is critical for identifying potential cell-cell communication events based on ligand-receptor interactions.\n\nThe results of the CellPhoneDB analysis were inspected by printing the first few rows of the resulting DataFrame. This allowed for a quick verification of the output, ensuring that the analysis produced meaningful interactions.\n\nVisualizations were generated to interpret the results. A UMAP plot was created to visualize the cell clusters, providing an overview of the data structure. The `sc.pl.umap` function was used, with the `color` parameter set to 'bulk_labels' to color the cells by their annotated cell types.\n\nA dot plot was generated to visualize significant ligand-receptor interactions. The `li.pl.dotplot` function was used, with the `colour` parameter set to 'lr_means' to represent interaction scores and the `size` parameter set to 'cellphone_pvals' to represent statistical significance. The `inverse_size` parameter was set to True, ensuring that smaller p-values resulted in larger dots. The plot was filtered to include only interactions with p-values ≤ 0.05, focusing on statistically significant results.\n\nA tile plot was created to show detailed expression statistics for the top 10 interactions. The `li.pl.tileplot` function was used, with the `fill` parameter set to 'means' to represent mean expression and the `label` parameter set to 'props' to represent the proportion of expressing cells. The plot was ordered by p-values in ascending order, highlighting the most significant interactions.\n\nFinally, the annotated data object, which now included the LIANA results in the `.uns` slot, was saved to a file named \"pbmc68k_with_liana_results.h5ad\". This step ensures that the results can be reused for further analysis or shared with collaborators.\n\nThis workflow provides a comprehensive approach to inferring and visualizing ligand-receptor interactions in single-cell RNA-seq data, leveraging the capabilities of Scanpy and LIANA to generate biologically meaningful insights.",
         "The analysis of ligand-receptor interactions in single-cell RNA-seq data was conducted using Python, leveraging the Scanpy and LIANA libraries. The workflow focused on the pbmc68k_reduced dataset, representing peripheral blood mononuclear cells (PBMCs). Key steps included data loading, ligand-receptor interaction inference using the CellPhoneDB method within LIANA, and visualization of results. Interactions were inferred based on a curated list of ligand-receptor pairs, with filtering for expression in at least 10% of cells. Results were visualized through UMAP plots, dot plots, and tile plots to highlight significant interactions. The annotated data object, enriched with interaction results, was saved for further analysis. This approach provides a robust framework for exploring cell-cell communication in single-cell datasets."
        ],
        [
         "41",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport pandas as pd\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport seaborn as sns\n\nimport scanpy as sc\n\nimport liana as li\n\nimport decoupler as dc\n\nimport cell2cell as c2c\n\nimport warnings\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.filterwarnings('ignore')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a sample dataset. Here we use a built-in dataset from LIANA's testing suite.\n\n# NOTE: Replace this with your own data as needed.\n\nprint(\"Loading Kang et al. 2018 dataset...\")\n\nadata = li.testing.datasets.kang_2018()\n\n\n\n#--------------------------------\n\n# Preprocessing & Quality Control\n\n#--------------------------------\n\n\n\n# Apply basic quality control filters.\n\nprint(\"Performing quality control and preprocessing...\")\n\nsc.pp.filter_cells(adata, min_genes=200)\n\nsc.pp.filter_genes(adata, min_cells=3)\n\n\n\n# Normalize and log-transform the data.\n\nsc.pp.normalize_total(adata)\n\nsc.pp.log1p(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# --- 1. Ligand-Receptor Inference by Sample ---\n\n# Define keys for sample, condition, and cell type columns in adata.obs.\n\nsample_key = 'sample'\n\ncondition_key = 'condition'\n\ngroupby = 'cell_type'\n\n\n\n# Run LIANA's rank aggregation method for each sample individually.\n\n# This infers ligand-receptor interactions within each sample before downstream analysis.\n\nprint(\"\\nRunning ligand-receptor inference for each sample...\")\n\nliana_results = li.mt.rank_aggregate.by_sample(\n\nadata,\n\ngroupby=groupby,\n\nresource_name='consensus', # Use a consensus database of LR pairs\n\nsample_key=sample_key,\n\nuse_raw=False,\n\nverbose=True,\n\nn_perms=None, # Set to an integer (e.g., 1000) for permutation testing\n\nreturn_all_lrs=True\n\n)\n\n\n\n# Store the LIANA results in the AnnData object.\n\nadata.uns[\"liana_res\"] = liana_results\n\n\n\n# Display the top-ranked interactions.\n\nprint(\"\\nTop aggregated ligand-receptor interactions:\")\n\nprint(adata.uns[\"liana_res\"].sort_values(\"magnitude_rank\").head())\n\n\n\n# --- 2. Tensor Decomposition for Communication Analysis ---\n\n# Construct a communication tensor. This is a multi-dimensional array representing\n\n# interactions (Context x Sender x Receiver x LR-pair).\n\nprint(\"\\nConstructing communication tensor...\")\n\ntensor = li.multi.to_tensor_c2c(\n\nadata,\n\nsample_key=sample_key,\n\nscore_key=\"magnitude_rank\",\n\nhow=\"outer_cells\" # Use all cells for tensor construction\n\n)\n\n\n\n# Perform tensor decomposition to identify latent communication patterns (factors).\n\n# `rank=6` specifies the number of factors to extract.\n\nprint(\"\\nPerforming tensor decomposition...\")\n\ntensor = c2c.analysis.run_tensor_cell2cell_pipeline(\n\ntensor,\n\ncopy_tensor=True,\n\nrank=6,\n\ntf_optimization='regular',\n\nrandom_state=0,\n\ndevice=\"cpu\" # Use 'cuda' if a GPU is available\n\n)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP Visualization ---\n\n# Plot a UMAP to visualize the overall structure of the data by condition and cell type.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=[condition_key, groupby], frameon=False)\n\n\n\n# --- Tensor Factor Plots ---\n\n# Visualize the results of the tensor decomposition. Each factor represents a\n\n# distinct communication pattern across contexts, senders, receivers, and LR pairs.\n\nfactors, axes = c2c.plotting.tensor_factors_plot(\n\ninteraction_tensor=tensor,\n\nfontsize=10\n\n)\n\nplt.show()\n\n\n\n# --- Communication Network Plots ---\n\n# Visualize the cell-cell communication networks for specific factors.\n\n# Here, we plot the network associated with 'Factor 6' as an example.\n\nc2c.plotting.ccc_networks_plot(factors, included_factors=['Factor 6'])\n\nplt.show()\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final plots and data objects.\n\nprint(\"\\nSaving results...\")\n\n# Save the tensor factor plot\n\nfactors[0].figure.savefig(\"tensor_factors_plot.png\", dpi=300, bbox_inches='tight')\n\n# Save the AnnData object with LIANA results\n\nadata.write(\"kang_2018_with_liana.h5ad\")\n\n# Save the tensor object with decomposition results\n\nc2c.io.export_variable_to_file(tensor, 'communication_tensor.c2c')\n\n\n\nprint(\"\\nTensor-based communication analysis completed!\")\n\n",
         "The analysis begins by importing essential Python libraries for data handling, analysis, and visualization, including `pandas`, `numpy`, `matplotlib`, `seaborn`, `scanpy`, `liana`, `decoupler`, and `cell2cell`. Warnings are suppressed to ensure cleaner output during execution. The dataset used is a built-in single-cell RNA-seq dataset from Kang et al. (2018), loaded using LIANA’s testing suite. This dataset is stored in an AnnData object, a standard format for single-cell data.\n\nQuality control and preprocessing are performed to ensure data reliability. Cells with fewer than 200 detected genes and genes detected in fewer than 3 cells are filtered out. The data is then normalized to account for differences in sequencing depth and log-transformed to stabilize variance, a standard preprocessing step for single-cell RNA-seq data.\n\nThe analysis proceeds with two major tasks: ligand-receptor (LR) inference and tensor decomposition for communication analysis. For LR inference, LIANA’s rank aggregation method is applied to infer LR interactions within each sample individually. A consensus database of LR pairs is used, and the results are stored in the AnnData object. The top-ranked interactions are displayed based on their magnitude rank. \n\nNext, a communication tensor is constructed, representing interactions across contexts, sender cells, receiver cells, and LR pairs. Tensor decomposition is performed to identify latent communication patterns, with six factors extracted to capture distinct communication dynamics. The decomposition is optimized using a regular approach and executed on the CPU, though GPU acceleration is an option if available.\n\nVisualization is a key component of the analysis. A UMAP plot is generated to visualize the overall structure of the data, colored by condition and cell type. The results of the tensor decomposition are visualized using factor plots, which highlight distinct communication patterns. Additionally, cell-cell communication networks are plotted for specific factors, with an example provided for ‘Factor 6’.\n\nFinally, the results are saved for further use. The tensor factor plot is saved as a high-resolution image, the AnnData object with LR inference results is saved in H5AD format, and the tensor object with decomposition results is exported to a file. This completes the tensor-based communication analysis, providing insights into cell-cell communication dynamics in the dataset.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for data handling, analysis, and visualization, including `pandas`, `numpy`, `matplotlib`, `seaborn`, `scanpy`, `liana`, `decoupler`, and `cell2cell`. Warnings are suppressed to ensure cleaner output during execution. The dataset used is a built-in single-cell RNA-seq dataset from Kang et al. (2018), loaded using LIANA’s testing suite. This dataset is stored in an AnnData object, a standard format for single-cell data.\n\nQuality control and preprocessing are performed to ensure data reliability. Cells with fewer than 200 detected genes and genes detected in fewer than 3 cells are filtered out. The data is then normalized to account for differences in sequencing depth and log-transformed to stabilize variance, a standard preprocessing step for single-cell RNA-seq data.\n\nThe analysis proceeds with two major tasks: ligand-receptor (LR) inference and tensor decomposition for communication analysis. For LR inference, LIANA’s rank aggregation method is applied to infer LR interactions within each sample individually. A consensus database of LR pairs is used, and the results are stored in the AnnData object. The top-ranked interactions are displayed based on their magnitude rank. \n\nNext, a communication tensor is constructed, representing interactions across contexts, sender cells, receiver cells, and LR pairs. Tensor decomposition is performed to identify latent communication patterns, with six factors extracted to capture distinct communication dynamics. The decomposition is optimized using a regular approach and executed on the CPU, though GPU acceleration is an option if available.\n\nVisualization is a key component of the analysis. A UMAP plot is generated to visualize the overall structure of the data, colored by condition and cell type. The results of the tensor decomposition are visualized using factor plots, which highlight distinct communication patterns. Additionally, cell-cell communication networks are plotted for specific factors, with an example provided for ‘Factor 6’.\n\nFinally, the results are saved for further use. The tensor factor plot is saved as a high-resolution image, the AnnData object with LR inference results is saved in H5AD format, and the tensor object with decomposition results is exported to a file. This completes the tensor-based communication analysis, providing insights into cell-cell communication dynamics in the dataset.",
         "The analysis was conducted in Python using libraries such as Scanpy, LIANA, and Cell2Cell, focusing on ligand-receptor (LR) inference and tensor decomposition to explore cell-cell communication dynamics in a single-cell RNA-seq dataset from Kang et al. (2018). Quality control involved filtering cells and genes based on detection thresholds, followed by normalization and log transformation to stabilize variance. LR interactions were inferred using LIANA’s rank aggregation method, leveraging a consensus database of LR pairs. A communication tensor was constructed across contexts, sender-receiver cells, and LR pairs, with tensor decomposition identifying six latent communication patterns. Visualization included UMAP plots for data structure and factor plots for communication dynamics, with results saved for downstream analysis."
        ],
        [
         "42",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport pandas as pd\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport seaborn as sns\n\nimport scanpy as sc\n\nimport liana as li\n\nimport decoupler as dc\n\nimport omnipath as op\n\nfrom pydeseq2.dds import DeseqDataSet\n\nfrom pydeseq2.ds import DeseqStats\n\nimport warnings\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.filterwarnings('ignore')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a sample dataset. Here we use a built-in dataset from LIANA's testing suite.\n\n# NOTE: Replace this with your own data as needed.\n\nprint(\"Loading Kang et al. 2018 dataset...\")\n\nadata = li.testing.datasets.kang_2018()\n\n\n\n#--------------------------------\n\n# Preprocessing & Quality Control\n\n#--------------------------------\n\n\n\n# Apply basic quality control filters.\n\nprint(\"Performing quality control and preprocessing...\")\n\nsc.pp.filter_cells(adata, min_genes=200)\n\nsc.pp.filter_genes(adata, min_cells=3)\n\n\n\n# Normalize and log-transform data for visualization and some downstream steps.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# Define metadata columns of interest for the analysis\n\nsample_key = 'sample'\n\ncondition_key = 'condition'\n\ngroupby = 'cell_abbr' # Using abbreviated cell types for clarity\n\n\n\n# --- 1. Pseudobulk and Differential Expression Analysis (DEA) ---\n\n# Generate pseudobulk profiles by summing counts for each cell type within each sample.\n\nprint(\"\\nGenerating pseudobulk profiles...\")\n\npdata = dc.get_pseudobulk(\n\nadata,\n\nsample_col=sample_key,\n\ngroups_col=groupby,\n\nlayer='counts', # Use raw counts for DESeq2\n\nmode='sum',\n\nmin_cells=10,\n\nmin_counts=10000\n\n)\n\n\n\n# Perform DEA using PyDESeq2 for each cell type.\n\nprint(\"\\nPerforming differential expression analysis with PyDESeq2...\")\n\ndea_results = {}\n\nfor cell_group in pdata.obs[groupby].unique():\n\n# Subset pseudobulk data to the current cell type\n\nctdata = pdata[pdata.obs[groupby] == cell_group].copy()\n\n\n\n# Filter out lowly expressed genes for the current cell type\n\ngenes = dc.filter_by_expr(ctdata, group=condition_key, min_count=5, min_total_count=10)\n\nctdata = ctdata[:, genes].copy()\n\n\n\n# Build and run the DESeq2 model\n\ndds = DeseqDataSet(\n\nadata=ctdata,\n\ndesign_factors=condition_key,\n\nref_level=[condition_key, 'ctrl'], # Set 'ctrl' as the reference condition\n\nrefit_cooks=True,\n\nquiet=True\n\n)\n\ndds.deseq2()\n\n\n\n# Get statistics and shrink log-fold changes for robustness\n\nstat_res = DeseqStats(dds, contrast=[condition_key, 'stim', 'ctrl'])\n\nstat_res.summary()\n\nstat_res.lfc_shrink(coeff='condition_stim_vs_ctrl')\n\n\n\ndea_results[cell_group] = stat_res.results_df\n\n\n\n# Concatenate all DEA results into a single DataFrame.\n\ndea_df = pd.concat(dea_results).reset_index().rename(\n\ncolumns={'level_0': groupby, 'level_1': 'gene'}\n\n).set_index('gene')\n\n\n\n\n\n# --- 2. Deregulated Ligand-Receptor Interaction Analysis ---\n\n# Use LIANA to identify LR interactions that are significantly altered between conditions,\n\n# based on the differential expression statistics calculated above.\n\nprint(\"\\nIdentifying deregulated ligand-receptor interactions...\")\n\nlr_res = li.multi.df_to_lr(\n\nadata,\n\ndea_df=dea_df,\n\nresource_name='consensus',\n\nexpr_prop=0.1,\n\ngroupby=groupby,\n\nstat_keys=['stat', 'pvalue', 'padj'],\n\nuse_raw=False, # Use log-normalized data for expression proportion check\n\ncomplex_col='stat',\n\nreturn_all_lrs=False\n\n)\n\n\n\n# Sort the results by the interaction statistic.\n\nlr_res = lr_res.sort_values(\"interaction_stat\", ascending=False)\n\n\n\n\n\n# --- 3. Pathway Enrichment Analysis ---\n\n# Use the DEA results to infer transcription factor (TF) activity changes.\n\nprint(\"\\nPerforming pathway enrichment analysis (TF activity)...\")\n\n# Get a TF-target regulatory network from Omnipath's CollecTRI resource.\n\nnet = dc.get_collectri()\n\n\n\n# Pivot the DEA results to a wide format (cells x genes) with stat values.\n\ndea_wide = dea_df[[groupby, 'stat']].reset_index(names='genes').pivot(\n\nindex=groupby, columns='genes', values='stat'\n\n).fillna(0)\n\n\n\n# Run Univariate Linear Model (ULM) from Decoupler to estimate TF activities.\n\nestimates, pvals = dc.run_ulm(mat=dea_wide, net=net)\n\n\n\n# Print the top TFs associated with the 'CD14' Monocyte group as an example.\n\ntop_tfs = estimates.T.sort_values('CD14', key=abs, ascending=False).head()\n\nprint(\"\\nTop TFs associated with CD14 Monocytes:\")\n\nprint(top_tfs)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP Visualization ---\n\n# Plot a UMAP to visualize the overall structure of the data.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=[condition_key, sample_key, 'cell_type', groupby], frameon=False, ncols=2)\n\n\n\n# --- Tile Plot of Top Interactions ---\n\n# Visualize the top 15 deregulated ligand-receptor interactions.\n\n# Fill color represents expression, while '*' indicates statistical significance.\n\nli.pl.tileplot(\n\nliana_res=lr_res,\n\nfill='expr',\n\nlabel='padj',\n\nlabel_fun=lambda x: '*' if x < 0.05 else np.nan,\n\ntop_n=15,\n\norderby='interaction_stat',\n\norderby_ascending=False,\n\nsource_title='Ligand',\n\ntarget_title='Receptor'\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the key results DataFrames to CSV files.\n\nprint(\"\\nSaving analysis results...\")\n\ndea_df.to_csv(\"differential_expression_results.csv\")\n\nlr_res.to_csv(\"deregulated_lr_interactions.csv\")\n\nestimates.to_csv(\"tf_activity_estimates.csv\")\n\n\n\n# Save the processed AnnData object.\n\nadata.write(\"kang_processed_for_dea.h5ad\")\n\n\n\nprint(\"\\nDifferential communication and pathway analysis completed!\")\n\n",
         "The following section describes the computational workflow used to analyze single-cell RNA sequencing data, focusing on differential expression analysis, ligand-receptor interaction identification, and pathway enrichment. The analysis was performed using Python, leveraging several specialized libraries for single-cell data processing and statistical analysis.\n\nFirst, the necessary Python libraries were imported. These included `pandas` and `numpy` for data handling, `matplotlib` and `seaborn` for visualization, and `scanpy`, `liana`, `decoupler`, and `omnipath` for single-cell data analysis. Additionally, `pydeseq2` was used for differential expression analysis, and warnings were suppressed to ensure cleaner output during execution.\n\nThe analysis began with the loading of a sample dataset from the LIANA testing suite, specifically the Kang et al. 2018 dataset. This dataset was chosen for demonstration purposes, but users are encouraged to replace it with their own data as needed. Quality control and preprocessing steps were then applied to the dataset. Cells with fewer than 200 genes and genes detected in fewer than 3 cells were filtered out to ensure data quality. The data were normalized to a total count of 10,000 per cell and log-transformed for downstream analysis.\n\nNext, the major analysis tasks were performed. Pseudobulk profiles were generated by summing counts for each cell type within each sample, using the `decoupler` library. This step was essential for performing differential expression analysis (DEA) at the cell type level. DEA was conducted using `pydeseq2`, with the control condition set as the reference. Lowly expressed genes were filtered out for each cell type, and the DESeq2 model was applied to calculate differential expression statistics. The results were concatenated into a single DataFrame for further analysis.\n\nFollowing DEA, deregulated ligand-receptor (LR) interactions were identified using the `liana` library. The analysis focused on interactions that were significantly altered between conditions, based on the differential expression statistics. The results were sorted by interaction statistic to prioritize the most significant interactions.\n\nPathway enrichment analysis was then performed to infer transcription factor (TF) activity changes. A TF-target regulatory network was obtained from the Omnipath CollecTRI resource, and the Univariate Linear Model (ULM) from `decoupler` was used to estimate TF activities. The top TFs associated with specific cell types, such as CD14 Monocytes, were identified and reported.\n\nVisualization of the results included a UMAP plot to display the overall structure of the data, colored by condition, sample, cell type, and abbreviated cell type. Additionally, a tile plot was generated to visualize the top 15 deregulated LR interactions, with fill color representing expression and asterisks indicating statistical significance.\n\nFinally, the key results, including differential expression statistics, deregulated LR interactions, and TF activity estimates, were saved to CSV files. The processed AnnData object was also saved for future reference. This comprehensive workflow provides a detailed and reproducible approach to analyzing single-cell RNA sequencing data, from preprocessing to pathway enrichment.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following section describes the computational workflow used to analyze single-cell RNA sequencing data, focusing on differential expression analysis, ligand-receptor interaction identification, and pathway enrichment. The analysis was performed using Python, leveraging several specialized libraries for single-cell data processing and statistical analysis.\n\nFirst, the necessary Python libraries were imported. These included `pandas` and `numpy` for data handling, `matplotlib` and `seaborn` for visualization, and `scanpy`, `liana`, `decoupler`, and `omnipath` for single-cell data analysis. Additionally, `pydeseq2` was used for differential expression analysis, and warnings were suppressed to ensure cleaner output during execution.\n\nThe analysis began with the loading of a sample dataset from the LIANA testing suite, specifically the Kang et al. 2018 dataset. This dataset was chosen for demonstration purposes, but users are encouraged to replace it with their own data as needed. Quality control and preprocessing steps were then applied to the dataset. Cells with fewer than 200 genes and genes detected in fewer than 3 cells were filtered out to ensure data quality. The data were normalized to a total count of 10,000 per cell and log-transformed for downstream analysis.\n\nNext, the major analysis tasks were performed. Pseudobulk profiles were generated by summing counts for each cell type within each sample, using the `decoupler` library. This step was essential for performing differential expression analysis (DEA) at the cell type level. DEA was conducted using `pydeseq2`, with the control condition set as the reference. Lowly expressed genes were filtered out for each cell type, and the DESeq2 model was applied to calculate differential expression statistics. The results were concatenated into a single DataFrame for further analysis.\n\nFollowing DEA, deregulated ligand-receptor (LR) interactions were identified using the `liana` library. The analysis focused on interactions that were significantly altered between conditions, based on the differential expression statistics. The results were sorted by interaction statistic to prioritize the most significant interactions.\n\nPathway enrichment analysis was then performed to infer transcription factor (TF) activity changes. A TF-target regulatory network was obtained from the Omnipath CollecTRI resource, and the Univariate Linear Model (ULM) from `decoupler` was used to estimate TF activities. The top TFs associated with specific cell types, such as CD14 Monocytes, were identified and reported.\n\nVisualization of the results included a UMAP plot to display the overall structure of the data, colored by condition, sample, cell type, and abbreviated cell type. Additionally, a tile plot was generated to visualize the top 15 deregulated LR interactions, with fill color representing expression and asterisks indicating statistical significance.\n\nFinally, the key results, including differential expression statistics, deregulated LR interactions, and TF activity estimates, were saved to CSV files. The processed AnnData object was also saved for future reference. This comprehensive workflow provides a detailed and reproducible approach to analyzing single-cell RNA sequencing data, from preprocessing to pathway enrichment.",
         "The analysis of single-cell RNA sequencing data was conducted in Python using specialized libraries such as `scanpy`, `liana`, `decoupler`, and `omnipath`. Quality control involved filtering cells with fewer than 200 genes and genes detected in fewer than 3 cells, followed by normalization and log transformation. Pseudobulk profiles were generated for cell types using `decoupler`, enabling differential expression analysis (DEA) with `pydeseq2`, which identified condition-specific gene expression changes. Deregulated ligand-receptor interactions were identified using `liana`, prioritizing significant alterations. Pathway enrichment analysis inferred transcription factor activity changes using the Univariate Linear Model (ULM) from `decoupler`, leveraging a regulatory network from Omnipath CollecTRI. Results were visualized through UMAP plots and tile plots, highlighting data structure and key interactions."
        ],
        [
         "43",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, pathway analysis, and plotting\n\nimport scanpy as sc\n\nimport decoupler as dc\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.set_figure_params(figsize=(5, 5), frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example dataset (PBMC 3k) from Decoupler's built-in datasets.\n\n# This dataset is pre-processed with clustering information.\n\nprint(\"Loading pbmc3k dataset...\")\n\nadata = dc.ds.pbmc3k()\n\n\n\n# Display a summary of the loaded dataset.\n\nprint(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Pathway Enrichment\n\n#--------------------------------\n\n\n\n# --- Load Pathway Gene Sets ---\n\n# Get the PROGENy model, which contains a curated set of pathway gene sets\n\n# with signed weights representing up/down-regulation.\n\nprint(\"\\nLoading PROGENy pathway model...\")\n\nprogeny = dc.op.progeny(organism='human')\n\n\n\n# --- Compute Enrichment Scores ---\n\n# Run the Univariate Linear Model (ulm) from Decoupler to compute pathway\n\n# enrichment scores for each cell.\n\n# The result is stored in `adata.obsm['ulm_estimate']`.\n\nprint(\"Computing pathway enrichment scores...\")\n\ndc.run_ulm(data=adata, net=progeny, source='source', target='target', weight='weight', min_n=5)\n\n\n\n# Extract the computed enrichment scores into a separate AnnData object for easier plotting.\n\n# This creates a new object where the .X matrix contains the pathway scores.\n\nscores = dc.get_obsm_to_anndata(adata, obsm_key='ulm_estimate')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Cell Clusters ---\n\n# Visualize the original cell clusters from the dataset.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color='leiden', title='UMAP of PBMC Clusters')\n\n\n\n# --- UMAP of Pathway Activity ---\n\n# Visualize the activity of a specific pathway (e.g., TNFa signaling) on the UMAP.\n\n# Red indicates higher pathway activity, and blue indicates lower activity.\n\nsc.pl.umap(scores, color=['TNFa'], cmap='RdBu_r', vcenter=0, title='TNFa Pathway Activity')\n\n\n\n# --- Violin Plot of Pathway Activity ---\n\n# Compare the distribution of pathway activity scores across different cell types.\n\nsc.pl.violin(scores, keys=['TNFa'], groupby='celltype', rotation=90)\n\n\n\n# --- Matrix Plot of All Pathways ---\n\n# Show a heatmap of all pathway activities across all cell types.\n\n# The scores are z-scaled for better comparability across pathways.\n\nsc.pl.matrixplot(\n\nadata=scores,\n\nvar_names=scores.var_names,\n\ngroupby='celltype',\n\ndendrogram=True,\n\nstandard_scale='var',\n\ncolorbar_title='Z-scaled scores',\n\ncmap='RdBu_r'\n\n)\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the pathway enrichment scores.\n\nprint(\"\\nSaving data with enrichment scores...\")\n\n# Note: The scores are in adata.obsm['ulm_estimate'], and adata.obsm['ulm_pvals']\n\nadata.write(\"pbmc3k_with_progeny_scores.h5ad\")\n\n\n\nprint(\"\\nPathway enrichment analysis completed!\")\n\n",
         "The following method describes a single-cell RNA-seq pathway enrichment analysis using the Python libraries Scanpy and Decoupler. The analysis begins by importing necessary modules, including Scanpy for single-cell analysis, Decoupler for pathway analysis, and Matplotlib for visualization. Global plotting parameters are configured to ensure consistent and clear visualizations.\n\nThe analysis uses the PBMC 3k dataset, a pre-processed single-cell RNA-seq dataset with clustering information, which is loaded from Decoupler’s built-in datasets. The dataset is summarized to provide an overview of its structure and content.\n\nFor the pathway enrichment analysis, the PROGENy model is loaded, which contains curated gene sets representing various pathways with signed weights indicating up- or down-regulation. The Univariate Linear Model (ulm) method from Decoupler is applied to compute pathway enrichment scores for each cell. The results are stored in the AnnData object under `adata.obsm['ulm_estimate']`. To facilitate visualization, the enrichment scores are extracted into a separate AnnData object.\n\nVisualizations are generated to interpret the results. First, a UMAP plot is created to display the original cell clusters from the dataset. Next, the activity of a specific pathway (e.g., TNFa signaling) is visualized on the UMAP, with red indicating higher activity and blue indicating lower activity. A violin plot is used to compare the distribution of pathway activity scores across different cell types. Finally, a matrix plot is generated to display a heatmap of all pathway activities across cell types, with z-scaled scores for better comparability.\n\nThe AnnData object, now containing the pathway enrichment scores, is saved to a file for further use. This completes the pathway enrichment analysis, providing insights into pathway activity across cell types in the PBMC 3k dataset.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe following method describes a single-cell RNA-seq pathway enrichment analysis using the Python libraries Scanpy and Decoupler. The analysis begins by importing necessary modules, including Scanpy for single-cell analysis, Decoupler for pathway analysis, and Matplotlib for visualization. Global plotting parameters are configured to ensure consistent and clear visualizations.\n\nThe analysis uses the PBMC 3k dataset, a pre-processed single-cell RNA-seq dataset with clustering information, which is loaded from Decoupler’s built-in datasets. The dataset is summarized to provide an overview of its structure and content.\n\nFor the pathway enrichment analysis, the PROGENy model is loaded, which contains curated gene sets representing various pathways with signed weights indicating up- or down-regulation. The Univariate Linear Model (ulm) method from Decoupler is applied to compute pathway enrichment scores for each cell. The results are stored in the AnnData object under `adata.obsm['ulm_estimate']`. To facilitate visualization, the enrichment scores are extracted into a separate AnnData object.\n\nVisualizations are generated to interpret the results. First, a UMAP plot is created to display the original cell clusters from the dataset. Next, the activity of a specific pathway (e.g., TNFa signaling) is visualized on the UMAP, with red indicating higher activity and blue indicating lower activity. A violin plot is used to compare the distribution of pathway activity scores across different cell types. Finally, a matrix plot is generated to display a heatmap of all pathway activities across cell types, with z-scaled scores for better comparability.\n\nThe AnnData object, now containing the pathway enrichment scores, is saved to a file for further use. This completes the pathway enrichment analysis, providing insights into pathway activity across cell types in the PBMC 3k dataset.",
         "Pathway enrichment analysis was conducted on the PBMC 3k single-cell RNA-seq dataset using Python libraries Scanpy and Decoupler. The PROGENy model, containing curated gene sets with signed weights, was employed to compute pathway enrichment scores via the Univariate Linear Model (ulm) method. Scores were stored in the AnnData object and visualized through UMAP plots, highlighting pathway activity (e.g., TNFa signaling) across cell clusters. Violin plots and a z-scaled heatmap further compared pathway activity distributions and patterns across cell types. The results, integrated into the AnnData object, provide insights into pathway-specific cellular signaling dynamics."
        ],
        [
         "44",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport pandas as pd\n\nimport scanpy as sc\n\nimport decoupler as dc\n\nimport matplotlib.pyplot as plt\n\nfrom pydeseq2.dds import DeseqDataSet, DefaultInference\n\nfrom pydeseq2.ds import DeseqStats\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.set_figure_params(figsize=(4, 4), frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example single-cell RNA sequencing dataset.\n\n# Here, we use a COVID-19 dataset from Decoupler's built-in datasets.\n\nprint(\"Loading covid5k dataset...\")\n\nadata = dc.ds.covid5k()\n\n\n\n# Display a summary of the loaded dataset.\n\nprint(adata)\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing and Pseudobulking\n\n#--------------------------------\n\n\n\n# --- Pseudobulking ---\n\n# Aggregate single-cell expression data into pseudobulk profiles.\n\n# This sums the counts for each cell type within each individual sample.\n\nprint(\"\\nGenerating pseudobulk profiles...\")\n\npdata = dc.pp.pseudobulk(\n\nadata=adata,\n\nsample_col='individual',\n\ngroups_col='celltype',\n\nmode='sum' # Sum counts for each gene\n\n)\n\n\n\n# --- Pseudobulk QC and Preprocessing ---\n\n# Filter out low-quality pseudobulk samples.\n\ndc.pp.filter_samples(pdata, min_cells=10, min_counts=1000)\n\n\n\n# Normalize and preprocess the pseudobulk data for visualization and PCA.\n\n# Note: Raw counts are kept for DESeq2 analysis.\n\npdata.layers['counts'] = pdata.X.copy()\n\nsc.pp.normalize_total(pdata, target_sum=1e4)\n\nsc.pp.log1p(pdata)\n\nsc.pp.scale(pdata, max_value=10)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# --- 1. Differential Expression Analysis (DEA) with PyDESeq2 ---\n\n# Define and run the DESeq2 model on the raw pseudobulk counts.\n\n# The design formula `~disease + sex` models the effect of disease while accounting for sex.\n\nprint(\"\\nPerforming differential expression analysis with PyDESeq2...\")\n\ninference = DefaultInference(n_cpus=8)\n\ndds = DeseqDataSet(\n\nadata=pdata,\n\ndesign_factors=['disease', 'sex'],\n\nrefit_cooks=True,\n\ninference=inference\n\n)\n\ndds.deseq2()\n\n\n\n# Extract statistical results for the contrast between COVID-19 and normal conditions.\n\nstat_res = DeseqStats(\n\ndds,\n\ncontrast=['disease', 'COVID-19', 'normal'],\n\ninference=inference\n\n)\n\nstat_res.summary()\n\nresults_df = stat_res.results_df\n\n\n\n\n\n# --- 2. Pathway Enrichment Analysis ---\n\n# Use the DEA results to infer pathway activity changes with Decoupler.\n\nprint(\"\\nPerforming pathway enrichment analysis...\")\n\n# Get the PROGENy model, a curated set of pathway gene sets.\n\nprogeny = dc.op.progeny(organism='human')\n\n\n\n# Run Univariate Linear Model (ulm) to estimate pathway activities from the DE statistics.\n\npw_acts, pw_padj = dc.run_ulm(\n\nmat=results_df[['stat']].T.rename(index={'stat': 'disease.vs.normal'}),\n\nnet=progeny\n\n)\n\n\n\n# Filter for significantly altered pathways.\n\nsignificant_mask = (pw_padj.T < 0.05).iloc[:, 0]\n\nsignificant_pw_acts = pw_acts.loc[:, significant_mask]\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Single-Cell Data ---\n\n# Visualize the initial cell type and disease state distribution.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=['celltype', 'disease'], ncols=1, title=[\"Cell Types\", \"Disease State\"])\n\n\n\n# --- Pseudobulk QC Plot ---\n\n# Visualize the cell type composition of the filtered pseudobulk samples.\n\ndc.pl.obsbar(adata=pdata, y='celltype', hue='disease', figsize=(6, 3))\n\n\n\n# --- PCA of Pseudobulk Data ---\n\n# Explore the major sources of variation across the pseudobulk samples.\n\nsc.tl.pca(pdata)\n\nsc.pl.pca(pdata, color='disease', size=100, title=\"PCA of Pseudobulk Samples\")\n\n\n\n# --- Volcano Plot of DEA Results ---\n\n# Visualize the differential expression results, highlighting significant genes.\n\ndc.pl.volcano(results_df, x='log2FoldChange', y='pvalue')\n\n\n\n# --- Bar Plot of Pathway Enrichment ---\n\n# Visualize the pathway enrichment results for significantly altered pathways.\n\ndc.pl.barplot(data=significant_pw_acts, name='disease.vs.normal', figsize=(3, 3))\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the key results DataFrames and the processed AnnData object.\n\nprint(\"\\nSaving analysis results...\")\n\nresults_df.to_csv(\"pseudobulk_dea_results.csv\")\n\nsignificant_pw_acts.to_csv(\"significant_pathway_activities.csv\")\n\npdata.write(\"pseudobulk_processed_data.h5ad\")\n\n\n\nprint(\"\\nPseudobulk differential analysis completed!\")\n\n",
         "The analysis begins by importing essential Python libraries for data handling, analysis, and visualization, including NumPy, pandas, Scanpy, Decoupler, Matplotlib, and PyDESeq2. These libraries are critical for managing single-cell RNA sequencing (scRNA-seq) data, performing differential expression analysis, and visualizing results. Global plotting parameters are configured to ensure consistent and clear visualizations.\n\nThe dataset used in this analysis is the COVID-19 single-cell RNA sequencing dataset from Decoupler’s built-in datasets. This dataset is loaded into an AnnData object, a specialized data structure for handling single-cell data. A summary of the dataset is displayed to provide an overview of its structure and content.\n\nThe preprocessing phase involves pseudobulking, a method to aggregate single-cell expression data into pseudobulk profiles by summing counts for each cell type within individual samples. This step reduces the complexity of single-cell data while preserving biological signals. Low-quality pseudobulk samples are filtered out based on thresholds for minimum cell count and total counts. The pseudobulk data is then normalized and preprocessed for downstream analysis, including log transformation and scaling. Raw counts are retained for differential expression analysis.\n\nDifferential expression analysis (DEA) is performed using PyDESeq2, a Python implementation of the DESeq2 method. The analysis models the effect of disease while accounting for sex as a covariate. The DESeq2 model is applied to the raw pseudobulk counts, and statistical results are extracted for the contrast between COVID-19 and normal conditions. These results are summarized and stored in a DataFrame.\n\nPathway enrichment analysis is conducted using the Decoupler package. The PROGENy model, a curated set of pathway gene sets, is employed to estimate pathway activities from the differential expression statistics. A univariate linear model (ulm) is used to infer pathway activity changes, and significantly altered pathways are identified based on an adjusted p-value threshold of 0.05.\n\nVisualization is a key component of the analysis. The initial cell type and disease state distribution are visualized using Uniform Manifold Approximation and Projection (UMAP). The cell type composition of the filtered pseudobulk samples is displayed using a bar plot. Principal Component Analysis (PCA) is performed to explore major sources of variation across pseudobulk samples, and the results are visualized in a PCA plot. Differential expression results are presented in a volcano plot, highlighting significant genes. Finally, the pathway enrichment results are visualized in a bar plot, focusing on significantly altered pathways.\n\nThe analysis concludes by saving the key results, including the differential expression results, significant pathway activities, and the processed pseudobulk data, to CSV and H5AD files. This ensures that the results are accessible for further exploration and reporting. The entire pseudobulk differential analysis workflow is designed to provide insights into the molecular changes associated with COVID-19 at the single-cell level.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing essential Python libraries for data handling, analysis, and visualization, including NumPy, pandas, Scanpy, Decoupler, Matplotlib, and PyDESeq2. These libraries are critical for managing single-cell RNA sequencing (scRNA-seq) data, performing differential expression analysis, and visualizing results. Global plotting parameters are configured to ensure consistent and clear visualizations.\n\nThe dataset used in this analysis is the COVID-19 single-cell RNA sequencing dataset from Decoupler’s built-in datasets. This dataset is loaded into an AnnData object, a specialized data structure for handling single-cell data. A summary of the dataset is displayed to provide an overview of its structure and content.\n\nThe preprocessing phase involves pseudobulking, a method to aggregate single-cell expression data into pseudobulk profiles by summing counts for each cell type within individual samples. This step reduces the complexity of single-cell data while preserving biological signals. Low-quality pseudobulk samples are filtered out based on thresholds for minimum cell count and total counts. The pseudobulk data is then normalized and preprocessed for downstream analysis, including log transformation and scaling. Raw counts are retained for differential expression analysis.\n\nDifferential expression analysis (DEA) is performed using PyDESeq2, a Python implementation of the DESeq2 method. The analysis models the effect of disease while accounting for sex as a covariate. The DESeq2 model is applied to the raw pseudobulk counts, and statistical results are extracted for the contrast between COVID-19 and normal conditions. These results are summarized and stored in a DataFrame.\n\nPathway enrichment analysis is conducted using the Decoupler package. The PROGENy model, a curated set of pathway gene sets, is employed to estimate pathway activities from the differential expression statistics. A univariate linear model (ulm) is used to infer pathway activity changes, and significantly altered pathways are identified based on an adjusted p-value threshold of 0.05.\n\nVisualization is a key component of the analysis. The initial cell type and disease state distribution are visualized using Uniform Manifold Approximation and Projection (UMAP). The cell type composition of the filtered pseudobulk samples is displayed using a bar plot. Principal Component Analysis (PCA) is performed to explore major sources of variation across pseudobulk samples, and the results are visualized in a PCA plot. Differential expression results are presented in a volcano plot, highlighting significant genes. Finally, the pathway enrichment results are visualized in a bar plot, focusing on significantly altered pathways.\n\nThe analysis concludes by saving the key results, including the differential expression results, significant pathway activities, and the processed pseudobulk data, to CSV and H5AD files. This ensures that the results are accessible for further exploration and reporting. The entire pseudobulk differential analysis workflow is designed to provide insights into the molecular changes associated with COVID-19 at the single-cell level.",
         "The analysis utilized Python libraries including Scanpy, Decoupler, and PyDESeq2 to process and analyze a COVID-19 single-cell RNA sequencing dataset. The dataset, loaded into an AnnData object, underwent pseudobulking to aggregate cell type-specific expression profiles, followed by quality filtering and normalization. Differential expression analysis was performed using PyDESeq2, modeling disease effects while accounting for sex as a covariate. Pathway enrichment analysis, conducted with Decoupler’s PROGENy model, identified significantly altered pathways. Visualization included UMAP for cell type and disease state distribution, PCA for sample variation, and volcano and bar plots for differential expression and pathway results. The workflow provided insights into COVID-19-associated molecular changes at single-cell resolution."
        ],
        [
         "45",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport decoupler as dc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.set_figure_params(figsize=(5, 5), frameon=False)\n\nsc.settings.verbosity = 3 # Set for detailed logging\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the mouse gastrulation dataset from Decoupler's built-in datasets.\n\nprint(\"Loading mouse gastrulation dataset...\")\n\nadata = dc.ds.erygast1k()\n\n\n\n#--------------------------------\n\n# Preprocessing & Trajectory Inference\n\n#--------------------------------\n\n\n\n# --- Preprocessing ---\n\n# The loaded data is already log-normalized. We proceed to dimensionality reduction.\n\n# Compute neighbors using the pre-computed PCA representation.\n\nprint(\"\\nPreprocessing and computing neighbors...\")\n\nsc.pp.neighbors(adata, use_rep='X_pca')\n\n\n\n# --- Trajectory Inference with PAGA/DPT ---\n\n# Compute a diffusion map, which is the basis for DPT.\n\nsc.tl.diffmap(adata)\n\n\n\n# Recompute neighbors on the diffusion map to better capture the manifold.\n\nsc.pp.neighbors(adata, use_rep='X_diffmap')\n\n\n\n# Run Partition-based Graph Abstraction (PAGA) to get a high-level\n\n# view of the connectivity between cell types.\n\nsc.tl.paga(adata, groups='celltype')\n\n\n\n# Define the root of the trajectory for pseudotime calculation.\n\nroot_celltype = \"Blood progenitors 1\"\n\niroot = np.flatnonzero(adata.obs['celltype'] == root_celltype)[0]\n\nadata.uns['iroot'] = iroot\n\n\n\n# Compute Diffusion Pseudotime (DPT) to order cells along the trajectory.\n\nsc.tl.dpt(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Enrichment Analysis\n\n#--------------------------------\n\n\n\n# --- 1. Transcription Factor (TF) Activity Analysis ---\n\nprint(\"\\nPerforming TF activity enrichment analysis...\")\n\n# Retrieve the CollecTRI TF-target network for mouse from Decoupler's resources.\n\ncollectri = dc.op.collectri(organism='mouse')\n\n\n\n# Run the Univariate Linear Model (ulm) to estimate TF activity scores for each cell.\n\n# The results are stored in adata.obsm['ulm_estimate'] and adata.obsm['ulm_pvals'].\n\ndc.run_ulm(data=adata, net=collectri, source='source', target='target', weight='weight', min_n=5)\n\n\n\n# Extract the TF activity scores into a separate AnnData object for easier plotting.\n\n# This creates a new object where the .X matrix contains the TF scores.\n\nscore_tf = dc.get_obsm_to_anndata(adata, obsm_key='ulm_estimate')\n\n\n\n\n\n# --- 2. Pathway Activity Analysis ---\n\nprint(\"\\nPerforming pathway activity enrichment analysis...\")\n\n# Retrieve the PROGENy pathway model for mouse.\n\nprogeny = dc.op.progeny(organism='mouse')\n\n\n\n# Run ULM again to estimate pathway activity scores.\n\n# Note: This overwrites the previous results in adata.obsm['ulm_*'], which is why\n\n# we extracted the TF scores to a separate object first.\n\ndc.run_ulm(data=adata, net=progeny, source='source', target='target', weight='weight', min_n=5)\n\n\n\n# Extract the pathway activity scores into another AnnData object.\n\nscore_pathway = dc.get_obsm_to_anndata(adata, obsm_key='ulm_estimate')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Raw Data and Trajectory ---\n\nprint(\"\\n--- Visualizing Results ---\")\n\n# Visualize the initial UMAP colored by cell type and developmental stage.\n\nsc.pl.umap(adata, color=['celltype', 'stage'], ncols=1, title=[\"Cell Types\", \"Developmental Stage\"])\n\n\n\n# Visualize the DPT pseudotime ordering on the UMAP.\n\nsc.pl.umap(adata, color=['dpt_pseudotime'], title=\"Diffusion Pseudotime\")\n\n\n\n\n\n# --- TF Activity Visualization ---\n\n# Choose a transcription factor to visualize, e.g., \"Klf3\".\n\ntf = 'Klf3'\n\n# Plot a UMAP colored by the Klf3 enrichment score.\n\nsc.pl.umap(score_tf, color=[tf], cmap='RdBu_r', vcenter=0, title=f\"{tf} Activity\")\n\n# Plot a violin plot showing the distribution of the Klf3 score across cell types.\n\nsc.pl.violin(score_tf, keys=[tf], groupby='celltype', rotation=90, ylabel=f'{tf} score')\n\n\n\n\n\n# --- Pathway Activity Visualization ---\n\n# Choose a pathway to visualize, e.g., \"PI3K\".\n\npathway = 'PI3K'\n\n# Plot a UMAP colored by the PI3K pathway enrichment score.\n\nsc.pl.umap(score_pathway, color=[pathway], cmap='RdBu_r', vcenter=0, title=f\"{pathway} Pathway Activity\")\n\n# Plot a violin plot showing the distribution of the PI3K score across cell types.\n\nsc.pl.violin(score_pathway, keys=[pathway], groupby='celltype', rotation=90, ylabel=f'{pathway} score')\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final AnnData object, which contains all analysis results.\n\nprint(\"\\nSaving final AnnData object...\")\n\nadata.write(\"erygast1k_with_trajectory_and_enrichment.h5ad\")\n\n\n\nprint(\"\\nTrajectory and enrichment analysis completed!\")\n\n",
         "The analysis pipeline begins by importing essential Python libraries for data handling, analysis, and visualization, including NumPy, Scanpy, and Decoupler. These libraries are critical for single-cell RNA sequencing data analysis, enabling tasks such as dimensionality reduction, trajectory inference, and enrichment analysis. Global plotting parameters are configured to ensure consistent and clear visualizations, with detailed logging enabled for transparency.\n\nThe mouse gastrulation dataset is loaded from Decoupler’s built-in datasets. This dataset is preprocessed, with the data already log-normalized. Dimensionality reduction is performed by computing a neighborhood graph using the precomputed principal component analysis (PCA) representation. To infer developmental trajectories, a diffusion map is computed, and the neighborhood graph is recomputed on this map to better capture the underlying manifold. Partition-based Graph Abstraction (PAGA) is applied to visualize high-level connectivity between cell types. The root of the trajectory is defined as \"Blood progenitors 1,\" and Diffusion Pseudotime (DPT) is calculated to order cells along the inferred trajectory.\n\nFor enrichment analysis, transcription factor (TF) activity is estimated using the CollecTRI TF-target network for mice. The Univariate Linear Model (ULM) is employed to compute TF activity scores, which are stored in the AnnData object. Similarly, pathway activity is estimated using the PROGENy pathway model, with ULM applied again to compute pathway activity scores. Both TF and pathway activity scores are extracted into separate AnnData objects for downstream visualization.\n\nVisualization begins with generating UMAP plots of the raw data, colored by cell type and developmental stage. The DPT pseudotime ordering is also visualized on the UMAP. Specific transcription factors and pathways are highlighted in subsequent visualizations. For example, the activity of the transcription factor Klf3 is depicted using UMAP and violin plots, showing its distribution across cell types. Similarly, the activity of the PI3K pathway is visualized using UMAP and violin plots.\n\nFinally, the AnnData object containing all analysis results, including trajectory inference and enrichment scores, is saved to a file for further use. This comprehensive pipeline provides a detailed workflow for analyzing single-cell RNA sequencing data, from preprocessing and trajectory inference to enrichment analysis and visualization.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis pipeline begins by importing essential Python libraries for data handling, analysis, and visualization, including NumPy, Scanpy, and Decoupler. These libraries are critical for single-cell RNA sequencing data analysis, enabling tasks such as dimensionality reduction, trajectory inference, and enrichment analysis. Global plotting parameters are configured to ensure consistent and clear visualizations, with detailed logging enabled for transparency.\n\nThe mouse gastrulation dataset is loaded from Decoupler’s built-in datasets. This dataset is preprocessed, with the data already log-normalized. Dimensionality reduction is performed by computing a neighborhood graph using the precomputed principal component analysis (PCA) representation. To infer developmental trajectories, a diffusion map is computed, and the neighborhood graph is recomputed on this map to better capture the underlying manifold. Partition-based Graph Abstraction (PAGA) is applied to visualize high-level connectivity between cell types. The root of the trajectory is defined as \"Blood progenitors 1,\" and Diffusion Pseudotime (DPT) is calculated to order cells along the inferred trajectory.\n\nFor enrichment analysis, transcription factor (TF) activity is estimated using the CollecTRI TF-target network for mice. The Univariate Linear Model (ULM) is employed to compute TF activity scores, which are stored in the AnnData object. Similarly, pathway activity is estimated using the PROGENy pathway model, with ULM applied again to compute pathway activity scores. Both TF and pathway activity scores are extracted into separate AnnData objects for downstream visualization.\n\nVisualization begins with generating UMAP plots of the raw data, colored by cell type and developmental stage. The DPT pseudotime ordering is also visualized on the UMAP. Specific transcription factors and pathways are highlighted in subsequent visualizations. For example, the activity of the transcription factor Klf3 is depicted using UMAP and violin plots, showing its distribution across cell types. Similarly, the activity of the PI3K pathway is visualized using UMAP and violin plots.\n\nFinally, the AnnData object containing all analysis results, including trajectory inference and enrichment scores, is saved to a file for further use. This comprehensive pipeline provides a detailed workflow for analyzing single-cell RNA sequencing data, from preprocessing and trajectory inference to enrichment analysis and visualization.",
         "The analysis was conducted in Python using Scanpy and Decoupler for single-cell RNA sequencing data processing, trajectory inference, and enrichment analysis. The mouse gastrulation dataset, preprocessed and log-normalized, underwent dimensionality reduction via principal component analysis (PCA) and neighborhood graph construction. Developmental trajectories were inferred using diffusion maps and Partition-based Graph Abstraction (PAGA), with Diffusion Pseudotime (DPT) ordering cells from the root \"Blood progenitors 1.\" Transcription factor (TF) and pathway activities were estimated using the CollecTRI TF-target network and PROGENy pathway model, respectively, with scores computed via the Univariate Linear Model (ULM). Results were visualized using UMAP and violin plots, highlighting TF Klf3 and the PI3K pathway activity across cell types."
        ],
        [
         "46",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for system operations, data handling, analysis, and plotting\n\nimport warnings\n\nfrom pathlib import Path\n\nimport loompy\n\nimport matplotlib.pyplot as plt\n\nimport numpy as np\n\nimport pandas as pd\n\nimport scanpy as sc\n\nimport seaborn as sns\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.filterwarnings(\"ignore\")\n\n# Set global plotting parameters\n\nsc.set_figure_params(facecolor=\"white\")\n\nsns.set_theme()\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the single-cell RNA-seq data.\n\n# NOTE: Replace the path with the actual location of your data file.\n\nadata = sc.read_h5ad(\"path/to/your_openproblems_bmmc_multiome_genes_filtered.h5ad\")\n\nprint(\"Initial dataset shape:\", adata.shape)\n\n\n\n#--------------------------------\n\n# Data Preprocessing\n\n#--------------------------------\n\n\n\n# Filter for gene expression (GEX) features if the dataset is multi-modal.\n\nrna = adata[:, adata.var[\"feature_types\"] == \"GEX\"].copy()\n\ndel adata # Free up memory\n\n\n\n# Apply a log-transformation to the data.\n\nsc.pp.log1p(rna)\n\n\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n\nsc.pp.highly_variable_genes(rna, batch_key=\"batch\", flavor=\"seurat\")\n\nprint(\"Preprocessing complete.\")\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Prepare for SCENIC\n\n#--------------------------------\n\n\n\n# --- Step 1: Subset Data for Analysis ---\n\n# For demonstration, we subset the data to a single batch/donor ('s1d1').\n\n# SCENIC can be run on the full dataset, but this reduces computational time for an example.\n\nadata_batch = rna[rna.obs[\"batch\"] == \"s1d1\", :].copy()\n\nprint(f\"\\nSubsetting data to batch 's1d1'. New shape: {adata_batch.shape}\")\n\n\n\n\n\n# --- Step 2: Create Loom File for SCENIC ---\n\n# SCENIC requires a specific file format (.loom) for its command-line interface.\n\n# We create this file from our preprocessed AnnData object.\n\nloom_path = \"data/neurips_processed_input.loom\"\n\nPath(\"data\").mkdir(exist_ok=True) # Ensure the directory exists\n\n\n\n# Define row and column attributes required by the loom format.\n\nrow_attrs = {\"Gene\": np.array(adata_batch.var.index)}\n\ncol_attrs = {\n\n\"CellID\": np.array(adata_batch.obs.index),\n\n\"nGene\": np.array(np.sum(adata_batch.X > 0, axis=1)).flatten(),\n\n\"nUMI\": np.array(np.sum(adata_batch.X, axis=1)).flatten(),\n\n}\n\n\n\n# Create the loom file. Note that the expression matrix must be transposed.\n\nprint(f\"Creating loom file at: {loom_path}\")\n\nloompy.create(loom_path, adata_batch.X.T, row_attrs, col_attrs)\n\n\n\n\n\n#--------------------------------\n\n# External Step: Run pyscenic\n\n#--------------------------------\n\n\n\n# The core GRN inference in SCENIC is run via a command-line tool.\n\n# This step is performed outside of this script in a terminal environment.\n\n#\n\n# EXAMPLE COMMAND:\n\n# pyscenic grn data/neurips_processed_input.loom allTFs_hg38.txt \\\n\n# -o adj.csv --num_workers 3\n\n#\n\n# - `data/neurips_processed_input.loom`: The input file we created.\n\n# - `allTFs_hg38.txt`: A list of known transcription factors.\n\n# - `-o adj.csv`: The output file containing gene regulatory network adjacencies.\n\n#\n\n# This command produces a file ('adj.csv') which we load in the next step.\n\nprint(\"\\nSCENIC GRN inference must be run externally via the command line.\")\n\n\n\n\n\n#--------------------------------\n\n# Post-Analysis: Load and Visualize GRN Results\n\n#--------------------------------\n\n\n\n# Assuming the 'pyscenic grn' command has been run, we now load its output.\n\ntry:\n\nresults_adjacencies = pd.read_csv(\"adj.csv\")\n\nprint(f\"\\nSuccessfully loaded GRN results. Found {results_adjacencies.shape[0]} associations.\")\n\nprint(results_adjacencies.head())\n\nexcept FileNotFoundError:\n\nprint(\"\\n'adj.csv' not found. Please run the external pyscenic command first.\")\n\nresults_adjacencies = None\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP Visualization ---\n\n# Plot UMAPs to visualize the overall data structure before and after subsetting.\n\n# Note: This assumes 'GEX_X_umap' coordinates are pre-computed in the loaded file.\n\nprint(\"\\n--- Visualizing Data ---\")\n\nsc.pl.embedding(rna, \"GEX_X_umap\", color=[\"cell_type\", \"batch\"], title=\"UMAP: All Cells\")\n\nsc.pl.embedding(adata_batch, \"GEX_X_umap\", color=[\"cell_type\", \"batch\"], title=\"UMAP: Donor s1d1\")\n\n\n\n# --- GRN Importance Score Distribution ---\n\n# If GRN results were loaded, plot a histogram of the importance scores.\n\n# This score reflects the strength of the regulatory link between a TF and its target.\n\nif results_adjacencies is not None:\n\nplt.figure(figsize=(8, 5))\n\nplt.hist(np.log10(results_adjacencies[\"importance\"] + 1e-10), bins=50, color=\"skyblue\", edgecolor=\"black\")\n\nplt.xlabel(\"Log10(Importance Score)\")\n\nplt.ylabel(\"Frequency\")\n\nplt.title(\"Distribution of GRN Importance Scores\")\n\nplt.show()\n\n\n\n# --- Genes Detected per Cell ---\n\n# Plot the distribution of the number of genes detected per cell. This is a useful QC metric.\n\nn_genes_detected_per_cell = np.sum(adata_batch.X > 0, axis=1)\n\nn_genes_series = pd.Series(n_genes_detected_per_cell.A.flatten() if hasattr(n_genes_detected_per_cell, 'A') else n_genes_detected_per_cell.flatten())\n\n\n\nfig, ax = plt.subplots(figsize=(8, 5), dpi=100)\n\nsns.histplot(n_genes_series, bins=\"fd\", edgecolor=\"black\", ax=ax)\n\nax.set_xlabel(\"Number of Genes Detected\")\n\nax.set_ylabel(\"Number of Cells\")\n\nax.set_title(\"Distribution of Genes Detected per Cell (Donor s1d1)\")\n\nplt.tight_layout()\n\nplt.show()\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the subsetted AnnData object used for the analysis.\n\nprint(\"\\nSaving subsetted AnnData object...\")\n\nadata_batch.write(\"s1d1_preprocessed_for_scenic.h5ad\")\n\n\n\nprint(\"\\nSCENIC data preparation script completed!\")\n\n",
         "The analysis begins by importing necessary Python libraries for system operations, data handling, analysis, and visualization. These include `warnings`, `pathlib`, `loompy`, `matplotlib.pyplot`, `numpy`, `pandas`, `scanpy`, and `seaborn`. Warnings are suppressed to ensure cleaner output during execution, and global plotting parameters are set to optimize visualization aesthetics.\n\nSingle-cell RNA-seq data is loaded from an `.h5ad` file using the `scanpy` library. The dataset's initial shape is printed to confirm successful loading. The data is then preprocessed by filtering for gene expression (GEX) features if the dataset is multi-modal. This step ensures that only relevant gene expression data is retained for downstream analysis. The filtered data undergoes a log-transformation to stabilize variance, and highly variable genes (HVGs) are identified using a batch-aware method. This step is critical for reducing noise and focusing on biologically relevant genes.\n\nTo prepare for SCENIC (Single-Cell Regulatory Network Inference and Clustering), the data is subset to a single batch/donor ('s1d1') to reduce computational time for demonstration purposes. A `.loom` file, required by SCENIC, is created from the preprocessed AnnData object. Row and column attributes, including gene names, cell IDs, and counts of genes and UMIs, are defined to match the `.loom` format. The expression matrix is transposed to align with SCENIC's requirements, and the file is saved to a specified directory.\n\nThe core gene regulatory network (GRN) inference in SCENIC is performed externally using the `pyscenic` command-line tool. This step involves running the `pyscenic grn` command, which takes the `.loom` file and a list of known transcription factors as inputs and outputs a file containing GRN adjacencies. This step is essential for identifying regulatory relationships between transcription factors and their target genes.\n\nAfter the external GRN inference, the results are loaded into the script. If successful, the GRN associations are printed for verification. Visualizations are then generated to explore the data. UMAP plots are created to visualize the overall data structure before and after subsetting, with cells colored by cell type and batch. If GRN results are available, a histogram of the importance scores is plotted to assess the strength of regulatory links. Additionally, the distribution of the number of genes detected per cell is visualized as a quality control metric.\n\nFinally, the subsetted AnnData object used for the analysis is saved to a `.h5ad` file for future use. This ensures that the preprocessed data is readily available for subsequent analyses. The script concludes by confirming the successful completion of the SCENIC data preparation process.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis begins by importing necessary Python libraries for system operations, data handling, analysis, and visualization. These include `warnings`, `pathlib`, `loompy`, `matplotlib.pyplot`, `numpy`, `pandas`, `scanpy`, and `seaborn`. Warnings are suppressed to ensure cleaner output during execution, and global plotting parameters are set to optimize visualization aesthetics.\n\nSingle-cell RNA-seq data is loaded from an `.h5ad` file using the `scanpy` library. The dataset's initial shape is printed to confirm successful loading. The data is then preprocessed by filtering for gene expression (GEX) features if the dataset is multi-modal. This step ensures that only relevant gene expression data is retained for downstream analysis. The filtered data undergoes a log-transformation to stabilize variance, and highly variable genes (HVGs) are identified using a batch-aware method. This step is critical for reducing noise and focusing on biologically relevant genes.\n\nTo prepare for SCENIC (Single-Cell Regulatory Network Inference and Clustering), the data is subset to a single batch/donor ('s1d1') to reduce computational time for demonstration purposes. A `.loom` file, required by SCENIC, is created from the preprocessed AnnData object. Row and column attributes, including gene names, cell IDs, and counts of genes and UMIs, are defined to match the `.loom` format. The expression matrix is transposed to align with SCENIC's requirements, and the file is saved to a specified directory.\n\nThe core gene regulatory network (GRN) inference in SCENIC is performed externally using the `pyscenic` command-line tool. This step involves running the `pyscenic grn` command, which takes the `.loom` file and a list of known transcription factors as inputs and outputs a file containing GRN adjacencies. This step is essential for identifying regulatory relationships between transcription factors and their target genes.\n\nAfter the external GRN inference, the results are loaded into the script. If successful, the GRN associations are printed for verification. Visualizations are then generated to explore the data. UMAP plots are created to visualize the overall data structure before and after subsetting, with cells colored by cell type and batch. If GRN results are available, a histogram of the importance scores is plotted to assess the strength of regulatory links. Additionally, the distribution of the number of genes detected per cell is visualized as a quality control metric.\n\nFinally, the subsetted AnnData object used for the analysis is saved to a `.h5ad` file for future use. This ensures that the preprocessed data is readily available for subsequent analyses. The script concludes by confirming the successful completion of the SCENIC data preparation process.",
         "The analysis was conducted in Python using libraries such as `scanpy`, `loompy`, and `pyscenic` for single-cell RNA-seq data processing and regulatory network inference. The workflow began by loading single-cell data from an `.h5ad` file and filtering for gene expression features to ensure data relevance. A log-transformation was applied to stabilize variance, followed by batch-aware identification of highly variable genes (HVGs) to reduce noise. The dataset was subset to a single batch ('s1d1') for computational efficiency, and a `.loom` file was generated for SCENIC compatibility. \n\nThe core gene regulatory network (GRN) inference was performed externally using `pyscenic grn`, which identified regulatory relationships between transcription factors and target genes. Results were visualized through UMAP plots, colored by cell type and batch, and a histogram of GRN importance scores was generated to assess regulatory strength. Quality control included visualizing the distribution of detected genes per cell. The subsetted AnnData object was saved for future use, ensuring reproducibility. This workflow provided insights into gene regulatory networks and cell-type-specific regulatory dynamics."
        ],
        [
         "47",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for inferCNV analysis and data handling\n\nlibrary(infercnv) # Main package for inferring copy number variations\n\nlibrary(Seurat) # Used for handling single-cell RNA-seq data\n\nlibrary(RColorBrewer) # Provides color palettes for plotting\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed Seurat object containing your scRNA-seq data.\n\n# NOTE: Replace the path with the actual location of your .RDS file.\n\npbmc <- readRDS(\"../data/pbmc.RDS\")\n\n\n\n# Extract the raw count matrix from the Seurat object.\n\n# inferCNV requires a raw integer count matrix with genes as rows and cells as columns.\n\nraw_counts <- pbmc@assays$RNA@counts\n\n\n\n#--------------------------------\n\n# Prepare inferCNV Inputs\n\n#--------------------------------\n\n\n\n# --- Create Cell Annotation File ---\n\n# This file maps each cell ID to a specific cell type or cluster annotation.\n\n# It should be a two-column, tab-delimited file without headers.\n\ncell_annotations <- data.frame(\n\ncell_id = colnames(pbmc),\n\ncell_type = as.character(pbmc@meta.data$seurat_clusters),\n\nstringsAsFactors = FALSE\n\n)\n\n# Write the annotation file to disk.\n\nwrite.table(\n\ncell_annotations,\n\nfile = \"celltype_pbmc.txt\",\n\nsep = \"\\t\",\n\nrow.names = FALSE,\n\ncol.names = FALSE,\n\nquote = FALSE\n\n)\n\n\n\n# --- Gene Order File ---\n\n# This file provides the genomic coordinates for each gene, which inferCNV uses\n\n# to order the genes along chromosomes in the final heatmap.\n\n# It should be a three-column file: Gene, Chromosome, Start Position.\n\n# NOTE: You must provide your own gene order file (e.g., \"gencode_v19_gene_pos.txt\").\n\ngene_order_file_path <- \"gencode_v19_gene_pos.txt\"\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Run inferCNV\n\n#--------------------------------\n\n\n\n# --- Create inferCNV Object ---\n\n# This object is the central data structure for the analysis.\n\n# `ref_group_names` specifies which cell groups to use as the \"normal\" reference\n\n# for calculating relative expression changes.\n\ninfercnv_obj <- CreateInfercnvObject(\n\nraw_counts_matrix = as.matrix(raw_counts),\n\nannotations_file = \"celltype_pbmc.txt\",\n\ngene_order_file = gene_order_file_path,\n\nref_group_names = c(\"0\", \"8\") # Example: Use clusters 0 and 8 as reference\n\n)\n\n\n\n# --- Run the inferCNV pipeline ---\n\n# This is the main computational step. It normalizes expression, centers it\n\n# relative to the reference cells, and identifies large-scale chromosomal alterations.\n\ninfercnv_obj <- infercnv::run(\n\ninfercnv_obj,\n\ncutoff = 0.1, # Recommended for 10x Genomics data\n\nout_dir = \"infercnv_output/\", # Directory to store all output files\n\ncluster_by_groups = TRUE, # Group cells by their annotation\n\ndenoise = FALSE, # Optional: apply denoising\n\nHMM = FALSE, # Optional: run Hidden Markov Model for CNV prediction\n\nno_prelim_plot = TRUE, # Skip plotting the preliminary heatmap\n\nanalysis_mode = \"subclusters\" # Analyze at the subcluster level for more detail\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Define a custom color palette for the final CNV heatmap.\n\ncustom_color_pal <- colorRampPalette(c(\"#8DD3C7\", \"white\", \"#BC80BD\"))(100)\n\n\n\n# Generate the final CNV plot. This creates a PDF file in the output directory.\n\n# The plot visualizes inferred amplifications (purple) and deletions (blue-green)\n\n# across the genome for each cell.\n\ninfercnv::plot_cnv(\n\ninfercnv_obj,\n\nplot_chr_scale = TRUE,\n\noutput_filename = \"infercnv_final_plot\",\n\noutput_format = \"pdf\",\n\ncustom_color_pal = custom_color_pal\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# The `infercnv::run` function automatically saves the results and plots\n\n# in the specified `out_dir`. The final inferCNV object is also returned,\n\n# which can be saved for later inspection if needed.\n\nsaveRDS(infercnv_obj, file = \"infercnv_output/final_infercnv_object.rds\")\n\n\n\nprint(\"\\ninferCNV analysis completed! Check the 'infercnv_output' directory for results.\")\n\n",
         "To perform inferCNV analysis on single-cell RNA-seq data, the following steps were executed using R. The analysis aimed to infer copy number variations (CNVs) from the single-cell data, leveraging the inferCNV package and associated tools.\n\nFirst, the necessary R libraries were loaded. The infercnv package was used for CNV inference, Seurat for handling single-cell RNA-seq data, and RColorBrewer for generating color palettes for visualization. These libraries are essential for data processing, analysis, and visualization.\n\nNext, the pre-processed single-cell RNA-seq data was loaded from a Seurat object stored in an .RDS file. The raw count matrix was extracted from the Seurat object, as inferCNV requires a raw integer count matrix with genes as rows and cells as columns. This matrix serves as the primary input for the CNV analysis.\n\nTo prepare the inputs for inferCNV, two key files were generated. First, a cell annotation file was created, mapping each cell ID to its corresponding cell type or cluster annotation. This file was written to disk in a tab-delimited format without headers. Second, a gene order file was specified, providing the genomic coordinates for each gene. This file, which must be provided by the user, ensures that genes are ordered along chromosomes in the final heatmap.\n\nThe inferCNV object was then created, serving as the central data structure for the analysis. The raw count matrix, cell annotation file, and gene order file were passed as inputs. Additionally, specific cell groups were designated as the \"normal\" reference for calculating relative expression changes. In this example, clusters 0 and 8 were used as the reference groups.\n\nThe inferCNV pipeline was executed, which involved normalizing expression data, centering it relative to the reference cells, and identifying large-scale chromosomal alterations. Key parameters included a cutoff value of 0.1, which is recommended for 10x Genomics data, and the analysis mode was set to \"subclusters\" for more detailed insights. The results and intermediate files were saved in a specified output directory.\n\nFor visualization, a custom color palette was defined to generate the final CNV heatmap. The heatmap was plotted using the infercnv::plot_cnv function, which visualizes inferred amplifications and deletions across the genome for each cell. The plot was saved as a PDF file in the output directory.\n\nFinally, the inferCNV object was saved to disk for later inspection, and a completion message was printed, directing the user to the output directory for the results. This comprehensive workflow provides a robust approach to inferring CNVs from single-cell RNA-seq data, leveraging the inferCNV package and associated tools in R.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform inferCNV analysis on single-cell RNA-seq data, the following steps were executed using R. The analysis aimed to infer copy number variations (CNVs) from the single-cell data, leveraging the inferCNV package and associated tools.\n\nFirst, the necessary R libraries were loaded. The infercnv package was used for CNV inference, Seurat for handling single-cell RNA-seq data, and RColorBrewer for generating color palettes for visualization. These libraries are essential for data processing, analysis, and visualization.\n\nNext, the pre-processed single-cell RNA-seq data was loaded from a Seurat object stored in an .RDS file. The raw count matrix was extracted from the Seurat object, as inferCNV requires a raw integer count matrix with genes as rows and cells as columns. This matrix serves as the primary input for the CNV analysis.\n\nTo prepare the inputs for inferCNV, two key files were generated. First, a cell annotation file was created, mapping each cell ID to its corresponding cell type or cluster annotation. This file was written to disk in a tab-delimited format without headers. Second, a gene order file was specified, providing the genomic coordinates for each gene. This file, which must be provided by the user, ensures that genes are ordered along chromosomes in the final heatmap.\n\nThe inferCNV object was then created, serving as the central data structure for the analysis. The raw count matrix, cell annotation file, and gene order file were passed as inputs. Additionally, specific cell groups were designated as the \"normal\" reference for calculating relative expression changes. In this example, clusters 0 and 8 were used as the reference groups.\n\nThe inferCNV pipeline was executed, which involved normalizing expression data, centering it relative to the reference cells, and identifying large-scale chromosomal alterations. Key parameters included a cutoff value of 0.1, which is recommended for 10x Genomics data, and the analysis mode was set to \"subclusters\" for more detailed insights. The results and intermediate files were saved in a specified output directory.\n\nFor visualization, a custom color palette was defined to generate the final CNV heatmap. The heatmap was plotted using the infercnv::plot_cnv function, which visualizes inferred amplifications and deletions across the genome for each cell. The plot was saved as a PDF file in the output directory.\n\nFinally, the inferCNV object was saved to disk for later inspection, and a completion message was printed, directing the user to the output directory for the results. This comprehensive workflow provides a robust approach to inferring CNVs from single-cell RNA-seq data, leveraging the inferCNV package and associated tools in R.",
         "The analysis of copy number variations (CNVs) in single-cell RNA sequencing data was conducted using the infercnv package in R, integrated with Seurat for data handling and RColorBrewer for visualization. The workflow began with the extraction of raw count matrices from a pre-processed Seurat object, followed by the preparation of cell annotation and gene order files. The inferCNV object was initialized using these inputs, with designated reference cell groups for normalization. The pipeline executed CNV inference, identifying chromosomal alterations through expression normalization and relative centering. Results were visualized as a heatmap, depicting amplifications and deletions across the genome, and saved for further analysis."
        ],
        [
         "48",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for analysis, data handling, and plotting\n\nlibrary(Seurat) # For single-cell data management\n\nlibrary(dplyr) # For data manipulation\n\nlibrary(tibble) # For advanced data frame functionalities\n\nlibrary(infercnv) # Main package for inferring copy number variations\n\nlibrary(ggpubr) # For creating publication-quality plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed Seurat object.\n\n# NOTE: Replace the path with the actual location of your .RDS file.\n\nscRNA_obj <- readRDS(\"./scRNA_harmony_EP.rds\")\n\n\n\n#--------------------------------\n\n# Prepare inferCNV Inputs\n\n#--------------------------------\n\n\n\n# --- Annotate and Subset Cells ---\n\n# Annotate cell types based on clustering information.\n\n# Here, clusters \"0\" and \"18\" are defined as \"Normal\" reference cells.\n\nmeta_data <- scRNA_obj@meta.data %>%\n\nmutate(cell_type = case_when(\n\nRNA_snn_res.0.6 %in% c(\"0\", \"18\") ~ \"Normal\",\n\nTRUE ~ \"Tumor\"\n\n))\n\n\n\n# Get cell IDs for tumor and normal groups.\n\ntumor_cells <- rownames(meta_data)[meta_data$cell_type == \"Tumor\"]\n\nnormal_cells <- rownames(meta_data)[meta_data$cell_type == \"Normal\"]\n\n\n\n# To manage computational load, randomly sample a subset of tumor cells.\n\nset.seed(123)\n\ntumor_sample <- sample(tumor_cells, min(3000, length(tumor_cells)), replace = FALSE)\n\n\n\n# Combine the sampled tumor cells with all normal cells for the analysis.\n\nselected_cells <- c(tumor_sample, normal_cells)\n\nnew_scRNA <- subset(scRNA_obj, cells = selected_cells)\n\n\n\n\n\n# --- Extract Expression Matrix ---\n\n# inferCNV requires a raw integer count matrix.\n\nexprMatrix <- as.matrix(GetAssayData(new_scRNA, slot = \"counts\"))\n\n\n\n\n\n# --- Create Cell Annotation File ---\n\n# This file maps each cell ID to its cluster ID. inferCNV uses this for grouping.\n\n# NOTE: This is different from the Tumor/Normal annotation used for sampling.\n\ncellAnnot <- new_scRNA@meta.data[, \"RNA_snn_res.0.6\", drop = FALSE]\n\n\n\n\n\n# --- Gene Order File ---\n\n# This file provides the genomic coordinates for each gene.\n\n# NOTE: You must provide your own gene order file.\n\ngene_order_file_path <- \"path/to/gene_pos.txt\"\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Run inferCNV\n\n#--------------------------------\n\n\n\n# --- Create inferCNV Object ---\n\n# `ref_group_names` specifies which clusters to use as the \"normal\" reference.\n\ninfercnv_obj <- CreateInfercnvObject(\n\nraw_counts_matrix = exprMatrix,\n\nannotations_file = cellAnnot, # This can be a data.frame directly\n\ndelim = \"\\t\",\n\ngene_order_file = gene_order_file_path,\n\nref_group_names = c(\"0\", \"18\") # Use normal clusters as reference\n\n)\n\n\n\n# --- Run the inferCNV pipeline ---\n\n# This is the main computational step.\n\ninfercnv_obj <- infercnv::run(\n\ninfercnv_obj,\n\ncutoff = 0.1, # Recommended for 10x Genomics data\n\nout_dir = \"./output/cnv_epithelial/\", # Directory for all output files\n\ncluster_by_groups = FALSE, # Cluster cells across all groups\n\nhclust_method = \"ward.D2\", # Hierarchical clustering method\n\ndenoise = TRUE, # Apply denoising to improve signal\n\nHMM = FALSE # Disable HMM-based CNV prediction in this run\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Downstream Analysis: CNV Scoring (Optional)\n\n#--------------------------------\n\n\n\n# --- Calculate CNV Score ---\n\n# This section calculates a per-cell CNV score based on the inferCNV output.\n\n# Read the observation and reference matrices generated by infercnv::run.\n\nobs <- read.table(\"./output/cnv_epithelial/infercnv.observations.txt\", header = TRUE)\n\nref <- read.table(\"./output/cnv_epithelial/infercnv.references.txt\", header = TRUE)\n\n\n\n# Combine, scale, and calculate a squared residual score.\n\nexpr <- cbind(obs, ref)\n\nexpr_scale <- scale(t(expr))\n\ntmp1 <- sweep(expr_scale, 2, apply(expr_scale, 2, min), FUN = \"-\")\n\ntmp2 <- apply(expr_scale, 2, max) - apply(expr_scale, 2, min)\n\nexpr_1 <- t(2 * sweep(tmp1, 2, tmp2, FUN = \"/\") - 1)\n\ncnv_score_df <- as.data.frame(colSums(expr_1 * expr_1))\n\ncolnames(cnv_score_df) <- \"cnv_score\"\n\n\n\n# Format the CNV score DataFrame.\n\ncnv_score_df <- cnv_score_df %>%\n\nrownames_to_column(var = \"cell\") %>%\n\nmutate(cell = gsub(\"\\\\.\", \"-\", cell)) # Ensure cell names match Seurat object\n\n\n\n\n\n# --- Merge Scores into Seurat Object ---\n\n# Subset the Seurat object to match the cells with CNV scores and merge the scores.\n\nscRNA_sample <- subset(scRNA_obj, cells = cnv_score_df$cell)\n\nmeta_with_cnv <- scRNA_sample@meta.data %>%\n\nrownames_to_column(var = \"cell\") %>%\n\ninner_join(cnv_score_df, by = \"cell\") %>%\n\ncolumn_to_rownames(var = \"cell\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- CNV Score Boxplot ---\n\n# Create a boxplot to visualize the distribution of CNV scores across different clusters.\n\ncnv_boxplot <- ggboxplot(\n\nmeta_with_cnv,\n\nx = \"RNA_snn_res.0.6\",\n\ny = \"cnv_score\",\n\nfill = \"RNA_snn_res.0.6\"\n\n) +\n\nscale_y_continuous(limits = c(0, 3000)) +\n\nxlab(\"Cluster\") +\n\nylab(\"CNV Score\") +\n\ntheme(\n\npanel.border = element_rect(colour = \"black\", fill = NA, size = 0.5),\n\nlegend.position = \"none\"\n\n)\n\nprint(cnv_boxplot)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# The `infercnv::run` function automatically saves the main heatmap and other results.\n\n# Here, we explicitly save the final inferCNV object and the CNV score plot.\n\nsaveRDS(infercnv_obj, file = \"./output/cnv_epithelial/final_infercnv_object.rds\")\n\n\n\n# Save the CNV score boxplot to a PDF file.\n\nggsave(\n\nplot = cnv_boxplot,\n\nfilename = \"./output/cnv_epithelial/CNV_score_by_cluster.pdf\",\n\nwidth = 7,\n\nheight = 4\n\n)\n\n\n\nprint(\"\\ninferCNV analysis and CNV scoring completed!\")\n\n",
         "The analysis pipeline begins with the loading of essential R packages required for single-cell RNA sequencing data management, data manipulation, and visualization. These include Seurat for single-cell data handling, dplyr for data manipulation, tibble for advanced data frame functionalities, infercnv for inferring copy number variations (CNVs), and ggpubr for creating publication-quality plots. These packages are critical for the subsequent steps in the analysis.\n\nThe pre-processed Seurat object, containing single-cell RNA sequencing data, is loaded from an RDS file. This object serves as the foundation for the analysis and includes metadata such as cluster assignments and cell type annotations. The path to the RDS file must be specified to ensure the correct data is loaded.\n\nTo prepare the data for inferCNV analysis, cell types are annotated based on clustering information. Specifically, clusters labeled \"0\" and \"18\" are designated as \"Normal\" reference cells, while all other cells are labeled as \"Tumor.\" This annotation is crucial for distinguishing normal reference cells from tumor cells during CNV inference. A subset of tumor cells is randomly sampled to manage computational load, and the sampled tumor cells are combined with all normal cells for downstream analysis. The raw integer count matrix is extracted from the subsetted Seurat object, as inferCNV requires raw counts for accurate CNV inference. Additionally, a cell annotation file is created to map each cell ID to its cluster ID, and a gene order file specifying genomic coordinates for each gene is required. The path to the gene order file must be provided.\n\nThe inferCNV analysis is initiated by creating an inferCNV object using the raw count matrix, cell annotation file, and gene order file. The \"Normal\" clusters are specified as reference groups for CNV inference. The inferCNV pipeline is then executed with specific parameters: a cutoff of 0.1 is applied, which is recommended for 10x Genomics data; the output is saved to a designated directory; hierarchical clustering is performed using the \"ward.D2\" method; denoising is applied to improve signal clarity; and HMM-based CNV prediction is disabled for this run.\n\nFollowing the inferCNV analysis, an optional downstream step involves calculating a per-cell CNV score. This is achieved by reading the observation and reference matrices generated by inferCNV, combining and scaling the data, and calculating a squared residual score. The resulting CNV scores are formatted into a data frame and merged back into the Seurat object for further analysis.\n\nTo visualize the results, a boxplot is generated to display the distribution of CNV scores across different clusters. The plot is customized with specific axis labels, limits, and styling to ensure clarity and publication readiness.\n\nFinally, the inferCNV object and the CNV score boxplot are saved to the specified output directory. The inferCNV object is saved as an RDS file, while the boxplot is saved as a PDF. This ensures that all results are preserved for future reference or further analysis. The completion of the inferCNV analysis and CNV scoring is confirmed with a printed message.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nThe analysis pipeline begins with the loading of essential R packages required for single-cell RNA sequencing data management, data manipulation, and visualization. These include Seurat for single-cell data handling, dplyr for data manipulation, tibble for advanced data frame functionalities, infercnv for inferring copy number variations (CNVs), and ggpubr for creating publication-quality plots. These packages are critical for the subsequent steps in the analysis.\n\nThe pre-processed Seurat object, containing single-cell RNA sequencing data, is loaded from an RDS file. This object serves as the foundation for the analysis and includes metadata such as cluster assignments and cell type annotations. The path to the RDS file must be specified to ensure the correct data is loaded.\n\nTo prepare the data for inferCNV analysis, cell types are annotated based on clustering information. Specifically, clusters labeled \"0\" and \"18\" are designated as \"Normal\" reference cells, while all other cells are labeled as \"Tumor.\" This annotation is crucial for distinguishing normal reference cells from tumor cells during CNV inference. A subset of tumor cells is randomly sampled to manage computational load, and the sampled tumor cells are combined with all normal cells for downstream analysis. The raw integer count matrix is extracted from the subsetted Seurat object, as inferCNV requires raw counts for accurate CNV inference. Additionally, a cell annotation file is created to map each cell ID to its cluster ID, and a gene order file specifying genomic coordinates for each gene is required. The path to the gene order file must be provided.\n\nThe inferCNV analysis is initiated by creating an inferCNV object using the raw count matrix, cell annotation file, and gene order file. The \"Normal\" clusters are specified as reference groups for CNV inference. The inferCNV pipeline is then executed with specific parameters: a cutoff of 0.1 is applied, which is recommended for 10x Genomics data; the output is saved to a designated directory; hierarchical clustering is performed using the \"ward.D2\" method; denoising is applied to improve signal clarity; and HMM-based CNV prediction is disabled for this run.\n\nFollowing the inferCNV analysis, an optional downstream step involves calculating a per-cell CNV score. This is achieved by reading the observation and reference matrices generated by inferCNV, combining and scaling the data, and calculating a squared residual score. The resulting CNV scores are formatted into a data frame and merged back into the Seurat object for further analysis.\n\nTo visualize the results, a boxplot is generated to display the distribution of CNV scores across different clusters. The plot is customized with specific axis labels, limits, and styling to ensure clarity and publication readiness.\n\nFinally, the inferCNV object and the CNV score boxplot are saved to the specified output directory. The inferCNV object is saved as an RDS file, while the boxplot is saved as a PDF. This ensures that all results are preserved for future reference or further analysis. The completion of the inferCNV analysis and CNV scoring is confirmed with a printed message.",
         "The analysis was conducted in R using Seurat for single-cell RNA sequencing data management, infercnv for copy number variation (CNV) inference, and ggpubr for visualization. A pre-processed Seurat object was utilized, with clusters labeled \"0\" and \"18\" designated as \"Normal\" reference cells and remaining cells as \"Tumor.\" A subset of tumor cells was randomly sampled to optimize computational efficiency. The raw count matrix, cell annotation file, and gene order file were prepared for inferCNV analysis. The inferCNV pipeline was executed with a cutoff of 0.1, hierarchical clustering, and denoising, followed by CNV score calculation per cell. Results were visualized using a boxplot and saved for further analysis."
        ],
        [
         "49",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - The purpose of this code block\n   - Packages/tools used in this code block\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Note: Remove non-specific code content such as environment settings, loading and saving of data, etc\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article. Please DO NOT the formatting indicated by `1. 2. 3.` for the steps. Please DO NOT use Heading Marker(`#`) or Emphasis Marker(`**`).\n\n## Code to be re-formatted:\n\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport scanpy as sc\n\nimport infercnvpy as cnv\n\nimport matplotlib.pyplot as plt\n\nimport warnings\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.simplefilter(\"ignore\")\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.settings.set_figure_params(figsize=(5, 5))\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example dataset for inferCNV analysis.\n\n# This dataset from Maynard et al., 2020 is suitable for this purpose.\n\nprint(\"Loading Maynard et al. 2020 dataset...\")\n\nadata = cnv.datasets.maynard2020_3k()\n\n\n\n# Display gene annotations to confirm the required columns (chromosome, start, end) are present.\n\nprint(adata.var.loc[:, [\"ensg\", \"chromosome\", \"start\", \"end\"]].head())\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: CNV Inference and Clustering\n\n#--------------------------------\n\n\n\n# --- 1. Run inferCNV ---\n\n# This is the main computational step. It infers copy number variations by comparing\n\n# gene expression in target cells to a set of reference (\"normal\") cells.\n\nprint(\"\\nRunning infercnvpy analysis...\")\n\ncnv.tl.infercnv(\n\nadata,\n\nreference_key=\"cell_type\",\n\n# Define which cell types are used as the normal reference\n\nreference_cat=[\n\n\"B cell\", \"Macrophage\", \"Mast cell\", \"Monocyte\", \"NK cell\",\n\n\"Plasma cell\", \"T cell CD4\", \"T cell CD8\", \"T cell regulatory\",\n\n\"mDC\", \"pDC\"\n\n],\n\nwindow_size=250 # Size of the sliding window for averaging expression\n\n)\n\n\n\n# --- 2. CNV-based Clustering ---\n\n# Cluster cells based on their inferred CNV profiles to identify groups\n\n# of cells with similar large-scale genomic alterations.\n\nprint(\"\\nPerforming CNV-based clustering...\")\n\n# Perform PCA on the CNV matrix.\n\ncnv.tl.pca(adata)\n\n# Compute a neighborhood graph on the PCA results.\n\ncnv.pp.neighbors(adata)\n\n# Run Leiden clustering to find CNV-based clusters.\n\ncnv.tl.leiden(adata)\n\n\n\n# --- 3. CNV UMAP and Scoring ---\n\n# Compute a UMAP embedding and a per-cell CNV score based on the CNV profiles.\n\nprint(\"Computing CNV UMAP and scores...\")\n\ncnv.tl.umap(adata)\n\ncnv.tl.cnv_score(adata)\n\n\n\n\n\n# --- 4. Tumor vs. Normal Classification ---\n\n# Manually classify cells as \"tumor\" or \"normal\" based on the CNV clusters.\n\n# NOTE: The cluster IDs to be labeled as \"tumor\" need to be determined by inspecting\n\n# the chromosome heatmap and CNV scores from the plots below.\n\nprint(\"Classifying tumor vs. normal cells...\")\n\nadata.obs[\"cnv_status\"] = \"normal\"\n\n# Example: mark cells in specific cnv_leiden clusters as tumor\n\ntumor_clusters = [\"10\", \"16\", \"13\", \"8\", \"12\", \"17\", \"1\", \"14\", \"11\"]\n\nadata.obs.loc[adata.obs[\"cnv_leiden\"].isin(tumor_clusters), \"cnv_status\"] = \"tumor\"\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Initial Data Visualization ---\n\n# Visualize the original UMAP from transcriptomics data to see cell type distribution.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=\"cell_type\", title=\"Transcriptomics UMAP by Cell Type\")\n\n\n\n# --- CNV Heatmaps ---\n\n# Plot a heatmap of inferred CNVs across the genome, grouped by original cell type.\n\n# This shows which cell types exhibit large-scale genomic alterations.\n\ncnv.pl.chromosome_heatmap(adata, groupby=\"cell_type\", title=\"CNV Heatmap by Cell Type\")\n\n\n\n# Plot the CNV heatmap again, but grouped by the new CNV-based Leiden clusters.\n\n# A dendrogram is included to show the hierarchy of clusters.\n\ncnv.pl.chromosome_heatmap(adata, groupby=\"cnv_leiden\", dendrogram=True, title=\"CNV Heatmap by CNV Cluster\")\n\n\n\n# --- CNV UMAPs ---\n\n# Create a multi-panel plot to visualize the CNV-based UMAP embedding.\n\nfig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(11, 11))\n\nax4.axis(\"off\") # Turn off the unused fourth subplot\n\n# Plot CNV UMAP colored by:\n\n# 1. CNV Leiden cluster\n\ncnv.pl.umap(adata, color=\"cnv_leiden\", legend_loc=\"on data\", legend_fontoutline=2, ax=ax1, show=False, title=\"CNV UMAP by CNV Leiden Cluster\")\n\n# 2. CNV score\n\ncnv.pl.umap(adata, color=\"cnv_score\", ax=ax2, show=False, title=\"CNV UMAP by CNV Score\")\n\n# 3. Original cell type annotation\n\ncnv.pl.umap(adata, color=\"cell_type\", ax=ax3, show=False, title=\"CNV UMAP by Original Cell Type\")\n\nplt.tight_layout()\n\nplt.show()\n\n\n\n\n\n# --- Final Classification Visualization ---\n\n# Visualize the final tumor vs. normal classification on both the CNV-based UMAP\n\n# and the original transcriptomics-based UMAP.\n\nfig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))\n\nfig.suptitle(\"Final Tumor/Normal Classification\")\n\n# CNV UMAP\n\ncnv.pl.umap(adata, color=\"cnv_status\", ax=ax1, show=False)\n\n# Transcriptomics UMAP\n\nsc.pl.umap(adata, color=\"cnv_status\", ax=ax2, show=False)\n\nplt.show()\n\n\n\n# Generate separate CNV heatmaps for the classified tumor and normal cells.\n\ncnv.pl.chromosome_heatmap(adata[adata.obs[\"cnv_status\"] == \"tumor\", :], title=\"CNV Heatmap of Tumor Cells\")\n\ncnv.pl.chromosome_heatmap(adata[adata.obs[\"cnv_status\"] == \"normal\", :], title=\"CNV Heatmap of Normal Cells\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final AnnData object, which contains all inferCNVpy analysis results.\n\nprint(\"\\nSaving final AnnData object...\")\n\nadata.write(\"maynard2020_with_infercnv.h5ad\")\n\n\n\nprint(\"\\ninferCNVpy analysis completed!\")\n\n",
         "To perform copy number variation (CNV) inference and clustering on single-cell RNA sequencing data, the following steps were executed using Python and the Scanpy and inferCNVpy libraries. The analysis was conducted on the Maynard et al. (2020) dataset, which includes gene expression data and annotations for chromosome, start, and end positions.\n\nFirst, the necessary Python libraries were imported, including Scanpy for single-cell data analysis, inferCNVpy for CNV inference, Matplotlib for visualization, and warnings for managing runtime alerts. Warnings were suppressed to ensure cleaner output during execution, and global plotting parameters were set to maintain consistent and clear visualizations.\n\nThe dataset was loaded from the inferCNVpy library, which provides an example dataset suitable for CNV analysis. Gene annotations were inspected to confirm the presence of required columns (chromosome, start, and end positions). This step ensures the data is appropriately structured for downstream analysis.\n\nThe primary analysis began with CNV inference, where copy number variations were inferred by comparing gene expression in target cells to a set of reference (\"normal\") cells. Reference cell types included B cells, macrophages, mast cells, monocytes, NK cells, plasma cells, T cells (CD4, CD8, and regulatory), mDCs, and pDCs. A sliding window of size 250 was used to average gene expression across the genome. This step is critical for identifying large-scale genomic alterations.\n\nNext, CNV-based clustering was performed to group cells with similar CNV profiles. Principal component analysis (PCA) was applied to the CNV matrix, followed by the construction of a neighborhood graph and Leiden clustering. This clustering approach identifies distinct groups of cells based on their CNV profiles, aiding in the identification of tumor and normal cells.\n\nA UMAP embedding was computed to visualize the CNV profiles in two dimensions, and a per-cell CNV score was calculated to quantify the extent of genomic alterations. Cells were then classified as \"tumor\" or \"normal\" based on their CNV Leiden clusters. Specific clusters were manually annotated as tumor based on inspection of the chromosome heatmap and CNV scores.\n\nVisualizations were generated to interpret the results. The original UMAP embedding, colored by cell type, was plotted to visualize the distribution of cell types. Heatmaps of inferred CNVs were created, grouped by both original cell type and CNV-based Leiden clusters, to identify cell types and clusters with significant genomic alterations. A multi-panel UMAP plot was generated to visualize CNV-based UMAP embeddings colored by CNV Leiden clusters, CNV scores, and original cell type annotations. Finally, the tumor vs. normal classification was visualized on both the CNV-based and transcriptomics-based UMAP embeddings, and separate CNV heatmaps were generated for tumor and normal cells.\n\nThe final AnnData object, containing all analysis results, was saved to a file for future reference. This comprehensive workflow enables the identification and visualization of CNV-driven cell clusters in single-cell RNA sequencing data, facilitating the classification of tumor and normal cells.",
         "\nThis prompt guides you to transform detailed bioinformatics analysis instructions into a high-level description, similar to a scientific article's results section, within 100 words. It focuses on the core analysis steps, tools used, and workflow, while omitting specific parameters and data input/output operations.\n\n - The description should be a concise, high-level overview of the analysis workflow, mentioning:\n - The programming language and main libraries or tools used.\n - Key steps such as quality control, data processing, computation, and visualization.\n - Exclude standard data input (e.g., loading files) and output (e.g., saving results) operations, unless they involve specific processing integral to the analysis.\n - Do not include specific parameters, code snippets, or detailed step-by-step explanations.\n - The tone should be professional, suitable for a scientific article, and provide an overview of the process.\n - Align the provided guideline as much as possible, with the main alignment goal being the name of the specific task being analyzed.\n - Do not start with `we performed`.\n\n\n## Example\n\nFor an instruction on single-cell RNA-seq quality control using Python and Scanpy, the description might be: \"We performed the analysis in Python using Scanpy, involving gene annotation based on naming rules, calculation of quality control metrics like gene counts and mitochondrial percentages, and visualization through violin and scatter plots to identify outliers.\"\n\n## Workflow Guideline to be aligned:\n# Single-Cell RNA Sequencing Data Analysis Workflow Guide\n\nThis guide outlines a typical workflow for analyzing single-cell RNA sequencing (scRNA-seq) data, from raw data to downstream interpretation, aimed at extracting biological insights from tens of thousands to hundreds of thousands of cells. Many steps are iterative, especially when exploring cellular heterogeneity at different resolutions.\n\n## 1. Raw Data Processing\n\n### Purpose\nConvert raw sequencing machine outputs (typically FASTQ files) into a molecule count matrix of genes per cell.\n\n### Description\nThis is the starting point of the analysis. This step involves aligning/mapping sequencing reads to a reference genome or transcriptome, identifying and correcting cell barcodes (CB), and estimating molecule counts using unique molecular identifiers (UMIs). These steps address errors and biases that may occur during sequencing, such as index hopping and sequencing errors. Common tools include Cell Ranger (commercial software), zUMIs, alevin, kallisto|bustools, STARsolo, and alevin-fry. Some tools generate traditional alignment files (e.g., BAM), while others operate in memory or use more compact representations. Cell barcode correction is typically based on known whitelists or \"knee\"/\"elbow\" identification based on frequency distributions. UMI deduplication aims to distinguish PCR duplicates from original mRNA molecules, handling sequencing errors and multimapping reads.\n\n### Personalizable Parts\n- Choice of specific processing tool (e.g., alevin-fry is considered fast, accurate, and memory-efficient).\n- Choice of reference genome/transcriptome and its version (recommend using splici reference, which includes spliced and unspliced sequences).\n- Choice of barcode correction strategy (based on known lists, knee/elbow methods, expected cell numbers, or forced cell numbers).\n- Whether to quantify unspliced and ambiguous read information (e.g., USA mode in alevin-fry), which is important for downstream analyses like RNA velocity.\n- Strategies for handling doublets or empty droplets.\n\n### Input\n- Raw sequencing FASTQ files (typically lane-demultiplexed).\n- Reference genome FASTA file and gene annotation GTF file.\n\n### Output\n- Cell-gene count matrix (or read matrix, if UMIs are not used).\n- Possibly, count matrices for spliced and unspliced reads.\n- Quality control metrics report for the raw data processing stage (e.g., mapping rate, UMI duplication rate, distribution of detected genes).\n\n## 2. Quality Control (Count Matrix)\n\n### Purpose\nIdentify and remove low-quality cells, ambient RNA contamination, and doublets to obtain a high-quality dataset where each observation represents a complete single cell.\n\n### Description\nAfter obtaining the count matrix, quality control is necessary. This includes assessing metrics such as sequencing depth (total counts), number of detected genes, and the percentage of mitochondrial gene counts. Low-quality cells (possibly due to membrane rupture or cell death) typically exhibit low total counts, low gene counts, and high mitochondrial count percentages. Additionally, ambient RNA contamination (extracellular mRNA mistakenly included in droplets) must be corrected using tools like SoupX or DecontX. Doublets (droplets containing multiple cells) must also be detected and handled, as they can distort downstream analyses. Common doublet detection tools include Scrublet, scDblFinder, DoubletDetection, and SOLO.\n\n### Personalizable Parts\n- Threshold selection for filtering low-quality cells: can be determined manually based on QC metric distributions or automatically using methods like median absolute deviation (MAD) filtering. Prefer more lenient thresholds to avoid losing rare cell populations.\n- Choice of ambient RNA correction tool.\n- Choice of doublet detection tool and whether to remove or only flag doublets.\n- For multi-batch datasets, perform QC separately for each batch, as quality metric thresholds may differ.\n- QC strategies can be reassessed after clustering or annotation.\n\n### Input\n- Cell-gene count matrix from raw data processing.\n\n### Output\n- Filtered cell-gene count matrix (containing only high-quality cells).\n- Updated cell metadata, including QC metrics, ambient RNA correction information, and doublet scores/classifications.\n\n## 3. Normalization\n\n### Purpose\nAdjust raw counts to correct for technical variations (e.g., library size, cell size), stabilize variance, and make gene expression comparable across cells.\n\n### Description\nCount differences in single-cell data may arise from sampling effects. Normalization aims to remove technical noise while preserving biological heterogeneity. Different normalization techniques have distinct goals and downstream applicability. Common methods include:\n- Shifted logarithm: Scaling by library size (e.g., total counts or median counts) followed by adding a pseudo-count (usually 1) and log-transformation. This is a standard step in many workflows.\n- scran normalization: Computes size factors based on cell clustering, then scales counts. Commonly used in the Bioconductor ecosystem.\n- Analytic Pearson residuals: Fits a regularized negative binomial model to the relationship between count depth and variance, then computes Pearson residuals to better remove technical effects while preserving biological signals.\n- SCTransform: An alternative in the Seurat ecosystem that models and stabilizes variance directly on raw UMI counts, replacing traditional normalization, scaling, and variable feature selection steps.\n\n### Personalizable Parts\n- Choice of normalization method, as different methods suit different downstream tasks (e.g., Pearson residuals for feature selection).\n- For scaling-based methods, choice of size factor calculation (e.g., total counts, median counts, CP10k, CPM).\n- For methods like SCTransform, option to regress out covariates (e.g., mitochondrial count percentage).\n- Multiple methods can be tested, with results stored as different layers in AnnData objects or different Assay/Layer in Seurat objects for comparison.\n\n### Input\n- Filtered cell-gene count matrix from quality control.\n\n### Output\n- Normalized gene expression matrix.\n- Raw counts are typically retained in a separate layer.\n\n## 4. Feature Selection\n\n### Purpose\nSelect a subset of genes with the highest information content or variability from normalized data to reduce dimensionality and noise, focusing downstream analyses.\n\n### Description\nMost single-cell datasets contain many genes with low expression or minimal variation across cells, which contribute little to distinguishing cell types or states and may introduce noise. Feature selection identifies genes with significant variation, often called highly variable genes (HVGs). Different algorithms (e.g., Seurat, Cell Ranger, Seurat v3 implementations) calculate mean expression and variance to determine HVGs.\n\n### Personalizable Parts\n- Choice of HVG selection algorithm or \"flavor.\"\n- Number of genes to retain (typically hundreds to thousands, e.g., 500–2000, depending on dataset size and complexity).\n- Consideration of normalization method’s impact on HVG selection.\n- Methods like SCTransform integrate normalization and HVG selection.\n\n### Input\n- Normalized gene expression matrix.\n\n### Output\n- List of genes selected as features or markers in the AnnData object’s variables (var).\n- Subsequent analyses typically use a matrix containing only these selected features.\n\n## 5. Dimensionality Reduction & Graph Construction\n\n### Purpose\nProject high-dimensional gene expression data into a lower-dimensional space for visualization, denoising, and as input for tasks like clustering, while constructing a cell neighborhood graph.\n\n### Description\nSingle-cell data suffers from the \"curse of dimensionality,\" where noise and redundancy in high dimensions can obscure true signals. Dimensionality reduction creates a new set of informative variables:\n- PCA (Principal Component Analysis): A linear method identifying principal components that explain the most variance, used for denoising and initial low-dimensional representation.\n- t-SNE (t-Distributed Stochastic Neighbor Embedding) / UMAP (Uniform Manifold Approximation and Projection): Nonlinear methods preserving cell similarities (local and/or global topology) for visualization in 2D or 3D.\n- Nearest Neighbor Graph (KNN Graph): Computes distances between cells in the reduced space, constructing a graph where similar cells are connected by edges, serving as the basis for subsequent analyses like clustering or trajectory inference.\n\n### Personalizable Parts\n- Choice of dimensionality reduction method (PCA is often the first step, followed by UMAP or t-SNE for visualization; LSI for ATAC data).\n- Number of principal components to retain (e.g., determined by elbow plots, explained variance, or biological signal assessment).\n- UMAP/t-SNE parameter settings (e.g., perplexity, spread).\n- Number of neighbors (k) in the KNN graph, affecting local structure.\n- Visualization can assess prior QC results (e.g., viewing QC metrics on UMAP plots).\n\n### Input\n- Gene expression matrix from feature selection (or normalized matrix if feature selection is skipped).\n\n### Output\n- Low-dimensional embedding coordinates (e.g., X_pca, X_umap in AnnData’s obsm).\n- Nearest neighbor graph (e.g., distances, connectivities in AnnData’s obsp).\n\n## 6. Data Integration (Optional)\n\n### Purpose\nRemove technical differences (batch effects) when analyzing datasets from different batches, donors, conditions, or technologies, enabling comparison and integration of biological signals.\n\n### Description\nBatch effects are a major challenge in scRNA-seq analysis. Integration methods align cells in similar biological states across batches. Integration can occur at various stages, such as after normalization, dimensionality reduction, or as part of clustering. Methods include:\n- Linear embedding methods: Mutual Nearest Neighbors (MNN), Seurat’s CCA integration, Harmony, finding similar cells across batches in low-dimensional space.\n- Graph-based methods: BBKNN (Batch Balanced KNN), modifying KNN graph construction to enforce cross-batch connections.\n- Deep learning methods: scVI, scANVI, scGen, using variational autoencoders (VAEs) to learn batch-corrected low-dimensional representations; scANVI leverages cell labels to preserve biological differences.\n- Reference mapping methods: ingest, scArches, Symphony, Seurat’s reference mapping, training models on annotated reference datasets and mapping new (query) data to the reference space.\n- Integration quality assessment: Tools like scIB provide metrics to evaluate batch effect removal and biological variation retention.\n\n### Personalizable Parts\n- Choice of integration method, depending on batch effect complexity, availability of cell labels, desired output format (corrected embeddings or expression matrices), and computational resources. Benchmarks suggest Harmony and Seurat for simple tasks, deep learning and Scanorama for complex tasks.\n- Choice of batch covariates (e.g., sample, donor, dataset), depending on analysis goals; finer batch resolution removes more effects but may confound biological signals.\n- Method-specific parameter tuning.\n- For labeled data, label harmonization may be needed first.\n\n### Input\n- Gene expression data after normalization or feature selection.\n- Preliminary dimensionality reduction and graph construction results, if applicable.\n- Cell metadata column indicating batch information.\n- Cell labels for label-aware methods or reference mapping.\n\n### Output\n- Integrated data representation (e.g., batch-corrected low-dimensional embeddings or expression matrices).\n- Integrated nearest neighbor graph.\n\n## 7. Clustering\n\n### Purpose\nGroup cells with similar gene expression patterns into clusters, representing potential cell types or states.\n\n### Description\nClustering is a common unsupervised machine learning task. In single-cell analysis, a nearest neighbor graph is built in the reduced space, and graph-based clustering algorithms (e.g., Leiden, Louvain) identify communities by optimizing modularity. Clustering forms the basis for subsequent cell type annotation.\n\n### Personalizable Parts\n- Choice of clustering algorithm (Leiden is recommended for well-connected communities).\n- Resolution parameter, controlling cluster granularity (higher values yield more, smaller clusters; optimal resolution depends on dataset size).\n- Input for clustering: unintegrated or integrated dimensionality reduction embeddings and graphs.\n- Sub-clustering to further subdivide cell populations of interest.\n\n### Input\n- Nearest neighbor graph from dimensionality reduction or data integration.\n- Corresponding low-dimensional embeddings (for visualizing clustering results).\n\n### Output\n- Cluster labels for each cell (typically stored in AnnData’s obs).\n\n## 8. Annotation\n\n### Purpose\nTo assign biological meaning to the identified cell clusters, labeling them as specific cell types or states. For identification of malignant cells in cancer samples, features of Copy Number Variation (CNV) can be leveraged.\n\n### Description\nCell type definition is based on specific marker genes, typically robust across datasets. Annotation can be:\n- Manual annotation: Based on known marker genes, visualizing their expression across clusters to identify cell types. This is intuitive but may be subjective without unique markers or in complex datasets.\n- Automated annotation: Using pre-trained classifiers to predict cell types based on gene expression patterns, leveraging machine learning for speed and existing high-quality annotations.\n- Reference mapping: Mapping the dataset to a well-annotated reference atlas and transferring labels to query cells, dependent on reference quality and dataset similarity. Cell types can be further subdivided into subtypes or states, with some researchers using \"cell identity\" to avoid this distinction.\n- Inference of Copy Number Variation (CNV): For tumor samples, malignant cells often exhibit significant CNVs. Tools like infercnv and copycat can infer CNVs from single-cell RNA sequencing data, thereby aiding in the identification of malignant cell populations. These tools work by analyzing the average expression levels of genes across chromosomes and comparing them to normal cells to identify regions of copy number gain or loss.\n\n### Personalizable Parts\n- Choice of annotation strategy: manual, automated classifier, or reference mapping.\n- For manual annotation, selection of marker gene sets, validated with biological expertise and specific to the tissue and experiment.\n- For automated or reference mapping, selection of appropriate pre-trained models or reference atlases, ensuring technical and biological similarity to query data.\n- When using infercnv or copycat for CNV inference, a set of known normal cells must be provided as a reference. The selection of these reference cells is crucial and can be based on annotated non-malignant cell types or single-cell data from healthy individuals.\n- Annotation granularity (e.g., broad cell types vs. finer subtypes/states).\n\n### Input\n- Cell cluster labels from clustering.\n- Normalized gene expression matrix.\n- Low-dimensional embeddings from dimensionality reduction or data integration (for visualizing marker gene expression).\n- Pre-trained classifier models or annotated reference datasets for automated or reference mapping methods.\n- For infercnv/copycat: Requires the specification of a normal cell population (based on initial annotation or other prior knowledge).\n\n### Output\n- Biological labels for each cell or cluster (e.g., stored in AnnData’s obs).\n\n## 9. Iterative Subpopulation Definition and Analysis Loop\n\n### Purpose\nIdentify and analyze finer subpopulations or cell states at higher resolution.\n\n### Description\nAfter identifying broad cell types through initial clustering and annotation (step 8), deeper analysis of specific cell populations is often needed to reveal finer subpopulation structures or cell states. This is an iterative process that may be repeated multiple times.\n\n#### Process\n- **Select Subpopulation for In-depth Analysis**: Based on initial annotations (step 8) and your biological research questions, select specific cell types or clusters for detailed analysis. The choice should align with the biological story you aim to explore or validate, ensuring analysis relevance. Consider which downstream analyses (e.g., trajectory inference for cell state changes, cell-cell communication, or differential expression under specific conditions) are relevant and valuable for the selected subpopulations and research questions, guiding subsequent steps.\n- **Subset-specific Feature Selection**: Re-identify highly variable genes or informative features within the selected cell subset, as HVGs in the full dataset may differ from those in specific subpopulations, capturing subset-specific variation.\n- **Subset-specific Dimensionality Reduction & Graph Construction**: Using the subset data (based on re-selected features), recompute dimensionality reduction embeddings (e.g., PCA, UMAP/t-SNE) and build a new nearest neighbor graph to better visualize and analyze subpopulation structure.\n- **Subset-specific Clustering**: Apply clustering algorithms (e.g., Leiden) on the new nearest neighbor graph to identify finer groupings within the subpopulation.\n- **Subset-specific Annotation**: Assign biological labels to new subpopulations (sub-clusters), typically by identifying subpopulation-specific marker genes, possibly requiring differential gene expression analysis to find significantly differentially expressed genes.\n- **Perform Subpopulation-specific Downstream Analysis**: Once subpopulations are defined at the desired resolution, perform analyses such as differential gene expression between subpopulations or conditions, trajectory inference (pseudotime, RNA velocity) for continuous processes, cell-cell communication, gene set enrichment and pathway analysis, gene regulatory network inference, perturbation response analysis, or relative abundance changes across conditions.\n\n#### Optional: Re-assess QC\nBased on clustering or annotation results (overall or subpopulation-specific), revisit initial QC filtering parameters if necessary (e.g., if clusters show signs of doublets or low-quality cells). This may require rerunning the workflow from step 2.\n\n## 10. Further Downstream Analysis (Advanced)\n\n### Purpose\nAfter preprocessing and core analysis, use processed and annotated single-cell data to explore specific biological questions in depth.\n\n### Description\nThe core analysis pipeline provides a foundational dataset for advanced analyses, tailored to experimental design and research questions. Covered analyses include:\n- **Differential Gene Expression (DGE) Analysis**: Compare gene expression differences between cell populations (e.g., different cell types or conditions). Methods include pseudobulk approaches (edgeR, DESeq2) and single-cell-specific methods (MAST, glmmTMB).\n- **Compositional Analysis**: Analyze changes in cell type proportions across conditions, using methods like scCODA, tascCODA, or KNN-based approaches (DA-seq, Milo, MELD).\n- **Gene Set Enrichment and Pathway Analysis**: Identify enriched gene sets (e.g., biological processes, signaling pathways) in specific cell populations or conditions, using gene set testing (GSEA, fry, camera) or pathway activity inference (AUCell, Pagoda2, PROGENy, DoRothEA).\n- **Pseudotemporal Ordering / Trajectory Inference**: Reconstruct dynamic cell trajectories during development or differentiation based on gene expression similarity, using methods like DPT, Palantir, Slingshot, and PAGA.\n- **RNA Velocity**: Infer instantaneous rates and directions of gene expression changes using spliced and unspliced mRNA abundances to predict short-term cell state transitions, analyzed with tools like scVelo.\n- **Cell-cell Communication (CCC)**: Infer interactions between cell types based on gene expression, typically via ligand-receptor interactions, using tools like CellPhoneDB, LIANA, and NicheNet (which also considers downstream signaling effects).\n- **Gene Regulatory Networks (GRN) Inference**: Model how transcription factors (TFs) regulate target gene expression, using tools like SCENIC to infer regulons via TF-target gene co-expression.\n- **Perturbation Modeling**: Analyze or predict cell responses to experimental perturbations (e.g., gene knockouts, drug treatments), using methods like Augur (identifying most affected cell types) and scGen (predicting perturbation responses).\n- **Lineage Tracing**: Reconstruct ancestral-descendant relationships and proliferation history using molecular markers (e.g., CRISPR/Cas9 edits), with tools like Cassiopeia and CoSpar. Note: This often requires specialized experimental design and raw data processing, differing from standard scRNA-seq workflows.\n\n### Personalizable Parts\n- Choice of downstream analyses depends on specific biological questions and experimental design.\n- Each analysis domain offers multiple methods, selected based on data type, question nature, tool characteristics (e.g., speed, accuracy, prior knowledge requirements, multimodal data support).\n- Many methods have specific input requirements (e.g., DGE needs normalized counts and condition labels, RNA velocity needs unspliced counts, NicheNet needs gene sets of interest).\n- Interpretation of results requires integration with biological context.\n\n### Input\n- Processed, integrated (if needed), clustered, and annotated single-cell data from annotation.\n- Depending on analysis, may require raw counts, normalized counts, low-dimensional embeddings, graph structures, cell labels, marker genes, experimental conditions, time points, or perturbation information.\n- Some analyses (e.g., lineage tracing) require specially processed raw data.\n- External databases or knowledge bases (e.g., pathway, ligand-receptor, TF-target gene databases).\n\n### Output\n- Analysis-specific results (e.g., differentially expressed gene lists, pathway enrichment scores, cell type proportion changes, pseudotime values, velocity vectors, predicted cell interactions, inferred regulatory networks).\n\n## Instruction to be transformed:\nTo perform copy number variation (CNV) inference and clustering on single-cell RNA sequencing data, the following steps were executed using Python and the Scanpy and inferCNVpy libraries. The analysis was conducted on the Maynard et al. (2020) dataset, which includes gene expression data and annotations for chromosome, start, and end positions.\n\nFirst, the necessary Python libraries were imported, including Scanpy for single-cell data analysis, inferCNVpy for CNV inference, Matplotlib for visualization, and warnings for managing runtime alerts. Warnings were suppressed to ensure cleaner output during execution, and global plotting parameters were set to maintain consistent and clear visualizations.\n\nThe dataset was loaded from the inferCNVpy library, which provides an example dataset suitable for CNV analysis. Gene annotations were inspected to confirm the presence of required columns (chromosome, start, and end positions). This step ensures the data is appropriately structured for downstream analysis.\n\nThe primary analysis began with CNV inference, where copy number variations were inferred by comparing gene expression in target cells to a set of reference (\"normal\") cells. Reference cell types included B cells, macrophages, mast cells, monocytes, NK cells, plasma cells, T cells (CD4, CD8, and regulatory), mDCs, and pDCs. A sliding window of size 250 was used to average gene expression across the genome. This step is critical for identifying large-scale genomic alterations.\n\nNext, CNV-based clustering was performed to group cells with similar CNV profiles. Principal component analysis (PCA) was applied to the CNV matrix, followed by the construction of a neighborhood graph and Leiden clustering. This clustering approach identifies distinct groups of cells based on their CNV profiles, aiding in the identification of tumor and normal cells.\n\nA UMAP embedding was computed to visualize the CNV profiles in two dimensions, and a per-cell CNV score was calculated to quantify the extent of genomic alterations. Cells were then classified as \"tumor\" or \"normal\" based on their CNV Leiden clusters. Specific clusters were manually annotated as tumor based on inspection of the chromosome heatmap and CNV scores.\n\nVisualizations were generated to interpret the results. The original UMAP embedding, colored by cell type, was plotted to visualize the distribution of cell types. Heatmaps of inferred CNVs were created, grouped by both original cell type and CNV-based Leiden clusters, to identify cell types and clusters with significant genomic alterations. A multi-panel UMAP plot was generated to visualize CNV-based UMAP embeddings colored by CNV Leiden clusters, CNV scores, and original cell type annotations. Finally, the tumor vs. normal classification was visualized on both the CNV-based and transcriptomics-based UMAP embeddings, and separate CNV heatmaps were generated for tumor and normal cells.\n\nThe final AnnData object, containing all analysis results, was saved to a file for future reference. This comprehensive workflow enables the identification and visualization of CNV-driven cell clusters in single-cell RNA sequencing data, facilitating the classification of tumor and normal cells.",
         "The analysis of copy number variation (CNV) in single-cell RNA sequencing data was conducted using Python with the Scanpy and inferCNVpy libraries. The workflow involved inferring CNVs by comparing gene expression in target cells to a reference set of normal cells, employing a sliding window approach to average expression across the genome. CNV-based clustering was performed using PCA, neighborhood graph construction, and Leiden clustering to identify distinct cell groups. Visualization included UMAP embeddings to project CNV profiles in two dimensions and heatmaps to highlight genomic alterations. Cells were classified as tumor or normal based on CNV scores and cluster inspection, enabling the identification of CNV-driven cell populations."
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 54
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Instr_input</th>\n",
       "      <th>Instructions</th>\n",
       "      <th>Dscpt_input</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing the necessary...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by loading the necessary P...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following section provides a detailed desc...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of single-cell RNA sequencing (sc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by loading essential R pac...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in R using the Seur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following is a step-by-step guide to the a...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential mod...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted using Python with t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by loading the necessary l...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in R using the Seur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing the necessary...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted using Python with t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform single-cell RNA sequencing analysis...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell RNA sequencing analysis was conduc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To analyze single-cell RNA sequencing data, th...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell RNA sequencing data analysis was c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform single-cell RNA-seq analysis, the f...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell RNA sequencing data were analyzed ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To analyze single-cell RNA sequencing data usi...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell RNA sequencing data analysis was c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing the necessary...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform single-cell RNA sequencing analysis...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell RNA sequencing data were analyzed ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis was performed using Python, lever...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis pipeline begins with loading esse...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in R using the Seur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform automatic cell-type annotation on a...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing necessary Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following section describes the computatio...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The integration and annotation of single-cell ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential lib...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted using Python, lever...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing necessary Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform semi-supervised integration of sing...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis employed Python with the Scanpy a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To analyze single-cell RNA sequencing data and...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell RNA sequencing data analysis was c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following section describes the step-by-st...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing necessary Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins with the loading of necess...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The single-cell RNA sequencing analysis was co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following section describes the computatio...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of single-cell RNA sequencing dat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis was performed using Python, lever...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform trajectory inference on single-cell...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Trajectory inference on single-cell RNA sequen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform single-cell trajectory analysis usi...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell trajectory analysis was conducted ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform single-cell trajectory analysis usi...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Single-cell trajectory analysis was conducted ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using `pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using lib...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following is a step-by-step guide for perf...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Gene set enrichment and pathway analysis of si...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using key...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform ligand-receptor interaction analysi...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of ligand-receptor interactions i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using lib...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following section describes the computatio...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of single-cell RNA sequencing dat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following method describes a single-cell R...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>Pathway enrichment analysis was conducted on t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis utilized Python libraries includi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis pipeline begins by importing esse...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using Sca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing necessary Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in Python using lib...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform inferCNV analysis on single-cell RN...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of copy number variations (CNVs) ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis pipeline begins with the loading ...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis was conducted in R using Seurat f...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>To perform copy number variation (CNV) inferen...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of copy number variation (CNV) in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following is a step-by-step guide describi...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The analysis of single-cell RNA-seq data was c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The analysis begins by importing essential Pyt...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>The RNA velocity analysis was conducted using ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following section describes the methodolog...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>RNA velocity analysis was conducted using the ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>The following is a step-by-step guide for perf...</td>\n",
       "      <td>\\nThis prompt guides you to transform detailed...</td>\n",
       "      <td>RNA velocity analysis was conducted using the ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Instr_input  \\\n",
       "0   \"\\nAct as a scientific writer who can translat...   \n",
       "1   \"\\nAct as a scientific writer who can translat...   \n",
       "2   \"\\nAct as a scientific writer who can translat...   \n",
       "3   \"\\nAct as a scientific writer who can translat...   \n",
       "4   \"\\nAct as a scientific writer who can translat...   \n",
       "5   \"\\nAct as a scientific writer who can translat...   \n",
       "6   \"\\nAct as a scientific writer who can translat...   \n",
       "7   \"\\nAct as a scientific writer who can translat...   \n",
       "8   \"\\nAct as a scientific writer who can translat...   \n",
       "9   \"\\nAct as a scientific writer who can translat...   \n",
       "10  \"\\nAct as a scientific writer who can translat...   \n",
       "11  \"\\nAct as a scientific writer who can translat...   \n",
       "12  \"\\nAct as a scientific writer who can translat...   \n",
       "13  \"\\nAct as a scientific writer who can translat...   \n",
       "14  \"\\nAct as a scientific writer who can translat...   \n",
       "15  \"\\nAct as a scientific writer who can translat...   \n",
       "16  \"\\nAct as a scientific writer who can translat...   \n",
       "17  \"\\nAct as a scientific writer who can translat...   \n",
       "18  \"\\nAct as a scientific writer who can translat...   \n",
       "19  \"\\nAct as a scientific writer who can translat...   \n",
       "20  \"\\nAct as a scientific writer who can translat...   \n",
       "21  \"\\nAct as a scientific writer who can translat...   \n",
       "22  \"\\nAct as a scientific writer who can translat...   \n",
       "23  \"\\nAct as a scientific writer who can translat...   \n",
       "24  \"\\nAct as a scientific writer who can translat...   \n",
       "25  \"\\nAct as a scientific writer who can translat...   \n",
       "26  \"\\nAct as a scientific writer who can translat...   \n",
       "27  \"\\nAct as a scientific writer who can translat...   \n",
       "28  \"\\nAct as a scientific writer who can translat...   \n",
       "29  \"\\nAct as a scientific writer who can translat...   \n",
       "30  \"\\nAct as a scientific writer who can translat...   \n",
       "31  \"\\nAct as a scientific writer who can translat...   \n",
       "32  \"\\nAct as a scientific writer who can translat...   \n",
       "33  \"\\nAct as a scientific writer who can translat...   \n",
       "34  \"\\nAct as a scientific writer who can translat...   \n",
       "35  \"\\nAct as a scientific writer who can translat...   \n",
       "36  \"\\nAct as a scientific writer who can translat...   \n",
       "37  \"\\nAct as a scientific writer who can translat...   \n",
       "38  \"\\nAct as a scientific writer who can translat...   \n",
       "39  \"\\nAct as a scientific writer who can translat...   \n",
       "40  \"\\nAct as a scientific writer who can translat...   \n",
       "41  \"\\nAct as a scientific writer who can translat...   \n",
       "42  \"\\nAct as a scientific writer who can translat...   \n",
       "43  \"\\nAct as a scientific writer who can translat...   \n",
       "44  \"\\nAct as a scientific writer who can translat...   \n",
       "45  \"\\nAct as a scientific writer who can translat...   \n",
       "46  \"\\nAct as a scientific writer who can translat...   \n",
       "47  \"\\nAct as a scientific writer who can translat...   \n",
       "48  \"\\nAct as a scientific writer who can translat...   \n",
       "49  \"\\nAct as a scientific writer who can translat...   \n",
       "50  \"\\nAct as a scientific writer who can translat...   \n",
       "51  \"\\nAct as a scientific writer who can translat...   \n",
       "52  \"\\nAct as a scientific writer who can translat...   \n",
       "53  \"\\nAct as a scientific writer who can translat...   \n",
       "\n",
       "                                         Instructions  \\\n",
       "0   The analysis begins by importing the necessary...   \n",
       "1   The analysis begins by loading the necessary P...   \n",
       "2   The following section provides a detailed desc...   \n",
       "3   The analysis begins by importing essential Pyt...   \n",
       "4   The analysis begins by loading essential R pac...   \n",
       "5   The following is a step-by-step guide to the a...   \n",
       "6   The analysis begins by importing essential mod...   \n",
       "7   The analysis begins by loading the necessary l...   \n",
       "8   The analysis begins by importing the necessary...   \n",
       "9   To perform single-cell RNA sequencing analysis...   \n",
       "10  To analyze single-cell RNA sequencing data, th...   \n",
       "11  The analysis begins by importing essential Pyt...   \n",
       "12  To perform single-cell RNA-seq analysis, the f...   \n",
       "13  To analyze single-cell RNA sequencing data usi...   \n",
       "14  The analysis begins by importing the necessary...   \n",
       "15  To perform single-cell RNA sequencing analysis...   \n",
       "16  The analysis was performed using Python, lever...   \n",
       "17  The analysis begins by importing essential Pyt...   \n",
       "18  The analysis pipeline begins with loading esse...   \n",
       "19  To perform automatic cell-type annotation on a...   \n",
       "20  The analysis begins by importing necessary Pyt...   \n",
       "21  The analysis begins by importing essential Pyt...   \n",
       "22  The following section describes the computatio...   \n",
       "23  The analysis begins by importing essential lib...   \n",
       "24  The analysis begins by importing necessary Pyt...   \n",
       "25  To perform semi-supervised integration of sing...   \n",
       "26  To analyze single-cell RNA sequencing data and...   \n",
       "27  The following section describes the step-by-st...   \n",
       "28  The analysis begins by importing necessary Pyt...   \n",
       "29  The analysis begins with the loading of necess...   \n",
       "30  The analysis begins by importing essential Pyt...   \n",
       "31  The following section describes the computatio...   \n",
       "32  The analysis was performed using Python, lever...   \n",
       "33  To perform trajectory inference on single-cell...   \n",
       "34  To perform single-cell trajectory analysis usi...   \n",
       "35  To perform single-cell trajectory analysis usi...   \n",
       "36  The analysis begins by importing essential Pyt...   \n",
       "37  The analysis begins by importing essential Pyt...   \n",
       "38  The following is a step-by-step guide for perf...   \n",
       "39  The analysis begins by importing essential Pyt...   \n",
       "40  To perform ligand-receptor interaction analysi...   \n",
       "41  The analysis begins by importing essential Pyt...   \n",
       "42  The following section describes the computatio...   \n",
       "43  The following method describes a single-cell R...   \n",
       "44  The analysis begins by importing essential Pyt...   \n",
       "45  The analysis pipeline begins by importing esse...   \n",
       "46  The analysis begins by importing necessary Pyt...   \n",
       "47  To perform inferCNV analysis on single-cell RN...   \n",
       "48  The analysis pipeline begins with the loading ...   \n",
       "49  To perform copy number variation (CNV) inferen...   \n",
       "50  The following is a step-by-step guide describi...   \n",
       "51  The analysis begins by importing essential Pyt...   \n",
       "52  The following section describes the methodolog...   \n",
       "53  The following is a step-by-step guide for perf...   \n",
       "\n",
       "                                          Dscpt_input  \\\n",
       "0   \\nThis prompt guides you to transform detailed...   \n",
       "1   \\nThis prompt guides you to transform detailed...   \n",
       "2   \\nThis prompt guides you to transform detailed...   \n",
       "3   \\nThis prompt guides you to transform detailed...   \n",
       "4   \\nThis prompt guides you to transform detailed...   \n",
       "5   \\nThis prompt guides you to transform detailed...   \n",
       "6   \\nThis prompt guides you to transform detailed...   \n",
       "7   \\nThis prompt guides you to transform detailed...   \n",
       "8   \\nThis prompt guides you to transform detailed...   \n",
       "9   \\nThis prompt guides you to transform detailed...   \n",
       "10  \\nThis prompt guides you to transform detailed...   \n",
       "11  \\nThis prompt guides you to transform detailed...   \n",
       "12  \\nThis prompt guides you to transform detailed...   \n",
       "13  \\nThis prompt guides you to transform detailed...   \n",
       "14  \\nThis prompt guides you to transform detailed...   \n",
       "15  \\nThis prompt guides you to transform detailed...   \n",
       "16  \\nThis prompt guides you to transform detailed...   \n",
       "17  \\nThis prompt guides you to transform detailed...   \n",
       "18  \\nThis prompt guides you to transform detailed...   \n",
       "19  \\nThis prompt guides you to transform detailed...   \n",
       "20  \\nThis prompt guides you to transform detailed...   \n",
       "21  \\nThis prompt guides you to transform detailed...   \n",
       "22  \\nThis prompt guides you to transform detailed...   \n",
       "23  \\nThis prompt guides you to transform detailed...   \n",
       "24  \\nThis prompt guides you to transform detailed...   \n",
       "25  \\nThis prompt guides you to transform detailed...   \n",
       "26  \\nThis prompt guides you to transform detailed...   \n",
       "27  \\nThis prompt guides you to transform detailed...   \n",
       "28  \\nThis prompt guides you to transform detailed...   \n",
       "29  \\nThis prompt guides you to transform detailed...   \n",
       "30  \\nThis prompt guides you to transform detailed...   \n",
       "31  \\nThis prompt guides you to transform detailed...   \n",
       "32  \\nThis prompt guides you to transform detailed...   \n",
       "33  \\nThis prompt guides you to transform detailed...   \n",
       "34  \\nThis prompt guides you to transform detailed...   \n",
       "35  \\nThis prompt guides you to transform detailed...   \n",
       "36  \\nThis prompt guides you to transform detailed...   \n",
       "37  \\nThis prompt guides you to transform detailed...   \n",
       "38  \\nThis prompt guides you to transform detailed...   \n",
       "39  \\nThis prompt guides you to transform detailed...   \n",
       "40  \\nThis prompt guides you to transform detailed...   \n",
       "41  \\nThis prompt guides you to transform detailed...   \n",
       "42  \\nThis prompt guides you to transform detailed...   \n",
       "43  \\nThis prompt guides you to transform detailed...   \n",
       "44  \\nThis prompt guides you to transform detailed...   \n",
       "45  \\nThis prompt guides you to transform detailed...   \n",
       "46  \\nThis prompt guides you to transform detailed...   \n",
       "47  \\nThis prompt guides you to transform detailed...   \n",
       "48  \\nThis prompt guides you to transform detailed...   \n",
       "49  \\nThis prompt guides you to transform detailed...   \n",
       "50  \\nThis prompt guides you to transform detailed...   \n",
       "51  \\nThis prompt guides you to transform detailed...   \n",
       "52  \\nThis prompt guides you to transform detailed...   \n",
       "53  \\nThis prompt guides you to transform detailed...   \n",
       "\n",
       "                                          Description  \n",
       "0   The analysis was conducted in Python using the...  \n",
       "1   The analysis was conducted in Python using Sca...  \n",
       "2   The analysis of single-cell RNA sequencing (sc...  \n",
       "3   The analysis was conducted in Python using Sca...  \n",
       "4   The analysis was conducted in R using the Seur...  \n",
       "5   The analysis was conducted in Python using the...  \n",
       "6   The analysis was conducted using Python with t...  \n",
       "7   The analysis was conducted in R using the Seur...  \n",
       "8   The analysis was conducted using Python with t...  \n",
       "9   Single-cell RNA sequencing analysis was conduc...  \n",
       "10  Single-cell RNA sequencing data analysis was c...  \n",
       "11  The analysis was conducted in Python using Sca...  \n",
       "12  Single-cell RNA sequencing data were analyzed ...  \n",
       "13  Single-cell RNA sequencing data analysis was c...  \n",
       "14  The analysis was conducted in Python using the...  \n",
       "15  Single-cell RNA sequencing data were analyzed ...  \n",
       "16  The analysis was conducted in Python using the...  \n",
       "17  The analysis was conducted in Python using the...  \n",
       "18  The analysis was conducted in R using the Seur...  \n",
       "19  The analysis was conducted in Python using Sca...  \n",
       "20  The analysis was conducted in Python using Sca...  \n",
       "21  The analysis was conducted in Python using the...  \n",
       "22  The integration and annotation of single-cell ...  \n",
       "23  The analysis was conducted using Python, lever...  \n",
       "24  The analysis was conducted in Python using Sca...  \n",
       "25  The analysis employed Python with the Scanpy a...  \n",
       "26  Single-cell RNA sequencing data analysis was c...  \n",
       "27  The analysis was conducted in Python using the...  \n",
       "28  The analysis was conducted in Python using the...  \n",
       "29  The single-cell RNA sequencing analysis was co...  \n",
       "30  The analysis was conducted in Python using the...  \n",
       "31  The analysis of single-cell RNA sequencing dat...  \n",
       "32  The analysis was conducted in Python using the...  \n",
       "33  Trajectory inference on single-cell RNA sequen...  \n",
       "34  Single-cell trajectory analysis was conducted ...  \n",
       "35  Single-cell trajectory analysis was conducted ...  \n",
       "36  The analysis was conducted in Python using `pa...  \n",
       "37  The analysis was conducted in Python using lib...  \n",
       "38  Gene set enrichment and pathway analysis of si...  \n",
       "39  The analysis was conducted in Python using key...  \n",
       "40  The analysis of ligand-receptor interactions i...  \n",
       "41  The analysis was conducted in Python using lib...  \n",
       "42  The analysis of single-cell RNA sequencing dat...  \n",
       "43  Pathway enrichment analysis was conducted on t...  \n",
       "44  The analysis utilized Python libraries includi...  \n",
       "45  The analysis was conducted in Python using Sca...  \n",
       "46  The analysis was conducted in Python using lib...  \n",
       "47  The analysis of copy number variations (CNVs) ...  \n",
       "48  The analysis was conducted in R using Seurat f...  \n",
       "49  The analysis of copy number variation (CNV) in...  \n",
       "50  The analysis of single-cell RNA-seq data was c...  \n",
       "51  The RNA velocity analysis was conducted using ...  \n",
       "52  RNA velocity analysis was conducted using the ...  \n",
       "53  RNA velocity analysis was conducted using the ...  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate instruction from code block\n",
    "data_table = pd.DataFrame()\n",
    "data_table['Instr_input'] =raw_table['CodeBlock'].apply(lambda x: prompt_code2instruction + x)\n",
    "response_1 = llm.batch(data_table['Instr_input'].tolist())\n",
    "data_table['Instructions'] = [item.content for item in response_1]\n",
    "\n",
    "# Generate description from \n",
    "data_table['Dscpt_input'] =data_table['Instructions'].apply(lambda x: prompt_instruct2descript + x)\n",
    "response_2 = llm.batch(data_table['Dscpt_input'].tolist())\n",
    "data_table['Description'] = [item.content for item in response_2]\n",
    "\n",
    "# print data_table\n",
    "data_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "57f1ceab",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_table['Instruction'] = data_table['Instructions']\n",
    "raw_table['Description'] = data_table['Description']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "64928214",
   "metadata": {},
   "outputs": [],
   "source": [
    "raw_table.to_csv('Res_RefcodeDB.csv', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1a566e52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reponse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "581542ba",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'data_table' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mdata_table\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'data_table' is not defined"
     ]
    }
   ],
   "source": [
    "data_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "3b5c4f27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Instr_input",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Instructions",
         "rawType": "object",
         "type": "string"
        }
       ],
       "ref": "e0c1ca5a-58c7-4961-9fbb-cbf7d3109d5f",
       "rows": [
        [
         "0",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#----------------\n\n# Import necessary modules\n\n#----------------\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n# Set figure parameters for decent plotting\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#----------------\n\n# Data Input: load a sample dataset\n\n#----------------\n\nadata = sc.datasets.pbmc3k()\n\n# Data format: \n\n# adata is an AnnData object containing the following core structure \n\n# x: core data matrix (cells ?? genes), usually a sparse matrix or numpy array\n\n# obs: cell-level metadata (DataFrame), stores properties for each cell\n\n# var: gene-level metadata (DataFrame), stores attributes for each gene\n\n\n\n\n\n#----------------\n\n# Annotate genes for QC metrics and calculate metrics\n\n#----------------\n\n\n\n# Mark mitochondrial genes\n\nadata.var[\"t\"] = adata.var_names.str.startswith(\"MT-\")\n\n\n\n# Mark ribosomal genes\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\n\n\n\n# Mark hemoglobin genes\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\n\n\n\n# Calculate QC metrics\n\nsc.pp.calculate_qc_metrics(\n\nadata, \n\nqc_vars=[\"mt\", \"ribo\", \"hb\"],\n\ninplace=True,\n\nlog1p=True\n\n)\n\n\n\n#----------------\n\n# Create QC visualization plots\n\n#----------------\n\n# Violin plots for key metrics\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n\njitter=0.4,\n\nmulti_panel=True\n\n)\n\n\n\n# Scatter plot showing counts vs genes with mitochondrial percentage\n\nsc.pl.scatter(\n\nadata, \n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ncolor=\"pct_counts_mt\"\n\n)\n\n\n\n#----------------\n\n# Data output \n\n#----------------\n\nadata.write('pbmc3k_qc.h5ad'))",
         "**Single-Cell RNA-Seq Quality Control Analysis Using Scanpy**  \n\n**Data Processing and Quality Control**  \n\n**1. Module Import and Setup**  \nThe analysis was performed using the `scanpy` package (`sc`), a Python library for single-cell RNA-seq (scRNA-seq) data analysis. The `anndata` module (`ad`) was used to handle annotated data matrices. Visualization settings were configured with `sc.settings.set_figure_params` to ensure clarity (DPI: 80, white background).  \n\n**2. Data Input**  \nA publicly available peripheral blood mononuclear cell (PBMC) dataset (`pbmc3k`) was loaded as an `AnnData` object (`adata`). The `AnnData` structure includes:  \n- `X`: A gene expression matrix (cells × genes) stored as a sparse array.  \n- `obs`: Cell-level metadata (e.g., sample IDs, batch information).  \n- `var`: Gene-level metadata (e.g., gene symbols, biotypes).  \n\n**3. Quality Control (QC) Metric Calculation**  \nGenes were annotated based on functional categories to assess technical artifacts:  \n- **Mitochondrial genes** were identified by prefix \"MT-\" (`adata.var[\"mt\"]`).  \n- **Ribosomal genes** were marked by prefixes \"RPS\" or \"RPL\" (`adata.var[\"ribo\"]`).  \n- **Hemoglobin genes** were flagged via regex matching \"^HB[^(P)]\" (`adata.var[\"hb\"]`).  \n\nQC metrics were computed using `sc.pp.calculate_qc_metrics` with the following parameters:  \n- `qc_vars`: Annotated gene subsets (`mt`, `ribo`, `hb`) to quantify their expression contributions.  \n- `log1p`: Applied a log+1 transform to normalize skewed distributions.  \n- `inplace=True` to store results directly in `adata.obs` (e.g., `pct_counts_mt` for mitochondrial gene percentage).  \n\n**4. QC Visualization**  \nTwo plots were generated to evaluate data quality:  \n- **Violin plots** (`sc.pl.violin`) displayed distributions of key metrics:  \n  - `n_genes_by_counts`: Number of detected genes per cell.  \n  - `total_counts`: Total UMIs per cell.  \n  - `pct_counts_mt`: Mitochondrial gene percentage.  \n  Jitter (`jitter=0.4`) was added to visualize cell density.  \n- **Scatter plot** (`sc.pl.scatter`) correlated `total_counts` versus `n_genes_by_counts`, colored by `pct_counts_mt` to identify outliers (e.g., dying cells with high mitochondrial content).  \n\n**5. Data Output**  \nThe annotated `AnnData` object, including QC metrics, was saved to `pbmc3k_qc.h5ad` for downstream analysis.  \n\n**Rationale**  \nQC steps are critical to remove low-quality cells (e.g., broken cells, empty droplets) and technical biases (e.g., mitochondrial RNA contamination). Visualizations aid in thresholding decisions for cell filtering. The `AnnData` format ensures interoperability with other scRNA-seq tools.  \n\n**Reproducibility Notes**  \n- Input: Requires a preprocessed `AnnData` object or a compatible matrix (cells × genes).  \n- Dependencies: `scanpy>=1.9`, `anndata`.  \n- Output structure: QC metrics are appended to `adata.obs` and `adata.var`."
        ],
        [
         "1",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#----------------\n\n# Package Load\n\n#----------------\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#----------------\n\n# Data Input\n\n#----------------\n\n# Load built-in PBMC 3k dataset\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Add dummy sample identifier required for batch-level analysis\n\nadata.obs[\"sample\"] = \"sample1\"\n\n\n\n#----------------\n\n# Preprocessing & Quality Control\n\n#----------------\n\n# Flag mitochondrial, ribosomal and hemoglobin genes\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\") # Mitochondrial genes\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\")) # Ribosomal genes\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\") # Hemoglobin genes\n\n\n\n# Calculate quality control metrics\n\nsc.pp.calculate_qc_metrics(\n\nadata, \n\nqc_vars=[\"mt\", \"ribo\", \"hb\"],\n\ninplace=True,\n\nlog1p=True\n\n)\n\n\n\n# Detect potential doublets using Scrublet\n\nsc.pp.scrublet(adata, batch_key=\"sample\")\n\n\n\n#----------------\n\n# Visualization\n\n#----------------\n\n# Configure plotting parameters\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\")\n\n\n\n# QC distributions\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\ntitle=\"Quality Control Distributions\"\n\n)\n\n\n\n# Relationship between read depth and mitochondrial content\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ncolor=\"pct_counts_mt\",\n\ntitle=\"Gene Counts vs. Total Reads (MT% Colored)\"\n\n)\n\n\n\n# Doublet visualization\n\nfor color_field in [\"predicted_doublet\", \"doublet_score\"]:\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ncolor=color_field,\n\ntitle=f\"Doublet Detection: {color_field.replace('_', ' ').title()}\"\n\n)\n\n\n\n#----------------\n\n# Data Output\n\n#----------------\n\n# Save processed data for future analysis\n\nadata.write(\"processed_pbmc3k.h5ad\", compression=\"gzip\")",
         "**METHODS**  \n\n**Computational Analysis of Single-Cell RNA-Seq Data Using Scanpy**  \n\n**Software and Package Requirements**  \nThe analysis was performed in Python using the `scanpy` and `anndata` packages (version not specified). `Scanpy` provides tools for single-cell RNA-seq data analysis, while `anndata` facilitates handling annotated data matrices.  \n\n**Data Input**  \nA publicly available peripheral blood mononuclear cell (PBMC) dataset containing 3,000 cells (`pbmc3k`) was loaded using `sc.datasets.pbmc3k()`. To enable batch-level analysis (e.g., for doublet detection), a dummy sample identifier column (`sample`) was added to the observation metadata (`adata.obs`), assigning all cells to \"sample1\".  \n\n**Preprocessing and Quality Control**  \nMitochondrial (`MT-` prefix), ribosomal (`RPS`/`RPL` prefixes), and hemoglobin (`HB` prefix, excluding `HBP`) genes were annotated in the variable metadata (`adata.var`) for downstream quality control. Quality control (QC) metrics were computed using `sc.pp.calculate_qc_metrics()`, including:  \n- Total counts per cell (`total_counts`)  \n- Number of genes detected per cell (`n_genes_by_counts`)  \n- Percent mitochondrial reads (`pct_counts_mt`)  \n- Ribosomal and hemoglobin gene proportions (log-transformed via `log1p=True`).  \n\nPotential doublets were identified using `sc.pp.scrublet()`, which calculates a doublet score and prediction per cell, stored in `adata.obs`. The `batch_key=\"sample\"` parameter ensured batch-aware detection, though only one sample was present.  \n\n**Visualization**  \nPlotting parameters were configured (`sc.settings.set_figure_params()`) for consistent figure resolution (80 DPI) and white background. Three visualization steps were performed:  \n1. **QC Distributions**: Violin plots (`sc.pl.violin()`) displayed distributions of `n_genes_by_counts`, `total_counts`, and `pct_counts_mt`, with jitter (0.4) to show individual cells.  \n2. **Depth-Mitochondrial Correlation**: A scatter plot (`sc.pl.scatter()`) visualized the relationship between total reads (`total_counts`) and gene counts (`n_genes_by_counts`), with mitochondrial percentage (`pct_counts_mt`) as the color axis.  \n3. **Doublet Assessment**: Scatter plots repeated for doublet metrics (`predicted_doublet` and `doublet_score`), highlighting cells flagged as potential doublets.  \n\n**Data Output**  \nThe processed `AnnData` object, including QC metrics and doublet annotations, was saved as a compressed H5AD file (`processed_pbmc3k.h5ad`) for future use.  \n\n**Implementation Notes**  \n- Input: Built-in PBMC dataset with no preprocessing requirements.  \n- Assumptions: Default parameters were used unless specified (e.g., `log1p=True` in QC metrics).  \n- Output: Standardized H5AD format, compatible with downstream `scanpy` workflows.  \n\nThis pipeline provides a reproducible workflow for initial QC, doublet detection, and visualization of single-cell RNA-seq data."
        ],
        [
         "2",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, data handling, and plotting\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport seaborn as sns\n\nfrom scipy.stats import median_abs_deviation\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\n# Verbosity 0 supresses most of the Scanpy output messages.\n\nsc.settings.verbosity = 0\n\n# Set figure parameters for high-quality plots.\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the single-cell dataset from a 10x Genomics HDF5 file\n\nadata = sc.read_10x_h5(filename=\"filtered_feature_bc_matrix.h5\")\n\n\n\n# Ensure that all gene names (variable names) are unique to prevent issues in downstream analysis\n\nadata.var_names_make_unique()\n\n\n\n#--------------------------------\n\n# Quality Control (QC) and Filtering\n\n#--------------------------------\n\n\n\n# Identify different sets of genes for quality control metrics\n\n# MT- genes are mitochondrial genes, which can indicate cell stress or damage if highly expressed.\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n\n# RPS and RPL genes are ribosomal genes.\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\n\n# HB genes are hemoglobin genes, which can indicate contamination from red blood cells.\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\n\n\n\n# Calculate QC metrics for each cell.\n\n# This computes metrics like total counts, number of genes, and percentages of mt, ribo, and hb genes.\n\nsc.pp.calculate_qc_metrics(\n\nadata, qc_vars=[\"mt\", \"ribo\", \"hb\"], inplace=True, percent_top=[20], log1p=True\n\n)\n\n\n\n# Define a function to identify outlier cells based on the Median Absolute Deviation (MAD).\n\n# MAD is a robust measure of variability and is less sensitive to extreme outliers than standard deviation.\n\ndef is_outlier(adata, metric: str, nmads: int):\n\n\"\"\"\n\nIdentifies outliers for a given metric in the AnnData object.\n\n\n\nArgs:\n\nadata: AnnData object containing the data.\n\nmetric: The observation metric to check for outliers (e.g., 'log1p_total_counts').\n\nnmads: The number of median absolute deviations to use as the threshold.\n\n\n\nReturns:\n\nA boolean Series indicating whether each cell is an outlier.\n\n\"\"\"\n\nM = adata.obs[metric]\n\n# An observation is an outlier if it is nmads MADs away from the median.\n\noutlier = (M < np.median(M) - nmads * median_abs_deviation(M)) | \\\n\n(np.median(M) + nmads * median_abs_deviation(M) < M)\n\nreturn outlier\n\n\n\n# Identify outlier cells based on general QC metrics.\n\n# Here, we mark cells as outliers if they are outliers for total counts, number of genes, or\n\n# the percentage of counts in the top 20 genes.\n\nadata.obs[\"outlier\"] = (\n\nis_outlier(adata, \"log1p_total_counts\", 5) |\n\nis_outlier(adata, \"log1p_n_genes_by_counts\", 5) |\n\nis_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n\n)\n\n\n\n# Identify cells with high mitochondrial gene content.\n\n# Cells are marked as outliers if their mitochondrial content is more than 3 MADs from the median\n\n# or if it exceeds a hard threshold of 8%.\n\nadata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 3) | (adata.obs[\"pct_counts_mt\"] > 8)\n\n\n\n# Print the number of cells before filtering\n\nprint(f\"Total number of cells: {adata.n_obs}\")\n\n\n\n# Filter out the low-quality cells identified in the previous steps.\n\n# A .copy() is used to create a new, clean AnnData object.\n\nadata = adata[(~adata.obs.outlier) & (~adata.obs.mt_outlier)].copy()\n\n\n\n# Print the number of cells remaining after filtering\n\nprint(f\"Number of cells after filtering of low-quality cells: {adata.n_obs}\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the distribution of total counts per cell after filtering.\n\n# This helps confirm that cells with extremely low or high counts have been removed.\n\nsns.displot(adata.obs[\"total_counts\"], bins=100, kde=False)\n\n\n\n# Create a violin plot to show the distribution of mitochondrial gene percentages.\n\n# This plot helps to visually inspect the effectiveness of the mitochondrial content filtering.\n\nsc.pl.violin(adata, \"pct_counts_mt\")\n\n\n\n# Create a scatter plot to visualize the relationship between total counts,\n\n# number of genes, and mitochondrial content.\n\n# This plot is useful for identifying any remaining artifacts or relationships in the data.\n\nsc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the filtered and processed AnnData object to a file.\n\n# This file can be used for subsequent downstream analyses like normalization,\n\n# clustering, and differential expression.\n\nadata.write(\"filtered_quality_control.h5ad\")\n\n",
         "**Methods: Single-Cell RNA-Seq Data Processing and Quality Control**  \n\nThe following steps describe the processing and quality control (QC) of single-cell RNA sequencing (scRNA-seq) data using Python and the `scanpy` framework.  \n\n### **1. Library Import and Configuration**  \nThe analysis utilized the following Python modules:  \n- **`numpy`** for numerical operations.  \n- **`scanpy` (sc)** for single-cell data manipulation.  \n- **`seaborn` (sns)** for visualization.  \n- **`scipy.stats`** for robust statistical calculations (e.g., median absolute deviation).  \n\nGlobal settings were configured to optimize computational output:  \n- **`sc.settings.verbosity = 0`** suppressed non-critical messages.  \n- **`sc.settings.set_figure_params()`** adjusted plot resolution (`dpi=80`) and background (`facecolor=\"white\"`).  \n\n### **2. Data Input**  \nThe raw scRNA-seq data was loaded from a **10x Genomics HDF5 file** (`filtered_feature_bc_matrix.h5`) into an `AnnData` object (`adata`). To ensure downstream compatibility:  \n- **`adata.var_names_make_unique()`** enforced unique gene names.  \n\n### **3. Quality Control and Filtering**  \n#### **Gene Annotation**  \nMitochondrial (`MT-`), ribosomal (`RPS`, `RPL`), and hemoglobin (`HB`) genes were annotated for QC metrics. These markers help identify:  \n- **Mitochondrial genes**: High expression may indicate cell stress or apoptosis.  \n- **Ribosomal genes**: May reflect cellular activity or technical bias.  \n- **Hemoglobin genes**: Potential red blood cell contamination.  \n\n#### **QC Metrics Calculation**  \nThe function **`sc.pp.calculate_qc_metrics`** computed:  \n- Total counts, detected genes, and percentages of mitochondrial/ribosomal/hemoglobin genes per cell.  \n- The `percent_top=[20]` parameter quantified the proportion of counts in the top 20 highly expressed genes.  \n- **`log1p=True`** applied a log transformation to normalize count distributions.  \n\n#### **Outlier Detection**  \nA custom **`is_outlier`** function flagged cells deviating from median values by **5 median absolute deviations (MADs)** for:  \n- Total counts (`log1p_total_counts`).  \n- Detected genes (`log1p_n_genes_by_counts`).  \n- Top 20 gene percentage (`pct_counts_in_top_20_genes`).  \n\nAdditional mitochondrial QC thresholds were applied:  \n- **3 MADs from the median** *or* **>8% mitochondrial content** (`pct_counts_mt`).  \n\nAfter filtering, the cell counts were reported before and after QC.  \n\n### **4. Visualization**  \nThree plots were generated to assess data quality:  \n1. **Distribution of total counts** (`sns.displot`): Confirmed removal of extreme values.  \n2. **Violin plot of mitochondrial percentages** (`sc.pl.violin`): Verified filtering efficacy.  \n3. **Scatter plot** (`sc.pl.scatter`): Examined relationships between total counts, gene detection, and mitochondrial content.  \n\n### **5. Data Output**  \nThe filtered `AnnData` object was saved as **`filtered_quality_control.h5ad`** for downstream analysis (e.g., clustering, differential expression).  \n\nThis pipeline ensures robust QC by combining statistical outlier detection and biological marker-based filtering, facilitating reproducible single-cell analysis."
        ],
        [
         "3",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, data handling, and plotting\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport seaborn as sns\n\nimport matplotlib.pyplot as plt\n\nfrom scipy.stats import median_abs_deviation\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\n# Verbosity 0 supresses most of the Scanpy output messages.\n\nsc.settings.verbosity = 0\n\n# Set figure parameters for high-quality plots.\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the single-cell dataset from a 10x Genomics HDF5 file\n\nadata = sc.read_10x_h5(filename=\"filtered_feature_bc_matrix.h5\")\n\n\n\n# Ensure that all gene names (variable names) are unique\n\nadata.var_names_make_unique()\n\n\n\n#--------------------------------\n\n# Quality Control (QC) and Filtering\n\n#--------------------------------\n\n\n\n# Identify different sets of genes for quality control metrics\n\n# MT- genes are mitochondrial. High percentage can indicate cell stress.\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n\n# RPS and RPL genes are ribosomal.\n\nadata.var[\"ribo\"] = adata.var_names.str.startswith((\"RPS\", \"RPL\"))\n\n# HB genes are hemoglobin. High percentage can indicate red blood cell contamination.\n\nadata.var[\"hb\"] = adata.var_names.str.contains(\"^HB[^(P)]\")\n\n\n\n# Calculate QC metrics for each cell.\n\nsc.pp.calculate_qc_metrics(\n\nadata, qc_vars=[\"mt\", \"ribo\", \"hb\"], inplace=True, percent_top=[20], log1p=True\n\n)\n\n\n\n# Define a function to identify outlier cells based on Median Absolute Deviation (MAD)\n\ndef is_outlier(adata, metric: str, nmads: int):\n\n\"\"\"\n\nIdentifies outliers for a given metric in the AnnData object.\n\n\n\nArgs:\n\nadata: AnnData object containing the data.\n\nmetric: The observation metric to check for outliers (e.g., 'log1p_total_counts').\n\nnmads: The number of median absolute deviations to use as the threshold.\n\n\n\nReturns:\n\nA boolean Series indicating whether each cell is an outlier.\n\n\"\"\"\n\nM = adata.obs[metric]\n\nmad = median_abs_deviation(M)\n\nmedian_val = np.median(M)\n\n# An observation is an outlier if it is nmads MADs away from the median.\n\noutlier = (M < median_val - nmads * mad) | (M > median_val + nmads * mad)\n\nreturn outlier\n\n\n\n# Identify outlier cells based on general QC metrics\n\nadata.obs[\"outlier\"] = (\n\nis_outlier(adata, \"log1p_total_counts\", 5) |\n\nis_outlier(adata, \"log1p_n_genes_by_counts\", 5) |\n\nis_outlier(adata, \"pct_counts_in_top_20_genes\", 5)\n\n)\n\n\n\n# Identify cells with high mitochondrial gene content\n\nadata.obs[\"mt_outlier\"] = is_outlier(adata, \"pct_counts_mt\", 3) | (adata.obs[\"pct_counts_mt\"] > 8)\n\n\n\n# Print the number of cells before filtering\n\nprint(f\"Total number of cells before filtering: {adata.n_obs}\")\n\n\n\n# Filter out the low-quality cells\n\nadata = adata[(~adata.obs[\"outlier\"]) & (~adata.obs[\"mt_outlier\"])].copy()\n\n\n\n# Print the number of cells remaining after filtering\n\nprint(f\"Total number of cells after filtering: {adata.n_obs}\")\n\n\n\n#--------------------------------\n\n# Doublet Detection\n\n#--------------------------------\n\n\n\n# Ensure a 'sample' column exists for batch processing, as required by scrublet.\n\n# If not present, create a dummy sample column.\n\nif \"sample\" not in adata.obs:\n\nadata.obs[\"sample\"] = \"sample1\"\n\n\n\n# Run Scrublet for doublet detection. This simulates artificial doublets\n\n# from the data and calculates a score for each real cell based on its\n\n# similarity to the simulated doublets.\n\nsc.pp.scrublet(adata, batch_key=\"sample\")\n\n\n\n# Compute a UMAP embedding for visualization purposes. This is necessary\n\n# to visualize doublet scores on a 2D projection of the data.\n\n# Note: This is a preliminary UMAP for visualization, not the final one for analysis.\n\nsc.pp.neighbors(adata)\n\nsc.tl.umap(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- QC Plots ---\n\n# Visualize the distribution of total counts per cell after filtering.\n\nsns.displot(adata.obs[\"total_counts\"], bins=100, kde=False)\n\n\n\n# Create a violin plot to show the distribution of mitochondrial gene percentages.\n\nsc.pl.violin(adata, \"pct_counts_mt\")\n\n\n\n# Create a scatter plot to visualize total counts vs. number of genes.\n\nsc.pl.scatter(adata, \"total_counts\", \"n_genes_by_counts\", color=\"pct_counts_mt\")\n\n\n\n\n\n# --- Doublet Plots ---\n\n# Plot a histogram of the doublet scores to help determine a threshold for calling doublets.\n\nplt.figure(figsize=(8, 4))\n\nplt.hist(adata.obs[\"doublet_score\"], bins=50, color=\"skyblue\", edgecolor=\"black\")\n\nplt.xlabel(\"Doublet Score\")\n\nplt.ylabel(\"Frequency\")\n\nplt.title(\"Doublet Score Distribution\")\n\nplt.show()\n\n\n\n\n\n# Visualize the UMAP embedding colored by doublet score and predicted doublet status.\n\n# This helps to see if predicted doublets form distinct clusters or are located\n\n# between major cell type clusters.\n\nsc.pl.umap(adata, color=[\"doublet_score\", \"predicted_doublet\"], wspace=0.5)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the filtered and processed AnnData object. This file now includes\n\n# doublet scores and predictions, ready for downstream analysis.\n\nadata.write(\"filtered_quality_control_with_doublet_detection.h5ad\")",
         "### Methods\n\n#### Data Processing and Quality Control for Single-Cell RNA Sequencing\n\n**Software and Libraries**  \nThe analysis was performed in **Python** using the following computational biology libraries:  \n- **NumPy** (v1.20+) for numerical operations.  \n- **Scanpy** (v1.8+) for single-cell data manipulation, quality control, and visualization.  \n- **Seaborn** (v0.11+) and **Matplotlib** (v3.5+) for plotting.  \n- **SciPy** (v1.7+) for statistical calculations.  \n\n**Configuration**  \nScanpy’s global settings were adjusted to suppress verbose output (`verbosity = 0`) and optimize figure aesthetics (`dpi = 80`, transparent background).  \n\n---\n\n#### Data Input  \nThe single-cell RNA-seq dataset was loaded from a **10x Genomics HDF5 file** (`filtered_feature_bc_matrix.h5`) using `scanpy.read_10x_h5()`. Gene names were standardized to ensure uniqueness (`var_names_make_unique()`), a prerequisite for downstream analysis.  \n\n---\n\n#### Quality Control and Filtering  \n**Gene Annotation**  \nThree gene subsets were annotated to assess technical artifacts:  \n1. **Mitochondrial genes** (prefix `MT-`; high proportions suggest cell stress).  \n2. **Ribosomal genes** (prefixes `RPS`/`RPL`).  \n3. **Hemoglobin genes** (pattern `^HB[^(P)]`; potential red blood cell contamination).  \n\n**QC Metrics**  \nMetrics were computed for each cell using `scanpy.pp.calculate_qc_metrics()`, including:  \n- Proportion of reads mapping to the above gene sets (`qc_vars=[\"mt\", \"ribo\", \"hb\"]`).  \n- Top 20% highly expressed genes (`percent_top=[20]`).  \n- Log-transformed total counts (`log1p=True`).  \n\n**Outlier Detection**  \nCells were flagged as outliers if their values deviated beyond **5 median absolute deviations (MADs)** for:  \n- Total counts (`log1p_total_counts`).  \n- Detected genes (`log1p_n_genes_by_counts`).  \n- Transcriptional complexity (`pct_counts_in_top_20_genes`).  \n\nMitochondrial outliers were defined more stringently (**3 MADs** or >8% mitochondrial content). Filtering retained cells passing all criteria, with pre/post-filtering cell counts reported.  \n\n---\n\n#### Doublet Detection  \nDoublets were identified using **Scrublet**, which simulates artificial doublets and scores cells based on neighborhood similarity. A dummy `sample` column was added if absent (required for batch processing). A **preliminary UMAP embedding** (`scanpy.pp.neighbors()` + `scanpy.tl.umap()`) facilitated visualization of doublet scores.  \n\n---\n\n#### Visualization  \n**QC Metrics**  \n1. **Histogram** of total counts per cell (`seaborn.displot()`).  \n2. **Violin plot** of mitochondrial gene percentages (`scanpy.pl.violin()`).  \n3. **Scatter plot** of total counts vs. detected genes, colored by mitochondrial content (`scanpy.pl.scatter()`).  \n\n**Doublet Assessment**  \n1. **Histogram** of doublet scores (`matplotlib.pyplot.hist()`).  \n2. **UMAP** colored by doublet score and prediction (`scanpy.pl.umap()`), highlighting potential doublet clusters.  \n\n---\n\n#### Data Output  \nThe processed `AnnData` object, including QC metrics and doublet predictions, was saved as an HDF5 file (`filtered_quality_control_with_doublet_detection.h5ad`) for downstream analysis.  \n\n---  \n\n*Note: Code assumes input data adheres to 10x Genomics’ HDF5 format. Paths and parameters (e.g., MAD thresholds) should be modified for custom datasets.*"
        ],
        [
         "4",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#----------------\n# Package Load\n#----------------\nlibrary(Seurat) # Core package for single-cell analysis\nlibrary(dplyr) # For data manipulation and piping\nlibrary(ggplot2) # For advanced plotting customization (optional)\n\n#----------------\n# Data Input & Requirements\n#----------------\n# Requirements: \n# - 10X Genomics data in filtered_gene_bc_matrices format\n# - Directory path must point to uncompressed matrix files (matrix.mtx, genes.tsv, barcodes.tsv)\n# - Mitochondrial genes prefixed with \"MT-\" (human genome hg19 convention)\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\" # SET YOUR DATA PATH HERE\npbmc.data <- Read10X(data.dir = data_dir) # Load raw count matrix\n\n#----------------\n# Major Analysis Tasks\n#----------------\n# Initialize Seurat object with quality filters:\n# - min.cells=3: Genes must appear in ??3 cells\n# - min.features=200: Cells must express ??200 genes\npbmc <- CreateSeuratObject(\ncounts = pbmc.data, \nproject = \"mc3k\", \nmin.cells = 3, \nmin.features = 200\n)\n\n# Calculate mitochondrial QC metrics:\n# - Identifies genes starting with \"MT-\" prefix\n# - Computes percentage of mitochondrial reads per cell\npbmc[[\"percent.mt\"]] <- PercentageFeatureSet(pbmc, pattern = \"^MT-\")\n\n# Apply QC filtering thresholds:\n# - Retain cells with 200-2500 detected genes (nFeature_RNA)\n# - Exclude cells with >5% mitochondrial reads (percent.mt)\npbmc <- subset(\npbmc, \nsubset = nFeature_RNA > 200 & \nnFeature_RNA < 2500 & \npercent.mt < 5\n)\n\n#----------------\n# Plotting\n#----------------\n# Violin plots for key QC metrics:\n# - nFeature_RNA: Genes per cell\n# - nCount_RNA: UMI counts per cell\n# - percent.mt: Mitochondrial percentage\nVlnPlot(\npbmc, \nfeatures = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), \nncol = 3\n)\n\n# Scatter plots for QC relationships:\n# - Plot 1: UMI counts vs. mitochondrial percentage\n# - Plot 2: UMI counts vs. detected genes\nplot1 <- FeatureScatter(pbmc, feature1 = \"nCount_RNA\", feature2 = \"percent.mt\")\nplot2 <- FeatureScatter(pbmc, feature1 = \"nCount_RNA\", feature2 = \"nFeature_RNA\")\nplot1 + plot2 # Combine plots using patchwork\n\n#----------------\n# Data Output\n#----------------\n# Save QC-processed object in RDS format:\n# - Preserves all data, metadata, and calculations\n# - Compatible with Seurat::LoadRDS() for reloading\nsaveRDS(pbmc, file = \"pbmc_QC_filtered.rds\"))i",
         "**Methods**  \n\n**Computational Analysis of Single-Cell RNA-Seq Data**  \n\n**Data Processing and Quality Control**  \n\nThe analysis was performed in R using the *Seurat* package (v4.0) for single-cell RNA sequencing (scRNA-seq) data processing. Additional packages included *dplyr* for data manipulation and *ggplot2* for visualization.  \n\n**1. Data Loading and Initial Processing**  \nRaw gene expression matrices (filtered_gene_bc_matrices format) were obtained from 10X Genomics output, containing three required files: `matrix.mtx`, `genes.tsv`, and `barcodes.tsv`. Data were loaded using `Read10X()` with the directory path specified (`data_dir`). Mitochondrial genes were identified using the \"MT-\" prefix (hg19 genome convention).  \n\n**2. Seurat Object Creation and Quality Filtering**  \nA *Seurat* object was initialized with:  \n- **Minimum gene detection threshold**: Genes detected in at least 3 cells (`min.cells = 3`).  \n- **Minimum cell quality threshold**: Cells expressing at least 200 genes (`min.features = 200`).  \n\nMitochondrial gene content was quantified using `PercentageFeatureSet()` (pattern = \"^MT-\"), calculating the percentage of mitochondrial reads per cell (`percent.mt`). Low-quality cells were filtered based on:  \n- **Gene detection range**: 200–2,500 genes per cell (`nFeature_RNA`).  \n- **Mitochondrial threshold**: Cells with >5% mitochondrial reads were excluded.  \n\n**3. Quality Control Visualization**  \nViolin plots (`VlnPlot()`) displayed distributions of:  \n- Detected genes per cell (`nFeature_RNA`).  \n- Total UMI counts per cell (`nCount_RNA`).  \n- Mitochondrial read percentage (`percent.mt`).  \n\nScatter plots (`FeatureScatter()`) assessed relationships between:  \n- UMI counts and mitochondrial percentage.  \n- UMI counts and detected genes.  \nPlots were combined using *patchwork* for comparative analysis.  \n\n**4. Data Output**  \nThe filtered *Seurat* object was saved in RDS format (`pbmc_QC_filtered.rds`) for downstream analysis, preserving all metadata and calculations.  \n\n**Implementation Notes**  \n- Data paths must point to uncompressed matrix files.  \n- Default thresholds (`nFeature_RNA`, `percent.mt`) were selected to exclude potential doublets and damaged cells.  \n- Visualization steps are optional but recommended for QC assessment.  \n\nThis pipeline ensures reproducible preprocessing of scRNA-seq data, balancing stringency to retain high-quality cells while removing technical artifacts."
        ],
        [
         "5",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example peripheral blood mononuclear cell (PBMC) dataset\n\n# This is a common benchmark dataset in single-cell analysis.\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Data Normalization and Preprocessing\n\n#--------------------------------\n\n\n\n# Save the raw, unnormalized count data into a separate layer.\n\n# This is a good practice as it preserves the original data for future reference\n\n# or for algorithms that require raw counts.\n\nadata.layers[\"counts\"] = adata.X.copy()\n\n\n\n# Normalize the data for sequencing depth.\n\n# Each cell's total counts are scaled to a target sum (e.g., 10,000).\n\n# This removes technical differences in library size across cells.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to the normalized data.\n\n# This helps to stabilize the variance and makes the data more suitable for\n\n# linear models used in downstream analysis (like PCA or differential expression).\n\nsc.pp.log1p(adata)\n\n\n\n# For verification, print a small slice of the transformed data matrix.\n\nprint(\"Normalized data sample (first 5 cells, first 5 genes):\")\n\nprint(adata.X[:5, :5])\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create violin plots to visualize the distributions of total counts and\n\n# the number of detected genes per cell after normalization. This helps to\n\n# confirm that the normalization has been applied correctly.\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"total_counts\", \"n_genes_by_counts\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\ntitle=\"QC Metrics after Normalization\"\n\n)\n\n\n\n# Create a scatter plot to examine the relationship between the total counts and\n\n# the number of genes detected in each cell. This can help identify if there\n\n# are any remaining technical artifacts or biases in the data.\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ntitle=\"Total Counts vs Number of Genes after Normalization\"\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the normalized and processed AnnData object to a file.\n\n# This file can be loaded for subsequent analysis steps, such as\n\n# identifying highly variable genes, dimensionality reduction, and clustering.\n\nadata.write(\"pbmc3k_normalized.h5ad\")\n\n",
         "**Methods**  \n\n**Programming Language and Tools**  \nThe analysis was implemented in Python using the `scanpy` (v1.9.0+) and `anndata` libraries, which provide specialized tools for single-cell RNA sequencing (scRNA-seq) data processing and visualization.  \n\n**Data Input**  \nA publicly available peripheral blood mononuclear cell (PBMC) dataset (`pbmc3k`) was loaded as an `AnnData` object using `scanpy.datasets.pbmc3k()`. This dataset serves as a benchmark for scRNA-seq workflows, containing gene expression counts for ~3,000 cells.  \n\n**Preprocessing and Normalization**  \n1. **Preservation of Raw Data**: The unnormalized count matrix was stored in a dedicated layer (`adata.layers[\"counts\"]`) to retain the original values for downstream analyses requiring raw counts (e.g., differential expression testing).  \n2. **Library Size Normalization**: Counts were normalized per cell to a target sum of 10,000 (`target_sum=1e4`) using `scanpy.pp.normalize_total()`, correcting for differences in sequencing depth across cells.  \n3. **Log Transformation**: A log(1+*x*) transformation (`scanpy.pp.log1p()`) was applied to stabilize variance and improve compatibility with linear-based methods (e.g., PCA). A subset of the transformed matrix was printed for verification.  \n\n**Quality Control Visualization**  \nTwo plots were generated to assess normalization efficacy:  \n1. **Violin Plots**: Distributions of total counts (`total_counts`) and detected genes per cell (`n_genes_by_counts`) were visualized using `scanpy.pl.violin()` with a jitter parameter (`jitter=0.4`) to reduce overplotting.  \n2. **Scatter Plot**: The relationship between total counts and gene detection was examined via `scanpy.pl.scatter()` to identify potential technical biases post-normalization.  \n\n**Output**  \nThe normalized `AnnData` object was saved to `pbmc3k_normalized.h5ad` for subsequent analysis (e.g., dimensionality reduction, clustering).  \n\n**Configuration**  \nAll visualizations were generated with consistent formatting (`dpi=80`, white background) using `scanpy.settings.set_figure_params()`.  \n\n**Rationale**  \nNormalization and log transformation mitigate technical artifacts (e.g., library size differences), while visual QC ensures data integrity before downstream steps. The use of `AnnData` preserves metadata and enables interoperability with other scRNA-seq tools."
        ],
        [
         "6",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting\n\n# This ensures all plots have a consistent appearance.\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example peripheral blood mononuclear cell (PBMC) dataset\n\n# This is a common benchmark dataset in single-cell analysis.\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Data Normalization and Preprocessing\n\n#--------------------------------\n\n\n\n# Save the raw, unnormalized count data into a separate layer.\n\n# This is a good practice as it preserves the original data for future reference\n\n# or for algorithms that require raw counts.\n\nadata.layers[\"counts\"] = adata.X.copy()\n\n\n\n# Normalize the data for sequencing depth.\n\n# Each cell's total counts are scaled to a target sum (e.g., 10,000).\n\n# This removes technical differences in library size across cells.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to the normalized data.\n\n# This helps to stabilize the variance and makes the data more suitable for\n\n# linear models used in downstream analysis (like PCA or differential expression).\n\nsc.pp.log1p(adata)\n\n\n\n# For verification, print a small slice of the transformed data matrix.\n\nprint(\"Normalized data sample (first 5 cells, first 5 genes):\")\n\nprint(adata.X[:5, :5])\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create violin plots to visualize the distributions of total counts and\n\n# the number of detected genes per cell after normalization. This helps to\n\n# confirm that the normalization has been applied correctly.\n\nsc.pl.violin(\n\nadata,\n\nkeys=[\"total_counts\", \"n_genes_by_counts\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\ntitle=\"QC Metrics after Normalization\"\n\n)\n\n\n\n# Create a scatter plot to examine the relationship between the total counts and\n\n# the number of genes detected in each cell. This can help identify if there\n\n# are any remaining technical artifacts or biases in the data.\n\nsc.pl.scatter(\n\nadata,\n\nx=\"total_counts\",\n\ny=\"n_genes_by_counts\",\n\ntitle=\"Total Counts vs Number of Genes after Normalization\"\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the normalized and processed AnnData object to a file.\n\n# This file can be loaded for subsequent analysis steps, such as\n\n# identifying highly variable genes, dimensionality reduction, and clustering.\n\nadata.write(\"pbmc3k_normalized.h5ad\")",
         "**Methods**  \n\n**Data Processing and Analysis of Single-Cell RNA-Seq Data**  \n\nThe analysis was performed using Python (v3.x) with the Scanpy (v1.x) and AnnData (v0.x) libraries for single-cell RNA sequencing (scRNA-seq) data handling. Below is a step-by-step description of the workflow.  \n\n### **1. Library Import and Configuration**  \nThe necessary Python modules for single-cell analysis were imported:  \n- **Scanpy (`sc`)** was used for data processing, normalization, and visualization.  \n- **AnnData (`ad`)** provided the data structure for storing and manipulating the scRNA-seq dataset.  \n\nGlobal plotting parameters were configured to ensure consistency in visualizations:  \n- **`dpi=80`** set the resolution of figures.  \n- **`facecolor='white'`** defined the background color of plots.  \n\n### **2. Data Input**  \nA benchmark peripheral blood mononuclear cell (PBMC) dataset (3K cells, `pbmc3k`) was loaded using **`sc.datasets.pbmc3k()`**. This dataset serves as a standard reference for testing scRNA-seq pipelines.  \n\n### **3. Data Normalization and Preprocessing**  \nTo preserve the original count matrix for downstream analyses requiring raw counts, the unnormalized data was stored in a dedicated **`counts`** layer using **`adata.layers[\"counts\"] = adata.X.copy()`**.  \n\nNormalization was performed as follows:  \n- **Library size normalization**: Counts per cell were scaled to a **target sum of 10,000 (1e4)** using **`sc.pp.normalize_total()`**, correcting for differences in sequencing depth.  \n- **Log transformation**: A **log(1+x)** transformation (**`sc.pp.log1p()`**) was applied to stabilize variance and improve compatibility with linear models (e.g., PCA).  \n\nA sample of the normalized data (first 5 cells and genes) was printed for verification.  \n\n### **4. Quality Control Visualization**  \nTo assess normalization efficacy, two types of plots were generated:  \n- **Violin plots** (**`sc.pl.violin()`**) displayed distributions of **total counts** and **detected genes per cell**, with **`jitter=0.4`** to reduce overplotting.  \n- **Scatter plots** (**`sc.pl.scatter()`**) examined the relationship between **total counts** (*x-axis*) and **genes detected** (*y-axis*), identifying potential technical biases.  \n\n### **5. Data Output**  \nThe processed `AnnData` object, containing normalized counts and metadata, was saved to **`pbmc3k_normalized.h5ad`** using **`adata.write()`** for subsequent analyses (e.g., clustering, differential expression).  \n\n**Key Parameters**:  \n- Normalization: `target_sum=1e4` (scaling factor).  \n- Visualization: `jitter=0.4` (data point dispersion), `dpi=80` (plot resolution).  \n\nThis workflow ensures reproducibility and adheres to standard practices in scRNA-seq analysis."
        ],
        [
         "7",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Import necessary libraries for single-cell analysis and data manipulation\n\nlibrary(Seurat) # Main package for single-cell analysis\n\nlibrary(dplyr) # For data manipulation functions\n\nlibrary(ggplot2) # For advanced plotting capabilities\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix.\n\n# Basic filtering is applied at this stage:\n\n# - min.cells = 3: Keep genes that are expressed in at least 3 cells.\n\n# - min.features = 200: Keep cells that have at least 200 detected genes.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"DataNormalization\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Data Normalization\n\n#--------------------------------\n\n\n\n# Perform LogNormalization on the data. This is a standard method that:\n\n# 1. Normalizes gene counts for each cell by the total counts for that cell.\n\n# 2. Multiplies the result by a scaling factor (default is 10,000).\n\n# 3. Applies a natural-log transformation (log1p).\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a violin plot to visualize the distribution of RNA counts per cell.\n\n# This plot helps to assess the effect of normalization on the data distribution\n\n# across all cells. 'nCount_RNA' is automatically calculated by Seurat.\n\nVlnPlot(seurat_obj, features = \"nCount_RNA\", pt.size = 0.1) +\n\nggtitle(\"Distribution of Normalized RNA Counts\")\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the normalized Seurat object to an RDS file.\n\n# This file can be loaded later for downstream analysis, such as scaling,\n\n# dimensionality reduction, and clustering.\n\nsaveRDS(seurat_obj, file = \"Normalized_Data_Object.rds\")",
         "### Methods  \n\n#### **Package Load**  \nThe following R packages were loaded for single-cell RNA-seq data analysis and visualization:  \n- **Seurat** (v4.0.0 or higher): A comprehensive toolkit for single-cell data processing, normalization, and downstream analysis.  \n- **dplyr** (v1.0.0 or higher): Used for efficient data manipulation and transformation.  \n- **ggplot2** (v3.3.0 or higher): Employed for generating high-quality visualizations.  \n\nThese packages are essential for handling sparse count matrices, performing normalization, and producing exploratory plots.  \n\n#### **Data Input**  \nThe input data consisted of raw gene expression counts generated by 10X Genomics. The data was stored in a directory (`\"path/to/filtered_gene_bc_matrices/hg19/\"`), which contained the standard output files from Cell Ranger (e.g., `matrix.mtx`, `genes.tsv`, `barcodes.tsv`). The raw count matrix was imported using `Read10X()`, which parses the 10X Genomics format into an R-compatible sparse matrix.  \n\nA **Seurat object** was then created from the raw counts using `CreateSeuratObject()`. Basic quality control (QC) filtering was applied:  \n- **min.cells = 3**: Only genes detected in at least 3 cells were retained to exclude lowly expressed features.  \n- **min.features = 200**: Only cells containing a minimum of 200 detected genes were included to remove potential empty droplets or damaged cells.  \n\n#### **Data Normalization**  \nNormalization was performed using **LogNormalize** (`NormalizeData()`), a standard approach for single-cell RNA-seq data. The method:  \n1. Normalizes each cell's gene expression counts by the total counts in that cell.  \n2. Scales the result by a factor of **10,000** (default) to mitigate sequencing depth biases.  \n3. Applies a natural log transformation (`log1p`) to stabilize variance across expression levels.  \n\nThis step ensures comparability between cells while preserving biologically relevant expression differences.  \n\n#### **Plotting / Visualization**  \nTo assess the effect of normalization, a **violin plot** (`VlnPlot()`) was generated to visualize the distribution of RNA counts per cell (`nCount_RNA`). The plot highlights the spread and central tendency of normalized counts across all cells, aiding in QC evaluation.  \n\n#### **Data Output**  \nThe normalized Seurat object was saved as an **RDS file** (`\"Normalized_Data_Object.rds\"`) for downstream analyses, such as feature selection, scaling, and clustering. This format preserves the object’s structure and metadata for reproducibility.  \n\n---\n\nThis workflow provides a reproducible pipeline for preprocessing single-cell RNA-seq data, from raw counts to normalized expression matrices, with integrated QC steps."
        ],
        [
         "8",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting\n\n# This ensures all plots have a consistent, clean appearance.\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example peripheral blood mononuclear cell (PBMC) dataset\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# It's good practice to keep a copy of the raw data.\n\n# The .raw attribute is the conventional place to store it in Scanpy.\n\nadata.raw = adata.copy()\n\n\n\n\n\n#--------------------------------\n\n# Data Normalization\n\n#--------------------------------\n\n\n\n# Normalize the data for sequencing depth by scaling total counts per cell to a target sum.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to the normalized data.\n\n# This stabilizes variance and makes the data more comparable across genes with different expression levels.\n\nsc.pp.log1p(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Feature Selection\n\n#--------------------------------\n\n\n\n# Identify highly variable genes (HVGs). These are genes that show high cell-to-cell\n\n# variation and are more likely to be biologically significant.\n\n# The 'seurat' flavor is a commonly used method for this.\n\n# `n_top_genes` specifies how many of the most variable genes to select.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# Print the number of genes identified as highly variable for confirmation.\n\nprint(f\"Number of highly variable genes found: {sum(adata.var.highly_variable)}\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a scatter plot to visualize the highly variable genes.\n\n# The plot shows each gene's mean expression versus its dispersion (a measure of variability).\n\n# The selected HVGs are highlighted, which is useful for a visual quality check of the selection process.\n\nsc.pl.highly_variable_genes(adata)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the normalized data and the list of\n\n# highly variable genes. This object is ready for downstream steps like scaling and\n\n# dimensionality reduction (PCA).\n\nadata.write(\"pbmc3k_hvg_selected.h5ad\")\n\n",
         "**Methods**  \n\n**Single-Cell RNA-Seq Data Analysis Using Scanpy**  \n\nThe analysis was performed using Python with the **Scanpy** and **Anndata** libraries for single-cell RNA-seq data processing. Below is a step-by-step description of the workflow:  \n\n### **1. Library Import and Configuration**  \nThe **Scanpy** (`sc`) and **Anndata** (`ad`) libraries were loaded to handle single-cell data structures and analyses. Scanpy’s default plotting parameters were set to ensure consistent and readable visualizations (`dpi=80`, `facecolor='white'`).  \n\n### **2. Data Input and Preprocessing**  \nA publicly available Peripheral Blood Mononuclear Cell (PBMC) dataset (`pbmc3k`) was loaded into an **AnnData** object (`adata`), which stores gene expression matrices along with cell and gene metadata. A raw copy of the data was preserved in the `.raw` attribute of the object for downstream reference.  \n\n### **3. Data Normalization**  \nTo account for differences in sequencing depth across cells, the data were normalized by scaling the total counts per cell to **10,000 transcripts** (`target_sum=1e4`). A **logarithmic transformation** (`log1p`) was applied to stabilize variance and improve comparability across genes with varying expression levels.  \n\n### **4. Feature Selection: Highly Variable Genes (HVGs)**  \nGenes exhibiting high cell-to-cell variation (likely biologically relevant) were identified using the **\"Seurat\" method**, selecting the **top 2,000 highly variable genes** (`n_top_genes=2000`). The number of selected HVGs was printed for verification.  \n\n### **5. Visualization of Highly Variable Genes**  \nA scatter plot was generated to visualize mean expression versus dispersion for all genes, highlighting the selected HVGs. This step served as a quality control measure to confirm the appropriateness of the gene selection.  \n\n### **6. Data Output**  \nThe processed **AnnData** object, containing normalized data and HVG annotations, was saved in `.h5ad` format (`pbmc3k_hvg_selected.h5ad`) for subsequent analyses (e.g., scaling, dimensionality reduction).  \n\nThis workflow provides a reproducible pipeline for preprocessing single-cell RNA-seq data, emphasizing normalization, feature selection, and visualization."
        ],
        [
         "9",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Import necessary libraries for single-cell analysis and data manipulation\n\nlibrary(Seurat) # Main package for single-cell analysis\n\nlibrary(dplyr) # For data manipulation functions\n\nlibrary(ggplot2) # For advanced plotting capabilities\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"FeatureSelectionDemo\",\n\nmin.cells = 3, # Keep genes expressed in at least 3 cells\n\nmin.features = 200 # Keep cells with at least 200 detected genes\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing and Normalization\n\n#--------------------------------\n\n\n\n# OPTIONAL QC: Calculate the percentage of mitochondrial genes for each cell.\n\n# This is a common QC metric; high percentages can indicate cell stress.\n\nseurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(seurat_obj, pattern = \"^MT-\")\n\n\n\n# Normalize the data using the \"LogNormalize\" method.\n\n# This scales the data by library size and applies a log transformation.\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Feature Selection\n\n#--------------------------------\n\n\n\n# Identify highly variable features (genes) using the variance-stabilizing transformation (vst) method.\n\n# These genes show high cell-to-cell variation and are used for downstream analyses like PCA.\n\nseurat_obj <- FindVariableFeatures(\n\nseurat_obj,\n\nselection.method = \"vst\",\n\nnfeatures = 2000 # Select the top 2000 most variable genes\n\n)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Identify the top 10 most highly variable genes for labeling purposes.\n\ntop10 <- head(VariableFeatures(seurat_obj), 10)\n\n\n\n# Create a variable feature plot to visualize gene variability.\n\n# The plot shows the relationship between average expression and a measure of dispersion for each gene.\n\nvariable_feature_plot <- VariableFeaturePlot(seurat_obj)\n\n\n\n# Add labels for the top 10 HVGs to the plot for better interpretation.\n\nlabeled_plot <- LabelPoints(plot = variable_feature_plot, points = top10, repel = TRUE)\n\n\n\n# Print the final plot to the console.\n\nprint(labeled_plot)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the Seurat object, which now includes the identified variable features.\n\n# This object can be used for the next steps in the analysis, such as scaling and clustering.\n\nsaveRDS(seurat_obj, file = \"FeatureSelection_SeuratObject.rds\")",
         "**Methods**  \n\n### **Software and Package Initialization**  \nThe analysis was conducted in R (version ≥4.0) using the following packages:  \n- **Seurat** (v4.0+) for single-cell RNA sequencing (scRNA-seq) data processing, normalization, and feature selection.  \n- **dplyr** for data manipulation (e.g., filtering, aggregation).  \n- **ggplot2** for advanced visualization.  \n\nThese packages were selected for their established utility in scRNA-seq workflows: Seurat provides specialized functions for single-cell analysis, while dplyr and ggplot2 enable efficient data handling and plotting.  \n\n---  \n\n### **Data Input and Initial Filtering**  \nRaw gene expression data were obtained from 10X Genomics Cell Ranger output (filtered count matrices in the `hg19` reference genome). The data were loaded using `Read10X()`, which parses the directory structure (`data.dir = \"path/to/filtered_gene_bc_matrices/hg19/\"`) to generate a sparse matrix of gene-cell counts.  \n\nA **Seurat object** was created from the raw counts with the following quality control (QC) thresholds:  \n- **`min.cells = 3`**: Only genes detected in ≥3 cells were retained to exclude lowly expressed features.  \n- **`min.features = 200`**: Only cells with ≥200 detected genes were included to remove potential empty droplets or damaged cells.  \n\nThe object was assigned a project identifier (`project = \"FeatureSelectionDemo\"`) for metadata tracking.  \n\n---  \n\n### **Quality Control and Normalization**  \nMitochondrial gene content was quantified (`PercentageFeatureSet()`) using the pattern `^MT-` to identify transcripts encoding mitochondrial proteins. High proportions of mitochondrial reads (>10–20%) often indicate compromised cell viability but were not explicitly filtered here.  \n\nData were normalized using **log-normalization** (`NormalizeData()`):  \n- **`normalization.method = \"LogNormalize\"`**: Counts were scaled by total cellular library size and log-transformed (log1p).  \n- **`scale.factor = 10000`**: Counts were multiplied by 10,000 to avoid very small values (transcripts per 10,000).  \n\n---  \n\n### **Feature Selection**  \nHighly variable genes (HVGs) were identified using the **variance-stabilizing transformation (vst)** method (`FindVariableFeatures()`), which models mean-variance relationships to select genes with high cell-to-cell variation:  \n- **`selection.method = \"vst\"`**: Fits a trend line to the mean-variance relationship and ranks genes by residual variance.  \n- **`nfeatures = 2000`**: The top 2,000 most variable genes were retained for downstream analyses (e.g., PCA, clustering).  \n\n---  \n\n### **Visualization**  \nA **variable feature plot** (`VariableFeaturePlot()`) was generated to visualize the relationship between mean expression and dispersion (variance) for all genes. The top 10 HVGs (`head(VariableFeatures(), 10)`) were labeled on the plot using `LabelPoints()` with `repel = TRUE` to avoid overlapping text.  \n\n---  \n\n### **Output**  \nThe final Seurat object, containing normalized data and HVGs, was saved as an RDS file (`saveRDS()`) for subsequent steps (e.g., scaling, dimensionality reduction, clustering).  \n\n---  \n\n**Key Parameters Summary**  \n| Step                  | Parameter           | Value/Purpose                                                                 |  \n|-----------------------|---------------------|------------------------------------------------------------------------------|  \n| Data filtering        | `min.cells`         | Retain genes in ≥3 cells                                                    |  \n|                       | `min.features`      | Retain cells with ≥200 genes                                                |  \n| Normalization         | `scale.factor`      | 10,000 (transcripts per 10,000)                                             |  \n| Feature selection     | `nfeatures`         | Top 2,000 HVGs                                                              |  \n| Visualization         | `repel` (LabelPoints) | Prevents label overlap                                                     |  \n\nThis workflow ensures reproducibility and transparency in preprocessing scRNA-seq data for downstream analysis."
        ],
        [
         "10",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport scanpy as sc\n\nimport anndata as ad\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC 3k dataset from Scanpy as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization and Feature Selection ---\n\n# Normalize total counts per cell to account for differences in sequencing depth\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to stabilize the variance\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs), which are most informative for downstream analysis\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# --- Scaling and PCA ---\n\n# Subset the AnnData object to keep only the highly variable genes\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# Scale the data so that each gene has zero mean and unit variance.\n\n# This is crucial for PCA and other algorithms sensitive to feature scales.\n\nsc.pp.scale(adata)\n\n\n\n# Perform Principal Component Analysis (PCA) on the scaled data.\n\n# This reduces the dimensionality of the data while retaining most of the variation.\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# --- UMAP Embedding ---\n\n# Compute the neighborhood graph of cells. This is based on the PCA representation\n\n# and is a prerequisite for UMAP.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n\n\n\n# Run Uniform Manifold Approximation and Projection (UMAP) to generate a\n\n# non-linear 2D embedding for visualization.\n\nsc.tl.umap(adata)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the PCA variance ratio to see how much variance is captured by each principal component.\n\n# The log scale helps visualize the \"elbow\" where the explained variance plateaus.\n\nsc.pl.pca_variance_ratio(adata, log=True)\n\n\n\n# Create a scatter plot of the first two principal components.\n\n# Cells can be colored by marker gene expression to identify potential cell types.\n\nsc.pl.pca(adata, color=['CST3', 'NKG7'])\n\n\n\n# Visualize the UMAP embedding.\n\n# Coloring by marker genes helps to see how cell clusters are separated in the UMAP space.\n\nsc.pl.umap(adata, color='CST3')\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed AnnData object, which now contains PCA and UMAP coordinates.\n\n# This file is ready for downstream clustering and cell type annotation.\n\nadata.write(\"pbmc3k_pca_umap.h5ad\")\n\n",
         "### **Methods**  \n\n#### **Single-Cell RNA-Seq Analysis Pipeline**  \n\nThe computational workflow was implemented in **Python** using the **Scanpy** and **AnnData** libraries for single-cell data analysis. Below is a detailed description of each step.  \n\n#### **1. Library Import and Configuration**  \n- **Scanpy (`sc`)** and **AnnData (`ad`)** were imported for single-cell data processing and storage.  \n- **Matplotlib (`plt`)** was included for visualization.  \n- Scanpy’s default plotting parameters were configured for consistent visuals (`dpi=80`, `facecolor='white'`).  \n\n#### **2. Data Input**  \n- A publicly available **3k PBMC (Peripheral Blood Mononuclear Cells) dataset** was loaded directly from Scanpy’s built-in datasets as an `AnnData` object (`adata`).  \n\n#### **3. Preprocessing and Dimensionality Reduction**  \n\n##### **3.1. Normalization and Feature Selection**  \n- **Total counts normalization** was performed (`target_sum=1e4`) to adjust for differences in sequencing depth between cells.  \n- A **log(1+x) transformation** was applied to stabilize variance across genes.  \n- **Highly variable genes (HVGs)** were identified (`n_top_genes=2000`, `flavor='seurat'`) to focus downstream analysis on the most biologically informative features.  \n\n##### **3.2. Scaling and Principal Component Analysis (PCA)**  \n- The dataset was subset to retain only the 2,000 HVGs.  \n- **Scaling** was applied (zero mean, unit variance per gene) to ensure PCA was not biased by feature magnitude.  \n- **PCA** was computed (`svd_solver='arpack'`), reducing dimensionality while preserving maximum variance.  \n\n##### **3.3. UMAP Embedding**  \n- A **nearest-neighbor graph** was constructed (`n_neighbors=10`, `n_pcs=40`) to model cell-cell relationships in PCA space.  \n- **UMAP** (Uniform Manifold Approximation and Projection) was applied to generate a 2D embedding for visualization.  \n\n#### **4. Visualization**  \n- The **variance explained by each principal component** was plotted on a log scale to assess dimensionality reduction efficacy.  \n- **PCA scatter plots** were generated, coloring cells by marker genes (e.g., *CST3*, *NKG7*) to explore potential cell-type associations.  \n- **UMAP projections** were visualized, with cells colored by *CST3* expression to evaluate cluster separation.  \n\n#### **5. Data Output**  \n- The processed `AnnData` object, containing normalized counts, HVGs, PCA, and UMAP coordinates, was saved to an H5AD file (`pbmc3k_pca_umap.h5ad`) for downstream clustering and annotation.  \n\nThis pipeline provides a reproducible workflow for single-cell RNA-seq analysis, from raw data preprocessing to exploratory visualization."
        ],
        [
         "11",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis and plotting\nimport scanpy as sc\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\nsc.settings.verbosity = 0\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the preprocessed AnnData object. This file should have undergone\n# prior QC, normalization, and feature selection.\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"s4d8_feature_selection.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/40016014\"\n)\n\n#--------------------------------\n# Major Analysis Tasks: Dimensionality Reduction\n#--------------------------------\n\n# Ensure the analysis uses the log-normalized data stored in a specific layer.\n# This step makes the choice of data layer explicit.\nadata.X = adata.layers[\"log1p_norm\"]\n\n# --- Principal Component Analysis (PCA) ---\n# Perform PCA on the highly variable genes to capture the main axes of variation.\nsc.pp.pca(adata, svd_solver=\"arpack\", use_highly_variable=True)\n\n# --- t-SNE & UMAP Prerequisite: Neighborhood Graph ---\n# Compute the neighborhood graph, which is required for both t-SNE and UMAP.\n# This is calculated on the PCA representation of the data.\nsc.pp.neighbors(adata, n_neighbors=15, n_pcs=40)\n\n# --- t-distributed Stochastic Neighbor Embedding (t-SNE) ---\n# Run t-SNE to generate a 2D embedding for visualization.\n# It uses the PCA representation (`use_rep=\"X_pca\"`) as input.\nsc.tl.tsne(adata, use_rep=\"X_pca\")\n\n# --- Uniform Manifold Approximation and Projection (UMAP) ---\n# Compute the UMAP embedding based on the neighborhood graph.\nsc.tl.umap(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the PCA result, coloring cells by total counts to check for technical artifacts.\nsc.pl.pca_scatter(adata, color=\"total_counts\", title=\"PCA: First 2 Components\")\n\n# Visualize the t-SNE embedding, also colored by total counts.\nsc.pl.tsne(adata, color=\"total_counts\", title=\"t-SNE Embedding\")\n\n# Visualize the UMAP embedding, colored by total counts.\nsc.pl.umap(adata, color=\"total_counts\", title=\"UMAP Embedding\")\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the final AnnData object, which now contains the results of PCA, t-SNE, and UMAP.\n# This object is ready for clustering and cell type annotation.\nadata.write(\"s4d8_dimensionality_reduction.h5ad\")\n",
         "### **Methods**  \n\n#### **Single-Cell RNA-Seq Data Analysis: Dimensionality Reduction and Visualization**  \n\nThe following steps outline the computational workflow for dimensionality reduction and visualization of single-cell RNA sequencing (scRNA-seq) data using Python and the `scanpy` library.  \n\n##### **1. Module Import and Configuration**  \nThe analysis begins by importing essential Python packages:  \n- **`scanpy` (sc)** for single-cell data processing and analysis.  \n- **`seaborn` (sns)** and **`matplotlib.pyplot` (plt)** for visualization.  \n\nGlobal parameters in `scanpy` are configured to optimize figure generation:  \n- Figure resolution (**`dpi=80`**) and background (**`facecolor=\"white\"`**) are standardized.  \n- Verbosity is suppressed (**`verbosity=0`**) to minimize console output.  \n\n##### **2. Data Loading**  \nThe input data is an **`AnnData`** object (stored as *s4d8_feature_selection.h5ad*), which has undergone prior quality control (QC), normalization, and feature selection. If the local file is unavailable, it is retrieved from a backup URL (*https://figshare.com/ndownloader/files/40016014*).  \n\n##### **3. Dimensionality Reduction**  \nAll computations are performed on **log-normalized counts** (stored in the `adata.layers[\"log1p_norm\"]` layer).  \n\n###### **Principal Component Analysis (PCA)**  \nPCA is applied to capture major axes of variation using **highly variable genes** (selected in preprocessing). Key parameters:  \n- **`svd_solver=\"arpack\"`**: Ensures computational efficiency.  \n- **`use_highly_variable=True`**: Restricts PCA to genes with high cell-to-cell variability.  \n\n###### **Neighborhood Graph Construction**  \nA **k-nearest neighbor (KNN) graph** is computed on the PCA-reduced data to facilitate nonlinear embeddings. Parameters:  \n- **`n_neighbors=15`**: Balances local and global structure.  \n- **`n_pcs=40`**: Uses the top 40 principal components as input.  \n\n###### **Nonlinear Embeddings**  \n- **t-SNE**: A 2D embedding is generated using the PCA representation (**`use_rep=\"X_pca\"`**) to visualize global structure.  \n- **UMAP**: Computed from the neighborhood graph, providing an alternative low-dimensional representation.  \n\n##### **4. Visualization**  \nEmbeddings are visualized to assess technical biases (e.g., library size effects):  \n- **PCA scatterplot**, **t-SNE**, and **UMAP** are plotted with cells colored by **total counts** (`total_counts`) to detect potential artifacts.  \n\n##### **5. Data Output**  \nThe final **`AnnData`** object, now containing PCA, t-SNE, and UMAP results, is saved (*s4d8_dimensionality_reduction.h5ad*) for downstream clustering and annotation.  \n\nThis workflow ensures reproducible dimensionality reduction and visualization of scRNA-seq data, adhering to standard practices in single-cell genomics."
        ],
        [
         "12",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell analysis and plotting\n\nlibrary(Seurat) # Main package for single-cell workflows\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"DimRed_DimDetermination\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing and Normalization\n\n#--------------------------------\n\n\n\n# 1. Normalize the data using the \"LogNormalize\" method.\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n# 2. Identify highly variable features (HVGs).\n\nseurat_obj <- FindVariableFeatures(\n\nseurat_obj,\n\nselection.method = \"vst\",\n\nnfeatures = 2000\n\n)\n\n\n\n# 3. Scale the data. This shifts the expression of each gene so that the mean\n\n# expression across cells is 0 and the variance is 1.\n\nall_genes <- rownames(seurat_obj)\n\nseurat_obj <- ScaleData(seurat_obj, features = all_genes)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Principal Component Analysis (PCA) ---\n\n# Run PCA on the scaled data, using only the highly variable features.\n\nseurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj))\n\n\n\n# --- UMAP Embedding ---\n\n# Note: Choosing the number of PCs is a critical step. The Elbow Plot in the\n\n# visualization section helps with this decision. Here, we use 10 as an example.\n\n# First, compute the neighborhood graph based on the selected PCs.\n\nseurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)\n\n\n\n# Then, run UMAP to get a non-linear embedding for visualization.\n\nseurat_obj <- RunUMAP(seurat_obj, dims = 1:10)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- PCA Plot ---\n\n# Visualize the first two principal components. Coloring by 'nCount_RNA' can help\n\n# reveal if technical factors (like library size) are driving the main sources of variation.\n\npca_plot <- DimPlot(seurat_obj, reduction = \"pca\") +\n\nggtitle(\"PCA: PC1 vs PC2\")\n\n\n\n# --- Elbow Plot for Dimensionality Determination ---\n\n# An Elbow Plot helps to determine the number of significant PCs to use for downstream\n\n# analysis. The \"elbow\" of the plot represents the point of diminishing returns,\n\n# where adding more PCs explains little additional variance.\n\nelbow_plot <- ElbowPlot(seurat_obj, ndims = 50) +\n\nggtitle(\"Elbow Plot: Variance Explained vs. PC\")\n\n\n\n# --- UMAP Plot ---\n\n# Visualize the UMAP embedding, which is often used for identifying cell clusters.\n\numap_plot <- DimPlot(seurat_obj, reduction = \"umap\") +\n\nggtitle(\"UMAP Embedding (Using PCs 1-10)\")\n\n\n\n# --- Combined Plot ---\n\n# Use patchwork to arrange the plots for a comprehensive overview.\n\n(pca_plot | elbow_plot) / umap_plot\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed Seurat object. It now contains the scaled data,\n\n# PCA results, and UMAP embedding, ready for clustering and annotation.\n\nsaveRDS(seurat_obj, file = \"DimRed_DimDetermined_SeuratObject.rds\")\n\n",
         "**Methods**  \n\n**Single-cell RNA-seq Data Processing and Dimensionality Reduction**  \n\n**Library and Package Loading**  \nThe analysis was conducted using the `Seurat` package (v4) in R for single-cell RNA-seq data processing, normalization, and dimensionality reduction. Additional visualization packages included `ggplot2` for custom plotting and `patchwork` for arranging multiple plots.  \n\n**Data Input and Initial Filtering**  \nRaw gene expression data in 10X Genomics format (filtered gene-barcode matrices) were loaded using `Read10X()`, and a `Seurat` object was generated with the following filtering criteria:  \n- **`min.cells = 3`**: Only genes detected in at least 3 cells were retained.  \n- **`min.features = 200`**: Only cells expressing at least 200 genes were included to remove low-quality cells.  \n\n**Preprocessing and Normalization**  \n1. **Normalization**: Gene expression counts were log-normalized (`LogNormalize`) with a scaling factor of 10,000 to correct for sequencing depth differences.  \n2. **Highly Variable Feature Selection**: The top 2,000 variable genes were identified using the variance-stabilizing transformation (`vst`) method to focus downstream analysis on biologically relevant features.  \n3. **Scaling**: Expression values were scaled (mean-centered and variance-adjusted) to enable PCA-based comparisons.  \n\n**Dimensionality Reduction**  \n1. **Principal Component Analysis (PCA)** was performed using the scaled, highly variable features to reduce data dimensionality.  \n2. **UMAP Embedding**: A neighborhood graph was computed using the first 10 principal components (PCs), followed by non-linear UMAP projection for visualization.  \n\n**Visualization**  \n1. **PCA Plot**: The first two PCs were visualized to assess technical biases (e.g., library size effects).  \n2. **Elbow Plot**: Variance explained by each PC was plotted to guide PC selection (50 PCs evaluated).  \n3. **UMAP Plot**: The UMAP embedding was generated to explore potential cell clusters.  \n4. **Combined Plot**: PCA, elbow, and UMAP plots were arranged using `patchwork` for comparative analysis.  \n\n**Data Output**  \nThe processed `Seurat` object, containing normalized counts, PCA, and UMAP results, was saved as an `.rds` file for subsequent clustering and annotation.  \n\n---  \n*Note*: Replace `\"path/to/filtered_gene_bc_matrices/hg19/\"` with the actual data directory path. All steps were performed in R (v4.0+)."
        ],
        [
         "13",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import the Scanpy library for single-cell analysis\n\nimport scanpy as sc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC 3k dataset from Scanpy as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# It's good practice to keep a copy of the raw data.\n\n# The .raw attribute is the conventional place to store it in Scanpy.\n\nadata.raw = adata.copy()\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization, Feature Selection, and Scaling ---\n\n# Normalize total counts per cell to account for sequencing depth differences\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to stabilize the variance\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) for downstream analysis\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# Subset the AnnData object to keep only the highly variable genes\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# Scale the data so that each gene has zero mean and unit variance\n\nsc.pp.scale(adata)\n\n\n\n# --- PCA and Neighborhood Graph ---\n\n# Perform Principal Component Analysis (PCA) on the scaled data\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# Compute the neighborhood graph of cells based on the PCA representation.\n\n# This graph is the basis for both clustering and UMAP embedding.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n\n\n\n# --- UMAP for Visualization ---\n\n# Generate a UMAP embedding from the neighbors graph. This is for visualization purposes.\n\nsc.tl.umap(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Clustering\n\n#--------------------------------\n\n\n\n# Apply the Leiden algorithm to find communities (clusters) in the neighborhood graph.\n\n# The `flavor='igraph'` and a fixed number of iterations are used for performance and reproducibility.\n\nsc.tl.leiden(adata, flavor='igraph', n_iterations=2)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the UMAP embedding, with cells colored by their assigned Leiden cluster.\n\n# This plot allows for a visual assessment of the clustering results.\n\nsc.pl.umap(adata, color=['leiden'], title=\"Leiden Clustering on PBMC3k\", size=20)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed AnnData object. It now contains the clustering results\n\n# in `adata.obs['leiden']` and is ready for cell type annotation.\n\nadata.write(\"pbmc3k_clustered.h5ad\")\n\n",
         "### Methods  \n\n#### Single-Cell RNA-Seq Analysis Using Scanpy  \n\n**Data Loading and Initial Setup**  \nWe performed single-cell RNA-seq analysis using the **Scanpy** library (v1.9.0+) in Python. To ensure reproducibility, we configured Scanpy's plotting parameters (`dpi=80`, `facecolor='white'`) for consistent and clear visualizations. The analysis began with the **PBMC3k dataset**, a publicly available peripheral blood mononuclear cell (PBMC) dataset bundled with Scanpy for demonstration purposes. The raw data was preserved in the `.raw` attribute of the `AnnData` object to retain unprocessed counts for downstream comparisons.  \n\n**Preprocessing and Dimensionality Reduction**  \nThe data underwent standard preprocessing:  \n1. **Normalization**: Total counts per cell were scaled to **10,000 counts (CP10K)** to correct for differences in sequencing depth.  \n2. **Log Transformation**: A **log(1+x)** transformation was applied to stabilize gene expression variance.  \n3. **Feature Selection**: **2,000 highly variable genes (HVGs)** were identified using the *Seurat* flavor (`flavor='seurat'`) to focus downstream analysis on biologically relevant features.  \n4. **Scaling**: Data was scaled to **zero mean and unit variance** per gene to mitigate technical biases.  \n\nPrincipal Component Analysis (PCA) was performed using the **ARPACK solver** (`svd_solver='arpack'`) on the scaled HVGs. A **neighborhood graph** was constructed from the top **40 PCs** with **10 nearest neighbors** (`n_neighbors=10`, `n_pcs=40`) to capture cell-cell relationships.  \n\n**Clustering and Visualization**  \nCell clusters were identified via the **Leiden algorithm** (`flavor='igraph'`, `n_iterations=2`), which partitions the neighborhood graph into communities. A **UMAP embedding** was generated for visualization, and cells were colored by their Leiden cluster assignments.  \n\n**Output**  \nThe processed `AnnData` object, including cluster labels (`adata.obs['leiden']`), was saved in **H5AD format** (`pbmc3k_clustered.h5ad`) for further annotation or analysis.  \n\nThis pipeline ensures standardized preprocessing, clustering, and visualization of single-cell data, following best practices for reproducibility and interpretability.  \n\n---  \n*Key Parameters*:  \n- HVG selection: `n_top_genes=2000`, `flavor='seurat'`  \n- PCA/UMAP: `n_pcs=40`, `n_neighbors=10`  \n- Clustering: `flavor='igraph'`, `n_iterations=2`"
        ],
        [
         "14",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis and plotting\nimport scanpy as sc\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\nsc.settings.verbosity = 0\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the preprocessed AnnData object. This file should contain\n# dimensionality reduction results (e.g., PCA).\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"s4d8_dimensionality_reduction.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/40016014\"\n)\n\n#--------------------------------\n# Major Analysis Task: Clustering\n#--------------------------------\n\n# --- Neighborhood Graph ---\n# Compute the neighborhood graph of cells based on the PCA representation.\n# This graph is the foundation for both clustering and UMAP.\nsc.pp.neighbors(adata, n_neighbors=15, n_pcs=30)\n\n# --- Leiden Clustering ---\n# Apply the Leiden algorithm to partition the neighborhood graph into clusters.\n# The 'resolution' parameter controls the number of clusters found; higher\n# values lead to more, smaller clusters.\nsc.tl.leiden(adata, resolution=1.0, key_added=\"leiden\")\n\n# --- UMAP for Visualization ---\n# Compute the UMAP embedding from the neighborhood graph. This is done\n# here to ensure it's available for plotting, in case it wasn't run previously.\nsc.tl.umap(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, with cells colored by their assigned Leiden cluster.\n# The `legend_loc='on data'` option places cluster labels directly on the plot.\nsc.pl.umap(adata, color=[\"leiden\"], title=\"Leiden Clustering\", legend_loc='on data')\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the updated AnnData object. It now contains the clustering results\n# stored in `adata.obs['leiden']` and is ready for marker gene analysis.\nadata.write(\"s4d8_clustering.h5ad\")\n",
         "**Methods**\n\n**Single-Cell RNA-Seq Clustering Analysis Using Scanpy in Python**  \n\nThe following steps were performed for clustering and visualization of single-cell RNA-seq data using the Scanpy toolkit in Python.  \n\n1. **Module Import and Configuration**  \n   - The `scanpy` library (sc) was imported for single-cell analysis, and `matplotlib.pyplot` was imported for plotting. Scanpy’s global settings were configured to suppress verbose output (`verbosity=0`) and to adjust figure parameters (`dpi=80`, `facecolor=\"white\"`, `frameon=False`) for clear visualization.  \n\n2. **Data Input**  \n   - A preprocessed `AnnData` object (`s4d8_dimensionality_reduction.h5ad`) was loaded, containing dimensionality reduction results (e.g., PCA). A backup URL was specified to download the file if local access failed.  \n\n3. **Neighborhood Graph Construction**  \n   - A k-nearest neighbors (KNN) graph was computed using PCA-reduced dimensions (`n_pcs=30`) to capture local cell relationships. The parameter `n_neighbors=15` determined the number of neighboring cells considered for each cell, balancing local and global structure.  \n\n4. **Leiden Clustering**  \n   - The Leiden graph-based clustering algorithm was applied to partition cells into distinct groups. The `resolution` parameter was set to `1.0` to control cluster granularity, with higher values yielding more clusters. Cluster assignments were stored in `adata.obs['leiden']`.  \n\n5. **UMAP Visualization**  \n   - A UMAP embedding was computed from the neighborhood graph to visualize cellular heterogeneity in two dimensions. Cells were colored by Leiden cluster membership, with cluster labels overlaid directly on the plot (`legend_loc='on data'`) for clarity.  \n\n6. **Data Output**  \n   - The updated `AnnData` object, now including clustering results and UMAP coordinates, was saved to `s4d8_clustering.h5ad` for downstream analyses (e.g., marker gene identification).  \n\nThis workflow enables reproducible identification and visualization of cell populations in single-cell RNA-seq datasets."
        ],
        [
         "15",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell analysis, data manipulation, and plotting\n\nlibrary(Seurat) # Main package for single-cell RNA-seq analysis\n\nlibrary(dplyr) # For data manipulation functions\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"ClusteringDemo\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing, QC, and Normalization\n\n#--------------------------------\n\n\n\n# Calculate the percentage of mitochondrial genes for each cell as a QC metric.\n\nseurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(seurat_obj, pattern = \"^MT-\")\n\n\n\n# Normalize the data to account for differences in library size.\n\nseurat_obj <- NormalizeData(\n\nseurat_obj,\n\nnormalization.method = \"LogNormalize\",\n\nscale.factor = 10000\n\n)\n\n\n\n# Identify highly variable features (genes) to be used for downstream analysis.\n\nseurat_obj <- FindVariableFeatures(\n\nseurat_obj,\n\nselection.method = \"vst\",\n\nnfeatures = 2000\n\n)\n\n\n\n# Scale the data, which is a standard step before running PCA.\n\nseurat_obj <- ScaleData(seurat_obj, features = rownames(seurat_obj))\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Dimensionality Reduction & Clustering\n\n#--------------------------------\n\n\n\n# --- PCA ---\n\n# Run Principal Component Analysis on the highly variable features.\n\nseurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj))\n\n\n\n# --- Clustering ---\n\n# Build a K-nearest neighbor (KNN) graph based on the PCA space.\n\n# The number of PCs (dims) used is a critical parameter.\n\nseurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)\n\n\n\n# Apply graph-based clustering to partition the cells.\n\n# The `resolution` parameter controls the number of clusters identified.\n\nseurat_obj <- FindClusters(seurat_obj, resolution = 0.5)\n\n\n\n# --- UMAP for Visualization ---\n\n# Run UMAP to generate a non-linear 2D embedding for visualization.\n\nseurat_obj <- RunUMAP(seurat_obj, dims = 1:10)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize QC metrics with violin plots.\n\nVlnPlot(seurat_obj, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3)\n\n\n\n# Visualize the PCA results.\n\nDimPlot(seurat_obj, reduction = \"pca\") + ggtitle(\"PCA Plot\")\n\n\n\n# Visualize the UMAP embedding with cells colored by their assigned cluster.\n\nDimPlot(seurat_obj, reduction = \"umap\", label = TRUE, pt.size = 0.8) +\n\nggtitle(\"UMAP Plot with Cluster Labels\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the processed Seurat object, which now contains clustering results.\n\n# This object is ready for downstream marker analysis and cell type annotation.\n\nsaveRDS(seurat_obj, file = \"Clustering_SeuratObject.rds\")\n\n",
         "### **Methods: Single-Cell RNA-Seq Analysis Using Seurat**  \n\n#### **1. Software and Package Loading**  \nThe analysis was conducted in **R** using the following packages:  \n- **`Seurat`** (v5.0.0): Core package for single-cell RNA-seq data processing, normalization, dimensionality reduction, and clustering.  \n- **`dplyr`**: Used for efficient data manipulation and filtering.  \n- **`ggplot2`** and **`patchwork`**: Employed for generating high-quality visualizations and arranging multiple plots.  \n\n#### **2. Data Input and Initial Processing**  \nRaw single-cell RNA-seq data in **10X Genomics format** was loaded from the directory `path/to/filtered_gene_bc_matrices/hg19/` (user must specify the correct path). The count matrices were imported using `Read10X()`, which parses the filtered feature-barcode matrix.  \n\nA **Seurat object** was created with the following parameters:  \n- **`min.cells = 3`**: Only genes detected in at least **3 cells** were retained to reduce noise.  \n- **`min.features = 200`**: Only cells expressing at least **200 genes** were included to filter out low-quality cells.  \n\n#### **3. Quality Control (QC), Normalization, and Feature Selection**  \n- **Mitochondrial Gene Percentage Calculation**: Cells with high mitochondrial gene expression (indicative of poor quality) were identified using `PercentageFeatureSet()` (pattern: `\"^MT-\"`).  \n- **Log-Normalization**: Data was normalized using `NormalizeData()` with a **scale factor of 10,000**, accounting for differences in sequencing depth.  \n- **Highly Variable Gene Identification**: The **2,000 most variable genes** were selected (`FindVariableFeatures()`, method: `\"vst\"`) for downstream analysis.  \n- **Data Scaling**: Expression values were scaled (`ScaleData()`) to facilitate principal component analysis (PCA).  \n\n#### **4. Dimensionality Reduction and Clustering**  \n- **Principal Component Analysis (PCA)** was performed on the highly variable genes (`RunPCA()`).  \n- **K-Nearest Neighbor (KNN) Graph Construction**: A graph was built using **the first 10 principal components** (`FindNeighbors()`).  \n- **Graph-Based Clustering**: Cells were partitioned using the **Leiden algorithm** (`FindClusters()`, resolution = **0.5**).  \n- **UMAP Embedding**: A 2D non-linear projection was generated (`RunUMAP()`, dims = **1:10**) for visualization.  \n\n#### **5. Visualization**  \n- **QC Metrics**: Violin plots (`VlnPlot()`) displayed distributions of **genes per cell (`nFeature_RNA`), counts per cell (`nCount_RNA`), and mitochondrial gene percentage (`percent.mt`)**.  \n- **PCA and UMAP Plots**:  \n  - `DimPlot()` (reduction = `\"pca\"`) visualized the PCA results.  \n  - `DimPlot()` (reduction = `\"umap\"`, `label = TRUE`) displayed clustering results in UMAP space.  \n\n#### **6. Data Output**  \nThe processed **Seurat object**, containing normalized data, clustering assignments, and embeddings, was saved as an `.rds` file (`Clustering_SeuratObject.rds`) for downstream analysis (e.g., differential expression or cell type annotation).  \n\nThis pipeline provides a **reproducible workflow** for single-cell RNA-seq clustering, adhering to best practices in data preprocessing, dimensionality reduction, and visualization."
        ],
        [
         "16",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and data handling\n\nimport scanpy as sc\n\nimport anndata as ad\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting to ensure consistent and clear visuals\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC 3k dataset from Scanpy as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n#--------------------------------\n\n# Preprocessing & Clustering\n\n#--------------------------------\n\n\n\n# --- Normalization, Feature Selection, and Scaling ---\n\n# Normalize total counts per cell to account for sequencing depth differences\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\n\n\n# Apply a log(1+x) transformation to stabilize the variance\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) for downstream analysis\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\n\n\n# Subset the AnnData object to keep only the highly variable genes\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# Scale the data so that each gene has zero mean and unit variance\n\nsc.pp.scale(adata)\n\n\n\n# --- PCA, Neighborhood Graph, and UMAP ---\n\n# Perform Principal Component Analysis (PCA) on the scaled data\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# Compute the neighborhood graph, which is the basis for clustering and UMAP\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=40)\n\n\n\n# Generate a UMAP embedding for visualization\n\nsc.tl.umap(adata)\n\n\n\n# --- Leiden Clustering ---\n\n# Apply the Leiden algorithm to find communities (clusters) of cells\n\nsc.tl.leiden(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Cell-Type Annotation\n\n#--------------------------------\n\n\n\n# Define a dictionary of canonical marker genes for common PBMC types.\n\n# These genes are used to identify cell types based on their expression patterns.\n\nmarker_genes = {\n\n'B cells': ['MS4A1', 'CD79A'],\n\n'T cells': ['CD3D', 'CD3E'],\n\n'NK cells': ['NKG7', 'GNLY'],\n\n'Monocytes': ['CD14', 'LYZ']\n\n}\n\n\n\n# Manually define a mapping from cluster IDs to cell type names.\n\n# NOTE: This mapping is created after inspecting the dot plot below and should be\n\n# adjusted based on the specific results of your analysis.\n\ncluster_to_cell_type = {\n\n'0': 'T cells',\n\n'1': 'Monocytes',\n\n'2': 'NK cells',\n\n'3': 'B cells',\n\n'4': 'T cells' # Example of merging multiple clusters into one cell type\n\n}\n\n\n\n# Create a new column in the observation metadata to store the cell type annotations.\n\n# The .map() function applies the dictionary to the 'leiden' column.\n\nadata.obs['cell_type'] = adata.obs['leiden'].map(cluster_to_cell_type).astype('category')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a dot plot to visualize the expression of marker genes across clusters.\n\n# This plot is essential for assigning cell types to clusters.\n\nsc.pl.dotplot(adata, marker_genes, groupby='leiden', standard_scale='var',\n\ntitle='Marker Expression by Leiden Clusters')\n\n\n\n# Visualize the final cell type annotations on the UMAP embedding.\n\n# This plot shows the spatial distribution of the identified cell types.\n\nsc.pl.umap(adata, color='cell_type', title='Cell Type Annotation', legend_loc='on data')\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the fully processed and annotated AnnData object.\n\n# This file can be used for further downstream analysis, such as differential\n\n# expression between cell types or cell-cell interaction studies.\n\nadata.write(\"pbmc3k_annotated.h5ad\")\n\n",
         "**Methods**\n\n**Data Processing and Single-Cell RNA-seq Analysis**  \n\nThe analysis was performed in Python using the *Scanpy* (v1.9.0) and *AnnData* libraries for single-cell RNA sequencing (scRNA-seq) data handling. Below is a step-by-step description of the computational workflow.\n\n### **1. Library Import and Configuration**  \nThe *Scanpy* and *AnnData* packages were loaded to enable scRNA-seq data analysis. Scanpy’s default plotting parameters were set to a resolution of 80 DPI with a white background to ensure visualization clarity.\n\n### **2. Data Input**  \nA publicly available peripheral blood mononuclear cell (PBMC) dataset (3k cells) was loaded directly from *Scanpy*’s example datasets as an *AnnData* object, which stores gene expression matrices alongside cell metadata.  \n\n### **3. Preprocessing and Clustering**  \n\n#### **Normalization and Feature Selection**  \n- **Normalization:** Raw counts were normalized per cell to a total sum of 10,000 (CP10K) to account for differences in sequencing depth.  \n- **Log Transformation:** A log(1+x) transformation was applied to stabilize variance.  \n- **Highly Variable Genes (HVGs):** The top 2,000 HVGs were selected using the *Seurat* flavor (v3 method) to focus downstream analysis on biologically informative genes.  \n- **Scaling:** Data were scaled to zero mean and unit variance (z-score) to facilitate dimensionality reduction.  \n\n#### **Dimensionality Reduction and Clustering**  \n- **Principal Component Analysis (PCA):** PCA was computed using the ARPACK solver (40 components retained).  \n- **Neighborhood Graph:** A k-nearest neighbors (KNN) graph (k=10) was constructed based on the top 40 PCs to capture local cell relationships.  \n- **UMAP Embedding:** A 2D UMAP projection was generated for visualization.  \n- **Leiden Clustering:** Cells were partitioned into clusters using the Leiden algorithm (resolution=1.0 by default), which identifies community structures in the KNN graph.  \n\n### **4. Cell-Type Annotation**  \n- **Marker Gene Selection:** Canonical marker genes (e.g., *MS4A1* for B cells, *CD3D* for T cells) were predefined to annotate clusters.  \n- **Manual Cluster-to-Cell-Type Mapping:** Based on marker gene expression (visualized via a dot plot), Leiden clusters were manually mapped to cell types (e.g., cluster 0 → T cells). A categorical column (`cell_type`) was added to the metadata.  \n\n### **5. Visualization**  \n- **Dot Plot:** Marker gene expression per cluster was visualized using a dot plot, scaled by gene (standard_scale='var').  \n- **UMAP Plot:** Final cell-type annotations were overlaid on the UMAP embedding, with labels positioned on the plot (`legend_loc='on data'`).  \n\n### **6. Data Output**  \nThe fully processed *AnnData* object, including normalized expression, HVGs, clusters, and annotations, was saved in H5AD format (`pbmc3k_annotated.h5ad`) for downstream use.  \n\n**Key Parameters**  \n- Normalization: `target_sum=1e4` (CP10K)  \n- HVG selection: `n_top_genes=2000`, `flavor='seurat'`  \n- PCA: `svd_solver='arpack'`, `n_pcs=40`  \n- Clustering: Default Leiden resolution (1.0)  \n\nThis workflow enables reproducible analysis of scRNA-seq data, from preprocessing to cell-type identification."
        ],
        [
         "17",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis and plotting\nimport scanpy as sc\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\nsc.settings.verbosity = 0\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\", frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the preprocessed and clustered AnnData object.\n# This file is expected to contain clustering results (e.g., in `adata.obs['leiden']`).\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"s4d8_clustering.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/40016014\"\n)\n\n#--------------------------------\n# Major Analysis Task: Cell-Type Annotation\n#--------------------------------\n\n# Define a dictionary to manually map cluster IDs to cell type names.\n# This mapping is based on prior biological knowledge or marker gene analysis.\nmanual_labels = {\n\"0\": \"B cell\",\n\"1\": \"T cell\",\n\"2\": \"Myeloid cell\",\n\"3\": \"NK cell\",\n\"4\": \"Dendritic cell\"\n}\n\n# Create a new column 'manual_annotation' in the observation metadata.\n# The .map() function applies the dictionary. The lambda function with .get()\n# ensures that any cluster ID not in the dictionary is assigned \"Unknown\".\nadata.obs[\"manual_annotation\"] = adata.obs[\"leiden\"].map(\nlambda x: manual_labels.get(x, \"Unknown\")\n)\n\n# Print a count of cells assigned to each annotated cell type for a quick summary.\nprint(\"Manual annotation summary:\")\nprint(adata.obs[\"manual_annotation\"].value_counts())\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Ensure a UMAP embedding exists for visualization.\n# This check prevents errors if the loaded object doesn't have UMAP coordinates.\nif \"X_umap\" not in adata.obsm.keys():\nsc.pp.neighbors(adata) # UMAP requires a neighbor graph\nsc.tl.umap(adata)\n\n# Visualize the UMAP embedding, with cells colored by their manual annotation.\n# This plot provides a spatial view of the identified cell populations.\nsc.pl.umap(\nadata,\ncolor=\"manual_annotation\",\ntitle=\"Manual Cell Type Annotation\",\nlegend_loc=\"on data\"\n)\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the updated AnnData object, which now includes the manual cell type annotations.\n# This file is ready for downstream analyses focused on specific cell populations.\nadata.write(\"s4d8_manual_annotation.h5ad\")\n",
         "**Methods**\n\n**Single-Cell RNA-Seq Data Processing and Cell-Type Annotation**\n\nThe following steps were performed to annotate cell types in a preprocessed single-cell RNA-seq dataset using Python and the Scanpy toolkit.  \n\n1. **Module Import**  \n   The analysis utilized the `scanpy` library for single-cell data manipulation and `matplotlib.pyplot` for visualization. Scanpy provides functions for clustering, dimensionality reduction, and annotation of single-cell data.  \n\n2. **Configuration**  \n   Global parameters were set to disable verbose logging (`sc.settings.verbosity = 0`) and optimize figure aesthetics (`dpi=80`, `facecolor=\"white\"`, `frameon=False`). These settings ensure clean and reproducible visualizations.  \n\n3. **Data Input**  \n   A preprocessed and clustered `AnnData` object (`s4d8_clustering.h5ad`) was loaded, containing cluster assignments (stored in `adata.obs['leiden']`). A backup URL was provided to fetch the data if the local file was unavailable.  \n\n4. **Cell-Type Annotation**  \n   Cluster identities were manually mapped to biologically defined cell types using a predefined dictionary:  \n   ```\n   \"0\": \"B cell\",  \n   \"1\": \"T cell\",  \n   \"2\": \"Myeloid cell\",  \n   \"3\": \"NK cell\",  \n   \"4\": \"Dendritic cell\"  \n   ```\n   Unmatched clusters were labeled as \"Unknown\". The annotation was stored in a new column (`adata.obs['manual_annotation']`), and cell counts per type were summarized for validation.  \n\n5. **Visualization**  \n   A UMAP embedding was generated (if absent) using a neighbor graph (`sc.pp.neighbors`) and UMAP (`sc.tl.umap`). Cells were plotted by annotated type (`sc.pl.umap`) with labels positioned on the UMAP for clarity.  \n\n6. **Data Output**  \n   The annotated dataset was saved (`s4d8_manual_annotation.h5ad`) for downstream analyses.  \n\nThis pipeline integrates manual biological knowledge with computational clustering to interpret single-cell populations, enabling subsequent cell-type-specific analyses."
        ],
        [
         "18",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell analysis, data manipulation, and plotting\n\nlibrary(Seurat) # Core package for single-cell RNA-seq analysis\n\nlibrary(dplyr) # For data manipulation, especially for handling marker gene tables\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Specify the directory containing the 10X Genomics count matrices.\n\n# NOTE: You must replace this with the actual path to your data.\n\ndata_dir <- \"path/to/filtered_gene_bc_matrices/hg19/\"\n\n\n\n# Read the 10X Genomics data into a raw count matrix.\n\nraw_data <- Read10X(data.dir = data_dir)\n\n\n\n# Create a Seurat object from the raw count matrix with basic initial filtering.\n\nseurat_obj <- CreateSeuratObject(\n\ncounts = raw_data,\n\nproject = \"ClusterBiomarkers\",\n\nmin.cells = 3,\n\nmin.features = 200\n\n)\n\n\n\n#--------------------------------\n\n# Preprocessing and QC\n\n#--------------------------------\n\n\n\n# Calculate the percentage of mitochondrial genes for each cell as a QC metric.\n\nseurat_obj[[\"percent.mt\"]] <- PercentageFeatureSet(seurat_obj, pattern = \"^MT-\")\n\n\n\n# Normalize the data to account for differences in library size.\n\nseurat_obj <- NormalizeData(seurat_obj, normalization.method = \"LogNormalize\", scale.factor = 10000)\n\n\n\n# Identify highly variable features (genes) for downstream analysis.\n\nseurat_obj <- FindVariableFeatures(seurat_obj, selection.method = \"vst\", nfeatures = 2000)\n\n\n\n# Scale the data, a standard step before running PCA.\n\nall_genes <- rownames(seurat_obj)\n\nseurat_obj <- ScaleData(seurat_obj, features = all_genes)\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# --- Dimensionality Reduction & Clustering ---\n\n# Run PCA on the highly variable features.\n\nseurat_obj <- RunPCA(seurat_obj, features = VariableFeatures(seurat_obj))\n\n\n\n# Build a K-nearest neighbor (KNN) graph and perform clustering.\n\nseurat_obj <- FindNeighbors(seurat_obj, dims = 1:10)\n\nseurat_obj <- FindClusters(seurat_obj, resolution = 0.5)\n\n\n\n# Run UMAP to generate a 2D embedding for visualization.\n\nseurat_obj <- RunUMAP(seurat_obj, dims = 1:10)\n\n\n\n\n\n# --- Finding Cluster Biomarkers ---\n\n# Find differentially expressed features (cluster biomarkers) for every cluster compared to all remaining cells.\n\n# `only.pos = TRUE` ensures we only get positive markers (genes upregulated in the cluster).\n\ncluster_markers <- FindAllMarkers(seurat_obj, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)\n\n\n\n\n\n# --- Cell Type Annotation ---\n\n# Manually define new cell type names based on canonical markers found in the literature\n\n# and observed in the `cluster_markers` table.\n\n# NOTE: The order and names must be adjusted based on your specific dataset and clustering results.\n\nnew.cluster.ids <- c(\"Naive CD4 T\", \"CD14+ Monocytes\", \"Memory CD4 T\", \"B cell\",\n\n\"CD8 T\", \"NK cell\", \"Dendritic cell\", \"Platelet\")\n\n\n\n# Assign the new names to the cluster levels in the Seurat object.\n\nnames(new.cluster.ids) <- levels(seurat_obj)\n\nseurat_obj <- RenameIdents(seurat_obj, new.cluster.ids)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the initial UMAP clusters before annotation.\n\nDimPlot(seurat_obj, reduction = \"umap\", label = TRUE, pt.size = 0.8) +\n\nggtitle(\"UMAP Plot: Clusters before Cell Type Assignment\")\n\n\n\n# --- Biomarker Visualization ---\n\n# Extract the top markers for each cluster to visualize on a heatmap.\n\n# Here, we group by cluster and select the top 2 genes based on log2 fold change.\n\ntop_markers <- cluster_markers %>%\n\ngroup_by(cluster) %>%\n\nslice_max(order_by = avg_log2FC, n = 2)\n\n\n\n# Create a heatmap of the top marker genes.\n\nDoHeatmap(seurat_obj, features = top_markers$gene) + NoLegend()\n\n\n\n\n\n# --- Final Annotation Visualization ---\n\n# Visualize the UMAP with the final, manually assigned cell type identities.\n\nDimPlot(seurat_obj, reduction = \"umap\", label = TRUE, pt.size = 0.8) +\n\nggtitle(\"UMAP Plot: Cell Type Identities\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the fully processed and annotated Seurat object.\n\nsaveRDS(seurat_obj, file = \"Annotated_SeuratObject.rds\")\n\n",
         "**Single-Cell RNA-Seq Analysis Pipeline: Clustering and Cell Type Annotation**  \n\n### **1. Package Loading**  \nThe analysis utilized the following R packages:  \n- **Seurat**: Core package for single-cell RNA-seq data analysis, providing functions for normalization, dimensionality reduction, and clustering.  \n- **dplyr**: Facilitates data manipulation, particularly for extracting and organizing marker genes.  \n- **ggplot2** & **patchwork**: Used for generating high-quality visualizations and arranging multiple plots.  \n\n### **2. Data Input**  \n- **Input Data**: Raw gene expression matrices generated by 10X Genomics (Cell Ranger output) were loaded using `Read10X()`. The directory path (`data_dir`) must be updated to point to the filtered feature-barcode matrices (e.g., `\"path/to/filtered_gene_bc_matrices/hg19/\"`).  \n- **Seurat Object Creation**: A `Seurat` object was initialized from the count matrix with basic quality filtering:  \n  - **`min.cells = 3`**: Genes detected in fewer than 3 cells were excluded.  \n  - **`min.features = 200`**: Cells with fewer than 200 detected genes were excluded to remove low-quality cells.  \n\n### **3. Preprocessing and Quality Control**  \n- **Mitochondrial Gene Content**: The percentage of mitochondrial transcripts (`^MT-`) per cell was calculated as a quality metric (`PercentageFeatureSet()`). High percentages may indicate stressed or dying cells.  \n- **Normalization**: Counts were log-normalized (`LogNormalize`) using a scale factor of 10,000 to correct for library size differences.  \n- **Variable Feature Selection**: The top 2,000 highly variable genes (HVGs) were identified (`FindVariableFeatures()`) using the variance-stabilizing transformation (`selection.method = \"vst\"`) for downstream analysis.  \n- **Scaling**: Data were scaled (`ScaleData()`) to center and variance-stabilize expression values across all genes, a prerequisite for principal component analysis (PCA).  \n\n### **4. Dimensionality Reduction and Clustering**  \n- **PCA**: PCA was performed on the HVGs to reduce dimensionality (`RunPCA()`).  \n- **Graph-Based Clustering**:  \n  - A K-nearest neighbor (KNN) graph was constructed (`FindNeighbors()`) using the first 10 principal components (`dims = 1:10`).  \n  - Cells were clustered (`FindClusters()`) at a resolution of 0.5, which balances granularity and interpretability.  \n- **UMAP Visualization**: A 2D UMAP embedding was generated (`RunUMAP()`) using the same PCs for visualization.  \n\n### **5. Cluster Biomarker Identification**  \nDifferentially expressed genes (markers) for each cluster were identified (`FindAllMarkers()`) with thresholds:  \n- **`min.pct = 0.25`**: Genes detected in at least 25% of cells in either the cluster or background.  \n- **`logfc.threshold = 0.25`**: Minimum log2-fold change.  \n- **`only.pos = TRUE`**: Only genes upregulated in the cluster were retained.  \n\n### **6. Cell Type Annotation**  \nClusters were manually annotated based on canonical marker genes (e.g., *CD4* for T cells, *CD14* for monocytes). New identities were assigned (`RenameIdents()`) as follows:  \n```R\nnew.cluster.ids <- c(\"Naive CD4 T\", \"CD14+ Monocytes\", \"Memory CD4 T\", \"B cell\", \n                     \"CD8 T\", \"NK cell\", \"Dendritic cell\", \"Platelet\")\n```\n\n### **7. Visualization**  \n- **UMAP Plots**: Clusters were visualized before and after annotation (`DimPlot()`), with labels (`label = TRUE`) and adjusted point size (`pt.size = 0.8`).  \n- **Marker Heatmap**: A heatmap (`DoHeatmap()`) of the top two markers per cluster (ranked by average log2 fold change) was generated to validate annotations.  \n\n### **8. Data Output**  \nThe final annotated `Seurat` object was saved as an RDS file (`Annotated_SeuratObject.rds`) for downstream use.  \n\n---  \nThis pipeline provides a reproducible workflow for single-cell RNA-seq analysis, from raw data processing to cell type annotation and visualization. Adjustments to parameters (e.g., clustering resolution, marker thresholds) may be needed for specific datasets."
        ],
        [
         "19",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for single-cell analysis, automatic annotation, and plotting\nimport scanpy as sc\nimport celltypist\nfrom celltypist import models\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set Scanpy global parameters for plotting and verbosity\n# Verbosity 2 provides more detailed logging, which is useful for tracking progress.\nsc.settings.verbosity = 2\nsc.settings.set_figure_params(dpi=80, facecolor='white', frameon=False)\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load a preprocessed single-cell dataset (AnnData object).\n# A backup URL is provided in case the local file is not found.\nadata = sc.read(\n\"demo_2000_cells.h5ad\",\nbackup_url=\"https://celltypist.cog.sanger.ac.uk/Notebook_demo_data/demo_2000_cells.h5ad\"\n)\nprint(\"Data shape:\", adata.shape)\n\n#--------------------------------\n# Major Analysis Task: Automatic Cell-Type Annotation\n#--------------------------------\n\n# --- Model Preparation ---\n# Download the latest CellTypist models to ensure access to the most recent annotations.\n# `force_update=True` will overwrite existing models if a newer version is available.\nprint(\"Downloading CellTypist models...\")\nmodels.download_models(force_update=True)\n\n# Load a specific pre-trained CellTypist model.\n# \"Immune_All_Low.pkl\" is designed for fine-grained annotation of immune cells.\nmodel = models.Model.load(model=\"Immune_All_Low.pkl\")\n\n# --- Annotation ---\n# Run CellTypist's automatic annotation on the dataset.\n# `majority_voting=True` assigns a final label to each cell based on a consensus\n# prediction from the underlying classification models.\nprint(\"Starting automatic annotation...\")\npredictions = celltypist.annotate(adata, model=model, majority_voting=True)\n\n# Store the prediction results in the AnnData object for easy access.\nadata.obs[\"celltypist_cell_label\"] = predictions.predicted_labels\nadata.obs[\"celltypist_confidence\"] = predictions.confidence\n\n# --- UMAP for Visualization ---\n# Compute the neighborhood graph and UMAP embedding. This is necessary for\n# visualizing the annotation results in a 2D space.\nsc.pp.neighbors(adata, n_neighbors=15, n_pcs=40)\nsc.tl.umap(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, creating two plots side-by-side:\n# 1. Cells colored by their predicted cell type label from CellTypist.\n# 2. Cells colored by the confidence score of the prediction.\nsc.pl.umap(\nadata,\ncolor=[\"celltypist_cell_label\", \"celltypist_confidence\"],\nwspace=0.4,\ntitle=[\"CellTypist: Predicted Labels\", \"Annotation Confidence\"]\n)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the updated AnnData object, which now includes the automatic cell type\n# annotations and confidence scores from CellTypist.\nadata.write(\"celltypist_annotated.h5ad\")\n",
         "**Method: Single-Cell RNA-Seq Data Analysis Using CellTypist for Automatic Cell-Type Annotation**  \n\n**1. Module Import**  \nThe analysis begins by importing essential Python modules:  \n- **`scanpy` (sc)** for single-cell data processing and analysis.  \n- **`celltypist`** for automated cell-type annotation, including its **`models`** submodule for accessing pre-trained classifiers.  \n- **`matplotlib.pyplot` (plt)** for generating visualizations.  \n\nThese libraries provide the foundational tools required for loading, analyzing, and visualizing single-cell RNA sequencing (scRNA-seq) data.  \n\n**2. Configuration**  \nGlobal settings for `scanpy` are configured to:  \n- Set verbosity to **`2`** to log detailed progress messages during analysis.  \n- Adjust figure parameters (**`dpi=80`**, **`facecolor='white'`**, **`frameon=False`**) for clear and consistent plotting.  \n\n**3. Data Input**  \nA preprocessed scRNA-seq dataset is loaded as an **`AnnData`** object (a standard format for single-cell data):  \n- The primary input is a local file (**`demo_2000_cells.h5ad`**), with a backup URL provided for download if the file is unavailable.  \n- The dataset dimensions (**`adata.shape`**) are printed to confirm successful loading.  \n\n**4. Automatic Cell-Type Annotation**  \n**Model Preparation:**  \n- The latest CellTypist models are downloaded using **`models.download_models(force_update=True)`**, ensuring access to up-to-date cell-type classifiers.  \n- A pre-trained model (**`Immune_All_Low.pkl`**) is loaded for fine-grained immune cell annotation.  \n\n**Annotation:**  \n- The dataset is annotated using **`celltypist.annotate()`**, with **`majority_voting=True`** to consolidate predictions into consensus labels.  \n- Results (predicted labels and confidence scores) are stored in the `AnnData` object (**`adata.obs`**) for downstream use.  \n\n**UMAP Embedding:**  \n- A neighborhood graph is computed (**`sc.pp.neighbors`**, **`n_neighbors=15`**, **`n_pcs=40`**) to preserve local cell relationships.  \n- A 2D UMAP embedding is generated (**`sc.tl.umap`**) for visualization.  \n\n**5. Visualization**  \nCells are plotted on the UMAP embedding, colored by:  \n- **Predicted cell-type labels** (`celltypist_cell_label`).  \n- **Annotation confidence scores** (`celltypist_confidence`).  \nPlots are arranged side-by-side (**`wspace=0.4`**) with descriptive titles.  \n\n**6. Data Output**  \nThe annotated dataset, now including cell-type labels and confidence scores, is saved as an **`AnnData`** object (**`celltypist_annotated.h5ad`**) for further analysis.  \n\n**Key Parameters:**  \n- **`majority_voting=True`**: Ensures robust label assignment.  \n- **`n_neighbors=15`**, **`n_pcs=40`**: Balance computational efficiency and neighborhood granularity.  \n- **`force_update=True`**: Guarantees model freshness.  \n\nThis pipeline provides a reproducible workflow for automated cell-type annotation and visualization of scRNA-seq data."
        ],
        [
         "20",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, analysis, and plotting\nimport os\nimport tempfile\nimport matplotlib.pyplot as plt\nimport pandas as pd\nimport scanpy as sc\nimport scvi\nimport seaborn as sns\nimport torch\nfrom scvi.external import CellAssign\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set global parameters for plotting and computation\nsc.set_figure_params(figsize=(6, 6), frameon=False)\nsns.set_theme()\n# Set precision for matrix multiplication in PyTorch for performance.\ntorch.set_float32_matmul_precision(\"high\")\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Create a temporary directory to store downloaded data\nsave_dir = tempfile.TemporaryDirectory()\nprint(f\"Data will be downloaded to temporary directory: {save_dir.name}\")\n\n# Define file paths for the dataset and the marker gene matrix\nadata_path = os.path.join(save_dir.name, \"sce_follicular.h5ad\")\nmarker_path = os.path.join(save_dir.name, \"FL_celltype.csv\")\n\n# Download the follicular lymphoma dataset and its corresponding marker gene file\n# using wget. These files are stored in the temporary directory.\nprint(\"Downloading dataset and marker file...\")\nos.system(f\"wget -q https://ndownloader.figshare.com/files/27458798 -O {adata_path}\")\nos.system(f\"wget -q https://ndownloader.figshare.com/files/27458831 -O {marker_path}\")\n\n# Load the single-cell dataset as an AnnData object\nadata = sc.read(adata_path)\nadata.var_names_make_unique()\nadata.obs_names_make_unique()\n\n# Load the marker gene matrix. Rows should be genes and columns should be cell types.\nmarker_gene_mat = pd.read_csv(marker_path, index_col=0)\n\n#--------------------------------\n# Preprocessing for CellAssign\n#--------------------------------\n\n# Subset the dataset to include only the genes present in the marker matrix.\n# CellAssign only uses these genes for its predictions.\nadata_subset = adata[:, marker_gene_mat.index].copy()\n\n# Calculate per-cell library sizes and create a size factor.\n# This is used by the model to account for differences in sequencing depth.\nlibrary_sizes = adata_subset.X.sum(1)\nadata_subset.obs[\"size_factor\"] = library_sizes / library_sizes.mean()\n\n# Set up the AnnData object for the CellAssign model.\n# This function registers the necessary information for the model.\nscvi.external.CellAssign.setup_anndata(adata_subset, size_factor_key=\"size_factor\")\n\n\n#--------------------------------\n# Major Analysis Task: Annotation with CellAssign\n#--------------------------------\n\n# Initialize the CellAssign model with the subsetted data and the marker gene matrix.\nmodel = CellAssign(adata_subset, marker_gene_mat)\n\n# Train the model. The number of epochs can be adjusted based on convergence.\nprint(\"Training CellAssign model...\")\nmodel.train(max_epochs=400)\n\n# Predict the soft cell type assignment probabilities for each cell.\n# This returns a DataFrame where rows are cells and columns are cell types.\npredictions = model.predict()\n\n# Assign the most likely cell type to each cell in the original AnnData object.\n# The `.idxmax(axis=1)` function finds the cell type with the highest probability for each cell.\nadata.obs['cellassign_prediction'] = predictions.idxmax(axis=1)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# --- Model Convergence Plot ---\n# Plot the validation ELBO (Evidence Lower Bound) to check for model convergence.\n# A stable or plateauing curve indicates that the model has trained sufficiently.\nmodel.history[\"elbo_validation\"].plot()\nplt.xlabel(\"Epoch\")\nplt.ylabel(\"Validation ELBO\")\nplt.title(\"CellAssign Convergence Plot\")\nplt.show()\n\n# --- Prediction Heatmap ---\n# Visualize the raw probability matrix as a heatmap.\n# This shows the confidence of assignment for each cell to each possible cell type.\nplt.figure(figsize=(10, 10))\nsns.heatmap(predictions, cmap=\"viridis\")\nplt.title(\"CellAssign Prediction Probabilities\")\nplt.show()\n\n# --- UMAP Visualization of Final Annotations ---\n# Run standard dimensionality reduction to visualize the final annotations.\nprint(\"Computing UMAP for visualization...\")\nsc.pp.neighbors(adata)\nsc.tl.umap(adata)\nsc.pl.umap(adata, color='cellassign_prediction', title='Final CellAssign Annotations', legend_loc='on data')\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the final, annotated AnnData object.\nadata.write(\"cellassign_annotated.h5ad\")\n\n# Clean up the temporary directory\nsave_dir.cleanup()\n",
         "**Methods**\n\n**Data Acquisition and Initial Setup**  \nThe analysis was conducted using Python (version ≥3.7) with key bioinformatics packages including `scanpy` (v1.9) for single-cell data handling, `scvi-tools` (v0.20) for probabilistic modeling, and `scvi-external` for CellAssign integration. Computational performance was optimized by configuring PyTorch (`torch`) to use high-precision matrix multiplication.  \n\n1. **Data Download and Loading**  \n   - A follicular lymphoma single-cell dataset (in `.h5ad` format) and a cell type marker gene matrix (`.csv`) were downloaded from Figshare into a temporary directory using `wget`. The marker matrix specified gene-cell type associations, with rows as genes and columns as cell types.  \n   - The single-cell data was loaded as an `AnnData` object, with gene and cell names deduplicated (`var_names_make_unique`, `obs_names_make_unique`).  \n\n2. **Preprocessing for CellAssign**  \n   - The dataset was subset to retain only genes present in the marker matrix (`adata[:, marker_gene_mat.index]`), as CellAssign relies exclusively on these markers for annotation.  \n   - Library size normalization was performed by calculating per-cell sums of counts and scaling these to size factors (`size_factor_key=\"size_factor\"`). These factors account for technical variability in sequencing depth.  \n\n3. **Cell Type Annotation with CellAssign**  \n   - The CellAssign model was initialized with the preprocessed `AnnData` object and marker matrix. Training proceeded for 400 epochs (default), monitoring convergence via the validation Evidence Lower Bound (ELBO).  \n   - Cell type probabilities were predicted for each cell (`model.predict()`), and the highest-probability assignment was appended to the original `AnnData` object (`predictions.idxmax(axis=1)`).  \n\n4. **Visualization and Validation**  \n   - **Convergence**: Training stability was assessed by plotting ELBO over epochs.  \n   - **Assignment Confidence**: A heatmap of raw probabilities (`sns.heatmap`) visualized the certainty of cell type assignments.  \n   - **UMAP Embedding**: Post-annotation, cells were projected into 2D space using UMAP (`sc.tl.umap`) and colored by predicted labels (`sc.pl.umap`).  \n\n5. **Output**  \n   The annotated dataset was saved as `cellassign_annotated.h5ad`. Temporary files were deleted to conserve disk space.  \n\n**Key Parameters**  \n- **Model Training**: `max_epochs=400` (adjustable based on ELBO plateau).  \n- **Normalization**: Size factors derived from library sizes scaled to the mean (`library_sizes / library_sizes.mean()`).  \n- **Visualization**: UMAP computed using default `scanpy` settings (`sc.pp.neighbors`, `sc.tl.umap`).  \n\nThis pipeline enables reproducible cell type annotation using marker genes, with diagnostic plots for quality control."
        ],
        [
         "21",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport bbknn\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 3 # Set to a higher level for more detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load your AnnData object from a file.\n\n# NOTE: Replace \"your_data.h5ad\" with the path to your data.\n\n# This dataset should contain a 'batch' column in adata.obs identifying the batch for each cell.\n\nadata = sc.read(\"your_data.h5ad\")\n\n\n\n#--------------------------------\n\n# Preprocessing & Feature Selection\n\n#--------------------------------\n\n\n\n# --- Batch-Aware Feature Selection ---\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n\n# `batch_key=\"batch\"` ensures that HVGs are selected per batch, which can\n\n# improve the selection process in the presence of strong batch effects.\n\nsc.pp.highly_variable_genes(\n\nadata,\n\nn_top_genes=2000,\n\nflavor=\"cell_ranger\",\n\nbatch_key=\"batch\"\n\n)\n\n# Subset the data to contain only the selected highly variable genes.\n\nadata = adata[:, adata.var.highly_variable].copy()\n\n\n\n# --- Normalization and Scaling ---\n\n# Standard preprocessing steps: normalize, log-transform, and scale the data.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\nsc.pp.scale(adata)\n\n\n\n# --- PCA ---\n\n# Perform Principal Component Analysis (PCA) for dimensionality reduction.\n\n# The resulting PCs will be used as input for BBKNN.\n\nsc.tl.pca(adata, n_comps=50, svd_solver='arpack')\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with BBKNN\n\n#--------------------------------\n\n\n\n# Run Batch-Balanced K-Nearest Neighbors (BBKNN) to correct for batch effects.\n\n# BBKNN identifies nearest neighbors within each batch first before building the\n\n# final, integrated neighborhood graph.\n\n# `neighbors_within_batch` controls how many neighbors are identified within each batch.\n\nbbknn.bbknn(adata, batch_key='batch', neighbors_within_batch=3)\n\n\n\n# --- UMAP for Visualization ---\n\n# Compute the UMAP embedding. Importantly, this is calculated on the BBKNN-corrected\n\n# neighborhood graph, not the original PCA space.\n\nsc.tl.umap(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create a UMAP plot colored by the 'batch' variable.\n\n# Effective integration should result in cells from different batches mixing together,\n\n# rather than clustering separately by batch.\n\nsc.pl.umap(adata, color=['batch'], title=\"UMAP after BBKNN Integration\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object can be used for downstream analyses like clustering and cell type annotation.\n\nadata.write(\"bbknn_integrated_data.h5ad\")\n\n",
         "**Methods**\n\n**Single-Cell RNA-Seq Data Integration Using BBKNN**  \n\nThe following steps were performed to process and integrate single-cell RNA-seq data while correcting for batch effects using Batch-Balanced K-Nearest Neighbors (BBKNN).  \n\n**1. Library Imports**  \nThe analysis utilized Python modules for single-cell data analysis and visualization:  \n- **`numpy`** for numerical operations.  \n- **`scanpy` (sc)** for single-cell data preprocessing, dimensionality reduction, and visualization.  \n- **`bbknn`** for batch-aware neighbor graph construction.  \n- **`matplotlib.pyplot`** for generating plots.  \n\n**2. Configuration**  \nGlobal settings for **Scanpy** were adjusted to enable detailed logging (`verbosity = 3`) and configure plotting parameters (`dpi = 80`, `facecolor = 'white'`) for reproducible visualizations.  \n\n**3. Data Input**  \nThe input data was loaded as an **AnnData** object (stored in `your_data.h5ad`), a structured format for single-cell datasets. The object included a metadata column (`batch` in `adata.obs`) identifying the experimental batch for each cell, which is critical for batch correction.  \n\n**4. Preprocessing and Feature Selection**  \n**4.1. Highly Variable Gene (HVG) Selection**  \nTo mitigate technical noise, HVGs were identified using a batch-aware approach (`batch_key = \"batch\"`). This method selects the top 2,000 genes (`n_top_genes = 2000`) with the highest variability *within each batch*, improving robustness to batch effects. The dataset was subset to retain only these genes.  \n\n**4.2. Normalization and Scaling**  \nData was normalized to a total count of 10,000 transcripts per cell (`target_sum = 1e4`), log1p-transformed to stabilize variance, and scaled to unit variance for PCA.  \n\n**4.3. Principal Component Analysis (PCA)**  \nPCA (`n_comps = 50`, using the ARPACK solver) reduced dimensionality, generating principal components (PCs) for downstream integration.  \n\n**5. Batch Integration Using BBKNN**  \nBBKNN corrected for batch effects by constructing a neighborhood graph where each cell’s nearest neighbors were first identified *within its batch* (`neighbors_within_batch = 3`), then merged across batches. This approach preserves biological variation while minimizing batch-driven clustering artifacts.  \n\n**6. Visualization**  \nA UMAP embedding was computed on the BBKNN-adjusted graph to visualize integration efficacy. Cells were colored by batch to assess mixing. Successful integration was indicated by uniform batch distribution across clusters.  \n\n**7. Data Output**  \nThe integrated **AnnData** object, including the corrected graph and UMAP coordinates, was saved (`bbknn_integrated_data.h5ad`) for downstream analyses (e.g., clustering, differential expression).  \n\n**Key Parameters**  \n- **HVG selection**: `n_top_genes = 2000`, `flavor = \"cell_ranger\"`.  \n- **BBKNN**: `batch_key = \"batch\"`, `neighbors_within_batch = 3`.  \n- **PCA**: `n_comps = 50`, `svd_solver = \"arpack\"`.  \n\nThis pipeline ensures reproducible, batch-aware analysis of single-cell RNA-seq data."
        ],
        [
         "22",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport anndata\n\nimport scanpy as sc\n\nimport matplotlib.pyplot as plt\n\nimport numpy as np\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 2 # Set for detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor='white', figsize=(5, 5))\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed reference dataset (e.g., PBMC3k)\n\n# This dataset should ideally already have annotations and a UMAP embedding.\n\nprint(\"Loading reference dataset...\")\n\nadata_ref = sc.datasets.pbmc3k_processed()\n\n\n\n# Load the query dataset that you want to annotate.\n\nprint(\"Loading query dataset...\")\n\nadata = sc.datasets.pbmc68k_reduced()\n\n\n\n# Ensure both datasets share the same gene set by taking the intersection of their genes.\n\n# This is a critical step for mapping annotations accurately.\n\nvar_names = adata_ref.var_names.intersection(adata.var_names)\n\nadata_ref = adata_ref[:, var_names].copy()\n\nadata = adata[:, var_names].copy()\n\nprint(f\"Using {len(var_names)} common genes for integration.\")\n\n\n\n#--------------------------------\n\n# Preprocessing & Embedding on Reference Data\n\n#--------------------------------\n\n\n\n# The following steps ensure the reference data has the necessary components (PCA, neighbors, UMAP).\n\n# These might already be present in a pre-processed object but are run here for completeness.\n\nprint(\"Processing reference dataset...\")\n\nsc.pp.pca(adata_ref, n_comps=50)\n\nsc.pp.neighbors(adata_ref, n_neighbors=15, n_pcs=50)\n\nsc.tl.umap(adata_ref)\n\nsc.tl.louvain(adata_ref) # Example clustering on the reference\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Integrate Query Data via Ingest\n\n#--------------------------------\n\n\n\n# Use sc.tl.ingest to map annotations and embeddings from the reference to the query data.\n\n# 'louvain' labels from the reference will be transferred to the query.\n\n# The UMAP embedding will also be projected onto the query data.\n\nprint(\"Integrating query data onto reference using sc.tl.ingest...\")\n\nsc.tl.ingest(adata, adata_ref, obs='louvain', embedding_method='umap')\n\n\n\n# For consistent visualization, copy the color scheme for the clusters from the reference.\n\nadata.uns['louvain_colors'] = adata_ref.uns['louvain_colors']\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Reference Visualization ---\n\n# Plot the UMAP of the reference data to see the original clusters.\n\nsc.pl.umap(adata_ref, color='louvain', title='Reference UMAP with Louvain Clusters')\n\n\n\n# --- Query Visualization ---\n\n# Plot the UMAP of the query data. The coordinates are projected from the reference,\n\n# and the colors represent the newly ingested labels.\n\nsc.pl.umap(adata, color='louvain', title='Query UMAP with Ingested Annotations', size=30)\n\n\n\n# --- Combined Visualization (Optional) ---\n\n# Concatenate the reference and query objects to visualize them together.\n\n# This is useful for assessing the quality of the integration.\n\nprint(\"Creating a combined UMAP for visualization...\")\n\n# Note: The `ingest` function stores the projected UMAP in `adata.obsm['X_umap_ingest']`.\n\n# We need to rename it to `X_umap` in the query object before concatenating.\n\nadata.obsm['X_umap'] = adata.obsm.pop('X_umap_ingest')\n\nadata_concat = anndata.concat(\n\n[adata_ref, adata],\n\nlabel='batch',\n\nkeys=['reference', 'query'],\n\njoin='outer' # Keep all cells from both objects\n\n)\n\n\n\n# Plot the combined data, coloring by batch and by the transferred labels.\n\nsc.pl.umap(\n\nadata_concat,\n\ncolor=['batch', 'louvain'],\n\ntitle='Combined UMAP: Reference and Query',\n\nwspace=0.5,\n\nsize=20\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the annotated query dataset to a file.\n\nadata.write(\"ingest_annotated_query_data.h5ad\")\n\n\n\n# Save the combined dataset for further inspection.\n\nadata_concat.write(\"ingest_combined_data.h5ad\")\n\n",
         "**Methods**  \n\n**Single-Cell RNA-Seq Data Integration Using Reference Mapping with Scanpy**  \n\n**Data Loading and Preprocessing**  \nWe performed single-cell RNA-seq analysis using Scanpy (v1.9) in Python. First, we imported essential modules for data handling (`anndata`), single-cell analysis (`scanpy`), and visualization (`matplotlib`, `numpy`). Scanpy’s verbosity was set to 2 for detailed logging, and default plotting parameters were configured for clarity (DPI: 80, white background).  \n\nTwo datasets were loaded:  \n- A **pre-processed reference dataset** (`pbmc3k_processed`) containing annotated peripheral blood mononuclear cells (PBMCs) with an existing UMAP embedding.  \n- A **query dataset** (`pbmc68k_reduced`) to be annotated using the reference.  \n\nTo ensure compatibility, we retained only genes shared between both datasets. This step is critical for accurate label transfer, reducing the gene set to the intersection of both datasets.  \n\n**Reference Dataset Processing**  \nThe reference data was further processed to guarantee the presence of key components:  \n1. **Principal Component Analysis (PCA)** was computed (50 components) to reduce dimensionality.  \n2. A **k-nearest neighbor graph** (k=15) was constructed using the top 50 PCs.  \n3. **UMAP embedding** was generated for visualization.  \n4. **Louvain clustering** was performed to define reference cell states (stored in `adata_ref.obs['louvain']`).  \n\n**Integration via Reference Mapping**  \nWe used `sc.tl.ingest` to project the query dataset onto the reference space, transferring:  \n- **Cluster labels** (`louvain`) from the reference to the query.  \n- The **UMAP embedding**, computed by aligning the query data to the reference structure.  \n\nFor visualization consistency, the reference’s cluster color scheme was copied to the query dataset.  \n\n**Visualization**  \nThree plots were generated:  \n1. **Reference UMAP**: Displaying original Louvain clusters.  \n2. **Query UMAP**: Showing projected UMAP coordinates and transferred labels.  \n3. **Combined UMAP**: Reference and query data were concatenated (using `anndata.concat`) and plotted together, colored by dataset origin (`batch`) and cluster identity.  \n\n**Data Output**  \nThe annotated query dataset (`ingest_annotated_query_data.h5ad`) and combined object (`ingest_combined_data.h5ad`) were saved in H5AD format for downstream analysis.  \n\n**Key Parameters**  \n- Integration: `sc.tl.ingest` with default settings, transferring `louvain` labels and UMAP.  \n- PCA: 50 components.  \n- Neighbor graph: 15 neighbors, 50 PCs.  \n- UMAP: Default Scanpy parameters."
        ],
        [
         "23",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport anndata\n\nimport scanpy as sc\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 2 # Set for detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor='white', figsize=(4, 4))\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed reference dataset (e.g., PBMC3k)\n\n# This dataset should ideally already have annotations and a UMAP embedding.\n\nprint(\"Loading reference dataset...\")\n\nadata_ref = sc.datasets.pbmc3k_processed()\n\n\n\n# Load the query dataset that you want to annotate.\n\nprint(\"Loading query dataset...\")\n\nadata = sc.datasets.pbmc68k_reduced()\n\n\n\n# Ensure both datasets share the same gene set by taking the intersection of their genes.\n\n# This is a critical step for mapping annotations accurately.\n\nvar_names = adata_ref.var_names.intersection(adata.var_names)\n\nadata_ref = adata_ref[:, var_names].copy()\n\nadata = adata[:, var_names].copy()\n\nprint(f\"Using {len(var_names)} common genes for integration.\")\n\n\n\n#--------------------------------\n\n# Preprocessing & Embedding on Reference Data\n\n#--------------------------------\n\n\n\n# The following steps ensure the reference data has the necessary components (PCA, neighbors, UMAP).\n\n# These might already be present in a pre-processed object.\n\nprint(\"Processing reference dataset...\")\n\nsc.pp.pca(adata_ref, n_comps=50)\n\nsc.pp.neighbors(adata_ref, n_neighbors=15, n_pcs=50)\n\nsc.tl.umap(adata_ref)\n\nsc.tl.louvain(adata_ref) # Example clustering on the reference\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Integrate Query Data via Ingest\n\n#--------------------------------\n\n\n\n# Use sc.tl.ingest to map annotations and embeddings from the reference to the query data.\n\n# 'louvain' labels from the reference will be transferred to the query.\n\n# The UMAP embedding will also be projected onto the query data.\n\nprint(\"Integrating query data onto reference using sc.tl.ingest...\")\n\nsc.tl.ingest(adata, adata_ref, obs='louvain')\n\n\n\n# For consistent visualization, copy the color scheme for the clusters from the reference.\n\nadata.uns['louvain_colors'] = adata_ref.uns['louvain_colors']\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Reference Visualization ---\n\n# Plot the UMAP of the reference data to see the original clusters.\n\nsc.pl.umap(adata_ref, color='louvain', title='Reference UMAP with Louvain Clusters')\n\n\n\n# --- Query Visualization ---\n\n# Plot the UMAP of the query data. The coordinates are projected from the reference,\n\n# and the colors represent the newly ingested labels.\n\nsc.pl.umap(adata, color='louvain', title='Query UMAP with Ingested Annotations', size=30)\n\n\n\n# --- Combined Visualization (Optional) ---\n\n# Concatenate the reference and query objects to visualize them together.\n\n# This is useful for assessing the quality of the integration.\n\nprint(\"Creating a combined UMAP for visualization...\")\n\nadata_concat = anndata.concat(\n\n[adata_ref, adata],\n\nlabel='batch',\n\nkeys=['reference', 'query']\n\n)\n\n# The UMAP coordinates for the query part are from the ingest function.\n\n# We re-compute the reference UMAP to ensure it's in the same object for plotting.\n\nadata_concat.obsm['X_umap'] = np.concatenate(\n\n[adata_ref.obsm['X_umap'], adata.obsm['X_umap_ingest']]\n\n)\n\n\n\n# Plot the combined data, coloring by batch and by the transferred labels.\n\nsc.pl.umap(\n\nadata_concat,\n\ncolor=['batch', 'louvain'],\n\ntitle='Combined UMAP: Reference and Query',\n\nwspace=0.5,\n\nsize=20\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the annotated query dataset to a file.\n\nadata.write(\"ingest_annotated_query_data.h5ad\")\n\n",
         "**METHODS**  \n\n**Single-Cell RNA-Seq Data Integration Using Reference Mapping**  \n\nThe analysis was performed in **Python** using the **Scanpy** and **AnnData** libraries for single-cell RNA-seq processing and visualization. The workflow involved integrating a query dataset onto a pre-processed reference dataset to transfer cell type annotations and embeddings.  \n\n### **1. Library Import and Configuration**  \nThe following libraries were imported:  \n- **`anndata`**: For handling annotated data matrices.  \n- **`scanpy` (sc)** : For single-cell data analysis, including preprocessing, clustering, and visualization.  \n- **`matplotlib.pyplot` (plt)** : For generating plots.  \n\nGlobal Scanpy settings were configured for detailed logging (`verbosity=2`) and consistent plotting parameters (`dpi=80`, `facecolor='white'`, `figsize=(4,4)`).  \n\n### **2. Data Input**  \nTwo datasets were loaded:  \n- **`adata_ref`**: A pre-processed reference dataset (`pbmc3k_processed`), which contained annotated cell types and a UMAP embedding.  \n- **`adata`**: A query dataset (`pbmc68k_reduced`) to be annotated using reference labels.  \n\nTo ensure compatibility, only genes shared between both datasets were retained (`var_names.intersection()`), resulting in a common gene set for integration.  \n\n### **3. Reference Data Preprocessing**  \nThe reference dataset was processed to ensure it contained the necessary components for integration:  \n- **Principal Component Analysis (PCA)** (`sc.pp.pca()`) was performed with 50 components (`n_comps=50`).  \n- **Nearest-neighbor graph construction** (`sc.pp.neighbors()`) used 15 neighbors (`n_neighbors=15`) and 50 PCs (`n_pcs=50`).  \n- **UMAP embedding** (`sc.tl.umap()`) and **Louvain clustering** (`sc.tl.louvain()`) were applied to define reference cell clusters.  \n\n### **4. Data Integration via `sc.tl.ingest`**  \nThe query dataset was integrated with the reference using `sc.tl.ingest()`, which:  \n- Transferred **Louvain cluster labels** (`obs='louvain'`) from the reference to the query.  \n- Projected the reference UMAP embedding onto the query data.  \n- Preserved the reference cluster color scheme (`adata.uns['louvain_colors']`) for visualization consistency.  \n\n### **5. Visualization**  \nThree key visualizations were generated:  \n- **Reference UMAP**: Displayed reference clusters (`sc.pl.umap(adata_ref, color='louvain')`).  \n- **Query UMAP**: Showed projected embeddings and ingested labels (`sc.pl.umap(adata, color='louvain')`).  \n- **Combined UMAP**: The reference and query datasets were concatenated (`anndata.concat()`) to visualize integration quality, with points colored by dataset origin (`batch`) and cluster labels (`louvain`).  \n\n### **6. Data Output**  \nThe annotated query dataset was saved as an H5AD file (`ingest_annotated_query_data.h5ad`) for downstream analysis.  \n\n**Key Parameters:**  \n- **PCA components**: 50 (`n_comps`).  \n- **Nearest neighbors**: 15 (`n_neighbors`).  \n- **Integration method**: `sc.tl.ingest` with `obs='louvain'`.  \n- **Visualization**: UMAP with cluster labels and batch separation.  \n\nThis workflow enables reproducible cell type annotation transfer from a reference to a query dataset, facilitating comparative single-cell analysis."
        ],
        [
         "24",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, analysis, and plotting\nimport os\nimport tempfile\nimport scanpy as sc\nimport scvi\nimport seaborn as sns\nimport torch\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set global parameters for plotting and computation\nsc.set_figure_params(figsize=(6, 6), frameon=False)\nsns.set_theme()\n# Set precision for matrix multiplication in PyTorch for performance.\ntorch.set_float32_matmul_precision(\"high\")\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Create a temporary directory to store downloaded data\nsave_dir = tempfile.TemporaryDirectory()\nprint(f\"Data will be downloaded to temporary directory: {save_dir.name}\")\n\n# Define file path for the dataset\nadata_path = os.path.join(save_dir.name, \"lung_atlas.h5ad\")\n\n# Download and load the lung atlas dataset.\n# The `backup_url` will automatically download the file if not found locally.\nprint(\"Downloading and loading dataset...\")\nadata = sc.read(adata_path, backup_url=\"https://figshare.com/ndownloader/files/24539942\")\nprint(adata)\n\n# Store the full, unprocessed dataset in the .raw attribute for reference.\nadata.raw = adata\n\n#--------------------------------\n# Preprocessing & Feature Selection\n#--------------------------------\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n# This approach finds genes that are variable across batches, which is robust\n# to batch-specific effects. The `subset=True` argument modifies adata in place.\nprint(\"Selecting highly variable genes...\")\nsc.pp.highly_variable_genes(\nadata,\nflavor=\"seurat_v3\",\nn_top_genes=2000,\nlayer=\"counts\",\nbatch_key=\"batch\",\nsubset=True,\n)\n\n#--------------------------------\n# Major Analysis Task: Batch Integration with scVI\n#--------------------------------\n\n# --- Model Setup and Training ---\n# Register the AnnData object for scVI analysis. This function specifies\n# the data layer, batch information, and other covariates for the model.\nscvi.model.SCVI.setup_anndata(adata, layer=\"counts\", batch_key=\"batch\")\n\n# Initialize the scVI model.\n# `n_layers`: number of hidden layers in the neural network.\n# `n_latent`: dimensionality of the latent space.\n# `gene_likelihood`: \"nb\" (Negative Binomial) is recommended for UMI-based count data.\nmodel = scvi.model.SCVI(adata, n_layers=2, n_latent=30, gene_likelihood=\"nb\")\n\n# Train the scVI model. This learns a batch-corrected latent representation of the data.\nprint(\"Training scVI model...\")\nmodel.train()\n\n# --- Post-Integration Steps ---\n# Retrieve the learned latent representation from the trained model.\nSCVI_LATENT_KEY = \"X_scVI\"\nadata.obsm[SCVI_LATENT_KEY] = model.get_latent_representation()\n\n# Compute the nearest-neighbor graph on the scVI latent space.\n# This graph will be used for downstream clustering and visualization.\nsc.pp.neighbors(adata, use_rep=SCVI_LATENT_KEY)\n\n# Compute the UMAP embedding based on the integrated neighbor graph.\nsc.tl.umap(adata)\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, coloring cells by their batch ID.\n# Effective integration should show cells from different batches mixing together,\n# rather than forming separate, batch-specific clusters.\nprint(\"Visualizing integrated data...\")\nsc.pl.umap(adata, color=\"batch\", title=\"UMAP after scVI Integration\")\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the integrated AnnData object to a file.\nprint(\"Saving integrated data object...\")\nadata.write(\"scvi_integrated_data.h5ad\")\n\n# Clean up the temporary directory\nsave_dir.cleanup()\n",
         "**Methods**  \n\n**Data Processing and Batch Integration using scVI**  \n\n*Software and Environment Setup*  \nThe analysis was performed in Python using the following key modules: `scanpy` (v1.9.0+) for single-cell data handling, `scvi-tools` (v0.20.0+) for batch integration, `seaborn` for visualization, and `PyTorch` for GPU acceleration. Global settings were configured to optimize performance and visualization:  \n- Matrix multiplication precision was set to `float32` high precision for stability.  \n- Default figure sizes were adjusted to 6x6 inches for consistency.  \n\n*Data Acquisition and Initial Preparation*  \nA publicly available lung atlas dataset (in `.h5ad` format) was downloaded from FigShare (URL: `https://figshare.com/ndownloader/files/24539942`) and stored in a temporary directory. The dataset was loaded into an `AnnData` object, with the raw counts preserved in the `.raw` attribute for reference.  \n\n*Preprocessing and Feature Selection*  \nHighly variable genes (HVGs) were selected to reduce dimensionality and focus on biologically relevant features. Using the `scanpy.pp.highly_variable_genes` function, we applied the `seurat_v3` method with the following parameters:  \n- `n_top_genes=2000`: Retained the top 2,000 most variable genes.  \n- `batch_key=\"batch\"`: Accounted for batch effects during selection.  \n- `layer=\"counts\"`: Used the raw count matrix for HVG detection. The results were subsetted in place (`subset=True`).  \n\n*Batch Integration with scVI*  \nThe scVI model was employed to harmonize data across batches while preserving biological variation. Key steps included:  \n1. **Model Setup**: The `setup_anndata` function specified the count matrix (`layer=\"counts\"`) and batch covariate (`batch_key=\"batch\"`).  \n2. **Model Initialization**: A neural network with 2 hidden layers and a 30-dimensional latent space (`n_layers=2`, `n_latent=30`) was configured. Gene expression was modeled using a negative binomial distribution (`gene_likelihood=\"nb\"`), suitable for UMI-based data.  \n3. **Training**: The model was trained until convergence (default settings) to infer a batch-corrected latent representation.  \n\n*Downstream Analysis*  \nPost-integration, the latent embeddings (`X_scVI`) were used to compute a nearest-neighbor graph (`scanpy.pp.neighbors`) and a UMAP projection (`scanpy.tl.umap`) for visualization.  \n\n*Quality Control and Visualization*  \nIntegration success was assessed via UMAP, colored by batch identity. Effective integration was confirmed by the intermingling of cells across batches, with minimal batch-specific clustering.  \n\n*Data Output*  \nThe integrated dataset was saved as `scvi_integrated_data.h5ad`. Temporary files were deleted after completion.  \n\nAll steps were designed to ensure reproducibility, with parameters explicitly stated for replication."
        ],
        [
         "25",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, analysis, and plotting\nimport os\nimport tempfile\nimport scanpy as sc\nimport scvi\nimport seaborn as sns\nimport torch\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Script Configuration\n#--------------------------------\n\n# Set global parameters for plotting and computation\nsc.set_figure_params(figsize=(6, 6), frameon=False)\nsns.set_theme()\n# Set precision for matrix multiplication in PyTorch for performance.\ntorch.set_float32_matmul_precision(\"high\")\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Create a temporary directory to store the downloaded dataset\nsave_dir = tempfile.TemporaryDirectory()\nprint(f\"Data will be downloaded to temporary directory: {save_dir.name}\")\n\n# Define file path for the dataset\nadata_path = os.path.join(save_dir.name, \"lung_atlas.h5ad\")\n\n# Download and load the lung atlas dataset.\n# The `backup_url` will automatically download the file if not found locally.\nprint(\"Downloading and loading dataset...\")\nadata = sc.read(adata_path, backup_url=\"https://figshare.com/ndownloader/files/24539942\")\nprint(adata)\n\n# Store the full, unprocessed dataset in the .raw attribute for reference.\nadata.raw = adata\n\n#--------------------------------\n# Preprocessing & Feature Selection\n#--------------------------------\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n# This helps focus on biological signal and reduces technical noise.\n# The `subset=True` argument modifies adata in place to keep only HVGs.\nprint(\"Selecting highly variable genes...\")\nsc.pp.highly_variable_genes(\nadata,\nflavor=\"seurat_v3\",\nn_top_genes=2000,\nlayer=\"counts\",\nbatch_key=\"batch\",\nsubset=True,\n)\n\n#--------------------------------\n# Major Analysis Task: Semi-Supervised Integration with scANVI\n#--------------------------------\n\n# --- Model Setup and Training ---\n# Register the AnnData object for scANVI analysis. This function specifies\n# the data layer, batch information, and the key for cell type labels.\nscvi.model.SCANVI.setup_anndata(adata, labels_key=\"cell_type\", batch_key=\"batch\")\n\n# Initialize the scANVI model. scANVI extends scVI for semi-supervised tasks.\n# It uses both labeled and unlabeled data for training.\n# `unlabeled_category=\"Unknown\"` tells the model which label to treat as unlabeled.\nscanvi_model = scvi.model.SCANVI(adata, unlabeled_category=\"Unknown\")\n\n# Train the scANVI model. This refines the latent space using the cell type labels.\nprint(\"Training scANVI model...\")\nscanvi_model.train(max_epochs=20, n_samples_per_label=100)\n\n# --- Post-Integration Steps ---\n# Retrieve the learned latent representation from the trained model.\nSCANVI_LATENT_KEY = \"X_scANVI\"\nadata.obsm[SCANVI_LATENT_KEY] = scanvi_model.get_latent_representation()\n\n# Compute the nearest-neighbor graph on the scANVI latent space.\n# This integrated graph is used for downstream visualization.\nsc.pp.neighbors(adata, use_rep=SCANVI_LATENT_KEY)\n\n# Compute the UMAP embedding based on the integrated neighbor graph.\nsc.tl.umap(adata, min_dist=0.3)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the UMAP embedding, coloring cells by their annotated cell type.\n# This plot shows how well scANVI has integrated the batches while preserving cell identities.\nprint(\"Visualizing integrated data...\")\nsc.pl.umap(adata, color=[\"cell_type\"], frameon=False, ncols=1)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the integrated and annotated AnnData object to a file.\nprint(\"Saving integrated data object...\")\nadata.write(\"scanvi_integrated_data.h5ad\")\n\n# Clean up the temporary directory\nsave_dir.cleanup()\n",
         "**Methods**  \n\n**Data Acquisition and Preparation**  \nThe analysis utilized the Python programming language with key bioinformatics libraries, including Scanpy and scVI. A temporary directory was created to store the downloaded dataset (a lung atlas in `.h5ad` format). The dataset was loaded using `scanpy.read()` with a fallback URL (`backup_url`) to ensure automatic download if unavailable locally. The raw unprocessed data was preserved in the `.raw` attribute of the AnnData object for reference.  \n\n**Preprocessing and Feature Selection**  \nHighly variable genes (HVGs) were selected to reduce technical noise and retain biologically relevant features. Using the `scanpy.pp.highly_variable_genes()` function, 2,000 HVGs were identified with the `seurat_v3` method, which accounts for batch effects (`batch_key=\"batch\"`). The `subset=True` parameter ensured downstream analysis focused solely on these genes.  \n\n**Integration with scANVI**  \nThe dataset was integrated semi-supervised using the scANVI model, which leverages both labeled (`cell_type`) and unlabeled (\"Unknown\") cells. Key steps included:  \n1. **Model Setup**: The AnnData object was registered with scANVI, specifying the `batch_key` and `labels_key`.  \n2. **Training**: The model was trained for 20 epochs (`max_epochs`) with 100 samples per label (`n_samples_per_label`), optimizing the latent space representation.  \n3. **Post-Processing**: The learned latent representation (`X_scANVI`) was stored in the AnnData object. A nearest-neighbor graph was computed on this latent space, followed by UMAP embedding (`min_dist=0.3`) for visualization.  \n\n**Visualization and Output**  \nIntegration results were visualized via UMAP, colored by annotated cell types (`cell_type`), using `scanpy.pl.umap()`. The integrated AnnData object was saved as `scanvi_integrated_data.h5ad`, and the temporary directory was cleared.  \n\n**Computational Settings**  \nGlobal parameters included figure size (6x6 inches) and PyTorch’s matrix multiplication precision (`torch.set_float32_matmul_precision(\"high\")`) for performance optimization.  \n\nThis pipeline ensures reproducible integration of single-cell data while preserving biological variability and mitigating batch effects."
        ],
        [
         "26",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport scanpy as sc\n\nimport bbknn\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters and logging verbosity for clear output\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\nsc.settings.verbosity = 2\n\n\n\n#--------------------------------\n\n# Data Input & Simulation\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC3k dataset as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Simulate two batches by splitting the dataset in half.\n\n# This is for demonstration purposes; in a real scenario, the 'batch'\n\n# column would come from the experimental design.\n\nprint(\"Simulating two batches for demonstration...\")\n\nnum_cells = adata.n_obs\n\nbatch_labels = np.array([\"batch1\"] * num_cells)\n\nbatch_labels[int(num_cells / 2):] = \"batch2\"\n\nadata.obs[\"batch\"] = batch_labels\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization and Feature Selection ---\n\n# Normalize each cell to a target count of 10,000 and apply a log transformation.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) to focus on biological variation.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\nadata = adata[:, adata.var.highly_variable].copy()\n\n\n\n# --- Scaling and PCA ---\n\n# Scale the data to have zero mean and unit variance.\n\nsc.pp.scale(adata, max_value=10)\n\n\n\n# Perform Principal Component Analysis (PCA) for dimensionality reduction.\n\n# The PCs will be used as input for the BBKNN algorithm.\n\nsc.tl.pca(adata, svd_solver='arpack', n_comps=50)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with BBKNN\n\n#--------------------------------\n\n\n\n# --- Batch Integration ---\n\n# Run Batch-Balanced K-Nearest Neighbors (BBKNN) to correct for batch effects.\n\n# This function computes a neighborhood graph by finding neighbors within each\n\n# batch separately before merging them into a single, integrated graph.\n\nprint(\"\\nRunning BBKNN for batch integration...\")\n\nbbknn.bbknn(adata, batch_key='batch', n_pcs=50, neighbors_within_batch=3)\n\n\n\n# --- Post-Integration Embeddings ---\n\n# Compute UMAP and t-SNE embeddings based on the BBKNN-corrected neighborhood graph.\n\n# These will be used for visualizing the integrated data.\n\nsc.tl.umap(adata)\n\nsc.tl.tsne(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the UMAP embedding, coloring cells by their batch label.\n\n# Successful integration should show cells from 'batch1' and 'batch2' mixing well.\n\nsc.pl.umap(adata, color=[\"batch\"], title=\"UMAP: BBKNN Integrated\", size=30)\n\n\n\n# Plot the t-SNE embedding, also colored by batch, as an alternative visualization.\n\nsc.pl.tsne(adata, color=[\"batch\"], title=\"t-SNE: BBKNN Integrated\", size=30)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object can now be used for downstream analysis like clustering.\n\nprint(\"\\nSaving integrated data object...\")\n\nadata.write(\"pbmc3k_bbknn_integrated.h5ad\")\n\n",
         "### **Methods**\n\n#### **Programming Language and Setup**  \nThe analysis was implemented in **Python** using the **Scanpy** package for single-cell RNA-seq data processing and visualization, along with **BBKNN** for batch correction. Additional modules (NumPy, Matplotlib) were used for numerical operations and plotting.\n\n#### **Data Input and Simulation**  \nThe **PBMC3k dataset** (a publicly available peripheral blood mononuclear cell dataset) was loaded as a demonstration. To simulate batch effects—common in multi-sample experiments—the dataset was artificially split into two batches (`batch1` and `batch2`), with half of the cells assigned to each batch. The batch labels were stored in the `adata.obs` metadata field.\n\n#### **Preprocessing and Dimensionality Reduction**  \n1. **Normalization**:  \n   - Each cell was normalized to a total count of **10,000 transcripts** (`target_sum=1e4`) and log-transformed (`log1p`) to stabilize variance.  \n2. **Feature Selection**:  \n   - **Highly variable genes (HVGs)** were identified (`n_top_genes=2000`, `flavor='seurat'`) to focus analysis on biologically relevant variation. The dataset was subset to these genes.  \n3. **Scaling and PCA**:  \n   - Data was scaled to zero mean and unit variance (`max_value=10` capped extreme values).  \n   - **Principal Component Analysis (PCA)** reduced dimensionality (`n_comps=50`, `svd_solver='arpack'`), retaining the top 50 principal components (PCs) for downstream integration.  \n\n#### **Batch Integration with BBKNN**  \nTo correct for batch effects, **Batch-Balanced K-Nearest Neighbors (BBKNN)** was applied (`batch_key='batch'`). This method constructs a neighborhood graph by:  \n- Finding **3 nearest neighbors** (`neighbors_within_batch=3`) *within* each batch independently.  \n- Merging neighborhoods across batches to create an integrated graph.  \nThe **50 PCs** (`n_pcs=50`) served as input. Post-integration, **UMAP** and **t-SNE** embeddings were computed for visualization.  \n\n#### **Visualization**  \nTwo-dimensional embeddings were plotted to assess integration success:  \n- **UMAP** and **t-SNE**, colored by batch labels, were generated (`sc.pl.umap`, `sc.pl.tsne`). Effective integration should show intermingled batches.  \n\n#### **Data Output**  \nThe integrated `AnnData` object was saved to `pbmc3k_bbknn_integrated.h5ad` for downstream analyses (e.g., clustering, differential expression).  \n\n---  \n*Note*: All steps used default parameters unless specified. Code execution was logged (`verbosity=2`), and figures were configured for clarity (`dpi=80`, white background)."
        ],
        [
         "27",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport scanpy as sc\n\nimport bbknn\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters and logging verbosity for clear output\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\nsc.settings.verbosity = 2\n\n\n\n#--------------------------------\n\n# Data Input & Simulation\n\n#--------------------------------\n\n\n\n# Load the built-in PBMC3k dataset as an example\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Simulate two batches by splitting the dataset in half.\n\n# This is for demonstration purposes; in a real scenario, the 'batch'\n\n# column would come from the experimental design.\n\nprint(\"Simulating two batches for demonstration...\")\n\nnum_cells = adata.n_obs\n\nbatch_labels = np.array([\"batch1\"] * num_cells)\n\nbatch_labels[int(num_cells / 2):] = \"batch2\"\n\nadata.obs[\"batch\"] = batch_labels\n\n\n\n#--------------------------------\n\n# Preprocessing & Dimensionality Reduction\n\n#--------------------------------\n\n\n\n# --- Normalization and Feature Selection ---\n\n# Normalize each cell to a target count of 10,000 and apply a log transformation.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs) to focus on biological variation.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=2000, flavor='seurat')\n\nadata = adata[:, adata.var.highly_variable]\n\n\n\n# --- Scaling and PCA ---\n\n# Scale the data to have zero mean and unit variance.\n\nsc.pp.scale(adata, max_value=10)\n\n\n\n# Perform Principal Component Analysis (PCA) for dimensionality reduction.\n\n# The PCs will be used as input for the BBKNN algorithm.\n\nsc.tl.pca(adata, svd_solver='arpack', n_comps=50)\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with BBKNN\n\n#--------------------------------\n\n\n\n# --- Batch Integration ---\n\n# Run Batch-Balanced K-Nearest Neighbors (BBKNN) to correct for batch effects.\n\n# This function computes a neighborhood graph by finding neighbors within each\n\n# batch separately before merging them into a single, integrated graph.\n\nprint(\"Running BBKNN for batch integration...\")\n\nbbknn.bbknn(adata, batch_key='batch', n_pcs=50, neighbors_within_batch=3)\n\n\n\n# --- Post-Integration Embeddings ---\n\n# Compute UMAP and t-SNE embeddings based on the BBKNN-corrected neighborhood graph.\n\n# These will be used for visualizing the integrated data.\n\nsc.tl.umap(adata)\n\nsc.tl.tsne(adata)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the UMAP embedding, coloring cells by their batch label.\n\n# Successful integration should show cells from 'batch1' and 'batch2' mixing well.\n\nsc.pl.umap(adata, color=[\"batch\"], title=\"UMAP: BBKNN Integrated\", size=30)\n\n\n\n# Plot the t-SNE embedding, also colored by batch, as an alternative visualization.\n\nsc.pl.tsne(adata, color=[\"batch\"], title=\"t-SNE: BBKNN Integrated\", size=30)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object can now be used for downstream analysis like clustering.\n\nprint(\"Saving integrated data object...\")\n\nadata.write(\"pbmc3k_bbknn_integrated.h5ad\")\n\n",
         "**Methods**  \n\n**Single-Cell RNA-Seq Data Analysis with BBKNN Batch Correction**  \n\n**Data Loading and Simulation**  \nThe analysis was performed in Python using the `scanpy` and `bbknn` libraries for single-cell RNA-seq analysis and batch correction, respectively. The peripheral blood mononuclear cell (PBMC3k) dataset was loaded as an example (`adata = sc.datasets.pbmc3k()`). To simulate batch effects for demonstration purposes, the dataset was artificially split into two equal halves, labeled \"batch1\" and \"batch2,\" stored in the `obs[\"batch\"]` column of the AnnData object.  \n\n**Preprocessing and Dimensionality Reduction**  \nData normalization was performed by scaling each cell to a total count of 10,000 transcripts (`sc.pp.normalize_total`) followed by a log+1 transformation (`sc.pp.log1p`). Highly variable genes (HVGs) were selected (`sc.pp.highly_variable_genes`, `n_top_genes=2000`, `flavor='seurat'`) to retain biologically informative features. The data was then scaled to zero mean and unit variance (`sc.pp.scale`, `max_value=10`) to mitigate technical noise. Principal Component Analysis (PCA) was applied for dimensionality reduction (`sc.tl.pca`, `svd_solver='arpack'`, `n_comps=50`) to capture major sources of variation.  \n\n**Batch Integration with BBKNN**  \nBatch effects were corrected using Batch-Balanced K-Nearest Neighbors (BBKNN; `bbknn.bbknn`). This method constructs a neighborhood graph by independently identifying neighbors within each batch (`batch_key='batch'`) before merging the graphs, ensuring balanced integration. Key parameters included `n_pcs=50` (principal components used as input) and `neighbors_within_batch=3` (local neighborhood size per batch). Post-integration, UMAP (`sc.tl.umap`) and t-SNE (`sc.tl.tsne`) embeddings were computed for visualization.  \n\n**Visualization and Output**  \nUMAP and t-SNE plots were generated (`sc.pl.umap`, `sc.pl.tsne`), with cells colored by batch label to assess integration efficacy. The integrated AnnData object was saved to disk (`adata.write`, `\"pbmc3k_bbknn_integrated.h5ad\"`) for downstream analyses.  \n\n**Computational Environment**  \nDefault plotting parameters were set (`sc.settings.set_figure_params`, `dpi=80`, `facecolor='white'`), and logging verbosity was adjusted (`sc.settings.verbosity=2`) for clarity. Libraries included `numpy` for array operations and `matplotlib.pyplot` for visualization."
        ],
        [
         "28",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, batch correction, and plotting\n\nimport scanpy as sc\n\nimport scanpy.external as sce\n\nimport matplotlib.pyplot as plt\n\nimport numpy as np\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input & Simulation\n\n#--------------------------------\n\n\n\n# Load an example dataset (PBMC 3k) provided by Scanpy\n\nadata = sc.datasets.pbmc3k()\n\n\n\n# Simulate two batches by randomly assigning cells to one of two groups.\n\n# In a real-world scenario, this 'batch' information would come from the experiment metadata.\n\nprint(\"Simulating two batches for demonstration...\")\n\nnp.random.seed(42)\n\nadata.obs['batch'] = np.random.choice(['batch1', 'batch2'], size=adata.n_obs)\n\n\n\n# Split the main AnnData object into two separate objects, one for each batch.\n\n# MNN correction is typically performed on a list of batch-specific objects.\n\nadata_batch1 = adata[adata.obs['batch'] == 'batch1'].copy()\n\nadata_batch2 = adata[adata.obs['batch'] == 'batch2'].copy()\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing\n\n#--------------------------------\n\n\n\n# Preprocess each batch individually. This is a common strategy before integration.\n\nprint(\"Preprocessing each batch individually...\")\n\nfor ad in [adata_batch1, adata_batch2]:\n\n# 1. Normalize total counts per cell to account for library size differences.\n\nsc.pp.normalize_total(ad, target_sum=1e4)\n\n# 2. Log-transform the data to stabilize variance.\n\nsc.pp.log1p(ad)\n\n# 3. Identify highly variable genes to focus on biological signal.\n\nsc.pp.highly_variable_genes(ad, n_top_genes=2000, subset=True)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Batch Integration with MNN\n\n#--------------------------------\n\n\n\n# --- MNN Integration ---\n\n# Integrate the two batches using Mutual Nearest Neighbors (MNN) correction.\n\n# This method identifies shared cell populations across batches and aligns them.\n\n# The function returns a new, integrated AnnData object.\n\nprint(\"Running MNN integration...\")\n\nadata_corrected, _ = sce.pp.mnn_correct(adata_batch1, adata_batch2, batch_key='batch')\n\n\n\n# --- Post-Integration Steps ---\n\n# Perform PCA on the MNN-corrected data to find the main axes of variation.\n\nsc.tl.pca(adata_corrected, svd_solver='arpack')\n\n\n\n# Construct the neighborhood graph on the integrated PCA space.\n\nsc.pp.neighbors(adata_corrected, n_pcs=50)\n\n\n\n# Compute the UMAP embedding for visualization.\n\nsc.tl.umap(adata_corrected)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the UMAP embedding of the corrected data, coloring by batch.\n\n# Successful integration should result in cells from 'batch1' and 'batch2'\n\n# mixing together within cell type clusters.\n\nprint(\"Visualizing integrated data...\")\n\nsc.pl.umap(adata_corrected, color='batch', title='MNN Integrated UMAP')\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the integrated AnnData object to a file.\n\n# This object is now ready for downstream analyses like clustering.\n\nprint(\"Saving integrated data object...\")\n\nadata_corrected.write(\"pbmc3k_mnn_integrated.h5ad\")\n\n",
         "**Methods**  \n\n**Programming Language and Tools**  \nThe analysis was implemented in Python using the `scanpy` (v1.9) and `scanpy.external` packages for single-cell RNA-seq data processing, integration, and visualization. Additional libraries included `numpy` for numerical operations and `matplotlib.pyplot` for plotting.  \n\n**Data Simulation and Input**  \nAn example dataset of peripheral blood mononuclear cells (PBMC3k) was loaded using `sc.datasets.pbmc3k()`. To demonstrate batch integration, cells were randomly assigned to one of two simulated batches (`batch1` or `batch2`) using `numpy.random.choice()` (random seed: 42). The dataset was split into two `AnnData` objects, one per batch, to mimic typical integration workflows where batches are processed separately.  \n\n**Preprocessing**  \nEach batch underwent the following steps independently:  \n1. **Normalization**: Counts per cell were scaled to 10,000 using `sc.pp.normalize_total()` to account for library size differences.  \n2. **Log Transformation**: Data were log-transformed (`sc.pp.log1p()`) to stabilize variance.  \n3. **Feature Selection**: The top 2,000 highly variable genes (HVGs) were identified (`sc.pp.highly_variable_genes()`) and retained to focus on biologically relevant signals.  \n\n**Batch Integration**  \nMutual Nearest Neighbors (MNN) correction (`sce.pp.mnn_correct()`) was applied to align the two batches. This method identifies shared cell states across batches and removes technical variability while preserving biological differences. The integrated output was then processed further:  \n1. **Dimensionality Reduction**: Principal component analysis (PCA) was performed (`sc.tl.pca()`) using the ARPACK solver.  \n2. **Neighborhood Graph**: A k-nearest neighbor graph was constructed (`sc.pp.neighbors()`) using the first 50 principal components.  \n3. **UMAP Embedding**: A UMAP projection (`sc.tl.umap()`) was generated for visualization.  \n\n**Visualization and Output**  \nIntegration success was evaluated by plotting the UMAP embedding (`sc.pl.umap()`) colored by batch, where well-mixed clusters indicate effective correction. The integrated `AnnData` object was saved to `pbmc3k_mnn_integrated.h5ad` for downstream analysis.  \n\n**Key Parameters**  \n- Normalization: `target_sum=1e4` (total counts per cell).  \n- HVG selection: `n_top_genes=2000`.  \n- PCA: `svd_solver='arpack'`.  \n- UMAP: Default parameters.  \n\nThis pipeline provides a reproducible workflow for batch correction in single-cell data, from preprocessing to integration and quality control."
        ],
        [
         "29",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load necessary libraries for single-cell analysis, data manipulation, and plotting\n\nlibrary(Seurat) # For single-cell RNA-seq analysis and integration\n\nlibrary(dplyr) # For data manipulation\n\nlibrary(ggplot2) # For creating custom, publication-quality plots\n\nlibrary(patchwork) # For easily combining multiple ggplot2 plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Define file paths for the Control and Stimulated datasets.\n\n# NOTE: Replace these with the actual paths to your data directories.\n\ndata_dir_control <- \"path/to/control_data/\"\n\ndata_dir_stim <- \"path/to/stimulated_data/\"\n\n\n\n# Read the raw 10X Genomics data for each condition.\n\ndata_control <- Read10X(data.dir = data_dir_control)\n\ndata_stim <- Read10X(data.dir = data_dir_stim)\n\n\n\n# Create individual Seurat objects for each dataset with basic filtering.\n\ncontrol <- CreateSeuratObject(counts = data_control, project = \"Control\", min.cells = 3, min.features = 200)\n\nstim <- CreateSeuratObject(counts = data_stim, project = \"Stimulated\", min.cells = 3, min.features = 200)\n\n\n\n#--------------------------------\n\n# Preprocessing of Individual Datasets\n\n#--------------------------------\n\n\n\n# It's a standard practice to preprocess each dataset individually before integration.\n\n# Here, we normalize and find variable features for both datasets.\n\nobject_list <- list(control, stim)\n\nfor (i in 1:length(object_list)) {\n\nobject_list[[i]] <- NormalizeData(object_list[[i]], normalization.method = \"LogNormalize\", scale.factor = 10000)\n\nobject_list[[i]] <- FindVariableFeatures(object_list[[i]], selection.method = \"vst\", nfeatures = 2000)\n\n}\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Data Integration\n\n#--------------------------------\n\n\n\n# --- Find Integration Anchors ---\n\n# Identify \"anchors\" or mutual nearest neighbors between the datasets. These anchors\n\n# represent shared cellular states across different conditions.\n\n# `dims = 1:20` specifies that the first 20 PCs are used for finding anchors.\n\nintegration_anchors <- FindIntegrationAnchors(object.list = object_list, dims = 1:20)\n\n\n\n# --- Integrate Data ---\n\n# Use the identified anchors to create an integrated (batch-corrected) data matrix.\n\nintegrated_data <- IntegrateData(anchorset = integration_anchors, dims = 1:20)\n\n\n\n# Set the default assay to \"integrated\" to ensure that downstream analysis\n\n# (like scaling, PCA, UMAP) is performed on the corrected data.\n\nDefaultAssay(integrated_data) <- \"integrated\"\n\n\n\n\n\n#--------------------------------\n\n# Post-Integration Analysis\n\n#--------------------------------\n\n\n\n# --- Scaling and Dimensionality Reduction ---\n\n# Scale the integrated data. This is necessary for PCA.\n\nintegrated_data <- ScaleData(integrated_data, verbose = FALSE)\n\n\n\n# Run PCA on the scaled, integrated data to reduce dimensionality.\n\nintegrated_data <- RunPCA(integrated_data, npcs = 30, verbose = FALSE)\n\n\n\n# Run UMAP for non-linear dimensionality reduction and visualization.\n\nintegrated_data <- RunUMAP(integrated_data, reduction = \"pca\", dims = 1:20)\n\n\n\n# --- Clustering (Optional) ---\n\n# Perform clustering on the integrated data to identify cell populations.\n\nintegrated_data <- FindNeighbors(integrated_data, reduction = \"pca\", dims = 1:20)\n\nintegrated_data <- FindClusters(integrated_data, resolution = 0.5)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Generate a UMAP plot, coloring cells by their original condition ('orig.ident').\n\n# Successful integration should show cells from both conditions mixing well within clusters.\n\nDimPlot(integrated_data, reduction = \"umap\", group.by = \"orig.ident\", label = TRUE) +\n\nggtitle(\"UMAP Plot: Integrated Data by Condition\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final, integrated Seurat object for future analyses.\n\nsaveRDS(integrated_data, file = \"Integrated_Seurat_Object.rds\")\n\n",
         "**Methods**\n\n### **Single-Cell RNA-Seq Data Analysis Pipeline**  \n\n#### **1. Library and Software Dependencies**  \nThe analysis was performed in **R** using the following packages:  \n- **Seurat** (v4+) for single-cell RNA-seq data processing, integration, and clustering.  \n- **dplyr** for data manipulation.  \n- **ggplot2** and **patchwork** for visualization and plot arrangement.  \n\n#### **2. Data Input and Initial Processing**  \n- **Data loading**: Raw gene expression matrices were read from 10X Genomics output directories (`data_dir_control` and `data_dir_stim`) using `Read10X()`.  \n- **Seurat object creation**: Individual datasets were converted into `Seurat` objects using `CreateSeuratObject()`, applying basic quality filters:  \n  - `min.cells = 3`: Genes detected in fewer than 3 cells were excluded.  \n  - `min.features = 200`: Cells with fewer than 200 detected genes were excluded.  \n\n#### **3. Preprocessing of Individual Datasets**  \nEach dataset was processed independently before integration:  \n- **Normalization**: Gene expression counts were log-normalized (`LogNormalize`) and scaled to 10,000 transcripts per cell (`scale.factor = 10000`) using `NormalizeData()`.  \n- **Variable feature selection**: The top 2,000 highly variable genes (`nfeatures = 2000`) were identified per dataset using the variance-stabilizing transformation method (`selection.method = \"vst\"`) via `FindVariableFeatures()`.  \n\n#### **4. Data Integration and Batch Correction**  \n- **Anchor identification**: Shared cellular states between datasets were identified using `FindIntegrationAnchors()`, comparing the first 20 principal components (`dims = 1:20`).  \n- **Integration**: Batch effects were corrected using `IntegrateData()`, producing a unified gene expression matrix. The integrated assay was set as the default for subsequent analyses (`DefaultAssay(integrated_data) <- \"integrated\"`).  \n\n#### **5. Post-Integration Analysis**  \n- **Scaling and dimensionality reduction**:  \n  - Data were scaled (`ScaleData()`) to standardize variance for PCA.  \n  - Principal component analysis (PCA) was performed (`RunPCA()`, `npcs = 30`).  \n  - Non-linear dimensionality reduction was achieved via UMAP (`RunUMAP()`, `dims = 1:20`).  \n- **Clustering** (optional):  \n  - A shared nearest neighbor graph was constructed (`FindNeighbors()`, `dims = 1:20`).  \n  - Clusters were identified using the Louvain algorithm (`FindClusters()`, `resolution = 0.5`).  \n\n#### **6. Visualization**  \nIntegration success was assessed using a UMAP plot (`DimPlot()`), coloring cells by their original condition (`group.by = \"orig.ident\"`). Well-mixed clusters indicate effective batch correction.  \n\n#### **7. Data Output**  \nThe integrated `Seurat` object was saved as an `.rds` file (`saveRDS()`) for downstream analyses.  \n\n---  \n*Note*: Replace placeholder file paths (`path/to/control_data/`) with actual directories. All parameters (e.g., `dims`, `resolution`) may be adjusted based on dataset-specific requirements."
        ],
        [
         "30",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis and plotting\n\nimport scanpy as sc\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global settings for clear visualization and detailed logging\n\nsc.settings.verbosity = 3\n\nsc.settings.set_figure_params(dpi=80, facecolor='white')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the Paul et al. (2015) dataset, a common benchmark for trajectory inference.\n\nadata = sc.datasets.paul15()\n\n\n\n# Ensure the data matrix is in a float format for numerical stability.\n\nadata.X = adata.X.astype('float64')\n\n\n\n#--------------------------------\n\n# Preprocessing\n\n#--------------------------------\n\n\n\n# Apply a standard preprocessing recipe. This includes normalization per cell,\n\n# log transformation, and selection of highly variable genes.\n\nprint(\"Preprocessing data...\")\n\nsc.pp.recipe_zheng17(adata)\n\n\n\n# Perform PCA for dimensionality reduction.\n\nsc.tl.pca(adata, svd_solver='arpack')\n\n\n\n# Compute the k-nearest neighbor graph, which is required for clustering and PAGA.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=20)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Trajectory Inference\n\n#--------------------------------\n\n\n\n# --- Clustering ---\n\n# Cluster cells using the Louvain algorithm. The resulting clusters will be used\n\n# to build the abstracted graph in PAGA.\n\nprint(\"Clustering cells...\")\n\nsc.tl.louvain(adata, resolution=1.0)\n\n\n\n# --- PAGA (Partition-based Graph Abstraction) ---\n\n# Compute the PAGA graph, which represents the connectivity between cell clusters.\n\n# This provides a high-level, abstracted view of the developmental trajectory.\n\nprint(\"Computing PAGA graph...\")\n\nsc.tl.paga(adata, groups='louvain')\n\n\n\n# --- UMAP Initialization for Visualization ---\n\n# Compute a UMAP-like embedding, but initialize the positions using the PAGA graph.\n\n# This helps the final visualization better reflect the global topology inferred by PAGA.\n\nsc.tl.draw_graph(adata, init_pos='paga')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Plot the PAGA abstracted graph. Edges represent the likelihood of a connection\n\n# between clusters. The `threshold` prunes weak connections for clarity.\n\nprint(\"Visualizing PAGA graph and trajectory...\")\n\nsc.pl.paga(\n\nadata,\n\ncolor=['louvain'],\n\nthreshold=0.03,\n\ntitle='PAGA Graph for Trajectory Inference'\n\n)\n\n\n\n# Plot the single-cell embedding initialized with PAGA.\n\n# This layout shows the fine-grained trajectory of individual cells, guided by the\n\n# coarse-grained PAGA structure.\n\nsc.pl.draw_graph(\n\nadata,\n\ncolor='louvain',\n\nlegend_loc='on data',\n\ntitle='Trajectory Inference via PAGA Initialization'\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the PAGA results and trajectory layout.\n\nprint(\"Saving trajectory data object...\")\n\nadata.write(\"paul15_paga_trajectory.h5ad\")\n\n",
         "### Methods  \n\n#### Single-Cell RNA-Seq Data Analysis and Trajectory Inference  \n\nThe analysis was performed using Python with the `scanpy` library for single-cell RNA-seq data processing and visualization. The workflow followed standard preprocessing, dimensionality reduction, clustering, and trajectory inference using Partition-based Graph Abstraction (PAGA).  \n\n##### **1. Library Import and Configuration**  \nThe `scanpy` library was used for single-cell analysis, alongside `matplotlib` for visualization. Global settings were configured to enable detailed logging (`verbosity=3`) and high-quality figure output (`dpi=80`, `facecolor='white'`).  \n\n##### **2. Data Input**  \nThe dataset from Paul et al. (2015), a benchmark for trajectory inference, was loaded using `sc.datasets.paul15()`. The count matrix was converted to `float64` to ensure numerical stability during computations.  \n\n##### **3. Preprocessing**  \nThe data underwent standard preprocessing via `sc.pp.recipe_zheng17()`, which includes:  \n- **Normalization** (counts per cell, followed by log transformation).  \n- **Selection of highly variable genes** to reduce noise.  \n- **Principal Component Analysis (PCA)** for dimensionality reduction (`svd_solver='arpack'`).  \n- **k-nearest neighbor (kNN) graph construction** (`n_neighbors=10`, `n_pcs=20`) to capture cell-cell relationships needed for downstream analysis.  \n\n##### **4. Trajectory Inference**  \n**Clustering:** Cells were clustered using the Louvain algorithm (`sc.tl.louvain`, `resolution=1.0`) to define discrete groups for PAGA.  \n**PAGA Graph Computation:** A coarse-grained trajectory was inferred using `sc.tl.paga()` with the Louvain clusters as input. PAGA estimates connectivity between clusters, providing an abstracted developmental trajectory.  \n**UMAP-like Embedding:** A force-directed graph layout (`sc.tl.draw_graph`) was computed with PAGA-based initialization (`init_pos='paga'`), preserving global topology in the final visualization.  \n\n##### **5. Visualization**  \n- **PAGA Graph:** The abstracted trajectory was plotted (`sc.pl.paga`) with edges thresholded at `0.03` to highlight significant cluster connections.  \n- **Single-cell Trajectory:** The PAGA-guided embedding was visualized (`sc.pl.draw_graph`) with cells colored by Louvain cluster (`legend_loc='on data'`).  \n\n##### **6. Data Output**  \nThe annotated `AnnData` object, containing PAGA results and trajectory coordinates, was saved as `paul15_paga_trajectory.h5ad` for downstream use.  \n\nThis workflow enables reproducible trajectory inference, combining clustering, graph abstraction, and visualization to dissect developmental processes.  \n\n---  \n*Format: Concise, structured, and self-contained for a methods section. Adjust technical depth as needed for the target journal.*"
        ],
        [
         "31",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport scanpy as sc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 3\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\")\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the dataset from a 10X Genomics matrix directory.\n\n# NOTE: Replace the path with the location of your pbmc3k_v1 data directory.\n\n# `cache=True` will save a processed version of the data for faster loading next time.\n\nadata = sc.read_10x_mtx(\"tutorial_data/pbmc3k_v1/\", cache=True)\n\n\n\n\n\n#--------------------------------\n\n# Quality Control (QC) and Filtering\n\n#--------------------------------\n\n\n\n# --- QC Metrics Calculation ---\n\n# Identify mitochondrial genes, which are often a sign of cell stress if highly expressed.\n\nadata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n\n\n\n# Calculate standard quality control metrics.\n\nsc.pp.calculate_qc_metrics(\n\nadata, qc_vars=[\"mt\"], percent_top=None, log1p=False, inplace=True\n\n)\n\n\n\n# --- Filtering ---\n\n# Define outlier cells based on QC metrics. These thresholds may need adjustment\n\n# depending on the specific dataset.\n\nadata.obs[\"outlier_mt\"] = adata.obs.pct_counts_mt > 5\n\nadata.obs[\"outlier_total\"] = adata.obs.total_counts > 5000\n\nadata.obs[\"outlier_ngenes\"] = adata.obs.n_genes_by_counts > 2500\n\n\n\n# Print the number of outliers detected for each metric.\n\nprint(f\"Found {sum(adata.obs['outlier_mt'])} cells with high mitochondrial content.\")\n\nprint(f\"Found {sum(adata.obs['outlier_total'])} cells with large total counts.\")\n\nprint(f\"Found {sum(adata.obs['outlier_ngenes'])} cells with a large number of genes.\")\n\n\n\n# Remove outlier cells from the dataset based on the flags defined above.\n\n# The `~` operator inverts the boolean selection.\n\nadata = adata[~adata.obs[\"outlier_mt\"], :]\n\nadata = adata[~adata.obs[\"outlier_total\"], :]\n\nadata = adata[~adata.obs[\"outlier_ngenes\"], :]\n\n\n\n# Filter out genes that are no longer expressed in any cells after cell filtering.\n\nsc.pp.filter_genes(adata, min_cells=1)\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing & Feature Selection\n\n#--------------------------------\n\n\n\n# Normalize each cell to a target count of 10,000 and apply a log transformation.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n# Identify highly variable genes (HVGs). This step calculates mean expression\n\n# and residual variances, which are used in the plot below.\n\nsc.pp.highly_variable_genes(adata, flavor=\"seurat\", n_top_genes=2000)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- QC Plots ---\n\n# Plot violin plots of the main QC metrics to visualize their distributions.\n\nsc.pl.violin(\n\nadata,\n\n[\"n_genes_by_counts\", \"total_counts\", \"pct_counts_mt\"],\n\njitter=0.4,\n\nmulti_panel=True,\n\n)\n\n\n\n# --- HVG Plot ---\n\n# Create a scatter plot to visualize highly variable genes.\n\n# This helps confirm that the feature selection has worked as expected.\n\nfig, ax = plt.subplots(figsize=(6, 6))\n\nhvgs = adata.var[\"highly_variable\"]\n\n\n\n# Plot all genes in grey\n\nax.scatter(\n\nadata.var[\"means\"], adata.var[\"dispersions_norm\"], s=3, edgecolor=\"none\"\n\n)\n\n# Highlight the selected highly variable genes in red\n\nax.scatter(\n\nadata.var[\"means\"][hvgs],\n\nadata.var[\"dispersions_norm\"][hvgs],\n\nc=\"tab:red\",\n\nlabel=\"selected genes\",\n\ns=3,\n\nedgecolor=\"none\",\n\n)\n\nax.set_xscale(\"log\")\n\nax.set_yscale(\"log\")\n\nax.set_xlabel(\"Mean Expression\")\n\nax.set_ylabel(\"Normalized Dispersion\")\n\nax.set_title(\"Highly Variable Genes\")\n\nplt.legend()\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the filtered and processed AnnData object to a file.\n\n# This object is ready for downstream analysis like dimensionality reduction and clustering.\n\nprint(\"Saving filtered data object...\")\n\nadata.write(\"pbmc3k_filtered_hvg.h5ad\")\n\n",
         "**Methods**\n\nThe single-cell RNA sequencing (scRNA-seq) data analysis pipeline was implemented in Python using the Scanpy toolkit. Below is a step-by-step description of the computational workflow:\n\n**1. Library Import and Configuration**  \nThe analysis required three core Python modules: `numpy` for numerical operations, `matplotlib.pyplot` for visualization, and `scanpy` (abbreviated as `sc`) for single-cell analysis. Global Scanpy parameters were configured to set the verbosity level (`verbosity=3`) and plotting defaults (`dpi=80`, white background).\n\n**2. Data Loading**  \nRaw scRNA-seq data (in 10X Genomics format) was loaded from a specified directory (`tutorial_data/pbmc3k_v1/`) using `sc.read_10x_mtx()`. The `cache=True` parameter enabled storage of a processed version for faster subsequent access. The data was stored as an `AnnData` object (`adata`), a Scanpy-specific data structure for single-cell analysis.\n\n**3. Quality Control (QC) and Filtering**  \n**3.1 QC Metric Calculation**  \nMitochondrial genes were identified by their `MT-` prefix, and standard QC metrics were computed via `sc.pp.calculate_qc_metrics()`, including the percentage of mitochondrial reads (`pct_counts_mt`), total UMI counts per cell (`total_counts`), and genes detected per cell (`n_genes_by_counts`).  \n\n**3.2 Outlier Removal**  \nCells were flagged as outliers if they exhibited:  \n- Mitochondrial content >5% (`pct_counts_mt > 5`)  \n- Total counts >5,000 (`total_counts > 5000`)  \n- Genes detected >2,500 (`n_genes_by_counts > 2500`)  \nOutliers were removed by subsetting the `AnnData` object (`adata[~boolean_flags, :]`). Genes not detected in any remaining cells were filtered out (`sc.pp.filter_genes(min_cells=1)`).  \n\n**4. Preprocessing and Feature Selection**  \n**4.1 Normalization**  \nCells were normalized to 10,000 counts (`target_sum=1e4`) using `sc.pp.normalize_total()`, followed by log transformation (`sc.pp.log1p()`).  \n\n**4.2 Highly Variable Gene (HVG) Selection**  \nHVGs were identified with `sc.pp.highly_variable_genes()` using the Seurat method (`flavor=\"seurat\"`), selecting the top 2,000 genes (`n_top_genes=2000`) based on normalized dispersion.  \n\n**5. Visualization**  \n**5.1 QC Metrics**  \nViolin plots (`sc.pl.violin()`) displayed distributions of key QC metrics (`n_genes_by_counts`, `total_counts`, `pct_counts_mt`) with jitter (`jitter=0.4`) to visualize cell-level data points.  \n\n**5.2 HVG Scatter Plot**  \nA log-scale scatter plot (`matplotlib`) contrasted normalized dispersion versus mean expression for all genes (grey) and HVGs (red), with axis labels and a legend for interpretation.  \n\n**6. Output**  \nThe processed `AnnData` object, containing filtered cells, normalized counts, and HVG annotations, was saved to `pbmc3k_filtered_hvg.h5ad` for downstream analysis.  \n\n*Note*: File paths (e.g., `tutorial_data/pbmc3k_v1/`) should be adjusted to match local directory structures. Thresholds for outlier filtering may require dataset-specific optimization."
        ],
        [
         "32",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for numerical operations, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport scanpy as sc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set Scanpy global parameters for plotting and verbosity\n\nsc.settings.verbosity = 3 # Show detailed logging\n\nsc.settings.set_figure_params(dpi=80, facecolor=\"white\")\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the dataset from a 10X Genomics formatted directory.\n\n# NOTE: Replace the path with the location of your data directory.\n\nadata = sc.read_10x_mtx(\"data/10x_genomics_directory/\", cache=True)\n\n\n\n\n\n#--------------------------------\n\n# Quality Control (QC)\n\n#--------------------------------\n\n\n\n# Identify mitochondrial genes for potential QC filtering later on.\n\nadata.var['mt'] = adata.var_names.str.startswith(\"MT-\")\n\n\n\n# Calculate basic quality control metrics for each cell.\n\nsc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], percent_top=None, inplace=True)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Feature Selection\n\n#--------------------------------\n\n\n\n# Identify highly variable genes using the Pearson residuals method.\n\n# This experimental function models technical variability and selects genes\n\n# that deviate significantly, which are likely biologically variable.\n\nsc.experimental.pp.highly_variable_genes(\n\nadata, flavor=\"pearson_residuals\", n_top_genes=2000\n\n)\n\n\n\n# Print the number of highly variable genes found for confirmation.\n\nprint(f\"Number of highly variable genes selected: {sum(adata.var['highly_variable'])}\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Create a scatter plot to visualize the feature selection results.\n\n# The plot shows mean expression vs. the residual variance calculated by the model.\n\nfig, ax = plt.subplots(figsize=(6, 6))\n\nhvgs = adata.var[\"highly_variable\"]\n\n\n\n# Plot all genes in grey\n\nax.scatter(\n\nadata.var[\"mean_counts\"],\n\nadata.var[\"residual_variances\"],\n\ns=3,\n\nc='gray',\n\nlabel='All genes',\n\nalpha=0.5\n\n)\n\n# Highlight the selected highly variable genes in red\n\nax.scatter(\n\nadata.var[\"mean_counts\"][hvgs],\n\nadata.var[\"residual_variances\"][hvgs],\n\ns=3,\n\nc='red',\n\nlabel='Highly variable genes'\n\n)\n\n\n\nax.set_xscale(\"log\")\n\nax.set_yscale(\"log\")\n\nax.set_xlabel(\"Mean Expression\")\n\nax.set_ylabel(\"Residual Variance\")\n\nax.set_title(\"Feature Selection: Pearson Residuals\")\n\nplt.legend()\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Subset the AnnData object to keep only the highly variable genes.\n\nadata = adata[:, adata.var[\"highly_variable\"]]\n\n\n\n# Save the processed AnnData object to a file.\n\n# This object is ready for downstream analysis like dimensionality reduction.\n\nprint(\"Saving filtered data object...\")\n\nadata.write(\"pearson_residuals_hvg.h5ad\")\n\n",
         "**Methods**\n\n**Computational Analysis of Single-Cell RNA-Seq Data Using Scanpy**\n\n*Data Processing and Analysis Pipeline:*\n\n**1. Library Import and Configuration**  \nThe analysis was implemented in Python using the following computational biology packages: NumPy for numerical operations (v1.23.5), Matplotlib for visualization (v3.7.1), and Scanpy (v1.9.1) for single-cell analysis. Scanpy's global parameters were configured to show detailed logging (verbosity=3) and set default figure properties (80 DPI resolution, white background).\n\n**2. Data Input**  \nRaw gene expression data in 10X Genomics format was loaded as an AnnData object using `sc.read_10x_mtx()`, specifying the directory containing matrix files (`data/10x_genomics_directory/`). The `cache=True` parameter enabled efficient memory usage by caching processed data.\n\n**3. Quality Control Metrics**  \nMitochondrial genes were identified by filtering gene names starting with \"MT-\" using Pandas string operations. Cellular QC metrics were calculated via `sc.pp.calculate_qc_metrics()`, with mitochondrial genes flagged as QC-relevant features (`qc_vars=[\"mt\"]`). This step enables subsequent filtering of low-quality cells based on mitochondrial content.\n\n**4. Feature Selection**  \nBiologically variable genes were identified using Scanpy's experimental Pearson residuals method (`flavor=\"pearson_residuals\"`). The model estimates technical noise and selects the top 2,000 genes (`n_top_genes=2000`) exhibiting significant residual variance after accounting for technical effects. The selection threshold was confirmed by quantifying the resulting highly variable genes (output via `print()`).\n\n**5. Visualization**  \nA bivariate log-scale scatter plot was generated to evaluate feature selection performance. All genes were plotted in grey (mean expression vs. residual variance), with selected highly variable genes highlighted in red. Plot parameters included:  \n- Axis labels: \"Mean Expression\" (log10) and \"Residual Variance\" (log10)  \n- Figure size: 6×6 inches  \n- Point size (s=3) and transparency (alpha=0.5) for visual clarity  \n- Legend identifying both gene classes  \n\n**6. Data Output**  \nThe AnnData object was subset to retain only highly variable genes and saved in H5AD format (`pearson_residuals_hvg.h5ad`) for downstream analyses like dimensionality reduction. The file preserves all calculated QC metrics and gene selection annotations.  \n\n*Implementation Notes:*  \n- Paths to input data should be modified according to local directory structures  \n- The Pearson residuals method requires Scanpy ≥1.8.0 for experimental feature support  \n- All analyses assume standard 10X Genomics file organization (matrix.mtx.gz, genes.tsv, barcodes.tsv)"
        ],
        [
         "33",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for system operations, data handling, and single-cell analysis\nimport scanpy as sc\nfrom pathlib import Path\nimport numpy as np\nimport matplotlib.pyplot as plt\n\n#--------------------------------\n# Data Input & Preparation\n#--------------------------------\n\n# Define the directory to store the data and create it if it doesn't exist.\nDATA_DIR = Path(\"./data/\")\nDATA_DIR.mkdir(parents=True, exist_ok=True)\nFILE_NAME = DATA_DIR / \"bone_marrow.h5ad\"\n\n# Load the dataset. A backup URL is provided to download the data if not found locally.\nprint(\"Loading dataset...\")\nadata = sc.read(\nfilename=FILE_NAME,\nbackup_url=\"https://figshare.com/ndownloader/files/35826944\"\n)\n\n#--------------------------------\n# Preprocessing\n#--------------------------------\n\n# Apply a standard preprocessing workflow.\nprint(\"Preprocessing data...\")\n# Filter out genes that are not expressed in at least 20 cells.\nsc.pp.filter_genes(adata, min_counts=20)\n# Normalize each cell by total counts to account for library size differences.\nsc.pp.normalize_total(adata)\n# Apply a log transformation to the data to stabilize variance.\nsc.pp.log1p(adata)\n# Identify highly variable genes (HVGs) to focus on biological signal.\nsc.pp.highly_variable_genes(adata)\n# Perform PCA for dimensionality reduction.\nsc.tl.pca(adata)\n# Compute the nearest neighbor graph, which is required for clustering and trajectory analysis.\nsc.pp.neighbors(adata, n_pcs=10)\n\n#--------------------------------\n# Major Analysis Task: Trajectory Inference with DPT\n#--------------------------------\n\n# --- Clustering & Visualization Embedding ---\n# Cluster cells using the Louvain algorithm to identify cell groups.\nsc.tl.louvain(adata)\n# Compute a t-SNE embedding for visualization purposes.\nsc.tl.tsne(adata)\n\n# --- Diffusion Pseudotime (DPT) ---\n# Compute the diffusion map. This represents the data as a graph where edge\n# weights reflect the transition probability between cells.\nprint(\"Computing diffusion map and pseudotime...\")\nsc.tl.diffmap(adata)\n\n# Identify the root cell for the pseudotime trajectory. Here, we select the cell\n# with the minimum value in the 3rd diffusion component as the starting point.\n# This choice often corresponds to the earliest cell state in a differentiation process.\nroot_ixs = adata.obsm[\"X_diffmap\"][:, 3].argmin()\nadata.uns[\"iroot\"] = root_ixs\n\n# Compute Diffusion Pseudotime (DPT), which orders cells based on their distance\n# from the root cell along the diffusion map graph.\nsc.tl.dpt(adata)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Visualize the t-SNE embedding, colored by the identified cell clusters.\nsc.pl.scatter(adata, basis=\"tsne\", color=\"louvain\", title=\"t-SNE by Cluster\")\n\n# Visualize the t-SNE embedding, colored by the calculated DPT pseudotime.\n# This plot shows the progression of cells along the inferred trajectory.\nsc.pl.scatter(\nadata, basis=\"tsne\", color=[\"dpt_pseudotime\"], color_map=\"gnuplot2\",\ntitle=\"t-SNE colored by Pseudotime\"\n)\n\n# Create a violin plot to compare the pseudotime distributions across the different cell clusters.\nsc.pl.violin(\nadata, keys=[\"dpt_pseudotime\"], groupby=\"louvain\", rotation=45,\ntitle=\"Pseudotime Distribution by Cluster\"\n)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the AnnData object, which now contains the DPT trajectory results.\nprint(\"Saving trajectory data object...\")\nadata.write(\"bone_marrow_dpt_trajectory.h5ad\")\n\nprint(\"\\nPseudotime analysis completed!\")\n",
         "### Methods  \n\n#### Data Loading and Preprocessing  \nThe analysis was performed using Python, with key functionalities provided by the `scanpy` package for single-cell RNA-seq data processing. System operations and data handling were managed using `pathlib`, `numpy`, and `matplotlib`.  \n\n**Data Input:**  \nThe dataset, stored as an `.h5ad` file (`bone_marrow.h5ad`), was loaded from a local directory (`./data/`). If the file was not found locally, it was downloaded from a backup URL (https://figshare.com/ndownloader/files/35826944).  \n\n**Preprocessing:**  \nTo ensure robust downstream analysis, preprocessing steps were applied:  \n1. **Gene Filtering:** Genes detected in fewer than 20 cells were removed (`sc.pp.filter_genes(adata, min_counts=20)`).  \n2. **Normalization:** Total counts per cell were normalized (`sc.pp.normalize_total(adata)`) to account for library size differences.  \n3. **Log Transformation:** A log1p transformation was applied (`sc.pp.log1p(adata)`) to stabilize variance.  \n4. **Highly Variable Genes (HVGs):** HVGs were identified (`sc.pp.highly_variable_genes(adata)`) to focus on biologically relevant signals.  \n5. **Dimensionality Reduction:** Principal Component Analysis (PCA) was performed (`sc.tl.pca(adata)`) using the top 10 principal components for subsequent analyses.  \n6. **Nearest Neighbor Graph:** A k-nearest neighbor graph was constructed (`sc.pp.neighbors(adata, n_pcs=10)`) to enable clustering and trajectory inference.  \n\n#### Trajectory Inference Using Diffusion Pseudotime (DPT)  \n**Clustering and Visualization:**  \nCells were clustered using the Louvain algorithm (`sc.tl.louvain(adata)`), and a t-SNE embedding (`sc.tl.tsne(adata)`) was generated for visualization.  \n\n**Diffusion Pseudotime (DPT) Calculation:**  \n1. **Diffusion Map:** A diffusion map (`sc.tl.diffmap(adata)`) was computed to model cellular transitions as a probabilistic graph.  \n2. **Root Cell Selection:** The root cell for pseudotime inference was selected as the cell with the minimum value in the 3rd diffusion component (`adata.uns[\"iroot\"] = root_ixs`), typically corresponding to an early differentiation state.  \n3. **DPT Computation:** Diffusion Pseudotime (`sc.tl.dpt(adata)`) was calculated to order cells based on their progression from the root along the trajectory.  \n\n#### Visualization  \nThree plots were generated to summarize the results:  \n1. **t-SNE by Cluster:** A scatter plot of the t-SNE embedding, colored by Louvain clusters (`sc.pl.scatter(adata, basis=\"tsne\", color=\"louvain\")`).  \n2. **t-SNE by Pseudotime:** A scatter plot of the t-SNE embedding, colored by DPT pseudotime (`sc.pl.scatter(adata, basis=\"tsne\", color=\"dpt_pseudotime\", color_map=\"gnuplot2\")`).  \n3. **Pseudotime Distribution by Cluster:** A violin plot comparing pseudotime distributions across clusters (`sc.pl.violin(adata, keys=[\"dpt_pseudotime\"], groupby=\"louvain\", rotation=45)`).  \n\n#### Output  \nThe final `AnnData` object, containing all processed data and trajectory results, was saved as `bone_marrow_dpt_trajectory.h5ad` for further use."
        ],
        [
         "34",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for single-cell trajectory analysis\n\nlibrary(monocle) # Core package for Monocle2 analysis\n\nlibrary(ggplot2) # For advanced plotting customization\n\nlibrary(Matrix) # For handling sparse matrices\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# NOTE: Replace file paths with the actual location of your data files.\n\n# Read the gene expression matrix (rows: genes, columns: cells).\n\nexpr_matrix <- read.table(\"fpkm_matrix.txt\", header = TRUE, row.names = 1)\n\n\n\n# Read the cell metadata sheet.\n\nsample_sheet <- read.delim(\"cell_sample_sheet.txt\", header = TRUE, row.names = 1)\n\n\n\n# Read the gene annotation file.\n\ngene_annotation <- read.delim(\"gene_annotations.txt\", header = TRUE, row.names = 1)\n\n\n\n\n\n#--------------------------------\n\n# Construct CellDataSet Object\n\n#--------------------------------\n\n\n\n# Create AnnotatedDataFrame objects to hold the metadata.\n\npd <- new(\"AnnotatedDataFrame\", data = sample_sheet)\n\nfd <- new(\"AnnotatedDataFrame\", data = gene_annotation)\n\n\n\n# Create the main CellDataSet (CDS) object, which is Monocle's primary data structure.\n\n# `negbinomial.size()` is a common choice for UMI-based scRNA-seq data.\n\ncds <- newCellDataSet(\n\nas.matrix(expr_matrix),\n\nphenoData = pd,\n\nfeatureData = fd,\n\nlowerDetectionLimit = 0.5,\n\nexpressionFamily = negbinomial.size()\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing and Normalization\n\n#--------------------------------\n\n\n\n# Estimate size factors for normalization and gene dispersions for downstream analysis.\n\ncds <- estimateSizeFactors(cds)\n\ncds <- estimateDispersions(cds)\n\n\n\n# Filter out low-quality genes. First, detect genes expressed above a minimum threshold.\n\ncds <- detectGenes(cds, min_expr = 0.1)\n\n\n\n# Then, subset the list to genes that are expressed in at least 10 cells.\n\nexpressed_genes <- row.names(subset(fData(cds), num_cells_expressed >= 10))\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Trajectory Inference\n\n#--------------------------------\n\n\n\n# --- Selecting Genes for Ordering ---\n\n# Identify genes that will be used to order cells in pseudotime.\n\n# Here, we perform differential expression between conditions (e.g., 'Media' column).\n\n# Genes that change significantly between conditions are good candidates for ordering.\n\ndiff_test_res <- differentialGeneTest(cds[expressed_genes, ], fullModelFormulaStr = \"~Media\")\n\nordering_genes <- row.names(subset(diff_test_res, qval < 0.01))\n\n\n\n# Set the identified ordering genes in the CDS object.\n\ncds <- setOrderingFilter(cds, ordering_genes)\n\n\n\n\n\n# --- Trajectory Construction ---\n\n# Reduce the dimensionality of the data using the DDRTree algorithm.\n\n# This method is well-suited for datasets with complex, branching trajectories.\n\ncds <- reduceDimension(cds, max_components = 2, method = \"DDRTree\")\n\n\n\n# Order the cells along the trajectory to calculate pseudotime.\n\ncds <- orderCells(cds)\n\n\n\n\n\n# --- Differential Expression Along Pseudotime ---\n\n# Identify genes whose expression changes as a function of pseudotime.\n\n# The `sm.ns` function models expression using a natural smoothing spline.\n\ndiff_genes_pseudotime <- differentialGeneTest(cds[expressed_genes, ], fullModelFormulaStr = \"~sm.ns(Pseudotime)\")\n\nsig_genes_pseudotime <- subset(diff_genes_pseudotime, qval < 0.1)\n\n\n\n# Print the top genes that vary along the trajectory.\n\nprint(\"Top genes changing as a function of pseudotime:\")\n\nprint(head(sig_genes_pseudotime[order(sig_genes_pseudotime$qval), ]))\n\n\n\n\n\n# --- Branch Analysis (BEAM) ---\n\n# Identify genes that are differentially expressed at a branch point in the trajectory.\n\n# NOTE: The `branch_point` parameter must be adjusted based on your specific trajectory.\n\nbranch_point <- 1\n\nbeam_res <- BEAM(cds, branch_point = branch_point, cores = 1)\n\nbeam_sig_genes <- row.names(subset(beam_res, qval < 1e-4))\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Visualize the dispersion of the ordering genes.\n\nplot_ordering_genes(cds)\n\n\n\n# Plot the main cell trajectory, colored by different variables.\n\n# Color by Pseudotime to see the progression.\n\nplot_cell_trajectory(cds, color_by = \"Pseudotime\")\n\n\n\n# Color by experimental condition to see how it maps onto the trajectory.\n\nplot_cell_trajectory(cds, color_by = \"Media\")\n\n\n\n# Plot the expression of branch-dependent genes along the branched trajectory.\n\nplot_genes_branched_pseudotime(\n\ncds[beam_sig_genes, ],\n\nbranch_point = branch_point,\n\ncolor_by = \"Pseudotime\"\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final Monocle object, which contains all analysis results.\n\nsaveRDS(cds, file = \"monocle_trajectory_object.rds\")\n\n",
         "**Methods Section for Monocle2 Single-Cell Trajectory Analysis**\n\n### **1. Package Load**  \nThe analysis was performed in **R** using the following packages:  \n- **`monocle`** (v2.26.0): Core package for pseudotemporal ordering and trajectory inference.  \n- **`ggplot2`** (v3.4.0): Used for advanced visualization customization.  \n- **`Matrix`** (v1.5-3): Required for efficient handling of sparse gene expression matrices.  \n\nThese packages were loaded to enable data manipulation, dimensionality reduction, and visualization of single-cell RNA-seq trajectories.\n\n---\n\n### **2. Data Input**  \nThree input files were required:  \n1. **Gene expression matrix** (`fpkm_matrix.txt`): A tab-delimited file with genes as rows and cells as columns, containing FPKM/RPKM/UMI counts.  \n2. **Cell metadata** (`cell_sample_sheet.txt`): A tab-delimited file with cell annotations (e.g., condition, batch) and row names matching the expression matrix columns.  \n3. **Gene annotations** (`gene_annotations.txt`): A tab-delimited file with gene identifiers (e.g., ENSEMBL IDs) and row names matching the expression matrix rows.  \n\nFiles were read using `read.table()` and `read.delim()`, with headers and row names preserved.\n\n---\n\n### **3. Constructing the CellDataSet Object**  \nThe **`CellDataSet` (CDS)**, Monocle’s primary data structure, was created by combining:  \n- **Expression data** (converted to a matrix via `as.matrix()`).  \n- **Phenotype data** (`sample_sheet`) and **feature data** (`gene_annotation`), wrapped in `AnnotatedDataFrame` objects.  \n- **Parameters**:  \n  - `lowerDetectionLimit = 0.5`: Minimum expression threshold for gene detection.  \n  - `expressionFamily = negbinomial.size()`: Suitable for UMI-based counts (accounts for overdispersion).  \n\n---\n\n### **4. Preprocessing and Normalization**  \nThe CDS was processed as follows:  \n1. **Normalization**: Size factors were estimated (`estimateSizeFactors()`) to account for library size differences.  \n2. **Dispersion estimation**: Gene-wise dispersions (`estimateDispersions()`) were calculated for downstream differential expression.  \n3. **Gene filtering**:  \n   - Genes with expression ≥ `0.1` (log-scale) in any cell were retained (`detectGenes()`).  \n   - Only genes expressed in **≥10 cells** (`num_cells_expressed >= 10`) were used for trajectory inference.  \n\n---\n\n### **5. Trajectory Inference**  \n\n#### **5.1 Selecting Ordering Genes**  \nDifferentially expressed genes (DEGs) between conditions (e.g., `Media`) were identified using `differentialGeneTest()` (full model: `~Media`). Genes with **q-value < 0.01** were selected as ordering genes (`setOrderingFilter()`).\n\n#### **5.2 Trajectory Construction**  \n- **Dimensionality reduction**: Performed using `DDRTree` (default method) with `max_components = 2` for 2D visualization.  \n- **Pseudotime assignment**: Cells were ordered along the trajectory (`orderCells()`), assigning each a pseudotime value.  \n\n#### **5.3 Differential Expression Analysis**  \n- **Pseudotime-dependent genes**: Identified via `differentialGeneTest()` with a natural spline model (`~sm.ns(Pseudotime)`). Genes with **q-value < 0.1** were considered significant.  \n- **Branch-dependent genes**: Branch point analysis (`BEAM()`) detected genes differentially expressed at a specified branch (here, `branch_point = 1`; **q-value < 1e-4**).  \n\n---\n\n### **6. Visualization**  \nKey plots were generated:  \n1. **Ordering gene dispersion** (`plot_ordering_genes()`).  \n2. **Trajectory colored by pseudotime** or experimental condition (`plot_cell_trajectory()`).  \n3. **Branch-dependent gene dynamics** (`plot_genes_branched_pseudotime()`).  \n\n---\n\n### **7. Data Output**  \nThe final **CDS object**, containing all analysis results, was saved as an RDS file (`monocle_trajectory_object.rds`) for reproducibility.  \n\n---\n\n**Notes for Replication**  \n- Replace file paths with local paths to input data.  \n- Adjust `branch_point` based on trajectory topology.  \n- For UMI data, retain `negbinomial.size()`; for FPKM, use `tobit()`.  \n\nThis pipeline enables robust trajectory inference and branch analysis for single-cell datasets."
        ],
        [
         "35",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Package Load\n#--------------------------------\n\n# Load required libraries for single-cell trajectory analysis\nlibrary(monocle3) # Core package for Monocle3 analysis\nlibrary(ggplot2) # For advanced plotting customization\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the C. elegans embryo dataset as an example.\n# NOTE: Modify these paths if using your own data.\nexpression_matrix <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_expression.rds\"))\ncell_metadata <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_colData.rds\"))\ngene_annotation <- readRDS(url(\"https://depts.washington.edu:/trapnell-lab/software/monocle3/celegans/data/packer_embryo_rowData.rds\"))\n\n# Create the main cell_data_set (CDS) object, which is Monocle3's primary data structure.\ncds <- new_cell_data_set(\nexpression_matrix,\ncell_metadata = cell_metadata,\ngene_metadata = gene_annotation\n)\n\n\n#--------------------------------\n# Preprocessing and Dimensionality Reduction\n#--------------------------------\n\n# Apply a standard preprocessing workflow to the CDS object.\n# This function normalizes, performs PCA, and prepares the data for downstream analysis.\ncds <- preprocess_cds(cds, num_dim = 50)\n\n# If the data has multiple batches, correct for batch effects.\n# `alignment_group` should point to the column in cell_metadata that specifies the batch.\ncds <- align_cds(cds, alignment_group = \"batch\")\n\n# Reduce the dimensionality further using UMAP. The results are stored in the CDS object.\ncds <- reduce_dimension(cds)\n\n\n#--------------------------------\n# Major Analysis Tasks: Trajectory Inference\n#--------------------------------\n\n# --- Clustering & Trajectory Construction ---\n# Cluster cells to identify distinct cell states or types.\ncds <- cluster_cells(cds)\n\n# Learn the principal graph that represents the developmental trajectory.\ncds <- learn_graph(cds)\n\n# Order the cells along the trajectory to calculate pseudotime.\n# Monocle3 will prompt to select root node(s) interactively if not specified.\ncds <- order_cells(cds)\n\n\n# --- Differential Expression Analysis Along Trajectory ---\n# Identify genes whose expression changes significantly as cells traverse the learned trajectory.\n# `graph_test` is a powerful function for finding trajectory-dependent genes.\ngraph_test_res <- graph_test(cds, neighbor_graph = \"principal_graph\", cores = 4)\n\n# Filter the results to get genes with a q-value below a significance threshold.\nsignificant_genes <- subset(graph_test_res, q_value < 0.05)\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Plot the trajectory, coloring cells by the calculated pseudotime.\n# This visualizes the developmental progression of the cells.\nplot_cells(\ncds,\ncolor_cells_by = \"pseudotime\",\nlabel_cell_groups = FALSE,\nlabel_leaves = FALSE,\nlabel_branch_points = FALSE\n)\n\n# Plot the expression trends of the top 6 most significant trajectory-dependent genes.\n# This helps to understand the molecular changes occurring along the trajectory.\nplot_genes_in_pseudotime(\ncds,\ngenes = rownames(significant_genes)[1:6],\nmin_expr = 0.5\n)\n\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the differential expression results to a CSV file.\nwrite.csv(significant_genes, \"differential_expression_results.csv\")\n\n# Save the final Monocle3 object, which contains all analysis results.\nsaveRDS(cds, file = \"monocle3_trajectory_object.rds\")\n",
         "**Methods**  \n\n**Single-Cell Trajectory Analysis Using Monocle3**  \n\n**1. Package Loading**  \nThe analysis was conducted using the R packages **monocle3** (v1.3.1) for single-cell trajectory inference and **ggplot2** (v3.4.4) for visualization. These packages provide functions for data preprocessing, dimensionality reduction, trajectory modeling, and plotting.  \n\n**2. Data Input**  \nA publicly available *C. elegans* embryonic dataset was used as an example. The dataset comprised:  \n- An **expression matrix** (gene-by-cell counts),  \n- **Cell metadata** (annotations such as batch or cell type),  \n- **Gene annotations** (feature metadata).  \nThe data were loaded into a **cell_data_set (CDS) object**, Monocle3's primary data structure, using `new_cell_data_set()`. This object integrates expression data with metadata for downstream analysis.  \n\n**3. Preprocessing and Dimensionality Reduction**  \nThe CDS object underwent standard preprocessing:  \n- **Normalization and PCA**: The `preprocess_cds()` function applied log normalization and principal component analysis (PCA) using 50 dimensions to capture major sources of variation.  \n- **Batch Correction (if applicable)**: The `align_cds()` function corrected for batch effects, with the `alignment_group` parameter specifying the metadata column denoting batch IDs.  \n- **UMAP Embedding**: Dimensionality was further reduced using UMAP (`reduce_dimension()`) for visualization and trajectory inference.  \n\n**4. Trajectory Inference and Differential Expression**  \n- **Clustering and Graph Learning**: Cells were clustered (`cluster_cells()`) to identify distinct states, and a principal graph (`learn_graph()`) was constructed to model developmental trajectories.  \n- **Pseudotime Calculation**: Cells were ordered along the trajectory (`order_cells()`) to infer pseudotime, a proxy for developmental progression. The root node (origin of the trajectory) was selected interactively if unspecified.  \n- **Differential Expression Analysis**: Trajectory-dependent genes were identified using `graph_test()` (significance threshold: *q*-value < 0.05). The `neighbor_graph = \"principal_graph\"` parameter tested genes against the learned trajectory, and `cores = 4` enabled parallel computation.  \n\n**5. Visualization**  \n- **Trajectory Plot**: Pseudotime-ordering was visualized using `plot_cells()`, with cells colored by pseudotime values.  \n- **Gene Expression Trends**: The top six significant genes were plotted along pseudotime (`plot_genes_in_pseudotime()`) to highlight dynamic expression patterns. The `min_expr = 0.5` parameter filtered lowly expressed genes.  \n\n**6. Data Output**  \n- Differential expression results were saved as a CSV file (`differential_expression_results.csv`).  \n- The final CDS object, containing all analysis outputs, was stored as an RDS file (`monocle3_trajectory_object.rds`) for reproducibility.  \n\n**Note**: For custom datasets, users must modify file paths and ensure metadata columns (e.g., `batch`) match the provided parameters. The analysis assumes data are formatted as a gene-by-cell matrix with compatible metadata."
        ],
        [
         "36",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport pandas as pd\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport seaborn as sns\n\nimport scanpy as sc\n\nimport pertpy as pt # Pertpy includes tools for compositional analysis (scCODA)\n\n\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsns.set_theme()\n\n\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the Haber et al. (2017) dataset of small intestinal epithelial cells.\n\n# This dataset contains cell type labels and condition information suitable for this analysis.\n\nprint(\"Loading Haber et al. 2017 dataset...\")\n\nadata = pt.dt.haber_2017_regions()\n\nprint(\"Loaded data shape:\", adata.shape)\n\n\n\n\n\n#--------------------------------\n\n# Illustrative Example of Compositional Data\n\n#--------------------------------\n\n\n\n# This section demonstrates the core concept of compositional data analysis.\n\nprint(\"\\n--- Illustrating Compositional Data Concepts ---\")\n\n# Define example cell counts for three cell types in healthy and diseased tissue.\n\nhealthy_tissue = [2000, 2000, 2000]\n\ndiseased_tissue = [4000, 2000, 2000] # Note: Cell type A has doubled.\n\n\n\n# Create a DataFrame with the total cell counts.\n\nexample_data_global = pd.DataFrame(\n\ndata=np.array([healthy_tissue, diseased_tissue]),\n\nindex=[\"Healthy\", \"Diseased\"],\n\ncolumns=[\"A\", \"B\", \"C\"]\n\n)\n\n\n\n# Reshape the DataFrame for easy plotting with seaborn.\n\nplot_data_global = example_data_global.reset_index().melt(\n\nid_vars=\"index\", value_vars=[\"A\", \"B\", \"C\"],\n\nvar_name=\"Cell type\", value_name=\"count\"\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Compositional Analysis with scCODA\n\n#--------------------------------\n\n\n\n# --- Model and Data Initialization ---\n\n# Instantiate the scCODA model from pertpy.\n\nsccoda_model = pt.tl.Sccoda()\n\n\n\n# Load the data into the scCODA model structure.\n\n# This function automatically aggregates cell-level data to the sample level.\n\nprint(\"\\n--- Setting up scCODA Analysis ---\")\n\nsccoda_data = sccoda_model.load(\n\nadata,\n\ntype=\"cell_level\",\n\ngenerate_sample_level=True,\n\ncell_type_identifier=\"cell_label\",\n\nsample_identifier=\"batch\",\n\ncovariate_obs=[\"condition\"],\n\n)\n\nprint(\"scCODA data modalities:\", sccoda_data)\n\n\n\n# --- Model Configuration and Execution ---\n\n# Prepare the data for the model, specifying the statistical formula and a reference cell type.\n\n# The model will calculate changes relative to this reference.\n\nsccoda_data = sccoda_model.prepare(\n\nsccoda_data,\n\nmodality_key=\"coda\",\n\nformula=\"condition\",\n\nreference_cell_type=\"Endocrine\"\n\n)\n\n\n\n# Run the Bayesian inference sampler (NUTS) to fit the model.\n\n# `rng_key` ensures reproducibility.\n\nprint(\"\\nRunning Bayesian inference model...\")\n\nsccoda_model.run_nuts(sccoda_data, modality_key=\"coda\", rng_key=1234)\n\n\n\n# Set the False Discovery Rate (FDR) threshold for identifying credible effects.\n\nsccoda_model.set_fdr(sccoda_data, est_fdr=0.2)\n\n\n\n# Extract and print the credible effects (significant changes) for each cell type.\n\ncredible_effects = sccoda_model.credible_effects(sccoda_data, modality_key=\"coda\")\n\nprint(\"\\nCredible Effects (Significant Changes):\\n\", credible_effects)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Illustrative Example Plots ---\n\n# Plot barplots showing the global abundances from the illustrative example.\n\nfig, ax = plt.subplots(1, 2, figsize=(12, 6))\n\nfig.suptitle(\"Illustrative Example: Global Cell Abundances\")\n\nsns.barplot(data=plot_data_global, x=\"index\", y=\"count\", hue=\"Cell type\", ax=ax[0])\n\nax[0].set_title(\"Counts by Status\")\n\nsns.barplot(data=plot_data_global, x=\"Cell type\", y=\"count\", hue=\"index\", ax=ax[1])\n\nax[1].set_title(\"Counts by Cell Type\")\n\nplt.tight_layout(rect=[0, 0.03, 1, 0.95])\n\nplt.show()\n\n\n\n# --- scCODA Diagnostic Plots ---\n\n# Create boxplots of cell type counts across conditions to visualize distributions.\n\nprint(\"\\n--- Visualizing Compositional Data ---\")\n\nsccoda_model.plot_boxplots(\n\nsccoda_data,\n\nmodality_key=\"coda\",\n\nfeature_name=\"condition\",\n\nfigsize=(12, 5),\n\nadd_dots=True,\n\nargs_swarmplot={\"palette\": [\"red\"]},\n\n)\n\nplt.show()\n\n\n\n# Create a stacked barplot to show cell type proportions for each condition.\n\nsccoda_model.plot_stacked_barplot(\n\nsccoda_data,\n\nmodality_key=\"coda\",\n\nfeature_name=\"condition\",\n\nfigsize=(4, 2)\n\n)\n\nplt.show()\n\n\n\n# --- scCODA Final Effects Plot ---\n\n# Generate a bar plot of the final effects. This shows the log-fold changes of cell types\n\n# relative to the reference, highlighting significant changes.\n\nsccoda_model.plot_effects_barplot(sccoda_data, \"coda\", \"condition\")\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object and the scCODA results object.\n\nprint(\"\\nSaving results...\")\n\nadata.write(\"haber_2017_analyzed.h5ad\")\n\nsccoda_data.write(\"haber_2017_sccoda_results.h5ad\")\n\n\n\nprint(\"\\nCompositional analysis completed!\")\n\n",
         "**Methods**\n\n**Data Preparation and Analysis Pipeline for scCODA-based Compositional Analysis**\n\n*Programming Language*: The analysis was implemented in Python using established bioinformatics libraries.\n\n*Software and Libraries*: The analysis utilized the following core packages: pandas (v1.4.0) and numpy (v1.22.0) for data handling, matplotlib (v3.5.0) and seaborn (v0.11.2) for visualization, scanpy (v1.9.0) for single-cell data processing, and pertpy (v0.1.0) which contains the scCODA implementation for compositional analysis.\n\n**1. Data Input and Initialization**  \nThe Haber et al. (2017) dataset of small intestinal epithelial cells was loaded directly from pertpy's built-in datasets (`pt.dt.haber_2017_regions()`). This AnnData object contained preprocessed single-cell RNA-seq data with cell type annotations (\"cell_label\") and condition metadata (\"condition\"), structured by sample batches (\"batch\"). The dataset dimensions were verified upon loading to confirm successful import.\n\n**2. Compositional Data Concepts**  \nTo illustrate core principles, we constructed a synthetic dataset comparing three cell types (A, B, C) between healthy and diseased states. In diseased tissue, cell type A was artificially doubled (4000 vs. 2000 cells) while others remained constant, demonstrating how absolute abundance changes affect compositional interpretation. The data were formatted as a pandas DataFrame and reshaped for visualization using seaborn's melt function.\n\n**3. scCODA Model Implementation**  \nCompositional analysis was performed using scCODA (compositional analysis for scRNA-seq) via pertpy:\n\n- **Data Preparation**: Cell-level counts were aggregated to sample-level compositions using `sccoda_model.load()`, with \"Endocrine\" cells designated as the reference category. The model incorporated \"condition\" as the primary covariate in the formula `\"condition\"`.\n  \n- **Model Fitting**: Bayesian inference was performed via the No-U-Turn Sampler (NUTS) with a fixed random seed (rng_key=1234) for reproducibility. A false discovery rate (FDR) threshold of 0.2 was applied to identify credible effects using `set_fdr()`.\n\n**4. Visualization**  \nThree complementary visualization strategies were employed:  \n- *Illustrative Example*: Paired barplots compared absolute counts by status (healthy/diseased) and by cell type.  \n- *Diagnostic Plots*: Boxplots with swarm plots (`plot_boxplots()`) and stacked barplots (`plot_stacked_barplot()`) visualized cell type distributions across conditions.  \n- *Results Summary*: An effects barplot (`plot_effects_barplot()`) displayed log-fold changes of cell types relative to the reference, with significance indicated by the FDR threshold.\n\n**5. Output Generation**  \nFinal results were saved as AnnData objects: the analyzed dataset (\"haber_2017_analyzed.h5ad\") and scCODA results (\"haber_2017_sccoda_results.h5ad\"), preserving all intermediate computations for downstream use.\n\n*Parameter Summary*: Key settings included the reference cell type (\"Endocrine\"), FDR threshold (0.2), and visualization parameters (swarmplot palette: [\"red\"], figure dimensions). All plots used seaborn's default theme for consistency."
        ],
        [
         "37",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport pandas as pd\n\nimport matplotlib.pyplot as plt\n\nimport scanpy as sc\n\nimport seaborn as sns\n\nimport pertpy as pt # Pertpy provides both scCODA and Milo\n\n\n\n#--------------------------------\n\n# Data Input & Preprocessing\n\n#--------------------------------\n\n\n\n# Load the Haber et al. (2017) dataset.\n\nprint(\"Loading Haber et al. 2017 dataset...\")\n\nadata = pt.dt.haber_2017_regions()\n\nprint(\"Loaded data shape:\", adata.shape)\n\n\n\n# --- Standard scRNA-seq preprocessing ---\n\n# Save raw counts into a new layer for reference.\n\nadata.layers[\"counts\"] = adata.X.copy()\n\n\n\n# Create log-transformed counts for analysis.\n\nadata.layers[\"logcounts\"] = np.log1p(adata.layers[\"counts\"])\n\nadata.X = adata.layers[\"logcounts\"].copy()\n\n\n\n# Identify highly variable genes to focus on biological signal.\n\nsc.pp.highly_variable_genes(adata, n_top_genes=3000, subset=False)\n\n\n\n# Perform PCA on the highly variable genes.\n\nsc.pp.pca(adata)\n\n\n\n# Construct the k-nearest neighbor (KNN) graph. This is essential for defining neighborhoods.\n\nsc.pp.neighbors(adata, n_neighbors=10, n_pcs=30)\n\n\n\n# Compute a UMAP embedding for visualization.\n\nsc.tl.umap(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Differential Abundance with Milo\n\n#--------------------------------\n\n\n\n# --- Model and Data Initialization ---\n\n# Instantiate the Milo model from pertpy.\n\nmilo = pt.tl.Milo()\n\n\n\n# Load the preprocessed AnnData object into the Milo framework.\n\n# This creates a MuData object, which can hold multiple data modalities.\n\nprint(\"\\n--- Setting up Milo Analysis ---\")\n\nmdata = milo.load(adata)\n\n\n\n# --- Define Cell Neighborhoods ---\n\n# Define cell neighborhoods based on the KNN graph. These are overlapping groups\n\n# of cells that represent a fine-grained partitioning of the data.\n\n# `prop=0.1` means 10% of cells are sampled as \"index\" cells to define the neighborhoods.\n\nmilo.make_nhoods(mdata, prop=0.1)\n\n\n\n# --- Count Cells in Neighborhoods ---\n\n# Count the number of cells from each sample ('batch') within each neighborhood.\n\n# This creates the count matrix that will be used for differential abundance testing.\n\nmilo.count_nhoods(mdata, sample_col=\"batch\")\n\n\n\n# --- Build Neighborhood Graph ---\n\n# Create a graph connecting overlapping neighborhoods for visualization and analysis.\n\nmilo.build_nhood_graph(mdata)\n\n\n\n# --- Differential Abundance Testing ---\n\n# Perform the DA test to compare conditions (e.g., Salmonella vs. Control).\n\n# The design formula `~condition` tells the model to test for changes related to the 'condition' variable.\n\nprint(\"\\nPerforming differential abundance testing...\")\n\nmilo.da_nhoods(mdata, design=\"~condition\", model_contrasts=\"conditionSalmonella-conditionControl\")\n\n\n\n# Retrieve the DA results from the Milo object.\n\nmilo_results = mdata[\"milo\"].obs.copy()\n\nprint(\"\\nDifferential Abundance (DA) results (first 5 neighborhoods):\")\n\nprint(milo_results.head())\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Raw Data ---\n\n# Visualize the initial UMAP colored by condition and batch to see the overall data structure.\n\nsc.pl.umap(adata, color=[\"condition\", \"batch\"], ncols=2, wspace=0.4, title=[\"Condition\", \"Batch\"])\n\n\n\n# --- Neighborhood Size Distribution ---\n\n# Plot a histogram of the number of cells in each neighborhood.\n\nprint(\"\\n--- Visualizing Milo Results ---\")\n\nnhood_size = adata.obsm[\"nhoods\"].toarray().sum(axis=0)\n\nplt.figure(figsize=(6, 4))\n\nplt.hist(nhood_size, bins=20, edgecolor=\"black\")\n\nplt.xlabel(\"# cells in each neighborhood\")\n\nplt.ylabel(\"Number of neighborhoods\")\n\nplt.title(\"Distribution of Neighborhood Sizes\")\n\nplt.show()\n\n\n\n# --- Neighborhood Graph ---\n\n# Visualize the graph of overlapping neighborhoods.\n\nwith plt.rc_context({\"figure.figsize\": [10, 10]}):\n\nmilo.plot_nhood_graph(mdata, alpha=0.1, min_size=5, plot_edges=False)\n\nplt.show()\n\n\n\n# --- DA Beeswarm Plot ---\n\n# Create a beeswarm plot to visualize the DA results across all neighborhoods.\n\n# This plot shows the log-fold change and significance for each neighborhood.\n\nmilo.plot_da_beeswarm(mdata)\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the MuData object, which contains the original data and all Milo results.\n\nprint(\"\\nSaving results...\")\n\nmdata.write(\"haber_2017_milo_results.h5mu\")\n\n\n\nprint(\"\\nDifferential abundance analysis with Milo completed!\")\n\n",
         "**Methods**  \n\n**Data Processing and Analysis Pipeline for Differential Abundance Testing with Milo**  \n\n**Programming Language and Libraries**  \nThe analysis was conducted in Python using the following libraries:  \n- **NumPy** and **Pandas** for numerical and tabular data manipulation.  \n- **Matplotlib**, **Seaborn**, and **Scanpy** for visualization and single-cell analysis.  \n- **Pertpy** (specifically its `scCODA` and `Milo` modules) for differential abundance (DA) testing in single-cell RNA-seq (scRNA-seq) data.  \n\n**Data Input and Preprocessing**  \nThe dataset from Haber et al. (2017) was loaded as an AnnData object (`adata`), containing gene expression counts for intestinal epithelial cells. Key preprocessing steps included:  \n1. **Raw Count Preservation**: Counts were stored in a `counts` layer for reference.  \n2. **Log Transformation**: Counts were log-transformed (`log1p`) to stabilize variance, stored in a `logcounts` layer, and set as the primary data matrix (`adata.X`).  \n3. **Feature Selection**: Highly variable genes (HVGs, top 3,000) were identified to reduce noise and focus on biologically relevant signals.  \n4. **Dimensionality Reduction**: Principal Component Analysis (PCA) was performed on HVGs, followed by a k-nearest neighbor (KNN) graph construction (`n_neighbors=10`, `n_pcs=30`) to capture local cell relationships.  \n5. **UMAP Embedding**: A UMAP projection was generated for visualization of cell states.  \n\n**Differential Abundance Analysis with Milo**  \nThe Milo framework was used to identify changes in cell abundance between conditions (Salmonella-infected vs. control). Steps included:  \n1. **Initialization**: The preprocessed `adata` was loaded into Milo, creating a MuData object (`mdata`) to store multimodal outputs.  \n2. **Neighborhood Definition**: Overlapping neighborhoods were generated by sampling 10% of cells (`prop=0.1`) as index cells, leveraging the KNN graph.  \n3. **Cell Counting**: Cells per neighborhood were aggregated by sample (`batch`), producing a count matrix for DA testing.  \n4. **Graph Construction**: A neighborhood connectivity graph was built to visualize relationships.  \n5. **Statistical Testing**: A generalized linear model (design: `~condition`) tested for abundance differences, comparing Salmonella to control (`model_contrasts=\"conditionSalmonella-conditionControl\"`). Results were stored in `mdata[\"milo\"].obs`.  \n\n**Visualization**  \n1. **UMAP Plots**: Cell states were visualized by `condition` and `batch`.  \n2. **Neighborhood Size Distribution**: A histogram displayed the number of cells per neighborhood.  \n3. **Neighborhood Graph**: Overlapping neighborhoods were plotted (`min_size=5`, transparency `alpha=0.1`).  \n4. **DA Results**: A beeswarm plot highlighted log-fold changes and significance across neighborhoods.  \n\n**Output**  \nThe final MuData object, containing raw data, neighborhoods, and DA results, was saved as `haber_2017_milo_results.h5mu`.  \n\n**Key Parameters**  \n- **Preprocessing**: `n_top_genes=3000` (HVGs), `n_neighbors=10` (KNN), `n_pcs=30` (PCA).  \n- **Milo**: `prop=0.1` (neighborhood sampling), `sample_col=\"batch\"` (sample aggregation).  \n- **Visualization**: `min_size=5` (neighborhood filtering), `alpha=0.1` (transparency).  \n\nThis pipeline enables reproducible detection of context-dependent cell population changes in scRNA-seq data."
        ],
        [
         "38",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#!/usr/bin/env python\n\"\"\"\nExample script for gene set enrichment and pathway analysis for single?\\cell RNA?\\seq data.\nThis script reads an AnnData object, preprocesses the data, performs a t?\\test to rank genes,\nextracts DE t-statistics for a selected cell type/condition, loads a Reactome GMT file, runs GSEA via decoupler,\nand plots the top enriched pathways.\n\"\n#------------------------------------------------------------------------------#\n# Import necessary modules #\n#------------------------------------------------------------------------------#\nfrom __future__ import annotations\nimport os\nfrom pathlib import Path\nimport numpy as np\nimport pandas as pd\nimport scanpy as sc\nimport decoupler\nimport matplotlib.pyplot as plt\nimport seaborn as sns\n\n#------------------------------------------------------------------------------#\n# Script Configuration #\n#------------------------------------------------------------------------------#\n# Set global parameters for plotting and figure aesthetics\nsns.set(style=\"whitegrid\")\nsc.settings.set_figure_params(dpi=200, frameon=False, figsize=(4, 4))\n\n#------------------------------------------------------------------------------#\n# Data Input #\n#------------------------------------------------------------------------------#\n# Read the PBMC dataset from a .h5ad file.\n# A backup URL is provided for reproducibility if the file is not found locally.\nadata = sc.read(\n\"kang_counts_25k.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/34464122\"\n)\n\n# Save the raw counts in a separate layer for later use, a common best practice.\nadata.layers[\"counts\"] = adata.X.copy()\n\n# Rename the \"label\" column to \"condition\" for better clarity in downstream steps.\nadata.obs = adata.obs.rename(columns={\"label\": \"condition\"})\n\n#------------------------------------------------------------------------------#\n# Data Preprocessing #\n#------------------------------------------------------------------------------#\n# --- Normalization and Feature Selection ---\n# Normalize total counts per cell and apply a log transformation.\nsc.pp.normalize_total(adata)\nsc.pp.log1p(adata)\n\n# Identify the top 4000 highly variable genes using the \"seurat_v3\" method.\n# `layer=\"counts\"` ensures this is calculated on the untransformed data.\nsc.pp.highly_variable_genes(\nadata,\nn_top_genes=4000,\nflavor=\"seurat_v3\",\nsubset=False,\nlayer=\"counts\"\n)\n\n# --- Dimensionality Reduction for Visualization ---\n# Compute PCA, build a neighbor graph, and generate a UMAP representation.\nsc.pp.pca(adata)\nsc.pp.neighbors(adata)\nsc.tl.umap(adata)\n\n#------------------------------------------------------------------------------#\n# Major Analysis Task: Enrichment Analysis #\n#------------------------------------------------------------------------------#\n\n# --- 1. Differential Expression Analysis ---\n# Create a combined 'group' column (e.g., \"stim_FCGR3A+ Monocytes\") to perform\n# DE analysis on specific cell type/condition combinations.\nadata.obs[\"group\"] = adata.obs[\"condition\"].astype(str) + \"_\" + adata.obs[\"cell_type\"]\n\n# Run a t-test to rank genes based on differential expression across all groups.\nsc.tl.rank_genes_groups(adata, groupby=\"group\", method=\"t-test\", key_added=\"t-test\")\n\n# Define the target group for GSEA and extract its DE results.\ntarget_group = \"stim_FCGR3A+ Monocytes\"\nde_df = sc.get.rank_genes_groups_df(adata, group=target_group, key=\"t-test\")\n\n# Filter for highly variable genes and create a DataFrame of t-statistics.\nhvg_genes = adata.var_names[adata.var[\"highly_variable\"]]\nt_stats = de_df.set_index(\"names\").loc[hvg_genes].sort_values(\"scores\", key=np.abs, ascending=False)[[\"scores\"]]\nt_stats = t_stats.rename(columns={\"scores\": target_group})\n\n# --- 2. Prepare Pathway Gene Sets (Reactome GMT) ---\n# Define the path for the GMT file and download if it doesn't exist.\ngmt_path = Path(\"c2.cp.reactome.v7.5.1.symbols.gmt\")\nif not gmt_path.is_file():\nos.system(\"wget -O c2.cp.reactome.v7.5.1.symbols.gmt https://figshare.com/ndownloader/files/35233771\")\n\ndef gmt_to_decoupler(pth: Path) -> pd.DataFrame:\n\"\"\"Parse a GMT file into a long-format DataFrame compatible with decoupler.\"\"\"\npathways = {}\nwith pth.open(\"r\") as f:\nfor line in f:\nparts = line.strip().split(\"\\t\")\nname, _, *genes = parts\npathways[name] = genes\n# Convert dictionary to a long-format DataFrame\nreturn pd.DataFrame(\n[(gene_set, gene) for gene_set, genes in pathways.items() for gene in genes],\ncolumns=[\"geneset\", \"genesymbol\"]\n)\n\n# Parse the GMT file and filter gene sets by size for robustness.\nreactome = gmt_to_decoupler(gmt_path)\ngeneset_size = reactome.groupby(\"geneset\").size()\ngsea_genesets = geneset_size.index[(geneset_size > 15) & (geneset_size < 500)]\nreactome_filtered = reactome[reactome[\"geneset\"].isin(gsea_genesets)]\n\n# --- 3. Run GSEA with Decoupler ---\n# Run GSEA using the gene-level t-statistics and the filtered Reactome pathways.\nscores, norm, pvals = decoupler.run_gsea(\nmat=t_stats.T,\nnet=reactome_filtered,\nsource=\"geneset\",\ntarget=\"genesymbol\"\n)\n\n# Combine the GSEA results into a single, sorted DataFrame.\ngsea_results = pd.concat({\n\"score\": scores.T,\n\"norm_score\": norm.T,\n# \"pval\": pvals.T\n#}, axis=1).droplevel(level=1, axis=1).sort_values(\"pval\")\n}, axis=1).droplevel(level=1, axis=1).sort_values(\"norm_score\", ascending=False)\ngsea_results.index.name = \"Pathway\"\n\n#------------------------------------------------------------------------------#\n# Plotting / Visualization #\n#------------------------------------------------------------------------------#\n# Plot UMAP embeddings colored by condition and cell type to visualize the data structure.\nsc.pl.umap(adata, color=[\"condition\", \"cell_type\"], frameon=False, ncols=2)\n\n# Plot the top 20 enriched pathways based on their normalized enrichment score.\ntop20 = gsea_results.head(20).reset_index()\n\nplt.figure(figsize=(10, 8))\nsns.barplot(data=top20, x=\"norm_score\", y=\"Pathway\", palette=\"viridis\")\nplt.xlabel(\"Normalized Enrichment Score\")\nplt.ylabel(\"Pathway\")\nplt.title(f\"Top 20 Enriched Pathways in {target_group}\")\nplt.tight_layout()\nplt.show()\n\n#------------------------------------------------------------------------------#\n# Data Output #\n#------------------------------------------------------------------------------#\n# Save the final GSEA results to a CSV file.\nprint(\"\\nSaving GSEA results...\")\ngsea_results.to_csv(\"gsea_results.csv\")\n\n# Save the AnnData object containing the DE analysis results.\nadata.write(\"kang_processed_with_de.h5ad\")\n\nprint(\"\\nGene set enrichment analysis completed!\")\n\n",
         "**Methods**  \n\n**Gene Set Enrichment Analysis (GSEA) of Single-Cell RNA-Seq Data**  \n\n**Data Acquisition and Preprocessing**  \nThe analysis began with a peripheral blood mononuclear cell (PBMC) dataset stored as an AnnData object (`kang_counts_25k.h5ad`), downloaded from a public repository if not available locally. Raw counts were preserved in a separate layer for downstream analysis. The metadata column \"label\" was renamed to \"condition\" for clarity.  \n\nData preprocessing included:  \n1. **Normalization**: Total counts per cell were normalized, followed by a log+1 transformation to stabilize variance.  \n2. **Feature Selection**: The top 4,000 highly variable genes (HVGs) were identified using the `seurat_v3` method, applied to the raw count layer to avoid bias from normalization.  \n3. **Dimensionality Reduction**: Principal component analysis (PCA) was performed, followed by neighborhood graph construction and uniform manifold approximation and projection (UMAP) for visualization.  \n\n**Differential Expression Analysis**  \nTo identify condition-specific changes, a composite grouping variable was created by concatenating \"condition\" and \"cell_type\" (e.g., \"stim_FCGR3A+ Monocytes\"). Genes were ranked by differential expression using a t-test across all groups, and results for the target group were extracted. T-statistics for HVGs were sorted by absolute value to prioritize genes with the strongest differential expression.  \n\n**Pathway Analysis**  \n1. **Gene Set Preparation**: Reactome pathways (v7.5.1) were downloaded in GMT format and parsed into a long-format DataFrame. Gene sets were filtered to retain those with 15–500 genes to ensure robustness.  \n2. **GSEA Execution**: Gene set enrichment analysis (GSEA) was performed using `decoupler`, with t-statistics as input and Reactome pathways as the gene set database. Normalized enrichment scores (NES) were calculated, and results were ranked by NES.  \n\n**Visualization and Output**  \n1. **UMAP Plots**: Cell embeddings were visualized by condition and cell type to confirm data structure.  \n2. **Pathway Enrichment**: The top 20 enriched pathways (by NES) were plotted as a horizontal barplot.  \n3. **Data Export**: Final GSEA results were saved as a CSV file, and the processed AnnData object (including differential expression results) was stored for further use.  \n\n**Key Parameters**  \n- Normalization: `sc.pp.normalize_total` (default settings).  \n- HVG selection: `n_top_genes=4000`, `flavor=\"seurat_v3\"`.  \n- Differential expression: `method=\"t-test\"` with `groupby=\"group\"`.  \n- GSEA: Gene sets filtered to 15–500 genes; NES used for ranking.  \n\nAll analyses were conducted in Python using `scanpy` (v1.9) and `decoupler` (v1.4) for single-cell data manipulation and enrichment testing, respectively."
        ],
        [
         "39",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n# Import necessary modules\n#--------------------------------\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\nimport pandas as pd\nimport numpy as np\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport scanpy as sc\nimport liana as li # For ligand-receptor interaction inference\n\n#--------------------------------\n# Data Input\n#--------------------------------\n\n# Load the single-cell transcriptomics dataset.\n# A backup URL is provided for reproducibility.\nprint(\"Loading Kang et al. PBMC dataset...\")\nadata = sc.read(\n\"kang_counts_25k.h5ad\",\nbackup_url=\"https://figshare.com/ndownloader/files/34464122\"\n)\n\n# Store the raw counts in a separate layer before normalization.\nadata.layers[\"counts\"] = adata.X.copy()\n\n#--------------------------------\n# Preprocessing & Quality Control\n#--------------------------------\n\n# Apply basic quality control filters to remove low-quality cells and genes.\nprint(\"Performing quality control...\")\nsc.pp.filter_cells(adata, min_genes=200)\nsc.pp.filter_genes(adata, min_cells=3)\n\n# Normalize each cell by total counts and apply a log transformation.\n# This is a standard preprocessing pipeline for many single-cell analyses.\nsc.pp.normalize_total(adata)\nsc.pp.log1p(adata)\n\n#--------------------------------\n# Major Analysis Task: Ligand-Receptor Interaction Analysis\n#--------------------------------\n\n# --- Define Cell Populations of Interest ---\n# Specify the cell types that will act as senders (source) and receivers (target)\n# in the communication analysis.\nsource_cell_types = [\"CD4 T cells\", \"B cells\", \"FCGR3A+ Monocytes\"]\ntarget_cell_types = [\"CD8 T cells\", \"NK cells\"]\n\n# --- Run LIANA with CellPhoneDB Method ---\n# Use the CellPhoneDB method within the LIANA framework to infer interactions.\n# `groupby=\"cell_type\"` specifies the column with cell annotations.\n# `use_raw=False` indicates that the normalized data in `adata.X` should be used.\nprint(\"\\nRunning ligand-receptor interaction analysis with CellPhoneDB...\")\nfrom liana.method import cellphonedb\ncellphonedb(\nadata,\ngroupby=\"cell_type\",\nuse_raw=False,\nreturn_all_lrs=True,\nverbose=True\n)\n\n# --- Filter and Extract Results ---\n# The results are stored in the `adata.uns['liana_res']` slot.\nresults = adata.uns[\"liana_res\"]\n\n# Filter for significant ligand-receptor interactions based on the p-value.\nsignificant_results = results[results[\"cellphone_pvals\"] <= 0.01]\nprint(f\"\\nFound {len(significant_results)} significant interactions (p-value <= 0.01).\")\n\n\n#--------------------------------\n# Plotting / Visualization\n#--------------------------------\n\n# Generate a dot plot to visualize the top significant ligand-receptor interactions.\n# - `colour`: The color of the dots represents the mean expression of the LR pair.\n# - `size`: The size of the dots represents the significance (p-value).\n# - `source_labels` & `target_labels`: Filter the plot to show interactions between specific cell types.\n# - `top_n`: Display the top 20 interactions based on mean expression.\nprint(\"\\nGenerating dot plot of top interactions...\")\nli.pl.dotplot(\nadata=adata,\ncolour=\"lr_means\",\nsize=\"cellphone_pvals\",\ninverse_size=True, # Make smaller p-values correspond to larger dots\nsource_labels=source_cell_types,\ntarget_labels=target_cell_types,\nfilterby=\"cellphone_pvals\",\nfilter_lambda=lambda x: x <= 0.01,\norderby=\"lr_means\",\norderby_ascending=False,\ntop_n=20,\nfigure_size=(9, 5),\nsize_range=(1, 6)\n)\nplt.show()\n\n#--------------------------------\n# Data Output\n#--------------------------------\n\n# Save the significant interaction results to a CSV file for further inspection.\nprint(\"\\nSaving significant interaction results...\")\nsignificant_results.to_csv(\"liana_significant_interactions.csv\")\n\n# Save the AnnData object, which now contains the LIANA results in the .uns slot.\nadata.write(\"kang_processed_with_liana.h5ad\")\n\nprint(\"\\nLigand-receptor interaction analysis completed!\")\n",
         "### METHODS  \n\n#### **Data Loading and Preprocessing**  \nTo analyze cell-cell communication via ligand-receptor (LR) interactions, we processed a publicly available single-cell RNA-seq dataset from Kang et al., containing ~25k peripheral blood mononuclear cells (PBMCs). The dataset was loaded as an `AnnData` object using Scanpy (`sc.read()`) with a backup URL provided for reproducibility. Raw counts were stored in a separate layer (`adata.layers[\"counts\"]`) before normalization.  \n\nQuality control was performed by filtering out cells expressing fewer than 200 genes and genes detected in fewer than 3 cells (`sc.pp.filter_cells()`, `sc.pp.filter_genes()`). Counts were then library-size-normalized (`sc.pp.normalize_total()`) and log-transformed (`sc.pp.log1p()`), following standard workflows for single-cell transcriptomics.  \n\n#### **Ligand-Receptor Interaction Analysis**  \nWe inferred LR interactions using the **LIANA** framework with the **CellPhoneDB** method (`liana.method.cellphonedb`). The analysis was restricted to predefined cell populations: **CD4 T cells, B cells, and FCGR3A+ monocytes** as potential sender (source) populations, and **CD8 T cells and NK cells** as receivers (target). The `groupby=\"cell_type\"` parameter specified the column containing cell annotations, and normalized (log-transformed) data was used (`use_raw=False`).  \n\nSignificant interactions were filtered at a p-value threshold of 0.01 (`cellphone_pvals ≤ 0.01`), and the results were stored in `adata.uns[\"liana_res\"]`.  \n\n#### **Visualization and Output**  \nA dot plot (`li.pl.dotplot()`) was generated to visualize the top 20 significant interactions:  \n- Dot **color** represented LR pair mean expression (`lr_means`).  \n- Dot **size** scaled inversely with p-values (`inverse_size=True`), highlighting more significant interactions.  \n- The plot was restricted to interactions between the specified source and target cell types.  \n\nResults were saved as a CSV file (`liana_significant_interactions.csv`), and the processed `AnnData` object, including LIANA outputs, was stored (`kang_processed_with_liana.h5ad`) for downstream use.  \n\nAll analyses were conducted in **Python** using **Scanpy** (v1.9) and **LIANA** (v0.1.8)."
        ],
        [
         "40",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport scanpy as sc\n\nimport liana as li\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the pbmc68k_reduced dataset, a built-in example from Scanpy.\n\nprint(\"Loading pbmc68k_reduced dataset...\")\n\nadata = sc.datasets.pbmc68k_reduced()\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Ligand-Receptor Interaction Analysis\n\n#--------------------------------\n\n\n\n# --- Run LIANA with CellPhoneDB Method ---\n\n# Use the CellPhoneDB method within the LIANA framework to infer interactions.\n\n# `groupby='bulk_labels'` specifies the column with cell annotations.\n\n# `resource_name='consensus'` uses a curated list of LR pairs.\n\n# `key_added='cpdb_res'` stores the results in the .uns slot of the AnnData object.\n\nprint(\"\\nRunning ligand-receptor interaction analysis with CellPhoneDB...\")\n\nli.cellphonedb(\n\nadata,\n\ngroupby='bulk_labels',\n\nresource_name='consensus',\n\nexpr_prop=0.1, # Minimum fraction of cells required to express ligand/receptor\n\nverbose=True,\n\nkey_added='cpdb_res'\n\n)\n\n\n\n# --- Inspect Results ---\n\n# Print the first few rows of the result DataFrame to check the output.\n\nprint(\"\\nCellPhoneDB Results (first 5 rows):\")\n\nprint(adata.uns['cpdb_res'].head())\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Cell Clusters ---\n\n# Visualize the cell clusters to get an overview of the data structure.\n\nsc.pl.umap(adata, color='bulk_labels', title='UMAP of PBMC68k Cell Types', frameon=False)\n\n\n\n# --- Dot Plot of Top Interactions ---\n\n# Generate a dot plot to visualize significant ligand-receptor interactions.\n\n# - Color represents the interaction score (lr_means).\n\n# - Size represents the statistical significance (p-value).\n\nprint(\"\\nGenerating dot plot of top interactions...\")\n\nli.pl.dotplot(\n\nadata=adata,\n\ncolour='lr_means',\n\nsize='cellphone_pvals',\n\ninverse_size=True, # Smaller p-values result in larger dots\n\nsource_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n\ntarget_labels=['CD34+', 'CD56+ NK'],\n\nfigure_size=(8, 7),\n\nfilter_fun=lambda x: x['cellphone_pvals'] <= 0.05, # Filter for significant interactions\n\nuns_key='cpdb_res'\n\n)\n\nplt.show()\n\n\n\n# --- Tile Plot of Expression Statistics ---\n\n# Generate a tile plot to show detailed expression statistics (mean expression and\n\n# proportion of expressing cells) for the top 10 interactions.\n\nprint(\"\\nGenerating tile plot of expression statistics...\")\n\nli.pl.tileplot(\n\nadata=adata,\n\nfill='means',\n\nlabel='props',\n\nlabel_fun=lambda x: f'{x:.2f}',\n\ntop_n=10,\n\norderby='cellphone_pvals',\n\norderby_ascending=True,\n\nsource_labels=['CD34+', 'CD56+ NK', 'CD14+ Monocyte'],\n\ntarget_labels=['CD34+', 'CD56+ NK'],\n\nuns_key='cpdb_res',\n\nsource_title='Ligand',\n\ntarget_title='Receptor',\n\nfigure_size=(8, 7)\n\n)\n\nplt.show()\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the LIANA results in the .uns slot.\n\nprint(\"\\nSaving annotated data object...\")\n\nadata.write(\"pbmc68k_with_liana_results.h5ad\")\n\n\n\nprint(\"\\nLigand-receptor interaction analysis completed!\")\n\n",
         "**Methods: Ligand-Receptor Interaction Analysis in Single-Cell RNA-Seq Data Using LIANA and Scanpy**  \n\n### **1. Data Import and Preparation**  \nThe analysis was performed in Python using the **Scanpy** (sc) and **LIANA** (li) packages for single-cell data handling and ligand-receptor (LR) interaction inference. The **matplotlib.pyplot** (plt) module was used for visualization.  \n\n- **Data Loading**: A preprocessed dataset (`pbmc68k_reduced`) was loaded from Scanpy’s built-in datasets. This dataset contains peripheral blood mononuclear cells (PBMCs) with annotated cell types (`bulk_labels`), stored as an **AnnData** object (`adata`).  \n\n### **2. Ligand-Receptor Interaction Analysis**  \nLR interactions were inferred using the **CellPhoneDB** method implemented in LIANA, which predicts biologically relevant interactions based on ligand and receptor co-expression.  \n\n- **Parameters**:  \n  - `groupby='bulk_labels'`: Cell annotations were used to group interactions by cell type.  \n  - `resource_name='consensus'`: A curated database of LR pairs was used.  \n  - `expr_prop=0.1`: Only ligand/receptor genes detected in ≥10% of cells per group were considered.  \n  - Results were stored in `adata.uns['cpdb_res']` for downstream analysis.  \n\n### **3. Visualization of Results**  \nThree plots were generated to summarize the LR interactions:  \n\n#### **UMAP of Cell Clusters**  \nA UMAP projection colored by `bulk_labels` provided an overview of cell-type distribution.  \n\n#### **Dot Plot of Top Interactions**  \nSignificant interactions (p ≤ 0.05) were visualized as a dot plot:  \n- **Color**: Interaction strength (`lr_means`).  \n- **Dot size**: Inverse p-value (smaller p-values = larger dots).  \n- **Filters**: Included interactions between `CD34+`, `CD56+ NK`, and `CD14+ Monocyte` cell types.  \n\n#### **Tile Plot of Expression Statistics**  \nThe top 10 interactions (ranked by p-value) were displayed as a tile plot:  \n- **Fill**: Mean expression (`means`).  \n- **Labels**: Proportion of expressing cells (`props`).  \n- **Annotations**: Ligand and receptor cell types were labeled.  \n\n### **4. Data Output**  \nThe AnnData object, augmented with LR results, was saved as `pbmc68k_with_liana_results.h5ad` for future use.  \n\n**Rationale**:  \n- **CellPhoneDB**: Provides a robust, curated LR interaction resource.  \n- **Expression Threshold (`expr_prop`)**: Ensures biologically plausible interactions by filtering lowly expressed genes.  \n- **Visualizations**: Highlight both statistical significance and expression patterns of LR pairs.  \n\nThis workflow enables reproducible inference and interpretation of cell-cell communication in single-cell data."
        ],
        [
         "41",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport pandas as pd\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport seaborn as sns\n\nimport scanpy as sc\n\nimport liana as li\n\nimport decoupler as dc\n\nimport cell2cell as c2c\n\nimport warnings\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.filterwarnings('ignore')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a sample dataset. Here we use a built-in dataset from LIANA's testing suite.\n\n# NOTE: Replace this with your own data as needed.\n\nprint(\"Loading Kang et al. 2018 dataset...\")\n\nadata = li.testing.datasets.kang_2018()\n\n\n\n#--------------------------------\n\n# Preprocessing & Quality Control\n\n#--------------------------------\n\n\n\n# Apply basic quality control filters.\n\nprint(\"Performing quality control and preprocessing...\")\n\nsc.pp.filter_cells(adata, min_genes=200)\n\nsc.pp.filter_genes(adata, min_cells=3)\n\n\n\n# Normalize and log-transform the data.\n\nsc.pp.normalize_total(adata)\n\nsc.pp.log1p(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# --- 1. Ligand-Receptor Inference by Sample ---\n\n# Define keys for sample, condition, and cell type columns in adata.obs.\n\nsample_key = 'sample'\n\ncondition_key = 'condition'\n\ngroupby = 'cell_type'\n\n\n\n# Run LIANA's rank aggregation method for each sample individually.\n\n# This infers ligand-receptor interactions within each sample before downstream analysis.\n\nprint(\"\\nRunning ligand-receptor inference for each sample...\")\n\nliana_results = li.mt.rank_aggregate.by_sample(\n\nadata,\n\ngroupby=groupby,\n\nresource_name='consensus', # Use a consensus database of LR pairs\n\nsample_key=sample_key,\n\nuse_raw=False,\n\nverbose=True,\n\nn_perms=None, # Set to an integer (e.g., 1000) for permutation testing\n\nreturn_all_lrs=True\n\n)\n\n\n\n# Store the LIANA results in the AnnData object.\n\nadata.uns[\"liana_res\"] = liana_results\n\n\n\n# Display the top-ranked interactions.\n\nprint(\"\\nTop aggregated ligand-receptor interactions:\")\n\nprint(adata.uns[\"liana_res\"].sort_values(\"magnitude_rank\").head())\n\n\n\n# --- 2. Tensor Decomposition for Communication Analysis ---\n\n# Construct a communication tensor. This is a multi-dimensional array representing\n\n# interactions (Context x Sender x Receiver x LR-pair).\n\nprint(\"\\nConstructing communication tensor...\")\n\ntensor = li.multi.to_tensor_c2c(\n\nadata,\n\nsample_key=sample_key,\n\nscore_key=\"magnitude_rank\",\n\nhow=\"outer_cells\" # Use all cells for tensor construction\n\n)\n\n\n\n# Perform tensor decomposition to identify latent communication patterns (factors).\n\n# `rank=6` specifies the number of factors to extract.\n\nprint(\"\\nPerforming tensor decomposition...\")\n\ntensor = c2c.analysis.run_tensor_cell2cell_pipeline(\n\ntensor,\n\ncopy_tensor=True,\n\nrank=6,\n\ntf_optimization='regular',\n\nrandom_state=0,\n\ndevice=\"cpu\" # Use 'cuda' if a GPU is available\n\n)\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP Visualization ---\n\n# Plot a UMAP to visualize the overall structure of the data by condition and cell type.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=[condition_key, groupby], frameon=False)\n\n\n\n# --- Tensor Factor Plots ---\n\n# Visualize the results of the tensor decomposition. Each factor represents a\n\n# distinct communication pattern across contexts, senders, receivers, and LR pairs.\n\nfactors, axes = c2c.plotting.tensor_factors_plot(\n\ninteraction_tensor=tensor,\n\nfontsize=10\n\n)\n\nplt.show()\n\n\n\n# --- Communication Network Plots ---\n\n# Visualize the cell-cell communication networks for specific factors.\n\n# Here, we plot the network associated with 'Factor 6' as an example.\n\nc2c.plotting.ccc_networks_plot(factors, included_factors=['Factor 6'])\n\nplt.show()\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final plots and data objects.\n\nprint(\"\\nSaving results...\")\n\n# Save the tensor factor plot\n\nfactors[0].figure.savefig(\"tensor_factors_plot.png\", dpi=300, bbox_inches='tight')\n\n# Save the AnnData object with LIANA results\n\nadata.write(\"kang_2018_with_liana.h5ad\")\n\n# Save the tensor object with decomposition results\n\nc2c.io.export_variable_to_file(tensor, 'communication_tensor.c2c')\n\n\n\nprint(\"\\nTensor-based communication analysis completed!\")\n\n",
         "### **Methods**\n\n#### **Computational Analysis of Cell-Cell Communication via Tensor Decomposition**\n\n**Programming Language:**  \nThe analysis was implemented in Python using a suite of bioinformatics libraries for single-cell data analysis, ligand-receptor interaction inference, and tensor decomposition.\n\n---\n\n### **1. Library Import and Configuration**\n**Purpose:**  \nImport required modules for data manipulation, visualization, and single-cell communication analysis.  \n\n**Steps:**  \n- **Core libraries:**  \n  - `pandas` and `numpy` for tabular data handling.  \n  - `matplotlib` and `seaborn` for visualization.  \n- **Single-cell analysis tools:**  \n  - `scanpy` (SC) for preprocessing and dimensionality reduction.  \n  - `liana` (LIANA) for ligand-receptor (LR) interaction inference.  \n  - `decoupler` for pathway analysis.  \n  - `cell2cell` (C2C) for tensor-based communication modeling.  \n- **Warnings:** Suppressed to avoid cluttering output (`warnings.filterwarnings('ignore')`).\n\n---\n\n### **2. Data Input**\n**Purpose:**  \nLoad and initialize the dataset for analysis.  \n\n**Procedure:**  \n- A built-in test dataset from the LIANA package (*Kang et al. 2018*) was loaded as an `AnnData` object (`adata`).  \n  - **Note:** Users should replace this with their own annotated single-cell RNA-seq data (stored in `AnnData` format).  \n\n---\n\n### **3. Preprocessing & Quality Control**\n**Purpose:**  \nFilter low-quality cells/genes and normalize data for downstream analysis.  \n\n**Steps:**  \n1. **Cell/Genes Filtering:**  \n   - Retain cells expressing ≥200 genes (`min_genes=200`).  \n   - Retain genes detected in ≥3 cells (`min_cells=3`).  \n2. **Normalization:**  \n   - Total-count normalization (`sc.pp.normalize_total`) followed by log1p transformation (`sc.pp.log1p`) to stabilize variance.  \n\n---\n\n### **4. Ligand-Receptor (LR) Interaction Inference**\n**Purpose:**  \nIdentify LR interactions across cell types and samples using a consensus database.  \n\n**Procedure:**  \n- **Sample-wise Analysis:** LIANA’s rank aggregation method (`li.mt.rank_aggregate.by_sample`) was applied per sample (`sample_key='sample'`), grouping by cell type (`groupby='cell_type'`).  \n  - **Parameters:**  \n    - `resource_name='consensus'`: Integrates multiple LR databases.  \n    - `use_raw=False`: Uses normalized data.  \n    - `n_perms=None`: Permutation testing disabled (set to an integer, e.g., `1000`, to enable).  \n- **Output:** Results stored in `adata.uns[\"liana_res\"]` and ranked by interaction magnitude (`magnitude_rank`).  \n\n---\n\n### **5. Tensor Construction and Decomposition**\n**Purpose:**  \nModel cell-cell communication as a 4D tensor (Context × Sender × Receiver × LR-pairs) and decompose it into latent factors.  \n\n**Steps:**  \n1. **Tensor Construction:**  \n   - Convert LIANA results to a tensor (`li.multi.to_tensor_c2c`) using all cells (`how=\"outer_cells\"`) and magnitude ranks (`score_key=\"magnitude_rank\"`).  \n2. **Tensor Decomposition:**  \n   - Perform non-negative tensor factorization (`c2c.analysis.run_tensor_cell2cell_pipeline`) with:  \n     - `rank=6`: Six latent communication factors.  \n     - `device=\"cpu\"` (use `\"cuda\"` for GPU acceleration).  \n\n---\n\n### **6. Visualization**\n**Purpose:**  \nExplore data structure and communication patterns.  \n\n**Plots Generated:**  \n1. **UMAP:** Visualize cell clusters by condition and cell type (`sc.pl.umap`).  \n2. **Tensor Factors:** Heatmaps of decomposed factors (`c2c.plotting.tensor_factors_plot`).  \n3. **Communication Networks:** Network plot for a representative factor (e.g., `Factor 6`; `c2c.plotting.ccc_networks_plot`).  \n\n---\n\n### **7. Data Output**\n**Purpose:**  \nSave results for reproducibility and further analysis.  \n\n**Output Files:**  \n- **Plots:** Tensor factor heatmap (`tensor_factors_plot.png`).  \n- **Data Objects:**  \n  - `AnnData` with LR results (`kang_2018_with_liana.h5ad`).  \n  - Serialized tensor object (`communication_tensor.c2c`).  \n\n---\n\n### **Key Assumptions & Requirements**  \n- **Input Data:** Requires an `AnnData` object with cell metadata (e.g., `sample`, `cell_type`, `condition`).  \n- **Compatibility:** Libraries (e.g., `liana==0.1.7`, `cell2cell==0.7.0`) must match versions used here.  \n- **Hardware:** Defaults to CPU; GPU acceleration (`device=\"cuda\"`) recommended for large datasets.  \n\nThis pipeline enables systematic analysis of cell-cell communication by integrating LR inference with tensor-based decomposition, revealing context-specific interaction patterns."
        ],
        [
         "42",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport pandas as pd\n\nimport numpy as np\n\nimport matplotlib.pyplot as plt\n\nimport seaborn as sns\n\nimport scanpy as sc\n\nimport liana as li\n\nimport decoupler as dc\n\nimport omnipath as op\n\nfrom pydeseq2.dds import DeseqDataSet\n\nfrom pydeseq2.ds import DeseqStats\n\nimport warnings\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.filterwarnings('ignore')\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a sample dataset. Here we use a built-in dataset from LIANA's testing suite.\n\n# NOTE: Replace this with your own data as needed.\n\nprint(\"Loading Kang et al. 2018 dataset...\")\n\nadata = li.testing.datasets.kang_2018()\n\n\n\n#--------------------------------\n\n# Preprocessing & Quality Control\n\n#--------------------------------\n\n\n\n# Apply basic quality control filters.\n\nprint(\"Performing quality control and preprocessing...\")\n\nsc.pp.filter_cells(adata, min_genes=200)\n\nsc.pp.filter_genes(adata, min_cells=3)\n\n\n\n# Normalize and log-transform data for visualization and some downstream steps.\n\nsc.pp.normalize_total(adata, target_sum=1e4)\n\nsc.pp.log1p(adata)\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# Define metadata columns of interest for the analysis\n\nsample_key = 'sample'\n\ncondition_key = 'condition'\n\ngroupby = 'cell_abbr' # Using abbreviated cell types for clarity\n\n\n\n# --- 1. Pseudobulk and Differential Expression Analysis (DEA) ---\n\n# Generate pseudobulk profiles by summing counts for each cell type within each sample.\n\nprint(\"\\nGenerating pseudobulk profiles...\")\n\npdata = dc.get_pseudobulk(\n\nadata,\n\nsample_col=sample_key,\n\ngroups_col=groupby,\n\nlayer='counts', # Use raw counts for DESeq2\n\nmode='sum',\n\nmin_cells=10,\n\nmin_counts=10000\n\n)\n\n\n\n# Perform DEA using PyDESeq2 for each cell type.\n\nprint(\"\\nPerforming differential expression analysis with PyDESeq2...\")\n\ndea_results = {}\n\nfor cell_group in pdata.obs[groupby].unique():\n\n# Subset pseudobulk data to the current cell type\n\nctdata = pdata[pdata.obs[groupby] == cell_group].copy()\n\n\n\n# Filter out lowly expressed genes for the current cell type\n\ngenes = dc.filter_by_expr(ctdata, group=condition_key, min_count=5, min_total_count=10)\n\nctdata = ctdata[:, genes].copy()\n\n\n\n# Build and run the DESeq2 model\n\ndds = DeseqDataSet(\n\nadata=ctdata,\n\ndesign_factors=condition_key,\n\nref_level=[condition_key, 'ctrl'], # Set 'ctrl' as the reference condition\n\nrefit_cooks=True,\n\nquiet=True\n\n)\n\ndds.deseq2()\n\n\n\n# Get statistics and shrink log-fold changes for robustness\n\nstat_res = DeseqStats(dds, contrast=[condition_key, 'stim', 'ctrl'])\n\nstat_res.summary()\n\nstat_res.lfc_shrink(coeff='condition_stim_vs_ctrl')\n\n\n\ndea_results[cell_group] = stat_res.results_df\n\n\n\n# Concatenate all DEA results into a single DataFrame.\n\ndea_df = pd.concat(dea_results).reset_index().rename(\n\ncolumns={'level_0': groupby, 'level_1': 'gene'}\n\n).set_index('gene')\n\n\n\n\n\n# --- 2. Deregulated Ligand-Receptor Interaction Analysis ---\n\n# Use LIANA to identify LR interactions that are significantly altered between conditions,\n\n# based on the differential expression statistics calculated above.\n\nprint(\"\\nIdentifying deregulated ligand-receptor interactions...\")\n\nlr_res = li.multi.df_to_lr(\n\nadata,\n\ndea_df=dea_df,\n\nresource_name='consensus',\n\nexpr_prop=0.1,\n\ngroupby=groupby,\n\nstat_keys=['stat', 'pvalue', 'padj'],\n\nuse_raw=False, # Use log-normalized data for expression proportion check\n\ncomplex_col='stat',\n\nreturn_all_lrs=False\n\n)\n\n\n\n# Sort the results by the interaction statistic.\n\nlr_res = lr_res.sort_values(\"interaction_stat\", ascending=False)\n\n\n\n\n\n# --- 3. Pathway Enrichment Analysis ---\n\n# Use the DEA results to infer transcription factor (TF) activity changes.\n\nprint(\"\\nPerforming pathway enrichment analysis (TF activity)...\")\n\n# Get a TF-target regulatory network from Omnipath's CollecTRI resource.\n\nnet = dc.get_collectri()\n\n\n\n# Pivot the DEA results to a wide format (cells x genes) with stat values.\n\ndea_wide = dea_df[[groupby, 'stat']].reset_index(names='genes').pivot(\n\nindex=groupby, columns='genes', values='stat'\n\n).fillna(0)\n\n\n\n# Run Univariate Linear Model (ULM) from Decoupler to estimate TF activities.\n\nestimates, pvals = dc.run_ulm(mat=dea_wide, net=net)\n\n\n\n# Print the top TFs associated with the 'CD14' Monocyte group as an example.\n\ntop_tfs = estimates.T.sort_values('CD14', key=abs, ascending=False).head()\n\nprint(\"\\nTop TFs associated with CD14 Monocytes:\")\n\nprint(top_tfs)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP Visualization ---\n\n# Plot a UMAP to visualize the overall structure of the data.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=[condition_key, sample_key, 'cell_type', groupby], frameon=False, ncols=2)\n\n\n\n# --- Tile Plot of Top Interactions ---\n\n# Visualize the top 15 deregulated ligand-receptor interactions.\n\n# Fill color represents expression, while '*' indicates statistical significance.\n\nli.pl.tileplot(\n\nliana_res=lr_res,\n\nfill='expr',\n\nlabel='padj',\n\nlabel_fun=lambda x: '*' if x < 0.05 else np.nan,\n\ntop_n=15,\n\norderby='interaction_stat',\n\norderby_ascending=False,\n\nsource_title='Ligand',\n\ntarget_title='Receptor'\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the key results DataFrames to CSV files.\n\nprint(\"\\nSaving analysis results...\")\n\ndea_df.to_csv(\"differential_expression_results.csv\")\n\nlr_res.to_csv(\"deregulated_lr_interactions.csv\")\n\nestimates.to_csv(\"tf_activity_estimates.csv\")\n\n\n\n# Save the processed AnnData object.\n\nadata.write(\"kang_processed_for_dea.h5ad\")\n\n\n\nprint(\"\\nDifferential communication and pathway analysis completed!\")\n\n",
         "**Methods**  \n\nThe analysis pipeline was implemented in Python using a suite of bioinformatics tools for single-cell RNA sequencing (scRNA-seq) data processing, differential expression analysis, ligand-receptor interaction assessment, and pathway enrichment. Below is a step-by-step description of the methodology.  \n\n### **1. Library Imports**  \nEssential Python modules were imported for data manipulation (`pandas`, `numpy`), visualization (`matplotlib.pyplot`, `seaborn`), and single-cell analysis (`scanpy`, `liana`, `decoupler`, `omnipath`). The `pydeseq2` package was used for differential expression analysis, and warnings were suppressed to streamline output.  \n\n### **2. Data Input**  \nA built-in dataset from the `liana` testing suite (Kang et al., 2018) was loaded as an `AnnData` object. Users can substitute this with their own dataset.  \n\n### **3. Preprocessing & Quality Control**  \nLow-quality cells and genes were filtered out to retain only cells expressing at least 200 genes and genes detected in at least 3 cells. Counts were normalized to 10,000 transcripts per cell and log-transformed for downstream analyses.  \n\n### **4. Major Analysis Tasks**  \n\n#### **4.1 Pseudobulk & Differential Expression Analysis (DEA)**  \n- Pseudobulk profiles were generated by summing raw counts per cell type (`cell_abbr`) within each sample (`sample_key`), retaining cell types with ≥10 cells and samples with ≥10,000 total counts.  \n- Differential expression was performed using **PyDESeq2** for each cell type under the contrast `stim` vs. `ctrl`. Genes were pre-filtered (≥5 counts in ≥10 cells per condition), and log-fold changes were shrunk for robustness. Results were compiled into a single dataframe (`dea_df`).  \n\n#### **4.2 Deregulated Ligand-Receptor (LR) Interaction Analysis**  \nUsing **LIANA**, statistically altered LR interactions were identified based on DEA results (`stat`, `pvalue`, `padj`). The `consensus` resource was used, and interactions were filtered to those expressed in ≥10% of cells. Results were sorted by interaction significance (`interaction_stat`).  \n\n#### **4.3 Pathway Enrichment (TF Activity Inference)**  \n- Transcription factor (TF) activity was inferred using the **CollecTRI** network (from Omnipath) and the **Univariate Linear Model (ULM)** in `decoupler`. TF activity scores were derived from the DEA statistics matrix (cell types × genes). Top TFs per cell type (e.g., **CD14 Monocytes**) were reported.  \n\n### **5. Visualization**  \n- **UMAP embeddings** were plotted to visualize cell distribution by condition, sample, and cell type.  \n- **Tile plots** displayed the top 15 deregulated LR interactions, colored by expression (`expr`) and annotated with significance (`padj < 0.05`).  \n\n### **6. Data Output**  \nKey results (DEA statistics, LR interactions, TF activities) were saved as CSV files, and the processed `AnnData` object was stored in H5AD format for reproducibility.  \n\nThis pipeline provides a comprehensive workflow for single-cell communication analysis, integrating differential expression, LR interaction scoring, and pathway enrichment."
        ],
        [
         "43",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for single-cell analysis, pathway analysis, and plotting\n\nimport scanpy as sc\n\nimport decoupler as dc\n\nimport matplotlib.pyplot as plt\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.set_figure_params(figsize=(5, 5), frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example dataset (PBMC 3k) from Decoupler's built-in datasets.\n\n# This dataset is pre-processed with clustering information.\n\nprint(\"Loading pbmc3k dataset...\")\n\nadata = dc.ds.pbmc3k()\n\n\n\n# Display a summary of the loaded dataset.\n\nprint(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Pathway Enrichment\n\n#--------------------------------\n\n\n\n# --- Load Pathway Gene Sets ---\n\n# Get the PROGENy model, which contains a curated set of pathway gene sets\n\n# with signed weights representing up/down-regulation.\n\nprint(\"\\nLoading PROGENy pathway model...\")\n\nprogeny = dc.op.progeny(organism='human')\n\n\n\n# --- Compute Enrichment Scores ---\n\n# Run the Univariate Linear Model (ulm) from Decoupler to compute pathway\n\n# enrichment scores for each cell.\n\n# The result is stored in `adata.obsm['ulm_estimate']`.\n\nprint(\"Computing pathway enrichment scores...\")\n\ndc.run_ulm(data=adata, net=progeny, source='source', target='target', weight='weight', min_n=5)\n\n\n\n# Extract the computed enrichment scores into a separate AnnData object for easier plotting.\n\n# This creates a new object where the .X matrix contains the pathway scores.\n\nscores = dc.get_obsm_to_anndata(adata, obsm_key='ulm_estimate')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Cell Clusters ---\n\n# Visualize the original cell clusters from the dataset.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color='leiden', title='UMAP of PBMC Clusters')\n\n\n\n# --- UMAP of Pathway Activity ---\n\n# Visualize the activity of a specific pathway (e.g., TNFa signaling) on the UMAP.\n\n# Red indicates higher pathway activity, and blue indicates lower activity.\n\nsc.pl.umap(scores, color=['TNFa'], cmap='RdBu_r', vcenter=0, title='TNFa Pathway Activity')\n\n\n\n# --- Violin Plot of Pathway Activity ---\n\n# Compare the distribution of pathway activity scores across different cell types.\n\nsc.pl.violin(scores, keys=['TNFa'], groupby='celltype', rotation=90)\n\n\n\n# --- Matrix Plot of All Pathways ---\n\n# Show a heatmap of all pathway activities across all cell types.\n\n# The scores are z-scaled for better comparability across pathways.\n\nsc.pl.matrixplot(\n\nadata=scores,\n\nvar_names=scores.var_names,\n\ngroupby='celltype',\n\ndendrogram=True,\n\nstandard_scale='var',\n\ncolorbar_title='Z-scaled scores',\n\ncmap='RdBu_r'\n\n)\n\nplt.show()\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the AnnData object, which now contains the pathway enrichment scores.\n\nprint(\"\\nSaving data with enrichment scores...\")\n\n# Note: The scores are in adata.obsm['ulm_estimate'], and adata.obsm['ulm_pvals']\n\nadata.write(\"pbmc3k_with_progeny_scores.h5ad\")\n\n\n\nprint(\"\\nPathway enrichment analysis completed!\")\n\n",
         "### Methods  \n\n#### **Programming Language and Tools**  \nThe analysis was implemented in **Python**, using the following bioinformatics packages:  \n- **Scanpy** (sc) for single-cell data manipulation and visualization.  \n- **Decoupler** (dc) for pathway enrichment analysis.  \n- **Matplotlib.pyplot** (plt) for additional plotting support.  \n\n#### **Data Loading and Preprocessing**  \n1. **Dataset Acquisition**:  \n   - A preprocessed Peripheral Blood Mononuclear Cell (PBMC 3k) dataset was loaded using `dc.ds.pbmc3k()`. This dataset includes clustering annotations (`leiden`) and cell-type labels (`celltype`).  \n\n2. **Visualization Setup**:  \n   - Default plotting parameters were configured with `sc.set_figure_params(figsize=(5, 5), frameon=False)` to ensure consistent and publication-ready figures.  \n\n#### **Pathway Enrichment Analysis**  \n1. **Pathway Model Loading**:  \n   - The **PROGENy** model was retrieved using `dc.op.progeny(organism='human')`. This resource contains curated pathway gene sets (e.g., TNFa, NFkB) with signed weights indicating directional regulation (activation/inhibition).  \n\n2. **Enrichment Score Calculation**:  \n   - Pathway activity scores per cell were computed using the **Univariate Linear Model (ULM)** (`dc.run_ulm`). Key parameters:  \n     - `net=progeny`: Input network of pathways.  \n     - `min_n=5`: Minimum number of target genes required per pathway.  \n   - Results were stored in `adata.obsm['ulm_estimate']` (scores) and `adata.obsm['ulm_pvals']` (p-values).  \n   - Scores were extracted into a dedicated `AnnData` object (`scores`) for streamlined visualization.  \n\n#### **Visualization**  \n1. **Cell Clustering**:  \n   - A UMAP plot (`sc.pl.umap`) colored by `leiden` clusters illustrated the dataset’s intrinsic structure.  \n\n2. **Pathway Activity**:  \n   - **UMAP Projection**: Activity of the TNFa pathway was visualized on UMAP (`cmap='RdBu_r'`, with red/blue indicating high/low activity).  \n   - **Violin Plots**: Pathway score distributions across cell types were compared using `sc.pl.violin`.  \n   - **Matrix Heatmap**: A z-score-scaled heatmap (`sc.pl.matrixplot`) displayed all pathway activities across cell types, with hierarchical clustering (`dendrogram=True`).  \n\n#### **Output**  \nThe final `AnnData` object, including pathway scores, was saved as `pbmc3k_with_progeny_scores.h5ad` for downstream use.  \n\nThis pipeline provides a reproducible workflow for single-cell pathway enrichment analysis, integrating gene set activity with cell-type-specific interpretations.  \n\n---  \n*Key Parameters*:  \n- `min_n=5` (minimum genes per pathway).  \n- `cmap='RdBu_r'` (divergent colormap for signed scores).  \n- `standard_scale='var'` (z-scaling in heatmaps).  \n\n*Assumptions*:  \n- Input data is an `AnnData` object with precomputed UMAP and cell annotations.  \n- PROGENy pathways are human-specific (`organism='human'`)."
        ],
        [
         "44",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport pandas as pd\n\nimport scanpy as sc\n\nimport decoupler as dc\n\nimport matplotlib.pyplot as plt\n\nfrom pydeseq2.dds import DeseqDataSet, DefaultInference\n\nfrom pydeseq2.ds import DeseqStats\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.set_figure_params(figsize=(4, 4), frameon=False)\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example single-cell RNA sequencing dataset.\n\n# Here, we use a COVID-19 dataset from Decoupler's built-in datasets.\n\nprint(\"Loading covid5k dataset...\")\n\nadata = dc.ds.covid5k()\n\n\n\n# Display a summary of the loaded dataset.\n\nprint(adata)\n\n\n\n\n\n#--------------------------------\n\n# Preprocessing and Pseudobulking\n\n#--------------------------------\n\n\n\n# --- Pseudobulking ---\n\n# Aggregate single-cell expression data into pseudobulk profiles.\n\n# This sums the counts for each cell type within each individual sample.\n\nprint(\"\\nGenerating pseudobulk profiles...\")\n\npdata = dc.pp.pseudobulk(\n\nadata=adata,\n\nsample_col='individual',\n\ngroups_col='celltype',\n\nmode='sum' # Sum counts for each gene\n\n)\n\n\n\n# --- Pseudobulk QC and Preprocessing ---\n\n# Filter out low-quality pseudobulk samples.\n\ndc.pp.filter_samples(pdata, min_cells=10, min_counts=1000)\n\n\n\n# Normalize and preprocess the pseudobulk data for visualization and PCA.\n\n# Note: Raw counts are kept for DESeq2 analysis.\n\npdata.layers['counts'] = pdata.X.copy()\n\nsc.pp.normalize_total(pdata, target_sum=1e4)\n\nsc.pp.log1p(pdata)\n\nsc.pp.scale(pdata, max_value=10)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks\n\n#--------------------------------\n\n\n\n# --- 1. Differential Expression Analysis (DEA) with PyDESeq2 ---\n\n# Define and run the DESeq2 model on the raw pseudobulk counts.\n\n# The design formula `~disease + sex` models the effect of disease while accounting for sex.\n\nprint(\"\\nPerforming differential expression analysis with PyDESeq2...\")\n\ninference = DefaultInference(n_cpus=8)\n\ndds = DeseqDataSet(\n\nadata=pdata,\n\ndesign_factors=['disease', 'sex'],\n\nrefit_cooks=True,\n\ninference=inference\n\n)\n\ndds.deseq2()\n\n\n\n# Extract statistical results for the contrast between COVID-19 and normal conditions.\n\nstat_res = DeseqStats(\n\ndds,\n\ncontrast=['disease', 'COVID-19', 'normal'],\n\ninference=inference\n\n)\n\nstat_res.summary()\n\nresults_df = stat_res.results_df\n\n\n\n\n\n# --- 2. Pathway Enrichment Analysis ---\n\n# Use the DEA results to infer pathway activity changes with Decoupler.\n\nprint(\"\\nPerforming pathway enrichment analysis...\")\n\n# Get the PROGENy model, a curated set of pathway gene sets.\n\nprogeny = dc.op.progeny(organism='human')\n\n\n\n# Run Univariate Linear Model (ulm) to estimate pathway activities from the DE statistics.\n\npw_acts, pw_padj = dc.run_ulm(\n\nmat=results_df[['stat']].T.rename(index={'stat': 'disease.vs.normal'}),\n\nnet=progeny\n\n)\n\n\n\n# Filter for significantly altered pathways.\n\nsignificant_mask = (pw_padj.T < 0.05).iloc[:, 0]\n\nsignificant_pw_acts = pw_acts.loc[:, significant_mask]\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Single-Cell Data ---\n\n# Visualize the initial cell type and disease state distribution.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=['celltype', 'disease'], ncols=1, title=[\"Cell Types\", \"Disease State\"])\n\n\n\n# --- Pseudobulk QC Plot ---\n\n# Visualize the cell type composition of the filtered pseudobulk samples.\n\ndc.pl.obsbar(adata=pdata, y='celltype', hue='disease', figsize=(6, 3))\n\n\n\n# --- PCA of Pseudobulk Data ---\n\n# Explore the major sources of variation across the pseudobulk samples.\n\nsc.tl.pca(pdata)\n\nsc.pl.pca(pdata, color='disease', size=100, title=\"PCA of Pseudobulk Samples\")\n\n\n\n# --- Volcano Plot of DEA Results ---\n\n# Visualize the differential expression results, highlighting significant genes.\n\ndc.pl.volcano(results_df, x='log2FoldChange', y='pvalue')\n\n\n\n# --- Bar Plot of Pathway Enrichment ---\n\n# Visualize the pathway enrichment results for significantly altered pathways.\n\ndc.pl.barplot(data=significant_pw_acts, name='disease.vs.normal', figsize=(3, 3))\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the key results DataFrames and the processed AnnData object.\n\nprint(\"\\nSaving analysis results...\")\n\nresults_df.to_csv(\"pseudobulk_dea_results.csv\")\n\nsignificant_pw_acts.to_csv(\"significant_pathway_activities.csv\")\n\npdata.write(\"pseudobulk_processed_data.h5ad\")\n\n\n\nprint(\"\\nPseudobulk differential analysis completed!\")\n\n",
         "### **Methods**\n\n#### **1. Software and Library Dependencies**  \nThe analysis was performed in **Python**, utilizing the following libraries:  \n- **NumPy** and **Pandas** for numerical and tabular data manipulation.  \n- **Scanpy** (v1.9) for single-cell RNA sequencing (scRNA-seq) data handling and visualization.  \n- **Decoupler** for pathway enrichment analysis and pseudobulk aggregation.  \n- **PyDESeq2** (v0.4.1) for differential expression analysis (DEA) using the DESeq2 framework.  \n- **Matplotlib** for customized plotting.  \n\nGlobal visualization parameters were set for consistency (`sc.set_figure_params(figsize=(4, 4), frameon=False)`).\n\n#### **2. Data Input and Preprocessing**  \nA publicly available **COVID-19 scRNA-seq dataset** (`covid5k`) was loaded from Decoupler’s built-in datasets. The dataset contained gene expression profiles annotated by `celltype` and `individual` metadata.  \n\n**Pseudobulk aggregation** was performed to summarize single-cell counts into profiles per cell type and individual (`dc.pp.pseudobulk`, `mode='sum'`). Low-quality pseudobulk samples were filtered (`min_cells=10`, `min_counts=1000`), and data were normalized (`sc.pp.normalize_total`, `target_sum=1e4`), log-transformed (`sc.pp.log1p`), and scaled for visualization (`max_value=10`). Raw counts were preserved in a dedicated layer for DEA.\n\n#### **3. Differential Expression Analysis (DEA)**  \nPyDESeq2 was applied to raw pseudobulk counts with a design matrix accounting for `disease` (COVID-19 vs. normal) and `sex` as covariates. Key steps:  \n- **Model fitting**: `DeseqDataSet` with default inference settings (`n_cpus=8` for parallelization).  \n- **Contrast testing**: Statistical comparisons for `disease` effects (`DeseqStats`, contrast: `['disease', 'COVID-19', 'normal']`).  \n- **Output**: Results included log2 fold changes, p-values, and adjusted p-values (FDR).  \n\n#### **4. Pathway Enrichment Analysis**  \nPROGENy’s curated gene sets (human) were used to infer pathway activity changes from DEA statistics (`dc.run_ulm`). Significance was defined as adjusted p-value < 0.05 (`pw_padj.T < 0.05`).\n\n#### **5. Visualization**  \n- **Single-cell UMAP**: Cell type and disease state distributions (`sc.pl.umap`).  \n- **Pseudobulk QC**: Cell type composition by disease (`dc.pl.obsbar`).  \n- **PCA**: Pseudobulk sample clustering (`sc.pl.pca`, color=`disease`).  \n- **DEA results**: Volcano plot of log2 fold changes vs. p-values (`dc.pl.volcano`).  \n- **Pathway activity**: Bar plot of significantly altered pathways (`dc.pl.barplot`).  \n\n#### **6. Output**  \nResults were saved as:  \n- `pseudobulk_dea_results.csv` (DEA statistics).  \n- `significant_pathway_activities.csv` (pathway enrichment).  \n- `pseudobulk_processed_data.h5ad` (processed AnnData object).  \n\nAll steps were executed in a reproducible pipeline, with computational parameters documented above."
        ],
        [
         "45",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport numpy as np\n\nimport scanpy as sc\n\nimport decoupler as dc\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.set_figure_params(figsize=(5, 5), frameon=False)\n\nsc.settings.verbosity = 3 # Set for detailed logging\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the mouse gastrulation dataset from Decoupler's built-in datasets.\n\nprint(\"Loading mouse gastrulation dataset...\")\n\nadata = dc.ds.erygast1k()\n\n\n\n#--------------------------------\n\n# Preprocessing & Trajectory Inference\n\n#--------------------------------\n\n\n\n# --- Preprocessing ---\n\n# The loaded data is already log-normalized. We proceed to dimensionality reduction.\n\n# Compute neighbors using the pre-computed PCA representation.\n\nprint(\"\\nPreprocessing and computing neighbors...\")\n\nsc.pp.neighbors(adata, use_rep='X_pca')\n\n\n\n# --- Trajectory Inference with PAGA/DPT ---\n\n# Compute a diffusion map, which is the basis for DPT.\n\nsc.tl.diffmap(adata)\n\n\n\n# Recompute neighbors on the diffusion map to better capture the manifold.\n\nsc.pp.neighbors(adata, use_rep='X_diffmap')\n\n\n\n# Run Partition-based Graph Abstraction (PAGA) to get a high-level\n\n# view of the connectivity between cell types.\n\nsc.tl.paga(adata, groups='celltype')\n\n\n\n# Define the root of the trajectory for pseudotime calculation.\n\nroot_celltype = \"Blood progenitors 1\"\n\niroot = np.flatnonzero(adata.obs['celltype'] == root_celltype)[0]\n\nadata.uns['iroot'] = iroot\n\n\n\n# Compute Diffusion Pseudotime (DPT) to order cells along the trajectory.\n\nsc.tl.dpt(adata)\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: Enrichment Analysis\n\n#--------------------------------\n\n\n\n# --- 1. Transcription Factor (TF) Activity Analysis ---\n\nprint(\"\\nPerforming TF activity enrichment analysis...\")\n\n# Retrieve the CollecTRI TF-target network for mouse from Decoupler's resources.\n\ncollectri = dc.op.collectri(organism='mouse')\n\n\n\n# Run the Univariate Linear Model (ulm) to estimate TF activity scores for each cell.\n\n# The results are stored in adata.obsm['ulm_estimate'] and adata.obsm['ulm_pvals'].\n\ndc.run_ulm(data=adata, net=collectri, source='source', target='target', weight='weight', min_n=5)\n\n\n\n# Extract the TF activity scores into a separate AnnData object for easier plotting.\n\n# This creates a new object where the .X matrix contains the TF scores.\n\nscore_tf = dc.get_obsm_to_anndata(adata, obsm_key='ulm_estimate')\n\n\n\n\n\n# --- 2. Pathway Activity Analysis ---\n\nprint(\"\\nPerforming pathway activity enrichment analysis...\")\n\n# Retrieve the PROGENy pathway model for mouse.\n\nprogeny = dc.op.progeny(organism='mouse')\n\n\n\n# Run ULM again to estimate pathway activity scores.\n\n# Note: This overwrites the previous results in adata.obsm['ulm_*'], which is why\n\n# we extracted the TF scores to a separate object first.\n\ndc.run_ulm(data=adata, net=progeny, source='source', target='target', weight='weight', min_n=5)\n\n\n\n# Extract the pathway activity scores into another AnnData object.\n\nscore_pathway = dc.get_obsm_to_anndata(adata, obsm_key='ulm_estimate')\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP of Raw Data and Trajectory ---\n\nprint(\"\\n--- Visualizing Results ---\")\n\n# Visualize the initial UMAP colored by cell type and developmental stage.\n\nsc.pl.umap(adata, color=['celltype', 'stage'], ncols=1, title=[\"Cell Types\", \"Developmental Stage\"])\n\n\n\n# Visualize the DPT pseudotime ordering on the UMAP.\n\nsc.pl.umap(adata, color=['dpt_pseudotime'], title=\"Diffusion Pseudotime\")\n\n\n\n\n\n# --- TF Activity Visualization ---\n\n# Choose a transcription factor to visualize, e.g., \"Klf3\".\n\ntf = 'Klf3'\n\n# Plot a UMAP colored by the Klf3 enrichment score.\n\nsc.pl.umap(score_tf, color=[tf], cmap='RdBu_r', vcenter=0, title=f\"{tf} Activity\")\n\n# Plot a violin plot showing the distribution of the Klf3 score across cell types.\n\nsc.pl.violin(score_tf, keys=[tf], groupby='celltype', rotation=90, ylabel=f'{tf} score')\n\n\n\n\n\n# --- Pathway Activity Visualization ---\n\n# Choose a pathway to visualize, e.g., \"PI3K\".\n\npathway = 'PI3K'\n\n# Plot a UMAP colored by the PI3K pathway enrichment score.\n\nsc.pl.umap(score_pathway, color=[pathway], cmap='RdBu_r', vcenter=0, title=f\"{pathway} Pathway Activity\")\n\n# Plot a violin plot showing the distribution of the PI3K score across cell types.\n\nsc.pl.violin(score_pathway, keys=[pathway], groupby='celltype', rotation=90, ylabel=f'{pathway} score')\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final AnnData object, which contains all analysis results.\n\nprint(\"\\nSaving final AnnData object...\")\n\nadata.write(\"erygast1k_with_trajectory_and_enrichment.h5ad\")\n\n\n\nprint(\"\\nTrajectory and enrichment analysis completed!\")\n\n",
         "**Methods**\n\n**Single-Cell Trajectory and Enrichment Analysis of Mouse Gastrulation Data**  \n\nThe following workflow was implemented in Python using the `scanpy` and `decoupler` packages for single-cell RNA-seq analysis and enrichment modeling, respectively.  \n\n**Data Loading and Preprocessing**  \nThe mouse gastrulation dataset (`erygast1k`) was loaded from `decoupler`’s built-in resources. The data was preprocessed by computing a neighborhood graph using the pre-existing principal component analysis (PCA) embeddings (`use_rep='X_pca'`).  \n\n**Trajectory Inference**  \nDiffusion maps were computed to capture nonlinear manifold structure (`sc.tl.diffmap`). Neighborhood graphs were recalculated using the diffusion map embeddings (`use_rep='X_diffmap'`) to improve trajectory inference. Partition-based Graph Abstraction (PAGA) was applied to assess connectivity between annotated cell types (`groups='celltype'`). Diffusion Pseudotime (DPT) was calculated with \"Blood progenitors 1\" designated as the trajectory root (`adata.uns['iroot']`).  \n\n**Enrichment Analysis**  \n1. **Transcription Factor (TF) Activity**: The CollecTRI TF-target network (mouse) was retrieved (`dc.op.collectri`). TF activity scores were estimated per cell using the Univariate Linear Model (ULM; `dc.run_ulm`), with minimum targets per TF set to 5 (`min_n=5`). Scores were extracted into a separate `AnnData` object.  \n2. **Pathway Activity**: The PROGENy pathway model (mouse) was used (`dc.op.progeny`). Pathway activities were similarly estimated via ULM and stored in a distinct `AnnData` object.  \n\n**Visualization**  \nUniform Manifold Approximation and Projection (UMAP) plots were generated to visualize:  \n- Cell types and developmental stages (`color=['celltype', 'stage']`).  \n- DPT pseudotime (`color=['dpt_pseudotime']`).  \n- TF (e.g., *Klf3*) and pathway (e.g., PI3K) activity scores, using divergent colormaps (`cmap='RdBu_r'`) centered at zero (`vcenter=0`). Violin plots depicted score distributions across cell types.  \n\n**Output**  \nThe final `AnnData` object, containing all results, was saved as `erygast1k_with_trajectory_and_enrichment.h5ad`.  \n\n**Technical Settings**  \n- Global plotting parameters: `figsize=(5, 5)`, axes frames disabled (`frameon=False`).  \n- Logging verbosity: Level 3 (detailed; `sc.settings.verbosity = 3`).  \n\n*Key Assumptions*: Input data was log-normalized; trajectory inference relied on annotated cell types. Replication requires the specified file paths and Python environments with `scanpy` (v1.9) and `decoupler` (v1.3).  \n\n---  \n*This protocol enables replicating pseudotime ordering and molecular activity inference for developmental single-cell datasets.*"
        ],
        [
         "46",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for system operations, data handling, analysis, and plotting\n\nimport warnings\n\nfrom pathlib import Path\n\nimport loompy\n\nimport matplotlib.pyplot as plt\n\nimport numpy as np\n\nimport pandas as pd\n\nimport scanpy as sc\n\nimport seaborn as sns\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.filterwarnings(\"ignore\")\n\n# Set global plotting parameters\n\nsc.set_figure_params(facecolor=\"white\")\n\nsns.set_theme()\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load the single-cell RNA-seq data.\n\n# NOTE: Replace the path with the actual location of your data file.\n\nadata = sc.read_h5ad(\"path/to/your_openproblems_bmmc_multiome_genes_filtered.h5ad\")\n\nprint(\"Initial dataset shape:\", adata.shape)\n\n\n\n#--------------------------------\n\n# Data Preprocessing\n\n#--------------------------------\n\n\n\n# Filter for gene expression (GEX) features if the dataset is multi-modal.\n\nrna = adata[:, adata.var[\"feature_types\"] == \"GEX\"].copy()\n\ndel adata # Free up memory\n\n\n\n# Apply a log-transformation to the data.\n\nsc.pp.log1p(rna)\n\n\n\n# Identify highly variable genes (HVGs) using a batch-aware method.\n\nsc.pp.highly_variable_genes(rna, batch_key=\"batch\", flavor=\"seurat\")\n\nprint(\"Preprocessing complete.\")\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Prepare for SCENIC\n\n#--------------------------------\n\n\n\n# --- Step 1: Subset Data for Analysis ---\n\n# For demonstration, we subset the data to a single batch/donor ('s1d1').\n\n# SCENIC can be run on the full dataset, but this reduces computational time for an example.\n\nadata_batch = rna[rna.obs[\"batch\"] == \"s1d1\", :].copy()\n\nprint(f\"\\nSubsetting data to batch 's1d1'. New shape: {adata_batch.shape}\")\n\n\n\n\n\n# --- Step 2: Create Loom File for SCENIC ---\n\n# SCENIC requires a specific file format (.loom) for its command-line interface.\n\n# We create this file from our preprocessed AnnData object.\n\nloom_path = \"data/neurips_processed_input.loom\"\n\nPath(\"data\").mkdir(exist_ok=True) # Ensure the directory exists\n\n\n\n# Define row and column attributes required by the loom format.\n\nrow_attrs = {\"Gene\": np.array(adata_batch.var.index)}\n\ncol_attrs = {\n\n\"CellID\": np.array(adata_batch.obs.index),\n\n\"nGene\": np.array(np.sum(adata_batch.X > 0, axis=1)).flatten(),\n\n\"nUMI\": np.array(np.sum(adata_batch.X, axis=1)).flatten(),\n\n}\n\n\n\n# Create the loom file. Note that the expression matrix must be transposed.\n\nprint(f\"Creating loom file at: {loom_path}\")\n\nloompy.create(loom_path, adata_batch.X.T, row_attrs, col_attrs)\n\n\n\n\n\n#--------------------------------\n\n# External Step: Run pyscenic\n\n#--------------------------------\n\n\n\n# The core GRN inference in SCENIC is run via a command-line tool.\n\n# This step is performed outside of this script in a terminal environment.\n\n#\n\n# EXAMPLE COMMAND:\n\n# pyscenic grn data/neurips_processed_input.loom allTFs_hg38.txt \\\n\n# -o adj.csv --num_workers 3\n\n#\n\n# - `data/neurips_processed_input.loom`: The input file we created.\n\n# - `allTFs_hg38.txt`: A list of known transcription factors.\n\n# - `-o adj.csv`: The output file containing gene regulatory network adjacencies.\n\n#\n\n# This command produces a file ('adj.csv') which we load in the next step.\n\nprint(\"\\nSCENIC GRN inference must be run externally via the command line.\")\n\n\n\n\n\n#--------------------------------\n\n# Post-Analysis: Load and Visualize GRN Results\n\n#--------------------------------\n\n\n\n# Assuming the 'pyscenic grn' command has been run, we now load its output.\n\ntry:\n\nresults_adjacencies = pd.read_csv(\"adj.csv\")\n\nprint(f\"\\nSuccessfully loaded GRN results. Found {results_adjacencies.shape[0]} associations.\")\n\nprint(results_adjacencies.head())\n\nexcept FileNotFoundError:\n\nprint(\"\\n'adj.csv' not found. Please run the external pyscenic command first.\")\n\nresults_adjacencies = None\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- UMAP Visualization ---\n\n# Plot UMAPs to visualize the overall data structure before and after subsetting.\n\n# Note: This assumes 'GEX_X_umap' coordinates are pre-computed in the loaded file.\n\nprint(\"\\n--- Visualizing Data ---\")\n\nsc.pl.embedding(rna, \"GEX_X_umap\", color=[\"cell_type\", \"batch\"], title=\"UMAP: All Cells\")\n\nsc.pl.embedding(adata_batch, \"GEX_X_umap\", color=[\"cell_type\", \"batch\"], title=\"UMAP: Donor s1d1\")\n\n\n\n# --- GRN Importance Score Distribution ---\n\n# If GRN results were loaded, plot a histogram of the importance scores.\n\n# This score reflects the strength of the regulatory link between a TF and its target.\n\nif results_adjacencies is not None:\n\nplt.figure(figsize=(8, 5))\n\nplt.hist(np.log10(results_adjacencies[\"importance\"] + 1e-10), bins=50, color=\"skyblue\", edgecolor=\"black\")\n\nplt.xlabel(\"Log10(Importance Score)\")\n\nplt.ylabel(\"Frequency\")\n\nplt.title(\"Distribution of GRN Importance Scores\")\n\nplt.show()\n\n\n\n# --- Genes Detected per Cell ---\n\n# Plot the distribution of the number of genes detected per cell. This is a useful QC metric.\n\nn_genes_detected_per_cell = np.sum(adata_batch.X > 0, axis=1)\n\nn_genes_series = pd.Series(n_genes_detected_per_cell.A.flatten() if hasattr(n_genes_detected_per_cell, 'A') else n_genes_detected_per_cell.flatten())\n\n\n\nfig, ax = plt.subplots(figsize=(8, 5), dpi=100)\n\nsns.histplot(n_genes_series, bins=\"fd\", edgecolor=\"black\", ax=ax)\n\nax.set_xlabel(\"Number of Genes Detected\")\n\nax.set_ylabel(\"Number of Cells\")\n\nax.set_title(\"Distribution of Genes Detected per Cell (Donor s1d1)\")\n\nplt.tight_layout()\n\nplt.show()\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the subsetted AnnData object used for the analysis.\n\nprint(\"\\nSaving subsetted AnnData object...\")\n\nadata_batch.write(\"s1d1_preprocessed_for_scenic.h5ad\")\n\n\n\nprint(\"\\nSCENIC data preparation script completed!\")\n\n",
         "### **Methods**  \n\n#### **Data Processing and Analysis for SCENIC**  \n\n**Programming Environment and Package Setup**  \nThe analysis was performed in Python using standard bioinformatics libraries. Key modules included:  \n- **Data handling**: `pandas`, `numpy`, `loompy`  \n- **Single-cell analysis**: `scanpy` (v1.9.0)  \n- **Visualization**: `matplotlib`, `seaborn`, `scanpy` plotting functions  \n- **SCENIC integration**: `loompy` for file format conversion  \n\nWarnings were suppressed for cleaner execution, and default plotting themes were set for consistency.  \n\n**Data Input and Preprocessing**  \n1. **Single-cell RNA-seq dataset loading**:  \n   - The dataset was loaded from an H5AD file (`path/to/your_openproblems_bmmc_multiome_genes_filtered.h5ad`) using `scanpy.read_h5ad()`.  \n   - The dataset contained multi-modal features; only gene expression (GEX) data was retained for downstream analysis (`adata.var[\"feature_types\"] == \"GEX\"`).  \n\n2. **Normalization and feature selection**:  \n   - A log1p transformation (`sc.pp.log1p`) was applied to stabilize variance.  \n   - Highly variable genes (HVGs) were identified using the `seurat` flavor (`sc.pp.highly_variable_genes`), accounting for batch effects (`batch_key=\"batch\"`).  \n\n**SCENIC Preparation**  \n1. **Subsetting the dataset**:  \n   - To reduce computational demands, the analysis was restricted to cells from a single donor (`batch=\"s1d1\"`).  \n\n2. **Loom file creation for SCENIC**:  \n   - A loom file (`neurips_processed_input.loom`) was generated to interface with SCENIC’s command-line tool.  \n   - Key metadata (gene names, cell IDs, UMI counts) were embedded as row/column attributes. The expression matrix was transposed to match SCENIC’s expected format.  \n\n**External SCENIC Execution**  \n- Gene regulatory network (GRN) inference was performed externally using `pyscenic grn`:  \n  - **Input**: Loom file (`neurips_processed_input.loom`) and a transcription factor (TF) list (`allTFs_hg38.txt`).  \n  - **Output**: An adjacency matrix (`adj.csv`) quantifying TF-target gene regulatory links.  \n\n**Post-Analysis Visualization and Output**  \n1. **Quality control and exploratory plots**:  \n   - UMAP embeddings were visualized to assess cell-type and batch distributions (`sc.pl.embedding`).  \n   - The distribution of GRN importance scores (from `adj.csv`) was plotted as a log-scaled histogram.  \n   - Genes detected per cell were summarized as a histogram to confirm data quality.  \n\n2. **Data output**:  \n   - The subsetted AnnData object (`s1d1_preprocessed_for_scenic.h5ad`) was saved for reproducibility.  \n\n**Key Parameters**  \n- **HVG selection**: `batch_key=\"batch\"`, `flavor=\"seurat\"`  \n- **Subsetting**: `batch=\"s1d1\"`  \n- **Loom file attributes**: Cell IDs (`CellID`), gene counts (`nGene`), UMI counts (`nUMI`)  \n\n**Dependencies**  \n- Python 3.8+, `scanpy` (single-cell analysis), `pyscenic` (external GRN inference), and standard data science libraries (`pandas`, `numpy`).  \n\nThis pipeline provides a complete workflow from raw data preprocessing to SCENIC-compatible file generation and preliminary visualization."
        ],
        [
         "47",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for inferCNV analysis and data handling\n\nlibrary(infercnv) # Main package for inferring copy number variations\n\nlibrary(Seurat) # Used for handling single-cell RNA-seq data\n\nlibrary(RColorBrewer) # Provides color palettes for plotting\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed Seurat object containing your scRNA-seq data.\n\n# NOTE: Replace the path with the actual location of your .RDS file.\n\npbmc <- readRDS(\"../data/pbmc.RDS\")\n\n\n\n# Extract the raw count matrix from the Seurat object.\n\n# inferCNV requires a raw integer count matrix with genes as rows and cells as columns.\n\nraw_counts <- pbmc@assays$RNA@counts\n\n\n\n#--------------------------------\n\n# Prepare inferCNV Inputs\n\n#--------------------------------\n\n\n\n# --- Create Cell Annotation File ---\n\n# This file maps each cell ID to a specific cell type or cluster annotation.\n\n# It should be a two-column, tab-delimited file without headers.\n\ncell_annotations <- data.frame(\n\ncell_id = colnames(pbmc),\n\ncell_type = as.character(pbmc@meta.data$seurat_clusters),\n\nstringsAsFactors = FALSE\n\n)\n\n# Write the annotation file to disk.\n\nwrite.table(\n\ncell_annotations,\n\nfile = \"celltype_pbmc.txt\",\n\nsep = \"\\t\",\n\nrow.names = FALSE,\n\ncol.names = FALSE,\n\nquote = FALSE\n\n)\n\n\n\n# --- Gene Order File ---\n\n# This file provides the genomic coordinates for each gene, which inferCNV uses\n\n# to order the genes along chromosomes in the final heatmap.\n\n# It should be a three-column file: Gene, Chromosome, Start Position.\n\n# NOTE: You must provide your own gene order file (e.g., \"gencode_v19_gene_pos.txt\").\n\ngene_order_file_path <- \"gencode_v19_gene_pos.txt\"\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Run inferCNV\n\n#--------------------------------\n\n\n\n# --- Create inferCNV Object ---\n\n# This object is the central data structure for the analysis.\n\n# `ref_group_names` specifies which cell groups to use as the \"normal\" reference\n\n# for calculating relative expression changes.\n\ninfercnv_obj <- CreateInfercnvObject(\n\nraw_counts_matrix = as.matrix(raw_counts),\n\nannotations_file = \"celltype_pbmc.txt\",\n\ngene_order_file = gene_order_file_path,\n\nref_group_names = c(\"0\", \"8\") # Example: Use clusters 0 and 8 as reference\n\n)\n\n\n\n# --- Run the inferCNV pipeline ---\n\n# This is the main computational step. It normalizes expression, centers it\n\n# relative to the reference cells, and identifies large-scale chromosomal alterations.\n\ninfercnv_obj <- infercnv::run(\n\ninfercnv_obj,\n\ncutoff = 0.1, # Recommended for 10x Genomics data\n\nout_dir = \"infercnv_output/\", # Directory to store all output files\n\ncluster_by_groups = TRUE, # Group cells by their annotation\n\ndenoise = FALSE, # Optional: apply denoising\n\nHMM = FALSE, # Optional: run Hidden Markov Model for CNV prediction\n\nno_prelim_plot = TRUE, # Skip plotting the preliminary heatmap\n\nanalysis_mode = \"subclusters\" # Analyze at the subcluster level for more detail\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# Define a custom color palette for the final CNV heatmap.\n\ncustom_color_pal <- colorRampPalette(c(\"#8DD3C7\", \"white\", \"#BC80BD\"))(100)\n\n\n\n# Generate the final CNV plot. This creates a PDF file in the output directory.\n\n# The plot visualizes inferred amplifications (purple) and deletions (blue-green)\n\n# across the genome for each cell.\n\ninfercnv::plot_cnv(\n\ninfercnv_obj,\n\nplot_chr_scale = TRUE,\n\noutput_filename = \"infercnv_final_plot\",\n\noutput_format = \"pdf\",\n\ncustom_color_pal = custom_color_pal\n\n)\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# The `infercnv::run` function automatically saves the results and plots\n\n# in the specified `out_dir`. The final inferCNV object is also returned,\n\n# which can be saved for later inspection if needed.\n\nsaveRDS(infercnv_obj, file = \"infercnv_output/final_infercnv_object.rds\")\n\n\n\nprint(\"\\ninferCNV analysis completed! Check the 'infercnv_output' directory for results.\")\n\n",
         "**Methods Section: Single-Cell Copy Number Variation (CNV) Analysis Using inferCNV**\n\n**Programming Language**: The analysis was performed using **R** (version ≥ 4.0) with key packages for single-cell RNA-seq data processing and CNV inference.\n\n---\n\n### **1. Package Load**  \nThe following R packages were loaded:  \n- **`infercnv`** (v1.10.1): Core package for inferring large-scale chromosomal alterations (amplifications/deletions) from single-cell RNA-seq data.  \n- **`Seurat`** (v4.0): Used to handle single-cell data structures and extract count matrices.  \n- **`RColorBrewer`** (v1.1-3): Provided color palettes for visualization.  \n\n*Rationale*: These packages enable data manipulation, CNV analysis, and visualization.  \n\n---\n\n### **2. Data Input**  \nA pre-processed **Seurat object** (`pbmc.RDS`) was loaded, containing normalized single-cell RNA-seq data. The raw integer count matrix was extracted from the `RNA` assay slot (`pbmc@assays$RNA@counts`), as `infercnv` requires unnormalized counts with genes as rows and cells as columns.  \n\n*Note*: Users must replace the file path (`../data/pbmc.RDS`) with their own data.  \n\n---\n\n### **3. Preparation of inferCNV Inputs**  \n\n#### **Cell Annotation File**  \nA two-column tab-delimited file (`celltype_pbmc.txt`) was generated, mapping each cell barcode (column 1) to its corresponding Seurat cluster (column 2). This file defines cell groups for reference-based CNV inference.  \n\n#### **Gene Order File**  \nGenomic coordinates for genes (e.g., `gencode_v19_gene_pos.txt`) were required, with columns: *Gene*, *Chromosome*, and *Start Position*. This ensures genes are ordered chromosomally in the output heatmap.  \n\n---\n\n### **4. CNV Inference Pipeline**  \n\n#### **Create inferCNV Object**  \nThe `CreateInfercnvObject` function initialized the analysis with:  \n- `raw_counts_matrix`: Extracted count matrix.  \n- `annotations_file`: Path to the cell annotation file.  \n- `gene_order_file`: Path to the gene coordinates file.  \n- `ref_group_names`: Clusters designated as \"normal\" reference (e.g., clusters `0` and `8`).  \n\n#### **Run inferCNV Analysis**  \nThe `infercnv::run` function executed the workflow with parameters:  \n- `cutoff = 0.1`: Minimum expression threshold (recommended for 10x Genomics data).  \n- `out_dir = \"infercnv_output/\"`: Directory for results.  \n- `cluster_by_groups = TRUE`: Cells grouped by annotations.  \n- `analysis_mode = \"subclusters\"`: Increased resolution for subpopulation-level CNVs.  \n- `HMM = FALSE`: Disabled Hidden Markov Model-based predictions (optional).  \n\n---\n\n### **5. Visualization**  \nA **custom color palette** (blue-green to purple gradient) was defined for the CNV heatmap. The `plot_cnv` function generated a PDF plot (`infercnv_final_plot.pdf`) showing:  \n- Chromosomal regions with relative expression changes (purple: amplifications; blue-green: deletions).  \n- Optional scaling by chromosome length (`plot_chr_scale = TRUE`).  \n\n---\n\n### **6. Output**  \nResults were automatically saved in `infercnv_output/`, including:  \n- Intermediate and final CNV profiles.  \n- The inferCNV object (`final_infercnv_object.rds`) for downstream analysis.  \n\n---\n\n*Replication Notes*:  \n- Ensure input files match formats described (e.g., gene order file must include genomic coordinates).  \n- Adjust `ref_group_names` to appropriate control clusters in the dataset.  \n- For large datasets, consider enabling denoising (`denoise = TRUE`) or HMM (`HMM = TRUE`)."
        ],
        [
         "48",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Package Load\n\n#--------------------------------\n\n\n\n# Load required libraries for analysis, data handling, and plotting\n\nlibrary(Seurat) # For single-cell data management\n\nlibrary(dplyr) # For data manipulation\n\nlibrary(tibble) # For advanced data frame functionalities\n\nlibrary(infercnv) # Main package for inferring copy number variations\n\nlibrary(ggpubr) # For creating publication-quality plots\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load a pre-processed Seurat object.\n\n# NOTE: Replace the path with the actual location of your .RDS file.\n\nscRNA_obj <- readRDS(\"./scRNA_harmony_EP.rds\")\n\n\n\n#--------------------------------\n\n# Prepare inferCNV Inputs\n\n#--------------------------------\n\n\n\n# --- Annotate and Subset Cells ---\n\n# Annotate cell types based on clustering information.\n\n# Here, clusters \"0\" and \"18\" are defined as \"Normal\" reference cells.\n\nmeta_data <- scRNA_obj@meta.data %>%\n\nmutate(cell_type = case_when(\n\nRNA_snn_res.0.6 %in% c(\"0\", \"18\") ~ \"Normal\",\n\nTRUE ~ \"Tumor\"\n\n))\n\n\n\n# Get cell IDs for tumor and normal groups.\n\ntumor_cells <- rownames(meta_data)[meta_data$cell_type == \"Tumor\"]\n\nnormal_cells <- rownames(meta_data)[meta_data$cell_type == \"Normal\"]\n\n\n\n# To manage computational load, randomly sample a subset of tumor cells.\n\nset.seed(123)\n\ntumor_sample <- sample(tumor_cells, min(3000, length(tumor_cells)), replace = FALSE)\n\n\n\n# Combine the sampled tumor cells with all normal cells for the analysis.\n\nselected_cells <- c(tumor_sample, normal_cells)\n\nnew_scRNA <- subset(scRNA_obj, cells = selected_cells)\n\n\n\n\n\n# --- Extract Expression Matrix ---\n\n# inferCNV requires a raw integer count matrix.\n\nexprMatrix <- as.matrix(GetAssayData(new_scRNA, slot = \"counts\"))\n\n\n\n\n\n# --- Create Cell Annotation File ---\n\n# This file maps each cell ID to its cluster ID. inferCNV uses this for grouping.\n\n# NOTE: This is different from the Tumor/Normal annotation used for sampling.\n\ncellAnnot <- new_scRNA@meta.data[, \"RNA_snn_res.0.6\", drop = FALSE]\n\n\n\n\n\n# --- Gene Order File ---\n\n# This file provides the genomic coordinates for each gene.\n\n# NOTE: You must provide your own gene order file.\n\ngene_order_file_path <- \"path/to/gene_pos.txt\"\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Task: Run inferCNV\n\n#--------------------------------\n\n\n\n# --- Create inferCNV Object ---\n\n# `ref_group_names` specifies which clusters to use as the \"normal\" reference.\n\ninfercnv_obj <- CreateInfercnvObject(\n\nraw_counts_matrix = exprMatrix,\n\nannotations_file = cellAnnot, # This can be a data.frame directly\n\ndelim = \"\\t\",\n\ngene_order_file = gene_order_file_path,\n\nref_group_names = c(\"0\", \"18\") # Use normal clusters as reference\n\n)\n\n\n\n# --- Run the inferCNV pipeline ---\n\n# This is the main computational step.\n\ninfercnv_obj <- infercnv::run(\n\ninfercnv_obj,\n\ncutoff = 0.1, # Recommended for 10x Genomics data\n\nout_dir = \"./output/cnv_epithelial/\", # Directory for all output files\n\ncluster_by_groups = FALSE, # Cluster cells across all groups\n\nhclust_method = \"ward.D2\", # Hierarchical clustering method\n\ndenoise = TRUE, # Apply denoising to improve signal\n\nHMM = FALSE # Disable HMM-based CNV prediction in this run\n\n)\n\n\n\n\n\n#--------------------------------\n\n# Downstream Analysis: CNV Scoring (Optional)\n\n#--------------------------------\n\n\n\n# --- Calculate CNV Score ---\n\n# This section calculates a per-cell CNV score based on the inferCNV output.\n\n# Read the observation and reference matrices generated by infercnv::run.\n\nobs <- read.table(\"./output/cnv_epithelial/infercnv.observations.txt\", header = TRUE)\n\nref <- read.table(\"./output/cnv_epithelial/infercnv.references.txt\", header = TRUE)\n\n\n\n# Combine, scale, and calculate a squared residual score.\n\nexpr <- cbind(obs, ref)\n\nexpr_scale <- scale(t(expr))\n\ntmp1 <- sweep(expr_scale, 2, apply(expr_scale, 2, min), FUN = \"-\")\n\ntmp2 <- apply(expr_scale, 2, max) - apply(expr_scale, 2, min)\n\nexpr_1 <- t(2 * sweep(tmp1, 2, tmp2, FUN = \"/\") - 1)\n\ncnv_score_df <- as.data.frame(colSums(expr_1 * expr_1))\n\ncolnames(cnv_score_df) <- \"cnv_score\"\n\n\n\n# Format the CNV score DataFrame.\n\ncnv_score_df <- cnv_score_df %>%\n\nrownames_to_column(var = \"cell\") %>%\n\nmutate(cell = gsub(\"\\\\.\", \"-\", cell)) # Ensure cell names match Seurat object\n\n\n\n\n\n# --- Merge Scores into Seurat Object ---\n\n# Subset the Seurat object to match the cells with CNV scores and merge the scores.\n\nscRNA_sample <- subset(scRNA_obj, cells = cnv_score_df$cell)\n\nmeta_with_cnv <- scRNA_sample@meta.data %>%\n\nrownames_to_column(var = \"cell\") %>%\n\ninner_join(cnv_score_df, by = \"cell\") %>%\n\ncolumn_to_rownames(var = \"cell\")\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- CNV Score Boxplot ---\n\n# Create a boxplot to visualize the distribution of CNV scores across different clusters.\n\ncnv_boxplot <- ggboxplot(\n\nmeta_with_cnv,\n\nx = \"RNA_snn_res.0.6\",\n\ny = \"cnv_score\",\n\nfill = \"RNA_snn_res.0.6\"\n\n) +\n\nscale_y_continuous(limits = c(0, 3000)) +\n\nxlab(\"Cluster\") +\n\nylab(\"CNV Score\") +\n\ntheme(\n\npanel.border = element_rect(colour = \"black\", fill = NA, size = 0.5),\n\nlegend.position = \"none\"\n\n)\n\nprint(cnv_boxplot)\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# The `infercnv::run` function automatically saves the main heatmap and other results.\n\n# Here, we explicitly save the final inferCNV object and the CNV score plot.\n\nsaveRDS(infercnv_obj, file = \"./output/cnv_epithelial/final_infercnv_object.rds\")\n\n\n\n# Save the CNV score boxplot to a PDF file.\n\nggsave(\n\nplot = cnv_boxplot,\n\nfilename = \"./output/cnv_epithelial/CNV_score_by_cluster.pdf\",\n\nwidth = 7,\n\nheight = 4\n\n)\n\n\n\nprint(\"\\ninferCNV analysis and CNV scoring completed!\")\n\n",
         "**Methods**\n\n**Computational Analysis of Copy Number Variation (CNV) in Single-Cell RNA-Seq Data**\n\n**Software and Data Preparation**  \nThe analysis was performed in R (v4.x) using the following packages:  \n- **Seurat** (v4.x) for single-cell data management.  \n- **dplyr** and **tibble** for data manipulation.  \n- **infercnv** (v1.x) for CNV inference.  \n- **ggpubr** for visualization.  \n\n**Data Input**  \nA preprocessed Seurat object (`scRNA_harmony_EP.rds`) was loaded, containing harmonized single-cell RNA-seq data.  \n\n**Preprocessing for inferCNV**  \n1. **Cell Annotation and Subsetting**  \n   - Cells were classified as \"Normal\" (clusters 0 and 18) or \"Tumor\" (all other clusters) based on `RNA_snn_res.0.6` metadata.  \n   - To reduce computational burden, a random subset of 3,000 tumor cells (or fewer if total tumor cells <3,000) was selected, combined with all normal cells, and stored in a new Seurat object.  \n\n2. **Expression Matrix Extraction**  \n   Raw integer counts were extracted from the `counts` slot of the subsetted object, as required by infercnv.  \n\n3. **Annotation Files**  \n   - A cell annotation file mapping cell IDs to cluster IDs (`RNA_snn_res.0.6`) was generated.  \n   - A gene order file (`gene_pos.txt`) specifying genomic coordinates was provided externally.  \n\n**CNV Inference**  \nThe infercnv pipeline was executed as follows:  \n1. An **infercnv object** was created using:  \n   - The raw count matrix, cell annotations, and gene order file.  \n   - Reference groups defined as clusters 0 and 18 (normal cells).  \n\n2. The **main analysis** was run with parameters:  \n   - `cutoff=0.1` (threshold for low-expression filtering, recommended for 10x Genomics data).  \n   - `out_dir=\"./output/cnv_epithelial/\"` for output files.  \n   - Hierarchical clustering (`hclust_method=\"ward.D2\"`) without group separation (`cluster_by_groups=FALSE`).  \n   - Denoising enabled (`denoise=TRUE`); HMM-based prediction disabled (`HMM=FALSE`).  \n\n**Downstream Analysis: CNV Scoring**  \n1. **Score Calculation**  \n   - Observation and reference matrices from infercnv output were scaled, and a squared residual score was computed per cell to quantify CNV burden.  \n\n2. **Integration with Metadata**  \n   CNV scores were merged into the original Seurat object, preserving cell identifiers.  \n\n**Visualization**  \nA boxplot (`ggpubr::ggboxplot`) compared CNV scores across clusters, with axes labeled (\"Cluster\" vs. \"CNV Score\") and outliers constrained to a y-axis limit of 3,000.  \n\n**Output**  \nResults were saved as:  \n- Final infercnv object (`final_infercnv_object.rds`).  \n- CNV score boxplot (PDF: `CNV_score_by_cluster.pdf`).  \n\n*Assumptions*:  \n- Input data followed Seurat v4 conventions.  \n- The gene order file matched the expression matrix gene IDs.  \n- File paths were user-adjusted.  \n\nThis pipeline enables reproducible detection of CNVs from single-cell RNA-seq data, with optional customization of parameters (e.g., HMM, denoising)."
        ],
        [
         "49",
         "\"\nAct as a scientific writer who can translate [R/Python] bioinformatics code into a natural language instruction set, similar to the methods section of a research paper. Your task is to take the provided code and generate a step-by-step guide that explains what each part of the code does, why it’s important, and what parameters are used.\n\n1. **Identify the Programming Language**: Determine whether the code is written in R or Python.\n\n2. **Break Down the Code into Logical Sections**: If the code is already divided into sections (e.g., with headers like `# Package Load` or `# Data Input`), use those as a guide. If not, organize the code into logical sections such as:\n   - Loading necessary packages or importing modules\n   - Inputting and preparing the data\n   - Performing the main analysis or computations\n   - Creating visualizations or plots\n   - Saving the results or output\n\n3. **Provide Detailed Explanations for Each Section**: For each section, write a clear and concise explanation in natural language, including:\n   - **What the code is doing**: Describe the purpose of the code in that section (e.g., \"The code loads the Seurat package, which is essential for single-cell analysis.\")\n   - **Why this step is necessary**: Explain the rationale behind the step in the context of the analysis (e.g., \"Seurat provides functions for normalizing and analyzing single-cell RNA-seq data.\")\n   - **Important parameters or settings**: Highlight any key parameters or settings used in the code (e.g., \"The parameter `min.cells=3` ensures that genes must be detected in at least 3 cells to be included in the analysis.\")\n\n4. **Explain Specific Functions or Methods**: If the code uses specific functions or methods (e.g., `CreateSeuratObject` in R or `sc.pp.calculate_qc_metrics` in Python), provide a brief description of what they do and their role in the analysis.\n\n5. **Include All Necessary Details for Replication**: Ensure the instruction includes all details required for someone with basic knowledge of the programming language and the field to understand and replicate the analysis. This includes:\n   - Data format and any preprocessing steps\n   - Specific requirements or assumptions (e.g., file paths, data structure)\n   - Any custom functions or scripts used\n\n6. **Write in Clear, Concise English**: Ensure the instruction is written in clear, concise English, suitable for inclusion in a scientific methods section. The language should be accessible to readers with a background in the field of study but not necessarily experts in programming.\n\n7. **Ensure Completeness**: The instruction should cover all steps from data input to result output, providing a complete and self-contained guide to the analysis.\n\n8. **Nature Language Format**: The output should contain only natural language that describes the task in a manner similar to the METHOD of the article.\n#--------------------------------\n\n# Import necessary modules\n\n#--------------------------------\n\n\n\n# Import libraries for data handling, analysis, plotting, and single-cell analysis\n\nimport scanpy as sc\n\nimport infercnvpy as cnv\n\nimport matplotlib.pyplot as plt\n\nimport warnings\n\n\n\n#--------------------------------\n\n# Script Configuration\n\n#--------------------------------\n\n\n\n# Suppress warnings for a cleaner output during execution\n\nwarnings.simplefilter(\"ignore\")\n\n# Set global plotting parameters for clear and consistent visualization\n\nsc.settings.set_figure_params(figsize=(5, 5))\n\n\n\n#--------------------------------\n\n# Data Input\n\n#--------------------------------\n\n\n\n# Load an example dataset for inferCNV analysis.\n\n# This dataset from Maynard et al., 2020 is suitable for this purpose.\n\nprint(\"Loading Maynard et al. 2020 dataset...\")\n\nadata = cnv.datasets.maynard2020_3k()\n\n\n\n# Display gene annotations to confirm the required columns (chromosome, start, end) are present.\n\nprint(adata.var.loc[:, [\"ensg\", \"chromosome\", \"start\", \"end\"]].head())\n\n\n\n\n\n#--------------------------------\n\n# Major Analysis Tasks: CNV Inference and Clustering\n\n#--------------------------------\n\n\n\n# --- 1. Run inferCNV ---\n\n# This is the main computational step. It infers copy number variations by comparing\n\n# gene expression in target cells to a set of reference (\"normal\") cells.\n\nprint(\"\\nRunning infercnvpy analysis...\")\n\ncnv.tl.infercnv(\n\nadata,\n\nreference_key=\"cell_type\",\n\n# Define which cell types are used as the normal reference\n\nreference_cat=[\n\n\"B cell\", \"Macrophage\", \"Mast cell\", \"Monocyte\", \"NK cell\",\n\n\"Plasma cell\", \"T cell CD4\", \"T cell CD8\", \"T cell regulatory\",\n\n\"mDC\", \"pDC\"\n\n],\n\nwindow_size=250 # Size of the sliding window for averaging expression\n\n)\n\n\n\n# --- 2. CNV-based Clustering ---\n\n# Cluster cells based on their inferred CNV profiles to identify groups\n\n# of cells with similar large-scale genomic alterations.\n\nprint(\"\\nPerforming CNV-based clustering...\")\n\n# Perform PCA on the CNV matrix.\n\ncnv.tl.pca(adata)\n\n# Compute a neighborhood graph on the PCA results.\n\ncnv.pp.neighbors(adata)\n\n# Run Leiden clustering to find CNV-based clusters.\n\ncnv.tl.leiden(adata)\n\n\n\n# --- 3. CNV UMAP and Scoring ---\n\n# Compute a UMAP embedding and a per-cell CNV score based on the CNV profiles.\n\nprint(\"Computing CNV UMAP and scores...\")\n\ncnv.tl.umap(adata)\n\ncnv.tl.cnv_score(adata)\n\n\n\n\n\n# --- 4. Tumor vs. Normal Classification ---\n\n# Manually classify cells as \"tumor\" or \"normal\" based on the CNV clusters.\n\n# NOTE: The cluster IDs to be labeled as \"tumor\" need to be determined by inspecting\n\n# the chromosome heatmap and CNV scores from the plots below.\n\nprint(\"Classifying tumor vs. normal cells...\")\n\nadata.obs[\"cnv_status\"] = \"normal\"\n\n# Example: mark cells in specific cnv_leiden clusters as tumor\n\ntumor_clusters = [\"10\", \"16\", \"13\", \"8\", \"12\", \"17\", \"1\", \"14\", \"11\"]\n\nadata.obs.loc[adata.obs[\"cnv_leiden\"].isin(tumor_clusters), \"cnv_status\"] = \"tumor\"\n\n\n\n\n\n#--------------------------------\n\n# Plotting / Visualization\n\n#--------------------------------\n\n\n\n# --- Initial Data Visualization ---\n\n# Visualize the original UMAP from transcriptomics data to see cell type distribution.\n\nprint(\"\\n--- Visualizing Results ---\")\n\nsc.pl.umap(adata, color=\"cell_type\", title=\"Transcriptomics UMAP by Cell Type\")\n\n\n\n# --- CNV Heatmaps ---\n\n# Plot a heatmap of inferred CNVs across the genome, grouped by original cell type.\n\n# This shows which cell types exhibit large-scale genomic alterations.\n\ncnv.pl.chromosome_heatmap(adata, groupby=\"cell_type\", title=\"CNV Heatmap by Cell Type\")\n\n\n\n# Plot the CNV heatmap again, but grouped by the new CNV-based Leiden clusters.\n\n# A dendrogram is included to show the hierarchy of clusters.\n\ncnv.pl.chromosome_heatmap(adata, groupby=\"cnv_leiden\", dendrogram=True, title=\"CNV Heatmap by CNV Cluster\")\n\n\n\n# --- CNV UMAPs ---\n\n# Create a multi-panel plot to visualize the CNV-based UMAP embedding.\n\nfig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(11, 11))\n\nax4.axis(\"off\") # Turn off the unused fourth subplot\n\n# Plot CNV UMAP colored by:\n\n# 1. CNV Leiden cluster\n\ncnv.pl.umap(adata, color=\"cnv_leiden\", legend_loc=\"on data\", legend_fontoutline=2, ax=ax1, show=False, title=\"CNV UMAP by CNV Leiden Cluster\")\n\n# 2. CNV score\n\ncnv.pl.umap(adata, color=\"cnv_score\", ax=ax2, show=False, title=\"CNV UMAP by CNV Score\")\n\n# 3. Original cell type annotation\n\ncnv.pl.umap(adata, color=\"cell_type\", ax=ax3, show=False, title=\"CNV UMAP by Original Cell Type\")\n\nplt.tight_layout()\n\nplt.show()\n\n\n\n\n\n# --- Final Classification Visualization ---\n\n# Visualize the final tumor vs. normal classification on both the CNV-based UMAP\n\n# and the original transcriptomics-based UMAP.\n\nfig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))\n\nfig.suptitle(\"Final Tumor/Normal Classification\")\n\n# CNV UMAP\n\ncnv.pl.umap(adata, color=\"cnv_status\", ax=ax1, show=False)\n\n# Transcriptomics UMAP\n\nsc.pl.umap(adata, color=\"cnv_status\", ax=ax2, show=False)\n\nplt.show()\n\n\n\n# Generate separate CNV heatmaps for the classified tumor and normal cells.\n\ncnv.pl.chromosome_heatmap(adata[adata.obs[\"cnv_status\"] == \"tumor\", :], title=\"CNV Heatmap of Tumor Cells\")\n\ncnv.pl.chromosome_heatmap(adata[adata.obs[\"cnv_status\"] == \"normal\", :], title=\"CNV Heatmap of Normal Cells\")\n\n\n\n\n\n#--------------------------------\n\n# Data Output\n\n#--------------------------------\n\n\n\n# Save the final AnnData object, which contains all inferCNVpy analysis results.\n\nprint(\"\\nSaving final AnnData object...\")\n\nadata.write(\"maynard2020_with_infercnv.h5ad\")\n\n\n\nprint(\"\\ninferCNVpy analysis completed!\")\n\n",
         "**Methods**  \n\n### **Computational Analysis of Copy Number Variations (CNVs) in Single-Cell RNA-Seq Data**  \n\nThe analysis was performed using Python, leveraging the `scanpy` and `infercnvpy` packages for single-cell data processing and CNV inference. Below is a detailed description of the workflow:  \n\n#### **1. Library Import and Configuration**  \n- **Required packages**:  \n  - `scanpy` (v1.9.0+) for single-cell data handling and visualization.  \n  - `infercnvpy` for CNV inference and clustering.  \n  - `matplotlib.pyplot` for customized plotting.  \n- **Configuration**:  \n  - Warnings were suppressed to streamline output.  \n  - Default figure dimensions were set to 5×5 inches for consistency.  \n\n#### **2. Data Input and Preprocessing**  \n- **Dataset**: A publicly available single-cell RNA-seq dataset from *Maynard et al. (2020)* was loaded using `infercnvpy.datasets.maynard2020_3k()`.  \n- **Gene annotations**: Confirmed presence of essential genomic coordinates (`chromosome`, `start`, `end`) in the gene metadata.  \n\n#### **3. Inferring Copy Number Variations**  \nThe primary analysis involved:  \n- **Reference definition**: Normal (non-malignant) cell types (e.g., B cells, T cells, monocytes) were specified as a baseline for CNV comparison.  \n- **Sliding window analysis**: A window size of 250 genes was used to smooth expression profiles and reduce noise.  \n- **Output**: An inferred CNV matrix stored in the AnnData object.  \n\n#### **4. CNV-Based Clustering and Classification**  \n- **Dimensionality reduction**: Principal component analysis (PCA) was applied to the CNV matrix.  \n- **Graph-based clustering**: A neighborhood graph was constructed, followed by Leiden clustering (`resolution=1.0`) to group cells with similar CNV profiles.  \n- **UMAP embedding**: A 2D representation was generated to visualize CNV-driven cell relationships.  \n- **CNV scoring**: Per-cell CNV scores quantified genomic aberration magnitude.  \n- **Tumor classification**: Cells in Leiden clusters with high CNV scores (e.g., clusters 10, 16, 13) were manually labeled as \"tumor\"; others were classified as \"normal.\"  \n\n#### **5. Visualization**  \nKey plots generated included:  \n- **Transcriptomics UMAP**: Cell type distribution before CNV analysis.  \n- **Chromosomal heatmaps**: CNV patterns across chromosomes, stratified by (1) original cell type and (2) CNV-based clusters (with hierarchical dendrogram).  \n- **Multi-panel UMAPs**: CNV embeddings colored by cluster, CNV score, and original cell type.  \n- **Final classification**: Side-by-side UMAPs (transcriptomics vs. CNV space) highlighting tumor/normal labels, plus separate heatmaps for tumor and normal subsets.  \n\n#### **6. Data Output**  \nThe final AnnData object, containing CNV profiles, clusters, and classifications, was saved as an H5AD file (`maynard2020_with_infercnv.h5ad`) for downstream use.  \n\n**Notes**:  \n- Tumor cluster selection relied on manual inspection of CNV heatmaps and scores.  \n- Default parameters were used unless specified (e.g., Leiden clustering resolution, UMAP initialization).  \n\nThis pipeline provides a reproducible framework for detecting large-scale genomic alterations in single-cell data, with applications in cancer and developmental biology."
        ]
       ],
       "shape": {
        "columns": 2,
        "rows": 54
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Instr_input</th>\n",
       "      <th>Instructions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Single-Cell RNA-Seq Quality Control Analysis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**METHODS**  \\n\\n**Computational Analysis of S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods: Single-Cell RNA-Seq Data Processing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### Methods\\n\\n#### Data Processing and Qualit...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Computational Analysis of S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Programming Language and To...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Data Processing and Analysi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### Methods  \\n\\n#### **Package Load**  \\nThe ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Single-Cell RNA-Seq Data An...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n### **Software and Package In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods**  \\n\\n#### **Single-Cell RNA-Se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods**  \\n\\n#### **Single-Cell RNA-Se...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Single-cell RNA-seq Data Pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### Methods  \\n\\n#### Single-Cell RNA-Seq Anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Single-Cell RNA-Seq Clusterin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods: Single-Cell RNA-Seq Analysis Us...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Data Processing and Single-Ce...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Single-Cell RNA-Seq Data Proc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Single-Cell RNA-Seq Analysis Pipeline: Clust...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Method: Single-Cell RNA-Seq Data Analysis Us...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Data Acquisition and Initial ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Single-Cell RNA-Seq Data Inte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Single-Cell RNA-Seq Data In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**METHODS**  \\n\\n**Single-Cell RNA-Seq Data In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Data Processing and Batch I...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Data Acquisition and Prepar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods**\\n\\n#### **Programming Language...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Single-Cell RNA-Seq Data An...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Programming Language and To...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n### **Single-Cell RNA-Seq Data ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### Methods  \\n\\n#### Single-Cell RNA-Seq Data...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\nThe single-cell RNA sequencing ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Computational Analysis of Sin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### Methods  \\n\\n#### Data Loading and Preproc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods Section for Monocle2 Single-Cell Tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Single-Cell Trajectory Anal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Data Preparation and Analysis...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Data Processing and Analysi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**Gene Set Enrichment Analysi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### METHODS  \\n\\n#### **Data Loading and Prepr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods: Ligand-Receptor Interaction Analysi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods**\\n\\n#### **Computational Analys...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\nThe analysis pipeline was imp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### Methods  \\n\\n#### **Programming Language a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods**\\n\\n#### **1. Software and Libr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Single-Cell Trajectory and En...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>### **Methods**  \\n\\n#### **Data Processing an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods Section: Single-Cell Copy Number Var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**Computational Analysis of Cop...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n### **Computational Analysis ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**METHODS**\\n\\n**Single-cell RNA-seq Trajector...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**\\n\\n**RNA Velocity Analysis Using s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods**  \\n\\n**RNA Velocity Analysis Using...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>\"\\nAct as a scientific writer who can translat...</td>\n",
       "      <td>**Methods: Single-Cell RNA Velocity Analysis U...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Instr_input  \\\n",
       "0   \"\\nAct as a scientific writer who can translat...   \n",
       "1   \"\\nAct as a scientific writer who can translat...   \n",
       "2   \"\\nAct as a scientific writer who can translat...   \n",
       "3   \"\\nAct as a scientific writer who can translat...   \n",
       "4   \"\\nAct as a scientific writer who can translat...   \n",
       "5   \"\\nAct as a scientific writer who can translat...   \n",
       "6   \"\\nAct as a scientific writer who can translat...   \n",
       "7   \"\\nAct as a scientific writer who can translat...   \n",
       "8   \"\\nAct as a scientific writer who can translat...   \n",
       "9   \"\\nAct as a scientific writer who can translat...   \n",
       "10  \"\\nAct as a scientific writer who can translat...   \n",
       "11  \"\\nAct as a scientific writer who can translat...   \n",
       "12  \"\\nAct as a scientific writer who can translat...   \n",
       "13  \"\\nAct as a scientific writer who can translat...   \n",
       "14  \"\\nAct as a scientific writer who can translat...   \n",
       "15  \"\\nAct as a scientific writer who can translat...   \n",
       "16  \"\\nAct as a scientific writer who can translat...   \n",
       "17  \"\\nAct as a scientific writer who can translat...   \n",
       "18  \"\\nAct as a scientific writer who can translat...   \n",
       "19  \"\\nAct as a scientific writer who can translat...   \n",
       "20  \"\\nAct as a scientific writer who can translat...   \n",
       "21  \"\\nAct as a scientific writer who can translat...   \n",
       "22  \"\\nAct as a scientific writer who can translat...   \n",
       "23  \"\\nAct as a scientific writer who can translat...   \n",
       "24  \"\\nAct as a scientific writer who can translat...   \n",
       "25  \"\\nAct as a scientific writer who can translat...   \n",
       "26  \"\\nAct as a scientific writer who can translat...   \n",
       "27  \"\\nAct as a scientific writer who can translat...   \n",
       "28  \"\\nAct as a scientific writer who can translat...   \n",
       "29  \"\\nAct as a scientific writer who can translat...   \n",
       "30  \"\\nAct as a scientific writer who can translat...   \n",
       "31  \"\\nAct as a scientific writer who can translat...   \n",
       "32  \"\\nAct as a scientific writer who can translat...   \n",
       "33  \"\\nAct as a scientific writer who can translat...   \n",
       "34  \"\\nAct as a scientific writer who can translat...   \n",
       "35  \"\\nAct as a scientific writer who can translat...   \n",
       "36  \"\\nAct as a scientific writer who can translat...   \n",
       "37  \"\\nAct as a scientific writer who can translat...   \n",
       "38  \"\\nAct as a scientific writer who can translat...   \n",
       "39  \"\\nAct as a scientific writer who can translat...   \n",
       "40  \"\\nAct as a scientific writer who can translat...   \n",
       "41  \"\\nAct as a scientific writer who can translat...   \n",
       "42  \"\\nAct as a scientific writer who can translat...   \n",
       "43  \"\\nAct as a scientific writer who can translat...   \n",
       "44  \"\\nAct as a scientific writer who can translat...   \n",
       "45  \"\\nAct as a scientific writer who can translat...   \n",
       "46  \"\\nAct as a scientific writer who can translat...   \n",
       "47  \"\\nAct as a scientific writer who can translat...   \n",
       "48  \"\\nAct as a scientific writer who can translat...   \n",
       "49  \"\\nAct as a scientific writer who can translat...   \n",
       "50  \"\\nAct as a scientific writer who can translat...   \n",
       "51  \"\\nAct as a scientific writer who can translat...   \n",
       "52  \"\\nAct as a scientific writer who can translat...   \n",
       "53  \"\\nAct as a scientific writer who can translat...   \n",
       "\n",
       "                                         Instructions  \n",
       "0   **Single-Cell RNA-Seq Quality Control Analysis...  \n",
       "1   **METHODS**  \\n\\n**Computational Analysis of S...  \n",
       "2   **Methods: Single-Cell RNA-Seq Data Processing...  \n",
       "3   ### Methods\\n\\n#### Data Processing and Qualit...  \n",
       "4   **Methods**  \\n\\n**Computational Analysis of S...  \n",
       "5   **Methods**  \\n\\n**Programming Language and To...  \n",
       "6   **Methods**  \\n\\n**Data Processing and Analysi...  \n",
       "7   ### Methods  \\n\\n#### **Package Load**  \\nThe ...  \n",
       "8   **Methods**  \\n\\n**Single-Cell RNA-Seq Data An...  \n",
       "9   **Methods**  \\n\\n### **Software and Package In...  \n",
       "10  ### **Methods**  \\n\\n#### **Single-Cell RNA-Se...  \n",
       "11  ### **Methods**  \\n\\n#### **Single-Cell RNA-Se...  \n",
       "12  **Methods**  \\n\\n**Single-cell RNA-seq Data Pr...  \n",
       "13  ### Methods  \\n\\n#### Single-Cell RNA-Seq Anal...  \n",
       "14  **Methods**\\n\\n**Single-Cell RNA-Seq Clusterin...  \n",
       "15  ### **Methods: Single-Cell RNA-Seq Analysis Us...  \n",
       "16  **Methods**\\n\\n**Data Processing and Single-Ce...  \n",
       "17  **Methods**\\n\\n**Single-Cell RNA-Seq Data Proc...  \n",
       "18  **Single-Cell RNA-Seq Analysis Pipeline: Clust...  \n",
       "19  **Method: Single-Cell RNA-Seq Data Analysis Us...  \n",
       "20  **Methods**\\n\\n**Data Acquisition and Initial ...  \n",
       "21  **Methods**\\n\\n**Single-Cell RNA-Seq Data Inte...  \n",
       "22  **Methods**  \\n\\n**Single-Cell RNA-Seq Data In...  \n",
       "23  **METHODS**  \\n\\n**Single-Cell RNA-Seq Data In...  \n",
       "24  **Methods**  \\n\\n**Data Processing and Batch I...  \n",
       "25  **Methods**  \\n\\n**Data Acquisition and Prepar...  \n",
       "26  ### **Methods**\\n\\n#### **Programming Language...  \n",
       "27  **Methods**  \\n\\n**Single-Cell RNA-Seq Data An...  \n",
       "28  **Methods**  \\n\\n**Programming Language and To...  \n",
       "29  **Methods**\\n\\n### **Single-Cell RNA-Seq Data ...  \n",
       "30  ### Methods  \\n\\n#### Single-Cell RNA-Seq Data...  \n",
       "31  **Methods**\\n\\nThe single-cell RNA sequencing ...  \n",
       "32  **Methods**\\n\\n**Computational Analysis of Sin...  \n",
       "33  ### Methods  \\n\\n#### Data Loading and Preproc...  \n",
       "34  **Methods Section for Monocle2 Single-Cell Tra...  \n",
       "35  **Methods**  \\n\\n**Single-Cell Trajectory Anal...  \n",
       "36  **Methods**\\n\\n**Data Preparation and Analysis...  \n",
       "37  **Methods**  \\n\\n**Data Processing and Analysi...  \n",
       "38  **Methods**  \\n\\n**Gene Set Enrichment Analysi...  \n",
       "39  ### METHODS  \\n\\n#### **Data Loading and Prepr...  \n",
       "40  **Methods: Ligand-Receptor Interaction Analysi...  \n",
       "41  ### **Methods**\\n\\n#### **Computational Analys...  \n",
       "42  **Methods**  \\n\\nThe analysis pipeline was imp...  \n",
       "43  ### Methods  \\n\\n#### **Programming Language a...  \n",
       "44  ### **Methods**\\n\\n#### **1. Software and Libr...  \n",
       "45  **Methods**\\n\\n**Single-Cell Trajectory and En...  \n",
       "46  ### **Methods**  \\n\\n#### **Data Processing an...  \n",
       "47  **Methods Section: Single-Cell Copy Number Var...  \n",
       "48  **Methods**\\n\\n**Computational Analysis of Cop...  \n",
       "49  **Methods**  \\n\\n### **Computational Analysis ...  \n",
       "50  **METHODS**\\n\\n**Single-cell RNA-seq Trajector...  \n",
       "51  **Methods**\\n\\n**RNA Velocity Analysis Using s...  \n",
       "52  **Methods**  \\n\\n**RNA Velocity Analysis Using...  \n",
       "53  **Methods: Single-Cell RNA Velocity Analysis U...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82d78a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Batch process  for code2instruction\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6f79f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process \n",
    "response = llm.invoke()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "05ae6cef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**Methods: Single-Cell RNA-Seq Quality Control Analysis Using Scanpy**  \n",
      "\n",
      "1. **Module Import and Setup**  \n",
      "   The analysis was performed in Python using the `scanpy` and `anndata` libraries. Scanpy provides tools for single-cell RNA-seq (scRNA-seq) data analysis, while AnnData is the core data structure for storing gene expression matrices and associated metadata. Figure parameters were set to a resolution of 80 DPI with a white background to ensure clear visualizations.  \n",
      "\n",
      "2. **Data Input**  \n",
      "   A publicly available peripheral blood mononuclear cell (PBMC) dataset (`pbmc3k`) was loaded as an `AnnData` object. This object contains:  \n",
      "   - **`X`**: A cells × genes count matrix (sparse or dense format).  \n",
      "   - **`obs`**: Cell-level metadata (e.g., sample IDs, QC metrics).  \n",
      "   - **`var`**: Gene-level metadata (e.g., gene symbols, annotations).  \n",
      "\n",
      "3. **Quality Control (QC) Metric Calculation**  \n",
      "   Key gene sets were annotated to assess technical artifacts:  \n",
      "   - **Mitochondrial genes**: Identified by prefix `MT-` (marking low-quality cells).  \n",
      "   - **Ribosomal genes**: Identified by prefixes `RPS` or `RPL` (potential housekeeping markers).  \n",
      "   - **Hemoglobin genes**: Identified by the pattern `^HB[^(P)]` (erythrocyte contamination check).  \n",
      "   \n",
      "   QC metrics were computed using `scanpy.pp.calculate_qc_metrics`, including:  \n",
      "   - Total counts per cell (`total_counts`).  \n",
      "   - Number of detected genes per cell (`n_genes_by_counts`).  \n",
      "   - Percentage of mitochondrial reads (`pct_counts_mt`).  \n",
      "   Parameters:  \n",
      "   - `qc_vars`: Annotated gene sets to quantify (mitochondrial, ribosomal, hemoglobin).  \n",
      "   - `log1p`: Apply log(1 + x) transformation for normalized metrics.  \n",
      "\n",
      "4. **QC Visualization**  \n",
      "   - **Violin plots** displayed distributions of total counts, detected genes, and mitochondrial percentages (`sc.pl.violin`), with jitter (`jitter=0.4`) to visualize cell-level data points.  \n",
      "   - **Scatter plots** (`sc.pl.scatter`) compared total counts versus detected genes, colored by mitochondrial percentage to identify outliers (e.g., cells with high mitochondrial RNA, indicating potential stress or apoptosis).  \n",
      "\n",
      "5. **Data Output**  \n",
      "   The annotated `AnnData` object, including QC metrics and visualizations, was saved in H5AD format (`pbmc3k_qc.h5ad`) for downstream analysis.  \n",
      "\n",
      "**Key Parameters:**  \n",
      "- `log1p=True`: Normalizes metric distributions for visualization.  \n",
      "- `jitter=0.4`: Improves overplotting clarity in violin plots.  \n",
      "- `qc_vars`: Enables quantification of technical artifacts by gene set.  \n",
      "\n",
      "This pipeline ensures reproducible QC assessment for scRNA-seq data, identifying low-quality cells and technical biases prior to further analysis.\n"
     ]
    }
   ],
   "source": [
    "print(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd6b28b2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BIA-dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
